<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>Peoteomics | Using MaxQuant to analyze LC-MS/MS data | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</h1><div class="post-meta">2022-03-09<span> | </span><span class="category"><a href="/categories/Proteomics/">Proteomics</a></span></div><div class="post-content"><h3 id="一、蛋白质谱的背景知识"><a href="#一、蛋白质谱的背景知识" class="headerlink" title="一、蛋白质谱的背景知识"></a>一、蛋白质谱的背景知识</h3><p>LC-MS/MS</p>
<p>Liquid Chromatography-Mass Spectrometry（LC-MS/MS ，液相色谱-质谱串联）可用于残留化合物检测、有机小分子检测、鉴定和定量污染物以及在医药和食品领域添加剂检测和生物小分子等检测。。</p>
<p>LC-MS/MS一般包含五个步骤：</p>
<p>样本制备；<br>样本分离：使用液相色谱方法分离；<br>质谱上机：离子化、LUMOS原理、采集模式（DDA、DIA、SRM/PRM）；<br>质谱鉴定：谱图格式（Raw、MzXML、MGF）、搜库、打分算法、FDR质控、蛋白推断；<br>生信分析：谱图校验和下游数据统计分析。</p>
<p><img src="https://zouhua.top/images/Omics_Proteomics_01/LCMSMS_concept.jpg"></p>
<p><em>液相色谱（HPLC）分离肽段的原理是什么？</em> 色谱法的分离原理是：溶于流动相(mobile phase)中的各组分经过固定相时，由于与固定相(station phase)发生作用(吸附、分配、排阻、亲和)的大小、强弱不同，在固定相中滞留时间不同，从而先后从固定相中流出。又称为色层法、层析法。（<a target="_blank" rel="noopener" href="https://www.sohu.com/a/471080306_121044873">参考资料</a>）</p>
<p><em>为什么质谱上机之前必须去盐？</em> 上机前的最后一步就是要保证多肽经过脱盐纯化，脱盐不充分不仅会影响目的肽段的鉴定，而且容易损坏仪器。此时可使用反相（RP）树脂来去除盐和缓冲液，其中C18基质是用于捕获疏水肽的理想基质。多肽在高水相流动相中与反相柱结合，同时盐和缓冲液被洗去，然后用高有机相流动相洗脱多肽。Pierce™ C18系列不仅提供了适合大体积样本的离心柱（column），还提供了适合小体积样本的吸头（tip）。（<a target="_blank" rel="noopener" href="http://www.ebiotrade.com/newsf/2019-7/201972164022321.htm">参考资料</a>）</p>
<p><em>离子化（Ionization）是什么？</em> 只有分子转化为气相离子后，质谱仪才能测量其质量。为了达到这一目的，质谱仪使分子带上电荷，然后将带电离子流转化为数据系统能够识别的成比例电流。数据系统将这一电流转化为数字信息，得到质谱图。（<a target="_blank" rel="noopener" href="https://www.waters.com/waters/zh_CN/What-is-MS-and-How-does-it-Work%3F/nav.htm?locale=zh_CN&cid=10073253">参考资料</a>）离子源的性能决定了离子化效率，很大程度上决定了质谱仪的灵敏度。常见的离子化方式有两种：一种是样品在离子源中以气体的形式被离子化，另一种为从固体表面或溶液中溅射出带电离子。在很多情况下进样和离子化同时进行。比如，电子轰击电离(EI)是指气化后的样品分子进入离子化室后，受到由钨或铼灯丝发射并加速的电子流的轰击产生正离子。离子化室压力保持在10-4～10-6mmHg。轰击电子的能量大于样品分子的电离能，使样品分子电离或碎裂。电子轰击质谱能提供有机化合物最丰富的结构信息，有较好的重现性，其裂解规律的研究也最为完善，已经建立了数万种有机化合物的标准谱图库可供检索。其缺点在于不适用于难挥发和热稳定性差的样品。还有其他离子化方法，这里未摘录。（<a target="_blank" rel="noopener" href="https://shiyanjia.com/knowledge/articleinfo-1611.html">参考资料</a>）</p>
<h3 id="一、使用-MaxQuant-软件对-label-free-LC-MS-MS-数据进行定量分析"><a href="#一、使用-MaxQuant-软件对-label-free-LC-MS-MS-数据进行定量分析" class="headerlink" title="一、使用 MaxQuant 软件对 label free LC-MS/MS 数据进行定量分析"></a>一、使用 MaxQuant 软件对 label free LC-MS/MS 数据进行定量分析</h3><p>质谱数据下机之后，需要用软件来进行搜库（根据物种的蛋白质数据库来进行序列比对检索），有多种软件可以选择，以 Orbitrap 下机数据为例，可以用 ProteinDiscover 软件里面调用相应的搜索引擎进行分析，这里不做阐述。下面主要介绍 MaxQuant，一方面是免费，另一方面操作界面人性化，上手快。</p>
<p>2016年，德国马普所的 Cox 和蛋白质组学领域巨擘 Matthias Mann 合作开发了 MaxQuant 软件，并发表在 nbt 上，protocol 也相应发表在 nature protocols。不足五年，引用已高达上万次，其中不乏CNS级别文章。毫不夸张地说，MaxQuant 是业界良心。（摘自：<a target="_blank" rel="noopener" href="https://www.cnblogs.com/jessepeng/p/13580371.html%EF%BC%89">https://www.cnblogs.com/jessepeng/p/13580371.html）</a></p>
<p>主要特点：</p>
<p>支持多种质谱仪厂商产生原始数据，如 raw/wiff；<br>免费不开源，基于 Andromeda 搜索引擎；；<br>支持标记定量和非标定量，如 labelfree（优势）/ICAT/SILAC/TMT/iTRAQ 等；<br>蛋白鉴定数多，定量准确性高（主要是因为非线性质量校正和 Match Between Runs）；<br>有较完整的结果查看和分析界面；<br>主要部署在 Windows 系统（Linux 需要 mono，MacOS 需借助 Parallels 或 Bootcamp）；<br>有配套的结果后处理软件 Perseus。</p>
<p><em>下载安装</em></p>
<p>官网：<a target="_blank" rel="noopener" href="https://www.maxquant.org/%EF%BC%8C%E6%8C%89%E7%85%A7%E5%AE%98%E6%96%B9%E6%96%87%E6%A1%A3%E8%BF%9B%E8%A1%8C%E5%AE%89%E8%A3%85%E3%80%82MQ%E7%9A%84%E4%B8%8B%E8%BD%BD%E9%9C%80%E8%A6%81%E7%94%A8%E9%82%AE%E7%AE%B1%E6%B3%A8%E5%86%8C%EF%BC%8C%E5%B9%B6%E5%A1%AB%E5%86%99%E6%B3%A8%E5%86%8C%E7%A0%81%E3%80%82%E4%BE%9D%E8%B5%96.NET">https://www.maxquant.org/，按照官方文档进行安装。MQ的下载需要用邮箱注册，并填写注册码。依赖.NET</a> Framework（基于.net框架编写，有些没安装的电脑需要安装） 和 MSFileReader（识别质谱raw文件）。解压后无需安装，点击文件目录下的 MaxQuant.exe 即可使用。</p>
<p><em>配置与运行</em></p>
<p>MaxQuant使用很简单，参数界面设置也不像PD那么凌乱。我的系统是win 11 中文版，软件的菜单有一部分成了中文。</p>
<p><img src="https://i.imgur.com/Xaw0jTc.jpeg"></p>
<p>6 大参数配置模块，我这里只选一些重要的参数，大部分默认就好：</p>
<ol>
<li>Raw files：导入原始数据和设计实验，包括组分和样本，Group（指不同的实验，如label/SILAC同时进行即为2个Group0和Group1，一般跑一个就好）；</li>
<li>Group-specific parameters：Group参数的设置，一般只是Group0；（忘了在哪里看到这么一句话：“Type” will usually be “Standard” (obviously), and the “Multiplicity” will be 1 for label-free quantification, 2 if you have light and heavy labels, and 3 if you have light, medium, and heavy.）</li>
<li>Global parameters：全局参数配置；</li>
<li>Performance：显示正在运行的具体步骤和动态；</li>
<li>Viewer：数据可视化模块， 能在labelfree定量中呈现三维图形。我认为没啥用，无需设置，拖慢速度；</li>
<li>Andromeda configuration：用于配置修饰、酶，以及数据库ID规则设置等，一般无额外需求，无需设置。</li>
</ol>
<p>所以，实际需要配置的就是前面三项，以 Labelfree 定量为例，写下 Group-specific parameters 和 Global parameters 中一些重要的参数：</p>
<p><img src="https://i.imgur.com/V0CLPDb.jpeg"></p>
<p>备注：我在实际操作中遇到的一个问题，在“组特定参数配置”菜单里，Label-free quantification 里面的 LFQ 选上的话，点击开始分析的时候会报错“LFQ requires at least two experiments”。</p>
<p><img src="https://i.imgur.com/aSnmScM.jpeg"></p>
<p>后来我发现，在“全局参数”菜单下面，也有一个 Label-free quantification 栏目。这里也有 LFQ 字样，不过我并没有测试。这里的 iBAQ 我是选择了的，跑完数据能够得到 iBAQ 值。</p>
<p><img src="https://i.imgur.com/RpMgWls.jpeg"></p>
<p>一篇文献中提到的使用 maxquant 进行蛋白组数据分析：</p>
<p><img src="https://i.imgur.com/dekTX7j.png"></p>
<p>搜库完成之后，在raw文件路径下，会生成一个combined目录，里面就有结果文件：proteinGroups.txt，可以用excel打开这个文件，并另存为xlsx格式，这个文件里面就有各个蛋白的iBAQ值。</p>
<p>在一个实践项目中，同一批数据分别用 PD 和 maxquant 跑，出来的结果大体一致，但是想让 maxquant 显示更多结果，如何调整参数？如下图，我将 FDR 从 0.01 调整到 0.05，另外 min pipetides 从1调整到0，这样出来的结果要从一千多个蛋白变成两千多个蛋白。</p>
<p><img src="https://i.imgur.com/RfL4mwM.jpeg"></p>
<p><img src="https://i.imgur.com/FYZpr0W.jpeg"></p>
<p>combined文件夹包含了搜库结果，如果是DIA建库，导入的即是这个文件夹。里面包含了txt文件夹，即鉴定和定量的全部结果。</p>
<p>mqpar.xml是所有参数配置文件，若二次搜库，可直接导入MaxQuant中，无需重复设置。所以，一般建议保留下来，这样也可对结果的设置进行追溯。</p>
<h3 id="二、蛋白质组定量结果之组间差异分析"><a href="#二、蛋白质组定量结果之组间差异分析" class="headerlink" title="二、蛋白质组定量结果之组间差异分析"></a>二、蛋白质组定量结果之组间差异分析</h3><p>查资料看到的，可以用 Perseus 进行后处理。Perseus也是上面两位大佬开发的软件，发在nature methods上。一开始是专门针对MQ数据处理的。后来对所有矩阵类型的表达数据，甚至是多组学数据都适合。</p>
<p>实际操作中，我要对两组蛋白组学样品进行差异分析，两组各有4个生物学重复。</p>
<p>在知乎看到一篇文章总结了三个工具：limma，DEP 和 DEqMS。（参考：<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/448479536%EF%BC%89">https://zhuanlan.zhihu.com/p/448479536）</a></p>
<p>其中 DEP 可以直接以 maxquant 的结果文件（ProteinGroups.txt）作为输入，而且分析之后的结果有命令可以直接画图，真是值得试试。</p>
<p>同时，我看到另一篇文章（<a target="_blank" rel="noopener" href="https://www.bioinfo-scrounger.com/archives/541/%EF%BC%89%E4%B8%AD%EF%BC%8C%E4%BD%9C%E8%80%85%E8%AE%A4%E4%B8%BA">https://www.bioinfo-scrounger.com/archives/541/）中，作者认为</a> limma 使用结果也不错的。</p>
<p>良心作者，甚至还提供一个label-free MS的数据矩阵文件：<a target="_blank" rel="noopener" href="http://www.bioinfo-scrounger.com/data/labelfree.csv">http://www.bioinfo-scrounger.com/data/labelfree.csv</a></p>
<p>下面摘抄该作者使用 limma 对两组蛋白质谱数据进行差异分析：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"># 如果就测试数据两组比较而言，limma需要先做log2转化，然后设定分组矩阵，最后用线性模型拟合再用Empirical Bayes test求出检验的P值</span><br><span class="line"></span><br><span class="line">library(limma)</span><br><span class="line"></span><br><span class="line">df1 &lt;- log2(df + 1)</span><br><span class="line">group_list &lt;- factor(c(rep(&quot;A&quot;,3), rep(&quot;B&quot;,3)))</span><br><span class="line">design &lt;- model.matrix(~group_list)</span><br><span class="line">colnames(design) &lt;- levels(group_list)</span><br><span class="line">rownames(design) &lt;- colnames(df1)</span><br><span class="line"></span><br><span class="line">fit &lt;- lmFit(df1, design)</span><br><span class="line">fit &lt;- eBayes(fit, trend&#x3D;TRUE)</span><br><span class="line">result_limma &lt;- topTable(fit, coef&#x3D;2,n&#x3D;Inf)</span><br><span class="line">&gt;sum(result_limma$P.Value &lt; 0.05 &amp; abs(result_limma$logFC) &gt; 1)</span><br><span class="line">[1] 952</span><br><span class="line"># Limma的差异蛋白数目就相比以上就略多了，满足阈值的有952个</span><br></pre></td></tr></table></figure>

<p>下面是我的测试过程，并且用自己的蛋白组数据做了组间差异分析。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br></pre></td><td class="code"><pre><span class="line">### 蛋白质组学差异分析的博文：《浅谈蛋白组的差异蛋白分析》https:&#x2F;&#x2F;www.bioinfo-scrounger.com&#x2F;archives&#x2F;541&#x2F;</span><br><span class="line">###  这篇文章也差不多：《蛋白质组差异分析的三个R包》 https:&#x2F;&#x2F;zhuanlan.zhihu.com&#x2F;p&#x2F;448479536</span><br><span class="line"></span><br><span class="line">##############################################################</span><br><span class="line"></span><br><span class="line">#测试数据下载：labelfree.csv (http:&#x2F;&#x2F;www.bioinfo-scrounger.com&#x2F;data&#x2F;labelfree.csv)</span><br><span class="line"></span><br><span class="line">#首先做一步过滤处理，以50%原则过滤掉缺失值较多的Protein，这步主要针对后续T检验和limma的预处理，而MSstats则有其默认的预处理方式</span><br><span class="line"></span><br><span class="line">library(ggplot2)</span><br><span class="line">library(dplyr)</span><br><span class="line"></span><br><span class="line">#data &lt;- read.table(file &#x3D; &quot;labelfree.csv&quot;, sep &#x3D; &quot;,&quot;, header &#x3D; T, row.names &#x3D; 1)</span><br><span class="line">data &lt;- read.table(file &#x3D; &quot;labelfree-edited.txt&quot;, sep &#x3D; &quot;\t&quot;, header &#x3D; T, row.names &#x3D; 1)</span><br><span class="line">rm &lt;- apply(data, 1, function(x)&#123;  # 这个函数的功能是去掉：超过50%的样品的表达值为0的样品</span><br><span class="line">  sum(x &#x3D;&#x3D; 0) &gt; 3</span><br><span class="line">&#125;)</span><br><span class="line">df &lt;- data[!rm,]</span><br><span class="line">#&gt;nrow(df)</span><br><span class="line">#[1] 5325</span><br><span class="line"></span><br><span class="line">##################################################################</span><br><span class="line"></span><br><span class="line">#接着则是做log转化，这步有没有必要呢？Label-free技术是用峰谱积分计算丰度的，其丰度的数量级在10的7次到10的11次不等，从下面每个样本的均值和标准差可看出，丰度值数据分布非常离散</span><br><span class="line"></span><br><span class="line">Mean &lt;- apply(df, 2, mean)</span><br><span class="line">SD &lt;- apply(df, 2, sd)</span><br><span class="line">#&gt; Mean</span><br><span class="line">#A1         A2         A3         B1         B2         B3 </span><br><span class="line">#2366243926 2351546722 2165457844 2175212680 2708233812 2290120209</span><br><span class="line">#&gt; SD</span><br><span class="line">#A1          A2          A3          B1          B2          B3 </span><br><span class="line">#13098498790 14392930144 14179115278 11151997873 12092333164 10344859562</span><br><span class="line">#那么做log转化有什么用呢，一般作用为消除异方差；减少自变量数量级不一致的情况；使非线性变量关系转化为线性关系；最主要的还是为了使数据的呈现方式接近我们所希望的假设，从而更好的进行统计推断（以上来自知乎）</span><br><span class="line"></span><br><span class="line">#下面左边是原始数据AB两组的散点图，右边是做了对数转化后的</span><br><span class="line"></span><br><span class="line">#---------------</span><br><span class="line">#   T检验</span><br><span class="line">#---------------</span><br><span class="line">#在做Two-Sample T检验前，需要样本总体符合正态分布，并且方差齐性，以测试数据为例（未做对数转化），先做个F检验</span><br><span class="line"></span><br><span class="line">fvalue &lt;- apply(df, 1, function(x)&#123;</span><br><span class="line">  a &lt;- factor(c(rep(&quot;A&quot;,3), rep(&quot;B&quot;,3)))</span><br><span class="line">  var.test(x~a)</span><br><span class="line">&#125;)</span><br><span class="line">sum(unlist(lapply(fvalue, function(x)&#123;</span><br><span class="line">  x$p.value &gt; 0.05</span><br><span class="line">&#125;)))</span><br><span class="line">#[1] 4605</span><br><span class="line">#可看出5325个蛋白有4605个是方差齐性的</span><br><span class="line"></span><br><span class="line">#接下来用t.test函数做成组T检验，双尾（默认），方差不齐性的用var.equal &#x3D; FALSE</span><br><span class="line"></span><br><span class="line">pvalue &lt;- apply(df, 1, function(x)&#123;</span><br><span class="line">  a &lt;- factor(c(rep(&quot;A&quot;,3), rep(&quot;B&quot;,3)))</span><br><span class="line">  fvalue &lt;- var.test(x~a)</span><br><span class="line">  if (fvalue$p.value &gt; 0.05)&#123;</span><br><span class="line">    t.test(x~a, var.equal &#x3D; T)</span><br><span class="line">  &#125;else&#123;</span><br><span class="line">    t.test(x~a, var.equal &#x3D; F)</span><br><span class="line">  &#125;</span><br><span class="line">&#125;)</span><br><span class="line"></span><br><span class="line">result_ttest &lt;- data.frame(ID&#x3D;row.names(df), </span><br><span class="line">                           Pvalue &#x3D; as.numeric(unlist(lapply(pvalue, function(x) x$p.value))),</span><br><span class="line">                           log2FC &#x3D; log2(as.numeric(unlist(lapply(pvalue, function(x) x$estimate[1]&#x2F;x$estimate[2]))))</span><br><span class="line">)</span><br><span class="line">sum(result_ttest$Pvalue &lt; 0.05 &amp; abs(result_ttest$log2FC) &gt; 1)</span><br><span class="line">#[1] 704</span><br><span class="line">#可看出差异蛋白数目为704个；如果是预处理时做了对数转化的话，代码同上（其中有略微区别，logFC是作相减，而不是上面代码的相除），结果符合相同的阈值的差异蛋白有694个</span><br><span class="line"></span><br><span class="line">#---------------</span><br><span class="line">     Limma</span><br><span class="line">#---------------</span><br><span class="line">#Limma就比较熟悉了，如果就测试数据两组比较而言，limma需要先做log2转化，然后设定分组矩阵，最后用线性模型拟合再用Empirical Bayes test求出检验的P值</span><br><span class="line">library(limma)</span><br><span class="line">df1 &lt;- log2(df + 1)</span><br><span class="line">group_list &lt;- factor(c(rep(&quot;A&quot;,3), rep(&quot;B&quot;,3)))</span><br><span class="line">design &lt;- model.matrix(~group_list)     ### 这是什么函数？？？</span><br><span class="line">colnames(design) &lt;- levels(group_list)</span><br><span class="line">rownames(design) &lt;- colnames(df1)</span><br><span class="line">fit &lt;- lmFit(df1, design)</span><br><span class="line">fit &lt;- eBayes(fit, trend&#x3D;TRUE)</span><br><span class="line">result_limma &lt;- topTable(fit, coef&#x3D;2,n&#x3D;Inf)</span><br><span class="line">sum(result_limma$P.Value &lt; 0.05 &amp; abs(result_limma$logFC) &gt; 1)</span><br><span class="line">#[1] 952</span><br><span class="line">#Limma的差异蛋白数目就相比以上就略多了，满足阈值的有952个</span><br><span class="line">write.table(result_limma, file &#x3D; &quot;limma_output.xls&quot;, sep &#x3D; &quot;\t&quot;)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">################################################################################</span><br><span class="line">############</span><br><span class="line">############   下面用自己的数据调用limma做差异分析</span><br><span class="line">############</span><br><span class="line">################################################################################</span><br><span class="line"></span><br><span class="line"># 用Perl写脚本，将不同样品的ProteinGroups.txt文件一起合并成一个表达谱矩阵。第一列为基因名，其余各列为表达值。类似于RNA Seq的表达矩阵。</span><br><span class="line"></span><br><span class="line">#首先做一步过滤处理，以50%原则（可以自己调整）过滤掉缺失值较多的Protein，这步主要针对后续T检验和limma的预处理，而MSstats则有其默认的预处理方式</span><br><span class="line"></span><br><span class="line">library(ggplot2)</span><br><span class="line">library(dplyr)</span><br><span class="line"></span><br><span class="line">#data &lt;- read.table(file &#x3D; &quot;labelfree.csv&quot;, sep &#x3D; &quot;,&quot;, header &#x3D; T, row.names &#x3D; 1)</span><br><span class="line">data &lt;- read.table(file &#x3D; &quot;merged_iBAQ_table_v3_input_for_R.txt&quot;, sep &#x3D; &quot;\t&quot;, header &#x3D; T, row.names &#x3D; 1)</span><br><span class="line">rm &lt;- apply(data, 1, function(x)&#123;  # 这个函数的功能是去掉：超过50%的样品的表达值为0的样品</span><br><span class="line">  sum(x &#x3D;&#x3D; 0) &gt; 7</span><br><span class="line">&#125;)</span><br><span class="line">df &lt;- data[!rm,]</span><br><span class="line">#&gt;nrow(df)</span><br><span class="line">#[1] 5325</span><br><span class="line"></span><br><span class="line">library(limma)</span><br><span class="line">df1 &lt;- log2(df + 1)</span><br><span class="line">group_list &lt;- factor(c(rep(&quot;A&quot;,4), rep(&quot;B&quot;,4)))</span><br><span class="line">design &lt;- model.matrix(~group_list)     ### 这是什么函数？？？</span><br><span class="line">colnames(design) &lt;- levels(group_list)</span><br><span class="line">rownames(design) &lt;- colnames(df1)</span><br><span class="line"></span><br><span class="line">fit &lt;- lmFit(df1, design)</span><br><span class="line">fit &lt;- eBayes(fit, trend&#x3D;TRUE)</span><br><span class="line">result_limma &lt;- topTable(fit, coef&#x3D;2,n&#x3D;Inf)</span><br><span class="line">sum(result_limma$P.Value &lt; 0.05 &amp; abs(result_limma$logFC) &gt; 1)</span><br><span class="line">#[1] 952</span><br><span class="line">#Limma的差异蛋白数目就相比以上就略多了，满足阈值的有952个</span><br><span class="line"></span><br><span class="line">write.table(result_limma, file &#x3D; &quot;limma_output_FREE1.xls&quot;, sep &#x3D; &quot;\t&quot;)</span><br><span class="line">### 如果下次重新打开R画图，则直接读入文件：</span><br><span class="line">#result_limma &lt;- read.table(&quot;limma_output_FREE1.xls&quot;, sep &#x3D; &quot;\t&quot;, head&#x3D;T)</span><br><span class="line"></span><br><span class="line">####</span><br><span class="line"># Volcano Plot</span><br><span class="line">#logFC AveExpr         t      P.Value   adj.P.Val         B</span><br><span class="line">#AT5G09660.4 -24.86099 12.4305 -13.67679 2.515145e-07 0.000174047 -1.497474</span><br><span class="line">#AT5G09660.3 -24.86099 12.4305 -13.67679 2.515145e-07 0.000174047 -1.497474</span><br><span class="line"></span><br><span class="line">#----------------     计算-logFC     ----------------</span><br><span class="line">#prop &lt;- apply(df, 2,function(x,y) (x&#x2F;y)*100, df$Total)</span><br><span class="line">#x &lt;- apply(result_limma, 2, function(x) -x, result_limma$logFC)</span><br><span class="line">result_limma$x &lt;- -result_limma$logFC</span><br><span class="line">head(result_limma)</span><br><span class="line">#----------------    计算-log10(Pvalue)   -----------</span><br><span class="line">result_limma$y &lt;- -log10(result_limma$P.Value)</span><br><span class="line"></span><br><span class="line">#install.packages(&#39;ggrepel&#39;)</span><br><span class="line">library(ggrepel)</span><br><span class="line"></span><br><span class="line">library(ggplot2)</span><br><span class="line"></span><br><span class="line">####  Volcano Plot</span><br><span class="line"></span><br><span class="line">ggplot() +</span><br><span class="line">  #ggplot(data&#x3D;a, aes(x&#x3D;normalized_log2Fab, y&#x3D;log2_mark_1_2_PrSM))</span><br><span class="line">  geom_hline(yintercept &#x3D; 0, linetype&#x3D;&quot;dashed&quot;, color &#x3D; &quot;grey&quot;, size&#x3D;1.5) +</span><br><span class="line">  geom_hline(yintercept &#x3D; -log10(0.05), linetype&#x3D;&quot;dashed&quot;, color &#x3D; &quot;grey&quot;, size&#x3D;1.5) +</span><br><span class="line">  geom_vline(xintercept &#x3D; 0, linetype&#x3D;&quot;dashed&quot;, color &#x3D; &quot;grey&quot;, size&#x3D;1.5) +</span><br><span class="line">  #geom_point(color&#x3D;&quot;#9167a9&quot;, size&#x3D;4, alpha&#x3D;0.7)+</span><br><span class="line">  geom_point(data&#x3D;filter(result_limma, P.Value&lt;&#x3D;0.05 &amp; x&gt;0), aes(x&#x3D;x, y&#x3D;y), size &#x3D; 2, color&#x3D;&quot;red&quot;, alpha&#x3D;0.2) +</span><br><span class="line">  geom_point(data&#x3D;filter(result_limma, P.Value&lt;&#x3D;0.05 &amp; x&lt;0), aes(x&#x3D;x, y&#x3D;y), size &#x3D; 2, color&#x3D;&quot;green&quot;, alpha&#x3D;0.2) +</span><br><span class="line">  geom_point(data&#x3D;filter(result_limma, P.Value&gt;0.05), aes(x&#x3D;x, y&#x3D;y), size &#x3D; 2, color&#x3D;&quot;darkblue&quot;, alpha&#x3D;0.2) +</span><br><span class="line">  #scale_size(range &#x3D; c(0.5, 10))  + </span><br><span class="line">  theme_bw() +</span><br><span class="line">  theme(panel.grid.major &#x3D; element_blank(), panel.grid.minor &#x3D; element_blank(), axis.line &#x3D; element_line(colour &#x3D; &quot;black&quot;)) +</span><br><span class="line">  #theme(axis.text.x &#x3D; element_text(size &#x3D; 15, color &#x3D; &quot;black&quot;, face &#x3D; &quot;bold&quot;, vjust &#x3D; 0.5, hjust &#x3D; 0.5, angle &#x3D; 45)) +</span><br><span class="line">  theme(axis.text.x &#x3D; element_text(size &#x3D; 15, color &#x3D; &quot;black&quot;, face &#x3D; &quot;bold&quot;)) +</span><br><span class="line">  theme(axis.text.y &#x3D; element_text(color &#x3D; &quot;black&quot;, size &#x3D; 15)) +</span><br><span class="line">  ylab(&quot;-log10(P value)&quot;) +</span><br><span class="line">  xlab(&quot;log10(Fold Change)&quot;) +</span><br><span class="line">  theme(axis.title.y &#x3D; element_text(size &#x3D; 18)) +</span><br><span class="line">  theme(axis.title.x &#x3D; element_text(size &#x3D; 18))</span><br><span class="line"></span><br></pre></td></tr></table></figure>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2022/04/01/mcscanx-%E6%B5%8B%E8%AF%95/">MCScanX 测试</a><a class="next" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/Bioinformatics-Ancestral-Genome-evolution/">Bioinformatics | Ancestral Genome evolution</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/mcscanx-%E6%B5%8B%E8%AF%95/">MCScanX 测试</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>