<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="关注这个问题的缘由是对转录组进行基因定量分析的过程中，发现 stringtie 和 htseq-count 产生的结果不一致。 stringtie 程序里面并没有提供 MAPQ 筛选的参数，也就是输入的 bam 文件中，所有 reads 都会被计数。（如果需要干预，需要提前用 samtools view -q 参数筛选 reads） 而 htseq-cuont 默认会使用 MAPQ&gt;10 来">
<meta property="og:type" content="article">
<meta property="og:title" content="基因组比对的SAM&#x2F;BAM文件中，怎么理解MAPQ值">
<meta property="og:url" content="http://example.com/2020/12/28/Bioinfo-SAM-BAM-MAPQ-value-in-samtools/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:description" content="关注这个问题的缘由是对转录组进行基因定量分析的过程中，发现 stringtie 和 htseq-count 产生的结果不一致。 stringtie 程序里面并没有提供 MAPQ 筛选的参数，也就是输入的 bam 文件中，所有 reads 都会被计数。（如果需要干预，需要提前用 samtools view -q 参数筛选 reads） 而 htseq-cuont 默认会使用 MAPQ&gt;10 来">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://miro.medium.com/max/875/0*CePvh8XpBcbC1HjZ.png">
<meta property="article:published_time" content="2020-12-28T03:06:55.000Z">
<meta property="article:modified_time" content="2020-12-29T01:29:27.703Z">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://miro.medium.com/max/875/0*CePvh8XpBcbC1HjZ.png">

<link rel="canonical" href="http://example.com/2020/12/28/Bioinfo-SAM-BAM-MAPQ-value-in-samtools/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>基因组比对的SAM/BAM文件中，怎么理解MAPQ值 | FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/28/Bioinfo-SAM-BAM-MAPQ-value-in-samtools/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          基因组比对的SAM/BAM文件中，怎么理解MAPQ值
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-28 11:06:55" itemprop="dateCreated datePublished" datetime="2020-12-28T11:06:55+08:00">2020-12-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-29 09:29:27" itemprop="dateModified" datetime="2020-12-29T09:29:27+08:00">2020-12-29</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>关注这个问题的缘由是对转录组进行基因定量分析的过程中，发现 stringtie 和 htseq-count 产生的结果不一致。</p>
<p>stringtie 程序里面并没有提供 MAPQ 筛选的参数，也就是输入的 bam 文件中，所有 reads 都会被计数。（如果需要干预，需要提前用 samtools view -q 参数筛选 reads）</p>
<p>而 htseq-cuont 默认会使用 MAPQ&gt;10 来做筛选，即只考虑唯一比对的 reads。</p>
<p>实际上，在植物转录组测序的 reads 中，multiple mapping 的 reads 会占据很大的比例。比如有的基因会有多个拷贝，而且表达量还特别高，那么分析的时候是需要考虑这些问题的。</p>
<h3 id="sambam-文件结构简介"><a class="markdownIt-Anchor" href="#sambam-文件结构简介"></a> SAM/BAM 文件结构简介</h3>
<p><img src="https://miro.medium.com/max/875/0*CePvh8XpBcbC1HjZ.png" alt="Example of the SAM file" /></p>
<p>sam/bam文件一共12列，每列具体信息如下。第五例为MAPQ值。</p>
<p>Each alignment line has 11 mandatory fields for essential alignment information such as mapping position, and variable number of optional fields for flexible or aligner specific information:</p>
<ol>
<li>Read Name</li>
<li>SAM flag</li>
<li>chromosome (if read is has no alignment, there will be a “*” here)</li>
<li>position (1-based index, “left end of read”)</li>
<li>MAPQ (mapping quality — describes the uniqueness of the alignment, 0=non-unique, &gt;10 probably unique)</li>
<li>CIGAR string (describes the position of insertions/deletions/matches in the alignment, encodes splice junctions, for example)</li>
<li>Name of mate (mate pair information for paired-end sequencing, often “=”)</li>
<li>Position of mate (mate pair information)</li>
<li>Template length (always zero for me)</li>
<li>Read Sequence</li>
<li>Read Quality</li>
<li>Program specific Flags (i.e. AS is an alignment score, NH is a number of reported alignments that contains the query in the current record)</li>
</ol>
<h3 id="mapq是怎么定义的"><a class="markdownIt-Anchor" href="#mapq是怎么定义的"></a> MAPQ是怎么定义的？</h3>
<p>Samtools<a target="_blank" rel="noopener" href="https://samtools.github.io/hts-specs/SAMv1.pdf">文档</a>中对MAPQ的定义是：</p>
<p>MAPQ: MAPping Quality. It equals −10 log10 Pr{mapping position is wrong}, rounded to the nearest<br />
integer. A value 255 indicates that the mapping quality is not available.</p>
<p>如果正确比对的可能性为 0, 那么错误率 P = 1 - 0 = 1<br />
MAPQ = -10*log10(1) = 0</p>
<p>如果正确比对的可能性为 0.9, 那么错误率 P = 1 - 0.9 = 0.1<br />
MAPQ = -10*log10(0.1) = 10</p>
<p>如果正确比对的可能性为 0.99, 那么错误率 P = 1 - 0.99 = 0.01<br />
MAPQ = -10*log10(0.01) = 20</p>
<p>如果正确比对的可能性为 0.999，那么错误率 P = 1 - 0.999 = 0.001<br />
MAPQ = -10*log10(0.001) = 30</p>
<h3 id="mapq0-的情况到底是指-unmapped-reads-还是指-non-unique-reads"><a class="markdownIt-Anchor" href="#mapq0-的情况到底是指-unmapped-reads-还是指-non-unique-reads"></a> MAPQ=0 的情况到底是指 unmapped reads 还是指 non unique reads？</h3>
<p>MAPQ=0 表示 non unique reads。我查看了 bowtie2 比对之后的 bam 文件，里面 MAPQ=0 的 reads 就是比对到了基因组的多个位置。</p>
<h3 id="mapq0-与-mapq1-有什么区别"><a class="markdownIt-Anchor" href="#mapq0-与-mapq1-有什么区别"></a> MAPQ=0 与 MAPQ=1 有什么区别？</h3>
<p>随机检查了一下 bam 文件中 MAPQ=0 与 MAPQ=1 的两个 PE reads，它们均比对到了基因组的两个不同位置。<br />
应该是比对区域的mismatch导致的MAPQ值不同。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">samtools view File1.bam | grep &#x27;V300018696L3C002R0430736588&#x27;</span><br><span class="line">V300018696L3C002R0430736588     99      Chr01   64085   0       90M     =       66302   331     CCCAGCAGTCTGAGTCCGCAGTATCCACCTTATCGGCGGCTGGACGGAACCCCTTCGCTAAGACATCTTGATACCTTAATTTTTCAATCT      FFFFFFFFFEFF@FFFFFFFFFAFFFFFFFFFFFFFFFF;EB-EC+F%6DDDF7DF&#x27;)DDCD&lt;B@D=@D0&#x27;FE(:@FAC8FF$D&lt;(;C3D     AS:i:-7 ZS:i:-7 XN:i:0  XM:i:3  XO:i:0  XG:i:0  NM:i:3  MD:Z:42A14A24A7 YS:i:-13        YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0430736588     355     Chr01   66206   0       90M     =       66302   186     CCCAGCAGTCTGAGTCCGCAGTATCCACCTTATCGGCGGCTGGACGGAACCCCTTCGCTAAGACATCTTGATACCTTAATTTTTCAATCT      FFFFFFFFFEFF@FFFFFFFFFAFFFFFFFFFFFFFFFF;EB-EC+F%6DDDF7DF&#x27;)DDCD&lt;B@D=@D0&#x27;FE(:@FAC8FF$D&lt;(;C3D     AS:i:-7 ZS:i:-7 XN:i:0  XM:i:3  XO:i:0  XG:i:0  NM:i:3  MD:Z:42A14A24A7 YS:i:-13        YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0430736588     147     Chr01   66302   0       19M1007N58M13S  =       64085   -331    CTTCTAGCTGTTATCACTCCATCATCCCTCTCTTGCATCTTCTGCTTTTCCTTTGGTGTGGACTTCAACTGGACAACAGGTCCCATGCTC      EFDEFFFFEFEEFEFFFEFFFBFFDFFFFFDFEEFGEEFFEFFFFDEEFFFFDFFFFFFFEEFFDFFFFFFEFFEEFFFFAFEFEFFFEF     AS:i:-13        XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:77 YS:i:-7 YT:Z:CP XS:A:-  NH:i:2</span><br><span class="line">V300018696L3C002R0430736588     403     Chr01   66302   0       19M1007N58M13S  =       66206   -186    CTTCTAGCTGTTATCACTCCATCATCCCTCTCTTGCATCTTCTGCTTTTCCTTTGGTGTGGACTTCAACTGGACAACAGGTCCCATGCTC      EFDEFFFFEFEEFEFFFEFFFBFFDFFFFFDFEEFGEEFFEFFFFDEEFFFFDFFFFFFFEEFFDFFFFFFEFFEEFFFFAFEFEFFFEF     AS:i:-13        XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:77 YS:i:-7 YT:Z:CP XS:A:-  NH:i:2</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">samtools view File1.bam | grep &#x27;V300018696L3C002R0110944536&#x27;</span><br><span class="line">V300018696L3C002R0110944536     99      Chr01   33996   1       90M     =       34061   155     GGAGAGAATTCCTATGCTGCCATATTGTAATAGATAGAGCAATCCACCACATGCCCCGCCTTCTGTGATTACTCGTGGCAACAAATCCAT      FEDFEFFFFEFFFFFDFFFFFFFFFFFCFFFFFEFFFDFEFFFEFFEFFF3FFFFFFEFFFFEFFDFDFE3AE?D6FDEFEEFEFECFFF     AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:90 YS:i:0  YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0110944536     147     Chr01   34061   1       90M     =       33996   -155    TGATTACTCGTGGCAACAAATCCATCACAGAATTGGATAAAATGGCTAGCCGGATCAGCTGAAAGAGCTCCTATACCCCGGATCCATCTC      DEBFF2E,DFED&gt;FDDFFCFDDFFEFDAFFEEEEBFCFFDEFECEFFEEFDFFFDEDFFFFEEDEFFFCEFFFFDFEBEFFECFFEEFDF     AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:90 YS:i:0  YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0110944536     355     Chr03   35175098        1       90M     =       35175163        155     GGAGAGAATTCCTATGCTGCCATATTGTAATAGATAGAGCAATCCACCACATGCCCCGCCTTCTGTGATTACTCGTGGCAACAAATCCAT      FEDFEFFFFEFFFFFDFFFFFFFFFFFCFFFFFEFFFDFEFFFEFFEFFF3FFFFFFEFFFFEFFDFDFE3AE?D6FDEFEEFEFECFFF     AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:90 YS:i:0  YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0110944536     403     Chr03   35175163        1       90M     =       35175098        -155    TGATTACTCGTGGCAACAAATCCATCACAGAATTGGATAAAATGGCTAGCCGGATCAGCTGAAAGAGCTCCTATACCCCGGATCCATCTC      DEBFF2E,DFED&gt;FDDFFCFDDFFEFDAFFEEEEBFCFFDEFECEFFEEFDFFFDEDFFFFEEDEFFFCEFFFFDFEBEFFECFFEEFDF     AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:90 YS:i:0  YT:Z:CP NH:i:2</span><br></pre></td></tr></table></figure>
<h3 id="为什么bwa和bowtie2生成的结果中mapq不一样"><a class="markdownIt-Anchor" href="#为什么bwa和bowtie2生成的结果中mapq不一样"></a> 为什么BWA和bowtie2生成的结果中MAPQ不一样？</h3>
<p>In SAM file the quality of mapped reads is defined in by so-called MAPQ values — MAPping Quality. It equals −10 log10 Probability {mapping position is wrong}, rounded to the nearest integer. A value 255 indicates that the mapping quality is not available. Now, it is very important to remember that MAPQ values generated by different aligners (e.g. Bowtie2, TopHat, BBMap, BWA) are not exactly comparable. For example, the maximum value of MAPQ score in Bowtie2 is 42, whereas in BWA — 37. You can read a great piece about why this happens in the ACGT blog post.</p>
<p>Another important thing not everyone knows is how MAPQ value is actually calculated. In our of our previous tutorial, we mentioned that it defines how “uniquely” a particular read maps to a reference. In reality, of course, it is more complex than that and you should check out great post here here to understand the exact procedure on how score gets calculated.</p>
<p>Lastly, remember that TopHat outputs only MAPQ scores of 0, 1, 3, or 50. The first three values indicate mappings to 5 or more locations, 3–4 locations, or 2 locations, whereas a value of 50 represents a unique match. Also, in older versions of TopHat unique matches were identified with a value of 255. Therefore, you must understand that values of Bowtie and Bowtie2 (used internally by TopHat to do read mapping) produce a different range of MAPQ scores (0–42).</p>
<p>Now, as we understand a little bit better MAPQ scores, we can filter BAM/SAM files on the mapping quality. eg. getting all reads with a mapping quality larger than 30 (you could also use 49 if you wanted to:). This will remove all reads mapped the undesirable mapping qualities and will only keep uniquely mapped reads. (<a target="_blank" rel="noopener" href="https://medium.com/@shilparaopradeep/samtools-guide-learning-how-to-filter-and-manipulate-with-sam-bam-files-2c28b25d29e8">source</a>)</p>

    </div>

    
    
    

      <footer class="post-footer">

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2020/12/25/Linux-vim-setting/" rel="prev" title="Linux：vim设置">
      <i class="fa fa-chevron-left"></i> Linux：vim设置
    </a></div>
      <div class="post-nav-item">
    <a href="/2020/12/28/Bioinfo-Snakemake-for-bioinfomatics-pipeline/" rel="next" title="使用Snakemake搭建生物信息分析流程">
      使用Snakemake搭建生物信息分析流程 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-3"><a class="nav-link" href="#sambam-%E6%96%87%E4%BB%B6%E7%BB%93%E6%9E%84%E7%AE%80%E4%BB%8B"><span class="nav-number">1.</span> <span class="nav-text"> SAM&#x2F;BAM 文件结构简介</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#mapq%E6%98%AF%E6%80%8E%E4%B9%88%E5%AE%9A%E4%B9%89%E7%9A%84"><span class="nav-number">2.</span> <span class="nav-text"> MAPQ是怎么定义的？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#mapq0-%E7%9A%84%E6%83%85%E5%86%B5%E5%88%B0%E5%BA%95%E6%98%AF%E6%8C%87-unmapped-reads-%E8%BF%98%E6%98%AF%E6%8C%87-non-unique-reads"><span class="nav-number">3.</span> <span class="nav-text"> MAPQ&#x3D;0 的情况到底是指 unmapped reads 还是指 non unique reads？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#mapq0-%E4%B8%8E-mapq1-%E6%9C%89%E4%BB%80%E4%B9%88%E5%8C%BA%E5%88%AB"><span class="nav-number">4.</span> <span class="nav-text"> MAPQ&#x3D;0 与 MAPQ&#x3D;1 有什么区别？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E4%B8%BA%E4%BB%80%E4%B9%88bwa%E5%92%8Cbowtie2%E7%94%9F%E6%88%90%E7%9A%84%E7%BB%93%E6%9E%9C%E4%B8%ADmapq%E4%B8%8D%E4%B8%80%E6%A0%B7"><span class="nav-number">5.</span> <span class="nav-text"> 为什么BWA和bowtie2生成的结果中MAPQ不一样？</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">176</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
