<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>基因组浏览器 JBrowse 安装（基于 Nginx 的 Web 服务器） | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">基因组浏览器 JBrowse 安装（基于 Nginx 的 Web 服务器）</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">基因组浏览器 JBrowse 安装（基于 Nginx 的 Web 服务器）</h1><div class="post-meta">2020-10-03<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a><a href="/categories/Linux/">Linux</a></span></div><div class="post-content"><p>需求：将基因组浏览器 JBrowse 安装在 Linux 服务器，在内网中用客户端电脑的网页浏览器可以打开 JBrowse 进行操作。安装过程如下，以 CentOS 7 系统为例，需要 root 权限。</p>
<p>JBrowse 是一款快速的，可以嵌入的基因组浏览器，它采用 JavaScript 和 HTML5 开发，同时提供了 Perl 脚本处理数据。JBrowse 采用了 Ajax 技术，主要的计算在客户端上完成， 对服务器资源的占用非常低，速度非常快，可以轻松部署上 G 的数据。JBrowse 支持 GFF3, BED, FASTA, Wiggle, BigWig, BAM, VCF (with tabix), REST 等众多数据格式，BAM, BigWig，VCF 格式的数据可以不需要转换而直接使用。 JBrowse 对系统依赖非常低，只需要一个 web 服务器，因此部署起来非常方便。Nginx 是一款轻量级、高性能 web 服务器，它占用内存少，并发能力强。越来越多的网站使用 Nginx 作为 web 服务器。</p>
<h3 id="1-安装-Jbrowse-依赖的软件"><a href="#1-安装-Jbrowse-依赖的软件" class="headerlink" title="1 安装 Jbrowse 依赖的软件"></a>1 安装 Jbrowse 依赖的软件</h3><p>参考资料：<a target="_blank" rel="noopener" href="https://jbrowse.org/docs/installation.html">https://jbrowse.org/docs/installation.html</a></p>
<p>需要先安装依赖的一些工具，即“system pre-requisites”。我用的 CentOS 7系统，操作如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sudo yum groupinstall &quot;Development Tools&quot;</span><br><span class="line">sudo yum install zlib-devel perl-ExtUtils-MakeMaker</span><br></pre></td></tr></table></figure>

<p>安装 node 和 npm<br>源代码下载 <a target="_blank" rel="noopener" href="https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz">https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz</a><br>xz文件解压：<code>tar -xvJf node-v8.11.1-linux-x64.tar.xz</code><br>解压之后文件目录里面包含bin目录（node，npm，npx 三个程序），将bin目录加入环境变量。<br><code>export PATH=/xxx/xxx/bin:$PATH</code><br>source ~/.bashrc 之后即可测试 node -h 与 npm -h，均可以运行。</p>
<h3 id="2-安装基因组浏览器-Jbrowe"><a href="#2-安装基因组浏览器-Jbrowe" class="headerlink" title="2 安装基因组浏览器 Jbrowe"></a>2 安装基因组浏览器 Jbrowe</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"># make a directory that this user can write to</span><br><span class="line"># for ubuntu&#x2F;Debian that is &#x2F;var&#x2F;www&#x2F;; for centos is &#x2F;var&#x2F;www&#x2F;html</span><br><span class="line">sudo mkdir &#x2F;var&#x2F;www&#x2F;html&#x2F;jbrowse;</span><br><span class="line">sudo chown &#96;whoami&#96; &#x2F;var&#x2F;www&#x2F;html&#x2F;jbrowse;</span><br><span class="line"></span><br><span class="line"># cd into it</span><br><span class="line">cd &#x2F;var&#x2F;www&#x2F;html&#x2F;jbrowse;</span><br><span class="line"></span><br><span class="line"># fetch a JBrowse release zip file</span><br><span class="line">curl -O http:&#x2F;&#x2F;jbrowse.org&#x2F;releases&#x2F;jbrowse-1.16.10-release.tar.gz</span><br><span class="line"></span><br><span class="line"># unzip it and cd into it</span><br><span class="line">unzip JBrowse-1.12.3.zip</span><br><span class="line">cd JBrowse-1.12.3</span><br><span class="line"></span><br><span class="line"># run setup.sh, quick start with example data</span><br><span class="line">.&#x2F;setup.sh</span><br></pre></td></tr></table></figure>

<p>运行 setup.sh 自动安装成功，setup.log 文件如下所示。./setup.sh 这一步需要安装很多 perl 依赖库，需要的时间可能有点久。如果报错，请查看 setup.log 文件，一般是系统可能缺某个依赖库。</p>
<p><img src="https://genehub.files.wordpress.com/2020/10/jbrowse.jpg?w=1024"></p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5ae89e8a385jbrowse.png?w=1024"></p>
<p>到这一步是无法从其他客户端访问服务器上的 JBrowse 的，需要在服务器上面安装 Web 服务器。比如 appache/xampp/nginx 等。</p>
<p>xampp案例见：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/1dddec8dc29c">https://www.jianshu.com/p/1dddec8dc29c</a><br>nginx案例见：<a target="_blank" rel="noopener" href="https://yq.aliyun.com/articles/650480">https://yq.aliyun.com/articles/650480</a></p>
<h3 id="3-安装-Web-服务器-Nginx"><a href="#3-安装-Web-服务器-Nginx" class="headerlink" title="3 安装 Web 服务器 Nginx"></a>3 安装 Web 服务器 Nginx</h3><p>我使用 nginx 做测试的。下载 ngnix 源文件 <a target="_blank" rel="noopener" href="http://nginx.org/download/nginx-1.18.0.tar.gz">nginx-1.18.0.tar.gz</a>，root 权限下，解压在 <code>/usr/local/ngnix</code> 下。</p>
<p>运行 <code>./configure &amp;&amp; make &amp;&amp; make install</code> 成功安装nginx（默认web端口是80），然后就可以在客户端浏览器输入服务器地址，看到初始页面：</p>
<p><img src="https://genehub.files.wordpress.com/2020/10/nginx-default-page.jpg?w=834"></p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003182915.jpg?w=795"></p>
<p>上方的截图所显示的 <em>Welcome to Nginx</em> 初始网页，实际上是由 nginx 安装目录下的 nginx.conf 配置文件控制的，红色标记处，<code>location</code> 字样下的 <code>index.html</code> 是指默认的 <code>/usr/local/nginx/html/index.html</code>。</p>
<p>nginx安装参考：<a target="_blank" rel="noopener" href="https://blog.csdn.net/t8116189520/article/details/81909574">https://blog.csdn.net/t8116189520/article/details/81909574</a> （实测：修改nginx.conf里面端口号之后暂时无法访问。需要linux系开放相应的端口。见下文。）</p>
<p>下面是启动与关闭 nginx 的操作。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"># 停止运行 nginx </span><br><span class="line">&#x2F;usr&#x2F;local&#x2F;nginx&#x2F;sbin&#x2F;nginx -s stop  </span><br><span class="line"></span><br><span class="line"># 重启（reload）如果报错，就要先输入下面的命令&#96;</span><br><span class="line">&#x2F;usr&#x2F;local&#x2F;nginx&#x2F;sbin&#x2F;nginx -c &#x2F;usr&#x2F;local&#x2F;nginx&#x2F;conf&#x2F;nginx.conf  </span><br><span class="line"></span><br><span class="line"># 重启 nginx</span><br><span class="line">&#x2F;usr&#x2F;local&#x2F;nginx&#x2F;sbin&#x2F;nginx -s reload</span><br></pre></td></tr></table></figure>

<h3 id="4-设置-nginx，使-Jbrowse-可以在-Web-浏览器访问"><a href="#4-设置-nginx，使-Jbrowse-可以在-Web-浏览器访问" class="headerlink" title="4 设置 nginx，使 Jbrowse 可以在 Web 浏览器访问"></a>4 设置 nginx，使 Jbrowse 可以在 Web 浏览器访问</h3><p>那么如何将 Jbrowse 相关文档经过 Nginx 平台呈现在客户端的web界面？需要修改 nginx的配置文件：<code>/usr/local/nginx/conf/nginx.conf</code>，root权限下vim这个配置文件，在 http 内容里面加入下面的文字。端口号可以自己设置，比如8080、8082等。location指向<code>/var/www/html/jbrowse/index.html</code></p>
<p>查询资料的过程中发现其他人安装nginx之后有这两个目录 <code>/etc/nginx/sites-available/</code> 和 <code>/etc/nginx/sites-enabled</code>，我安装之后是没有的，后面发现并没有影响。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">server &#123;</span><br><span class="line">    listen     8000;</span><br><span class="line">    server\_name  192.168.1.108;</span><br><span class="line">    location &#x2F; &#123;</span><br><span class="line">        root   &#x2F;var&#x2F;www&#x2F;html&#x2F;jbrowse;</span><br><span class="line">        index  index.html;</span><br><span class="line">    &#125;</span><br></pre></td></tr></table></figure>

<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003183445.jpg"></p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003183654.jpg"></p>
<p>按照上图修改并保存，重启 nginx 理论上可以在客户端打开 192.168.1.108:8000 即可看到 jbrowse，可是我遇到了打不开的问题。原来不仅需要在 nginx 里面设置端口，还需要linux系开放相应的端口：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">iptables -A INPUT -ptcp --dport 8000 -j ACCEPT</span><br></pre></td></tr></table></figure>

<p>然后就顺利打开了 Jbrowse 浏览器，可以看到欢迎页面。</p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003184126.jpg?w=852"></p>
<p>访问 <a target="_blank" rel="noopener" href="http://192.168.1.108/index.html?data=sample_data/json/volvox">http://192.168.1.108/index.html?data=sample_data/json/volvox</a> 可以看到示例页面：</p>
<p><img src="http://blog.biochen.com/wp-content/uploads/2016/06/JBrowse_1.png"></p>
<h3 id="5-如何配置自己的基因组"><a href="#5-如何配置自己的基因组" class="headerlink" title="5 如何配置自己的基因组"></a>5 如何配置自己的基因组</h3><p>基因组浏览器一般展示如下文件</p>
<ol>
<li>reference genome sequence (.fasta)</li>
<li>gene/mRNA/ncRNA annotation file (.gff3)</li>
<li>ChIP Peak file (.bed)</li>
<li>RNA expression file (?)</li>
</ol>
<p>下图所示，JBrowse 浏览器窗口左侧是一个目录，即 Available Tracks，下面通常包括 Reference、Annotation、RNA expression、Histone PTM等栏目，是需要逐个配置的。控制这些“栏目”的配置文件是<code>/var/www/html/jbrowse/data/trackList.json</code>。</p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003184221.jpg?w=1024"></p>
<p><strong>载入参考基因组</strong></p>
<p>我们需要准备一个物种的基因组 fasta 文件，在其他应用场景下 RNA 和 protein 都没有问题。通过 <code>prepare-refseqs.pl</code> 格式化生成的track，这为后续所有添加的文件提供一个坐标系，在网页端参考序列就可以被加载浏览，放大后连碱基序列可以显示出来。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># jbrowse家目录：&#x2F;var&#x2F;www&#x2F;html&#x2F;jbrowse</span><br><span class="line"># 以下操作在jbrowse家目录，序列文件根据实际情况修改</span><br><span class="line">.&#x2F;bin&#x2F;prepare-refseqs.pl --fasta ~&#x2F;database&#x2F;Glycine_max&#x2F;Gmax_275_Wm82_a2.v1&#x2F;Gmax_275_v2.0.softmasked.fa</span><br><span class="line"># .&#x2F;bin&#x2F;prepare-refseqs.pl --fasta ~&#x2F;lyrata&#x2F;Sequence&#x2F;Alyrata_384_v1.fa</span><br></pre></td></tr></table></figure>
<p>这就会在当前目录（/var/www/html/jbrowse）生成一个 data 文件夹，直接访问IP地址所看到的序列就来源于该文件夹。</p>
<p><strong>载入基因注释信息</strong></p>
<p>特征序列一般以”gff|gbk|bed”格式存放，用于注明序列的信息。所需工具为 flatfile-to-json.pl，注意 <code>--trackLabel</code> 参数可以设置 Tracks。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">.&#x2F;bin&#x2F;flatfile-to-json.pl --gff &#x2F;home&#x2F;fenglei&#x2F;database&#x2F;Glycine_max&#x2F;Gmax_275_Wm82_a2.v1&#x2F;Gmax_275_Wm82.a2.v1.gene.gff3 --trackLabel &#39;Gene_a2.v1&#39;</span><br><span class="line"># .&#x2F;bin&#x2F;flatfile-to-json.pl --gff ~&#x2F;Athalina&#x2F;TAIR10&#x2F;TAIR10_GFF3_genes.gtf --tracklabel gene --out .&#x2F;Athaliana&#x2F;   </span><br><span class="line"># 结果是在当前目录下生成data，data里包括序列track配置文件. 同样可以用--out参数输出到指定文件夹。</span><br></pre></td></tr></table></figure>

<p>注意，基因组浏览器有一个搜索框，在完成上述操作之后可以浏览基因组和基因信息，也可以输入 <code>Chr02:123-456</code> 这样的坐标查看我们感兴趣的区域，但是还不能使用搜索功能。需要在上两步的基础上运行下面程序才可。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">.&#x2F;bin&#x2F;generate-names.pl</span><br></pre></td></tr></table></figure>

<p> <strong>载入bed格式数据</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">.&#x2F;bin&#x2F;flatfile-to-json.pl --bed .&#x2F;ChIP_peaks&#x2F;CCK9_peaks.xls.bed --trackType CanvasFeatures --trackLabel BED_CCK9_p0.01</span><br><span class="line">.&#x2F;bin&#x2F;flatfile-to-json.pl --bed .&#x2F;ChIP_peaks&#x2F;CSK9_peaks.xls.bed --trackType CanvasFeatures --trackLabel BED_CSK9_p0.01</span><br><span class="line">.&#x2F;bin&#x2F;flatfile-to-json.pl --bed .&#x2F;ChIP_peaks&#x2F;CCK9_peaks.unfilter.bed --trackType CanvasFeatures --trackLabel BED_CCK9_all</span><br><span class="line">.&#x2F;bin&#x2F;flatfile-to-json.pl --bed .&#x2F;ChIP_peaks&#x2F;CSK9_peaks.unfilter.bed --trackType CanvasFeatures --trackLabel BED_CSK9_all</span><br></pre></td></tr></table></figure>
<p><strong>载入bigwig格式的peak信息</strong></p>
<p>直接在data目录下新建一个 histone_BigWig 文件夹，把需要展示的bigwig文件放置在里面。然后编辑 /var/www/html/jbrowse/data/trackList.json 文件，逐个设置。</p>
<p>每一个bigwig文件，都需要增加下面这样的代码块。我针对样品设计和网页端展示效果，做了很多次调整。下面仅仅列出某个文件的配置信息。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">&#123;</span><br><span class="line">   &quot;autoscale&quot; : &quot;local&quot;,</span><br><span class="line">   &quot;category&quot; : &quot;Histone&quot;,  # 设定文件所属的 Track</span><br><span class="line">   &quot;key&quot; : &quot;BigWig input control&quot;,</span><br><span class="line">   &quot;label&quot; : &quot;BigWig input control&quot;, # 设定在网页上显示的标签</span><br><span class="line">   &quot;logScaleOption&quot; : true,</span><br><span class="line">   &quot;maxExportSpan&quot; : 500000,</span><br><span class="line">   &quot;max_score&quot; : 6,  # 设定所需展示的最大数值。大于这个阈值的数字就按照该值展示。消除极大值带来的展示问题。</span><br><span class="line">   &quot;metadata&quot; : &#123;&#125;,</span><br><span class="line">   &quot;min_score&quot; : 0,</span><br><span class="line">   &quot;scale&quot; : &quot;log&quot;,  # 设定是否以log形式展示数值。因为测序数据经常有个别极大值，log处理更利于展示</span><br><span class="line">   &quot;storeClass&quot; : &quot;JBrowse&#x2F;Store&#x2F;SeqFeature&#x2F;BigWig&quot;,  # 什么意思？表示这个代码块需要展示的是bigWig格式的文件。</span><br><span class="line">   &quot;style&quot; : &#123;</span><br><span class="line">      &quot;height&quot; : 60,  # 设定展示条带的高度。因为有多个文件要展示，如何设定其高度，将使浏览器打开后的视角比较舒适。</span><br><span class="line">      &quot;neg_color&quot; : &quot;red&quot;,  # 负值的颜色。</span><br><span class="line">      &quot;origin_color&quot; : &quot;#888&quot;,</span><br><span class="line">      &quot;pos_color&quot; : &quot;#DDB6A1&quot;,  # 正数值的颜色。</span><br><span class="line">      &quot;variance_band_color&quot; : &quot;rgba(0,0,0,0.3)&quot;</span><br><span class="line">   &#125;,</span><br><span class="line">   &quot;type&quot; : &quot;JBrowse&#x2F;View&#x2F;Track&#x2F;Wiggle&#x2F;XYPlot&quot;,</span><br><span class="line">   &quot;urlTemplate&quot; : &quot;histone_BigWig&#x2F;CCinput.bw&quot;   # 这里指定文件路径。</span><br><span class="line">&#125;,</span><br></pre></td></tr></table></figure>
<p>配置完成之后就类似下面的结果了。</p>
<p><img src="https://ics.hutton.ac.uk/wp-content/uploads/2015/07/jbrowse.png"></p>
<p><strong>参考资料</strong><br><a target="_blank" rel="noopener" href="https://www.plob.org/article/14590.html">JBrowse使用说明：参考基因组格式化</a></p>
<p>-END-</p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2020/10/15/nas-%E5%AE%89%E8%A3%85%E4%B8%8E%E4%BD%BF%E7%94%A8/">NAS 安装与使用</a><a class="next" href="/2020/09/16/ai-%E4%BD%9C%E5%9B%BE%E7%AC%94%E8%AE%B0/">AI 作图笔记</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/01/08/Bioinfo-Installation-of-BUSCO-v4/">Bioinfo | Installation of BUSCO v4</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/05/Linux-Installation-of-Tomcat-9/">Linux | Installation of Tomcat 9, Java 8, and MySQL 5.7.20</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/04/Remove-Background-from-Image/">Remove Background from Image</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/01/%E5%9C%A8Linux%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E4%BD%BF%E7%94%A8Nginx%E9%83%A8%E7%BD%B2Hexo%E7%AB%99%E7%82%B9/">在Linux服务器上使用Nginx部署Hexo站点</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/29/Linux%EF%BC%9A%E4%BD%BF%E7%94%A8Modules%E7%AE%A1%E7%90%86%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/">Linux：使用Environment Modules管理环境变量</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E4%BD%BF%E7%94%A8Snakemake%E6%90%AD%E5%BB%BA%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/">使用Snakemake搭建生物信息分析流程</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9%E7%9A%84SAM-BAM%E6%96%87%E4%BB%B6%E4%B8%AD%EF%BC%8C%E6%80%8E%E4%B9%88%E7%90%86%E8%A7%A3MAPQ%E5%80%BC/">基因组比对的SAM/BAM文件中，怎么理解MAPQ值</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/25/Linux%EF%BC%9Avim%E8%AE%BE%E7%BD%AE/">Linux：vim设置</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/Hi-C%EF%BC%9A3d-dna%E6%8C%82%E8%BD%BD%E6%9F%93%E8%89%B2%E4%BD%93/">Hi-C：3d-dna挂载染色体</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/%E4%B8%80%E4%BA%9BFASTA-FASTQ%E5%BA%8F%E5%88%97%E5%A4%84%E7%90%86%E5%B7%A5%E5%85%B7/">一些FASTA/FASTQ序列处理工具</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>