<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta property="og:type" content="website">
<meta property="og:title" content="FENG&#39;s Notes">
<meta property="og:url" content="http://example.com/page/9/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/page/9/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2019/08/08/%E6%A4%8D%E7%89%A9%E5%9F%BA%E5%9B%A0%E7%BB%84ncrna%E9%A2%84%E6%B5%8B%EF%BC%88trna%E3%80%81rrna%E3%80%81snrna%E3%80%81mirna%EF%BC%89/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2019/08/08/%E6%A4%8D%E7%89%A9%E5%9F%BA%E5%9B%A0%E7%BB%84ncrna%E9%A2%84%E6%B5%8B%EF%BC%88trna%E3%80%81rrna%E3%80%81snrna%E3%80%81mirna%EF%BC%89/" class="post-title-link" itemprop="url">植物基因组ncRNA预测（tRNA、rRNA、snRNA、miRNA）</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2019-08-08 15:49:22" itemprop="dateCreated datePublished" datetime="2019-08-08T15:49:22+08:00">2019-08-08</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>引用文献 <img src="https://genehub.files.wordpress.com/2019/08/e883a1e69da8ncrnae989b4e5ae9ae696b9e6b395.png" alt="胡杨ncRNA鉴定方法" /></p>
<h2 id="一-trna"><a class="markdownIt-Anchor" href="#一-trna"></a> 一、tRNA</h2>
<p>先安装tRNAscan-SE，然后运行如下命令即可。</p>
<p>/xxx/biosoftwares/tRNAscan-SE-2.0/tRNAscan-SE -G -o gDNA.tRNA.out -f gDNA.tRNA.scdstr -m gDNA.tRNA.stat genome.sm.fa</p>
<p>结果示意如下</p>
<p>Sequence tRNA Bounds tRNA Anti Intron Bounds Cove<br />
Name tRNA # Begin End Type Codon Begin End Score</p>
<hr />
<p>ChrM 1 12693 12766 Met CAT 0 0 63.37<br />
ChrM 2 75688 75758 Cys GCA 0 0 48.76<br />
ChrM 3 77742 77815 His GTG 0 0 56.52<br />
ChrM 4 90837 90909 Cys GCA 0 0 59.37<br />
ChrM 5 102551 102624 Asp GTC 0 0 68.25<br />
ChrM 6 184436 184507 Gly GCC 0 0 70.59<br />
ChrM 7 197972 198043 Glu TTC 0 0 65.49<br />
ChrM 8 198963 199035 Met CAT 0 0 70.15<br />
ChrM 9 237157 237228 Asn GTT 0 0 71.98<br />
ChrM 10 238160 238242 Tyr GTA 0 0 62.80<br />
ChrM 11 261058 261136 Lys TTT 261096 261101 34.22<br />
ChrM 12 288825 288906 Met CAT 0 0 73.07<br />
ChrM 13 312625 312696 Gln TTG 0 0 64.58<br />
ChrM 14 335330 335403 Met CAT 0 0 73.76<br />
ChrM 15 317035 316963 Phe AAA 0 0 30.83<br />
ChrM 16 285436 285349 Ser GCT 0 0 55.78<br />
ChrM 17 284968 284895 Phe GAA 0 0 67.07<br />
ChrM 18 284630 284556 Pro TGG 0 0 64.26<br />
ChrM 19 260713 260640 Trp CCA 0 0 76.23<br />
ChrM 20 3494 3422 Lys TTT 0 0 81.82</p>
<h2 id="二-rrna"><a class="markdownIt-Anchor" href="#二-rrna"></a> 二、rRNA</h2>
<p>不同植物中的rRNA序列高度保守，基于blast比对即可预测新物种的rRNA。植物基因组rRNA包含18S、28S、5S和5.8S四种。我下载了大豆或拟南芥的rRNA序列作为比对文件。</p>
<p>&gt;5S rRNA 556bp X15199.1 Soybean 5S rDNA gene for 5S rRNA</p>
<blockquote>
<p>5.8S rRNA Arabidopsis thaliana 15172 AT2G01020.1 Chr2+ 5782-5945<br />
18S rRNA 1807bp X02623.1 Soybean (Glycine max) 18S ribosomal RNA<br />
28S rRNA XR_003264281.1 PREDICTED: Glycine max 28S ribosomal RNA (LOC112999102), rRNA</p>
</blockquote>
<p>比对方法如下，结果还需要根据实际情况进行过滤，将部分匹配的基因组片段过滤掉即可。</p>
<p>blastn -db /xxx_references/index/genome.sm.fa -query Glycine_max_rRNA.fa -evalue 1e-5 -outfmt 7 -num_threads 4 -out Glycine_max_rRNA.fa__genome.7<br />
blastn -db /xxx_references/index/genome.sm.fa -query Glycine_max_rRNA.fa -evalue 1e-5 -outfmt 0 -num_threads 4 -out Glycine_max_rRNA.fa__genome.0</p>
<p>三、snRNA snRNA是真核细胞核内的一组小分子RNA，分子较小，约含50—200nt，其主要作用是与有关蛋白结合形成小核糖核蛋白体(snRNP)，对RNA的前体进行加工。snRNA在核内转录，但snRNP在胞浆组装，发挥功能则又在核内，所以需要跨核膜转运。 snRNA的特点 ① 高度保守，如人和爪蟾的U1snRNA有90%的序列相同。② 其长度在哺乳动物中约为100-215个核苷酸，共分为7类，由于含U丰富，故编号为U1~U7。③ 分析方法是用rfam数据库（cm格式）作参考，通过cmscan比对程序进行比对预测。 先安装infernal程序（内置cmscan）。</p>
<p>wget <a target="_blank" rel="noopener" href="http://eddylab.org/infernal/infernal-1.1.2.tar.gz">http://eddylab.org/infernal/infernal-1.1.2.tar.gz</a><br />
`tar -zxvf infernal-1.1.2.tar.gz <br />
cd infernal-1.1.2<br />
./configure  --prefix=/home/user/local<br />
make<br />
make install<br />
cd easel; make install</p>
<h1 id="测试"><a class="markdownIt-Anchor" href="#测试"></a> 测试`</h1>
<p>$ cmscan --version<br />
Failed to parse command line: No such option “–version”.<br />
Usage: cmscan [-options] <cmdb> <seqfile><br />
where basic options are:<br />
-h        : show brief help on version and usage<br />
-g        : configure CM for glocal alignment [default: local]<br />
-Z <x>    : set search space size in *Mb* to <x> for E-value calculations  (x&gt;0)<br />
–devhelp : show list of otherwise hidden developer/expert options<br />
To see more help on available options, do cmscan -h</p>
<p>再下载rfam数据库，并将Rfam.cm做成索引文件。</p>
<p>wget <a href="ftp://ftp.ebi.ac.uk/pub/databases/Rfam/14.1/Rfam.clanin">ftp://ftp.ebi.ac.uk/pub/databases/Rfam/14.1/Rfam.clanin</a><br />
wget <a href="ftp://ftp.ebi.ac.uk/pub/databases/Rfam/14.1/Rfam.cm.gz">ftp://ftp.ebi.ac.uk/pub/databases/Rfam/14.1/Rfam.cm.gz</a><br />
gunzip Rfam.cm.gz<br />
cmpress <a target="_blank" rel="noopener" href="http://Rfam.cm">Rfam.cm</a></p>
<h1 id="出现如下文件"><a class="markdownIt-Anchor" href="#出现如下文件"></a> 出现如下文件</h1>
<p>ls Rfam.*<br />
Rfam.clanin <a target="_blank" rel="noopener" href="http://Rfam.cm">Rfam.cm</a> Rfam.cm.i1f Rfam.cm.i1i Rfam.cm.i1m Rfam.cm.i1p</p>
<p>计算基因组大小，后面cmscan要用到-G参数</p>
<p>esl-seqstat my-genome.fa</p>
<p>比如基因组1000M，那么-G参数为 1000M*2/1M=2000，这个数据根据实际情况调整。 运行比对程序进行ncRNA预测</p>
<p># Rfam12.2.claninfo 为下载的claninfo文件，需提供所在路径</p>
<h1 id="rfamcm"><a class="markdownIt-Anchor" href="#rfamcm"></a> <a target="_blank" rel="noopener" href="http://Rfam.cm">Rfam.cm</a> 下载的cm文件</h1>
<h1 id="my-genomefa-待查询序列"><a class="markdownIt-Anchor" href="#my-genomefa-待查询序列"></a> my-genome.fa 待查询序列</h1>
<h1 id="my-genomecmscan-输出结果"><a class="markdownIt-Anchor" href="#my-genomecmscan-输出结果"></a> my-genome.cmscan 输出结果</h1>
<h1 id="my-genometblout-有一个输出结果"><a class="markdownIt-Anchor" href="#my-genometblout-有一个输出结果"></a> my-genome.tblout 有一个输出结果</h1>
<h1 id="对500m大小的输入序列48线程需要7个小时最好放入后台"><a class="markdownIt-Anchor" href="#对500m大小的输入序列48线程需要7个小时最好放入后台"></a> 对500M大小的输入序列，48线程，需要7个小时，最好放入后台</h1>
<h1 id="esl-seqstat-my-genomefa-awk-if0~total-print-int421000000-1751"><a class="markdownIt-Anchor" href="#esl-seqstat-my-genomefa-awk-if0~total-print-int421000000-1751"></a> esl-seqstat my-genome.fa  awk ‘{if($0~/^Total/) print int($4*2/1000000);}’ -----&gt; 1751</h1>
<p>cmscan -Z 1751 --cut_ga --rfam --nohmmonly --tblout my-genome.tblout --fmt 2 --clanin Rfam.clanin <a target="_blank" rel="noopener" href="http://Rfam.cm">Rfam.cm</a> my-genome.fa &gt; my-genome.cmscan</p>
<p>cmscan结果文件示意如下，包含预测的ncRNA列表和RNA比对结构示意，其实不仅包含了snRNA，还有miRNA、rRNA和tRNA。 另有一个my-genome.tblout文件，就是tab分割的表格，更加简洁。</p>
<p>Query: scaffold4 [L=2607366]<br />
Hit scores:<br />
rank E-value score bias modelname start end mdl trunc gc description</p>
<hr />
<p>(1) ! 6.3e-19 81.7 0.0 MIR408 2312792 2312644 - cm no 0.48 microRNA MIR408<br />
(2) ! 1.9e-17 88.4 0.0 snoZ107_R87 2081687 2081578 - cm no 0.46 Small nucleolar RNA Z107/R87<br />
(3) ! 2.4e-16 73.0 0.0 MIR164 1105583 1105483 - cm no 0.54 microRNA MIR164<br />
(4) ! 1.2e-13 71.7 0.0 snoR30 2081522 2081420 - cm no 0.44 Small nucleolar RNA R30/Z108<br />
(5) ! 1.3e-11 65.1 0.0 tRNA 2140920 2140992 + cm no 0.59 tRNA<br />
(6) ! 1.3e-11 65.1 0.0 tRNA 2245335 2245407 + cm no 0.59 tRNA<br />
(7) ! 1.6e-11 64.8 0.0 tRNA 223678 223759 + cm no 0.56 tRNA<br />
(8) ! 1.7e-11 64.7 0.0 tRNA 374728 374800 + cm no 0.56 tRNA<br />
(9) ! 4e-11 63.3 0.0 tRNA 482023 481943 - cm no 0.53 tRNA<br />
(10) ! 7.3e-11 71.3 0.0 snoR41 705999 705902 - cm no 0.40 Small nucleolar RNA R41<br />
(11) ! 3.2e-10 60.1 0.0 tRNA 624224 624152 - cm no 0.58 tRNA<br />
(12) ! 4.3e-10 59.6 0.0 tRNA 793379 793309 - cm no 0.59 tRNA<br />
(13) ! 1.7e-08 53.9 0.0 tRNA 767498 767569 + cm no 0.57 tRNA<br />
(14) ! 0.00018 43.5 0.1 SNORD34 2081329 2081243 - cm no 0.33 Small nucleolar RNA SNORD34</p>
<p>Hit alignments:</p>
<blockquote>
<blockquote>
<p>MIR408 microRNA MIR408<br />
rank E-value score bias mdl mdl from mdl to seq from seq to acc trunc gc</p>
</blockquote>
</blockquote>
<hr />
<p>(1) ! 6.3e-19 81.7 0.0 cm 1 136 [] 2312792 2312644 - … 0.95 no 0.48</p>
<p>v NC<br />
::::::::::::::::&lt;&lt;&lt;-&lt;&lt;&lt;&lt;&lt;&lt;&lt;-&lt;&lt;&lt;&lt;&lt;&lt;-&lt;&lt;&lt;&lt;&lt;-&lt;—&lt;&lt;_____________…______________ CS<br />
MIR408 1 aauUGGugAugAgAugGAGACaGGGAagAgGCAGaGCAuGaGAUGGaGCUaucAAcaAau…ugUGAgAaaaagaU 74<br />
A+U G+GA +AGA + AG AGGGAA AGGCAGAGCAUG :AUG AGC AUCAACA A GUGAG +AA++<br />
scaffold4 2312792 AGUCAGAGACAAGACAAAGGUAGGGAACAGGCAGAGCAUG-GAUGGAGCCAUCAACAGAAuagagucaagaaaccgagAGUGAGGGAAGA-- 2312704<br />
****************************************.***************88779*************9988777777777766… PP</p>
<p>v NC<br />
______________&gt;&gt;----&gt;&gt;&gt;&gt;&gt;&gt;-&gt;&gt;&gt;&gt;&gt;&gt;-&gt;&gt;&gt;&gt;&gt;&gt;&gt;-&gt;&gt;&gt;::::::::::::::::: CS<br />
MIR408 75 GgaauUGUUgUuGCuCcCuCCCaUGCACUGCcUCuUCCCuGGCUCCuuucuCuccucuuUuu 136</p>
<ul>
<li>++UGUU +GCU C + :CAUGCACUGCCUCUUCCCU GCU U +U+U+C+ U+U +<br />
scaffold4 2312703 -AGUCUGUUACGGCUUC-AAUCAUGCACUGCCUCUUCCCUGGCUCUAUCUUUUUCCAUCUCC 2312644<br />
.8899**********99.999***********************8899999999******** PP</li>
</ul>
<blockquote>
<blockquote>
<p>snoZ107_R87 Small nucleolar RNA Z107/R87<br />
rank E-value score bias mdl mdl from mdl to seq from seq to acc trunc gc</p>
</blockquote>
</blockquote>
<hr />
<p>(2) ! 1.9e-17 88.4 0.0 cm 1 117 [] 2081687 2081578 - … 0.94 no 0.46</p>
<p>NC<br />
&lt;&lt;&lt;&lt;&lt;&lt;________________.___________________________________________________________________ CS<br />
snoZ107_R87 1 GAugGCaGUGAuGACUuGGUAA.uAUUCAAGCUCAACAGACCaaAuuacAGgucUUUCUCuauggcuuuuccuaugGGAUuGaUUUGUaU 89<br />
GAU::CAGUGAUGAC UAA +AUUCAAGCUCAACAGACC+ AUU CA G +UUUCUCU+U ++U G GAUU UUUG +U<br />
scaffold4 2081687 GAUCACAGUGAUGACCACAUAAaAAUUCAAGCUCAACAGACCGGAUUGCACGGUUUUCUCUCU-------UUUGGGAGAUUACUUUGCGU 2081605<br />
***************999987615677*******************************99665…9999**************** PP</p>
<p>NC<br />
______________________&gt;&gt;&gt;&gt;&gt;&gt; CS<br />
snoZ107_R87 90 GcCGAUaaUcCCGCUGAACuGAGCcaUC 117<br />
G+CGA A CCCGCUGAACUGAG::AUC<br />
scaffold4 2081604 GUCGAC-ACCCCGCUGAACUGAGUGAUC 2081578<br />
*****8.799****************** PP</p>
<p>如何对结果进行分类？首先下载Rfam家族的注释，点击<a target="_blank" rel="noopener" href="http://rfam.xfam.org/search#tabview=tab5">http://rfam.xfam.org/search#tabview=tab5</a>，选择所有复选框，提交，把得到的表格拷贝下来，整理成TAB键分割的格式。并把第三列拆开，取出类型, 存储为 <code>Rfam_anno.txt</code>。</p>
<p>Accession ID Type Description<br />
RF00001 5S_rRNA Gene; rRNA 5S ribosomal RNA<br />
RF00002 5_8S_rRNA Gene; rRNA 5.8S ribosomal RNA<br />
RF00003 U1 Gene; snRNA; splicing U1 spliceosomal RNA<br />
RF00004 U2 Gene; snRNA; splicing U2 spliceosomal RNA<br />
RF00005 tRNA Gene; tRNA tRNA<br />
RF00006 Vault Gene Vault RNA<br />
RF00007 U12 Gene; snRNA; splicing U12 minor spliceosomal RNA<br />
RF00008 Hammerhead_3 Gene; ribozyme Hammerhead ribozyme (type III)<br />
RF00009 RNaseP_nuc Gene; ribozyme Nuclear RNase P<br />
RF00010 RNaseP_bact_a Gene; ribozyme Bacterial RNase P class A<br />
RF00011 RNaseP_bact_b Gene; ribozyme Bacterial RNase P class B<br />
RF00012 U3 Gene; snRNA; snoRNA; CD-box Small nucleolar RNA U3<br />
RF00013 6S Gene 6S / SsrS RNA<br />
RF00014 DsrA Gene; sRNA DsrA RNA<br />
RF00015 U4 Gene; snRNA; splicing U4 spliceosomal RNA<br />
RF00016 SNORD14 Gene; snRNA; snoRNA; CD-box Small nucleolar RNA SNORD14<br />
RF00017 Metazoa_SRP Gene Metazoan signal recognition particle RNA</p>
<p>grep其中snRNA如下</p>
<p>$ cat xfam_family_types.tab  grep snRNA  awk -F “\t” ‘{print $3}’  sort  uniq<br />
Cis-reg<br />
Gene; antisense<br />
Gene; snRNA<br />
Gene; snRNA; snoRNA; CD-box<br />
Gene; snRNA; snoRNA; HACA-box<br />
Gene; snRNA; snoRNA; scaRNA<br />
Gene; snRNA; splicing<br />
Gene; sRNA</p>
<p>将CD-box、HACA-box和splicing对应的部分统计出来就可以做成如下文献对应的图了。 <img src="https://genehub.files.wordpress.com/2019/08/e883a1e69da8ncrnae989b4e5ae9ae7bb93e69e9c.png" alt="胡杨ncRNA鉴定结果" /></p>
<h2 id="四-mirna"><a class="markdownIt-Anchor" href="#四-mirna"></a> 四、miRNA</h2>
<p>可以直接用上面的结果，也可以用miRNA测序的数据。   参考资料： [1] Chen Tong: <a target="_blank" rel="noopener" href="http://blog.genesino.com/2017/06/Rfam/">http://blog.genesino.com/2017/06/Rfam/</a> [2] Chen Lianfu: <a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=2185">http://www.chenlianfu.com/?p=2185</a> [3] tiehan: <a target="_blank" rel="noopener" href="http://blog.sina.com.cn/s/blog%5C_670445240102uxhu.html">http://blog.sina.com.cn/s/blog\_670445240102uxhu.html</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2019/07/31/Linux-how-to-upgrade-git/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2019/07/31/Linux-how-to-upgrade-git/" class="post-title-link" itemprop="url">Linux更新git</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2019-07-31 11:46:27" itemprop="dateCreated datePublished" datetime="2019-07-31T11:46:27+08:00">2019-07-31</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/" itemprop="url" rel="index"><span itemprop="name">未分类</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>系统自带的git无法下载：</p>
<p>error: error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protocol</p>
<p>查询之后据说是 GitHub 不再支持TLSv1/TLSv1.1了，公告地址：<a target="_blank" rel="noopener" href="https://githubengineering.com/crypto-removal-notice/">Weak cryptographic standards removal notice</a> 于是安装新版git（<a target="_blank" rel="noopener" href="https://mirrors.edge.kernel.org/pub/software/scm/git/git-2.15.0.tar.gz%EF%BC%89%EF%BC%8Cmake%E7%9A%84%E8%BF%87%E7%A8%8B%E4%B8%AD%E6%8A%A5%E9%94%99%EF%BC%8C%E6%98%BE%E7%A4%BA%E7%B3%BB%E7%BB%9F%E7%9A%84openssl%E6%9C%89%E9%97%AE%E9%A2%98%E3%80%82">https://mirrors.edge.kernel.org/pub/software/scm/git/git-2.15.0.tar.gz），make的过程中报错，显示系统的openssl有问题。</a></p>
<p>error OPENSSL_ALGORITHM_DEFINES no longer supported</p>
<p>重新安装openssl到/usr/local/ 安装openssl到 /usr/local/openssl 目录，安装之后，编译；</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">./config shared zlib  --prefix=/usr/local/openssl &amp;&amp; make &amp;&amp; make install</span><br></pre></td></tr></table></figure>
<p>安装openssl就重新编译git，还报错（实际上上面写成<code>--prefix=/usr/local应该就没问题了</code>）</p>
<p>/usr/local/include/openssl/opensslconf.h  error OPENSSL_ALGORITHM_DEFINES no longer supported</p>
<p>于是进去这个目录，将opensslconf.h做了个软连接到新安装的文件目录/usr/local/openssl/include/openssl/opensslconf.h</p>
<p>ln -s /usr/local/openssl/include/openssl/opensslconf.h /usr/local/include/openssl/opensslconf.h</p>
<p>接下来make的过程还遇到了perl的问题。在将git安装文件里面的Makefile，将里面的/usr/bin/perl修改成自己目录的perl就行了。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2019/04/01/%E8%BD%AC%EF%BC%9A%E8%87%AA%E5%8A%A8%E5%8C%96%E8%BF%90%E8%A1%8Corthomcl/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2019/04/01/%E8%BD%AC%EF%BC%9A%E8%87%AA%E5%8A%A8%E5%8C%96%E8%BF%90%E8%A1%8Corthomcl/" class="post-title-link" itemprop="url">转：自动化运行OrthoMCL</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2019-04-01 13:05:14" itemprop="dateCreated datePublished" datetime="2019-04-01T13:05:14+08:00">2019-04-01</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
                  , 
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>OrthoMCL (<a target="_blank" rel="noopener" href="https://link.jianshu.com/?t=http://orthomcl.org/orthomcl/">http://orthomcl.org/orthomcl/</a>) 是现在用的最多的一款来找直系同源基因（Orthologs）以及旁系同源基因 (Paralog) 的软件。</p>
<p>OrthoMCL 的最新版本是2013年7月公布的v2.0版本，已经很久没更新过了。根据官网的教程至少得十多步才能完成整个运行流程，包括Mysql数据库配置、修改OrthoMCL配置文件、转换序列格式、过滤、比对、解析结果和聚类等步骤，特别麻烦。 OrthoMCL Pipeline (<a target="_blank" rel="noopener" href="https://link.jianshu.com?t=https://github.com/apetkau/orthomcl-pipeline">https://github.com/apetkau/orthomcl-pipeline</a>) 可以很好的帮我们解决这个痛点。Pipeline安装有点复杂，但是安装完成后，使用就方便了。</p>
<p>作者：亮亮就是亮 链接：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/449a51fa3d18">https://www.jianshu.com/p/449a51fa3d18</a> 来源：简书 简书著作权归作者所有，任何形式的转载都请联系作者获得授权并注明出处。</p>
<p>使用OrthoMCL Pipeline 来自动化运行OrthoMCL，简化操作步骤。</p>
<h2 id="安装"><a class="markdownIt-Anchor" href="#安装"></a> 安装</h2>
<p>1</p>
<p>git clone <a target="_blank" rel="noopener" href="https://github.com/apetkau/orthomcl-pipeline.git">https://github.com/apetkau/orthomcl-pipeline.git</a></p>
<p>或者直接下载：<a target="_blank" rel="noopener" href="https://github.com/apetkau/orthomcl-pipeline">https://github.com/apetkau/orthomcl-pipeline</a></p>
<h3 id="perl模块"><a class="markdownIt-Anchor" href="#perl模块"></a> Perl模块</h3>
<p>通过<code>cpanm</code>安装</p>
<p>1</p>
<p>cpanm BioPerl DBD::mysql DBI Parallel::ForkManager YAML::Tiny Set::Scalar Text::Table Exception::Class Test::Most Test::Warn Test::Exception Test::Deep Moose SVG Algorithm::Combinatorics</p>
<p>我是在linux命令行界面输入 perl -MCPAN -e shell ，使用instlall DBI这样的命令安装perl模块的。</p>
<h3 id="依赖软件"><a class="markdownIt-Anchor" href="#依赖软件"></a> 依赖软件</h3>
<ul>
<li><a target="_blank" rel="noopener" href="http://orthomcl.org/common/downloads/software/v2.0/">OrthoMCL</a> 或 <a target="_blank" rel="noopener" href="https://github.com/apetkau/orthomclsoftware-custom">OrthoMCL Custom</a> （能够自定义序列识别符的修改版本）。OrhtoMCL下载并解压之后，bin目录下的Perl程序首行是#!/usr/bin/perl，我将其全部改为用户的perl了，因为我自己安装的perl位置是/home/fenglei/local/bin/perl。</li>
<li><a target="_blank" rel="noopener" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">BLAST</a> (blastall, formatdb) 注意不是NCBI-blast+, 推荐使用<code>2.2.26</code>版本。由于我之前下载过interproscan，在interproscan安装目录下游blastall和formatdb等程序，直接将其路径加入环境变量即可！</li>
<li><a target="_blank" rel="noopener" href="http://www.micans.org/mcl/index.html">MCL</a></li>
</ul>
<p>依赖环境解决后能够顺利运行设置脚本</p>
<p>1<br />
2<br />
3<br />
4<br />
5<br />
6<br />
7<br />
8<br />
9</p>
<p>$ perl scripts/orthomcl-pipeline-setup.pl<br />
Checking for Software dependencies…<br />
Checking for OthoMCL … OK<br />
Checking for formatdb … OK<br />
Checking for blastall … OK<br />
Checking for mcl … OK<br />
Wrote new configuration to orthomcl-pipeline/scripts/…/etc/orthomcl-pipeline.conf<br />
Wrote executable file to orthomcl-pipeline/scripts/…/bin/orthomcl-pipeline<br />
Please add directory orthomcl-pipeline/scripts/…/bin to PATH</p>
<h3 id="数据库设置"><a class="markdownIt-Anchor" href="#数据库设置"></a> 数据库设置</h3>
<h4 id="创建orthomcl用户"><a class="markdownIt-Anchor" href="#创建orthomcl用户"></a> 创建orthomcl用户</h4>
<p>1<br />
2<br />
3<br />
4<br />
5</p>
<p>$ mysql -u root -p<br />
Enter password:<br />
mysql&gt; GRANT SELECT, INSERT, UPDATE, DELETE, CREATE, CREATE VIEW, INDEX, DROP on . to orthomcl;  #创建用户并授权<br />
mysql&gt; set password for orthomcl@localhost = password(‘orthomcl’);  #设置用户密码<br />
mysql&gt; quit;</p>
<h4 id="创建数据库并生成配置脚本"><a class="markdownIt-Anchor" href="#创建数据库并生成配置脚本"></a> 创建数据库并生成配置脚本</h4>
<p>1<br />
2<br />
3<br />
4</p>
<p>$ perl scripts/orthomcl-setup-database.pl --user orthomcl --password orthomcl --host localhost --database orthomcl --outfile orthomcl.conf<br />
Connecting to mysql and creating database **orthmcldb** on host orthodb with user orthomcl …OK<br />
database orthmcl created …OK<br />
Config file **orthomcl.conf** created.</p>
<p>若已有orthomcl数据库，则可添加<code>--no-create-database</code>参数，不新建数据库。</p>
<h2 id="测试数据"><a class="markdownIt-Anchor" href="#测试数据"></a> 测试数据</h2>
<p>1<br />
2<br />
3<br />
4<br />
5<br />
6<br />
7<br />
8<br />
9</p>
<p>$ perl t/test_pipeline.pl -m orthomcl.conf -s fork -t /tmp<br />
Test using scheduler fork</p>
<p>TESTING NON-COMPLIANT INPUT<br />
TESTING FULL PIPELINE RUN 3<br />
README:<br />
Tests case of one gene (in 1.fasta and 2.fasta) not present in other files.<br />
ok 1 - Expected matched returned groups file<br />
…</p>
<h2 id="运行"><a class="markdownIt-Anchor" href="#运行"></a> 运行</h2>
<p>1<br />
2<br />
3<br />
4</p>
<p>$ ./bin/orthomcl-pipeline<br />
Error: no input-dir defined<br />
Usage: orthomcl-pipeline -i [input dir] -o [output dir] -m [orthmcl config] [Options]<br />
…</p>
<p>1</p>
<p>加上–nocompliant就可以将输入文件中的基因名自动调整为具备唯一性的字符串（and <code>--nocompliant</code> adjusts gene names in fasta files to make them unique.）。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ ./bin/orthomcl-pipeline -i ~/test_data/zbl -o ~/test_data/zbl-out -m orthomcl.conf --nocompliant</span><br></pre></td></tr></table></figure>
<p>zbl文件夹里面放入fasta格式的蛋白序列文件，一个基因组一个文件，例如genome1.fasta, genome2.fasta, 文件后缀名必需是.fasta。zbl-out文件夹里面是所有的结果，包括聚类完成的groups文件，orthologs文件，inparalogs文件以及coorthologs文件。</p>
<h2 id="参考来源"><a class="markdownIt-Anchor" href="#参考来源"></a> 参考来源</h2>
<p><a target="_blank" rel="noopener" href="https://github.com/apetkau/orthomcl-pipeline/blob/master/INSTALL.md">https://github.com/apetkau/orthomcl-pipeline/blob/master/INSTALL.md</a> <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/449a51fa3d18">https://www.jianshu.com/p/449a51fa3d18 （亮亮就是亮）</a> <a target="_blank" rel="noopener" href="https://sr-c.github.io/2018/09/18/OrthoMCL-Pipeline/">https://sr-c.github.io/2018/09/18/OrthoMCL-Pipeline/</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2019/03/30/Linux-install-mysql-5-7-25/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2019/03/30/Linux-install-mysql-5-7-25/" class="post-title-link" itemprop="url">Linux安装mysql-5.7.25</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2019-03-30 20:38:24" itemprop="dateCreated datePublished" datetime="2019-03-30T20:38:24+08:00">2019-03-30</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>下面的操作完全是照搬其他人的经验（<a target="_blank" rel="noopener" href="https://segmentfault.com/a/1190000012703513%EF%BC%89%EF%BC%8C%E6%AD%A4%E5%A4%84%E8%AE%B0%E5%BD%95%E5%A4%87%E4%BB%BD%E7%94%A8%E3%80%82%E4%B9%9F%E5%8F%AF%E5%8F%82%E8%80%83%EF%BC%9Ahttps://www.jellythink.com/archives/14">https://segmentfault.com/a/1190000012703513），此处记录备份用。也可参考：https://www.jellythink.com/archives/14</a> ---------------------------------- 由于mysql 5.7.17版本以后 support_files文件夹中无 <strong>my_default.cnf</strong> 文件，所以今天给大家详细描述一下 mysql 5.7.20版本(目前官方最新版)的安装步骤。 第一步：下载mysql最新版</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget http://dev.mysql.com/get/Downloads/MySQL-5.7/mysql-5.7.20-linux-glibc2.12-x86_64.tar.gz</span><br></pre></td></tr></table></figure>
<p>第二步：在/usr/local/中解压压缩包，并改名为mysql</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">cd /usr/local/</span><br><span class="line"></span><br><span class="line">tar -xzvf /data/software/mysql-5.7.13-linux-glibc2.5-x86_64.tar.gz (我的下载目录为 data/software)</span><br><span class="line"></span><br><span class="line">mv mysql-5.7.13-linux-glibc2.5-x86_64 mysql</span><br></pre></td></tr></table></figure>
<p>第三步：创建用户组mysql，创建用户mysql并将其添加到用户组mysql中，并赋予读写权限</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">groupadd mysql</span><br><span class="line"></span><br><span class="line">useradd -r -g mysql mysql</span><br><span class="line"></span><br><span class="line">chown -R mysql mysql/</span><br><span class="line"></span><br><span class="line">chgrp -R mysql mysql/</span><br></pre></td></tr></table></figure>
<p>第四步：创建配置文件</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">vim /etc/my.cnf</span><br><span class="line"></span><br><span class="line">#复制以下内容</span><br><span class="line"></span><br><span class="line">[client]</span><br><span class="line">port = 3306</span><br><span class="line">socket = /tmp/mysql.sock</span><br><span class="line"></span><br><span class="line">[mysqld]</span><br><span class="line">character_set_server=utf8</span><br><span class="line">init_connect=&#x27;SET NAMES utf8&#x27;</span><br><span class="line">basedir=/usr/local/mysql</span><br><span class="line">datadir=/usr/local/mysql/data</span><br><span class="line">socket=/tmp/mysql.sock</span><br><span class="line">log-error=/var/log/mysqld.log</span><br><span class="line">pid-file=/var/run/mysqld/mysqld.pid</span><br><span class="line">#不区分大小写</span><br><span class="line">lower_case_table_names = 1</span><br><span class="line"></span><br><span class="line">sql_mode=STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION</span><br><span class="line"></span><br><span class="line">max_connections=5000</span><br><span class="line"></span><br><span class="line">default-time_zone = &#x27;+8:00&#x27;</span><br></pre></td></tr></table></figure>
<p>按ESC保存并关闭，输入如下命令</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">:wq!</span><br></pre></td></tr></table></figure>
<p>第五步：初始化数据库</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">#先安装一下这个东东，要不然初始化有可能会报错</span><br><span class="line">yum install libaio</span><br><span class="line">#手动编辑一下日志文件，什么也不用写，直接保存退出</span><br><span class="line">cd /var/log/</span><br><span class="line"></span><br><span class="line">vim mysqld.log</span><br><span class="line">：wq</span><br><span class="line"></span><br><span class="line">chmod 777 mysqld.log</span><br><span class="line">chown mysql:mysql mysqld.log</span><br><span class="line"></span><br><span class="line">/usr/local/mysql/bin/mysqld --initialize --user=mysql --basedir=/usr/local/mysql --datadir=/usr/local/mysql/data --lc_messages_dir=/usr/local/mysql/share --lc_messages=en_US</span><br></pre></td></tr></table></figure>
<p>第六步：查看初始密码</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cat /var/log/mysqld.log</span><br></pre></td></tr></table></figure>
<p>执行后关注最后一点：root@localhost: 这里就是初始密码 第七步：启动服务，进入mysql，修改初始密码，运行远程连接（这里执行完后，密码将变成：你设置的新密码）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">#如果提示必须要修改密码才可以进行操作的话则执行下面操作</span><br><span class="line">set password=password(&#x27;新密码&#x27;);</span><br><span class="line"></span><br><span class="line">flush privileges;</span><br><span class="line"></span><br><span class="line">UPDATE `mysql`.`user` SET `Host` = &#x27;%&#x27;,  `User` = &#x27;root&#x27;  WHERE (`Host` = &#x27;localhost&#x27;) AND (`User` = &#x27;root&#x27;);</span><br><span class="line"></span><br><span class="line">#然后执行如下操作开启mysql服务，以及设置相关权限</span><br><span class="line">cd /var/run/</span><br><span class="line"></span><br><span class="line">mkdir mysqld</span><br><span class="line"></span><br><span class="line">chmod 777 mysqld</span><br><span class="line"></span><br><span class="line">cd mysqld</span><br><span class="line"></span><br><span class="line">touch mysqld.pid</span><br><span class="line"></span><br><span class="line">chmod 777 mysqld.pid</span><br><span class="line"></span><br><span class="line">chown mysql:mysql mysqld.pid </span><br><span class="line"></span><br><span class="line">/usr/local/mysql/support-files/mysql.server start</span><br><span class="line"></span><br><span class="line">/usr/local/mysql/bin/mysql -uroot -p 你在上面看到的初始密码</span><br><span class="line"></span><br><span class="line"># 以下是进入数据库之后的sql语句</span><br><span class="line"> use mysql;</span><br><span class="line"></span><br><span class="line"> UPDATE `mysql`.`user` SET `Host`=&#x27;%&#x27;, `User`=&#x27;root&#x27;, `Select_priv`=&#x27;Y&#x27;, `Insert_priv`=&#x27;Y&#x27;, `Update_priv`=&#x27;Y&#x27;, `Delete_priv`=&#x27;Y&#x27;, `Create_priv`=&#x27;Y&#x27;, `Drop_priv`=&#x27;Y&#x27;, `Reload_priv`=&#x27;Y&#x27;, `Shutdown_priv`=&#x27;Y&#x27;, `Process_priv`=&#x27;Y&#x27;, `File_priv`=&#x27;Y&#x27;, `Grant_priv`=&#x27;Y&#x27;, `References_priv`=&#x27;Y&#x27;, `Index_priv`=&#x27;Y&#x27;, `Alter_priv`=&#x27;Y&#x27;, `Show_db_priv`=&#x27;Y&#x27;, `Super_priv`=&#x27;Y&#x27;, `Create_tmp_table_priv`=&#x27;Y&#x27;, `Lock_tables_priv`=&#x27;Y&#x27;, `Execute_priv`=&#x27;Y&#x27;, `Repl_slave_priv`=&#x27;Y&#x27;, `Repl_client_priv`=&#x27;Y&#x27;, `Create_view_priv`=&#x27;Y&#x27;, `Show_view_priv`=&#x27;Y&#x27;, `Create_routine_priv`=&#x27;Y&#x27;, `Alter_routine_priv`=&#x27;Y&#x27;, `Create_user_priv`=&#x27;Y&#x27;, `Event_priv`=&#x27;Y&#x27;, `Trigger_priv`=&#x27;Y&#x27;, `Create_tablespace_priv`=&#x27;Y&#x27;, `ssl_type`=&#x27;&#x27;, `ssl_cipher`=&#x27;&#x27;, `x509_issuer`=&#x27;&#x27;, `x509_subject`=&#x27;&#x27;, `max_questions`=&#x27;0&#x27;, `max_updates`=&#x27;0&#x27;, `max_connections`=&#x27;0&#x27;, `max_user_connections`=&#x27;0&#x27;, `plugin`=&#x27;mysql_native_password&#x27;, `authentication_string`=&#x27;*6BB4837EB74329105EE4568DDA7DC67ED2CA2AD9&#x27;, `password_expired`=&#x27;N&#x27;, `password_last_changed`=&#x27;2017-11-20 12:41:07&#x27;, `password_lifetime`=NULL, `account_locked`=&#x27;N&#x27; WHERE  (`User`=&#x27;root&#x27;);</span><br><span class="line"></span><br><span class="line"> flush privileges;</span><br></pre></td></tr></table></figure>
<p>第八步：开机自启</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">cd /usr/local/mysql/support-files</span><br><span class="line"></span><br><span class="line">cp mysql.server /etc/init.d/mysqld</span><br><span class="line"></span><br><span class="line">chkconfig --add mysqld</span><br></pre></td></tr></table></figure>
<p>第九步：使用service mysqld命令启动/停止服务</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">su - mysql</span><br><span class="line"></span><br><span class="line">service mysqld start/stop/restart</span><br></pre></td></tr></table></figure>
<p>登陆，修改密码 root密码是123456</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 登陆mysql</span><br><span class="line">mysql -uroot -p</span><br><span class="line"></span><br><span class="line"># 修改root用户密码</span><br><span class="line">set password for root@localhost=password(&quot;123456&quot;);</span><br></pre></td></tr></table></figure>
<p>远程用户建立</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">grant all privileges on *.* to &#x27;新用户名&#x27;@&#x27;%&#x27; identified by &#x27;新密码&#x27;;</span><br><span class="line"></span><br><span class="line">flush privileges;</span><br></pre></td></tr></table></figure>
<p><img src="https://genehub.files.wordpress.com/2019/03/e5beaee4bfa1e688aae59bbe_20190330204936.png" alt="微信截图_20190330204936" /> 添加系统路径</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">vim /etc/profile</span><br><span class="line"></span><br><span class="line">export PATH=/usr/local/mysql/bin:$PATH</span><br><span class="line"></span><br><span class="line">source /etc/profile</span><br></pre></td></tr></table></figure>
<p>以上步骤都执行完毕后，大吉大利，今晚吃鸡！！！！！！ 如果可以使用navicat 使用你刚才设置的用户名和密码访问，那么你的mysql就已经安装成功啦！ ===== 20190814 更新 增加pasa用户，设置只读权限。用于基因预测pasa软件流程。参考 <a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=1133">http://www.chenlianfu.com/?p=1133</a></p>
<p>$ mysql -uroot -p<br />
Enter password: ******** [# 输入root用户的密码]<br />
Welcome to the MySQL monitor. Commands end with ; or \g.<br />
Your MySQL connection id is 6<br />
Server version: 5.7.25 MySQL Community Server (GPL)</p>
<p>Copyright © 2000, 2019, Oracle and/or its affiliates. All rights reserved.</p>
<p>Oracle is a registered trademark of Oracle Corporation and/or its<br />
affiliates. Other names may be trademarks of their respective<br />
owners.</p>
<p>Type ‘help;’ or ‘\h’ for help. Type ‘\c’ to clear the current input statement.</p>
<h2 id="新建pasa用户只设置可读权限"><a class="markdownIt-Anchor" href="#新建pasa用户只设置可读权限"></a> 新建pasa用户，只设置可读权限</h2>
<p>mysql&gt; CREATE USER ‘pasa’@‘localhost’ IDENTIFIED BY ‘123456’;<br />
Query OK, 0 rows affected (0.00 sec)</p>
<p>mysql&gt; GRANT SELECT ON *.* TO ‘pasa’@‘%’ IDENTIFIED BY ‘123456’;<br />
Query OK, 0 rows affected, 1 warning (0.00 sec)</p>
<p>mysql&gt; FLUSH PRIVILEGES;<br />
Query OK, 0 rows affected (0.00 sec)</p>
<p>mysql&gt;</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2019/02/28/Bioinfo-install-lastz/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2019/02/28/Bioinfo-install-lastz/" class="post-title-link" itemprop="url">安装LASTZ</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2019-02-28 17:45:24" itemprop="dateCreated datePublished" datetime="2019-02-28T17:45:24+08:00">2019-02-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>LASTZ是序列比对工具，可以生成AXT格式的比对文件。 安装LASTZ方法如下 下载地址：<a target="_blank" rel="noopener" href="http://www.bx.psu.edu/~rsharris/lastz/">http://www.bx.psu.edu/~rsharris/lastz/</a> If you have received the distribution as a packed archive, unpack it by whatever means are appropriate for your computer. The result should be a directory <code>/lastz‑distrib‑X.XX.XX</code> that contains a <code>src</code> subdirectory (and some others). You may find it convenient to remove the revision number (<code>‑X.XX.XX</code>) from the directory name. Before building or installing any of the programs, you will need to tell the installer where to put the executable, either by setting the shell variable<code>$LASTZ_INSTALL</code>, or by editing the <code>make‑include.mak</code> file to set the definition of <code>installDir</code>. Also, be sure to add the directory you choose to your <code>$PATH</code>. Then to build the LASTZ executable, enter the following commands from bash or a similar command-line shell (Solaris users should substitute <code>gmake</code> for <code>make</code>). This will build two executables (<code>lastz</code> and <code>lastz_D</code>) and copy them into your <code>installDir</code>.</p>
<pre><code>cd /lastz-distrib-X.XX.XX/src
make
make install
</code></pre>
<p>The two executables are basically the same program; the only difference is that <code>lastz</code> uses integer scores, while <code>lastz_D</code> uses floating-point scores. The build process should not report any warnings or errors. Because of this, the Makefile is set up so that warnings are considered errors and will stop the build. If you encounter this situation, you can modify the Makefile, removing “-Werror” from the variable definedForAll. This should allow the build to complete, while still reporting the warnings. You’ll need to decide whether the warnings indicate something is really wrong. Usually they don’t, but please report them to the author regardless. A simple self test is included so you can test whether the build succeeded. To run it, enter the following command:</p>
<pre><code>make test
</code></pre>
<p>If the test is successful, you will see no output from this command. Otherwise, you will see the differences between the expected output and the output of your build, plus a line that looks like this:</p>
<pre><code>make: \*\*\* \[test\] Error 1
</code></pre>
<p>我在安装的过程中遇到下面的问题</p>
<p>cd src &amp;&amp; make lastz lastz_D<br />
make[1]: Entering directory ‘/home/david/src/lastz/src’<br />
gcc -c -O3 -Wall -Wextra -Werror -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DVERSION_MAJOR=“\“1”\” -DVERSION_MINOR=“\“04”\” -DVERSION_SUBMINOR=“\“00”\” -DREVISION_DATE=“\“20170312”\” -DSUBVERSION_REV=“\“1881:1893M”\” -Dscore_type=\‘I\’ lastz.c -o lastz.o<br />
lastz.c: In function ‘parse_options_loop’:<br />
lastz.c:6525:6: error: this ‘if’ clause does not guard… [-Werror=misleading-indentation]<br />
if (*scan != ‘.’) goto cant_understand; scan++;<br />
^~<br />
lastz.c:6525:47: note: …this statement, but the latter is misleadingly indented as if it is guarded by the ‘if’<br />
if (*scan != ‘.’) goto cant_understand; scan++;<br />
^~~~<br />
lastz.c:6585:6: error: this ‘if’ clause does not guard… [-Werror=misleading-indentation]<br />
if (*scan != ‘.’) goto cant_understand; scan++;<br />
^~<br />
lastz.c:6585:47: note: …this statement, but the latter is misleadingly indented as if it is guarded by the ‘if’<br />
if (*scan != ‘.’) goto cant_understand; scan++;<br />
^~~~<br />
lastz.c:6644:6: error: this ‘if’ clause does not guard… [-Werror=misleading-indentation]<br />
if (*scan != ‘.’) goto cant_understand; scan++;<br />
^~<br />
lastz.c:6644:47: note: …this statement, but the latter is misleadingly indented as if it is guarded by the ‘if’<br />
if (*scan != ‘.’) goto cant_understand; scan++;<br />
^~~~<br />
cc1: all warnings being treated as errors<br />
Makefile:94: recipe for target ‘lastz.o’ failed<br />
make[1]: *** [lastz.o] Error 1<br />
make[1]: Leaving directory ‘/home/david/src/lastz/src’<br />
Makefile:14: recipe for target ‘build_lastz’ failed<br />
make: *** [build_lastz] Error 2</p>
<p>解决办法 You can get around this by deleting “<em>-Werror</em>” from the CFLAGS listed in <em>definedForAll</em> in src/Makefile (overriding CFLAGS from the command line just causing new errors). 参考资料 <a target="_blank" rel="noopener" href="https://github.com/lastz/lastz/issues/5">https://github.com/lastz/lastz/issues/5</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2018/11/27/Python-python3-python2-in-a-system/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2018/11/27/Python-python3-python2-in-a-system/" class="post-title-link" itemprop="url">Python：python3与python2共存</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2018-11-27 21:53:40" itemprop="dateCreated datePublished" datetime="2018-11-27T21:53:40+08:00">2018-11-27</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-08 16:56:42" itemprop="dateModified" datetime="2020-12-08T16:56:42+08:00">2020-12-08</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p><strong>20201208更新：本文已过期。</strong></p>
<p>Linux系统自带了Python2.6，自己又安装了Python2.7和Anaconda3（Python3.6） 今天运行python3.6报错：</p>
<p>[fenglei@localhost lib64]$ /home/fenglei/local/app/anaconda3/bin/python3<br />
Failed to import the site module<br />
Traceback (most recent call last):<br />
File “/usr/local/lib/python2.7/site-packages/site.py”, line 75, in <module><br />
__boot()<br />
File “/usr/local/lib/python2.7/site-packages/site.py”, line 27, in __boot<br />
import imp # Avoid import loop in Python &gt;= 3.3<br />
File “/home/fenglei/local/app/anaconda3/lib/python3.6/imp.py”, line 27, in <module><br />
import tokenize<br />
File “/home/fenglei/local/app/anaconda3/lib/python3.6/tokenize.py”, line 33, in <module><br />
import re<br />
File “/home/fenglei/local/app/anaconda3/lib/python3.6/re.py”, line 142, in <module><br />
class RegexFlag(enum.IntFlag):<br />
AttributeError: module ‘enum’ has no attribute ‘IntFlag’</p>
<p>通过vim进入~/.bashrc文件进行编辑：</p>
<p>export PYTHONPATH=/home/fenglei/local/app/anaconda3/lib/python3.6/site-packages</p>
<p>即可运行 python3.6了。 如果需要运行python2.7，则做下修改：</p>
<p>export PYTHONPATH=/usr/local/lib/python2.7/site-packages</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2018/07/09/Bioinfo-diamond-vs-blast/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2018/07/09/Bioinfo-diamond-vs-blast/" class="post-title-link" itemprop="url">Diamond: 比BLAST更快速的比对工具</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2018-07-09 13:36:29" itemprop="dateCreated datePublished" datetime="2018-07-09T13:36:29+08:00">2018-07-09</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p><strong>Diamond简介</strong> 序列比对软件，能输出与BLAST类似的结果，但是速度快了数百倍。 <strong>安装</strong> 二进制可以直接解压使用：</p>
<p>wget <a target="_blank" rel="noopener" href="http://github.com/bbuchfink/diamond/releases/download/v0.9.22/diamond-linux64.tar.gz">http://github.com/bbuchfink/diamond/releases/download/v0.9.22/diamond-linux64.tar.gz</a><br />
tar xzf diamond-linux64.tar.gz</p>
<p>源码则用cmake安装：</p>
<p>wget <a target="_blank" rel="noopener" href="http://github.com/bbuchfink/diamond/archive/v0.9.22.tar.gz">http://github.com/bbuchfink/diamond/archive/v0.9.22.tar.gz</a><br />
tar xzf v0.9.22.tar.gz<br />
cd diamond-0.9.22<br />
mkdir bin<br />
cd bin<br />
cmake …<br />
make install</p>
<p><strong>使用</strong> 1 建立index</p>
<p>diamond makedb --in swissprot.fasta -d swissprot</p>
<p>2 比对</p>
<p>diamond blastp --db ./index/swissprot --query putative_proteins.fa --threads 10 --out proteins-BLASTP-swissprot.m8 -e 0.00001 --max-target-seqs 1</p>
<p>建立index的过程就6秒钟（Processed 557491 sequences, 199978344 letters. Total time = 6.64576s）。比对过程也极快，几分钟跑完了。 <strong>与其他软件的比较</strong> “生信技能树”论坛有人讲Diamond与BLAST进行比较，26740个蛋白序列与NR的蛋白库（animal）作比对，运行DIAMOND仅需要3小时，运行BLASTP则需要2周。存在的问题是同一个蛋白序列在两个比对软件下得到的最高分比对结果通常不一样（可能跟参数设置有关？）。第二个问题是输出的tabular格式结果不能被blast2go识别（可能需要调整输出格式的参数）。 <strong>参考资料</strong> [1] <a target="_blank" rel="noopener" href="https://github.com/bbuchfink/diamond/releases/">https://github.com/bbuchfink/diamond/releases/</a> [2] <a target="_blank" rel="noopener" href="http://www.biotrainee.com/thread-1465-1-1.html">http://www.biotrainee.com/thread-1465-1-1.html</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2018/05/16/Bioinfo-augustus-3-3-installation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2018/05/16/Bioinfo-augustus-3-3-installation/" class="post-title-link" itemprop="url">Augustus-3.3的安装</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2018-05-16 21:09:09" itemprop="dateCreated datePublished" datetime="2018-05-16T21:09:09+08:00">2018-05-16</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>Augustus 是流行的基因预测软件。在安装的过程中，遇到了 make 报错的问题，查询资料之后成功安装。错误信息提示 Augustus 在 bam2hints 编译的时候不通过，导致 make 失败。原来是bam2hints、filterBam 和 bam2wig 这三个目录下的 Makefile 中对其依赖程序的调用需要手动更改。下面部分文字摘抄自 <a target="_blank" rel="noopener" href="http://iamphioxus.org">iamphioxus.org</a> 。 Augustus is a very popular tool for gene annotation, however its installation process can be a bit tricky. For example, if we just download and install <em>Augustus</em> like below, it will not work.</p>
<p>wget <a target="_blank" rel="noopener" href="http://bioinf.uni-greifswald.de/augustus/binaries/augustus-3.3.1.tar.gz">http://bioinf.uni-greifswald.de/augustus/binaries/augustus-3.3.1.tar.gz</a><br />
tar xvzf augustus-3.3.1.tar.gz<br />
cd augustus-3.3.1<br />
make</p>
<p>You will probably get these errors:</p>
<p>cd auxprogs &amp;&amp; make<br />
make[1]: Entering directory ‘/home/jxyue/Tools/augustus-3.2.3/auxprogs’<br />
cd bam2hints; make;<br />
make[2]: Entering directory ‘/home/jxyue/Tools/augustus-3.2.3/auxprogs/bam2hints’<br />
g++ -Wall -O2 -c <a target="_blank" rel="noopener" href="http://bam2hints.cc">bam2hints.cc</a> -o bam2hints.o -I/usr/include/bamtools<br />
bam2hints.cc:16:27: fatal error: api/BamReader.h: No such file or directory<br />
#include &lt;api/BamReader.h&gt;<br />
^<br />
compilation terminated.<br />
Makefile:29: recipe for target ‘bam2hints.o’ failed<br />
make[2]: *** [bam2hints.o] Error 1<br />
make[2]: Leaving directory ‘/home/jxyue/Tools/augustus-3.2.3/auxprogs/bam2hints’<br />
Makefile:7: recipe for target ‘all’ failed<br />
make[1]: *** [all] Error 2<br />
make[1]: Leaving directory ‘/home/jxyue/Tools/augustus-3.2.3/auxprogs’<br />
Makefile:7: recipe for target ‘all’ failed<br />
make: *** [all] Error 2<br />
This is because the auxiliary tools, bam2hints and filterBam, depend on the pre-installation of bamtools with proper library path configured.</p>
<p>Therefore, let’s install bamtools first. It will be much easier if we just do system-wide installation like: 首先安装 bamtools，我在使用 redhat 系统，安装 bamtools-2.4.1 要用到 cmake 3.0以上版本。我尝试安装了 cmake 3.0 的源码，但是编译失败，于是下载了二进制文件，cmake 可以运行，随后顺利编译 bamtools 2.4.1。</p>
<p>tar -zxvf bamtools-2.4.1.tar.gz<br />
cd bamtools-2.4.1<br />
mkdir build<br />
cd build<br />
cmake -DCMAKE_INSTALL_PREFIX=/home/fenglei/local …<br />
make<br />
make install</p>
<h1 id="此时运行-bamtools-发现报错提示无法用-libbamtoolsso241于是在下载目录把该文件拷贝到了安装目录下的-lib-和-lib64-中再运行bamtools就成了"><a class="markdownIt-Anchor" href="#此时运行-bamtools-发现报错提示无法用-libbamtoolsso241于是在下载目录把该文件拷贝到了安装目录下的-lib-和-lib64-中再运行bamtools就成了"></a> 此时运行 bamtools 发现报错，提示无法用 libbamtools.so.2.4.1，于是在下载目录把该文件拷贝到了安装目录下的 lib 和 lib64 中，再运行bamtools就成了。</h1>
<h1 id="find-~local-name-libbamtoolsso241"><a class="markdownIt-Anchor" href="#find-~local-name-libbamtoolsso241"></a> find ~/local -name libbamtools.so.2.4.1</h1>
<p>cp /home/fenglei/local/app/bamtools-2.4.1/lib/libbamtools.so.2.4.1 /home/fenglei/local/lib/<br />
cp /home/fenglei/local/app/bamtools-2.4.1/lib/libbamtools.so.2.4.1  /home/fenglei/local/lib64</p>
<p>Now bamtools should have been correctly installed. Next, we need to modify the Makefiles of bam2hints and filterBam to adapt them with our manually installed bamtools.</p>
<p>First, go to the “./auxprogs/bam2hints” directory and make the following changes for the Makefile:</p>
<p>Add:<br />
BAMTOOLS = /home/fenglei/local</p>
<p>Replace:<br />
INCLUDES = /usr/include/bamtools<br />
By:<br />
INCLUDES = $(BAMTOOLS)/include</p>
<p>Replace:<br />
LIBS = -lbamtools -lz<br />
By:<br />
LIBS = $(BAMTOOLS)/lib64/libbamtools.a -lz</p>
<h1 id="注意-libs-中-lib64-也可能是-lib要通过搜索-libbamtoolsa-查看其实际位置"><a class="markdownIt-Anchor" href="#注意-libs-中-lib64-也可能是-lib要通过搜索-libbamtoolsa-查看其实际位置"></a> 注意 LIBS 中 lib64 也可能是 lib，要通过搜索 libbamtools.a 查看其实际位置。</h1>
<p>Then, go to the “augustus-3.2.3/auxprogs/filterBam/src” directory and make the following changes for the Makefile:</p>
<p>Replace:</p>
<p>BAMTOOLS = /usr/include/bamtools<br />
By:<br />
BAMTOOLS = /home/fenglei/local</p>
<p>Replace:<br />
INCLUDES = -I<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="false">(</mo><mi>B</mi><mi>A</mi><mi>M</mi><mi>T</mi><mi>O</mi><mi>O</mi><mi>L</mi><mi>S</mi><mo stretchy="false">)</mo><mo>−</mo><mi>I</mi><mi>h</mi><mi>e</mi><mi>a</mi><mi>d</mi><mi>e</mi><mi>r</mi><mi>s</mi><mo>−</mo><mi>I</mi><mi mathvariant="normal">.</mi><mi mathvariant="normal">/</mi><mi>b</mi><mi>a</mi><mi>m</mi><mi>t</mi><mi>o</mi><mi>o</mi><mi>l</mi><mi>s</mi><mi>B</mi><mi>y</mi><mo>:</mo><mi>I</mi><mi>N</mi><mi>C</mi><mi>L</mi><mi>U</mi><mi>D</mi><mi>E</mi><mi>S</mi><mo>=</mo><mo>−</mo><mi>I</mi></mrow><annotation encoding="application/x-tex">(BAMTOOLS) -Iheaders -I./bamtools
By:
INCLUDES = -I</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="mord mathnormal">A</span><span class="mord mathnormal" style="margin-right:0.10903em;">M</span><span class="mord mathnormal" style="margin-right:0.13889em;">T</span><span class="mord mathnormal" style="margin-right:0.02778em;">O</span><span class="mord mathnormal" style="margin-right:0.02778em;">O</span><span class="mord mathnormal">L</span><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.77777em;vertical-align:-0.08333em;"></span><span class="mord mathnormal" style="margin-right:0.07847em;">I</span><span class="mord mathnormal">h</span><span class="mord mathnormal">e</span><span class="mord mathnormal">a</span><span class="mord mathnormal">d</span><span class="mord mathnormal">e</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">s</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.07847em;">I</span><span class="mord">.</span><span class="mord">/</span><span class="mord mathnormal">b</span><span class="mord mathnormal">a</span><span class="mord mathnormal">m</span><span class="mord mathnormal">t</span><span class="mord mathnormal">o</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">s</span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">:</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.68333em;vertical-align:0em;"></span><span class="mord mathnormal" style="margin-right:0.07847em;">I</span><span class="mord mathnormal" style="margin-right:0.10903em;">N</span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mord mathnormal">L</span><span class="mord mathnormal" style="margin-right:0.10903em;">U</span><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="mord mathnormal" style="margin-right:0.05764em;">E</span><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.76666em;vertical-align:-0.08333em;"></span><span class="mord">−</span><span class="mord mathnormal" style="margin-right:0.07847em;">I</span></span></span></span>(BAMTOOLS)/include -Iheaders -I./bamtools</p>
<p>Replace:<br />
LIBS = -lbamtools -lz<br />
By:<br />
LIBS = $(BAMTOOLS)/lib64/libbamtools.a -lz</p>
<p>进入 augustus-3.3.1/auxprogs/bam2wig，编辑 Makefile。</p>
<p># SAMTOOLS=$(TOOLDIR)/samtools/<br />
SAMTOOLS=/home/fenglei/local/app/samtools</p>
<p>#HTSLIB=$(TOOLDIR)/htslib/<br />
HTSLIB=/home/fenglei/local/app/htslib</p>
<h1 id="bcftoolstooldirbcftools"><a class="markdownIt-Anchor" href="#bcftoolstooldirbcftools"></a> BCFTOOLS=$(TOOLDIR)/bcftools/</h1>
<p>BCFTOOLS=/home/fenglei/local/app/bcftools</p>
<h1 id="tabixtooldirtabix"><a class="markdownIt-Anchor" href="#tabixtooldirtabix"></a> TABIX=$(TOOLDIR)/tabix/</h1>
<p>TABIX=/home/fenglei/local/app/htslib</p>
<h2 id="上述注释掉的内容是-makefile-中的原始设置分别改一下就行了-tabix-是-htslib-下的一程序"><a class="markdownIt-Anchor" href="#上述注释掉的内容是-makefile-中的原始设置分别改一下就行了-tabix-是-htslib-下的一程序"></a> 上述注释掉的内容是 Makefile 中的原始设置，分别改一下就行了。tabix 是 htslib 下的一程序。</h2>
<p>[20201027 update] 在新的机器上重新安装Augustus，虽然按照上述步骤修改MakeFile，但是bam2wig还是无法make成功。于是将这个文件夹改掉，单独下载bam2wig程序（<a target="_blank" rel="noopener" href="https://github.com/MikeAxtell/bam2wig%EF%BC%89%EF%BC%8C%E5%B7%B2%E7%BB%8F%E6%98%AF%E4%BA%8C%E8%BF%9B%E5%88%B6%E6%96%87%E4%BB%B6%EF%BC%8C%E4%B8%8D%E9%9C%80%E8%A6%81%E7%BC%96%E8%AF%91%EF%BC%8C%E5%8F%AF%E4%BB%A5%E7%9B%B4%E6%8E%A5%E4%BD%BF%E7%94%A8%EF%BC%81%E7%84%B6%E5%90%8E%E8%BF%9B%E5%85%A5auxprogs%E7%9B%AE%E5%BD%95%EF%BC%8C%E4%BF%AE%E6%94%B9MakeFile%E6%96%87%E4%BB%B6%EF%BC%8C%E5%B0%B1%E5%B0%86bam2wig%E5%AD%97%E7%AC%A6%E9%82%A3%E4%B8%A4%E8%A1%8C%E6%B3%A8%E9%87%8A%E6%8E%89%E3%80%82%E7%84%B6%E5%90%8E%E5%9B%9E%E5%88%B0Augustus%E7%9B%AE%E5%BD%95%E4%B8%8BMake%E5%B0%B1OK%E4%BA%86%E3%80%82">https://github.com/MikeAxtell/bam2wig），已经是二进制文件，不需要编译，可以直接使用！然后进入auxprogs目录，修改MakeFile文件，就将bam2wig字符那两行注释掉。然后回到Augustus目录下Make就OK了。</a> Now, we are finally ready to compile Augustus. Get back to the “augustus-3.3.1” directory and type “make”, 成功! References <a target="_blank" rel="noopener" href="https://github.com/pezmaster31/bamtools/wiki/Building-and-installing">https://github.com/pezmaster31/bamtools/wiki/Building-and-installing</a> <a target="_blank" rel="noopener" href="https://www.biostars.org/p/189048/">https://www.biostars.org/p/189048/</a> <a target="_blank" rel="noopener" href="http://tiramisutes.github.io/2017/01/06/GLIBC.html">http://tiramisutes.github.io/2017/01/06/GLIBC.html</a> <a target="_blank" rel="noopener" href="https://iamphioxus.org/2017/05/08/installing-augustus-with-manual-bamtools-installation/">https://iamphioxus.org/2017/05/08/installing-augustus-with-manual-bamtools-installation/</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2018/04/25/R-volcano-plot-in-R/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2018/04/25/R-volcano-plot-in-R/" class="post-title-link" itemprop="url">R：火山图 volcano plot</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2018-04-25 16:54:11" itemprop="dateCreated datePublished" datetime="2018-04-25T16:54:11+08:00">2018-04-25</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-08 11:44:04" itemprop="dateModified" datetime="2020-12-08T11:44:04+08:00">2020-12-08</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>两个实验组之间的若干个基因的差异倍数比较，可以通过 Volcano Plot 图形化展示。 <img src="https://genehub.files.wordpress.com/2018/04/qqe688aae59bbe20180425163433.jpg" alt="QQ截图20180425163433" /> 例如 DEseq2 对转录组测序的样品做差异基因分析之后，会得到一个基因表达差异的表格的。案例使用数据的来源是 GitHub：<a target="_blank" rel="noopener" href="https://github.com/hbc/fabio-splicing/blob/master/new-samples/2016-07-29%5C_fabio-newlines/summary/control-vs-e7107-deseq2.csv">https://github.com/hbc/fabio-splicing/blob/master/new-samples/2016-07-29\_fabio-newlines/summary/control-vs-e7107-deseq2.csv</a> 下载上述数据之后，运行下面的命令即可得到文章起始的图。</p>
<p>setwd(“C:\\Users\\FengLei\\Desktop\\scatter-plot”)<br />
library(ggplot2)<br />
library(ggthemes)<br />
library(Cairo)</p>
<p>data=read.csv(“control-vs-e7107-deseq2.csv”, head=T)</p>
<h3 id="火山图"><a class="markdownIt-Anchor" href="#火山图"></a> 火山图</h3>
<p>data<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mi>h</mi><mi>r</mi><mi>e</mi><mi>s</mi><mi>h</mi><mi>o</mi><mi>l</mi><mi>d</mi><mo>=</mo><mi>a</mi><mi>s</mi><mi mathvariant="normal">.</mi><mi>f</mi><mi>a</mi><mi>c</mi><mi>t</mi><mi>o</mi><mi>r</mi><mo stretchy="false">(</mo><mi>i</mi><mi>f</mi><mi>e</mi><mi>l</mi><mi>s</mi><mi>e</mi><mo stretchy="false">(</mo><mi>d</mi><mi>a</mi><mi>t</mi><mi>a</mi></mrow><annotation encoding="application/x-tex">threshold = as.factor(ifelse(data</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.69444em;vertical-align:0em;"></span><span class="mord mathnormal">t</span><span class="mord mathnormal">h</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mord mathnormal">h</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">d</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal">s</span><span class="mord">.</span><span class="mord mathnormal" style="margin-right:0.10764em;">f</span><span class="mord mathnormal">a</span><span class="mord mathnormal">c</span><span class="mord mathnormal">t</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mopen">(</span><span class="mord mathnormal">i</span><span class="mord mathnormal" style="margin-right:0.10764em;">f</span><span class="mord mathnormal">e</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">s</span><span class="mord mathnormal">e</span><span class="mopen">(</span><span class="mord mathnormal">d</span><span class="mord mathnormal">a</span><span class="mord mathnormal">t</span><span class="mord mathnormal">a</span></span></span></span>pvalue =1, ‘Up’, ifelse(data<span class='katex-error' title='ParseError: KaTeX parse error: Expected &#039;EOF&#039;, got &#039;&amp;&#039; at position 15: pvalue &lt; 0.05 &amp;̲ data'>pvalue &lt; 0.05 &amp; data</span>log2FoldChange &lt;= -1, ‘Down’, ‘Not’)))<br />
ggplot(data=data, aes(x=log2FoldChange, y=-log10(pvalue), colour=threshold, fill=threshold)) +<br />
scale_color_manual(values=c(“blue”, “grey”,“red”))+<br />
geom_point(alpha=0.4, size=1.2) +</p>
<h1 id="xlimc-4-4"><a class="markdownIt-Anchor" href="#xlimc-4-4"></a> xlim(c(-4, 4)) +</h1>
<h1 id="ylimc0-75"><a class="markdownIt-Anchor" href="#ylimc0-75"></a> ylim(c(0, 7.5)) +</h1>
<p>theme_bw(base_size = 12, base_family = “Times”) +<br />
geom_vline(xintercept=c(-1,1),lty=4,col=“grey”,lwd=0.6)+<br />
geom_hline(yintercept = -log10(0.05),lty=4,col=“grey”,lwd=0.6)+<br />
theme(legend.position=“right”,<br />
panel.grid=element_blank(),<br />
legend.title = element_blank(),<br />
legend.text= element_text(face=“bold”, color=“black”,family = “Times”, size=8),<br />
plot.title = element_text(hjust = 0.5),<br />
axis.text.x = element_text(face=“bold”, color=“black”, size=12),<br />
axis.text.y = element_text(face=“bold”, color=“black”, size=12),<br />
axis.title.x = element_text(face=“bold”, color=“black”, size=12),<br />
axis.title.y = element_text(face=“bold”,color=“black”, size=12))+<br />
labs(x=“log2 (Fold Change)”,y=“-log10 (P-value)”,title=“Volcano Plot”)</p>
<p>注意 R 语言中 ifelse() 函数的使用方法。其基本语法格式如下： if(con,statement1,statement2) con 是逻辑条件，当逻辑条件的值为 TRUE 时，则输出 statement1 的值，否则输出 statement2 的值。 也就是说 ifelse() 函数只能输出两个类型的值，然而本文使用的转录组案例中，组间基因变化有三种情况，分别为上调、下调与无明显变化，这里可以使用 ifelse() 的嵌套结构来实现三种判断值的输出。</p>
<blockquote>
<p># ifelse() 函数使用例1</p>
<blockquote>
<p>x y # 如果向量 x 中的元素值非0，就输出0，否则输出1<br />
y<br />
[1] 0 1 0 1 0 0 1 1</p>
<h1 id="ifelse-函数使用例2-嵌套结构的应用"><a class="markdownIt-Anchor" href="#ifelse-函数使用例2-嵌套结构的应用"></a> ifelse() 函数使用例2: 嵌套结构的应用</h1>
<p>x y0, 2*x-1, ifelse(x==0,0,3*x-10))</p>
<h1 id="如向量-x-中元素值大于0等于0和小于0分三种情况做转换"><a class="markdownIt-Anchor" href="#如向量-x-中元素值大于0等于0和小于0分三种情况做转换"></a> 如向量 x 中元素值大于0，等于0和小于0，分三种情况做转换</h1>
<p>y<br />
[1] 5 19 0 -13 -40</p>
</blockquote>
</blockquote>
<p>参考资料 [1] ggplot 与火山图：<a target="_blank" rel="noopener" href="https://www.tanboyu.com/ggplot2-for-volcano.html">https://www.tanboyu.com/ggplot2-for-volcano.html</a> [2] R 语言 ifelse() 函数使用：<a target="_blank" rel="noopener" href="http://www.biye5u.com/article/R/2017/6352.html">http://www.biye5u.com/article/R/2017/6352.html</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2018/03/28/mitogenome-unfinished/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2018/03/28/mitogenome-unfinished/" class="post-title-link" itemprop="url">mitogenome (unfinished)</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2018-03-28 00:37:16" itemprop="dateCreated datePublished" datetime="2018-03-28T00:37:16+08:00">2018-03-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>tRNA tRNA sequences  整理成 fasta 格式 通过 blastn mapping  到 mitogenome，输出 format=6。 编程序转换成 features table 格式的文件。 注意： 1）部分 tRNA 有多个 copies，例如 trnC(gca) 与 trnfM(cat) 各有两个copies； 2）部分 tRNA 具有同样的 anticodon，但是其名称和序列不同，例如 trnI(cat) 和 trnM(cat)； 3）tRNA都没有内含子，所以可以较为简单的转换格式； 4）anticodon 如何自动注释？</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/8/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/8/">8</a><span class="page-number current">9</span><a class="page-number" href="/page/10/">10</a><span class="space">&hellip;</span><a class="page-number" href="/page/19/">19</a><a class="extend next" rel="next" href="/page/10/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">185</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2023</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
