<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></h1><div class="post-meta">2022-03-09</div><div class="post-content"><h3 id="一、蛋白质谱的背景知识"><a href="#一、蛋白质谱的背景知识" class="headerlink" title="一、蛋白质谱的背景知识"></a>一、蛋白质谱的背景知识</h3><p>LC-MS/MS<br>Liquid Chromatography-Mass Spectrometry（LC-MS/MS ，液相色谱-质谱串联）可用于残留化合物检测、有机小分子检测、鉴定和定量污染物以及在医药和食品领域添加剂检测和生物小分子等检测。。</p></div><p class="readmore"><a href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></h1><div class="post-meta">2021-10-31</div><div class="post-content"><p>本文总结粳稻（japonica）与籼稻（indica）的分类，以及相对应的数据库的事项。</p>
<p>作者：稻我家安全食材<br>链接：<a target="_blank" rel="noopener" href="https://www.zhihu.com/question/28272395/answer/513511209">https://www.zhihu.com/question/28272395/answer/513511209</a><br>来源：知乎<br>著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。</p></div><p class="readmore"><a href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></h1><div class="post-meta">2021-10-27</div><div class="post-content"><p>Linux 系统硬盘似乎出问题，报错信息是 I/O error。网上找资料说可以直接做硬盘修复，记录如下。</p>
<p>切换到 root 账户，执行 badblocks 命令查看出错的扇区数量，-b 指定每次扫描的区块大小，设置为和扇区大小一致，后面的数字 <del>1729898440</del> 为日志提示出错的扇区位置，前面的数字 <del>1729898499</del> 指定扫描扇区的终止位置，扫描后可以看到共有 <del>8</del> 个扇区出错。这里的几个数字要根据实际情况来配置。</p></div><p class="readmore"><a href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&amp;Tag sequencing and ChIP sequencing</a></h1><div class="post-meta">2021-10-20</div><div class="post-content"><p>CUT&amp;Tag 技术研究蛋白与染色质结合，相对于 ChIP Seq 技术具有显著优势，这个方法的原理是什么？先需要提到的是染色质免疫切割技术。CUT&amp;RUN 是一项对染色质免疫切割技术（ChIC）在全基因组范围的扩展方法，由 Laemmli 实验室开发（Schmid et al., 2004）。ChIC 使用蛋白 A-MNase 融合蛋白切割与靶蛋白相关并被特异性抗体识别的 DNA 区域。但是，ChIC 只能通过 southern blot 对特定位点进行分析。与此有关的一种方法是染色质内源性切割（ChEC），这种方法通过融合目标蛋白和 MNase 来分析目标蛋白在全基因组范围的结合位点（Schmid et al., 2004; Zentner et al ., 2015）。这种方法有一个明显的缺点，那就是需要为各目标蛋白生成特异性融合蛋白。CUT&amp;RUN 在这两种方法的基础上有了重大改进，它通过抗体将重组蛋白 A/G-MNase 固定在了目标蛋白处（Meers et al., 2019; Skene and Henikoff, 2017）。</p></div><p class="readmore"><a href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></h1><div class="post-meta">2021-10-11</div><div class="post-content"><p>DIAMOND 是一种快速批量序列比对程序，可将 DNA 文件与蛋白质参考序列文件（如 NCBI-nr）进行比较。它以 C++ 实现，旨在在多核服务器上快速运行。该程序明确地设计为，利用具有大内存容量和许多内核的现代计算机体系结构。那么为什么它那么快呢，因为它使用了种子延伸方法。额外的算法成分是使用缩小的字母，间隔种子和双索引。算法简单了解一下就可以了，具体的算法的内容比较难懂就不深入讨论了。其主要的特点包括(为什么该工具那么受欢迎):在比较NCBI-nr数据库的短DNA reads时，DIAMOND比BLASTX 快4个数量级，并且在e-value &lt; 0.001 的比对上保持相当的灵敏度水平。简单一句话就是又快又准。可用于移位分析的长reads比对资源要求低，适合在标准台式机或笔记本电脑上运行(入门门槛低，适合各类的玩家)各种输出格式，包括BLAST对比格式，例如格式6的tabular分隔形式和格式5的XML格式 (很多下游分析，或者脚本都是基于BLAST的输出结果，diamond能够直接输出和Blast相同的格式不能不说是最大的优点之一)（摘自：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/d2407ec4648a">lakeseafly</a></p></div><p class="readmore"><a href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></h1><div class="post-meta">2021-10-01</div><div class="post-content"><p>总结正向选择分析的一些知识。2017 年做过一次线粒体基因的正向选择，相关内容已经<a target="_blank" rel="noopener" href="https://link.springer.com/article/10.1007%2Fs00606-019-01578-2#Sec18">发表</a>，并且我在Supplementary Files 里附上了我的<a target="_blank" rel="noopener" href="https://static-content.springer.com/esm/art%3A10.1007%2Fs00606-019-01578-2/MediaObjects/606_2019_1578_MOESM1_ESM.pdf">分析代码</a>。2021年再次需要做正向选择分析，再次回顾以前的工作，发现其中也有一个问题，就是没有做统计检验，推算P值，再次学习正向选择分析，并摘抄一些网络资料并整理记录。Codeml 的学习成本比较高，原理与操作都不易理解，学习的过程中发现不少好资源，比如 B 站有<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1Qx411m7Bn?from=search&seid=9774626637377853773&spm_id_from=333.337.0.0">视频</a>讲解如何使用 codeml 推断特定分支的 ω 值（dN/dS），以及使用卡方检验来计算 P 值。另外有国内的团队以 codeml 为内核，用 java 编写了更友好的操作界面：<a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/home.php?mod=space&uid=460481&do=blog&id=1163040">EasyCodeML</a>。</p></div><p class="readmore"><a href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></h1><div class="post-meta">2021-09-20</div><div class="post-content"><h3 id="1-workflow"><a href="#1-workflow" class="headerlink" title="1 workflow"></a>1 workflow</h3><ol>
<li>crude histone isolation from cells or tissues;</li></div><p class="readmore"><a href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></h1><div class="post-meta">2021-09-16</div><div class="post-content"><p>转录组测序如果有多个时间点，可以做表达趋势分析，即基因表达模式聚类分析，可以用 STEM、Mfuzz、TCseq 软件等工具。</p>
<p>Mfuzz 采用了一种新的聚类算法 fuzzy c-means algorithm，在文献中称这种聚类算法为 soft clustering 算法，相比 K-means 等 hard clustering 算法，一定程度上降低了噪声对聚类结果的干扰，而且这种算法有效的定义了基因和 cluster 之间的关系，即基因是否属于某个 cluster, 对应的值为 memebership。</p></div><p class="readmore"><a href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Bioinfo | LTR identification and analysis</a></h1><div class="post-meta">2021-07-17</div><div class="post-content"><p>在植物基因组中，I 类转座因子，LTR-RT（LTR retrotransposons）是基因组扩张的主要原因。完整的LTR长度在 85 – 5000 bp 之间，下图表示的是一个完整的 LTR-RT，灰色框表示 TSD（target site duplications）, 红色三角形表示 LTR motif（长度在 2 bp 左右），蓝色框表示 LTR。LTR 中间序列长度在 1,000 – 15,000 之间波动。</p></div><p class="readmore"><a href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a></h1><div class="post-meta">2021-06-29</div><div class="post-content"><p>下面是我做的大豆GO数据库（version Gmax_508_Wm82_a4.v1）构建与富集分析。数据文件在：<a target="_blank" rel="noopener" href="https://github.com/bioinfx/Glycine_max_Gene_Ontology">https://github.com/bioinfx/Glycine_max_Gene_Ontology</a>。</p></div><p class="readmore"><a href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Read More</a></p></div><nav class="page-navigator"><span class="page-number current">1</span><a class="page-number" href="/page/2/">2</a><a class="page-number" href="/page/3/">3</a><span class="space">&hellip;</span><a class="page-number" href="/page/17/">17</a><a class="extend next" rel="next" href="/page/2/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Bioinfo | LTR identification and analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>