<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>Bioinfo | LTR identification and analysis | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">Bioinfo | LTR identification and analysis</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">Bioinfo | LTR identification and analysis</h1><div class="post-meta">2021-07-17<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>在植物基因组中，I 类转座因子，LTR-RT（LTR retrotransposons）是基因组扩张的主要原因。完整的LTR长度在 85 – 5000 bp 之间，下图表示的是一个完整的 LTR-RT，灰色框表示 TSD（target site duplications）, 红色三角形表示 LTR motif（长度在 2 bp 左右），蓝色框表示 LTR。LTR 中间序列长度在 1,000 – 15,000 之间波动。</p>
<p>完整的LTR-RT主要归为两大类: Gypsy 和 Copia。如果 LTR 中间的序列不包含开放阅读框（ORF）, 那么所属的 LTR-RT 就无法独立的转座。</p>
<p><img src="https://i.imgur.com/pDbuH66.jpeg"></p>
<h3 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h3><p>LTR_retriever 不是一个独立的工具，他的主要作用就是整合 LTRharvest, LTR_FINDER, MGEScan 3.0.0, LTR_STRUC, 和 LtrDetector 的结果，过滤其中的假阳性 LTR-RT，得到高质量的LTR-RT库。</p>
<p>先下载 LTR_retriever：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https:&#x2F;&#x2F;github.com&#x2F;oushujun&#x2F;LTR_retriever.git</span><br></pre></td></tr></table></figure>
<p>之后修改 LTR_retriever 下的 paths, 提供 BLAST+, RepeatMasker， HMMER， CDHIT 这些工具的路径。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">BLAST+&#x3D;&#x2F;your_path_to&#x2F;BLAST+2.2.30&#x2F;bin&#x2F;</span><br><span class="line">RepeatMasker&#x3D;&#x2F;your_path_to&#x2F;RepeatMasker4.0.0&#x2F;</span><br><span class="line">HMMER&#x3D;&#x2F;your_path_to&#x2F;HMMER3.1b2&#x2F;bin&#x2F;</span><br><span class="line">CDHIT&#x3D;&#x2F;your_path_to&#x2F;CDHIT4.6.1&#x2F;</span><br><span class="line">BLAST&#x3D;&#x2F;your_path_to&#x2F;BLAST2.2.26&#x2F;bin&#x2F; #not required if CDHIT provided</span><br></pre></td></tr></table></figure>

<p>下面是我实际操作中填写的案例。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"> 1 ##This file will provide LTR_retriever paths to dependent programs.</span><br><span class="line"> 2 ##You can leave the respective paths empty if programs are accessible through ENV (i.e. exported to .bashrc)</span><br><span class="line"> 3 ##If you specify a path, please make sure that the required program(s) is directly contained in that path but not in any child directories.</span><br><span class="line"> 4 ##e.g. BLAST+&#x3D;&#x2F;opt&#x2F;software&#x2F;BLAST+&#x2F;2.2.30--GCC-4.4.5&#x2F;bin&#x2F;</span><br><span class="line"> 5 ##LTR_retriever is build based on GenomeTools&#x2F;1.5.4, BLAST+&#x2F;2.2.28, BLAST&#x2F;2.2.26, CDHIT&#x2F;4.6.1c, HMMER&#x2F;3.1b2, RepeatMasker&#x2F;4.0.0 and Tandem Repeats Finder 4.07b</span><br><span class="line"> 6 </span><br><span class="line"> 7 BLAST+&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;blast-2.11.0&#x2F;bin     #a path that contains makeblastdb, blastn, blastx</span><br><span class="line"> 8 RepeatMasker&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;RepeatMasker   #a path that contains RepeatMasker</span><br><span class="line"> 9 HMMER&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;bin       #a path that contains hmmsearch</span><br><span class="line">10 CDHIT&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;cdhit     #a path that contains cd-hit-est (preferred). CDHIT and BLAST are replaceable</span><br><span class="line">11 BLAST&#x3D;      #a path that contains blastclust (optional)</span><br></pre></td></tr></table></figure>

<p>更加方便的安装方法用 Bioconda 安装好 cd-hit repeatmasker， 然后下载 LTR_retriever:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">conda create -n LTR_retriever</span><br><span class="line">source activate LTR_retriever</span><br><span class="line">conda install -c conda-forge perl perl-text-soundex</span><br><span class="line">conda install -c bioconda cd-hit</span><br><span class="line">conda install -c bioconda&#x2F;label&#x2F;cf201901 repeatmasker</span><br><span class="line"></span><br><span class="line">git clone https:&#x2F;&#x2F;github.com&#x2F;oushujun&#x2F;LTR_retriever.git</span><br><span class="line">.&#x2F;LTR_retriever&#x2F;LTR_retriever -h</span><br></pre></td></tr></table></figure>

<p>此外还需要额外安装 LTRharvest, LTR_FINDER 和 MGEScan_LTR。</p>
<p>LTRharverst: <a target="_blank" rel="noopener" href="http://genometools.org/">http://genometools.org/</a><br>LTR_FINDER: <a target="_blank" rel="noopener" href="https://github.com/xzhub/LTR_Finder">https://github.com/xzhub/LTR_Finder</a><br>LTR_FINDER_parallel: <a target="_blank" rel="noopener" href="https://github.com/oushujun/LTR_FINDER_parallel">https://github.com/oushujun/LTR_FINDER_parallel</a><br>修改版MGEScan_LTR: <a target="_blank" rel="noopener" href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net</a></p>
<p>由于MGEScan_LTR装起来比我想象中麻烦，所以本文就仅使用LTRharverst和LTR_FINDER</p>
<h3 id="使用篇"><a href="#使用篇" class="headerlink" title="使用篇"></a>使用篇</h3><p>尽管 LTR_retriever 支持多个 LTR 工具的输入，但其实上 LTRharverst 和 LTR_FINDER 的结果就已经很不错了。</p>
<p>以拟南芥的基因组序列为例，分别使用 LTRharverst 和 LTR_FINDER 来寻找拟南芥中潜在 LTR 序列，之后用 LTR_retreiver 来合并结果。<br>注意，也可以用 LTR-finder-parallel，应该会更快。</p>
<p>注意：LTRharvest 输出的结果中会将染色体名称成数字编号，可能会对后续分析造成问题。Unfortunately, LTRharvest has an annoying habit of not using the scaffold/chromosome names you give it (it outputs scaffolds as “seqX” where X is the 0-base number of your scaffold), so I remove all scaffolds in my fasta file, remove newline (“\n”) characters, and add a single scaffold name so that getting the fasta sequence for regions in the output gff3 is painless. 有人的解决方案是将自己的基因组手动拼成一条序列，因为只需要找LTR序列而已。 This is useful if you just want the distribution of insertion times of your LTRs, but if you need location information you’ll either have to rerun LTRharvest with the regular fasta, or use the length of each scaffold to figure out the real scaffold from your concatenated fasta. </p>
<p>20210719 备注：经过测试，LTRharvest输出的scn文件虽然有上面提到的问题，但是对后续导入LTR-retriever的整合分析不会造成影响。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"># LTRharvest</span><br><span class="line">gt suffixerator \</span><br><span class="line">  -db TAIR10.fa \</span><br><span class="line">  -indexname TAIR10 \</span><br><span class="line">  -tis -suf -lcp -des -ssp -sds -dna</span><br><span class="line">gt ltrharvest \</span><br><span class="line">  -index TAIR10 \</span><br><span class="line">  -similar 90 -vic 10 -seed 20 -seqids yes \</span><br><span class="line">  -minlenltr 100 -maxlenltr 7000 -mintsd 4 -maxtsd 6 \</span><br><span class="line">  -motif TGCA -motifmis 1  &gt; TAIR10.harvest.scn &amp;</span><br><span class="line"></span><br><span class="line"># LTR_FINDER</span><br><span class="line"># ltr_finder -D 15000 -d 1000 -L 7000 -l 100 -p 20 -C -M 0.9 TAIR10.fa &gt; TAIR10.finder.scn &amp;</span><br><span class="line">~&#x2F;local&#x2F;app&#x2F;LTR_FINDER_parallel&#x2F;LTR_FINDER_parallel -seq genome.fasta -threads 10 -harvest_out -size 1000000 -time 300</span><br></pre></td></tr></table></figure>

<p>LTR_retriever支持单个候选的LTR：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LTR_retriever -genome TAIR10.fa -inharvest TAIR10.harvest.scn</span><br></pre></td></tr></table></figure>

<p>也支持多个候选LTR输入：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LTR_retriever -genome TAIR10.fa -inharvest TAIR10.harvest.scn -infinder TAIR10.finder.scn -threads 20</span><br></pre></td></tr></table></figure>

<p>输出文件如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br></pre></td><td class="code"><pre><span class="line">(base) [fenglei@localhost 20210719_LTR_retriever_test]$ LTR_retriever -genome genome.fasta -inharvest genome.fasta.rawLTR.scn -threads 10</span><br><span class="line"></span><br><span class="line">##########################</span><br><span class="line">### LTR_retriever v2.9.0 ###</span><br><span class="line">##########################</span><br><span class="line"></span><br><span class="line">Contributors: Shujun Ou, Ning Jiang</span><br><span class="line"></span><br><span class="line">For LTR_retriever, please cite:</span><br><span class="line"></span><br><span class="line">        Ou S and Jiang N (2018). LTR_retriever: A Highly Accurate and Sensitive Program for Identification of Long Terminal Repeat Retrotransposons. Plant Physiol. 176(2): 1410-1422.</span><br><span class="line"></span><br><span class="line">For LAI, please cite:</span><br><span class="line"></span><br><span class="line">        Ou S, Chen J, Jiang N (2018). Assessing genome assembly quality using the LTR Assembly Index (LAI). Nucleic Acids Res. 2018;46(21):e126.</span><br><span class="line"></span><br><span class="line">Parameters: -genome genome.fasta -inharvest genome.fasta.rawLTR.scn -threads 10</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:09:05 CST 2021    Dependency checking: All passed!</span><br><span class="line">Mon Jul 19 09:09:24 CST 2021    LTR_retriever is starting from the Init step.</span><br><span class="line">Mon Jul 19 09:09:26 CST 2021    Start to convert inputs...</span><br><span class="line">                                Total candidates: 1110</span><br><span class="line">                                Total uniq candidates: 1045</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:09:28 CST 2021    Module 1: Start to clean up candidates...</span><br><span class="line">                                Sequences with 10 missing bp or 0.8 missing data rate will be discarded.</span><br><span class="line">                                Sequences containing tandem repeats will be discarded.</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:09:48 CST 2021    950 clean candidates remained</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:09:48 CST 2021    Modules 2-5: Start to analyze the structure of candidates...</span><br><span class="line">                                The terminal motif, TSD, boundary, orientation, age, and superfamily will be identified in this step.</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:16:20 CST 2021    Intact LTR-RT found: 230</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:16:24 CST 2021    Module 6: Start to analyze truncated LTR-RTs...</span><br><span class="line">                                Truncated LTR-RTs without the intact version will be retained in the LTR-RT library.</span><br><span class="line">                                Use -notrunc if you don&#39;t want to keep them.</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:16:24 CST 2021    66 truncated LTR-RTs found</span><br><span class="line">Mon Jul 19 09:16:41 CST 2021    29 truncated LTR sequences have added to the library</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:16:41 CST 2021    Module 5: Start to remove DNA TE and LINE transposases, and remove plant protein sequences...</span><br><span class="line">                                Total library sequences: 355</span><br><span class="line">Mon Jul 19 09:18:52 CST 2021    Retained clean sequence: 348</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:18:52 CST 2021    Sequence clustering for genome.fasta.ltrTE ...</span><br><span class="line">Mon Jul 19 09:18:52 CST 2021    Unique lib sequence: 347</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:18:55 CST 2021    Module 6: Start to remove nested insertions in internal regions...</span><br><span class="line">Mon Jul 19 09:27:23 CST 2021    Raw internal region size (bit): 961029</span><br><span class="line">                                Clean internal region size (bit): 912840</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:24 CST 2021    Sequence number of the redundant LTR-RT library: 659</span><br><span class="line">                                The redundant LTR-RT library size (bit): 1422033</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:24 CST 2021    Module 8: Start to make non-redundant library...</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:27 CST 2021    Final LTR-RT library entries: 342</span><br><span class="line">                                Final LTR-RT library size (bit): 1025212</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:27 CST 2021    Total intact LTR-RTs found: 206</span><br><span class="line">                                Total intact non-TGCA LTR-RTs found: 43</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:28 CST 2021    Start to annotate whole-genome LTR-RTs...</span><br><span class="line">                                Use -noanno if you don&#39;t want whole-genome LTR-RT annotation.</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">######################################</span><br><span class="line">### LTR Assembly Index (LAI) beta3.2 ###</span><br><span class="line">######################################</span><br><span class="line"></span><br><span class="line">Developer: Shujun Ou</span><br><span class="line"></span><br><span class="line">Please cite:</span><br><span class="line"></span><br><span class="line">Ou S., Chen J. and Jiang N. (2018). Assessing genome assembly quality using the LTR Assembly Index (LAI). Nucleic Acids Res. gky730: https:&#x2F;&#x2F;doi.org&#x2F;10.1093&#x2F;nar&#x2F;gky730</span><br><span class="line"></span><br><span class="line">Parameters: -genome genome.fasta -intact genome.fasta.pass.list -all genome.fasta.out -t 10 -q -blast &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;blast-2.11.0&#x2F;bin&#x2F;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:32:57 CST 2021    Dependency checking: Passed!</span><br><span class="line">Mon Jul 19 09:32:57 CST 2021    Calculation of LAI will be based on the whole genome.</span><br><span class="line">                                Please use the -mono parameter if your genome is a recent ployploid, for high identity between homeologues will overcorrect raw LAI scores.</span><br><span class="line">Mon Jul 19 09:32:57 CST 2021    Estimate the identity of LTR sequences in the genome: quick mode</span><br><span class="line">Mon Jul 19 09:33:08 CST 2021    The identity of LTR sequences: 93.5158904793606%</span><br><span class="line">Mon Jul 19 09:33:08 CST 2021    Calculate LAI:</span><br><span class="line"></span><br><span class="line">                                                Done!</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:33:11 CST 2021    Result file: genome.fasta.out.LAI</span><br><span class="line"></span><br><span class="line">                                You may use either raw_LAI or LAI for intraspecific comparison</span><br><span class="line">                                but please use ONLY LAI for interspecific comparison</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:33:11 CST 2021    All analyses were finished!</span><br><span class="line"></span><br><span class="line">##############################</span><br><span class="line">####### Result files #########</span><br><span class="line">##############################</span><br><span class="line"></span><br><span class="line">Table output for intact LTR-RTs (detailed info)</span><br><span class="line">        genome.fasta.pass.list (All LTR-RTs)</span><br><span class="line">        genome.fasta.nmtf.pass.list (Non-TGCA LTR-RTs)</span><br><span class="line">        genome.fasta.pass.list.gff3 (GFF3 format for intact LTR-RTs)</span><br><span class="line"></span><br><span class="line">LTR-RT library</span><br><span class="line">        genome.fasta.LTRlib.redundant.fa (All LTR-RTs with redundancy)</span><br><span class="line">        genome.fasta.LTRlib.fa (All non-redundant LTR-RTs)</span><br><span class="line">        genome.fasta.nmtf.LTRlib.fa (Non-TGCA LTR-RTs)</span><br><span class="line"></span><br><span class="line">Whole-genome LTR-RT annotation by the non-redundant library</span><br><span class="line">        genome.fasta.LTR.gff3 (GFF3 format)</span><br><span class="line">        genome.fasta.out.fam.size.list (LTR family summary)</span><br><span class="line">        genome.fasta.out.superfam.size.list (LTR superfamily summary)</span><br><span class="line"></span><br><span class="line">LTR Assembly Index (LAI)</span><br><span class="line">        genome.fasta.out.LAI</span><br></pre></td></tr></table></figure>

<h3 id="LTR-RTs-insertion-time"><a href="#LTR-RTs-insertion-time" class="headerlink" title="LTR-RTs insertion time"></a>LTR-RTs insertion time</h3><p>根据LTR-RT的结构特点，即插入基因组新位置之后，反转录转座子的两端的LTR会独立进化，各自积累突变，计算出他们之间的遗传距离与进化时间，就可推断基因组里面大量LTR-RT出现的规律。“LTR-RTs are a case where we can exploit biology to provide insight into larger processes. LTR-RTs are a class of retrotransposon with long terminal repeats, which are identical upon insertion at the beginning and end of the TE’s sequence. Over time, mutations accumulate in these terminal repeats, roughly following a molecular clock. This allows us to more or less count the differences between the two LTRs of a LTR-RT, and calculate the time (in generations ago) it was inserted. If we do this for all LTR-RTs in a genome, we get a distribution of insertion times, and shifts in this distribution can tell us about the history of TE activity and/or selective pressures against TEs in our favourite species or between species.”</p>
<p>计算LTR插入时间可以参考：<a target="_blank" rel="noopener" href="https://github.com/SIWLab/Lab_Info/wiki/Ageing-LTR-insertions">https://github.com/SIWLab/Lab_Info/wiki/Ageing-LTR-insertions</a></p>
<p>按照上面的流程，运行LTRretriever之后得到一系列文件，其中genome.fasta.pass.list.gff3是最终的LTR列表，包含三种类型：Copia、Gypsy和unknown。一般文献会分别统计Copia、Gypsy和non-autonomous这三个类型的LTR。那么non-autonomous是否就是unknown呢？我将这个gff3文件导入IGV浏览器，发现unknown LTR的特点就是具有双端LTR结构，但是中间区域很短，意味着其很可能不具备gag和pol序列，无法独立执行这个转座子单元的反转录的过程。</p>
<p><img src="https://i.imgur.com/AAVN4k5.jpeg"></p>
<p>下截图所示，是gff3的文件内容，每个LTR-RT一般有6行，第3和第5行就是两端的LTR区域，将这两行提取出来，存入一个文件。我写了一个Perl脚本，逐个LTR-RT分别存入新的文件，每个文件两行，比如LTR_1.gff3、LTR_2.gff3、LTR_3.gff3 …… 然后基于这些gff3文件，用bedtools分别提取fasta序列，再用muscle软件做序列比对（注意muscle软件只能按照两条序列自然的方向作比较，他不会考虑两条序列是否反向之后匹配更好，当然，LTR-RT的两端的LTR序列本来就是方向一致的，所以不用担心。），muscle输出对齐的fasta格式文件，然后运用程序计算双端LTR之间的距离K，再依据 T=K/(2r) 公式就能推测到插入时间。将“插入时间”数据统一存入一个文本，导入R，用ggplot2的geom_histogram功能就可以得到下面的图。</p>
<p>上面公式中r的意思是“synonymous nucleotide substitution rate”，“r is the average substitution rate, taken as the Ks value estimated above”，<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1456784/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1456784/</a>，7e-9 substitutions/site/year, with standard deviation of 2.6e−9, was used as the rate of synonymous substitutions (r). 根据这篇文章，豆科的平均替换率是7e-9 substitutions/site/year，我在我研究的物种中也用这个数值。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">#########################################################################</span><br><span class="line"># File Name: ltr.sh</span><br><span class="line"># Author: fenglei</span><br><span class="line"># Mail: fengleiluck@gmail.com</span><br><span class="line"># Created Time: Wed 21 Jul 2021 01:57:33 PM CST</span><br><span class="line">#########################################################################</span><br><span class="line">#!&#x2F;usr&#x2F;bin&#x2F;bash</span><br><span class="line">set -e</span><br><span class="line">set -u</span><br><span class="line">set -o pipefail</span><br><span class="line"></span><br><span class="line">for IN in &#96;find .&#x2F;loop&#x2F; -name &quot;LTR*.gff3&quot;&#96;; do</span><br><span class="line">        echo [&#96;date +&quot;%Y-%m-%d %H:%M:%S&quot;&#96;] &quot;   * processing $IN&quot;</span><br><span class="line">        path&#x3D;$&#123;IN%&#x2F;*&#125;    #  .&#x2F;loop&#x2F;LTR_85.gff3 -&gt; .&#x2F;loop</span><br><span class="line">        file&#x3D;$&#123;IN##*&#x2F;&#125;   #  .&#x2F;loop&#x2F;LTR_85.gff3 -&gt; LTR_85.gff3</span><br><span class="line">        name&#x3D;$&#123;file%.*&#125;  #  LTR_85.gff3 -&gt; LTR_85</span><br><span class="line">        # cat genome.fasta.finder.combine.scn genome.fasta.harvest.scn_new &gt;genome.fasta.rawLTR.scn </span><br><span class="line">        # awk &#39;NR &#x3D;&#x3D; 9 || NR &#x3D;&#x3D; 11&#39; genome.fasta.pass.list.gff3 &gt; LTR_1.gff3</span><br><span class="line">        # bedtools getfasta -fi genome.fasta -bed LTR_1.gff3 &gt; $&#123;name&#125;.fasta </span><br><span class="line">        bedtools getfasta -fi genome.fasta -bed $IN &gt; .&#x2F;loop&#x2F;$&#123;name&#125;.fasta</span><br><span class="line">        ~&#x2F;local&#x2F;app&#x2F;muscle3.8.31&#x2F;muscle3.8.31_i86linux64 -in .&#x2F;loop&#x2F;$&#123;name&#125;.fasta -out .&#x2F;loop&#x2F;$&#123;name&#125;.afa</span><br><span class="line">        # cat LTR_1.pair_diff</span><br><span class="line">        python ..&#x2F;..&#x2F;20210720_LTR_evolution&#x2F;LTR_pairwise_differences.py --input .&#x2F;loop&#x2F;$&#123;name&#125;.afa --output .&#x2F;loop&#x2F;$&#123;name&#125;.pair_diff</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<p><img src="https://i.imgur.com/Xkpiivl.jpeg"></p>
<p>附画图R语言代码：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)</span><br><span class="line">setwd(&quot;C:&#x2F;&#x2F;R-data&quot;)</span><br><span class="line">ltr &lt;- read.table(&quot;Arabidopsis_LTR.txt&quot;, head&#x3D;F)</span><br><span class="line">head(ltr)</span><br><span class="line">colnames(ltr) &lt;- c(&quot;locus&quot;, &quot;K&quot;, &quot;T&quot;)</span><br><span class="line">head(ltr$locus)</span><br><span class="line">ltr$newT &lt;- ltr$T&#x2F;1000000</span><br><span class="line">head(ltr)</span><br><span class="line">ggplot() +</span><br><span class="line">  geom_histogram(data&#x3D;ltr, aes(x&#x3D;newT), fill&#x3D;&quot;#0570c0&quot;, colour&#x3D;&quot;black&quot;) +</span><br><span class="line">  theme_bw() +</span><br><span class="line">  theme(panel.grid.major &#x3D; element_blank(), panel.grid.minor &#x3D; element_blank(), axis.line &#x3D; element_line(colour &#x3D; &quot;black&quot;)) +</span><br><span class="line">  theme(axis.text.x &#x3D; element_text(size &#x3D; 15, color &#x3D; &quot;black&quot;, vjust &#x3D; 0.5 , hjust &#x3D; 0.5, angle &#x3D; 0)) +</span><br><span class="line">  theme(axis.text.y &#x3D; element_text(color &#x3D; &quot;black&quot;, size &#x3D; 15)) +</span><br><span class="line">  ylab(&quot;Copy Number&quot;) +</span><br><span class="line">  xlab(&quot;Insertion time (Myr)&quot;) +</span><br><span class="line">  #ggtitle(&quot;4DTv density of paralogs gene pairs in legumes&quot;) +</span><br><span class="line">  theme(axis.title.y &#x3D; element_text(size &#x3D; 18, color &#x3D; &quot;black&quot;)) +</span><br><span class="line">  theme(axis.title.x &#x3D; element_text(size &#x3D; 18, color &#x3D; &quot;black&quot;))</span><br></pre></td></tr></table></figure>

<h3 id="LTR-RTs-系统发生树"><a href="#LTR-RTs-系统发生树" class="headerlink" title="LTR-RTs 系统发生树"></a>LTR-RTs 系统发生树</h3><ol>
<li>genome.fasta.pass.list.gff3 文件，用bedtools提取对应的LTR-RT的完整DNA序列。</li>
<li>一个LTR-RT中有多个蛋白质domain，如何提取这些domain？用TEsorter可以轻松实现，TEsorter用conda即可安装，我在python3.7环境下安装的。得到不同功能域（GAG、INT、RT、RH……）的氨基酸序列。<br>（一开始不知道TEsorter存在，自己用blastx和hmmer尝试，效果都不理想）</li>
<li>通过cdhit软件进行序列聚类，将高度近似或相同的序列归类，大大减少下一步比对的序列数目。我从2800聚集成400多条序列。</li>
<li>序列比对。可以用mafft或muscle。我用这windows版本的MEGA软件，MEGA X 内置muscle，，运行数秒即可得到下图。</li>
</ol>
<p><img src="https://i.imgur.com/WMphP1J.jpeg"></p>
<ol start="5">
<li>直接在MEGA软件里面选择系统进化分析，选择ML（Maximum Likelihood），其余参数默认，运行很快就得到下图。</li>
</ol>
<p><img src="https://i.imgur.com/ksgYYlM.jpeg"></p>
<ol start="6">
<li><p>上面的树，导出为Newick格式（.nwk），然后可以用其他软件来美化这个进化树。注意导出的时候可以设定是否导出树枝长度，若不选，那么出来的树就是类似二叉树的形象，树枝都是一样的长度。</p>
</li>
<li><p>用Figtree软件打开这个树，可以美化。参考资料 <a target="_blank" rel="noopener" href="https://www.omicsclass.com/article/62">https://www.omicsclass.com/article/62</a>. 比如某个分支需要加颜色背景，就需要点击目录上方的“clade”按钮，接下来点击树枝，就能选中那个分支，再点击目录上方的“Hilight”按钮，就可以设置颜色背景，效果图像如下！</p>
</li>
</ol>
<p><img src="https://i.imgur.com/FNCS9nl.jpeg"></p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a><a class="next" href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/mcscanx-%E6%B5%8B%E8%AF%95/">MCScanX 测试</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Bioinfo | LTR identification and analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>