<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta property="og:type" content="website">
<meta property="og:title" content="FENG&#39;s Notes">
<meta property="og:url" content="http://example.com/page/6/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/page/6/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/12/Plant-Soybean-introduction/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/12/Plant-Soybean-introduction/" class="post-title-link" itemprop="url">大豆产业现状</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2020-12-12 12:27:45 / Modified: 19:51:16" itemprop="dateCreated datePublished" datetime="2020-12-12T12:27:45+08:00">2020-12-12</time>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>全球大豆年产量 <a target="_blank" rel="noopener" href="http://www.mczx.agri.cn/scts/gxyb/201910/t20191008_7206083.htm">3.42 亿吨</a>，主要为表中的6个国家生产。2019年全年，中国大豆播种面积 <a target="_blank" rel="noopener" href="http://www.stats.gov.cn/tjsj/zxfb/201912/t20191206_1716156.html">1.40 亿亩</a>（约 2300 万英亩，或 9.33 万平方公里，或 933 万公顷），产量 <a target="_blank" rel="noopener" href="http://www.cngrain.com/Publish/1/660407.html">1810 万吨</a>。中国 2019 年进口大豆  <a target="_blank" rel="noopener" href="https://www.thepaper.cn/newsDetail_forward_6206738">8851 万吨</a>，从巴西、美国、阿根廷三国进口大豆比例分别为 <a target="_blank" rel="noopener" href="http://pdf.dfcfw.com/pdf/H3_AP202004121377900207_1.PDF">65.1%、19.1%、9.9%</a>。也就是说，2019年全年中国自产加进口的大豆总量是 10661 万吨，占世界全年产量的 31.17%。</p>
<p><strong>为什么中国需要这么多大豆？</strong> 大豆主要用途是什么？<a target="_blank" rel="noopener" href="http://www.gov.cn/xinwen/2020-06/22/content_5520960.htm">我国是世界最大的大豆消费国，每年大豆消费量1亿吨以上，其中进口大豆消费量9000多万吨。目前，我国大豆消费已形成了食用和压榨两个相对独立的市场。其中，国产大豆主要用于直接食用、加工豆制品和大豆蛋白；进口大豆主要用于满足国内植物油和蛋白粕需求。</a> <a target="_blank" rel="noopener" href="https://www.yicai.com/news/100377221.html">国内居民对动物源蛋白需求增加，刺激了我国养殖行业快速增长，而动物饲料中的重要成分就是豆粕。</a><a target="_blank" rel="noopener" href="https://www.sohu.com/a/340624292_100032755">全球 1.2 亿吨猪肉产量，欧盟 2000 多万吨，中国是 5400 多万吨，美国不到 1200 万吨，我们占接近50%，欧盟占20%，美国占不到10%。目前我国人均年消费猪肉量为38kg，全年国内猪肉消费量大概在5500万吨</a>。</p>
<p><strong>大豆的单位面积产值并不高</strong> 大豆净利润远低于玉米：2012年，黑龙江省大豆每亩净利润256元，而玉米则为495元，粳稻为695元——<a target="_blank" rel="noopener" href="http://ww2.usc.cuhk.edu.hk/PaperCollection/Details.aspx?id=9316">难怪该地越来越多农民不愿意种大豆</a>。</p>
<p><strong>进口大豆价格</strong> “<a target="_blank" rel="noopener" href="http://chinawto.mofcom.gov.cn/article/ap/p/201907/20190702882188.shtml">2019年1至6月（截止到6月21日）大豆价格维持在325美元/吨左右，仍在2009年以来的低位徘徊，刚刚超过308美元/吨（约为每千克2.1元人民币）的成本线。</a>” 2019年4月4日，中国发布对美国的关税反制措施，将对原产于美国 106 项商品加征关税，其中大豆、玉米、小麦、牛肉、汽车、飞机和部分化学品等在列。同时，商务部公告称，中国加征关税的商品金额约为 500 亿美元。大豆是中美贸易的重要组成部分，此次公告对进口自美国的大豆加征 25%关税，标志着中美贸易摩擦再次升级。“<a target="_blank" rel="noopener" href="http://pg.jrj.com.cn/acc/Res/CN_RES/INDUS/2018/4/10/573a9dd8-293d-4246-a7e8-13c9f1a53b57.pdf">目前美豆到岸完税价约为 3300~3400 元/吨，根据测算，加征 25%关税之后到岸完税价提高至 4100 元/吨左右，进口成本增加 700 多元/吨（上涨幅度 22%左右）。</a>” <a target="_blank" rel="noopener" href="http://www.gov.cn/xinwen/2020-06/22/content_5520960.htm">2020年6月16日，9月份船期美湾大豆到我国港口完税成本约为每吨3190元，比1月初每吨下跌170元。国产大豆与进口大豆价格差距进一步拉大，目前国产大豆价格比进口大豆每吨高2000多元。</a></p>
<p><strong>国产大豆价格</strong> <a target="_blank" rel="noopener" href="http://www.gov.cn/xinwen/2020-06/22/content_5520960.htm">2020年6月12日，中储粮油脂公司委托中储粮网竞价销售储备轮换国产大豆60486吨，起拍价格为每吨4800元，最低成交价5140元，最高成交价5270元。</a></p>
<p><strong>大豆亩产量</strong> 中国大豆亩产量比美国低。“<a target="_blank" rel="noopener" href="https://www.guancha.cn/economy/2019_05_29_503615.shtml?s=zwyxgtjdt">根据我国大豆振兴计划，到2020年，全国大豆平均亩产达到135公斤，到2022年，全国大豆平均亩产达到140公斤，逐步缩小与世界大豆主产国的单产差距。</a>” <a target="_blank" rel="noopener" href="https://world.huanqiu.com/article/9CaKrnKbg1t">在整个黑龙江垦区，大豆平均亩产为167.5公斤，距离美国大豆的亩产200公斤差距已经不大。</a>根据媒体报道的大豆总产量计算，中国2019年大豆种植面积 1.4 亿亩，年产量 1810 万吨，平均亩产 129 公斤，按照 5000 元/吨售出价计算，合646元人民币。美国大豆种植面积由3100万公顷增至3566万公顷，年均增长1.8%；产量由9066万吨增至1.24亿吨，年均增长4%。美国2019年大豆种植面积料为 <a target="_blank" rel="noopener" href="http://chinawto.mofcom.gov.cn/article/ap/p/201907/20190702882188.shtml">3566万公顷</a>（5.35亿亩），产量 1.24 亿吨，平均亩产 232 公斤。</p>
<p>如果中国想要完全由自己来生产每年所消费的大豆，每年需要另外播种多少亩大豆，占用全国总播种面积多少？每年需要另外播种约四亿亩以上的大豆，占到全国总播种面积的15％～20％。这意味着必须减少如此幅度的谷物或高值新农业产品，实际上根本就不可考虑。这是国内大豆产量没有显著增加而大豆进口快速增长的基本原因。</p>
<p>2019年2月19日，新华社受权发布《中共中央、国务院关于坚持农业农村优先发展做好“三农”工作的若干意见》要求，严守<a target="_blank" rel="noopener" href="https://www.yicai.com/news/100121079.html">18亿亩耕地红线</a>，全面落实永久基本农田特殊保护制度，确保永久基本农田保持在15.46亿亩以上。按照18亿亩耕地的红线，这个面积高达120万平方公里。<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/wtNo0KQmHdWkv-gzm81dNg">一亩土地是666.67平米，乘以18亿亩，刚好是120万平方公里整数</a>，约等于西藏的面积。</p>
<p>如果按照当前中国大豆亩产量 129 公斤/亩来计算，如果需要满足国内总需求量 10661 万吨的生产，需要耕地面积 8.2 亿亩，合 54.67 万平方公里，比整个黑龙江省的面积（47.7 万平方公里）略大。</p>
<p><strong>中国和美国的大豆有什么不一样？</strong><br />
学者曾比较过中美大豆生产的投入费用。根据其分析，2009年至2012年，中国大豆生产的物质和人工平均费用，分别为166.96、133.18元；美国大豆生产的物质和人工平均费用，分别是236.54、21.26元。美国大豆生产的物质费用比中国高69.58元，人工费用比中国却低111.92元。<a target="_blank" rel="noopener" href="https://news.fx168.com/opinion/column/hujiajun/1806/2541828.shtml">两项合计，美国比中国大豆生产成本整体上低了42.34元。</a> 更为关键的是，美国大豆生产不仅投入成本低，亩产量还比中国高。2009年至2012年，中国大豆亩产量平均为142.46公斤，美国大豆亩产量平均则为201.32公斤，较中国高出58.86公斤。一进一出，美国大豆取得了绝对的价格优势。</p>
<p><strong>美国大豆单位面积产量为什么比中国大豆更高？</strong></p>
<p><strong>转基因大豆的原理</strong><br />
1976年，美国孟山都公司（Monsanto Company）推出一款高效除草剂，名为农达草甘膦（Roundup）。草甘膦能够抑制深根杂草生长，大大减少农作物的田间管理强度。对于大规模农场来说，农达是一个巨大福音。但是，具体到大豆农户，则喜忧参半，因为草甘膦在消灭杂草的同时，也会杀死豆苗。为了解决这个难题，孟山都公司又组织科学家，利用前期的转基因研究，1993年培育出了抗农达大豆品种。农户买了他们的大豆种子，即使喷洒草甘膦，也不会影响豆苗生长。</p>
<p>大豆基础研究：</p>
<ol>
<li>增加产量</li>
<li>提高抗虫性</li>
<li>提高含油量</li>
<li>提高植物蛋白含量</li>
</ol>
<table>
<thead>
<tr>
<th>国家</th>
<th>种植面积（英亩）</th>
<th>产量</th>
<th>出口量</th>
<th>进口量</th>
</tr>
</thead>
<tbody>
<tr>
<td>巴西</td>
<td>9509万</td>
<td>1.2500亿吨</td>
<td>7700万吨</td>
<td>-</td>
</tr>
<tr>
<td>美国</td>
<td>8670万</td>
<td>1.2400亿吨</td>
<td>4967万吨</td>
<td>-</td>
</tr>
<tr>
<td>阿根廷</td>
<td>4820万</td>
<td>5100万吨</td>
<td>820万吨</td>
<td>-</td>
</tr>
<tr>
<td>中国</td>
<td>2187万</td>
<td>1800万吨</td>
<td>-</td>
<td>8859万吨</td>
</tr>
<tr>
<td>印度</td>
<td>2680万</td>
<td>1346万吨</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td>巴拉圭</td>
<td>940万</td>
<td>1030万吨</td>
<td>630万吨</td>
<td>-</td>
</tr>
<tr>
<td>其他</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>-</td>
</tr>
<tr>
<td>总计</td>
<td>30641万</td>
<td>3.42亿吨</td>
<td>1.5亿吨</td>
<td>-</td>
</tr>
</tbody>
</table>
<p><img src="http://img.zhitongcaijing.com/images/contentformat/10a8eb25dc1e1cc86384b6337f9cb2f3.jpg" alt="大豆种植面积" /></p>
<p><img src="http://img.zhitongcaijing.com/images/contentformat/42907666545e4dd2a32c1065f9eda879.jpg" alt="大豆产量" /></p>
<p><img src="http://www.dwjmsw.com/Uploads/Article/image/20191118/20191118100510_56129.jpg" alt="中国大豆产量、进口量、消费量" /></p>
<p><img src="http://img.zhitongcaijing.com/images/contentformat/0905bdc7968f303cb5c4ca1f54d74f4a.jpg" alt="中国进口大豆来源" /></p>
<h3 id="参考资料"><a class="markdownIt-Anchor" href="#参考资料"></a> 参考资料</h3>
<ol>
<li><a target="_blank" rel="noopener" href="https://www.zhitongcaijing.com/content/detail/162164.html">西部证券研究报告：大豆之争，渔翁得利 2018-11-07</a></li>
<li><a target="_blank" rel="noopener" href="http://pdf.dfcfw.com/pdf/H3_AP202004121377900207_1.PDF">全球大豆价格还能一路上涨吗？2020-04-11</a></li>
<li><a target="_blank" rel="noopener" href="http://chinawto.mofcom.gov.cn/article/ap/p/201907/20190702882188.shtml">农民日报：美国大豆产业的发展现状如何？2019-07-16</a></li>
<li><a target="_blank" rel="noopener" href="http://www.xzfutures.com/files/document/20170510/6363000333771339151814280.pdf">兴证期货：大豆气象周报 2017-05-10</a></li>
<li><a href="">《开放时代》2014年第1期：大豆生产和进口的经济逻辑 2014-01-01</a></li>
<li><a target="_blank" rel="noopener" href="https://www.guancha.cn/economy/2019_05_29_503615.shtml?s=zwyxgtjdt">国产大豆振兴 2020年大豆种植面积将达到1.4亿亩 </a></li>
<li><a target="_blank" rel="noopener" href="https://www.qianzhan.com/analyst/detail/220/191120-8ccd4a58.html">前瞻经济学人：2019年全球大豆行业市场发展现状分析 全球大豆产量稳步增长 2019-11-21</a></li>
<li><a target="_blank" rel="noopener" href="http://finance.ifeng.com/topic/money/ddzz/index.shtml">凤凰网财经专题：大豆战争 2012年</a></li>
<li><a target="_blank" rel="noopener" href="http://www.cciee.org.cn/archiver/cciee/UpFile/Files/qqh/20191009140421154928.pdf">《全球化》2019年第5期：中国大豆产业发展：主要问题、原因以及对策建议</a></li>
<li><a target="_blank" rel="noopener" href="http://jjckb.xinhuanet.com/2010-05/07/content_220528.htm">新华网：中洋大豆之争 实乃产业链的抗衡 2010-05-07</a></li>
<li><a target="_blank" rel="noopener" href="http://www.cqvip.com/qk/91584x/201009/33584422.html">《求是》2010年 第9期 | 万宝瑞: 振兴我国大豆产业的根本途径</a></li>
<li><a target="_blank" rel="noopener" href="https://www.yicai.com/news/345460.html">第一财经：我国大豆产业急需发展专业合作社 2010-05-05</a></li>
<li><a target="_blank" rel="noopener" href="https://news.fx168.com/opinion/column/hujiajun/1806/2541828.shtml">财经报社：美国大豆是如何“征服”世界的？ 2018-06-09</a></li>
</ol>
<h3 id="附注"><a class="markdownIt-Anchor" href="#附注"></a> 附注</h3>
<p><strong>草甘膦的作用机理</strong></p>
<p>草甘膦是由著名的美国公司孟山都（Monsanto）研制成功的灭生性茎叶处理除草剂，具有高效 、广谱 、低毒（草甘膦的急性经口毒性比食盐都低，但后来有报道草甘膦具有肾毒性，我还没有认真查找相关的学术报道，先不做评论。）、低残留 、易于被微生物分解 ,不破坏土壤环境 ,对大多数植物具有灭生性等优点。后来又得益于抗草甘膦转基因作物的发展和推广，这种除草剂一直稳坐世界第一大农药的宝座。    草甘膦作为一种非选择性除草剂，它的作用机理是竞争性地抑制莽草酸途径中 5 - 烯醇丙酮莽草酸 - 3 - 磷酸合成酶 ( EPSPS)的活性 。该合成酶是真菌 、细菌 、藻类和高等植物体内芳香族氨基酸 (包括色氨酸 、酪氨酸 、苯丙氨酸 )生物合成过程中一个关键性的酶。具体的生理生化过程见下图。用草甘膦处理植物以后，植物体内含有这类氨基酸的蛋白质合成就会受阻，进而影响一系列代谢过程，导致植物死亡。草甘膦使用时一般将其制成异丙胺盐或钠盐。草甘膦除草性能优异，极易通过叶面少量通过根吸收，并运送到植物生长点。<br />
<img src="https://pic2.zhimg.com/80/fccf586c8c431cd3b71435d878b4851c_720w.jpg?source=1940ef5c" alt="" /></p>
<p><strong>转基因大豆抵御草甘膦的机理</strong></p>
<p>目前研究者主要从三个方面入手来改善作物的抗草甘膦特性：1. 通过使草甘膦抑制的酶过量表达，进而让植物吸收草甘膦之后能够进行正常代谢。但是目前用这个方法得到的抗草甘膦作物暴露在草甘膦之下时仍表现出生长抑制，未达到能够商业化的水平。2. 引入酶系统，在草甘膦进入植物体后迅速将其降解。目前这一策略也没有在生产上大量应用。3. 而应用于生产实践的上的主要方法是对草甘膦作用的酶进行修饰，降低它对草甘膦的敏感性。这里插一句背景知识，生物体内的酶大部分都是由氨基酸按照一定顺序排列组成的大分子物质，具有一定的空间结构，能够由特定酶催化，和特定酶反应的分子需要和这种酶的结构位点相结合才能发挥作用。改变了草甘膦作用的酶的结构，就可以降低甚至消除草甘膦对这种酶的作用效力。目前研究者们从大肠杆菌( <em>Escherichia coli</em>)、沙门氏杆菌( <em>Salmonella typhimurium</em>)和农杆菌( <em>Agrobacterium tumefaciens</em>) 等微生物中分离出了抗草甘膦的突变基因，而应用于生产的是从农杆菌菌株( <em>A. tumefaciens</em> sp. CP4) 上分离出的CP4基因。在大豆上导入了这种抗草甘膦基因，在大豆植株中表达出的CP4-EPSP合成酶对草甘膦的敏感性比较低，使得转基因植物体内的莽草酸途径可以正常进行。这样以来大豆就表现出了对草甘膦的抗性。</p>
<p><strong>转基因植物抗虫的机理</strong><br />
以转基因Bt抗虫作物为例, Bt蛋白为什么能够杀死昆虫？Bt进入昆虫的消化道, 昆虫的消化道是碱性环境, Bt蛋白和昆虫消化道上的受体结合, 引起肠穿孔, 从而导致害虫生长缓慢, 直至死亡。人的消化道结构和昆虫不一样, 而且人的肠道环境是酸性的, Bt蛋白进入胃中后很快在胃液中被消化降解为氨基酸, 对人体消化道没有任何作用。也就是说, Bt基因的杀虫功能具有很强的选择性, 只在鳞翅目昆虫的碱性消化道中发挥作用, 对人和哺乳动物都不会产生影响(Sheltonet al., 2002)。抗除草剂性状由于其巨大的经济效益和环境效益, 成为作物的第一大改良性状。2011年抗除草剂大豆、玉米、油菜、棉花、甜菜以及苜蓿种植面积达9 390万公顷, 占全球转基因作物种植总面积的59%。其中使用最为广泛的抗除草剂基因是从根瘤农杆菌中分离的编码455个氨基酸的抗草甘膦基因CP4-EP- SPS。安全性评价表明, CP4-EPSPS蛋白在哺乳动物消化系统中极易分解, 小鼠经口食入高达572 mg·kg–1(高于转基因大豆、玉米等植物中的1 300倍)的CP4-EPSPS蛋白, 亦无不良影响, 其在胃中的半衰期小于15秒, 在肠系统中小于10分钟。以上结果结合种子成分分析及动物营养研究表明, 转基因抗草甘膦大豆对人和动物是安全的(Harrison et al., 1998)。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/07/Life-initial-demo-page/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/07/Life-initial-demo-page/" class="post-title-link" itemprop="url">My Fast HexoPages</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-07 16:57:42" itemprop="dateCreated datePublished" datetime="2020-12-07T16:57:42+08:00">2020-12-07</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-08 14:09:25" itemprop="dateModified" datetime="2020-12-08T14:09:25+08:00">2020-12-08</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Life/" itemprop="url" rel="index"><span itemprop="name">Life</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>abstract</p>
          <!--noindex-->
            <div class="post-button">
              <a class="btn" href="/2020/12/07/Life-initial-demo-page/#more" rel="contents">
                Read more &raquo;
              </a>
            </div>
          <!--/noindex-->
        
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/06/Bioinfo-HiC-pro-installation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/06/Bioinfo-HiC-pro-installation/" class="post-title-link" itemprop="url">安装HiC-Pro （HiC数据预处理工具）与绘制HiC染色体互作图谱</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-06 13:17:53" itemprop="dateCreated datePublished" datetime="2020-12-06T13:17:53+08:00">2020-12-06</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-05-29 17:27:59" itemprop="dateModified" datetime="2021-05-29T17:27:59+08:00">2021-05-29</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>HiC-Pro可以用来处理Hi-C数据，从原始的fastq文件(配对端Illumina数据)到标准化的交互图谱。简单的来说就是将Hi-C数据比对到拼装好的参考基因组上，并形成交互文件去存储Hi-C数据。</p>
<p>下载 HiC-Pro VERSION 2.11.4 安装包，依赖 Python 2。（后续更新：HiC-Pro version 3.0.0 要求 Python 3）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/nservant/HiC-Pro.git </span><br></pre></td></tr></table></figure>
<h3 id="安装-hic-pro-version-2114"><a class="markdownIt-Anchor" href="#安装-hic-pro-version-2114"></a> 安装 HiC-Pro VERSION 2.11.4</h3>
<p>由于需要python 2 （2.7以上），我之前已经在系统安装Anaconda3，并在里面分别安装了python3.7和python2.7，(<a target="_blank" rel="noopener" href="https://genehub.wordpress.com/2020/04/21/1070/">原文链接</a>, 这里只需要切换到python2.7环境即可。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">source activate python27</span><br></pre></td></tr></table></figure>
<p>再直接用conda命令安装相关软件</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">conda install -y samtools bowtie2 R </span><br><span class="line">conda install -y pysam bx-python numpy scipy </span><br><span class="line">#打开R </span><br><span class="line">install.packages(c(&#x27;ggplot2&#x27;,&#x27;RColorBrewer&#x27;)) </span><br></pre></td></tr></table></figure>
<p>通过 conda 安装 samtools，samtools 运行会报错。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools: error while loading shared libraries: libcrypto.so.1.0.0: cannot open shared object file: No such file or directory</span><br></pre></td></tr></table></figure>
<p>现在samtools的版本已经在1.9以上了，但是conda安装的samtools版本依然是1.7，这是出现问题的原因。<a target="_blank" rel="noopener" href="https://blog.csdn.net/zhangjunya/article/details/108235796">解决办法</a>：指定另一个更新的版本，重新安装如下。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -c bioconda samtools=1.9 --force-reinstall</span><br></pre></td></tr></table></figure>
<p>或者刚开始安装的时候：<code>conda install samtools=1.9</code>，就不会再出现上述问题。</p>
<p>不知道为什么bx-python安装失败，换成下面的方法，可以安装。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -c conda-forge -c bioconda bx-python</span><br></pre></td></tr></table></figure>
<p>然后进入HiC-Pro目录，修改config-install 文件，设置依赖软件的路径。PREFIX不要设置当前目录，会被覆盖。CLUSTER_SYS有四个选项，分别是TORQUE, SGE, SLURM 或 LSF，我随便填了一个TORQUE。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#</span><br><span class="line"> Paths and Settings  - Start editing here !</span><br><span class="line"> #</span><br><span class="line"> PREFIX = /home/user/local/scripts/HiC-Pro</span><br><span class="line"> BOWTIE2_PATH = /home/user/local/app/anaconda3/envs/python27/bin/</span><br><span class="line"> SAMTOOLS_PATH = /home/user/local/bin/</span><br><span class="line"> R_PATH = /home/user/local/app/anaconda3/envs/python27/bin</span><br><span class="line"> PYTHON_PATH = /home/user/local/app/anaconda3/envs/python27/bin</span><br><span class="line"> CLUSTER_SYS = TORQUE</span><br></pre></td></tr></table></figure>
<p>接下里 make configure ，通过之后屏幕显示如下信息</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">(python27) xxxxx/HiC-Pro$ make configure       </span><br><span class="line"> </span><br><span class="line">make -f ./scripts/install/Makefile CONFIG_SYS=./config-install.txt</span><br><span class="line"> make[1]: Entering directory &#x27;/home/-/local/app/HiC-Pro&#x27;</span><br><span class="line"> ./scripts/install/install_dependencies.sh -c ./config-install.txt -o  /home/-/local/scripts/HiC-Pro/HiC-Pro_2.11.4 -q</span><br><span class="line"> Make sure internet connection works for your shell prompt under current user&#x27;s privilege …</span><br><span class="line"> Starting HiC-Pro installation …</span><br><span class="line"> export /home/--/local/app/anaconda3/envs/python27/bin/ in PATH</span><br><span class="line"> export /home/-/local/bin/ in PATH</span><br><span class="line"> export /home/-/local/app/anaconda3/envs/python27/bin in PATH</span><br><span class="line"> export /home/-/local/app/anaconda3/envs/python27/bin in PATH</span><br><span class="line"> Checking dependencies … </span><br><span class="line"> Checking Python libraries …</span><br><span class="line"> The required Python libraries appear to be already installed. </span><br><span class="line"> Checking R installation …</span><br><span class="line"> The required R packages appear to be already installed. </span><br><span class="line"> Bowtie2 Aligner appears to be already installed. </span><br><span class="line"> Samtools appears to be already installed. </span><br><span class="line"> Dependencies checked !</span><br><span class="line"> Check HiC-Pro configuration … </span><br><span class="line"> Configuration for TORQUE/PBS system.</span><br><span class="line"> make[1]: Leaving directory &#x27;/home/-/local/app/HiC-Pro&#x27;</span><br></pre></td></tr></table></figure>
<p>然后 make，通过即安装成功。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">(some information are not shown here due to size limitation)</span><br><span class="line">cp -Ri /home/-/local/app/HiC-Pro /home/-/local/scripts/HiC-Pro/HiC-Pro_2.11.4 </span><br><span class="line">HiC-Pro installed in /home/-/local/scripts/HiC-Pro/HiC-Pro_2.11.4 !</span><br></pre></td></tr></table></figure>
<p>测试</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">(python27) XXX:~/local/scripts/HiC-Pro/HiC-Pro_2.11.4/bin$ ./HiC-Pro -h</span><br><span class="line"> usage : HiC-Pro -i INPUT -o OUTPUT -c CONFIG [-s ANALYSIS_STEP] [-p] [-h] [-v]</span><br><span class="line"> Use option -h--help for more information</span><br><span class="line"> HiC-Pro 2.11.4</span><br><span class="line"> OPTIONS</span><br><span class="line"> -i--input INPUT : input data folder; Must contains a folder per sample with input files</span><br><span class="line">    -o--output OUTPUT : output folder</span><br><span class="line">    -c--conf CONFIG : configuration file for Hi-C processing</span><br><span class="line">    [-p--parallel] : if specified run HiC-Pro on a cluster</span><br><span class="line">    [-s--step ANALYSIS_STEP] : run only a subset of the HiC-Pro workflow; if not specified the complete workflow is run</span><br><span class="line">       mapping: perform reads alignment - require fast files</span><br><span class="line">       proc_hic: perform Hi-C filtering - require BAM files</span><br><span class="line">       quality_checks: run Hi-C quality control plots</span><br><span class="line">       merge_persample: merge multiple inputs and remove duplicates if specified - require .validPairs files</span><br><span class="line">       build_contact_maps: Build raw inter/intrachromosomal contact maps - require .allValidPairs files</span><br><span class="line">       ice_norm : run ICE normalization on contact maps - require .matrix files</span><br></pre></td></tr></table></figure>
<h3 id="optional-安装-hic-pro-version-300"><a class="markdownIt-Anchor" href="#optional-安装-hic-pro-version-300"></a> <s>Optional: 安装 HiC-Pro VERSION 3.0.0</s></h3>
<p><s>conda create --name python38 python=3.8</s><br />
<s>source activate python38</s><br />
<s>conda install conda</s><br />
<s># 前面的conda是anaconda3的conda，后面的conda是python3.8的conda</s></p>
<p><s>conda install -c bioconda bowtie2</s><br />
<s>conda install -c bioconda samtools=1.9</s><br />
<s>conda install -c conda-forge -c bioconda bx-python</s></p>
<p><s>在conda里面安装R和pysam遇到很多问题，更换python版本3.7/3.8/3.9都没成功，真是浪费时间，放弃。</s></p>
<h3 id="使用hic-pro对hic数据进行分析的过程"><a class="markdownIt-Anchor" href="#使用hic-pro对hic数据进行分析的过程"></a> 使用HiC-Pro对HiC数据进行分析的过程</h3>
<p>1. 在工作目录（例如HiC-analysis-workdir），建立若干文件夹，例如index、rawdata，方便管理，当然这些数据也可以在其他位置。</p>
<p>2. 在index路径下，对参考基因组或者scaffolds建立index。HiC-Pro调用的比对软件是bowtie2。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bowtie2-build --threads 20 ref.fasta ref.fasta</span><br></pre></td></tr></table></figure>
<p>3. 在工作目录，建立一个文本文件，内含参考基因组（或scaffolds）各个染色体名称和长度的信息。可以用samtools fadix，也可以用iTools或R语言等工具。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ cat chromosome_size.txt</span><br><span class="line"> Chr01   11111</span><br><span class="line"> Chr02   1233333</span><br><span class="line"> Chr03   4444</span><br><span class="line"> Chr04   105314844</span><br><span class="line">(continue ...)</span><br></pre></td></tr></table></figure>
<p>4. 在工作目录，新建一个文件夹（例如rawdata），用来存放作为input的测序数据。注意不能直接将测序文件放到rawdata目录下，而是要按照样品名先建立子文件夹（例如sample1、sample2）。sample1目录下即存放样品的测序文件，包含F/R两个测序文件，文件应该是.fastq(.gz)的命名，不可写成fq(.gz)；可以是软链接的形式。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207122951.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207122951.png?w=506" alt="" /></a></p>
<p>4. 基因组的酶切位点识别。我用的MboI（酶切位点：^GATC）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python2.7 XXX/HiC-Pro/HiC-Pro_2.11.4/bin/utils/digest_genome.py -r ^GATC -o mboi_resfrag_genome.bed Ref_genome.fa</span><br></pre></td></tr></table></figure>
<p>得到的酶切信息：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"> Chr01   0       3282    HIC_Chr01_1     0       +</span><br><span class="line"> Chr01   3282    3609    HIC_Chr01_2     0       +</span><br><span class="line"> Chr01   3609    3625    HIC_Chr01_3     0       +</span><br><span class="line"> Chr01   3625    4485    HIC_Chr01_4     0       +</span><br><span class="line"> Chr01   4485    5312    HIC_Chr01_5     0       +</span><br><span class="line"> Chr01   5312    5327    HIC_Chr01_6     0       +</span><br><span class="line"> Chr01   5327    5958    HIC_Chr01_7     0       +</span><br><span class="line"> Chr01   5958    6475    HIC_Chr01_8     0       +</span><br><span class="line"> Chr01   6475    7545    HIC_Chr01_9     0       +</span><br><span class="line">(continue ...)</span><br></pre></td></tr></table></figure>
<p>5. 将HiC-Pro程序安装包目录下的 config-hicpro.txt 文件拷贝到工作目录，对其进行编辑。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"># N_CPU，例如N_CPU = 24</span><br><span class="line"># JOB_NAME，例如JOB_NAME = Ga</span><br><span class="line"># JOB_WALLTIME，例如JOB_WALLTIME = 11:00:00</span><br><span class="line"># JOB_QUEUE，例如JOB_QUEUE = normal</span><br><span class="line"># JOB_MAIL，例如JOB_MAIL = 1570447120@qq.com</span><br><span class="line"># BOWTIE2_IDX_PATH，填写用bowtie2对reference建立索引的目录，注意是目录，而不是索引路径!如/PATH/to/Ga.ref</span><br><span class="line"># REFERENCE_GENOME，例如Ga 注意：这里REFERENCE_GENOME一定要和bowtie2建立的索引对应上,名字为物种名</span><br><span class="line"># GENOME_SIZE，即生成的文件Ga.fasta.size（最好使用绝对路径）</span><br><span class="line"># GENOME_FRAGMENT，用digest_genome.py程序生成的文件Ga.HindIII.txt</span><br><span class="line"># LIGATION_SITE，酶的序列，例如HindIII，则为AAGCTAGCTT；如果是MboI则序列为GATCGATC。</span><br><span class="line"># MIN_FRAG_SIZE，例如MIN_FRAG_SIZE = 100</span><br><span class="line"># MAX_FRAG_SIZE，例如MAX_FRAG_SIZE = 160000</span><br><span class="line"># MIN_INSERT_SIZE，例如MIN_INSERT_SIZE = 200</span><br><span class="line"># MAX_INSERT_SIZE，例如MAX_INSERT_SIZE = 600</span><br><span class="line"># GET_PROCESS_SAM，例如GET_PROCESS_SAM = 1</span><br></pre></td></tr></table></figure>
<p>6. 运行</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">/PATH/TO/HiC-Pro_2.11.1/bin/HiC-Pro -i /PATH/TO/rawdata --output hicpro_output --conf config-hicpro.txt </span><br><span class="line"># --input reads所在的文件</span><br><span class="line"># --output 输出的文件夹</span><br><span class="line"># --conf 配置文件</span><br></pre></td></tr></table></figure>
<p>7. 运行结果</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br></pre></td><td class="code"><pre><span class="line">.</span><br><span class="line">├── bowtie_results</span><br><span class="line">│   ├── bwt2</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.bwt2merged.bam</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.mapstat</span><br><span class="line">│   │       ├── mysample_R2_V3_20201206.fa.bwt2merged.bam</span><br><span class="line">│   │       ├── mysample_R2_V3_20201206.fa.mapstat</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.bam</span><br><span class="line">│   │       └── _V3_20201206.fa.bwt2pairs.pairstat</span><br><span class="line">│   ├── bwt2_global</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.bwt2glob.bam</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap.fastq</span><br><span class="line">│   │       ├── mysample_R2_V3_20201206.fa.bwt2glob.bam</span><br><span class="line">│   │       └── mysample_R2_V3_20201206.fa.bwt2glob.unmap.fastq</span><br><span class="line">│   └── bwt2_local</span><br><span class="line">│       └── sample1</span><br><span class="line">│           ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_bwt2loc.bam</span><br><span class="line">│           ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_trimmed.fastq</span><br><span class="line">│           ├── mysample_R2_V3_20201206.fa.bwt2glob.unmap_bwt2loc.bam</span><br><span class="line">│           └── mysample_R2_V3_20201206.fa.bwt2glob.unmap_trimmed.fastq</span><br><span class="line">├── config-hicpro.txt</span><br><span class="line">├── hic_results</span><br><span class="line">│   ├── data</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── sample1.allValidPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.DEPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.DumpPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.FiltPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.REPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.RSstat</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.SCPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.SinglePairs</span><br><span class="line">│   │       └── _V3_20201206.fa.bwt2pairs.validPairs</span><br><span class="line">│   ├── matrix</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── iced</span><br><span class="line">│   │       │   └── 200000</span><br><span class="line">│   │       │       ├── sample1_200000_iced.matrix</span><br><span class="line">│   │       │       └── sample1_200000_iced.matrix.biases</span><br><span class="line">│   │       └── raw</span><br><span class="line">│   │           └── 200000</span><br><span class="line">│   │               ├── sample1_200000_abs.bed</span><br><span class="line">│   │               └── sample1_200000.matrix</span><br><span class="line">│   ├── pic</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── plotHiCContactRanges_sample1.pdf</span><br><span class="line">│   │       ├── plotHiCFragment_sample1.pdf</span><br><span class="line">│   │       ├── plotHiCFragmentSize_sample1.pdf</span><br><span class="line">│   │       ├── plotMappingPairing_sample1.pdf</span><br><span class="line">│   │       └── plotMapping_sample1.pdf</span><br><span class="line">│   └── stats</span><br><span class="line">│       └── sample1</span><br><span class="line">│           ├── sample1_allValidPairs.mergestat</span><br><span class="line">│           ├── sample1mysample_R1.mmapstat</span><br><span class="line">│           ├── sample1mysample_R2.mmapstat</span><br><span class="line">│           ├── sample1.mpairstat</span><br><span class="line">│           └── sample1.mRSstat</span><br><span class="line">├── logs</span><br><span class="line">│   └── sample1</span><br><span class="line">│       ├── build_raw_maps.log</span><br><span class="line">│       ├── mysample_R1_bowtie2.log</span><br><span class="line">│       ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_bowtie2.log</span><br><span class="line">│       ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_readsTrimming.log</span><br><span class="line">│       ├── mysample_R2_bowtie2.log</span><br><span class="line">│       ├── mysample_R2_V3_20201206.fa.bwt2glob.unmap_bowtie2.log</span><br><span class="line">│       ├── mysample_R2_V3_20201206.fa.bwt2glob.unmap_readsTrimming.log</span><br><span class="line">│       ├── ice_200000.log</span><br><span class="line">│       ├── make_Rplots.log</span><br><span class="line">│       ├── mapped_2hic_fragments.log</span><br><span class="line">│       ├── mapping_combine.log</span><br><span class="line">│       ├── mapping_stats.log</span><br><span class="line">│       ├── mapping_step1.log</span><br><span class="line">│       ├── mapping_step2.log</span><br><span class="line">│       ├── mergeSAM.log</span><br><span class="line">│       ├── merge_stats.log</span><br><span class="line">│       ├── merge_valid_interactions.log</span><br><span class="line">│       ├── plot_hic_contacts.Rout</span><br><span class="line">│       ├── plot_hic_fragment.Rout</span><br><span class="line">│       ├── plot_mapping_portion.Rout</span><br><span class="line">│       └── plot_pairing_portion.Rout</span><br><span class="line">├── rawdata -&gt; /PATH/TO//rawdata/</span><br><span class="line">└── tmp</span><br></pre></td></tr></table></figure>
<p><strong>画图</strong></p>
<p>HiC-Pro运行结果中，sample1.allValidPairs文件就表示 pair end reads 在染色体或 scaffolds 上面位置的信息，该文件是文本格式，需要进行格式转化成为.hic文件，才能放到 juicebox 软件中得到可视化的互作图谱。</p>
<p>如何转换成.hic文件？使用HiC-Pro安装路径下的<code>hicpro2juicebox.sh</code>程序按照下面的方式运行即可得到.hic文件。<br />
需要下载 juicer_tools_1.22.01.jar 文件作为输入文件。下载地址：<a target="_blank" rel="noopener" href="https://github.com/aidenlab/juicer/wiki/Download">https://github.com/aidenlab/juicer/wiki/Download</a><br />
备注：我一开始下载错了，去<a target="_blank" rel="noopener" href="https://github.com/aidenlab/Juicebox/wiki/Download">https://github.com/aidenlab/Juicebox/wiki/Download</a> 页面下了个Juicebox.jar，结果不能用于下面的步骤。原来这是 JuiceBox，不是 Juicer！从论坛发现遇到这个问题的不止我一个人<a target="_blank" rel="noopener" href="https://groups.google.com/g/hic-pro/c/DF0zfZOivsE?pli=1">https://groups.google.com/g/hic-pro/c/DF0zfZOivsE?pli=1</a>。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">/PATH/TO/HiC-Pro/HiC-Pro_2.11.4/bin/utils/hicpro2juicebox.sh -i /PATH/TO//sample1.allValidPairs -g /PATH/TO//chromosome_sizes -j /PATH/TO/juicer_tools_1.22.01.jar -r /PATH/TO//mboi_resfrag_AMgenome.bed -t temp -o /PATH/TO/output</span><br></pre></td></tr></table></figure>
<p>用上面的命令就得到了 .hic 文件，然后从 Linux 服务器将该文件下载到本地客户端（客户端即我个人电脑用的 windows 系统），直接运行 Juicebox_1.11.08.exe 二进制程序，双击运行，菜单选 open，选择 .hic 文件，就得到下面的图谱，可以 export 成 pdf 或 svg 格式的图片。</p>
<p><img src="https://gblobscdn.gitbook.com/assets%2F-LijQZQE8sc7HbVdoxEA%2F-Lj2gavu4NTsjkOUMYaT%2F-Lj2js_V0JRQoDXKpVvE%2FScreen%20Shot%202019-07-05%20at%2012.51.13%20PM.png?alt=media&amp;token=fdae559d-237e-45dd-a0dd-a12a22ff7506" alt=".hic文件导入juicebox软件之后的示意图" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/05/Bioinfo-compare-two-genome-sequences-by-mummer/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/05/Bioinfo-compare-two-genome-sequences-by-mummer/" class="post-title-link" itemprop="url">两个基因组序列相互比较：Mummer</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-05 21:11:00" itemprop="dateCreated datePublished" datetime="2020-12-05T21:11:00+08:00">2020-12-05</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-02-25 17:42:04" itemprop="dateModified" datetime="2021-02-25T17:42:04+08:00">2021-02-25</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>直接对两个基因组进行比较，用Mummer：</p>
<p>例如，将细菌1（H_pyloriJ99_Eslice.fasta）与细菌2（H_pylori26695_Eslice.fasta）进行比较：</p>
<p>mummer -mum -b -c H_pylori26695_Eslice.fasta H_pyloriJ99_Eslice.fasta &gt; mummer.mums<br />
mummerplot -x [0,275287] -y [0,265111] -postscript -p mummer mummer.mums</p>
<p>结果：</p>
<p><img src="https://genehub.files.wordpress.com/2019/07/mummer.png" alt="mummer.png" /></p>
<p>如果换成两个大基因组呢？800M以上的基因组，结果报错了，尺寸太大，程序默认不支持这么多，只允许400M。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># reading input file &quot;Jxxx.fasta&quot; of length 854295539</span><br><span class="line"># construct suffix tree for sequence of length 854295539</span><br><span class="line"># (maximum reference length is 536870908)</span><br><span class="line"># (maximum query length is 4294967295)</span><br><span class="line"># process 8542955 characters per dot</span><br><span class="line">mummer: suffix tree construction failed: textlen=854295539 larger than maximal textlen=536870908</span><br></pre></td></tr></table></figure>
<p>我的版本是：MUMmer3.23</p>
<p>需要对mummer添加参数重新编译：</p>
<p>The size of your reference sequence is larger than is supported by default.<br />
You can either break your reference into multiple files and run each<br />
individually, or recompile the mummer package with:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">make clean</span><br><span class="line">make CPPFLAGS=&quot;-O3 -DSIXTYFOURBITS&quot;</span><br></pre></td></tr></table></figure>
<p><strong>20201204 更新</strong></p>
<p>MUMMER v4 可以直接对大型基因组进行比较，他的功能已经升级了。<a target="_blank" rel="noopener" href="https://github.com/mummer4/mummer/releases/download/v4.0.0rc1/mummer-4.0.0rc1.tar.gz">mummer-4.0.0rc1.tar.gz下载链接</a></p>
<p>如何对同一个物种的不同版本基因组序列比较？</p>
<p>以大豆<em>Glycine max</em>为例，大豆20条染色体，基因组尺寸约1 Gb，最新基因组版本Gmax_508_a4.v1，上一个版本是Gmax_275_a2.v1。</p>
<p><strong>一、先用nucmer对两个genome序列进行比较，生成delta文件。</strong><br />
我设定的最小匹配单元长度是1 kb，也就是1 kb内不允许错配。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">nucmer --prefix=GM508_GM275_genome --minmatch=1000 --threads=12 ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa ~/databases/Glycine_max/phytozome.12/Gmax_275_v2.0.softmasked.fa</span><br></pre></td></tr></table></figure>
<p><strong>二、然后可以delta-filter自定义筛选。</strong><br />
比如说我只想看contig和reference两者唯一匹配，并且长度在1000，相似度大于90.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">delta-filter -i 89 -l 1000 -1 GM508_GM275_genome.delta &gt;GM508_GM275_genome_i89_l1000_1.delta</span><br></pre></td></tr></table></figure>
<p><strong>三、再用mummerplot命令画图。</strong><br />
注意Rfile和Qfile主要功能是提供两个基因组各自包含的若干染色体（或者scaffold、contig）序列，不加这个参数，出来的图片没有染色体名称标记的。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mummerplot -Rfile ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa -Qfile ~/databases/Glycine_max/phytozome.12/Gmax_275_v2.0.softmasked.fa -postscript -p GM508_GM275_genome_i89_l1000_1 GM508_GM275_genome_i89_l1000_1.delta</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># `gnuplot 5.4 patchlevel 1  </span><br><span class="line"># Reading delta file GM508_GM275_genome_i89_l1000_1.delta  </span><br><span class="line"># Writing plot files GM508_GM275_genome_i89_l1000_1.fplot, GM508_GM275_genome_i89_l1000_1.rplot  </span><br><span class="line"># Writing gnuplot script GM508_GM275_genome_i89_l1000_1.gp  </span><br><span class="line"># Rendering plot GM508_GM275_genome_i89_l1000_1.ps`</span><br></pre></td></tr></table></figure>
<p><strong>20201230 备注</strong></p>
<p>更换一台机器之后，运行画图遇到错误。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ mummerplot -Rfile ./AM_genome_V5_20201211.fa -Qfile ./20201228_3ddna.fa -postscript -p plot_test V5_vs_3dV1_genome_i89_l1000_1.delta </span><br><span class="line">gnuplot 4.6 patchlevel 2</span><br><span class="line">Reading delta file V5_vs_3dV1_genome_i89_l1000_1.delta</span><br><span class="line">Writing plot files plot_test.fplot, plot_test.rplot</span><br><span class="line">Writing gnuplot script plot_test.gp</span><br><span class="line">Rendering plot plot_test.ps</span><br><span class="line">WARNING: Unable to run &#x27;false plot_test.gp&#x27;, Inappropriate ioctl for device</span><br></pre></td></tr></table></figure>
<p>切换到 python37 或者 python27 环境，用conda安装了更高版本的 gnuplot ，再安装 libjpeg-turbo 就可以运行 gnuplot 了。但是仍然无法运行mummerplot。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"># 用conda安装gnuplot</span><br><span class="line">conda install -c bioconda gnuplot</span><br><span class="line"></span><br><span class="line"># 启动gnuplot遇到错误</span><br><span class="line">$ gnuplot --version</span><br><span class="line">gnuplot: error while loading shared libraries: libjpeg.so.8: cannot open shared object file: No such file or directory</span><br><span class="line"></span><br><span class="line"># 安装 libjpeg-turbo，就可以运行gnuplot了</span><br><span class="line">$ conda install -c conda-forge libjpeg-turbo</span><br><span class="line"></span><br><span class="line"># 成功运行gnuplot</span><br><span class="line">(python27) [fenglei@localhost ~]$ gnuplot --version</span><br><span class="line">gnuplot 5.0 patchlevel 3</span><br></pre></td></tr></table></figure>
<p><strong>四、生成比对图片</strong><br />
把ps文件放到windows系统，用Photoshop软件打开上面生成的ps文件即可看到图。当然也可以直接在linux里面用convert命令将ps文件转换为pdf/png文件，但是我的linux系统没有convert命令。如图可以见到两个版本基因组的差异。11和14号染色体有明显的结构差异。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome_i89_l1000_1.jpg"><img src="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome_i89_l1000_1.jpg?w=1024" alt="" /></a></p>
<p><strong>五、如果不对delta文件做自定义筛选呢？</strong><br />
可见到大量的点，浅蓝色是反向匹配，深蓝色是正向匹配。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome.jpg"><img src="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome.jpg?w=1024" alt="" /></a></p>
<p><strong>六、如果两个物种相比较呢？</strong><br />
要注意minmatch参数不能设太大，我把两个物种（Glycine max and Glycine soja）的minmatch设置为1000000，虽然运行时间比较短，但是结果输出一个空白delta文件。设置为minmatch=100测试，需要的时间更长，得到一个满屏铺满蓝点的图，根本看不清。设置minmatch=1000，并用delta-filter对比对的序列做筛选，进而得到下图。可见到11和13号染色体的区别。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">nucmer --prefix=GM508_Gsoja_genome --minmatch=1000 --threads=12 ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa ~/databases/03_Glycine_soja_W05_2019/Gsoja.W05.genome.fadelta-filter -i 89 -l 1000 -1 GM508_Gsoja_genome.delta &gt; GM508_Gsoja_genome_i89_l1000_1.deltamummerplot -Rfile ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa -Qfile ~/databases/03_Glycine_soja_W05_2019/Gsoja.W05.genome.fa -postscript -p GM508_Gsoja_genome_i89_l1000_1 GM508_Gsoja_genome_i89_l1000_1.delta</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/gm508_gsoja_genome_i89_l1000_1.jpg"><img src="https://genehub.files.wordpress.com/2020/12/gm508_gsoja_genome_i89_l1000_1.jpg?w=1024" alt="" /></a></p>
<p>补充：画图可以用dotPlotly包，参考 <a target="_blank" rel="noopener" href="https://github.com/tpoorten/dotPlotly/tree/master/example">https://github.com/tpoorten/dotPlotly/tree/master/example</a></p>
<p>主要是R语言：<a target="_blank" rel="noopener" href="https://github.com/tpoorten/dotPlotly/blob/master/mummerCoordsDotPlotly.R">https://github.com/tpoorten/dotPlotly/blob/master/mummerCoordsDotPlotly.R</a></p>
<p>补充：如何理解上述作图过程中的几个文件呢？</p>
<p>delta文件（在<a target="_blank" rel="noopener" href="http://mummer.sourceforge.net/manual/#HumanCompare">http://mummer.sourceforge.net/manual/#HumanCompare</a> 页面的The delta format section下面有详细解释）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">/path/Gmax_508_v4.0.softmasked.fa /path/G.max-mitochondrion.fasta</span><br><span class="line"> NUCMER</span><br><span class="line">   Gm13 ChrM 45225048 402558     # 匹配区段所在的染色体</span><br><span class="line">   21773198 21776178 114322 117301 3 3 0   # 前四个数字：两个染色体上面的区段坐标。后三个数字：错误数(不相同碱基+indel碱基数)，相似错误(非正匹配得分) 终止密码子(NUCMER为0)。 如果结束大于起始，表示在负链。英文版解释：The three digits following the location coordinates are the number of errors (non-identities + indels), similarity errors (non-positive match scores), and stop codons (does not apply to DNA alignments, will be &quot;0&quot;)</span><br><span class="line">   8  # query的8位置有“缺失”？？？如果前面有负号，则表示此处为“插入”？相对ref序列而言。</span><br><span class="line">   0  # 表示当前匹配结束</span><br><span class="line">   Gm20 ChrM 47846026 402558</span><br><span class="line">   14299558 14301282 188578 190302 3 3 0</span><br><span class="line">   0 </span><br><span class="line">   Gm17 ChrM 41740656 402558</span><br><span class="line">   23669908 23674616 270387 265667 78 78 0</span><br><span class="line">   -499</span><br><span class="line">   -1   # 这连续的“-1”是什么意思？</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -106</span><br><span class="line">   -2</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -2</span><br><span class="line">   -2182</span><br><span class="line">   1263</span><br><span class="line">   -230</span><br><span class="line">   0 </span><br><span class="line">(output continues ...)</span><br></pre></td></tr></table></figure>
<p><strong>七、输出适合肉眼观察的Tabular格式比对结果</strong><br />
上面的delta文件不便于我们肉眼阅读，可用show-coords -r ${prefix}.delta &gt; ${prefix}.coords命令之后得到更适合阅读的格式（coords文件）。<br />
有点类似BLAST的“-outfmt 6”参数输出的Tabular格式<br />
如下，以ref序列的染色体排序的。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">   [S1]     [E1]       [S2]     [E2]    [LEN 1]  [LEN 2]    [% IDY]   [TAGS]</span><br><span class="line">41311147 41314402     161362   158107       3256     3256      99.91   Gm08 ChrM</span><br><span class="line">24034371 24036869     210827   213325       2499     2499      99.88   Gm12 ChrM</span><br><span class="line">24038545 24043058      13087     8578       4514     4510      99.78   Gm12 ChrM</span><br><span class="line">21773198 21776178     114322   117301       2981     2980      99.90   Gm13 ChrM</span><br><span class="line">23669908 23674616     270387   265667       4709     4721      98.35   Gm17 ChrM</span><br><span class="line">14299558 14301282     188578   190302       1725     1725      99.83   Gm20 ChrM</span><br></pre></td></tr></table></figure>
<p><strong>八、输出适合肉眼观察的Pairwise格式比对结果</strong><br />
可指定特定的染色体或scaffold用于输出到屏幕，输出格式类似BLAST的“-outfmt 1”参数。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">show-aligns seqA_vs_seqB_genome_i89_l1000_1.delta Chrxxx Chryyy | less</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">============================================================</span><br><span class="line">-- Alignments between xxx and yyy</span><br><span class="line"></span><br><span class="line">-- BEGIN alignment [ +1 1 - 245034 | -1 94962171 - 94717077 ]</span><br><span class="line"></span><br><span class="line">                  10|       20|       30|       40|</span><br><span class="line">1          cccaaccccgagccctaaaccctaaaccccaaacccccaaccctgagcc</span><br><span class="line">94962171   cccaaccccgagccctaaaccctaaaccccaaacccccaaccctgagcc</span><br><span class="line">                                                            </span><br><span class="line">                  10|       20|       30|       40|</span><br><span class="line">50         tcgagccctaagccttgagcccgagccctaagccccgatccctaagccc</span><br><span class="line">94962122   tcgagccctaagccttgagcccgagccctaagccccgatccctaagccc</span><br><span class="line">                                                            </span><br><span class="line">                  10|       20|       30|       40|</span><br><span class="line">99         tgaaccccgaaccctcaaccccgaaccccaaaccctcaaccccaaaccc</span><br><span class="line">94962073   tgaaccccgaaccctcaaccccgaaccccaaaccctcaaccccaaaccc</span><br></pre></td></tr></table></figure>
<p>参考：</p>
<ol>
<li>MUMMER 两个基因组间比较 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/c12f2a117892">https://www.jianshu.com/p/c12f2a117892</a></li>
<li>MUMMER 两个基因组间比较 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/9701c9add82d">https://www.jianshu.com/p/9701c9add82d</a></li>
<li>最详细的MUMMER中文文档 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/2e184e5c15b7">https://www.jianshu.com/p/2e184e5c15b7</a></li>
<li>How to understand MUMMER delta format？<a target="_blank" rel="noopener" href="http://mummer.sourceforge.net/manual/#HumanCompare">http://mummer.sourceforge.net/manual/#HumanCompare</a></li>
<li>用R绘制点状图：<a target="_blank" rel="noopener" href="https://github.com/tpoorten/dotPlotly">https://github.com/tpoorten/dotPlotly</a></li>
<li>用R绘制mummer的结果图 <a target="_blank" rel="noopener" href="https://blog.csdn.net/u012110870/article/details/102492711?utm%5C_medium=distribute.pc%5C_relevant.none-task-blog-title-3&amp;spm=1001.2101.3001.4242">https://blog.csdn.net/u012110870/article/details/102492711?utm\_medium=distribute.pc\_relevant.none-task-blog-title-3&amp;spm=1001.2101.3001.4242</a></li>
</ol>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/05/Linux-shell-perl-mute-multiple-lines/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/05/Linux-shell-perl-mute-multiple-lines/" class="post-title-link" itemprop="url">shell脚本和perl脚本的段注释</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-05 12:53:15" itemprop="dateCreated datePublished" datetime="2020-12-05T12:53:15+08:00">2020-12-05</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
                  , 
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>shell脚本段注释</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">#!/bin/bash </span><br><span class="line">echo &quot;Say Something&quot; </span><br><span class="line">&lt;&lt;COMMENT     </span><br><span class="line">  your comment 1     </span><br><span class="line">  comment 2     </span><br><span class="line">  blah </span><br><span class="line">COMMENT </span><br><span class="line">echo &quot;Do something else&quot;</span><br></pre></td></tr></table></figure>
<p>perl脚本段注释。测试了一下，“=”符号和“pod”之间不能加空格，pod不能大写。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">=pod</span><br><span class="line">while(&lt;file&gt;)&#123;</span><br><span class="line">  chomp $_;</span><br><span class="line">  print &quot;$_\n;</span><br><span class="line">&#125;</span><br><span class="line">=cut</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/11/16/R-corrplot-coefficient-correlation-/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/11/16/R-corrplot-coefficient-correlation-/" class="post-title-link" itemprop="url">R：corrplot 相关系数矩阵</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-11-16 12:03:31" itemprop="dateCreated datePublished" datetime="2020-11-16T12:03:31+08:00">2020-11-16</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">install.packages(&quot;corrplot&quot;)</span><br><span class="line">library(&quot;corrplot&quot;)</span><br><span class="line">M = cor(mtcars)</span><br><span class="line">#corrplot(M, order = &quot;hclust&quot;, addrect = 2)</span><br><span class="line">#head(mtcars)</span><br><span class="line">#</span><br><span class="line">#corrplot(M, order = &quot;hclust&quot;, addrect = 2)</span><br><span class="line"># 只需要下半部分三角形</span><br><span class="line">corrplot(M, order = &quot;hclust&quot;, addrect = 2, type=&quot;lower&quot;)</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116095756.png"><img src="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116095756.png?w=894" alt="" /></a></p>
<p>参考资料：</p>
<p>[1] corrplot函数参数调整 <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/28076189">https://zhuanlan.zhihu.com/p/28076189</a></p>
<p>[2] 图形元素设定 <a target="_blank" rel="noopener" href="https://www.cnblogs.com/Mao1518202/p/11270767.html">https://www.cnblogs.com/Mao1518202/p/11270767.html</a></p>
<p>搜资料的时候看到与上述图形类似的点阵图（dot plot），还可以结合聚类分析，顺便记录一下。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"># install.packages(&#x27;tidyverse&#x27;)</span><br><span class="line">library(&#x27;tidyverse&#x27;)</span><br><span class="line"># -- Attaching packages ----------------------------------------------------# -------------------------------------------------------- tidyverse 1.3.0</span><br><span class="line"># √ tibble  3.0.4     √ stringr 1.4.0</span><br><span class="line"># √ readr   1.4.0     √ forcats 0.5.0</span><br><span class="line"># √ purrr   0.3.4     </span><br><span class="line"># -- Conflicts -------------------------------------------------------------#-------------------------------------------------- tidyverse_conflicts() --</span><br><span class="line"># x dplyr::filter() masks stats::filter()</span><br><span class="line"># x dplyr::lag()    masks stats::lag()</span><br><span class="line"></span><br><span class="line"># install.packages(&#x27;ggdendro&#x27;)</span><br><span class="line">library(ggdendro)</span><br><span class="line"></span><br><span class="line">library(cowplot)</span><br><span class="line"></span><br><span class="line"># BiocManager::install(&#x27;ggtree&#x27;)</span><br><span class="line">library(ggtree)</span><br><span class="line"></span><br><span class="line">library(patchwork) </span><br><span class="line"></span><br><span class="line">gene_cluster &lt;- read_tsv(&#x27;https://github.com/davemcg/davemcg.github.io/raw/master/content/post/scRNA_dotplot_data.tsv.gz&#x27;)</span><br><span class="line"></span><br><span class="line">gene_cluster %&gt;% sample_n(5)</span><br><span class="line"></span><br><span class="line">markers &lt;- gene_cluster$Gene %&gt;% unique()</span><br></pre></td></tr></table></figure>
<p>#    gene_cluster %&gt;% sample_n(5)</p>
<h1 id="gene_cluster-6列-888行-的矩阵下面显示前5行"><a class="markdownIt-Anchor" href="#gene_cluster-6列-888行-的矩阵下面显示前5行"></a> gene_cluster 6列 888行 的矩阵，下面显示前5行</h1>
<p>A tibble: 5 x 6<br />
Gene    cluster cell_ct cell_exp_ct   count Group</p>
<p>1 CELA1   c16        1278          87 0.0826  Neuron<br />
2 NDUFA1  c13        1725           5 0.00290 Neuron<br />
3 IGFBP7  c13        1725        1013 1.54    Neuron<br />
4 IGFBP7  c09        2124          29 0.0154  Astrocyte<br />
5 SLC10A7 c13        1725          59 0.0413  Neuron</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">   gene_cluster %&gt;% filter(Gene %in% markers) %&gt;% </span><br><span class="line">      mutate(% Expressing = (cell_exp_ct/cell_ct) * 100) %&gt;% </span><br><span class="line">   ggplot(aes(x=cluster, y = Gene, color = count, size = % Expressing)) + </span><br><span class="line">   geom_point()     </span><br><span class="line"># mutate函数来自dplyr，功能是Create, modify, and delete columns</span><br><span class="line"># mutate函数计算一个比例数值，作为矩阵第七列，名称为“% Expression”。</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/01.png"><img src="https://genehub.files.wordpress.com/2020/11/01.png?w=1024" alt="" /></a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 用dplyr包的filter功能，筛选条件：矩阵中“count”列中大于0，“% expression”列中大于1%。</span><br><span class="line">gene_cluster %&gt;% filter(Gene %in% markers) %&gt;%    </span><br><span class="line">mutate(`% Expressing` = (cell_exp_ct/cell_ct) * 100) %&gt;%    </span><br><span class="line">filter(count &gt; 0, `% Expressing` &gt; 1) %&gt;%    </span><br><span class="line">ggplot(aes(x=cluster, y = Gene, color = count, size = `% Expressing`)) +    geom_point() </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/02.png"><img src="https://genehub.files.wordpress.com/2020/11/02.png?w=1024" alt="" /></a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"># 图形调整与美化。使用“theme(axis.line  = element_blank())”取消显示xy坐标轴。“theme(axis.ticks = element_blank()) ”去掉坐标轴的刻度值。</span><br><span class="line"># cowplot包是ggplot2的简单附加组件。它旨在为ggplot2提供一个出版物就绪的主题，一个需要最小量的轴标签尺寸，情节背景等。它的主要目的是方便制作符合要求的图片。除了提供修改的绘图主题外，此包还提供了对ggplot2绘图的自定义注释的功能。</span><br><span class="line"></span><br><span class="line">gene_cluster %&gt;% filter(Gene %in% markers) %&gt;%    </span><br><span class="line">mutate(`% Expressing` = (cell_exp_ct/cell_ct) * 100) %&gt;%    </span><br><span class="line">filter(count &gt; 0, `% Expressing` &gt; 1) %&gt;%    </span><br><span class="line">ggplot(aes(x=cluster, y = Gene, color = count, size = `% Expressing`)) +    geom_point() +    </span><br><span class="line">scale_color_viridis_c(name = &#x27;log2 (count + 1)&#x27;) +    cowplot::theme_cowplot() +    </span><br><span class="line">theme(axis.line  = element_blank()) +   </span><br><span class="line">theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +   ylab(&#x27;&#x27;) +   </span><br><span class="line">theme(axis.ticks = element_blank()) </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/03.png"><img src="https://genehub.files.wordpress.com/2020/11/03.png?w=1024" alt="" /></a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">gene_cluster %&gt;% </span><br><span class="line">filter(Gene %in% markers) %&gt;%    </span><br><span class="line">mutate(`% Expressing` = (cell_exp_ct/cell_ct) * 100) %&gt;%    </span><br><span class="line">filter(count &gt; 0, `% Expressing` &gt; 1) %&gt;%    </span><br><span class="line">ggplot(aes(x=cluster, y = Gene, color = count, size = `% Expressing`)) +    geom_point() +    </span><br><span class="line">cowplot::theme_cowplot() +    </span><br><span class="line">theme(axis.line  = element_blank()) +   </span><br><span class="line">theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +   ylab(&#x27;&#x27;) +   </span><br><span class="line">theme(axis.ticks = element_blank()) +   </span><br><span class="line">scale_color_gradientn(colours = viridis::viridis(20), limits = c(0,4), oob = scales::squish, name = &#x27;log2 (count + 1)&#x27;)</span><br></pre></td></tr></table></figure>
<p>前面的图中，对所有圆点颜色进行观察，可见整体偏向深蓝，只有一个位于坐标c15 * KCNQ1 的圆点是黄色，其count数值4.94，比其他坐标的数值大了不少。通常绘制热图的时候也会因为这种值（甚至更大的极值，因为大部分基因的表达值FPKM通常小于100，但是少数基因的表达值会接近10000）导致画面颜色不够均匀。 参考原文的做法，调用 <code>scale_color_gradientn</code> 将 <code>c(0,4)</code> 数字范围的圆点进行着色，对于大于4的极值， 调用scales函数的squish功能将其“压碎”成为4（即“squished” down） <code>oob = scales::squish</code></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/05.png"><img src="https://genehub.files.wordpress.com/2020/11/05.png?w=1024" alt="" /></a></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/04.png"><img src="https://genehub.files.wordpress.com/2020/11/04.png?w=1024" alt="" /></a></p>
<p>进一步的，可以将基因进行聚类，并调用ggtree进行树形图绘制。先对原始数据进行整理，提取三列：Gene、cluster、count。调用pivot_wider功能，将其转换为行为gene，列为cluster的矩阵。（tidyr包的 <code>pivot_wider</code> 函数可以将长表变成宽表。 这适用于将多个变量保存到了一列的情况。—— 摘抄自<a target="_blank" rel="noopener" href="https://www.math.pku.edu.cn/teachers/lidf/docs/Rbook/html/%5C_Rbook/summary-manip.html%EF%BC%89">https://www.math.pku.edu.cn/teachers/lidf/docs/Rbook/html/\_Rbook/summary-manip.html）</a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"># make data square to calculate euclidean distance </span><br><span class="line">mat &lt;- gene_cluster %&gt;%    </span><br><span class="line">filter(Gene %in% markers) %&gt;%    </span><br><span class="line">select(-cell_ct, -cell_exp_ct, -Group) %&gt;%  # drop unused columns to faciliate widening   </span><br><span class="line">pivot_wider(names_from = cluster, values_from = count) %&gt;%    </span><br><span class="line">data.frame() # make df as tibbles -&gt; matrix annoying </span><br><span class="line"></span><br><span class="line">row.names(mat) &lt;- mat$Gene  # put gene in `row` </span><br><span class="line"></span><br><span class="line">mat &lt;- mat[,-1] #drop gene column as now in rows </span><br><span class="line"></span><br><span class="line">clust &lt;- hclust(dist(mat %&gt;% as.matrix())) # hclust with distance matrix</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/06.png"><img src="https://genehub.files.wordpress.com/2020/11/06.png?w=1024" alt="" /></a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">ddgram &lt;- as.dendrogram(clust) # create dendrogram </span><br><span class="line">ggtree_plot &lt;- ggtree::ggtree(ddgram) </span><br><span class="line">ggtree_plot</span><br><span class="line">ggtree::ggtree(ddgram, size=1) + geom_tiplab() </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/07-1.png"><img src="https://genehub.files.wordpress.com/2020/11/07-1.png?w=1024" alt="" /></a></p>
<p>（右侧的名称有问题）</p>
<p>将树形图和点状图拼接组合。下面的操作只是将两幅图拼起来，基因名并没有对应关系。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">dotplot &lt;- gene_cluster %&gt;% filter(Gene %in% markers) %&gt;% </span><br><span class="line">   mutate(% Expressing = (cell_exp_ct/cell_ct) * 100) %&gt;% </span><br><span class="line">   filter(count &gt; 0, % Expressing &gt; 1) %&gt;% </span><br><span class="line">   ggplot(aes(x=cluster, y = Gene, color = count, size = % Expressing)) + </span><br><span class="line">   geom_point() + </span><br><span class="line">   cowplot::theme_cowplot() + </span><br><span class="line">   theme(axis.line  = element_blank()) +</span><br><span class="line">   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +</span><br><span class="line">   ylab(&#x27;&#x27;) +</span><br><span class="line">   theme(axis.ticks = element_blank()) +</span><br><span class="line">   scale_color_gradientn(colours = viridis::viridis(20), limits = c(0,4), oob = scales::squish, name = &#x27;log2 (count + 1)&#x27;)   </span><br><span class="line"></span><br><span class="line"></span><br><span class="line">   plot_grid(ggtree_plot, dotplot, nrow = 1, rel_widths = c(0.5,2), align = &#x27;h&#x27;) </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/08.png"><img src="https://genehub.files.wordpress.com/2020/11/08.png?w=1024" alt="" /></a></p>
<p>使用</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">dotplot &lt;- gene_cluster %&gt;% filter(Gene %in% markers) %&gt;% </span><br><span class="line">mutate(% Expressing = (cell_exp_ct/cell_ct) * 100,</span><br><span class="line">   Gene = factor(Gene, levels = clust$labels[clust$order])) %&gt;% </span><br><span class="line">   #filter(count &gt; 0, % Expressing &gt; 1) %&gt;% </span><br><span class="line">   ggplot(aes(x=cluster, y = Gene, color = count, size = % Expressing)) + </span><br><span class="line">   geom_point() + </span><br><span class="line">   cowplot::theme_cowplot() + </span><br><span class="line">   theme(axis.line  = element_blank()) +</span><br><span class="line">   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +</span><br><span class="line">   ylab(&#x27;&#x27;) +</span><br><span class="line">   theme(axis.ticks = element_blank()) +</span><br><span class="line">   scale_color_gradientn(colours = viridis::viridis(20), limits = c(0,4), oob = scales::squish, name = &#x27;log2 (count + 1)&#x27;)   </span><br><span class="line">   plot_grid(ggtree_plot, NULL, dotplot, nrow = 1, rel_widths = c(0.5,-0.05, 2), align = &#x27;h&#x27;) </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/10.png"><img src="https://genehub.files.wordpress.com/2020/11/10.png?w=1024" alt="" /></a></p>
<p>代码摘抄自：<a target="_blank" rel="noopener" href="https://davemcg.github.io/post/lets-plot-scrna-dotplots/">https://davemcg.github.io/post/lets-plot-scrna-dotplots/</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/11/16/R-Empirical-cumulative-distribution/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/11/16/R-Empirical-cumulative-distribution/" class="post-title-link" itemprop="url">R：经验分布函数</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-11-16 09:45:08" itemprop="dateCreated datePublished" datetime="2020-11-16T09:45:08+08:00">2020-11-16</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-08 16:47:12" itemprop="dateModified" datetime="2020-12-08T16:47:12+08:00">2020-12-08</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>经验分布函数（Empirical cumulative distribution (ECDF)），是统计学中一个与样本<strong>经验</strong>测度有关的<strong>分布函数</strong>。 该累积<strong>分布函数</strong>是在所有n个数据点上都跳跃 1/n的阶跃<strong>函数</strong>。 对被测变量的某个值而言，该值的<strong>分布函数</strong>值表示所有观测样本中小于或等于该值的样本所占的比例。</p>
<p>在一个网站看到大量ggplot的画图案例，值得收藏：<a target="_blank" rel="noopener" href="https://www.datanovia.com/en/blog/ggplot-examples-best-reference/">https://www.datanovia.com/en/blog/ggplot-examples-best-reference/</a></p>
<p>以 R 内置数据集 iris 为例。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ggplot(iris, aes(Sepal.Length)) +</span><br><span class="line">stat_ecdf(aes(color = Species), size=3) +</span><br><span class="line">scale_color_viridis_d() </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116093707.png"><img src="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116093707.png?w=708" alt="" /></a></p>
<p>便于理解，用了一个简化的数据集画图。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">k=read.table(&quot;test.txt&quot;, head=T)</span><br><span class="line">ggplot(k, aes(Sepal.Length)) +</span><br><span class="line">   stat_ecdf(aes(color = Species), size=3) +</span><br><span class="line">   scale_color_viridis_d()    </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116094244.png"><img src="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116094244.png?w=697" alt="" /></a></p>
<p>计算公式</p>
<p><img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/68f10704c974630c3b15d009b884366d2bcbf685" alt="" /></p>
<p>可以理解为将数据按照从小到大排序，然后分布统计不同大小的数字出现的比例，（其实就是histgram？），那么横坐标就是数据从小到大的范围，纵坐标就是0~100%。图形相当于把histgram的柱形顶端和左侧画出来了。</p>
<p>附数据集test.txt：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br></pre></td><td class="code"><pre><span class="line">Sepal.Length    Species</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 1    versicolor</span><br><span class="line"> 2    versicolor</span><br><span class="line"> 3    versicolor</span><br><span class="line"> 4    versicolor</span><br><span class="line"> 5    versicolor</span><br><span class="line"> 6    versicolor</span><br><span class="line"> 7    versicolor</span><br><span class="line"> 8    versicolor</span><br><span class="line"> 9    versicolor</span><br><span class="line"> 10    versicolor</span><br><span class="line"> 11    versicolor</span><br><span class="line"> 12    versicolor</span><br><span class="line"> 13    versicolor</span><br><span class="line"> 14    versicolor</span><br><span class="line"> 15    versicolor</span><br><span class="line"> 16    versicolor</span><br><span class="line"> 17    versicolor</span><br><span class="line"> 18    versicolor</span><br><span class="line"> 19    versicolor</span><br><span class="line"> 20    versicolor</span><br><span class="line"> 1    virginica</span><br><span class="line"> 1    virginica</span><br><span class="line"> 2    virginica</span><br><span class="line"> 2    virginica</span><br><span class="line"> 3    virginica</span><br><span class="line"> 2    virginica</span><br><span class="line"> 4    virginica</span><br><span class="line"> 5    virginica</span><br><span class="line"> 4    virginica</span><br><span class="line"> 3    virginica</span><br><span class="line"> 2    virginica</span><br><span class="line"> 4    virginica</span><br><span class="line"> 5    virginica</span><br><span class="line"> 7    virginica</span><br><span class="line"> 6    virginica</span><br><span class="line"> 8    virginica</span><br><span class="line"> 6    virginica</span><br><span class="line"> 15    virginica</span><br><span class="line"> 16    virginica</span><br><span class="line"> 19    virginica</span><br><span class="line"> 13    kaka</span><br><span class="line"> 14    kaka</span><br><span class="line"> 16    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 17    kaka</span><br><span class="line"> 18    kaka</span><br><span class="line"> 19    kaka</span><br><span class="line"> 20    kaka</span><br><span class="line"> 13    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 12    kaka</span><br><span class="line"> 16    kaka</span><br><span class="line"> 18    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 13    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 2    kaka</span><br><span class="line"> 4    kaka</span><br><span class="line"> 5    kaka</span><br><span class="line"> 1    kaka</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/11/14/R-How-to-remove-rows-with-NA-values/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/11/14/R-How-to-remove-rows-with-NA-values/" class="post-title-link" itemprop="url">R：删除包含缺失值NA的行</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-11-14 18:08:01" itemprop="dateCreated datePublished" datetime="2020-11-14T18:08:01+08:00">2020-11-14</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-08 17:18:12" itemprop="dateModified" datetime="2020-12-08T17:18:12+08:00">2020-12-08</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p><strong>R数据框，如何删除包含NA的行？</strong></p>
<p>在进行统计建模时，通常需要用到的因变量和自变量都不包含缺失值。tidyr 包的 drop_na() 函数可以对数据框 指定一到多个变量，删去指定的变量有缺失值的行。不指定变量时有任何变量缺失的行都会被删去。</p>
<p>例如，将 NHANES 中所有存在缺失值的行删去后数出保留的行数，原来有 10000 行：</p>
<p>NHANES <strong>%&gt;%</strong> <strong>drop_na</strong>() <strong>%&gt;% nrow</strong>()</p>
<p>## [1] 0 可见所有行都有缺失值。</p>
<p>下面仅剔除 AlcoholDay 缺失的观测并计数：</p>
<p>NHANES <strong>%&gt;% drop_na</strong>(AlcoholDay) <strong>%&gt;% nrow</strong>()</p>
<p>## [1] 4914</p>
<p>基本 stats 包的 complete.cases 函数返回是否无缺失值的逻辑向量，na.omit 函数则返回无缺失值的观测的子集。</p>
<p><strong>如何删除全部为NA的行？只要有非缺省值，可保留对应行。</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line"> testmatrix &lt;- matrix(nrow=6, ncol=4)</span><br><span class="line">testmatrix</span><br><span class="line">[,1] [,2] [,3] [,4]</span><br><span class="line">[1,] NA NA NA NA</span><br><span class="line">[2,] NA NA NA NA</span><br><span class="line">[3,] NA NA NA NA</span><br><span class="line">[4,] NA NA NA NA</span><br><span class="line">[5,] NA NA NA NA</span><br><span class="line">[6,] NA NA NA NA</span><br><span class="line">testmatrix[2:5,2:3] &lt;- seq(2)</span><br><span class="line">testmatrix</span><br><span class="line">[,1] [,2] [,3] [,4]</span><br><span class="line">[1,] NA NA NA NA</span><br><span class="line">[2,] NA 1 1 NA</span><br><span class="line">[3,] NA 2 2 NA</span><br><span class="line">[4,] NA 1 1 NA</span><br><span class="line">[5,] NA 2 2 NA</span><br><span class="line">[6,] NA NA NA NA</span><br><span class="line">tm1&lt;-testmatrix[,-which(apply(testmatrix,2,function(x)all(is.na(x))))]</span><br><span class="line">tm1</span><br><span class="line">[,1] [,2]</span><br><span class="line">[1,] NA NA</span><br><span class="line">[2,] 1 1</span><br><span class="line">[3,] 2 2</span><br><span class="line">[4,] 1 1</span><br><span class="line">[5,] 2 2</span><br><span class="line">[6,] NA NA</span><br><span class="line">tm2&lt;-tm1[-which(apply(testmatrix,1,function(x)all(is.na(x)))),]</span><br><span class="line">tm2</span><br><span class="line">[,1] [,2]</span><br><span class="line">[1,] 1 1</span><br><span class="line">[2,] 2 2</span><br><span class="line">[3,] 1 1</span><br><span class="line">[4,] 2 2</span><br><span class="line">testmatrix[-which(apply(testmatrix,1,function(x)all(is.na(x)))), -which(apply(testmatrix,2,function(x)all(is.na(x))))]</span><br><span class="line">[,1] [,2]</span><br><span class="line">[1,] 1 1</span><br><span class="line">[2,] 2 2</span><br><span class="line">[3,] 1 1</span><br><span class="line">[4,] 2 2 </span><br></pre></td></tr></table></figure>
<p>Ref: <a target="_blank" rel="noopener" href="http://azaleasays.com/2008/02/06/r-remove-row-column-all-na-value/">http://azaleasays.com/2008/02/06/r-remove-row-column-all-na-value/</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/10/28/Linux-Run-windows-programs-inlinux-mono/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/10/28/Linux-Run-windows-programs-inlinux-mono/" class="post-title-link" itemprop="url">在Linux平台运行windows程序：mono</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-10-28 11:11:10" itemprop="dateCreated datePublished" datetime="2020-10-28T11:11:10+08:00">2020-10-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-20 02:11:00" itemprop="dateModified" datetime="2020-12-20T02:11:00+08:00">2020-12-20</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
                  , 
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>在Windows系统，最常见的就是Windows Forms程序（以下简称WinForm程序），也就是窗体应用程序，比如MS Office，媒体播放器软件等，WinForm程序比起Web应用程序，明显在实时性，用户亲和度和操作便捷能力方面强很多。在Linux系统下，虽然也有桌面窗体应用程序，但应用很少。能否将WinForm程序移植到Linux系统上面呢？目前，Novell公司的开源项目mono提供了将.NET程序运行在各种非Windows平台的能力，例如控制台程序，WinForm程序和ASP.NET等。(摘抄自：<a target="_blank" rel="noopener" href="https://blog.csdn.net/qq%5C_21153619/article/details/81459359">https://blog.csdn.net/qq\_21153619/article/details/81459359</a>)</p>
<p>如何在linux安装mono？我主要目的是在Linux下调用Windows程序做批量文件的处理。因为windows的cmd窗口操作真的不好用，相比之下Linux命令行就好用多了。</p>
<h3 id="1-安装依赖组件"><a class="markdownIt-Anchor" href="#1-安装依赖组件"></a> 1 安装依赖组件</h3>
<p>yum -y install gcc gcc-c++ bison pkgconfig glib2-devel gettext make libpng-devel libjpeg-devel libtiff-devel libexif-devel giflib-devel libX11-devel freetype-devel fontconfig-devel  cairo-devel</p>
<h3 id="2-安装libgdiplus"><a class="markdownIt-Anchor" href="#2-安装libgdiplus"></a> 2 安装libgdiplus</h3>
<ul>
<li><code>wget http://download.mono-project.com/sources/libgdiplus/libgdiplus-4.2.tar.gz</code></li>
<li><code>tar zxf libgdiplus-4.2.tar.gz</code></li>
<li><code>cd libgdiplus-4.2</code></li>
<li><code>./configure --prefix=/home/fenglei/local</code></li>
<li><code>make</code></li>
<li><code>make install</code></li>
</ul>
<h3 id="3-安装mono"><a class="markdownIt-Anchor" href="#3-安装mono"></a> 3 安装mono</h3>
<p>版本选择：<a target="_blank" rel="noopener" href="http://download.mono-project.com/sources/mono/">http://download.mono-project.com/sources/mono/</a></p>
<ul>
<li><code>wget http://download.mono-project.com/sources/mono/mono-4.6.0.125.tar.bz2</code></li>
<li><code>tar jxf mono-4.6.0.125.tar.bz2</code></li>
<li><code>cd mono-4.6.0</code></li>
<li><code>./configure --prefix=/home/fenglei/local</code></li>
<li><code>make -j2</code></li>
<li><code>make install</code></li>
</ul>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028105649.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028105649.png?w=1015" alt="" /></a></p>
<h3 id="4-软件测试我用了informedproteomics和rawtools软件做测试"><a class="markdownIt-Anchor" href="#4-软件测试我用了informedproteomics和rawtools软件做测试"></a> 4 软件测试，我用了InformedProteomics和rawTools软件做测试</h3>
<p>下载rawTools: <a target="_blank" rel="noopener" href="https://github.com/kevinkovalchik/RawTools/releases/download/2.0.2/RawTools-2.0.2.zip">https://github.com/kevinkovalchik/RawTools/releases/download/2.0.2/RawTools-2.0.2.zip</a></p>
<p>rawTools其他版本选择：<a target="_blank" rel="noopener" href="https://github.com/kevinkovalchik/RawTools/releases/tag/2.0.2">https://github.com/kevinkovalchik/RawTools/releases/tag/2.0.2</a></p>
<p><code>mkdir RawTools</code></p>
<p><code>unzip RawTools-2.0.2.zip -d RawTools</code></p>
<p><code>mono ~/local/app/RawTools/RawTools.exe</code></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110306.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110306.png?w=1024" alt="" /></a></p>
<p>下载InformedProteomics：<a target="_blank" rel="noopener" href="https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics/releases/download/v1.0.7569/InformedProteomics%5C_Program.zip">https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics/releases/download/v1.0.7569/InformedProteomics\_Program.zip</a></p>
<p>版本选择：<a target="_blank" rel="noopener" href="https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics/releases">https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics/releases</a></p>
<p><code>mkdir InformedProteomics</code></p>
<p><code>unzip InformedProteomics_Program.zip -d InformedProteomics</code></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110657.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110657.png?w=1024" alt="" /></a></p>
<p>mono ~/local/app/InformedProteomics/ProMex.exe</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110926.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110926.png?w=1024" alt="" /></a></p>
<p>mono ~/local/app/InformedProteomics/MSPathFinderT.exe</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110955.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110955.png?w=1024" alt="" /></a></p>
<h3 id="报错"><a class="markdownIt-Anchor" href="#报错"></a> 报错</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">mono ~/local/app/InformedProteomics/ProMex.exe -i xxx.raw -csv n -maxThreads 6                                                       </span><br><span class="line">************ ProMex version 1.0.7569 (September 21, 2020) ************</span><br><span class="line">InputPath    xxx.raw</span><br><span class="line">OutputPath   </span><br><span class="line">MinMass       2,000 Da</span><br><span class="line">MaxMass      50,000 Da</span><br><span class="line">MinCharge     1</span><br><span class="line">MaxCharge    60</span><br><span class="line">FeatureMap   True</span><br><span class="line">ScoreReport  False</span><br><span class="line">LikelihoodRatioThreshold -10</span><br><span class="line">MaxThreads   4</span><br><span class="line">Exception while processing: The type initializer for &#x27;InformedProteomics.Backend.MassSpecData.MassSpecDataReaderFactory&#x27; threw an exception.</span><br><span class="line">  at InformedProteomics.FeatureFinding.LcMsFeatureFinderLauncher.Run () [0x00011] in &lt;filename unknown&gt;:0 </span><br><span class="line">  at ProMex.Program.Main (System.String[] args) [0x000f1] in &lt;filename unknown&gt;:0 </span><br><span class="line">[MVID] 0604fddcd448487d905c1e71f6809d88 0</span><br><span class="line">[MVID] 072679dfee414eaab286f6a82270167b 1</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/10/15/Others-NAS-set-up/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/10/15/Others-NAS-set-up/" class="post-title-link" itemprop="url">NAS 安装与使用</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-10-15 11:38:06" itemprop="dateCreated datePublished" datetime="2020-10-15T11:38:06+08:00">2020-10-15</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Life/" itemprop="url" rel="index"><span itemprop="name">Life</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>我主要用NAS做数据存储和备份，一些生物实验的原始数据比较大。<br />
<strong>一 组装</strong><br />
NAS型号：群晖 DS918+，有四个硬盘卡槽，底部还有一个内存卡插槽。买了两块IronWolf的NAS专用硬盘，各8 TB，一起16 TB。网络上有大量教程参考，组装完毕就是下图。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/81yfruavezl._ac_sl1500_.jpg"><img src="https://genehub.files.wordpress.com/2020/10/81yfruavezl._ac_sl1500_.jpg?w=1024" alt="" /></a></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/synology-ds918-installing-hard-drive.jpg"><img src="https://genehub.files.wordpress.com/2020/10/synology-ds918-installing-hard-drive.jpg?w=800" alt="" /></a></p>
<p><strong>二 系统安装与配置</strong></p>
<p>在官网下载相应的系统安装文件 <a target="_blank" rel="noopener" href="https://www.synology.com/zh-tw/support/download/DS918+#firmware">https://www.synology.com/zh-tw/support/download/DS918+#firmware</a> 。安裝 DSM （DiskStation Manager）。</p>
<p>[安装过程：<a target="_blank" rel="noopener" href="https://ningselect.com/11397/40/%5C">https://ningselect.com/11397/40/\</a>]</p>
<p><strong>三 使用</strong></p>
<p>1 在任意位置使用web登入NAS服务器</p>
<p>2 在内网web浏览器输入IP即可登录使用</p>
<p>3 FTP连接。在FTP客户端（FileZilla）输入 sftp://admin@192.168.x.y 即可，后面不用加端口号。</p>
<p>4 将NAS挂载到本地网络的一个个人电脑下。这样相当于将NAS作为个人电脑的硬盘使用。先在NAS管理界面设置NFS，参考：<a target="_blank" rel="noopener" href="https://www.synology.com/zh-cn/knowledgebase/DSM/tutorial/File%5C_Sharing/How%5C_to%5C_access%5C_files%5C_on%5C_Synology%5C_NAS%5C_within%5C_the%5C_local%5C_network%5C_NFS%E3%80%82%E7%84%B6%E5%90%8E%E5%9C%A8%E4%B8%AA%E4%BA%BA%E7%94%B5%E8%84%91%EF%BC%88%E6%88%91%E7%94%A8%E7%9A%84windows">https://www.synology.com/zh-cn/knowledgebase/DSM/tutorial/File\_Sharing/How\_to\_access\_files\_on\_Synology\_NAS\_within\_the\_local\_network\_NFS。然后在个人电脑（我用的windows</a> 10）设置，将群晖NAS映射为网络驱动器。在“我的电脑”右键，选择“映射网络驱动器”，驱动器选“Y”（随便选都可，一般考虑Z-&gt;A的顺序），文件夹地址输入“\\192.168.x.x\share_name”（192.168.x.x是NAS的IP地址，视具体情况而定；share_name就是NAS里面盘阵的名称，视具体情况而定），点击“完成”并输入账户名和密码。依次映射了两个硬盘，我得到了下面的界面，可见“网络位置”下方出现了相应的磁盘，这样就像使用本机硬盘一样的，很方便。</p>
<p>参考 <a target="_blank" rel="noopener" href="https://jingyan.baidu.com/article/e75057f22d7901ebc81a8966.html">https://jingyan.baidu.com/article/e75057f22d7901ebc81a8966.html</a></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201015112922.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201015112922.png?w=702" alt="" /></a></p>
<p><strong>四 空间清理</strong></p>
<p>收到系统提醒说存储空间将满，删除了一部分文件，并清空回收站，但是空间还是没有释放。查询资料之后发现是 Synology Drive 这个应用的缘故，它会将已删除的文件保留备份，所以占用大量空间，将其卸载可以释放哪些空间。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201026162623.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201026162623.png?w=831" alt="" /></a></p>
<p>Use a terminal (eg. Putty, SecureCRT) to connect Synology through SSH.<br />
ie: 192.168.x.x port 22<br />
Login ID: admin; Password: admin password.<br />
Switch to ROOT by “sudo -i”, type a new password for root user. Then detect space usage by “du -h -d 1”</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201026160121.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201026160121.png?w=1024" alt="" /></a></p>
<p>进入 Synology Drive Sever，选择“版本资源管理器”，勾选“显示已删除文件”，可见到那些被删除的文件实际上还保存在硬盘里，所以占用大量空间。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/5/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/5/">5</a><span class="page-number current">6</span><a class="page-number" href="/page/7/">7</a><span class="space">&hellip;</span><a class="page-number" href="/page/18/">18</a><a class="extend next" rel="next" href="/page/7/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">179</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
