<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2018/03/28/mitogenome-unfinished/">mitogenome (unfinished)</a></h1><div class="post-meta">2018-03-28</div><div class="post-content"><p>  tRNA tRNA sequences  整理成 fasta 格式 通过 blastn mapping  到 mitogenome，输出 format=6。 编程序转换成 features table 格式的文件。 注意： 1）部分 tRNA 有多个 copies，例如 trnC(gca) 与 trnfM(cat) 各有两个copies； 2）部分 tRNA 具有同样的 anticodon，但是其名称和序列不同，例如 trnI(cat) 和 trnM(cat)； 3）tRNA都没有内含子，所以可以较为简单的转换格式； 4）anticodon 如何自动注释？</p></div><p class="readmore"><a href="/2018/03/28/mitogenome-unfinished/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2018/02/08/R%EF%BC%9A%E8%B6%85%E5%87%A0%E4%BD%95%E5%88%86%E5%B8%83%E4%B8%8E%E5%9F%BA%E5%9B%A0%E5%8A%9F%E8%83%BD%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90-/">R：超几何分布与基因功能富集分析</a></h1><div class="post-meta">2018-02-08</div><div class="post-content"><p>GO功能显著性富集分析给出与基因组背景相比，在差异表达基因中显著富集的GO功能条目，从而给出差异表达基因与哪些生物学功能显著相关。该分析首先把所有差异表达基因向Gene Ontology数据库（<a target="_blank" rel="noopener" href="http://www.geneontology.org/">http://www.geneontology.org/</a>）的各个term映射，计算每个term的基因数目，然后应用超几何检验，找出与整个基因组背景相比，在差异表达基因中显著富集的GO条目，其计算公式为 <img src="https://genehub.files.wordpress.com/2018/02/hypergeometric.jpg" alt="hypergeometric"> 其中，N为所有基因中具有GO注释的基因数目；n为N中差异表达基因的数目；M为所有基因中注释为某特定GO term的基因数目；m为注释为某特定GO term的差异表达基因数目。计算得到的p-value通过Bonferroni校正之后，以corrected p-value≤0.05为阈值，满足此条件的GO term定义为在差异表达基因中显著富集的GO term。通过GO功能显著性富集分析能确定差异表达基因行使的主要生物学功能。 例如，如下的结果： <img src="https://genehub.files.wordpress.com/2018/02/qqe688aae59bbe20180208115638.jpg" alt="QQ截图20180208115638"> P值计算方法为（在R语言程序中调用phyper命令）：</p></div><p class="readmore"><a href="/2018/02/08/R%EF%BC%9A%E8%B6%85%E5%87%A0%E4%BD%95%E5%88%86%E5%B8%83%E4%B8%8E%E5%9F%BA%E5%9B%A0%E5%8A%9F%E8%83%BD%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90-/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2018/01/09/%E5%AE%89%E8%A3%85vcftools/">安装vcftools</a></h1><div class="post-meta">2018-01-09</div><div class="post-content"><p>configure的过程报错：Zlib找不到。 实际上已经安装zlib。 解决办法</p>
<p>export ZLIB_LIBS=’-L/home/fenglei/local/lib -lz’<br>export ZLIB_CFLAGS=-I/home/fenglei/local/include<br>echo $ZLIB_LIBS $ZLIB_CFLAGS<br>./configure –prefix=/home/fenglei/local/<br>make<br>make install</p></div><p class="readmore"><a href="/2018/01/09/%E5%AE%89%E8%A3%85vcftools/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2018/01/02/gene-ontology-analysis-for-degs-of-arabidopsis/">Gene Ontology analysis for DEGs of Arabidopsis</a></h1><div class="post-meta">2018-01-02</div><div class="post-content"><blockquote>
<p>try http:// if https:// URLs are not supported<br> source(“<a target="_blank" rel="noopener" href="https://bioconductor.org/biocLite.R&quot;">https://bioconductor.org/biocLite.R&quot;</a>)<br> biocLite(“clusterProfiler”)<br> biocLite(“DOSE”)<br> biocLite(“tibble”)<br> library(clusterProfiler)<br> biocLite(“topGO”)<br> library(topGO)<br> biocLite(“org.At.tair.db”)<br> library(org.At.tair.db)</p></div><p class="readmore"><a href="/2018/01/02/gene-ontology-analysis-for-degs-of-arabidopsis/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2017/09/20/%E6%A0%B9%E6%8D%AE%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AF%B9contigs%E8%BF%9B%E8%A1%8C%E6%8E%92%E5%88%97/">根据参考基因组对contigs进行排列</a></h1><div class="post-meta">2017-09-20</div><div class="post-content"><p>Multiple programs have been developed for reference-assisted chromosome assembly: Bambus [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR10">10</a>], BACCardI [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR11">11</a>], Projector2 [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR12">12</a>], OSLay [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR13">13</a>], ABACAS [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR14">14</a>], MeDuSa [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR15">15</a>], AlignGraph [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR16">16</a>], Ragout [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR17">17</a>], SyMap [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR18">18</a>] and RACA [<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6#CR19">19</a>]. Most of the listed tools were designed for bacterial or small genomes. For example, ABACAS is a convenient bacterial genome contiguation tool that may also be used for small eukaryotic genomes such as <em>Saccharomyces cerevisiae</em> (12.1 mega base pairs). However, ABACAS is not efficiently scaled to use with the large genomes typical of vertebrate species. ABACAS对小型基因组很实用。但是注意：contig连接处是由99个N连接起来的，实际上contig的末端室友重叠关系的。 针对大型基因组，可以使用Chromosomer: a reference-based genome arrangement tool for producing draft chromosome sequences</p></div><p class="readmore"><a href="/2017/09/20/%E6%A0%B9%E6%8D%AE%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%AF%B9contigs%E8%BF%9B%E8%A1%8C%E6%8E%92%E5%88%97/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2017/09/20/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%BA%8F%E5%88%97%E7%9A%84%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90/">基因组序列的共线性分析</a></h1><div class="post-meta">2017-09-20</div><div class="post-content"><h3 id="1-MUMMER"><a href="#1-MUMMER" class="headerlink" title="1 MUMMER"></a>1 MUMMER</h3><p>基因组序列直接比较的首选工具。参考资料很多。</p>
<h3 id="2-LASTZ"><a href="#2-LASTZ" class="headerlink" title="2 LASTZ"></a>2 LASTZ</h3><p><img src="https://genehub.files.wordpress.com/2017/09/13742_2016_141_fig5_html.gif" alt="13742_2016_141_fig5_html"><br>Ref：<a target="_blank" rel="noopener" href="http://www.bx.psu.edu/miller/_lab/dist/README.lastz-1.02.00/README.lastz-1.02.00a.html">http://www.bx.psu.edu/miller\_lab/dist/README.lastz-1.02.00/README.lastz-1.02.00a.html</a><br>Paper：<a target="_blank" rel="noopener" href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6">https://gigascience.biomedcentral.com/articles/10.1186/s13742-016-0141-6</a> </p></div><p class="readmore"><a href="/2017/09/20/%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%BA%8F%E5%88%97%E7%9A%84%E5%85%B1%E7%BA%BF%E6%80%A7%E5%88%86%E6%9E%90/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2017/08/28/sspace-cant-locate-getopts-pl/">SSPACE: Can't locate getopts.pl</a></h1><div class="post-meta">2017-08-28</div><div class="post-content"><p>问题：运行SSPACE_Standard_v3.0.pl报错。 Can’t locate getopts.pl in @INC (@INC contains: /xxx/SSPACE-STANDARD/dotlib/ /home/fenglei-cuhk/perl5/lib/perl5 /etc/perl /usr/local/lib/perl/5.18.2 /usr/local/share/perl/5.18.2 /usr/lib/perl5 /usr/share/perl5 /usr/lib/perl/5.18 /usr/share/perl/5.18 /usr/local/lib/site_perl .) at /home/xx/SSPACE-STANDARD/SSPACE_Standard_v3.0.pl line 124. 原因： getopts.pl is a Perl 4 core library but no longer included in current Perl 5 distributions. 解决办法： 在Linux命令行模式下输入“perl -MCPAN -e shell”进入perl界面，然后输入“install Perl4::CoreLibs”即可。</p></div><p class="readmore"><a href="/2017/08/28/sspace-cant-locate-getopts-pl/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2017/08/23/shell%E7%BC%96%E7%A8%8B%EF%BC%9A%E8%8E%B7%E5%8F%96%E6%96%87%E4%BB%B6%E5%89%8D%E7%BC%80%E5%AD%97%E7%AC%A6%E4%B8%B2/">Shell编程：获取文件前缀字符串</a></h1><div class="post-meta">2017-08-23</div><div class="post-content"><p>比如处理名为 /xxx/yyy/zzz/abc2017.1.fq.gz 的文件，我们需要获取abc2017这个文件名。</p>
<p>i=/xxx/yyy/zzz/abc2017.1.fq.gz</p>
<p>IN=$i      </p>
<h1 id="read1文件绝对路径：-xxx-yyy-zzz-abc2017-1-fq-gz"><a href="#read1文件绝对路径：-xxx-yyy-zzz-abc2017-1-fq-gz" class="headerlink" title="read1文件绝对路径：/xxx/yyy/zzz/abc2017.1.fq.gz"></a>read1文件绝对路径：/xxx/yyy/zzz/abc2017.1.fq.gz</h1><p>path=${IN%/*}        </p></div><p class="readmore"><a href="/2017/08/23/shell%E7%BC%96%E7%A8%8B%EF%BC%9A%E8%8E%B7%E5%8F%96%E6%96%87%E4%BB%B6%E5%89%8D%E7%BC%80%E5%AD%97%E7%AC%A6%E4%B8%B2/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2017/08/23/shellcheck-%E8%87%AA%E5%8A%A8%E6%A3%80%E6%B5%8Bshell%E8%84%9A%E6%9C%AC%E7%9A%84%E8%AF%AD%E6%B3%95/">shellcheck: 自动检测shell脚本的语法</a></h1><div class="post-meta">2017-08-23</div><div class="post-content"><p>生物信息分析工作中，可以将数据操作命令写入shell脚本，一方面方便自动化运行，另一方面是将操作记录存档，便于日后查阅。 但是shell脚本也会因人为疏忽而存在错误，我想有没有程序可以检测脚本的正确性呢，一查果然有shellcheck之类的程序。shellcheck有在线版：<a target="_blank" rel="noopener" href="http://www.shellcheck.net/">http://www.shellcheck.net/</a> shellcheck也有本地版，安装有多种方法，可参考官方网站：<a target="_blank" rel="noopener" href="https://github.com/koalaman/shellcheck">https://github.com/koalaman/shellcheck</a> 我直接下载了编译好的程序：<a target="_blank" rel="noopener" href="https://storage.googleapis.com/shellcheck/shellcheck-latest.linux.x86/_64.tar.xz">https://storage.googleapis.com/shellcheck/shellcheck-latest.linux.x86\_64.tar.xz</a> 解压之后，即可使用。 使用方法：shellcheck myscript.sh</p></div><p class="readmore"><a href="/2017/08/23/shellcheck-%E8%87%AA%E5%8A%A8%E6%A3%80%E6%B5%8Bshell%E8%84%9A%E6%9C%AC%E7%9A%84%E8%AF%AD%E6%B3%95/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2017/08/15/shell%E8%84%9A%E6%9C%AC-set-illegal-option-o-pipefail/">Shell脚本 \&quot;set Illegal option -o pipefail\&quot;</a></h1><div class="post-meta">2017-08-15</div><div class="post-content"><p>在Linux编写Shell 脚本程序的时候，set -o pipefail是很实用的命令。因为Shell脚本中通常有一连串的动作需要执行，一旦某个环节出错，如果设定了pipefail，脚本会立即中断运行；而如果没有设定，脚本将继续执行后续命令，可能造成错误。 然而在Ubuntu中，当使用“set -o pipefail”的时候，一直报错：</p>
<p>set Illegal option -o pipefail</p>
<p>经查，原因是Ubuntu的 shell 默认安装的是 dash，而不是 bash。 运行以下命令查看 sh 的详细信息，确认 shell 对应的程序是哪个： $ls -al /bin/sh dash 比 bash 更轻，更快。但 bash 却更常用。 如果一些命令、脚本等总不能正常执行，有可能是 dash 的原因。 比如编译 Android 源代码的时候，如果使用 dash，则有可能编译出错，或者编译的系统不能启动。 通过以下方式可以使 shell 切换回 bash： $sudo dpkg-reconfigure dash 然后选择 no 或者 否 ，并确认。 这样做将重新配置 dash，并使其不作为默认的 shell 工具。 也可以直接修改 /bin/sh 链接文件，将其指定到 /bin/bash： $sudo ln -fs /bin/bash /bin/sh</p></div><p class="readmore"><a href="/2017/08/15/shell%E8%84%9A%E6%9C%AC-set-illegal-option-o-pipefail/">Read More</a></p></div><nav class="page-navigator"><a class="extend prev" rel="prev" href="/page/5/">Previous</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/4/">4</a><a class="page-number" href="/page/5/">5</a><span class="page-number current">6</span><a class="page-number" href="/page/7/">7</a><a class="page-number" href="/page/8/">8</a><span class="space">&hellip;</span><a class="page-number" href="/page/15/">15</a><a class="extend next" rel="next" href="/page/7/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-What-the-differences-between-DDA-and-DIA-in-mass-spectrometer/">Proteomics | 质谱分析中DDA与DIA有什么不同？</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-Protein-Persulfidation/">Proteomics | 蛋白质硫巯基化修饰</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/Proteomics-Common-reagents-and-their-functions-in-LC-MS-MS-proteomics/">Proteomics | Common reagents and their functions in LC-MS/MS proteomics</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/R-Using-ggrepel-to-repel-overlapping-text-labels-in-ggplot2/">R | Using ggrepel to repel overlapping text labels in ggplot2</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/08/Bioinfo-Installation-of-BUSCO-v4/">Bioinfo | Installation of BUSCO v4</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/05/Linux-Installation-of-Tomcat-9/">Linux | Installation of Tomcat 9, Java 8, and MySQL 5.7.20</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/04/Remove-Background-from-Image/">Remove Background from Image</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/01/%E5%9C%A8Linux%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E4%BD%BF%E7%94%A8Nginx%E9%83%A8%E7%BD%B2Hexo%E7%AB%99%E7%82%B9/">在Linux服务器上使用Nginx部署Hexo站点</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/29/Linux%EF%BC%9A%E4%BD%BF%E7%94%A8Modules%E7%AE%A1%E7%90%86%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/">Linux：使用Environment Modules管理环境变量</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E4%BD%BF%E7%94%A8Snakemake%E6%90%AD%E5%BB%BA%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/">使用Snakemake搭建生物信息分析流程</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>