<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta property="og:type" content="website">
<meta property="og:title" content="FENG&#39;s Notes">
<meta property="og:url" content="http://example.com/page/6/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/page/6/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/28/Bioinfo-Snakemake-for-bioinfomatics-pipeline/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/28/Bioinfo-Snakemake-for-bioinfomatics-pipeline/" class="post-title-link" itemprop="url">使用Snakemake搭建生物信息分析流程</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-28 18:08:56" itemprop="dateCreated datePublished" datetime="2020-12-28T18:08:56+08:00">2020-12-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-29 17:32:28" itemprop="dateModified" datetime="2020-12-29T17:32:28+08:00">2020-12-29</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>snakemake 简介：Snakemake is a workflow engine that provides a readable Python-based workflow definition language and a powerful execution environment that scales from single-core workstations to compute clusters without modifying the workflow. It is the first system to support the use of automatically inferred multiple named wildcards (or variables) in input and output filenames. (Johannes Köster, Sven Rahmann, 2012, Bioinformatics 28, 19:2520–2522)</p>
<h3 id="snakemake-安装"><a class="markdownIt-Anchor" href="#snakemake-安装"></a> snakemake 安装</h3>
<p>安装snakemake的方法有多种，snakemake官方推荐的是conda，安装方法如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -c bioconda snakemake</span><br></pre></td></tr></table></figure>
<p>snakemake对python版本有要求，我用python3.7就不能安装，遇到下面的错误。使用 <code>conda create --name python36 python=3.6</code> 新安装了一个python3.6。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">UnsatisfiableError: The following specifications were found</span><br><span class="line">to be incompatible with the existing python installation in your environment:</span><br><span class="line"></span><br><span class="line">Specifications:</span><br><span class="line">  - snakemake -&gt; python[version=&#x27;3.4.*|3.5.*|3.6.*|&gt;=3.5,&lt;3.6.0a0|&gt;=3.6,&lt;3.7.0a0&#x27;]</span><br><span class="line"></span><br><span class="line">Your python: python=3.7</span><br></pre></td></tr></table></figure>
<h3 id="一个简单的snakemake脚本"><a class="markdownIt-Anchor" href="#一个简单的snakemake脚本"></a> 一个简单的snakemake脚本</h3>
<p>虽然snakemake广泛的应用于生物信息方面的流程编写，但是snakemake的应用并不局限于编写生物信息学的流程，这里以一个简单的合并文件的例子开始介绍snakemake的简单使用。</p>
<p>首先我们建立两个文件</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ echo &quot;Here is hello.&quot; &gt; hello.txt</span><br><span class="line">$ echo &quot;Here is world.&quot; &gt; world.txt</span><br></pre></td></tr></table></figure>
<p>接下来开始编写 <code>Snakefile</code></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">rule concat:  # 这里的rule可视为snakemake定义的关键字，concat使我们自定义的这一步任务的名称</span><br><span class="line">    input:   # input同样是snakemake的关键字，定义了在这个任务中的输入文件 </span><br><span class="line">        expand(&quot;&#123;file&#125;.txt&quot;, file=[&quot;hello&quot;, &quot;world&quot;]) #expand是一个snakemake定义的替换命令</span><br><span class="line">    output:   # output也是snakemake的关键字，定义输出结果的保存文件</span><br><span class="line">        &quot;merged.txt&quot;</span><br><span class="line">    shell:  # 这里表示我们下面的命令将在命令行中执行</span><br><span class="line">        &quot;cat &#123;input&#125; &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure>
<p>最后就可以在Snakefile的路径执行snakemake命令即可</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">(python36) [fenglei@localhost test]$ snakemake</span><br><span class="line">Provided cores: 1</span><br><span class="line">Rules claiming more threads will be scaled down.</span><br><span class="line">Job counts:</span><br><span class="line">        count   jobs</span><br><span class="line">        1       concat</span><br><span class="line">        1</span><br><span class="line"></span><br><span class="line">rule concat:</span><br><span class="line">    input: hello.txt, world.txt</span><br><span class="line">    output: merged.txt</span><br><span class="line">    jobid: 0</span><br><span class="line"></span><br><span class="line">Finished job 0.</span><br><span class="line">1 of 1 steps (100%) done</span><br><span class="line"></span><br><span class="line">$ cat merge.txt</span><br><span class="line">Here is hello.</span><br><span class="line">Here is world.</span><br></pre></td></tr></table></figure>
<p>在上面的Snakefile脚本中，rule、input、output、shell、expand均为snakemake中的关键字或者命令。同时Snakefile中的每一个rule其实都可以看作是一个简单的shell脚本，通过Snakefile将多个rule组织在一起并按照我们定义的顺序来执行。另外，在output中的结果文件可以是未存在目录中的文件,这时会自动创建不存在的目录。</p>
<h3 id="snakemake中的一些命令与规则"><a class="markdownIt-Anchor" href="#snakemake中的一些命令与规则"></a> snakemake中的一些命令与规则</h3>
<p><em>1、rule</em><br />
rule是Snakefile中最主要的部分。如上面的例子所说，每一个rule定义了一系列pipe中的一步，每一个rule都可以当作一个shell脚本来处理，一般主要包括input、output、shell3个部分。同时还有许多上面没有列出来的用法：</p>
<p>rule all。不同于其他的rule，在rule all里面一般不会去定义要执行的命令，他一般用来定义最后的输出结果文件。除了rule all中定义的文件外最后输出结果不会保存任何中间文件。例如将上面的脚本改成如下文件则没有输出结果：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">rule all:</span><br><span class="line">    input:</span><br><span class="line">         #&quot;merged.txt&quot;  取消注释后，则能正常输出文件</span><br><span class="line">rule concat:</span><br><span class="line">    input:</span><br><span class="line">        expand(&quot;&#123;file&#125;.txt&quot;, file=[&quot;hello&quot;, &quot;world&quot;])</span><br><span class="line">    output:</span><br><span class="line">        &quot;merge.txt&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;cat &#123;input&#125; &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure>
<p>wildcards。用来获取通配符匹配到的部分，例如对于通配符&quot;{dataset}/file.{group}.txt&quot;匹配到文件101/file.A.txt，则{wildcards.dataset}就是101，{wildcards.group}就是A。<br />
threads。通过在rule里面指定threads参数来指定分配给程序的线程数，egthreads: 8。<br />
resources。可用来指定程序运行的内存，eg. resources: mem_mb=800。<br />
message。使用message参数可以指定每运行到一个rule时，在终端中给出提示信息，eg.message: “starting mapping …”。<br />
priority。可用来指定程序运行的优先级，默认为0，eg.priority: 20。<br />
log。用来指定生成的日志文件，eg.log: “logs/concat.log”。<br />
params。指定程序运行的参数，eg.params: cat=“-n”,调用方法为{params.cat}。<br />
run。在run的缩进区域里面可以输入并执行python代码。<br />
scripts。用来执行指定脚本，eg.scripts: “rm_dup.py”<br />
temp。通过temp方法可以在所有rule运行完后删除指定的中间文件，eg.output: temp(“f1.bam”)。<br />
protected。用来指定某些中间文件是需要保留的，eg.output: protected(“f1.bam”)。<br />
ancient。重复运行执行某个Snakefile时，snakemake会通过比较输入文件的时间戳是否更改(比原来的新)来决定是否重新执行程序生成文件，使用ancient方法可以强制使得结果文件一旦生成就不会再次重新生成覆盖，即便输入文件时间戳已经更新，eg.input: ancient(“f1.fastq”)。<br />
Rule Dependencies。可通过快捷方式指定前一个rule的输出文件为此rule的输入文件：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">rule a:</span><br><span class="line">    input:  &quot;path/to/input&quot;</span><br><span class="line">    output: &quot;path/to/output&quot;</span><br><span class="line">    shell:  ...</span><br><span class="line"></span><br><span class="line">rule b:</span><br><span class="line">    input:  rules.a.output   #直接通过rules.a.output 指定rule a的输出</span><br><span class="line">    output: &quot;path/to/output/of/b&quot;</span><br><span class="line">    shell:  ...</span><br></pre></td></tr></table></figure>
<p>report。使用snakemake定义的report函数可以方便的将结果嵌入到一个HTML文件中进行查看。</p>
<p><em>2. configuration</em></p>
<p>每计算一次数据都要重写一次Snakefile有时可能会显得有些繁琐，我们可以将那些改动写入配置文件，使用相同流程计算时，将输入文件的文件名写入配置文件然后通过Snakefile读入即可。<br />
配置文件有两种书写格式——json和yaml。在Snakefile中读入配置文件使用如下方式：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">configfile: &quot;path/to/config.json&quot; </span><br><span class="line">configfile: &quot;path/to/config.yaml&quot;</span><br><span class="line"></span><br><span class="line"># 也可直接在执行snakemake命令时指定配置</span><br><span class="line">$ snakemake --config yourparam=1.5</span><br></pre></td></tr></table></figure>
<p>在shell命令中直接调用config文件中的内容的话，不需要引号，如config[a]而不是config[“a”]。</p>
<h3 id="snakemake-的执行"><a class="markdownIt-Anchor" href="#snakemake-的执行"></a> snakemake 的执行</h3>
<p>一般讲所有的参数配置写入Snakefile后直接在Snakefile所在路径执行snakemake命令即可开始执行流程任务。一些常用的参数：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">--snakefile, -s 指定Snakefile，否则是当前目录下的Snakefile</span><br><span class="line">--dryrun, -n  不真正执行，一般用来查看Snakefile是否有错</span><br><span class="line">--printshellcmds, -p   输出要执行的shell命令</span><br><span class="line">--reason, -r  输出每条rule执行的原因,默认FALSE</span><br><span class="line">--cores, --jobs, -j  指定运行的核数，若不指定，则使用最大的核数</span><br><span class="line">--force, -f 重新运行第一条rule或指定的rule</span><br><span class="line">--forceall, -F 重新运行所有的rule，不管是否已经有输出结果</span><br><span class="line">--forcerun, -R 重新执行Snakefile，当更新了rule时候使用此命令</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">#一些可视化命令</span><br><span class="line">$ snakemake --dag | dot -Tpdf &gt; dag.pdf</span><br><span class="line"></span><br><span class="line">#集群投递</span><br><span class="line">snakemake --cluster &quot;qsub -V -cwd -q 节点队列&quot; -j 10</span><br><span class="line"># --cluster /-c CMD: 集群运行指令</span><br><span class="line"># qusb -V -cwd -q， 表示输出当前环境变量(-V),在当前目录下运行(-cwd), 投递到指定的队列(-q), 如果不指定则使用任何可用队列</span><br><span class="line"># --local-cores N: 在每个集群中最多并行N核</span><br><span class="line"># --cluster-config/-u FILE: 集群配置文件</span><br></pre></td></tr></table></figure>
<h2 id="u"><a class="markdownIt-Anchor" href="#u"></a> <a target="_blank" rel="noopener" href="https://vidotto.top/post/%E4%BD%BF%E7%94%A8-snakemake-%E7%BC%96%E5%86%99-rna-seq-%E6%B5%81%E7%A8%8B/"><u>使用 snakemake 编写 RNA seq 流程</u></a></h2>
<p>学习生信两年多了，自己也积累了一些测序数据分析流程，后来数据越来越多就有把流程自动化以方便分析工作。开始是将流程写成 shell 脚本，但是有时候还需要根据不同的数据去更改一些步骤的参数，所以每次用还是要检查一下脚本。后来发现 snakemake 流程管理工具，它完全能够满足我目前的需求。</p>
<p>我以搭建转录组流程为例子介绍如何使用 snakemake 搭建适合你的流程。</p>
<h3 id="创建一个单独的文件夹存放测序数据"><a class="markdownIt-Anchor" href="#创建一个单独的文件夹存放测序数据"></a> 创建一个单独的文件夹存放测序数据</h3>
<p>创建名为 input 的文件夹存放转录组测序数据，我使用的数据如下：</p>
<p>s1-1_RRL18212-V_1.fq.gz——Control(forward)<br />
s1-1_RRL18212-V_2.fq.gz——Control(reverse)<br />
s1-2_RRL18214-V_1.fq.gz——Treated(forward)<br />
s1-2_RRL18214-V_2.fq.gz——Treated(reverse)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">input</span><br><span class="line">├── s1-1_RRL18212-V_1.fq.gz</span><br><span class="line">├── s1-1_RRL18212-V_2.fq.gz</span><br><span class="line">├── s1-2_RRL18214-V_1.fq.gz</span><br><span class="line">└── s1-2_RRL18214-V_2.fq.gz</span><br></pre></td></tr></table></figure>
<h3 id="为-star-比对步骤准备-index"><a class="markdownIt-Anchor" href="#为-star-比对步骤准备-index"></a> 为 STAR 比对步骤准备 index</h3>
<p>我们需要把测序数据比对到小鼠参考基因组上，需要先下载小鼠（GRCm38.p6）参考基因组文件并使用 STAR 的 genomeGenerate 创建 index 。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">mkdir genome</span><br><span class="line">wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M20/GRCm38.p6.genome.fa.gz</span><br><span class="line">wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M20/gencode.vM20.annotation.gtf.gz</span><br><span class="line">pigz -d GRCm38.p6.genome.fa.gz gencode.vM20.annotation.gtf.gz</span><br><span class="line"></span><br><span class="line">STAR \</span><br><span class="line">--runMode genomeGenerate \</span><br><span class="line">--genomeDir star_index \</span><br><span class="line">--genomeFastaFiles genome/GRCm38.p6.genome.fa \</span><br><span class="line">--sjdbGTFfile annotation/gencode.vM20.annotation.gtf \</span><br><span class="line">--runThreadN 4</span><br></pre></td></tr></table></figure>
<h3 id="编写-snakefile"><a class="markdownIt-Anchor" href="#编写-snakefile"></a> 编写 snakefile</h3>
<p>创建一个文件名为 snakefile 的文件，我们的流程主要编写在这个文件里。</p>
<p>我们先看一个基础 snakefile 是什么样子的：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">rule bwa_map:</span><br><span class="line">    input:</span><br><span class="line">        &quot;data/genome.fa&quot;,</span><br><span class="line">        &quot;data/samples/A.fastq&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;mapped_reads/A.bam&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;bwa mem &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure>
<p>每一个 rule 表示不同的任务，需要通过 input 和 output 指定任务需要的输入文件和会产生的输出文件，run 就指明该任务的将要运行的命令行。</p>
<h3 id="第一个任务fastqc"><a class="markdownIt-Anchor" href="#第一个任务fastqc"></a> 第一个任务：fastqc</h3>
<p>第一个任务我们检测测序数据质量，按照 snakemake 规则如下写法：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">rule fastqc:</span><br><span class="line">    input:</span><br><span class="line">        &quot;input/&#123;sample&#125;_1.fq.gz&quot;,</span><br><span class="line">        &quot;input/&#123;sample&#125;_2.fq.gz&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;results/1_initial_qc/&#123;sample&#125;/&#123;sample&#125;_1_fastqc.zip&quot;,</span><br><span class="line">        &quot;results/1_initial_qc/&#123;sample&#125;/&#123;sample&#125;_2_fastqc.zip&quot;,</span><br><span class="line">        &quot;results/1_initial_qc/&#123;sample&#125;/&#123;sample&#125;_1_fastqc.html&quot;,</span><br><span class="line">        &quot;results/1_initial_qc/&#123;sample&#125;/&#123;sample&#125;_2_fastqc.html&quot;</span><br><span class="line">    log:</span><br><span class="line">        &quot;log/&#123;sample&#125;_fastqc&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;&quot;&quot;</span><br><span class="line">        fastqc -o results/1_initial_qc/&#123;wildcards.sample&#125; --noextract &#123;input[0]&#125; &#123;input[1]&#125; 2&gt; &#123;log&#125;</span><br><span class="line">        &quot;&quot;&quot;</span><br></pre></td></tr></table></figure>
<p>因为是双端测序数据，我每次运行将一个样品的两端数据同时检测。可以发现在 input 和 output 里我都使用了 {sample} 字样，这是snakemake 中的通配方式，比如 input/{sample}_1.fq.gz 就可以通配到：s1-1_RRL18212-V_1.fq.gz 和 s1-2_RRL18214-V_1.fq.gz 两个文件，这一点是 snakemake 的精髓，这样我们一行命令就能把所以样品都跑完流程。</p>
<h3 id="第二个任务fastp"><a class="markdownIt-Anchor" href="#第二个任务fastp"></a> 第二个任务：fastp</h3>
<p>对测序数据进行质量控制步骤，主要修剪一些接头序列和去除低质量 reads 。这里我一般使用 fastp 自动化完成这一步。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">rule fastp:</span><br><span class="line">    input:</span><br><span class="line">        &quot;input/&#123;sample&#125;_1.fq.gz&quot;,</span><br><span class="line">        &quot;input/&#123;sample&#125;_2.fq.gz&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;results/2_fastp_output/&#123;sample&#125;_1_good.fq.gz&quot;,</span><br><span class="line">        &quot;results/2_fastp_output/&#123;sample&#125;_2_good.fq.gz&quot;,</span><br><span class="line">        &quot;results/2_fastp_output/&#123;sample&#125;.html&quot;,</span><br><span class="line">        &quot;results/2_fastp_output/&#123;sample&#125;.json&quot;</span><br><span class="line">    log:</span><br><span class="line">        &quot;log/&#123;sample&#125;_fastp&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;&quot;&quot;</span><br><span class="line">        fastp -w 10 -h results/2_fastp_output/&#123;wildcards.sample&#125;.html -j results/2_fastp_output/&#123;wildcards.sample&#125;.json -i &#123;input[0]&#125; -I &#123;input[1]&#125; -o &#123;output[0]&#125; -O &#123;output[1]&#125; 2&gt; &#123;log&#125;</span><br><span class="line">        &quot;&quot;&quot;</span><br></pre></td></tr></table></figure>
<p>仔细阅读 snakemake 的文档之后就会发现它的语法非常简单，并且有很强的阅读性。这个任务中需要对 shell 部分做一点解释。<br />
首先，我想每个样品都设置有样品名的结果文件（html 和json），我使用 {wildcards.sample} 获得通配名。<br />
其二，shell 命令里面需要两个设定两个输入和输出，这里和 python 中的切片类似，或者我们可以在 input 和 output 这一步骤就对文件进行命令，这样可读性更好，比如：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">rule fastp:</span><br><span class="line">    input:</span><br><span class="line">        fwd:&quot;input/&#123;sample&#125;_1.fq.gz&quot;,</span><br><span class="line">        rev:&quot;input/&#123;sample&#125;_2.fq.gz&quot;</span><br><span class="line">    output:</span><br><span class="line">        fwwd:&quot;results/2_fastp_output/&#123;sample&#125;_1_good.fq.gz&quot;,</span><br><span class="line">        rev:&quot;results/2_fastp_output/&#123;sample&#125;_2_good.fq.gz&quot;,</span><br><span class="line">        &quot;results/2_fastp_output/&#123;sample&#125;.html&quot;,</span><br><span class="line">        &quot;results/2_fastp_output/&#123;sample&#125;.json&quot;</span><br><span class="line">    log:</span><br><span class="line">        &quot;log/&#123;sample&#125;_fastp&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;&quot;&quot;</span><br><span class="line">        fastp -w 10 -h results/2_fastp_output/&#123;wildcards.sample&#125;.html -j results/2_fastp_output/&#123;wildcards.sample&#125;.json -i &#123;input[fwd]&#125; -I &#123;input[rev]&#125; -o &#123;output[fwd]&#125; -O &#123;output[rev]&#125; 2&gt; &#123;log&#125;</span><br><span class="line">        &quot;&quot;&quot;</span><br></pre></td></tr></table></figure>
<h3 id="第三个任务star"><a class="markdownIt-Anchor" href="#第三个任务star"></a> 第三个任务：STAR</h3>
<p>完成质量控制部分后就可以进行数据比对这一步骤了，上一任务的输出文件是这一任务的输入文件，snakemake 利用不同任务之间文件的关系来自动理顺任务关系。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">rule STAR:</span><br><span class="line">    input:</span><br><span class="line">        &quot;results/2_fastp_output/&#123;sample&#125;_1_good.fq.gz&quot;,</span><br><span class="line">        &quot;results/2_fastp_output/&#123;sample&#125;_2_good.fq.gz&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;results/3_aligned_sequences/&#123;sample&#125;Aligned.sortedByCoord.out.bam&quot;,</span><br><span class="line">        &quot;results/3_aligned_sequences/&#123;sample&#125;Log.final.out&quot;,</span><br><span class="line">        &quot;results/3_aligned_sequences/&#123;sample&#125;Log.out&quot;</span><br><span class="line">    params:&quot;&#123;sample&#125;&quot;</span><br><span class="line">    log:</span><br><span class="line">        &quot;log/&#123;sample&#125;_STAR&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;&quot;&quot;</span><br><span class="line">        STAR --genomeDir &#123;config[star_index]&#125;  --readFilesCommand zcat --readFilesIn &#123;input[0]&#125; &#123;input[1]&#125; --runThreadN 10 \</span><br><span class="line">        --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix results/3_aligned_sequences/&#123;params&#125; 2&gt; &#123;log&#125;</span><br><span class="line">        &quot;&quot;&quot;</span><br></pre></td></tr></table></figure>
<p>我在这个任务中使用了配置文件中的参数，在开始我没有说明这点，我们可以将软件运行参数单独存放在配置文件里，每次使用 snakemake 的时候只需要修改配置文件而不需要对主流程进行修改。<br />
在snakefile 的开始要对配置文件进行声明configfile:“config.yaml”。<br />
在这个任务中我还设置了 params ，我们可以将命令中的一些静态参数写在这里以提高易读性。在 rule 中的 params 可以直接在shell 中调用。</p>
<h3 id="第四个任务featurecounts"><a class="markdownIt-Anchor" href="#第四个任务featurecounts"></a> 第四个任务：featureCounts</h3>
<p>最后就要对各个样品进行定量了。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">rule featureCounts:</span><br><span class="line">    input:</span><br><span class="line">        expand(&quot;results/3_aligned_sequences/&#123;sample&#125;Aligned.sortedByCoord.out.bam&quot;, sample=SAMPLES)</span><br><span class="line">    output:</span><br><span class="line">        &quot;results/4_final_counts/final_counts.txt&quot;,</span><br><span class="line">        &quot;results/4_final_counts/final_counts.txt.summary&quot;</span><br><span class="line">    log:</span><br><span class="line">        &quot;log/featurecounts.log&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;&quot;&quot; </span><br><span class="line">        ls results/3_aligned_sequences/*bam | xargs featureCounts -a &#123;config[gtf]&#125; -o &#123;output[0]&#125; \</span><br><span class="line">        -T 5 -t exon -g gene_id 2&gt; &#123;log&#125;</span><br><span class="line">        &quot;&quot;&quot;</span><br></pre></td></tr></table></figure>
<p>snakemake 另外一个很简便的地方就在于它将列表推导式简化为 expand 函数，比如这个任务中的 input 就会生成所有符合条件的 bam 文件。</p>
<h3 id="总结"><a class="markdownIt-Anchor" href="#总结"></a> 总结</h3>
<p>这只是使用 snakemake 编写的简单的流程，主要想记录下在学习 snakemake 中的几个知识点，snakemake 还有其他的一些使用技巧我会在日后其他的流程中慢慢总结。如果你有写生信分析流程的需要还是非常推荐这个工具，它还有很多高级技巧有待我细细挖掘。</p>
<p><strong>全文来源</strong><br />
<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/14b9eccc0c0e"><u>井底蛙蛙呱呱呱</u></a><br />
<a target="_blank" rel="noopener" href="https://vidotto.top/post/%E4%BD%BF%E7%94%A8-snakemake-%E7%BC%96%E5%86%99-rna-seq-%E6%B5%81%E7%A8%8B/"><u>Vidotto’s blog：使用 snakemake 编写 RNA seq 流程</u></a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/28/Bioinfo-SAM-BAM-MAPQ-value-in-samtools/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/28/Bioinfo-SAM-BAM-MAPQ-value-in-samtools/" class="post-title-link" itemprop="url">基因组比对的SAM/BAM文件中，怎么理解MAPQ值</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-28 11:06:55" itemprop="dateCreated datePublished" datetime="2020-12-28T11:06:55+08:00">2020-12-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-29 09:29:27" itemprop="dateModified" datetime="2020-12-29T09:29:27+08:00">2020-12-29</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>关注这个问题的缘由是对转录组进行基因定量分析的过程中，发现 stringtie 和 htseq-count 产生的结果不一致。</p>
<p>stringtie 程序里面并没有提供 MAPQ 筛选的参数，也就是输入的 bam 文件中，所有 reads 都会被计数。（如果需要干预，需要提前用 samtools view -q 参数筛选 reads）</p>
<p>而 htseq-cuont 默认会使用 MAPQ&gt;10 来做筛选，即只考虑唯一比对的 reads。</p>
<p>实际上，在植物转录组测序的 reads 中，multiple mapping 的 reads 会占据很大的比例。比如有的基因会有多个拷贝，而且表达量还特别高，那么分析的时候是需要考虑这些问题的。</p>
<h3 id="sambam-文件结构简介"><a class="markdownIt-Anchor" href="#sambam-文件结构简介"></a> SAM/BAM 文件结构简介</h3>
<p><img src="https://miro.medium.com/max/875/0*CePvh8XpBcbC1HjZ.png" alt="Example of the SAM file" /></p>
<p>sam/bam文件一共12列，每列具体信息如下。第五例为MAPQ值。</p>
<p>Each alignment line has 11 mandatory fields for essential alignment information such as mapping position, and variable number of optional fields for flexible or aligner specific information:</p>
<ol>
<li>Read Name</li>
<li>SAM flag</li>
<li>chromosome (if read is has no alignment, there will be a “*” here)</li>
<li>position (1-based index, “left end of read”)</li>
<li>MAPQ (mapping quality — describes the uniqueness of the alignment, 0=non-unique, &gt;10 probably unique)</li>
<li>CIGAR string (describes the position of insertions/deletions/matches in the alignment, encodes splice junctions, for example)</li>
<li>Name of mate (mate pair information for paired-end sequencing, often “=”)</li>
<li>Position of mate (mate pair information)</li>
<li>Template length (always zero for me)</li>
<li>Read Sequence</li>
<li>Read Quality</li>
<li>Program specific Flags (i.e. AS is an alignment score, NH is a number of reported alignments that contains the query in the current record)</li>
</ol>
<h3 id="mapq是怎么定义的"><a class="markdownIt-Anchor" href="#mapq是怎么定义的"></a> MAPQ是怎么定义的？</h3>
<p>Samtools<a target="_blank" rel="noopener" href="https://samtools.github.io/hts-specs/SAMv1.pdf">文档</a>中对MAPQ的定义是：</p>
<p>MAPQ: MAPping Quality. It equals −10 log10 Pr{mapping position is wrong}, rounded to the nearest<br />
integer. A value 255 indicates that the mapping quality is not available.</p>
<p>如果正确比对的可能性为 0, 那么错误率 P = 1 - 0 = 1<br />
MAPQ = -10*log10(1) = 0</p>
<p>如果正确比对的可能性为 0.9, 那么错误率 P = 1 - 0.9 = 0.1<br />
MAPQ = -10*log10(0.1) = 10</p>
<p>如果正确比对的可能性为 0.99, 那么错误率 P = 1 - 0.99 = 0.01<br />
MAPQ = -10*log10(0.01) = 20</p>
<p>如果正确比对的可能性为 0.999，那么错误率 P = 1 - 0.999 = 0.001<br />
MAPQ = -10*log10(0.001) = 30</p>
<h3 id="mapq0-的情况到底是指-unmapped-reads-还是指-non-unique-reads"><a class="markdownIt-Anchor" href="#mapq0-的情况到底是指-unmapped-reads-还是指-non-unique-reads"></a> MAPQ=0 的情况到底是指 unmapped reads 还是指 non unique reads？</h3>
<p>MAPQ=0 表示 non unique reads。我查看了 bowtie2 比对之后的 bam 文件，里面 MAPQ=0 的 reads 就是比对到了基因组的多个位置。</p>
<h3 id="mapq0-与-mapq1-有什么区别"><a class="markdownIt-Anchor" href="#mapq0-与-mapq1-有什么区别"></a> MAPQ=0 与 MAPQ=1 有什么区别？</h3>
<p>随机检查了一下 bam 文件中 MAPQ=0 与 MAPQ=1 的两个 PE reads，它们均比对到了基因组的两个不同位置。<br />
应该是比对区域的mismatch导致的MAPQ值不同。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">samtools view File1.bam | grep &#x27;V300018696L3C002R0430736588&#x27;</span><br><span class="line">V300018696L3C002R0430736588     99      Chr01   64085   0       90M     =       66302   331     CCCAGCAGTCTGAGTCCGCAGTATCCACCTTATCGGCGGCTGGACGGAACCCCTTCGCTAAGACATCTTGATACCTTAATTTTTCAATCT      FFFFFFFFFEFF@FFFFFFFFFAFFFFFFFFFFFFFFFF;EB-EC+F%6DDDF7DF&#x27;)DDCD&lt;B@D=@D0&#x27;FE(:@FAC8FF$D&lt;(;C3D     AS:i:-7 ZS:i:-7 XN:i:0  XM:i:3  XO:i:0  XG:i:0  NM:i:3  MD:Z:42A14A24A7 YS:i:-13        YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0430736588     355     Chr01   66206   0       90M     =       66302   186     CCCAGCAGTCTGAGTCCGCAGTATCCACCTTATCGGCGGCTGGACGGAACCCCTTCGCTAAGACATCTTGATACCTTAATTTTTCAATCT      FFFFFFFFFEFF@FFFFFFFFFAFFFFFFFFFFFFFFFF;EB-EC+F%6DDDF7DF&#x27;)DDCD&lt;B@D=@D0&#x27;FE(:@FAC8FF$D&lt;(;C3D     AS:i:-7 ZS:i:-7 XN:i:0  XM:i:3  XO:i:0  XG:i:0  NM:i:3  MD:Z:42A14A24A7 YS:i:-13        YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0430736588     147     Chr01   66302   0       19M1007N58M13S  =       64085   -331    CTTCTAGCTGTTATCACTCCATCATCCCTCTCTTGCATCTTCTGCTTTTCCTTTGGTGTGGACTTCAACTGGACAACAGGTCCCATGCTC      EFDEFFFFEFEEFEFFFEFFFBFFDFFFFFDFEEFGEEFFEFFFFDEEFFFFDFFFFFFFEEFFDFFFFFFEFFEEFFFFAFEFEFFFEF     AS:i:-13        XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:77 YS:i:-7 YT:Z:CP XS:A:-  NH:i:2</span><br><span class="line">V300018696L3C002R0430736588     403     Chr01   66302   0       19M1007N58M13S  =       66206   -186    CTTCTAGCTGTTATCACTCCATCATCCCTCTCTTGCATCTTCTGCTTTTCCTTTGGTGTGGACTTCAACTGGACAACAGGTCCCATGCTC      EFDEFFFFEFEEFEFFFEFFFBFFDFFFFFDFEEFGEEFFEFFFFDEEFFFFDFFFFFFFEEFFDFFFFFFEFFEEFFFFAFEFEFFFEF     AS:i:-13        XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:77 YS:i:-7 YT:Z:CP XS:A:-  NH:i:2</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">samtools view File1.bam | grep &#x27;V300018696L3C002R0110944536&#x27;</span><br><span class="line">V300018696L3C002R0110944536     99      Chr01   33996   1       90M     =       34061   155     GGAGAGAATTCCTATGCTGCCATATTGTAATAGATAGAGCAATCCACCACATGCCCCGCCTTCTGTGATTACTCGTGGCAACAAATCCAT      FEDFEFFFFEFFFFFDFFFFFFFFFFFCFFFFFEFFFDFEFFFEFFEFFF3FFFFFFEFFFFEFFDFDFE3AE?D6FDEFEEFEFECFFF     AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:90 YS:i:0  YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0110944536     147     Chr01   34061   1       90M     =       33996   -155    TGATTACTCGTGGCAACAAATCCATCACAGAATTGGATAAAATGGCTAGCCGGATCAGCTGAAAGAGCTCCTATACCCCGGATCCATCTC      DEBFF2E,DFED&gt;FDDFFCFDDFFEFDAFFEEEEBFCFFDEFECEFFEEFDFFFDEDFFFFEEDEFFFCEFFFFDFEBEFFECFFEEFDF     AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:90 YS:i:0  YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0110944536     355     Chr03   35175098        1       90M     =       35175163        155     GGAGAGAATTCCTATGCTGCCATATTGTAATAGATAGAGCAATCCACCACATGCCCCGCCTTCTGTGATTACTCGTGGCAACAAATCCAT      FEDFEFFFFEFFFFFDFFFFFFFFFFFCFFFFFEFFFDFEFFFEFFEFFF3FFFFFFEFFFFEFFDFDFE3AE?D6FDEFEEFEFECFFF     AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:90 YS:i:0  YT:Z:CP NH:i:2</span><br><span class="line">V300018696L3C002R0110944536     403     Chr03   35175163        1       90M     =       35175098        -155    TGATTACTCGTGGCAACAAATCCATCACAGAATTGGATAAAATGGCTAGCCGGATCAGCTGAAAGAGCTCCTATACCCCGGATCCATCTC      DEBFF2E,DFED&gt;FDDFFCFDDFFEFDAFFEEEEBFCFFDEFECEFFEEFDFFFDEDFFFFEEDEFFFCEFFFFDFEBEFFECFFEEFDF     AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:90 YS:i:0  YT:Z:CP NH:i:2</span><br></pre></td></tr></table></figure>
<h3 id="为什么bwa和bowtie2生成的结果中mapq不一样"><a class="markdownIt-Anchor" href="#为什么bwa和bowtie2生成的结果中mapq不一样"></a> 为什么BWA和bowtie2生成的结果中MAPQ不一样？</h3>
<p>In SAM file the quality of mapped reads is defined in by so-called MAPQ values — MAPping Quality. It equals −10 log10 Probability {mapping position is wrong}, rounded to the nearest integer. A value 255 indicates that the mapping quality is not available. Now, it is very important to remember that MAPQ values generated by different aligners (e.g. Bowtie2, TopHat, BBMap, BWA) are not exactly comparable. For example, the maximum value of MAPQ score in Bowtie2 is 42, whereas in BWA — 37. You can read a great piece about why this happens in the ACGT blog post.</p>
<p>Another important thing not everyone knows is how MAPQ value is actually calculated. In our of our previous tutorial, we mentioned that it defines how “uniquely” a particular read maps to a reference. In reality, of course, it is more complex than that and you should check out great post here here to understand the exact procedure on how score gets calculated.</p>
<p>Lastly, remember that TopHat outputs only MAPQ scores of 0, 1, 3, or 50. The first three values indicate mappings to 5 or more locations, 3–4 locations, or 2 locations, whereas a value of 50 represents a unique match. Also, in older versions of TopHat unique matches were identified with a value of 255. Therefore, you must understand that values of Bowtie and Bowtie2 (used internally by TopHat to do read mapping) produce a different range of MAPQ scores (0–42).</p>
<p>Now, as we understand a little bit better MAPQ scores, we can filter BAM/SAM files on the mapping quality. eg. getting all reads with a mapping quality larger than 30 (you could also use 49 if you wanted to:). This will remove all reads mapped the undesirable mapping qualities and will only keep uniquely mapped reads. (<a target="_blank" rel="noopener" href="https://medium.com/@shilparaopradeep/samtools-guide-learning-how-to-filter-and-manipulate-with-sam-bam-files-2c28b25d29e8">source</a>)</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/25/Linux-vim-setting/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/25/Linux-vim-setting/" class="post-title-link" itemprop="url">Linux：vim设置</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2020-12-25 23:26:27 / Modified: 23:47:25" itemprop="dateCreated datePublished" datetime="2020-12-25T23:26:27+08:00">2020-12-25</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>以前是用spf13-vim方案的，运行一个安装命令就可以使用了。最喜欢的功能是自动缩进，还有代码块高亮显示（比如while循环下面的每行子命令左侧都会有一个高亮的方块）。</p>
<p>今天不知道为什么在tmux下面用vim出现问题，很多字段有莫名其妙的红色背景，而且鼠标左键选中字符的时候无法选中，反而是将光标移过去了。</p>
<p>于是换了一个vim设置，如下的代码（<a target="_blank" rel="noopener" href="https://blog.csdn.net/geng823/article/details/41804399">代码来源</a>），新建<code>~/.vimrc</code>文件，将下面的代码写入这个文件，保存退出即可，代码中有几行是作者email等信息可以改掉。测试了一下，下方的代码可以用，至少有了我需要自动缩进功能，且不会有上述问题，另外还有新建.sh代码的时候，会自动生成一个包含作者信息和新建文件时间的几行注释信息，很实用的小功能。记录在此。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br><span class="line">229</span><br><span class="line">230</span><br><span class="line">231</span><br><span class="line">232</span><br><span class="line">233</span><br><span class="line">234</span><br><span class="line">235</span><br><span class="line">236</span><br><span class="line">237</span><br><span class="line">238</span><br><span class="line">239</span><br><span class="line">240</span><br><span class="line">241</span><br><span class="line">242</span><br><span class="line">243</span><br><span class="line">244</span><br><span class="line">245</span><br><span class="line">246</span><br><span class="line">247</span><br><span class="line">248</span><br><span class="line">249</span><br><span class="line">250</span><br><span class="line">251</span><br><span class="line">252</span><br><span class="line">253</span><br><span class="line">254</span><br><span class="line">255</span><br><span class="line">256</span><br><span class="line">257</span><br><span class="line">258</span><br><span class="line">259</span><br><span class="line">260</span><br><span class="line">261</span><br><span class="line">262</span><br><span class="line">263</span><br><span class="line">264</span><br><span class="line">265</span><br><span class="line">266</span><br><span class="line">267</span><br><span class="line">268</span><br><span class="line">269</span><br><span class="line">270</span><br><span class="line">271</span><br><span class="line">272</span><br><span class="line">273</span><br><span class="line">274</span><br><span class="line">275</span><br><span class="line">276</span><br><span class="line">277</span><br><span class="line">278</span><br><span class="line">279</span><br><span class="line">280</span><br><span class="line">281</span><br><span class="line">282</span><br><span class="line">283</span><br><span class="line">284</span><br><span class="line">285</span><br><span class="line">286</span><br><span class="line">287</span><br><span class="line">288</span><br><span class="line">289</span><br><span class="line">290</span><br><span class="line">291</span><br><span class="line">292</span><br><span class="line">293</span><br><span class="line">294</span><br><span class="line">295</span><br><span class="line">296</span><br><span class="line">297</span><br><span class="line">298</span><br><span class="line">299</span><br><span class="line">300</span><br><span class="line">301</span><br><span class="line">302</span><br><span class="line">303</span><br><span class="line">304</span><br><span class="line">305</span><br><span class="line">306</span><br><span class="line">307</span><br><span class="line">308</span><br><span class="line">309</span><br><span class="line">310</span><br><span class="line">311</span><br><span class="line">312</span><br><span class="line">313</span><br><span class="line">314</span><br><span class="line">315</span><br><span class="line">316</span><br><span class="line">317</span><br><span class="line">318</span><br><span class="line">319</span><br><span class="line">320</span><br><span class="line">321</span><br><span class="line">322</span><br><span class="line">323</span><br><span class="line">324</span><br><span class="line">325</span><br><span class="line">326</span><br></pre></td><td class="code"><pre><span class="line">&quot; An example for a vimrc file.</span><br><span class="line">&quot;</span><br><span class="line">&quot; Maintainer:	Bram Moolenaar &lt;Bram@vim.org&gt;</span><br><span class="line">&quot; Last change:	2001 Jul 18</span><br><span class="line">&quot;</span><br><span class="line">&quot; To use it, copy it to</span><br><span class="line">&quot;     for Unix and OS/2:  ~/.vimrc</span><br><span class="line">&quot;	      for Amiga:  s:.vimrc</span><br><span class="line">&quot;  for MS-DOS and Win32:  $VIM\_vimrc</span><br><span class="line">&quot;	    for OpenVMS:  sys$login:.vimrc</span><br><span class="line"></span><br><span class="line">set encoding=utf-8</span><br><span class="line">set fileencodings=utf-8,gbk,gb2312,gb18030,ucs-bom,latin1</span><br><span class="line"></span><br><span class="line">set tags=tags;</span><br><span class="line">set shiftwidth=4 </span><br><span class="line">set ts=4</span><br><span class="line">&quot;set nu</span><br><span class="line">set vb t_vb=</span><br><span class="line">&quot; When started as &quot;evim&quot;, evim.vim will already have done these settings.</span><br><span class="line">if v:progname =~? &quot;evim&quot;</span><br><span class="line">  finish</span><br><span class="line">endif</span><br><span class="line"></span><br><span class="line">&quot; Use Vim settings, rather then Vi settings (much better!).</span><br><span class="line">&quot; This must be first, because it changes other options as a side effect.</span><br><span class="line">set nocompatible</span><br><span class="line"></span><br><span class="line">&quot; allow backspacing over everything in insert mode</span><br><span class="line">set backspace=indent,eol,start</span><br><span class="line"></span><br><span class="line">set autoindent		&quot; always set autoindenting on</span><br><span class="line">set nobackup		&quot; do not keep a backup file, use versions instead</span><br><span class="line">&quot;if has(&quot;vms&quot;)</span><br><span class="line">&quot;  set nobackup		&quot; do not keep a backup file, use versions instead</span><br><span class="line">&quot;else</span><br><span class="line">&quot;  set backup		&quot; keep a backup file</span><br><span class="line">&quot;endif</span><br><span class="line">set history=50		&quot; keep 50 lines of command line history</span><br><span class="line">set ruler		&quot; show the cursor position all the time</span><br><span class="line">set showcmd		&quot; display incomplete commands</span><br><span class="line">set incsearch		&quot; do incremental searching</span><br><span class="line"></span><br><span class="line">set nobackup		&quot; do not keep a backup file, use versions instead</span><br><span class="line"></span><br><span class="line">&quot; For Win32 GUI: remove &#x27;t&#x27; flag from &#x27;guioptions&#x27;: no tearoff menu entries</span><br><span class="line">&quot; let &amp;guioptions = substitute(&amp;guioptions, &quot;t&quot;, &quot;&quot;, &quot;g&quot;)</span><br><span class="line"></span><br><span class="line">&quot; Don&#x27;t use Ex mode, use Q for formatting</span><br><span class="line">map Q gq</span><br><span class="line"></span><br><span class="line">&quot; Make p in Visual mode replace the selected text with the &quot;&quot; register.</span><br><span class="line">vnoremap p &lt;Esc&gt;:let current_reg = @&quot;&lt;CR&gt;gvs&lt;C-R&gt;=current_reg&lt;CR&gt;&lt;Esc&gt;</span><br><span class="line"></span><br><span class="line">&quot; This is an alternative that also works in block mode, but the deleted</span><br><span class="line">&quot; text is lost and it only works for putting the current register.</span><br><span class="line">&quot;vnoremap p &quot;_dp</span><br><span class="line"></span><br><span class="line">&quot; Switch syntax highlighting on, when the terminal has colors</span><br><span class="line">&quot; Also switch on highlighting the last used search pattern.</span><br><span class="line">if &amp;t_Co &gt; 2 || has(&quot;gui_running&quot;)</span><br><span class="line">  syntax on</span><br><span class="line">  set hlsearch</span><br><span class="line">endif</span><br><span class="line"></span><br><span class="line">&quot; Only do this part when compiled with support for autocommands.</span><br><span class="line">if has(&quot;autocmd&quot;)</span><br><span class="line"></span><br><span class="line">  &quot; Enable file type detection.</span><br><span class="line">  &quot; Use the default filetype settings, so that mail gets &#x27;tw&#x27; set to 72,</span><br><span class="line">  &quot; &#x27;cindent&#x27; is on in C files, etc.</span><br><span class="line">  &quot; Also load indent files, to automatically do language-dependent indenting.</span><br><span class="line">  filetype plugin indent on</span><br><span class="line"></span><br><span class="line">  &quot; For all text files set &#x27;textwidth&#x27; to 78 characters.</span><br><span class="line">  autocmd FileType text setlocal textwidth=78</span><br><span class="line"></span><br><span class="line">  &quot; When editing a file, always jump to the last known cursor position.</span><br><span class="line">  &quot; Don&#x27;t do it when the position is invalid or when inside an event handler</span><br><span class="line">  &quot; (happens when dropping a file on gvim).</span><br><span class="line">  autocmd BufReadPost *</span><br><span class="line">    \ if line(&quot;&#x27;\&quot;&quot;) &gt; 0 &amp;&amp; line(&quot;&#x27;\&quot;&quot;) &lt;= line(&quot;$&quot;) |</span><br><span class="line">    \   exe &quot;normal g`\&quot;&quot; |</span><br><span class="line">    \ endif</span><br><span class="line"></span><br><span class="line">endif &quot; has(&quot;autocmd&quot;)</span><br><span class="line"></span><br><span class="line">if has(&quot;cscope&quot;)</span><br><span class="line">	let current = &quot;.&quot;</span><br><span class="line">	let num = 1</span><br><span class="line">	while num &lt; 20 </span><br><span class="line">		if filereadable(current . &quot;/cscope.out&quot;)</span><br><span class="line">			let $CSCOPE_DB = current . &quot;/cscope.out&quot;</span><br><span class="line">			cs add $CSCOPE_DB</span><br><span class="line">			break</span><br><span class="line">		else</span><br><span class="line">			let current = current . &quot;/..&quot;</span><br><span class="line">			let num = num + 1</span><br><span class="line">		endif</span><br><span class="line">	endwhile</span><br><span class="line">endif</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">syntax enable</span><br><span class="line">syntax on</span><br><span class="line">colorscheme desert</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">filetype plugin indent on</span><br><span class="line">set completeopt=longest,menu</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">set cst</span><br><span class="line">set csto=1</span><br><span class="line">set cscopequickfix=s-,c-,d-,i-,t-,e-,f-</span><br><span class="line">&quot; cs add /home/yangxl/readcode/cscope-kernel/cscope.out</span><br><span class="line">&quot; cs add /home/yangxl/readcode/cscope-app/cscope.out</span><br><span class="line">let Tlist_Enable_Fold_Column = 0</span><br><span class="line">let Tlist_WinWidth = 30</span><br><span class="line">let Tlist_Show_One_File = 1</span><br><span class="line">let g:miniBufExplMapCTabSwitchBufs = 1</span><br><span class="line">let g:miniBufExplMapWindowNavVim = 1</span><br><span class="line">let g:miniBufExplMapWindowNavArrows = 1</span><br><span class="line">let g:winManagerWindowLayout=&#x27;FileExplorer|TagList&#x27;</span><br><span class="line">let g:SuperTabRetainCompletionType=2</span><br><span class="line">nmap wm :WMToggle&lt;cr&gt;</span><br><span class="line">set tabstop=4</span><br><span class="line">&quot;nmap &lt;F2&gt; :cs find d &lt;C-R&gt;&lt;C-W&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;F3&gt; :cs find c &lt;C-R&gt;&lt;C-W&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;F4&gt; :cs find t &lt;C-R&gt;&lt;C-W&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;F5&gt; :cs find e &lt;C-R&gt;&lt;C-W&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;F6&gt; :cs find f &lt;C-R&gt;&lt;C-W&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;F7&gt; :cs find i &lt;C-R&gt;&lt;C-W&gt;&lt;CR&gt;</span><br><span class="line">nmap &lt;F5&gt; :TlistToggle &lt;CR&gt;</span><br><span class="line"></span><br><span class="line">nmap &lt;F2&gt; :cs find c &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">nmap &lt;F3&gt; :cs find s &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">nmap &lt;F4&gt; :cs find g &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">nmap &lt;F6&gt; :cs find f </span><br><span class="line">nmap &lt;C-n&gt; :cn&lt;CR&gt;</span><br><span class="line">nmap &lt;C-p&gt; :cp&lt;CR&gt;</span><br><span class="line">&quot;nmap  :cs find t &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap  :cs find e &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap  :cs find f &lt;C-R&gt;=expand(&quot;&lt;cfile&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap  :cs find i ^&lt;C-R&gt;=expand(&quot;&lt;cfile&gt;&quot;)&lt;CR&gt;$&lt;CR&gt;</span><br><span class="line">&quot;nmap  :cs find d &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line"></span><br><span class="line">&quot; Using &#x27;CTRL-spacebar&#x27; then a search type makes the vim window</span><br><span class="line">&quot; split horizontally, with search result displayed in</span><br><span class="line">&quot; the new window.</span><br><span class="line"></span><br><span class="line">&quot;nmap &lt;C-Space&gt;s :scs find s &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;g :scs find g &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;c :scs find c &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;t :scs find t &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;e :scs find e &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;f :scs find f &lt;C-R&gt;=expand(&quot;&lt;cfile&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;i :scs find i ^&lt;C-R&gt;=expand(&quot;&lt;cfile&gt;&quot;)&lt;CR&gt;$&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;F2&gt; :scs find d &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line"></span><br><span class="line">&quot; Hitting CTRL-space *twice* before the search type does a vertical</span><br><span class="line">&quot; split instead of a horizontal one</span><br><span class="line"></span><br><span class="line">&quot;nmap &lt;C-Space&gt;&lt;C-Space&gt;s</span><br><span class="line">        \:vert scs find s &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;&lt;C-Space&gt;g</span><br><span class="line">        \:vert scs find g &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;&lt;C-Space&gt;c</span><br><span class="line">        \:vert scs find c &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;&lt;C-Space&gt;t</span><br><span class="line">        \:vert scs find t &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;&lt;C-Space&gt;e</span><br><span class="line">        \:vert scs find e &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;C-Space&gt;&lt;C-Space&gt;i</span><br><span class="line">        \:vert scs find i ^&lt;C-R&gt;=expand(&quot;&lt;cfile&gt;&quot;)&lt;CR&gt;$&lt;CR&gt;</span><br><span class="line">&quot;nmap &lt;F2&gt; </span><br><span class="line">        \:vert scs find d &lt;C-R&gt;=expand(&quot;&lt;cword&gt;&quot;)&lt;CR&gt;&lt;CR&gt;</span><br><span class="line"></span><br><span class="line">set nu</span><br><span class="line">set mouse=a</span><br><span class="line">set autoindent</span><br><span class="line">set cindent</span><br><span class="line"></span><br><span class="line">filetype plugin indent on </span><br><span class="line">&quot;打开文件类型检测, 加了这句才可以用智能补全</span><br><span class="line">set completeopt=longest,menu</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span><br><span class="line">&quot;&quot;&quot;&quot;&quot;新文件标题</span><br><span class="line">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span><br><span class="line">&quot;新建.c,.h,.sh,.java文件，自动插入文件头 </span><br><span class="line">autocmd BufNewFile *.cpp,*.[ch],*.sh,*.java exec &quot;:call SetTitle()&quot; </span><br><span class="line">&quot;&quot;定义函数SetTitle，自动插入文件头 </span><br><span class="line">func SetTitle() </span><br><span class="line">	&quot;如果文件类型为.sh文件 </span><br><span class="line">	if &amp;filetype == &#x27;sh&#x27; </span><br><span class="line">		call setline(1,&quot;\#########################################################################&quot;) </span><br><span class="line">		call append(line(&quot;.&quot;), &quot;\# File Name: &quot;.expand(&quot;%&quot;)) </span><br><span class="line">		call append(line(&quot;.&quot;)+1, &quot;\# Author: xx&quot;) </span><br><span class="line">		call append(line(&quot;.&quot;)+2, &quot;\# Mail: xx@163.com&quot;) </span><br><span class="line">		call append(line(&quot;.&quot;)+3, &quot;\# Created Time: &quot;.strftime(&quot;%c&quot;)) </span><br><span class="line">		call append(line(&quot;.&quot;)+4, &quot;\#########################################################################&quot;) </span><br><span class="line">		call append(line(&quot;.&quot;)+5, &quot;\#!/bin/bash&quot;) </span><br><span class="line">		call append(line(&quot;.&quot;)+6, &quot;&quot;) </span><br><span class="line">	else </span><br><span class="line">		call setline(1, &quot;/*************************************************************************&quot;) </span><br><span class="line">		call append(line(&quot;.&quot;), &quot;	&gt; File Name: &quot;.expand(&quot;%&quot;)) </span><br><span class="line">		call append(line(&quot;.&quot;)+1, &quot;	&gt; Author: xx&quot;) </span><br><span class="line">		call append(line(&quot;.&quot;)+2, &quot;	&gt; Mail: xx@163.com&quot;) </span><br><span class="line">		call append(line(&quot;.&quot;)+3, &quot;	&gt; Created Time: &quot;.strftime(&quot;%c&quot;)) </span><br><span class="line">		call append(line(&quot;.&quot;)+4, &quot; ************************************************************************/&quot;) </span><br><span class="line">		call append(line(&quot;.&quot;)+5, &quot;&quot;)</span><br><span class="line">	endif</span><br><span class="line">	if &amp;filetype == &#x27;cpp&#x27;</span><br><span class="line">		call append(line(&quot;.&quot;)+6, &quot;#include&lt;iostream&gt;&quot;)</span><br><span class="line">		call append(line(&quot;.&quot;)+7, &quot;using namespace std;&quot;)</span><br><span class="line">		call append(line(&quot;.&quot;)+8, &quot;&quot;)</span><br><span class="line">	endif</span><br><span class="line">	if &amp;filetype == &#x27;c&#x27;</span><br><span class="line">		call append(line(&quot;.&quot;)+6, &quot;#include&lt;stdio.h&gt;&quot;)</span><br><span class="line">		call append(line(&quot;.&quot;)+7, &quot;&quot;)</span><br><span class="line">	endif</span><br><span class="line">	&quot;	if &amp;filetype == &#x27;java&#x27;</span><br><span class="line">	&quot;		call append(line(&quot;.&quot;)+6,&quot;public class &quot;.expand(&quot;%&quot;))</span><br><span class="line">	&quot;		call append(line(&quot;.&quot;)+7,&quot;&quot;)</span><br><span class="line">	&quot;	endif</span><br><span class="line">	&quot;新建文件后，自动定位到文件末尾</span><br><span class="line">	autocmd BufNewFile * normal G</span><br><span class="line">endfunc </span><br><span class="line"></span><br><span class="line"></span><br><span class="line">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span><br><span class="line">&quot; 显示相关  </span><br><span class="line">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span><br><span class="line">&quot;set shortmess=atI   &quot; 启动的时候不显示那个援助乌干达儿童的提示  </span><br><span class="line">&quot;winpos 5 5          &quot; 设定窗口位置  </span><br><span class="line">&quot;set lines=40 columns=155    &quot; 设定窗口大小  </span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">&quot;</span><br><span class="line">set go=             &quot; 不要图形按钮  </span><br><span class="line">&quot;color asmanian2     &quot; 设置背景主题  </span><br><span class="line">&quot;set guifont=Courier_New:h10:cANSI   &quot; 设置字体  </span><br><span class="line">&quot;syntax on           &quot; 语法高亮  </span><br><span class="line">&quot;</span><br><span class="line">&quot;autocmd InsertLeave * se nocul  &quot; 用浅色高亮当前行  </span><br><span class="line">&quot;autocmd InsertEnter * se cul    &quot; 用浅色高亮当前行 </span><br><span class="line">set ruler           &quot; 显示标尺  </span><br><span class="line">set showcmd         &quot; 输入的命令显示出来，看的清楚些  </span><br><span class="line">&quot;set cmdheight=1     &quot; 命令行（在状态行下）的高度，设置为1  </span><br><span class="line">&quot;set whichwrap+=&lt;,&gt;,h,l   &quot; 允许backspace和光标键跨越行边界(不建议)  </span><br><span class="line">&quot;set scrolloff=3     &quot; 光标移动到buffer的顶部和底部时保持3行距离  </span><br><span class="line">&quot;</span><br><span class="line">set novisualbell    &quot; 不要闪烁(不明白)  </span><br><span class="line">set statusline=%F%m%r%h%w\ [FORMAT=%&#123;&amp;ff&#125;]\ [TYPE=%Y]\ [POS=%l,%v][%p%%]\ %&#123;strftime(\&quot;%d/%m/%y\ -\ %H:%M\&quot;)&#125;   &quot;状态行显示的内容  </span><br><span class="line">set laststatus=1    &quot; 启动显示状态行(1),总是显示状态行(2)  </span><br><span class="line">set foldenable      &quot; 允许折叠  </span><br><span class="line">set foldmethod=manual   &quot; 手动折叠  </span><br><span class="line">&quot;set background=dark &quot;背景使用黑色 </span><br><span class="line">&quot;</span><br><span class="line">set nocompatible  &quot;去掉讨厌的有关vi一致性模式，避免以前版本的一些bug和局限  </span><br><span class="line">&quot; 显示中文帮助</span><br><span class="line">if version &gt;= 603</span><br><span class="line">	set helplang=cn</span><br><span class="line">	set encoding=utf-8</span><br><span class="line">endif</span><br><span class="line">&quot; 设置配色方案</span><br><span class="line">&quot;colorscheme murphy</span><br><span class="line">&quot;字体 </span><br><span class="line">&quot;if (has(&quot;gui_running&quot;)) </span><br><span class="line">&quot;   set guifont=Bitstream\ Vera\ Sans\ Mono\ 10 </span><br><span class="line">&quot;endif </span><br><span class="line"></span><br><span class="line">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span><br><span class="line">&quot;键盘命令</span><br><span class="line">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span><br><span class="line"></span><br><span class="line">nmap &lt;leader&gt;w :w!&lt;cr&gt;</span><br><span class="line">nmap &lt;leader&gt;f :find&lt;cr&gt;</span><br><span class="line"></span><br><span class="line">&quot; 映射全选+复制 ctrl+a</span><br><span class="line">map &lt;C-A&gt; ggVGY</span><br><span class="line">map! &lt;C-A&gt; &lt;Esc&gt;ggVGY</span><br><span class="line">map &lt;F12&gt; gg=G</span><br><span class="line">&quot; 选中状态下 Ctrl+c 复制</span><br><span class="line">vmap &lt;C-c&gt; &quot;+y</span><br><span class="line">&quot;去空行  </span><br><span class="line">nnoremap &lt;F2&gt; :g/^\s*$/d&lt;CR&gt; </span><br><span class="line">&quot;比较文件  </span><br><span class="line">nnoremap &lt;C-F2&gt; :vert diffsplit </span><br><span class="line">&quot;新建标签  </span><br><span class="line">map &lt;M-F2&gt; :tabnew&lt;CR&gt;  </span><br><span class="line">&quot;列出当前目录文件  </span><br><span class="line">map &lt;F3&gt; :tabnew .&lt;CR&gt;  </span><br><span class="line">&quot;打开树状文件目录  </span><br><span class="line">map &lt;C-F3&gt; \be  </span><br><span class="line">&quot;C，C++ 按F5编译运行</span><br><span class="line">map &lt;F5&gt; :call CompileRunGcc()&lt;CR&gt;</span><br><span class="line">func! CompileRunGcc()</span><br><span class="line">	exec &quot;w&quot;</span><br><span class="line">	if &amp;filetype == &#x27;c&#x27;</span><br><span class="line">		exec &quot;!g++ % -o %&lt;&quot;</span><br><span class="line">		exec &quot;! ./%&lt;&quot;</span><br><span class="line">	elseif &amp;filetype == &#x27;cpp&#x27;</span><br><span class="line">		exec &quot;!g++ % -o %&lt;&quot;</span><br><span class="line">		exec &quot;! ./%&lt;&quot;</span><br><span class="line">	elseif &amp;filetype == &#x27;java&#x27; </span><br><span class="line">		exec &quot;!javac %&quot; </span><br><span class="line">		exec &quot;!java %&lt;&quot;</span><br><span class="line">	elseif &amp;filetype == &#x27;sh&#x27;</span><br><span class="line">		:!./%</span><br><span class="line">	elseif &amp;filetype == &#x27;py&#x27;</span><br><span class="line">		exec &quot;!python %&quot;</span><br><span class="line">		exec &quot;!python %&lt;&quot;</span><br><span class="line">	endif</span><br><span class="line">endfunc</span><br><span class="line">&quot;C,C++的调试</span><br><span class="line">map &lt;F8&gt; :call Rungdb()&lt;CR&gt;</span><br><span class="line">func! Rungdb()</span><br><span class="line">	exec &quot;w&quot;</span><br><span class="line">	exec &quot;!g++ % -g -o %&lt;&quot;</span><br><span class="line">	exec &quot;!gdb ./%&lt;&quot;</span><br><span class="line">endfunc</span><br></pre></td></tr></table></figure>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/15/Bioinfo-HiC-3d-dna-for-chromosome-assembly/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/15/Bioinfo-HiC-3d-dna-for-chromosome-assembly/" class="post-title-link" itemprop="url">Hi-C：3d-dna挂载染色体</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-15 14:24:17" itemprop="dateCreated datePublished" datetime="2020-12-15T14:24:17+08:00">2020-12-15</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-29 18:01:27" itemprop="dateModified" datetime="2020-12-29T18:01:27+08:00">2020-12-29</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>使用二代数据或三代数据得到contig后，下一步就是将contig提升到染色体水平。有很多策略可以做到这一点，比如说遗传图谱，BioNano, HiC, 参考近源物种。如果利用HiC进行准染色体水平，那么目前常见的组装软件有下面几个。</p>
<ul>
<li>HiRise: 2015年后的GitHub就不再更新</li>
<li>LACHESIS: 发表在NBT，2017年后不再更新</li>
<li>SALSA: 发表在BMC genomics, 仍在更新中</li>
<li>3D-DNA: 发表在science，仍在更新中</li>
<li>ALLHiC: 发表在Nature Plants, 用于解决植物多倍体组装问题</li>
</ul>
<p>对于二倍体物种而言，目前3D-DNA应该是组装效果最好的一个软件。<a target="_blank" rel="noopener" href="https://github.com/theaidenlab/3d-dna">3d-dna</a>是一款简单、方便的处理Hi-C软件，可将contig提升到染色体水平。3D-DNA流程简介:</p>
<ul>
<li>将Hi-C数据比对到draft.genome.fa。（利用Juicer分析Hi-C数据）</li>
<li>利用自动化流程进行纠错（misjoin），排序（order），确定正确方向（orient），最后scaffolding，得到染色体水平的组装结果（3D-DNA分析）</li>
<li>Juicebox 进行人工纠错</li>
</ul>
<h3 id="一-分析软件安装"><a class="markdownIt-Anchor" href="#一-分析软件安装"></a> 一、分析软件安装</h3>
<p>依赖的软件</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">LastZ (version 1.03.73 released 20150708) – for diploid mode only</span><br><span class="line">Java version &gt;=1.8</span><br><span class="line">Bash &gt;=4</span><br><span class="line">GNU Awk &gt;=4.0.2</span><br><span class="line">GNU coreutils sort &gt;=8.11</span><br><span class="line">Python &gt;=2.7 - for chromosome number-aware splitter module only</span><br><span class="line">scipy numpy matplotlib - for chromosome number-aware splitter module only</span><br><span class="line">GUN Parallel &gt;=20150322 (可选，建议装)</span><br><span class="line">bwa</span><br></pre></td></tr></table></figure>
<p>安装Juicer</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/theaidenlab/juicer.git</span><br><span class="line">cd juicer</span><br><span class="line">ln -s CPU scripts</span><br><span class="line">cd scripts/common</span><br><span class="line">wget https://hicfiles.tc4ga.com/public/juicer/juicer_tools.1.9.9_jcuda.0.8.jar</span><br><span class="line"># wget https://s3.amazonaws.com/hicfiles.tc4ga.com/public/juicer/juicer_tools_1.22.01.jar</span><br><span class="line">ln -s juicer_tools.1.9.9_jcuda.0.8.jar  juicer_tools.jar</span><br></pre></td></tr></table></figure>
<p>安装3D-DNA</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/theaidenlab/3d-dna.git</span><br></pre></td></tr></table></figure>
<p>数据准备<br />
<code>ref</code>文件夹： 存放<code>draft.genome.fa</code><br />
<code>fastq</code>文件夹: 存放Hi-C测序双端reads, 注意reads文件名的格式 <code>*_R1.fastq, *_R2.fastq</code>。<br />
否则报错<code>Failed to find any files matching /xxx/fastq/*_R*.fastq*</code>。</p>
<h3 id="二-利用-juicer-分析将hi-c数据比对到contigsscaffolds"><a class="markdownIt-Anchor" href="#二-利用-juicer-分析将hi-c数据比对到contigsscaffolds"></a> 二、利用 Juicer 分析将Hi-C数据比对到contigs/scaffolds</h3>
<ol>
<li>基因组建立索引</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bwa index draft.genome.fa</span><br></pre></td></tr></table></figure>
<ol start="2">
<li>创建可能的酶切位点文件</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">python ~/software/juicer/misc/generate_site_positions.py  MboI  draft.genome  draft.genome.fa</span><br><span class="line">\# 如果使用的是 HindIII 进行酶切，就设定为HindIII；这里我设定为MboI</span><br></pre></td></tr></table></figure>
<ol start="3">
<li>获取每条contig的长度</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk &#x27;BEGIN&#123;OFS=&quot;\t&quot;&#125;&#123;print $1, $NF&#125;&#x27; draft.genome_MboI.txt &gt; draft.genome.chrom.sizes</span><br></pre></td></tr></table></figure>
<p>设置当前工作目录的文件结构如下，目录存在fastq和ref文件夹。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">ref</span><br><span class="line">├── draft.genome.chrom.sizes</span><br><span class="line">├── draft.genome.fa -&gt; ../draft.genome.fa</span><br><span class="line">├── draft.genome.fa.amb</span><br><span class="line">├── draft.genome.fa.ann</span><br><span class="line">├── draft.genome.fa.bwt</span><br><span class="line">├── draft.genome.fa.pac</span><br><span class="line">├── draft.genome.fa.sa</span><br><span class="line">├── draft.genome_MboI.txt</span><br><span class="line">└── opt.sh</span><br><span class="line">fastq</span><br><span class="line">├── XXX_R1.fastq.gz -&gt; /PATH/TO/XXX_R1.fastq.gz</span><br><span class="line">└── XXX_R2.fastq.gz -&gt; /PATH/TO/XXX_R2.fastq.gz</span><br></pre></td></tr></table></figure>
<ol start="4">
<li>运行juicer<br />
注意：必须在当前目录存在fastq和ref文件夹， -z,-p,-y,-D是必须参数</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">~/software/juicer/scripts/juicer.sh -g draft_genome -s HindIII -z ./ref/draft.genome.fa -y ./ref/draft.genome_HindIII.txt -p ./ref/draft.genome.chrom.sizes -D /PATH/TO/juicer -t 8</span><br><span class="line"></span><br><span class="line">\## juicer 参数</span><br><span class="line">-g： 定义一个物种名</span><br><span class="line">-s：酶切类型, HindIII(AAGCTAGCTT), MboI(GATCGATC) , DpnII(GATCGATC), NcoI(CCATGCATGG)</span><br><span class="line">-z : 参考基因组文件</span><br><span class="line">-y: 限制性酶切位点可能出现位置文件</span><br><span class="line">-p: 染色体大小文件</span><br><span class="line">-C: 将原来的文件进行拆分，必须是4的倍数，默认是90000000, 即22.5M reads</span><br><span class="line">-S: 和任务重运行有关，从中途的某一步开始,&quot;merge&quot;, &quot;dedup&quot;, &quot;final&quot;, &quot;postproc&quot; 或 &quot;early&quot;</span><br><span class="line">-d: juicer的目录</span><br><span class="line">-D: juicer scripts的目录</span><br><span class="line">-t: 线程数</span><br></pre></td></tr></table></figure>
<p>运行记录。（5G Hi-C数据 + 900 Mb contigs 耗时6 h。）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">Using ./ref/draft.genome_MboI.txt as site file</span><br><span class="line">(-: Looking for fastq files...fastq files exist</span><br><span class="line">(-: Aligning files matching /PATH/TO/WORK/DIRECTORY/fastq/\*_R\*.fastq*</span><br><span class="line"> to genome draft_genome with site file ./ref/draft.genome_MboI.txt</span><br><span class="line">(-: Created /PATH/TO/WORK/DIRECTORY/splits and /PATH/TO/WORK/DIRECTORY/aligned.</span><br><span class="line">(...  ...)</span><br><span class="line">(-: Postprocessing successfully completed, maps too sparse to annotate or GPUs unavailable (-:</span><br><span class="line">(-: Pipeline successfully completed (-:</span><br><span class="line">Run cleanup.sh to remove the splits directory</span><br><span class="line">Check /PATH/TO/WORK/DIRECTORY/aligned for results</span><br></pre></td></tr></table></figure>
<p>结果：结果文件在aligned目录下，其中“<code>merged_nodups.txt</code>”就是下一步3D-DNA的输入文件之一。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">splits/</span><br><span class="line">├── MySample.fastq.gz_abnorm.sam</span><br><span class="line">├── MySample.fastq.gz_linecount.txt</span><br><span class="line">├── MySample.fastq.gz_norm.txt.res.txt</span><br><span class="line">├── MySample.fastq.gz.sam</span><br><span class="line">├── MySample.fastq.gz.sort.txt</span><br><span class="line">├── MySample.fastq.gz_unmapped.sam</span><br><span class="line">├── MySample_R1.fastq.gz -&gt; /PATH/TO/WORK/DIRECTORY/fastq/MySample_R1.fastq.gz</span><br><span class="line">└── MySample_R2.fastq.gz -&gt; /PATH/TO/WORK/DIRECTORY/fastq/MySample_R2.fastq.gz</span><br><span class="line">aligned/</span><br><span class="line">├── abnormal.sam</span><br><span class="line">├── collisions_nodups.txt</span><br><span class="line">├── collisions.txt</span><br><span class="line">├── dups.txt</span><br><span class="line">├── header</span><br><span class="line">├── inter_30_contact_domains</span><br><span class="line">├── inter_30.hic</span><br><span class="line">├── inter_30_hists.m</span><br><span class="line">├── inter_30.txt</span><br><span class="line">├── inter.hic</span><br><span class="line">├── inter_hists.m</span><br><span class="line">├── inter.txt</span><br><span class="line">├── merged_nodups.txt</span><br><span class="line">├── merged_sort.txt</span><br><span class="line">├── opt_dups.txt</span><br><span class="line">└── unmapped.sam</span><br></pre></td></tr></table></figure>
<p>Note: 测试的过程中遇到报错信息，原来是默认的“/opt/juicer/scripts/”需要根据实际安装路径，通过设定<code>-D</code>参数修改之。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">(-:  Align of /PATH/TO/WORK/DIRECTORY/splits/XXX.fastq.gz.sam done successfully</span><br><span class="line">awk: fatal: can&#x27;t open source file `/opt/juicer/scripts/common/chimeric_blacklist.awk&#x27; for reading (No such file or directory)</span><br><span class="line">***! Failure during chimera handling of /PATH/TO/WORK/DIRECTORY/splits/MySample.fastq.gz</span><br></pre></td></tr></table></figure>
<p>原来是<code>-D</code>参数指定juicer目录（包含script文件夹那个路径），<code>vi juicer.sh</code> 可见到：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">75 # Juicer directory, contains scripts/, references/, and restriction_sites/</span><br><span class="line">76 # can also be set in options via -D</span><br><span class="line">77 juiceDir=&quot;/opt/juicer&quot;</span><br></pre></td></tr></table></figure>
<h3 id="三-运行3d-dna对contigscaffolds进行组装"><a class="markdownIt-Anchor" href="#三-运行3d-dna对contigscaffolds进行组装"></a> 三、运行3D-DNA，对contig/scaffolds进行组装</h3>
<p>使用默认参数进行3D-DNA</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">~/software/3d-dna/run-asm-pipeline.sh ./ref/draft.genome.fa ./aligned/merged_nodups.txt</span><br><span class="line"># 通过-i参数设定允许挂载的最小尺寸</span><br><span class="line">~/software/3d-dna/run-asm-pipeline.sh -i 8000 ./ref/draft.genome.fa ./aligned/merged_nodups.txt</span><br></pre></td></tr></table></figure>
<p>最后输出文件中，包含 final 字样的文件就是我们需要的结果：<code>draft.genome.final.hic</code> 和 <code>draft.genome.final.assembly</code>。将这两个文件从服务器下载到客户端（我是win 10），并导入 JuiceBox 浏览器。</p>
<p>注意，生成的文件中，draft.genome.final.assembly 与 draft.genome.FINAL.assembly 是不同的文件。“FINAL” – chromosome-length scaffolds; “final” – input with all the misjoin correction introduced。<br />
这两个文件： draft.genome.final.hic 和 draft.genome.final.assembly，同时导入 JuiceBox 才可以进行人工手动矫正。我最初错误地导入了 draft.genome.FINAL.assembly 文件，就遇到了问题。</p>
<h3 id="四-juicebox进行手动纠错"><a class="markdownIt-Anchor" href="#四-juicebox进行手动纠错"></a> 四、JuiceBox进行手动纠错</h3>
<p>不少人将软件名写成juicerbox，其实正确写法是juicebox。下载该软件：<a target="_blank" rel="noopener" href="https://github.com/aidenlab/Juicebox/wiki/Download">https://github.com/aidenlab/Juicebox/wiki/Download</a>。直接下载exe可执行程序，不用安装，双击即可进入界面。</p>
<p><img src="https://oscimg.oschina.net/oscnet/d8e47314-9ca7-409e-9293-614ba5f561b3.png" alt="juicebox" /></p>
<p>一般组装错误为：misjoin、translocations、inversions 和 chromosome boundaries。<br />
关于该软件用法，可看视频：油管地址 <a target="_blank" rel="noopener" href="https://www.youtube.com/watch?v=Nj7RhQZHM18">Tutorial video for Juicebox Assembly Tools</a> 或 <a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av65134634">B 站</a></p>
<p>通过<code>File</code>-&gt;<code>open</code>可以导入后缀为.hic的文件<code>draft.genome.final.hic</code>，随后<code>Assembly</code>-&gt;<code>select</code>，选择<code>draft.genome.final.assembly</code>文件，这时候能看到“染色体”组装图（还不是完美的染色体，只是初步的scaffolds排序，但是我的数据里可看到9条染色体雏形，此外还有一部分没有定位到染色体上的contigs/scaffolds，但是从heatmap中可以看到他们中的大部分与某条染色体会有互作）。</p>
<p>怎么手动调整？左上角到右下角这一条线作为对角线，在对角线任选取两个点，就可以形成一个正方形，两点之间的线段就是“染色体”的一段。如果需要移动某个方形区域，按住shift键不动，在需要挪动的染色体区域左上角点击一次，即可看到一个黑色十字固定在此处，随后在右下角再单击一次，再次出现一个黑色十字，松开shift键，两个黑色十字即包围了这一段待调整的染色体，可以适当调整黑框大小。此时将鼠标移到另一个染色体位置，可见到鼠标白色图标变成黑色箭头，单击，上述的黑色框便移动到了这里，完成了移动选取的操作。如果需要对选择进行翻转呢？则只需要将鼠标移动到黑色框的右上角区域，这时候鼠标图标会变成黑色环状，单击，即可使之翻转。</p>
<p>纠错完以后，单击上方功能区<code>Assembly</code>-&gt;<code>Export Assembly</code>，导出<code>draft.genome.final.review.assembly</code>，或命名为<code>genome.review.assembly</code>，用于下一步的分析</p>
<h3 id="五-再次运行3d-dna"><a class="markdownIt-Anchor" href="#五-再次运行3d-dna"></a> 五、再次运行3D-DNA</h3>
<p>Generate a fasta sequence after JBAT review. In addition to the chromosome-length genome sequence (<em>.FINAL.fasta</em>), the pipeline will produce a new <em>.final.hic</em> file for concluding quality control.</p>
<p>会重新生成 draft.genome.rawchrom.* 和 draft.genome.final.* 这些文件，覆盖掉原来的文件。如果要避免覆盖。就需要另起一个目录运行程序。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/software/3d-dna/run-asm-pipeline-post-review.sh -r genome.review.assembly ./ref/draft.genome.fa aligned/merged_nodups.txt</span><br></pre></td></tr></table></figure>
<p>结果文件：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">draft.genome.FINAL.fasta</span><br><span class="line">draft.genome.FINAL.hic</span><br><span class="line">draft.genome.FINAL_asm.superscaf_track.txt</span><br><span class="line">draft.genome.FINAL_asm.scaffold_track.txt</span><br><span class="line">draft.genome.FINAL.assembly</span><br><span class="line">draft.genome.rawchrom.fasta</span><br><span class="line">draft.genome.rawchrom.hic</span><br><span class="line">draft.genome.final_asm.superscaf_track.txt</span><br><span class="line">draft.genome.final_asm.scaffold_track.txt</span><br><span class="line">draft.genome.final.asm</span><br><span class="line">draft.genome.mnd.txt</span><br><span class="line">draft.genome.rawchrom.asm</span><br><span class="line">draft.genome.rawchrom.cprops</span><br><span class="line">draft.genome.60G.final.review.assembly</span><br><span class="line">draft.genome.cprops</span><br></pre></td></tr></table></figure>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/15/Bioinfo-FASTA-FASTQ-tools/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/15/Bioinfo-FASTA-FASTQ-tools/" class="post-title-link" itemprop="url">一些FASTA/FASTQ序列处理工具</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2020-12-15 11:47:04 / Modified: 13:20:23" itemprop="dateCreated datePublished" datetime="2020-12-15T11:47:04+08:00">2020-12-15</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>序列分析的常见操作包括：序列长度统计（包括总长度、N50 length等），序列提取（根据序列名称或ID、根据序列坐标范围），序列格式转换（fasta-&gt;tab，调整fasta每行碱基数，序列翻转），序列重新命名，序列重新排序，序列切割与重新组合。下面的几个工具就像瑞士军刀一样灵活，基本能满足这些分析需求。</p>
<h3 id="1-seqtk"><a class="markdownIt-Anchor" href="#1-seqtk"></a> 1. seqtk</h3>
<p>Seqtk is a fast and lightweight tool for processing sequences in the FASTA or FASTQ format.<br />
It seamlessly parses both FASTA and FASTQ files which can also be optionally compressed by gzip.<br />
To install seqtk,</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/lh3/seqtk.git;</span><br><span class="line">cd seqtk; make</span><br></pre></td></tr></table></figure>
<p>Usage:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">Usage:   seqtk &lt;command&gt; &lt;arguments&gt;</span><br><span class="line">Version: 1.3-r114-dirty</span><br><span class="line"></span><br><span class="line">Command: seq       common transformation of FASTA/Q</span><br><span class="line">         comp      get the nucleotide composition of FASTA/Q</span><br><span class="line">         sample    subsample sequences</span><br><span class="line">         subseq    extract subsequences from FASTA/Q</span><br><span class="line">         fqchk     fastq QC (base/quality summary)</span><br><span class="line">         mergepe   interleave two PE FASTA/Q files</span><br><span class="line">         trimfq    trim FASTQ using the Phred algorithm</span><br><span class="line"></span><br><span class="line">         hety      regional heterozygosity</span><br><span class="line">         gc        identify high- or low-GC regions</span><br><span class="line">         mutfa     point mutate FASTA at specified positions</span><br><span class="line">         mergefa   merge two FASTA/Q files</span><br><span class="line">         famask    apply a X-coded FASTA to a source FASTA</span><br><span class="line">         dropse    drop unpaired from interleaved PE FASTA/Q</span><br><span class="line">         rename    rename sequence names</span><br><span class="line">         randbase  choose a random base from hets</span><br><span class="line">         cutN      cut sequence at long N</span><br><span class="line">         gap       get the gap locations</span><br><span class="line">         listhet   extract the position of each het</span><br></pre></td></tr></table></figure>
<h3 id="2-seqkit"><a class="markdownIt-Anchor" href="#2-seqkit"></a> 2. seqkit</h3>
<p>To install, type <code>conda install -c bioconda seqkit</code>.<br />
Usage:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">SeqKit -- a cross-platform and ultrafast toolkit for FASTA/Q file manipulation</span><br><span class="line"></span><br><span class="line">Version: 0.14.0</span><br><span class="line"></span><br><span class="line">Author: Wei Shen &lt;shenwei356&gt;</span><br><span class="line"></span><br><span class="line">Documents  : http://bioinf.shenwei.me/seqkit</span><br><span class="line">Source code: https://github.com/shenwei356/seqkit</span><br><span class="line">Please cite: https://doi.org/10.1371/journal.pone.0163962</span><br><span class="line"></span><br><span class="line">Usage:</span><br><span class="line">  seqkit [command]</span><br><span class="line"></span><br><span class="line">Available Commands:</span><br><span class="line">  amplicon        retrieve amplicon (or specific region around it) via primer(s)</span><br><span class="line">  bam             monitoring and online histograms of BAM record features</span><br><span class="line">  common          find common sequences of multiple files by id/name/sequence</span><br><span class="line">  concat          concatenate sequences with same ID from multiple files</span><br><span class="line">  convert         convert FASTQ quality encoding between Sanger, Solexa and Illumina</span><br><span class="line">  duplicate       duplicate sequences N times</span><br><span class="line">  faidx           create FASTA index file and extract subsequence</span><br><span class="line">  fish            look for short sequences in larger sequences using local alignment</span><br><span class="line">  fq2fa           convert FASTQ to FASTA</span><br><span class="line">  fx2tab          convert FASTA/Q to tabular format (with length/GC content/GC skew)</span><br><span class="line">  genautocomplete generate shell autocompletion script</span><br><span class="line">  grep            search sequences by ID/name/sequence/sequence motifs, mismatch allowed</span><br><span class="line">  head            print first N FASTA/Q records</span><br><span class="line">  help            Help about any command</span><br><span class="line">  locate          locate subsequences/motifs, mismatch allowed</span><br><span class="line">  mutate          edit sequence (point mutation, insertion, deletion)</span><br><span class="line">  pair            match up paired-end reads from two fastq files</span><br><span class="line">  range           print FASTA/Q records in a range (start:end)</span><br><span class="line">  rename          rename duplicated IDs</span><br><span class="line">  replace         replace name/sequence by regular expression</span><br><span class="line">  restart         reset start position for circular genome</span><br><span class="line">  rmdup           remove duplicated sequences by id/name/sequence</span><br><span class="line">  sample          sample sequences by number or proportion</span><br><span class="line">  sana            sanitize broken single line fastq files</span><br><span class="line">  scat            real time recursive concatenation and streaming of fastx files</span><br><span class="line">  seq             transform sequences (revserse, complement, extract ID...)</span><br><span class="line">  shuffle         shuffle sequences</span><br><span class="line">  sliding         sliding sequences, circular genome supported</span><br><span class="line">  sort            sort sequences by id/name/sequence/length</span><br><span class="line">  split           split sequences into files by id/seq region/size/parts (mainly for FASTA)</span><br><span class="line">  split2          split sequences into files by size/parts (FASTA, PE/SE FASTQ)</span><br><span class="line">  stats           simple statistics of FASTA/Q files</span><br><span class="line">  subseq          get subsequences by region/gtf/bed, including flanking sequences</span><br><span class="line">  tab2fx          convert tabular format to FASTA/Q format</span><br><span class="line">  translate       translate DNA/RNA to protein sequence (supporting ambiguous bases)</span><br><span class="line">  version         print version information and check for update</span><br><span class="line">  watch           monitoring and online histograms of sequence features</span><br></pre></td></tr></table></figure>
<h3 id="3-itools-reseqtools"><a class="markdownIt-Anchor" href="#3-itools-reseqtools"></a> 3. iTools (Reseqtools)</h3>
<p>To install, use <code>git clone git@github.com:BGI-shenzhen/Reseqtools.git</code>.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">Program: iTools (ReSeqtools)</span><br><span class="line">Version: 0.25   author:hewm2008      Oct 23 2018</span><br><span class="line"></span><br><span class="line">        Usage:</span><br><span class="line"></span><br><span class="line">                Fatools        Tools For Fasta</span><br><span class="line">                Fqtools        Tools For Fastq</span><br><span class="line">                SOAPtools      Tools For SOAP </span><br><span class="line">                Vartools       Tools For SOAP Variant</span><br><span class="line">                CNStools       Tools For CNS</span><br><span class="line">                Xamtools       Tools For Sam/Bam</span><br><span class="line">                Gfftools       Tools For Gff</span><br><span class="line">                Formtools      Tools For Form convert</span><br><span class="line">                Filetools      Tools For Specified File</span><br><span class="line">                Othertools     Tools For Other</span><br><span class="line">                Gametools      Tools For Game</span><br></pre></td></tr></table></figure>
<h3 id="4-ngsqctoolkit"><a class="markdownIt-Anchor" href="#4-ngsqctoolkit"></a> 4. NGSQCToolkit</h3>
<p>A toolkit for the quality control (QC) of next generation sequencing (NGS) data. The toolkit comprises of user-friendly stand alone tools for quality control of the sequence data generated using Illumina and Roche 454 platforms with detailed results in the form of tables and graphs, and filtering of high-quality sequence data. It also includes few other tools, which are helpful in NGS data quality control and analysis.<br />
原下载链接已经失效：<a target="_blank" rel="noopener" href="http://www.nipgr.ac.in/ngsqctoolkit.html">http://www.nipgr.ac.in/ngsqctoolkit.html</a><br />
我把它上传到GitHub了：<code>git clone git@github.com:bioinfx/NGSQCToolkit.git</code></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">perl ~/local/app/NGSQCToolkit/Statistics/N50Stat.pl   </span><br><span class="line">Usage: perl /home/fenglei/local/app/NGSQCToolkit/Statistics/N50Stat.pl &lt;options&gt;</span><br><span class="line"></span><br><span class="line">\### Input reads/sequences (FASTA) (Required)</span><br><span class="line">  -i &lt;Read/Sequence file&gt;</span><br><span class="line">    Read/Sequence in fasta format</span><br><span class="line">\### Other options [Optional]</span><br><span class="line">  -h | -help</span><br><span class="line">    Prints this help</span><br><span class="line">  -o | -outputFile &lt;Output file name&gt;</span><br><span class="line">    Output will be stored in the given file</span><br><span class="line">    default: By default, N50 statistics file will be stored where the input file is</span><br></pre></td></tr></table></figure>
<h3 id="5-fastx-toolkit"><a class="markdownIt-Anchor" href="#5-fastx-toolkit"></a> 5. FASTX-Toolkit</h3>
<p><a target="_blank" rel="noopener" href="http://hannonlab.cshl.edu/fastx_toolkit/">http://hannonlab.cshl.edu/fastx_toolkit/</a><br />
Usage:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">FASTQ-to-FASTA converter</span><br><span class="line">	Convert FASTQ files to FASTA files.</span><br><span class="line">FASTQ Information</span><br><span class="line">	Chart Quality Statistics and Nucleotide Distribution</span><br><span class="line">FASTQ/A Collapser</span><br><span class="line">	Collapsing identical sequences in a FASTQ/A file into a single sequence (while maintaining reads counts)</span><br><span class="line">FASTQ/A Trimmer</span><br><span class="line">	Shortening reads in a FASTQ or FASTQ files (removing barcodes or noise).</span><br><span class="line">FASTQ/A Renamer</span><br><span class="line">	Renames the sequence identifiers in FASTQ/A file.</span><br><span class="line">FASTQ/A Clipper</span><br><span class="line">	Removing sequencing adapters / linkers</span><br><span class="line">FASTQ/A Reverse-Complement</span><br><span class="line">	Producing the Reverse-complement of each sequence in a FASTQ/FASTA file.</span><br><span class="line">FASTQ/A Barcode splitter</span><br><span class="line">	Splitting a FASTQ/FASTA files containning multiple samples</span><br><span class="line">FASTA Formatter</span><br><span class="line">	changes the width of sequences line in a FASTA file</span><br><span class="line">FASTA Nucleotide Changer</span><br><span class="line">	Convets FASTA sequences from/to RNA/DNA</span><br><span class="line">FASTQ Quality Filter</span><br><span class="line">	Filters sequences based on quality</span><br><span class="line">FASTQ Quality Trimmer</span><br><span class="line">	Trims (cuts) sequences based on quality</span><br><span class="line">FASTQ Masker</span><br><span class="line">	Masks nucleotides with &#x27;N&#x27; (or other character) based on quality</span><br></pre></td></tr></table></figure>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/12/Plant-Soybean-introduction/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/12/Plant-Soybean-introduction/" class="post-title-link" itemprop="url">大豆产业现状</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2020-12-12 12:27:45 / Modified: 19:51:16" itemprop="dateCreated datePublished" datetime="2020-12-12T12:27:45+08:00">2020-12-12</time>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>全球大豆年产量 <a target="_blank" rel="noopener" href="http://www.mczx.agri.cn/scts/gxyb/201910/t20191008_7206083.htm">3.42 亿吨</a>，主要为表中的6个国家生产。2019年全年，中国大豆播种面积 <a target="_blank" rel="noopener" href="http://www.stats.gov.cn/tjsj/zxfb/201912/t20191206_1716156.html">1.40 亿亩</a>（约 2300 万英亩，或 9.33 万平方公里，或 933 万公顷），产量 <a target="_blank" rel="noopener" href="http://www.cngrain.com/Publish/1/660407.html">1810 万吨</a>。中国 2019 年进口大豆  <a target="_blank" rel="noopener" href="https://www.thepaper.cn/newsDetail_forward_6206738">8851 万吨</a>，从巴西、美国、阿根廷三国进口大豆比例分别为 <a target="_blank" rel="noopener" href="http://pdf.dfcfw.com/pdf/H3_AP202004121377900207_1.PDF">65.1%、19.1%、9.9%</a>。也就是说，2019年全年中国自产加进口的大豆总量是 10661 万吨，占世界全年产量的 31.17%。</p>
<p><strong>为什么中国需要这么多大豆？</strong> 大豆主要用途是什么？<a target="_blank" rel="noopener" href="http://www.gov.cn/xinwen/2020-06/22/content_5520960.htm">我国是世界最大的大豆消费国，每年大豆消费量1亿吨以上，其中进口大豆消费量9000多万吨。目前，我国大豆消费已形成了食用和压榨两个相对独立的市场。其中，国产大豆主要用于直接食用、加工豆制品和大豆蛋白；进口大豆主要用于满足国内植物油和蛋白粕需求。</a> <a target="_blank" rel="noopener" href="https://www.yicai.com/news/100377221.html">国内居民对动物源蛋白需求增加，刺激了我国养殖行业快速增长，而动物饲料中的重要成分就是豆粕。</a><a target="_blank" rel="noopener" href="https://www.sohu.com/a/340624292_100032755">全球 1.2 亿吨猪肉产量，欧盟 2000 多万吨，中国是 5400 多万吨，美国不到 1200 万吨，我们占接近50%，欧盟占20%，美国占不到10%。目前我国人均年消费猪肉量为38kg，全年国内猪肉消费量大概在5500万吨</a>。</p>
<p><strong>大豆的单位面积产值并不高</strong> 大豆净利润远低于玉米：2012年，黑龙江省大豆每亩净利润256元，而玉米则为495元，粳稻为695元——<a target="_blank" rel="noopener" href="http://ww2.usc.cuhk.edu.hk/PaperCollection/Details.aspx?id=9316">难怪该地越来越多农民不愿意种大豆</a>。</p>
<p><strong>进口大豆价格</strong> “<a target="_blank" rel="noopener" href="http://chinawto.mofcom.gov.cn/article/ap/p/201907/20190702882188.shtml">2019年1至6月（截止到6月21日）大豆价格维持在325美元/吨左右，仍在2009年以来的低位徘徊，刚刚超过308美元/吨（约为每千克2.1元人民币）的成本线。</a>” 2019年4月4日，中国发布对美国的关税反制措施，将对原产于美国 106 项商品加征关税，其中大豆、玉米、小麦、牛肉、汽车、飞机和部分化学品等在列。同时，商务部公告称，中国加征关税的商品金额约为 500 亿美元。大豆是中美贸易的重要组成部分，此次公告对进口自美国的大豆加征 25%关税，标志着中美贸易摩擦再次升级。“<a target="_blank" rel="noopener" href="http://pg.jrj.com.cn/acc/Res/CN_RES/INDUS/2018/4/10/573a9dd8-293d-4246-a7e8-13c9f1a53b57.pdf">目前美豆到岸完税价约为 3300~3400 元/吨，根据测算，加征 25%关税之后到岸完税价提高至 4100 元/吨左右，进口成本增加 700 多元/吨（上涨幅度 22%左右）。</a>” <a target="_blank" rel="noopener" href="http://www.gov.cn/xinwen/2020-06/22/content_5520960.htm">2020年6月16日，9月份船期美湾大豆到我国港口完税成本约为每吨3190元，比1月初每吨下跌170元。国产大豆与进口大豆价格差距进一步拉大，目前国产大豆价格比进口大豆每吨高2000多元。</a></p>
<p><strong>国产大豆价格</strong> <a target="_blank" rel="noopener" href="http://www.gov.cn/xinwen/2020-06/22/content_5520960.htm">2020年6月12日，中储粮油脂公司委托中储粮网竞价销售储备轮换国产大豆60486吨，起拍价格为每吨4800元，最低成交价5140元，最高成交价5270元。</a></p>
<p><strong>大豆亩产量</strong> 中国大豆亩产量比美国低。“<a target="_blank" rel="noopener" href="https://www.guancha.cn/economy/2019_05_29_503615.shtml?s=zwyxgtjdt">根据我国大豆振兴计划，到2020年，全国大豆平均亩产达到135公斤，到2022年，全国大豆平均亩产达到140公斤，逐步缩小与世界大豆主产国的单产差距。</a>” <a target="_blank" rel="noopener" href="https://world.huanqiu.com/article/9CaKrnKbg1t">在整个黑龙江垦区，大豆平均亩产为167.5公斤，距离美国大豆的亩产200公斤差距已经不大。</a>根据媒体报道的大豆总产量计算，中国2019年大豆种植面积 1.4 亿亩，年产量 1810 万吨，平均亩产 129 公斤，按照 5000 元/吨售出价计算，合646元人民币。美国大豆种植面积由3100万公顷增至3566万公顷，年均增长1.8%；产量由9066万吨增至1.24亿吨，年均增长4%。美国2019年大豆种植面积料为 <a target="_blank" rel="noopener" href="http://chinawto.mofcom.gov.cn/article/ap/p/201907/20190702882188.shtml">3566万公顷</a>（5.35亿亩），产量 1.24 亿吨，平均亩产 232 公斤。</p>
<p>如果中国想要完全由自己来生产每年所消费的大豆，每年需要另外播种多少亩大豆，占用全国总播种面积多少？每年需要另外播种约四亿亩以上的大豆，占到全国总播种面积的15％～20％。这意味着必须减少如此幅度的谷物或高值新农业产品，实际上根本就不可考虑。这是国内大豆产量没有显著增加而大豆进口快速增长的基本原因。</p>
<p>2019年2月19日，新华社受权发布《中共中央、国务院关于坚持农业农村优先发展做好“三农”工作的若干意见》要求，严守<a target="_blank" rel="noopener" href="https://www.yicai.com/news/100121079.html">18亿亩耕地红线</a>，全面落实永久基本农田特殊保护制度，确保永久基本农田保持在15.46亿亩以上。按照18亿亩耕地的红线，这个面积高达120万平方公里。<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/wtNo0KQmHdWkv-gzm81dNg">一亩土地是666.67平米，乘以18亿亩，刚好是120万平方公里整数</a>，约等于西藏的面积。</p>
<p>如果按照当前中国大豆亩产量 129 公斤/亩来计算，如果需要满足国内总需求量 10661 万吨的生产，需要耕地面积 8.2 亿亩，合 54.67 万平方公里，比整个黑龙江省的面积（47.7 万平方公里）略大。</p>
<p><strong>中国和美国的大豆有什么不一样？</strong><br />
学者曾比较过中美大豆生产的投入费用。根据其分析，2009年至2012年，中国大豆生产的物质和人工平均费用，分别为166.96、133.18元；美国大豆生产的物质和人工平均费用，分别是236.54、21.26元。美国大豆生产的物质费用比中国高69.58元，人工费用比中国却低111.92元。<a target="_blank" rel="noopener" href="https://news.fx168.com/opinion/column/hujiajun/1806/2541828.shtml">两项合计，美国比中国大豆生产成本整体上低了42.34元。</a> 更为关键的是，美国大豆生产不仅投入成本低，亩产量还比中国高。2009年至2012年，中国大豆亩产量平均为142.46公斤，美国大豆亩产量平均则为201.32公斤，较中国高出58.86公斤。一进一出，美国大豆取得了绝对的价格优势。</p>
<p><strong>美国大豆单位面积产量为什么比中国大豆更高？</strong></p>
<p><strong>转基因大豆的原理</strong><br />
1976年，美国孟山都公司（Monsanto Company）推出一款高效除草剂，名为农达草甘膦（Roundup）。草甘膦能够抑制深根杂草生长，大大减少农作物的田间管理强度。对于大规模农场来说，农达是一个巨大福音。但是，具体到大豆农户，则喜忧参半，因为草甘膦在消灭杂草的同时，也会杀死豆苗。为了解决这个难题，孟山都公司又组织科学家，利用前期的转基因研究，1993年培育出了抗农达大豆品种。农户买了他们的大豆种子，即使喷洒草甘膦，也不会影响豆苗生长。</p>
<p>大豆基础研究：</p>
<ol>
<li>增加产量</li>
<li>提高抗虫性</li>
<li>提高含油量</li>
<li>提高植物蛋白含量</li>
</ol>
<table>
<thead>
<tr>
<th>国家</th>
<th>种植面积（英亩）</th>
<th>产量</th>
<th>出口量</th>
<th>进口量</th>
</tr>
</thead>
<tbody>
<tr>
<td>巴西</td>
<td>9509万</td>
<td>1.2500亿吨</td>
<td>7700万吨</td>
<td>-</td>
</tr>
<tr>
<td>美国</td>
<td>8670万</td>
<td>1.2400亿吨</td>
<td>4967万吨</td>
<td>-</td>
</tr>
<tr>
<td>阿根廷</td>
<td>4820万</td>
<td>5100万吨</td>
<td>820万吨</td>
<td>-</td>
</tr>
<tr>
<td>中国</td>
<td>2187万</td>
<td>1800万吨</td>
<td>-</td>
<td>8859万吨</td>
</tr>
<tr>
<td>印度</td>
<td>2680万</td>
<td>1346万吨</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td>巴拉圭</td>
<td>940万</td>
<td>1030万吨</td>
<td>630万吨</td>
<td>-</td>
</tr>
<tr>
<td>其他</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>-</td>
</tr>
<tr>
<td>总计</td>
<td>30641万</td>
<td>3.42亿吨</td>
<td>1.5亿吨</td>
<td>-</td>
</tr>
</tbody>
</table>
<p><img src="http://img.zhitongcaijing.com/images/contentformat/10a8eb25dc1e1cc86384b6337f9cb2f3.jpg" alt="大豆种植面积" /></p>
<p><img src="http://img.zhitongcaijing.com/images/contentformat/42907666545e4dd2a32c1065f9eda879.jpg" alt="大豆产量" /></p>
<p><img src="http://www.dwjmsw.com/Uploads/Article/image/20191118/20191118100510_56129.jpg" alt="中国大豆产量、进口量、消费量" /></p>
<p><img src="http://img.zhitongcaijing.com/images/contentformat/0905bdc7968f303cb5c4ca1f54d74f4a.jpg" alt="中国进口大豆来源" /></p>
<h3 id="参考资料"><a class="markdownIt-Anchor" href="#参考资料"></a> 参考资料</h3>
<ol>
<li><a target="_blank" rel="noopener" href="https://www.zhitongcaijing.com/content/detail/162164.html">西部证券研究报告：大豆之争，渔翁得利 2018-11-07</a></li>
<li><a target="_blank" rel="noopener" href="http://pdf.dfcfw.com/pdf/H3_AP202004121377900207_1.PDF">全球大豆价格还能一路上涨吗？2020-04-11</a></li>
<li><a target="_blank" rel="noopener" href="http://chinawto.mofcom.gov.cn/article/ap/p/201907/20190702882188.shtml">农民日报：美国大豆产业的发展现状如何？2019-07-16</a></li>
<li><a target="_blank" rel="noopener" href="http://www.xzfutures.com/files/document/20170510/6363000333771339151814280.pdf">兴证期货：大豆气象周报 2017-05-10</a></li>
<li><a href="">《开放时代》2014年第1期：大豆生产和进口的经济逻辑 2014-01-01</a></li>
<li><a target="_blank" rel="noopener" href="https://www.guancha.cn/economy/2019_05_29_503615.shtml?s=zwyxgtjdt">国产大豆振兴 2020年大豆种植面积将达到1.4亿亩 </a></li>
<li><a target="_blank" rel="noopener" href="https://www.qianzhan.com/analyst/detail/220/191120-8ccd4a58.html">前瞻经济学人：2019年全球大豆行业市场发展现状分析 全球大豆产量稳步增长 2019-11-21</a></li>
<li><a target="_blank" rel="noopener" href="http://finance.ifeng.com/topic/money/ddzz/index.shtml">凤凰网财经专题：大豆战争 2012年</a></li>
<li><a target="_blank" rel="noopener" href="http://www.cciee.org.cn/archiver/cciee/UpFile/Files/qqh/20191009140421154928.pdf">《全球化》2019年第5期：中国大豆产业发展：主要问题、原因以及对策建议</a></li>
<li><a target="_blank" rel="noopener" href="http://jjckb.xinhuanet.com/2010-05/07/content_220528.htm">新华网：中洋大豆之争 实乃产业链的抗衡 2010-05-07</a></li>
<li><a target="_blank" rel="noopener" href="http://www.cqvip.com/qk/91584x/201009/33584422.html">《求是》2010年 第9期 | 万宝瑞: 振兴我国大豆产业的根本途径</a></li>
<li><a target="_blank" rel="noopener" href="https://www.yicai.com/news/345460.html">第一财经：我国大豆产业急需发展专业合作社 2010-05-05</a></li>
<li><a target="_blank" rel="noopener" href="https://news.fx168.com/opinion/column/hujiajun/1806/2541828.shtml">财经报社：美国大豆是如何“征服”世界的？ 2018-06-09</a></li>
</ol>
<h3 id="附注"><a class="markdownIt-Anchor" href="#附注"></a> 附注</h3>
<p><strong>草甘膦的作用机理</strong></p>
<p>草甘膦是由著名的美国公司孟山都（Monsanto）研制成功的灭生性茎叶处理除草剂，具有高效 、广谱 、低毒（草甘膦的急性经口毒性比食盐都低，但后来有报道草甘膦具有肾毒性，我还没有认真查找相关的学术报道，先不做评论。）、低残留 、易于被微生物分解 ,不破坏土壤环境 ,对大多数植物具有灭生性等优点。后来又得益于抗草甘膦转基因作物的发展和推广，这种除草剂一直稳坐世界第一大农药的宝座。    草甘膦作为一种非选择性除草剂，它的作用机理是竞争性地抑制莽草酸途径中 5 - 烯醇丙酮莽草酸 - 3 - 磷酸合成酶 ( EPSPS)的活性 。该合成酶是真菌 、细菌 、藻类和高等植物体内芳香族氨基酸 (包括色氨酸 、酪氨酸 、苯丙氨酸 )生物合成过程中一个关键性的酶。具体的生理生化过程见下图。用草甘膦处理植物以后，植物体内含有这类氨基酸的蛋白质合成就会受阻，进而影响一系列代谢过程，导致植物死亡。草甘膦使用时一般将其制成异丙胺盐或钠盐。草甘膦除草性能优异，极易通过叶面少量通过根吸收，并运送到植物生长点。<br />
<img src="https://pic2.zhimg.com/80/fccf586c8c431cd3b71435d878b4851c_720w.jpg?source=1940ef5c" alt="" /></p>
<p><strong>转基因大豆抵御草甘膦的机理</strong></p>
<p>目前研究者主要从三个方面入手来改善作物的抗草甘膦特性：1. 通过使草甘膦抑制的酶过量表达，进而让植物吸收草甘膦之后能够进行正常代谢。但是目前用这个方法得到的抗草甘膦作物暴露在草甘膦之下时仍表现出生长抑制，未达到能够商业化的水平。2. 引入酶系统，在草甘膦进入植物体后迅速将其降解。目前这一策略也没有在生产上大量应用。3. 而应用于生产实践的上的主要方法是对草甘膦作用的酶进行修饰，降低它对草甘膦的敏感性。这里插一句背景知识，生物体内的酶大部分都是由氨基酸按照一定顺序排列组成的大分子物质，具有一定的空间结构，能够由特定酶催化，和特定酶反应的分子需要和这种酶的结构位点相结合才能发挥作用。改变了草甘膦作用的酶的结构，就可以降低甚至消除草甘膦对这种酶的作用效力。目前研究者们从大肠杆菌( <em>Escherichia coli</em>)、沙门氏杆菌( <em>Salmonella typhimurium</em>)和农杆菌( <em>Agrobacterium tumefaciens</em>) 等微生物中分离出了抗草甘膦的突变基因，而应用于生产的是从农杆菌菌株( <em>A. tumefaciens</em> sp. CP4) 上分离出的CP4基因。在大豆上导入了这种抗草甘膦基因，在大豆植株中表达出的CP4-EPSP合成酶对草甘膦的敏感性比较低，使得转基因植物体内的莽草酸途径可以正常进行。这样以来大豆就表现出了对草甘膦的抗性。</p>
<p><strong>转基因植物抗虫的机理</strong><br />
以转基因Bt抗虫作物为例, Bt蛋白为什么能够杀死昆虫？Bt进入昆虫的消化道, 昆虫的消化道是碱性环境, Bt蛋白和昆虫消化道上的受体结合, 引起肠穿孔, 从而导致害虫生长缓慢, 直至死亡。人的消化道结构和昆虫不一样, 而且人的肠道环境是酸性的, Bt蛋白进入胃中后很快在胃液中被消化降解为氨基酸, 对人体消化道没有任何作用。也就是说, Bt基因的杀虫功能具有很强的选择性, 只在鳞翅目昆虫的碱性消化道中发挥作用, 对人和哺乳动物都不会产生影响(Sheltonet al., 2002)。抗除草剂性状由于其巨大的经济效益和环境效益, 成为作物的第一大改良性状。2011年抗除草剂大豆、玉米、油菜、棉花、甜菜以及苜蓿种植面积达9 390万公顷, 占全球转基因作物种植总面积的59%。其中使用最为广泛的抗除草剂基因是从根瘤农杆菌中分离的编码455个氨基酸的抗草甘膦基因CP4-EP- SPS。安全性评价表明, CP4-EPSPS蛋白在哺乳动物消化系统中极易分解, 小鼠经口食入高达572 mg·kg–1(高于转基因大豆、玉米等植物中的1 300倍)的CP4-EPSPS蛋白, 亦无不良影响, 其在胃中的半衰期小于15秒, 在肠系统中小于10分钟。以上结果结合种子成分分析及动物营养研究表明, 转基因抗草甘膦大豆对人和动物是安全的(Harrison et al., 1998)。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/07/Life-initial-demo-page/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/07/Life-initial-demo-page/" class="post-title-link" itemprop="url">My Fast HexoPages</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-07 16:57:42" itemprop="dateCreated datePublished" datetime="2020-12-07T16:57:42+08:00">2020-12-07</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-08 14:09:25" itemprop="dateModified" datetime="2020-12-08T14:09:25+08:00">2020-12-08</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Life/" itemprop="url" rel="index"><span itemprop="name">Life</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>abstract</p>
          <!--noindex-->
            <div class="post-button">
              <a class="btn" href="/2020/12/07/Life-initial-demo-page/#more" rel="contents">
                Read more &raquo;
              </a>
            </div>
          <!--/noindex-->
        
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/06/Bioinfo-HiC-pro-installation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/06/Bioinfo-HiC-pro-installation/" class="post-title-link" itemprop="url">安装HiC-Pro （HiC数据预处理工具）与绘制HiC染色体互作图谱</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-06 13:17:53" itemprop="dateCreated datePublished" datetime="2020-12-06T13:17:53+08:00">2020-12-06</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-05-29 17:27:59" itemprop="dateModified" datetime="2021-05-29T17:27:59+08:00">2021-05-29</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>HiC-Pro可以用来处理Hi-C数据，从原始的fastq文件(配对端Illumina数据)到标准化的交互图谱。简单的来说就是将Hi-C数据比对到拼装好的参考基因组上，并形成交互文件去存储Hi-C数据。</p>
<p>下载 HiC-Pro VERSION 2.11.4 安装包，依赖 Python 2。（后续更新：HiC-Pro version 3.0.0 要求 Python 3）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/nservant/HiC-Pro.git </span><br></pre></td></tr></table></figure>
<h3 id="安装-hic-pro-version-2114"><a class="markdownIt-Anchor" href="#安装-hic-pro-version-2114"></a> 安装 HiC-Pro VERSION 2.11.4</h3>
<p>由于需要python 2 （2.7以上），我之前已经在系统安装Anaconda3，并在里面分别安装了python3.7和python2.7，(<a target="_blank" rel="noopener" href="https://genehub.wordpress.com/2020/04/21/1070/">原文链接</a>, 这里只需要切换到python2.7环境即可。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">source activate python27</span><br></pre></td></tr></table></figure>
<p>再直接用conda命令安装相关软件</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">conda install -y samtools bowtie2 R </span><br><span class="line">conda install -y pysam bx-python numpy scipy </span><br><span class="line">#打开R </span><br><span class="line">install.packages(c(&#x27;ggplot2&#x27;,&#x27;RColorBrewer&#x27;)) </span><br></pre></td></tr></table></figure>
<p>通过 conda 安装 samtools，samtools 运行会报错。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools: error while loading shared libraries: libcrypto.so.1.0.0: cannot open shared object file: No such file or directory</span><br></pre></td></tr></table></figure>
<p>现在samtools的版本已经在1.9以上了，但是conda安装的samtools版本依然是1.7，这是出现问题的原因。<a target="_blank" rel="noopener" href="https://blog.csdn.net/zhangjunya/article/details/108235796">解决办法</a>：指定另一个更新的版本，重新安装如下。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -c bioconda samtools=1.9 --force-reinstall</span><br></pre></td></tr></table></figure>
<p>或者刚开始安装的时候：<code>conda install samtools=1.9</code>，就不会再出现上述问题。</p>
<p>不知道为什么bx-python安装失败，换成下面的方法，可以安装。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -c conda-forge -c bioconda bx-python</span><br></pre></td></tr></table></figure>
<p>然后进入HiC-Pro目录，修改config-install 文件，设置依赖软件的路径。PREFIX不要设置当前目录，会被覆盖。CLUSTER_SYS有四个选项，分别是TORQUE, SGE, SLURM 或 LSF，我随便填了一个TORQUE。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#</span><br><span class="line"> Paths and Settings  - Start editing here !</span><br><span class="line"> #</span><br><span class="line"> PREFIX = /home/user/local/scripts/HiC-Pro</span><br><span class="line"> BOWTIE2_PATH = /home/user/local/app/anaconda3/envs/python27/bin/</span><br><span class="line"> SAMTOOLS_PATH = /home/user/local/bin/</span><br><span class="line"> R_PATH = /home/user/local/app/anaconda3/envs/python27/bin</span><br><span class="line"> PYTHON_PATH = /home/user/local/app/anaconda3/envs/python27/bin</span><br><span class="line"> CLUSTER_SYS = TORQUE</span><br></pre></td></tr></table></figure>
<p>接下里 make configure ，通过之后屏幕显示如下信息</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">(python27) xxxxx/HiC-Pro$ make configure       </span><br><span class="line"> </span><br><span class="line">make -f ./scripts/install/Makefile CONFIG_SYS=./config-install.txt</span><br><span class="line"> make[1]: Entering directory &#x27;/home/-/local/app/HiC-Pro&#x27;</span><br><span class="line"> ./scripts/install/install_dependencies.sh -c ./config-install.txt -o  /home/-/local/scripts/HiC-Pro/HiC-Pro_2.11.4 -q</span><br><span class="line"> Make sure internet connection works for your shell prompt under current user&#x27;s privilege …</span><br><span class="line"> Starting HiC-Pro installation …</span><br><span class="line"> export /home/--/local/app/anaconda3/envs/python27/bin/ in PATH</span><br><span class="line"> export /home/-/local/bin/ in PATH</span><br><span class="line"> export /home/-/local/app/anaconda3/envs/python27/bin in PATH</span><br><span class="line"> export /home/-/local/app/anaconda3/envs/python27/bin in PATH</span><br><span class="line"> Checking dependencies … </span><br><span class="line"> Checking Python libraries …</span><br><span class="line"> The required Python libraries appear to be already installed. </span><br><span class="line"> Checking R installation …</span><br><span class="line"> The required R packages appear to be already installed. </span><br><span class="line"> Bowtie2 Aligner appears to be already installed. </span><br><span class="line"> Samtools appears to be already installed. </span><br><span class="line"> Dependencies checked !</span><br><span class="line"> Check HiC-Pro configuration … </span><br><span class="line"> Configuration for TORQUE/PBS system.</span><br><span class="line"> make[1]: Leaving directory &#x27;/home/-/local/app/HiC-Pro&#x27;</span><br></pre></td></tr></table></figure>
<p>然后 make，通过即安装成功。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">(some information are not shown here due to size limitation)</span><br><span class="line">cp -Ri /home/-/local/app/HiC-Pro /home/-/local/scripts/HiC-Pro/HiC-Pro_2.11.4 </span><br><span class="line">HiC-Pro installed in /home/-/local/scripts/HiC-Pro/HiC-Pro_2.11.4 !</span><br></pre></td></tr></table></figure>
<p>测试</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">(python27) XXX:~/local/scripts/HiC-Pro/HiC-Pro_2.11.4/bin$ ./HiC-Pro -h</span><br><span class="line"> usage : HiC-Pro -i INPUT -o OUTPUT -c CONFIG [-s ANALYSIS_STEP] [-p] [-h] [-v]</span><br><span class="line"> Use option -h--help for more information</span><br><span class="line"> HiC-Pro 2.11.4</span><br><span class="line"> OPTIONS</span><br><span class="line"> -i--input INPUT : input data folder; Must contains a folder per sample with input files</span><br><span class="line">    -o--output OUTPUT : output folder</span><br><span class="line">    -c--conf CONFIG : configuration file for Hi-C processing</span><br><span class="line">    [-p--parallel] : if specified run HiC-Pro on a cluster</span><br><span class="line">    [-s--step ANALYSIS_STEP] : run only a subset of the HiC-Pro workflow; if not specified the complete workflow is run</span><br><span class="line">       mapping: perform reads alignment - require fast files</span><br><span class="line">       proc_hic: perform Hi-C filtering - require BAM files</span><br><span class="line">       quality_checks: run Hi-C quality control plots</span><br><span class="line">       merge_persample: merge multiple inputs and remove duplicates if specified - require .validPairs files</span><br><span class="line">       build_contact_maps: Build raw inter/intrachromosomal contact maps - require .allValidPairs files</span><br><span class="line">       ice_norm : run ICE normalization on contact maps - require .matrix files</span><br></pre></td></tr></table></figure>
<h3 id="optional-安装-hic-pro-version-300"><a class="markdownIt-Anchor" href="#optional-安装-hic-pro-version-300"></a> <s>Optional: 安装 HiC-Pro VERSION 3.0.0</s></h3>
<p><s>conda create --name python38 python=3.8</s><br />
<s>source activate python38</s><br />
<s>conda install conda</s><br />
<s># 前面的conda是anaconda3的conda，后面的conda是python3.8的conda</s></p>
<p><s>conda install -c bioconda bowtie2</s><br />
<s>conda install -c bioconda samtools=1.9</s><br />
<s>conda install -c conda-forge -c bioconda bx-python</s></p>
<p><s>在conda里面安装R和pysam遇到很多问题，更换python版本3.7/3.8/3.9都没成功，真是浪费时间，放弃。</s></p>
<h3 id="使用hic-pro对hic数据进行分析的过程"><a class="markdownIt-Anchor" href="#使用hic-pro对hic数据进行分析的过程"></a> 使用HiC-Pro对HiC数据进行分析的过程</h3>
<p>1. 在工作目录（例如HiC-analysis-workdir），建立若干文件夹，例如index、rawdata，方便管理，当然这些数据也可以在其他位置。</p>
<p>2. 在index路径下，对参考基因组或者scaffolds建立index。HiC-Pro调用的比对软件是bowtie2。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bowtie2-build --threads 20 ref.fasta ref.fasta</span><br></pre></td></tr></table></figure>
<p>3. 在工作目录，建立一个文本文件，内含参考基因组（或scaffolds）各个染色体名称和长度的信息。可以用samtools fadix，也可以用iTools或R语言等工具。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ cat chromosome_size.txt</span><br><span class="line"> Chr01   11111</span><br><span class="line"> Chr02   1233333</span><br><span class="line"> Chr03   4444</span><br><span class="line"> Chr04   105314844</span><br><span class="line">(continue ...)</span><br></pre></td></tr></table></figure>
<p>4. 在工作目录，新建一个文件夹（例如rawdata），用来存放作为input的测序数据。注意不能直接将测序文件放到rawdata目录下，而是要按照样品名先建立子文件夹（例如sample1、sample2）。sample1目录下即存放样品的测序文件，包含F/R两个测序文件，文件应该是.fastq(.gz)的命名，不可写成fq(.gz)；可以是软链接的形式。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207122951.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207122951.png?w=506" alt="" /></a></p>
<p>4. 基因组的酶切位点识别。我用的MboI（酶切位点：^GATC）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python2.7 XXX/HiC-Pro/HiC-Pro_2.11.4/bin/utils/digest_genome.py -r ^GATC -o mboi_resfrag_genome.bed Ref_genome.fa</span><br></pre></td></tr></table></figure>
<p>得到的酶切信息：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"> Chr01   0       3282    HIC_Chr01_1     0       +</span><br><span class="line"> Chr01   3282    3609    HIC_Chr01_2     0       +</span><br><span class="line"> Chr01   3609    3625    HIC_Chr01_3     0       +</span><br><span class="line"> Chr01   3625    4485    HIC_Chr01_4     0       +</span><br><span class="line"> Chr01   4485    5312    HIC_Chr01_5     0       +</span><br><span class="line"> Chr01   5312    5327    HIC_Chr01_6     0       +</span><br><span class="line"> Chr01   5327    5958    HIC_Chr01_7     0       +</span><br><span class="line"> Chr01   5958    6475    HIC_Chr01_8     0       +</span><br><span class="line"> Chr01   6475    7545    HIC_Chr01_9     0       +</span><br><span class="line">(continue ...)</span><br></pre></td></tr></table></figure>
<p>5. 将HiC-Pro程序安装包目录下的 config-hicpro.txt 文件拷贝到工作目录，对其进行编辑。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"># N_CPU，例如N_CPU = 24</span><br><span class="line"># JOB_NAME，例如JOB_NAME = Ga</span><br><span class="line"># JOB_WALLTIME，例如JOB_WALLTIME = 11:00:00</span><br><span class="line"># JOB_QUEUE，例如JOB_QUEUE = normal</span><br><span class="line"># JOB_MAIL，例如JOB_MAIL = 1570447120@qq.com</span><br><span class="line"># BOWTIE2_IDX_PATH，填写用bowtie2对reference建立索引的目录，注意是目录，而不是索引路径!如/PATH/to/Ga.ref</span><br><span class="line"># REFERENCE_GENOME，例如Ga 注意：这里REFERENCE_GENOME一定要和bowtie2建立的索引对应上,名字为物种名</span><br><span class="line"># GENOME_SIZE，即生成的文件Ga.fasta.size（最好使用绝对路径）</span><br><span class="line"># GENOME_FRAGMENT，用digest_genome.py程序生成的文件Ga.HindIII.txt</span><br><span class="line"># LIGATION_SITE，酶的序列，例如HindIII，则为AAGCTAGCTT；如果是MboI则序列为GATCGATC。</span><br><span class="line"># MIN_FRAG_SIZE，例如MIN_FRAG_SIZE = 100</span><br><span class="line"># MAX_FRAG_SIZE，例如MAX_FRAG_SIZE = 160000</span><br><span class="line"># MIN_INSERT_SIZE，例如MIN_INSERT_SIZE = 200</span><br><span class="line"># MAX_INSERT_SIZE，例如MAX_INSERT_SIZE = 600</span><br><span class="line"># GET_PROCESS_SAM，例如GET_PROCESS_SAM = 1</span><br></pre></td></tr></table></figure>
<p>6. 运行</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">/PATH/TO/HiC-Pro_2.11.1/bin/HiC-Pro -i /PATH/TO/rawdata --output hicpro_output --conf config-hicpro.txt </span><br><span class="line"># --input reads所在的文件</span><br><span class="line"># --output 输出的文件夹</span><br><span class="line"># --conf 配置文件</span><br></pre></td></tr></table></figure>
<p>7. 运行结果</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br></pre></td><td class="code"><pre><span class="line">.</span><br><span class="line">├── bowtie_results</span><br><span class="line">│   ├── bwt2</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.bwt2merged.bam</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.mapstat</span><br><span class="line">│   │       ├── mysample_R2_V3_20201206.fa.bwt2merged.bam</span><br><span class="line">│   │       ├── mysample_R2_V3_20201206.fa.mapstat</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.bam</span><br><span class="line">│   │       └── _V3_20201206.fa.bwt2pairs.pairstat</span><br><span class="line">│   ├── bwt2_global</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.bwt2glob.bam</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap.fastq</span><br><span class="line">│   │       ├── mysample_R2_V3_20201206.fa.bwt2glob.bam</span><br><span class="line">│   │       └── mysample_R2_V3_20201206.fa.bwt2glob.unmap.fastq</span><br><span class="line">│   └── bwt2_local</span><br><span class="line">│       └── sample1</span><br><span class="line">│           ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_bwt2loc.bam</span><br><span class="line">│           ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_trimmed.fastq</span><br><span class="line">│           ├── mysample_R2_V3_20201206.fa.bwt2glob.unmap_bwt2loc.bam</span><br><span class="line">│           └── mysample_R2_V3_20201206.fa.bwt2glob.unmap_trimmed.fastq</span><br><span class="line">├── config-hicpro.txt</span><br><span class="line">├── hic_results</span><br><span class="line">│   ├── data</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── sample1.allValidPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.DEPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.DumpPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.FiltPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.REPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.RSstat</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.SCPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.SinglePairs</span><br><span class="line">│   │       └── _V3_20201206.fa.bwt2pairs.validPairs</span><br><span class="line">│   ├── matrix</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── iced</span><br><span class="line">│   │       │   └── 200000</span><br><span class="line">│   │       │       ├── sample1_200000_iced.matrix</span><br><span class="line">│   │       │       └── sample1_200000_iced.matrix.biases</span><br><span class="line">│   │       └── raw</span><br><span class="line">│   │           └── 200000</span><br><span class="line">│   │               ├── sample1_200000_abs.bed</span><br><span class="line">│   │               └── sample1_200000.matrix</span><br><span class="line">│   ├── pic</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── plotHiCContactRanges_sample1.pdf</span><br><span class="line">│   │       ├── plotHiCFragment_sample1.pdf</span><br><span class="line">│   │       ├── plotHiCFragmentSize_sample1.pdf</span><br><span class="line">│   │       ├── plotMappingPairing_sample1.pdf</span><br><span class="line">│   │       └── plotMapping_sample1.pdf</span><br><span class="line">│   └── stats</span><br><span class="line">│       └── sample1</span><br><span class="line">│           ├── sample1_allValidPairs.mergestat</span><br><span class="line">│           ├── sample1mysample_R1.mmapstat</span><br><span class="line">│           ├── sample1mysample_R2.mmapstat</span><br><span class="line">│           ├── sample1.mpairstat</span><br><span class="line">│           └── sample1.mRSstat</span><br><span class="line">├── logs</span><br><span class="line">│   └── sample1</span><br><span class="line">│       ├── build_raw_maps.log</span><br><span class="line">│       ├── mysample_R1_bowtie2.log</span><br><span class="line">│       ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_bowtie2.log</span><br><span class="line">│       ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_readsTrimming.log</span><br><span class="line">│       ├── mysample_R2_bowtie2.log</span><br><span class="line">│       ├── mysample_R2_V3_20201206.fa.bwt2glob.unmap_bowtie2.log</span><br><span class="line">│       ├── mysample_R2_V3_20201206.fa.bwt2glob.unmap_readsTrimming.log</span><br><span class="line">│       ├── ice_200000.log</span><br><span class="line">│       ├── make_Rplots.log</span><br><span class="line">│       ├── mapped_2hic_fragments.log</span><br><span class="line">│       ├── mapping_combine.log</span><br><span class="line">│       ├── mapping_stats.log</span><br><span class="line">│       ├── mapping_step1.log</span><br><span class="line">│       ├── mapping_step2.log</span><br><span class="line">│       ├── mergeSAM.log</span><br><span class="line">│       ├── merge_stats.log</span><br><span class="line">│       ├── merge_valid_interactions.log</span><br><span class="line">│       ├── plot_hic_contacts.Rout</span><br><span class="line">│       ├── plot_hic_fragment.Rout</span><br><span class="line">│       ├── plot_mapping_portion.Rout</span><br><span class="line">│       └── plot_pairing_portion.Rout</span><br><span class="line">├── rawdata -&gt; /PATH/TO//rawdata/</span><br><span class="line">└── tmp</span><br></pre></td></tr></table></figure>
<p><strong>画图</strong></p>
<p>HiC-Pro运行结果中，sample1.allValidPairs文件就表示 pair end reads 在染色体或 scaffolds 上面位置的信息，该文件是文本格式，需要进行格式转化成为.hic文件，才能放到 juicebox 软件中得到可视化的互作图谱。</p>
<p>如何转换成.hic文件？使用HiC-Pro安装路径下的<code>hicpro2juicebox.sh</code>程序按照下面的方式运行即可得到.hic文件。<br />
需要下载 juicer_tools_1.22.01.jar 文件作为输入文件。下载地址：<a target="_blank" rel="noopener" href="https://github.com/aidenlab/juicer/wiki/Download">https://github.com/aidenlab/juicer/wiki/Download</a><br />
备注：我一开始下载错了，去<a target="_blank" rel="noopener" href="https://github.com/aidenlab/Juicebox/wiki/Download">https://github.com/aidenlab/Juicebox/wiki/Download</a> 页面下了个Juicebox.jar，结果不能用于下面的步骤。原来这是 JuiceBox，不是 Juicer！从论坛发现遇到这个问题的不止我一个人<a target="_blank" rel="noopener" href="https://groups.google.com/g/hic-pro/c/DF0zfZOivsE?pli=1">https://groups.google.com/g/hic-pro/c/DF0zfZOivsE?pli=1</a>。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">/PATH/TO/HiC-Pro/HiC-Pro_2.11.4/bin/utils/hicpro2juicebox.sh -i /PATH/TO//sample1.allValidPairs -g /PATH/TO//chromosome_sizes -j /PATH/TO/juicer_tools_1.22.01.jar -r /PATH/TO//mboi_resfrag_AMgenome.bed -t temp -o /PATH/TO/output</span><br></pre></td></tr></table></figure>
<p>用上面的命令就得到了 .hic 文件，然后从 Linux 服务器将该文件下载到本地客户端（客户端即我个人电脑用的 windows 系统），直接运行 Juicebox_1.11.08.exe 二进制程序，双击运行，菜单选 open，选择 .hic 文件，就得到下面的图谱，可以 export 成 pdf 或 svg 格式的图片。</p>
<p><img src="https://gblobscdn.gitbook.com/assets%2F-LijQZQE8sc7HbVdoxEA%2F-Lj2gavu4NTsjkOUMYaT%2F-Lj2js_V0JRQoDXKpVvE%2FScreen%20Shot%202019-07-05%20at%2012.51.13%20PM.png?alt=media&amp;token=fdae559d-237e-45dd-a0dd-a12a22ff7506" alt=".hic文件导入juicebox软件之后的示意图" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/05/Bioinfo-compare-two-genome-sequences-by-mummer/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/05/Bioinfo-compare-two-genome-sequences-by-mummer/" class="post-title-link" itemprop="url">两个基因组序列相互比较：Mummer</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-05 21:11:00" itemprop="dateCreated datePublished" datetime="2020-12-05T21:11:00+08:00">2020-12-05</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-02-25 17:42:04" itemprop="dateModified" datetime="2021-02-25T17:42:04+08:00">2021-02-25</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>直接对两个基因组进行比较，用Mummer：</p>
<p>例如，将细菌1（H_pyloriJ99_Eslice.fasta）与细菌2（H_pylori26695_Eslice.fasta）进行比较：</p>
<p>mummer -mum -b -c H_pylori26695_Eslice.fasta H_pyloriJ99_Eslice.fasta &gt; mummer.mums<br />
mummerplot -x [0,275287] -y [0,265111] -postscript -p mummer mummer.mums</p>
<p>结果：</p>
<p><img src="https://genehub.files.wordpress.com/2019/07/mummer.png" alt="mummer.png" /></p>
<p>如果换成两个大基因组呢？800M以上的基因组，结果报错了，尺寸太大，程序默认不支持这么多，只允许400M。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># reading input file &quot;Jxxx.fasta&quot; of length 854295539</span><br><span class="line"># construct suffix tree for sequence of length 854295539</span><br><span class="line"># (maximum reference length is 536870908)</span><br><span class="line"># (maximum query length is 4294967295)</span><br><span class="line"># process 8542955 characters per dot</span><br><span class="line">mummer: suffix tree construction failed: textlen=854295539 larger than maximal textlen=536870908</span><br></pre></td></tr></table></figure>
<p>我的版本是：MUMmer3.23</p>
<p>需要对mummer添加参数重新编译：</p>
<p>The size of your reference sequence is larger than is supported by default.<br />
You can either break your reference into multiple files and run each<br />
individually, or recompile the mummer package with:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">make clean</span><br><span class="line">make CPPFLAGS=&quot;-O3 -DSIXTYFOURBITS&quot;</span><br></pre></td></tr></table></figure>
<p><strong>20201204 更新</strong></p>
<p>MUMMER v4 可以直接对大型基因组进行比较，他的功能已经升级了。<a target="_blank" rel="noopener" href="https://github.com/mummer4/mummer/releases/download/v4.0.0rc1/mummer-4.0.0rc1.tar.gz">mummer-4.0.0rc1.tar.gz下载链接</a></p>
<p>如何对同一个物种的不同版本基因组序列比较？</p>
<p>以大豆<em>Glycine max</em>为例，大豆20条染色体，基因组尺寸约1 Gb，最新基因组版本Gmax_508_a4.v1，上一个版本是Gmax_275_a2.v1。</p>
<p><strong>一、先用nucmer对两个genome序列进行比较，生成delta文件。</strong><br />
我设定的最小匹配单元长度是1 kb，也就是1 kb内不允许错配。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">nucmer --prefix=GM508_GM275_genome --minmatch=1000 --threads=12 ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa ~/databases/Glycine_max/phytozome.12/Gmax_275_v2.0.softmasked.fa</span><br></pre></td></tr></table></figure>
<p><strong>二、然后可以delta-filter自定义筛选。</strong><br />
比如说我只想看contig和reference两者唯一匹配，并且长度在1000，相似度大于90.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">delta-filter -i 89 -l 1000 -1 GM508_GM275_genome.delta &gt;GM508_GM275_genome_i89_l1000_1.delta</span><br></pre></td></tr></table></figure>
<p><strong>三、再用mummerplot命令画图。</strong><br />
注意Rfile和Qfile主要功能是提供两个基因组各自包含的若干染色体（或者scaffold、contig）序列，不加这个参数，出来的图片没有染色体名称标记的。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mummerplot -Rfile ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa -Qfile ~/databases/Glycine_max/phytozome.12/Gmax_275_v2.0.softmasked.fa -postscript -p GM508_GM275_genome_i89_l1000_1 GM508_GM275_genome_i89_l1000_1.delta</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># `gnuplot 5.4 patchlevel 1  </span><br><span class="line"># Reading delta file GM508_GM275_genome_i89_l1000_1.delta  </span><br><span class="line"># Writing plot files GM508_GM275_genome_i89_l1000_1.fplot, GM508_GM275_genome_i89_l1000_1.rplot  </span><br><span class="line"># Writing gnuplot script GM508_GM275_genome_i89_l1000_1.gp  </span><br><span class="line"># Rendering plot GM508_GM275_genome_i89_l1000_1.ps`</span><br></pre></td></tr></table></figure>
<p><strong>20201230 备注</strong></p>
<p>更换一台机器之后，运行画图遇到错误。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ mummerplot -Rfile ./AM_genome_V5_20201211.fa -Qfile ./20201228_3ddna.fa -postscript -p plot_test V5_vs_3dV1_genome_i89_l1000_1.delta </span><br><span class="line">gnuplot 4.6 patchlevel 2</span><br><span class="line">Reading delta file V5_vs_3dV1_genome_i89_l1000_1.delta</span><br><span class="line">Writing plot files plot_test.fplot, plot_test.rplot</span><br><span class="line">Writing gnuplot script plot_test.gp</span><br><span class="line">Rendering plot plot_test.ps</span><br><span class="line">WARNING: Unable to run &#x27;false plot_test.gp&#x27;, Inappropriate ioctl for device</span><br></pre></td></tr></table></figure>
<p>切换到 python37 或者 python27 环境，用conda安装了更高版本的 gnuplot ，再安装 libjpeg-turbo 就可以运行 gnuplot 了。但是仍然无法运行mummerplot。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"># 用conda安装gnuplot</span><br><span class="line">conda install -c bioconda gnuplot</span><br><span class="line"></span><br><span class="line"># 启动gnuplot遇到错误</span><br><span class="line">$ gnuplot --version</span><br><span class="line">gnuplot: error while loading shared libraries: libjpeg.so.8: cannot open shared object file: No such file or directory</span><br><span class="line"></span><br><span class="line"># 安装 libjpeg-turbo，就可以运行gnuplot了</span><br><span class="line">$ conda install -c conda-forge libjpeg-turbo</span><br><span class="line"></span><br><span class="line"># 成功运行gnuplot</span><br><span class="line">(python27) [fenglei@localhost ~]$ gnuplot --version</span><br><span class="line">gnuplot 5.0 patchlevel 3</span><br></pre></td></tr></table></figure>
<p><strong>四、生成比对图片</strong><br />
把ps文件放到windows系统，用Photoshop软件打开上面生成的ps文件即可看到图。当然也可以直接在linux里面用convert命令将ps文件转换为pdf/png文件，但是我的linux系统没有convert命令。如图可以见到两个版本基因组的差异。11和14号染色体有明显的结构差异。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome_i89_l1000_1.jpg"><img src="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome_i89_l1000_1.jpg?w=1024" alt="" /></a></p>
<p><strong>五、如果不对delta文件做自定义筛选呢？</strong><br />
可见到大量的点，浅蓝色是反向匹配，深蓝色是正向匹配。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome.jpg"><img src="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome.jpg?w=1024" alt="" /></a></p>
<p><strong>六、如果两个物种相比较呢？</strong><br />
要注意minmatch参数不能设太大，我把两个物种（Glycine max and Glycine soja）的minmatch设置为1000000，虽然运行时间比较短，但是结果输出一个空白delta文件。设置为minmatch=100测试，需要的时间更长，得到一个满屏铺满蓝点的图，根本看不清。设置minmatch=1000，并用delta-filter对比对的序列做筛选，进而得到下图。可见到11和13号染色体的区别。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">nucmer --prefix=GM508_Gsoja_genome --minmatch=1000 --threads=12 ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa ~/databases/03_Glycine_soja_W05_2019/Gsoja.W05.genome.fadelta-filter -i 89 -l 1000 -1 GM508_Gsoja_genome.delta &gt; GM508_Gsoja_genome_i89_l1000_1.deltamummerplot -Rfile ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa -Qfile ~/databases/03_Glycine_soja_W05_2019/Gsoja.W05.genome.fa -postscript -p GM508_Gsoja_genome_i89_l1000_1 GM508_Gsoja_genome_i89_l1000_1.delta</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/gm508_gsoja_genome_i89_l1000_1.jpg"><img src="https://genehub.files.wordpress.com/2020/12/gm508_gsoja_genome_i89_l1000_1.jpg?w=1024" alt="" /></a></p>
<p>补充：画图可以用dotPlotly包，参考 <a target="_blank" rel="noopener" href="https://github.com/tpoorten/dotPlotly/tree/master/example">https://github.com/tpoorten/dotPlotly/tree/master/example</a></p>
<p>主要是R语言：<a target="_blank" rel="noopener" href="https://github.com/tpoorten/dotPlotly/blob/master/mummerCoordsDotPlotly.R">https://github.com/tpoorten/dotPlotly/blob/master/mummerCoordsDotPlotly.R</a></p>
<p>补充：如何理解上述作图过程中的几个文件呢？</p>
<p>delta文件（在<a target="_blank" rel="noopener" href="http://mummer.sourceforge.net/manual/#HumanCompare">http://mummer.sourceforge.net/manual/#HumanCompare</a> 页面的The delta format section下面有详细解释）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">/path/Gmax_508_v4.0.softmasked.fa /path/G.max-mitochondrion.fasta</span><br><span class="line"> NUCMER</span><br><span class="line">   Gm13 ChrM 45225048 402558     # 匹配区段所在的染色体</span><br><span class="line">   21773198 21776178 114322 117301 3 3 0   # 前四个数字：两个染色体上面的区段坐标。后三个数字：错误数(不相同碱基+indel碱基数)，相似错误(非正匹配得分) 终止密码子(NUCMER为0)。 如果结束大于起始，表示在负链。英文版解释：The three digits following the location coordinates are the number of errors (non-identities + indels), similarity errors (non-positive match scores), and stop codons (does not apply to DNA alignments, will be &quot;0&quot;)</span><br><span class="line">   8  # query的8位置有“缺失”？？？如果前面有负号，则表示此处为“插入”？相对ref序列而言。</span><br><span class="line">   0  # 表示当前匹配结束</span><br><span class="line">   Gm20 ChrM 47846026 402558</span><br><span class="line">   14299558 14301282 188578 190302 3 3 0</span><br><span class="line">   0 </span><br><span class="line">   Gm17 ChrM 41740656 402558</span><br><span class="line">   23669908 23674616 270387 265667 78 78 0</span><br><span class="line">   -499</span><br><span class="line">   -1   # 这连续的“-1”是什么意思？</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -106</span><br><span class="line">   -2</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -2</span><br><span class="line">   -2182</span><br><span class="line">   1263</span><br><span class="line">   -230</span><br><span class="line">   0 </span><br><span class="line">(output continues ...)</span><br></pre></td></tr></table></figure>
<p><strong>七、输出适合肉眼观察的Tabular格式比对结果</strong><br />
上面的delta文件不便于我们肉眼阅读，可用show-coords -r ${prefix}.delta &gt; ${prefix}.coords命令之后得到更适合阅读的格式（coords文件）。<br />
有点类似BLAST的“-outfmt 6”参数输出的Tabular格式<br />
如下，以ref序列的染色体排序的。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">   [S1]     [E1]       [S2]     [E2]    [LEN 1]  [LEN 2]    [% IDY]   [TAGS]</span><br><span class="line">41311147 41314402     161362   158107       3256     3256      99.91   Gm08 ChrM</span><br><span class="line">24034371 24036869     210827   213325       2499     2499      99.88   Gm12 ChrM</span><br><span class="line">24038545 24043058      13087     8578       4514     4510      99.78   Gm12 ChrM</span><br><span class="line">21773198 21776178     114322   117301       2981     2980      99.90   Gm13 ChrM</span><br><span class="line">23669908 23674616     270387   265667       4709     4721      98.35   Gm17 ChrM</span><br><span class="line">14299558 14301282     188578   190302       1725     1725      99.83   Gm20 ChrM</span><br></pre></td></tr></table></figure>
<p><strong>八、输出适合肉眼观察的Pairwise格式比对结果</strong><br />
可指定特定的染色体或scaffold用于输出到屏幕，输出格式类似BLAST的“-outfmt 1”参数。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">show-aligns seqA_vs_seqB_genome_i89_l1000_1.delta Chrxxx Chryyy | less</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">============================================================</span><br><span class="line">-- Alignments between xxx and yyy</span><br><span class="line"></span><br><span class="line">-- BEGIN alignment [ +1 1 - 245034 | -1 94962171 - 94717077 ]</span><br><span class="line"></span><br><span class="line">                  10|       20|       30|       40|</span><br><span class="line">1          cccaaccccgagccctaaaccctaaaccccaaacccccaaccctgagcc</span><br><span class="line">94962171   cccaaccccgagccctaaaccctaaaccccaaacccccaaccctgagcc</span><br><span class="line">                                                            </span><br><span class="line">                  10|       20|       30|       40|</span><br><span class="line">50         tcgagccctaagccttgagcccgagccctaagccccgatccctaagccc</span><br><span class="line">94962122   tcgagccctaagccttgagcccgagccctaagccccgatccctaagccc</span><br><span class="line">                                                            </span><br><span class="line">                  10|       20|       30|       40|</span><br><span class="line">99         tgaaccccgaaccctcaaccccgaaccccaaaccctcaaccccaaaccc</span><br><span class="line">94962073   tgaaccccgaaccctcaaccccgaaccccaaaccctcaaccccaaaccc</span><br></pre></td></tr></table></figure>
<p>参考：</p>
<ol>
<li>MUMMER 两个基因组间比较 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/c12f2a117892">https://www.jianshu.com/p/c12f2a117892</a></li>
<li>MUMMER 两个基因组间比较 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/9701c9add82d">https://www.jianshu.com/p/9701c9add82d</a></li>
<li>最详细的MUMMER中文文档 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/2e184e5c15b7">https://www.jianshu.com/p/2e184e5c15b7</a></li>
<li>How to understand MUMMER delta format？<a target="_blank" rel="noopener" href="http://mummer.sourceforge.net/manual/#HumanCompare">http://mummer.sourceforge.net/manual/#HumanCompare</a></li>
<li>用R绘制点状图：<a target="_blank" rel="noopener" href="https://github.com/tpoorten/dotPlotly">https://github.com/tpoorten/dotPlotly</a></li>
<li>用R绘制mummer的结果图 <a target="_blank" rel="noopener" href="https://blog.csdn.net/u012110870/article/details/102492711?utm%5C_medium=distribute.pc%5C_relevant.none-task-blog-title-3&amp;spm=1001.2101.3001.4242">https://blog.csdn.net/u012110870/article/details/102492711?utm\_medium=distribute.pc\_relevant.none-task-blog-title-3&amp;spm=1001.2101.3001.4242</a></li>
</ol>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/05/Linux-shell-perl-mute-multiple-lines/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/05/Linux-shell-perl-mute-multiple-lines/" class="post-title-link" itemprop="url">shell脚本和perl脚本的段注释</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-05 12:53:15" itemprop="dateCreated datePublished" datetime="2020-12-05T12:53:15+08:00">2020-12-05</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
                  , 
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>shell脚本段注释</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">#!/bin/bash </span><br><span class="line">echo &quot;Say Something&quot; </span><br><span class="line">&lt;&lt;COMMENT     </span><br><span class="line">  your comment 1     </span><br><span class="line">  comment 2     </span><br><span class="line">  blah </span><br><span class="line">COMMENT </span><br><span class="line">echo &quot;Do something else&quot;</span><br></pre></td></tr></table></figure>
<p>perl脚本段注释。测试了一下，“=”符号和“pod”之间不能加空格，pod不能大写。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">=pod</span><br><span class="line">while(&lt;file&gt;)&#123;</span><br><span class="line">  chomp $_;</span><br><span class="line">  print &quot;$_\n;</span><br><span class="line">&#125;</span><br><span class="line">=cut</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/5/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/5/">5</a><span class="page-number current">6</span><a class="page-number" href="/page/7/">7</a><span class="space">&hellip;</span><a class="page-number" href="/page/19/">19</a><a class="extend next" rel="next" href="/page/7/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">184</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
