<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>安装HiC-Pro （HiC数据预处理工具） | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">安装HiC-Pro （HiC数据预处理工具）</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">安装HiC-Pro （HiC数据预处理工具）</h1><div class="post-meta">2020-12-06<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>HiC-Pro可以用来处理Hi-C数据，从原始的fastq文件(配对端Illumina数据)到标准化的交互图谱。简单的来说就是将Hi-C数据比对到拼装好的参考基因组上，并形成交互文件去存储Hi-C数据。</p>
<p>下载 HiC-Pro VERSION 2.11.4 安装包，依赖 Python 2。（后续更新：HiC-Pro version 3.0.0 要求 Python 3）</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https:&#x2F;&#x2F;github.com&#x2F;nservant&#x2F;HiC-Pro.git </span><br></pre></td></tr></table></figure>

<h3 id="安装-HiC-Pro-VERSION-2-11-4"><a href="#安装-HiC-Pro-VERSION-2-11-4" class="headerlink" title="安装 HiC-Pro VERSION 2.11.4"></a>安装 HiC-Pro VERSION 2.11.4</h3><p>由于需要python 2 （2.7以上），我之前已经在系统安装Anaconda3，并在里面分别安装了python3.7和python2.7，(<a target="_blank" rel="noopener" href="https://genehub.wordpress.com/2020/04/21/1070/">原文链接</a>, 这里只需要切换到python2.7环境即可。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">source activate python27</span><br></pre></td></tr></table></figure>

<p>再直接用conda命令安装相关软件</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">conda install -y samtools bowtie2 R </span><br><span class="line">conda install -y pysam bx-python numpy scipy </span><br><span class="line">#打开R </span><br><span class="line">install.packages(c(&#39;ggplot2&#39;,&#39;RColorBrewer&#39;)) </span><br></pre></td></tr></table></figure>

<p>通过 conda 安装 samtools，samtools 运行会报错。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools: error while loading shared libraries: libcrypto.so.1.0.0: cannot open shared object file: No such file or directory</span><br></pre></td></tr></table></figure>

<p>现在samtools的版本已经在1.9以上了，但是conda安装的samtools版本依然是1.7，这是出现问题的原因。<a target="_blank" rel="noopener" href="https://blog.csdn.net/zhangjunya/article/details/108235796">解决办法</a>：指定另一个更新的版本，重新安装如下。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -c bioconda samtools&#x3D;1.9 --force-reinstall</span><br></pre></td></tr></table></figure>

<p>或者刚开始安装的时候：<code>conda install samtools=1.9</code>，就不会再出现上述问题。</p>
<p>不知道为什么bx-python安装失败，换成下面的方法，可以安装。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -c conda-forge -c bioconda bx-python</span><br></pre></td></tr></table></figure>

<p>然后进入HiC-Pro目录，修改config-install 文件，设置依赖软件的路径。PREFIX不要设置当前目录，会被覆盖。CLUSTER_SYS有四个选项，分别是TORQUE, SGE, SLURM 或 LSF，我随便填了一个TORQUE。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#</span><br><span class="line"> Paths and Settings  - Start editing here !</span><br><span class="line"> #</span><br><span class="line"> PREFIX &#x3D; &#x2F;home&#x2F;user&#x2F;local&#x2F;scripts&#x2F;HiC-Pro</span><br><span class="line"> BOWTIE2_PATH &#x3D; &#x2F;home&#x2F;user&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python27&#x2F;bin&#x2F;</span><br><span class="line"> SAMTOOLS_PATH &#x3D; &#x2F;home&#x2F;user&#x2F;local&#x2F;bin&#x2F;</span><br><span class="line"> R_PATH &#x3D; &#x2F;home&#x2F;user&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python27&#x2F;bin</span><br><span class="line"> PYTHON_PATH &#x3D; &#x2F;home&#x2F;user&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python27&#x2F;bin</span><br><span class="line"> CLUSTER_SYS &#x3D; TORQUE</span><br></pre></td></tr></table></figure>

<p>接下里 make configure ，通过之后屏幕显示如下信息</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">(python27) xxxxx&#x2F;HiC-Pro$ make configure       </span><br><span class="line"> </span><br><span class="line">make -f .&#x2F;scripts&#x2F;install&#x2F;Makefile CONFIG_SYS&#x3D;.&#x2F;config-install.txt</span><br><span class="line"> make[1]: Entering directory &#39;&#x2F;home&#x2F;-&#x2F;local&#x2F;app&#x2F;HiC-Pro&#39;</span><br><span class="line"> .&#x2F;scripts&#x2F;install&#x2F;install_dependencies.sh -c .&#x2F;config-install.txt -o  &#x2F;home&#x2F;-&#x2F;local&#x2F;scripts&#x2F;HiC-Pro&#x2F;HiC-Pro_2.11.4 -q</span><br><span class="line"> Make sure internet connection works for your shell prompt under current user&#39;s privilege …</span><br><span class="line"> Starting HiC-Pro installation …</span><br><span class="line"> export &#x2F;home&#x2F;--&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python27&#x2F;bin&#x2F; in PATH</span><br><span class="line"> export &#x2F;home&#x2F;-&#x2F;local&#x2F;bin&#x2F; in PATH</span><br><span class="line"> export &#x2F;home&#x2F;-&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python27&#x2F;bin in PATH</span><br><span class="line"> export &#x2F;home&#x2F;-&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python27&#x2F;bin in PATH</span><br><span class="line"> Checking dependencies … </span><br><span class="line"> Checking Python libraries …</span><br><span class="line"> The required Python libraries appear to be already installed. </span><br><span class="line"> Checking R installation …</span><br><span class="line"> The required R packages appear to be already installed. </span><br><span class="line"> Bowtie2 Aligner appears to be already installed. </span><br><span class="line"> Samtools appears to be already installed. </span><br><span class="line"> Dependencies checked !</span><br><span class="line"> Check HiC-Pro configuration … </span><br><span class="line"> Configuration for TORQUE&#x2F;PBS system.</span><br><span class="line"> make[1]: Leaving directory &#39;&#x2F;home&#x2F;-&#x2F;local&#x2F;app&#x2F;HiC-Pro&#39;</span><br></pre></td></tr></table></figure>

<p>然后 make，通过即安装成功。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">(some information are not shown here due to size limitation)</span><br><span class="line">cp -Ri &#x2F;home&#x2F;-&#x2F;local&#x2F;app&#x2F;HiC-Pro &#x2F;home&#x2F;-&#x2F;local&#x2F;scripts&#x2F;HiC-Pro&#x2F;HiC-Pro_2.11.4 </span><br><span class="line">HiC-Pro installed in &#x2F;home&#x2F;-&#x2F;local&#x2F;scripts&#x2F;HiC-Pro&#x2F;HiC-Pro_2.11.4 !</span><br></pre></td></tr></table></figure>

<p>测试</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">(python27) XXX:~&#x2F;local&#x2F;scripts&#x2F;HiC-Pro&#x2F;HiC-Pro_2.11.4&#x2F;bin$ .&#x2F;HiC-Pro -h</span><br><span class="line"> usage : HiC-Pro -i INPUT -o OUTPUT -c CONFIG [-s ANALYSIS_STEP] [-p] [-h] [-v]</span><br><span class="line"> Use option -h--help for more information</span><br><span class="line"> HiC-Pro 2.11.4</span><br><span class="line"> OPTIONS</span><br><span class="line"> -i--input INPUT : input data folder; Must contains a folder per sample with input files</span><br><span class="line">    -o--output OUTPUT : output folder</span><br><span class="line">    -c--conf CONFIG : configuration file for Hi-C processing</span><br><span class="line">    [-p--parallel] : if specified run HiC-Pro on a cluster</span><br><span class="line">    [-s--step ANALYSIS_STEP] : run only a subset of the HiC-Pro workflow; if not specified the complete workflow is run</span><br><span class="line">       mapping: perform reads alignment - require fast files</span><br><span class="line">       proc_hic: perform Hi-C filtering - require BAM files</span><br><span class="line">       quality_checks: run Hi-C quality control plots</span><br><span class="line">       merge_persample: merge multiple inputs and remove duplicates if specified - require .validPairs files</span><br><span class="line">       build_contact_maps: Build raw inter&#x2F;intrachromosomal contact maps - require .allValidPairs files</span><br><span class="line">       ice_norm : run ICE normalization on contact maps - require .matrix files</span><br></pre></td></tr></table></figure>

<h3 id="Optional-安装-HiC-Pro-VERSION-3-0-0"><a href="#Optional-安装-HiC-Pro-VERSION-3-0-0" class="headerlink" title="Optional: 安装 HiC-Pro VERSION 3.0.0"></a><del>Optional: 安装 HiC-Pro VERSION 3.0.0</del></h3><p><del>conda create –name python38 python=3.8</del><br><del>source activate python38</del><br><del>conda install conda</del><br><del># 前面的conda是anaconda3的conda，后面的conda是python3.8的conda</del></p>
<p><del>conda install -c bioconda bowtie2</del><br><del>conda install -c bioconda samtools=1.9</del><br><del>conda install -c conda-forge -c bioconda bx-python</del></p>
<p><del>在conda里面安装R和pysam遇到很多问题，更换python版本3.7/3.8/3.9都没成功，真是浪费时间，放弃。</del></p>
<h3 id="使用HiC-Pro对HiC数据进行分析的过程"><a href="#使用HiC-Pro对HiC数据进行分析的过程" class="headerlink" title="使用HiC-Pro对HiC数据进行分析的过程"></a>使用HiC-Pro对HiC数据进行分析的过程</h3><p>1. 在工作目录（例如HiC-analysis-workdir），建立若干文件夹，例如index、rawdata，方便管理，当然这些数据也可以在其他位置。</p>
<p>2. 在index路径下，对参考基因组或者scaffolds建立index。HiC-Pro调用的比对软件是bowtie2。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bowtie2-build --threads 20 ref.fasta ref.fasta</span><br></pre></td></tr></table></figure>

<p>3. 在工作目录，建立一个文本文件，内含参考基因组（或scaffolds）各个染色体名称和长度的信息。可以用samtools fadix，也可以用iTools或R语言等工具。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ cat chromosome_size.txt</span><br><span class="line"> Chr01   11111</span><br><span class="line"> Chr02   1233333</span><br><span class="line"> Chr03   4444</span><br><span class="line"> Chr04   105314844</span><br><span class="line">(continue ...)</span><br></pre></td></tr></table></figure>

<p>4. 在工作目录，新建一个文件夹（例如rawdata），用来存放作为input的测序数据。注意不能直接将测序文件放到rawdata目录下，而是要按照样品名先建立子文件夹（例如sample1、sample2）。sample1目录下即存放样品的测序文件，包含F/R两个测序文件，文件应该是.fastq(.gz)的命名，不可写成fq(.gz)；可以是软链接的形式。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207122951.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207122951.png?w=506"></a></p>
<p>4. 基因组的酶切位点识别。我用的MboI（酶切位点：^GATC）</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python2.7 XXX&#x2F;HiC-Pro&#x2F;HiC-Pro_2.11.4&#x2F;bin&#x2F;utils&#x2F;digest_genome.py -r ^GATC -o mboi_resfrag_genome.bed Ref_genome.fa</span><br></pre></td></tr></table></figure>

<p>得到的酶切信息：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"> Chr01   0       3282    HIC_Chr01_1     0       +</span><br><span class="line"> Chr01   3282    3609    HIC_Chr01_2     0       +</span><br><span class="line"> Chr01   3609    3625    HIC_Chr01_3     0       +</span><br><span class="line"> Chr01   3625    4485    HIC_Chr01_4     0       +</span><br><span class="line"> Chr01   4485    5312    HIC_Chr01_5     0       +</span><br><span class="line"> Chr01   5312    5327    HIC_Chr01_6     0       +</span><br><span class="line"> Chr01   5327    5958    HIC_Chr01_7     0       +</span><br><span class="line"> Chr01   5958    6475    HIC_Chr01_8     0       +</span><br><span class="line"> Chr01   6475    7545    HIC_Chr01_9     0       +</span><br><span class="line">(continue ...)</span><br></pre></td></tr></table></figure>

<p>5. 将HiC-Pro程序安装包目录下的 config-hicpro.txt 文件拷贝到工作目录，对其进行编辑。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"># N_CPU，例如N_CPU &#x3D; 24</span><br><span class="line"># JOB_NAME，例如JOB_NAME &#x3D; Ga</span><br><span class="line"># JOB_WALLTIME，例如JOB_WALLTIME &#x3D; 11:00:00</span><br><span class="line"># JOB_QUEUE，例如JOB_QUEUE &#x3D; normal</span><br><span class="line"># JOB_MAIL，例如JOB_MAIL &#x3D; 1570447120@qq.com</span><br><span class="line"># BOWTIE2_IDX_PATH，填写用bowtie2对reference建立索引的目录，注意是目录，而不是索引路径!如&#x2F;PATH&#x2F;to&#x2F;Ga.ref</span><br><span class="line"># REFERENCE_GENOME，例如Ga 注意：这里REFERENCE_GENOME一定要和bowtie2建立的索引对应上,名字为物种名</span><br><span class="line"># GENOME_SIZE，即生成的文件Ga.fasta.size（最好使用绝对路径）</span><br><span class="line"># GENOME_FRAGMENT，用digest_genome.py程序生成的文件Ga.HindIII.txt</span><br><span class="line"># LIGATION_SITE，酶的序列，例如HindIII，则为AAGCTAGCTT；如果是MboI则序列为GATCGATC。</span><br><span class="line"># MIN_FRAG_SIZE，例如MIN_FRAG_SIZE &#x3D; 100</span><br><span class="line"># MAX_FRAG_SIZE，例如MAX_FRAG_SIZE &#x3D; 160000</span><br><span class="line"># MIN_INSERT_SIZE，例如MIN_INSERT_SIZE &#x3D; 200</span><br><span class="line"># MAX_INSERT_SIZE，例如MAX_INSERT_SIZE &#x3D; 600</span><br><span class="line"># GET_PROCESS_SAM，例如GET_PROCESS_SAM &#x3D; 1</span><br></pre></td></tr></table></figure>

<p>6. 运行</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&#x2F;PATH&#x2F;TO&#x2F;HiC-Pro_2.11.1&#x2F;bin&#x2F;HiC-Pro -i &#x2F;PATH&#x2F;TO&#x2F;rawdata --output hicpro_output --conf config-hicpro.txt </span><br><span class="line"># --input reads所在的文件</span><br><span class="line"># --output 输出的文件夹</span><br><span class="line"># --conf 配置文件</span><br></pre></td></tr></table></figure>

<p>7. 运行结果</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br></pre></td><td class="code"><pre><span class="line">.</span><br><span class="line">├── bowtie_results</span><br><span class="line">│   ├── bwt2</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.bwt2merged.bam</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.mapstat</span><br><span class="line">│   │       ├── mysample_R2_V3_20201206.fa.bwt2merged.bam</span><br><span class="line">│   │       ├── mysample_R2_V3_20201206.fa.mapstat</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.bam</span><br><span class="line">│   │       └── _V3_20201206.fa.bwt2pairs.pairstat</span><br><span class="line">│   ├── bwt2_global</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.bwt2glob.bam</span><br><span class="line">│   │       ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap.fastq</span><br><span class="line">│   │       ├── mysample_R2_V3_20201206.fa.bwt2glob.bam</span><br><span class="line">│   │       └── mysample_R2_V3_20201206.fa.bwt2glob.unmap.fastq</span><br><span class="line">│   └── bwt2_local</span><br><span class="line">│       └── sample1</span><br><span class="line">│           ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_bwt2loc.bam</span><br><span class="line">│           ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_trimmed.fastq</span><br><span class="line">│           ├── mysample_R2_V3_20201206.fa.bwt2glob.unmap_bwt2loc.bam</span><br><span class="line">│           └── mysample_R2_V3_20201206.fa.bwt2glob.unmap_trimmed.fastq</span><br><span class="line">├── config-hicpro.txt</span><br><span class="line">├── hic_results</span><br><span class="line">│   ├── data</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── sample1.allValidPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.DEPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.DumpPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.FiltPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.REPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.RSstat</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.SCPairs</span><br><span class="line">│   │       ├── _V3_20201206.fa.bwt2pairs.SinglePairs</span><br><span class="line">│   │       └── _V3_20201206.fa.bwt2pairs.validPairs</span><br><span class="line">│   ├── matrix</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── iced</span><br><span class="line">│   │       │   └── 200000</span><br><span class="line">│   │       │       ├── sample1_200000_iced.matrix</span><br><span class="line">│   │       │       └── sample1_200000_iced.matrix.biases</span><br><span class="line">│   │       └── raw</span><br><span class="line">│   │           └── 200000</span><br><span class="line">│   │               ├── sample1_200000_abs.bed</span><br><span class="line">│   │               └── sample1_200000.matrix</span><br><span class="line">│   ├── pic</span><br><span class="line">│   │   └── sample1</span><br><span class="line">│   │       ├── plotHiCContactRanges_sample1.pdf</span><br><span class="line">│   │       ├── plotHiCFragment_sample1.pdf</span><br><span class="line">│   │       ├── plotHiCFragmentSize_sample1.pdf</span><br><span class="line">│   │       ├── plotMappingPairing_sample1.pdf</span><br><span class="line">│   │       └── plotMapping_sample1.pdf</span><br><span class="line">│   └── stats</span><br><span class="line">│       └── sample1</span><br><span class="line">│           ├── sample1_allValidPairs.mergestat</span><br><span class="line">│           ├── sample1mysample_R1.mmapstat</span><br><span class="line">│           ├── sample1mysample_R2.mmapstat</span><br><span class="line">│           ├── sample1.mpairstat</span><br><span class="line">│           └── sample1.mRSstat</span><br><span class="line">├── logs</span><br><span class="line">│   └── sample1</span><br><span class="line">│       ├── build_raw_maps.log</span><br><span class="line">│       ├── mysample_R1_bowtie2.log</span><br><span class="line">│       ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_bowtie2.log</span><br><span class="line">│       ├── mysample_R1_V3_20201206.fa.bwt2glob.unmap_readsTrimming.log</span><br><span class="line">│       ├── mysample_R2_bowtie2.log</span><br><span class="line">│       ├── mysample_R2_V3_20201206.fa.bwt2glob.unmap_bowtie2.log</span><br><span class="line">│       ├── mysample_R2_V3_20201206.fa.bwt2glob.unmap_readsTrimming.log</span><br><span class="line">│       ├── ice_200000.log</span><br><span class="line">│       ├── make_Rplots.log</span><br><span class="line">│       ├── mapped_2hic_fragments.log</span><br><span class="line">│       ├── mapping_combine.log</span><br><span class="line">│       ├── mapping_stats.log</span><br><span class="line">│       ├── mapping_step1.log</span><br><span class="line">│       ├── mapping_step2.log</span><br><span class="line">│       ├── mergeSAM.log</span><br><span class="line">│       ├── merge_stats.log</span><br><span class="line">│       ├── merge_valid_interactions.log</span><br><span class="line">│       ├── plot_hic_contacts.Rout</span><br><span class="line">│       ├── plot_hic_fragment.Rout</span><br><span class="line">│       ├── plot_mapping_portion.Rout</span><br><span class="line">│       └── plot_pairing_portion.Rout</span><br><span class="line">├── rawdata -&gt; &#x2F;PATH&#x2F;TO&#x2F;&#x2F;rawdata&#x2F;</span><br><span class="line">└── tmp</span><br></pre></td></tr></table></figure>

<p><strong>画图</strong><br>HiC-Pro运行结果中，sample1.allValidPairs文件就表示pair end reads在染色体/scaffolds上面位置的信息。<br>这时我们就可以把它转化为.hic文件，放到juicebox中就很好的可视化。<br>如何转换成.hic文件，并且用windows版本的Juicebox软件画成热图？使用HiC-Pro安装路径下的<code>hicpro2juicebox.sh</code>程序按照下面的方式运行即可得到.hic文件。<br>需要下载juicer_tools_1.22.01.jar文件作为输入文件。下载地址：<a target="_blank" rel="noopener" href="https://github.com/aidenlab/juicer/wiki/Download">https://github.com/aidenlab/juicer/wiki/Download</a><br>备注：我一开始下载错了，去<a target="_blank" rel="noopener" href="https://github.com/aidenlab/Juicebox/wiki/Download">https://github.com/aidenlab/Juicebox/wiki/Download</a> 页面下了个Juicebox.jar，结果不能用于下面的步骤。原来这是JuiceBox，不是Juicer！从论坛发现遇到这个问题的不止我一个人<a target="_blank" rel="noopener" href="https://groups.google.com/g/hic-pro/c/DF0zfZOivsE?pli=1">https://groups.google.com/g/hic-pro/c/DF0zfZOivsE?pli=1</a>。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">&#x2F;PATH&#x2F;TO&#x2F;HiC-Pro&#x2F;HiC-Pro_2.11.4&#x2F;bin&#x2F;utils&#x2F;hicpro2juicebox.sh -i &#x2F;PATH&#x2F;TO&#x2F;&#x2F;sample1.allValidPairs -g &#x2F;PATH&#x2F;TO&#x2F;&#x2F;chromosome_sizes -j &#x2F;PATH&#x2F;TO&#x2F;juicer_tools_1.22.01.jar -r &#x2F;PATH&#x2F;TO&#x2F;&#x2F;mboi_resfrag_AMgenome.bed -t temp -o &#x2F;PATH&#x2F;TO&#x2F;output</span><br></pre></td></tr></table></figure>


<p><img src="https://gblobscdn.gitbook.com/assets%2F-LijQZQE8sc7HbVdoxEA%2F-Lj2gavu4NTsjkOUMYaT%2F-Lj2js_V0JRQoDXKpVvE%2FScreen%20Shot%202019-07-05%20at%2012.51.13%20PM.png?alt=media&token=fdae559d-237e-45dd-a0dd-a12a22ff7506" alt=".hic文件导入juicebox软件之后的示意图"></p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2020/12/07/hello-world/">Hello World</a><a class="next" href="/2020/12/05/%E4%B8%A4%E4%B8%AA%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%BA%8F%E5%88%97%E7%9B%B8%E4%BA%92%E6%AF%94%E8%BE%83%EF%BC%9Amummer/">两个基因组序列相互比较：Mummer</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/01/08/Bioinfo-Installation-of-BUSCO-v4/">Bioinfo | Installation of BUSCO v4</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/05/Linux-Installation-of-Tomcat-9/">Linux | Installation of Tomcat 9, Java 8, and MySQL 5.7.20</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/04/Remove-Background-from-Image/">Remove Background from Image</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/01/%E5%9C%A8Linux%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E4%BD%BF%E7%94%A8Nginx%E9%83%A8%E7%BD%B2Hexo%E7%AB%99%E7%82%B9/">在Linux服务器上使用Nginx部署Hexo站点</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/29/Linux%EF%BC%9A%E4%BD%BF%E7%94%A8Modules%E7%AE%A1%E7%90%86%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/">Linux：使用Environment Modules管理环境变量</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E4%BD%BF%E7%94%A8Snakemake%E6%90%AD%E5%BB%BA%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/">使用Snakemake搭建生物信息分析流程</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9%E7%9A%84SAM-BAM%E6%96%87%E4%BB%B6%E4%B8%AD%EF%BC%8C%E6%80%8E%E4%B9%88%E7%90%86%E8%A7%A3MAPQ%E5%80%BC/">基因组比对的SAM/BAM文件中，怎么理解MAPQ值</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/25/Linux%EF%BC%9Avim%E8%AE%BE%E7%BD%AE/">Linux：vim设置</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/Hi-C%EF%BC%9A3d-dna%E6%8C%82%E8%BD%BD%E6%9F%93%E8%89%B2%E4%BD%93/">Hi-C：3d-dna挂载染色体</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/%E4%B8%80%E4%BA%9BFASTA-FASTQ%E5%BA%8F%E5%88%97%E5%A4%84%E7%90%86%E5%B7%A5%E5%85%B7/">一些FASTA/FASTQ序列处理工具</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>