<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta property="og:type" content="website">
<meta property="og:title" content="FENG&#39;s Notes">
<meta property="og:url" content="http://example.com/page/3/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/page/3/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/" class="post-title-link" itemprop="url">Bioinfo | Positive Selection Ka/Ks calculation</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-10-01 15:32:13" itemprop="dateCreated datePublished" datetime="2021-10-01T15:32:13+08:00">2021-10-01</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-10-27 14:00:20" itemprop="dateModified" datetime="2021-10-27T14:00:20+08:00">2021-10-27</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>总结正向选择分析的一些知识。2017 年做过一次线粒体基因的正向选择，相关内容已经<a target="_blank" rel="noopener" href="https://link.springer.com/article/10.1007%2Fs00606-019-01578-2#Sec18">发表</a>，并且我在Supplementary Files 里附上了我的<a target="_blank" rel="noopener" href="https://static-content.springer.com/esm/art%3A10.1007%2Fs00606-019-01578-2/MediaObjects/606_2019_1578_MOESM1_ESM.pdf">分析代码</a>。2021年再次需要做正向选择分析，再次回顾以前的工作，发现其中也有一个问题，就是没有做统计检验，推算P值，再次学习正向选择分析，并摘抄一些网络资料并整理记录。Codeml 的学习成本比较高，原理与操作都不易理解，学习的过程中发现不少好资源，比如 B 站有<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1Qx411m7Bn?from=search&amp;seid=9774626637377853773&amp;spm_id_from=333.337.0.0">视频</a>讲解如何使用 codeml 推断特定分支的 ω 值（dN/dS），以及使用卡方检验来计算 P 值。另外有国内的团队以 codeml 为内核，用 java 编写了更友好的操作界面：<a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/home.php?mod=space&amp;uid=460481&amp;do=blog&amp;id=1163040">EasyCodeML</a>。</p>
<p><img src="https://d3i71xaburhd42.cloudfront.net/aac1c83c6feff2575f6296ea4751f29c0fec42a1/12-Figure8-1.png" alt="我计算 8 个物种的线粒体基因的 Ka/Ks，但是没有计算 P 值" /></p>
<h3 id="基因正选择分析原理"><a class="markdownIt-Anchor" href="#基因正选择分析原理"></a> <a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3084">基因正选择分析原理</a></h3>
<p>下文由 <a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3084">chenlianfu.com</a> 发表于2019 年 11 月 18 日</p>
<p><strong>一、 正选择分析的目的</strong></p>
<p>两两基因的密码子序列进行比较，从而计算 dN/dS，即 omega（ω）值。若该值 &lt; 1，则表示纯化选择；omega = 1，则中性进化；omega &gt; 1，则正选择。若分析基因在两个物种中的序列，可以计算 dN/dS 的值，若 omega &gt; 1，即表明该基因在物种进化过程中，即由其祖先物种分化成这两个物种时，基因受到了正选择。对于两个物种/序列的正选择分析，比较简单。而实际情况中，要分析的物种数量很多，包含多个类群。这个时候的正选择分析相对复杂些。</p>
<p>对多个物种的基因序列进行正选择分析，若仍然按照两个物种时的要求，即分析该基因在物种进化中是否受到了正选择？这种结果可能不好说清楚。因为该基因可能在某一类群中序列很相似，其两两比较时，omega &lt;= 1；而在另外一类群中两两比较时，很多时候 omega &gt; 1。最后软件可以从总体上给一个 omega 值，该值不可以拿来简单地评价该基因是否受到了正选择。所以，对多个物种进行正选择分析时，没法直接评价该基因是否受到了正选择。正选择只有在进行两两序列比较的时候，才能计算 omega 值，从而得到结果。</p>
<p>对基因在多个物种上的正选择分析，分析的目的则是：比较某个分枝上祖先节点和后裔节点（可以理解成，对无根树上某分枝两侧的两组物种进行比较，依然属于两两比较），从而计算该分枝的 omega 值。而在实际数据中，基因在不同的进化分枝上具有不同的 omega 值，同时在序列不同的位点也具有不同的 omega 值。目标分枝两侧的物种数量较多时，可以对序列上的每个位点进行 omega 值分析，从而鉴定正选择位点。所以，对基因在多个物种上的正选择分析，需要同时分析目标分枝的 omega 值和序列位点的 omega 值，从而判断基因是否受到正选择压。</p>
<p><strong>二、使用 PAML 对基因进行正选择分析，有三种方法</strong></p>
<p>PAML site model: 主要用于检测基因中的正选择位点。该方法分析时，认为进化树中各分枝的 omega 值是一致的，并比较两种模型：(1) 模型 m1 是 null model，认为所有位点的 omega 值&lt;1 或 =1; (2) 模型 m2 是正选择模型，存在 omega &lt;1、=1 或 &gt;1 的位点。比较两个模型的似然值（lnL）差异，利用卡方检验（自由度为2）算出 p 值。若 p 值 &lt; 0.05，则否定 null model，认为存在正选择位点。此外，推荐采用比较模型 m7 和 m8，它们将 omega 值分成了 10 类，其 p 值结果比上一种比较方法更宽松，能检测到更多的正选择基因。使用 PAML site model 方法能在整体水平上检测基因的正选择位点，而不能表明基因在某个进化分枝上是否受到正选择压。</p>
<p><font color=blue>PAML branch-site model: 主要用于检测基因在某个进化枝上是否存在正选择位点。该分析方法认为目标分化枝具有一个 omega 值，其它所有分枝具有一个相同的 omega 值，然后再检测正选择位点。同样对两种模型进行比较：（1）第一种模型为模型2，将 omega 值分成 &lt;1、=1、&gt;1 的三类，这和 site model 中的一样；（2）第二种模型和前者一致，只是将 omega 固定成 1，作为 null model。比较两种模型的似然差异，利用卡方检验（自由度为2）算 p 值（chi2 命令算出的值除以2）。若 p 值 &lt; 0.05，则能通过 Bayes Empirical Bayes (BEB) 方法计算正选择位点的后验概率</font>，若存在概率值 &gt; 0.95 正选择位点，则表示基因在目标分枝上受到正选择压。<u><font color=red>PAML 软件在 branch-site 模式下，并不给出分枝上的 omega 值。</font></u>这表示 branch-site 模式虽然考虑了目标分枝上具有不同的 omega 值，但仍然以分析位点上的 omega 为主。值得注意的是，在 branch-site 模式下可能检测到正选择位点，但在目标分枝上的 omega 值仍然可能低于 1。可能软件作者基于这点考虑，就没有给出目标分枝上的 omega 值，以免影响一些人对正选择结果的判断。</p>
<p>PAML branch model: 主要用于检测在某个分枝上，其 omega 值是否显著高于背景分枝，即基因在目标分枝上进化速度加快。该方法认为基因序列上所有位点的 omega 值是一致的，对两种模型进行比较：（1）第一种模型为 null model，所有分枝具有相同的 omega 值；（2）第二种模型认为目标分枝具有一个 omega 值，其它所有分枝具有一个相同的 omega 值。比较两种模型的似然差异，利用卡方检验（自由度为1）算 p 值。若 p 值 &lt;= 0.05，且目标分枝上的 omega 值高于背景值，则认为该基因为快速进化基因。一般情况下，该方法计算得到的 p 值会低于第二种方法的结果。</p>
<p><strong>三、其它注意事项</strong></p>
<p>Branch-site model 相比于 site model 的优点是考虑了不同的分枝具有不同的选择压，即具有不同的 omega 值。该方法让目标分枝具有一个不同的 omega 值，并没有让所有分枝的 omega 值独立进行计算（理论上这样是最好的）。这样算法很复杂，程序运行非常非常消耗时间。但其实也没必要这样做，因为正选择分析其实是两条序列比较后，分析 dN/dS，再找正选择位点，其分析结果就应该是某个分枝上基因是否受到正选择，在序列那个位点上受到正选择。</p>
<p>若在目标分枝上，其 omega 值小于 1，但是却能找到正选择位点。即该基因在该分枝上的 dN/dS &lt; 1，但是在某些位点上，dN/dS &gt; 1。那么该基因是否属于正选择基因？我认为：属于。之所以为正选择基因，主要是因为基因的个别位点或多个位点存在正选择。当只有个别位点受到正选择压力时，而其它多个位点存在纯化选择时，可能导致整体上的omega值小于1。此时，该基因也应该是属于正选择基因。</p>
<p><strong>四、参考文献中的正选择分析方法描述</strong></p>
<p>Science 文章<a target="_blank" rel="noopener" href="https://science.sciencemag.org/content/364/6446/eaav6202"><em>《大规模反刍动物基因组测序与其独特性质的进化研究》</em></a> 中的正选择基因分析方法。</p>
<p>注：这篇文章做了几十个物种的从头测序与拼接，基本上由 Illumina 测序，说明该项目应该做了挺多年，基因组拼接的质量放在今天来说不算好，大部分物种的 contig N50 不到 50 Kb，后期的比较基因组学分析做的好。主要目的是研究反刍动物特性，例如用于反刍的瘤胃、可再生的骨质角、不同物种巨大的身高差异等。主要手段是研究基因的选择压力。分析思路值得学习，其<a target="_blank" rel="noopener" href="https://www.science.org/doi/suppl/10.1126/science.aav6202/suppl_file/aav6202-chen-sm.pdf"><em>附件</em></a>包含大量中间数据，可作为学习参考。</p>
<p><strong>Identification of positively selected genes (PSGs).</strong> We used a conserved genome synteny methodology to establish a highconfidence orthologous gene set that included camel, cat, dog, horse, human, minke whale, killer whale, pig, and the 51 ruminant species. Briefly, pairwise WGAs were constructed for relevant genomes using LAST, with the goat genome serving as the reference genome. To minimize the effect of annotation, sequencing and assembly errors, pseudogenes, nonorthologous alignments, and non-conserved gene structures 22 on subsequent evolutionary rate analyses, a series of rigorous filtering criteria were adopted: (1) the genes mapped to the goat genome via a single chain of sequence alignments including at least 80% of its coding sequence (CDS), and met the alignment length/score thresholds required for inclusion in the MULTIZ alignments; (2) frame-shift indels in CDSs were prohibited; (3) CDSs with premature stop codons were excluded and (4) genes with Ks values (synonymous substitutions per synonymous site) between each species and cattle larger than two were excluded. As the number of available species for each gene increased, the Codeml running time dramatically increases because of the power function relation between the number of tree leaves and branches. As a trade-off, we subsampled the ruminant species by using the following criteria: (1) if an ortholog existed in more than four species in the Cervidae or Bovidae lineage, we randomly selected 4 sequences to represent either family in order to reduce computational demand; (2) genes contained in less than 4 outgroup species were excluded; (3) genes recovered in fewer than 3 ruminant families were excluded and (4) final alignments shorter than 150 bp were discarded. After these filtering steps, a total of 12,977 genes were retained, and on average ~6.6 families containing ~19.8 species were retained for each ortholog <font color=blue>(fig. S51)</font>. The tree topology shown in Fig. 1A was used as the prior tree topology, and the “<em>ape</em>” package (164) was used to estimate the unrooted tree topology for each ortholog.</p>
<p>To estimate the lineage-specific evolutionary rate for each branch, the Codeml program in the PAML package (version 4.8) (134) with the free-ratio model (model = 1) was run for each ortholog. <font color=blue>Positive selection signals on genes along specific lineages were detected using the optimized branch-site model following the author’s recommendation. A likelihood ratio test (LRT) was conducted to compare a model that allowed sites to be under positive selection on the foreground branch with the null model in which sites could evolve either neutrally and under purifying selection. The p values were computed based on Chi-square statistics, and genes with p value less than 0.05 were treated as candidates that underwent positive selection.</font> We identified PSGs at the ancestral branch of Ruminantia (table S22), the ancestral branch of Pecora (table S23), each ancestral family branch of Ruminantia (tableS24), and each ancestral subfamily branch of Bovidae (table S24). We also compared the dN/dS values of Ruminantia families with outgroup mammals <font color=blue>(fig. S52)</font>.</p>
<p><img src="https://i.imgur.com/a9XwYdx.png" alt="" /></p>
<p><img src="https://i.imgur.com/he2bRKC.png" alt="" /></p>
<p>Science 文章<a target="_blank" rel="noopener" href="https://science.sciencemag.org/content/364/6446/eaav6202">《大规模反刍动物基因组测序与其独特性质的进化研究》</a> 中快速进化基因分析方法：The branch model in PAML was used, with the null model (model=0) assuming that all branches have been evolving at the same rate and the alternative model (model=2), allowing the foreground branch to evolve under a different rate. An LRT with df =1 was used to discriminate between alternative models for each ortholog in the gene set. Genes with a p value less than 0.05 and a higher ω value for the foreground than the background branches were considered as evolving with a significantly faster rate in the foreground branch.</p>
<p><a target="_blank" rel="noopener" href="https://journals.plos.org/plosntds/article?id=10.1371%2Fjournal.pntd.0007463">Identification of positively selected genes in human pathogenic treponemes: Syphilis-, yaws-, and bejel-causing strains differ in sets of genes showing adaptive evolution</a> 文献中对正选择基因的分析方法：</p>
<p><img src="https://i.imgur.com/35rVAXK.png" alt="" /></p>
<p>A calculation of mutational rate ratio ω between two gene sequences was the basis for the positive selection analysis. The ω was calculated as a ratio of nonsynonymous to synonymous mutational rates. The ratio indicates negative purifying selection (0 &lt; ω &lt; 1), neutral evolution (ω = 1), and positive selection (ω &gt; 1) [54]. A set of selected genes from complete genomes was tested relative to positive selection using the maximum likelihood method using the CODEML of the PAML software package [55]. PAML version 4 [56] and its user interface PAMLX [57] were used in our study. For each analyzed gene, its maximum likelihood phylogenetic tree was used as an input tree. The CODEML offers several different codon evolutionary models, and the statistical likelihood ratio test (LRT) was used to compare the codon evolutionary model to the null model. The Bayes empirical Bayes method (BEB) [58] was then used to evaluate the posterior probability of sites considered to have been positively selected.</p>
<p>The CODEML models could produce different results (i.e., a list of sites under positive selection) since they calculate different parameter estimates. Site models allow ω to vary in each site (codon) within the gene. Statistical testing was required for sites with ω &gt; 1. Two pairs of models were predominantly used since their LRTs have low false-positive rates. M1a (nearly neutral evolution) was compared to M2a (positive selection) [58,59] and M7 (beta) was compared to M8 (beta &amp; ω) [60]. Our preliminary testing found that the two model pairs gave the same or very similar results. Therefore we chose to use the M7-M8 model pair. The M7 model is a null model that allows 10 classes of sites with a ω beta-distribution within the interval 0 ≤ ω ≤ 1. Sites with ω &gt; 1 are not allowed. The alternative M8 model adds an eleventh class of sites with ω &gt; 1. Each site was tested to determine the class to which it belongs. The LRT compares twice the log-likelihood difference 2Δl = 2(l1-l0) between the M7 model (log likelihood value l0) and the M8 model (log likelihood value l1) to the χ2 distribution [61]. If the twice log-likelihood difference is above a critical χ2 value, then the null model is rejected, and the positive selection is statistically significant.</p>
<p>A considerable disadvantage of the site models is that ω was calculated as an average over all codons of the site. Therefore, the site models are not suitable for the data where ω also varies between lineages. In contrast, the branch-site models search for positive selection in sites and pre-specifies lineages where different rates of ω may occur [62]. Sequences of lineages are a priori divided into a group of foreground lineages where positive selection may occur and group of background lineages where only purifying selection or neutral evolution occurs. We used branch-site model A, which allows four classes of sites and different setups of foreground lineages to be tested depending on the gene phylogeny. In branch-site model A, all lineages under purifying selection with a low value of ω0 belong to site class 0. Weak purifying selection and neutral evolution with ω1 near to value 1 are allowed in site class 1. In site class 2a, a proportion of class 0 sites in foreground lineages is under positive selection with ω2 &gt; 1. Similarly, site class 2b is a proportion of class 1 sites under positive selection with ω2 &gt; 1. The null model for LRT has ω2 = 1. Critical values of LRT (2Δl) are 2.71 at 5% and 5.41 at 1% [63]. The posterior probabilities of suggested sites under positive selection were calculated using the BEB method.</p>
<h3 id="基因-positive-selecetion-分析过程"><a class="markdownIt-Anchor" href="#基因-positive-selecetion-分析过程"></a> 基因 positive selecetion 分析过程</h3>
<p>PAML 软件中的 <code>codeml</code> 命令可以使用 Maxmum Likelihood 方法对密码子或氨基酸的多序列比对结果进行分析和进化参数计算 。这两种数据共用一个命令，程序运行时的参数绝大部分是公用的，但仍需要分清两种数据各自的参数意义。PAML软件中进行正选择分析的示例数据位于 <code>examples/lysozyme/</code> 目录下。</p>
<ol>
<li>基因序列准备，对于选定的若干个物种，分别准备它们的基因的 CDS 序列（DNA 格式），要注意每个基因只保留一个转录本即可；另外对序列名进行手动调整，使序列名具有独特性，易识别性，主要是为了便于后续的程序自动循环处理。并将基因 CDS 分别翻译成蛋白序列，注意这一步可能会出现的问题，比如不少物种的基因组文章即使发表在高分杂志上，它的基因集并不一定是完美的，以花生为例，我发现有数个基因CDS直接翻译之后会在序列中间出现终止密码子，这种现象可能是基因组突变，也有可能是基因注释过程中引入的小错误，还有一个我发现有些基因的CDS序列起始并不是ATG这种起始密码子（不排除是 RNA editing 现象），那用普通翻译程序可能就会出现翻译错误。所以针对各个物种，最好分别检查，灵活处理，比如移除这些包含错误的基因，或者直接使用物种自带的蛋白序列集。</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">#########################################################################</span><br><span class="line"># File Name: loop.sh</span><br><span class="line"># Author: fenglei</span><br><span class="line"># Mail: fengleiluck@gmail.com</span><br><span class="line"># Created Time: Fri 08 Oct 2021 04:14:35 PM CST</span><br><span class="line">#########################################################################</span><br><span class="line">#!/usr/bin/bash</span><br><span class="line">set -e</span><br><span class="line">set -u</span><br><span class="line">set -o pipefail</span><br><span class="line"></span><br><span class="line">for i in Ammopiptanthus_nanus.cds.fa Lotus_japonicus.cds.fa Glycine_max.cds.fa Medicago_truncatula.cds.fa Vigna_angularis.cds.fa Arachis_ipaensis.cds.fa Lupinus_angustifolius.cds.fa; do</span><br><span class="line">        #prefix=$&#123;i#*.cds&#125;;</span><br><span class="line">        prefix=$&#123;i%%.cds*&#125;;</span><br><span class="line">        echo [`date +&quot;%Y-%m-%d %H:%M:%S&quot;`] &quot;#&gt; Processing $prefix&quot;</span><br><span class="line">        iTools Fatools CDS2Pep -InCDS $&#123;prefix&#125;.cds.fa -OutPut  $&#123;prefix&#125;.pep_iTools.fa # 可能翻译的序列种存在终止密码子</span><br><span class="line">        gunzip *gz</span><br><span class="line">        perl pep_seq_clean.pl $&#123;prefix&#125;.pep_iTools.fa $&#123;prefix&#125;.pep.fa</span><br><span class="line">done</span><br></pre></td></tr></table></figure>
<p>上面的循环脚本中调用了一个序列清理程序，如下，可以去掉中间含有星号的氨基酸序列，也可以移除蛋白序列尾部的终止密码子形成的星号。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">$ cat pep_seq_clean.pl</span><br><span class="line">#!/usr/bin/perl</span><br><span class="line"></span><br><span class="line">#/*************************************************************************</span><br><span class="line">#       &gt; File Name: seq_clean.pl</span><br><span class="line">#       &gt; Author: fenglei</span><br><span class="line">#       &gt; Mail: fengleiluck@gmail.com</span><br><span class="line">##      &gt; Created Time: Fri 08 Oct 2021 01:42:50 PM CST</span><br><span class="line"># ************************************************************************/</span><br><span class="line"></span><br><span class="line">#  如果氨基酸序列内部含有终止密码子，就去掉这个序列</span><br><span class="line">#  perl seq_clean.pl input.pep.fa output.pep.fa</span><br><span class="line"></span><br><span class="line">open INPUT, $ARGV[0] or die &quot;Can not open the file!\n $!&quot;;</span><br><span class="line">open OUTPUT, &quot;&gt;$ARGV[1]&quot; or die &quot;Can not open the file!\n $!&quot;;</span><br><span class="line">select OUTPUT;</span><br><span class="line"></span><br><span class="line">$/ = &quot;\&gt;&quot;;</span><br><span class="line">&lt;INPUT&gt;;</span><br><span class="line">while(&lt;INPUT&gt;)&#123;</span><br><span class="line">        chomp $_;</span><br><span class="line">        my $unit = $_;</span><br><span class="line">        my @array = split(/\n/, $unit);</span><br><span class="line">        my $seq_name = shift @array;</span><br><span class="line">        my $seq=&#x27;&#x27;;</span><br><span class="line">        foreach my $line (@array)&#123;</span><br><span class="line">                $seq = $seq . $line;</span><br><span class="line">        &#125;</span><br><span class="line">        #print  &quot;\&gt;$seq_name\n$seq\n&quot;; exit;</span><br><span class="line">        #print &quot;$_n&quot;; exit;</span><br><span class="line">        if($seq =~ /\*[A-Z]/i)&#123;    ### 如果中间出现终止密码子，将剔除这个序列，某物钟一共有大约100条这样的序列</span><br><span class="line">                #print &quot;\&gt;$seq_name\n$seq\n&quot;;</span><br><span class="line">        &#125;</span><br><span class="line">        else&#123;</span><br><span class="line">                #$seq =~ s/\*//;           ### 去掉尾部的星号</span><br><span class="line">                print &quot;\&gt;$seq_name\n&quot;;</span><br><span class="line">                #print &quot;\&gt;$seq_name\n$seq\n&quot;;</span><br><span class="line">                # 如何更美观的输出蛋白序列，每行100个氨基酸</span><br><span class="line">                my $len = length($seq)/100;</span><br><span class="line">                #print &quot;$len\n&quot;;</span><br><span class="line">                for(my $i=0; $i&lt;$len; $i++)&#123;</span><br><span class="line">                        my $sub_line = substr($seq, $i*100, 100);</span><br><span class="line">                        $seq =~ s/\*//;           ### 去掉尾部的可能的星号</span><br><span class="line">                        $sub_line =~ s/\*//;</span><br><span class="line">                        print &quot;$sub_line\n&quot;;</span><br><span class="line">                &#125;</span><br><span class="line">        &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">close INPUT;</span><br><span class="line">close OUTPUT;</span><br></pre></td></tr></table></figure>
<p>用 orthofindr 程序寻找物种间的同源基因。我切换到的 Python37 环境，因为 orthofinder 是在这个环境下安装的，之前的博文有记录。</p>
<p>运行命令 <code>orthofinder -t 8 -f orthofinder_data_4 -S diamond</code> 即可进行同源基因寻找。生成的目录下会包含同源基因的列表和 Group 分组情况。下面截取的是程序刚运行的目录，所以只有部分数据呈现。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">$ tree orthofinder_data_4</span><br><span class="line">orthofinder_data_4</span><br><span class="line">├── XXX.fasta -&gt; ../pep_database/XXXXXXXX.pep.fa</span><br><span class="line">├── Arachis_ipaensis.pep.fa -&gt; ../pep_database/Arachis_ipaensis.pep.fa</span><br><span class="line">├── Cajanus_cajan.pep.fa -&gt; ../pep_database/Cajanus_cajan.pep.fa</span><br><span class="line">├── Lja.fasta -&gt; ../pep_database/Lotus_japonicus.pep.fa</span><br><span class="line">├── Lupinus_angustifolius.pep.fa -&gt; ../pep_database/Lupinus_angustifolius.pep.fa</span><br><span class="line">├── Mtr.fasta -&gt; ../pep_database/Medicago_truncatula.pep.fa</span><br><span class="line">├── OrthoFinder</span><br><span class="line">│   └── Results_Oct17</span><br><span class="line">│       ├── Log.txt</span><br><span class="line">│       └── WorkingDirectory</span><br><span class="line">│           ├── diamondDBSpecies0.dmnd</span><br><span class="line">│           ├── diamondDBSpecies1.dmnd</span><br><span class="line">│           ├── diamondDBSpecies2.dmnd</span><br><span class="line">│           ├── diamondDBSpecies3.dmnd</span><br><span class="line">│           ├── diamondDBSpecies4.dmnd</span><br><span class="line">│           ├── diamondDBSpecies5.dmnd</span><br><span class="line">│           ├── diamondDBSpecies6.dmnd</span><br><span class="line">│           ├── SequenceIDs.txt</span><br><span class="line">│           ├── Species0.fa</span><br><span class="line">│           ├── Species1.fa</span><br><span class="line">│           ├── Species2.fa</span><br><span class="line">│           ├── Species3.fa</span><br><span class="line">│           ├── Species4.fa</span><br><span class="line">│           ├── Species5.fa</span><br><span class="line">│           ├── Species6.fa</span><br><span class="line">│           ├── Species7.fa</span><br><span class="line">│           └── SpeciesIDs.txt</span><br><span class="line">├── Psa.fasta -&gt; ../pep_database/Pisum_sativum.pep.fa</span><br><span class="line">└── Van.fasta -&gt; ../pep_database/Vigna_angularis.pep.fa</span><br></pre></td></tr></table></figure>
<ol start="2">
<li>序列比对，即 codon alignment，依据密码子来进行比对，比直接依据 DNA 比对更准确，比对结果存储为 paml 格式。PAL2NAL 软件可以设定输出格式，默认为 clustal 格式，需要设置为paml。</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">#!/bin/bash</span><br><span class="line">set -e</span><br><span class="line">set -u</span><br><span class="line">set -o pipefail</span><br><span class="line">for i in `cat commn_genes_list`; do</span><br><span class="line">        echo [`date +&quot;%Y-%m-%d %H:%M:%S&quot;`] &quot;#&gt; sequence convertion and alignment: $&#123;i&#125;.fa&quot;;</span><br><span class="line">        cp ../01_fasta_files/$&#123;i&#125;.fa $&#123;i&#125;.nuc;</span><br><span class="line">        echo [`date +&quot;%Y-%m-%d %H:%M:%S&quot;`] &quot;#&gt; Proceccing $&#123;i&#125;.nuc ...&quot;;</span><br><span class="line">        iTools Fatools CDS2Pep -InCDS $&#123;i&#125;.nuc -OutPut $&#123;i&#125;.pep;</span><br><span class="line">        gunzip *gz;</span><br><span class="line">        clustalw2 -TYPE=PROTEIN -INFILE=$&#123;i&#125;.pep -ALIGN -OUTFILE=$&#123;i&#125;.pep.clustalw;</span><br><span class="line">        ~/local/app/PAL2NAL/pal2nal.pl $&#123;i&#125;.pep.clustalw $&#123;i&#125;.nuc -output paml -nogap &gt; $&#123;i&#125;.codon</span><br><span class="line">        echo;</span><br><span class="line">done</span><br></pre></td></tr></table></figure>
<p>一个基因 codon 比对的结果 paml 格式（the PHYLIP format）如下所示，首行两个数字，一个为物种数目，另一个为碱基数目。随后几行分布是物种名和去除 gap 的碱基序列。对于多个基因，可以直接用 cat 命令连接成一个文件，作为下一步 codeml 的输入。The first line contains the number of species and the sequence length (possibly followed by option characters). For codon sequences (codeml with seqtype = 1), the sequence length in the sequence file refers to the number of nucleotides rather than the number of codons.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">7  390</span><br><span class="line">Hsa_Human        AAGGTCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Hla_gibbon       AAGGTCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Cgu/Can_colobus  AAGATCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Pne_langur       AAGATCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Mmu_rhesus       AAGATCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Ssc_squirrelM    AAGGTCTTCGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Cja_marmoset     AAGGTCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br></pre></td></tr></table></figure>
<ol start="3">
<li>准备一个无根树</li>
</ol>
<p>我之前用过的无根树如下。newick格式，注意一定以分号结尾，否则 codeml 运行的结果中不会出现 Omema 值和 lnL 值。据我自己的经验，分析正向选择，有6~8个物种为宜，少了可能不准确，多了可能时间会很长。如果用有根树，codeml 会报错。无根树怎判断，首先可以用 Figtree 软件打开下面这个文件“unrooted_tree.txt”文件，默认的“rectangular form layput”显示下，从左侧起有 3 个分支，其实也可以直接肉眼看下面的文本内容，从最外层的括号观察，里面包含了 3 个分支。而root tree的图从最左侧开始看，只有2个分支。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$cat unrooted_tree.txt</span><br><span class="line">(((Vicia_faba, Medicago_truncatula), Lotus_japonicus),(Millettia_pinnata, ((Vigna_angularis, Vigna_radiata), Glycine_max)), Ammopiptanthus_mongolicus #1);</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/home.php?mod=space&amp;uid=1339458&amp;do=blog&amp;id=1023521"><strong>其他人的经验 1</strong></a>：制作tree文件。tree文件最简单的方法就是直接用Clustal制作，推荐 <a target="_blank" rel="noopener" href="http://www.genome.jp/tools/clustalw/">http://www.genome.jp/tools/clustalw/</a>。将序列提交上去，点Execute Multiple Alignment，然后将结果里的dnd文件下载，改名为自己想要的文件+后缀名“.trees”。我这里保存为n1.trees。但是上Phylogenetics课程的时候问了老师一句关于拿Clustal建树用于PAML分析，得到的回答是:  Never, never use Clustal to build a tree。所以还是老老实实用正经的建树工具来做Newick格式的tree文件吧。</p>
<p><a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3036"><strong>其他人的经验 2</strong></a>：输入一个树文件input.trees。该文件可以是有根树，也可以是无根树，也可以带有枝长信息。codeml命令用于分析各分枝上的omega值，若是有根树的话，对于分析root节点所在的分枝，其结果可能是不对的。因此，作者推荐使用无根树进行正选择分析。无根树是3分叉的树，最少需要3个物种；而3个物种的无根数有3个分枝（2n – 3），对其进行分析，仅能得到某个物种所对应分枝的omega值，这种情况下没太大意义；而使用4个物种的无根树进行分析，可以得到两个物种共同祖先分枝的omega值。故作者推荐使用至少4~5个物种进行分析。一个树文件内容示例。文件内容分两行：第一行表述树中有多少个物种，共计1个树，两个数值之间用任意长度的空分割。此外，Newick格式的树尾部一定要有分号，没有的话程序可能不能正常运行。</p>
<p><a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1Qx411m7Bn?from=search&amp;seid=9774626637377853773&amp;spm_id_from=333.337.0.0"><strong>其他人的经验 3</strong></a>：如果要计算某个分支，可以直接在分支后面加一个“$1”，即美元符号。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">7  1</span><br><span class="line">((Hsa_Human, Hla_gibbon) #1,((Cgu/Can_colobus, Pne_langur) #1, Mmu_rhesus), (Ssc_squirrelM, Cja_marmoset));</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">((Hsa_Human, Hla_gibbon), ((Cgu/Can_colobus, Pne_langur) $1, Mmu_rhesus), (Ssc_squirrelM, Cja_marmoset));</span><br></pre></td></tr></table></figure>
<ol start="4">
<li>准备 codeml 的配置文件</li>
</ol>
<p>下面是从 &lt;<a target="_blank" rel="noopener" href="http://chenlianfu.com">chenlianfu.com</a>&gt; 摘抄的一个配置案例：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line">*输入输出参数：</span><br><span class="line">      seqfile = input.txt    *设置输入的多序列比对文件路径。</span><br><span class="line">     treefile = input.trees  *设置输入的树文件路径。</span><br><span class="line">      outfile = mlc          *设置输出文件路径。</span><br><span class="line">        noisy = 9            *设置输出到屏幕上的信息量等级：0，1，2，3，9。</span><br><span class="line">      verbose = 0            *设置输出到结果文件中的信息量等级：0，精简模式结果（推荐）；1，输出详细信息，包含碱基序列；2，输出更多信息。</span><br><span class="line">        getSE = 0            *设置是否计算并获得各参数的标准误：0，不需要；1，需要。</span><br><span class="line"> RateAncestor = 0            *设置是否计算序列中每个位点的替换率：0，不需要；1，需要。当设置成1时，会在结果文件rates中给出各个位点的碱基替换率；同时也会进行祖先序列的重构，在结果文件rst中有体现。</span><br><span class="line"></span><br><span class="line">*数据使用说明参数：</span><br><span class="line">      runmode = 0      *设置程序获取进化树拓扑结构的方法：0，直接从输入文件中获取进化树拓扑结构（推荐）；1，程序以输入文件中的多分叉树作为起始树，并使用星状分解法搜索最佳树；2，程序直接以星状树作为起始树，并使用星状分解法搜索最佳树；3，使用逐步添加法搜索最佳树；4，使用简约法获取起始树，再进行邻近分枝交换寻找最优树；5，以输入文件中的树作为起始树，再进行邻近分枝交换寻找最优树；-2，对密码子序列进行两两比较并使用ML方法计算DnDs，或对蛋白序列进行两两比较计算ML距离，而不进行其它参数（枝长和omega等）的计算。</span><br><span class="line">* fix_blength = 1      *设置程序处理输入树文件中枝长数据的方法：0，忽略输入树文件中的枝长信息；-1，使用一个随机起始进行进行计算；1，以输入的枝长信息作为初始值进行ML迭代分析，此时需要注意输入的枝长信息是碱基替换率，而不是碱基替换数；2，不需要使用ML迭代计算枝长，直接使用输入的枝长信息，需要注意，若枝长信息和序列信息不吻合可能导致程序崩溃；3，让ML计算出的枝长和输入的枝长呈正比。</span><br><span class="line">      seqtype = 1      *设置输入的多序列比对数据的类型：1，密码子数据；2，氨基酸数据；3，输入数据虽然为密码子序列，但先转换为氨基酸序列后再进行分析。</span><br><span class="line">    CodonFreq = 2      *设置密码子频率的计算算法：0，表示除三种终止密码子频率为0，其余密码值频率全部为1/61；1，程序分别计算三个密码子位点的四种碱基频率，再算四种碱基频率在三个位点上的算术平均值，计算61种密码子频率时，使用该均值进行计算，这61种密码子频率之和不等于1，然后对其数据标准化，使其和为1，得到所有密码子的频率；2，程序分别计算三个密码子位点的四种碱基频率，计算61种密码子频率时，使用相应位点的碱基频率进行计算，这61种密码子频率之和不等于1，然后对其数据标准化，使其和为1，得到所有密码子的频率；3，直接使用观测到的各密码子的总的频数/所有密码值的总数，得到所有密码子的频率。一般选择第三种方法，设置CodonFre的值为2。第一种方法让所有密码子频率相等，不符合实际轻卡；第二种方法没有考虑到三种位点各碱基频率的差异，得到的密码子频率不准确；第三种方法能比较准确计算出各密码子的频率；第四种方法由输入数据得到真实的各密码子频率，但有时候有些非终止密码子的频率为0，可能不利于后续的计算。</span><br><span class="line">    cleandata = 1      *设置是否移除不明确字符（N、？、W、R和Y等）或含以后gap的列后再进行数据分析：0，不需要，但在序列两两比较的时候，还是会去除后进行比较；1，需要。</span><br><span class="line">*       ndata = 1      *设置输入的多序列比对的数据个数。</span><br><span class="line">        clock = 0      *设置进化树中各分支上的变异速率是否一致，服从分子钟理论：0，变异速率不一致，不服从分子钟理论（当输入数据中物种相差较远时，各分支变异速率不一致）；1，所有分枝具有相同的变异速率，全局上服从分子钟理论；2，进化树局部符和分子钟理论，程序认为除了指定的分支具有不同的进化速率，其它分枝都具有相同的进化速率，这要求输入的进化树信息中使用#来指定分支；3，对多基因数据进行联合分析。</span><br><span class="line">        Mgene = 0      *设置是否有多个基因的多序列比对信息输入，以及多各基因之间的参数是否一致：0，输入的多序列比对文件中仅包含一个基因时，或多个基因具有相同的Kappa和Pi参数；1，输入文件包含多个基因，这些基因之间是相互独立的（这些基因之间具有不同的Kappa和Pi值，且其进化树的枝长也不相关）；2，输入文件包含多个基因，这些基因具有相同的Kappa值，不同的Pi值；3，输入文件包含多个基因，这些基因具有相同的Pi值，不同的Kappa值；4，输入文件包含多个基因，这些基因具有不同的Kappa值和不同的Pi值。当值是2、3或4时，多个基因的进化树枝长虽然长度不一样，但是呈正比关系。这点和参数值等于1是不一样的。</span><br><span class="line">        icode = 0      *设置遗传密码。其值1-10和NCBI的1-11遗传密码规则对应：0，表示通用的遗传密码。</span><br><span class="line">   Small_Diff = .5e-6  *设置一个很小的值，一般位于1e-8到1e-5之间。推荐检测设置不同的值在比较其结果，需要该参数值对结果没什么影响。</span><br><span class="line"></span><br><span class="line">*位点替换模型参数：</span><br><span class="line">        model = 1            *若输入数据是密码子序列，该参数用于设置branch models，即进化树各分枝的omega值的分布：0，进化树上所有分枝的omega值一致；1，对每个分枝单独进行omega计算；2，设置多类omega值，根据树文件中对分枝的编号信息来确定类别，具有相同编号的分枝具有相同的omega值，没有编号的分枝具有相同的omega值，程序分别计算各编号和没有编号的omega值。</span><br><span class="line">                             *若输入数据是蛋白序列，或数据是密码子序列且seqtype值是3时，该参数用于设置氨基酸替换模型：0，Poisson；1，氨基酸替换率和氨基酸的观测频率成正比；2，从aaRatefile参数指定的文件路径中读取氨基酸替换率信息，这些信息是根据经验获得，并记录到后缀为.dat的配置文件中。这些经验模型(Empirical Models)文件位于PAML软件安装目录中的dat目录下，例如，Dayhoff(dayhoff.dat)、WAG(wag.dat)、LG(lg.dat)、mtMAN(mtman.dat)和mtREV24(mtREV24.dat)等；</span><br><span class="line">   aaRatefile = dat/wag.dat  *当对蛋白数据进行分析，且model = 2时，该参数生效，用于设置氨基酸替换模型。</span><br><span class="line">       aaDist = 0            *设置氨基酸之间的距离。</span><br><span class="line">      NSsites = 0            *输入数据时密码子序列时生效，用于设置site model，即序列各位点的omega值的分布：0，所有位点具有相同的omega值；1，各位点上的omega值小于1或等于1（服从中性进化neutral）；2，各位点上的omega值小于1、等于1或大于1（选择性进化selection）；3，discrete；4，freq；5:gamma；6，2gamma；7，beta；8，beta&amp;w；9，beta&amp;gamma；10，beta&amp;gamma+1；11，beta&amp;normal&gt;1；12，0&amp;2normal&gt;1；13，3normal&gt;0。</span><br><span class="line">                             *可以一次输入多个模型进行计算并比较，其结果输出的rst文件中。</span><br><span class="line">    fix_alpha = 1            *序列中不同的位点具有不同的碱基替换率，服从discrete-gamma分布，该模型通过alpha（shape参数）和ncatG参数控制。该参数设置是否给定一个alpha值：0，使用ML方法对alpha值进行计算；1，使用下一个参数设置一个固定的alpha值。</span><br><span class="line">	                     *对于密码子序列，当NSsites参数值不为0或model不为0时，推荐设置fix_alpha = 1且alpha = 0，即不设置alpha值，认为位点间的变异速率一致，否则程序报错。若设置了alpha值，则程序认为不同密码子位点的变异速率不均匀，且同时所有位点的omega值一致，当然各分枝的omega值也会一致，这时要求NSsites和model参数值都设置为0（这一般不是我们需要的分析，它不能进行正选择分析了）。</span><br><span class="line">        alpha = 0            *设置一个固定的alpha值或初始alpha值（推荐设置为0.5）。该值小于1，表示只有少数热点位置的替换率较高；该值越小，表示位点替换率在各位点上越不均匀；若设置fix_alpha = 1且alpha = 0则表示所有位点的替换率是恒定一致的。</span><br><span class="line">       Malpha = 0            *当输入的多序列比对结果中有多基因时，设置这些基因间的alpha值是否相等：0，分别对每个基因单独计算alpha值；1，所有基因的alpha值保持一致。</span><br><span class="line">        ncatG = 5            *序列中不同位点的变异速率服从GAMMA分布的，ncatG是其一个参数，一般设置为5，4，8或10，且序列条数越多，该值设置越大。</span><br><span class="line">                             *对于密码子序列，当NSites设置为3时，ncatG设置为3；当NSites设置为4时，ncatG设置为5；当NSites值设置&gt;=5时，ncatG值设置为10。</span><br><span class="line">    fix_kappa = 0            *设置是否给定一个Kappa值：0，通过ML迭代来估算Kappa值；1，使用下一个参数设置一个固定的Kappa值。</span><br><span class="line">        kappa = 2            *设置一个固定的Kappa值，或一个初始的Kappa值。</span><br><span class="line">    fix_omega = 0            *设置是否给定一个omega值：0，通过ML迭代来估算Kappa值；1，使用下一个参数设置一个固定的omega值。 </span><br><span class="line">        omega = .4           *设置一个固定的omega值，或一个初始的omega值。</span><br><span class="line">       method = 0            *设置评估枝长的ML迭代算法：0，使用PAML的老算法同时计算所有枝长，在clock = 0下有效；1，PAML新加入的算法，一次对一个枝长进行计算，该算法仅在clock参数值为1，2或3下工作。</span><br></pre></td></tr></table></figure>
<p>下面是我自己用的一个配置</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">#$ cat codeml-M0.ctl</span><br><span class="line">      seqfile = mtGenes.aligned.DNA.paml      * sequence data filename</span><br><span class="line">     treefile = unrooted_tree.txt      * tree structure file name</span><br><span class="line">      outfile = mtGenes.substitution           * main result file name</span><br><span class="line"></span><br><span class="line">      runmode = 0  * 0: user tree;  1: semi-automatic;  2: automatic</span><br><span class="line">                   * 3: StepwiseAddition; (4,5):PerturbationNNI; -2: pairwise</span><br><span class="line"></span><br><span class="line">      seqtype = 1  * 1:codons; 2:AAs; 3:codons--&gt;AAs</span><br><span class="line">    CodonFreq = 2  * 0:1/61 each, 1:F1X4, 2:F3X4, 3:codon table</span><br><span class="line"></span><br><span class="line">        ndata = 30 * number of gene alignments to be analysed</span><br><span class="line">        clock = 0  * 0:no clock, 1:clock; 2:local clock; 3:CombinedAnalysis</span><br><span class="line"></span><br><span class="line">        model = 2  * models for codons: 0:one, 1:b, 2:2 or more dN/dS ratios for branches</span><br><span class="line"></span><br><span class="line">      NSsites = 0  * 0:one w;1:neutral;2:selection; 3:discrete;4:freqs;</span><br><span class="line">                   * 5:gamma;6:2gamma;7:beta;8:beta&amp;w;9:beta&amp;gamma;</span><br><span class="line">                   * 10:beta&amp;gamma+1; 11:beta&amp;normal&gt;1; 12:0&amp;2normal&gt;1;</span><br><span class="line">                   * 13:3normal&gt;0</span><br><span class="line"></span><br><span class="line">        icode = 0  * 0:universal code; 1:mammalian mt; 2-10:see below</span><br><span class="line"></span><br><span class="line">    fix_omega = 0  * 1: omega or omega_1 fixed, 0: estimate </span><br><span class="line">        omega = .4 * initial or fixed omega, for codons or codon-based AAs</span><br><span class="line"></span><br><span class="line">    cleandata = 0  * remove sites with ambiguity data (1:yes, 0:no)?</span><br><span class="line"></span><br><span class="line">* Genetic codes: 0:universal, 1:mammalian mt., 2:yeast mt., 3:mold mt.,</span><br><span class="line">* 4: invertebrate mt., 5: ciliate nuclear, 6: echinoderm mt., </span><br><span class="line">* 7: euplotid mt., 8: alternative yeast nu. 9: ascidian mt., </span><br><span class="line">* 10: blepharisma nu.</span><br><span class="line">* These codes correspond to transl_table 1 to 11 of GENEBANK.</span><br></pre></td></tr></table></figure>
<p>下面是结果的展示。可以看到 w (dN/dS) for branches:  0.37585 0.22840，尾部的0.2284就是我们设置#1分支的ω，而第一个0.37585是其他分支的ω。由于所设定的物种的ω小于1，所以这个基因在该物种不是正向选择。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br></pre></td><td class="code"><pre><span class="line">TREE #  1:  (((8, 7), 5), (4, ((1, 2), 3)), 6);   MP score: 160</span><br><span class="line">check convergence..</span><br><span class="line">lnL(ntime: 13  np: 16):  -3122.716855      +0.000000</span><br><span class="line">   9..10   10..11   11..8    11..7    10..5     9..12   12..4    12..13   13..14   14..1    14..2    13..3     9..6  </span><br><span class="line"> 0.008466 0.038271 0.303083 0.013331 0.024760 0.007420 0.033139 0.020543 0.021931 0.003298 0.005453 0.018905 0.029024 0.643248 0.375846 0.228401</span><br><span class="line"></span><br><span class="line">Note: Branch length is defined as number of nucleotide substitutions per codon (not per neucleotide site).</span><br><span class="line"></span><br><span class="line">tree length =   0.52763</span><br><span class="line"></span><br><span class="line">(((8: 0.303083, 7: 0.013331): 0.038271, 5: 0.024760): 0.008466, (4: 0.033139, ((1: 0.003298, 2: 0.005453): 0.021931, 3: 0.018905): 0.020543): 0.007420, 6: 0.029024);</span><br><span class="line"></span><br><span class="line">(((Vicia_faba: 0.303083, Medicago_truncatula: 0.013331): 0.038271, Lotus_japonicus: 0.024760): 0.008466, (Millettia_pinnata: 0.033139, ((Vigna_angularis: 0.003298, Vigna_radiata: 0.005453): 0.021931, Glycine_max: 0.018905): 0.020543): 0.007420, Ammopiptanthus_mongolicus: 0.029024);</span><br><span class="line"></span><br><span class="line">Detailed output identifying parameters</span><br><span class="line"></span><br><span class="line">kappa (ts/tv) =  0.64325</span><br><span class="line"></span><br><span class="line">w (dN/dS) for branches:  0.37585 0.22840</span><br><span class="line"></span><br><span class="line">dN &amp; dS for each branch</span><br><span class="line"></span><br><span class="line"> branch          t       N       S   dN/dS      dN      dS  N*dN  S*dS</span><br><span class="line"></span><br><span class="line">   9..10     0.008  1077.6   338.4  0.3758  0.0020  0.0054   2.2   1.8</span><br><span class="line">  10..11     0.038  1077.6   338.4  0.3758  0.0091  0.0243   9.8   8.2</span><br><span class="line">  11..8      0.303  1077.6   338.4  0.3758  0.0723  0.1924  77.9  65.1</span><br><span class="line">  11..7      0.013  1077.6   338.4  0.3758  0.0032  0.0085   3.4   2.9</span><br><span class="line">  10..5      0.025  1077.6   338.4  0.3758  0.0059  0.0157   6.4   5.3</span><br><span class="line">   9..12     0.007  1077.6   338.4  0.3758  0.0018  0.0047   1.9   1.6</span><br><span class="line">  12..4      0.033  1077.6   338.4  0.3758  0.0079  0.0210   8.5   7.1</span><br><span class="line">  12..13     0.021  1077.6   338.4  0.3758  0.0049  0.0130   5.3   4.4</span><br><span class="line">  13..14     0.022  1077.6   338.4  0.3758  0.0052  0.0139   5.6   4.7</span><br><span class="line">  14..1      0.003  1077.6   338.4  0.3758  0.0008  0.0021   0.8   0.7</span><br><span class="line">  14..2      0.005  1077.6   338.4  0.3758  0.0013  0.0035   1.4   1.2</span><br><span class="line">  13..3      0.019  1077.6   338.4  0.3758  0.0045  0.0120   4.9   4.1</span><br><span class="line">   9..6      0.029  1077.6   338.4  0.2284  0.0054  0.0234   5.8   7.9</span><br><span class="line"></span><br><span class="line">tree length for dN:       0.1243</span><br><span class="line">tree length for dS:       0.3400</span><br><span class="line">dS tree:</span><br><span class="line">(((Vicia_faba: 0.192424, Medicago_truncatula: 0.008464): 0.024298, Lotus_japonicus: 0.015720): 0.005375, (Millettia_pinnata: 0.021039, ((Vigna_angularis: 0.002094, Vigna_radiata: 0.003462): 0.013924, Glycine_max: 0.012003): 0.013043): 0.004711, Ammopiptanthus_mongolicus: 0.023436);</span><br><span class="line">dN tree:</span><br><span class="line">(((Vicia_faba: 0.072322, Medicago_truncatula: 0.003181): 0.009132, Lotus_japonicus: 0.005908): 0.002020, (Millettia_pinnata: 0.007908, ((Vigna_angularis: 0.000787, Vigna_radiata: 0.001301): 0.005233, Glycine_max: 0.004511): 0.004902): 0.001770, Ammopiptanthus_mongolicus: 0.005353);</span><br><span class="line"></span><br><span class="line">w ratios as labels for TreeView:</span><br><span class="line">(((Vicia_faba #0.3758 , Medicago_truncatula #0.3758 ) #0.3758 , Lotus_japonicus #0.3758 ) #0.3758 , (Millettia_pinnata #0.3758 , ((Vigna_angularis #0.3758 , Vigna_radiata #0.3758 ) #0.3758 , Glycine_max #0.3758 ) #0.3758 ) #0.3758 , Ammopiptanthus_mongolicus #0.2284 );</span><br><span class="line"></span><br><span class="line">Time used:  0:05</span><br></pre></td></tr></table></figure>
<p>上述 ctl 文件中 model = 2，运行得到一个结果文件。然后将其中的 model 参数改为 0，再次运行，得到新的结果文件。那么现在有三个文件是我需要的。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">all.genes.list</span><br><span class="line">result.branch_model_0.mcl</span><br><span class="line">result.branch_model_2.mcl</span><br></pre></td></tr></table></figure>
<p>我根据这三个文件，得到了下面的表格。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">$ perl chi2_calculator.pl </span><br><span class="line"># The number of &quot;genes, lnL0 number, lnL1 number&quot; are the same.</span><br><span class="line"># If you see this info, the data is probably OK.</span><br><span class="line">#Number Group   lnL0    lnL1    Omega_branch    Omega_controlP_value</span><br><span class="line">Number:1        atp1    lnL0:-3145.827397       lnL1:-3122.716855       Omega_branch:0.22840    Omega_control:0.37585   P_value:0.000000000</span><br><span class="line">Number:2        atp4    lnL0:-930.895378        lnL1:-930.038146        Omega_branch:0.45618    Omega_control:0.52692   P_value:0.424335018</span><br><span class="line">Number:3        atp6    lnL0:-1453.909109       lnL1:-1446.498209       Omega_branch:0.86281    Omega_control:0.37541   P_value:0.000604626</span><br><span class="line">Number:4        atp8    lnL0:-924.486129        lnL1:-914.009795        Omega_branch:1.75174    Omega_control:0.19043   P_value:0.000028196</span><br><span class="line">Number:5        atp9    lnL0:-333.236113        lnL1:-331.374307        Omega_branch:0.00010    Omega_control:0.39757   P_value:0.155391739</span><br><span class="line">Number:6        ccmB    lnL0:-1023.961013       lnL1:-1022.534291       Omega_branch:999.00000  Omega_control:0.56135   P_value:0.240094664</span><br><span class="line">Number:7        ccmC    lnL0:-1131.746218       lnL1:-1129.394721       Omega_branch:999.00000  Omega_control:0.39406   P_value:0.095226501</span><br><span class="line">Number:8        ccmFC   lnL0:-1950.197350       lnL1:-1948.956865       Omega_branch:0.42724    Omega_control:0.88398   P_value:0.289243901</span><br><span class="line">Number:9        ccmFN   lnL0:-2737.722421       lnL1:-2737.051425       Omega_branch:1.13399    Omega_control:0.65455   P_value:0.511199170</span><br><span class="line">Number:10       cob     lnL0:-1836.351940       lnL1:-1826.480046       Omega_branch:0.16318    Omega_control:0.22020   P_value:0.000051605</span><br><span class="line">Number:11       cox1    lnL0:-2381.442730       lnL1:-2354.196627       Omega_branch:0.09805    Omega_control:0.04898   P_value:0.000000000</span><br><span class="line">Number:12       cox3    lnL0:-1220.568961       lnL1:-1218.320955       Omega_branch:0.00010    Omega_control:0.53125   P_value:0.105609600</span><br><span class="line">Number:13       matR    lnL0:-3293.272506       lnL1:-3288.780620       Omega_branch:1.06998    Omega_control:0.46260   P_value:0.011199502</span><br><span class="line">Number:14       mttB    lnL0:-1110.395753       lnL1:-1109.356198       Omega_branch:1.07609    Omega_control:0.44712   P_value:0.353612004</span><br><span class="line">Number:15       nad1    lnL0:-1410.749338       lnL1:-1404.082741       Omega_branch:0.34165    Omega_control:0.13677   P_value:0.001272722</span><br><span class="line">Number:16       nad2    lnL0:-2199.786871       lnL1:-2187.259279       Omega_branch:0.80075    Omega_control:0.12534   P_value:0.000003625</span><br><span class="line">Number:17       nad3    lnL0:-537.855553        lnL1:-532.299091        Omega_branch:0.00010    Omega_control:0.27379   P_value:0.003862417</span><br><span class="line">Number:18       nad4    lnL0:-2549.647893       lnL1:-2546.428114       Omega_branch:0.25913    Omega_control:0.61043   P_value:0.039963889</span><br><span class="line">Number:19       nad4L   lnL0:-420.857873        lnL1:-413.049727        Omega_branch:0.00010    Omega_control:0.00010   P_value:0.000406411</span><br><span class="line">Number:20       nad5    lnL0:-2986.298356       lnL1:-2975.844248       Omega_branch:0.15621    Omega_control:0.26908   P_value:0.000028830</span><br><span class="line">Number:21       nad6    lnL0:-1099.896407       lnL1:-1092.361003       Omega_branch:0.54957    Omega_control:0.24816   P_value:0.000533846</span><br><span class="line">Number:22       nad7    lnL0:-1766.620230       lnL1:-1765.116602       Omega_branch:0.51002    Omega_control:0.42390   P_value:0.222322111</span><br><span class="line">Number:23       nad9    lnL0:-1032.089035       lnL1:-1031.583127       Omega_branch:999.00000  Omega_control:0.86278   P_value:0.602957841</span><br><span class="line">Number:24       rpl16   lnL0:-895.556806        lnL1:-891.758505        Omega_branch:0.27993    Omega_control:0.35303   P_value:0.022408812</span><br><span class="line">Number:25       rpl5    lnL0:-920.541028        lnL1:-919.686869        Omega_branch:0.86954    Omega_control:0.50538   P_value:0.425641005</span><br><span class="line">Number:26       rps10   lnL0:-603.758104        lnL1:-600.272712        Omega_branch:999.00000  Omega_control:0.30933   P_value:0.030641745</span><br><span class="line">Number:27       rps12   lnL0:-613.237087        lnL1:-609.335187        Omega_branch:999.00000  Omega_control:0.21780   P_value:0.020203488</span><br><span class="line">Number:28       rps14   lnL0:-493.189267        lnL1:-491.210035        Omega_branch:0.26530    Omega_control:0.31987   P_value:0.138175315</span><br><span class="line">Number:29       rps3    lnL0:-2982.678697       lnL1:-2978.575083       Omega_branch:0.54439    Omega_control:0.53311   P_value:0.016512890</span><br><span class="line">Number:30       rps4    lnL0:-1761.670696       lnL1:-1760.205396       Omega_branch:0.44191    Omega_control:0.58927   P_value:0.231008678</span><br></pre></td></tr></table></figure>
<p>附上 <code>chi2_calculator.pl</code> 程序的代码。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/perl</span><br><span class="line">#/*************************************************************************</span><br><span class="line">#	&gt; File Name: chi2_calculator.pl</span><br><span class="line">#	&gt; Author: fenglei</span><br><span class="line">#	&gt; Mail: fengleiluck@gmail.com</span><br><span class="line">#	&gt; Created Time: Thu 07 Oct 2021 01:26:04 PM CST</span><br><span class="line"># ************************************************************************/</span><br><span class="line"></span><br><span class="line">open GENE_LIST, &quot;&lt;30_genes.list&quot; or die &quot;Can not open the file!\n&quot;;</span><br><span class="line">open NONMODEL, &quot;&lt;mtGenes.branch_model_0.mcl&quot; or die &quot;Can not open the file!\n&quot;;  # 找到这样的行：lnL(ntime: 13  np: 16):  -3122.716855      +0.000000  </span><br><span class="line">open ALTERNATIVE, &quot;&lt;mtGenes.branch_model_2.mcl&quot; or die &quot;Can not open the file!\n&quot;;</span><br><span class="line"></span><br><span class="line">#####  perl chi2_calculator.pl &gt; out</span><br><span class="line">#####  依据上面三个文件，依次计算P值（通过似然率检验（LRT）（卡方检验）获得p值确定零假设模型和备选模型之间是否存在差异）</span><br><span class="line">#      能否同时输出 omega 值？  Alternative 文件里面有一行“w (dN/dS) for branches:  0.37585 0.22840” 第二个数字是我需要的</span><br><span class="line">my @genes;</span><br><span class="line">while(&lt;GENE_LIST&gt;)&#123;</span><br><span class="line">	chomp $_;</span><br><span class="line">	push @genes, $_;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">my @lnL0;</span><br><span class="line">while(&lt;NONMODEL&gt;)&#123;</span><br><span class="line">	chomp $_;</span><br><span class="line">	my $record = $_;</span><br><span class="line">	if($record =~ m/lnL/)&#123;</span><br><span class="line">		my @array = split(/\s+/, $record);</span><br><span class="line">		#print &quot;lnL0: $array[4]\n&quot;; exit;</span><br><span class="line">		push @lnL0, $array[4];</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">my @lnL1;</span><br><span class="line">my @omega_branch;</span><br><span class="line">my @omega_control;</span><br><span class="line">while(&lt;ALTERNATIVE&gt;)&#123;</span><br><span class="line">	chomp $_;</span><br><span class="line">	my $record = $_;</span><br><span class="line">	if($record =~ m/lnL/)&#123;   # 找到这样的行：lnL(ntime: 13  np: 16):  -3122.716855      +0.000000 </span><br><span class="line">		my @array = split(/\s+/, $record);</span><br><span class="line">		push @lnL1, $array[4];</span><br><span class="line">	&#125;</span><br><span class="line">	if($record =~ m/w\s\(dN\/dS\)\sfor\sbranches\:/)&#123;   # 找到这样的行：w (dN/dS) for branches:  0.37585 0.22840” 第二个数字是我需要的选定brach的omega</span><br><span class="line">		my @array = split(/\s+/, $record);</span><br><span class="line">		push @omega_branch,  $array[-1];</span><br><span class="line">		push @omega_control, $array[-2];</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">if($#genes == $#lnL0 &amp;&amp; $#genes == $#lnL1)&#123;</span><br><span class="line">	print &quot;# The number of \&quot;genes, lnL0 number, lnL1 number\&quot; are the same.\n# If you see this info, the data is probably OK.\n&quot;;</span><br><span class="line">	#exit;</span><br><span class="line">&#125;</span><br><span class="line">else&#123;</span><br><span class="line">	print &quot;# The number of \&quot;genes, lnL0 number, lnL1 number\&quot; are NOT the same.\n# If you see this info, the data is probably NOT OK.\n# You should have a check.\n&quot;;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">print &quot;#Number\tGroup\tlnL0\tlnL1\tOmega_branch\tOmega_control\tP_value\n&quot;;</span><br><span class="line"></span><br><span class="line">for(my $count=0; $count&lt;=$#genes; $count++)&#123;</span><br><span class="line">	$num = $count + 1;</span><br><span class="line">	print &quot;Number:$num\t$genes[$count]\tlnL0:$lnL0[$count]\tlnL1:$lnL1[$count]\tOmega_branch:$omega_branch[$count]\tOmega_control:$omega_control[$count]\t&quot;;</span><br><span class="line">	my $likelihood = 2 * abs( $lnL0[$count] - $lnL1[$count] );</span><br><span class="line">	my $df  = 2;  ### ??? df=1 or df=2 ???</span><br><span class="line">	system(&quot;~/local/app/PAML4.8/bin/chi2 $df $likelihood &gt; /tmp/chi2_result.txt&quot;);  # 三行的一个文件：首行空格，次行“df =  2  prob = 0.016512890 = 1.651e-02”，末行空格。需要提取P值</span><br><span class="line">	open CHISQUARE, &quot;/tmp/chi2_result.txt&quot; or die &quot;Can not open the file!\n&quot;;</span><br><span class="line">	my $p_value = 1;</span><br><span class="line">	while(&lt;CHISQUARE&gt;)&#123;</span><br><span class="line">		chomp $_;</span><br><span class="line">		my $record = $_;</span><br><span class="line">		if($record =~ m/prob/)&#123;</span><br><span class="line">			my @array = split(/\s+/, $record);</span><br><span class="line">			$p_value = $array[-3]/2;  ### PAML文档（http://abacus.gene.ucl.ac.uk/software/pamlDOC.pdf ; Page 30），p值应该是chi2程序的结果除以2。chenlianfu也有如此记录）</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	print &quot;P_value:$p_value\n&quot;;</span><br><span class="line">	close CHISQUARE;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">=pod</span><br><span class="line">system(&quot;~/local/app/PAML4.8/bin/chi2 2 55&quot;);</span><br><span class="line">$df = 2;</span><br><span class="line">$likelihood = 23;</span><br><span class="line">system(&quot;~/local/app/PAML4.8/bin/chi2 $df $likelihood &gt;&gt; P_value.txt&quot;);</span><br><span class="line">print &quot;test\n&quot;;</span><br><span class="line"># print &quot;$a\n&quot;;  # 这命令输出的是 0，也就是 system 命令输出在屏幕的内容，不能传递给变量。</span><br><span class="line">=cut</span><br><span class="line"></span><br><span class="line">close GENE_LIST;</span><br><span class="line">close NONMODEL;</span><br><span class="line">close ALTERNATIVE;</span><br></pre></td></tr></table></figure>
<p><u><font color=blue>按照前文 Science 杂志那个反刍动物的研究论文资料，我这里的 branch model 实际上找到的是指定分支上的快速进化基因 Rapidly evolved genes。如果要做正向选择分析，要使用 branch-site mode，即参考下文的PAML四大模型之第三个模型，分别设置ModelA：“model = 2, NSsites = 2, fix_omega = 0, omega = 1.5” 和 Model A Null：“model = 2, NSsites = 2, fix_omega = 1, omega = 1”。</font></u></p>
<p>附<a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?m=201910">chenlianfu</a>关于正向选择分析步骤的总结：</p>
<p>(1) 收集该基因在多个物种中的 CDS 和 protein 序列（可以使用 orthoMCL 分析结果）。<br />
(2) 对该基因的蛋白序列进行多序列比对，再根据 CDS 序列转换成 Codon 序列比对结果。这个步骤需要自己编写一些程序来实现。<br />
(3) 根据 Codon 序列的比对结果构建系统发育树。推荐使用 RAxML 软件使用 ML 算法对所有的单拷贝同源基因进行物种树构建。然后，输入物种树信息、Codon 多序列比对信息，使用 codeml 进行 omega 计算和选择压模型检验。<br />
(4) 快速过滤非正选择基因：设置 runmode = -2 运行 codeml 命令对两两序列比较，使用 ML 方法计算 dNdS，若存在 omega 值大于1，则认为该基因可能属于正选择基因，进行后续分析。<br />
(5) 正选择基因的初步鉴定方法：设置 model = 0、NSsites = 1 2 运行 codeml 命令（site models）分别对正选择模型(M2a)和近中性进化模型(M1a)进行检测，若两种模型的似然值相差较大，通过自由度为 2 的卡方检验，可以确定该基因是否为正选择基因。从程序结果中（例如：lnL(ntime: 11 np: 16): -870.867266 +0.000000）可以找到 M2a 模型的似然值 l1 和 M1a 模型的似然值 l0，再计算 2Δl = 2(l1-l0)，通过命令“chi2 2 2Δl”根据卡方检验算出 p 值。此外，也可以设置 NSsites = 7 8 进行计算，则表示分别对 M8(beta &amp; omega) 和 M7(beta) 模型进行检测，和前者类似，若两种模型的似然值相差较大，通过自由度为 2 的卡方检验，可以确定该基因是否为正选择基因。根据 PAML 作者所说，M2a 和 M1a 的比较，比 M7-M8 的比较更严格。即若想得到更多的正选择基因，可以使用 M7_VS_M8 分析。此外，根据 M2 和 M8 模型的 BEB（Bayes empirical Bayes）分析结果，还可以得到在多序列比对结果中第一条序列上的正选择位点。根据 PAM L作者所说，这种 site models 之间的比较，能用于检测是否在序列上不同的位点具有不同的 omega 值，而不是用于正选择检测（We suggest that The M0-M3 comparison should be used as a test of variable ω among sites rather than a test of positive selection）。<br />
(6) <font color=blue>指定分化枝上的正选择基因鉴定方法：经过上一步初步鉴定后，设置 model = 2、NSsites = 2、fix_omega = 0、omega = 2.0 运行codeml命令（branch-site model A）；再设置 model = 2、NSsites = 2、fix_omega = 1、omega = 1 运行 codeml 命令（modified branch-site model A / null model）。进行这两种模型分析时，要求输入的树文件中对目标分化枝进行标注。对这两种模型进行 LRT 分析，计算 2Δl = 2(l1-l0)，注意是前者的似然值减后者（null model）的似然值；再使用自由度为 1 的卡方检验，通过命令“chi2 1 2Δl”计算出的值再除以 2，即得到 p 值。</font><br />
(7) <font color=blue>目标分化枝上的快速进化基因（Rapidly evolving gene）鉴定方法：设置 model = 0、NSsites = 0 运行 codeml 命令，再设置 model = 2、NSsites = 0 运行 codeml 命令（这时要求输入的树文件中对目标分化枝进行标注）</font>。然后比较两种运行模式下的似然值，通过自由度为 1 的卡方检验，可以鉴定该基因在目标分化枝的 omega 值和背景差异显著。</p>
<h3 id="一些测试与思考"><a class="markdownIt-Anchor" href="#一些测试与思考"></a> 一些测试与思考</h3>
<p>对于一个基因或者固定数目的少数基因，我按照一个参考流程 <a target="_blank" rel="noopener" href="https://figshare.com/articles/journal_contribution/worked_example_pdf/3839583">https://figshare.com/articles/journal_contribution/worked_example_pdf/3839583</a>，即可计算出 omega 值。</p>
<p>该 demo 用的是 6 个物种的单拷贝 orthologous，有 3000 多个基因。而我的实验里，想用 16 个物种运行单拷贝基因鉴定，只有 322 个基因，这样范围太小，怎么改善？</p>
<p>对于无根树文件而言，如果要指定运算某个分支的进化速率，需要在那个分支上加一个“#1”。这里的意思是该物种有一个进化速率 omega，而其他物种有统一的 omega。在另一个资料中，我发现还可以在tree文件中给指定的分支设定“$1”，比如<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1Qx411m7Bn?from=search&amp;seid=9774626637377853773&amp;spm_id_from=333.337.0.0">作者</a>用到四种海洋哺乳动物，四种陆地动物的 myoglobin 基因做的正向选择分析，把四种海洋动物标记为“$1”，这样运行结果就是指这四个鲸鱼具有统一的 omega。</p>
<p><img src="https://i.imgur.com/feqs2ew.png" alt="" /></p>
<p>比对的基因序列文件中，物种数目少于tree文件中的物种数目，似乎也可以运行得到结果？（有待确认？如果真的是这样，那我如果对若干个orthologous基因的正向选择做计算，只需要一个tree文件就可以，对于个别基因而言，确实个别物种的orthologous不会影响结果。）2021.10.08 做了一个测试，如果tree文件里面有8个物种，但是phylip格式的序列文件里面只有7个物种，codeml程序会报错的。</p>
<p>Model=0; NSsites= 7 8; 这样结果里面确实有两个 1nL 值，可以用来计算 P 值，但是没有特定 branch 的 ω，而我需要的是特定分支的 ω。这个模式确实无法实现我的目的，因为 Model=0 就表示所有分支的 Omega 都是固定的。</p>
<p>对于 codeml 的配置文件而言，里面的 model 可以设置为 0、1 或 2，这三个参数有什么区别？这个参数主要设定不同分支的ω是否一致。<font color=blue>当model=0，NSsites=0，那么表示假设所有分支、所有氨基酸位点的ω都是一致的。当 model=2, NSsites=0，即假设前景支与背景支的ω不同，出来的结果特定分支ω跟其他分支ω不同，此时其他分支的ω一致。对这两个不同的假设来做卡方检验，即可推算出前者为假的可能性（即 P 值），当 P &lt; 0.05， 可信度高。</font></p>
<p><img src="https://i.imgur.com/h89xpWB.png" alt="Model=0，假设所有分支ω一致" /></p>
<p><img src="https://i.imgur.com/64Ec8kB.png" alt="Model=2，假设前景支与背景支的ω不同" /></p>
<p><img src="https://i.imgur.com/R4qryV2.png" alt="对上述两个假设做卡方检验" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">...</span><br><span class="line">Model 7: beta (10 categories)</span><br><span class="line"></span><br><span class="line">TREE #  1:  (((8, 7), 5), (4, ((1, 2), 3)), 6);   MP score: 160</span><br><span class="line">lnL(ntime: 13  np: 16):  -3062.689720      +0.000000</span><br><span class="line">   9..10   10..11   11..8    11..7    10..5     9..12   12..4    12..13   13..14   14..1    14..2    13..3     9..6  </span><br><span class="line"> 0.007984 0.037210 0.329993 0.012402 0.023400 0.007531 0.031297 0.019721 0.020914 0.003075 0.005174 0.017900 0.027104 0.522045 0.005000 0.019948</span><br><span class="line"></span><br><span class="line">Note: Branch length is defined as number of nucleotide substitutions per codon (not per neucleotide site).</span><br><span class="line">...</span><br><span class="line">Model 8: beta&amp;w&gt;1 (11 categories)</span><br><span class="line"></span><br><span class="line">TREE #  1:  (((8, 7), 5), (4, ((1, 2), 3)), 6);   MP score: 160</span><br><span class="line">lnL(ntime: 13  np: 18):  -3041.864857      +0.000000</span><br><span class="line">   9..10   10..11   11..8    11..7    10..5     9..12   12..4    12..13   13..14   14..1    14..2    13..3     9..6  </span><br><span class="line"> 0.008053 0.038859 0.392264 0.011943 0.023468 0.008401 0.030708 0.020215 0.021091 0.002819 0.005071 0.017885 0.027144 0.504233 0.894161 0.007220 0.044307 4.278923</span><br><span class="line"></span><br><span class="line">Note: Branch length is defined as number of nucleotide substitutions per codon (not per neucleotide site).</span><br><span class="line"></span><br><span class="line">tree length =   0.60792</span><br><span class="line"></span><br><span class="line">(((8: 0.392264, 7: 0.011943): 0.038859, 5: 0.023468): 0.008053, (4: 0.030708, ((1: 0.002819, 2: 0.005071): 0.021091, 3: 0.017885): 0.020215): 0.008401, 6: 0.027144);</span><br><span class="line"></span><br><span class="line">(((Vicia_faba: 0.392264, Medicago_truncatula: 0.011943): 0.038859, Lotus_japonicus: 0.023468): 0.008053, (Millettia_pinnata: 0.030708, ((Vigna_angularis: 0.002819, Vigna_radiata: 0.005071): 0.021091, Glycine_max: 0.017885): 0.020215): 0.008401, Ammopiptanthus_mongolicus: 0.027144);</span><br><span class="line"></span><br><span class="line">Detailed output identifying parameters</span><br><span class="line"></span><br><span class="line">kappa (ts/tv) =  0.50423</span><br><span class="line"></span><br><span class="line">Parameters in M8 (beta&amp;w&gt;1):</span><br><span class="line">  p0 =   0.89416  p =   0.00722 q =   0.04431</span><br><span class="line"> (p1 =   0.10584) w =   4.27892</span><br><span class="line"> ...</span><br></pre></td></tr></table></figure>
<p>下面的方法（来自<a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3036">chenlianfu.com</a>）可以检测指定分支的 ω, 并且计算 P 值。我查阅了 PAML 的文档，确实是这么计算 P 值。</p>
<p><font color=blue>指定分化枝上的正选择基因鉴定方法：经过上一步初步鉴定后，设置 <code>model = 2、NSsites = 2、fix_omega = 0、omega = 2.0</code> 运行 codeml 命令（branch-site model A）；再设置 <code>model = 2、NSsites = 2、fix_omega = 1、omega = 1</code> 运行 codeml 命令（modified branch-site model A / null model）。进行这两种模型分析时，要求输入的树文件中对目标分化枝进行标注。对这两种模型进行LRT分析，计算 2Δl = 2(l1-l0)，注意是前者（branch-site model A）的似然值减后者（null model）的似然值；再使用自由度为1的卡方检验，通过命令“chi2 1 2Δl”计算出的值再除以2，即得到p值。</font></p>
<p>20211011 备注：上面卡方检验直接 <code>chi2 2 2Δl</code> 即可？不需要除以二吧？—— 查阅文档之后看到，要除以二。</p>
<p><img src="https://i.imgur.com/TV6Sr9x.jpeg" alt="" /></p>
<p>上面截图中调用 <code>chi2 2 2.71</code> 命令，得到 “df =  2  prob = 0.257947294 = 2.579e-01”，再除以二就是 P 值。</p>
<p>其实也可以用 R 来做计算，在R界面输入命令 <code>pchisq(2.71, 2, lower.tail = FALSE)</code> 得到结果：“0.2579473”。跟上面调用 chi2 程序的结果是一致的。</p>
<p>卡方检验是一种用途很广的计数资料的假设检验方法，由卡尔·皮尔逊提出。它属于非参数检验的范畴，主要是比较两个及两个以上样本率( 构成比）以及两个分类变量的关联性分析。其根本思想就是在于比较理论频数和实际频数的吻合程度或拟合优度问题。比如<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/42803826">我们想知道不吃晚饭对体重下降有没有影响</a>，即探讨吃晚饭与体重的关系，将一个人群分为“吃晚饭组”与“不吃晚饭组”</p>
<h3 id="学习笔记怎么通过-chi2-计算-p-值"><a class="markdownIt-Anchor" href="#学习笔记怎么通过-chi2-计算-p-值"></a> 学习笔记：怎么通过 chi2 计算 P 值？</h3>
<p>一开始接触 codeml 的时候，并不清楚怎么计算 P 值，经过阅读一些相关材料，获得 P 值计算方法，大致记录如下。</p>
<p>上面的 codeml 运行结果文件有这么一行：lnL(ntime: 13  np: 16):  -3122.716855      +0.000000</p>
<p>这个“-3122.716855”就有用，我们修改 ctl 文件的参数之后，再次运行 codeml 程序，将生成一个新的 lnL 值。依据前后两次 lnL 值，用下文提到的公式，就可以计算 P 值。</p>
<p><a target="_blank" rel="noopener" href="http://blog.sina.com.cn/s/blog_65ba09d90102x4um.html">http://blog.sina.com.cn/s/blog_65ba09d90102x4um.html</a> 上回讲到了用codeml计算选择位点的时候，同时选择M7和M8模型，这两个模型的差异进近在于 M8(beta &amp; ω) 模型比 M7 （beta）多了一个参数，所以可以依靠比较这两个模型来进行LRT。公式是这个：LRT = 2dl = abs(2 X (l1-l0)) （abs=绝对值）。Step1: 打开之前用codeml生成的 mcl 文件，找到 Model7, 和 Model8。两个lnL值想减，取绝对值，乘以2，这里是：2|-10302.87869-（-10332.64273）| = 59.52 (小数点后影响不大) Step2：打开PAML程序包中的 Chi2 程序, 分别输入自由度和刚才得到的数，自由度直接取 2 就好（这个是杨子恒自己说的，在PAML的文档里有写），后面写上Step1算出来的数字，得到的 p 小于 0.05 则结果比较可靠，这里的 p 非常小。</p>
<p><a target="_blank" rel="noopener" href="https://bcb.unl.edu/yyin/codeml.html">https://bcb.unl.edu/yyin/codeml.html</a> I write some simple ways to test likelihood ratio test on variable codon rates. These two models assume codon rates varied among all branchs. (1). M1a vs. M2a: M1a(model=0, NSsites=1), M2a(model=0, NSsites=2), the df = 2, LRT = 2dl = abs(2 X (l1- l0)). (2). M7 vs. M8: M1a(model=0, NSsites=7), M2a(model=0, NSsites=8), the df = 2, LRT = 2dl = abs(2 X (l1-l0)). The chi-square value can be calucated by PROGRAM CHI2.EXE in PAML.  If the p-value &lt;0.05, then we can conclude that some sites are under positive selection.</p>
<h3 id="paml-分析四大模型"><a class="markdownIt-Anchor" href="#paml-分析四大模型"></a> PAML 分析四大模型</h3>
<p>在paml分析里面，主要涉及四种模型：位点模型（site-model）、支模型（branch model）、支位点模型（branch site model）以及进化支模型。位点模型通常适用于检测某一支系普遍性、广泛性的正选择，这种正选择是由于位点持续改变所引起的，例如，适应多种病原体；支模型主要是检测某一支系是否存在快速进化、选择压力约束以及正选择，但却无法检测到正选择位点；<font color=blue>支位点模型比较准确与稳定，适用于检测某一支系断点性的正选择事件，此结果是由于适应某一时期环境改变所引起的，通常会保留在后代中</font>；进化支模型（clade model），则主要是判别不同物种之间是否存受分化选择压力作用，不局限于正选择，它一次可以标记多个分支进行比较。</p>
<ol>
<li><strong>位点模型</strong></li>
</ol>
<p>M0：假设所有位点具有相同的 dN/dS 值；<br />
M1a：假设存在两类位点—保守位点0 &lt; dN/dS &lt; 1，中性进化位点dN/dS = 1，并且估算这两类位点的比率（p0，p1）和ω值（ω0，ω1）；<br />
M2a：假设存在三类位点——纯化选择位点 dN/dS &lt; 1，中性进化位点dN/dS = 1与正选择位点dN/dS &gt; 1，并估算三类位点的比率（p0，p1，p2）；<br />
M3: 离散模型，假设所有位点的ω值呈离散分布；<br />
M7：假设所有位点0 &lt; ω &lt;1且呈现beta分布；<br />
M8：在M7模型的基础上，增加一类正选择位点（ω &gt;1）;<br />
M8a：与M8类似，只不过是将新增的ω固定为1；</p>
<p>位点模型的 Codeml.ctl 参数设置：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">model = 0                 </span><br><span class="line">NSsites = 0 1 2 3 7 8</span><br><span class="line">即分别比较如下模型（* 为正选择模型，前提是需要LRTs显著）：</span><br><span class="line">M3（分散比率） vs. M0（负选择）</span><br><span class="line">M2a* （正选择）vs. M1a （选择约束放松）</span><br><span class="line">M8*（正选择） vs. M7 （选择约束放松）</span><br><span class="line">M8* （正选择）vs. M8a（选择约束放松）</span><br></pre></td></tr></table></figure>
<ol start="2">
<li><strong>支模型</strong></li>
</ol>
<p>one rario：假设所有的进化谱系都具有相同的ω值；<br />
free ratio：假设所有的支系都具有独立的ω值；<br />
two ratio：假设前景支与背景支的ω不同</p>
<p>分支模型的Codeml.ctl参数设置：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">two ratio: model = 2, NSsites = 0  </span><br><span class="line">one ratio: model = 0, NSsites = 0</span><br><span class="line">free ratio: model = 1, NSsites = 0</span><br><span class="line">               </span><br><span class="line">one ratio（负选择） vs. free ratio（自由比率）</span><br><span class="line">one ratio（负选择） vs. two ratio （正选择）</span><br><span class="line">two ratio natural（中性进化） vs. two ratio （正选择）</span><br></pre></td></tr></table></figure>
<ol start="3">
<li><strong>支位点模型</strong></li>
</ol>
<p>假定位点间的ω值是变化的，同时也假定支系间的ω值是变化的。该模型主要用于检测前景支中正选择作用对部分位点的影响。<br />
modelA null（零假设）：ω值设定为固定值1<br />
modelA（备择假设）：估算其ω值是否大于1<br />
背景支与前景支具有相同的位点ω值：<br />
K0：前景支与背景支中的位点受到纯化选择0 &lt; ω &lt; 1；<br />
K1：前景支与背景支中的位点处于中性进化0 &lt; ω = 1;<br />
背景支与前景支具有不同的位点ω值：<br />
K2a：前景支处于中性进化，而背景支处于纯化选择；<br />
K2b：前景支受到正选择压力（ ω&gt;1 ），而背景支处于中性进化；<br />
支位点模型的的Codeml.ctl参数设置：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">ModelA：</span><br><span class="line">model = 2, NSsites = 2, fix_omega = 0, omega = 1.5</span><br><span class="line">ModelAnull：</span><br><span class="line">model = 2, NSsites = 2, fix_omega = 1, omega = 1</span><br><span class="line">即比较如下模型（* 为正选择模型，前提是需要LRTs显著）：</span><br><span class="line">ModelA*（正选择） vs. ModelAnull （中性进化）</span><br></pre></td></tr></table></figure>
<ol start="4">
<li><strong>进化支模型</strong></li>
</ol>
<p>与枝位点模型类型，能同时检测多个进化枝（Clade），但是该模型并没有将背景支的dN/dS值约束在（0,1）。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">M2a_model: model = 0 NSsites = 22</span><br><span class="line">Clade Model C: model = 3 NSsites = 2</span><br><span class="line">Clade Model D: model = 3 NSsites = 3</span><br><span class="line">Clade Model C vs. M2a_model</span><br></pre></td></tr></table></figure>
<h3 id="怎么区分-unrooted-tree-与-rooted-tree"><a class="markdownIt-Anchor" href="#怎么区分-unrooted-tree-与-rooted-tree"></a> 怎么区分 unrooted tree 与 rooted tree</h3>
<p>有时通过构建进化树的软件获得进化树后，软件不会对进化树是有根树，还是无根树做说明。在不做额外编写脚本的基础上，其实可以直接从newick文件的内容格式上区分有根树和无根树］</p>
<p>假设有下面三种形式的进化树：</p>
<p>[1] (a, b, c);      无根树</p>
<p>[2] (a, b, c):1;   有根树</p>
<p>[3] ((a,b),c);      有根树</p>
<p>除第一种形式的进化树外，其它两种形式的进化树都是有根树。R 语言对有根树、无根树的判断是：</p>
<p>如果最外层大括号内只有两个分枝，即为有根树，如 [3]；</p>
<p>如果最外层大括号内有三个或以上分枝，一般为无根树，如 [1]；</p>
<p>但是，如果大括号外存在枝长参数，如 [2] 中的 :1，这种情况认为三个分枝以一定的枝长连接到根上，为有根树。</p>
<p>我想通过上述三个例子，可以先看最外层是否有枝长参数，再数一下最外层括号下有几个分枝，即可简单地、直接地判断目前绝大多数获得的newick格式的进化树是有根树，还是无根树。</p>
<h3 id="实践"><a class="markdownIt-Anchor" href="#实践"></a> 实践</h3>
<p>如何收集基因？如果我想找出一个物种与相邻的若干物种相比较，有哪些基因出现了正向选择。首先，用 OrthoFinder 软件可以找出 single copy genes。这里要注意，初始的物种数目越多，single copy genes 越少，后续用于分析得到的结果越少，我看文献中有用到 7 个物种，摘录于此：“To detect genes evolving under positive selection in jujube, we used gene clustering of the seven Rosales species to identify 1,420 single-copy orthologous genes and used the optimized branch-site mode<sup>l63</sup> to identify 254 genes that are positively selected in <em>Z. jujuba</em>.” 另外，我觉得可以用 jcvi 流程，他会基于基因组位置和基因序列两个信息来做 orthologous 分析，以感兴趣物种为参考，将其他若干物种的基因集分别与之比较，就可以得这一群物种的 orthologous genes（反刍动物的进化分析其实就是类似这个策略，只不过他们用的是lastal比对软件），然后要思考的一点是，是不是必须所有物种都出现的ortholog gene，才用于后续分析。</p>
<p>2021.10.08</p>
<ol>
<li>选择7个物种，准备好CDS序列，然后翻译成pep序列，将用于orthofinder流程。注意，同时考虑jcvi流程需要的gff文件和pep文件。（别用花生。感觉序列质量不好。）</li>
<li>运行orthofinder流程，预计找到1000多个single copy genes</li>
<li>在AM与其他物种依次运行jcvi，然后整合结果，预计这种方式可以找到几千个orthologous genes</li>
<li>分别运行codeml流程，计算Ka/Ks值。并计算P值。看最终有多少个基因会出现正向选择。</li>
</ol>
<h3 id="参考资料"><a class="markdownIt-Anchor" href="#参考资料"></a> 参考资料</h3>
<p><a target="_blank" rel="noopener" href="http://yuzhenpeng.github.io/2019/12/24/paml/">于振朋: Comparative Genomics and Adaptive Evolution</a><br />
<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1Qx411m7Bn?from=search&amp;seid=9774626637377853773&amp;spm_id_from=333.337.0.0">B 站 PAML 视频教程：利用 codeml 计算 dN/dS</a><br />
<a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3001">Chenlianfu: 使用 PAML 软件利用核酸序列进行系统发育参数分析</a><br />
<a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3036">Chenlianfu: 使用 PAML 软件利用密码子序列进行正选择分析</a><br />
<a target="_blank" rel="noopener" href="https://figshare.com/articles/journal_contribution/worked_example_pdf/3839583">A worked example of estimating ω and testing for adaptive evolution in six parasite species</a><br />
<a target="_blank" rel="noopener" href="http://wap.sciencenet.cn/blog-460481-1163040.html?mobile=1">高芳銮：选择压力分析之 EasyCodeML 完整篇（By Raindy)</a><br />
<a target="_blank" rel="noopener" href="https://static-content.springer.com/esm/art%3A10.1007%2Fs00606-019-01578-2/MediaObjects/606_2019_1578_MOESM1_ESM.pdf">Supplementary File 1 of “<u><em>Analyses of mitochondrial genomes of the genus Ammopiptanthus provide new insights into the evolution of legume plants</em></u>”</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/" class="post-title-link" itemprop="url">Proteomics | Top-down proteomics analysis of histones</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-09-20 13:31:13 / Modified: 16:28:51" itemprop="dateCreated datePublished" datetime="2021-09-20T13:31:13+08:00">2021-09-20</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Proteomics/" itemprop="url" rel="index"><span itemprop="name">Proteomics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h3 id="1-workflow"><a class="markdownIt-Anchor" href="#1-workflow"></a> 1 workflow</h3>
<ol>
<li>crude histone isolation from cells or tissues;</li>
<li>offline HPLC to separete core histones  (H2, H3.1, H3.2, H4)</li>
<li>LC-MS (raw data)</li>
<li>MS data analysis</li>
</ol>
<p>Note: 图见 RGC proposal。</p>
<p><img src="https://ars.els-cdn.com/content/image/1-s2.0-S1046202319301732-gr1.jpg" alt="图例" /></p>
<h3 id="2-offline-hplc"><a class="markdownIt-Anchor" href="#2-offline-hplc"></a> 2 Offline HPLC</h3>
<p>Note：图见 RGC proposal</p>
<p><img src="https://www.researchgate.net/profile/Yifan-Liu-41/publication/235775342/figure/fig2/AS:931921148276736@1599198724085/HPLC-and-SDS-PAGE-separation-of-core-histones_W640.jpg" alt="图例" /></p>
<h3 id="3-ms1-chromatograph"><a class="markdownIt-Anchor" href="#3-ms1-chromatograph"></a> 3 MS1 chromatograph</h3>
<p>首先要安装 Thermo Fisher 的 Xcalibur 软件。安装方法，<a target="_blank" rel="noopener" href="https://thermo.flexnetoperations.com/control/thmo/login">https://thermo.flexnetoperations.com/control/thmo/login</a> ，使用教育邮箱注册，xx@link.cuhk.edu.hk，注册并登入，即可看到一系列软件的列表，我们选择 Xcalibur <a target="_blank" rel="noopener" href="https://thermo.flexnetoperations.com/control/thmo/product?child_plneID=820487&amp;ver=CURRENT">https://thermo.flexnetoperations.com/control/thmo/product?child_plneID=820487&amp;ver=CURRENT</a>，可执行文件 exe 格式，将文件下载到windows电脑，再安装，安装的软件里面就包含了 freestype。</p>
<p>将 raw data 导入 freestyle 软件，即可浏览 MS1 的图。如何导出高清图片？</p>
<p><img src="" alt="图例" /></p>
<h3 id="raw-data-to-pbf-和-ms2ft-格式文件"><a class="markdownIt-Anchor" href="#raw-data-to-pbf-和-ms2ft-格式文件"></a> raw data to pbf 和 ms2ft 格式文件。</h3>
<p>使用 ProMex 软件，可以将 raw 格式的质谱文件转换成 pbf 和 ms2ft 格式文件。</p>
<p>已经有公开的文献报道 top down proteomics 数据的分析方法 <a target="_blank" rel="noopener" href="https://www.nature.com/articles/nmeth.4388">NATURE METHODS: Informed-Proteomics: open-source software package for top-down proteomics</a>，可以去 <a target="_blank" rel="noopener" href="https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics">github</a> 下载 Informed Proteomics，软件处于持续更新中，本笔记撰写的 2021 年仍有更新。</p>
<p>这些格式有什么不同呢？</p>
<p>raw:<br />
pbf: Creates a .pbf file This file contains spectra information and full chromatograms for MS1 and MSn data, allowing fast access to extracted ion chromatograms during the search.<br />
ms2ft:</p>
<p>“插播一句，这些格式转换的软件是 windows 平台的，如果要再 linux 下面跨平台运行，则需要用到 mono 软件，具体可以参考 <a target="_blank" rel="noopener" href="https://www.cnblogs.com/jessepeng/p/12452276.html">https://www.cnblogs.com/jessepeng/p/12452276.html</a>。mono 的安装并不容易，可能有些普通用户会碰到问题，主要是依赖libgdiplus库，只要这个解决了，源码安装三部曲就行了，这里不加以赘述。”</p>
<p>转换代码如下（参考：<a target="_blank" rel="noopener" href="https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics">https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics</a>）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">PbfGen.exe -s MyDataset.raw</span><br><span class="line">ProMex.exe -i MyDataset.pbf -minCharge 2 -maxCharge 60 -minMass 3000 -maxMass 50000 -score n -csv n -maxThreads 0</span><br></pre></td></tr></table></figure>
<h3 id="sequence-alignment-and-idntification-of-ptms"><a class="markdownIt-Anchor" href="#sequence-alignment-and-idntification-of-ptms"></a> sequence alignment and idntification of PTMs</h3>
<p>得到pbf 和 ms2ft 文件之后，就可以检索蛋白序列和可能的修饰情况。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">MSPathFinderT.exe -s MyDataset.pbf -feature MyDataset.ms1ft -d C:\FASTA\ProteinList.fasta -o C:\WorkDir -t 10 -f 10 -m 1 -tda 1 -minLength 21 -maxLength 300 -minCharge 2 -maxCharge 30 -minFragCharge 1 -maxFragCharge 15 -minMass 3000 -maxMass 50000 -mod MSPathFinder_Mods.txt</span><br></pre></td></tr></table></figure>
<p>上面的 <code>Mods.txt</code> 文件即是可能的修饰情况，以大豆组蛋白为例，文件内容如下。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"># This file is used to specify modifications for MSPathFinder</span><br><span class="line"></span><br><span class="line"># Max Number of Modifications per peptide</span><br><span class="line">NumMods=9</span><br><span class="line"></span><br><span class="line"># Static mods</span><br><span class="line"># None</span><br><span class="line"></span><br><span class="line"># Dynamic mods</span><br><span class="line"># Functional group, Amino acid, opt?, Position, Modification type</span><br><span class="line">O1,MC,opt,any,Oxidation</span><br><span class="line">HO3P,ST,opt,any,Phospho</span><br><span class="line">C2H2O1,K,opt,any,Acetyl</span><br><span class="line">CH2,KR,opt,any,Methyl</span><br><span class="line">C2H4,KR,opt,any,Dimethyl</span><br><span class="line">C3H6,K,opt,any,Trimethyl</span><br><span class="line">C2H2O,*,opt,Prot-N-term,Acetyl     # Acetylation Protein N-term (C2H2O can be replaced with &quot;H(2) C(2) O&quot;)  </span><br></pre></td></tr></table></figure>
<p>运行结果，将得到一个 tsv （表格）文本文件，列出了可能的蛋白序列以及修饰情况，另有一个图片文件 <code>xxx.ms1ft.png</code>，意思？。</p>
<h3 id="图片"><a class="markdownIt-Anchor" href="#图片"></a> 图片</h3>
<p>上面的分析中，mods 文件中设定了一条蛋白序列上修饰的数目为 9 个，怎么确定这个数字是合理的？我设定为 10 个，跑了一天也没有结果，所以少于这10 会比较合适。另外需要画一个统计图，即从1到9个修饰的密度图，如果类似正态分布，就可以说明我们的设定是合理的。</p>
<p><img src="https://www.researchgate.net/publication/329379093/figure/fig4/AS:699921351049216@1543885662248/Degree-distribution-All-the-networks-show-a-Poisson-distribution-as-corresponds-to_W640.jpg" alt="图例" /></p>
<h3 id="图片-2"><a class="markdownIt-Anchor" href="#图片-2"></a> 图片</h3>
<p>各个位点的修饰情况统计。</p>
<p>（D:\博后课题文件\PHD5 project\200831 top-down proteomics in soybean\R代码-冯磊\ptm_freq_diagram\version 2 custom color）</p>
<p><img src="https://ars.els-cdn.com/content/image/1-s2.0-S1046202319301732-gr5.jpg" alt="图例" /></p>
<h3 id="interplay-score"><a class="markdownIt-Anchor" href="#interplay-score"></a> Interplay Score</h3>
<p>何谓 interplay score：<a target="_blank" rel="noopener" href="https://www.sciencedirect.com/science/article/pii/S1046202319301732">https://www.sciencedirect.com/science/article/pii/S1046202319301732</a><br />
何为 co-existence？<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5240223/bin/nihms840427f5.jpg">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5240223/bin/nihms840427f5.jpg</a></p>
<p>两个修饰是否存在coexisence？</p>
<p><img src="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5240223/bin/nihms840427f5.jpg" alt="图例" /></p>
<h3 id="样品之间的比较定量的问题"><a class="markdownIt-Anchor" href="#样品之间的比较定量的问题"></a> 样品之间的比较，定量的问题？</h3>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/" class="post-title-link" itemprop="url">Bioinfo | TCseq Time Course sequencing data analysis</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-09-16 11:01:58 / Modified: 13:53:39" itemprop="dateCreated datePublished" datetime="2021-09-16T11:01:58+08:00">2021-09-16</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>转录组测序如果有多个时间点，可以做表达趋势分析，即基因表达模式聚类分析，可以用 STEM、Mfuzz、TCseq 软件等工具。</p>
<p>Mfuzz 采用了一种新的聚类算法 fuzzy c-means algorithm，在文献中称这种聚类算法为 soft clustering 算法，相比 K-means 等 hard clustering 算法，一定程度上降低了噪声对聚类结果的干扰，而且这种算法有效的定义了基因和 cluster 之间的关系，即基因是否属于某个 cluster, 对应的值为 memebership。</p>
<p>对于分析而言，我们只需要提供基因表达量的数据就可以了，需要注意的是，Mfuzz 默认你提供的数据是归一化之后的表达量，这意味着表达量必须可以直接在样本间进行比较，对于 FPKM, TPM 这两种定量方式而言，是可以直接在样本间进行比较的，但是对于 count 的定量结果，我们必须先进行归一化，可以使用 edgeR 或者 DESeq 先得到归一化之后的数据再进行后续分析。Mfuzz 的使用可参考：<a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1625288">生信修炼手册：使用 Mfuzz 进行时间序列表达模式聚类分析</a></p>
<p>TCseq 则可以用 count 定量来做输入，他自己会调用 edgeR 来做样品之间的归一化。分析之前准备三个文件：（1）基因表达矩阵文件，可以直接拿原始 read count 数据；（2）样品分组情况的文件；（3）基因坐标文件。</p>
<p>下面列出文件内容：</p>
<p>（1）基因表达矩阵文件，可以直接拿原始 read count 数据。我是用 htseq-count 工具来做的每个基因的 read count 统计，这里按下不表。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">gene-id	L1-1A	L1-2A	L1-3A	L2-1A	L2-2A	L2-3A	L3-1A	L3-2A	L3-3A	L4-1A	L4-2A	L4-3A</span><br><span class="line">Chr01.g00016	72	74	30	16	95	134	100	10	57	95	78	5</span><br><span class="line">Chr01.g00017	1645	2072	2672	791	1105	1726	3060	947	882	1576	2414	466</span><br><span class="line">Chr01.g00018	8	3	59	6	16	16	19	6	0	12	3	1</span><br><span class="line">Chr01.g00019	1	14	26	2	5	7	15	0	0	8	6	0</span><br><span class="line">Chr01.g00020	0	0	0	0	0	0	0	0	0	0	0	0</span><br><span class="line">Chr01.g00021	0	0	0	0	0	0	0	0	0	0	0	0</span><br><span class="line">Chr01.g00022	1744	1912	2568	1549	2708	1966	2198	2308	1995	1789	2255	1697</span><br><span class="line">Chr01.g00023	8	1	3	8	14	3	6	3	6	0	7	1</span><br><span class="line">Chr01.g00024	39	25	69	127	103	8	43	170	167	89	89	67</span><br><span class="line">Chr01.g00025	0	0	0	0	0	0	0	0	0	0	0	0</span><br><span class="line">Chr01.g00026	0	1	0	6	2	1	17	30	8	2	5	94</span><br><span class="line">Chr01.g00027	748	794	808	370	640	805	886	1169	786	677	864	817</span><br><span class="line">... ... (其余省略)</span><br></pre></td></tr></table></figure>
<p>（2）样品分组文件 (样品，时间点，分组三列)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">sampleid	timepoint	group</span><br><span class="line">L1-1A	0h	1</span><br><span class="line">L1-2A	0h	1</span><br><span class="line">L1-3A	0h	1</span><br><span class="line">L2-1A	2h	2</span><br><span class="line">L2-2A	2h	2</span><br><span class="line">L2-3A	2h	2</span><br><span class="line">L3-2A	24h	3</span><br><span class="line">L3-3A	24h	3</span><br><span class="line">L4-1A	48h	4</span><br><span class="line">L4-2A	48h	4</span><br><span class="line">L4-3A	48h	4</span><br></pre></td></tr></table></figure>
<p>（3）基因坐标文件 (染色体，起始及终止位点，基因id四列)</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">chr	start	end	id</span><br><span class="line">scaffold_1	760	5575	scaffold_1.g00001</span><br><span class="line">scaffold_1	81403	85562	scaffold_1.g00002</span><br><span class="line">scaffold_1	122872	124937	scaffold_1.g00003</span><br><span class="line">scaffold_1	138239	140514	scaffold_1.g00004</span><br><span class="line">scaffold_1	156823	159196	scaffold_1.g00005</span><br><span class="line">scaffold_1	160925	162845	scaffold_1.g00006</span><br><span class="line">scaffold_1	195980	199165	scaffold_1.g00007</span><br><span class="line">scaffold_1	199535	200289	scaffold_1.g00008</span><br></pre></td></tr></table></figure>
<h3 id="读取输入基因表达量与分组数据"><a class="markdownIt-Anchor" href="#读取输入基因表达量与分组数据"></a> 读取输入基因表达量与分组数据</h3>
<p>准备好上述三个文件，就可以在 Rstudio 界面开始加载 TCseq 包，读取上述三个数据文件，开始分析。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line">setwd(&quot;C:/R-data&quot;)</span><br><span class="line"></span><br><span class="line">#To install this package, start R (version &quot;4.1&quot;) and enter:</span><br><span class="line">  </span><br><span class="line"> if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))</span><br><span class="line">    install.packages(&quot;BiocManager&quot;)</span><br><span class="line"></span><br><span class="line">BiocManager::install(&quot;TCseq&quot;)</span><br><span class="line">library(TCseq)</span><br><span class="line"></span><br><span class="line">BiocManager::install(&quot;topGO&quot;)   ## 画GO树状图需要用</span><br><span class="line">library(topGO)</span><br><span class="line"></span><br><span class="line">library(dplyr)</span><br><span class="line"></span><br><span class="line"># To view documentation for the version of this package installed in your system, start R and enter:</span><br><span class="line">browseVignettes(&quot;TCseq&quot;)</span><br><span class="line"></span><br><span class="line">rm(list = ls())  # 清除 Rstudio 里面旧的变量</span><br><span class="line">ls()</span><br><span class="line"># 位置信息文件</span><br><span class="line">genomicIntervals &lt;- read.delim(&#x27;AMO.intervial.txt&#x27;, stringsAsFactors = FALSE)</span><br><span class="line">genomicIntervals$chr &lt;- as.factor(genomicIntervals$chr)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">## 表达量矩阵文件</span><br><span class="line">#count &lt;- read.delim(&#x27;all_read_count.txt&#x27;,header = T,row.names = 1) ### read count table</span><br><span class="line"></span><br><span class="line">count &lt;- read.delim(&#x27;all_read_count.txt&#x27;,header = T,row.names = 1) ### read count table</span><br><span class="line">count &lt;- count[,-7]   ### L3-1A 样品数据有问题，剔除  # Optional</span><br><span class="line"></span><br><span class="line"># as.integer 转换为整数型</span><br><span class="line">countsTable &lt;- count %&gt;%  mutate(across(where(is.numeric), as.integer))   # 依赖 dplyr 包</span><br><span class="line"># or</span><br><span class="line">#count1 &lt;- apply(count, 2, as.integer)</span><br><span class="line">rownames(countsTable) &lt;- rownames(count)</span><br><span class="line">##矩阵形式</span><br><span class="line">countsTable &lt;- as.matrix(countsTable)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># 分组文件</span><br><span class="line">experiment &lt;- read.delim(&#x27;leaf_group.txt&#x27;,header = T,stringsAsFactors = FALSE)</span><br><span class="line">experiment$sampleid &lt;- as.factor(experiment$sampleid)</span><br></pre></td></tr></table></figure>
<h3 id="创建tca对象"><a class="markdownIt-Anchor" href="#创建tca对象"></a> 创建TCA对象</h3>
<p>TCseq 使用 S4 类 TCA 存储所有输入数据以进行后续分析。根据以上三个文件创造即可。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">tca &lt;- TCA(design = experiment, genomicFeature = genomicIntervals, counts = countsTable)</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line">tca  # 此次摘抄自 凯凯何_Boy 链接：https://www.jianshu.com/p/26332e4da742</span><br><span class="line">  ##############</span><br><span class="line">  An object of class &quot;TCA&quot;</span><br><span class="line">@design</span><br><span class="line">  sampleid timepoint group</span><br><span class="line">1       s1        0h     1</span><br><span class="line">2       s2        0h     1</span><br><span class="line">3       s3        0h     1</span><br><span class="line">4       s4       24h     2</span><br><span class="line">5       s5       24h     2</span><br><span class="line">6       s6       24h     2</span><br><span class="line">7       s7       48h     3</span><br><span class="line">8       s8       48h     3</span><br><span class="line">9       s9       48h     3</span><br><span class="line"></span><br><span class="line">@counts</span><br><span class="line">                    s1  s2  s3  s4  s5  s6  s7  s8  s9</span><br><span class="line">ENSBTAG00000000005  60  58  70  64  60  74  53  55  50</span><br><span class="line">ENSBTAG00000000008   0   0   0   1   0   1   2   0   0</span><br><span class="line">ENSBTAG00000000009   0   1   3   0   1   0  11   9   9</span><br><span class="line">ENSBTAG00000000010 341 300 341 170 216 214 150 236 199</span><br><span class="line">ENSBTAG00000000011   4   2   3   5   3   1   6   2   3</span><br><span class="line">29749 more rows ...</span><br><span class="line"></span><br><span class="line">@genomicFeature</span><br><span class="line">   chr     start       end                 id</span><br><span class="line">1 chr1 100098899 100161382 ENSBTAG00000035710</span><br><span class="line">2 chr1 100754351 100754458 ENSBTAG00000043293</span><br><span class="line">3 chr1 101522743 101601659 ENSBTAG00000011139</span><br><span class="line">4 chr1 101619573 101646742 ENSBTAG00000053582</span><br><span class="line">5 chr1 101776678 101776956          MSTRG.445</span><br><span class="line">29749 more rows ...</span><br><span class="line"></span><br><span class="line">@clusterRes</span><br><span class="line">An object of class &quot;clust&quot;</span><br><span class="line">另一种创建的方式,是通过SummarizedExperiment对象：</span><br><span class="line"></span><br><span class="line">suppressWarnings(library(SummarizedExperiment))</span><br><span class="line">se &lt;- SummarizedExperiment(assays=list(counts = countsTable), colData = experiment)</span><br><span class="line">tca &lt;- TCAFromSummarizedExperiment(se = se, genomicFeature = genomicIntervals)</span><br></pre></td></tr></table></figure>
<p>现在已经基于这三个文件得到了TCA对象</p>
<h3 id="差异分析"><a class="markdownIt-Anchor" href="#差异分析"></a> 差异分析</h3>
<p>该包内置了我们常用的差异分析 R 包 edgeR，通过使用广义线性模型 (GLM) 方法去鉴定差异基因。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">tca &lt;- DBanalysis(tca）</span><br><span class="line">## 我们也可以增添一些参数进行过滤</span><br><span class="line">tca &lt;- DBanalysis(tca, filter.type = &quot;raw&quot;, filter.value = 10, samplePassfilter = 2)</span><br><span class="line">#比如上述步骤保留具有两个或多个已读取样本的基因表达量超过10的基因组区域</span><br><span class="line"></span><br><span class="line">tca &lt;- DBanalysis(tca, filter.type = &quot;raw&quot;, filter.value = 10, samplePassfilter = 2)  # 与上面的选一个即可，这个命令是加了筛选参数的。</span><br><span class="line">#比如上述步骤保留具有两个或多个已读取样本的基因表达量超过10的基因组区域</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="提取差异分析结果"><a class="markdownIt-Anchor" href="#提取差异分析结果"></a> 提取差异分析结果</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">#####################</span><br><span class="line">####提取差异分析结果</span><br><span class="line">#####################</span><br><span class="line"># 通过DBresult函数可以直接提取我们差异分析的结果，并通过$去查看各个组的内容。</span><br><span class="line">DBres &lt;- DBresult(tca, group1 = &quot;0h&quot;, group2 = c(&quot;2h&quot;,&quot;24h&quot;,&quot;48h&quot;))</span><br><span class="line">str(DBres, strict.width =  &quot;cut&quot;)</span><br><span class="line">head(DBres$`24hvs0h`)</span><br><span class="line"># 若只想提取满足显著差异的结果(一般abs(log2-fold &gt; 2)且adjust p-value&lt;0.05)，只需要加上top.sig = TRUE函数即可。</span><br><span class="line">## 显著差异</span><br><span class="line">DBres.sig &lt;- DBresult(tca, group1 = &quot;0h&quot;, group2 = c(&quot;2h&quot;,&quot;24h&quot;,&quot;48h&quot;), top.sig = TRUE)</span><br><span class="line">str(DBres.sig, strict.width =  &quot;cut&quot;)</span><br><span class="line">#############</span><br></pre></td></tr></table></figure>
<p>下面是摘抄他人的案例：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">通过DBresult函数可以直接提取我们差异分析的结果，在通过$去查看各个组的内容。</span><br><span class="line"></span><br><span class="line">DBres &lt;- DBresult(tca, group1 = &quot;0h&quot;, group2 = c(&quot;24h&quot;,&quot;48h&quot;))</span><br><span class="line">str(DBres, strict.width =  &quot;cut&quot;)</span><br><span class="line">head(DBres$`24hvs0h`)</span><br><span class="line">############</span><br><span class="line">GRanges object with 6 ranges and 4 metadata columns:</span><br><span class="line">                     seqnames              ranges strand |     logFC      PValue         paj</span><br><span class="line">                        &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span><br><span class="line">  ENSBTAG00000000005     chr1 100098899-100161382      * |  0.274010 1.35162e-01 2.81082e-01</span><br><span class="line">  ENSBTAG00000000009     chr1 101522743-101601659      * | -1.501457 2.26674e-01 4.03419e-01</span><br><span class="line">  ENSBTAG00000000010     chr1 101776678-101776956      * | -0.513593 3.36326e-06 3.80124e-05</span><br><span class="line">  ENSBTAG00000000011     chr1   10231295-10543983      * |  0.198446 7.93039e-01 8.90540e-01</span><br><span class="line">  ENSBTAG00000000012     chr1 102768932-102769966      * | -0.300941 5.04886e-03 2.18330e-02</span><br><span class="line">  ENSBTAG00000000013     chr1 105069395-105069910      * |  0.140489 1.63446e-01 3.21469e-01</span><br><span class="line">                                     id</span><br><span class="line">                            &lt;character&gt;</span><br><span class="line">  ENSBTAG00000000005 ENSBTAG00000035710</span><br><span class="line">  ENSBTAG00000000009 ENSBTAG00000011139</span><br><span class="line">  ENSBTAG00000000010          MSTRG.445</span><br><span class="line">  ENSBTAG00000000011 ENSBTAG00000017753</span><br><span class="line">  ENSBTAG00000000012 ENSBTAG00000048551</span><br><span class="line">  ENSBTAG00000000013          MSTRG.449</span><br><span class="line">  -------</span><br><span class="line">  seqinfo: 206 sequences from an unspecified genome; no seqlengths</span><br><span class="line"></span><br><span class="line"># 作者：凯凯何_Boy https://www.jianshu.com/p/26332e4da742</span><br></pre></td></tr></table></figure>
<h3 id="时间趋势分析"><a class="markdownIt-Anchor" href="#时间趋势分析"></a> 时间趋势分析</h3>
<p>接下来就到了趋势分析的换件了，为了检测数据的时序模式，TCseq包使用了非监督聚类的方法,首先通过聚类创建一个时程表，行为基因组区域信息，列为时间点，表中的数值可以选择标准化后的表达量值或者是基于初始时间点所有样品点的LogFC值，代码中通过value参数来调整，另外若设置filter = TRUE时候，将自动过滤在任何两个时间段都没有显著差异的基因组区域，结果通过tcTable函数进行查看</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">##通过标准化后的表达量来做聚类</span><br><span class="line"># tca &lt;- timecourseTable(tca, value = &quot;expression&quot;, norm.method = &quot;rpkm&quot;, filter = TRUE)  # Aug 9 用的这个</span><br><span class="line">tca &lt;- timecourseTable(tca, value = &quot;expression&quot;, lib.norm = TRUE, norm.method = &quot;rpkm&quot;, filter = TRUE, pvalue.threshold =  0.1, abs.fold = 1.5)</span><br><span class="line">#tca &lt;- timecourseTable(tca, value = &quot;expression&quot;, norm.method = &quot;rpkm&quot;, filter = FALSE)</span><br><span class="line"></span><br><span class="line">## 通过LogFC值来做聚类</span><br><span class="line">#tca &lt;- timecourseTable(tca, value = &quot;FC&quot;, norm.method = &quot;rpkm&quot;, filter = TRUE)</span><br><span class="line"></span><br><span class="line">## 查看生成结果</span><br><span class="line">t &lt;- tcTable(tca)</span><br><span class="line">head(t)</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">## 通过LogFC值</span><br><span class="line">tca &lt;- timecourseTable(tca, value = &quot;FC&quot;, norm.method = &quot;rpkm&quot;, filter = TRUE)</span><br><span class="line"></span><br><span class="line">##通过标准化后的表达量</span><br><span class="line">tca &lt;- timecourseTable(tca, value = &quot;expression&quot;, norm.method = &quot;rpkm&quot;, filter = TRUE)</span><br><span class="line"></span><br><span class="line">## 查看生成结果</span><br><span class="line">t &lt;- tcTable(tca)</span><br><span class="line">head(t)</span><br><span class="line">                             0h          24h          48h</span><br><span class="line">ENSBTAG00000000425 0.0000000000  0.030165698 0.0075187827</span><br><span class="line">ENSBTAG00000000706 2.6013480888 11.521986790 1.7768887654</span><br><span class="line">ENSBTAG00000000936 0.0007283438  0.003506178 0.0003475789</span><br><span class="line">ENSBTAG00000001325 0.0002585125  0.001290036 0.0001705818</span><br><span class="line">ENSBTAG00000001687 0.3065966903  2.485377556 2.6233245852</span><br><span class="line">ENSBTAG00000001745 0.2496824763  2.120932031 4.7101417152</span><br></pre></td></tr></table></figure>
<h3 id="聚类分析"><a class="markdownIt-Anchor" href="#聚类分析"></a> 聚类分析</h3>
<p>两种聚类算法可供选择: hard clustering（包含hierachical,pam,kmeans） 及 soft clustering (fuzzy cmeans)，算法的选择大家自己参考文献选择，这里就拿文档中cmeans算法进行分析聚类。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">tca &lt;- timeclust(tca, algo = &quot;cm&quot;, k = 6, standardize = TRUE)</span><br><span class="line"></span><br><span class="line">###################################################</span><br><span class="line">p &lt;- timeclustplot(tca, value = &quot;z-score(RPKM)&quot;, cols = 3)</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/PVjl6Hy.jpeg" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line"># 若是想查看某个聚类有哪些基因参与，使用tca@clusterRes查看聚类结果即可</span><br><span class="line"></span><br><span class="line">a &lt;- as.data.frame(tca@clusterRes@cluster)</span><br><span class="line">names(a) &lt;- &#x27;Cluster&#x27;</span><br><span class="line">## 查看各个聚类中的基因数目</span><br><span class="line">table(a)</span><br><span class="line">dim(a)</span><br><span class="line">table(a)</span><br><span class="line"></span><br><span class="line">#&gt; table(a)</span><br><span class="line">#### 过滤标准： tca &lt;- DBanalysis(tca, filter.type = &quot;raw&quot;, filter.value = 10, samplePassfilter = 2)</span><br><span class="line"># 1    2    3    4    5    6 </span><br><span class="line"># 1751  380 1122  599  905  554   # Aug 11 基于read count。过滤了一部分数据：</span><br><span class="line"></span><br><span class="line">#--- 过往记录：</span><br><span class="line"># 1    2    3    4    5    6 </span><br><span class="line"># 418 1397  993  602 1033 1042    ### Aug 10 </span><br><span class="line"># table(a)</span><br><span class="line"># filter.value = 1, samplePassfilter = 2) # Aug 9 用的这个 筛选条件之后，基因数目少了很多</span><br><span class="line">#1   2   3   4   5   6 </span><br><span class="line">#736 222 569 810 812 452 </span><br><span class="line"># 干旱处理 叶片的基因分组情况。没有设置筛选的情况下，基因数目多很多，但是手动查询发现有的基因并不明显</span><br><span class="line">#1    2    3    4    5    6 </span><br><span class="line">#3431 2156 3690 5167 2368 3566 </span><br><span class="line">##1   2   3   4   5   6 </span><br><span class="line">##511 147 165 292 180 255 </span><br><span class="line">##筛选 。可以分别导出每个cluster的基因列表 </span><br><span class="line"></span><br><span class="line">Cluster1 &lt;- subset(a,Cluster == 1)</span><br><span class="line">Cluster2 &lt;- subset(a,Cluster == 2)</span><br><span class="line">Cluster3 &lt;- subset(a,Cluster == 3)</span><br><span class="line">Cluster4 &lt;- subset(a,Cluster == 4)</span><br><span class="line">Cluster5 &lt;- subset(a,Cluster == 5)</span><br><span class="line">Cluster6 &lt;- subset(a,Cluster == 6)</span><br><span class="line">#Cluster7 &lt;- subset(a,Cluster == 7)</span><br><span class="line"></span><br><span class="line">write.csv(as.data.frame( tca@clusterRes@cluster ), file=&quot;6_clusters.csv&quot;)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">#上面分6次执行提取cluster命令比较繁琐，另一个方法是批量导出每个聚类所包含的基因，上面这种重复写6次的命令太繁复，尝试用for循环来解决</span><br><span class="line">dir.create(path=&quot;tcseq_cluster&quot;,recursive = TRUE)  # 在工作目录下新建一个文件夹</span><br><span class="line">for(i in 1:6)&#123;</span><br><span class="line">  assign(paste(&quot;Cluster&quot;,i,sep=&quot;&quot;),subset(a,Cluster == i)) # Cluster2 &lt;- subset(a,Cluster == 2)</span><br><span class="line">&#125;</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="对个别-cluster-进行功能富集分析"><a class="markdownIt-Anchor" href="#对个别-cluster-进行功能富集分析"></a> 对个别 Cluster 进行功能富集分析</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">library(clusterProfiler)</span><br><span class="line">library(org.XXXX.eg.db)</span><br><span class="line">library(stringr)</span><br><span class="line">library(topGO)</span><br><span class="line">org &lt;- org.XXXX.eg.db</span><br><span class="line"></span><br><span class="line">geneList=rownames(Cluster1)  # Cluster 1</span><br><span class="line">bp1 &lt;-enrichGO(gene = geneList, OrgDb = org, keyType = &quot;GID&quot;, ont = &quot;BP&quot;)</span><br><span class="line">pdf(&quot;TCseq_Cluster1_GO_BP.pdf&quot;)</span><br><span class="line">barplot(bp1, drop=TRUE, showCategory=30)</span><br><span class="line">dev.off()</span><br><span class="line">write.csv(as.data.frame( bp1@result ), file=&quot;TCseq_Cluster1_GO_BP.csv&quot;)</span><br><span class="line"></span><br><span class="line">geneList=rownames(Cluster2)  # Cluster 2</span><br><span class="line">bp2 &lt;-enrichGO(gene = geneList, OrgDb = org, keyType = &quot;GID&quot;, ont = &quot;BP&quot;)</span><br><span class="line">pdf(&quot;TCseq_Cluster2_GO_BP.pdf&quot;)</span><br><span class="line">barplot(bp2, drop=TRUE, showCategory=30)</span><br><span class="line">dev.off()</span><br><span class="line">write.csv(as.data.frame( bp2@result ), file=&quot;TCseq_Cluster2_GO_BP.csv&quot;)</span><br><span class="line">#----- 上面采用默认参数，得到的GO条目太少，于是放松筛选标准，画的图更好看</span><br><span class="line">bp2 &lt;-enrichGO(gene = geneList, OrgDb = org, keyType = &quot;GID&quot;, ont = &quot;BP&quot;, pvalueCutoff = 0.5, qvalueCutoff = 0.9)</span><br><span class="line">pdf(&quot;TCseq_Cluster2_GO_BP_low_standard.pdf&quot;)</span><br><span class="line">barplot(bp2, drop=TRUE, showCategory=30)</span><br><span class="line">dev.off()</span><br><span class="line">#------</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/CUmLAMP.jpeg" alt="" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/07/17/Bioinfo-LTR-identification-and-analysis/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/" class="post-title-link" itemprop="url">Bioinfo | LTR identification and analysis</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-07-17 02:22:18" itemprop="dateCreated datePublished" datetime="2021-07-17T02:22:18+08:00">2021-07-17</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-07-27 11:08:06" itemprop="dateModified" datetime="2021-07-27T11:08:06+08:00">2021-07-27</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>在植物基因组中，I 类转座因子，LTR-RT（LTR retrotransposons）是基因组扩张的主要原因。完整的LTR长度在 85 – 5000 bp 之间，下图表示的是一个完整的 LTR-RT，灰色框表示 TSD（target site duplications）, 红色三角形表示 LTR motif（长度在 2 bp 左右），蓝色框表示 LTR。LTR 中间序列长度在 1,000 – 15,000 之间波动。</p>
<p>完整的LTR-RT主要归为两大类: Gypsy 和 Copia。如果 LTR 中间的序列不包含开放阅读框（ORF）, 那么所属的 LTR-RT 就无法独立的转座。</p>
<p><img src="https://i.imgur.com/pDbuH66.jpeg" alt="" /></p>
<h3 id="安装"><a class="markdownIt-Anchor" href="#安装"></a> 安装</h3>
<p>LTR_retriever 不是一个独立的工具，他的主要作用就是整合 LTRharvest, LTR_FINDER, MGEScan 3.0.0, LTR_STRUC, 和 LtrDetector 的结果，过滤其中的假阳性 LTR-RT，得到高质量的LTR-RT库。</p>
<p>先下载 LTR_retriever：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/oushujun/LTR_retriever.git</span><br></pre></td></tr></table></figure>
<p>之后修改 LTR_retriever 下的 paths, 提供 BLAST+, RepeatMasker， HMMER， CDHIT 这些工具的路径。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">BLAST+=/your_path_to/BLAST+2.2.30/bin/</span><br><span class="line">RepeatMasker=/your_path_to/RepeatMasker4.0.0/</span><br><span class="line">HMMER=/your_path_to/HMMER3.1b2/bin/</span><br><span class="line">CDHIT=/your_path_to/CDHIT4.6.1/</span><br><span class="line">BLAST=/your_path_to/BLAST2.2.26/bin/ #not required if CDHIT provided</span><br></pre></td></tr></table></figure>
<p>下面是我实际操作中填写的案例。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"> 1 ##This file will provide LTR_retriever paths to dependent programs.</span><br><span class="line"> 2 ##You can leave the respective paths empty if programs are accessible through ENV (i.e. exported to .bashrc)</span><br><span class="line"> 3 ##If you specify a path, please make sure that the required program(s) is directly contained in that path but not in any child directories.</span><br><span class="line"> 4 ##e.g. BLAST+=/opt/software/BLAST+/2.2.30--GCC-4.4.5/bin/</span><br><span class="line"> 5 ##LTR_retriever is build based on GenomeTools/1.5.4, BLAST+/2.2.28, BLAST/2.2.26, CDHIT/4.6.1c, HMMER/3.1b2, RepeatMasker/4.0.0 and Tandem Repeats Finder 4.07b</span><br><span class="line"> 6 </span><br><span class="line"> 7 BLAST+=/home/fenglei/local/app/blast-2.11.0/bin     #a path that contains makeblastdb, blastn, blastx</span><br><span class="line"> 8 RepeatMasker=/home/fenglei/local/app/RepeatMasker   #a path that contains RepeatMasker</span><br><span class="line"> 9 HMMER=/home/fenglei/local/bin       #a path that contains hmmsearch</span><br><span class="line">10 CDHIT=/home/fenglei/local/app/cdhit     #a path that contains cd-hit-est (preferred). CDHIT and BLAST are replaceable</span><br><span class="line">11 BLAST=      #a path that contains blastclust (optional)</span><br></pre></td></tr></table></figure>
<p>更加方便的安装方法用 Bioconda 安装好 cd-hit repeatmasker， 然后下载 LTR_retriever:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">conda create -n LTR_retriever</span><br><span class="line">source activate LTR_retriever</span><br><span class="line">conda install -c conda-forge perl perl-text-soundex</span><br><span class="line">conda install -c bioconda cd-hit</span><br><span class="line">conda install -c bioconda/label/cf201901 repeatmasker</span><br><span class="line"></span><br><span class="line">git clone https://github.com/oushujun/LTR_retriever.git</span><br><span class="line">./LTR_retriever/LTR_retriever -h</span><br></pre></td></tr></table></figure>
<p>此外还需要额外安装 LTRharvest, LTR_FINDER 和 MGEScan_LTR。</p>
<p>LTRharverst: <a target="_blank" rel="noopener" href="http://genometools.org/">http://genometools.org/</a><br />
LTR_FINDER: <a target="_blank" rel="noopener" href="https://github.com/xzhub/LTR_Finder">https://github.com/xzhub/LTR_Finder</a><br />
LTR_FINDER_parallel: <a target="_blank" rel="noopener" href="https://github.com/oushujun/LTR_FINDER_parallel">https://github.com/oushujun/LTR_FINDER_parallel</a><br />
修改版MGEScan_LTR: <a target="_blank" rel="noopener" href="http://dawgpaws.sourceforge.net">http://dawgpaws.sourceforge.net</a></p>
<p>由于MGEScan_LTR装起来比我想象中麻烦，所以本文就仅使用LTRharverst和LTR_FINDER</p>
<h3 id="使用篇"><a class="markdownIt-Anchor" href="#使用篇"></a> 使用篇</h3>
<p>尽管 LTR_retriever 支持多个 LTR 工具的输入，但其实上 LTRharverst 和 LTR_FINDER 的结果就已经很不错了。</p>
<p>以拟南芥的基因组序列为例，分别使用 LTRharverst 和 LTR_FINDER 来寻找拟南芥中潜在 LTR 序列，之后用 LTR_retreiver 来合并结果。<br />
注意，也可以用 LTR-finder-parallel，应该会更快。</p>
<p>注意：LTRharvest 输出的结果中会将染色体名称成数字编号，可能会对后续分析造成问题。Unfortunately, LTRharvest has an annoying habit of not using the scaffold/chromosome names you give it (it outputs scaffolds as “seqX” where X is the 0-base number of your scaffold), so I remove all scaffolds in my fasta file, remove newline (“\n”) characters, and add a single scaffold name so that getting the fasta sequence for regions in the output gff3 is painless. 有人的解决方案是将自己的基因组手动拼成一条序列，因为只需要找LTR序列而已。 This is useful if you just want the distribution of insertion times of your LTRs, but if you need location information you’ll either have to rerun LTRharvest with the regular fasta, or use the length of each scaffold to figure out the real scaffold from your concatenated fasta.</p>
<p>20210719 备注：经过测试，LTRharvest输出的scn文件虽然有上面提到的问题，但是对后续导入LTR-retriever的整合分析不会造成影响。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"># LTRharvest</span><br><span class="line">gt suffixerator \</span><br><span class="line">  -db TAIR10.fa \</span><br><span class="line">  -indexname TAIR10 \</span><br><span class="line">  -tis -suf -lcp -des -ssp -sds -dna</span><br><span class="line">gt ltrharvest \</span><br><span class="line">  -index TAIR10 \</span><br><span class="line">  -similar 90 -vic 10 -seed 20 -seqids yes \</span><br><span class="line">  -minlenltr 100 -maxlenltr 7000 -mintsd 4 -maxtsd 6 \</span><br><span class="line">  -motif TGCA -motifmis 1  &gt; TAIR10.harvest.scn &amp;</span><br><span class="line"></span><br><span class="line"># LTR_FINDER</span><br><span class="line"># ltr_finder -D 15000 -d 1000 -L 7000 -l 100 -p 20 -C -M 0.9 TAIR10.fa &gt; TAIR10.finder.scn &amp;</span><br><span class="line">~/local/app/LTR_FINDER_parallel/LTR_FINDER_parallel -seq genome.fasta -threads 10 -harvest_out -size 1000000 -time 300</span><br></pre></td></tr></table></figure>
<p>LTR_retriever支持单个候选的LTR：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LTR_retriever -genome TAIR10.fa -inharvest TAIR10.harvest.scn</span><br></pre></td></tr></table></figure>
<p>也支持多个候选LTR输入：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LTR_retriever -genome TAIR10.fa -inharvest TAIR10.harvest.scn -infinder TAIR10.finder.scn -threads 20</span><br></pre></td></tr></table></figure>
<p>输出文件如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br></pre></td><td class="code"><pre><span class="line">(base) [fenglei@localhost 20210719_LTR_retriever_test]$ LTR_retriever -genome genome.fasta -inharvest genome.fasta.rawLTR.scn -threads 10</span><br><span class="line"></span><br><span class="line">##########################</span><br><span class="line">### LTR_retriever v2.9.0 ###</span><br><span class="line">##########################</span><br><span class="line"></span><br><span class="line">Contributors: Shujun Ou, Ning Jiang</span><br><span class="line"></span><br><span class="line">For LTR_retriever, please cite:</span><br><span class="line"></span><br><span class="line">        Ou S and Jiang N (2018). LTR_retriever: A Highly Accurate and Sensitive Program for Identification of Long Terminal Repeat Retrotransposons. Plant Physiol. 176(2): 1410-1422.</span><br><span class="line"></span><br><span class="line">For LAI, please cite:</span><br><span class="line"></span><br><span class="line">        Ou S, Chen J, Jiang N (2018). Assessing genome assembly quality using the LTR Assembly Index (LAI). Nucleic Acids Res. 2018;46(21):e126.</span><br><span class="line"></span><br><span class="line">Parameters: -genome genome.fasta -inharvest genome.fasta.rawLTR.scn -threads 10</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:09:05 CST 2021    Dependency checking: All passed!</span><br><span class="line">Mon Jul 19 09:09:24 CST 2021    LTR_retriever is starting from the Init step.</span><br><span class="line">Mon Jul 19 09:09:26 CST 2021    Start to convert inputs...</span><br><span class="line">                                Total candidates: 1110</span><br><span class="line">                                Total uniq candidates: 1045</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:09:28 CST 2021    Module 1: Start to clean up candidates...</span><br><span class="line">                                Sequences with 10 missing bp or 0.8 missing data rate will be discarded.</span><br><span class="line">                                Sequences containing tandem repeats will be discarded.</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:09:48 CST 2021    950 clean candidates remained</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:09:48 CST 2021    Modules 2-5: Start to analyze the structure of candidates...</span><br><span class="line">                                The terminal motif, TSD, boundary, orientation, age, and superfamily will be identified in this step.</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:16:20 CST 2021    Intact LTR-RT found: 230</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:16:24 CST 2021    Module 6: Start to analyze truncated LTR-RTs...</span><br><span class="line">                                Truncated LTR-RTs without the intact version will be retained in the LTR-RT library.</span><br><span class="line">                                Use -notrunc if you don&#x27;t want to keep them.</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:16:24 CST 2021    66 truncated LTR-RTs found</span><br><span class="line">Mon Jul 19 09:16:41 CST 2021    29 truncated LTR sequences have added to the library</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:16:41 CST 2021    Module 5: Start to remove DNA TE and LINE transposases, and remove plant protein sequences...</span><br><span class="line">                                Total library sequences: 355</span><br><span class="line">Mon Jul 19 09:18:52 CST 2021    Retained clean sequence: 348</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:18:52 CST 2021    Sequence clustering for genome.fasta.ltrTE ...</span><br><span class="line">Mon Jul 19 09:18:52 CST 2021    Unique lib sequence: 347</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:18:55 CST 2021    Module 6: Start to remove nested insertions in internal regions...</span><br><span class="line">Mon Jul 19 09:27:23 CST 2021    Raw internal region size (bit): 961029</span><br><span class="line">                                Clean internal region size (bit): 912840</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:24 CST 2021    Sequence number of the redundant LTR-RT library: 659</span><br><span class="line">                                The redundant LTR-RT library size (bit): 1422033</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:24 CST 2021    Module 8: Start to make non-redundant library...</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:27 CST 2021    Final LTR-RT library entries: 342</span><br><span class="line">                                Final LTR-RT library size (bit): 1025212</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:27 CST 2021    Total intact LTR-RTs found: 206</span><br><span class="line">                                Total intact non-TGCA LTR-RTs found: 43</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:27:28 CST 2021    Start to annotate whole-genome LTR-RTs...</span><br><span class="line">                                Use -noanno if you don&#x27;t want whole-genome LTR-RT annotation.</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">######################################</span><br><span class="line">### LTR Assembly Index (LAI) beta3.2 ###</span><br><span class="line">######################################</span><br><span class="line"></span><br><span class="line">Developer: Shujun Ou</span><br><span class="line"></span><br><span class="line">Please cite:</span><br><span class="line"></span><br><span class="line">Ou S., Chen J. and Jiang N. (2018). Assessing genome assembly quality using the LTR Assembly Index (LAI). Nucleic Acids Res. gky730: https://doi.org/10.1093/nar/gky730</span><br><span class="line"></span><br><span class="line">Parameters: -genome genome.fasta -intact genome.fasta.pass.list -all genome.fasta.out -t 10 -q -blast /home/fenglei/local/app/blast-2.11.0/bin/</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:32:57 CST 2021    Dependency checking: Passed!</span><br><span class="line">Mon Jul 19 09:32:57 CST 2021    Calculation of LAI will be based on the whole genome.</span><br><span class="line">                                Please use the -mono parameter if your genome is a recent ployploid, for high identity between homeologues will overcorrect raw LAI scores.</span><br><span class="line">Mon Jul 19 09:32:57 CST 2021    Estimate the identity of LTR sequences in the genome: quick mode</span><br><span class="line">Mon Jul 19 09:33:08 CST 2021    The identity of LTR sequences: 93.5158904793606%</span><br><span class="line">Mon Jul 19 09:33:08 CST 2021    Calculate LAI:</span><br><span class="line"></span><br><span class="line">                                                Done!</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:33:11 CST 2021    Result file: genome.fasta.out.LAI</span><br><span class="line"></span><br><span class="line">                                You may use either raw_LAI or LAI for intraspecific comparison</span><br><span class="line">                                but please use ONLY LAI for interspecific comparison</span><br><span class="line"></span><br><span class="line">Mon Jul 19 09:33:11 CST 2021    All analyses were finished!</span><br><span class="line"></span><br><span class="line">##############################</span><br><span class="line">####### Result files #########</span><br><span class="line">##############################</span><br><span class="line"></span><br><span class="line">Table output for intact LTR-RTs (detailed info)</span><br><span class="line">        genome.fasta.pass.list (All LTR-RTs)</span><br><span class="line">        genome.fasta.nmtf.pass.list (Non-TGCA LTR-RTs)</span><br><span class="line">        genome.fasta.pass.list.gff3 (GFF3 format for intact LTR-RTs)</span><br><span class="line"></span><br><span class="line">LTR-RT library</span><br><span class="line">        genome.fasta.LTRlib.redundant.fa (All LTR-RTs with redundancy)</span><br><span class="line">        genome.fasta.LTRlib.fa (All non-redundant LTR-RTs)</span><br><span class="line">        genome.fasta.nmtf.LTRlib.fa (Non-TGCA LTR-RTs)</span><br><span class="line"></span><br><span class="line">Whole-genome LTR-RT annotation by the non-redundant library</span><br><span class="line">        genome.fasta.LTR.gff3 (GFF3 format)</span><br><span class="line">        genome.fasta.out.fam.size.list (LTR family summary)</span><br><span class="line">        genome.fasta.out.superfam.size.list (LTR superfamily summary)</span><br><span class="line"></span><br><span class="line">LTR Assembly Index (LAI)</span><br><span class="line">        genome.fasta.out.LAI</span><br></pre></td></tr></table></figure>
<h3 id="ltr-rts-insertion-time"><a class="markdownIt-Anchor" href="#ltr-rts-insertion-time"></a> LTR-RTs insertion time</h3>
<p>根据LTR-RT的结构特点，即插入基因组新位置之后，反转录转座子的两端的LTR会独立进化，各自积累突变，计算出他们之间的遗传距离与进化时间，就可推断基因组里面大量LTR-RT出现的规律。“LTR-RTs are a case where we can exploit biology to provide insight into larger processes. LTR-RTs are a class of retrotransposon with long terminal repeats, which are identical upon insertion at the beginning and end of the TE’s sequence. Over time, mutations accumulate in these terminal repeats, roughly following a molecular clock. This allows us to more or less count the differences between the two LTRs of a LTR-RT, and calculate the time (in generations ago) it was inserted. If we do this for all LTR-RTs in a genome, we get a distribution of insertion times, and shifts in this distribution can tell us about the history of TE activity and/or selective pressures against TEs in our favourite species or between species.”</p>
<p>计算LTR插入时间可以参考：<a target="_blank" rel="noopener" href="https://github.com/SIWLab/Lab_Info/wiki/Ageing-LTR-insertions">https://github.com/SIWLab/Lab_Info/wiki/Ageing-LTR-insertions</a></p>
<p>按照上面的流程，运行LTRretriever之后得到一系列文件，其中genome.fasta.pass.list.gff3是最终的LTR列表，包含三种类型：Copia、Gypsy和unknown。一般文献会分别统计Copia、Gypsy和non-autonomous这三个类型的LTR。那么non-autonomous是否就是unknown呢？我将这个gff3文件导入IGV浏览器，发现unknown LTR的特点就是具有双端LTR结构，但是中间区域很短，意味着其很可能不具备gag和pol序列，无法独立执行这个转座子单元的反转录的过程。</p>
<p><img src="https://i.imgur.com/AAVN4k5.jpeg" alt="" /></p>
<p>下截图所示，是gff3的文件内容，每个LTR-RT一般有6行，第3和第5行就是两端的LTR区域，将这两行提取出来，存入一个文件。我写了一个Perl脚本，逐个LTR-RT分别存入新的文件，每个文件两行，比如LTR_1.gff3、LTR_2.gff3、LTR_3.gff3 …… 然后基于这些gff3文件，用bedtools分别提取fasta序列，再用muscle软件做序列比对（注意muscle软件只能按照两条序列自然的方向作比较，他不会考虑两条序列是否反向之后匹配更好，当然，LTR-RT的两端的LTR序列本来就是方向一致的，所以不用担心。），muscle输出对齐的fasta格式文件，然后运用程序计算双端LTR之间的距离K，再依据 T=K/(2r) 公式就能推测到插入时间。将“插入时间”数据统一存入一个文本，导入R，用ggplot2的geom_histogram功能就可以得到下面的图。</p>
<p>上面公式中r的意思是“synonymous nucleotide substitution rate”，“r is the average substitution rate, taken as the Ks value estimated above”，<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1456784/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1456784/</a>，7e-9 substitutions/site/year, with standard deviation of 2.6e−9, was used as the rate of synonymous substitutions ®. 根据这篇文章，豆科的平均替换率是7e-9 substitutions/site/year，我在我研究的物种中也用这个数值。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">#########################################################################</span><br><span class="line"># File Name: ltr.sh</span><br><span class="line"># Author: fenglei</span><br><span class="line"># Mail: fengleiluck@gmail.com</span><br><span class="line"># Created Time: Wed 21 Jul 2021 01:57:33 PM CST</span><br><span class="line">#########################################################################</span><br><span class="line">#!/usr/bin/bash</span><br><span class="line">set -e</span><br><span class="line">set -u</span><br><span class="line">set -o pipefail</span><br><span class="line"></span><br><span class="line">for IN in `find ./loop/ -name &quot;LTR*.gff3&quot;`; do</span><br><span class="line">        echo [`date +&quot;%Y-%m-%d %H:%M:%S&quot;`] &quot;   * processing $IN&quot;</span><br><span class="line">        path=$&#123;IN%/*&#125;    #  ./loop/LTR_85.gff3 -&gt; ./loop</span><br><span class="line">        file=$&#123;IN##*/&#125;   #  ./loop/LTR_85.gff3 -&gt; LTR_85.gff3</span><br><span class="line">        name=$&#123;file%.*&#125;  #  LTR_85.gff3 -&gt; LTR_85</span><br><span class="line">        # cat genome.fasta.finder.combine.scn genome.fasta.harvest.scn_new &gt;genome.fasta.rawLTR.scn </span><br><span class="line">        # awk &#x27;NR == 9 || NR == 11&#x27; genome.fasta.pass.list.gff3 &gt; LTR_1.gff3</span><br><span class="line">        # bedtools getfasta -fi genome.fasta -bed LTR_1.gff3 &gt; $&#123;name&#125;.fasta </span><br><span class="line">        bedtools getfasta -fi genome.fasta -bed $IN &gt; ./loop/$&#123;name&#125;.fasta</span><br><span class="line">        ~/local/app/muscle3.8.31/muscle3.8.31_i86linux64 -in ./loop/$&#123;name&#125;.fasta -out ./loop/$&#123;name&#125;.afa</span><br><span class="line">        # cat LTR_1.pair_diff</span><br><span class="line">        python ../../20210720_LTR_evolution/LTR_pairwise_differences.py --input ./loop/$&#123;name&#125;.afa --output ./loop/$&#123;name&#125;.pair_diff</span><br><span class="line">done</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/Xkpiivl.jpeg" alt="" /></p>
<p>附画图R语言代码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)</span><br><span class="line">setwd(&quot;C://R-data&quot;)</span><br><span class="line">ltr &lt;- read.table(&quot;Arabidopsis_LTR.txt&quot;, head=F)</span><br><span class="line">head(ltr)</span><br><span class="line">colnames(ltr) &lt;- c(&quot;locus&quot;, &quot;K&quot;, &quot;T&quot;)</span><br><span class="line">head(ltr$locus)</span><br><span class="line">ltr$newT &lt;- ltr$T/1000000</span><br><span class="line">head(ltr)</span><br><span class="line">ggplot() +</span><br><span class="line">  geom_histogram(data=ltr, aes(x=newT), fill=&quot;#0570c0&quot;, colour=&quot;black&quot;) +</span><br><span class="line">  theme_bw() +</span><br><span class="line">  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = &quot;black&quot;)) +</span><br><span class="line">  theme(axis.text.x = element_text(size = 15, color = &quot;black&quot;, vjust = 0.5 , hjust = 0.5, angle = 0)) +</span><br><span class="line">  theme(axis.text.y = element_text(color = &quot;black&quot;, size = 15)) +</span><br><span class="line">  ylab(&quot;Copy Number&quot;) +</span><br><span class="line">  xlab(&quot;Insertion time (Myr)&quot;) +</span><br><span class="line">  #ggtitle(&quot;4DTv density of paralogs gene pairs in legumes&quot;) +</span><br><span class="line">  theme(axis.title.y = element_text(size = 18, color = &quot;black&quot;)) +</span><br><span class="line">  theme(axis.title.x = element_text(size = 18, color = &quot;black&quot;))</span><br></pre></td></tr></table></figure>
<h3 id="ltr-rts-系统发生树"><a class="markdownIt-Anchor" href="#ltr-rts-系统发生树"></a> LTR-RTs 系统发生树</h3>
<ol>
<li>genome.fasta.pass.list.gff3 文件，用bedtools提取对应的LTR-RT的完整DNA序列。</li>
<li>一个LTR-RT中有多个蛋白质domain，如何提取这些domain？用TEsorter可以轻松实现，TEsorter用conda即可安装，我在python3.7环境下安装的。得到不同功能域（GAG、INT、RT、RH……）的氨基酸序列。<br />
（一开始不知道TEsorter存在，自己用blastx和hmmer尝试，效果都不理想）</li>
<li>通过cdhit软件进行序列聚类，将高度近似或相同的序列归类，大大减少下一步比对的序列数目。我从2800聚集成400多条序列。</li>
<li>序列比对。可以用mafft或muscle。我用这windows版本的MEGA软件，MEGA X 内置muscle，，运行数秒即可得到下图。</li>
</ol>
<p><img src="https://i.imgur.com/WMphP1J.jpeg" alt="" /></p>
<ol start="5">
<li>直接在MEGA软件里面选择系统进化分析，选择ML（Maximum Likelihood），其余参数默认，运行很快就得到下图。</li>
</ol>
<p><img src="https://i.imgur.com/ksgYYlM.jpeg" alt="" /></p>
<ol start="6">
<li>
<p>上面的树，导出为Newick格式（.nwk），然后可以用其他软件来美化这个进化树。注意导出的时候可以设定是否导出树枝长度，若不选，那么出来的树就是类似二叉树的形象，树枝都是一样的长度。</p>
</li>
<li>
<p>用Figtree软件打开这个树，可以美化。参考资料 <a target="_blank" rel="noopener" href="https://www.omicsclass.com/article/62">https://www.omicsclass.com/article/62</a>. 比如某个分支需要加颜色背景，就需要点击目录上方的“clade”按钮，接下来点击树枝，就能选中那个分支，再点击目录上方的“Hilight”按钮，就可以设置颜色背景，效果图像如下！</p>
</li>
</ol>
<p><img src="https://i.imgur.com/FNCS9nl.jpeg" alt="" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/" class="post-title-link" itemprop="url">Bioinfo | Gene Ontology - Enrichment Analysis</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-06-29 11:14:12 / Modified: 23:44:19" itemprop="dateCreated datePublished" datetime="2021-06-29T11:14:12+08:00">2021-06-29</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>下面是我做的大豆GO数据库（version Gmax_508_Wm82_a4.v1）构建与富集分析。数据文件在：<a target="_blank" rel="noopener" href="https://github.com/bioinfx/Glycine_max_Gene_Ontology">https://github.com/bioinfx/Glycine_max_Gene_Ontology</a>。</p>
<p>经过前期数据比对和整理（比如依据 swissprot、blast2GO，或 GOMAP），得到下面“基因——GO号——Ontology 类别——证据”这四列格式的文件 GMA_GO_TERM.txt，就可以在R程序里面用于 GO 数据库的构建。</p>
<h3 id="数据库源文件准备"><a class="markdownIt-Anchor" href="#数据库源文件准备"></a> 数据库源文件准备</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ zcat GMA_GO_TERM.txt.gz | head</span><br><span class="line">GID     GO      ONTOLOGY        EVIDENCE</span><br><span class="line">Glyma.01G000100 GO:0006412      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009063      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009234      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0016021      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0031969      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0016835      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0030976      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0070204      F       IEA</span><br><span class="line">Glyma.01G000137 GO:0016043      P       IEA</span><br></pre></td></tr></table></figure>
<h3 id="数据库导入-r"><a class="markdownIt-Anchor" href="#数据库导入-r"></a> 数据库导入 R</h3>
<p>构建过程如下，调用 makeOrgPackage 函数构建 GO 数据库。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><span class="line">## Makes an organism package for Glycine max data.frames:</span><br><span class="line"></span><br><span class="line">library(RSQLite)</span><br><span class="line">library(AnnotationForge)</span><br><span class="line"></span><br><span class="line">## Now prepare some data.frames</span><br><span class="line">fGO &lt;- read.table(&quot;../3_final-result/GMA_GO_TERM.txt&quot;, sep=&quot;\t&quot;, head=TRUE)</span><br><span class="line"></span><br><span class="line">## remove duplication in the table</span><br><span class="line">go_df &lt;- fGO[!duplicated(fGO),]</span><br><span class="line"></span><br><span class="line">## only tree coloums are needed: GID, GO, EVIDENCE</span><br><span class="line">go_df &lt;- go_df[,c(1,2,4)]</span><br><span class="line"></span><br><span class="line">## Making Organism packages </span><br><span class="line">makeOrgPackage(go=go_df,</span><br><span class="line">    version = &quot;0.0.1&quot;,</span><br><span class="line">    maintainer = &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">    author = &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">    outputDir = &quot;./&quot;,</span><br><span class="line">    tax_id = &quot;3847&quot;,</span><br><span class="line">    genus = &quot;Glycine&quot;,</span><br><span class="line">    species = &quot;max&quot;,</span><br><span class="line">    goTable = &quot;go&quot;</span><br><span class="line">    )</span><br><span class="line"></span><br><span class="line">## then you can call install.packages based on the return value</span><br><span class="line">install.packages(&quot;./org.Gmax.eg.db&quot;, repos = NULL, type=&quot;source&quot;)</span><br><span class="line">library(org.Gmax.eg.db)</span><br><span class="line">org &lt;- org.Gmax.eg.db</span><br><span class="line"></span><br><span class="line">## A demo for using the database</span><br><span class="line"># BiocManager::install(&quot;Rgraphviz&quot;)</span><br><span class="line"># BiocManager::install(&quot;topGO&quot;)</span><br><span class="line">library()</span><br><span class="line">geneTable = read.table(&quot;./Gmax_ABA_exp.xls&quot;, head=T)</span><br><span class="line">geneList=geneTable[,5]</span><br><span class="line">geneList=geneList[!duplicated(geneList)]</span><br><span class="line"># Biology process</span><br><span class="line">ego_up &lt;-enrichGO(gene = geneList, OrgDb = org, keyType = &quot;GID&quot;, ont = &quot;BP&quot;)</span><br><span class="line">pdf(&quot;up_GO_BP_barplot.pdf&quot;, height=10, width=15)</span><br><span class="line">barplot(ego_up, drop=TRUE, showCategory=12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;up_GO_BP_dotPlot.pdf&quot;, height=10, width=15)</span><br><span class="line">dotplot(ego_up)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;up_GO_BP_tree.pdf&quot;)</span><br><span class="line">plotGOgraph(ego_up)</span><br><span class="line">dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_up@result ), file=&quot;up_GO_BP.csv&quot;)</span><br></pre></td></tr></table></figure>
<h1 id="数据库开启与富集分析"><a class="markdownIt-Anchor" href="#数据库开启与富集分析"></a> 数据库开启与富集分析</h1>
<p>下面代码示范如何引用上构建的数据库，并且开展 GO 分析：BP、MF 和 CC 三个类别。<br />
图片可输出为柱状图 barplot 和点状图 dotplot，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br></pre></td><td class="code"><pre><span class="line">library(org.Gmax.eg.db)</span><br><span class="line">org &lt;- org.Gmax.eg.db</span><br><span class="line"></span><br><span class="line">## A demo for using the database</span><br><span class="line"># BiocManager::install(&quot;Rgraphviz&quot;)</span><br><span class="line"># BiocManager::install(&quot;topGO&quot;)</span><br><span class="line">library(topGO)</span><br><span class="line">library(clusterProfiler)</span><br><span class="line">geneTable = read.table(&quot;./gene_table.txt&quot;, head=T, sep=&quot;\t&quot;)</span><br><span class="line">#geneTable = geneTable[!apply(geneTable == &quot;&quot;, 1, all),]</span><br><span class="line">geneList=geneTable[,5]</span><br><span class="line"></span><br><span class="line">geneList=geneList[!duplicated(geneList)]</span><br><span class="line"># Biology process</span><br><span class="line">b=geneList</span><br><span class="line"></span><br><span class="line"># biological progress</span><br><span class="line">ego_up &lt;-enrichGO(</span><br><span class="line">                  gene = b, </span><br><span class="line">                  OrgDb = org, </span><br><span class="line">                  keyType = &quot;GID&quot;, </span><br><span class="line">                  ont = &quot;BP&quot;</span><br><span class="line">                  )</span><br><span class="line">pdf(&quot;test_GO_BP_barplot.pdf&quot;, height=10, width=15)</span><br><span class="line">barplot(ego_up, drop=TRUE, showCategory=12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;test_GO_BP_dotPlot.pdf&quot;, height=10, width=15)</span><br><span class="line">dotplot(ego_up)</span><br><span class="line">dev.off()</span><br><span class="line">#pdf(&quot;up_GO_BP_tree.pdf&quot;)</span><br><span class="line">#plotGOgraph(ego_up)</span><br><span class="line">#dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_up@result ), file=&quot;test_GO_BP.csv&quot;)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># molecular function</span><br><span class="line">ego_down &lt;-enrichGO(</span><br><span class="line">                    gene = b,</span><br><span class="line">                    OrgDb = org,</span><br><span class="line">                    keyType = &quot;GID&quot;,</span><br><span class="line">                    ont = &quot;MF&quot;</span><br><span class="line">                    )</span><br><span class="line">pdf(&quot;test_GO_MF_barplot.pdf&quot;, height=10, width=15)</span><br><span class="line">barplot(ego_down, drop=TRUE, showCategory=12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;test_GO_MF_dotPlot.pdf&quot;, height=10, width=15)</span><br><span class="line">dotplot(ego_down)</span><br><span class="line">dev.off()</span><br><span class="line">#pdf(&quot;test_GO_MF_tree.pdf&quot;)</span><br><span class="line">#plotGOgraph(ego_down)</span><br><span class="line">#dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_down@result), file=&quot;test_GO_MF.csv&quot;)</span><br><span class="line"></span><br><span class="line"># cellular component</span><br><span class="line">ego_down &lt;-enrichGO(</span><br><span class="line">                    gene = b,</span><br><span class="line">                    OrgDb = org,</span><br><span class="line">                    keyType = &quot;GID&quot;,</span><br><span class="line">                    ont = &quot;CC&quot;</span><br><span class="line">                    )</span><br><span class="line">pdf(&quot;test_GO_CC_barplot.pdf&quot;, height=10, width=15)</span><br><span class="line">barplot(ego_down, drop=TRUE, showCategory=12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;test_GO_CC_dotPlot.pdf&quot;, height=10, width=15)</span><br><span class="line">dotplot(ego_down)</span><br><span class="line">dev.off()</span><br><span class="line">#pdf(&quot;test_GO_CC_tree.pdf&quot;)</span><br><span class="line">#plotGOgraph(ego_down)</span><br><span class="line">#dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_down@result), file=&quot;test_GO_CC.csv&quot;)</span><br></pre></td></tr></table></figure>
<p>有时候 GO 项目的名称过长，基于默认作图代码出来的图片会出现文字将图形区域挤压得无法显示。</p>
<p><img src="https://i.imgur.com/FlARCCO.jpeg" alt="" /></p>
<p>有两种优化方案，一是将长的文字换行；二是只截取一部分文字显示。</p>
<h3 id="一过长的文字换行显示"><a class="markdownIt-Anchor" href="#一过长的文字换行显示"></a> 一：过长的文字换行显示，</h3>
<p>参考资料：<a target="_blank" rel="noopener" href="https://www.cnblogs.com/songbiao/articles/12717670.html">songbiao: GO term name太长的处理</a><br />
在使用clusterProfiler进行绘制GO dotplot时，会遇到 GO term 的 name 过长，压缩了图形区域的情况。<br />
因为 clusterProfiler 作图是基于 ggplot2 的，所以这个处理办法也适用于其他 ggplot2 作图遇到此种情况。<br />
解决文本过长的办法就是，用stringr包的str_wrap来完成文本自动换行。<br />
效果如下图所示，其实还是不太完美，需要后期手工调整才美观。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)  # 需要开启 ggplot2 才可以使用 scale_y_discrete 函数</span><br><span class="line">library(stringr)</span><br><span class="line">dotplot(mf, showCategory=30) + scale_y_discrete(labels=function(mf) str_wrap(mf, width=40))</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/JRkomgm.jpeg" alt="" /></p>
<h3 id="二只截取一部分文字显示"><a class="markdownIt-Anchor" href="#二只截取一部分文字显示"></a> 二：只截取一部分文字显示</h3>
<p>参考资料：<a target="_blank" rel="noopener" href="https://r.bio-spring.info/2018/11/26/enrichplot-fancy-label/">在enrichplot中使用更好的Label</a><br />
该方法是定义一个截取短 label 的函数, 将其传递给 scale_y_discrete() 即可. 在输出 ggplot 对象时做修改。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line">#&#x27; Truncate string vector of ggplot axis label</span><br><span class="line">#&#x27;</span><br><span class="line">#&#x27; @param label    a ordered string vector</span><br><span class="line">#&#x27; @param maxLen   max length of character (nchar) to show in label</span><br><span class="line">#&#x27; @param maxWord  max count of words allowed to show in label</span><br><span class="line">#&#x27; @param pattern  Word separater</span><br><span class="line">#&#x27; @param dot      If true, three dots will added to truncated label</span><br><span class="line">#&#x27;</span><br><span class="line">#&#x27; @return a vector of truncated strings</span><br><span class="line">#&#x27; @export</span><br><span class="line">#&#x27;</span><br><span class="line">#&#x27; @examples</span><br><span class="line"># 默认最多显示 50 个字符, 5 个单词.</span><br><span class="line">short_label &lt;- function(label, maxLen = 50, maxWord = 5, pattern = &quot; &quot;, dot = TRUE)&#123;</span><br><span class="line">  l &lt;- strsplit(label, pattern)</span><br><span class="line">  short_label &lt;- vector(&quot;character&quot;,length(l))</span><br><span class="line">  </span><br><span class="line">  for (i in seq_along(l))&#123;</span><br><span class="line">    truncated &lt;- FALSE</span><br><span class="line">    s &lt;- l[[i]]</span><br><span class="line">    if (length(s) &gt; maxWord)&#123;</span><br><span class="line">      ll &lt;- paste(s[1:maxWord], collapse = &quot; &quot;)</span><br><span class="line">      truncated &lt;- TRUE</span><br><span class="line">    &#125;</span><br><span class="line">    else&#123;</span><br><span class="line">      ll &lt;- paste(s, collapse = &quot; &quot;)</span><br><span class="line">    &#125;</span><br><span class="line">    </span><br><span class="line">    if (nchar(ll) &gt; maxLen)&#123;</span><br><span class="line">      ll &lt;- substr(ll, 1, maxLen)</span><br><span class="line">      truncated &lt;- TRUE</span><br><span class="line">    &#125;</span><br><span class="line">    </span><br><span class="line">    if (dot &amp; truncated) ll &lt;- paste(ll, &quot;...&quot;,sep = &quot; &quot;)</span><br><span class="line">    </span><br><span class="line">    short_label[[i]] &lt;- ll</span><br><span class="line">  &#125;</span><br><span class="line">  attr(short_label, &quot;pos&quot;) &lt;- attr(label,&quot;pos&quot;)</span><br><span class="line">  return(short_label)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>画图代码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)</span><br><span class="line">dotplot(mf) + scale_y_discrete(label=short_label)</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/pALA2H9.jpeg" alt="" /></p>
<p>注意上面的图，其实很多项目都是高度近似的，怎么优化？对于模式物种而言，CytoScape的插件ClueGO有一个选项可以实现，就是设置Ontology的级别。因为Ontology的结构就是树状，从根部到.</p>
<p>构建GO数据库的过程中，如果成功会输出下面的信息。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(go_df)</span><br><span class="line">           GID         GO EVIDENCE</span><br><span class="line">1 Chr01.g00016 GO:0003676      IEA</span><br><span class="line">2 Chr01.g00016 GO:0003964      IEA</span><br><span class="line">3 Chr01.g00016 GO:0004190      IEA</span><br><span class="line">4 Chr01.g00016 GO:0004519      IEA</span><br><span class="line">5 Chr01.g00016 GO:0006310      IEA</span><br><span class="line">6 Chr01.g00016 GO:0015074      IEA</span><br><span class="line">&gt; makeOrgPackage(go=go_df,</span><br><span class="line">+                version = &quot;0.0.1&quot;,</span><br><span class="line">+                maintainer = &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">+                author = &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">+                outputDir = &quot;.&quot;,</span><br><span class="line">+                tax_id = &quot;126911&quot;,</span><br><span class="line">+                genus = &quot;Ammopiptanthus&quot;,</span><br><span class="line">+                species = &quot;mongolicus&quot;,</span><br><span class="line">+                goTable = &quot;go&quot;</span><br><span class="line">+ )</span><br><span class="line">Populating genes table:</span><br><span class="line">genes table filled</span><br><span class="line">Populating go table:</span><br><span class="line">go table filled</span><br><span class="line">table metadata filled</span><br><span class="line">&#x27;select()&#x27; returned many:1 mapping between keys and columns</span><br><span class="line">Dropping GO IDs that are too new for the current GO.db</span><br><span class="line">Populating go table:</span><br><span class="line">go table filled</span><br><span class="line">Populating go_bp table:</span><br><span class="line">go_bp table filled</span><br><span class="line">Populating go_cc table:</span><br><span class="line">go_cc table filled</span><br><span class="line">Populating go_mf table:</span><br><span class="line">go_mf table filled</span><br><span class="line">&#x27;select()&#x27; returned many:1 mapping between keys and columns</span><br><span class="line">Populating go_bp_all table:</span><br><span class="line">go_bp_all table filled</span><br><span class="line">Populating go_cc_all table:</span><br><span class="line">go_cc_all table filled</span><br><span class="line">Populating go_mf_all table:</span><br><span class="line">go_mf_all table filled</span><br><span class="line">Populating go_all table:</span><br><span class="line">go_all table filled</span><br><span class="line">Creating package in ./org.Amongolicus.eg.db </span><br><span class="line">Now deleting temporary database file</span><br><span class="line">[1] &quot;./org.Amongolicus.eg.db&quot;</span><br><span class="line">There were 50 or more warnings (use warnings() to see the first 50)</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/17/Linux-Gene-Ontology-GOMAP/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/17/Linux-Gene-Ontology-GOMAP/" class="post-title-link" itemprop="url">Bioinfo | Gene Ontology - GOMAP</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-06-17 10:59:17" itemprop="dateCreated datePublished" datetime="2021-06-17T10:59:17+08:00">2021-06-17</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-06-29 11:40:34" itemprop="dateModified" datetime="2021-06-29T11:40:34+08:00">2021-06-29</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>Gene Ontology Meta Annotator for Plants</p>
<p>GOMAP 主要是给植物基因集做 Gene Ontology 注释信息匹配，它综合了多个数据库，所以得到的信息会全面一些。早年将基因对应到GO用的工具是Blast2GO，这个软件linux版本安装比较麻烦，基于mysql数据库，而且后面它商业化之后要收费。后来的策略是用diamond的快速blastp将基因集与SwissProt数据库进行比较，得到每个基因最相近的Uniprot ID，再依靠id-mapping文件将基因与GO对应起来，本文最后有简单介绍基于SwissProt做GO注释的流程。</p>
<p>What is GOMAP-Singularity? GOMAP-Singularity is the containerized version of the Gene Ontology Meta Annotator for Plants (GOMAP) pipeline. GOMAP is a high-throughput pipeline to annotate GO terms to plant protein sequences. The pipeline uses three different approaches to annotate GO terms to plant protein sequences, and compile the annotations together to generate an aggregated dataset. GOMAP uses Python code to run the component tools to generate the GO annotations, and R code to clean and aggregate the GO annotations from the component tools.</p>
<p>How do I use GOMAP-Singularity?</p>
<p>Please check &lt;<a target="_blank" rel="noopener" href="http://bioinformapping.com/gomap">bioinformapping.com/gomap</a>&gt; for detailed documentation on how to use the container.</p>
<h3 id="install-system-dependencies"><a class="markdownIt-Anchor" href="#install-system-dependencies"></a> Install system dependencies</h3>
<p>You must first install development tools and libraries to your host.<br />
On Debian-based systems:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ sudo apt-get update &amp;&amp; \</span><br><span class="line">  sudo apt-get install -y build-essential \</span><br><span class="line">  libseccomp-dev pkg-config squashfs-tools cryptsetup</span><br></pre></td></tr></table></figure>
<p>On CentOS/RHEL:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sudo yum groupinstall -y &#x27;Development Tools&#x27; &amp;&amp; \</span><br><span class="line">  sudo yum install -y epel-release &amp;&amp; \</span><br><span class="line">  sudo yum install -y golang libseccomp-devel \  </span><br><span class="line">  squashfs-tools cryptsetup</span><br><span class="line">  yum remove golang # 后面手动安装</span><br></pre></td></tr></table></figure>
<h3 id="install-golang"><a class="markdownIt-Anchor" href="#install-golang"></a> Install Golang</h3>
<p>Golang 有多种安装方法，我是在其官网(<a target="_blank" rel="noopener" href="https://golang.org/doc/install#download">https://golang.org/doc/install#download</a>)下载<a target="_blank" rel="noopener" href="https://golang.org/dl/go1.16.5.linux-amd64.tar.gz">源文件</a>到本地，然后安装到用户目录。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">wget https://golang.org/dl/go1.16.5.linux-amd64.tar.gz</span><br><span class="line">tar -zxvf go1.16.5.linux-amd64.tar.gz</span><br><span class="line"></span><br><span class="line"># 将 “export PATH=&quot;/home/fenglei/local/app/go/bin&quot;:$PATH” 写入bashrc</span><br><span class="line">source ~/.bashrc</span><br><span class="line">which go</span><br><span class="line"># ~/local/app/go/bin/go</span><br><span class="line"></span><br><span class="line"># Verify that you&#x27;ve installed Go by opening a command prompt and typing the following command:</span><br><span class="line">go version</span><br><span class="line"># go version go1.16.5 linux/amd64</span><br></pre></td></tr></table></figure>
<h3 id="compiling-singularityce"><a class="markdownIt-Anchor" href="#compiling-singularityce"></a> Compiling SingularityCE</h3>
<p>参考安装方法如下</p>
<p>You can build SingularityCE using the following commands:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ cd $&#123;GOPATH&#125;/src/github.com/sylabs/singularity &amp;&amp; \</span><br><span class="line">  ./mconfig &amp;&amp; \</span><br><span class="line">  cd ./builddir &amp;&amp; \</span><br><span class="line">  make &amp;&amp; \</span><br><span class="line">  sudo make install</span><br></pre></td></tr></table></figure>
<p>And that’s it! Now you can check your SingularityCE version by running:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ singularity version</span><br></pre></td></tr></table></figure>
<p>To build in a different folder and to set the install prefix to a different path:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ ./mconfig -b ./buildtree -p /usr/local</span><br></pre></td></tr></table></figure>
<p>我的安装如下。</p>
<p>系统自带的gcc（版本4.8.5）编译报错，提示找不到config.c文件。我切换成用户的gcc（版本10.2.0），就可以安装。<br />
启动用户gcc的办法，修改bashrc文件，将下面三行的注释符号去掉，然后source ~/.bashrc即可。以后有必要再回来将这几行注释掉，就恢复系统自带的老版本gcc。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">export PATH=/home/fenglei/local/gcc-10.2.0/bin/:$PATH</span><br><span class="line">export LD_LIBRARY_PATH=/home/fenglei/local/gcc-10.2.0/lib/:/home/fenglei/local/gcc-10.2.0/lib64/:$LD_LIBRARY_PATH</span><br><span class="line">export LD_LIBRARY_PATH=/home/fenglei/local/lib/:/home/fenglei/local/lib64/:$LD_LIBRARY_PATH</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">$ git clone https://github.com/sylabs/singularity.git</span><br><span class="line">$ cd singularity  # 该目录下有一个 mconfig 文件</span><br><span class="line"># ./mconfig -b ./buildtree -p /home/fenglei/local</span><br><span class="line"># ./mconfig -b ./buildtree -p /home/fenglei/local --without-suid</span><br><span class="line">$ ./mconfig    # 默认的安装位置是 /usr/local</span><br><span class="line">$ cd /home/fenglei/local/app/singularity/buildtree</span><br><span class="line">$ make</span><br><span class="line"># make install</span><br><span class="line">$ sudo make install</span><br><span class="line"></span><br><span class="line">$ which singularity</span><br><span class="line">/usr/local/bin/singularity</span><br></pre></td></tr></table></figure>
<p>测试：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ singularity version                                                                                        </span><br><span class="line">3.8.0-rc.1+107-g8b21e5a16</span><br></pre></td></tr></table></figure>
<h3 id="gomap-singularity"><a class="markdownIt-Anchor" href="#gomap-singularity"></a> GOMAP-singularity</h3>
<ol>
<li>Singularity 安装</li>
</ol>
<p>上面已经完成。也可参考：<a target="_blank" rel="noopener" href="https://bioinformapping.com/gomap/master/RUNNING.html">https://bioinformapping.com/gomap/master/RUNNING.html</a></p>
<ol start="2">
<li>从 git 下载 GOMAP-singularity</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/Dill-PICL/GOMAP-singularity.git</span><br><span class="line">cd GOMAP-singularity</span><br></pre></td></tr></table></figure>
<ol start="3">
<li>配置并下载相关参考数据集</li>
</ol>
<p>Run the setup step to make necessary directories and download data files from CyVerse. The pipeline download is large and would require ~40GB of free hard drive space during the setup step.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">$ ./setup.sh </span><br><span class="line"># 由于GO没安好，导致这一步报错，后面重新安装了GO，并重装singularity</span><br><span class="line"># ERROR  : No setuid installation found, for unprivileged installation use: ./mconfig --without-suid</span><br><span class="line"></span><br><span class="line">$ ./setup.sh </span><br><span class="line">0/1 -  0.00% of files done   0.000/39254.793 MB -  0.00% of file sizes done</span><br><span class="line">Processing GOMAP.sif - 39254.793 MB   2021-06-17.15:58:52</span><br><span class="line">GOMAP.sif - 40.000/39254.793 MB -  0.10% done   2021-06-17.16:00:53</span><br><span class="line">GOMAP.sif - 680.000/39254.793 MB -  1.73% done   2021-06-17.16:02:47</span><br><span class="line">GOMAP.sif - 1360.000/39254.793 MB -  3.46% done   2021-06-17.16:04:49</span><br><span class="line">GOMAP.sif - 2000.000/39254.793 MB -  5.09% done   2021-06-17.16:06:40</span><br><span class="line">GOMAP.sif - 2400.000/39254.793 MB -  6.11% done   2021-06-17.16:08:32</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">...</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">GOMAP.sif - 30880.000/39254.793 MB - 78.67% done   2021-06-17.17:59:30</span><br><span class="line">GOMAP.sif - 31520.000/39254.793 MB - 80.30% done   2021-06-17.18:01:43</span><br><span class="line">GOMAP.sif - 32040.000/39254.793 MB - 81.62% done   2021-06-17.18:03:45</span><br><span class="line">GOMAP.sif - 32640.000/39254.793 MB - 83.15% done   2021-06-17.18:05:31</span><br><span class="line">GOMAP.sif - 33200.000/39254.793 MB - 84.58% done   2021-06-17.18:07:27</span><br><span class="line">GOMAP.sif - 33373.426/39254.793 MB - 85.02% done   2021-06-17.18:08:24</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">GOMAP.sif - 39254.793/39254.793 MB - 100.00% done   2021-06-17.18:45:46</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<ol start="4">
<li>测试 GOMAP 注释流程</li>
</ol>
<p>[optional]: Test whether the container and the data files are working as intended.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">$ ./test.sh</span><br><span class="line"></span><br><span class="line">Running Sequence-similarity based Annotation Step</span><br><span class="line">/usr/local/lib/python2.7/dist-packages/Bio/SearchIO/__init__.py:211: BiopythonExperimentalWarning: Bio.SearchIO is an experimental submodule which may undergo significant changes prior to its future official release.</span><br><span class="line">  BiopythonExperimentalWarning)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Building a new DB, current time: 06/17/2021 20:59:09</span><br><span class="line">New DB name:   /workdir/test/GOMAP-0.3_GOMAP-input/input/0.3_GOMAP-input.fa</span><br><span class="line">New DB title:  /workdir/test/GOMAP-0.3_GOMAP-input/input/0.3_GOMAP-input.fa</span><br><span class="line">Sequence type: Protein</span><br><span class="line">...</span><br><span class="line">Checking gaf file for simple errors and fixing them</span><br><span class="line">There were 34 warnings (use warnings() to see them)</span><br><span class="line">[1] &quot;Checking if data/data/go/go.obo.data exists&quot;</span><br><span class="line">[1] &quot;data/data/go/go.obo.data exists so loading R object&quot;</span><br><span class="line">   user  system elapsed </span><br><span class="line">  2.547   0.078   2.624 </span><br><span class="line">[1] &quot;Reading All Datasets&quot;</span><br><span class="line">Checking gaf file for simple errors and fixing them</span><br><span class="line">[1] 1714</span><br><span class="line">[1] &quot;Removing Redundancy&quot;</span><br><span class="line">[1] 18</span><br><span class="line">Compiling Non-Redundant Comprehensive Data</span><br><span class="line">There were 50 or more warnings (use warnings() to see the first 50)</span><br></pre></td></tr></table></figure>
<ol start="5">
<li>实战篇：学习自定义配置文件 Edit the config file</li>
</ol>
<p>a. Declare export GOMAP_LOC environment variable. Edit <code>~/.bashrc</code> as below:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">export GOPATH=&quot;/home/fenglei/local/app/go&quot;</span><br><span class="line">export GOMAP_LOC=&quot;/home/fenglei/local/app/GOMAP-singularity/install/location&quot;</span><br></pre></td></tr></table></figure>
<p>b. Download the <a target="_blank" rel="noopener" href="https://bioinformapping.com/gomap/master/_static/min-config.yml">config.yml</a> file and make necessary changes. Change the highlighted lines to fit your input data.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">#Input section </span><br><span class="line">input:</span><br><span class="line">  #input fasta file name</span><br><span class="line">  fasta: GOMAP-input.fa</span><br><span class="line">  # output file basename</span><br><span class="line">  basename: GOMAP_output</span><br><span class="line">  #input NCBI taxonomy id</span><br><span class="line">  taxon: &quot;123456&quot;</span><br><span class="line">  # Name of the species</span><br><span class="line">  species: &quot;Glycine max&quot;</span><br><span class="line">  # Email is mandatory</span><br><span class="line">  email: myname#gmail.com</span><br><span class="line">  #Number of CPUs used for tools</span><br><span class="line">  cpus: 4</span><br><span class="line">  #Whether openmpi should be used</span><br><span class="line">  mpi: False</span><br><span class="line">  #what the name of the temporary directory is</span><br><span class="line">  tmpdir: &quot;/tmpdir&quot;</span><br></pre></td></tr></table></figure>
<ol start="6">
<li>Run pipeline</li>
</ol>
<p>可直接修改 <a target="_blank" rel="noopener" href="http://test.sh">test.sh</a> 并运行该脚本。也可参考软件手册分步骤运行。<br />
下面摘抄的是在单机（SINGLE）运行的模式，软件手册里也有在集群（MPI）运行的示范代码。<br />
<a target="_blank" rel="noopener" href="http://test.sh">test.sh</a> 与 test目录在同一个路径。GOMAP-input.fa 与 config.yml 都在test目录下。输入的蛋白序列，其尾部如果有表示stop codon的星号，要移除星号。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"># seqsim</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=seqsim --config=test/config.yml</span><br><span class="line"># domain</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=domain --config=test/config.yml</span><br><span class="line"># fanngo</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=fanngo --config=test/config.yml</span><br><span class="line"># mixmeth-blast</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=mixmeth-blast --config=test/config.yml</span><br><span class="line"># mixmeth-preproc</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=mixmeth-preproc --config=test/config.yml</span><br><span class="line"># mixmeth</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=mixmeth --config=test/config.yml</span><br><span class="line"># aggregate</span><br><span class="line"># Attention: Please wait for all your Argot2.5 jobs to finish before running this step. You will get emails from Argot2.5 when your jobs are submitted and when they are finished. You can also check the status of all current jobs from all users here.</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=aggregate --config=test/config.yml</span><br></pre></td></tr></table></figure>
<h3 id="附注用-swissprot-与-idmapping-来做-go"><a class="markdownIt-Anchor" href="#附注用-swissprot-与-idmapping-来做-go"></a> 附注：用 SwissProt 与 idmapping 来做 GO</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">## SwissProt 文件是氨基酸序列，比较大</span><br><span class="line">wget ftp://ftp.ncbi.nih.gov/blast/db/FASTA/swissprot.gz</span><br><span class="line">## idmapping.tb.gz 文件可能下载链接有变化</span><br><span class="line">wget ftp://ftp.pir.georgetown.edu/databases/idmapping/idmapping.tb.gz</span><br><span class="line">## swissprot index 建立</span><br><span class="line">diamond makedb --in swissprot -d swissprot</span><br><span class="line">## 序列比对</span><br><span class="line">diamond blastp -d /PATH/TO/index/swissprot -q input.proteins.fa -k 1 -e 0.00001 -o swissprot_blastp.out</span><br><span class="line">## idmapping 文件整理，下文有摘抄文件具体内容</span><br><span class="line">awk -F &quot;\t&quot;  &#x27;&#123;print $1&quot;\t&quot;$8&#125;&#x27; idmapping.tb &gt; uniprotID-GOnumber.tab</span><br><span class="line">## 将“基因-Uniprot-GO”的关系建立起来</span><br><span class="line">perl UniProt2GO_annotate.pl uniprotID-GOnumber.tab swiss_dia_matches.m8 swiss_GO_matches.tab</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ head idmapping.tb </span><br><span class="line">Q6GZX4  001R_FRG3G      2947773 YP_031579.1     81941549; 49237298              PF04947 GO:0006355; GO:0046782; GO:0006351                      UniRef100_Q6GZX4        UniRef90_Q6GZX4 UniRef50_Q6GZX4 UPI00003B0FD4           65492415165820 AY548484        AAT09660.1</span><br><span class="line">Q6GZX3  002L_FRG3G      2947774 YP_031580.1     49237299; 81941548; 47060117            PF03003 GO:0033644; GO:0016021                  UniRef100_Q6GZX3        UniRef90_Q6GZX3 UniRef50_Q6GZX3 UPI00003B0FD5           654924        15165820 AY548484        AAT09661.1</span><br><span class="line">Q197F8  002R_IIV3       4156251 YP_654574.1     109287880; 123808694; 106073503                                         UniRef100_Q197F8        UniRef90_Q197F8 UniRef50_Q197F8 UPI0000D83464           345201                        16912294 DQ643392        ABF82032.1</span><br><span class="line">Q197F7  003L_IIV3       4156252 YP_654575.1     106073504; 109287881; 123808693                                         UniRef100_Q197F7        UniRef90_Q197F7 UniRef50_Q197F7 UPI0000D83465           345201                        16912294 DQ643392        ABF82033.1</span><br><span class="line">Q6GZX2  003R_FRG3G      2947775 YP_031581.1     81941547; 49237300                                              UniRef100_Q6GZX2        UniRef90_Q6GZX2 UniRef50_Q6GZX2 UPI00003B0FD6           654924                          15165820       AY548484        AAT09662.1</span><br><span class="line">Q6GZX1  004R_FRG3G      2947776 YP_031582.1     49237301; 81941546                      GO:0033644; GO:0016021                  UniRef100_Q6GZX1        UniRef90_Q6GZX1 UniRef50_Q6GZX1 UPI00003B0FD7           654924                15165820 AY548484        AAT09663.1</span><br><span class="line">Q197F5  005L_IIV3       4156254 YP_654577.1     123808691; 109287883; 106073506                                         UniRef100_Q197F5        UniRef90_Q197F5 UniRef50_Q197F5 UPI0000D83467           345201                        16912294 DQ643392        ABF82035.1</span><br><span class="line">Q6GZX0  005R_FRG3G      2947777 YP_031583.1     47060120; 49237302; 81941545            PF02393                         UniRef100_Q6GZX0        UniRef90_Q6GZX0 UniRef50_Q6GZX0 UPI00003B0FD8           654924                        15165820 AY548484        AAT09664.1</span><br><span class="line">Q91G88  006L_IIV6       1733056 NP_149469.1     82012333; 15078718; 15042163            PF04383; PF12299                                UniRef100_Q91G88        UniRef90_Q91G88 UniRef50_Q91G88 UPI00000F3D32           176652        2820141; 3959991; 7698884; 8492091; 17239238; 8021587; 10456793; 8121799; 9926400; 16789247; 11448171; 1475907; 3201750; 9482589; 1549908; 8073636       AF303741        AAK81943.1</span><br><span class="line">Q6GZW9  006R_FRG3G      2947778 YP_031584.1     81941544; 49237303                                              UniRef100_Q6GZW9        UniRef90_Q6GZW9 UniRef50_Q6GZW9 UPI00003B0FD9           654924                          15165820       AY548484        AAT09665.1</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ head swiss_GO_matches.tab</span><br><span class="line">Q6GZX4  GO:0006355; GO:0046782; GO:0006351</span><br><span class="line">Q6GZX3  GO:0033644; GO:0016021</span><br><span class="line">Q197F8</span><br><span class="line">Q197F7</span><br><span class="line">Q6GZX2</span><br><span class="line">Q6GZX1  GO:0033644; GO:0016021</span><br><span class="line">Q197F5</span><br><span class="line">Q6GZX0</span><br><span class="line">Q91G88</span><br><span class="line">Q6GZW9</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ head swiss_GO_matches.tab</span><br><span class="line">gene1 Q9M9M9  GO:0050897; GO:0009055; GO:0008137; GO:0009735; GO:0006979; GO:0005747; GO:0005739</span><br><span class="line">gene2 Q9LP77  GO:0005524; GO:0004672; GO:0016021; GO:0005886; GO:0009506; GO:0005774</span><br><span class="line">gene3 Q9CAL2  GO:0005524; GO:0004674; GO:0016021; GO:0005886</span><br><span class="line">gene4 A4FLF6</span><br><span class="line">gene5 Q8RY67  GO:0005524; GO:0004674; GO:0007166; GO:0016021; GO:0005886</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">$ cat UniProt2GO_annotate.pl</span><br><span class="line">#!/usr/bin/perl</span><br><span class="line">use warnings;</span><br><span class="line">use strict;</span><br><span class="line"></span><br><span class="line">open IDMAPPING, $ARGV[0] or die $!;</span><br><span class="line">open BLASTOUT, $ARGV[1] or die $!;</span><br><span class="line">open OUT, &quot;&gt;$ARGV[2]&quot; or die $!;</span><br><span class="line">select OUT;</span><br><span class="line"></span><br><span class="line">my %id2go;</span><br><span class="line"></span><br><span class="line">while(&lt;IDMAPPING&gt;)&#123;</span><br><span class="line">        chomp $_;</span><br><span class="line">        my @array1=split(/\t/);</span><br><span class="line">        $id2go&#123;$array1[0]&#125;=$array1[1];</span><br><span class="line">#       print &quot;$array1[1]\n&quot;; exit;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">while(&lt;BLASTOUT&gt;)&#123;</span><br><span class="line">        chomp $_;</span><br><span class="line">        my @array2=split(/\t/);</span><br><span class="line">        my $geneid=$array2[0];</span><br><span class="line">        my $UniprotID=$array2[1];</span><br><span class="line">#       print &quot;$UniprotID\n&quot;;</span><br><span class="line">        $UniprotID =~ s/\.[0-9]$//;</span><br><span class="line">#       print &quot;$geneid\t$&#123;UniprotID&#125;test\n&quot;; exit;</span><br><span class="line">        if(exists $id2go&#123;$UniprotID&#125;)&#123; print &quot;$geneid\t$UniprotID\t$id2go&#123;$UniprotID&#125;\n&quot;;&#125;</span><br><span class="line">        else&#123;print &quot;$geneid\t$UniprotID\n&quot;;&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">close IDMAPPING;</span><br><span class="line">close BLASTOUT;</span><br><span class="line">close OUT;</span><br></pre></td></tr></table></figure>
<p>经过一系列的转换，得到下面格式的文件，就可以在R程序里面用于GO数据库的构建。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ zcat GMA_GO_TERM.txt.gz | head</span><br><span class="line">GID     GO      ONTOLOGY        EVIDENCE</span><br><span class="line">Glyma.01G000100 GO:0006412      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009063      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009234      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0016021      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0031969      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0016835      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0030976      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0070204      F       IEA</span><br><span class="line">Glyma.01G000137 GO:0016043      P       IEA</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/08/Linux-PKG-CONFIG-PATH/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/08/Linux-PKG-CONFIG-PATH/" class="post-title-link" itemprop="url">Linux | 认识 PKG_CONFIG_PATH</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-06-08 15:55:09 / Modified: 16:21:29" itemprop="dateCreated datePublished" datetime="2021-06-08T15:55:09+08:00">2021-06-08</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>安装 <a target="_blank" rel="noopener" href="https://github.com/agordon/fastx_toolkit/releases/download/0.0.14/fastx_toolkit-0.0.14.tar.bz2">Fastx-Toolkit</a> 的过程中遇问题：No package ‘gtextutils’ found. Consider adjusting the PKG_CONFIG_PATH environment variable if you.</p>
<p>实际上我已经安装 <a target="_blank" rel="noopener" href="https://github.com/agordon/libgtextutils/releases/download/0.7/libgtextutils-0.7.tar.gz">libgtextutils</a></p>
<p>解决办法：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sudo cp /usr/local/lib/pkgconfig/gtextutils.pc /usr/lib64/pkgconfig/</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=1490">参考资料</a>：</p>
<p>安装源码包，进行 configure 步骤，提示某依赖的软件包找不到或软件包版本太低；或某一个库文件找不到。而实际上却已经安装好了相应的软件。这两种情况的处理方法分别为：</p>
<ol>
<li>设置环境变量 PKG_CONFIG_PATH</li>
</ol>
<p>.pc 文件是安装软件后生成的一个文件，其中包含该软件的信息；该 .pc 文件一般统一存放于名为 pkgconfig 的目录下。命令 pkg-config 按照系统设置路径（CentOS6 x64_86 系统默认为 /usr/lib64/pkgconfig/）的先后顺序来搜索 .pc 文件，从而或得到软件的版本号、头文件和库文件位置等信息。</p>
<p>因此当把软件包安装到默认位置 /usr/local/ 位置后，则其 .pc 文件会生成到 /usr/local/lib/pkgconfig/ 或 /usr/local/lib64/pkgconfig/ 文件夹下，此时只需要将相应的 .pc 文件复制到 /usr/lib64/pkgconfig/ 中即可。</p>
<p>或者，在安装软件包的时候，使用 “./configure –prefix=/usr/” 将软件直接安装到 /usr/ 目录下。这时有可能会将 .pc 文件生成到 /usr/lib/pkgconfig/ 中，依然导致问题出现，只需将该目录中的 .pc 文件复制到 /usr/lib64/pkgconfig/ 中即可。</p>
<p>或者，设置环境变量PKG_CONFIG_PATH，加入含有相应的.pc文件的路径。注意要使用命令 ‘export’ 。比如:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig/:$PKG_CONFIG_PATH</span><br></pre></td></tr></table></figure>
<p>使用以上3种方法，能解决软件版本不符或系统找不到依赖的软件包的问题。</p>
<ol start="2">
<li>设置环境变量 LD_LIBRARY_PATH</li>
</ol>
<p>Linux 系统默认把 /lib 和 /usr/lib 两个目录作为默认的库搜索路径。而修改库搜索路径有两个方法：</p>
<p>(1) 在 /etc/ld.so.conf 文件中添加库的搜索路径。</p>
<p>(2) 在环境变量 LD_LIBRARY_PATH 中指明库的搜索路径。</p>
<p>将库的路径设置好后，需要使用 /sbin/ldconfig 命令来将这些搜索路径下的共享库文件集中在一起而生成 /etc/ld.so.cache。该 cache 用于快速定位共享库。</p>
<p>特别是新安装了软件后，即使其库文件明明就在库搜索路径下，但是编译的时候依然报错，提示缺少该库。则需要使用 ldconfig 命令来更新缓存。</p>
<p>更详细信息，可参考：<a target="_blank" rel="noopener" href="http://www.mike.org.cn/articles/description-configure-pkg-config-pkg_config_path-of-the-relations-between/">简述configure、pkg-config、pkg_config_path三者的关系</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/03/Bioinfo-jcvi-installation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/03/Bioinfo-jcvi-installation/" class="post-title-link" itemprop="url">jcvi 安装与使用，报错以及解决办法</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-06-03 17:36:50" itemprop="dateCreated datePublished" datetime="2021-06-03T17:36:50+08:00">2021-06-03</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-06-07 23:04:42" itemprop="dateModified" datetime="2021-06-07T23:04:42+08:00">2021-06-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>jcvi 是 Haibao Tang 开发的基因组组装注释与比较基因组分析工具，比如可以用于可视化 synteny blocks。原理是依据物种之间基因编码序列相似性比较，如果若干个连续的基因在两个基因组之间都有出现，那这若干个基因就是一个 synteny block，可能是从祖先染色体上继承下来的，物种之间的亲缘关系越近，synteny blocks 数目越多。</p>
<p>之前<a target="_blank" rel="noopener" href="https://genehub.wordpress.com/2019/07/05/mcscanx-%e6%b5%8b%e8%af%95/">测试过 MCScanX</a>，其实 jcvi 和 MCScanX 就是一回事，是同一个作者开发的，jcvi 是升级版，操作更加便利，生成的图片更加精美。回去 <a target="_blank" rel="noopener" href="https://github.com/wyp1125/MCScanX#installation">MCScanX</a> 看了一下开发时间，最后更新大约是 2014 年，但是里面有的程序还是很实用的，比如可以计算共线性区块之间的基因对的 Ka/Ks value（add_kaks_to_synteny.pl）.</p>
<h3 id="安装-jcvi"><a class="markdownIt-Anchor" href="#安装-jcvi"></a> 安装 jcvi</h3>
<p>jcvi 软件推荐使用 pip 安装，也可以选择使用 conda 安装。我是 anaconda3 下面建立了 python 2.7 和 python 3.7 两个环境，分别进行了测试，最后发现 python2.7 环境下 jcvi 可以正常运行。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">source activate python27</span><br><span class="line">conda install -c bioconda jcvi</span><br><span class="line">conda install -c bioconda last</span><br><span class="line">conda install scipy</span><br><span class="line"># 但是运行序列比对时候报错！--&gt; 经过查询是 lastal 的问题，也就是别用 conda 安装的 last，自己去 last 网站下载源码，安装到用户目录下，然后将其bin路径写入环境变量bashrc文件，然后就可以解决比对报错的问题。</span><br></pre></td></tr></table></figure>
<h3 id="安装-lastal"><a class="markdownIt-Anchor" href="#安装-lastal"></a> 安装 lastal</h3>
<p>手动安装 last 软件，并将其路径写入环境变量。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">git clone https://gitlab.com/mcfrith/last.git</span><br><span class="line">cd last</span><br><span class="line">mdke</span><br><span class="line"># then we can see lastal in ./last/bin/</span><br><span class="line"># put /PATH/TO/last/bin into ~/.bashrc</span><br><span class="line">which lastal</span><br><span class="line">(python37) [fenglei@localhost cds]$ which lastal</span><br><span class="line">~/local/app/anaconda3/envs/python37/bin/lastal</span><br><span class="line">(python37) [fenglei@localhost cds]$ source ~/.bashrc</span><br><span class="line">(base) [fenglei@localhost cds]$ which lastal</span><br><span class="line">~/local/app/last/bin/lastal</span><br></pre></td></tr></table></figure>
<p>下面开始做一个数据测试，参考：<a target="_blank" rel="noopener" href="https://github.com/tanghaibao/jcvi/wiki/MCscan-(Python-version)">https://github.com/tanghaibao/jcvi/wiki/MCscan-(Python-version)</a></p>
<h3 id="基因坐标与基因序列数据准备"><a class="markdownIt-Anchor" href="#基因坐标与基因序列数据准备"></a> 基因坐标与基因序列数据准备</h3>
<p>原始数据：（1）gff3 格式的基因坐标；（2）fasta 格式的基因 CDS 序列。使用 jcvi 内置程序对这两个文件做格式整理。以大豆（Glycine max）和野生豆（Glycine soja）为例，这一步需要得到四个文件：gma.bed gma.cds gso.bed gso.cds</p>
<p>如下图所示，要注意同一个物种的 bed 文件和 cds 文件的基因 ID 要相互匹配。由于基因 ID 通常带有“.1”或“.2p”这样的后缀，需要注意，必要的时候自己写脚本处理数据。</p>
<p><img src="https://i.imgur.com/RYAdtK3.jpg" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line">### BED FORMAT FILES FOR GENES</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=transcript_id Gmax_508_Wm82.a4.v1.gene.gff3.gz &gt; gma.bed</span><br><span class="line">11:12:31 [base] Load file `Gmax_508_Wm82.a4.v1.gene.gff3.gz`</span><br><span class="line">11:13:30 [gff] Extracted 86256 features (type=mRNA id=transcript_id)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Gmax_508_Wm82.a4.v1.gene.gff3.gz &gt; gma.bed2</span><br><span class="line">11:20:14 [base] Load file `Gmax_508_Wm82.a4.v1.gene.gff3.gz`</span><br><span class="line">11:21:13 [gff] Extracted 86256 features (type=mRNA id=Name)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=transcript_id Gsoja.W05.gene.gff.gz &gt; gso.bed</span><br><span class="line">11:14:15 [base] Load file `Gsoja.W05.gene.gff.gz`</span><br><span class="line">11:15:31 [gff] Extracted 89477 features (type=mRNA id=transcript_id)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Gsoja.W05.gene.gff.gz &gt; gso.bed2             </span><br><span class="line">11:22:16 [base] Load file `Gsoja.W05.gene.gff.gz`</span><br><span class="line">11:23:32 [gff] Extracted 89477 features (type=mRNA id=Name)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=ID Gsoja.W05.gene.gff.gz &gt; gso.bed3    </span><br><span class="line">11:27:42 [base] Load file `Gsoja.W05.gene.gff.gz`</span><br><span class="line">11:28:58 [gff] Extracted 89477 features (type=mRNA id=ID)</span><br><span class="line"></span><br><span class="line">### REMOVE DUPLICATIONS</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.bed uniq gma.bed2 </span><br><span class="line">11:35:34 [base] Load file `gma.bed2`</span><br><span class="line">11:35:37 [bed] Imported: 86256, Exported: 51013</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.bed uniq gso.bed3</span><br><span class="line">11:37:25 [base] Load file `gso.bed3`</span><br><span class="line">11:37:29 [bed] Imported: 89477, Exported: 55103</span><br><span class="line"></span><br><span class="line">### CDS FORMAT FILE FOR UNIQ GENES</span><br><span class="line"></span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gma.uniq.bed ) Gmax_508_Wm82.a4.v1.cds.fa.gz |  seqkit seq -i &gt; gma.cds</span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gso.uniq.bed ) Gsoja.W05.gene.cds.fa.gz |  seqkit seq -i &gt; gso.cds</span><br><span class="line"></span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gma.uniq.bed ) Gmax_508_Wm82.a4.v1.protein.fa |  seqkit seq -i &gt; gma.pep  # ERROR due to id isuues</span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gso.uniq.bed ) Gsoja.W05.gene.protein.fa.gz |  seqkit seq -i &gt; gso.pep</span><br><span class="line"></span><br><span class="line"># TROUBLESHOOTING</span><br><span class="line">awk &#x27;&#123;if($0~/&gt;/)&#123;gsub(/\.p/,&quot;&quot;,$1)&#125;; &#123;print&#125;&#125;&#x27; Gmax_508_Wm82.a4.v1.protein.fa &gt; Gmax_508_Wm82.a4.v1.protein.new.fa</span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gma.uniq.bed ) Gmax_508_Wm82.a4.v1.protein.new.fa |  seqkit seq -i &gt; gma.pep</span><br></pre></td></tr></table></figure>
<h3 id="apairwise_synteny_searchsh"><a class="markdownIt-Anchor" href="#apairwise_synteny_searchsh"></a> a.pairwise_synteny_search.sh</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/local/app/anaconda3/envs/python27/bin/python -m jcvi.compara.catalog ortholog gma gso --no_strip_names</span><br></pre></td></tr></table></figure>
<p>该命令直接对物种间基因序列进行比较，将生成一个 gma.gso.pdf 文件，打开如下。</p>
<p><img src="https://i.imgur.com/Ly4FSpT.png" alt="" /></p>
<p>同时生成一堆文件：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">-rw-rw-r-- 1 fenglei fenglei    139507 Jun  3 12:09 gma.gso.pdf</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   5333469 Jun  3 12:09 gma.gso.lifted.anchors</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   2931001 Jun  3 12:08 gma.gso.anchors</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   5953786 Jun  3 12:08 gma.gso.last.filtered</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  97488440 Jun  3 12:08 gma.gso.last</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  58121140 Jun  3 12:08 gso.bck</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei       515 Jun  3 12:08 gso.prj</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 264000792 Jun  3 12:08 gso.suf</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   1101303 Jun  3 12:07 gso.des</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    220416 Jun  3 12:07 gso.sds</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    220416 Jun  3 12:07 gso.ssp</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  66055302 Jun  3 12:07 gso.tis</span><br></pre></td></tr></table></figure>
<p>The <code>.last</code> file is raw LAST output, <code>.last.filtered</code> is filtered LAST output, <code>.anchors</code> is the seed synteny blocks (high quality), <code>.lifted.anchors</code> recruits additional anchors to form the final synteny blocks.</p>
<p>查看了一下 .lifted.anchors 与 .anchors 的区别，如下看到两个文件都有 551 个 synteny blocks，只是 blocks 内部的基因配对数目不同，应该就是后者有更严格的筛选条件。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.lifted.anchors | grep -v &#x27;###&#x27; | wc -l</span><br><span class="line">124133</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.lifted.anchors | grep &#x27;###&#x27; | wc -l   </span><br><span class="line">551</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.anchors | grep -v &#x27;###&#x27; | wc -l        </span><br><span class="line">68549</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.anchors | grep &#x27;###&#x27; | wc -l   </span><br><span class="line">551</span><br></pre></td></tr></table></figure>
<h3 id="bdepth_visualiztionsh"><a class="markdownIt-Anchor" href="#bdepth_visualiztionsh"></a> b.depth_visualiztion.sh</h3>
<p>We could also quick test if the synteny pattern is indeed 1:1, by running:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.compara.synteny depth --histogram gma.gso.anchors</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/BB2UsVk.jpg" alt="" /></p>
<h3 id="canchorssimple_file_generationsh"><a class="markdownIt-Anchor" href="#canchorssimple_file_generationsh"></a> c.anchors.simple_file_generation.sh</h3>
<p>This command generates a .simple file which is a more succinct form than the .anchors file.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.compara.synteny screen --minspan=30 --simple gma.gso.anchors gma.gso.anchors.new</span><br></pre></td></tr></table></figure>
<p>将 <code>gma.gso.anchors</code> 文件转化成 <code>gma.gso.anchors.simple</code> 文件。有什么不同？见下面的文本框。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">(python27) [fenglei@localhost gma-gso]$ head gma.gso.anchors</span><br><span class="line">###</span><br><span class="line">Glyma.01G000100.1       Glysoja.01G000001.1     496</span><br><span class="line">Glyma.01G000322.1       Glysoja.01G000002.1     775</span><br><span class="line">Glyma.01G000400.1       Glysoja.01G000003.1     2600</span><br><span class="line">Glyma.01G001000.2       Glysoja.01G000010.2     984</span><br><span class="line">Glyma.01G001100.3       Glysoja.01G000011.2     1490</span><br><span class="line">Glyma.01G001300.3       Glysoja.01G000013.2     3560</span><br><span class="line">Glyma.01G001350.1       Glysoja.01G000014.1     381</span><br><span class="line">Glyma.01G001500.1       Glysoja.01G000015.3     2880</span><br><span class="line">Glyma.01G001700.1       Glysoja.01G000017.1     1650</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ head gma.gso.anchors.simple</span><br><span class="line">Glyma.01G000100.1       Glyma.01G245700.6       Glysoja.01G000001.1     Glysoja.01G002529.1     2354    +</span><br><span class="line">Glyma.01G023500.1       Glyma.01G044800.1       Glysoja.02G002717.1     Glysoja.02G002929.1     209     -</span><br><span class="line">Glyma.01G045100.2       Glyma.01G071700.1       Glysoja.02G003545.1     Glysoja.02G003757.1     210     +</span><br><span class="line">Glyma.01G083700.1       Glyma.01G090200.1       Glysoja.02G003470.1     Glysoja.02G003510.1     46      +</span><br><span class="line">Glyma.01G098400.6       Glyma.01G106600.2       Glysoja.03G006294.1     Glysoja.03G006391.1     76      +</span><br><span class="line">Glyma.01G125600.2       Glyma.01G145000.1       Glysoja.03G005845.1     Glysoja.03G006096.1     198     -</span><br><span class="line">Glyma.01G118000.1       Glyma.01G122800.1       Glysoja.03G006181.1     Glysoja.03G006214.2     37      +</span><br><span class="line">Glyma.01G107000.1       Glyma.01G113400.1       Glysoja.03G006412.2     Glysoja.03G006472.1     56      +</span><br><span class="line">Glyma.01G003900.1       Glyma.01G008500.1       Glysoja.05G013057.1     Glysoja.05G013097.1     40      +</span><br><span class="line">Glyma.01G001900.2       Glyma.01G007300.1       Glysoja.07G017673.1     Glysoja.07G017734.1     56      -</span><br></pre></td></tr></table></figure>
<h3 id="dmacrosynteny_visualizationsh"><a class="markdownIt-Anchor" href="#dmacrosynteny_visualizationsh"></a> d.Macrosynteny_visualization.sh</h3>
<p>准备一个 <code>seqids</code> 文件，定义哪些染色体需要展示出来。文件内容是：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Gm01,Gm02,Gm03,Gm04,Gm05,Gm06,Gm07,Gm08,Gm09,Gm10,Gm11,Gm12,Gm13,Gm14,Gm15,Gm16,Gm17,Gm18,Gm19,Gm20</span><br><span class="line">Chr01,Chr02,Chr03,Chr04,Chr05,Chr06,Chr07,Chr08,Chr09,Chr10,Chr11,Chr12,Chr13,Chr14,Chr15,Chr16,Chr17,Chr18,Chr19,Chr20</span><br></pre></td></tr></table></figure>
<p>可以用 shell 代码生成上面的文件，我的命令如下。注意 seqids 内容上下行代表两个物种，跟下文出现的 layout 文件要顺序对应。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">for i in $(seq 1 20); do printf &quot;Gm&quot;%02d&quot;,&quot; $i ; done; echo </span><br><span class="line">for i in $(seq 1 20); do printf &quot;Chr&quot;%02d&quot;,&quot; $i ; done; echo</span><br></pre></td></tr></table></figure>
<p>再准备一个 <code>layout</code> 文件。定义染色体出现的位置，角度。文件内容是：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># y, xstart, xend, rotation, color, label, va,  bed</span><br><span class="line"> .6,     .1,    .8,       0,      , gma, top, gma.bed</span><br><span class="line"> .4,     .1,    .8,       0,      , gso, top, gso.bed</span><br><span class="line"># edges</span><br><span class="line">e, 0, 1, gma.gso.anchors.simple</span><br></pre></td></tr></table></figure>
<h3 id="ekaryotypesh"><a class="markdownIt-Anchor" href="#ekaryotypesh"></a> <a target="_blank" rel="noopener" href="http://e.karyotype.sh">e.karyotype.sh</a></h3>
<p>画图文件准备完毕，运行画图命令：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.graphics.karyotype seqids layout</span><br></pre></td></tr></table></figure>
<p>将生成一个 karyotype.pdf 文件，打开如下。</p>
<p><img src="https://i.imgur.com/Bugyz9W.jpeg" alt="" /></p>
<p>以上就是两个物种间的共线性比较分析。</p>
<h3 id="同一个物种内部的-synteny-呢"><a class="markdownIt-Anchor" href="#同一个物种内部的-synteny-呢"></a> 同一个物种内部的 synteny 呢？</h3>
<p>由于大豆在近期经历了 WGD 事件，导致很多基因都有双拷贝，所以内部就有 synteny，我做了测试分析，用与上文相同的代码，得到图片如下。看来 jcvi 程序会自动滤掉基因的自我比较，而是做 paralog 的比较。</p>
<p><img src="https://i.imgur.com/9QhsL89.jpg" alt="" /></p>
<h3 id="其他案例"><a class="markdownIt-Anchor" href="#其他案例"></a> 其他案例</h3>
<p>软件的文档里面还列出了三个物种的比较、对某些 synteny 区段进行高亮、只展示特定染色体区间的方法。下面直接摘抄原作者的图片，留作参考。</p>
<p><img src="https://camo.githubusercontent.com/833e373b369f1c221132d729e1cf0128a8cd638fc04c90782ad7d5e9d30867db/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f39766c337973336e6476696d6734632f67726170652d70656163682d636163616f2e706e673f7261773d31" alt="" /></p>
<p><img src="https://camo.githubusercontent.com/5b300c541bd65669e40cf15f7e496fd0a63a0a47ecb4c80a8f5641e9253c9981/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f736f3274643065716b6c626b6c6e6c2f67726170652e70656163682e6b6172796f747970652d677265656e2e706e673f7261773d31" alt="" /></p>
<p><img src="https://camo.githubusercontent.com/7a878ab0e1aa1221d9c4af36a4d0f131eb43f8ddb10caea79e53366d6797cd92/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f696d646f7539756b6e74743474616d2f67726170652d70656163682d636163616f2d626c6f636b732e706e673f7261773d31" alt="" /></p>
<h3 id="其他笔记"><a class="markdownIt-Anchor" href="#其他笔记"></a> 其他笔记</h3>
<p>下面是软件安装调试的一些笔记。python3.7 环境我也测试了，可以安装 jcvi，而且我发现 python3.7 的 jcvi 与 python2.7 下面的 jcvi 的屏幕输出不同。但是 pyton2.7 jcvi 运行过程中有些步骤报错（[Errno 13] Permission denied: ‘dvipng’）导致缺失一些数据结果，所以我放弃了python3.7 jcvi，转而用 python2.7 jcvi 来做分析。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">conda install scipy</span><br><span class="line">conda install -c bioconda last</span><br><span class="line">conda install -c bioconda jcvi  # 这一条就报错！</span><br><span class="line"># 于是从github手动下载软件尝试</span><br><span class="line">cd ~/local/app  # or any directory of your choice</span><br><span class="line">git clone git://github.com/tanghaibao/jcvi.git</span><br><span class="line">pip install -e .</span><br><span class="line"># 安装成功。能否运行数据比对？</span><br><span class="line"># 依然不能，经查是 lastal 的问题！！！</span><br><span class="line">## lastal: I was installed here with multi-threading disabled  # 运行到lastal就报错，根本没有比对结果输出。</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">conda create -n jcvi python=2.7</span><br><span class="line">source activate jcvi</span><br><span class="line"># 安装前需要确认存在bioconda和r channel，安装前先更新conda</span><br><span class="line"># conda --add-channel r</span><br><span class="line"># conda install jcvi last</span><br><span class="line">conda install -c bioconda last</span><br></pre></td></tr></table></figure>
<p>后续还可能会由于缺少 scipy 在绘图时报错 ImportError: No module named scipy.spatial ，因此可提<br />
前安装</p>
<p><strong>安装scipy</strong><br />
#conda-forge scipy-1.2.0<br />
conda install scipy</p>
<p><strong>问题一</strong></p>
<p>File “/biosoft/python/Python-v2.7.11/lib/python2.7/site-packages/subprocess32.py”, line 1393, in _execute_child<br />
    raise child_exception_type(errno_num, err_msg)<br />
OSError: [Errno 2] No such file or directory: ‘kpsewhich’</p>
<p>解决办法：linux中缺少kpsewhich命令导致的，于是安装该命令，CentOS系统方法如下：</p>
<p>sudo yum install tetex-latex tetex-doc tetex-fonts<br />
sudo yum install texlive tex-live-latex</p>
<p>安装好后再次运行，顺利通过。</p>
<p>我的系统是redhat，没有安yum，所以下载了textlive的源码编译，但是一直不能通过，后来重装了一个centOS系统才可以用。</p>
<p><strong>问题二</strong></p>
<p>原来在用户目录下安装了GCC-6.5.0，可以正常使用jcvi的；但是昨天升级到了GCC-9.2.0之后，今天用jcvi 画synteny图的时候遇到下面的问题。</p>
<p>ImportError: /home/fenglei/local/lib64/libstdc++.so.6: undefined symbol: libiconv</p>
<p>将libstdc++.so.6这个软连接重新指向旧版GCC对应的libstdc++.so.6.0.22，然后jcvi就没有报错了。</p>
<p><img src="https://genehub.files.wordpress.com/2019/09/e5beaee4bfa1e688aae59bbe_20190906173207.png" alt="微信截图_20190906173207" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/" class="post-title-link" itemprop="url">Bioinfo | Installation of OrthoMCL-pipeline</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-06-02 13:36:29" itemprop="dateCreated datePublished" datetime="2021-06-02T13:36:29+08:00">2021-06-02</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-14 18:33:24" itemprop="dateModified" datetime="2022-07-14T18:33:24+08:00">2022-07-14</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>OrthoMCL (<a target="_blank" rel="noopener" href="http://orthomcl.org/orthomcl/">http://orthomcl.org/orthomcl/</a>) 是现在用的最多的一款来找直系同源基因（Orthologs）以及旁系同源基因 (Paralog) 的软件。OrthoMCL 原始软件安装比较繁琐，根据官网的教程至少得十多步才能完成整个运行流程，包括Mysql数据库配置、修改OrthoMCL配置文件、转换序列格式、过滤、比对、解析结果和聚类等步骤，特别麻烦。有更方便的途径来完成安装：OrthoMCL Pipeline (<a target="_blank" rel="noopener" href="https://github.com/apetkau/orthomcl-pipeline">https://github.com/apetkau/orthomcl-pipeline</a>)。Pipeline安装有点复杂，但是安装完成后，使用就方便了。</p>
<p>两年多前使用过 OrthoMCL-pipeline 流程来寻找一个物种内部的同源基因（paralogs）和物种间的同源基因（othologs）。物种内部的同源基因之间的碱基差别可能是若干年演化的结果，比如染色体倍增或基因扩张，导致出现同源基因，随后由于演化导致某些突变，尤其是同义替换（Ks 或称为 synonymous）不会引起氨基酸序列变化，常作为分子钟，也比如基因组分析种常见的 4dtv（四倍简并位点），也是用需要依靠同源基因的序列来做分析。</p>
<blockquote>
<p>2021.06.25 备注：OrthoMCL-pipeline 安装测试失败。改用 OrthoFinder 则十分顺利，只需要切换到 python37 环境，用 conda 一键安装 OrthoFinder，然后一句命令即可运行物种之间的同源基因比较与归类。下面的代码示例，orthofinder_data 目录下存放蛋白序列，比如含有若干序列比较的物种的蛋白序列集合:ath.fasta, gma.fasta等等，每个物种可以用一个序列文件。</p>
</blockquote>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">orthofinder -t 6 -f orthofinder_data -S diamond</span><br></pre></td></tr></table></figure>
<p>有哪些文章的示例？可以便于记忆与理解。</p>
<p><img src="https://i.imgur.com/aDN8Pcn.jpg" alt="Fig. 1" /></p>
<p>上图是 2011 年发表于 Nature 期刊的土豆基因组（<a target="_blank" rel="noopener" href="https://www.nature.com/articles/nature10158">https://www.nature.com/articles/nature10158</a>），使用 OrthoMCL 分析 12 个植物之间的直系同源基因和旁系同源基因，统计数目并制作了条形图。</p>
<p><img src="https://i.imgur.com/XpDMtM8.jpg" alt="Fig. 2" /></p>
<p>上图是 2018 年发表于 PBJ 期刊的西葫芦基因组（<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5978595/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5978595/</a>）。使用 OrthoMCL 分析五个植物的同源基因，并且基于同源基因信息计算四倍简并位点（4tdv），推算基因组是否加倍。</p>
<h3 id="安装-orthomlc-pipeline"><a class="markdownIt-Anchor" href="#安装-orthomlc-pipeline"></a> 安装 OrthoMLC-pipeline</h3>
<p>下面是安装OrthoMCL的记录，其实可以不用看了。<br />
先下载 <strong>OrthoMCL-pipeline</strong>。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/apetkau/orthomcl-pipeline.git</span><br></pre></td></tr></table></figure>
<p>再安装 <strong>Perl 以及若干相关模块</strong>。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cpanm BioPerl DBD::mysql DBI Parallel::ForkManager YAML::Tiny Set::Scalar Text::Table Exception::Class Test::Most Test::Warn Test::Exception Test::Deep Moose SVG Algorithm::Combinatorics</span><br></pre></td></tr></table></figure>
<p>实际上在安装 Perl 模块的过程中遇到问题，主要是 BioPerl 安装失败，报错信息：<code>Installing the dependencies failed: Module 'Alien::Libxml2' is not installed</code>。并且我发现不能通过 cpanm 来安装 Alien::Libxml2。经过检索，在一个论坛（<a target="_blank" rel="noopener" href="https://github.com/PerlAlien/Alien-Libxml2/issues/7">https://github.com/PerlAlien/Alien-Libxml2/issues/7</a>）找到方法。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">$ which perl</span><br><span class="line"> ~/local/bin/perl</span><br><span class="line"></span><br><span class="line">$ perl --version</span><br><span class="line"> This is perl 5, version 30, subversion 3 (v5.30.3) built for x86_64-linux</span><br><span class="line"></span><br><span class="line">$ curl -O https://cpan.metacpan.org/authors/id/P/PL/PLICEASE/Alien-Build-1.71_01.tar.gz</span><br><span class="line">  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current</span><br><span class="line">                                 Dload  Upload   Total   Spent    Left  Speed</span><br><span class="line">100  257k  100  257k    0     0  1764k      0 --:--:-- --:--:-- --:--:-- 1775k</span><br><span class="line">$ tar xf Alien-Build-1.71_01.tar.gz </span><br><span class="line">$ cd Alien-Build-1.71_01</span><br><span class="line">$ perl Makefile.PL </span><br><span class="line">Checking if your kit is complete...</span><br><span class="line">Looks good</span><br><span class="line">Generating a Unix-style Makefile</span><br><span class="line">Writing Makefile for Alien::Build</span><br><span class="line">Writing MYMETA.yml and MYMETA.json</span><br><span class="line">...</span><br><span class="line">$ make</span><br><span class="line">$ make install</span><br><span class="line"># 最后这样就安装成功了</span><br><span class="line"># Appending installation info to /home/fenglei/local/lib/perl5/5.30.3/x86_64-linux/perllocal.pod</span><br></pre></td></tr></table></figure>
<p><strong><a target="_blank" rel="noopener" href="http://orthomcl.org/common/downloads/software/v2.0/">OrthoMCL</a></strong> 或 OrthoMCL Custom （能够自定义序列识别符的修改版本）。OrhtoMCL 下载并解压之后，bin目录下的Perl程序首行是 <code>#!/usr/bin/perl</code>，我将其全部改为用户的 perl 了，因为我自己安装的perl位置是 <code>/home/fenglei/local/bin/perl</code>。</p>
<p><strong><a target="_blank" rel="noopener" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">BLAST</a></strong> (blastall, formatdb) 注意不是NCBI-blast+, 推荐使用 2.2.26 版本。由于我之前下载过 interproscan，在 interproscan 安装目录下游 blastall 和 formatdb 等程序，直接将其路径加入环境变量即可！</p>
<p><strong><a target="_blank" rel="noopener" href="https://www.micans.org/mcl/src/">MCL</a></strong> The MCL algorithm is short for the Markov Cluster Algorithm, a fast and scalable unsupervised cluster algorithm for graphs (also known as networks) based on simulation of (stochastic) flow in graphs.</p>
<p>依赖环境解决后能够顺利运行设置脚本:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">$ perl scripts/orthomcl-pipeline-setup.pl</span><br><span class="line">Checking for Software dependencies...</span><br><span class="line">Checking for OthoMCL ... OK</span><br><span class="line">Checking for formatdb ... OK</span><br><span class="line">Checking for blastall ... OK</span><br><span class="line">Checking for mcl ... OK</span><br><span class="line">Wrote new configuration to orthomcl-pipeline/scripts/../etc/orthomcl-pipeline.conf</span><br><span class="line">Wrote executable file to orthomcl-pipeline/scripts/../bin/orthomcl-pipeline</span><br><span class="line">Please add directory orthomcl-pipeline/scripts/../bin to PATH</span><br></pre></td></tr></table></figure>
<p>然而，我重复以前的流程失败，还是卡在 BioPerl 安装失败。看来问题跟我使用 Anaconda3 有关，下面的错误提示信息种出现的 make 是来自 python37 环境。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">make: *** [Makefile:3290: test_dynamic] Error 255</span><br><span class="line">  CJFIELDS/BioPerl-1.7.8.tar.gz</span><br><span class="line">  /home/fenglei/local/app/anaconda3/envs/python37/bin/make test -- NOT OK</span><br><span class="line">//hint// to see the cpan-testers results for installing this module, try:</span><br><span class="line">  reports CJFIELDS/BioPerl-1.7.8.tar.gz</span><br><span class="line">Failed during this command:</span><br><span class="line"> (optional) CJFIELDS/BioPerl-1.7.8.tar.gz     : make_test NO</span><br></pre></td></tr></table></figure>
<h3 id="20210615-再次尝试安装测试"><a class="markdownIt-Anchor" href="#20210615-再次尝试安装测试"></a> 20210615 再次尝试安装测试</h3>
<p>由于系统自带了 Perl（/usr/bin/perl），我安装的Anaconda3下面的不同环境也分别有 Perl（<sub>/local/app/anaconda3/envs/python37/bin/perl、</sub>/local/app/anaconda3/envs/python27/bin/perl），另外自己又在用户目录下安装了 Perl（~/local/bin/perl），多个版本导致冲突。</p>
<p>软件测试：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">PERL5LIB=/home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0  </span><br><span class="line"># 上面其实应该设为 /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/</span><br><span class="line"># 此时直接用 cpan -i 或者 perl -MCPAN -s shell 均无法安装 perl 模块，于是尝试 conda，conda基本可以安装这些模块</span><br><span class="line"> conda install -c bioconda perl-Moo</span><br><span class="line"> conda install -c bioconda perl-DBD::mysql # ERROR: Wrong command</span><br><span class="line"> conda install -c bioconda perl-DBD-mysql</span><br><span class="line"> conda install -c bioconda perl-Parallel-ForkManager</span><br><span class="line"> conda install -c bioconda perl-YAML-Tiny</span><br><span class="line"> conda install -c bioconda perl-Set-Scalar</span><br><span class="line"> conda install -c bioconda perl-Text-Table  # ERROR: PackagesNotFoundError</span><br><span class="line"> conda install -c bioconda perl-Exception-Class</span><br><span class="line"> conda install -c bioconda perl-Test-Most</span><br><span class="line"> conda install -c bioconda perl-Test-Warn</span><br><span class="line"> conda install -c bioconda perl-Test-Exception</span><br><span class="line"> conda install -c bioconda perl-Test-Deep</span><br><span class="line"> conda install -c bioconda perl-Moose</span><br><span class="line"> conda install -c bioconda perl-SVG</span><br><span class="line"> conda install -c bioconda perl-Algorithm-Combinatorics  # ERROR: PackagesNotFoundError</span><br></pre></td></tr></table></figure>
<p>conda 安装perl模块有两个安不上，所以手动尝试</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">wget https://cpan.metacpan.org/authors/id/S/SH/SHLOMIF/Text-Aligner-0.16.tar.gz</span><br><span class="line">perl Makefile.PL &amp;&amp; make &amp;&amp; make install</span><br><span class="line">## Installing /home/fenglei/perl5/man/man3/Text::Aligner.3</span><br><span class="line">## Appending installation info to /home/fenglei/perl5/lib/perl5/x86_64-linux-thread-multi/perllocal.pod</span><br><span class="line"></span><br><span class="line">conda install -c bioconda  perl-Module-Build</span><br><span class="line"></span><br><span class="line">wget https://cpan.metacpan.org/authors/id/S/SH/SHLOMIF/Text-Table-1.134.tar.gz</span><br><span class="line">perl Makefile.PL &amp;&amp; make &amp;&amp; make install</span><br><span class="line">## Installing /home/fenglei/perl5/man/man3/Text::Table.3</span><br><span class="line">## Appending installation info to /home/fenglei/perl5/lib/perl5/x86_64-linux-thread-multi/perllocal.pod</span><br><span class="line"></span><br><span class="line">## 安装上述模块之后， perl lib 路径下根本没有出现pm文件，我手动拷贝过去，为了保险，拷贝到两个位置：</span><br><span class="line">cp Aligner.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Text/</span><br><span class="line">cp Aligner.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/Text/</span><br><span class="line">cp Table.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Text/</span><br><span class="line">cp Table.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/Text/</span><br><span class="line">PERL5LIB=/home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi/auto/:$PERL5LIB</span><br><span class="line"></span><br><span class="line">wget https://cpan.metacpan.org/authors/id/F/FX/FXN/Algorithm-Combinatorics-0.27.tar.gz</span><br><span class="line">## make 失败</span><br><span class="line">## 报错：unable to execute &#x27;x86_64-conda_cos6-linux-gnu-gcc&#x27;: No such file or directory  </span><br></pre></td></tr></table></figure>
<p>这次卡在了 Algorithm-Combinatorics。</p>
<p>在base环境下，自己安装的perl可以编译Algorithm-Combinatorics。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">source activate base</span><br><span class="line">which perl</span><br><span class="line"># ~/local/bin/perl</span><br><span class="line">PERL5LIB=/home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0  </span><br><span class="line">cp Combinatorics.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/</span><br><span class="line">cp Combinatorics.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Algorithm/</span><br><span class="line">cp Combinatorics.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/Algorithm/</span><br></pre></td></tr></table></figure>
<p>更改前后，都有报错，看来直接复制Combinatorics.pm也不行。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost orthomcl-pipeline]$ perl ./scripts/orthomcl-pipeline-setup.pl</span><br><span class="line">Can&#x27;t locate Algorithm/Combinatorics.pm in @INC (you may need to install the Algorithm::Combinatorics module) (@INC contains: /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi/auto/ /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0 /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2 /home/fenglei/local/app/anaconda3/envs/python37/lib/5.26.2/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/5.26.2 .) at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br><span class="line">BEGIN failed--compilation aborted at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br><span class="line"></span><br><span class="line">(python37) [fenglei@localhost orthomcl-pipeline]$ perl ./scripts/orthomcl-pipeline-setup.pl</span><br><span class="line">Can&#x27;t locate loadable object for module Algorithm::Combinatorics in @INC (@INC contains: /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi/auto/ /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0 /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2 /home/fenglei/local/app/anaconda3/envs/python37/lib/5.26.2/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/5.26.2 .) at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br><span class="line">Compilation failed in require at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br><span class="line">BEGIN failed--compilation aborted at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br></pre></td></tr></table></figure>
<h3 id="orthofinder-测试"><a class="markdownIt-Anchor" href="#orthofinder-测试"></a> orthofinder 测试</h3>
<p>安装可以很方便，就是conda：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">source  activate python37</span><br><span class="line">conda install -c bioconda -y orthofinder</span><br><span class="line"></span><br><span class="line">(python37) [fenglei@localhost Results_Apr24]$ orthofinder -help</span><br><span class="line"></span><br><span class="line">OrthoFinder version 2.5.2 Copyright (C) 2014 David Emms</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/" class="post-title-link" itemprop="url">Bioinfo | Using Circos for genome visualization</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-05-28 11:40:21" itemprop="dateCreated datePublished" datetime="2021-05-28T11:40:21+08:00">2021-05-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-06-03 21:11:14" itemprop="dateModified" datetime="2021-06-03T21:11:14+08:00">2021-06-03</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>circos 是以环形展示基因组序列和注释信息的工具，我首先按照 circos 官网的指引进行手动安装，即手动逐个安装 Perl 模块，最后再安装 circos 文件，但是卡在 GD 库，没法成功运行 circos。但是换了一台服务器，进入 anaconda3 下面的 python3.7 环境，直接用 conda 不到一分钟就成功安装 circos。</p>
<p>First, understand the <a target="_blank" rel="noopener" href="http://circos.ca/software/requirements">requirements</a>. If you are a Windows user, read about pertinent differences between UNIX and Windows.</p>
<p><a target="_blank" rel="noopener" href="http://circos.ca/software/download">Download</a> and <a target="_blank" rel="noopener" href="http://circos.ca/software/install">install</a> the software, then follow the <a target="_blank" rel="noopener" href="http://circos.ca/documentation/tutorials">tutorials</a>.</p>
<h3 id="installing-perl-modules"><a class="markdownIt-Anchor" href="#installing-perl-modules"></a> Installing Perl modules</h3>
<p>安装 Circos 的第一步，安装 Perl 以及相关的若干 Perl 模块。我 Linux 系统自带的 Perl，进入命令行界面时报错（<code>perl -MCPAN -e shell</code> Can’t locate <a target="_blank" rel="noopener" href="http://CPAN.pm">CPAN.pm</a> in @INC），所以我先尝试自己安装的 Perl，即 anaconda3/python37 里面的 Perl v5.26.2，但是在安装 Params::Validate 包时候无法通过。我再次尝试系统默认的 Perl v5.16.3，先安装 CPAN，再逐个安装包，成功安装 Params::Validate。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">perl --version</span><br><span class="line"># This is perl 5, version 26, subversion 2 (v5.26.2) built for x86_64-linux-thread-multi</span><br><span class="line">perl --version</span><br><span class="line">This is perl 5, version 16, subversion 3 (v5.16.3) built for x86_64-linux-thread-multi</span><br><span class="line">perl -MCPAN -e shell</span><br><span class="line"></span><br><span class="line">Config::General (v2.50 or later)</span><br><span class="line">Font::TTF</span><br><span class="line">GD</span><br><span class="line">List::MoreUtils</span><br><span class="line">Math::Bezier</span><br><span class="line">Math::Round</span><br><span class="line">Math::VecStat</span><br><span class="line">Params::Validate (自动安装失败，尝试手动安装)</span><br><span class="line">Readonly</span><br><span class="line">Regexp::Common</span><br><span class="line">Set::IntSpan (v1.16 or later)</span><br><span class="line">Text::Format</span><br></pre></td></tr></table></figure>
<p>安装 Params::Validate 的过程中，同时安装了下列 Perl 模块。均是模块安装包下载之后解压，进入模块目录后依次运行 <code>perl Makefile.PL</code>，<code>make Makefile</code>，<code>make install Makefile</code> 即可。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">wget https://cpan.metacpan.org/authors/id/D/DR/DROLSKY/Params-Validate-1.30.tar.gz</span><br><span class="line"># wget http://search.cpan.org/CPAN/authors/id/D/DR/DROLSKY/Params-Validate-0.95.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/E/ET/ETHER/Module-Metadata-1.000037.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/L/LE/LEONT/version-0.9929.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/B/BI/BINGOS/IPC-Cmd-1.04.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/B/BI/BINGOS/Params-Check-0.38.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/J/JE/JESSE/Locale-Maketext-Simple-0.21.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/B/BI/BINGOS/Module-Load-Conditional-0.74.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/B/BI/BINGOS/Module-Load-0.36.tar.gz</span><br></pre></td></tr></table></figure>
<h3 id="installing-libpng-freetype-libgd-and-gd"><a class="markdownIt-Anchor" href="#installing-libpng-freetype-libgd-and-gd"></a> INSTALLING libpng, freetype, libgd AND gd</h3>
<p>About half of the support questions in the Google Group are about installing Perl’s GD module, which is the interface to the libgd system graphics library.</p>
<p>The installation process is not robust because of dependencies and the possibility of their being different versions of these dependencies on your system. Below I show the process of installing GD on a new Mac OS Mavericks system (10.9.4).</p>
<p>Below I link to my local copies of the versions of each library that I installed. These are mature libraries and newer versions are likely to be minor bug releases. The exception is GD—I was unable to successfully compile v2.56 (boot_GD symbol warning which I could not fix). However v2.53 worked.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">&gt; tar xvfz libpng-1.6.14.tar.gz</span><br><span class="line">&gt; cd libpng-1.6.14</span><br><span class="line">&gt; ./configure —prefix=/usr/local</span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br><span class="line"></span><br><span class="line">&gt; tar xvfz jpegsrc.v9.tar.gz</span><br><span class="line">&gt; cd jpeg-9</span><br><span class="line">&gt; ./configure —prefix=/usr/local</span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br><span class="line"></span><br><span class="line">&gt; tar xvfz freetype-2.4.0.tar.gz</span><br><span class="line">&gt; cd freetype-2.4.0</span><br><span class="line">&gt; ./configure —prefix=/usr/local</span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br></pre></td></tr></table></figure>
<p>You should now have libraries in /usr/local/lib as we as some header files in /usr/local/include.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">-rwxr-xr-x  1 root  wheel   650336 21 Nov 11:15 libfreetype.6.dylib</span><br><span class="line">-rw-r--r--  1 root  wheel  3676456 21 Nov 11:15 libfreetype.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       19 21 Nov 11:15 libfreetype.dylib -&gt; libfreetype.6.dylib</span><br><span class="line">-rwxr-xr-x  1 root  wheel      957 21 Nov 11:15 libfreetype.la</span><br><span class="line">-rwxr-xr-x  1 root  wheel   275736 21 Nov 11:14 libjpeg.9.dylib</span><br><span class="line">-rw-r--r--  1 root  wheel  1492952 21 Nov 11:14 libjpeg.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       15 21 Nov 11:14 libjpeg.dylib -&gt; libjpeg.9.dylib</span><br><span class="line">-rwxr-xr-x  1 root  wheel      920 21 Nov 11:14 libjpeg.la</span><br><span class="line">lrwxr-xr-x  1 root  wheel       10 21 Nov 11:13 libpng.a -&gt; libpng16.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       14 21 Nov 11:13 libpng.dylib -&gt; libpng16.dylib</span><br><span class="line">lrwxr-xr-x  1 root  wheel       11 21 Nov 11:13 libpng.la -&gt; libpng16.la</span><br><span class="line">-rwxr-xr-x  1 root  wheel   240580 21 Nov 11:13 libpng16.16.dylib</span><br><span class="line">-rw-r--r--  1 root  wheel  1077240 21 Nov 11:13 libpng16.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       17 21 Nov 11:13 libpng16.dylib -&gt; libpng16.16.dylib</span><br><span class="line">-rwxr-xr-x  1 root  wheel      924 21 Nov 11:13 libpng16.la</span><br><span class="line">drwxr-xr-x  6 root  wheel      204 21 Nov 11:15 pkgconfig/</span><br></pre></td></tr></table></figure>
<p>Now install libgd, linking to the libraries installed above</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">&gt; tar xvfz libgd-2.1.0.tar.gz</span><br><span class="line">&gt; cd libgd-2.1.0</span><br><span class="line">&gt; ./configure --with-png=/usr/local --with-freetype=/usr/local --with-jpeg=/usr/local --prefix=/usr/local</span><br><span class="line">./configure --with-png=/home/fenglei/local --with-freetype=/home/fenglei/local --with-jpeg=/home/fenglei/local --prefix=/home/fenglei/local</span><br><span class="line">…</span><br><span class="line">** Configuration summary for libgd 2.1.0:</span><br><span class="line"></span><br><span class="line">   Support for Zlib:                 yes</span><br><span class="line">   Support for PNG library:          yes</span><br><span class="line">   Support for JPEG library:         yes</span><br><span class="line">   Support for VPX library:          no</span><br><span class="line">   Support for TIFF library:         no</span><br><span class="line">   Support for Freetype 2.x library: yes</span><br><span class="line">   Support for Fontconfig library:   no</span><br><span class="line">   Support for Xpm library:          no</span><br><span class="line">   Support for pthreads:             yes</span><br><span class="line">…</span><br><span class="line"></span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br></pre></td></tr></table></figure>
<p>You now have libgd in /usr/local/lib</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">-rwxr-xr-x  1 root  wheel   389372 19 Nov 14:47 libgd.3.dylib</span><br><span class="line">-rw-r--r--  1 root  wheel  1217200 19 Nov 14:47 libgd.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       13 19 Nov 14:47 libgd.dylib -&gt; libgd.3.dylib</span><br><span class="line">-rwxr-xr-x  1 root  wheel     1139 19 Nov 14:47 libgd.la</span><br></pre></td></tr></table></figure>
<p>as well as some binaries in /usr/local/bin. In particular, you have /usr/local/bin/glib-config, which provides the configuration for your libgd installation</p>
<p>The --with-* parameters during the configure stage of gdlib installation sets the sources of the dependencies that will be build into gdlib. If you used a different -prefix in compiling these dependencies (see above), adjust these parameters accordingly. For example, if you compiled libpng with -prefix=/my/path then use --with-png=/my/path/.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">/usr/local/bin/gdlib-config --all</span><br><span class="line"></span><br><span class="line">GD library  2.1.0</span><br><span class="line">includedir: /usr/local/include</span><br><span class="line">cflags:     -I/usr/local/include</span><br><span class="line">ldflags:     -L/usr/local/lib</span><br><span class="line">libs:       -ljpeg -lz  -L/usr/local/lib -lpng16 -L/usr/local/lib -lfreetype -lz -liconv</span><br><span class="line">libdir:     /usr/local/lib</span><br><span class="line">features:   GD_JPEG GD_FREETYPE GD_PNG GD_GIF GD_GIFANIM GD_OPENPOLYGON</span><br></pre></td></tr></table></figure>
<p>由于我安装的 gd 时候是配置到了用户目录下，所以得到下面的信息。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">GD library  2.1.0</span><br><span class="line">includedir: /home/fenglei/local/include</span><br><span class="line">cflags:     -I/home/fenglei/local/include</span><br><span class="line">ldflags:     -L/home/fenglei/local/lib</span><br><span class="line">libs:       -ljpeg -lz -lm  -L/home/fenglei/local/lib -lpng16 -L/home/fenglei/local/lib -lfreetype -lz -lfontconfig -lfreetype   -lfontconfig -lXpm -lX11   /home/fenglei/local/lib/libiconv.so -Wl,-rpath -Wl,/home/fenglei/local/lib</span><br><span class="line">libdir:     /home/fenglei/local/lib</span><br><span class="line">features:   GD_XPM GD_JPEG GD_FONTCONFIG GD_FREETYPE GD_PNG GD_GIF GD_GIFANIM GD_OPENPOLYGON</span><br></pre></td></tr></table></figure>
<p>You must have GD_FREETYPE and GD_PNG for Circos to run. The other features, such as support for JPEG and TIFF are optional. In this example, I’ve included the JPEG library in the installation.</p>
<p>Now, install the Perl interface to libgd — the GD module.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">&gt; tar xvfz GD-2.53.tar.gz</span><br><span class="line">&gt; perl Makefile.PL</span><br><span class="line"></span><br><span class="line">Configuring for libgd version 2.1.0.</span><br><span class="line">Checking for stray libgd header files...none found.</span><br><span class="line"></span><br><span class="line">Included Features:          GD_JPEG GD_FREETYPE GD_PNG GD_GIF GD_GIFANIM GD_OPENPOLYGON</span><br><span class="line">GD library used from:       /usr/local</span><br><span class="line">Checking if your kit is complete...</span><br><span class="line">Looks good</span><br><span class="line">Writing Makefile for GD</span><br><span class="line">Writing MYMETA.yml and MYMETA.json</span><br><span class="line"></span><br><span class="line">During the installation GD will look for libgd-config (see above) for libgd configuration. </span><br><span class="line"></span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br><span class="line"></span><br><span class="line">&gt; perl -MGD -e &#x27;print $GD::VERSION,”\n”’</span><br><span class="line">&gt; 2.53</span><br></pre></td></tr></table></figure>
<p>Test that the modules are installed using</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&gt; circos --modules</span><br><span class="line">…</span><br><span class="line">ok       0.39 Font::TTF::Font</span><br><span class="line">ok       2.53 GD</span><br><span class="line">ok        0.2 GD::Polyline</span><br><span class="line">ok       2.39 Getopt::Long</span><br><span class="line">ok       1.16 IO::File</span><br><span class="line">…</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line">wget http://circos.ca/distribution/circos-0.69-9.tgz</span><br><span class="line">tar -zxvf circos-0.69-9.tgz</span><br><span class="line">./bin/circos --modules</span><br><span class="line"></span><br><span class="line">ok       1.26 Carp</span><br><span class="line">missing            Clone</span><br><span class="line">ok       2.63 Config::General</span><br><span class="line">ok       3.40 Cwd</span><br><span class="line">ok      2.145 Data::Dumper</span><br><span class="line">ok       2.52 Digest::MD5</span><br><span class="line">ok       2.84 File::Basename</span><br><span class="line">ok       3.40 File::Spec::Functions</span><br><span class="line">ok     0.2301 File::Temp</span><br><span class="line">ok       1.51 FindBin</span><br><span class="line">ok       0.39 Font::TTF::Font</span><br><span class="line">missing            GD</span><br><span class="line">missing            GD::Polyline</span><br><span class="line">ok        2.4 Getopt::Long</span><br><span class="line">ok       1.16 IO::File</span><br><span class="line">ok      0.430 List::MoreUtils</span><br><span class="line">ok       1.27 List::Util</span><br><span class="line">ok       0.01 Math::Bezier</span><br><span class="line">ok      1.997 Math::BigFloat</span><br><span class="line">ok       0.07 Math::Round</span><br><span class="line">ok       0.08 Math::VecStat</span><br><span class="line">ok       1.02 Memoize</span><br><span class="line">ok       1.30 POSIX</span><br><span class="line">ok       1.30 Params::Validate</span><br><span class="line">ok       1.63 Pod::Usage</span><br><span class="line">ok       2.05 Readonly</span><br><span class="line">ok 2017060201 Regexp::Common</span><br><span class="line">missing            SVG</span><br><span class="line">ok       1.19 Set::IntSpan</span><br><span class="line">missing            Statistics::Basic</span><br><span class="line">ok       2.45 Storable</span><br><span class="line">ok       1.16 Sys::Hostname</span><br><span class="line">ok       2.02 Text::Balanced</span><br><span class="line">ok       0.62 Text::Format</span><br><span class="line">ok     1.9725 Time::HiRes</span><br></pre></td></tr></table></figure>
<p>Install missing packages</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cpan -i SVG</span><br><span class="line">cpan -i Clone</span><br><span class="line">cpan -i Statistics::Basic</span><br></pre></td></tr></table></figure>
<h3 id="以上做完依然报错"><a class="markdownIt-Anchor" href="#以上做完依然报错"></a> 以上做完依然报错</h3>
<p>运行 <code>circos --mudules</code> 显示 GD 模块缺失。（已经分别安装 libgd 和 Perl 的 GD 模块，为什么circos还是显示没有 GD 模块？）</p>
<h3 id="使用-conda-安装-circos"><a class="markdownIt-Anchor" href="#使用-conda-安装-circos"></a> 使用 conda 安装 circos</h3>
<p>换一台服务器，<code>source activate python37</code> 切换成 python3.7 环境。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&gt; which perl</span><br><span class="line">~/local/app/anaconda3/envs/python37/bin/perl</span><br><span class="line">&gt; perl --version</span><br><span class="line">This is perl 5, version 26, subversion 2 (v5.26.2) built for x86_64-linux-thread-multi</span><br></pre></td></tr></table></figure>
<p>再运行 <code>conda install -c bioconda circos</code> 顺利安装！下面是模块测试。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost ~]$ circos --modules</span><br><span class="line">ok       1.38 Carp</span><br><span class="line">ok       0.42 Clone</span><br><span class="line">ok       2.63 Config::General</span><br><span class="line">ok       3.75 Cwd</span><br><span class="line">ok      2.167 Data::Dumper</span><br><span class="line">ok       2.55 Digest::MD5</span><br><span class="line">ok       2.85 File::Basename</span><br><span class="line">ok       3.67 File::Spec::Functions</span><br><span class="line">ok     0.2304 File::Temp</span><br><span class="line">ok       1.51 FindBin</span><br><span class="line">ok       0.39 Font::TTF::Font</span><br><span class="line">ok       2.71 GD</span><br><span class="line">ok        0.2 GD::Polyline</span><br><span class="line">ok       2.49 Getopt::Long</span><br><span class="line">ok       1.16 IO::File</span><br><span class="line">ok      0.428 List::MoreUtils</span><br><span class="line">ok     1.4602 List::Util</span><br><span class="line">ok       0.01 Math::Bezier</span><br><span class="line">ok   1.999806 Math::BigFloat</span><br><span class="line">ok       0.07 Math::Round</span><br><span class="line">ok       0.08 Math::VecStat</span><br><span class="line">ok    1.03_01 Memoize</span><br><span class="line">ok       1.76 POSIX</span><br><span class="line">ok       1.29 Params::Validate</span><br><span class="line">ok       1.69 Pod::Usage</span><br><span class="line">ok       2.05 Readonly</span><br><span class="line">ok 2017060201 Regexp::Common</span><br><span class="line">ok       2.84 SVG</span><br><span class="line">ok       1.19 Set::IntSpan</span><br><span class="line">ok     1.6611 Statistics::Basic</span><br><span class="line">ok       2.62 Storable</span><br><span class="line">ok       1.20 Sys::Hostname</span><br><span class="line">ok       2.03 Text::Balanced</span><br><span class="line">ok       0.59 Text::Format</span><br><span class="line">ok     1.9741 Time::HiRes</span><br></pre></td></tr></table></figure>
<p>再以 circos 自带的案例做画图测试。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">wget http://circos.ca/distribution/circos-0.69-9.tgz</span><br><span class="line">tar -zxvf circos-0.69-9.tgz</span><br><span class="line">cd circos-0.69-9</span><br><span class="line">cd example</span><br><span class="line"># 在 example 目录下有名为 run 的脚本：../bin/circos -conf etc/circos.conf -debug_group summary,timer &gt; run.out</span><br><span class="line"># 由于我调用的是 anaconda3 里面的 circos，所以用下面的命令，几秒钟即可运行完毕，顺利得到 circos.svg &amp; circos.png 图。</span><br><span class="line">circos -conf etc/circos.conf -debug_group summary,timer &gt; run.out</span><br></pre></td></tr></table></figure>
<p><img src="http://i.imgur.com/ZkOgjPN.jpg" alt="example/circos.png" /></p>
<p>注：<br />
另一台服务器上，通过 conda 可以安装 circos，但是运行的时候报错：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; circos</span><br><span class="line">perl: symbol lookup error: /home/fenglei/perl5/lib/perl5/x86_64-linux-thread-multi/auto/Clone/Clone.so: undefined symbol: Perl_xs_apiversion_bootcheck</span><br></pre></td></tr></table></figure>
<p>看来是 Perl 环境变量的问题。</p>
<p>Circos 的文件结构如下图所示，conf 文件即图片属性设置，txt 文件就是统计数据，circos 将二者的结合就是图片。</p>
<p><img src="http://qiubio.com/wp-content/uploads/2011/11/circos-architecture.png" alt="" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/2/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><a class="page-number" href="/page/2/">2</a><span class="page-number current">3</span><a class="page-number" href="/page/4/">4</a><span class="space">&hellip;</span><a class="page-number" href="/page/18/">18</a><a class="extend next" rel="next" href="/page/4/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">179</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
