<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta property="og:type" content="website">
<meta property="og:title" content="FENG&#39;s Notes">
<meta property="og:url" content="http://example.com/page/3/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/page/3/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/06/10/NAS-adding-new-harddisks-20220608/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2022/06/10/NAS-adding-new-harddisks-20220608/" class="post-title-link" itemprop="url">Notes | 群晖 Synology NAS 新硬盘的安装</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-06-10 19:32:13" itemprop="dateCreated datePublished" datetime="2022-06-10T19:32:13+08:00">2022-06-10</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-22 10:06:27" itemprop="dateModified" datetime="2022-07-22T10:06:27+08:00">2022-07-22</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Notes/" itemprop="url" rel="index"><span itemprop="name">Notes</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>新买一个 8 TB 硬盘，需要插入群晖 NAS 的空盘位，并且作为独立的一个存储空间用来备份文件，跟之前的两个存储空间不影响。</p>
          <!--noindex-->
            <div class="post-button">
              <a class="btn" href="/2022/06/10/NAS-adding-new-harddisks-20220608/#more" rel="contents">
                Read more &raquo;
              </a>
            </div>
          <!--/noindex-->
        
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/05/26/Linux-add-new-harddisk/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2022/05/26/Linux-add-new-harddisk/" class="post-title-link" itemprop="url">Linux | 给Linux工作站更换新硬盘</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-05-26 15:32:13" itemprop="dateCreated datePublished" datetime="2022-05-26T15:32:13+08:00">2022-05-26</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-14 20:51:43" itemprop="dateModified" datetime="2022-07-14T20:51:43+08:00">2022-07-14</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>之前已经挂载过硬盘，就是在root界面先<code>fdisk -l</code>查看机箱里有哪些硬盘，比如<code>/dev/sda</code>和<code>/dev/sdb</code>，如果<code>/dev/sdb</code>需要挂载，就用<code>fdisk /dev/sdb</code>命令进行硬盘分区，依次输入<code>n</code>（创建新分区），<code>p</code>（创建主分区）,<code>w</code>（保存新建分区）就可以将一块硬盘作为一个新的分区，比如<code>/dev/sdb1</code>。这样我们就创建完一个分区，如果这块硬盘要创建更多分区，比如<code>/dev/sdb2</code>，可以照上面的步骤继续创建。</p>
<p>现在问题来了，我新买了一块4TB硬盘，按照上面的步骤操作，挂载之后怎么只有2TB呢？</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost fenglei]# fdisk -l</span><br><span class="line"></span><br><span class="line">Disk /dev/sdb: 4000.8 GB, 4000787030016 bytes, 7814037168 sectors</span><br><span class="line">Units = sectors of 1 * 512 = 512 bytes</span><br><span class="line">Sector size (logical/physical): 512 bytes / 512 bytes</span><br><span class="line">I/O size (minimum/optimal): 512 bytes / 512 bytes</span><br><span class="line">Disk label type: dos</span><br><span class="line">Disk identifier: 0xabe30878</span><br><span class="line"></span><br><span class="line">   Device Boot      Start         End      Blocks   Id  System</span><br><span class="line">/dev/sdb1            2048  4294967294  2147482623+  83  Linux</span><br><span class="line">WARNING: fdisk GPT support is currently new, and therefore in an experimental phase. Use at your own discretion.</span><br></pre></td></tr></table></figure>
<p>查看<a target="_blank" rel="noopener" href="https://blog.frytea.com/archives/597/">资料</a>才知道：进行分区前需要先明确采用分区表的格式，目前主流的有 MBR 和 GPT ，二者的区别可以自行搜索，总结两点：MBR 兼容性较好，兼容所有 windows，但单盘最大 2TB ；GPT 是一种新格式，最大支持 18EB 的大容量，但并不是所有的 windows 都支持。知道上面两点就够了，这里我使用在 Linux 服务器上，不需要考虑 windows 兼容性，此外是一块 4T 盘，因此采用 GPT 进行分区，下面两种方式请根据自己的需要选择。</p>
<p>看来默认的是 MBR，而不是 GPT。方法也很简单，就是在<code>fdisk /dev/sdb</code>界面先输入<code>d</code>，删除现有分区信息，然后输入<code>g</code>，创建一张新的空 GPT 格式分区表。如下所示，顺利将一整块硬盘的空间新建一个分区。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br></pre></td><td class="code"><pre><span class="line">$ fdisk /dev/sdb</span><br><span class="line">Command (m for help): p</span><br><span class="line"></span><br><span class="line">Disk /dev/sdb: 4000.8 GB, 4000787030016 bytes, 7814037168 sectors</span><br><span class="line">Units = sectors of 1 * 512 = 512 bytes</span><br><span class="line">Sector size (logical/physical): 512 bytes / 512 bytes</span><br><span class="line">I/O size (minimum/optimal): 512 bytes / 512 bytes</span><br><span class="line">Disk label type: dos</span><br><span class="line">Disk identifier: 0xabe30878</span><br><span class="line"></span><br><span class="line">   Device Boot      Start         End      Blocks   Id  System</span><br><span class="line">/dev/sdb1            2048  4294967294  2147482623+  83  Linux</span><br><span class="line"></span><br><span class="line">Command (m for help): d</span><br><span class="line">Selected partition 1</span><br><span class="line">Partition 1 is deleted</span><br><span class="line"></span><br><span class="line">Command (m for help): p               </span><br><span class="line"></span><br><span class="line">Disk /dev/sdb: 4000.8 GB, 4000787030016 bytes, 7814037168 sectors</span><br><span class="line">Units = sectors of 1 * 512 = 512 bytes</span><br><span class="line">Sector size (logical/physical): 512 bytes / 512 bytes</span><br><span class="line">I/O size (minimum/optimal): 512 bytes / 512 bytes</span><br><span class="line">Disk label type: dos</span><br><span class="line">Disk identifier: 0xabe30878</span><br><span class="line"></span><br><span class="line">   Device Boot      Start         End      Blocks   Id  System</span><br><span class="line"></span><br><span class="line">Command (m for help): g</span><br><span class="line">Building a new GPT disklabel (GUID: A2443A99-8E33-4353-947A-0244414D75A4)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Command (m for help): n</span><br><span class="line">Partition number (1-128, default 1): </span><br><span class="line">First sector (2048-7814037134, default 2048): </span><br><span class="line">Last sector, +sectors or +size&#123;K,M,G,T,P&#125; (2048-7814037134, default 7814037134): </span><br><span class="line">Created partition 1</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Command (m for help): p</span><br><span class="line"></span><br><span class="line">Disk /dev/sdb: 4000.8 GB, 4000787030016 bytes, 7814037168 sectors</span><br><span class="line">Units = sectors of 1 * 512 = 512 bytes</span><br><span class="line">Sector size (logical/physical): 512 bytes / 512 bytes</span><br><span class="line">I/O size (minimum/optimal): 512 bytes / 512 bytes</span><br><span class="line">Disk label type: gpt</span><br><span class="line">Disk identifier: A2443A99-8E33-4353-947A-0244414D75A4</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">#         Start          End    Size  Type            Name</span><br><span class="line"> 1         2048   7814037134    3.7T  Linux filesyste </span><br><span class="line"></span><br><span class="line">Command (m for help): w</span><br><span class="line">The partition table has been altered!</span><br><span class="line"></span><br><span class="line">Calling ioctl() to re-read partition table.</span><br><span class="line">Syncing disks.</span><br><span class="line">[root@localhost fenglei]# </span><br></pre></td></tr></table></figure>
<h3 id="硬盘格式化"><a class="markdownIt-Anchor" href="#硬盘格式化"></a> 硬盘格式化</h3>
<p>先看看分区信息</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost fenglei]# fdisk -l</span><br><span class="line">WARNING: fdisk GPT support is currently new, and therefore in an experimental phase. Use at your own discretion.</span><br><span class="line"></span><br><span class="line">Disk /dev/sdb: 4000.8 GB, 4000787030016 bytes, 7814037168 sectors</span><br><span class="line">Units = sectors of 1 * 512 = 512 bytes</span><br><span class="line">Sector size (logical/physical): 512 bytes / 512 bytes</span><br><span class="line">I/O size (minimum/optimal): 512 bytes / 512 bytes</span><br><span class="line">Disk label type: gpt</span><br><span class="line">Disk identifier: A2443A99-8E33-4353-947A-0244414D75A4</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">#         Start          End    Size  Type            Name</span><br><span class="line"> 1         2048   7814037134    3.7T  Linux filesyste </span><br><span class="line">WARNING: fdisk GPT support is currently new, and therefore in an experimental phase. Use at your own discretion.</span><br></pre></td></tr></table></figure>
<p>没问题后进行格式化，如果没有特殊需求就采用 ext4 ：第四代扩展文件系统（英语：Fourth extended filesystem，缩写为 ext4）是 Linux 系统下的日志文件系统，是 ext3 文件系统的后继版本。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost fenglei]#  mkfs.ext4 /dev/sdb1</span><br><span class="line">mke2fs 1.42.9 (28-Dec-2013)</span><br><span class="line">Filesystem label=</span><br><span class="line">OS type: Linux</span><br><span class="line">Block size=4096 (log=2)</span><br><span class="line">Fragment size=4096 (log=2)</span><br><span class="line">Stride=0 blocks, Stripe width=0 blocks</span><br><span class="line">244195328 inodes, 976754385 blocks</span><br><span class="line">48837719 blocks (5.00%) reserved for the super user</span><br><span class="line">First data block=0</span><br><span class="line">Maximum filesystem blocks=3124756480</span><br><span class="line">29809 block groups</span><br><span class="line">32768 blocks per group, 32768 fragments per group</span><br><span class="line">8192 inodes per group</span><br><span class="line">Superblock backups stored on blocks: </span><br><span class="line">        32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632, 2654208, </span><br><span class="line">        4096000, 7962624, 11239424, 20480000, 23887872, 71663616, 78675968, </span><br><span class="line">        102400000, 214990848, 512000000, 550731776, 644972544</span><br><span class="line"></span><br><span class="line">Allocating group tables:     0/29809</span><br><span class="line">done                            </span><br><span class="line">Writing inode tables: done                            </span><br><span class="line">Creating journal (32768 blocks): done</span><br><span class="line">Writing superblocks and filesystem accounting information: done </span><br></pre></td></tr></table></figure>
<p>上面的格式化只需要几秒钟！</p>
<p>以前的ext3分区需要几个小时，如下所示，下面这个命令不要再用了。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">#mkfs -t ext3 -c /dev/sdb1  //如果有多个分区，则分区修改为sdb2这样</span><br><span class="line">（格式化2TB的硬盘，耗时约3小时。20151125 FL备注）</span><br></pre></td></tr></table></figure>
<h3 id="挂载"><a class="markdownIt-Anchor" href="#挂载"></a> 挂载</h3>
<p>下面挂载到 <code>/mnt/sdb</code>。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost fenglei]# mount /dev/sdb1 /mnt/sdb</span><br><span class="line">[root@localhost fenglei]# df -h</span><br><span class="line">Filesystem      Size  Used Avail Use% Mounted on</span><br><span class="line">devtmpfs         32G     0   32G   0% /dev</span><br><span class="line">tmpfs            32G     0   32G   0% /dev/shm</span><br><span class="line">tmpfs            32G  9.7M   32G   1% /run</span><br><span class="line">tmpfs            32G     0   32G   0% /sys/fs/cgroup</span><br><span class="line">/dev/sda5       3.2T  3.0T  209G  94% /</span><br><span class="line">/dev/sda3       466G  466G   20K 100% /home</span><br><span class="line">/dev/sda2       197M  186M   11M  95% /boot</span><br><span class="line">tmpfs           6.3G     0  6.3G   0% /run/user/1000</span><br><span class="line">/dev/sdc1       1.8T  707G  1.1T  41% /mnt/sdc</span><br><span class="line">/dev/sdb1       3.6T   89M  3.4T   1% /mnt/sdb</span><br></pre></td></tr></table></figure>
<h3 id="设置开机自动挂载硬盘"><a class="markdownIt-Anchor" href="#设置开机自动挂载硬盘"></a> 设置开机自动挂载硬盘</h3>
<p>直接挂载，重启后配置会丢失，可以修改 /etc/fstab 配置文件配置开机自动挂载，该配置文件的格式如下：<br />
（下面是参考资料直接抄来的）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">$ cat /etc/fstab </span><br><span class="line"># /etc/fstab: static file system information.</span><br><span class="line">#</span><br><span class="line"># Use &#x27;blkid&#x27; to print the universally unique identifier for a</span><br><span class="line"># device; this may be used with UUID= as a more robust way to name devices</span><br><span class="line"># that works even if disks are added and removed. See fstab(5).</span><br><span class="line">#</span><br><span class="line"># &lt;file system&gt; &lt;mount point&gt;   &lt;type&gt;  &lt;options&gt;       &lt;dump&gt;  &lt;pass&gt;</span><br><span class="line"># / was on /dev/sda7 during installation</span><br><span class="line">UUID=4ae9f128-7cee-48fd-998c-4257e394fa4a /               ext4    errors=remount-ro 0       1</span><br><span class="line"># /boot/efi was on /dev/sda1 during installation</span><br><span class="line">UUID=A5EB-573A  /boot/efi       vfat    umask=0077      0       1</span><br><span class="line">/swapfile                                 none            swap    sw              0       0</span><br><span class="line"></span><br><span class="line"># /dev/sda1</span><br><span class="line">/dev/sda1    /    ext4    rw,relatime    0    1</span><br><span class="line">UUID=4ae9f128-7cee-48fd-998c-4257e394fa4a /               ext4    errors=remount-ro 0       1</span><br></pre></td></tr></table></figure>
<p>格式大概是这样的：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># &lt;file system&gt; &lt;mount point&gt;   &lt;type&gt;  &lt;options&gt;       &lt;dump&gt;  &lt;pass&gt;</span><br><span class="line"># 要挂载的分区设备号    挂载点    文件系统类型    挂载选项    是否备份    是否检测</span><br></pre></td></tr></table></figure>
<p>如果需要设备号，可以使用 blkid 命令获取：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ blkid</span><br><span class="line">/dev/sda1: UUID=&quot;e943af1e-72cf-4b72-b444-859b9610256f&quot; TYPE=&quot;ext4&quot; PARTUUID=&quot;161a8c3f-871d-45f6-a25f-466bf16f6035&quot;</span><br></pre></td></tr></table></figure>
<p>实测下面这两种都是可以挂载的：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">UUID=4ae9f128-7cee-48fd-998c-4257e394fa4a /               ext4    errors=remount-ro     0       1</span><br><span class="line">/dev/sda1       /               ext4            rw,relatime     0       1</span><br></pre></td></tr></table></figure>
<p>我的实际操作如下，在 /etc/fstab  添加如下内容！</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># /dev/sdb1: UUID=&quot;42ece9ae-1c69-4a2d-8862-45c44c44ccac&quot; TYPE=&quot;ext4&quot; PARTUUID=&quot;b3035439-0e9e-4512-b56c-d5cfc7754706&quot;</span><br><span class="line">UUID=42ece9ae-1c69-4a2d-8862-45c44c44ccac       /mnt/sdb        ext4    defaults        0       1</span><br></pre></td></tr></table></figure>
<p>写好配置后可以使用该命令生效（挂载 /etc/fstab 中所有档案）：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mount -a</span><br></pre></td></tr></table></figure>
<p>下次重启设备就不需要再手动挂载该磁盘了。</p>
<p>磁盘状态</p>
<p>使用 lsblk 看一下当前硬盘的树形结构：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ lsblk </span><br><span class="line">NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT</span><br><span class="line">sda      8:0    0 465.8G  0 disk </span><br><span class="line">└─sda1   8:1    0 465.8G  0 part /</span><br><span class="line">sdc      8:32   0   3.7T  0 disk </span><br><span class="line">└─sdc1   8:33   0   3.7T  0 part /home/songtianlun/data</span><br></pre></td></tr></table></figure>
<p>使用 df-h 看看使用率：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ df -h</span><br><span class="line">文件系统        容量  已用  可用 已用% 挂载点</span><br><span class="line">/dev/sda1       459G  8.6G  427G    2% /</span><br><span class="line">...</span><br><span class="line">/dev/sdc1       3.6T   89M  3.4T    1% /home/songtianlun/data</span><br></pre></td></tr></table></figure>
<h3 id="参考资料"><a class="markdownIt-Anchor" href="#参考资料"></a> 参考资料</h3>
<p><a target="_blank" rel="noopener" href="https://blog.frytea.com/archives/597/">https://blog.frytea.com/archives/597/</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/05/15/Linux-adjust-space-in-home/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2022/05/15/Linux-adjust-space-in-home/" class="post-title-link" itemprop="url">Linux | 安装Linux系统之后还想调整home目录的空间怎么操作</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-05-15 15:32:13" itemprop="dateCreated datePublished" datetime="2022-05-15T15:32:13+08:00">2022-05-15</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-14 20:52:10" itemprop="dateModified" datetime="2022-07-14T20:52:10+08:00">2022-07-14</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>问题：2020 年的时候，服务器系统盘损坏导致系统无法开机，于是新加了一块硬盘作为 Linux 系统盘。系统盘 /dev/sda，拿一个 500 GB 的分区（/dev/sda3）作为系统安装位置，其余 3 TB 左右（/dev/sda5）则准备挂载到系统的某个目录下（比如/mnt/xxx）。但是同学习惯在 home 目录下做项目，于是 500 GB 的空间很快就不够用，同学采取 bind mount 的方法将其余空间也挂载到 home 目录下面。</p>
<p>确定四下无人后, 开始搬家(复制). 使用 cp 命令复制时, 记得带上 -p 参数, 保留文件权限设置. 使用 root 权限, 假设目标分区为 /new_disk :</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cp -p -r /home /home_new/</span><br></pre></td></tr></table></figure>
<p>新家已经建好, 我们先给老家弄个另外的门牌号, 以免把门牌号给新家后, 找不到老家了. 利用 mount –bind 把原 home 目录挂载到一个新目录:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">mkdir /home_bkp</span><br><span class="line">mount --bind /home /home_bkp</span><br></pre></td></tr></table></figure>
<p>这时我们就可以在 /home_bkp 这个目录下找到老家的所有文件. 可以把 /home 这个门牌给新家了:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mount --bind /home_new/home /home</span><br></pre></td></tr></table></figure>
<p>至此 home 目录就扩大了。mount bind 是什么原理？有人说“当我们使用 bind 的时候，是将一个目录 A 挂载到另一个目录 B ，目录 B 原有的内容就被屏蔽掉了，目录 B 里面的内容就是目录 A 里面的内容。这和我们挂在其他分区到挂载点目录一样，目录 B 的内容还是存在的，只不过是被屏蔽了，当我们 umount B 后，原内容就会复现。”</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mount --bind olddir newdir</span><br></pre></td></tr></table></figure>
<p>如果执行了上面这个命令，在 olddir 和 newdir 都可以访问相同的内容，并且如果对其中一个目录内的内容进行了修改，在另一个目录会有相同的显示。</p>
<p>最后，注意一点，如果机器重启，上面的挂载信息会丢失，造成不便，所以可以添加相关挂载信息到 /etc/fstab 文件。这样重启之后也会自动挂载。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">/home   /home_bkp   none    bind    0   0</span><br><span class="line">/home_new   /home   none    bind    0   0</span><br></pre></td></tr></table></figure>
<p>备注：这样操作之后，重启还是没有自动挂载！还是需要重新 mount。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost fenglei]# mount --bind /home /home_bkp</span><br><span class="line">[root@localhost fenglei]# mount --bind /home_new/home /home</span><br></pre></td></tr></table></figure>
<h3 id="参考资料"><a class="markdownIt-Anchor" href="#参考资料"></a> 参考资料：</h3>
<p><a target="_blank" rel="noopener" href="http://blog.csdn.net/liuxu0703/article/details/70225591">http://blog.csdn.net/liuxu0703/article/details/70225591</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/05/05/DOS-usual-commands/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2022/05/05/DOS-usual-commands/" class="post-title-link" itemprop="url">Bioinfo | DOS 界面下的基本操作命令</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-05-05 15:32:13" itemprop="dateCreated datePublished" datetime="2022-05-05T15:32:13+08:00">2022-05-05</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-14 20:48:34" itemprop="dateModified" datetime="2022-07-14T20:48:34+08:00">2022-07-14</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>偶尔需要用到 DOS 界面下调用软件来做数据处理，现对场用的命令行做一个记录。</p>
<ol>
<li>从默认 C 盘切换磁盘到 D 盘，直接输入 <code>d:</code>。</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Microsoft Windows [版本 10.0.19044.1645]</span><br><span class="line">(c) Microsoft Corporation。保留所有权利。</span><br><span class="line"></span><br><span class="line">C:\Users\fengl&gt;d:</span><br><span class="line"></span><br><span class="line">D:\&gt;</span><br></pre></td></tr></table></figure>
<ol start="2">
<li>
<p>显示当前目录下的文件列表：<code>dir</code>。</p>
</li>
<li>
<p>进入目录，输入 <code>cd folder_name/sub_folder</code>。</p>
</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cd    #显示当前目录</span><br><span class="line">cd\   #返回根目录</span><br><span class="line">cd..  #返回上一级目录</span><br></pre></td></tr></table></figure>
<ol start="4">
<li>
<p>运行 exe 程序，直接输入 <code>D:\path\to\xxx.exe</code> 或 <code>xxx.exe</code>。如果路径中有空格怎么办？用双引号处理：<code>&quot;D:\Program Files\path\to\xxx.exe&quot;</code>。</p>
</li>
<li>
<p>清理当前界面：<code>cls</code>。</p>
</li>
<li>
<p>在当前路径下创建新目录：<code>md folder_name</code>。</p>
</li>
<li>
<p>删除目录：<code>rd /s /q folder_name</code>。</p>
</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">rd folder_name  #rd 的另外一个写法是 rmdir。使用的方法也很简单：rd 文件夹名 即可，例如：rd test。</span><br><span class="line">rd /s folder_name #如果要删除的文件夹下面还有子文件，你得给rd加个 /s 参数，它才会执行，所以，正确的写法是：rd /s test。</span><br><span class="line">rd /s /q folder_name #当删除带子文件的文件夹的时候，CMD总是殷勤地问你：是否确认(Y/N)? 参数 /q ，取自 quiet，安静模式，带 /S 删除目录树时不要求确认</span><br></pre></td></tr></table></figure>
<ol start="8">
<li>
<p>在当前目录下删除特定的文件：<code>del file.txt</code>。删除指定的一个或多个文件，不能用于删除子目录。参数 /P 的功能是使 DOS 在删除每个文件之前，要求用户先认可，这样使得用户可以有选择地删除一些文件。</p>
</li>
<li>
<p>批处理，批处理文件（.bat）</p>
</li>
</ol>
<p>以 for 循环为例：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">@echo off</span><br><span class="line">for %%I in (ABC) do echo %%I</span><br><span class="line">pause</span><br></pre></td></tr></table></figure>
<p>批处理文件的for循环就是这样简单，下面来看看for语句的注意事项，并运行更复杂的for循环实例。<br />
for语句的形式变量I，可以换成26个字母中的任意一个，这些字母会区分大小写，也就是说，%%I和%%i会被认为不是同一个变量；形式变量I还可以换成其他的字符，但是，为了不与批处理中的%0～%9这10个形式变量发生冲突，请不要随意把%%I替换为%%0～%%9中的任意一个；<br />
in和do之间的command1表示的字符串或变量可以是一个，也可以是多个，每一个字符串或变量，我们称之为一个元素，每个元素之间，用空格键、跳格键、逗号、分号或等号分隔；<br />
for语句依次提取command1中的每一个元素，把它的值赋予形式变量I，带到do后的command2中参与命令的执行；并且每次只提取一个元素，然后执行一次do后的命令语句，而无论这个元素是否被带到command2中参与了command2的运行；当执行完一次do后的语句之后，再提取command1中的下一个元素，再执行一次command2，如此循环，直到command1中的所有元素都已经被提取完毕，该for语句才宣告执行结束。</p>
<p>有了以上的基础，我们再来看下面这个例子，这个例子修改了demo1中的部分内容(记为demo2)，结果将大不一样：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">@echo off</span><br><span class="line">for  %%I in (A,B,C) do echo %%I</span><br><span class="line">pause</span><br></pre></td></tr></table></figure>
<h3 id="参考资料"><a class="markdownIt-Anchor" href="#参考资料"></a> 参考资料</h3>
<p>追梦菜鸟：<a target="_blank" rel="noopener" href="https://blog.csdn.net/u013514928/article/details/79629937">https://blog.csdn.net/u013514928/article/details/79629937</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/04/20/Bioinfo-DRIMM-synteny-lab-notes/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2022/04/20/Bioinfo-DRIMM-synteny-lab-notes/" class="post-title-link" itemprop="url">Bioinfo | DRIMM synteny and ancestor genomes</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-04-20 15:32:13" itemprop="dateCreated datePublished" datetime="2022-04-20T15:32:13+08:00">2022-04-20</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-14 20:47:40" itemprop="dateModified" datetime="2022-07-14T20:47:40+08:00">2022-07-14</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h3 id="重新进行编译-drimm-程序"><a class="markdownIt-Anchor" href="#重新进行编译-drimm-程序"></a> 重新进行编译 drimm 程序</h3>
<p>安装 dotnet 的方法，进入 <a target="_blank" rel="noopener" href="https://dotnet.microsoft.com/zh-cn/download/dotnet/6.0">https://dotnet.microsoft.com/zh-cn/download/dotnet/6.0</a> ，选择 Wondows 10 栏目的 x64 版本，就可以得到名为“dotnet-sdk-6.0.202-win-x64.exe”的一个安装文件，双击之后默认安装在系统即可。</p>
<p>根据 IAGS 作者的建议：drimm 目录为 IAGS 团队整理的核心代码，因为各个 PC 的 dotnet 版本可能存在不同，我们建议重新进行编译，生成新的exe文件。<br />
在 ./drimm 路径下使用：dotnet publish -c Release -r win-x64 进行编译，可以在./drimm/drimm/bin/Release/netcoreapp3.1/win-x64路径下得到drimm.exe</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">D:\Program Files\processDRIMM\drimm&gt;dotnet publish -c Release -r win-x64</span><br><span class="line"></span><br><span class="line">欢迎使用 .NET 6.0!</span><br><span class="line">---------------------</span><br><span class="line">SDK 版本: 6.0.202</span><br><span class="line"></span><br><span class="line">遥测</span><br><span class="line">---------</span><br><span class="line">.NET 工具会收集用法数据，帮助我们改善你的体验。它由 Microsoft 收集并与社区共享。你可通过使用喜欢的 shell 将 DOTNET_CLI_TELEMETRY_OPTOUT 环境变量设置为 &quot;1&quot; 或 &quot;true&quot; 来选择退出遥测。</span><br><span class="line"></span><br><span class="line">阅读有关 .NET CLI 工具遥测的更多信息: https://aka.ms/dotnet-cli-telemetry</span><br><span class="line"></span><br><span class="line">----------------</span><br><span class="line">已安装 ASP.NET Core HTTPS 开发证书。</span><br><span class="line">若要信任该证书，请运行 &quot;dotnet dev-certs https --trust&quot; (仅限 Windows 和 macOS)。</span><br><span class="line">了解 HTTPS: https://aka.ms/dotnet-https</span><br><span class="line">----------------</span><br><span class="line">编写你的第一个应用: https://aka.ms/dotnet-hello-world</span><br><span class="line">查找新增功能: https://aka.ms/dotnet-whats-new</span><br><span class="line">浏览文档: https://aka.ms/dotnet-docs</span><br><span class="line">在 GitHub 上报告问题和查找源: https://github.com/dotnet/core</span><br><span class="line">使用 &quot;dotnet --help&quot; 查看可用命令或访问: https://aka.ms/dotnet-cli</span><br><span class="line">--------------------------------------------------------------------------------------</span><br><span class="line">用于 .NET 的 Microsoft (R) 生成引擎版本 17.1.1+a02f73656</span><br><span class="line">版权所有(C) Microsoft Corporation。保留所有权利。</span><br><span class="line"></span><br><span class="line">  正在确定要还原的项目…</span><br><span class="line">  已还原 D:\Program Files\processDRIMM\drimm\drimm\drimm.csproj (用时 13.61 sec)。</span><br><span class="line">  drimm -&gt; D:\Program Files\processDRIMM\drimm\drimm\bin\Release\netcoreapp3.1\win-x64\drimm.dll</span><br><span class="line">  drimm -&gt; D:\Program Files\processDRIMM\drimm\drimm\bin\Release\netcoreapp3.1\win-x64\publish\</span><br><span class="line"></span><br><span class="line">D:\Program Files\processDRIMM\drimm&gt;</span><br></pre></td></tr></table></figure>
<h3 id="orthofinder-的使用"><a class="markdownIt-Anchor" href="#orthofinder-的使用"></a> Orthofinder 的使用</h3>
<p>OrthoFinder 的使用需要注意几点：<br />
(1) gff 文件，只包含染色体信息即可，不必包含scaffolds，因为研究对象是染色体进化。染色体的命名要统一规范化，以大豆（Glycine max）为例，数据库自带的命名为Gm01、Gm02 …… 我统一改为：Gmax_chr01、Gmax_chr02 …… 这种“物种名+染色体编号的”格式。<br />
(2) 蛋白数据集，剔除来自scaffolds和细胞器的蛋白序列，只保留来自染色体的蛋白序列。要注意一个基因可能存在若干转录本的情况，Phytozome 下载的数据会提供 primary transcripts 文件，相当于每个基因只提供一个转录本。对于其他情况就需要自己去甄别并采用一定的脚本来做数据规范化处理。<br />
(3) OrthoFinder 运行时，input 目录下是各个物种的蛋白序列文件，注意命名为 Mtruncatula.fasta、Gmax.fasta 这样的格式，因为最后生成的 OrthoGroup.tsv 文件中，文件第一行就是 Mtruncatula、Gmax 字样，这对于后续处理是有益的。<br />
(4) 意外情况，花生的蛋白序列，很多错误，比如蛋白序列中含有点号，这是什么意思？OrthoFinder会报错：Error reading input stream at line 14: Invalid character (.) in sequence\n。</p>
<h3 id="将-orthofinder-的结果转化成下一步的-drimm-synteny-的输入文件"><a class="markdownIt-Anchor" href="#将-orthofinder-的结果转化成下一步的-drimm-synteny-的输入文件"></a> 将 OrthoFinder 的结果转化成下一步的 DRIMM-synteny 的输入文件</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost processDRIMM]$ ll example_fl</span><br><span class="line">total 10380</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 1235362 Apr 22 14:33 Brachy.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 1617027 Apr 22 14:33 Maize.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 1222139 Apr 22 14:33 Rice.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 1367402 Apr 22 14:33 Sorghum.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 2717575 Apr 22 14:33 Telongatum.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 2458531 Apr 22 14:33 Orthogroups.tsv</span><br></pre></td></tr></table></figure>
<p>下面运行 <code>./processOrthofinder.py</code> 程序。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">## 在 ./processOrthofinder.py 首行加上“#!/home/fenglei/local/app/anaconda3/envs/python37/bin/python3”</span><br><span class="line">dos2unix ./processOrthofinder.py    ### 作者原来的文档实在win系统撰写的，拿到linux里面运行会出现结尾符为^M的错误警告</span><br><span class="line">python3 ./processOrthofinder.py     ###</span><br></pre></td></tr></table></figure>
<p>下面检查生成的文件。processOrthofinder 会根据物种目标拷贝数对超过目标拷贝数的同源基因进行过滤，得到用来制备共线块的基因，结果为 *.sequence 文件（Maize.sequence）。之后需要将各个 *.sequence 文件进行合并，得到 drimm.sequence 文件作为 DRIMM 的输入。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost processDRIMM]$ ll example_fl/</span><br><span class="line">total 18068</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  418312 Apr 22 14:42 Telongatum.all.sequence.genename</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  142096 Apr 22 14:42 Telongatum.all.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   80934 Apr 22 14:42 Telongatum.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  455606 Apr 22 14:42 Sorghum.all.sequence.genename</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  128955 Apr 22 14:42 Sorghum.all.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   80002 Apr 22 14:42 Sorghum.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  376841 Apr 22 14:42 Rice.all.sequence.genename</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  127914 Apr 22 14:42 Rice.all.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   80069 Apr 22 14:42 Rice.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  488212 Apr 22 14:42 Maize.all.sequence.genename</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  140247 Apr 22 14:42 Maize.all.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   95546 Apr 22 14:42 Maize.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  442444 Apr 22 14:42 Brachy.all.sequence.genename</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  121269 Apr 22 14:42 Brachy.all.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   79332 Apr 22 14:42 Brachy.sequence</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 1732515 Apr 22 14:42 filter_group.xls</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 2841844 Apr 22 14:42 group.xls</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 1235362 Apr 22 14:33 Brachy.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 1617027 Apr 22 14:33 Maize.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 1222139 Apr 22 14:33 Rice.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 1367402 Apr 22 14:33 Sorghum.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 2717575 Apr 22 14:33 Telongatum.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 2458531 Apr 22 14:33 Orthogroups.tsv</span><br></pre></td></tr></table></figure>
<p>将上面的drimm.sequence文件拷贝到win系统的工作目录下，如下所示，我放在“D:\Program Files\processDRIMM\example_fl”，随后进入 DOS 界面。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">D:\Program Files\processDRIMM\example_fl&gt; dir</span><br><span class="line"> 驱动器 D 中的卷没有标签。</span><br><span class="line"> 卷的序列号是 7EAB-2AD7</span><br><span class="line"></span><br><span class="line"> D:\Program Files\processDRIMM\example_fl 的目录</span><br><span class="line"></span><br><span class="line">2022/04/22  15:43    &lt;DIR&gt;          .</span><br><span class="line">2022/04/22  15:43    &lt;DIR&gt;          ..</span><br><span class="line">2022/04/22  14:57           415,883 drimm.sequence</span><br><span class="line">               1 个文件        415,883 字节</span><br><span class="line">               2 个目录 444,340,531,200 可用字节</span><br></pre></td></tr></table></figure>
<p>这一步骤的操作引起问题，原因是路径中有空格。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">### &quot;D:\Program Files\Informed-Proteomics\ProMex.exe&quot;</span><br><span class="line">C:&gt;D</span><br><span class="line">cd &quot;D:\Program Files\processDRIMM\example_fl&quot;</span><br><span class="line">&gt;&quot;D:\Program Files\processDRIMM\drimm\drimm\bin\Release\netcoreapp3.1\win-x64\drimm.exe&quot;</span><br><span class="line">&quot;D:\Program Files\processDRIMM\example_fl&quot;</span><br></pre></td></tr></table></figure>
<p>在 D 盘新建 InferAncestorGenome 目录，将 processDRIMM 文件夹从 “D:\Program Files\processDRIMM” 复制一份到 D:\InferAncestorGenome 目录。<br />
现在 drimm.sequence 文件存放在 D:\InferAncestorGenome\processDRIMM\example_fl，如下方代码所示，我们在 DOS 界面下进入这个目录，然后执行 <code>dir</code> 命令就可以看到 drimm.sequence 文件；继而调用命令 <code>D:\InferAncestorGenome\processDRIMM\drimm\drimm\bin\Release\netcoreapp3.1\win-x64\drimm.exe</code> 即开始运行 DRIMM-synteny，这时候按照程序的提示分别设定如下内容：</p>
<p>(1) 输入文件：drimm.sequence；<br />
(2) 输出路径：D:\InferAncestorGenome\processDRIMM\example_fl\drimm；<br />
(3) cycleLengthThreshold参数控制产生共线块的大小，设置小参数时会考虑到一些小的变异产生共线块比较破碎打断会比较频繁，设置大参数时运行会比较慢会忽略一些小的变异保留更大的特征，DRIMM 原文中测试数据为酵母基因组选择默认参数为 20，IAGS 文章中处理植物基因组时选择参数为 60，这个参数可以根据研究需要进行调整，一般可以选择默认参数 20<br />
(4) dustThreshold 为最大同源基因多样性，会将同源基因个数中超过这个上限的同源基因组进行过滤，推荐设置为目标拷贝数总和 +1，<br />
例如 IAGS 文章中禾本科五个物种：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sp_list = [&#x27;Brachy&#x27;,&#x27;Maize&#x27;,&#x27;Rice&#x27;,&#x27;Sorghum&#x27;,&#x27;Telongatum&#x27;]  </span><br><span class="line">target_rate = &#x27;2:4:2:2:2&#x27;  </span><br><span class="line">dustThreshold为13（12+1）</span><br></pre></td></tr></table></figure>
<p>下面记载了 DOS 界面下运行 drimm.exe 的场景。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br></pre></td><td class="code"><pre><span class="line">D:\Program Files\processDRIMM\example_fl&gt;cd D:\InferAncestorGenome\processDRIMM\example_fl</span><br><span class="line"></span><br><span class="line">D:\InferAncestorGenome\processDRIMM\example_fl&gt;dir</span><br><span class="line"> 驱动器 D 中的卷没有标签。</span><br><span class="line"> 卷的序列号是 7EAB-2AD7</span><br><span class="line"></span><br><span class="line"> D:\InferAncestorGenome\processDRIMM\example_fl 的目录</span><br><span class="line"></span><br><span class="line">2022/04/22  15:59    &lt;DIR&gt;          .</span><br><span class="line">2022/04/22  15:59    &lt;DIR&gt;          ..</span><br><span class="line">2022/04/22  14:57           415,883 drimm.sequence</span><br><span class="line">               1 个文件        415,883 字节</span><br><span class="line">               2 个目录 444,175,998,976 可用字节</span><br><span class="line"></span><br><span class="line">D:\InferAncestorGenome\processDRIMM\example_fl&gt;D:\InferAncestorGenome\processDRIMM\drimm\drimm\bin\Release\netcoreapp3.1\win-x64\drimm.exe</span><br><span class="line">Enter inputfile</span><br><span class="line">drimm.sequence</span><br><span class="line">Enter outdir</span><br><span class="line">D:\InferAncestorGenome\processDRIMM\example_fl</span><br><span class="line">### D:\InferAncestorGenome\processDRIMM\example_fl\drimm</span><br><span class="line">Enter cycleLengthThreshold</span><br><span class="line">20</span><br><span class="line">Enter dustThreshold</span><br><span class="line">13</span><br><span class="line">Padding numb924</span><br><span class="line">Nodes: 14133</span><br><span class="line">Number of edges: 36364</span><br><span class="line">Nodes:14039</span><br><span class="line">Number of edges: 19134</span><br><span class="line">Nodes:13823</span><br><span class="line">Number of edges: 16765</span><br><span class="line">Nodes:13686</span><br><span class="line">Number of edges: 15998</span><br><span class="line">Nodes:13575</span><br><span class="line">Number of edges: 15636</span><br><span class="line">Nodes:13411</span><br><span class="line">Number of edges: 15326</span><br><span class="line">Nodes:13273</span><br><span class="line">Number of edges: 15035</span><br><span class="line">Nodes:12959</span><br><span class="line">Number of edges: 14540</span><br><span class="line">Nodes:12754</span><br><span class="line">Number of edges: 14244</span><br><span class="line">Nodes:12692</span><br><span class="line">Number of edges: 14152</span><br><span class="line">Nodes:12653</span><br><span class="line">Number of edges: 14099</span><br><span class="line">Nodes:12652</span><br><span class="line">Number of edges: 14090</span><br><span class="line">Nodes:12646</span><br><span class="line">Number of edges: 14084</span><br><span class="line">Nodes:12642</span><br><span class="line">Number of edges: 14075</span><br><span class="line">Nodes:12642</span><br><span class="line">Number of edges: 14102</span><br><span class="line">Nodes:13311</span><br></pre></td></tr></table></figure>
<p>执行 drimm.exe 之后的输出中有两个重要的文件 synteny.txt 以及 blocks.txt 文件。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">D:\InferAncestorGenome\processDRIMM\example_fl&gt;dir</span><br><span class="line"> 驱动器 D 中的卷没有标签。</span><br><span class="line"> 卷的序列号是 7EAB-2AD7</span><br><span class="line"></span><br><span class="line"> D:\InferAncestorGenome\processDRIMM\example_fl 的目录</span><br><span class="line"></span><br><span class="line">2022/04/22  16:03    &lt;DIR&gt;          .</span><br><span class="line">2022/04/22  16:03    &lt;DIR&gt;          ..</span><br><span class="line">2022/04/22  16:03            33,056 blocks.txt</span><br><span class="line">2022/04/22  14:57           415,883 drimm.sequence</span><br><span class="line">2022/04/22  16:03           474,580 modifiedSequence.txt</span><br><span class="line">2022/04/22  16:03           766,866 sequenceColor.txt</span><br><span class="line">2022/04/22  16:03             4,010 split.txt</span><br><span class="line">2022/04/22  16:03            97,352 synteny.txt</span><br><span class="line">               6 个文件      1,791,747 字节</span><br><span class="line">               2 个目录 444,174,614,528 可用字节</span><br></pre></td></tr></table></figure>
<p>其中 synteny.txt 文件截取展示如下，第一列为 block 的 ID，冒号后为在全部物种中出现次数，之后的若干个数字为 block 内部同源基因的 ID。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">...</span><br><span class="line">1534:11 7400 14370 14369 </span><br><span class="line">1535:11 9549 9548 </span><br><span class="line">1536:12 7957 </span><br><span class="line">1537:12 1306 1305 </span><br><span class="line">1538:12 3698 1612 5338 2860 1611 1610 </span><br><span class="line">1539:12 2336 </span><br><span class="line">1540:12 1320 </span><br><span class="line">1541:12 3089 6660 6659 12584 1800 6658 6657 12583 12582 6656 6655 12581 1799 </span><br><span class="line">1542:12 1782 6228 3674 6229 1783 6592 12460 12461 12462 12463 12464 6593 12465 12466 1784 2349 </span><br><span class="line">...</span><br></pre></td></tr></table></figure>
<p>blocks.txt 文件为全部物种的共线块信息，其中负号表示方向。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">...</span><br><span class="line">-666 -736 -1441 -781 -1435 -657 -1372 395 -607 -928 1461 -1333 ...</span><br><span class="line">-1003 1173 -1004 -990 -1271 1019 -897 1390 1033 671 1070 -1429 -1629</span><br><span class="line">...</span><br></pre></td></tr></table></figure>
<p>把 synteny.txt 以及 blocks.txt 文件拷贝到 Linux 服务器，同时修改 <a target="_blank" rel="noopener" href="http://processDrimm.py">processDrimm.py</a> 程序，然后运行就可以得到 IAGS 要求的 input files。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost processDRIMM]$ ./processDrimm.py</span><br><span class="line">-bash: ./processDrimm.py: /home/fenglei/local/app/anaconda3/envs/python37/bin/python3^M: bad interpreter: No such file or directory</span><br><span class="line">(python37) [fenglei@localhost processDRIMM]$ dos2unix processDrimm.py </span><br><span class="line">dos2unix: converting file processDrimm.py to Unix format ...</span><br><span class="line">(python37) [fenglei@localhost processDRIMM]$ ./processDrimm.py </span><br><span class="line">98</span><br><span class="line">98</span><br><span class="line"></span><br><span class="line">(python37) [fenglei@localhost processDRIMM]$ ll example_fl/drimm</span><br><span class="line">total 152</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  1094 Apr 22 16:09 Telongatum.block</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  1107 Apr 22 16:09 Sorghum.block</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  1101 Apr 22 16:09 Rice.block</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  2176 Apr 22 16:09 Maize.block</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  1111 Apr 22 16:09 Brachy.block</span><br><span class="line">-rw-r--r-- 1 fenglei fenglei 33056 Apr 22 16:03 blocks.txt</span><br><span class="line">-rw-r--r-- 1 fenglei fenglei 97352 Apr 22 16:03 synteny.txt</span><br></pre></td></tr></table></figure>
<p>此时打开 Rice.block，就是我们期望得到的格式！</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost drimm]$ cat Rice.block</span><br><span class="line">s -1629 -1628 1590 1591 1671 -1613 1648 -1657 -1660 -1659 -1654 -1550 -1566 -1577 -1582 -1583 1542 1578 -1558 1544 -1568 -1541 -1581 -1571 1631 1633 </span><br><span class="line">s 1553 1543 -1672 -1554 -1573 -1569 -1538 -1679 1625 -1697 -1684 1682 -1685 1658 -1696 1691 1627 1622 -1694 1592 -1677 1678 -1662 -1668 -1675 -1673 -1667 1666 -1664 -1655 -1551 </span><br><span class="line">s -1630 1632 1634 1636 1639 1643 1644 1572 1645 -1615 -1651 1623 -1602 1562 1580 1600 -1593 1621 1586 1609 1687 </span><br><span class="line">s 1625 -1697 -1684 1682 -1685 1658 -1696 1691 1627 1622 -1694 1592 -1677 1678 </span><br><span class="line">s 1631 1633 -1617 -1629 -1628 -1571 1581 1541 1568 -1544 1558 -1578 -1542 1583 1582 1577 1566 1550 1654 1659 1660 1657 -1648 1613 -1671 -1591 -1590 </span><br><span class="line">s 1551 1655 1664 -1666 1667 1673 1675 1668 1662 1679 1538 1569 1573 1554 1672 -1543 -1553 -1572 </span><br><span class="line">s 1586 1609 1687 1651 1615 </span><br><span class="line">s 1617 1595 -1603 1619 -1604 1656 -1607 -1608 -1610 1624 1606 1611 -1612 </span><br><span class="line">s 1595 -1603 1619 -1604 1656 -1607 -1608 -1610 1624 1606 1611 -1612 </span><br><span class="line">s 1634 1636 1639 -1644 -1643 1645 -1632 1630 </span><br><span class="line">s -1557 1560 1616 1597 -1559 1605 1601 </span><br><span class="line">s -1557 1560 1616 1597 -1559 1605 -1601 1623 -1602 1562 1580 1600 -1593 1621 </span><br></pre></td></tr></table></figure>
<h3 id="一个需要注意的地方"><a class="markdownIt-Anchor" href="#一个需要注意的地方"></a> 一个需要注意的地方</h3>
<p>drimm.sequence 一共 44 行，每行都是一个数列，所有物种的染色体数目加起来刚好是 44 行。但是这里文件里面是没有描述物种顺序的内容的。<br />
所以一定要注意，<a target="_blank" rel="noopener" href="http://processOrthofinder.py">processOrthofinder.py</a> 运行之后，有一个合并所有物种 xxx.sequence 文件的操作，一定要跟脚本中保持一致。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">### 下面内容摘取自：processOrthofinder.py</span><br><span class="line">sp = [&#x27;Brachy&#x27;,&#x27;Maize&#x27;,&#x27;Rice&#x27;,&#x27;Sorghum&#x27;,&#x27;Telongatum&#x27;]</span><br><span class="line">gff_list = [&#x27;Brachy.gff&#x27;,&#x27;Maize.gff&#x27;,&#x27;Rice.gff&#x27;,&#x27;Sorghum.gff&#x27;,&#x27;Telongatum.gff&#x27;]</span><br><span class="line">sp_ratio = [2,4,2,2,2]</span><br><span class="line"></span><br><span class="line">### 注意这里物种的顺序，需要跟上一步骤保持一致</span><br><span class="line">cat Brachy.sequence Maize.sequence Rice.sequence Sorghum.sequence Telongatum.sequence &gt; drimm.sequence</span><br><span class="line"></span><br><span class="line">### 下面内容摘取自：processDrimm.py</span><br><span class="line">chr_number = [5,10,12,10,7]</span><br><span class="line">sp_list = [&#x27;Brachy&#x27;,&#x27;Maize&#x27;,&#x27;Rice&#x27;,&#x27;Sorghum&#x27;,&#x27;Telongatum&#x27;]   ### 注意这里物种的顺序，需要跟上一步骤保持一致</span><br><span class="line">target_rate = &#x27;2:4:2:2:2&#x27;</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="cyclelengththreshold-参数测试"><a class="markdownIt-Anchor" href="#cyclelengththreshold-参数测试"></a> cycleLengthThreshold 参数测试</h3>
<p>之前将 cycleLengthThreshold 设置为 20，这次设定为 60 试试。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line">D:\InferAncestorGenome\processDRIMM\example_fl&gt;D:\InferAncestorGenome\processDRIMM\drimm\drimm\bin\Release\netcoreapp3.1\win-x64\drimm.exe</span><br><span class="line">Enter inputfile</span><br><span class="line">drimm.sequence</span><br><span class="line">Enter outdir</span><br><span class="line">D:\InferAncestorGenome\processDRIMM\example_fl\drimm</span><br><span class="line">Enter cycleLengthThreshold</span><br><span class="line">60</span><br><span class="line">Enter dustThreshold</span><br><span class="line">13</span><br><span class="line">Padding numb2684</span><br><span class="line">Nodes: 15893</span><br><span class="line">Number of edges: 38124</span><br><span class="line">Nodes:15757</span><br><span class="line">Number of edges: 21030</span><br><span class="line">Nodes:15446</span><br><span class="line">Number of edges: 18309</span><br><span class="line">Nodes:15106</span><br><span class="line">Number of edges: 17037</span><br><span class="line">Nodes:14781</span><br><span class="line">Number of edges: 16338</span><br><span class="line">Nodes:14416</span><br><span class="line">Number of edges: 15723</span><br><span class="line">Nodes:13923</span><br><span class="line">Number of edges: 15033</span><br><span class="line">Nodes:13519</span><br><span class="line">Number of edges: 14462</span><br><span class="line">Nodes:12991</span><br><span class="line">Number of edges: 13799</span><br><span class="line">Nodes:12805</span><br><span class="line">Number of edges: 13575</span><br><span class="line">Nodes:12713</span><br><span class="line">Number of edges: 13456</span><br><span class="line">Nodes:12635</span><br><span class="line">Number of edges: 13358</span><br><span class="line">Nodes:12624</span><br><span class="line">Number of edges: 13330</span><br><span class="line">Nodes:12609</span><br><span class="line">Number of edges: 13310</span><br><span class="line">Nodes:12599</span><br><span class="line">Number of edges: 13319</span><br><span class="line">Nodes:12801</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost processDRIMM]$ ./processDrimm.py</span><br><span class="line">27</span><br><span class="line">27</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/03/31/Bioinfo-MCScanX-genome-synteny-blocks/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2022/03/31/Bioinfo-MCScanX-genome-synteny-blocks/" class="post-title-link" itemprop="url">Bioinfo | MCScanX genome synteny blocks and 4dtv of paralogous genes</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-03-31 15:32:13" itemprop="dateCreated datePublished" datetime="2022-03-31T15:32:13+08:00">2022-03-31</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-04-02 23:28:56" itemprop="dateModified" datetime="2022-04-02T23:28:56+08:00">2022-04-02</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>原文记录于 2019-07-05，因 2022-04-01 再次测试，遇到问题，再次编辑记录。</p>
<p>安装如果遇到下面的问题：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">[fenglei@localhost MCScanX]$ make</span><br><span class="line">g++ struct.cc mcscan.cc read_data.cc out_utils.cc dagchainer.cc msa.cc permutation.cc -o MCScanX</span><br><span class="line">msa.cc: In function ‘void msa_main(const char*)’:</span><br><span class="line">msa.cc:289:22: error: ‘chdir’ was not declared in this scope</span><br><span class="line">if (chdir(html_fn)&lt;0)</span><br><span class="line">^</span><br><span class="line">make: *** [makefile:2: mcscanx] Error 1</span><br></pre></td></tr></table></figure>
<p>解决办法：</p>
<p>错误的原因是 MCScanX 不支持64位系统, 如果要在 64 位上运行, 需要在<code>msa.h</code>, <code>dissect_multiple_alignment.h</code>, and <code>detect_collinear_tandem_arrays.h</code>第一行加上</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">#include &lt;unistd.h&gt;</span><br></pre></td></tr></table></figure>
<p>如果要对两个基因组（以soybean和medicago为例）进行共线性分析，需要收集两个物种的数据编码基因的gff3文件和CDS序列文件，然后进行下面的操作。</p>
<p>首先利用gff3文件，得到每一个基因的起始和终止坐标信息，注意基因有多个transcripts，只需要保留“xxxx.1”的转录本，不同物种的序列用awk等工具灵活处理。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Gmax\_275\_Wm82.a2.v1.gene.gff3.gz -o soybean.bed</span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Mtruncatula\_285\_Mt4.0v1.gene.gff3 -o medicago.bed</span><br><span class="line"></span><br><span class="line">cat soybean.bed  awk &#x27;BEGIN&#123;OFS=&quot;\\t&quot;&#125;&#123;if(length($4) == 17 &amp;&amp; $4 ~ /\\.1$/)&#123;gsub(/\\.1$/,&quot;&quot;,$4);gsub(/Chr/,&quot;gm&quot;,$1); print $1,$4,$2,$3&#125;&#125;&#x27; &gt; soybean\_gff.tab</span><br><span class="line">cat medicago.bed  awk &#x27;BEGIN&#123;OFS=&quot;\\t&quot;&#125;&#123;if(length($4) == 15 &amp;&amp; $4 ~ /\\.1$/)&#123;gsub(/\\.1/,&quot;&quot;,$4);gsub(/chr/,&quot;mt&quot;,$1); print $1,$4,$2,$3&#125;&#125;&#x27; &gt; medicago\_gff.tab</span><br></pre></td></tr></table></figure>
<p>得到gff格式文件如下。 soybean.gff</p>
<p><strong>注意</strong> gff 文件中别包含 contigs 和 scaffolds，最好只包含 chromosomes。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">gm01 Glyma.01G000100 27354 28320</span><br><span class="line">gm01 Glyma.01G000200 58974 67527</span><br><span class="line">gm01 Glyma.01G000300 67769 69968</span><br><span class="line">gm01 Glyma.01G000400 90151 95947</span><br><span class="line">gm01 Glyma.01G000500 90288 91197</span><br><span class="line">gm01 Glyma.01G000600 116093 127845</span><br><span class="line">gm01 Glyma.01G000700 143466 155573</span><br><span class="line">gm01 Glyma.01G000800 157029 157772</span><br><span class="line">gm01 Glyma.01G000900 170533 193342</span><br><span class="line">gm01 Glyma.01G001000 196255 201895</span><br></pre></td></tr></table></figure>
<p>medicago.gff</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">mt1 Medtr1g004930 688 7332</span><br><span class="line">mt1 Medtr1g004940 6523 7366</span><br><span class="line">mt1 Medtr1g004950 14513 15729</span><br><span class="line">mt1 Medtr1g004960 16282 18382</span><br><span class="line">mt1 Medtr1g006530 28972 29214</span><br><span class="line">mt1 Medtr1g004980 31972 32344</span><br><span class="line">mt1 Medtr1g004990 35909 40554</span><br><span class="line">mt1 Medtr1g006490 44429 46280</span><br><span class="line">mt1 Medtr1g006590 46547 50437</span><br><span class="line">mt1 Medtr1g006600 51360 54977</span><br></pre></td></tr></table></figure>
<p>随后对 CDS 序列文件进行整理，保留 primary transcripts 的序列。序列名称跟上面 gff 文件里面要一致。并灵活选用工具将 CDS 序列转换成蛋白序列，或者一开始就用蛋白序列也可以。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">awk &#x27;&#123;print $1&#125;&#x27; ../jcvi\_190704/Gmax\_275\_Wm82.a2.v1.cds.fa &gt; Gmax\_275\_Wm82.a2.v1.cds.fa\_simplified</span><br><span class="line">awk &#x27;&#123;print $1&#125;&#x27; ../jcvi\_190704/Mtruncatula\_285\_Mt4.0v1.cds.fa &gt; Mtruncatula\_285\_Mt4.0v1.cds.fa\_simplified</span><br><span class="line"></span><br><span class="line">fasta\_formatter -t -i Mtruncatula\_285\_Mt4.0v1.cds.fa\_simplified -o Mtruncatula\_285\_Mt4.0v1.cds.tab</span><br><span class="line">fasta\_formatter -t -i Gmax\_275\_Wm82.a2.v1.cds.fa\_simplified -o Gmax\_275\_Wm82.a2.v1.cds.tab</span><br><span class="line"></span><br><span class="line">cat Gmax\_275\_Wm82.a2.v1.cds.tab  awk &#x27;BEGIN&#123;OFS=&quot;\\t&quot;&#125;&#123;if(length($1) == 17 &amp;&amp; $1 ~ /\\.1$/)&#123;gsub(/\\.1$/,&quot;&quot;,$1); print &quot;\\&gt;&quot;$1&quot;\\n&quot;$2&#125;&#125;&#x27; &gt; soybean\_cds.fa</span><br><span class="line">cat Mtruncatula\_285\_Mt4.0v1.cds.tab  awk &#x27;BEGIN&#123;OFS=&quot;\\t&quot;&#125;&#123;if(length($1) == 15 &amp;&amp; $1 ~ /\\.1$/)&#123;gsub(/\\.1/,&quot;&quot;,$1); print &quot;\\&gt;&quot;$1&quot;\\n&quot;$2&#125;&#125;&#x27; &gt; medicago.cds.fa</span><br><span class="line">iTools Fatools CDS2Pep -InCDS soybean\_cds.fa -OutPut soybean\_pep.fa</span><br><span class="line">iTools Fatools CDS2Pep -InCDS medicago.cds.fa -OutPut medicago.pep.fa</span><br></pre></td></tr></table></figure>
<p>然后用 blastp 对两个物种的蛋白序列进行比对，并用 MCScanX 命令检测共线性区间。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">makeblastdb -in medicago.pep.fa -dbtype prot -out index/mt -parse\_seqids</span><br><span class="line">## diamond makedb --in medicago.pep.fa --db ./diamond\_index/mt</span><br><span class="line">## diamond blastp --query soybean\_pep.fa --evalue 1e-5 --outfmt 6 --threads 24 --db ./diamond\_index/mt.dmnd --out soybean\_medicago.blast</span><br><span class="line">blastp -query soybean\_pep.fa -db index/mt -out gm\_mt.blast -evalue 1e-5 -num\_threads 24 -outfmt 6 -num\_alignments 5</span><br><span class="line">mkdir mcdata</span><br><span class="line">cp gm\_mt.blast mcdata</span><br><span class="line">cat soybean.gff medicago.gff &gt; mcdata/gm\_mt.gff</span><br><span class="line">MCScanX ./mcdata/gm\_mt</span><br></pre></td></tr></table></figure>
<p>产生的线性文件（gm_mt.collinearity）如下，可以用于下游画图。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">\[fenglei@localhost gm\_mt\_test\]$ head -n 50 gm\_mt.collinearity </span><br><span class="line">############### Parameters ###############</span><br><span class="line"># MATCH\_SCORE: 50</span><br><span class="line"># MATCH\_SIZE: 5</span><br><span class="line"># GAP\_PENALTY: -1</span><br><span class="line"># OVERLAP\_WINDOW: 5</span><br><span class="line"># E\_VALUE: 1e-05</span><br><span class="line"># MAX GAPS: 25</span><br><span class="line">############### Statistics ###############</span><br><span class="line"># Number of collinear genes: 51572, Percentage: 49.62</span><br><span class="line"># Number of all genes: 103927</span><br><span class="line">##########################################</span><br><span class="line">## Alignment 0: score=1008.0 e\_value=1.4e-69 N=23 gm01&amp;mt1 plus</span><br><span class="line">0- 0: Glyma01G232800 Medtr1g103160 0</span><br><span class="line">0- 1: Glyma01G234900 Medtr1g103450 8e-39</span><br><span class="line">0- 2: Glyma01G235100 Medtr1g103490 0.0001</span><br><span class="line">0- 3: Glyma01G236800 Medtr1g103500 2e-87</span><br><span class="line">0- 4: Glyma01G237000 Medtr1g103550 9e-47</span><br><span class="line">0- 5: Glyma01G237300 Medtr1g103570 2e-45</span><br><span class="line">0- 6: Glyma01G237400 Medtr1g103600 1e-144</span><br><span class="line">0- 7: Glyma01G237900 Medtr1g103690 9e-39</span><br><span class="line">0- 8: Glyma01G238200 Medtr1g103830 0</span><br><span class="line">0- 9: Glyma01G238400 Medtr1g104500 1e-170</span><br><span class="line">0- 10: Glyma01G238500 Medtr1g104520 6e-33</span><br><span class="line">0- 11: Glyma01G238700 Medtr1g104680 0</span><br><span class="line">0- 12: Glyma01G238800 Medtr1g104750 0</span><br><span class="line">0- 13: Glyma01G238900 Medtr1g104800 2e-80</span><br><span class="line">0- 14: Glyma01G239200 Medtr1g104870 0</span><br><span class="line">0- 15: Glyma01G239400 Medtr1g104930 1e-50</span><br><span class="line">0- 16: Glyma01G239800 Medtr1g105075 1e-32</span><br><span class="line">0- 17: Glyma01G239900 Medtr1g105305 5e-101</span><br><span class="line">0- 18: Glyma01G240000 Medtr1g105415 3e-91</span><br><span class="line">0- 19: Glyma01G240500 Medtr1g105555 5e-62</span><br><span class="line">0- 20: Glyma01G240600 Medtr1g105570 6e-06</span><br><span class="line">0- 21: Glyma01G240800 Medtr1g105650 2e-175</span><br><span class="line">0- 22: Glyma01G240900 Medtr1g105655 4e-164</span><br><span class="line">## Alignment 1: score=741.0 e\_value=1e-39 N=16 gm01&amp;mt1 plus</span><br><span class="line">1- 0: Glyma01G084800 Medtr1g093020 0</span><br><span class="line">1- 1: Glyma01G086700 Medtr1g093080 1e-138</span><br><span class="line">1- 2: Glyma01G086800 Medtr1g093095 2e-108</span><br><span class="line">1- 3: Glyma01G087500 Medtr1g093600 5e-175</span><br><span class="line">1- 4: Glyma01G087700 Medtr1g093630 0</span><br><span class="line">1- 5: Glyma01G088000 Medtr1g093650 8e-07</span><br><span class="line">1- 6: Glyma01G088200 Medtr1g093670 9e-115</span><br><span class="line">1- 7: Glyma01G088300 Medtr1g093750 1e-118</span><br><span class="line">1- 8: Glyma01G089100 Medtr1g093770 0</span><br><span class="line">1- 9: Glyma01G089300 Medtr1g093850 0</span><br><span class="line">1- 10: Glyma01G089900 Medtr1g093860 0</span><br><span class="line">1- 11: Glyma01G090200 Medtr1g093900 0</span><br><span class="line">1- 12: Glyma01G090400 Medtr1g093920 9e-22</span><br><span class="line">1- 13: Glyma01G091000 Medtr1g094010 0</span><br></pre></td></tr></table></figure>
<p>注意画图的配置文件不要出错，双斜杠（//）表示注释，注释号前面用tab符号跟染色体名称分隔，或者干脆去掉注释，我删掉注释才顺利画图，否则报错。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">[fenglei@localhost gm\_mt\_test\]$ cat dot.ctl</span><br><span class="line">2000</span><br><span class="line">2000</span><br><span class="line">gm01,gm02,gm03,gm04,gm05,gm06,gm07,gm08,gm09,gm10,gm11,gm12,gm13,gm14,gm15,gm16,gm17,gm18,gm19,gm20</span><br><span class="line">mt1,mt2,mt3,mt4,mt5,mt6,mt7,mt8</span><br></pre></td></tr></table></figure>
<p>画图命令。注意 dot_plotter.java 与 dot_plotter .class 要在一个文件夹下面。其他环状图做法也类似。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cd /PATH_TO_MCScanX/downstream_analyses</span><br><span class="line">DATA_PATH=/xx/my/mcdata  # 根据实际情况</span><br><span class="line">java dot_plotter -g $DATA_PATH/gm_mt.gff -s $DATA_PATH/gm_mt.collinearity -c $DATA_PATH/dot.ctl -o gm_mt.test.png</span><br></pre></td></tr></table></figure>
<p>画图结果：</p>
<p><img src="https://genehub.files.wordpress.com/2019/07/gm_mt.test_.png" alt="gm_mt.test" /></p>
<p>与文献（ Wang Jinpeng et al., 2017, Plant Physiol.）中的图片对比，基本是一样的效果。</p>
<p><img src="https://genehub.files.wordpress.com/2019/07/mt4_gm.order_.png" alt="mt4_gm.order" /></p>
<p>这个 MCScanX 共线性图跟更新开发的 jcvi 相比，画质要差一些。下面是 jcvi 的结果。</p>
<p><img src="https://i.imgur.com/Ly4FSpT.png" alt="jcvi 绘制的基因组共线性图实例" /></p>
<p>MCScanX 还可以将图画成圆形。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java circle_plotter -g /home/fenglei/projects/genome_synteny_20190704/MCScanX_soybean_medicago_190705/gm_mt_test/MADS-master/MADS.gff -s /home/fenglei/projects/genome_synteny_20190704/MCScanX_soybean_medicago_190705/gm_mt_test/MADS-master/MADS.collinearity -c /home/fenglei/projects/genome_synteny_20190704/MCScanX_soybean_medicago_190705/gm_mt_test/MADS-master/circle.ctl -o MADS.png</span><br></pre></td></tr></table></figure>
<p><strong>注意</strong> circle_plotter 的 dot.ctl 文件与 dot_plotter 的 dot.ctl 文件不一样。</p>
<p>For circle_plotter , put chromosome IDs in one line in dot.ctl file.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">2000</span><br><span class="line">Gm01,Gm02,Gm03,Gs01,Gs02,Gs03</span><br></pre></td></tr></table></figure>
<p>For dot_plotter, put chromosome IDs in two lines in dot.ctl file</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">2000</span><br><span class="line">2000</span><br><span class="line">Gm01,Gm02,Gm03</span><br><span class="line">Gs01,Gs02,Gs03</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ cat /home/fenglei/projects/genome_synteny_20190704/MCScanX_soybean_medicago_190705/gm_mt_test/MADS-master/circle.ctl</span><br><span class="line">1600    //plot width and height (in pixels)</span><br><span class="line">Chr1,Chr2,Chr3,Chr4,Chr5,Chr6,Chr7,Chr8,Chr9,Chr10,Chr11,Chr12,Chr13,Chr14,Chr15,Chr16,Chr17    //chromosomes in the circle</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/lRwZIhA.png" alt="" /></p>
<h3 id="mcscanx-与-jcvi-的比较"><a class="markdownIt-Anchor" href="#mcscanx-与-jcvi-的比较"></a> MCScanX 与 jcvi 的比较</h3>
<p>在基因组进化分析的 paralogous genes 鉴定中，MCScanX 与 jcvi 都可以用，有什么区别？从下面两幅图来看，jcvi 绘制的图片更有观赏性，而且 jcvi 能识别出这是一个物种内部（intra-genomic comparison）的 paralogous genes。</p>
<p>MCScanX：2019年基于a2v1的基因组数据。一共 1112 synteny blocks，35094 个两两配对。Number of collinear genes: 37190, Percentage: 66.90，Number of all genes: 55589。</p>
<p><img src="https://i.imgur.com/cAzbMcg.png" alt="MCScanX 分析 soybean 基因组内部的 synteny blocks" /></p>
<p>jcvi：2021年基于a4v1的基因组数据。一共 472 synteny blocks，18439个两两配对。</p>
<p><img src="https://i.imgur.com/6P2ig8r.jpeg" alt="jcvi 分析 soybean 基因组内部的 synteny blocks" /></p>
<p>看来 MCScanX 的结果更全面？以基因 <code>Glyma.01G210400</code> 为例，通过 MCScanX 可以找到另外三个旁系同源基因（paralogous），而通过 jcvi 只找到一个旁系同源基因。难怪在 2021 年的时候测试 4dtv 计算的时候，我发现 MCScanX 的结果更适合做 4dtv 计算，原因是它鉴定的 paralogous 更合理。当然，也有可能是 jcvi 的参数设计原因，我用的默认参数，没有进行额外设置。</p>
<p>MCScanX：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">(base) [fenglei@localhost mcdata]$ cat soybean.collinearity | grep &#x27;Glyma.01G210400&#x27;</span><br><span class="line"> 28- 43:        Glyma.01G210400 Glyma.05G058100       0</span><br><span class="line"> 71-432:        Glyma.01G210400 Glyma.11G031600   5e-82</span><br><span class="line"> 88-129:        Glyma.01G210400 Glyma.17G140600       0</span><br></pre></td></tr></table></figure>
<p>jcvi：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">(base) [fenglei@localhost gmo-gmo]$ cat gma.gma.anchors | grep &#x27;Glyma.01G210400&#x27;</span><br><span class="line">Glyma.01G210400.2       Glyma.17G140600.1       1030</span><br></pre></td></tr></table></figure>
<p><em>END</em></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/" class="post-title-link" itemprop="url">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-03-09 23:42:18" itemprop="dateCreated datePublished" datetime="2022-03-09T23:42:18+08:00">2022-03-09</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-04-01 13:50:20" itemprop="dateModified" datetime="2022-04-01T13:50:20+08:00">2022-04-01</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Proteomics/" itemprop="url" rel="index"><span itemprop="name">Proteomics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h3 id="一-蛋白质谱的背景知识"><a class="markdownIt-Anchor" href="#一-蛋白质谱的背景知识"></a> 一、蛋白质谱的背景知识</h3>
<p>LC-MS/MS</p>
<p>Liquid Chromatography-Mass Spectrometry（LC-MS/MS ，液相色谱-质谱串联）可用于残留化合物检测、有机小分子检测、鉴定和定量污染物以及在医药和食品领域添加剂检测和生物小分子等检测。。</p>
<p>LC-MS/MS一般包含五个步骤：</p>
<p>样本制备；<br />
样本分离：使用液相色谱方法分离；<br />
质谱上机：离子化、LUMOS原理、采集模式（DDA、DIA、SRM/PRM）；<br />
质谱鉴定：谱图格式（Raw、MzXML、MGF）、搜库、打分算法、FDR质控、蛋白推断；<br />
生信分析：谱图校验和下游数据统计分析。</p>
<p><img src="https://zouhua.top/images/Omics_Proteomics_01/LCMSMS_concept.jpg" alt="" /></p>
<p><em>液相色谱（HPLC）分离肽段的原理是什么？</em> 色谱法的分离原理是：溶于流动相(mobile phase)中的各组分经过固定相时，由于与固定相(station phase)发生作用(吸附、分配、排阻、亲和)的大小、强弱不同，在固定相中滞留时间不同，从而先后从固定相中流出。又称为色层法、层析法。（<a target="_blank" rel="noopener" href="https://www.sohu.com/a/471080306_121044873">参考资料</a>）</p>
<p><em>为什么质谱上机之前必须去盐？</em> 上机前的最后一步就是要保证多肽经过脱盐纯化，脱盐不充分不仅会影响目的肽段的鉴定，而且容易损坏仪器。此时可使用反相（RP）树脂来去除盐和缓冲液，其中C18基质是用于捕获疏水肽的理想基质。多肽在高水相流动相中与反相柱结合，同时盐和缓冲液被洗去，然后用高有机相流动相洗脱多肽。Pierce™ C18系列不仅提供了适合大体积样本的离心柱（column），还提供了适合小体积样本的吸头（tip）。（<a target="_blank" rel="noopener" href="http://www.ebiotrade.com/newsf/2019-7/201972164022321.htm">参考资料</a>）</p>
<p><em>离子化（Ionization）是什么？</em> 只有分子转化为气相离子后，质谱仪才能测量其质量。为了达到这一目的，质谱仪使分子带上电荷，然后将带电离子流转化为数据系统能够识别的成比例电流。数据系统将这一电流转化为数字信息，得到质谱图。（<a target="_blank" rel="noopener" href="https://www.waters.com/waters/zh_CN/What-is-MS-and-How-does-it-Work%3F/nav.htm?locale=zh_CN&amp;cid=10073253">参考资料</a>）离子源的性能决定了离子化效率，很大程度上决定了质谱仪的灵敏度。常见的离子化方式有两种：一种是样品在离子源中以气体的形式被离子化，另一种为从固体表面或溶液中溅射出带电离子。在很多情况下进样和离子化同时进行。比如，电子轰击电离(EI)是指气化后的样品分子进入离子化室后，受到由钨或铼灯丝发射并加速的电子流的轰击产生正离子。离子化室压力保持在10-4～10-6mmHg。轰击电子的能量大于样品分子的电离能，使样品分子电离或碎裂。电子轰击质谱能提供有机化合物最丰富的结构信息，有较好的重现性，其裂解规律的研究也最为完善，已经建立了数万种有机化合物的标准谱图库可供检索。其缺点在于不适用于难挥发和热稳定性差的样品。还有其他离子化方法，这里未摘录。（<a target="_blank" rel="noopener" href="https://shiyanjia.com/knowledge/articleinfo-1611.html">参考资料</a>）</p>
<h3 id="一-使用-maxquant-软件对-label-free-lc-msms-数据进行定量分析"><a class="markdownIt-Anchor" href="#一-使用-maxquant-软件对-label-free-lc-msms-数据进行定量分析"></a> 一、使用 MaxQuant 软件对 label free LC-MS/MS 数据进行定量分析</h3>
<p>质谱数据下机之后，需要用软件来进行搜库（根据物种的蛋白质数据库来进行序列比对检索），有多种软件可以选择，以 Orbitrap 下机数据为例，可以用 ProteinDiscover 软件里面调用相应的搜索引擎进行分析，这里不做阐述。下面主要介绍 MaxQuant，一方面是免费，另一方面操作界面人性化，上手快。</p>
<p>2016年，德国马普所的 Cox 和蛋白质组学领域巨擘 Matthias Mann 合作开发了 MaxQuant 软件，并发表在 nbt 上，protocol 也相应发表在 nature protocols。不足五年，引用已高达上万次，其中不乏CNS级别文章。毫不夸张地说，MaxQuant 是业界良心。（摘自：<a target="_blank" rel="noopener" href="https://www.cnblogs.com/jessepeng/p/13580371.html%EF%BC%89">https://www.cnblogs.com/jessepeng/p/13580371.html）</a></p>
<p>主要特点：</p>
<p>支持多种质谱仪厂商产生原始数据，如 raw/wiff；<br />
免费不开源，基于 Andromeda 搜索引擎；；<br />
支持标记定量和非标定量，如 labelfree（优势）/ICAT/SILAC/TMT/iTRAQ 等；<br />
蛋白鉴定数多，定量准确性高（主要是因为非线性质量校正和 Match Between Runs）；<br />
有较完整的结果查看和分析界面；<br />
主要部署在 Windows 系统（Linux 需要 mono，MacOS 需借助 Parallels 或 Bootcamp）；<br />
有配套的结果后处理软件 Perseus。</p>
<p><em>下载安装</em></p>
<p>官网：<a target="_blank" rel="noopener" href="https://www.maxquant.org/%EF%BC%8C%E6%8C%89%E7%85%A7%E5%AE%98%E6%96%B9%E6%96%87%E6%A1%A3%E8%BF%9B%E8%A1%8C%E5%AE%89%E8%A3%85%E3%80%82MQ%E7%9A%84%E4%B8%8B%E8%BD%BD%E9%9C%80%E8%A6%81%E7%94%A8%E9%82%AE%E7%AE%B1%E6%B3%A8%E5%86%8C%EF%BC%8C%E5%B9%B6%E5%A1%AB%E5%86%99%E6%B3%A8%E5%86%8C%E7%A0%81%E3%80%82%E4%BE%9D%E8%B5%96.NET">https://www.maxquant.org/，按照官方文档进行安装。MQ的下载需要用邮箱注册，并填写注册码。依赖.NET</a> Framework（基于.net框架编写，有些没安装的电脑需要安装） 和 MSFileReader（识别质谱raw文件）。解压后无需安装，点击文件目录下的 MaxQuant.exe 即可使用。</p>
<p><em>配置与运行</em></p>
<p>MaxQuant使用很简单，参数界面设置也不像PD那么凌乱。我的系统是win 11 中文版，软件的菜单有一部分成了中文。</p>
<p><img src="https://i.imgur.com/Xaw0jTc.jpeg" alt="" /></p>
<p>6 大参数配置模块，我这里只选一些重要的参数，大部分默认就好：</p>
<ol>
<li>Raw files：导入原始数据和设计实验，包括组分和样本，Group（指不同的实验，如label/SILAC同时进行即为2个Group0和Group1，一般跑一个就好）；</li>
<li>Group-specific parameters：Group参数的设置，一般只是Group0；（忘了在哪里看到这么一句话：“Type” will usually be “Standard” (obviously), and the “Multiplicity” will be 1 for label-free quantification, 2 if you have light and heavy labels, and 3 if you have light, medium, and heavy.）</li>
<li>Global parameters：全局参数配置；</li>
<li>Performance：显示正在运行的具体步骤和动态；</li>
<li>Viewer：数据可视化模块， 能在labelfree定量中呈现三维图形。我认为没啥用，无需设置，拖慢速度；</li>
<li>Andromeda configuration：用于配置修饰、酶，以及数据库ID规则设置等，一般无额外需求，无需设置。</li>
</ol>
<p>所以，实际需要配置的就是前面三项，以 Labelfree 定量为例，写下 Group-specific parameters 和 Global parameters 中一些重要的参数：</p>
<p><img src="https://i.imgur.com/V0CLPDb.jpeg" alt="" /></p>
<p>备注：我在实际操作中遇到的一个问题，在“组特定参数配置”菜单里，Label-free quantification 里面的 LFQ 选上的话，点击开始分析的时候会报错“LFQ requires at least two experiments”。</p>
<p><img src="https://i.imgur.com/aSnmScM.jpeg" alt="" /></p>
<p>后来我发现，在“全局参数”菜单下面，也有一个 Label-free quantification 栏目。这里也有 LFQ 字样，不过我并没有测试。这里的 iBAQ 我是选择了的，跑完数据能够得到 iBAQ 值。</p>
<p><img src="https://i.imgur.com/RpMgWls.jpeg" alt="" /></p>
<p>一篇文献中提到的使用 maxquant 进行蛋白组数据分析：</p>
<p><img src="https://i.imgur.com/dekTX7j.png" alt="" /></p>
<p>搜库完成之后，在raw文件路径下，会生成一个combined目录，里面就有结果文件：proteinGroups.txt，可以用excel打开这个文件，并另存为xlsx格式，这个文件里面就有各个蛋白的iBAQ值。</p>
<p>在一个实践项目中，同一批数据分别用 PD 和 maxquant 跑，出来的结果大体一致，但是想让 maxquant 显示更多结果，如何调整参数？如下图，我将 FDR 从 0.01 调整到 0.05，另外 min pipetides 从1调整到0，这样出来的结果要从一千多个蛋白变成两千多个蛋白。</p>
<p><img src="https://i.imgur.com/RfL4mwM.jpeg" alt="" /></p>
<p><img src="https://i.imgur.com/FYZpr0W.jpeg" alt="" /></p>
<p>combined文件夹包含了搜库结果，如果是DIA建库，导入的即是这个文件夹。里面包含了txt文件夹，即鉴定和定量的全部结果。</p>
<p>mqpar.xml是所有参数配置文件，若二次搜库，可直接导入MaxQuant中，无需重复设置。所以，一般建议保留下来，这样也可对结果的设置进行追溯。</p>
<h3 id="二-蛋白质组定量结果之组间差异分析"><a class="markdownIt-Anchor" href="#二-蛋白质组定量结果之组间差异分析"></a> 二、蛋白质组定量结果之组间差异分析</h3>
<p>查资料看到的，可以用 Perseus 进行后处理。Perseus也是上面两位大佬开发的软件，发在nature methods上。一开始是专门针对MQ数据处理的。后来对所有矩阵类型的表达数据，甚至是多组学数据都适合。</p>
<p>实际操作中，我要对两组蛋白组学样品进行差异分析，两组各有4个生物学重复。</p>
<p>在知乎看到一篇文章总结了三个工具：limma，DEP 和 DEqMS。（参考：<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/448479536%EF%BC%89">https://zhuanlan.zhihu.com/p/448479536）</a></p>
<p>其中 DEP 可以直接以 maxquant 的结果文件（ProteinGroups.txt）作为输入，而且分析之后的结果有命令可以直接画图，真是值得试试。</p>
<p>同时，我看到另一篇文章（<a target="_blank" rel="noopener" href="https://www.bioinfo-scrounger.com/archives/541/%EF%BC%89%E4%B8%AD%EF%BC%8C%E4%BD%9C%E8%80%85%E8%AE%A4%E4%B8%BA">https://www.bioinfo-scrounger.com/archives/541/）中，作者认为</a> limma 使用结果也不错的。</p>
<p>良心作者，甚至还提供一个label-free MS的数据矩阵文件：<a target="_blank" rel="noopener" href="http://www.bioinfo-scrounger.com/data/labelfree.csv">http://www.bioinfo-scrounger.com/data/labelfree.csv</a></p>
<p>下面摘抄该作者使用 limma 对两组蛋白质谱数据进行差异分析：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"># 如果就测试数据两组比较而言，limma需要先做log2转化，然后设定分组矩阵，最后用线性模型拟合再用Empirical Bayes test求出检验的P值</span><br><span class="line"></span><br><span class="line">library(limma)</span><br><span class="line"></span><br><span class="line">df1 &lt;- log2(df + 1)</span><br><span class="line">group_list &lt;- factor(c(rep(&quot;A&quot;,3), rep(&quot;B&quot;,3)))</span><br><span class="line">design &lt;- model.matrix(~group_list)</span><br><span class="line">colnames(design) &lt;- levels(group_list)</span><br><span class="line">rownames(design) &lt;- colnames(df1)</span><br><span class="line"></span><br><span class="line">fit &lt;- lmFit(df1, design)</span><br><span class="line">fit &lt;- eBayes(fit, trend=TRUE)</span><br><span class="line">result_limma &lt;- topTable(fit, coef=2,n=Inf)</span><br><span class="line">&gt;sum(result_limma$P.Value &lt; 0.05 &amp; abs(result_limma$logFC) &gt; 1)</span><br><span class="line">[1] 952</span><br><span class="line"># Limma的差异蛋白数目就相比以上就略多了，满足阈值的有952个</span><br></pre></td></tr></table></figure>
<p>下面是我的测试过程，并且用自己的蛋白组数据做了组间差异分析。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br></pre></td><td class="code"><pre><span class="line">### 蛋白质组学差异分析的博文：《浅谈蛋白组的差异蛋白分析》https://www.bioinfo-scrounger.com/archives/541/</span><br><span class="line">###  这篇文章也差不多：《蛋白质组差异分析的三个R包》 https://zhuanlan.zhihu.com/p/448479536</span><br><span class="line"></span><br><span class="line">##############################################################</span><br><span class="line"></span><br><span class="line">#测试数据下载：labelfree.csv (http://www.bioinfo-scrounger.com/data/labelfree.csv)</span><br><span class="line"></span><br><span class="line">#首先做一步过滤处理，以50%原则过滤掉缺失值较多的Protein，这步主要针对后续T检验和limma的预处理，而MSstats则有其默认的预处理方式</span><br><span class="line"></span><br><span class="line">library(ggplot2)</span><br><span class="line">library(dplyr)</span><br><span class="line"></span><br><span class="line">#data &lt;- read.table(file = &quot;labelfree.csv&quot;, sep = &quot;,&quot;, header = T, row.names = 1)</span><br><span class="line">data &lt;- read.table(file = &quot;labelfree-edited.txt&quot;, sep = &quot;\t&quot;, header = T, row.names = 1)</span><br><span class="line">rm &lt;- apply(data, 1, function(x)&#123;  # 这个函数的功能是去掉：超过50%的样品的表达值为0的样品</span><br><span class="line">  sum(x == 0) &gt; 3</span><br><span class="line">&#125;)</span><br><span class="line">df &lt;- data[!rm,]</span><br><span class="line">#&gt;nrow(df)</span><br><span class="line">#[1] 5325</span><br><span class="line"></span><br><span class="line">##################################################################</span><br><span class="line"></span><br><span class="line">#接着则是做log转化，这步有没有必要呢？Label-free技术是用峰谱积分计算丰度的，其丰度的数量级在10的7次到10的11次不等，从下面每个样本的均值和标准差可看出，丰度值数据分布非常离散</span><br><span class="line"></span><br><span class="line">Mean &lt;- apply(df, 2, mean)</span><br><span class="line">SD &lt;- apply(df, 2, sd)</span><br><span class="line">#&gt; Mean</span><br><span class="line">#A1         A2         A3         B1         B2         B3 </span><br><span class="line">#2366243926 2351546722 2165457844 2175212680 2708233812 2290120209</span><br><span class="line">#&gt; SD</span><br><span class="line">#A1          A2          A3          B1          B2          B3 </span><br><span class="line">#13098498790 14392930144 14179115278 11151997873 12092333164 10344859562</span><br><span class="line">#那么做log转化有什么用呢，一般作用为消除异方差；减少自变量数量级不一致的情况；使非线性变量关系转化为线性关系；最主要的还是为了使数据的呈现方式接近我们所希望的假设，从而更好的进行统计推断（以上来自知乎）</span><br><span class="line"></span><br><span class="line">#下面左边是原始数据AB两组的散点图，右边是做了对数转化后的</span><br><span class="line"></span><br><span class="line">#---------------</span><br><span class="line">#   T检验</span><br><span class="line">#---------------</span><br><span class="line">#在做Two-Sample T检验前，需要样本总体符合正态分布，并且方差齐性，以测试数据为例（未做对数转化），先做个F检验</span><br><span class="line"></span><br><span class="line">fvalue &lt;- apply(df, 1, function(x)&#123;</span><br><span class="line">  a &lt;- factor(c(rep(&quot;A&quot;,3), rep(&quot;B&quot;,3)))</span><br><span class="line">  var.test(x~a)</span><br><span class="line">&#125;)</span><br><span class="line">sum(unlist(lapply(fvalue, function(x)&#123;</span><br><span class="line">  x$p.value &gt; 0.05</span><br><span class="line">&#125;)))</span><br><span class="line">#[1] 4605</span><br><span class="line">#可看出5325个蛋白有4605个是方差齐性的</span><br><span class="line"></span><br><span class="line">#接下来用t.test函数做成组T检验，双尾（默认），方差不齐性的用var.equal = FALSE</span><br><span class="line"></span><br><span class="line">pvalue &lt;- apply(df, 1, function(x)&#123;</span><br><span class="line">  a &lt;- factor(c(rep(&quot;A&quot;,3), rep(&quot;B&quot;,3)))</span><br><span class="line">  fvalue &lt;- var.test(x~a)</span><br><span class="line">  if (fvalue$p.value &gt; 0.05)&#123;</span><br><span class="line">    t.test(x~a, var.equal = T)</span><br><span class="line">  &#125;else&#123;</span><br><span class="line">    t.test(x~a, var.equal = F)</span><br><span class="line">  &#125;</span><br><span class="line">&#125;)</span><br><span class="line"></span><br><span class="line">result_ttest &lt;- data.frame(ID=row.names(df), </span><br><span class="line">                           Pvalue = as.numeric(unlist(lapply(pvalue, function(x) x$p.value))),</span><br><span class="line">                           log2FC = log2(as.numeric(unlist(lapply(pvalue, function(x) x$estimate[1]/x$estimate[2]))))</span><br><span class="line">)</span><br><span class="line">sum(result_ttest$Pvalue &lt; 0.05 &amp; abs(result_ttest$log2FC) &gt; 1)</span><br><span class="line">#[1] 704</span><br><span class="line">#可看出差异蛋白数目为704个；如果是预处理时做了对数转化的话，代码同上（其中有略微区别，logFC是作相减，而不是上面代码的相除），结果符合相同的阈值的差异蛋白有694个</span><br><span class="line"></span><br><span class="line">#---------------</span><br><span class="line">     Limma</span><br><span class="line">#---------------</span><br><span class="line">#Limma就比较熟悉了，如果就测试数据两组比较而言，limma需要先做log2转化，然后设定分组矩阵，最后用线性模型拟合再用Empirical Bayes test求出检验的P值</span><br><span class="line">library(limma)</span><br><span class="line">df1 &lt;- log2(df + 1)</span><br><span class="line">group_list &lt;- factor(c(rep(&quot;A&quot;,3), rep(&quot;B&quot;,3)))</span><br><span class="line">design &lt;- model.matrix(~group_list)     ### 这是什么函数？？？</span><br><span class="line">colnames(design) &lt;- levels(group_list)</span><br><span class="line">rownames(design) &lt;- colnames(df1)</span><br><span class="line">fit &lt;- lmFit(df1, design)</span><br><span class="line">fit &lt;- eBayes(fit, trend=TRUE)</span><br><span class="line">result_limma &lt;- topTable(fit, coef=2,n=Inf)</span><br><span class="line">sum(result_limma$P.Value &lt; 0.05 &amp; abs(result_limma$logFC) &gt; 1)</span><br><span class="line">#[1] 952</span><br><span class="line">#Limma的差异蛋白数目就相比以上就略多了，满足阈值的有952个</span><br><span class="line">write.table(result_limma, file = &quot;limma_output.xls&quot;, sep = &quot;\t&quot;)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">################################################################################</span><br><span class="line">############</span><br><span class="line">############   下面用自己的数据调用limma做差异分析</span><br><span class="line">############</span><br><span class="line">################################################################################</span><br><span class="line"></span><br><span class="line"># 用Perl写脚本，将不同样品的ProteinGroups.txt文件一起合并成一个表达谱矩阵。第一列为基因名，其余各列为表达值。类似于RNA Seq的表达矩阵。</span><br><span class="line"></span><br><span class="line">#首先做一步过滤处理，以50%原则（可以自己调整）过滤掉缺失值较多的Protein，这步主要针对后续T检验和limma的预处理，而MSstats则有其默认的预处理方式</span><br><span class="line"></span><br><span class="line">library(ggplot2)</span><br><span class="line">library(dplyr)</span><br><span class="line"></span><br><span class="line">#data &lt;- read.table(file = &quot;labelfree.csv&quot;, sep = &quot;,&quot;, header = T, row.names = 1)</span><br><span class="line">data &lt;- read.table(file = &quot;merged_iBAQ_table_v3_input_for_R.txt&quot;, sep = &quot;\t&quot;, header = T, row.names = 1)</span><br><span class="line">rm &lt;- apply(data, 1, function(x)&#123;  # 这个函数的功能是去掉：超过50%的样品的表达值为0的样品</span><br><span class="line">  sum(x == 0) &gt; 7</span><br><span class="line">&#125;)</span><br><span class="line">df &lt;- data[!rm,]</span><br><span class="line">#&gt;nrow(df)</span><br><span class="line">#[1] 5325</span><br><span class="line"></span><br><span class="line">library(limma)</span><br><span class="line">df1 &lt;- log2(df + 1)</span><br><span class="line">group_list &lt;- factor(c(rep(&quot;A&quot;,4), rep(&quot;B&quot;,4)))</span><br><span class="line">design &lt;- model.matrix(~group_list)     ### 这是什么函数？？？</span><br><span class="line">colnames(design) &lt;- levels(group_list)</span><br><span class="line">rownames(design) &lt;- colnames(df1)</span><br><span class="line"></span><br><span class="line">fit &lt;- lmFit(df1, design)</span><br><span class="line">fit &lt;- eBayes(fit, trend=TRUE)</span><br><span class="line">result_limma &lt;- topTable(fit, coef=2,n=Inf)</span><br><span class="line">sum(result_limma$P.Value &lt; 0.05 &amp; abs(result_limma$logFC) &gt; 1)</span><br><span class="line">#[1] 952</span><br><span class="line">#Limma的差异蛋白数目就相比以上就略多了，满足阈值的有952个</span><br><span class="line"></span><br><span class="line">write.table(result_limma, file = &quot;limma_output_FREE1.xls&quot;, sep = &quot;\t&quot;)</span><br><span class="line">### 如果下次重新打开R画图，则直接读入文件：</span><br><span class="line">#result_limma &lt;- read.table(&quot;limma_output_FREE1.xls&quot;, sep = &quot;\t&quot;, head=T)</span><br><span class="line"></span><br><span class="line">####</span><br><span class="line"># Volcano Plot</span><br><span class="line">#logFC AveExpr         t      P.Value   adj.P.Val         B</span><br><span class="line">#AT5G09660.4 -24.86099 12.4305 -13.67679 2.515145e-07 0.000174047 -1.497474</span><br><span class="line">#AT5G09660.3 -24.86099 12.4305 -13.67679 2.515145e-07 0.000174047 -1.497474</span><br><span class="line"></span><br><span class="line">#----------------     计算-logFC     ----------------</span><br><span class="line">#prop &lt;- apply(df, 2,function(x,y) (x/y)*100, df$Total)</span><br><span class="line">#x &lt;- apply(result_limma, 2, function(x) -x, result_limma$logFC)</span><br><span class="line">result_limma$x &lt;- -result_limma$logFC</span><br><span class="line">head(result_limma)</span><br><span class="line">#----------------    计算-log10(Pvalue)   -----------</span><br><span class="line">result_limma$y &lt;- -log10(result_limma$P.Value)</span><br><span class="line"></span><br><span class="line">#install.packages(&#x27;ggrepel&#x27;)</span><br><span class="line">library(ggrepel)</span><br><span class="line"></span><br><span class="line">library(ggplot2)</span><br><span class="line"></span><br><span class="line">####  Volcano Plot</span><br><span class="line"></span><br><span class="line">ggplot() +</span><br><span class="line">  #ggplot(data=a, aes(x=normalized_log2Fab, y=log2_mark_1_2_PrSM))</span><br><span class="line">  geom_hline(yintercept = 0, linetype=&quot;dashed&quot;, color = &quot;grey&quot;, size=1.5) +</span><br><span class="line">  geom_hline(yintercept = -log10(0.05), linetype=&quot;dashed&quot;, color = &quot;grey&quot;, size=1.5) +</span><br><span class="line">  geom_vline(xintercept = 0, linetype=&quot;dashed&quot;, color = &quot;grey&quot;, size=1.5) +</span><br><span class="line">  #geom_point(color=&quot;#9167a9&quot;, size=4, alpha=0.7)+</span><br><span class="line">  geom_point(data=filter(result_limma, P.Value&lt;=0.05 &amp; x&gt;0), aes(x=x, y=y), size = 2, color=&quot;red&quot;, alpha=0.2) +</span><br><span class="line">  geom_point(data=filter(result_limma, P.Value&lt;=0.05 &amp; x&lt;0), aes(x=x, y=y), size = 2, color=&quot;green&quot;, alpha=0.2) +</span><br><span class="line">  geom_point(data=filter(result_limma, P.Value&gt;0.05), aes(x=x, y=y), size = 2, color=&quot;darkblue&quot;, alpha=0.2) +</span><br><span class="line">  #scale_size(range = c(0.5, 10))  + </span><br><span class="line">  theme_bw() +</span><br><span class="line">  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = &quot;black&quot;)) +</span><br><span class="line">  #theme(axis.text.x = element_text(size = 15, color = &quot;black&quot;, face = &quot;bold&quot;, vjust = 0.5, hjust = 0.5, angle = 45)) +</span><br><span class="line">  theme(axis.text.x = element_text(size = 15, color = &quot;black&quot;, face = &quot;bold&quot;)) +</span><br><span class="line">  theme(axis.text.y = element_text(color = &quot;black&quot;, size = 15)) +</span><br><span class="line">  ylab(&quot;-log10(P value)&quot;) +</span><br><span class="line">  xlab(&quot;log10(Fold Change)&quot;) +</span><br><span class="line">  theme(axis.title.y = element_text(size = 18)) +</span><br><span class="line">  theme(axis.title.x = element_text(size = 18))</span><br><span class="line"></span><br></pre></td></tr></table></figure>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/" class="post-title-link" itemprop="url">Plant | Rice: Oryza sativa subsp. japonica and indica</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-10-31 19:25:40" itemprop="dateCreated datePublished" datetime="2021-10-31T19:25:40+08:00">2021-10-31</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-03-09 23:44:12" itemprop="dateModified" datetime="2022-03-09T23:44:12+08:00">2022-03-09</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Plant/" itemprop="url" rel="index"><span itemprop="name">Plant</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>本文总结粳稻（japonica）与籼稻（indica）的分类，以及相对应的数据库的事项。</p>
<p>作者：稻我家安全食材<br />
链接：<a target="_blank" rel="noopener" href="https://www.zhihu.com/question/28272395/answer/513511209">https://www.zhihu.com/question/28272395/answer/513511209</a><br />
来源：知乎<br />
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。</p>
<p>籼稻（Indica rice）：有 20% 左右为直链淀粉。属中黏性。籼稻起源于亚热带，种植于热带和亚热带地区，生长期短，在无霜期长的地方一年可多次成熟。去壳成为籼米后，外观细长、透明度低。有的品种表皮发红，如中国江西出产的红米，煮熟后米饭较干、松。通常用于萝卜糕、米粉、炒饭。</p>
<p>粳稻（Japonica rice）：粳稻的直链淀粉较少，低于15%。种植于温带和寒带地区，生长期长，一般一年只能成熟一次。去壳成为粳米后，外观圆短、透明（部分品种米粒有局部白粉质）。煮食特性介于糯米与籼米之间。用途为一般食用米。籼稻和粳稻是长期适应不同生态条件，尤其是温度条件而形成的两种气候生态型，两者在形态生理特性方面部有明显差异因为籼稻和粳稻产自不同的地方，米粒的形状也不同，口感上更是相差很远。</p>
<p>那么二者的区别又在哪里呢？</p>
<ol>
<li>
<p>首先在品种上籼稻是由野生稻演变来的，而粳稻是由籼稻在人们不断向高纬度、高海拔地区引种的过程中不断演变和人工选育而来的。因此说籼稻是基本型，粳稻是变异型。</p>
</li>
<li>
<p>从形态特征和经济性状上看，一般籼稻的茎杆较粗，分蘖力较强，叶色较淡，谷粒细长，容易落粒，出米率较低。籼米的直链淀粉含量高，煮饭时胀饭性大，黏性小，米饭散落。粳稻一般茎杆较细，传统粳稻品种的分蘖力不如籼稻，叶色较深，谷粒短圆，不容易落粒，出米率较高，碎米少。粳米的直链淀粉含量低，米饭黏性大，胀饭性小。</p>
</li>
<li>
<p>从生理特征和适应性上看，籼稻一般吸肥性强，而耐肥力差，易倒伏，耐寒力较差，温度在12℃以上时才能发芽。粳稻则耐肥力强而吸肥性差，较抗倒伏，耐寒力较强，温度达到10℃即可发芽。在温度适宜的条件下，籼稻叶片的光合速率高于粳稻，繁茂性好，易早生快发。</p>
</li>
<li>
<p>从地区分布上看，籼稻适于在低纬度、低海拔的湿热地区栽培，如我国的南方稻区。粳稻则适于在高纬度、高海拔地区栽培，如我国的东北稻区、华北稻区和西北稻区，也可在云贵高原的高海拔地区和江淮流域双季稻区作中粳和晚粳种植。</p>
</li>
<li>
<p>从稻株外部形态进行区别：籼稻一般株高多在1米以上，茎秆较软，叶片宽，色泽淡绿，剑叶开度小。叶片茸毛多，有短小茸毛散生于颖壳。大多为无芒或短芒，谷粒细长而稍扁平，一般长度为宽度的二倍以上。谷粒易脱落，较耐湿、耐热和耐强光，但不耐寒。粳稻一般株高多为75-95厘米，茎秆较矮，叶子较窄，深绿色，米粒不易脱落，耐寒性强。</p>
</li>
</ol>
<p>因此，籼稻是由野生稻演变成的栽培稻，是基本型。粳稻分布范围广泛，从南方的高寒山区，云贵高原到秦岭，淮河以北的广大的地区均有栽培。粳稻具有耐寒，耐弱光的习性，粒形短圆，米质粘性较强，叶面少毛或无毛，颖毛长密，不易落粒等特性，与野生稻有较大差异。因此，可以说粳稻是人类将籼稻由南向北，由低向高引种后，逐渐适应低温的变异型。</p>
<h3 id="籼稻indica9311的基因组信息"><a class="markdownIt-Anchor" href="#籼稻indica9311的基因组信息"></a> 籼稻（indica，9311）的基因组信息</h3>
<p>数据下载</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget ftp://ftp.genomics.org.cn/pub/ricedb/SynVs9311/9311/Sequence/Chromosome/ChrAll.fasta.gz</span><br><span class="line">wget ftp://ftp.genomics.org.cn/pub/ricedb/rice_update_data/GLEAN_genes/Beijing_indica/GLEAN_genes/9311.glean.gff.cds.gz</span><br><span class="line">wget ftp://ftp.genomics.org.cn/pub/ricedb/rice_update_data/GLEAN_genes/Beijing_indica/GLEAN_genes/9311.glean.gff.gff.gz</span><br><span class="line">hisat2-build -p 8  genome.fa genome</span><br><span class="line">gffread 9311.glean.gff.gff -F -T -o 9311.glean.raw.gtf</span><br><span class="line">perl ~/local/app/NGSQCToolkit/Statistics/N50Stat.pl -i ChrAll.fasta -o ChrAll.fasta.stat</span><br></pre></td></tr></table></figure>
<p>基因命名</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">##gff-version 3</span><br><span class="line">##sequence-region Chr01 1 47283185</span><br><span class="line">Chr01   glean   gene    47275570        47278635        .       -       .       ID=BGIOSGA000001;Name=BGIOSGA000001;</span><br><span class="line">Chr01   glean   mRNA    47275570        47278635        0.798504        -       .       ID=BGIOSGA000001-TA;Name=BGIOSGA000001-TA;Parent=BGIOSGA000001;</span><br><span class="line">Chr01   glean   CDS     47278477        47278635        .       -       0       Parent=BGIOSGA000001-TA;</span><br><span class="line">Chr01   glean   CDS     47278262        47278399        .       -       0       Parent=BGIOSGA000001-TA;</span><br><span class="line">Chr01   glean   CDS     47278135        47278186        .       -       0       Parent=BGIOSGA000001-TA;</span><br><span class="line">Chr01   glean   CDS     47277979        47278058        .       -       2       Parent=BGIOSGA000001-TA;</span><br><span class="line">Chr01   glean   CDS     47276528        47276587        .       -       0       Parent=BGIOSGA000001-TA;</span><br><span class="line">Chr01   glean   CDS     47275842        47275922        .       -       0       Parent=BGIOSGA000001-TA;</span><br><span class="line">Chr01   glean   CDS     47275570        47275662        .       -       0       Parent=BGIOSGA000001-TA;</span><br><span class="line">Chr01   glean   gene    47247293        47259724        .       -       .       ID=BGIOSGA000002;Name=BGIOSGA000002;</span><br><span class="line">Chr01   glean   mRNA    47247293        47259724        0.997143        -       .       ID=BGIOSGA000002-TA;Name=BGIOSGA000002-TA;Parent=BGIOSGA000002;</span><br><span class="line">Chr01   glean   CDS     47259551        47259724        .       -       0       Parent=BGIOSGA000002-TA;</span><br><span class="line">Chr01   glean   CDS     47259199        47259261        .       -       0       Parent=BGIOSGA000002-TA;</span><br><span class="line">Chr01   glean   CDS     47258441        47258518        .       -       0       Parent=BGIOSGA000002-TA;</span><br><span class="line">Chr01   glean   CDS     47258327        47258376        .       -       0       Parent=BGIOSGA000002-TA;</span><br><span class="line">Chr01   glean   CDS     47257303        47257405        .       -       1       Parent=BGIOSGA000002-TA;</span><br><span class="line">Chr01   glean   CDS     47256255        47256407        .       -       0       Parent=BGIOSGA000002-TA;</span><br><span class="line">Chr01   glean   CDS     47255968        47256060        .       -       0       Parent=BGIOSGA000002-TA;</span><br></pre></td></tr></table></figure>
<h3 id="粳稻japonica或称日本晴var-nipponbare的基因组信息"><a class="markdownIt-Anchor" href="#粳稻japonica或称日本晴var-nipponbare的基因组信息"></a> 粳稻（japonica，或称“日本晴”，var Nipponbare）的基因组信息</h3>
<p>下载地址 <a target="_blank" rel="noopener" href="https://rapdb.dna.affrc.go.jp/download/irgsp1.html">https://rapdb.dna.affrc.go.jp/download/irgsp1.html</a>，从下面下载的文件来看，该数据集的更新日期为2020年12月。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">drwxrwxr-x 2 fenglei fenglei      4096 Mar 10  2021 index</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   2205461 Mar 10  2021 IRGSP-1.0_representative_transcript_exon_2020-12-02.gtf.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   2910542 Mar 10  2021 IRGSP-1.0_representative_annotation_2020-12-02.tsv.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  15479652 Mar 10  2021 IRGSP-1.0_representative_2020-12-02.tar.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   8071768 Mar 10  2021 IRGSP-1.0_protein_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  12921645 Mar 10  2021 IRGSP-1.0_cds_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  21521204 Mar 10  2021 IRGSP-1.0_transcript_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  14082338 Mar 10  2021 IRGSP-1.0_1kb-upstream_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  13852984 Mar 10  2021 IRGSP-1.0_1kb-downstream_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  27531029 Mar 10  2021 IRGSP-1.0_2kb-upstream_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  40892583 Mar 10  2021 IRGSP-1.0_gene_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei     65536 Mar 10  2021 irgsp1_repeat_unit.gff.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  27067699 Mar 10  2021 IRGSP-1.0_2kb-downstream_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  40469323 Mar 10  2021 IRGSP-1.0_3kb-upstream_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  39804961 Mar 10  2021 IRGSP-1.0_3kb-downstream_2020-12-02.fasta.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 208876624 Mar 10  2021 kasalath_genome.tar.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    828064 Mar 10  2021 IRGSP-1.0_predicted_2020-12-02.tar.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    212118 Mar 10  2021 IRGSP-1.0_predicted_transcript_exon_2020-12-02.gtf.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei     65407 Mar 10  2021 IRGSP-1.0_predicted_annotation_2020-12-02.tsv.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   1624227 Mar 10  2021 IRGSP-1.0_predicted-protein_2020-12-02.fasta</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei     18236 Mar 10  2021 irgsp1_rRNA_tRNA.gff.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    323084 Mar 10  2021 RAP-MSU_2020-12-02.txt.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    339560 Mar 10  2021 IRGSP-1.0_transcript-evidence_2012-04-11.txt.gz</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   4589349 Mar 10  2021 IRGSP-1.0_predicted-cds_2020-12-02.fasta</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    174987 Mar 10  2021 organelle.gff</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    650061 Mar 10  2021 Mt_Pt_genome.fasta</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   8054917 Mar 10  2021 Rice_vegetative_and_reproductive_stage-specific_sRNAs_and_networks.zip</span><br><span class="line">-rwxr-xr-x 1 fenglei fenglei        51 Mar 10  2021 opt.sh</span><br></pre></td></tr></table></figure>
<p>看看他的基因命名</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">(base) [fenglei@localhost O.sativa_japonica_2021]$ less IRGSP-1.0_representative_transcript_exon_2020-12-02.gtf.gz</span><br><span class="line">chr01   irgsp1_rep      transcript      2983    10815   .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    2983    3268    .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    3354    3616    .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    4357    4455    .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    5457    5560    .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    7136    7944    .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    8028    8150    .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    8232    8320    .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    8408    8608    .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    9210    9615    .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    10102   10187   .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    10274   10430   .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      exon    10504   10815   .       +       .       gene_id &quot;Os01g0100100&quot;; transcript_id &quot;Os01t0100100-01&quot;;</span><br><span class="line">chr01   irgsp1_rep      transcript      11218   12435   .       +       .       gene_id &quot;Os01g0100200&quot;; transcript_id &quot;Os01t0100200-01&quot;;</span><br></pre></td></tr></table></figure>
<h3 id="粳稻另一个版本"><a class="markdownIt-Anchor" href="#粳稻另一个版本"></a> 粳稻另一个版本</h3>
<p>下载地址 <a target="_blank" rel="noopener" href="http://rice.uga.edu/">http://rice.uga.edu/</a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">RGAP 7 Summary</span><br><span class="line">Gene Annotation</span><br><span class="line"></span><br><span class="line">Total loci	55,986</span><br><span class="line">Non-TE Loci	</span><br><span class="line">Number	39,045</span><br><span class="line">Gene models	49,066</span><br><span class="line">Gene size	2,853 bp</span><br><span class="line">Exons/gene	4.9</span><br><span class="line">Introns/gene	3.9</span><br><span class="line">TE Loci	</span><br><span class="line">Number	16,941</span><br><span class="line">Gene models	17,272</span><br><span class="line">Gene size	3,223 bp</span><br><span class="line">Exons/gene	4.2</span><br><span class="line">Introns/gene	3.2</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">$ le MSU_japonica_all.gff3</span><br><span class="line">##gff-version 3</span><br><span class="line">Chr1    MSU_osa1r7      gene    2903    10817   .       +       .       ID=LOC_Os01g01010;Name=LOC_Os01g01010;Note=TBC%20domain%20containing%20protein%2C%20expressed</span><br><span class="line">Chr1    MSU_osa1r7      mRNA    2903    10817   .       +       .       ID=LOC_Os01g01010.1;Name=LOC_Os01g01010.1;Parent=LOC_Os01g01010</span><br><span class="line">Chr1    MSU_osa1r7      exon    2903    3268    .       +       .       ID=LOC_Os01g01010.1:exon_1;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    3354    3616    .       +       .       ID=LOC_Os01g01010.1:exon_2;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    4357    4455    .       +       .       ID=LOC_Os01g01010.1:exon_3;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    5457    5560    .       +       .       ID=LOC_Os01g01010.1:exon_4;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    7136    7944    .       +       .       ID=LOC_Os01g01010.1:exon_5;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    8028    8150    .       +       .       ID=LOC_Os01g01010.1:exon_6;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    8232    8320    .       +       .       ID=LOC_Os01g01010.1:exon_7;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    8408    8608    .       +       .       ID=LOC_Os01g01010.1:exon_8;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    9210    9617    .       +       .       ID=LOC_Os01g01010.1:exon_9;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    10104   10187   .       +       .       ID=LOC_Os01g01010.1:exon_10;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    10274   10430   .       +       .       ID=LOC_Os01g01010.1:exon_11;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      exon    10504   10817   .       +       .       ID=LOC_Os01g01010.1:exon_12;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      five_prime_UTR  2903    3268    .       +       .       ID=LOC_Os01g01010.1:utr_1;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      five_prime_UTR  3354    3448    .       +       .       ID=LOC_Os01g01010.1:utr_2;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      CDS     3449    3616    .       +       .       ID=LOC_Os01g01010.1:cds_1;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      CDS     4357    4455    .       +       .       ID=LOC_Os01g01010.1:cds_2;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      CDS     5457    5560    .       +       .       ID=LOC_Os01g01010.1:cds_3;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      CDS     7136    7944    .       +       .       ID=LOC_Os01g01010.1:cds_4;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      CDS     8028    8150    .       +       .       ID=LOC_Os01g01010.1:cds_5;Parent=LOC_Os01g01010.1</span><br><span class="line">Chr1    MSU_osa1r7      CDS     8232    8320    .       +       .       ID=LOC_Os01g01010.1:cds_6;Parent=LOC_Os01g01010.1</span><br></pre></td></tr></table></figure>
<p>注意粳稻有两个版本的基因组，使用的染色体编号与基因编号均有不同，曾经因为参考基因组版本问题导致生成的比对 bam 格式文件无法浏览，后续查询才发觉问题。</p>
<h3 id="新的分析项目"><a class="markdownIt-Anchor" href="#新的分析项目"></a> 新的分析项目</h3>
<p>某分析项目的 gff 文件如下所示，这是哪个版本的基因组？</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">Chr1    MSU_osa1r7      gene    2903    10817   .       +       .       ID=LOC_Os01g01010;</span><br><span class="line">Chr1    MSU_osa1r7      mRNA    2903    10817   .       +       .       ID=LOC_Os01g01010.1;Parent=LOC_Os01g01010;</span><br><span class="line">Chr1    MSU_osa1r7      exon    2903    3268    .       +       .       ID=LOC_Os01g01010.1:exon_1;Parent=LOC_Os01g01010.1;</span><br><span class="line">Chr1    MSU_osa1r7      exon    3354    3616    .       +       .       ID=LOC_Os01g01010.1:exon_2;Parent=LOC_Os01g01010.1;</span><br><span class="line">Chr1    MSU_osa1r7      exon    4357    4455    .       +       .       ID=LOC_Os01g01010.1:exon_3;Parent=LOC_Os01g01010.1;</span><br><span class="line">Chr1    MSU_osa1r7      exon    5457    5560    .       +       .       ID=LOC_Os01g01010.1:exon_4;Parent=LOC_Os01g01010.1;</span><br><span class="line">Chr1    MSU_osa1r7      exon    7136    7944    .       +       .       ID=LOC_Os01g01010.1:exon_5;Parent=LOC_Os01g01010.1;</span><br><span class="line">Chr1    MSU_osa1r7      exon    8028    8150    .       +       .       ID=LOC_Os01g01010.1:exon_6;Parent=LOC_Os01g01010.1;</span><br><span class="line">Chr1    MSU_osa1r7      exon    8232    8320    .       +       .       ID=LOC_Os01g01010.1:exon_7;Parent=LOC_Os01g01010.1;</span><br></pre></td></tr></table></figure>
<p>经过比对，推测该数据来自：<a target="_blank" rel="noopener" href="http://rice.uga.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/">http://rice.uga.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/</a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">Index of /pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir</span><br><span class="line">[ICO]	Name	Last modified	Size	Description</span><br><span class="line">[PARENTDIR]	Parent Directory	 	-	 </span><br><span class="line">[TXT]	APK_ftp_file.txt	2011-10-31 10:12	3.9M	 </span><br><span class="line">[TXT]	POG_ftp_file.txt	2011-10-31 10:12	2.3M	 </span><br><span class="line">[   ]	all.GOSlim_assignment	2011-11-21 10:15	12M	 </span><br><span class="line">[   ]	all.cdna	2011-10-31 10:12	114M	 </span><br><span class="line">[   ]	all.cds	2011-10-31 10:12	91M	 </span><br><span class="line">[   ]	all.chrs.con	2011-10-31 10:11	363M	 </span><br><span class="line">[   ]	all.con	2011-10-31 10:12	364M	 </span><br><span class="line">[   ]	all.gff3	2012-02-07 09:41	78M	 </span><br><span class="line">[   ]	all.interpro	2011-10-31 10:12	62M	 </span><br><span class="line">[   ]	all.locus_brief_info..&gt;	2012-02-07 09:38	6.8M	 </span><br><span class="line">[   ]	all.masked.con.gz	2011-11-21 15:13	78M	 </span><br><span class="line">[   ]	all.pep	2011-10-31 10:11	33M	 </span><br><span class="line">[   ]	all.pfam	2011-10-31 10:12	4.9M	 </span><br><span class="line">[   ]	all.seq	2011-10-31 10:12	165M	 </span><br><span class="line">[TXT]	final_rice_v7_expres..&gt;	2011-10-31 10:12	10M	 </span><br><span class="line">[   ]	final_rice_v7_expres..&gt;	2011-10-31 10:12	38M	 </span><br><span class="line">[   ]	rice_osa1r7_rm.gff3.gz	2011-11-21 15:21	7.5M	 </span><br><span class="line">[   ]	rice_r7_all_tiling_p..&gt;	2011-10-31 10:12	412K	 </span><br></pre></td></tr></table></figure>
<p>将该目录下的文件全部下载</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget -r -np -nH -R index.html  http://rice.uga.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/</span><br></pre></td></tr></table></figure>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/" class="post-title-link" itemprop="url">Linux | Meet I/O error when using harddisk in Linux</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-10-27 09:55:01" itemprop="dateCreated datePublished" datetime="2021-10-27T09:55:01+08:00">2021-10-27</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-10-28 00:29:37" itemprop="dateModified" datetime="2021-10-28T00:29:37+08:00">2021-10-28</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>Linux 系统硬盘似乎出问题，报错信息是 I/O error。网上找资料说可以直接做硬盘修复，记录如下。</p>
<p>切换到 root 账户，执行 badblocks 命令查看出错的扇区数量，-b 指定每次扫描的区块大小，设置为和扇区大小一致，后面的数字 <s>1729898440</s> 为日志提示出错的扇区位置，前面的数字 <s>1729898499</s> 指定扫描扇区的终止位置，扫描后可以看到共有 <s>8</s> 个扇区出错。这里的几个数字要根据实际情况来配置。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost fenglei]# badblocks -b 512 /dev/sdb1 2766144199 2766144103     </span><br><span class="line">[root@localhost fenglei]# badblocks -b 512 /dev/sdb 2766144199 2766144103 </span><br><span class="line">2766144104</span><br><span class="line">2766144105</span><br><span class="line">2766144106</span><br><span class="line">2766144107</span><br><span class="line">2766144108</span><br><span class="line">2766144109</span><br><span class="line">2766144110</span><br><span class="line">2766144111</span><br><span class="line">2766144112</span><br><span class="line">2766144113</span><br><span class="line">2766144114</span><br><span class="line">2766144115</span><br><span class="line">2766144116</span><br><span class="line">2766144117</span><br><span class="line">2766144118</span><br><span class="line">2766144119</span><br><span class="line">2766144128</span><br><span class="line">2766144129</span><br><span class="line">2766144130</span><br><span class="line">2766144131</span><br><span class="line">2766144132</span><br><span class="line">2766144133</span><br><span class="line">2766144134</span><br><span class="line">2766144135</span><br><span class="line">2766144136</span><br><span class="line">2766144137</span><br><span class="line">2766144138</span><br><span class="line">2766144139</span><br><span class="line">2766144140</span><br><span class="line">2766144141</span><br><span class="line">2766144142</span><br><span class="line">2766144143</span><br><span class="line">2766144152</span><br><span class="line">2766144153</span><br><span class="line">2766144154</span><br><span class="line">2766144155</span><br><span class="line">2766144156</span><br><span class="line">2766144157</span><br><span class="line">2766144158</span><br><span class="line">2766144159</span><br></pre></td></tr></table></figure>
<p><strong>1 先备份数据（可选）</strong>：若修复的硬盘或分区的重要数据已备份，此部分可以省略。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">dd if=/dev/mapper/VolGroup-lv_home  skip=217874591 of=/tmp/217874591-217874595.dat count=5</span><br></pre></td></tr></table></figure>
<p><strong>2 修复</strong>：硬盘在使用时不能修复，否则可能存在写并发的问题，所以修复前需要 umount 对应分区，例如 <code>umount /mnt/sdb</code>。另外注意若为系统所在分区就没办法在线修复了，因为无法 umount。我在执行 umount 命令室收到提示说无法 umount，发现此时并没有打开窗口读取 /mnt/sdb 目录，查找后台才发现原来 /mnt/sda 目录下正在开展的一个分析项目，代码里面有读取 /mnt/sdb 下面的数据，停止这个进程之后，umount 就可以了。再操作下面的命令的时候就通过了。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost fenglei]# badblocks -b 512 -w /dev/sdb 2766144159 2766144104</span><br><span class="line">/dev/sdb is apparently in use by the system; it&#x27;s not safe to run badblocks!</span><br></pre></td></tr></table></figure>
<p>先备份数据（可选），若修复的硬盘或分区的重要数据已备份，此部分可以省略。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">dd if=/dev/mapper/VolGroup-lv_home  skip=217874591 of=/tmp/217874591-217874595.dat count=5</span><br></pre></td></tr></table></figure>
<p>在 umount 可能出现&quot;Device busy&quot;错误时，是因为有程序在使用这个分区，需要将这些进程都关闭。那么怎么知道哪些进程占用分区呢？使用 fuser（命令如下），其中 /home 是分区对应的挂载目录。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">fuser -m /home</span><br><span class="line">fuser -m -v -i -k /home</span><br></pre></td></tr></table></figure>
<p>第一条fuser命令列出使用/home的进程ID，第二条列出PID并kill掉进程（带有提示确认），建议先使用第一条命令列出PID，然后针对查看是哪些类型的进程，不要盲目杀死进程。</p>
<p>umount 分区成功后，修复命令如下，其中-s表示给出进度，-w表示写入修复的，后面是结束（END）和开始（START）块号，注意END在前，START在后。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">badblocks -s -w /dev/mapper/VolGroup-lv_home 217874595 217874591</span><br></pre></td></tr></table></figure>
<p>修复后再重新检查下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">badblocks -s -v /dev/mapper/VolGroup-lv_home 217874595 217874591</span><br></pre></td></tr></table></figure>
<p><strong>恢复数据（可选）</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">dd if=/tmp/217874591-217874595.dat of=/dev/mapper/VolGroup-lv_home </span><br></pre></td></tr></table></figure>
<p><strong>重新分区检查</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">badblocks -s -v -o /root/bb-home.log /dev/mapper/VolGroup-lv_home</span><br></pre></td></tr></table></figure>
<p>若没有坏道说明修复已完成，若有坏道可以尝试重复以上方法。</p>
<p>完成后重新mount分区</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mount /dev/mapper/VolGroup-lv_home /home</span><br></pre></td></tr></table></figure>
<p><strong>硬盘使用的一些建议</strong></p>
<p>硬RAID还是要有的。硬件RAID服务器自带，性能比软件RAID要高。根据对数据的安全性、硬盘大小，存储性能来评估该选择的RAID类型，常用的RAID类型如下：</p>
<p>RAID0：存储性能（读写效率）最高，连续的数据分散到多个磁盘上存储，提高并发访问。<br />
RAID1：数据安全性高，即同样的数据在另一块盘上备份一份，硬盘的容量也就减少一半。<br />
RAID5：兼具RAID0和RAID1的优点。存储性能大于RAID1，小于RAID0；数据安全性大于RAID0，小于RAID1。</p>
<p>读写频繁的操作不要放在/分区<br />
/ 分区是存放系统程序的，一般也比较小，读写频繁的日志不要放在根分区，这样损坏起码不会影响系统的稳定性。</p>
<p><strong>参考资料</strong><br />
1 <a target="_blank" rel="noopener" href="https://www.cnblogs.com/augusite/p/10931175.html">https://www.cnblogs.com/augusite/p/10931175.html</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/" class="post-title-link" itemprop="url">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-10-20 16:35:06" itemprop="dateCreated datePublished" datetime="2021-10-20T16:35:06+08:00">2021-10-20</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-10-22 10:23:20" itemprop="dateModified" datetime="2021-10-22T10:23:20+08:00">2021-10-22</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>CUT&amp;Tag 技术研究蛋白与染色质结合，相对于 ChIP Seq 技术具有显著优势，这个方法的原理是什么？先需要提到的是染色质免疫切割技术。CUT&amp;RUN 是一项对染色质免疫切割技术（ChIC）在全基因组范围的扩展方法，由 Laemmli 实验室开发（Schmid et al., 2004）。ChIC 使用蛋白 A-MNase 融合蛋白切割与靶蛋白相关并被特异性抗体识别的 DNA 区域。但是，ChIC 只能通过 southern blot 对特定位点进行分析。与此有关的一种方法是染色质内源性切割（ChEC），这种方法通过融合目标蛋白和 MNase 来分析目标蛋白在全基因组范围的结合位点（Schmid et al., 2004; Zentner et al ., 2015）。这种方法有一个明显的缺点，那就是需要为各目标蛋白生成特异性融合蛋白。CUT&amp;RUN 在这两种方法的基础上有了重大改进，它通过抗体将重组蛋白 A/G-MNase 固定在了目标蛋白处（Meers et al., 2019; Skene and Henikoff, 2017）。</p>
<p>重要的是，该方法可以回收 MNase 消化的片段，因此可用于基于测序的蛋白占用和组蛋白修饰定位的全基因组分析。与基于 ChIP 的方法相比，CUT&amp;RUN 具有以下优势：使用磁珠固定细胞核，简化了操作步骤；适用于新鲜组织样本和冷冻组织样本；缩短了生成 DNA 测序文库制备材料的时间（1-2 天）。此外，由于对目标蛋白两侧的 DNA 进行了原位靶向切割，因此只有靶点 DNA 片段才会从细胞核中释放并被收集，从而留下脱靶序列。因此，CUT&amp;RUN 产生的背景信号比 ChIP 少得多。自开发以来，CUT&amp;RUN 已被调整用于多种实验设置，包括自动化高通量表观遗传学分析（AutoCUT&amp;RUN）（Janssens et al., 2018）、不溶性染色质分析（如使用 CUT &amp; RUN.Salt (Thakur and Henikoff, 2018) 和 CUT&amp;RUN.ChIP 对着丝粒区域进行分析），用于检查 CUT &amp; RUN 消化 (Brahma and Henikoff, 2019) 释放的复合物中的特定蛋白组分。最显著的是，CUT&amp;RUN 对 input 的要求极低且信噪比高，这意味着其可用于单细胞分析，例如，用于研究单个小鼠胚胎细胞中的转录因子占有率（Hainer and Fazzio, 2019）。</p>
<p>CUT&amp;Tag 由 Henikoff 实验室（Kaya-Okur et al., 2019）于 2019 年发布，衍生自 CUT&amp;RUN 实验方案，可以加快文库制备并简化自动化操作。<font color=blue>CUT&amp;Tag 对天然或轻微固定的染色质进行抗体引导的标签化，并据此识别全基因组中目标蛋白占据的位置。CUT&amp;Tag 方法中用到了预装接头 DNA 并与蛋白 A 融合的超高活性 Tn5 转座酶。这种融合蛋白能与一抗结合，在抗体附近切割 DNA，并插入带有接头序列的短标签（标签化）。回收标签化的 DNA 片段后，使用引物识别添加的标签中的序列进行 PCR 扩增，以生成下一代测序（NGS）文库。特定区域的序列读取频率与靶蛋白占据的位置或组蛋白修饰的位置对应。</font></p>
<p>CUT&amp;RUN 和 CUT&amp;Tag 是两种操作简单、用途广泛且功能强大的 DNA-蛋白质相互作用分析方法，应成为每位分子生物学家的必备工具。这两种方法都有助于在全基因组范围内识别特定基因，以及以组蛋白修饰标记或与目标蛋白结合的顺式调控因子，从而深入了解基于染色质的基因控制机制。由于所需的起始材料极少，所以 CUT&amp;RUN 和 CUT&amp;Tag 在稀有细胞类型研究方面具有独特的优势。最重要的是，这两种方法的实验方案都简单易用，省时省力，可以快速周转，因此研究人员能够进行多项平行分析，从而更加全面地了解基因组调控的复杂性。</p>
<p>转录因子和组蛋白通过与DNA相互作用（DNA-Protein Interaction, DPI）发挥着调控基因复制、表达、重组和修复的重要作用，在实现细胞功能的过程中至关重要。其中转录因子组合决定了细胞的增值、分化以及死亡，是表观基因组学研究的重要组成部分。研究 DNA-蛋白质相互作用的实验方法主要有凝胶阻滞实验、DNase 1 足迹实验、甲基化干扰实验、体内足迹实验和 ChIP-seq 等。基于 NGS 研究手段的 ChIP-seq 技术可以真实反应靶蛋白在全基因组上的结合情况，是非常经典的 DNA 结合位点分析方法，至今仍然在广泛使用。</p>
<p>ChIP-Seq 的基本实验流程过程包括（图1）：甲醛交联将 DNA 结合蛋白和 DNA 紧密固定；然后抽提染色质，超声破碎 DNA 断裂，片段范围在 200-500 bp 之间；抗体孵育结合靶转录因子，Protein A/G 磁珠拉取抗体-蛋白-DNA 复合物；蛋白消化后再通过 PCR、芯片或者二代测序对 DNA 序列进行鉴定。</p>
<p><img src="https://i.imgur.com/l384XKj.jpeg" alt="图1 ChIP Seq 原理示意图" /></p>
<p>看似简单，实则有几个技术难题，如果实验条件摸索不成熟会消耗大量时间。首先，ChIP需要较多的细胞量，免疫共沉淀需要足够的靶标蛋白，而靶标蛋白有限的情况下就需要提供大量的细胞；<br />
其次，信噪比低，很多转录因子和染色体的结合相对松散，需要甲醛强交联，以防止后续洗涤过程破坏这些结合，这样导致很多非特异性信号，需要更高的数据量得到真实peak值；此外，实验的打断条件、重复性差等都需要系统的摸索才能得到较好的结果。这些因素综合起来导致ChIP对新手非常不友好，需要采购特定的设备、很长时间的预实验才能得到满意的结果。</p>
<p>为了克服这些难题，CUT&amp;Run 以及更加成熟 CUT&amp;Tag 被开发出来。2019 年，美国弗雷德哈钦森癌症研究中心的 Henikoff 博士在 Nature Communication 公开了 CUT&amp;Tag 技术的详细结果与实验方案。</p>
<p><font color=blue>CUT&amp;Tag（Cleavage Under Targets and Tagmentation, CUT&amp;Tag）技术，它免去了甲醛交联、超声破碎和免疫共沉淀的过程，这样既节省了初始实验材料又提高了信噪比、提升了实验重复性。</font>基本技术流程包括（图2）：首先，特异性抗体和靶标蛋白孵育结合；加入Tn5转座酶—Protein A复合物，其中Tn5转座酶两端已装载好建库接头引物；特异性抗体和Protein A相结合，转座酶Tn5被“拉”到转录因子附近；转座酶Tn5一次性完成附近DNA打断并完成NGS加接头过程，可直接建库[2]。基于CUT&amp;Tag，单细胞转录因子结合位点的研究也成为可能[2]，这也展示了CUT-Tag的潜能。<br />
转录组测序</p>
<p><img src="https://i.imgur.com/mSTAtC1.jpeg" alt="图2 CUT&amp;Tag seq 原理示意图" /></p>
<p>为证明Cut&amp;Tag技术的有效性，Henikoff 博士在对组蛋白H3K27me3进行研究时对比使用了 ChIP-seq、Cut&amp;RUN、Cut&amp;Tag 三种方法。从不同角度对三组数据进行比较后，发现 Cut&amp;Tag 技术具有显著优势，具体结果如下：</p>
<ol>
<li>CUT&amp;Tag 具有更好的信噪比和更低的背景<br />
研究者将三种方法的数据量都设置为8M Reads，结果显示三者得到的染色体pattern相似。但ChIP-seq的背景噪声非常高，增加数据量到50M Reads时，才能达到类似的峰图，因此ChIP-seq需要更大的测序深度才能达到去背景噪音的结果。相反，Cut＆RUN和Cut＆Tag均具有极低的背景噪声水平，其中Cut＆Tag的背景噪音最低。</li>
</ol>
<p><img src="https://i.imgur.com/dFuPanX.jpeg" alt="图3 ChIP-Seq 、CUT&amp;RUN、CUT&amp;Tag 实验结果对比" /></p>
<ol start="2">
<li>CUT&amp;Tag 在识别染色质特征时最高效</li>
</ol>
<p><img src="https://i.imgur.com/4kiBShb.jpeg" alt="Fig 4" /></p>
<p>图 4 左图为 CUT&amp;Tag、CUT&amp;RUN、ChIP-Seq 三种方法分析 H3K4me1 组蛋白修饰，CUT&amp;Tag 在识别染色质特征时最有效，能够给出 ChIP-Seq 无法读出的信息；右图为 peak-calling 的对比，使用默认参数调用每种方法上的峰值，结果显示 CUT&amp;Tag 比 CUT&amp;RUN、ChIP-seq 或 ATAC-seq 有更高的信号。</p>
<ol start="3">
<li>重复性好，灵敏度高</li>
</ol>
<p>三种方法对H3K4me1数据分析，并对结果进行分层聚类相关矩阵分析，结果显示显示CUT&amp;Tag重复样本的数据有高度的相似性，说明了CUT&amp;Tag的可重复性高。</p>
<p><img src="https://i.imgur.com/J9MfdEm.jpeg" alt="Fig 5" /></p>
<p>图5 a：CUT&amp;Tag、CUT&amp;RUN、ChIP-Seq三种方法分析H3K4me1组蛋白修饰：CUT&amp;Tag在识别染色质特征时最有效，能够给出ChIP-Seq无法读出的信息；b：peak-Calling的对比，使用默认参数调用每种方法上的峰值，结果显示CUT&amp;Tag比CUT&amp;RUN、ChIP-seq或ATAC-seq有更高的信号。</p>
<ol start="4">
<li>CUT&amp;Tag 能对极少量细胞甚至单细胞进行分析</li>
</ol>
<p>CUT&amp;Tag 能够对极少量细胞（60 个细胞）甚至单细胞进行分析。由于 PCR 之前的反应都在细胞内进行，Henikoff 博士通过分配单个细胞到 5184 纳米孔，再加入带随机标签的引物进行扩增的办法，实现单细胞测序。</p>
<p><img src="https://i.imgur.com/pgHzjVI.jpeg" alt="图6 CUT&amp;Tag 单细胞测序流程" /></p>
<p>Cut&amp;Tag注意事项：一个好的抗体是CUT&amp;Tag成功的前提！这个抗体要能识别天然构象蛋白且特异性强，最好选用文献验证过的抗体，实验前并进行IP验证。如果实在没有合适的抗体，可以选择标签蛋白，如FLAG和HA等经过有效验证的单克隆抗体。</p>
<p>参考文献：</p>
<p>genewiz: <a target="_blank" rel="noopener" href="https://www.genewiz.com.cn/Public/Resources/zxzx/0038/">https://www.genewiz.com.cn/Public/Resources/zxzx/0038/</a><br />
abcam: <a target="_blank" rel="noopener" href="https://www.abcam.cn/epigenetics/chromatin-profiling-guide-1">https://www.abcam.cn/epigenetics/chromatin-profiling-guide-1</a></p>
<p>[1] Kaya-Okur HS, Wu SJ, Codomo CA, Pledger ES, Bryson TD, Henikoff JG,Ahmad K, Henikoff S. CUT&amp;Tag for efficient epigenomic profilingof small samples and single cells. Nat Commun. 2019 Apr29;10(1):1930. doi: 10.1038/s41467-019-09982-5. PMID: 31036827;PMCID: PMC6488672.<br />
[2] Wang Q, Xiong H, Ai S, Yu X,Liu Y, Zhang J, He A. CoBATCH for High-Throughput Single-CellEpigenomic Profiling. Mol Cell. 2019 Oct 3;76(1):206-216.e7. doi:10.1016/j.molcel.2019.07.015. Epub 2019 Aug 27. PMID: 31471188.</p>
<p>CUT&amp;RUN 实验方案概述</p>
<p>运行 CUT&amp;RUN 过程中，需使用靶向目标蛋白（例如具有特定修饰的组蛋白或转录因子）的抗体将蛋白 A/G 标记的 MNase 引导至基因组中目标蛋白所在区域。</p>
<p>然后，活化 MNase，切除那些接近目标蛋白的 DNA 序列，这时小片段会从细胞核中释放出来。收集这些 DNA 片段，并通过低 input DNA 文库制备试剂盒生成 NGS 文库。</p>
<p><img src="https://i.imgur.com/Vwb1FCW.png" alt="" /></p>
<p>图 1. CUT&amp;RUN 实验方案示意图。细胞核附着在磁性伴刀豆球蛋白 A 磁珠上，以便进行细胞处理并在每次洗涤后安全除去液体。对细胞核进行通透处理，同时与靶向目标蛋白的抗体一起孵育。蛋白 A/G-MNase 融合蛋白与靶向目标蛋白的抗体结合。添加 Ca2+ 时，MNase 会切割所形成的复合物两侧的 DNA，并释放从细胞核中扩散出的 DNA 片段。提取 DNA 并在基于末端修复和接头连接制备 DNA 文库时使用。NGS 通过特定区域中序列的频率给出目标蛋白的结合信息。</p>
<p>该实验方案可分为五个部分：<br />
1.提取细胞核并与磁珠结合<br />
CUT&amp;RUN 通常使用未固定的新鲜样本作为起始材料，但是对实验方案做出调整后，也可以使用在 10% DMSO 中冷冻保存的样本（Janssens et al ., 2018; Skene et al ., 2018）。</p>
<p>进行该实验方案所需的材料很少，一般建议先使用不超过 500,000 个细胞（哺乳动物）。</p>
<p>用 Henikoff 推荐的细胞核提取缓冲液释放细胞或细胞核，并与伴刀豆球蛋白 A 磁珠结合，这种磁珠具有独特的糖结合特性。也可以使用其他兼容的细胞核提取实验方案（使用含 Triton X-100 的缓冲液时，请查看下述建议）。执行剩余实验方案时，可以使用磁力架轻松洗涤样本。</p>
<p>2.通透技术和抗体结合<br />
对与伴刀豆球蛋白 A 磁珠结合的细胞核作通透处理，同时在含洋地黄皂苷和 EDTA 的缓冲液中与靶向目的蛋白的抗体（一抗）一起孵育。EDTA 可迅速中断细胞代谢，从而抑制内源性 DNAse 的活性，保留染色质并降低总体背景信号。用户可自行调整该步骤的持续时间；通常为 2 小时到过夜不等。建议首先按 1:100 或 0.5 µg - 1.0 µg 稀释抗体，但需优化抗体的用量。建议加入阳性（α-H3K27me3）和阴性对照（同型对照 IgG）样本。</p>
<p>需要注意的是，抗体种属不同，蛋白 A 和蛋白 G 的结合效率也不同（点击查看）。虽然相较单独使用蛋白 A 而言，使用蛋白 A/G 可以增强抗体的兼容性，但在某些情况下可能需要使用二抗。二抗通过增加蛋白 A/G 结合区域的数量将 MNAse 引导至目标区域，从而帮助回收低丰度的目标序列。</p>
<p>3.MNase 结合和靶序列切割<br />
为了将蛋白 A/G-MNase 定位到抗体结合的基因组靶区域，需在含洋地黄皂苷的洗涤缓冲液中稀释融合蛋白并与细胞核一起孵育。然后洗去未结合的酶融合蛋白，于 0℃ 加入 Ca2+ 离子激活 MNase。虽然切割本身对温度不是特别敏感，但是切割后的 DNA 片段的后续扩散对温度很敏感，而且温度升高会导致更高的背景。如果最终物质中的高分子量片段过多或过少，可以调整消化时间。</p>
<p>4.DNA 回收<br />
添加含 EGTA 的终止缓冲液，终止 MNase 活性。该缓冲液可以选择性地包含异源 DNA 添加物，以便在数据处理过程中用于校准 CUT&amp;RUN 分析。通过提高孵育温度，促使细胞核释放 MNase 产生的片段，并用苯酚/氯仿/异戊醇提取这些片段，然后用乙醇进行沉淀。</p>
<p>修改实验方案可以进一步限制 MNase 融合蛋白从其结合位点过早释放、扩散以及非特异性切割。</p>
<p>该版本实验方案联合使用低盐缓冲液和高浓度 Ca2+ 溶液激活 MNase，非常适合主要在活性开放染色质中发现的靶点，但抗体显示高背景信号时，也可以使用。</p>
<p>5.文库制备和测序<br />
清除切割后的 DNA 片段后，根据制造商指南，使用标准末端修复和接头连接方法生成低 input DNA 文库。Henikoff 实验室最初建议使用 TRUseq 文库制备法，但是现在已有大量使用 NEBNext® Ultra™ II DNA 文库试剂盒的报告。与自动化 CUT&amp;RUN 实验方案（Jenssens et al. 2018）类似，您也可以省略苯酚/氯仿/异戊醇提取步骤，直接将释放的 DNA 片段用于基于末端修复和接头连接的实验方案。</p>
<p>文库的大小分布和浓度可通过毛细管电泳（例如 Bioanalyzer 或 TapeStation）确定。可以将多个文库合并，每个文库获得约 800 万配对末端测序读段。由于 CUT&amp;RUN 文库的背景较低，因此 800 万配对末端读段足以分析组蛋白修饰，甚至转录因子。</p>
<p>大部分常规 ChIP-seq 数据分析工具都可以分析 CUT&amp;RUN 数据。Henikoff 实验室也专门为 CUT&amp;RUN 数据设计了一些分析工具，例如 SEACR peak caller（Meers et al ., 2019b）。对于经校准的 CUT&amp;RUN，添加了终止缓冲液的异源 DNA 添加物可用于样本信号的归一化。此外，重组蛋白 A/G-MNase 残留的大肠杆菌 DNA 也可用作添加物。</p>
<p>CUT&amp;Tag 实验方案概述<br />
运行 CUT&amp;Tag 过程中，需使用抗体将蛋白 A 标记的 Tn5 转座酶引导至目标蛋白（例如具有特定修饰的组蛋白或转录因子）所在的基因组区域。Tn5 片段和标签靶向带有特定预定义核苷酸序列（接头）的 DNA。标签化的 DNA 可以轻松回收，并进行 PCR 扩增，从而生成 NGS 文库。整个实验方案仅需一天即可完成，具体取决于抗体孵育时间。</p>
<p><img src="https://i.imgur.com/OC0ezUh.png" alt="" /></p>
<p>​​图 2. CUT&amp;Tag 实验方案示意图。细胞核附着在磁性伴刀豆球蛋白 A 磁珠上，以便进行细胞处理并在每次洗涤后安全除去液体。对细胞核进行通透处理，同时与靶向目标蛋白的一抗一起孵育，然后与识别一抗的二抗一起孵育。蛋白 A-Tn5 融合蛋白与目标蛋白上形成的抗体复合物结合。添加 Mg2+ 离子时，Tn5 会切割所形成的复合物两侧的 DNA，并释放从细胞核中扩散出的已切割 DNA 片段。提取 DNA 并在基于 PCR 扩增制备 DNA 文库时使用。NGS 通过特定区域中序列的频率给出目标蛋白的结合信息。</p>
<p>该实验方案可分为五个部分：</p>
<ol>
<li>提取细胞核并与磁珠结合</li>
</ol>
<p>CUT&amp;Tag 通常使用未固定的新鲜样本作为起始材料，但对实验方案进行修改后，也可以使用冷冻样本，近期，还可使用提取细胞核后经福尔马林轻微固定的样本。固定步骤可以降低执行实验方案期间细胞核聚集在一起的可能性。使用由含 Triton X-100 的缓冲液产生的固定和冷冻保存样本时，可从所有缓冲液中除去洋地黄皂苷，这样可以进一步减少磁珠结块。但是，需要注意，在某些情况下，表位固定会干扰抗体的结合。通常建议先使用不超过 500,000 个细胞（哺乳动物）。使用新鲜或冷冻组织时，细胞核的制备与 CUT&amp;RUN 实验方案中的细胞核制备类似（Janssens et al ., 2018; Skene et al ., 2018）。对于 CUT&amp;RUN，细胞核与伴刀豆球蛋白 A 磁珠结合，可以使用磁力架轻松洗涤样本。但是，该步骤可以省略，并且整个实验方案可以按以下方式执行：每次洗涤后，温和离心，然后小心除去上清液，不干扰沉淀的细胞核。</p>
<ol start="2">
<li>通透技术和抗体结合</li>
</ol>
<p>对结合至伴刀豆球蛋白 A 磁珠的细胞核进行通透处理，同时在含洋地黄皂苷的缓冲液中与靶向目的蛋白的抗体（一抗）一起孵育。该步骤的持续时间可以调整，最短 2 小时，最长 5 天。抗体的推荐稀释度通常在 1:50 至 1:100（或 0.5 µg - 1.0 µg）之间，但需优化抗体的使用量。建议加入阳性（α-H3K27me3）和阴性对照（同型对照 IgG）样本。一抗孵育后，用含洋地黄皂苷的洗涤缓冲液快速洗涤，然后将细胞核与二抗一起孵育，这里使用的二抗需靶向一抗，充当桥接抗体并且可以增加蛋白 A 结合位点的数量。在 CUT&amp;RUN 实验方案中，这一步骤可以省略；但在 CUT&amp;Tag 实验方案中，必须执行这一步骤，以增强信号。需要注意的是，抗体类别不同，蛋白 A 的结合效率也不同，因此需要事先检查二抗与蛋白 A 的兼容性。</p>
<ol start="3">
<li>Tn5 结合和标签化</li>
</ol>
<p>为了将蛋白 A-Tn5 转座酶与目标抗体结合，需在含洋地黄皂苷的高盐缓冲液中稀释带接头的 Tn5 融合蛋白，并与细胞核一起孵育。在该步骤中增加盐浓度有助于减少脱靶标签化，主要是可接触的基因组区域的脱靶标签化，从而产生类似 ATAC 的峰。与 Tn5 融合蛋白一起孵育后，洗涤细胞核，以除去未结合的 Tn5，并在添加 Mg2+ 离子的高盐洋地黄皂苷缓冲液中于 37℃ 下孵育，从而激活标签化。</p>
<ol start="4">
<li>DNA 回收</li>
</ol>
<p>添加 EDTA 终止标签化，并用 SDS 和蛋白酶 K 裂解细胞核。然后可以通过多种方式清除 DNA 片段，原实验方案建议先用苯酚/氯仿/异戊醇提取这些片段，再用乙醇进行沉淀。您也可以使用 AMPure 磁珠来清除 DNA 片段。建议在沉淀步骤中避免使用任何载体，例如糖原，因为载体会降低后续 PCR 反应的效率。</p>
<ol start="5">
<li>通过 PCR 扩增和测序制备文库</li>
</ol>
<p>由于 Tn5 在标签化过程中会在 DNA 片段中引入兼容序列，所以使用通用 i5 引物和带条形码的 i7 引物进行的简单 PCR 反应就可以生成测序文库，让 DNA 文库的制备变得极为快速简便。该步骤与 ATAC-seq 类似，因此可以使用 ATAC 实验方案描述的引物序列（Buenrostro et al ., 2013）。但需要根据用户的机器调整循环程序。为了确保 PCR 有效，需要进行短时间的退火操作；如使用缓慢升温循环仪，则可以省略退火步骤，因为变性和延伸之间的冷却时间较长，足以退火。如使用快速升温机器，则必须增加退火步骤。通常，建议不超过 12 到 14 个扩增循环，否则，文库的复杂性将降低，而 PCR 的重复率升高。</p>
<p>在 PCR 扩增后，可用毛细管凝胶电泳法（例如 Bioanalyzer 或 TapeStation）对 CUT&amp;Tag 文库进行评估。CUT&amp;Tag 实验的失败通常表现在阳性对照样本中不存在核小体片段。</p>
<p>但是，如果只在转录因子 CUT&amp;Tag 文库中观察到非常弱的信号，实验仍然可以继续，这种情况很常见。通过凝胶电泳进行评估后，可以使用 SPRI 磁珠清理并浓缩文库。可以将多个文库合并，每个文库获得约 200 万配对末端测序读段。</p>
<p>大部分常规 ChIP-seq 数据分析工具都可以分析 CUT&amp;Tag 数据。Henikoff 实验室也专门为 CUT&amp;RUN 和 CUT&amp;Tag 数据设计了一些分析工具，例如 SEACR peak caller（Meers et al ., 2019b）。要想校准 CUT&amp;Tag 数据，可以将重组蛋白 A-Tn5 残留的大肠杆菌 DNA 用作普通添加物。</p>
<p>样本制备</p>
<p>由于 CUT&amp;RUN 和 CUT&amp;Tag 都可以使用未固定的新鲜样本，所以，它们的样本制备非常简单。但是需要以适合所用细胞类型的方式制备单细胞悬液，包括使用解离试剂（如 Accutase™）、从细胞培养皿上刮下细胞或机械解离组织等。</p>
<p>为了便于收集，可以在 10% DMSO 中冷冻保存样本，并在 Mr. Frosty 异丙醇冻存盒内冻存。两种方法均不需要固定样本，但是，如果磁珠在洗涤过程中结块，则应在孵育抗体前，于室温下使用 0.1% 福尔马林对样本进行轻固定，2 分钟即可。</p>
<p>抗体选择</p>
<p>至于 ChIP，并非所有抗体都可以用于 CUT&amp;RUN 和 CUT&amp;Tag。由于这两种方法较新，大多数抗体均未经过兼容性测试，因此终端用户需要对抗体进行测试和优化。ChIP 级抗体大都适用于 CUT&amp;RUN 和 CUT&amp;Tag，特别是适用于天然 ChIP 的 ChIP 级抗体。</p>
<p>试用非 ChIP 级抗体时，假设特异性已进行充分表征，可以首先选择能够识别目标蛋白天然形式的抗体；例如，在免疫沉淀或免疫细胞化学中起作用的抗体。同理，在考虑 CUT&amp;RUN 或 CUT&amp;Tag 实验所用的抗体浓度时，可以将免疫荧光法的推荐浓度作为起始浓度。需测试抗体与蛋白 A 和蛋白 G 的兼容性，必要时，可使用合适的二抗。</p>
<p>对照</p>
<p>与其他实验类型一样，添加合适的对照是确保实验符合预期并能在实验失败时轻松确定问题区域的关键。</p>
<p>与 ChIP 不同，该实验方案不需要 input 样本，因为非抗体引导的 MNase 处理或标签化仅可识别可接触的染色质。为了从样本中获取背景并为实验设置基线，应添加 IgG 对照。在抗体对照方面，Henikoff 实验室建议使用 H3K27me3 作为 CUT&amp;RUN 和 CUT&amp;Tag 实验的阳性对照。使用未修饰的总组蛋白对照，例如总组蛋白 H3，可以按比例显示组蛋白修饰。</p>
<p>优化</p>
<p>使用这些技术时，需要在多个阶段对 CUT&amp;RUN 和 CUT&amp;Tag 实验方案进行优化。</p>
<p>透化</p>
<p>初版实验方案要求对洗涤缓冲液中所用的洋地黄皂苷的浓度进行测定，以确保有效地对细胞核作通透处理。如果您对实验方案作了改进，具体来说就是在洗涤缓冲液中添加了 NP40，则有可能不需要进行上述测定。但是，在新材料上使用这两种技术时，应牢记细胞核的通透效率。</p>
<p>抗体浓度</p>
<p>采用滴定法测定每次反应使用的抗体量很重要，起始稀释度可以参考 ChIP 或免疫荧光检测法的推荐稀释度。</p>
<p>抗体孵育时间</p>
<p>室温下，一抗孵育 1 小时即可；但在冷藏室中，孵育时间可延长至 1 至 5 天。对于每项实验使用的抗体，可以对孵育时间进行优化；但应牢记，孵育时间过长可能会增加背景信号，最终导致总体信噪比不佳。</p>
<p>使用二抗</p>
<p>如果在只使用一抗的情况下，回收率较低，则强烈建议在 CUT&amp;Tag 中使用二抗，还可以考虑在 CUT&amp;RUN 中使用二抗。添加二抗的另一个原因是为了避免蛋白 A/G 与一抗的配对不符合预期，添加合适类别的二抗可以改善这一点。</p>
<p>消化和标签化时间</p>
<p>MNase 消化和 Tn5 标签化的时间可以根据目标蛋白进行调整。低丰度蛋白要想回收所有位点可能需要花费更多时间。要记住的是，消化或标签化时间过长可能会导致无针对性的切割，从而产生高背景信号。</p>
<p>常见问题</p>
<ul>
<li>
<p>起始材料应该使用多少个细胞？<br />
模式生物的基因组小于人/小鼠基因组时，是否需要增加细胞数量？建议每个样本不超过 500,000 个细胞。一般来说，起始量为 50,000 个细胞比较好。与 ChIP 不同，CUT &amp;RUN 和 CUT＆Tag 非常敏感，所以不需要上百万个细胞作为起始材料。细胞数量过多会降低收率，甚至降低文库的复杂性。</p>
</li>
<li>
<p>增加细胞用量时，是否应增加伴刀豆球蛋白 A 磁珠的用量？<br />
每次实验使用的细胞数不应超过 500,000 个。由于实验方案已对使用 50,000 至 500,000 个细胞时的磁珠数量进行了优化，因此不需要在实验方案建议的磁珠用量基础上另行增加。</p>
</li>
<li>
<p>如何冷冻保存细胞确保其可用于 CUT＆RUN 或 CUT &amp;Tag？<br />
建议在合适的缓冲液或培养基中添加 10％ DMSO，然后使用 Mr. Frosty 异丙醇冻存盒缓慢冻存。不建议快速冻存。</p>
</li>
<li>
<p>我有固定和冷冻 ChIP 细胞，可以将它们用于 CUT&amp;RUN 或 CUT＆Tag 吗？<br />
可以，但具体要看 ChIP 样本的固定和冷冻保存条件。由于 ChIP 样本的细胞数可能会超过 CUT&amp;RUN 和 CUT&amp;Tag 的推荐用量，并且使用强固定条件（如双重固定和淬灭）会损害 MNase 或 Tn5 的活性，所以标准实验方案可能不太适合。您可以将样本分成多份，这样可以轻松避免 ChIP 样本中细胞数超过 CUT&amp;RUN 或 CUT&amp;Tag 推荐起始量的问题。但是，样本固定和冻存等问题仍然存在。ChIP 实验方案的固定条件比较严格，很难与 CUT&amp;RUN 或 CUT&amp;Tag 兼容，因为严格的固定条件很可能会损害 MNase 或 Tn5 片段化/标签化目标序列的能力，并且已有大量报道暗示了这种不兼容性。另外，固定还有可能导致抗原表位被掩盖。因此，不建议固定样本。冷冻保存用于 CUT&amp;RUN 或 CUT&amp;Tag 的细胞时，需要使用 10% DMSO 和 Mr. Frosty 异丙醇冻存盒缓慢冻存样本。不建议在大多数 ChIP 实验方案中增加快速冻存步骤。</p>
</li>
<li>
<p>为什么洗涤过程中磁珠会结块？<br />
结块可能是因为细胞与磁珠的比例过高或含洋地黄皂苷的缓冲液中的细胞核/细胞发生裂解，从而释放 DNA 并导致结块。这种情况下，首先要检查细胞数量是否超过推荐数量。此外，在与抗体一起孵育前，建议先用福尔马林轻微固定样本（室温下使用 0.1% 福尔马林固定 2 分钟），以减少磁珠在含洋地黄皂苷的洗涤缓冲液中的结块。但需记住的是，固定可能会影响表位的可用性，因此需要对每种抗体进行测试。</p>
</li>
<li>
<p>为什么用于一抗孵育的缓冲液中含 EDTA？<br />
建议在通透和抗体孵育缓冲液中添加 EDTA，因为它会螯合 Mg2+ 离子，从而终止所有依赖 ATP 的细胞过程（包括复制和染色质重塑），同时终止内源性 DNase。</p>
</li>
<li>
<p>哪里可以得到蛋白 A/G-MNase 或蛋白 A-Tn5？<br />
这两种融合蛋白最初都由 Henikoff 实验室提供。现在已经有了制备融合蛋白的实验方案（Kaya-Okur et al ., 2019; Meers et al ., 2019a），而且质粒、蛋白 A-MNase、蛋白 A/G-MNase 和蛋白 A-Tn5 都可以从试剂供应商处购买。</p>
</li>
<li>
<p>在 CUT&amp;Tag 中，是否会得到非抗体引导的标签化，从而产生类似 ATAC 的峰？<br />
按照标准实验方案的建议增加标签化缓冲液中的盐浓度即可解决这一问题。在某些情况下，依据目标蛋白的情况，可能会人为发现一些峰看起来像 Kaya-Okur et al 2019 的报道中所示的 ATAC 峰。</p>
</li>
</ul>
<p>参考文献<br />
Brahma, S ., Henikoff, S. RSC-associated subnucleosomes define MNase-sensitive promoters in yeast . Mol Cell 73, 238–249, e233 (2019).</p>
<p>Buenrostro, J .D ., Giresi, P .G ., Zaba, L .C ., Chang, H .Y ., Greenleaf, W .J . Transposition of native chromatin for fast and sensitive epigenomic profiling of open chromatin, DNA-binding proteins and nucleosome position. Nat Methods 10, 1213–1218 (2013).</p>
<p>Hainer, S .J ., Fazzio, T .G. High-resolution chromatin profiling using CUT&amp;RUN. Curr Protoc Mol Biol 126, e85 (2019).</p>
<p>Janssens, D .H ., et al. Automated in situ chromatin profiling efficiently resolves cell types and gene regulatory programs. Epigenet Chromatin 11, 74 (2018).</p>
<p>Kaya-Okur, H .S ., et al. CUT&amp;Tag for efficient epigenomic profiling of small samples and single cells. Nat Commun 10, 1930 (2019).</p>
<p>Meers, M .P ., Bryson, T .D ., Henikoff, J .G ., Henikoff, S. Improved CUT&amp;RUN chromatin profiling tools. eLife 8: e46314 (2019).</p>
<p>Meers, M .P ., Tenenbaum, D ., Henikoff, S . Peak calling by Sparse Enrichment Analysis for CUT&amp;RUN chromatin profiling . Epigenet Chromatin 12, 42 (2019).</p>
<p>Schmid, M ., Durussel, T ., Laemmli, U .K. ChIC and ChEC; genomic mapping of chromatin proteins. Mol Cell 16, 147–157 (2004).</p>
<p>Skene, P .J ., Henikoff, J .G ., Henikoff, S. Targeted in situ genome-wide profiling with high efficiency for low cell numbers. Nat Protoc 13, 1006–1019 (2018).</p>
<p>Skene, P .J ., Henikoff, S. An efficient targeted nuclease strategy for high-resolution mapping of DNA binding sites. eLife 6: e21856 (2017).</p>
<p>Thakur, J ., Henikoff, S. Unexpected conformational variations of the human centromeric chromatin complex. Genes Dev 32, 20–25 (2018).</p>
<p>Zentner, G .E ., Kasinathan, S ., Xin, B ., Rohs, R ., Henikoff, S. ChEC-seq kinetics discriminates transcription factor binding sites by DNA sequence and shape in vivo. Nat Commun 6, 8733 (2015).</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/2/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><a class="page-number" href="/page/2/">2</a><span class="page-number current">3</span><a class="page-number" href="/page/4/">4</a><span class="space">&hellip;</span><a class="page-number" href="/page/19/">19</a><a class="extend next" rel="next" href="/page/4/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">190</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2023</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
