<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta property="og:type" content="website">
<meta property="og:title" content="FENG&#39;s Notes">
<meta property="og:url" content="http://example.com/page/3/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/page/3/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/" class="post-title-link" itemprop="url">Bioinfo | Gene Ontology - Enrichment Analysis</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-06-29 11:14:12 / Modified: 23:44:19" itemprop="dateCreated datePublished" datetime="2021-06-29T11:14:12+08:00">2021-06-29</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>下面是我做的大豆GO数据库（version Gmax_508_Wm82_a4.v1）构建与富集分析。数据文件在：<a target="_blank" rel="noopener" href="https://github.com/bioinfx/Glycine_max_Gene_Ontology">https://github.com/bioinfx/Glycine_max_Gene_Ontology</a>。</p>
<p>经过前期数据比对和整理（比如依据 swissprot、blast2GO，或 GOMAP），得到下面“基因——GO号——Ontology 类别——证据”这四列格式的文件 GMA_GO_TERM.txt，就可以在R程序里面用于 GO 数据库的构建。</p>
<h3 id="数据库源文件准备"><a class="markdownIt-Anchor" href="#数据库源文件准备"></a> 数据库源文件准备</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ zcat GMA_GO_TERM.txt.gz | head</span><br><span class="line">GID     GO      ONTOLOGY        EVIDENCE</span><br><span class="line">Glyma.01G000100 GO:0006412      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009063      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009234      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0016021      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0031969      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0016835      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0030976      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0070204      F       IEA</span><br><span class="line">Glyma.01G000137 GO:0016043      P       IEA</span><br></pre></td></tr></table></figure>
<h3 id="数据库导入-r"><a class="markdownIt-Anchor" href="#数据库导入-r"></a> 数据库导入 R</h3>
<p>构建过程如下，调用 makeOrgPackage 函数构建 GO 数据库。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><span class="line">## Makes an organism package for Glycine max data.frames:</span><br><span class="line"></span><br><span class="line">library(RSQLite)</span><br><span class="line">library(AnnotationForge)</span><br><span class="line"></span><br><span class="line">## Now prepare some data.frames</span><br><span class="line">fGO &lt;- read.table(&quot;../3_final-result/GMA_GO_TERM.txt&quot;, sep=&quot;\t&quot;, head=TRUE)</span><br><span class="line"></span><br><span class="line">## remove duplication in the table</span><br><span class="line">go_df &lt;- fGO[!duplicated(fGO),]</span><br><span class="line"></span><br><span class="line">## only tree coloums are needed: GID, GO, EVIDENCE</span><br><span class="line">go_df &lt;- go_df[,c(1,2,4)]</span><br><span class="line"></span><br><span class="line">## Making Organism packages </span><br><span class="line">makeOrgPackage(go=go_df,</span><br><span class="line">    version = &quot;0.0.1&quot;,</span><br><span class="line">    maintainer = &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">    author = &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">    outputDir = &quot;./&quot;,</span><br><span class="line">    tax_id = &quot;3847&quot;,</span><br><span class="line">    genus = &quot;Glycine&quot;,</span><br><span class="line">    species = &quot;max&quot;,</span><br><span class="line">    goTable = &quot;go&quot;</span><br><span class="line">    )</span><br><span class="line"></span><br><span class="line">## then you can call install.packages based on the return value</span><br><span class="line">install.packages(&quot;./org.Gmax.eg.db&quot;, repos = NULL, type=&quot;source&quot;)</span><br><span class="line">library(org.Gmax.eg.db)</span><br><span class="line">org &lt;- org.Gmax.eg.db</span><br><span class="line"></span><br><span class="line">## A demo for using the database</span><br><span class="line"># BiocManager::install(&quot;Rgraphviz&quot;)</span><br><span class="line"># BiocManager::install(&quot;topGO&quot;)</span><br><span class="line">library()</span><br><span class="line">geneTable = read.table(&quot;./Gmax_ABA_exp.xls&quot;, head=T)</span><br><span class="line">geneList=geneTable[,5]</span><br><span class="line">geneList=geneList[!duplicated(geneList)]</span><br><span class="line"># Biology process</span><br><span class="line">ego_up &lt;-enrichGO(gene = geneList, OrgDb = org, keyType = &quot;GID&quot;, ont = &quot;BP&quot;)</span><br><span class="line">pdf(&quot;up_GO_BP_barplot.pdf&quot;, height=10, width=15)</span><br><span class="line">barplot(ego_up, drop=TRUE, showCategory=12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;up_GO_BP_dotPlot.pdf&quot;, height=10, width=15)</span><br><span class="line">dotplot(ego_up)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;up_GO_BP_tree.pdf&quot;)</span><br><span class="line">plotGOgraph(ego_up)</span><br><span class="line">dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_up@result ), file=&quot;up_GO_BP.csv&quot;)</span><br></pre></td></tr></table></figure>
<h1 id="数据库开启与富集分析"><a class="markdownIt-Anchor" href="#数据库开启与富集分析"></a> 数据库开启与富集分析</h1>
<p>下面代码示范如何引用上构建的数据库，并且开展 GO 分析：BP、MF 和 CC 三个类别。<br />
图片可输出为柱状图 barplot 和点状图 dotplot，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br></pre></td><td class="code"><pre><span class="line">library(org.Gmax.eg.db)</span><br><span class="line">org &lt;- org.Gmax.eg.db</span><br><span class="line"></span><br><span class="line">## A demo for using the database</span><br><span class="line"># BiocManager::install(&quot;Rgraphviz&quot;)</span><br><span class="line"># BiocManager::install(&quot;topGO&quot;)</span><br><span class="line">library(topGO)</span><br><span class="line">library(clusterProfiler)</span><br><span class="line">geneTable = read.table(&quot;./gene_table.txt&quot;, head=T, sep=&quot;\t&quot;)</span><br><span class="line">#geneTable = geneTable[!apply(geneTable == &quot;&quot;, 1, all),]</span><br><span class="line">geneList=geneTable[,5]</span><br><span class="line"></span><br><span class="line">geneList=geneList[!duplicated(geneList)]</span><br><span class="line"># Biology process</span><br><span class="line">b=geneList</span><br><span class="line"></span><br><span class="line"># biological progress</span><br><span class="line">ego_up &lt;-enrichGO(</span><br><span class="line">                  gene = b, </span><br><span class="line">                  OrgDb = org, </span><br><span class="line">                  keyType = &quot;GID&quot;, </span><br><span class="line">                  ont = &quot;BP&quot;</span><br><span class="line">                  )</span><br><span class="line">pdf(&quot;test_GO_BP_barplot.pdf&quot;, height=10, width=15)</span><br><span class="line">barplot(ego_up, drop=TRUE, showCategory=12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;test_GO_BP_dotPlot.pdf&quot;, height=10, width=15)</span><br><span class="line">dotplot(ego_up)</span><br><span class="line">dev.off()</span><br><span class="line">#pdf(&quot;up_GO_BP_tree.pdf&quot;)</span><br><span class="line">#plotGOgraph(ego_up)</span><br><span class="line">#dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_up@result ), file=&quot;test_GO_BP.csv&quot;)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># molecular function</span><br><span class="line">ego_down &lt;-enrichGO(</span><br><span class="line">                    gene = b,</span><br><span class="line">                    OrgDb = org,</span><br><span class="line">                    keyType = &quot;GID&quot;,</span><br><span class="line">                    ont = &quot;MF&quot;</span><br><span class="line">                    )</span><br><span class="line">pdf(&quot;test_GO_MF_barplot.pdf&quot;, height=10, width=15)</span><br><span class="line">barplot(ego_down, drop=TRUE, showCategory=12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;test_GO_MF_dotPlot.pdf&quot;, height=10, width=15)</span><br><span class="line">dotplot(ego_down)</span><br><span class="line">dev.off()</span><br><span class="line">#pdf(&quot;test_GO_MF_tree.pdf&quot;)</span><br><span class="line">#plotGOgraph(ego_down)</span><br><span class="line">#dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_down@result), file=&quot;test_GO_MF.csv&quot;)</span><br><span class="line"></span><br><span class="line"># cellular component</span><br><span class="line">ego_down &lt;-enrichGO(</span><br><span class="line">                    gene = b,</span><br><span class="line">                    OrgDb = org,</span><br><span class="line">                    keyType = &quot;GID&quot;,</span><br><span class="line">                    ont = &quot;CC&quot;</span><br><span class="line">                    )</span><br><span class="line">pdf(&quot;test_GO_CC_barplot.pdf&quot;, height=10, width=15)</span><br><span class="line">barplot(ego_down, drop=TRUE, showCategory=12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;test_GO_CC_dotPlot.pdf&quot;, height=10, width=15)</span><br><span class="line">dotplot(ego_down)</span><br><span class="line">dev.off()</span><br><span class="line">#pdf(&quot;test_GO_CC_tree.pdf&quot;)</span><br><span class="line">#plotGOgraph(ego_down)</span><br><span class="line">#dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_down@result), file=&quot;test_GO_CC.csv&quot;)</span><br></pre></td></tr></table></figure>
<p>有时候 GO 项目的名称过长，基于默认作图代码出来的图片会出现文字将图形区域挤压得无法显示。</p>
<p><img src="https://i.imgur.com/FlARCCO.jpeg" alt="" /></p>
<p>有两种优化方案，一是将长的文字换行；二是只截取一部分文字显示。</p>
<h3 id="一过长的文字换行显示"><a class="markdownIt-Anchor" href="#一过长的文字换行显示"></a> 一：过长的文字换行显示，</h3>
<p>参考资料：<a target="_blank" rel="noopener" href="https://www.cnblogs.com/songbiao/articles/12717670.html">songbiao: GO term name太长的处理</a><br />
在使用clusterProfiler进行绘制GO dotplot时，会遇到 GO term 的 name 过长，压缩了图形区域的情况。<br />
因为 clusterProfiler 作图是基于 ggplot2 的，所以这个处理办法也适用于其他 ggplot2 作图遇到此种情况。<br />
解决文本过长的办法就是，用stringr包的str_wrap来完成文本自动换行。<br />
效果如下图所示，其实还是不太完美，需要后期手工调整才美观。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)  # 需要开启 ggplot2 才可以使用 scale_y_discrete 函数</span><br><span class="line">library(stringr)</span><br><span class="line">dotplot(mf, showCategory=30) + scale_y_discrete(labels=function(mf) str_wrap(mf, width=40))</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/JRkomgm.jpeg" alt="" /></p>
<h3 id="二只截取一部分文字显示"><a class="markdownIt-Anchor" href="#二只截取一部分文字显示"></a> 二：只截取一部分文字显示</h3>
<p>参考资料：<a target="_blank" rel="noopener" href="https://r.bio-spring.info/2018/11/26/enrichplot-fancy-label/">在enrichplot中使用更好的Label</a><br />
该方法是定义一个截取短 label 的函数, 将其传递给 scale_y_discrete() 即可. 在输出 ggplot 对象时做修改。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line">#&#x27; Truncate string vector of ggplot axis label</span><br><span class="line">#&#x27;</span><br><span class="line">#&#x27; @param label    a ordered string vector</span><br><span class="line">#&#x27; @param maxLen   max length of character (nchar) to show in label</span><br><span class="line">#&#x27; @param maxWord  max count of words allowed to show in label</span><br><span class="line">#&#x27; @param pattern  Word separater</span><br><span class="line">#&#x27; @param dot      If true, three dots will added to truncated label</span><br><span class="line">#&#x27;</span><br><span class="line">#&#x27; @return a vector of truncated strings</span><br><span class="line">#&#x27; @export</span><br><span class="line">#&#x27;</span><br><span class="line">#&#x27; @examples</span><br><span class="line"># 默认最多显示 50 个字符, 5 个单词.</span><br><span class="line">short_label &lt;- function(label, maxLen = 50, maxWord = 5, pattern = &quot; &quot;, dot = TRUE)&#123;</span><br><span class="line">  l &lt;- strsplit(label, pattern)</span><br><span class="line">  short_label &lt;- vector(&quot;character&quot;,length(l))</span><br><span class="line">  </span><br><span class="line">  for (i in seq_along(l))&#123;</span><br><span class="line">    truncated &lt;- FALSE</span><br><span class="line">    s &lt;- l[[i]]</span><br><span class="line">    if (length(s) &gt; maxWord)&#123;</span><br><span class="line">      ll &lt;- paste(s[1:maxWord], collapse = &quot; &quot;)</span><br><span class="line">      truncated &lt;- TRUE</span><br><span class="line">    &#125;</span><br><span class="line">    else&#123;</span><br><span class="line">      ll &lt;- paste(s, collapse = &quot; &quot;)</span><br><span class="line">    &#125;</span><br><span class="line">    </span><br><span class="line">    if (nchar(ll) &gt; maxLen)&#123;</span><br><span class="line">      ll &lt;- substr(ll, 1, maxLen)</span><br><span class="line">      truncated &lt;- TRUE</span><br><span class="line">    &#125;</span><br><span class="line">    </span><br><span class="line">    if (dot &amp; truncated) ll &lt;- paste(ll, &quot;...&quot;,sep = &quot; &quot;)</span><br><span class="line">    </span><br><span class="line">    short_label[[i]] &lt;- ll</span><br><span class="line">  &#125;</span><br><span class="line">  attr(short_label, &quot;pos&quot;) &lt;- attr(label,&quot;pos&quot;)</span><br><span class="line">  return(short_label)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>画图代码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)</span><br><span class="line">dotplot(mf) + scale_y_discrete(label=short_label)</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/pALA2H9.jpeg" alt="" /></p>
<p>注意上面的图，其实很多项目都是高度近似的，怎么优化？对于模式物种而言，CytoScape的插件ClueGO有一个选项可以实现，就是设置Ontology的级别。因为Ontology的结构就是树状，从根部到.</p>
<p>构建GO数据库的过程中，如果成功会输出下面的信息。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(go_df)</span><br><span class="line">           GID         GO EVIDENCE</span><br><span class="line">1 Chr01.g00016 GO:0003676      IEA</span><br><span class="line">2 Chr01.g00016 GO:0003964      IEA</span><br><span class="line">3 Chr01.g00016 GO:0004190      IEA</span><br><span class="line">4 Chr01.g00016 GO:0004519      IEA</span><br><span class="line">5 Chr01.g00016 GO:0006310      IEA</span><br><span class="line">6 Chr01.g00016 GO:0015074      IEA</span><br><span class="line">&gt; makeOrgPackage(go=go_df,</span><br><span class="line">+                version = &quot;0.0.1&quot;,</span><br><span class="line">+                maintainer = &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">+                author = &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">+                outputDir = &quot;.&quot;,</span><br><span class="line">+                tax_id = &quot;126911&quot;,</span><br><span class="line">+                genus = &quot;Ammopiptanthus&quot;,</span><br><span class="line">+                species = &quot;mongolicus&quot;,</span><br><span class="line">+                goTable = &quot;go&quot;</span><br><span class="line">+ )</span><br><span class="line">Populating genes table:</span><br><span class="line">genes table filled</span><br><span class="line">Populating go table:</span><br><span class="line">go table filled</span><br><span class="line">table metadata filled</span><br><span class="line">&#x27;select()&#x27; returned many:1 mapping between keys and columns</span><br><span class="line">Dropping GO IDs that are too new for the current GO.db</span><br><span class="line">Populating go table:</span><br><span class="line">go table filled</span><br><span class="line">Populating go_bp table:</span><br><span class="line">go_bp table filled</span><br><span class="line">Populating go_cc table:</span><br><span class="line">go_cc table filled</span><br><span class="line">Populating go_mf table:</span><br><span class="line">go_mf table filled</span><br><span class="line">&#x27;select()&#x27; returned many:1 mapping between keys and columns</span><br><span class="line">Populating go_bp_all table:</span><br><span class="line">go_bp_all table filled</span><br><span class="line">Populating go_cc_all table:</span><br><span class="line">go_cc_all table filled</span><br><span class="line">Populating go_mf_all table:</span><br><span class="line">go_mf_all table filled</span><br><span class="line">Populating go_all table:</span><br><span class="line">go_all table filled</span><br><span class="line">Creating package in ./org.Amongolicus.eg.db </span><br><span class="line">Now deleting temporary database file</span><br><span class="line">[1] &quot;./org.Amongolicus.eg.db&quot;</span><br><span class="line">There were 50 or more warnings (use warnings() to see the first 50)</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/17/Linux-Gene-Ontology-GOMAP/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/17/Linux-Gene-Ontology-GOMAP/" class="post-title-link" itemprop="url">Bioinfo | Gene Ontology - GOMAP</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-06-17 10:59:17" itemprop="dateCreated datePublished" datetime="2021-06-17T10:59:17+08:00">2021-06-17</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-06-29 11:40:34" itemprop="dateModified" datetime="2021-06-29T11:40:34+08:00">2021-06-29</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>Gene Ontology Meta Annotator for Plants</p>
<p>GOMAP 主要是给植物基因集做 Gene Ontology 注释信息匹配，它综合了多个数据库，所以得到的信息会全面一些。早年将基因对应到GO用的工具是Blast2GO，这个软件linux版本安装比较麻烦，基于mysql数据库，而且后面它商业化之后要收费。后来的策略是用diamond的快速blastp将基因集与SwissProt数据库进行比较，得到每个基因最相近的Uniprot ID，再依靠id-mapping文件将基因与GO对应起来，本文最后有简单介绍基于SwissProt做GO注释的流程。</p>
<p>What is GOMAP-Singularity? GOMAP-Singularity is the containerized version of the Gene Ontology Meta Annotator for Plants (GOMAP) pipeline. GOMAP is a high-throughput pipeline to annotate GO terms to plant protein sequences. The pipeline uses three different approaches to annotate GO terms to plant protein sequences, and compile the annotations together to generate an aggregated dataset. GOMAP uses Python code to run the component tools to generate the GO annotations, and R code to clean and aggregate the GO annotations from the component tools.</p>
<p>How do I use GOMAP-Singularity?</p>
<p>Please check &lt;<a target="_blank" rel="noopener" href="http://bioinformapping.com/gomap">bioinformapping.com/gomap</a>&gt; for detailed documentation on how to use the container.</p>
<h3 id="install-system-dependencies"><a class="markdownIt-Anchor" href="#install-system-dependencies"></a> Install system dependencies</h3>
<p>You must first install development tools and libraries to your host.<br />
On Debian-based systems:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ sudo apt-get update &amp;&amp; \</span><br><span class="line">  sudo apt-get install -y build-essential \</span><br><span class="line">  libseccomp-dev pkg-config squashfs-tools cryptsetup</span><br></pre></td></tr></table></figure>
<p>On CentOS/RHEL:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sudo yum groupinstall -y &#x27;Development Tools&#x27; &amp;&amp; \</span><br><span class="line">  sudo yum install -y epel-release &amp;&amp; \</span><br><span class="line">  sudo yum install -y golang libseccomp-devel \  </span><br><span class="line">  squashfs-tools cryptsetup</span><br><span class="line">  yum remove golang # 后面手动安装</span><br></pre></td></tr></table></figure>
<h3 id="install-golang"><a class="markdownIt-Anchor" href="#install-golang"></a> Install Golang</h3>
<p>Golang 有多种安装方法，我是在其官网(<a target="_blank" rel="noopener" href="https://golang.org/doc/install#download">https://golang.org/doc/install#download</a>)下载<a target="_blank" rel="noopener" href="https://golang.org/dl/go1.16.5.linux-amd64.tar.gz">源文件</a>到本地，然后安装到用户目录。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">wget https://golang.org/dl/go1.16.5.linux-amd64.tar.gz</span><br><span class="line">tar -zxvf go1.16.5.linux-amd64.tar.gz</span><br><span class="line"></span><br><span class="line"># 将 “export PATH=&quot;/home/fenglei/local/app/go/bin&quot;:$PATH” 写入bashrc</span><br><span class="line">source ~/.bashrc</span><br><span class="line">which go</span><br><span class="line"># ~/local/app/go/bin/go</span><br><span class="line"></span><br><span class="line"># Verify that you&#x27;ve installed Go by opening a command prompt and typing the following command:</span><br><span class="line">go version</span><br><span class="line"># go version go1.16.5 linux/amd64</span><br></pre></td></tr></table></figure>
<h3 id="compiling-singularityce"><a class="markdownIt-Anchor" href="#compiling-singularityce"></a> Compiling SingularityCE</h3>
<p>参考安装方法如下</p>
<p>You can build SingularityCE using the following commands:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ cd $&#123;GOPATH&#125;/src/github.com/sylabs/singularity &amp;&amp; \</span><br><span class="line">  ./mconfig &amp;&amp; \</span><br><span class="line">  cd ./builddir &amp;&amp; \</span><br><span class="line">  make &amp;&amp; \</span><br><span class="line">  sudo make install</span><br></pre></td></tr></table></figure>
<p>And that’s it! Now you can check your SingularityCE version by running:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ singularity version</span><br></pre></td></tr></table></figure>
<p>To build in a different folder and to set the install prefix to a different path:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ ./mconfig -b ./buildtree -p /usr/local</span><br></pre></td></tr></table></figure>
<p>我的安装如下。</p>
<p>系统自带的gcc（版本4.8.5）编译报错，提示找不到config.c文件。我切换成用户的gcc（版本10.2.0），就可以安装。<br />
启动用户gcc的办法，修改bashrc文件，将下面三行的注释符号去掉，然后source ~/.bashrc即可。以后有必要再回来将这几行注释掉，就恢复系统自带的老版本gcc。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">export PATH=/home/fenglei/local/gcc-10.2.0/bin/:$PATH</span><br><span class="line">export LD_LIBRARY_PATH=/home/fenglei/local/gcc-10.2.0/lib/:/home/fenglei/local/gcc-10.2.0/lib64/:$LD_LIBRARY_PATH</span><br><span class="line">export LD_LIBRARY_PATH=/home/fenglei/local/lib/:/home/fenglei/local/lib64/:$LD_LIBRARY_PATH</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">$ git clone https://github.com/sylabs/singularity.git</span><br><span class="line">$ cd singularity  # 该目录下有一个 mconfig 文件</span><br><span class="line"># ./mconfig -b ./buildtree -p /home/fenglei/local</span><br><span class="line"># ./mconfig -b ./buildtree -p /home/fenglei/local --without-suid</span><br><span class="line">$ ./mconfig    # 默认的安装位置是 /usr/local</span><br><span class="line">$ cd /home/fenglei/local/app/singularity/buildtree</span><br><span class="line">$ make</span><br><span class="line"># make install</span><br><span class="line">$ sudo make install</span><br><span class="line"></span><br><span class="line">$ which singularity</span><br><span class="line">/usr/local/bin/singularity</span><br></pre></td></tr></table></figure>
<p>测试：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ singularity version                                                                                        </span><br><span class="line">3.8.0-rc.1+107-g8b21e5a16</span><br></pre></td></tr></table></figure>
<h3 id="gomap-singularity"><a class="markdownIt-Anchor" href="#gomap-singularity"></a> GOMAP-singularity</h3>
<ol>
<li>Singularity 安装</li>
</ol>
<p>上面已经完成。也可参考：<a target="_blank" rel="noopener" href="https://bioinformapping.com/gomap/master/RUNNING.html">https://bioinformapping.com/gomap/master/RUNNING.html</a></p>
<ol start="2">
<li>从 git 下载 GOMAP-singularity</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/Dill-PICL/GOMAP-singularity.git</span><br><span class="line">cd GOMAP-singularity</span><br></pre></td></tr></table></figure>
<ol start="3">
<li>配置并下载相关参考数据集</li>
</ol>
<p>Run the setup step to make necessary directories and download data files from CyVerse. The pipeline download is large and would require ~40GB of free hard drive space during the setup step.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">$ ./setup.sh </span><br><span class="line"># 由于GO没安好，导致这一步报错，后面重新安装了GO，并重装singularity</span><br><span class="line"># ERROR  : No setuid installation found, for unprivileged installation use: ./mconfig --without-suid</span><br><span class="line"></span><br><span class="line">$ ./setup.sh </span><br><span class="line">0/1 -  0.00% of files done   0.000/39254.793 MB -  0.00% of file sizes done</span><br><span class="line">Processing GOMAP.sif - 39254.793 MB   2021-06-17.15:58:52</span><br><span class="line">GOMAP.sif - 40.000/39254.793 MB -  0.10% done   2021-06-17.16:00:53</span><br><span class="line">GOMAP.sif - 680.000/39254.793 MB -  1.73% done   2021-06-17.16:02:47</span><br><span class="line">GOMAP.sif - 1360.000/39254.793 MB -  3.46% done   2021-06-17.16:04:49</span><br><span class="line">GOMAP.sif - 2000.000/39254.793 MB -  5.09% done   2021-06-17.16:06:40</span><br><span class="line">GOMAP.sif - 2400.000/39254.793 MB -  6.11% done   2021-06-17.16:08:32</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">...</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">GOMAP.sif - 30880.000/39254.793 MB - 78.67% done   2021-06-17.17:59:30</span><br><span class="line">GOMAP.sif - 31520.000/39254.793 MB - 80.30% done   2021-06-17.18:01:43</span><br><span class="line">GOMAP.sif - 32040.000/39254.793 MB - 81.62% done   2021-06-17.18:03:45</span><br><span class="line">GOMAP.sif - 32640.000/39254.793 MB - 83.15% done   2021-06-17.18:05:31</span><br><span class="line">GOMAP.sif - 33200.000/39254.793 MB - 84.58% done   2021-06-17.18:07:27</span><br><span class="line">GOMAP.sif - 33373.426/39254.793 MB - 85.02% done   2021-06-17.18:08:24</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">cliReconnManager: Reconnecting clientState = 0</span><br><span class="line">The client/server socket connection has been renewed</span><br><span class="line">GOMAP.sif - 39254.793/39254.793 MB - 100.00% done   2021-06-17.18:45:46</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<ol start="4">
<li>测试 GOMAP 注释流程</li>
</ol>
<p>[optional]: Test whether the container and the data files are working as intended.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">$ ./test.sh</span><br><span class="line"></span><br><span class="line">Running Sequence-similarity based Annotation Step</span><br><span class="line">/usr/local/lib/python2.7/dist-packages/Bio/SearchIO/__init__.py:211: BiopythonExperimentalWarning: Bio.SearchIO is an experimental submodule which may undergo significant changes prior to its future official release.</span><br><span class="line">  BiopythonExperimentalWarning)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Building a new DB, current time: 06/17/2021 20:59:09</span><br><span class="line">New DB name:   /workdir/test/GOMAP-0.3_GOMAP-input/input/0.3_GOMAP-input.fa</span><br><span class="line">New DB title:  /workdir/test/GOMAP-0.3_GOMAP-input/input/0.3_GOMAP-input.fa</span><br><span class="line">Sequence type: Protein</span><br><span class="line">...</span><br><span class="line">Checking gaf file for simple errors and fixing them</span><br><span class="line">There were 34 warnings (use warnings() to see them)</span><br><span class="line">[1] &quot;Checking if data/data/go/go.obo.data exists&quot;</span><br><span class="line">[1] &quot;data/data/go/go.obo.data exists so loading R object&quot;</span><br><span class="line">   user  system elapsed </span><br><span class="line">  2.547   0.078   2.624 </span><br><span class="line">[1] &quot;Reading All Datasets&quot;</span><br><span class="line">Checking gaf file for simple errors and fixing them</span><br><span class="line">[1] 1714</span><br><span class="line">[1] &quot;Removing Redundancy&quot;</span><br><span class="line">[1] 18</span><br><span class="line">Compiling Non-Redundant Comprehensive Data</span><br><span class="line">There were 50 or more warnings (use warnings() to see the first 50)</span><br></pre></td></tr></table></figure>
<ol start="5">
<li>实战篇：学习自定义配置文件 Edit the config file</li>
</ol>
<p>a. Declare export GOMAP_LOC environment variable. Edit <code>~/.bashrc</code> as below:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">export GOPATH=&quot;/home/fenglei/local/app/go&quot;</span><br><span class="line">export GOMAP_LOC=&quot;/home/fenglei/local/app/GOMAP-singularity/install/location&quot;</span><br></pre></td></tr></table></figure>
<p>b. Download the <a target="_blank" rel="noopener" href="https://bioinformapping.com/gomap/master/_static/min-config.yml">config.yml</a> file and make necessary changes. Change the highlighted lines to fit your input data.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">#Input section </span><br><span class="line">input:</span><br><span class="line">  #input fasta file name</span><br><span class="line">  fasta: GOMAP-input.fa</span><br><span class="line">  # output file basename</span><br><span class="line">  basename: GOMAP_output</span><br><span class="line">  #input NCBI taxonomy id</span><br><span class="line">  taxon: &quot;123456&quot;</span><br><span class="line">  # Name of the species</span><br><span class="line">  species: &quot;Glycine max&quot;</span><br><span class="line">  # Email is mandatory</span><br><span class="line">  email: myname#gmail.com</span><br><span class="line">  #Number of CPUs used for tools</span><br><span class="line">  cpus: 4</span><br><span class="line">  #Whether openmpi should be used</span><br><span class="line">  mpi: False</span><br><span class="line">  #what the name of the temporary directory is</span><br><span class="line">  tmpdir: &quot;/tmpdir&quot;</span><br></pre></td></tr></table></figure>
<ol start="6">
<li>Run pipeline</li>
</ol>
<p>可直接修改 <a target="_blank" rel="noopener" href="http://test.sh">test.sh</a> 并运行该脚本。也可参考软件手册分步骤运行。<br />
下面摘抄的是在单机（SINGLE）运行的模式，软件手册里也有在集群（MPI）运行的示范代码。<br />
<a target="_blank" rel="noopener" href="http://test.sh">test.sh</a> 与 test目录在同一个路径。GOMAP-input.fa 与 config.yml 都在test目录下。输入的蛋白序列，其尾部如果有表示stop codon的星号，要移除星号。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"># seqsim</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=seqsim --config=test/config.yml</span><br><span class="line"># domain</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=domain --config=test/config.yml</span><br><span class="line"># fanngo</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=fanngo --config=test/config.yml</span><br><span class="line"># mixmeth-blast</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=mixmeth-blast --config=test/config.yml</span><br><span class="line"># mixmeth-preproc</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=mixmeth-preproc --config=test/config.yml</span><br><span class="line"># mixmeth</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=mixmeth --config=test/config.yml</span><br><span class="line"># aggregate</span><br><span class="line"># Attention: Please wait for all your Argot2.5 jobs to finish before running this step. You will get emails from Argot2.5 when your jobs are submitted and when they are finished. You can also check the status of all current jobs from all users here.</span><br><span class="line">./run-GOMAP-SINGLE.sh --step=aggregate --config=test/config.yml</span><br></pre></td></tr></table></figure>
<h3 id="附注用-swissprot-与-idmapping-来做-go"><a class="markdownIt-Anchor" href="#附注用-swissprot-与-idmapping-来做-go"></a> 附注：用 SwissProt 与 idmapping 来做 GO</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">## SwissProt 文件是氨基酸序列，比较大</span><br><span class="line">wget ftp://ftp.ncbi.nih.gov/blast/db/FASTA/swissprot.gz</span><br><span class="line">## idmapping.tb.gz 文件可能下载链接有变化</span><br><span class="line">wget ftp://ftp.pir.georgetown.edu/databases/idmapping/idmapping.tb.gz</span><br><span class="line">## swissprot index 建立</span><br><span class="line">diamond makedb --in swissprot -d swissprot</span><br><span class="line">## 序列比对</span><br><span class="line">diamond blastp -d /PATH/TO/index/swissprot -q input.proteins.fa -k 1 -e 0.00001 -o swissprot_blastp.out</span><br><span class="line">## idmapping 文件整理，下文有摘抄文件具体内容</span><br><span class="line">awk -F &quot;\t&quot;  &#x27;&#123;print $1&quot;\t&quot;$8&#125;&#x27; idmapping.tb &gt; uniprotID-GOnumber.tab</span><br><span class="line">## 将“基因-Uniprot-GO”的关系建立起来</span><br><span class="line">perl UniProt2GO_annotate.pl uniprotID-GOnumber.tab swiss_dia_matches.m8 swiss_GO_matches.tab</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ head idmapping.tb </span><br><span class="line">Q6GZX4  001R_FRG3G      2947773 YP_031579.1     81941549; 49237298              PF04947 GO:0006355; GO:0046782; GO:0006351                      UniRef100_Q6GZX4        UniRef90_Q6GZX4 UniRef50_Q6GZX4 UPI00003B0FD4           65492415165820 AY548484        AAT09660.1</span><br><span class="line">Q6GZX3  002L_FRG3G      2947774 YP_031580.1     49237299; 81941548; 47060117            PF03003 GO:0033644; GO:0016021                  UniRef100_Q6GZX3        UniRef90_Q6GZX3 UniRef50_Q6GZX3 UPI00003B0FD5           654924        15165820 AY548484        AAT09661.1</span><br><span class="line">Q197F8  002R_IIV3       4156251 YP_654574.1     109287880; 123808694; 106073503                                         UniRef100_Q197F8        UniRef90_Q197F8 UniRef50_Q197F8 UPI0000D83464           345201                        16912294 DQ643392        ABF82032.1</span><br><span class="line">Q197F7  003L_IIV3       4156252 YP_654575.1     106073504; 109287881; 123808693                                         UniRef100_Q197F7        UniRef90_Q197F7 UniRef50_Q197F7 UPI0000D83465           345201                        16912294 DQ643392        ABF82033.1</span><br><span class="line">Q6GZX2  003R_FRG3G      2947775 YP_031581.1     81941547; 49237300                                              UniRef100_Q6GZX2        UniRef90_Q6GZX2 UniRef50_Q6GZX2 UPI00003B0FD6           654924                          15165820       AY548484        AAT09662.1</span><br><span class="line">Q6GZX1  004R_FRG3G      2947776 YP_031582.1     49237301; 81941546                      GO:0033644; GO:0016021                  UniRef100_Q6GZX1        UniRef90_Q6GZX1 UniRef50_Q6GZX1 UPI00003B0FD7           654924                15165820 AY548484        AAT09663.1</span><br><span class="line">Q197F5  005L_IIV3       4156254 YP_654577.1     123808691; 109287883; 106073506                                         UniRef100_Q197F5        UniRef90_Q197F5 UniRef50_Q197F5 UPI0000D83467           345201                        16912294 DQ643392        ABF82035.1</span><br><span class="line">Q6GZX0  005R_FRG3G      2947777 YP_031583.1     47060120; 49237302; 81941545            PF02393                         UniRef100_Q6GZX0        UniRef90_Q6GZX0 UniRef50_Q6GZX0 UPI00003B0FD8           654924                        15165820 AY548484        AAT09664.1</span><br><span class="line">Q91G88  006L_IIV6       1733056 NP_149469.1     82012333; 15078718; 15042163            PF04383; PF12299                                UniRef100_Q91G88        UniRef90_Q91G88 UniRef50_Q91G88 UPI00000F3D32           176652        2820141; 3959991; 7698884; 8492091; 17239238; 8021587; 10456793; 8121799; 9926400; 16789247; 11448171; 1475907; 3201750; 9482589; 1549908; 8073636       AF303741        AAK81943.1</span><br><span class="line">Q6GZW9  006R_FRG3G      2947778 YP_031584.1     81941544; 49237303                                              UniRef100_Q6GZW9        UniRef90_Q6GZW9 UniRef50_Q6GZW9 UPI00003B0FD9           654924                          15165820       AY548484        AAT09665.1</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ head swiss_GO_matches.tab</span><br><span class="line">Q6GZX4  GO:0006355; GO:0046782; GO:0006351</span><br><span class="line">Q6GZX3  GO:0033644; GO:0016021</span><br><span class="line">Q197F8</span><br><span class="line">Q197F7</span><br><span class="line">Q6GZX2</span><br><span class="line">Q6GZX1  GO:0033644; GO:0016021</span><br><span class="line">Q197F5</span><br><span class="line">Q6GZX0</span><br><span class="line">Q91G88</span><br><span class="line">Q6GZW9</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ head swiss_GO_matches.tab</span><br><span class="line">gene1 Q9M9M9  GO:0050897; GO:0009055; GO:0008137; GO:0009735; GO:0006979; GO:0005747; GO:0005739</span><br><span class="line">gene2 Q9LP77  GO:0005524; GO:0004672; GO:0016021; GO:0005886; GO:0009506; GO:0005774</span><br><span class="line">gene3 Q9CAL2  GO:0005524; GO:0004674; GO:0016021; GO:0005886</span><br><span class="line">gene4 A4FLF6</span><br><span class="line">gene5 Q8RY67  GO:0005524; GO:0004674; GO:0007166; GO:0016021; GO:0005886</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">$ cat UniProt2GO_annotate.pl</span><br><span class="line">#!/usr/bin/perl</span><br><span class="line">use warnings;</span><br><span class="line">use strict;</span><br><span class="line"></span><br><span class="line">open IDMAPPING, $ARGV[0] or die $!;</span><br><span class="line">open BLASTOUT, $ARGV[1] or die $!;</span><br><span class="line">open OUT, &quot;&gt;$ARGV[2]&quot; or die $!;</span><br><span class="line">select OUT;</span><br><span class="line"></span><br><span class="line">my %id2go;</span><br><span class="line"></span><br><span class="line">while(&lt;IDMAPPING&gt;)&#123;</span><br><span class="line">        chomp $_;</span><br><span class="line">        my @array1=split(/\t/);</span><br><span class="line">        $id2go&#123;$array1[0]&#125;=$array1[1];</span><br><span class="line">#       print &quot;$array1[1]\n&quot;; exit;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">while(&lt;BLASTOUT&gt;)&#123;</span><br><span class="line">        chomp $_;</span><br><span class="line">        my @array2=split(/\t/);</span><br><span class="line">        my $geneid=$array2[0];</span><br><span class="line">        my $UniprotID=$array2[1];</span><br><span class="line">#       print &quot;$UniprotID\n&quot;;</span><br><span class="line">        $UniprotID =~ s/\.[0-9]$//;</span><br><span class="line">#       print &quot;$geneid\t$&#123;UniprotID&#125;test\n&quot;; exit;</span><br><span class="line">        if(exists $id2go&#123;$UniprotID&#125;)&#123; print &quot;$geneid\t$UniprotID\t$id2go&#123;$UniprotID&#125;\n&quot;;&#125;</span><br><span class="line">        else&#123;print &quot;$geneid\t$UniprotID\n&quot;;&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">close IDMAPPING;</span><br><span class="line">close BLASTOUT;</span><br><span class="line">close OUT;</span><br></pre></td></tr></table></figure>
<p>经过一系列的转换，得到下面格式的文件，就可以在R程序里面用于GO数据库的构建。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ zcat GMA_GO_TERM.txt.gz | head</span><br><span class="line">GID     GO      ONTOLOGY        EVIDENCE</span><br><span class="line">Glyma.01G000100 GO:0006412      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009063      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009234      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0016021      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0031969      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0016835      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0030976      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0070204      F       IEA</span><br><span class="line">Glyma.01G000137 GO:0016043      P       IEA</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/08/Linux-PKG-CONFIG-PATH/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/08/Linux-PKG-CONFIG-PATH/" class="post-title-link" itemprop="url">Linux | 认识 PKG_CONFIG_PATH</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-06-08 15:55:09 / Modified: 16:21:29" itemprop="dateCreated datePublished" datetime="2021-06-08T15:55:09+08:00">2021-06-08</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>安装 <a target="_blank" rel="noopener" href="https://github.com/agordon/fastx_toolkit/releases/download/0.0.14/fastx_toolkit-0.0.14.tar.bz2">Fastx-Toolkit</a> 的过程中遇问题：No package ‘gtextutils’ found. Consider adjusting the PKG_CONFIG_PATH environment variable if you.</p>
<p>实际上我已经安装 <a target="_blank" rel="noopener" href="https://github.com/agordon/libgtextutils/releases/download/0.7/libgtextutils-0.7.tar.gz">libgtextutils</a></p>
<p>解决办法：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sudo cp /usr/local/lib/pkgconfig/gtextutils.pc /usr/lib64/pkgconfig/</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=1490">参考资料</a>：</p>
<p>安装源码包，进行 configure 步骤，提示某依赖的软件包找不到或软件包版本太低；或某一个库文件找不到。而实际上却已经安装好了相应的软件。这两种情况的处理方法分别为：</p>
<ol>
<li>设置环境变量 PKG_CONFIG_PATH</li>
</ol>
<p>.pc 文件是安装软件后生成的一个文件，其中包含该软件的信息；该 .pc 文件一般统一存放于名为 pkgconfig 的目录下。命令 pkg-config 按照系统设置路径（CentOS6 x64_86 系统默认为 /usr/lib64/pkgconfig/）的先后顺序来搜索 .pc 文件，从而或得到软件的版本号、头文件和库文件位置等信息。</p>
<p>因此当把软件包安装到默认位置 /usr/local/ 位置后，则其 .pc 文件会生成到 /usr/local/lib/pkgconfig/ 或 /usr/local/lib64/pkgconfig/ 文件夹下，此时只需要将相应的 .pc 文件复制到 /usr/lib64/pkgconfig/ 中即可。</p>
<p>或者，在安装软件包的时候，使用 “./configure –prefix=/usr/” 将软件直接安装到 /usr/ 目录下。这时有可能会将 .pc 文件生成到 /usr/lib/pkgconfig/ 中，依然导致问题出现，只需将该目录中的 .pc 文件复制到 /usr/lib64/pkgconfig/ 中即可。</p>
<p>或者，设置环境变量PKG_CONFIG_PATH，加入含有相应的.pc文件的路径。注意要使用命令 ‘export’ 。比如:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig/:$PKG_CONFIG_PATH</span><br></pre></td></tr></table></figure>
<p>使用以上3种方法，能解决软件版本不符或系统找不到依赖的软件包的问题。</p>
<ol start="2">
<li>设置环境变量 LD_LIBRARY_PATH</li>
</ol>
<p>Linux 系统默认把 /lib 和 /usr/lib 两个目录作为默认的库搜索路径。而修改库搜索路径有两个方法：</p>
<p>(1) 在 /etc/ld.so.conf 文件中添加库的搜索路径。</p>
<p>(2) 在环境变量 LD_LIBRARY_PATH 中指明库的搜索路径。</p>
<p>将库的路径设置好后，需要使用 /sbin/ldconfig 命令来将这些搜索路径下的共享库文件集中在一起而生成 /etc/ld.so.cache。该 cache 用于快速定位共享库。</p>
<p>特别是新安装了软件后，即使其库文件明明就在库搜索路径下，但是编译的时候依然报错，提示缺少该库。则需要使用 ldconfig 命令来更新缓存。</p>
<p>更详细信息，可参考：<a target="_blank" rel="noopener" href="http://www.mike.org.cn/articles/description-configure-pkg-config-pkg_config_path-of-the-relations-between/">简述configure、pkg-config、pkg_config_path三者的关系</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/03/Bioinfo-jcvi-installation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/03/Bioinfo-jcvi-installation/" class="post-title-link" itemprop="url">jcvi 安装与使用，报错以及解决办法</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-06-03 17:36:50" itemprop="dateCreated datePublished" datetime="2021-06-03T17:36:50+08:00">2021-06-03</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-06-07 23:04:42" itemprop="dateModified" datetime="2021-06-07T23:04:42+08:00">2021-06-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>jcvi 是 Haibao Tang 开发的基因组组装注释与比较基因组分析工具，比如可以用于可视化 synteny blocks。原理是依据物种之间基因编码序列相似性比较，如果若干个连续的基因在两个基因组之间都有出现，那这若干个基因就是一个 synteny block，可能是从祖先染色体上继承下来的，物种之间的亲缘关系越近，synteny blocks 数目越多。</p>
<p>之前<a target="_blank" rel="noopener" href="https://genehub.wordpress.com/2019/07/05/mcscanx-%e6%b5%8b%e8%af%95/">测试过 MCScanX</a>，其实 jcvi 和 MCScanX 就是一回事，是同一个作者开发的，jcvi 是升级版，操作更加便利，生成的图片更加精美。回去 <a target="_blank" rel="noopener" href="https://github.com/wyp1125/MCScanX#installation">MCScanX</a> 看了一下开发时间，最后更新大约是 2014 年，但是里面有的程序还是很实用的，比如可以计算共线性区块之间的基因对的 Ka/Ks value（add_kaks_to_synteny.pl）.</p>
<h3 id="安装-jcvi"><a class="markdownIt-Anchor" href="#安装-jcvi"></a> 安装 jcvi</h3>
<p>jcvi 软件推荐使用 pip 安装，也可以选择使用 conda 安装。我是 anaconda3 下面建立了 python 2.7 和 python 3.7 两个环境，分别进行了测试，最后发现 python2.7 环境下 jcvi 可以正常运行。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">source activate python27</span><br><span class="line">conda install -c bioconda jcvi</span><br><span class="line">conda install -c bioconda last</span><br><span class="line">conda install scipy</span><br><span class="line"># 但是运行序列比对时候报错！--&gt; 经过查询是 lastal 的问题，也就是别用 conda 安装的 last，自己去 last 网站下载源码，安装到用户目录下，然后将其bin路径写入环境变量bashrc文件，然后就可以解决比对报错的问题。</span><br></pre></td></tr></table></figure>
<h3 id="安装-lastal"><a class="markdownIt-Anchor" href="#安装-lastal"></a> 安装 lastal</h3>
<p>手动安装 last 软件，并将其路径写入环境变量。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">git clone https://gitlab.com/mcfrith/last.git</span><br><span class="line">cd last</span><br><span class="line">mdke</span><br><span class="line"># then we can see lastal in ./last/bin/</span><br><span class="line"># put /PATH/TO/last/bin into ~/.bashrc</span><br><span class="line">which lastal</span><br><span class="line">(python37) [fenglei@localhost cds]$ which lastal</span><br><span class="line">~/local/app/anaconda3/envs/python37/bin/lastal</span><br><span class="line">(python37) [fenglei@localhost cds]$ source ~/.bashrc</span><br><span class="line">(base) [fenglei@localhost cds]$ which lastal</span><br><span class="line">~/local/app/last/bin/lastal</span><br></pre></td></tr></table></figure>
<p>下面开始做一个数据测试，参考：<a target="_blank" rel="noopener" href="https://github.com/tanghaibao/jcvi/wiki/MCscan-(Python-version)">https://github.com/tanghaibao/jcvi/wiki/MCscan-(Python-version)</a></p>
<h3 id="基因坐标与基因序列数据准备"><a class="markdownIt-Anchor" href="#基因坐标与基因序列数据准备"></a> 基因坐标与基因序列数据准备</h3>
<p>原始数据：（1）gff3 格式的基因坐标；（2）fasta 格式的基因 CDS 序列。使用 jcvi 内置程序对这两个文件做格式整理。以大豆（Glycine max）和野生豆（Glycine soja）为例，这一步需要得到四个文件：gma.bed gma.cds gso.bed gso.cds</p>
<p>如下图所示，要注意同一个物种的 bed 文件和 cds 文件的基因 ID 要相互匹配。由于基因 ID 通常带有“.1”或“.2p”这样的后缀，需要注意，必要的时候自己写脚本处理数据。</p>
<p><img src="https://i.imgur.com/RYAdtK3.jpg" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line">### BED FORMAT FILES FOR GENES</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=transcript_id Gmax_508_Wm82.a4.v1.gene.gff3.gz &gt; gma.bed</span><br><span class="line">11:12:31 [base] Load file `Gmax_508_Wm82.a4.v1.gene.gff3.gz`</span><br><span class="line">11:13:30 [gff] Extracted 86256 features (type=mRNA id=transcript_id)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Gmax_508_Wm82.a4.v1.gene.gff3.gz &gt; gma.bed2</span><br><span class="line">11:20:14 [base] Load file `Gmax_508_Wm82.a4.v1.gene.gff3.gz`</span><br><span class="line">11:21:13 [gff] Extracted 86256 features (type=mRNA id=Name)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=transcript_id Gsoja.W05.gene.gff.gz &gt; gso.bed</span><br><span class="line">11:14:15 [base] Load file `Gsoja.W05.gene.gff.gz`</span><br><span class="line">11:15:31 [gff] Extracted 89477 features (type=mRNA id=transcript_id)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Gsoja.W05.gene.gff.gz &gt; gso.bed2             </span><br><span class="line">11:22:16 [base] Load file `Gsoja.W05.gene.gff.gz`</span><br><span class="line">11:23:32 [gff] Extracted 89477 features (type=mRNA id=Name)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=ID Gsoja.W05.gene.gff.gz &gt; gso.bed3    </span><br><span class="line">11:27:42 [base] Load file `Gsoja.W05.gene.gff.gz`</span><br><span class="line">11:28:58 [gff] Extracted 89477 features (type=mRNA id=ID)</span><br><span class="line"></span><br><span class="line">### REMOVE DUPLICATIONS</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.bed uniq gma.bed2 </span><br><span class="line">11:35:34 [base] Load file `gma.bed2`</span><br><span class="line">11:35:37 [bed] Imported: 86256, Exported: 51013</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.bed uniq gso.bed3</span><br><span class="line">11:37:25 [base] Load file `gso.bed3`</span><br><span class="line">11:37:29 [bed] Imported: 89477, Exported: 55103</span><br><span class="line"></span><br><span class="line">### CDS FORMAT FILE FOR UNIQ GENES</span><br><span class="line"></span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gma.uniq.bed ) Gmax_508_Wm82.a4.v1.cds.fa.gz |  seqkit seq -i &gt; gma.cds</span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gso.uniq.bed ) Gsoja.W05.gene.cds.fa.gz |  seqkit seq -i &gt; gso.cds</span><br><span class="line"></span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gma.uniq.bed ) Gmax_508_Wm82.a4.v1.protein.fa |  seqkit seq -i &gt; gma.pep  # ERROR due to id isuues</span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gso.uniq.bed ) Gsoja.W05.gene.protein.fa.gz |  seqkit seq -i &gt; gso.pep</span><br><span class="line"></span><br><span class="line"># TROUBLESHOOTING</span><br><span class="line">awk &#x27;&#123;if($0~/&gt;/)&#123;gsub(/\.p/,&quot;&quot;,$1)&#125;; &#123;print&#125;&#125;&#x27; Gmax_508_Wm82.a4.v1.protein.fa &gt; Gmax_508_Wm82.a4.v1.protein.new.fa</span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gma.uniq.bed ) Gmax_508_Wm82.a4.v1.protein.new.fa |  seqkit seq -i &gt; gma.pep</span><br></pre></td></tr></table></figure>
<h3 id="apairwise_synteny_searchsh"><a class="markdownIt-Anchor" href="#apairwise_synteny_searchsh"></a> a.pairwise_synteny_search.sh</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/local/app/anaconda3/envs/python27/bin/python -m jcvi.compara.catalog ortholog gma gso --no_strip_names</span><br></pre></td></tr></table></figure>
<p>该命令直接对物种间基因序列进行比较，将生成一个 gma.gso.pdf 文件，打开如下。</p>
<p><img src="https://i.imgur.com/Ly4FSpT.png" alt="" /></p>
<p>同时生成一堆文件：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">-rw-rw-r-- 1 fenglei fenglei    139507 Jun  3 12:09 gma.gso.pdf</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   5333469 Jun  3 12:09 gma.gso.lifted.anchors</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   2931001 Jun  3 12:08 gma.gso.anchors</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   5953786 Jun  3 12:08 gma.gso.last.filtered</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  97488440 Jun  3 12:08 gma.gso.last</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  58121140 Jun  3 12:08 gso.bck</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei       515 Jun  3 12:08 gso.prj</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 264000792 Jun  3 12:08 gso.suf</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   1101303 Jun  3 12:07 gso.des</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    220416 Jun  3 12:07 gso.sds</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    220416 Jun  3 12:07 gso.ssp</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  66055302 Jun  3 12:07 gso.tis</span><br></pre></td></tr></table></figure>
<p>The <code>.last</code> file is raw LAST output, <code>.last.filtered</code> is filtered LAST output, <code>.anchors</code> is the seed synteny blocks (high quality), <code>.lifted.anchors</code> recruits additional anchors to form the final synteny blocks.</p>
<p>查看了一下 .lifted.anchors 与 .anchors 的区别，如下看到两个文件都有 551 个 synteny blocks，只是 blocks 内部的基因配对数目不同，应该就是后者有更严格的筛选条件。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.lifted.anchors | grep -v &#x27;###&#x27; | wc -l</span><br><span class="line">124133</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.lifted.anchors | grep &#x27;###&#x27; | wc -l   </span><br><span class="line">551</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.anchors | grep -v &#x27;###&#x27; | wc -l        </span><br><span class="line">68549</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.anchors | grep &#x27;###&#x27; | wc -l   </span><br><span class="line">551</span><br></pre></td></tr></table></figure>
<h3 id="bdepth_visualiztionsh"><a class="markdownIt-Anchor" href="#bdepth_visualiztionsh"></a> b.depth_visualiztion.sh</h3>
<p>We could also quick test if the synteny pattern is indeed 1:1, by running:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.compara.synteny depth --histogram gma.gso.anchors</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/BB2UsVk.jpg" alt="" /></p>
<h3 id="canchorssimple_file_generationsh"><a class="markdownIt-Anchor" href="#canchorssimple_file_generationsh"></a> c.anchors.simple_file_generation.sh</h3>
<p>This command generates a .simple file which is a more succinct form than the .anchors file.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.compara.synteny screen --minspan=30 --simple gma.gso.anchors gma.gso.anchors.new</span><br></pre></td></tr></table></figure>
<p>将 <code>gma.gso.anchors</code> 文件转化成 <code>gma.gso.anchors.simple</code> 文件。有什么不同？见下面的文本框。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">(python27) [fenglei@localhost gma-gso]$ head gma.gso.anchors</span><br><span class="line">###</span><br><span class="line">Glyma.01G000100.1       Glysoja.01G000001.1     496</span><br><span class="line">Glyma.01G000322.1       Glysoja.01G000002.1     775</span><br><span class="line">Glyma.01G000400.1       Glysoja.01G000003.1     2600</span><br><span class="line">Glyma.01G001000.2       Glysoja.01G000010.2     984</span><br><span class="line">Glyma.01G001100.3       Glysoja.01G000011.2     1490</span><br><span class="line">Glyma.01G001300.3       Glysoja.01G000013.2     3560</span><br><span class="line">Glyma.01G001350.1       Glysoja.01G000014.1     381</span><br><span class="line">Glyma.01G001500.1       Glysoja.01G000015.3     2880</span><br><span class="line">Glyma.01G001700.1       Glysoja.01G000017.1     1650</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ head gma.gso.anchors.simple</span><br><span class="line">Glyma.01G000100.1       Glyma.01G245700.6       Glysoja.01G000001.1     Glysoja.01G002529.1     2354    +</span><br><span class="line">Glyma.01G023500.1       Glyma.01G044800.1       Glysoja.02G002717.1     Glysoja.02G002929.1     209     -</span><br><span class="line">Glyma.01G045100.2       Glyma.01G071700.1       Glysoja.02G003545.1     Glysoja.02G003757.1     210     +</span><br><span class="line">Glyma.01G083700.1       Glyma.01G090200.1       Glysoja.02G003470.1     Glysoja.02G003510.1     46      +</span><br><span class="line">Glyma.01G098400.6       Glyma.01G106600.2       Glysoja.03G006294.1     Glysoja.03G006391.1     76      +</span><br><span class="line">Glyma.01G125600.2       Glyma.01G145000.1       Glysoja.03G005845.1     Glysoja.03G006096.1     198     -</span><br><span class="line">Glyma.01G118000.1       Glyma.01G122800.1       Glysoja.03G006181.1     Glysoja.03G006214.2     37      +</span><br><span class="line">Glyma.01G107000.1       Glyma.01G113400.1       Glysoja.03G006412.2     Glysoja.03G006472.1     56      +</span><br><span class="line">Glyma.01G003900.1       Glyma.01G008500.1       Glysoja.05G013057.1     Glysoja.05G013097.1     40      +</span><br><span class="line">Glyma.01G001900.2       Glyma.01G007300.1       Glysoja.07G017673.1     Glysoja.07G017734.1     56      -</span><br></pre></td></tr></table></figure>
<h3 id="dmacrosynteny_visualizationsh"><a class="markdownIt-Anchor" href="#dmacrosynteny_visualizationsh"></a> d.Macrosynteny_visualization.sh</h3>
<p>准备一个 <code>seqids</code> 文件，定义哪些染色体需要展示出来。文件内容是：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Gm01,Gm02,Gm03,Gm04,Gm05,Gm06,Gm07,Gm08,Gm09,Gm10,Gm11,Gm12,Gm13,Gm14,Gm15,Gm16,Gm17,Gm18,Gm19,Gm20</span><br><span class="line">Chr01,Chr02,Chr03,Chr04,Chr05,Chr06,Chr07,Chr08,Chr09,Chr10,Chr11,Chr12,Chr13,Chr14,Chr15,Chr16,Chr17,Chr18,Chr19,Chr20</span><br></pre></td></tr></table></figure>
<p>可以用 shell 代码生成上面的文件，我的命令如下。注意 seqids 内容上下行代表两个物种，跟下文出现的 layout 文件要顺序对应。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">for i in $(seq 1 20); do printf &quot;Gm&quot;%02d&quot;,&quot; $i ; done; echo </span><br><span class="line">for i in $(seq 1 20); do printf &quot;Chr&quot;%02d&quot;,&quot; $i ; done; echo</span><br></pre></td></tr></table></figure>
<p>再准备一个 <code>layout</code> 文件。定义染色体出现的位置，角度。文件内容是：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># y, xstart, xend, rotation, color, label, va,  bed</span><br><span class="line"> .6,     .1,    .8,       0,      , gma, top, gma.bed</span><br><span class="line"> .4,     .1,    .8,       0,      , gso, top, gso.bed</span><br><span class="line"># edges</span><br><span class="line">e, 0, 1, gma.gso.anchors.simple</span><br></pre></td></tr></table></figure>
<h3 id="ekaryotypesh"><a class="markdownIt-Anchor" href="#ekaryotypesh"></a> <a target="_blank" rel="noopener" href="http://e.karyotype.sh">e.karyotype.sh</a></h3>
<p>画图文件准备完毕，运行画图命令：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.graphics.karyotype seqids layout</span><br></pre></td></tr></table></figure>
<p>将生成一个 karyotype.pdf 文件，打开如下。</p>
<p><img src="https://i.imgur.com/Bugyz9W.jpeg" alt="" /></p>
<p>以上就是两个物种间的共线性比较分析。</p>
<h3 id="同一个物种内部的-synteny-呢"><a class="markdownIt-Anchor" href="#同一个物种内部的-synteny-呢"></a> 同一个物种内部的 synteny 呢？</h3>
<p>由于大豆在近期经历了 WGD 事件，导致很多基因都有双拷贝，所以内部就有 synteny，我做了测试分析，用与上文相同的代码，得到图片如下。看来 jcvi 程序会自动滤掉基因的自我比较，而是做 paralog 的比较。</p>
<p><img src="https://i.imgur.com/9QhsL89.jpg" alt="" /></p>
<h3 id="其他案例"><a class="markdownIt-Anchor" href="#其他案例"></a> 其他案例</h3>
<p>软件的文档里面还列出了三个物种的比较、对某些 synteny 区段进行高亮、只展示特定染色体区间的方法。下面直接摘抄原作者的图片，留作参考。</p>
<p><img src="https://camo.githubusercontent.com/833e373b369f1c221132d729e1cf0128a8cd638fc04c90782ad7d5e9d30867db/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f39766c337973336e6476696d6734632f67726170652d70656163682d636163616f2e706e673f7261773d31" alt="" /></p>
<p><img src="https://camo.githubusercontent.com/5b300c541bd65669e40cf15f7e496fd0a63a0a47ecb4c80a8f5641e9253c9981/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f736f3274643065716b6c626b6c6e6c2f67726170652e70656163682e6b6172796f747970652d677265656e2e706e673f7261773d31" alt="" /></p>
<p><img src="https://camo.githubusercontent.com/7a878ab0e1aa1221d9c4af36a4d0f131eb43f8ddb10caea79e53366d6797cd92/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f696d646f7539756b6e74743474616d2f67726170652d70656163682d636163616f2d626c6f636b732e706e673f7261773d31" alt="" /></p>
<h3 id="其他笔记"><a class="markdownIt-Anchor" href="#其他笔记"></a> 其他笔记</h3>
<p>下面是软件安装调试的一些笔记。python3.7 环境我也测试了，可以安装 jcvi，而且我发现 python3.7 的 jcvi 与 python2.7 下面的 jcvi 的屏幕输出不同。但是 pyton2.7 jcvi 运行过程中有些步骤报错（[Errno 13] Permission denied: ‘dvipng’）导致缺失一些数据结果，所以我放弃了python3.7 jcvi，转而用 python2.7 jcvi 来做分析。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">conda install scipy</span><br><span class="line">conda install -c bioconda last</span><br><span class="line">conda install -c bioconda jcvi  # 这一条就报错！</span><br><span class="line"># 于是从github手动下载软件尝试</span><br><span class="line">cd ~/local/app  # or any directory of your choice</span><br><span class="line">git clone git://github.com/tanghaibao/jcvi.git</span><br><span class="line">pip install -e .</span><br><span class="line"># 安装成功。能否运行数据比对？</span><br><span class="line"># 依然不能，经查是 lastal 的问题！！！</span><br><span class="line">## lastal: I was installed here with multi-threading disabled  # 运行到lastal就报错，根本没有比对结果输出。</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">conda create -n jcvi python=2.7</span><br><span class="line">source activate jcvi</span><br><span class="line"># 安装前需要确认存在bioconda和r channel，安装前先更新conda</span><br><span class="line"># conda --add-channel r</span><br><span class="line"># conda install jcvi last</span><br><span class="line">conda install -c bioconda last</span><br></pre></td></tr></table></figure>
<p>后续还可能会由于缺少 scipy 在绘图时报错 ImportError: No module named scipy.spatial ，因此可提<br />
前安装</p>
<p><strong>安装scipy</strong><br />
#conda-forge scipy-1.2.0<br />
conda install scipy</p>
<p><strong>问题一</strong></p>
<p>File “/biosoft/python/Python-v2.7.11/lib/python2.7/site-packages/subprocess32.py”, line 1393, in _execute_child<br />
    raise child_exception_type(errno_num, err_msg)<br />
OSError: [Errno 2] No such file or directory: ‘kpsewhich’</p>
<p>解决办法：linux中缺少kpsewhich命令导致的，于是安装该命令，CentOS系统方法如下：</p>
<p>sudo yum install tetex-latex tetex-doc tetex-fonts<br />
sudo yum install texlive tex-live-latex</p>
<p>安装好后再次运行，顺利通过。</p>
<p>我的系统是redhat，没有安yum，所以下载了textlive的源码编译，但是一直不能通过，后来重装了一个centOS系统才可以用。</p>
<p><strong>问题二</strong></p>
<p>原来在用户目录下安装了GCC-6.5.0，可以正常使用jcvi的；但是昨天升级到了GCC-9.2.0之后，今天用jcvi 画synteny图的时候遇到下面的问题。</p>
<p>ImportError: /home/fenglei/local/lib64/libstdc++.so.6: undefined symbol: libiconv</p>
<p>将libstdc++.so.6这个软连接重新指向旧版GCC对应的libstdc++.so.6.0.22，然后jcvi就没有报错了。</p>
<p><img src="https://genehub.files.wordpress.com/2019/09/e5beaee4bfa1e688aae59bbe_20190906173207.png" alt="微信截图_20190906173207" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/" class="post-title-link" itemprop="url">Bioinfo | Installation of OrthoMCL-pipeline</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-06-02 13:36:29" itemprop="dateCreated datePublished" datetime="2021-06-02T13:36:29+08:00">2021-06-02</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-14 18:33:24" itemprop="dateModified" datetime="2022-07-14T18:33:24+08:00">2022-07-14</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>OrthoMCL (<a target="_blank" rel="noopener" href="http://orthomcl.org/orthomcl/">http://orthomcl.org/orthomcl/</a>) 是现在用的最多的一款来找直系同源基因（Orthologs）以及旁系同源基因 (Paralog) 的软件。OrthoMCL 原始软件安装比较繁琐，根据官网的教程至少得十多步才能完成整个运行流程，包括Mysql数据库配置、修改OrthoMCL配置文件、转换序列格式、过滤、比对、解析结果和聚类等步骤，特别麻烦。有更方便的途径来完成安装：OrthoMCL Pipeline (<a target="_blank" rel="noopener" href="https://github.com/apetkau/orthomcl-pipeline">https://github.com/apetkau/orthomcl-pipeline</a>)。Pipeline安装有点复杂，但是安装完成后，使用就方便了。</p>
<p>两年多前使用过 OrthoMCL-pipeline 流程来寻找一个物种内部的同源基因（paralogs）和物种间的同源基因（othologs）。物种内部的同源基因之间的碱基差别可能是若干年演化的结果，比如染色体倍增或基因扩张，导致出现同源基因，随后由于演化导致某些突变，尤其是同义替换（Ks 或称为 synonymous）不会引起氨基酸序列变化，常作为分子钟，也比如基因组分析种常见的 4dtv（四倍简并位点），也是用需要依靠同源基因的序列来做分析。</p>
<blockquote>
<p>2021.06.25 备注：OrthoMCL-pipeline 安装测试失败。改用 OrthoFinder 则十分顺利，只需要切换到 python37 环境，用 conda 一键安装 OrthoFinder，然后一句命令即可运行物种之间的同源基因比较与归类。下面的代码示例，orthofinder_data 目录下存放蛋白序列，比如含有若干序列比较的物种的蛋白序列集合:ath.fasta, gma.fasta等等，每个物种可以用一个序列文件。</p>
</blockquote>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">orthofinder -t 6 -f orthofinder_data -S diamond</span><br></pre></td></tr></table></figure>
<p>有哪些文章的示例？可以便于记忆与理解。</p>
<p><img src="https://i.imgur.com/aDN8Pcn.jpg" alt="Fig. 1" /></p>
<p>上图是 2011 年发表于 Nature 期刊的土豆基因组（<a target="_blank" rel="noopener" href="https://www.nature.com/articles/nature10158">https://www.nature.com/articles/nature10158</a>），使用 OrthoMCL 分析 12 个植物之间的直系同源基因和旁系同源基因，统计数目并制作了条形图。</p>
<p><img src="https://i.imgur.com/XpDMtM8.jpg" alt="Fig. 2" /></p>
<p>上图是 2018 年发表于 PBJ 期刊的西葫芦基因组（<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5978595/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5978595/</a>）。使用 OrthoMCL 分析五个植物的同源基因，并且基于同源基因信息计算四倍简并位点（4tdv），推算基因组是否加倍。</p>
<h3 id="安装-orthomlc-pipeline"><a class="markdownIt-Anchor" href="#安装-orthomlc-pipeline"></a> 安装 OrthoMLC-pipeline</h3>
<p>下面是安装OrthoMCL的记录，其实可以不用看了。<br />
先下载 <strong>OrthoMCL-pipeline</strong>。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/apetkau/orthomcl-pipeline.git</span><br></pre></td></tr></table></figure>
<p>再安装 <strong>Perl 以及若干相关模块</strong>。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cpanm BioPerl DBD::mysql DBI Parallel::ForkManager YAML::Tiny Set::Scalar Text::Table Exception::Class Test::Most Test::Warn Test::Exception Test::Deep Moose SVG Algorithm::Combinatorics</span><br></pre></td></tr></table></figure>
<p>实际上在安装 Perl 模块的过程中遇到问题，主要是 BioPerl 安装失败，报错信息：<code>Installing the dependencies failed: Module 'Alien::Libxml2' is not installed</code>。并且我发现不能通过 cpanm 来安装 Alien::Libxml2。经过检索，在一个论坛（<a target="_blank" rel="noopener" href="https://github.com/PerlAlien/Alien-Libxml2/issues/7">https://github.com/PerlAlien/Alien-Libxml2/issues/7</a>）找到方法。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">$ which perl</span><br><span class="line"> ~/local/bin/perl</span><br><span class="line"></span><br><span class="line">$ perl --version</span><br><span class="line"> This is perl 5, version 30, subversion 3 (v5.30.3) built for x86_64-linux</span><br><span class="line"></span><br><span class="line">$ curl -O https://cpan.metacpan.org/authors/id/P/PL/PLICEASE/Alien-Build-1.71_01.tar.gz</span><br><span class="line">  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current</span><br><span class="line">                                 Dload  Upload   Total   Spent    Left  Speed</span><br><span class="line">100  257k  100  257k    0     0  1764k      0 --:--:-- --:--:-- --:--:-- 1775k</span><br><span class="line">$ tar xf Alien-Build-1.71_01.tar.gz </span><br><span class="line">$ cd Alien-Build-1.71_01</span><br><span class="line">$ perl Makefile.PL </span><br><span class="line">Checking if your kit is complete...</span><br><span class="line">Looks good</span><br><span class="line">Generating a Unix-style Makefile</span><br><span class="line">Writing Makefile for Alien::Build</span><br><span class="line">Writing MYMETA.yml and MYMETA.json</span><br><span class="line">...</span><br><span class="line">$ make</span><br><span class="line">$ make install</span><br><span class="line"># 最后这样就安装成功了</span><br><span class="line"># Appending installation info to /home/fenglei/local/lib/perl5/5.30.3/x86_64-linux/perllocal.pod</span><br></pre></td></tr></table></figure>
<p><strong><a target="_blank" rel="noopener" href="http://orthomcl.org/common/downloads/software/v2.0/">OrthoMCL</a></strong> 或 OrthoMCL Custom （能够自定义序列识别符的修改版本）。OrhtoMCL 下载并解压之后，bin目录下的Perl程序首行是 <code>#!/usr/bin/perl</code>，我将其全部改为用户的 perl 了，因为我自己安装的perl位置是 <code>/home/fenglei/local/bin/perl</code>。</p>
<p><strong><a target="_blank" rel="noopener" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">BLAST</a></strong> (blastall, formatdb) 注意不是NCBI-blast+, 推荐使用 2.2.26 版本。由于我之前下载过 interproscan，在 interproscan 安装目录下游 blastall 和 formatdb 等程序，直接将其路径加入环境变量即可！</p>
<p><strong><a target="_blank" rel="noopener" href="https://www.micans.org/mcl/src/">MCL</a></strong> The MCL algorithm is short for the Markov Cluster Algorithm, a fast and scalable unsupervised cluster algorithm for graphs (also known as networks) based on simulation of (stochastic) flow in graphs.</p>
<p>依赖环境解决后能够顺利运行设置脚本:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">$ perl scripts/orthomcl-pipeline-setup.pl</span><br><span class="line">Checking for Software dependencies...</span><br><span class="line">Checking for OthoMCL ... OK</span><br><span class="line">Checking for formatdb ... OK</span><br><span class="line">Checking for blastall ... OK</span><br><span class="line">Checking for mcl ... OK</span><br><span class="line">Wrote new configuration to orthomcl-pipeline/scripts/../etc/orthomcl-pipeline.conf</span><br><span class="line">Wrote executable file to orthomcl-pipeline/scripts/../bin/orthomcl-pipeline</span><br><span class="line">Please add directory orthomcl-pipeline/scripts/../bin to PATH</span><br></pre></td></tr></table></figure>
<p>然而，我重复以前的流程失败，还是卡在 BioPerl 安装失败。看来问题跟我使用 Anaconda3 有关，下面的错误提示信息种出现的 make 是来自 python37 环境。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">make: *** [Makefile:3290: test_dynamic] Error 255</span><br><span class="line">  CJFIELDS/BioPerl-1.7.8.tar.gz</span><br><span class="line">  /home/fenglei/local/app/anaconda3/envs/python37/bin/make test -- NOT OK</span><br><span class="line">//hint// to see the cpan-testers results for installing this module, try:</span><br><span class="line">  reports CJFIELDS/BioPerl-1.7.8.tar.gz</span><br><span class="line">Failed during this command:</span><br><span class="line"> (optional) CJFIELDS/BioPerl-1.7.8.tar.gz     : make_test NO</span><br></pre></td></tr></table></figure>
<h3 id="20210615-再次尝试安装测试"><a class="markdownIt-Anchor" href="#20210615-再次尝试安装测试"></a> 20210615 再次尝试安装测试</h3>
<p>由于系统自带了 Perl（/usr/bin/perl），我安装的Anaconda3下面的不同环境也分别有 Perl（<sub>/local/app/anaconda3/envs/python37/bin/perl、</sub>/local/app/anaconda3/envs/python27/bin/perl），另外自己又在用户目录下安装了 Perl（~/local/bin/perl），多个版本导致冲突。</p>
<p>软件测试：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">PERL5LIB=/home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0  </span><br><span class="line"># 上面其实应该设为 /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/</span><br><span class="line"># 此时直接用 cpan -i 或者 perl -MCPAN -s shell 均无法安装 perl 模块，于是尝试 conda，conda基本可以安装这些模块</span><br><span class="line"> conda install -c bioconda perl-Moo</span><br><span class="line"> conda install -c bioconda perl-DBD::mysql # ERROR: Wrong command</span><br><span class="line"> conda install -c bioconda perl-DBD-mysql</span><br><span class="line"> conda install -c bioconda perl-Parallel-ForkManager</span><br><span class="line"> conda install -c bioconda perl-YAML-Tiny</span><br><span class="line"> conda install -c bioconda perl-Set-Scalar</span><br><span class="line"> conda install -c bioconda perl-Text-Table  # ERROR: PackagesNotFoundError</span><br><span class="line"> conda install -c bioconda perl-Exception-Class</span><br><span class="line"> conda install -c bioconda perl-Test-Most</span><br><span class="line"> conda install -c bioconda perl-Test-Warn</span><br><span class="line"> conda install -c bioconda perl-Test-Exception</span><br><span class="line"> conda install -c bioconda perl-Test-Deep</span><br><span class="line"> conda install -c bioconda perl-Moose</span><br><span class="line"> conda install -c bioconda perl-SVG</span><br><span class="line"> conda install -c bioconda perl-Algorithm-Combinatorics  # ERROR: PackagesNotFoundError</span><br></pre></td></tr></table></figure>
<p>conda 安装perl模块有两个安不上，所以手动尝试</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">wget https://cpan.metacpan.org/authors/id/S/SH/SHLOMIF/Text-Aligner-0.16.tar.gz</span><br><span class="line">perl Makefile.PL &amp;&amp; make &amp;&amp; make install</span><br><span class="line">## Installing /home/fenglei/perl5/man/man3/Text::Aligner.3</span><br><span class="line">## Appending installation info to /home/fenglei/perl5/lib/perl5/x86_64-linux-thread-multi/perllocal.pod</span><br><span class="line"></span><br><span class="line">conda install -c bioconda  perl-Module-Build</span><br><span class="line"></span><br><span class="line">wget https://cpan.metacpan.org/authors/id/S/SH/SHLOMIF/Text-Table-1.134.tar.gz</span><br><span class="line">perl Makefile.PL &amp;&amp; make &amp;&amp; make install</span><br><span class="line">## Installing /home/fenglei/perl5/man/man3/Text::Table.3</span><br><span class="line">## Appending installation info to /home/fenglei/perl5/lib/perl5/x86_64-linux-thread-multi/perllocal.pod</span><br><span class="line"></span><br><span class="line">## 安装上述模块之后， perl lib 路径下根本没有出现pm文件，我手动拷贝过去，为了保险，拷贝到两个位置：</span><br><span class="line">cp Aligner.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Text/</span><br><span class="line">cp Aligner.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/Text/</span><br><span class="line">cp Table.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Text/</span><br><span class="line">cp Table.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/Text/</span><br><span class="line">PERL5LIB=/home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi/auto/:$PERL5LIB</span><br><span class="line"></span><br><span class="line">wget https://cpan.metacpan.org/authors/id/F/FX/FXN/Algorithm-Combinatorics-0.27.tar.gz</span><br><span class="line">## make 失败</span><br><span class="line">## 报错：unable to execute &#x27;x86_64-conda_cos6-linux-gnu-gcc&#x27;: No such file or directory  </span><br></pre></td></tr></table></figure>
<p>这次卡在了 Algorithm-Combinatorics。</p>
<p>在base环境下，自己安装的perl可以编译Algorithm-Combinatorics。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">source activate base</span><br><span class="line">which perl</span><br><span class="line"># ~/local/bin/perl</span><br><span class="line">PERL5LIB=/home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0  </span><br><span class="line">cp Combinatorics.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/</span><br><span class="line">cp Combinatorics.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Algorithm/</span><br><span class="line">cp Combinatorics.pm /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/Algorithm/</span><br></pre></td></tr></table></figure>
<p>更改前后，都有报错，看来直接复制Combinatorics.pm也不行。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost orthomcl-pipeline]$ perl ./scripts/orthomcl-pipeline-setup.pl</span><br><span class="line">Can&#x27;t locate Algorithm/Combinatorics.pm in @INC (you may need to install the Algorithm::Combinatorics module) (@INC contains: /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi/auto/ /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0 /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2 /home/fenglei/local/app/anaconda3/envs/python37/lib/5.26.2/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/5.26.2 .) at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br><span class="line">BEGIN failed--compilation aborted at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br><span class="line"></span><br><span class="line">(python37) [fenglei@localhost orthomcl-pipeline]$ perl ./scripts/orthomcl-pipeline-setup.pl</span><br><span class="line">Can&#x27;t locate loadable object for module Algorithm::Combinatorics in @INC (@INC contains: /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi/auto/ /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/perl5/site_perl/5.22.0 /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/site_perl/5.26.2 /home/fenglei/local/app/anaconda3/envs/python37/lib/5.26.2/x86_64-linux-thread-multi /home/fenglei/local/app/anaconda3/envs/python37/lib/5.26.2 .) at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br><span class="line">Compilation failed in require at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br><span class="line">BEGIN failed--compilation aborted at ./scripts/orthomcl-pipeline-setup.pl line 21.</span><br></pre></td></tr></table></figure>
<h3 id="orthofinder-测试"><a class="markdownIt-Anchor" href="#orthofinder-测试"></a> orthofinder 测试</h3>
<p>安装可以很方便，就是conda：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">source  activate python37</span><br><span class="line">conda install -c bioconda -y orthofinder</span><br><span class="line"></span><br><span class="line">(python37) [fenglei@localhost Results_Apr24]$ orthofinder -help</span><br><span class="line"></span><br><span class="line">OrthoFinder version 2.5.2 Copyright (C) 2014 David Emms</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/" class="post-title-link" itemprop="url">Bioinfo | Using Circos for genome visualization</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-05-28 11:40:21" itemprop="dateCreated datePublished" datetime="2021-05-28T11:40:21+08:00">2021-05-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-06-03 21:11:14" itemprop="dateModified" datetime="2021-06-03T21:11:14+08:00">2021-06-03</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>circos 是以环形展示基因组序列和注释信息的工具，我首先按照 circos 官网的指引进行手动安装，即手动逐个安装 Perl 模块，最后再安装 circos 文件，但是卡在 GD 库，没法成功运行 circos。但是换了一台服务器，进入 anaconda3 下面的 python3.7 环境，直接用 conda 不到一分钟就成功安装 circos。</p>
<p>First, understand the <a target="_blank" rel="noopener" href="http://circos.ca/software/requirements">requirements</a>. If you are a Windows user, read about pertinent differences between UNIX and Windows.</p>
<p><a target="_blank" rel="noopener" href="http://circos.ca/software/download">Download</a> and <a target="_blank" rel="noopener" href="http://circos.ca/software/install">install</a> the software, then follow the <a target="_blank" rel="noopener" href="http://circos.ca/documentation/tutorials">tutorials</a>.</p>
<h3 id="installing-perl-modules"><a class="markdownIt-Anchor" href="#installing-perl-modules"></a> Installing Perl modules</h3>
<p>安装 Circos 的第一步，安装 Perl 以及相关的若干 Perl 模块。我 Linux 系统自带的 Perl，进入命令行界面时报错（<code>perl -MCPAN -e shell</code> Can’t locate <a target="_blank" rel="noopener" href="http://CPAN.pm">CPAN.pm</a> in @INC），所以我先尝试自己安装的 Perl，即 anaconda3/python37 里面的 Perl v5.26.2，但是在安装 Params::Validate 包时候无法通过。我再次尝试系统默认的 Perl v5.16.3，先安装 CPAN，再逐个安装包，成功安装 Params::Validate。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">perl --version</span><br><span class="line"># This is perl 5, version 26, subversion 2 (v5.26.2) built for x86_64-linux-thread-multi</span><br><span class="line">perl --version</span><br><span class="line">This is perl 5, version 16, subversion 3 (v5.16.3) built for x86_64-linux-thread-multi</span><br><span class="line">perl -MCPAN -e shell</span><br><span class="line"></span><br><span class="line">Config::General (v2.50 or later)</span><br><span class="line">Font::TTF</span><br><span class="line">GD</span><br><span class="line">List::MoreUtils</span><br><span class="line">Math::Bezier</span><br><span class="line">Math::Round</span><br><span class="line">Math::VecStat</span><br><span class="line">Params::Validate (自动安装失败，尝试手动安装)</span><br><span class="line">Readonly</span><br><span class="line">Regexp::Common</span><br><span class="line">Set::IntSpan (v1.16 or later)</span><br><span class="line">Text::Format</span><br></pre></td></tr></table></figure>
<p>安装 Params::Validate 的过程中，同时安装了下列 Perl 模块。均是模块安装包下载之后解压，进入模块目录后依次运行 <code>perl Makefile.PL</code>，<code>make Makefile</code>，<code>make install Makefile</code> 即可。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">wget https://cpan.metacpan.org/authors/id/D/DR/DROLSKY/Params-Validate-1.30.tar.gz</span><br><span class="line"># wget http://search.cpan.org/CPAN/authors/id/D/DR/DROLSKY/Params-Validate-0.95.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/E/ET/ETHER/Module-Metadata-1.000037.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/L/LE/LEONT/version-0.9929.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/B/BI/BINGOS/IPC-Cmd-1.04.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/B/BI/BINGOS/Params-Check-0.38.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/J/JE/JESSE/Locale-Maketext-Simple-0.21.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/B/BI/BINGOS/Module-Load-Conditional-0.74.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/B/BI/BINGOS/Module-Load-0.36.tar.gz</span><br></pre></td></tr></table></figure>
<h3 id="installing-libpng-freetype-libgd-and-gd"><a class="markdownIt-Anchor" href="#installing-libpng-freetype-libgd-and-gd"></a> INSTALLING libpng, freetype, libgd AND gd</h3>
<p>About half of the support questions in the Google Group are about installing Perl’s GD module, which is the interface to the libgd system graphics library.</p>
<p>The installation process is not robust because of dependencies and the possibility of their being different versions of these dependencies on your system. Below I show the process of installing GD on a new Mac OS Mavericks system (10.9.4).</p>
<p>Below I link to my local copies of the versions of each library that I installed. These are mature libraries and newer versions are likely to be minor bug releases. The exception is GD—I was unable to successfully compile v2.56 (boot_GD symbol warning which I could not fix). However v2.53 worked.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">&gt; tar xvfz libpng-1.6.14.tar.gz</span><br><span class="line">&gt; cd libpng-1.6.14</span><br><span class="line">&gt; ./configure —prefix=/usr/local</span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br><span class="line"></span><br><span class="line">&gt; tar xvfz jpegsrc.v9.tar.gz</span><br><span class="line">&gt; cd jpeg-9</span><br><span class="line">&gt; ./configure —prefix=/usr/local</span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br><span class="line"></span><br><span class="line">&gt; tar xvfz freetype-2.4.0.tar.gz</span><br><span class="line">&gt; cd freetype-2.4.0</span><br><span class="line">&gt; ./configure —prefix=/usr/local</span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br></pre></td></tr></table></figure>
<p>You should now have libraries in /usr/local/lib as we as some header files in /usr/local/include.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">-rwxr-xr-x  1 root  wheel   650336 21 Nov 11:15 libfreetype.6.dylib</span><br><span class="line">-rw-r--r--  1 root  wheel  3676456 21 Nov 11:15 libfreetype.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       19 21 Nov 11:15 libfreetype.dylib -&gt; libfreetype.6.dylib</span><br><span class="line">-rwxr-xr-x  1 root  wheel      957 21 Nov 11:15 libfreetype.la</span><br><span class="line">-rwxr-xr-x  1 root  wheel   275736 21 Nov 11:14 libjpeg.9.dylib</span><br><span class="line">-rw-r--r--  1 root  wheel  1492952 21 Nov 11:14 libjpeg.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       15 21 Nov 11:14 libjpeg.dylib -&gt; libjpeg.9.dylib</span><br><span class="line">-rwxr-xr-x  1 root  wheel      920 21 Nov 11:14 libjpeg.la</span><br><span class="line">lrwxr-xr-x  1 root  wheel       10 21 Nov 11:13 libpng.a -&gt; libpng16.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       14 21 Nov 11:13 libpng.dylib -&gt; libpng16.dylib</span><br><span class="line">lrwxr-xr-x  1 root  wheel       11 21 Nov 11:13 libpng.la -&gt; libpng16.la</span><br><span class="line">-rwxr-xr-x  1 root  wheel   240580 21 Nov 11:13 libpng16.16.dylib</span><br><span class="line">-rw-r--r--  1 root  wheel  1077240 21 Nov 11:13 libpng16.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       17 21 Nov 11:13 libpng16.dylib -&gt; libpng16.16.dylib</span><br><span class="line">-rwxr-xr-x  1 root  wheel      924 21 Nov 11:13 libpng16.la</span><br><span class="line">drwxr-xr-x  6 root  wheel      204 21 Nov 11:15 pkgconfig/</span><br></pre></td></tr></table></figure>
<p>Now install libgd, linking to the libraries installed above</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">&gt; tar xvfz libgd-2.1.0.tar.gz</span><br><span class="line">&gt; cd libgd-2.1.0</span><br><span class="line">&gt; ./configure --with-png=/usr/local --with-freetype=/usr/local --with-jpeg=/usr/local --prefix=/usr/local</span><br><span class="line">./configure --with-png=/home/fenglei/local --with-freetype=/home/fenglei/local --with-jpeg=/home/fenglei/local --prefix=/home/fenglei/local</span><br><span class="line">…</span><br><span class="line">** Configuration summary for libgd 2.1.0:</span><br><span class="line"></span><br><span class="line">   Support for Zlib:                 yes</span><br><span class="line">   Support for PNG library:          yes</span><br><span class="line">   Support for JPEG library:         yes</span><br><span class="line">   Support for VPX library:          no</span><br><span class="line">   Support for TIFF library:         no</span><br><span class="line">   Support for Freetype 2.x library: yes</span><br><span class="line">   Support for Fontconfig library:   no</span><br><span class="line">   Support for Xpm library:          no</span><br><span class="line">   Support for pthreads:             yes</span><br><span class="line">…</span><br><span class="line"></span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br></pre></td></tr></table></figure>
<p>You now have libgd in /usr/local/lib</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">-rwxr-xr-x  1 root  wheel   389372 19 Nov 14:47 libgd.3.dylib</span><br><span class="line">-rw-r--r--  1 root  wheel  1217200 19 Nov 14:47 libgd.a</span><br><span class="line">lrwxr-xr-x  1 root  wheel       13 19 Nov 14:47 libgd.dylib -&gt; libgd.3.dylib</span><br><span class="line">-rwxr-xr-x  1 root  wheel     1139 19 Nov 14:47 libgd.la</span><br></pre></td></tr></table></figure>
<p>as well as some binaries in /usr/local/bin. In particular, you have /usr/local/bin/glib-config, which provides the configuration for your libgd installation</p>
<p>The --with-* parameters during the configure stage of gdlib installation sets the sources of the dependencies that will be build into gdlib. If you used a different -prefix in compiling these dependencies (see above), adjust these parameters accordingly. For example, if you compiled libpng with -prefix=/my/path then use --with-png=/my/path/.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">/usr/local/bin/gdlib-config --all</span><br><span class="line"></span><br><span class="line">GD library  2.1.0</span><br><span class="line">includedir: /usr/local/include</span><br><span class="line">cflags:     -I/usr/local/include</span><br><span class="line">ldflags:     -L/usr/local/lib</span><br><span class="line">libs:       -ljpeg -lz  -L/usr/local/lib -lpng16 -L/usr/local/lib -lfreetype -lz -liconv</span><br><span class="line">libdir:     /usr/local/lib</span><br><span class="line">features:   GD_JPEG GD_FREETYPE GD_PNG GD_GIF GD_GIFANIM GD_OPENPOLYGON</span><br></pre></td></tr></table></figure>
<p>由于我安装的 gd 时候是配置到了用户目录下，所以得到下面的信息。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">GD library  2.1.0</span><br><span class="line">includedir: /home/fenglei/local/include</span><br><span class="line">cflags:     -I/home/fenglei/local/include</span><br><span class="line">ldflags:     -L/home/fenglei/local/lib</span><br><span class="line">libs:       -ljpeg -lz -lm  -L/home/fenglei/local/lib -lpng16 -L/home/fenglei/local/lib -lfreetype -lz -lfontconfig -lfreetype   -lfontconfig -lXpm -lX11   /home/fenglei/local/lib/libiconv.so -Wl,-rpath -Wl,/home/fenglei/local/lib</span><br><span class="line">libdir:     /home/fenglei/local/lib</span><br><span class="line">features:   GD_XPM GD_JPEG GD_FONTCONFIG GD_FREETYPE GD_PNG GD_GIF GD_GIFANIM GD_OPENPOLYGON</span><br></pre></td></tr></table></figure>
<p>You must have GD_FREETYPE and GD_PNG for Circos to run. The other features, such as support for JPEG and TIFF are optional. In this example, I’ve included the JPEG library in the installation.</p>
<p>Now, install the Perl interface to libgd — the GD module.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">&gt; tar xvfz GD-2.53.tar.gz</span><br><span class="line">&gt; perl Makefile.PL</span><br><span class="line"></span><br><span class="line">Configuring for libgd version 2.1.0.</span><br><span class="line">Checking for stray libgd header files...none found.</span><br><span class="line"></span><br><span class="line">Included Features:          GD_JPEG GD_FREETYPE GD_PNG GD_GIF GD_GIFANIM GD_OPENPOLYGON</span><br><span class="line">GD library used from:       /usr/local</span><br><span class="line">Checking if your kit is complete...</span><br><span class="line">Looks good</span><br><span class="line">Writing Makefile for GD</span><br><span class="line">Writing MYMETA.yml and MYMETA.json</span><br><span class="line"></span><br><span class="line">During the installation GD will look for libgd-config (see above) for libgd configuration. </span><br><span class="line"></span><br><span class="line">&gt; make</span><br><span class="line">&gt; make install</span><br><span class="line"></span><br><span class="line">&gt; perl -MGD -e &#x27;print $GD::VERSION,”\n”’</span><br><span class="line">&gt; 2.53</span><br></pre></td></tr></table></figure>
<p>Test that the modules are installed using</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&gt; circos --modules</span><br><span class="line">…</span><br><span class="line">ok       0.39 Font::TTF::Font</span><br><span class="line">ok       2.53 GD</span><br><span class="line">ok        0.2 GD::Polyline</span><br><span class="line">ok       2.39 Getopt::Long</span><br><span class="line">ok       1.16 IO::File</span><br><span class="line">…</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line">wget http://circos.ca/distribution/circos-0.69-9.tgz</span><br><span class="line">tar -zxvf circos-0.69-9.tgz</span><br><span class="line">./bin/circos --modules</span><br><span class="line"></span><br><span class="line">ok       1.26 Carp</span><br><span class="line">missing            Clone</span><br><span class="line">ok       2.63 Config::General</span><br><span class="line">ok       3.40 Cwd</span><br><span class="line">ok      2.145 Data::Dumper</span><br><span class="line">ok       2.52 Digest::MD5</span><br><span class="line">ok       2.84 File::Basename</span><br><span class="line">ok       3.40 File::Spec::Functions</span><br><span class="line">ok     0.2301 File::Temp</span><br><span class="line">ok       1.51 FindBin</span><br><span class="line">ok       0.39 Font::TTF::Font</span><br><span class="line">missing            GD</span><br><span class="line">missing            GD::Polyline</span><br><span class="line">ok        2.4 Getopt::Long</span><br><span class="line">ok       1.16 IO::File</span><br><span class="line">ok      0.430 List::MoreUtils</span><br><span class="line">ok       1.27 List::Util</span><br><span class="line">ok       0.01 Math::Bezier</span><br><span class="line">ok      1.997 Math::BigFloat</span><br><span class="line">ok       0.07 Math::Round</span><br><span class="line">ok       0.08 Math::VecStat</span><br><span class="line">ok       1.02 Memoize</span><br><span class="line">ok       1.30 POSIX</span><br><span class="line">ok       1.30 Params::Validate</span><br><span class="line">ok       1.63 Pod::Usage</span><br><span class="line">ok       2.05 Readonly</span><br><span class="line">ok 2017060201 Regexp::Common</span><br><span class="line">missing            SVG</span><br><span class="line">ok       1.19 Set::IntSpan</span><br><span class="line">missing            Statistics::Basic</span><br><span class="line">ok       2.45 Storable</span><br><span class="line">ok       1.16 Sys::Hostname</span><br><span class="line">ok       2.02 Text::Balanced</span><br><span class="line">ok       0.62 Text::Format</span><br><span class="line">ok     1.9725 Time::HiRes</span><br></pre></td></tr></table></figure>
<p>Install missing packages</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cpan -i SVG</span><br><span class="line">cpan -i Clone</span><br><span class="line">cpan -i Statistics::Basic</span><br></pre></td></tr></table></figure>
<h3 id="以上做完依然报错"><a class="markdownIt-Anchor" href="#以上做完依然报错"></a> 以上做完依然报错</h3>
<p>运行 <code>circos --mudules</code> 显示 GD 模块缺失。（已经分别安装 libgd 和 Perl 的 GD 模块，为什么circos还是显示没有 GD 模块？）</p>
<h3 id="使用-conda-安装-circos"><a class="markdownIt-Anchor" href="#使用-conda-安装-circos"></a> 使用 conda 安装 circos</h3>
<p>换一台服务器，<code>source activate python37</code> 切换成 python3.7 环境。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&gt; which perl</span><br><span class="line">~/local/app/anaconda3/envs/python37/bin/perl</span><br><span class="line">&gt; perl --version</span><br><span class="line">This is perl 5, version 26, subversion 2 (v5.26.2) built for x86_64-linux-thread-multi</span><br></pre></td></tr></table></figure>
<p>再运行 <code>conda install -c bioconda circos</code> 顺利安装！下面是模块测试。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost ~]$ circos --modules</span><br><span class="line">ok       1.38 Carp</span><br><span class="line">ok       0.42 Clone</span><br><span class="line">ok       2.63 Config::General</span><br><span class="line">ok       3.75 Cwd</span><br><span class="line">ok      2.167 Data::Dumper</span><br><span class="line">ok       2.55 Digest::MD5</span><br><span class="line">ok       2.85 File::Basename</span><br><span class="line">ok       3.67 File::Spec::Functions</span><br><span class="line">ok     0.2304 File::Temp</span><br><span class="line">ok       1.51 FindBin</span><br><span class="line">ok       0.39 Font::TTF::Font</span><br><span class="line">ok       2.71 GD</span><br><span class="line">ok        0.2 GD::Polyline</span><br><span class="line">ok       2.49 Getopt::Long</span><br><span class="line">ok       1.16 IO::File</span><br><span class="line">ok      0.428 List::MoreUtils</span><br><span class="line">ok     1.4602 List::Util</span><br><span class="line">ok       0.01 Math::Bezier</span><br><span class="line">ok   1.999806 Math::BigFloat</span><br><span class="line">ok       0.07 Math::Round</span><br><span class="line">ok       0.08 Math::VecStat</span><br><span class="line">ok    1.03_01 Memoize</span><br><span class="line">ok       1.76 POSIX</span><br><span class="line">ok       1.29 Params::Validate</span><br><span class="line">ok       1.69 Pod::Usage</span><br><span class="line">ok       2.05 Readonly</span><br><span class="line">ok 2017060201 Regexp::Common</span><br><span class="line">ok       2.84 SVG</span><br><span class="line">ok       1.19 Set::IntSpan</span><br><span class="line">ok     1.6611 Statistics::Basic</span><br><span class="line">ok       2.62 Storable</span><br><span class="line">ok       1.20 Sys::Hostname</span><br><span class="line">ok       2.03 Text::Balanced</span><br><span class="line">ok       0.59 Text::Format</span><br><span class="line">ok     1.9741 Time::HiRes</span><br></pre></td></tr></table></figure>
<p>再以 circos 自带的案例做画图测试。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">wget http://circos.ca/distribution/circos-0.69-9.tgz</span><br><span class="line">tar -zxvf circos-0.69-9.tgz</span><br><span class="line">cd circos-0.69-9</span><br><span class="line">cd example</span><br><span class="line"># 在 example 目录下有名为 run 的脚本：../bin/circos -conf etc/circos.conf -debug_group summary,timer &gt; run.out</span><br><span class="line"># 由于我调用的是 anaconda3 里面的 circos，所以用下面的命令，几秒钟即可运行完毕，顺利得到 circos.svg &amp; circos.png 图。</span><br><span class="line">circos -conf etc/circos.conf -debug_group summary,timer &gt; run.out</span><br></pre></td></tr></table></figure>
<p><img src="http://i.imgur.com/ZkOgjPN.jpg" alt="example/circos.png" /></p>
<p>注：<br />
另一台服务器上，通过 conda 可以安装 circos，但是运行的时候报错：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; circos</span><br><span class="line">perl: symbol lookup error: /home/fenglei/perl5/lib/perl5/x86_64-linux-thread-multi/auto/Clone/Clone.so: undefined symbol: Perl_xs_apiversion_bootcheck</span><br></pre></td></tr></table></figure>
<p>看来是 Perl 环境变量的问题。</p>
<p>Circos 的文件结构如下图所示，conf 文件即图片属性设置，txt 文件就是统计数据，circos 将二者的结合就是图片。</p>
<p><img src="http://qiubio.com/wp-content/uploads/2011/11/circos-architecture.png" alt="" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/" class="post-title-link" itemprop="url">Bioinfo | Repeat sequence analysis by RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer)</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-04-23 12:01:58 / Modified: 12:24:09" itemprop="dateCreated datePublished" datetime="2021-04-23T12:01:58+08:00">2021-04-23</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>学习重复序列的过程中，读到2019年发表在《Mobile DNA》杂志的文献 “Systematic survey of plant LTR-retrotransposons elucidates phylogenetic relationships of their polyprotein domains and provides a reference for element classification”。</p>
<p>文中提到 We have established an improved classification system applicable to LTR-retrotransposons from a wide range of plant species. This system reflects phylogenetic relationships as well as distinct sequence and structural features of the elements. A comprehensive database of retrotransposon protein domains (REXdb) that reflects this classification provides a reference for efficient and unified annotation of LTR-retrotransposons in plant genomes. Access to REXdb related tools is implemented in the RepeatExplorer web server (<a target="_blank" rel="noopener" href="https://repeatexplorer-elixir.cerit-sc.cz/">https://repeatexplorer-elixir.cerit-sc.cz/</a>) or using a standalone version of REXdb that can be downloaded seaparately from RepeatExplorer web page (<a target="_blank" rel="noopener" href="http://repeatexplorer.org/">http://repeatexplorer.org/</a>).</p>
<p>A comprehensive database of retrotransposon protein domains (REXdb) that reflects this classification provides a reference for efficient and unified annotation of LTR-retrotransposons in plant genomes.</p>
<h3 id="installation"><a class="markdownIt-Anchor" href="#installation"></a> Installation</h3>
<p>To use RepeatExplorer without installation, we recommend our freely available Galaxy server at <a target="_blank" rel="noopener" href="https://repeatexplorer-elixir.cerit-sc.cz">https://repeatexplorer-elixir.cerit-sc.cz</a>. This server is provided in frame of the ELIXIR-CZ project. The Galaxy server includes also additional tools useful for data preprocessing, quality control and genome annotation.</p>
<p>For installing command-line version, follow the instruction below:</p>
<p>Download source code using git command:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">git clone https://bitbucket.org/petrnovak/repex_tarean.git</span><br><span class="line">cd repex_tarean</span><br></pre></td></tr></table></figure>
<p>We recommend to install dependencies using conda (conda can be installed using miniconda). The required environment can be prepared using command:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda env create -f environment.yml</span><br></pre></td></tr></table></figure>
<p>Then activate the prepared environment using:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda activate repeatexplorer</span><br></pre></td></tr></table></figure>
<p>In the repex_tarean direcory compile the source and prepare databases using:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">make</span><br></pre></td></tr></table></figure>
<p><s>Support for 32-bit executables is required. If you are using Ubuntu distribution you can add 32-bit support by running:</s></p>
<p><s>sudo dpkg --add-architecture i386</s><br />
<s>sudo apt-get update</s><br />
<s>sudo apt-get install libc6:i386 libncurses5:i386 libstdc++6:i386</s></p>
<p>To verify the installation, run RepeatExplorer clustering on the provided example data:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">./seqclust -p -v tmp/clustering_output test_data/LAS_paired_10k.fas</span><br></pre></td></tr></table></figure>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/04/16/Plant-Chromatin-immunoprecipitation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/04/16/Plant-Chromatin-immunoprecipitation/" class="post-title-link" itemprop="url">Plant | Chromatin immunoprecipitation</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-04-16 17:53:35 / Modified: 20:57:47" itemprop="dateCreated datePublished" datetime="2021-04-16T17:53:35+08:00">2021-04-16</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Plant/" itemprop="url" rel="index"><span itemprop="name">Plant</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>##Crosslinking</p>
<ol>
<li>Prepare crosslinking buffer (Extraction buffer 1). For each gram of leaves, 36 ml buffer is needed.</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Reagent  Amount (500 mL)</span><br><span class="line">0.4 M sucrose 68.46 g</span><br><span class="line">10 mM Tris-HCl, pH 8.0 5 mL of 1 M stock</span><br><span class="line">5 mM beta-ME  (Add prior to use) 176 uL of 14.2 M stock</span><br><span class="line">0.1 mM PMSF (Add prior to use) 500 uL of 0.1 M stock</span><br><span class="line">1% Formaldehyde (Add prior to use) 13.5 mL of 37% stock</span><br></pre></td></tr></table></figure>
<ol start="2">
<li>Prepare 2M glycine solution (for stopping the crosslinking reaction). For each gram of leaves, 2.5 ml solution is needed. Store at 4°C.</li>
<li>Collect beaker with appropriate capacity. Make sure the buffer volume will not exceed 60-70% of the beaker capacity (to avoid spillage).</li>
<li>Harvest soybean leaves (C08).</li>
<li>Cut leaves into 0.5-1 cm pieces and submerge them in Milli-Q water in beaker.</li>
<li>Rinse leaves with Milli-Q water twice.</li>
<li>Add 5 mM beta-ME into the cold extraction buffer 1. Pour buffer into the beaker with leaves.</li>
<li>Place the beaker into the transparent “bowl” for vacuum infiltration.</li>
<li>Disconnect centrivap with the cold trap.</li>
<li>Connect the bowl with the cold trap.</li>
<li>Add appropriate volume of formaldehyde solution to the buffer. 1 ml formaldehyde solution in 36 ml extraction buffer 1. working concentration ~ 1% formaldehyde</li>
<li>Stir the mixture with stirrer.</li>
<li>Place a piece of gauze under the bottom of a flask/beaker.</li>
<li>Place the flask/beaker with gauze onto the water surface in the beaker.</li>
<li>Cover the lid of the “bowl”.</li>
<li>Close the valve at the cold trap. (Turn the valve to a position perpendicular to the pipe).</li>
<li>Turn on the vacuum pump. After a while, the lid will be closed tightly and bubbles can be observed in the solution. Vacuum infiltrate for 5 minutes.</li>
<li>Turn off the pump. Open the valve and the lid.</li>
<li>Stir the solution again and make sure all leaves are submerged. Repeat step 16-18.</li>
<li>Add appropriate volume of 2M glycine solution to the buffer. For each gram of leaves, 2.5 ml solution is added.</li>
<li>Vacuum infiltrate for 5 minutes again.</li>
<li>Cover the opening with the gauze already used in vacuum infiltration and filter out all the buffer under running tap water in a sink.</li>
<li>Rinse the cross-linked leaves with Milli-Q water. (Cold Milli-Q water is preferred in some protocols.)</li>
<li>With the aid of a stirrer/rod, transfer cross-linked leaves onto layers of tissue paper.</li>
<li>Remove excess water by gently pressing cross-linked leaves with tissue paper.</li>
<li>Pack leaves with aluminum foil.</li>
<li>Freeze leaves with liquid nitrogen. Store at -80°C.</li>
</ol>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/04/08/Bioinfo-RepeatModeler/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/04/08/Bioinfo-RepeatModeler/" class="post-title-link" itemprop="url">Bioinfo | 重复序列预测工具 RepeatModeler</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-04-08 14:25:07" itemprop="dateCreated datePublished" datetime="2021-04-08T14:25:07+08:00">2021-04-08</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-04-16 17:02:46" itemprop="dateModified" datetime="2021-04-16T17:02:46+08:00">2021-04-16</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>RepeatMasker是重复序列检测的常用工具，通过与参考数据库的相似性比对来准确识别或屏蔽基因组中的重复序列，属于同源预测注释的方式。基因组组装完成后，进行基因预测和注释。由于基因组中存在重复序列结构区，特别是高等真核生物，重复序列占了相当大的比例，会影响基因预测的质量，也会带来不必要的资源消耗。因此在基因预测前，首先要检测并屏蔽基因组中的重复序列（真核生物很必要，原核生物可忽略）。对于这种情况，只是为了屏蔽重复序列降低后续基因预测的压力，重复序列的检测可以不必太纠结。但有时候我们注释基因组的重复序列结构，也可能是专注于某些特定研究，例如，某些重复元件可能参与了重要功能，我们期望定位它们的位置。这种情况下，需要识别精准。</p>
<p>RepeatModeler 的识别过程是怎么样的？根据<a target="_blank" rel="noopener" href="https://www.biorxiv.org/content/10.1101/856591v1.full.pdf">原始文献</a>的描述，它采用了两个不同的策略，第一个是 RepeatScout，第二个是 RECON。另外，新版的RepeatModeler2 针对 LTR 鉴定进行了功能升级，引入 LTRharvest 和 LTR_retriever 模块。RepeatScout 原理是先找出出现频率较高的“种子”序列，然后对“种子”序列进行延长来预测重复序列，这与BLAST的双序列比对过程类似，RepeatScout 能快速鉴定高频率的重复序列与出现时间比较晚的重复元件，这个过程中（Round-1），RepeatScout 从基因组随机取 40Mbp 序列，并分析其中的重复元件，通过与已知数据库进行比较能够快速鉴定已知的转座子，在 Round-1 目录下，有一个 sampleDB-1.fa 文件，这个文件对应的 blastdb 数据库文件，。对于 RECON 而言，它通过基因组之间的比较来鉴定转座子，更彻底的序列比对，也能鉴定出现时间比较早的转座子。RECON 先从基因组选取3Mbp序列作为样品，再依次增加为3倍，分别9Mbp、27Mbp、81Mbp、243Mbp，即一直从Round-2到Round-6（什么目的？）。</p>
<h3 id="软件依赖-安装"><a class="markdownIt-Anchor" href="#软件依赖-安装"></a> 软件依赖、安装</h3>
<ol>
<li><em>Perl</em> 5.8.0 or higher. 我是通过 Python3.7 的 conda 安装的：This is perl 5, version 26, subversion 2 (v5.26.2) built for x86_64-linux-thread-multi。</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install perl</span><br></pre></td></tr></table></figure>
<p>在Linux命令行模式下输入“perl -MCPAN -e shell”进入perl界面，然后输入“instal module-name”安装下面的包。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">install JSON</span><br><span class="line">install File::Which</span><br><span class="line">install URI</span><br><span class="line">install LWP::UserAgent</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://metacpan.org/pod/JSON">JSON</a>自动安装失败，改成手动从cpan下载安装。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">wget https://cpan.metacpan.org/authors/id/I/IS/ISHIGAKI/JSON-4.03.tar.gz</span><br><span class="line">tar -zxvf JSON-4.03.tar.gz</span><br><span class="line">cd JSON-4.03</span><br><span class="line">perl Makefile.PL</span><br><span class="line">make</span><br><span class="line">make install Makefile</span><br><span class="line"></span><br><span class="line">wget https://cpan.metacpan.org/authors/id/P/PL/PLICEASE/File-Which-1.24.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/O/OA/OALDERS/URI-5.09.tar.gz</span><br><span class="line">wget https://cpan.metacpan.org/authors/id/O/OA/OALDERS/libwww-perl-6.53.tar.gz</span><br></pre></td></tr></table></figure>
<ol start="2">
<li>Python 3 and the <em>h5py</em> python library. 我电脑已经安装 Python 的管理器 Anaconda 3，然后安装Python3.7，再通过 conda 安装 h5py：</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install h5py。</span><br></pre></td></tr></table></figure>
<ol start="3">
<li>Sequence Search Engine (RepeatMasker uses a sequence search engine to perform it’s search for repeats. Currently Cross_Match, <em>RMBlast</em> and WUBlast/ABBlast are supported. You will need to obtain one or the other of these and install them on your system.) Download Pre-compiled Package: Download the RMBlast package for your platform:</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wget http://www.repeatmasker.org/rmblast-2.11.0+-x64-linux.tar.gz</span><br><span class="line">tar -zxvf rmblast-2.11.0+-x64-linux.tar.gz</span><br></pre></td></tr></table></figure>
<ol start="4">
<li><em>TRF</em> - Tandem Repeat Finder, G. Benson et al.<br />
You can obtain a free copy at <a target="_blank" rel="noopener" href="http://tandem.bu.edu/trf/trf.html">http://tandem.bu.edu/trf/trf.html</a>. RepeatMasker was developed using TRF version 4.0.9</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/Benson-Genomics-Lab/TRF.git</span><br><span class="line">cd TRF</span><br><span class="line">mkdir build</span><br><span class="line">cd build</span><br><span class="line">../configure --prefix=/home/fenglei/local</span><br><span class="line">make</span><br><span class="line"># To install to user&#x27;s path</span><br><span class="line">make install</span><br></pre></td></tr></table></figure>
<p>安装TRF中途报错：/home/fenglei/local/app/TRF/missing: line 81: aclocal-1.16: command not found. WARNING: ‘aclocal-1.16’ is missing on your system.<br />
通过安装 automake 解决问题。</p>
<ol start="5">
<li><em>Repeat Database</em><br />
RepeatMasker can be used with custom libraries, or with Dfam out of the box. Dfam is an open database of transposable element (TE) profile HMM models and consensus sequences. The current release (Dfam 3.2) contains 6,900 TE families spanning five organisms: human, mouse, zebrafish, fruit fly, nematode, and a growing number of additional species. To supplement this databases we recommend obtaining the RepeatMasker edition of RepBase.<br />
To update the Dfam libraries contained in this release go to <a target="_blank" rel="noopener" href="http://www.dfam.org">http://www.dfam.org</a>.<br />
RepeatMasker软件自带有Dfam（RepBase数据库HMM版）的常见重复序列库，要更新完整的Dfam需去官网下载。该文件大小15G,下载时间比较久下载解压之后将Dfam文件转移到 RepeatMasker 安装目录下，替换掉RepeatMasker自带的Dfam数据集。如下方命令操作。<br />
RepeatMasker comes with a copy of the curated portion of Dfam ( Libraries/Dfam.h5 ) and will work out-of-the box with this library. This is a small library ( at this time ). There are two options for supplementing the main RepeatMasker library:<br />
The complete Dfam ( curated and uncurated families ) library may be downloaded from <a target="_blank" rel="noopener" href="http://www.dfam.org">www.dfam.org</a> in famdb HDF5 format and used to replace the distributed “Libraries/dfam.h5” file. For example:</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wget https://www.dfam.org/releases/Dfam_3.2/families/Dfam.h5.gz</span><br><span class="line">gunzip Dfam.h5.gz</span><br><span class="line"># mv Dfam.h5 /usr/local/RepeatMasker/Libraries</span><br><span class="line">NOTE: This will overwrite the distributed Dfam.h5 file.</span><br></pre></td></tr></table></figure>
<ol start="6">
<li>安装 <em>RepeatMasker</em> 按照这个网页：<a target="_blank" rel="noopener" href="https://www.repeatmasker.org/RepeatMasker/">https://www.repeatmasker.org/RepeatMasker/</a></li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget https://www.repeatmasker.org/RepeatMasker/RepeatMasker-4.1.2-p1.tar.gz</span><br><span class="line">tar -zxvf RepeatMasker-4.1.2-p1.tar.gz</span><br><span class="line">cd RepeatMasker</span><br><span class="line">perl ./configure</span><br><span class="line"># 输入TRF软件所在路径：/software/annotation/TRF-4.10/bin/trf</span><br><span class="line"># 选择搜索引擎：小编选的是2. RMBlast，其他搜索引擎可以根据需求选择（可多次操作），输入 RMBlast路径：/software/annotation/rmblast-2.10.0/bin 并确认默认搜索引擎</span><br></pre></td></tr></table></figure>
<ol start="7">
<li><em>RECON</em> - De Novo Repeat Finder, Bao Z. and Eddy S.R. Developed and tested with our patched version of RECON ( 1.08 ). The 1.08 version fixes problems with running RECON on 64 bit machines and supplies a workaround to a division by zero bug along with some buffer overrun fixes. The program is available at: <a target="_blank" rel="noopener" href="http://www.repeatmasker.org/RepeatModeler/RECON-1.08.tar.gz">http://www.repeatmasker.org/RepeatModeler/RECON-1.08.tar.gz</a>. The original version is available at <a target="_blank" rel="noopener" href="http://eddylab.org/software/recon/">http://eddylab.org/software/recon/</a>.</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget http://www.repeatmasker.org/RepeatModeler/RECON-1.08.tar.gz</span><br><span class="line">tar -zxvf RECON-1.08.tar.gz</span><br><span class="line">cd RECON-1.08</span><br><span class="line">cd src</span><br><span class="line">make</span><br><span class="line">make install</span><br></pre></td></tr></table></figure>
<ol start="8">
<li><em>RepeatScout</em> - De Novo Repeat Finder, Price A.L., Jones N.C. and Pevzner P.A. Developed and tested with our multiple sequence version of RepeatScout ( 1.0.6 ). This version is available at <a target="_blank" rel="noopener" href="http://www.repeatmasker.org/RepeatScout-1.0.6.tar.gz">http://www.repeatmasker.org/RepeatScout-1.0.6.tar.gz</a></li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">tar -zxvf RepeatScout-1.0.6.tar.gz</span><br><span class="line">cd RepeatScout</span><br><span class="line"># edit MakeFile:</span><br><span class="line"># INSTDIR = /home/fenglei/local/scripts/</span><br><span class="line">make</span><br></pre></td></tr></table></figure>
<ol start="9">
<li><em>GenomeTools</em> [optional]<br />
LtrHarvest - The LtrHarvest program is part of the GenomeTools suite. We have developed this release of RepeatModeler on GenomeTools version 1.5.9 available for download from here: <a target="_blank" rel="noopener" href="http://genometools.org/pub/">http://genometools.org/pub/</a> NOTE: use the “make threads=yes” build options to enable multi-threaded runs.<br />
参考 <a target="_blank" rel="noopener" href="http://events.jianshu.io/p/d2fc7a1de669">CentOS 8 安装genometools</a></li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">wget http://genometools.org/pub/genometools-1.6.1.tar.gz</span><br><span class="line"># make 报错：cairo.h: No such file or directory 报错，需要安装cairo</span><br><span class="line">sudo yum install cairo.x86_64 cairo-devel.x86_64 pango-devel.x86_64 pango.x86_64</span><br><span class="line">tar -zxvf genometools-1.6.1.tar.gz</span><br><span class="line">cd genometools-1.6.1</span><br><span class="line">~~make~~</span><br><span class="line"># make 的参数加上 threads=yes</span><br><span class="line"># 最开始没有加允许多线程的参数，导致程序允许时间特别长</span><br><span class="line"># 另注意，我在 Pytho3.7 环境下 make 失败，切换到系统默认的 base 环境才 make 成功</span><br><span class="line">make threads=yes</span><br></pre></td></tr></table></figure>
<ol start="10">
<li><em>Ltr_retriever</em> [optional]<br />
Ltr_retriever - A LTR discovery post-processing and filtering tool. We recommend using version 2.6 or higher from here: <a target="_blank" rel="noopener" href="https://github.com/oushujun/LTR_retriever/releases">https://github.com/oushujun/LTR_retriever/releases</a></li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># download LTR_retriever-2.9.0.tar.gz</span><br><span class="line">tar -zxvf LTR_retriever-2.9.0.tar.gz</span><br><span class="line">mv LTR_retriever-2.9.0 LTR_retriever</span><br><span class="line">/home/fenglei/local/app/LTR_retriever/LTR_retriever -h</span><br></pre></td></tr></table></figure>
<ol start="11">
<li><em>MAFFT</em> [optional]<br />
MAFFT - A multiple sequence alignment program. We developed and tested RepeatModeler using mafft version 7.407. Please use this verison or higher from here: <a target="_blank" rel="noopener" href="https://mafft.cbrc.jp/alignment/software/">https://mafft.cbrc.jp/alignment/software/</a></li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">wget https://mafft.cbrc.jp/alignment/software/mafft-7.475-with-extensions-src.tgz</span><br><span class="line">tar -zxvf mafft-7.475-with-extensions-src.tgz</span><br><span class="line">mv </span><br><span class="line">cd </span><br><span class="line">cd core</span><br><span class="line">make clean</span><br><span class="line">make</span><br><span class="line">vi Makefile</span><br><span class="line"># eidt 1st line. change &quot;/usr/local&quot; to &quot;/home/fenglei/local&quot;</span><br><span class="line">make clean</span><br><span class="line">make</span><br><span class="line">make install</span><br><span class="line">cd ../extention</span><br><span class="line">vi Makefile</span><br><span class="line"># Edit the 1st line. Change &quot;/usr/local&quot; to &quot;/home/fenglei/local&quot;</span><br><span class="line">make clean</span><br><span class="line">make</span><br><span class="line">make install</span><br></pre></td></tr></table></figure>
<ol start="12">
<li><em>Ninjia</em> [optional]<br />
Ninja - A tool for large-scale neighbor-joining phylogeny inference and clustering. We developed and tested RepeatModeler using Ninja version “0.95-cluster_only”. Please obtain a copy from: <a target="_blank" rel="noopener" href="https://github.com/TravisWheelerLab/NINJA/releases/tag/0.95-cluster_only">https://github.com/TravisWheelerLab/NINJA/releases/tag/0.95-cluster_only</a><br />
Download source code from <a target="_blank" rel="noopener" href="https://github.com/TravisWheelerLab/NINJA/releases/tag/0.95-cluster_only">https://github.com/TravisWheelerLab/NINJA/releases/tag/0.95-cluster_only</a></li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">tar -zxvf NINJA-0.95-cluster_only.tar.gz</span><br><span class="line">mv NINJA-0.95-cluster_only /home/fenglei/local/app/NINJA-0.95-cluster</span><br><span class="line">cd NINJA-0.95-cluster/NINJIA</span><br><span class="line">make</span><br></pre></td></tr></table></figure>
<ol start="13">
<li><em>cd-hit</em> [optional]<br />
CD-HIT - A sequence clustering package. We developed and tested RepeatModeler using version 4.8.1. Please use this version or higher from: <a target="_blank" rel="noopener" href="http://weizhongli-lab.org/cd-hit/">http://weizhongli-lab.org/cd-hit/</a></li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/weizhongli/cdhit.git</span><br><span class="line">sudo yum install zlib-devel</span><br><span class="line">cd cdhit</span><br><span class="line">make</span><br></pre></td></tr></table></figure>
<ol start="14">
<li><em>RepeatModeler</em> 按照这个网页 <a target="_blank" rel="noopener" href="http://www.repeatmasker.org/RepeatModeler/">http://www.repeatmasker.org/RepeatModeler/</a></li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wget http://www.repeatmasker.org/RepeatModeler/RepeatModeler-2.0.1.tar.gz</span><br><span class="line">tar -zxvf RepeatModeler-2.0.1.tar.gz</span><br><span class="line">cd RepeatModeler-2.0.1</span><br><span class="line">perl ./configure</span><br></pre></td></tr></table></figure>
<p>根据程序提示，依次输入下方各个程序所在路径。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">#perl</span><br><span class="line">/home/fenglei/local/app/anaconda3/envs/python37/bin/perl</span><br><span class="line">#RepeatMasker</span><br><span class="line">/home/fenglei/local/app/RepeatMasker</span><br><span class="line">#RECON</span><br><span class="line">/home/fenglei/local/app/RECON-1.08/bin</span><br><span class="line">#RepeatScount</span><br><span class="line">/home/fenglei/local/app/RepeatScout-1.0.6</span><br><span class="line">#TRF</span><br><span class="line">/home/fenglei/local/bin/trf</span><br><span class="line">#search engine: RMBlast</span><br><span class="line">/home/fenglei/local/app/rmblast-2.11.0/bin</span><br><span class="line">#genometools</span><br><span class="line">/home/fenglei/local/app/genometools-1.6.1/bin</span><br><span class="line">#Ltr_retriever</span><br><span class="line">/home/fenglei/local/app/LTR_retriever</span><br><span class="line">#MAFFT</span><br><span class="line">/home/fenglei/local/bin</span><br><span class="line">#NINJIA</span><br><span class="line">/home/fenglei/local/app/NINJA-0.95-cluster/NINJA</span><br><span class="line">#cd-hit</span><br><span class="line">/home/fenglei/local/app/cdhit</span><br><span class="line">### Congratulations!  RepeatModeler is now ready to use.</span><br></pre></td></tr></table></figure>
<h3 id="软件使用"><a class="markdownIt-Anchor" href="#软件使用"></a> 软件使用</h3>
<p>使用 REPEATMODELER 来通过基因组序列构建 LIBRARY。生成一个文件夹，名称为 RM_[PID].[DATE] ie. “RM_5098.MonMar141305172005″。该文件夹中的“consensi.fa.classified”即为 library，用于RepeatMasker的输入。值得注意的是，输入 fasta 文件序列间不能有空（空格和换行等），否则会程序出错。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">##改成自己的输入fasta名称</span><br><span class="line">fasta=/home/fenglei/databases/Arabidopsis/TAIR10_chr_all.fasta</span><br><span class="line">##改成自己想要的database名称</span><br><span class="line">database_name=Arabidopsis</span><br><span class="line">##改成RepeatModeler安装目录</span><br><span class="line">RepeatModelerPath=/home/fenglei/local/app/RepeatModeler-2.0.1</span><br><span class="line">## Build database</span><br><span class="line">$&#123;RepeatModelerPath&#125;/BuildDatabase -name $&#123;database_name&#125; -engine rmblast $&#123;fasta&#125;</span><br><span class="line">## Run RepeatModeler to detect repeats</span><br><span class="line">$&#123;RepeatModelerPath&#125;/RepeatModeler -pa 10 -database $&#123;database_name&#125; -LTRStruct</span><br></pre></td></tr></table></figure>
<p>第二步就调用 RepeatMasker，依据上一步鉴定的该物种的重复序列库（Library），即 consensi.fa.classified 文件，来对基因组进行标记，输出文件夹中的“*.fasta.masked”文件就是 softmasked 文件，重复序列会被标记成小写字母。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">#!/bin/bash</span><br><span class="line">set -e</span><br><span class="line">set -u</span><br><span class="line">set -o pipefail</span><br><span class="line"></span><br><span class="line">refgenome=/home/fenglei/databases/Arabidopsis/TAIR10_chr_all.fasta  #改成自己的输入fasta名称 不可以用压缩文件</span><br><span class="line">database_name=Arabidopsis  #改成自己想要的database名称</span><br><span class="line">RepeatMaskerHome=/home/fenglei/local/app/RepeatMasker/</span><br><span class="line"></span><br><span class="line">$&#123;RepeatMaskerHome&#125;/RepeatMasker -pa 12 \</span><br><span class="line">        -e rmblast \</span><br><span class="line">        -lib consensi.fa.classified \</span><br><span class="line">        -dir Repeat_result \</span><br><span class="line">        -gff \</span><br><span class="line">        -xsmall \</span><br><span class="line">        $refgenome</span><br></pre></td></tr></table></figure>
<p>如何对重复序列元件进行统计？比如每种元件的数目和总长度，从而知道其在基因组所占比例，这也是一般文章需要展示的地方。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># 从 RepeatMasker 分析结果找出重复元件的类别</span><br><span class="line">awk &#x27;NR&gt;3&#123;print $11&#125;&#x27; TAIR10_chr_all.fasta.out | sort | uniq -c &gt; repeat.class</span><br><span class="line"># 自编程序统计每种元件的数目和总长度</span><br><span class="line">perl ./statistics.pl  repeat.class TAIR10_chr_all.fasta.out &gt; repeat.stat</span><br><span class="line"># 调用 buildSummary.pl 对重复序列元件进行详细统计</span><br><span class="line">perl ~/local/app/RepeatMasker/util/buildSummary.pl TAIR10_chr_all.fasta.out &gt; TAIR10_chr_all.fasta.out_stat</span><br></pre></td></tr></table></figure>
<p>注：上面自编的statistics.pl程序内容如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line">#!/home/fenglei/local/bin/perl</span><br><span class="line">use warnings;</span><br><span class="line">use strict;</span><br><span class="line"></span><br><span class="line">### usage: ./statistics.pl  repeat.class scaffold_out.final.scaffolds.fasta.out &gt; repeat.stat</span><br><span class="line"></span><br><span class="line">open REPCLASS, $ARGV[0] or die $!;</span><br><span class="line">open REPEATMASKER, $ARGV[1] or die $!;</span><br><span class="line"></span><br><span class="line">while(&lt;REPCLASS&gt;)&#123;</span><br><span class="line">        chomp $_;</span><br><span class="line">        my @reparr=split(/\s+/);</span><br><span class="line">        print &quot;$reparr[2]\t&quot;;</span><br><span class="line">        my $totallength=0;</span><br><span class="line">        my $repeatNum=0;</span><br><span class="line">        while(&lt;REPEATMASKER&gt;)&#123;</span><br><span class="line">                chomp $_;</span><br><span class="line">                $_ =~ s/^\s+//;</span><br><span class="line">                my @repeatmask=split(/\s+/);</span><br><span class="line">                my $repeatlength=0;</span><br><span class="line">#               print &quot;$repeatmask[5]\t$repeatmask[6]\t$repeatmask[10]&quot;; exit;</span><br><span class="line">                if(length($_) == 0)&#123;next;&#125;</span><br><span class="line">                if($reparr[2] eq $repeatmask[10])&#123;</span><br><span class="line">#                       print &quot;$repeatmask[5]\t$repeatmask[6]\t$repeatmask[10]&quot;; exit;</span><br><span class="line">                        $repeatNum = $repeatNum + 1;</span><br><span class="line">                        $repeatlength = $repeatmask[6] - $repeatmask[5];</span><br><span class="line">                        $totallength = $totallength + $repeatlength;</span><br><span class="line">                &#125;</span><br><span class="line">        &#125;</span><br><span class="line">        print &quot;$repeatNum\t$totallength\n&quot;;</span><br><span class="line">        close REPEATMASKER;</span><br><span class="line">        open REPEATMASKER, $ARGV[1] or die $!;</span><br><span class="line">#       print &quot;\n&quot;;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">close REPCLASS;</span><br><span class="line">close REPEATMASKER;</span><br></pre></td></tr></table></figure>
<p>以拟南芥为例，下面是输出的部分结果。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line">Repeat Classes</span><br><span class="line">==============</span><br><span class="line">Total Sequences: 7</span><br><span class="line">Total Length: 119667750 bp</span><br><span class="line">Class                  Count        bpMasked    %masked</span><br><span class="line">=====                  =====        ========     =======</span><br><span class="line">DNA                    --           --           --   </span><br><span class="line">    CMC-EnSpm          878          720527       0.60% </span><br><span class="line">    MULE-MuDR          1202         1171346      0.98% </span><br><span class="line">    PIF-Harbinger      150          71815        0.06% </span><br><span class="line">    TcMar-Pogo         150          31269        0.03% </span><br><span class="line">    hAT-Ac             81           28117        0.02% </span><br><span class="line">    hAT-Tip100         94           86316        0.07% </span><br><span class="line">LINE                   --           --           --   </span><br><span class="line">    L1                 2259         1292456      1.08% </span><br><span class="line">LTR                    --           --           --   </span><br><span class="line">    Copia              1251         1270535      1.06% </span><br><span class="line">    Gypsy              4058         4618620      3.86% </span><br><span class="line">    Pao                55           10627        0.01% </span><br><span class="line">RC                     --           --           --   </span><br><span class="line">    Helitron           1213         556951       0.47% </span><br><span class="line">Unknown                23305        7697251      6.43% </span><br><span class="line">                      ---------------------------------</span><br><span class="line">    total interspersed 34696        17555830     14.67%</span><br><span class="line"></span><br><span class="line">Low_complexity         8617         414746       0.35% </span><br><span class="line">Simple_repeat          34644        1364479      1.14% </span><br><span class="line">rRNA                   265          338782       0.28% </span><br><span class="line">tRNA                   289          84920        0.07% </span><br><span class="line">---------------------------------------------------------</span><br><span class="line">Total                  78511        19758757     16.51%</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">By Sequence</span><br><span class="line">===========</span><br><span class="line">Seq            Count      bpMasked</span><br><span class="line">=====          =====      ========</span><br><span class="line">               2   13984      3678725 </span><br><span class="line">    mitochondria   163        89662   </span><br><span class="line">               1   18731      4168514 </span><br><span class="line">               3   15642      4105811 </span><br><span class="line">     chloroplast   77         49506   </span><br><span class="line">               5   17313      4206310 </span><br><span class="line">               4   12601      3460229 </span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/03/16/Bioinfo-understanding-sequence-identity/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/03/16/Bioinfo-understanding-sequence-identity/" class="post-title-link" itemprop="url">Bioinfo | 怎么理解序列一致性（Sequence identity）</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-03-16 12:06:40" itemprop="dateCreated datePublished" datetime="2021-03-16T12:06:40+08:00">2021-03-16</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-03-18 09:38:48" itemprop="dateModified" datetime="2021-03-18T09:38:48+08:00">2021-03-18</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>序列一致性（Sequence identity）是评估两个序列相似度的一个参考值，不同的软件可能会有不通过的计算方法。下面的案例摘抄自 <a target="_blank" rel="noopener" href="http://lh3.github.io/2018/11/25/on-the-definition-of-sequence-identity">Heng Li’s blog</a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">CCAGTGTGGCCGATACCCCAGGTTGGCACGCATCGTTGCCTTGGTAAGC</span><br><span class="line">CCAGTGTGGCCGATGCCCGTGCTACGCATCGTTGCCTTGGTAAGC</span><br></pre></td></tr></table></figure>
<p>经典的方法是先对两个序列做比对，并且设置打分规则：匹配 match=1，错配 mismatch=-2，缺失 gapOpen=-2，缺失延伸 gapExt=-1。依据这个打分规则，上述两条序列的最佳比对结果如下。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Ref+:  1 CCAGTGTGGCCGATaCCCcagGTtgGC-ACGCATCGTTGCCTTGGTAAGC 49</span><br><span class="line">         |||||||||||||| |||   ||  || ||||||||||||||||||||||</span><br><span class="line">Qry+:  1 CCAGTGTGGCCGATgCCC---GT--GCtACGCATCGTTGCCTTGGTAAGC 45</span><br></pre></td></tr></table></figure>
<p>Here we have 43 matching bases, 1 mismatch, 5 deletions and 1 insertion to the first/Ref sequence. The <a target="_blank" rel="noopener" href="https://www.drive5.com/usearch/manual/cigar.html">CIGAR</a> is 18M3D2M2D2M1I22M. CIGAR 的定义： Concise Idiosyncratic Gapped Alignment Report，上面的CIGAR值可翻译为：18 matches，3 deletions，2 matches，2 matches，1 deletions，22 mathces。</p>
<h3 id="不考虑-gap-的序列一致性"><a class="markdownIt-Anchor" href="#不考虑-gap-的序列一致性"></a> 不考虑 Gap 的“序列一致性”</h3>
<p>如果不考虑 Gap，则只考虑匹配 Match 和错配 Mismatch。序列一致性等于 “#matches / (#matches + #mismatches)” = 43/(43+1) = 97.7%。这里的序列一致性并没有考虑 Gap，有一定局限性，早期基因组相关报道中提到的“大猩猩基因组与人类基因组的相似程度为98%”应该就是这种情况。摘抄：An obvious problem with this definition is that it doesn’t count gaps. However, it is an often used definition. We may hear that the chimpanzee and human genome differ by a couple of percent. Here we are referring to such gap-excluded identity. The exact sentence in the first chimpanzee genome paper is “Single-nucleotide substitutions occur at a mean rate of 1.23% between copies of the human and chimpanzee genome”.</p>
<h3 id="考虑-gap-的序列一致性以-blast-为例"><a class="markdownIt-Anchor" href="#考虑-gap-的序列一致性以-blast-为例"></a> 考虑 Gap 的“序列一致性”（以 BLAST 为例）</h3>
<p>BLAST 中对序列一致性的算法是最常见的算法，序列一致性等于匹配数 43（matchings bases）除以比对区间总长度 50，得到 86%。需要指出，虽然 Ref 序列的长度是 49，但是总比对长度却是 50（就是要多加 1 个插入碱基的长度）。局限在于，虽然考虑的 Gap 的存在，但是没有考虑 Gap 是否连续。一个连续的 Gap，与若干个分散的 Gap，如果总长度一样，计算得到的序列一致性也一样，这种计算方法就有不够科学的地方。原作者举例，如果一个 query 序列长度为 1000bp，比对到参考序列上面发现中间有连续 300bp 是 insertion 导致的 Gap，这个情况下序列一致性为 (1000-300)/1000 = 70%，这个结果会被过滤掉，而在进化（evolution）的过程中，这个连续的 300bp 可能只是一次序列插入事件，而非 300 次独立的单碱基插入事件。BLAST identity is defined as the number of matching bases over the number of alignment columns. In this example, there are 50 columns, so the identity is 43/50=86%. In a SAM file, the number of columns can be calculated by summing over the lengths of M/I/D CIGAR operators. The number of matching bases equals the column length minus the NM tag. Here is a Perl one-liner to calculate BLAST identity:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">perl -ane &#x27;if(/NM:i:(\d+)/)&#123;$n=$1; $l=0; $l+=$1 while/(\d+)[MID]/g; print(($l-$n)/$l,&quot;\n&quot;)&#125;&#x27;</span><br><span class="line"># Where variable $n is the sum of mismatches and gaps and $l is the alignment length. </span><br><span class="line">#In the PAF format, column 10 divived by column 11 gives the BLAST identity.</span><br></pre></td></tr></table></figure>
<p>核酸序列比对程序 BLASTN 可以输出 SAM 格式结果，如下所示。上面的 Perl 命令行即可以读取包含“NM:i:[\d+]”的行，NM即$1=17；识别M、I、D 字符对应的数字，相同字母对应的数值分别累加，累加之后分别为 match，M=551；insertion，I=5；deletion，D=9。那么序列一致性等于 (M+I+D-NM)/(M+I+D)=(565-17)/565=97%。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">@HD     VN:1.2  SO:coordinate   GO:reference</span><br><span class="line">@SQ     SN:Query_1      LN:916</span><br><span class="line">@PG     ID:0    VN:2.9.0+       CL:blastn -db /home/fenglei/database/Glycine_max/Gmax_275_Wm82_a2.v1/index/genome.fa -query GNAT_mutate.fa -out GNAT_mutate.fa-BLASTN-Wm82-275.m17 -outfmt 17 -window_size 10 -word_size 10     PN:blastn</span><br><span class="line">Chr02   0       Query_1 1       255     2929886H24M3D83M3D83M3D83M3D291M3I13M2I57M45646980H     *       0       0       *       *       AS:i:591        EV:f:0  NM:i:17 PI:f:100.00     BS:f:1092.49</span><br></pre></td></tr></table></figure>
<h3 id="压缩-gap-的序列一致性以-minimap2-为例"><a class="markdownIt-Anchor" href="#压缩-gap-的序列一致性以-minimap2-为例"></a> 压缩 Gap 的“序列一致性”（以 minimap2 为例）</h3>
<p>上文提到，连续若干个碱基的 Gap 可能是进化过程中单次插入导致的，那么可以通过压缩 Gap 将上述序列比对转化成下面的情形。Gap 的数目没有变化，但是长度全部变为 1，这样总长度也发生了变化。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Ref+:  CCAGTGTGGCCGATaCCCcGTtGCTACGCATC-TTGCCTTGGTAAGC</span><br><span class="line">       |||||||||||||| ||| || |||||||||| ||||||||||||||</span><br><span class="line">Qry+:  CCAGTGTGGCCGATgCCC-GT-GCTACGCATCgTTGCCTTGGTAAGC</span><br></pre></td></tr></table></figure>
<p>Gap 压缩之后的序列一致性为 43/(50-2-1)=91.5%。对于比对软件 minimap2，即是采用这种算法。下面是Perl 命令计算一致性数值的例子。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">perl -ane &#x27;if(/NM:i:(\d+)/)&#123;$n=$1; $m=$g=$o=0; $m+=$1 while/(\d+)M/g; $g+=$1,++$o while/(\d+)[ID]/g; print(1-($n-$g+$o)/($m+$o),&quot;\n&quot;)&#125;&#x27;</span><br><span class="line">#where $m is the sum of M operations, $g the sum of I and D operations and $o the number of gap opens.</span><br></pre></td></tr></table></figure>
<p>参考资料</p>
<ol>
<li><a target="_blank" rel="noopener" href="http://lh3.github.io/2018/11/25/on-the-definition-of-sequence-identity">On the definition of sequence identity</a></li>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/9539f44ecf61">怎样定义 sequences 比对的相似度</a></li>
<li><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/55324279">知乎专栏：序列一致性</a></li>
</ol>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/2/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><a class="page-number" href="/page/2/">2</a><span class="page-number current">3</span><a class="page-number" href="/page/4/">4</a><span class="space">&hellip;</span><a class="page-number" href="/page/18/">18</a><a class="extend next" rel="next" href="/page/4/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">175</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
