<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>Bioinfo | Gene Ontology - GOMAP | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">Bioinfo | Gene Ontology - GOMAP</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">Bioinfo | Gene Ontology - GOMAP</h1><div class="post-meta">2021-06-17<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>Gene Ontology Meta Annotator for Plants</p>
<p>GOMAP 主要是给植物基因集做 Gene Ontology 注释信息匹配，它综合了多个数据库，所以得到的信息会全面一些。早年将基因对应到GO用的工具是Blast2GO，这个软件linux版本安装比较麻烦，基于mysql数据库，而且后面它商业化之后要收费。后来的策略是用diamond的快速blastp将基因集与SwissProt数据库进行比较，得到每个基因最相近的Uniprot ID，再依靠id-mapping文件将基因与GO对应起来，本文最后有简单介绍基于SwissProt做GO注释的流程。</p>
<p>What is GOMAP-Singularity? GOMAP-Singularity is the containerized version of the Gene Ontology Meta Annotator for Plants (GOMAP) pipeline. GOMAP is a high-throughput pipeline to annotate GO terms to plant protein sequences. The pipeline uses three different approaches to annotate GO terms to plant protein sequences, and compile the annotations together to generate an aggregated dataset. GOMAP uses Python code to run the component tools to generate the GO annotations, and R code to clean and aggregate the GO annotations from the component tools.</p>
<p>How do I use GOMAP-Singularity?</p>
<p>Please check &lt;bioinformapping.com/gomap&gt; for detailed documentation on how to use the container.</p>
<h3 id="Install-system-dependencies"><a href="#Install-system-dependencies" class="headerlink" title="Install system dependencies"></a>Install system dependencies</h3><p>You must first install development tools and libraries to your host.<br>On Debian-based systems:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ sudo apt-get update &amp;&amp; \</span><br><span class="line">  sudo apt-get install -y build-essential \</span><br><span class="line">  libseccomp-dev pkg-config squashfs-tools cryptsetup</span><br></pre></td></tr></table></figure>

<p>On CentOS/RHEL:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sudo yum groupinstall -y &#39;Development Tools&#39; &amp;&amp; \</span><br><span class="line">  sudo yum install -y epel-release &amp;&amp; \</span><br><span class="line">  sudo yum install -y golang libseccomp-devel \  </span><br><span class="line">  squashfs-tools cryptsetup</span><br><span class="line">  yum remove golang # 后面手动安装</span><br></pre></td></tr></table></figure>

<h3 id="Install-Golang"><a href="#Install-Golang" class="headerlink" title="Install Golang"></a>Install Golang</h3><p>Golang 有多种安装方法，我是在其官网(<a target="_blank" rel="noopener" href="https://golang.org/doc/install#download">https://golang.org/doc/install#download</a>)下载<a target="_blank" rel="noopener" href="https://golang.org/dl/go1.16.5.linux-amd64.tar.gz">源文件</a>到本地，然后安装到用户目录。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">wget https:&#x2F;&#x2F;golang.org&#x2F;dl&#x2F;go1.16.5.linux-amd64.tar.gz</span><br><span class="line">tar -zxvf go1.16.5.linux-amd64.tar.gz</span><br><span class="line"></span><br><span class="line"># 将 “export PATH&#x3D;&quot;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;go&#x2F;bin&quot;:$PATH” 写入bashrc</span><br><span class="line">source ~&#x2F;.bashrc</span><br><span class="line">which go</span><br><span class="line"># ~&#x2F;local&#x2F;app&#x2F;go&#x2F;bin&#x2F;go</span><br><span class="line"></span><br><span class="line"># Verify that you&#39;ve installed Go by opening a command prompt and typing the following command:</span><br><span class="line">go version</span><br><span class="line"># go version go1.16.5 linux&#x2F;amd64</span><br></pre></td></tr></table></figure>

<h3 id="Compiling-SingularityCE"><a href="#Compiling-SingularityCE" class="headerlink" title="Compiling SingularityCE"></a>Compiling SingularityCE</h3><p>参考安装方法如下</p>
<p>You can build SingularityCE using the following commands:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ cd $&#123;GOPATH&#125;&#x2F;src&#x2F;github.com&#x2F;sylabs&#x2F;singularity &amp;&amp; \</span><br><span class="line">  .&#x2F;mconfig &amp;&amp; \</span><br><span class="line">  cd .&#x2F;builddir &amp;&amp; \</span><br><span class="line">  make &amp;&amp; \</span><br><span class="line">  sudo make install</span><br></pre></td></tr></table></figure>

<p>And that’s it! Now you can check your SingularityCE version by running:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ singularity version</span><br></pre></td></tr></table></figure>
<p>To build in a different folder and to set the install prefix to a different path:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ .&#x2F;mconfig -b .&#x2F;buildtree -p &#x2F;usr&#x2F;local</span><br></pre></td></tr></table></figure>

<p>我的安装如下。</p>
<p>系统自带的gcc（版本4.8.5）编译报错，提示找不到config.c文件。我切换成用户的gcc（版本10.2.0），就可以安装。<br>启动用户gcc的办法，修改bashrc文件，将下面三行的注释符号去掉，然后source ~/.bashrc即可。以后有必要再回来将这几行注释掉，就恢复系统自带的老版本gcc。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">export PATH&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;gcc-10.2.0&#x2F;bin&#x2F;:$PATH</span><br><span class="line">export LD_LIBRARY_PATH&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;gcc-10.2.0&#x2F;lib&#x2F;:&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;gcc-10.2.0&#x2F;lib64&#x2F;:$LD_LIBRARY_PATH</span><br><span class="line">export LD_LIBRARY_PATH&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;lib&#x2F;:&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;lib64&#x2F;:$LD_LIBRARY_PATH</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">$ git clone https:&#x2F;&#x2F;github.com&#x2F;sylabs&#x2F;singularity.git</span><br><span class="line">$ cd singularity  # 该目录下有一个 mconfig 文件</span><br><span class="line"># .&#x2F;mconfig -b .&#x2F;buildtree -p &#x2F;home&#x2F;fenglei&#x2F;local</span><br><span class="line"># .&#x2F;mconfig -b .&#x2F;buildtree -p &#x2F;home&#x2F;fenglei&#x2F;local --without-suid</span><br><span class="line">$ .&#x2F;mconfig    # 默认的安装位置是 &#x2F;usr&#x2F;local</span><br><span class="line">$ cd &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;singularity&#x2F;buildtree</span><br><span class="line">$ make</span><br><span class="line"># make install</span><br><span class="line">$ sudo make install</span><br><span class="line"></span><br><span class="line">$ which singularity</span><br><span class="line">&#x2F;usr&#x2F;local&#x2F;bin&#x2F;singularity</span><br></pre></td></tr></table></figure>

<p>测试：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ singularity version                                                                                        </span><br><span class="line">3.8.0-rc.1+107-g8b21e5a16</span><br></pre></td></tr></table></figure>

<h3 id="GOMAP-singularity"><a href="#GOMAP-singularity" class="headerlink" title="GOMAP-singularity"></a>GOMAP-singularity</h3><ol>
<li>Singularity 安装</li>
</ol>
<p>上面已经完成。也可参考：<a target="_blank" rel="noopener" href="https://bioinformapping.com/gomap/master/RUNNING.html">https://bioinformapping.com/gomap/master/RUNNING.html</a></p>
<ol start="2">
<li>从 git 下载 GOMAP-singularity</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">git clone https:&#x2F;&#x2F;github.com&#x2F;Dill-PICL&#x2F;GOMAP-singularity.git</span><br><span class="line">cd GOMAP-singularity</span><br></pre></td></tr></table></figure>
<ol start="3">
<li>配置并下载相关参考数据集</li>
</ol>
<p>Run the setup step to make necessary directories and download data files from CyVerse. The pipeline download is large and would require ~40GB of free hard drive space during the setup step.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">$ .&#x2F;setup.sh </span><br><span class="line"># 由于GO没安好，导致这一步报错，后面重新安装了GO，并重装singularity</span><br><span class="line"># ERROR  : No setuid installation found, for unprivileged installation use: .&#x2F;mconfig --without-suid</span><br><span class="line"></span><br><span class="line">$ .&#x2F;setup.sh </span><br><span class="line">0&#x2F;1 -  0.00% of files done   0.000&#x2F;39254.793 MB -  0.00% of file sizes done</span><br><span class="line">Processing GOMAP.sif - 39254.793 MB   2021-06-17.15:58:52</span><br><span class="line">GOMAP.sif - 40.000&#x2F;39254.793 MB -  0.10% done   2021-06-17.16:00:53</span><br><span class="line">GOMAP.sif - 680.000&#x2F;39254.793 MB -  1.73% done   2021-06-17.16:02:47</span><br><span class="line">GOMAP.sif - 1360.000&#x2F;39254.793 MB -  3.46% done   2021-06-17.16:04:49</span><br><span class="line">GOMAP.sif - 2000.000&#x2F;39254.793 MB -  5.09% done   2021-06-17.16:06:40</span><br><span class="line">GOMAP.sif - 2400.000&#x2F;39254.793 MB -  6.11% done   2021-06-17.16:08:32</span><br><span class="line">cliReconnManager: Reconnecting clientState &#x3D; 0</span><br><span class="line">...</span><br><span class="line">cliReconnManager: Reconnecting clientState &#x3D; 0</span><br><span class="line">The client&#x2F;server socket connection has been renewed</span><br><span class="line">GOMAP.sif - 30880.000&#x2F;39254.793 MB - 78.67% done   2021-06-17.17:59:30</span><br><span class="line">GOMAP.sif - 31520.000&#x2F;39254.793 MB - 80.30% done   2021-06-17.18:01:43</span><br><span class="line">GOMAP.sif - 32040.000&#x2F;39254.793 MB - 81.62% done   2021-06-17.18:03:45</span><br><span class="line">GOMAP.sif - 32640.000&#x2F;39254.793 MB - 83.15% done   2021-06-17.18:05:31</span><br><span class="line">GOMAP.sif - 33200.000&#x2F;39254.793 MB - 84.58% done   2021-06-17.18:07:27</span><br><span class="line">GOMAP.sif - 33373.426&#x2F;39254.793 MB - 85.02% done   2021-06-17.18:08:24</span><br><span class="line">cliReconnManager: Reconnecting clientState &#x3D; 0</span><br><span class="line">The client&#x2F;server socket connection has been renewed</span><br><span class="line">cliReconnManager: Reconnecting clientState &#x3D; 0</span><br><span class="line">The client&#x2F;server socket connection has been renewed</span><br><span class="line">cliReconnManager: Reconnecting clientState &#x3D; 0</span><br><span class="line">The client&#x2F;server socket connection has been renewed</span><br><span class="line">cliReconnManager: Reconnecting clientState &#x3D; 0</span><br><span class="line">The client&#x2F;server socket connection has been renewed</span><br><span class="line">GOMAP.sif - 39254.793&#x2F;39254.793 MB - 100.00% done   2021-06-17.18:45:46</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<ol start="4">
<li>测试 GOMAP 注释流程</li>
</ol>
<p>[optional]: Test whether the container and the data files are working as intended.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">$ .&#x2F;test.sh</span><br><span class="line"></span><br><span class="line">Running Sequence-similarity based Annotation Step</span><br><span class="line">&#x2F;usr&#x2F;local&#x2F;lib&#x2F;python2.7&#x2F;dist-packages&#x2F;Bio&#x2F;SearchIO&#x2F;__init__.py:211: BiopythonExperimentalWarning: Bio.SearchIO is an experimental submodule which may undergo significant changes prior to its future official release.</span><br><span class="line">  BiopythonExperimentalWarning)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Building a new DB, current time: 06&#x2F;17&#x2F;2021 20:59:09</span><br><span class="line">New DB name:   &#x2F;workdir&#x2F;test&#x2F;GOMAP-0.3_GOMAP-input&#x2F;input&#x2F;0.3_GOMAP-input.fa</span><br><span class="line">New DB title:  &#x2F;workdir&#x2F;test&#x2F;GOMAP-0.3_GOMAP-input&#x2F;input&#x2F;0.3_GOMAP-input.fa</span><br><span class="line">Sequence type: Protein</span><br><span class="line">...</span><br><span class="line">Checking gaf file for simple errors and fixing them</span><br><span class="line">There were 34 warnings (use warnings() to see them)</span><br><span class="line">[1] &quot;Checking if data&#x2F;data&#x2F;go&#x2F;go.obo.data exists&quot;</span><br><span class="line">[1] &quot;data&#x2F;data&#x2F;go&#x2F;go.obo.data exists so loading R object&quot;</span><br><span class="line">   user  system elapsed </span><br><span class="line">  2.547   0.078   2.624 </span><br><span class="line">[1] &quot;Reading All Datasets&quot;</span><br><span class="line">Checking gaf file for simple errors and fixing them</span><br><span class="line">[1] 1714</span><br><span class="line">[1] &quot;Removing Redundancy&quot;</span><br><span class="line">[1] 18</span><br><span class="line">Compiling Non-Redundant Comprehensive Data</span><br><span class="line">There were 50 or more warnings (use warnings() to see the first 50)</span><br></pre></td></tr></table></figure>

<ol start="5">
<li><p>实战篇：学习自定义配置文件 Edit the config file</p>
<p>a. Declare export GOMAP_LOC environment variable. Edit <code>~/.bashrc</code> as below:</p>
</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">export GOPATH&#x3D;&quot;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;go&quot;</span><br><span class="line">export GOMAP_LOC&#x3D;&quot;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;GOMAP-singularity&#x2F;install&#x2F;location&quot;</span><br></pre></td></tr></table></figure>

<p> b. Download the <a target="_blank" rel="noopener" href="https://bioinformapping.com/gomap/master/_static/min-config.yml">config.yml</a> file and make necessary changes. Change the highlighted lines to fit your input data.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">#Input section </span><br><span class="line">input:</span><br><span class="line">  #input fasta file name</span><br><span class="line">  fasta: GOMAP-input.fa</span><br><span class="line">  # output file basename</span><br><span class="line">  basename: GOMAP_output</span><br><span class="line">  #input NCBI taxonomy id</span><br><span class="line">  taxon: &quot;123456&quot;</span><br><span class="line">  # Name of the species</span><br><span class="line">  species: &quot;Glycine max&quot;</span><br><span class="line">  # Email is mandatory</span><br><span class="line">  email: myname#gmail.com</span><br><span class="line">  #Number of CPUs used for tools</span><br><span class="line">  cpus: 4</span><br><span class="line">  #Whether openmpi should be used</span><br><span class="line">  mpi: False</span><br><span class="line">  #what the name of the temporary directory is</span><br><span class="line">  tmpdir: &quot;&#x2F;tmpdir&quot;</span><br></pre></td></tr></table></figure>
<ol start="6">
<li>Run pipeline</li>
</ol>
<p>可直接修改 test.sh 并运行该脚本。也可参考软件手册分步骤运行。<br>下面摘抄的是在单机（SINGLE）运行的模式，软件手册里也有在集群（MPI）运行的示范代码。<br>test.sh 与 test目录在同一个路径。GOMAP-input.fa 与 config.yml 都在test目录下。输入的蛋白序列，其尾部如果有表示stop codon的星号，要移除星号。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"># seqsim</span><br><span class="line">.&#x2F;run-GOMAP-SINGLE.sh --step&#x3D;seqsim --config&#x3D;test&#x2F;config.yml</span><br><span class="line"># domain</span><br><span class="line">.&#x2F;run-GOMAP-SINGLE.sh --step&#x3D;domain --config&#x3D;test&#x2F;config.yml</span><br><span class="line"># fanngo</span><br><span class="line">.&#x2F;run-GOMAP-SINGLE.sh --step&#x3D;fanngo --config&#x3D;test&#x2F;config.yml</span><br><span class="line"># mixmeth-blast</span><br><span class="line">.&#x2F;run-GOMAP-SINGLE.sh --step&#x3D;mixmeth-blast --config&#x3D;test&#x2F;config.yml</span><br><span class="line"># mixmeth-preproc</span><br><span class="line">.&#x2F;run-GOMAP-SINGLE.sh --step&#x3D;mixmeth-preproc --config&#x3D;test&#x2F;config.yml</span><br><span class="line"># mixmeth</span><br><span class="line">.&#x2F;run-GOMAP-SINGLE.sh --step&#x3D;mixmeth --config&#x3D;test&#x2F;config.yml</span><br><span class="line"># aggregate</span><br><span class="line"># Attention: Please wait for all your Argot2.5 jobs to finish before running this step. You will get emails from Argot2.5 when your jobs are submitted and when they are finished. You can also check the status of all current jobs from all users here.</span><br><span class="line">.&#x2F;run-GOMAP-SINGLE.sh --step&#x3D;aggregate --config&#x3D;test&#x2F;config.yml</span><br></pre></td></tr></table></figure>

<h3 id="附注：用-SwissProt-与-idmapping-来做-GO"><a href="#附注：用-SwissProt-与-idmapping-来做-GO" class="headerlink" title="附注：用 SwissProt 与 idmapping 来做 GO"></a>附注：用 SwissProt 与 idmapping 来做 GO</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">## SwissProt 文件是氨基酸序列，比较大</span><br><span class="line">wget ftp:&#x2F;&#x2F;ftp.ncbi.nih.gov&#x2F;blast&#x2F;db&#x2F;FASTA&#x2F;swissprot.gz</span><br><span class="line">## idmapping.tb.gz 文件可能下载链接有变化</span><br><span class="line">wget ftp:&#x2F;&#x2F;ftp.pir.georgetown.edu&#x2F;databases&#x2F;idmapping&#x2F;idmapping.tb.gz</span><br><span class="line">## swissprot index 建立</span><br><span class="line">diamond makedb --in swissprot -d swissprot</span><br><span class="line">## 序列比对</span><br><span class="line">diamond blastp -d &#x2F;PATH&#x2F;TO&#x2F;index&#x2F;swissprot -q input.proteins.fa -k 1 -e 0.00001 -o swissprot_blastp.out</span><br><span class="line">## idmapping 文件整理，下文有摘抄文件具体内容</span><br><span class="line">awk -F &quot;\t&quot;  &#39;&#123;print $1&quot;\t&quot;$8&#125;&#39; idmapping.tb &gt; uniprotID-GOnumber.tab</span><br><span class="line">## 将“基因-Uniprot-GO”的关系建立起来</span><br><span class="line">perl UniProt2GO_annotate.pl uniprotID-GOnumber.tab swiss_dia_matches.m8 swiss_GO_matches.tab</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ head idmapping.tb </span><br><span class="line">Q6GZX4  001R_FRG3G      2947773 YP_031579.1     81941549; 49237298              PF04947 GO:0006355; GO:0046782; GO:0006351                      UniRef100_Q6GZX4        UniRef90_Q6GZX4 UniRef50_Q6GZX4 UPI00003B0FD4           65492415165820 AY548484        AAT09660.1</span><br><span class="line">Q6GZX3  002L_FRG3G      2947774 YP_031580.1     49237299; 81941548; 47060117            PF03003 GO:0033644; GO:0016021                  UniRef100_Q6GZX3        UniRef90_Q6GZX3 UniRef50_Q6GZX3 UPI00003B0FD5           654924        15165820 AY548484        AAT09661.1</span><br><span class="line">Q197F8  002R_IIV3       4156251 YP_654574.1     109287880; 123808694; 106073503                                         UniRef100_Q197F8        UniRef90_Q197F8 UniRef50_Q197F8 UPI0000D83464           345201                        16912294 DQ643392        ABF82032.1</span><br><span class="line">Q197F7  003L_IIV3       4156252 YP_654575.1     106073504; 109287881; 123808693                                         UniRef100_Q197F7        UniRef90_Q197F7 UniRef50_Q197F7 UPI0000D83465           345201                        16912294 DQ643392        ABF82033.1</span><br><span class="line">Q6GZX2  003R_FRG3G      2947775 YP_031581.1     81941547; 49237300                                              UniRef100_Q6GZX2        UniRef90_Q6GZX2 UniRef50_Q6GZX2 UPI00003B0FD6           654924                          15165820       AY548484        AAT09662.1</span><br><span class="line">Q6GZX1  004R_FRG3G      2947776 YP_031582.1     49237301; 81941546                      GO:0033644; GO:0016021                  UniRef100_Q6GZX1        UniRef90_Q6GZX1 UniRef50_Q6GZX1 UPI00003B0FD7           654924                15165820 AY548484        AAT09663.1</span><br><span class="line">Q197F5  005L_IIV3       4156254 YP_654577.1     123808691; 109287883; 106073506                                         UniRef100_Q197F5        UniRef90_Q197F5 UniRef50_Q197F5 UPI0000D83467           345201                        16912294 DQ643392        ABF82035.1</span><br><span class="line">Q6GZX0  005R_FRG3G      2947777 YP_031583.1     47060120; 49237302; 81941545            PF02393                         UniRef100_Q6GZX0        UniRef90_Q6GZX0 UniRef50_Q6GZX0 UPI00003B0FD8           654924                        15165820 AY548484        AAT09664.1</span><br><span class="line">Q91G88  006L_IIV6       1733056 NP_149469.1     82012333; 15078718; 15042163            PF04383; PF12299                                UniRef100_Q91G88        UniRef90_Q91G88 UniRef50_Q91G88 UPI00000F3D32           176652        2820141; 3959991; 7698884; 8492091; 17239238; 8021587; 10456793; 8121799; 9926400; 16789247; 11448171; 1475907; 3201750; 9482589; 1549908; 8073636       AF303741        AAK81943.1</span><br><span class="line">Q6GZW9  006R_FRG3G      2947778 YP_031584.1     81941544; 49237303                                              UniRef100_Q6GZW9        UniRef90_Q6GZW9 UniRef50_Q6GZW9 UPI00003B0FD9           654924                          15165820       AY548484        AAT09665.1</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ head swiss_GO_matches.tab</span><br><span class="line">Q6GZX4  GO:0006355; GO:0046782; GO:0006351</span><br><span class="line">Q6GZX3  GO:0033644; GO:0016021</span><br><span class="line">Q197F8</span><br><span class="line">Q197F7</span><br><span class="line">Q6GZX2</span><br><span class="line">Q6GZX1  GO:0033644; GO:0016021</span><br><span class="line">Q197F5</span><br><span class="line">Q6GZX0</span><br><span class="line">Q91G88</span><br><span class="line">Q6GZW9</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ head swiss_GO_matches.tab</span><br><span class="line">gene1 Q9M9M9  GO:0050897; GO:0009055; GO:0008137; GO:0009735; GO:0006979; GO:0005747; GO:0005739</span><br><span class="line">gene2 Q9LP77  GO:0005524; GO:0004672; GO:0016021; GO:0005886; GO:0009506; GO:0005774</span><br><span class="line">gene3 Q9CAL2  GO:0005524; GO:0004674; GO:0016021; GO:0005886</span><br><span class="line">gene4 A4FLF6</span><br><span class="line">gene5 Q8RY67  GO:0005524; GO:0004674; GO:0007166; GO:0016021; GO:0005886</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">$ cat UniProt2GO_annotate.pl</span><br><span class="line">#!&#x2F;usr&#x2F;bin&#x2F;perl</span><br><span class="line">use warnings;</span><br><span class="line">use strict;</span><br><span class="line"></span><br><span class="line">open IDMAPPING, $ARGV[0] or die $!;</span><br><span class="line">open BLASTOUT, $ARGV[1] or die $!;</span><br><span class="line">open OUT, &quot;&gt;$ARGV[2]&quot; or die $!;</span><br><span class="line">select OUT;</span><br><span class="line"></span><br><span class="line">my %id2go;</span><br><span class="line"></span><br><span class="line">while(&lt;IDMAPPING&gt;)&#123;</span><br><span class="line">        chomp $_;</span><br><span class="line">        my @array1&#x3D;split(&#x2F;\t&#x2F;);</span><br><span class="line">        $id2go&#123;$array1[0]&#125;&#x3D;$array1[1];</span><br><span class="line">#       print &quot;$array1[1]\n&quot;; exit;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">while(&lt;BLASTOUT&gt;)&#123;</span><br><span class="line">        chomp $_;</span><br><span class="line">        my @array2&#x3D;split(&#x2F;\t&#x2F;);</span><br><span class="line">        my $geneid&#x3D;$array2[0];</span><br><span class="line">        my $UniprotID&#x3D;$array2[1];</span><br><span class="line">#       print &quot;$UniprotID\n&quot;;</span><br><span class="line">        $UniprotID &#x3D;~ s&#x2F;\.[0-9]$&#x2F;&#x2F;;</span><br><span class="line">#       print &quot;$geneid\t$&#123;UniprotID&#125;test\n&quot;; exit;</span><br><span class="line">        if(exists $id2go&#123;$UniprotID&#125;)&#123; print &quot;$geneid\t$UniprotID\t$id2go&#123;$UniprotID&#125;\n&quot;;&#125;</span><br><span class="line">        else&#123;print &quot;$geneid\t$UniprotID\n&quot;;&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">close IDMAPPING;</span><br><span class="line">close BLASTOUT;</span><br><span class="line">close OUT;</span><br></pre></td></tr></table></figure>

<p>经过一系列的转换，得到下面格式的文件，就可以在R程序里面用于GO数据库的构建。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ zcat GMA_GO_TERM.txt.gz | head</span><br><span class="line">GID     GO      ONTOLOGY        EVIDENCE</span><br><span class="line">Glyma.01G000100 GO:0006412      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009063      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009234      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0016021      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0031969      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0016835      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0030976      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0070204      F       IEA</span><br><span class="line">Glyma.01G000137 GO:0016043      P       IEA</span><br></pre></td></tr></table></figure></div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a><a class="next" href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Linux | 认识 PKG_CONFIG_PATH</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/Bioinfo-Ancestral-Genome-evolution/">Bioinfo | Ancestral Genome evolution</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/31/Bioinfo-MCScanX-genome-synteny-blocks/">Bioinfo | MCScanX genome synteny blocks and 4dtv of paralogous genes</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>