<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>Bioinfo | Positive Selection Ka/Ks calculation | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">Bioinfo | Positive Selection Ka/Ks calculation</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">Bioinfo | Positive Selection Ka/Ks calculation</h1><div class="post-meta">2021-10-01<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>总结正向选择分析的一些知识。2017 年做过一次线粒体基因的正向选择，相关内容已经<a target="_blank" rel="noopener" href="https://link.springer.com/article/10.1007%2Fs00606-019-01578-2#Sec18">发表</a>，并且我在Supplementary Files 里附上了我的<a target="_blank" rel="noopener" href="https://static-content.springer.com/esm/art%3A10.1007%2Fs00606-019-01578-2/MediaObjects/606_2019_1578_MOESM1_ESM.pdf">分析代码</a>。2021年再次需要做正向选择分析，再次回顾以前的工作，发现其中也有一个问题，就是没有做统计检验，推算P值，再次学习正向选择分析，并摘抄一些网络资料并整理记录。Codeml 的学习成本比较高，原理与操作都不易理解，学习的过程中发现不少好资源，比如 B 站有<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1Qx411m7Bn?from=search&seid=9774626637377853773&spm_id_from=333.337.0.0">视频</a>讲解如何使用 codeml 推断特定分支的 ω 值（dN/dS），以及使用卡方检验来计算 P 值。另外有国内的团队以 codeml 为内核，用 java 编写了更友好的操作界面：<a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/home.php?mod=space&uid=460481&do=blog&id=1163040">EasyCodeML</a>。</p>
<p><img src="https://d3i71xaburhd42.cloudfront.net/aac1c83c6feff2575f6296ea4751f29c0fec42a1/12-Figure8-1.png" alt="我计算 8 个物种的线粒体基因的 Ka/Ks，但是没有计算 P 值"></p>
<h3 id="基因正选择分析原理"><a href="#基因正选择分析原理" class="headerlink" title="基因正选择分析原理"></a><a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3084">基因正选择分析原理</a></h3><p>下文由 <a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3084">chenlianfu.com</a> 发表于2019 年 11 月 18 日</p>
<p><strong>一、 正选择分析的目的</strong></p>
<p>两两基因的密码子序列进行比较，从而计算 dN/dS，即 omega（ω）值。若该值 &lt; 1，则表示纯化选择；omega = 1，则中性进化；omega &gt; 1，则正选择。若分析基因在两个物种中的序列，可以计算 dN/dS 的值，若 omega &gt; 1，即表明该基因在物种进化过程中，即由其祖先物种分化成这两个物种时，基因受到了正选择。对于两个物种/序列的正选择分析，比较简单。而实际情况中，要分析的物种数量很多，包含多个类群。这个时候的正选择分析相对复杂些。</p>
<p>对多个物种的基因序列进行正选择分析，若仍然按照两个物种时的要求，即分析该基因在物种进化中是否受到了正选择？这种结果可能不好说清楚。因为该基因可能在某一类群中序列很相似，其两两比较时，omega &lt;= 1；而在另外一类群中两两比较时，很多时候 omega &gt; 1。最后软件可以从总体上给一个 omega 值，该值不可以拿来简单地评价该基因是否受到了正选择。所以，对多个物种进行正选择分析时，没法直接评价该基因是否受到了正选择。正选择只有在进行两两序列比较的时候，才能计算 omega 值，从而得到结果。</p>
<p>对基因在多个物种上的正选择分析，分析的目的则是：比较某个分枝上祖先节点和后裔节点（可以理解成，对无根树上某分枝两侧的两组物种进行比较，依然属于两两比较），从而计算该分枝的 omega 值。而在实际数据中，基因在不同的进化分枝上具有不同的 omega 值，同时在序列不同的位点也具有不同的 omega 值。目标分枝两侧的物种数量较多时，可以对序列上的每个位点进行 omega 值分析，从而鉴定正选择位点。所以，对基因在多个物种上的正选择分析，需要同时分析目标分枝的 omega 值和序列位点的 omega 值，从而判断基因是否受到正选择压。</p>
<p><strong>二、使用 PAML 对基因进行正选择分析，有三种方法</strong></p>
<p>PAML site model: 主要用于检测基因中的正选择位点。该方法分析时，认为进化树中各分枝的 omega 值是一致的，并比较两种模型：(1) 模型 m1 是 null model，认为所有位点的 omega 值&lt;1 或 =1; (2) 模型 m2 是正选择模型，存在 omega &lt;1、=1 或 &gt;1 的位点。比较两个模型的似然值（lnL）差异，利用卡方检验（自由度为2）算出 p 值。若 p 值 &lt; 0.05，则否定 null model，认为存在正选择位点。此外，推荐采用比较模型 m7 和 m8，它们将 omega 值分成了 10 类，其 p 值结果比上一种比较方法更宽松，能检测到更多的正选择基因。使用 PAML site model 方法能在整体水平上检测基因的正选择位点，而不能表明基因在某个进化分枝上是否受到正选择压。</p>
<p><font color=blue>PAML branch-site model: 主要用于检测基因在某个进化枝上是否存在正选择位点。该分析方法认为目标分化枝具有一个 omega 值，其它所有分枝具有一个相同的 omega 值，然后再检测正选择位点。同样对两种模型进行比较：（1）第一种模型为模型2，将 omega 值分成 &lt;1、=1、&gt;1 的三类，这和 site model 中的一样；（2）第二种模型和前者一致，只是将 omega 固定成 1，作为 null model。比较两种模型的似然差异，利用卡方检验（自由度为2）算 p 值（chi2 命令算出的值除以2）。若 p 值 &lt; 0.05，则能通过 Bayes Empirical Bayes (BEB) 方法计算正选择位点的后验概率</font>，若存在概率值 &gt; 0.95 正选择位点，则表示基因在目标分枝上受到正选择压。<u><font color=red>PAML 软件在 branch-site 模式下，并不给出分枝上的 omega 值。</font></u>这表示 branch-site 模式虽然考虑了目标分枝上具有不同的 omega 值，但仍然以分析位点上的 omega 为主。值得注意的是，在 branch-site 模式下可能检测到正选择位点，但在目标分枝上的 omega 值仍然可能低于 1。可能软件作者基于这点考虑，就没有给出目标分枝上的 omega 值，以免影响一些人对正选择结果的判断。</p>
<p>PAML branch model: 主要用于检测在某个分枝上，其 omega 值是否显著高于背景分枝，即基因在目标分枝上进化速度加快。该方法认为基因序列上所有位点的 omega 值是一致的，对两种模型进行比较：（1）第一种模型为 null model，所有分枝具有相同的 omega 值；（2）第二种模型认为目标分枝具有一个 omega 值，其它所有分枝具有一个相同的 omega 值。比较两种模型的似然差异，利用卡方检验（自由度为1）算 p 值。若 p 值 &lt;= 0.05，且目标分枝上的 omega 值高于背景值，则认为该基因为快速进化基因。一般情况下，该方法计算得到的 p 值会低于第二种方法的结果。</p>
<p><strong>三、其它注意事项</strong></p>
<p>Branch-site model 相比于 site model 的优点是考虑了不同的分枝具有不同的选择压，即具有不同的 omega 值。该方法让目标分枝具有一个不同的 omega 值，并没有让所有分枝的 omega 值独立进行计算（理论上这样是最好的）。这样算法很复杂，程序运行非常非常消耗时间。但其实也没必要这样做，因为正选择分析其实是两条序列比较后，分析 dN/dS，再找正选择位点，其分析结果就应该是某个分枝上基因是否受到正选择，在序列那个位点上受到正选择。</p>
<p>若在目标分枝上，其 omega 值小于 1，但是却能找到正选择位点。即该基因在该分枝上的 dN/dS &lt; 1，但是在某些位点上，dN/dS &gt; 1。那么该基因是否属于正选择基因？我认为：属于。之所以为正选择基因，主要是因为基因的个别位点或多个位点存在正选择。当只有个别位点受到正选择压力时，而其它多个位点存在纯化选择时，可能导致整体上的omega值小于1。此时，该基因也应该是属于正选择基因。</p>
<p><strong>四、参考文献中的正选择分析方法描述</strong></p>
<p>Science 文章<a target="_blank" rel="noopener" href="https://science.sciencemag.org/content/364/6446/eaav6202"><em>《大规模反刍动物基因组测序与其独特性质的进化研究》</em></a> 中的正选择基因分析方法。</p>
<p> 注：这篇文章做了几十个物种的从头测序与拼接，基本上由 Illumina 测序，说明该项目应该做了挺多年，基因组拼接的质量放在今天来说不算好，大部分物种的 contig N50 不到 50 Kb，后期的比较基因组学分析做的好。主要目的是研究反刍动物特性，例如用于反刍的瘤胃、可再生的骨质角、不同物种巨大的身高差异等。主要手段是研究基因的选择压力。分析思路值得学习，其<a target="_blank" rel="noopener" href="https://www.science.org/doi/suppl/10.1126/science.aav6202/suppl_file/aav6202-chen-sm.pdf"><em>附件</em></a>包含大量中间数据，可作为学习参考。</p>
<p><strong>Identification of positively selected genes (PSGs).</strong> We used a conserved genome synteny methodology to establish a highconfidence orthologous gene set that included camel, cat, dog, horse, human, minke whale, killer whale, pig, and the 51 ruminant species. Briefly, pairwise WGAs were constructed for relevant genomes using LAST, with the goat genome serving as the reference genome. To minimize the effect of annotation, sequencing and assembly errors, pseudogenes, nonorthologous alignments, and non-conserved gene structures 22 on subsequent evolutionary rate analyses, a series of rigorous filtering criteria were adopted: (1) the genes mapped to the goat genome via a single chain of sequence alignments including at least 80% of its coding sequence (CDS), and met the alignment length/score thresholds required for inclusion in the MULTIZ alignments; (2) frame-shift indels in CDSs were prohibited; (3) CDSs with premature stop codons were excluded and (4) genes with Ks values (synonymous substitutions per synonymous site) between each species and cattle larger than two were excluded. As the number of available species for each gene increased, the Codeml running time dramatically increases because of the power function relation between the number of tree leaves and branches. As a trade-off, we subsampled the ruminant species by using the following criteria: (1) if an ortholog existed in more than four species in the Cervidae or Bovidae lineage, we randomly selected 4 sequences to represent either family in order to reduce computational demand; (2) genes contained in less than 4 outgroup species were excluded; (3) genes recovered in fewer than 3 ruminant families were excluded and (4) final alignments shorter than 150 bp were discarded. After these filtering steps, a total of 12,977 genes were retained, and on average ~6.6 families containing ~19.8 species were retained for each ortholog <font color=blue>(fig. S51)</font>. The tree topology shown in Fig. 1A was used as the prior tree topology, and the “<em>ape</em>” package (164) was used to estimate the unrooted tree topology for each ortholog.</p>
<p>To estimate the lineage-specific evolutionary rate for each branch, the Codeml program in the PAML package (version 4.8) (134) with the free-ratio model (model = 1) was run for each ortholog. <font color=blue>Positive selection signals on genes along specific lineages were detected using the optimized branch-site model following the author’s recommendation. A likelihood ratio test (LRT) was conducted to compare a model that allowed sites to be under positive selection on the foreground branch with the null model in which sites could evolve either neutrally and under purifying selection. The p values were computed based on Chi-square statistics, and genes with p value less than 0.05 were treated as candidates that underwent positive selection.</font> We identified PSGs at the ancestral branch of Ruminantia (table S22), the ancestral branch of Pecora (table S23), each ancestral family branch of Ruminantia (tableS24), and each ancestral subfamily branch of Bovidae (table S24). We also compared the dN/dS values of Ruminantia families with outgroup mammals <font color=blue>(fig. S52)</font>.</p>
<p><img src="https://i.imgur.com/a9XwYdx.png"></p>
<p><img src="https://i.imgur.com/he2bRKC.png"></p>
<p>Science 文章<a target="_blank" rel="noopener" href="https://science.sciencemag.org/content/364/6446/eaav6202">《大规模反刍动物基因组测序与其独特性质的进化研究》</a> 中快速进化基因分析方法：The branch model in PAML was used, with the null model (model=0) assuming that all branches have been evolving at the same rate and the alternative model (model=2), allowing the foreground branch to evolve under a different rate. An LRT with df =1 was used to discriminate between alternative models for each ortholog in the gene set. Genes with a p value less than 0.05 and a higher ω value for the foreground than the background branches were considered as evolving with a significantly faster rate in the foreground branch.</p>
<p><a target="_blank" rel="noopener" href="https://journals.plos.org/plosntds/article?id=10.1371/journal.pntd.0007463">Identification of positively selected genes in human pathogenic treponemes: Syphilis-, yaws-, and bejel-causing strains differ in sets of genes showing adaptive evolution</a> 文献中对正选择基因的分析方法：</p>
<p><img src="https://i.imgur.com/35rVAXK.png"></p>
<p>A calculation of mutational rate ratio ω between two gene sequences was the basis for the positive selection analysis. The ω was calculated as a ratio of nonsynonymous to synonymous mutational rates. The ratio indicates negative purifying selection (0 &lt; ω &lt; 1), neutral evolution (ω = 1), and positive selection (ω &gt; 1) [54]. A set of selected genes from complete genomes was tested relative to positive selection using the maximum likelihood method using the CODEML of the PAML software package [55]. PAML version 4 [56] and its user interface PAMLX [57] were used in our study. For each analyzed gene, its maximum likelihood phylogenetic tree was used as an input tree. The CODEML offers several different codon evolutionary models, and the statistical likelihood ratio test (LRT) was used to compare the codon evolutionary model to the null model. The Bayes empirical Bayes method (BEB) [58] was then used to evaluate the posterior probability of sites considered to have been positively selected.</p>
<p>The CODEML models could produce different results (i.e., a list of sites under positive selection) since they calculate different parameter estimates. Site models allow ω to vary in each site (codon) within the gene. Statistical testing was required for sites with ω &gt; 1. Two pairs of models were predominantly used since their LRTs have low false-positive rates. M1a (nearly neutral evolution) was compared to M2a (positive selection) [58,59] and M7 (beta) was compared to M8 (beta &amp; ω) [60]. Our preliminary testing found that the two model pairs gave the same or very similar results. Therefore we chose to use the M7-M8 model pair. The M7 model is a null model that allows 10 classes of sites with a ω beta-distribution within the interval 0 ≤ ω ≤ 1. Sites with ω &gt; 1 are not allowed. The alternative M8 model adds an eleventh class of sites with ω &gt; 1. Each site was tested to determine the class to which it belongs. The LRT compares twice the log-likelihood difference 2Δl = 2(l1-l0) between the M7 model (log likelihood value l0) and the M8 model (log likelihood value l1) to the χ2 distribution [61]. If the twice log-likelihood difference is above a critical χ2 value, then the null model is rejected, and the positive selection is statistically significant.</p>
<p>A considerable disadvantage of the site models is that ω was calculated as an average over all codons of the site. Therefore, the site models are not suitable for the data where ω also varies between lineages. In contrast, the branch-site models search for positive selection in sites and pre-specifies lineages where different rates of ω may occur [62]. Sequences of lineages are a priori divided into a group of foreground lineages where positive selection may occur and group of background lineages where only purifying selection or neutral evolution occurs. We used branch-site model A, which allows four classes of sites and different setups of foreground lineages to be tested depending on the gene phylogeny. In branch-site model A, all lineages under purifying selection with a low value of ω0 belong to site class 0. Weak purifying selection and neutral evolution with ω1 near to value 1 are allowed in site class 1. In site class 2a, a proportion of class 0 sites in foreground lineages is under positive selection with ω2 &gt; 1. Similarly, site class 2b is a proportion of class 1 sites under positive selection with ω2 &gt; 1. The null model for LRT has ω2 = 1. Critical values of LRT (2Δl) are 2.71 at 5% and 5.41 at 1% [63]. The posterior probabilities of suggested sites under positive selection were calculated using the BEB method.</p>
<h3 id="基因-positive-selecetion-分析过程"><a href="#基因-positive-selecetion-分析过程" class="headerlink" title="基因 positive selecetion 分析过程"></a>基因 positive selecetion 分析过程</h3><p>PAML 软件中的 <code>codeml</code> 命令可以使用 Maxmum Likelihood 方法对密码子或氨基酸的多序列比对结果进行分析和进化参数计算 。这两种数据共用一个命令，程序运行时的参数绝大部分是公用的，但仍需要分清两种数据各自的参数意义。PAML软件中进行正选择分析的示例数据位于 <code>examples/lysozyme/</code> 目录下。</p>
<ol>
<li>基因序列准备，对于选定的若干个物种，分别准备它们的基因的 CDS 序列（DNA 格式），要注意每个基因只保留一个转录本即可；另外对序列名进行手动调整，使序列名具有独特性，易识别性，主要是为了便于后续的程序自动循环处理。并将基因 CDS 分别翻译成蛋白序列，注意这一步可能会出现的问题，比如不少物种的基因组文章即使发表在高分杂志上，它的基因集并不一定是完美的，以花生为例，我发现有数个基因CDS直接翻译之后会在序列中间出现终止密码子，这种现象可能是基因组突变，也有可能是基因注释过程中引入的小错误，还有一个我发现有些基因的CDS序列起始并不是ATG这种起始密码子（不排除是 RNA editing 现象），那用普通翻译程序可能就会出现翻译错误。所以针对各个物种，最好分别检查，灵活处理，比如移除这些包含错误的基因，或者直接使用物种自带的蛋白序列集。</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">#########################################################################</span><br><span class="line"># File Name: loop.sh</span><br><span class="line"># Author: fenglei</span><br><span class="line"># Mail: fengleiluck@gmail.com</span><br><span class="line"># Created Time: Fri 08 Oct 2021 04:14:35 PM CST</span><br><span class="line">#########################################################################</span><br><span class="line">#!&#x2F;usr&#x2F;bin&#x2F;bash</span><br><span class="line">set -e</span><br><span class="line">set -u</span><br><span class="line">set -o pipefail</span><br><span class="line"></span><br><span class="line">for i in Ammopiptanthus_nanus.cds.fa Lotus_japonicus.cds.fa Glycine_max.cds.fa Medicago_truncatula.cds.fa Vigna_angularis.cds.fa Arachis_ipaensis.cds.fa Lupinus_angustifolius.cds.fa; do</span><br><span class="line">        #prefix&#x3D;$&#123;i#*.cds&#125;;</span><br><span class="line">        prefix&#x3D;$&#123;i%%.cds*&#125;;</span><br><span class="line">        echo [&#96;date +&quot;%Y-%m-%d %H:%M:%S&quot;&#96;] &quot;#&gt; Processing $prefix&quot;</span><br><span class="line">        iTools Fatools CDS2Pep -InCDS $&#123;prefix&#125;.cds.fa -OutPut  $&#123;prefix&#125;.pep_iTools.fa # 可能翻译的序列种存在终止密码子</span><br><span class="line">        gunzip *gz</span><br><span class="line">        perl pep_seq_clean.pl $&#123;prefix&#125;.pep_iTools.fa $&#123;prefix&#125;.pep.fa</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<p>上面的循环脚本中调用了一个序列清理程序，如下，可以去掉中间含有星号的氨基酸序列，也可以移除蛋白序列尾部的终止密码子形成的星号。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">$ cat pep_seq_clean.pl</span><br><span class="line">#!&#x2F;usr&#x2F;bin&#x2F;perl</span><br><span class="line"></span><br><span class="line">#&#x2F;*************************************************************************</span><br><span class="line">#       &gt; File Name: seq_clean.pl</span><br><span class="line">#       &gt; Author: fenglei</span><br><span class="line">#       &gt; Mail: fengleiluck@gmail.com</span><br><span class="line">##      &gt; Created Time: Fri 08 Oct 2021 01:42:50 PM CST</span><br><span class="line"># ************************************************************************&#x2F;</span><br><span class="line"></span><br><span class="line">#  如果氨基酸序列内部含有终止密码子，就去掉这个序列</span><br><span class="line">#  perl seq_clean.pl input.pep.fa output.pep.fa</span><br><span class="line"></span><br><span class="line">open INPUT, $ARGV[0] or die &quot;Can not open the file!\n $!&quot;;</span><br><span class="line">open OUTPUT, &quot;&gt;$ARGV[1]&quot; or die &quot;Can not open the file!\n $!&quot;;</span><br><span class="line">select OUTPUT;</span><br><span class="line"></span><br><span class="line">$&#x2F; &#x3D; &quot;\&gt;&quot;;</span><br><span class="line">&lt;INPUT&gt;;</span><br><span class="line">while(&lt;INPUT&gt;)&#123;</span><br><span class="line">        chomp $_;</span><br><span class="line">        my $unit &#x3D; $_;</span><br><span class="line">        my @array &#x3D; split(&#x2F;\n&#x2F;, $unit);</span><br><span class="line">        my $seq_name &#x3D; shift @array;</span><br><span class="line">        my $seq&#x3D;&#39;&#39;;</span><br><span class="line">        foreach my $line (@array)&#123;</span><br><span class="line">                $seq &#x3D; $seq . $line;</span><br><span class="line">        &#125;</span><br><span class="line">        #print  &quot;\&gt;$seq_name\n$seq\n&quot;; exit;</span><br><span class="line">        #print &quot;$_n&quot;; exit;</span><br><span class="line">        if($seq &#x3D;~ &#x2F;\*[A-Z]&#x2F;i)&#123;    ### 如果中间出现终止密码子，将剔除这个序列，某物钟一共有大约100条这样的序列</span><br><span class="line">                #print &quot;\&gt;$seq_name\n$seq\n&quot;;</span><br><span class="line">        &#125;</span><br><span class="line">        else&#123;</span><br><span class="line">                #$seq &#x3D;~ s&#x2F;\*&#x2F;&#x2F;;           ### 去掉尾部的星号</span><br><span class="line">                print &quot;\&gt;$seq_name\n&quot;;</span><br><span class="line">                #print &quot;\&gt;$seq_name\n$seq\n&quot;;</span><br><span class="line">                # 如何更美观的输出蛋白序列，每行100个氨基酸</span><br><span class="line">                my $len &#x3D; length($seq)&#x2F;100;</span><br><span class="line">                #print &quot;$len\n&quot;;</span><br><span class="line">                for(my $i&#x3D;0; $i&lt;$len; $i++)&#123;</span><br><span class="line">                        my $sub_line &#x3D; substr($seq, $i*100, 100);</span><br><span class="line">                        $seq &#x3D;~ s&#x2F;\*&#x2F;&#x2F;;           ### 去掉尾部的可能的星号</span><br><span class="line">                        $sub_line &#x3D;~ s&#x2F;\*&#x2F;&#x2F;;</span><br><span class="line">                        print &quot;$sub_line\n&quot;;</span><br><span class="line">                &#125;</span><br><span class="line">        &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">close INPUT;</span><br><span class="line">close OUTPUT;</span><br></pre></td></tr></table></figure>

<p>用 orthofindr 程序寻找物种间的同源基因。我切换到的 Python37 环境，因为 orthofinder 是在这个环境下安装的，之前的博文有记录。</p>
<p>运行命令 <code>orthofinder -t 8 -f orthofinder_data_4 -S diamond</code> 即可进行同源基因寻找。生成的目录下会包含同源基因的列表和 Group 分组情况。下面截取的是程序刚运行的目录，所以只有部分数据呈现。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">$ tree orthofinder_data_4</span><br><span class="line">orthofinder_data_4</span><br><span class="line">├── XXX.fasta -&gt; ..&#x2F;pep_database&#x2F;XXXXXXXX.pep.fa</span><br><span class="line">├── Arachis_ipaensis.pep.fa -&gt; ..&#x2F;pep_database&#x2F;Arachis_ipaensis.pep.fa</span><br><span class="line">├── Cajanus_cajan.pep.fa -&gt; ..&#x2F;pep_database&#x2F;Cajanus_cajan.pep.fa</span><br><span class="line">├── Lja.fasta -&gt; ..&#x2F;pep_database&#x2F;Lotus_japonicus.pep.fa</span><br><span class="line">├── Lupinus_angustifolius.pep.fa -&gt; ..&#x2F;pep_database&#x2F;Lupinus_angustifolius.pep.fa</span><br><span class="line">├── Mtr.fasta -&gt; ..&#x2F;pep_database&#x2F;Medicago_truncatula.pep.fa</span><br><span class="line">├── OrthoFinder</span><br><span class="line">│   └── Results_Oct17</span><br><span class="line">│       ├── Log.txt</span><br><span class="line">│       └── WorkingDirectory</span><br><span class="line">│           ├── diamondDBSpecies0.dmnd</span><br><span class="line">│           ├── diamondDBSpecies1.dmnd</span><br><span class="line">│           ├── diamondDBSpecies2.dmnd</span><br><span class="line">│           ├── diamondDBSpecies3.dmnd</span><br><span class="line">│           ├── diamondDBSpecies4.dmnd</span><br><span class="line">│           ├── diamondDBSpecies5.dmnd</span><br><span class="line">│           ├── diamondDBSpecies6.dmnd</span><br><span class="line">│           ├── SequenceIDs.txt</span><br><span class="line">│           ├── Species0.fa</span><br><span class="line">│           ├── Species1.fa</span><br><span class="line">│           ├── Species2.fa</span><br><span class="line">│           ├── Species3.fa</span><br><span class="line">│           ├── Species4.fa</span><br><span class="line">│           ├── Species5.fa</span><br><span class="line">│           ├── Species6.fa</span><br><span class="line">│           ├── Species7.fa</span><br><span class="line">│           └── SpeciesIDs.txt</span><br><span class="line">├── Psa.fasta -&gt; ..&#x2F;pep_database&#x2F;Pisum_sativum.pep.fa</span><br><span class="line">└── Van.fasta -&gt; ..&#x2F;pep_database&#x2F;Vigna_angularis.pep.fa</span><br></pre></td></tr></table></figure>

<ol start="2">
<li>序列比对，即 codon alignment，依据密码子来进行比对，比直接依据 DNA 比对更准确，比对结果存储为 paml 格式。PAL2NAL 软件可以设定输出格式，默认为 clustal 格式，需要设置为paml。</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">#!&#x2F;bin&#x2F;bash</span><br><span class="line">set -e</span><br><span class="line">set -u</span><br><span class="line">set -o pipefail</span><br><span class="line">for i in &#96;cat commn_genes_list&#96;; do</span><br><span class="line">        echo [&#96;date +&quot;%Y-%m-%d %H:%M:%S&quot;&#96;] &quot;#&gt; sequence convertion and alignment: $&#123;i&#125;.fa&quot;;</span><br><span class="line">        cp ..&#x2F;01_fasta_files&#x2F;$&#123;i&#125;.fa $&#123;i&#125;.nuc;</span><br><span class="line">        echo [&#96;date +&quot;%Y-%m-%d %H:%M:%S&quot;&#96;] &quot;#&gt; Proceccing $&#123;i&#125;.nuc ...&quot;;</span><br><span class="line">        iTools Fatools CDS2Pep -InCDS $&#123;i&#125;.nuc -OutPut $&#123;i&#125;.pep;</span><br><span class="line">        gunzip *gz;</span><br><span class="line">        clustalw2 -TYPE&#x3D;PROTEIN -INFILE&#x3D;$&#123;i&#125;.pep -ALIGN -OUTFILE&#x3D;$&#123;i&#125;.pep.clustalw;</span><br><span class="line">        ~&#x2F;local&#x2F;app&#x2F;PAL2NAL&#x2F;pal2nal.pl $&#123;i&#125;.pep.clustalw $&#123;i&#125;.nuc -output paml -nogap &gt; $&#123;i&#125;.codon</span><br><span class="line">        echo;</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<p>一个基因 codon 比对的结果 paml 格式（the PHYLIP format）如下所示，首行两个数字，一个为物种数目，另一个为碱基数目。随后几行分布是物种名和去除 gap 的碱基序列。对于多个基因，可以直接用 cat 命令连接成一个文件，作为下一步 codeml 的输入。The first line contains the number of species and the sequence length (possibly followed by option characters). For codon sequences (codeml with seqtype = 1), the sequence length in the sequence file refers to the number of nucleotides rather than the number of codons.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">7  390</span><br><span class="line">Hsa_Human        AAGGTCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Hla_gibbon       AAGGTCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Cgu&#x2F;Can_colobus  AAGATCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Pne_langur       AAGATCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Mmu_rhesus       AAGATCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Ssc_squirrelM    AAGGTCTTCGAAAGGTGTGAGTTGGCCAGAACT...</span><br><span class="line">Cja_marmoset     AAGGTCTTTGAAAGGTGTGAGTTGGCCAGAACT...</span><br></pre></td></tr></table></figure>


<ol start="3">
<li>准备一个无根树</li>
</ol>
<p>我之前用过的无根树如下。newick格式，注意一定以分号结尾，否则 codeml 运行的结果中不会出现 Omema 值和 lnL 值。据我自己的经验，分析正向选择，有6~8个物种为宜，少了可能不准确，多了可能时间会很长。如果用有根树，codeml 会报错。无根树怎判断，首先可以用 Figtree 软件打开下面这个文件“unrooted_tree.txt”文件，默认的“rectangular form layput”显示下，从左侧起有 3 个分支，其实也可以直接肉眼看下面的文本内容，从最外层的括号观察，里面包含了 3 个分支。而root tree的图从最左侧开始看，只有2个分支。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$cat unrooted_tree.txt</span><br><span class="line">(((Vicia_faba, Medicago_truncatula), Lotus_japonicus),(Millettia_pinnata, ((Vigna_angularis, Vigna_radiata), Glycine_max)), Ammopiptanthus_mongolicus #1);</span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/home.php?mod=space&uid=1339458&do=blog&id=1023521"><strong>其他人的经验 1</strong></a>：制作tree文件。tree文件最简单的方法就是直接用Clustal制作，推荐 <a target="_blank" rel="noopener" href="http://www.genome.jp/tools/clustalw/">http://www.genome.jp/tools/clustalw/</a>。将序列提交上去，点Execute Multiple Alignment，然后将结果里的dnd文件下载，改名为自己想要的文件+后缀名“.trees”。我这里保存为n1.trees。但是上Phylogenetics课程的时候问了老师一句关于拿Clustal建树用于PAML分析，得到的回答是:  Never, never use Clustal to build a tree。所以还是老老实实用正经的建树工具来做Newick格式的tree文件吧。</p>
<p><a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3036"><strong>其他人的经验 2</strong></a>：输入一个树文件input.trees。该文件可以是有根树，也可以是无根树，也可以带有枝长信息。codeml命令用于分析各分枝上的omega值，若是有根树的话，对于分析root节点所在的分枝，其结果可能是不对的。因此，作者推荐使用无根树进行正选择分析。无根树是3分叉的树，最少需要3个物种；而3个物种的无根数有3个分枝（2n – 3），对其进行分析，仅能得到某个物种所对应分枝的omega值，这种情况下没太大意义；而使用4个物种的无根树进行分析，可以得到两个物种共同祖先分枝的omega值。故作者推荐使用至少4~5个物种进行分析。一个树文件内容示例。文件内容分两行：第一行表述树中有多少个物种，共计1个树，两个数值之间用任意长度的空分割。此外，Newick格式的树尾部一定要有分号，没有的话程序可能不能正常运行。</p>
<p><a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1Qx411m7Bn?from=search&seid=9774626637377853773&spm_id_from=333.337.0.0"><strong>其他人的经验 3</strong></a>：如果要计算某个分支，可以直接在分支后面加一个“$1”，即美元符号。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">7  1</span><br><span class="line">((Hsa_Human, Hla_gibbon) #1,((Cgu&#x2F;Can_colobus, Pne_langur) #1, Mmu_rhesus), (Ssc_squirrelM, Cja_marmoset));</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">((Hsa_Human, Hla_gibbon), ((Cgu&#x2F;Can_colobus, Pne_langur) $1, Mmu_rhesus), (Ssc_squirrelM, Cja_marmoset));</span><br></pre></td></tr></table></figure>

<ol start="4">
<li>准备 codeml 的配置文件</li>
</ol>
<p>下面是从 &lt;chenlianfu.com&gt; 摘抄的一个配置案例：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line">*输入输出参数：</span><br><span class="line">      seqfile &#x3D; input.txt    *设置输入的多序列比对文件路径。</span><br><span class="line">     treefile &#x3D; input.trees  *设置输入的树文件路径。</span><br><span class="line">      outfile &#x3D; mlc          *设置输出文件路径。</span><br><span class="line">        noisy &#x3D; 9            *设置输出到屏幕上的信息量等级：0，1，2，3，9。</span><br><span class="line">      verbose &#x3D; 0            *设置输出到结果文件中的信息量等级：0，精简模式结果（推荐）；1，输出详细信息，包含碱基序列；2，输出更多信息。</span><br><span class="line">        getSE &#x3D; 0            *设置是否计算并获得各参数的标准误：0，不需要；1，需要。</span><br><span class="line"> RateAncestor &#x3D; 0            *设置是否计算序列中每个位点的替换率：0，不需要；1，需要。当设置成1时，会在结果文件rates中给出各个位点的碱基替换率；同时也会进行祖先序列的重构，在结果文件rst中有体现。</span><br><span class="line"></span><br><span class="line">*数据使用说明参数：</span><br><span class="line">      runmode &#x3D; 0      *设置程序获取进化树拓扑结构的方法：0，直接从输入文件中获取进化树拓扑结构（推荐）；1，程序以输入文件中的多分叉树作为起始树，并使用星状分解法搜索最佳树；2，程序直接以星状树作为起始树，并使用星状分解法搜索最佳树；3，使用逐步添加法搜索最佳树；4，使用简约法获取起始树，再进行邻近分枝交换寻找最优树；5，以输入文件中的树作为起始树，再进行邻近分枝交换寻找最优树；-2，对密码子序列进行两两比较并使用ML方法计算DnDs，或对蛋白序列进行两两比较计算ML距离，而不进行其它参数（枝长和omega等）的计算。</span><br><span class="line">* fix_blength &#x3D; 1      *设置程序处理输入树文件中枝长数据的方法：0，忽略输入树文件中的枝长信息；-1，使用一个随机起始进行进行计算；1，以输入的枝长信息作为初始值进行ML迭代分析，此时需要注意输入的枝长信息是碱基替换率，而不是碱基替换数；2，不需要使用ML迭代计算枝长，直接使用输入的枝长信息，需要注意，若枝长信息和序列信息不吻合可能导致程序崩溃；3，让ML计算出的枝长和输入的枝长呈正比。</span><br><span class="line">      seqtype &#x3D; 1      *设置输入的多序列比对数据的类型：1，密码子数据；2，氨基酸数据；3，输入数据虽然为密码子序列，但先转换为氨基酸序列后再进行分析。</span><br><span class="line">    CodonFreq &#x3D; 2      *设置密码子频率的计算算法：0，表示除三种终止密码子频率为0，其余密码值频率全部为1&#x2F;61；1，程序分别计算三个密码子位点的四种碱基频率，再算四种碱基频率在三个位点上的算术平均值，计算61种密码子频率时，使用该均值进行计算，这61种密码子频率之和不等于1，然后对其数据标准化，使其和为1，得到所有密码子的频率；2，程序分别计算三个密码子位点的四种碱基频率，计算61种密码子频率时，使用相应位点的碱基频率进行计算，这61种密码子频率之和不等于1，然后对其数据标准化，使其和为1，得到所有密码子的频率；3，直接使用观测到的各密码子的总的频数&#x2F;所有密码值的总数，得到所有密码子的频率。一般选择第三种方法，设置CodonFre的值为2。第一种方法让所有密码子频率相等，不符合实际轻卡；第二种方法没有考虑到三种位点各碱基频率的差异，得到的密码子频率不准确；第三种方法能比较准确计算出各密码子的频率；第四种方法由输入数据得到真实的各密码子频率，但有时候有些非终止密码子的频率为0，可能不利于后续的计算。</span><br><span class="line">    cleandata &#x3D; 1      *设置是否移除不明确字符（N、？、W、R和Y等）或含以后gap的列后再进行数据分析：0，不需要，但在序列两两比较的时候，还是会去除后进行比较；1，需要。</span><br><span class="line">*       ndata &#x3D; 1      *设置输入的多序列比对的数据个数。</span><br><span class="line">        clock &#x3D; 0      *设置进化树中各分支上的变异速率是否一致，服从分子钟理论：0，变异速率不一致，不服从分子钟理论（当输入数据中物种相差较远时，各分支变异速率不一致）；1，所有分枝具有相同的变异速率，全局上服从分子钟理论；2，进化树局部符和分子钟理论，程序认为除了指定的分支具有不同的进化速率，其它分枝都具有相同的进化速率，这要求输入的进化树信息中使用#来指定分支；3，对多基因数据进行联合分析。</span><br><span class="line">        Mgene &#x3D; 0      *设置是否有多个基因的多序列比对信息输入，以及多各基因之间的参数是否一致：0，输入的多序列比对文件中仅包含一个基因时，或多个基因具有相同的Kappa和Pi参数；1，输入文件包含多个基因，这些基因之间是相互独立的（这些基因之间具有不同的Kappa和Pi值，且其进化树的枝长也不相关）；2，输入文件包含多个基因，这些基因具有相同的Kappa值，不同的Pi值；3，输入文件包含多个基因，这些基因具有相同的Pi值，不同的Kappa值；4，输入文件包含多个基因，这些基因具有不同的Kappa值和不同的Pi值。当值是2、3或4时，多个基因的进化树枝长虽然长度不一样，但是呈正比关系。这点和参数值等于1是不一样的。</span><br><span class="line">        icode &#x3D; 0      *设置遗传密码。其值1-10和NCBI的1-11遗传密码规则对应：0，表示通用的遗传密码。</span><br><span class="line">   Small_Diff &#x3D; .5e-6  *设置一个很小的值，一般位于1e-8到1e-5之间。推荐检测设置不同的值在比较其结果，需要该参数值对结果没什么影响。</span><br><span class="line"></span><br><span class="line">*位点替换模型参数：</span><br><span class="line">        model &#x3D; 1            *若输入数据是密码子序列，该参数用于设置branch models，即进化树各分枝的omega值的分布：0，进化树上所有分枝的omega值一致；1，对每个分枝单独进行omega计算；2，设置多类omega值，根据树文件中对分枝的编号信息来确定类别，具有相同编号的分枝具有相同的omega值，没有编号的分枝具有相同的omega值，程序分别计算各编号和没有编号的omega值。</span><br><span class="line">                             *若输入数据是蛋白序列，或数据是密码子序列且seqtype值是3时，该参数用于设置氨基酸替换模型：0，Poisson；1，氨基酸替换率和氨基酸的观测频率成正比；2，从aaRatefile参数指定的文件路径中读取氨基酸替换率信息，这些信息是根据经验获得，并记录到后缀为.dat的配置文件中。这些经验模型(Empirical Models)文件位于PAML软件安装目录中的dat目录下，例如，Dayhoff(dayhoff.dat)、WAG(wag.dat)、LG(lg.dat)、mtMAN(mtman.dat)和mtREV24(mtREV24.dat)等；</span><br><span class="line">   aaRatefile &#x3D; dat&#x2F;wag.dat  *当对蛋白数据进行分析，且model &#x3D; 2时，该参数生效，用于设置氨基酸替换模型。</span><br><span class="line">       aaDist &#x3D; 0            *设置氨基酸之间的距离。</span><br><span class="line">      NSsites &#x3D; 0            *输入数据时密码子序列时生效，用于设置site model，即序列各位点的omega值的分布：0，所有位点具有相同的omega值；1，各位点上的omega值小于1或等于1（服从中性进化neutral）；2，各位点上的omega值小于1、等于1或大于1（选择性进化selection）；3，discrete；4，freq；5:gamma；6，2gamma；7，beta；8，beta&amp;w；9，beta&amp;gamma；10，beta&amp;gamma+1；11，beta&amp;normal&gt;1；12，0&amp;2normal&gt;1；13，3normal&gt;0。</span><br><span class="line">                             *可以一次输入多个模型进行计算并比较，其结果输出的rst文件中。</span><br><span class="line">    fix_alpha &#x3D; 1            *序列中不同的位点具有不同的碱基替换率，服从discrete-gamma分布，该模型通过alpha（shape参数）和ncatG参数控制。该参数设置是否给定一个alpha值：0，使用ML方法对alpha值进行计算；1，使用下一个参数设置一个固定的alpha值。</span><br><span class="line">	                     *对于密码子序列，当NSsites参数值不为0或model不为0时，推荐设置fix_alpha &#x3D; 1且alpha &#x3D; 0，即不设置alpha值，认为位点间的变异速率一致，否则程序报错。若设置了alpha值，则程序认为不同密码子位点的变异速率不均匀，且同时所有位点的omega值一致，当然各分枝的omega值也会一致，这时要求NSsites和model参数值都设置为0（这一般不是我们需要的分析，它不能进行正选择分析了）。</span><br><span class="line">        alpha &#x3D; 0            *设置一个固定的alpha值或初始alpha值（推荐设置为0.5）。该值小于1，表示只有少数热点位置的替换率较高；该值越小，表示位点替换率在各位点上越不均匀；若设置fix_alpha &#x3D; 1且alpha &#x3D; 0则表示所有位点的替换率是恒定一致的。</span><br><span class="line">       Malpha &#x3D; 0            *当输入的多序列比对结果中有多基因时，设置这些基因间的alpha值是否相等：0，分别对每个基因单独计算alpha值；1，所有基因的alpha值保持一致。</span><br><span class="line">        ncatG &#x3D; 5            *序列中不同位点的变异速率服从GAMMA分布的，ncatG是其一个参数，一般设置为5，4，8或10，且序列条数越多，该值设置越大。</span><br><span class="line">                             *对于密码子序列，当NSites设置为3时，ncatG设置为3；当NSites设置为4时，ncatG设置为5；当NSites值设置&gt;&#x3D;5时，ncatG值设置为10。</span><br><span class="line">    fix_kappa &#x3D; 0            *设置是否给定一个Kappa值：0，通过ML迭代来估算Kappa值；1，使用下一个参数设置一个固定的Kappa值。</span><br><span class="line">        kappa &#x3D; 2            *设置一个固定的Kappa值，或一个初始的Kappa值。</span><br><span class="line">    fix_omega &#x3D; 0            *设置是否给定一个omega值：0，通过ML迭代来估算Kappa值；1，使用下一个参数设置一个固定的omega值。 </span><br><span class="line">        omega &#x3D; .4           *设置一个固定的omega值，或一个初始的omega值。</span><br><span class="line">       method &#x3D; 0            *设置评估枝长的ML迭代算法：0，使用PAML的老算法同时计算所有枝长，在clock &#x3D; 0下有效；1，PAML新加入的算法，一次对一个枝长进行计算，该算法仅在clock参数值为1，2或3下工作。</span><br></pre></td></tr></table></figure>

<p>下面是我自己用的一个配置</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">#$ cat codeml-M0.ctl</span><br><span class="line">      seqfile &#x3D; mtGenes.aligned.DNA.paml      * sequence data filename</span><br><span class="line">     treefile &#x3D; unrooted_tree.txt      * tree structure file name</span><br><span class="line">      outfile &#x3D; mtGenes.substitution           * main result file name</span><br><span class="line"></span><br><span class="line">      runmode &#x3D; 0  * 0: user tree;  1: semi-automatic;  2: automatic</span><br><span class="line">                   * 3: StepwiseAddition; (4,5):PerturbationNNI; -2: pairwise</span><br><span class="line"></span><br><span class="line">      seqtype &#x3D; 1  * 1:codons; 2:AAs; 3:codons--&gt;AAs</span><br><span class="line">    CodonFreq &#x3D; 2  * 0:1&#x2F;61 each, 1:F1X4, 2:F3X4, 3:codon table</span><br><span class="line"></span><br><span class="line">        ndata &#x3D; 30 * number of gene alignments to be analysed</span><br><span class="line">        clock &#x3D; 0  * 0:no clock, 1:clock; 2:local clock; 3:CombinedAnalysis</span><br><span class="line"></span><br><span class="line">        model &#x3D; 2  * models for codons: 0:one, 1:b, 2:2 or more dN&#x2F;dS ratios for branches</span><br><span class="line"></span><br><span class="line">      NSsites &#x3D; 0  * 0:one w;1:neutral;2:selection; 3:discrete;4:freqs;</span><br><span class="line">                   * 5:gamma;6:2gamma;7:beta;8:beta&amp;w;9:beta&amp;gamma;</span><br><span class="line">                   * 10:beta&amp;gamma+1; 11:beta&amp;normal&gt;1; 12:0&amp;2normal&gt;1;</span><br><span class="line">                   * 13:3normal&gt;0</span><br><span class="line"></span><br><span class="line">        icode &#x3D; 0  * 0:universal code; 1:mammalian mt; 2-10:see below</span><br><span class="line"></span><br><span class="line">    fix_omega &#x3D; 0  * 1: omega or omega_1 fixed, 0: estimate </span><br><span class="line">        omega &#x3D; .4 * initial or fixed omega, for codons or codon-based AAs</span><br><span class="line"></span><br><span class="line">    cleandata &#x3D; 0  * remove sites with ambiguity data (1:yes, 0:no)?</span><br><span class="line"></span><br><span class="line">* Genetic codes: 0:universal, 1:mammalian mt., 2:yeast mt., 3:mold mt.,</span><br><span class="line">* 4: invertebrate mt., 5: ciliate nuclear, 6: echinoderm mt., </span><br><span class="line">* 7: euplotid mt., 8: alternative yeast nu. 9: ascidian mt., </span><br><span class="line">* 10: blepharisma nu.</span><br><span class="line">* These codes correspond to transl_table 1 to 11 of GENEBANK.</span><br></pre></td></tr></table></figure>

<p>下面是结果的展示。可以看到 w (dN/dS) for branches:  0.37585 0.22840，尾部的0.2284就是我们设置#1分支的ω，而第一个0.37585是其他分支的ω。由于所设定的物种的ω小于1，所以这个基因在该物种不是正向选择。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br></pre></td><td class="code"><pre><span class="line">TREE #  1:  (((8, 7), 5), (4, ((1, 2), 3)), 6);   MP score: 160</span><br><span class="line">check convergence..</span><br><span class="line">lnL(ntime: 13  np: 16):  -3122.716855      +0.000000</span><br><span class="line">   9..10   10..11   11..8    11..7    10..5     9..12   12..4    12..13   13..14   14..1    14..2    13..3     9..6  </span><br><span class="line"> 0.008466 0.038271 0.303083 0.013331 0.024760 0.007420 0.033139 0.020543 0.021931 0.003298 0.005453 0.018905 0.029024 0.643248 0.375846 0.228401</span><br><span class="line"></span><br><span class="line">Note: Branch length is defined as number of nucleotide substitutions per codon (not per neucleotide site).</span><br><span class="line"></span><br><span class="line">tree length &#x3D;   0.52763</span><br><span class="line"></span><br><span class="line">(((8: 0.303083, 7: 0.013331): 0.038271, 5: 0.024760): 0.008466, (4: 0.033139, ((1: 0.003298, 2: 0.005453): 0.021931, 3: 0.018905): 0.020543): 0.007420, 6: 0.029024);</span><br><span class="line"></span><br><span class="line">(((Vicia_faba: 0.303083, Medicago_truncatula: 0.013331): 0.038271, Lotus_japonicus: 0.024760): 0.008466, (Millettia_pinnata: 0.033139, ((Vigna_angularis: 0.003298, Vigna_radiata: 0.005453): 0.021931, Glycine_max: 0.018905): 0.020543): 0.007420, Ammopiptanthus_mongolicus: 0.029024);</span><br><span class="line"></span><br><span class="line">Detailed output identifying parameters</span><br><span class="line"></span><br><span class="line">kappa (ts&#x2F;tv) &#x3D;  0.64325</span><br><span class="line"></span><br><span class="line">w (dN&#x2F;dS) for branches:  0.37585 0.22840</span><br><span class="line"></span><br><span class="line">dN &amp; dS for each branch</span><br><span class="line"></span><br><span class="line"> branch          t       N       S   dN&#x2F;dS      dN      dS  N*dN  S*dS</span><br><span class="line"></span><br><span class="line">   9..10     0.008  1077.6   338.4  0.3758  0.0020  0.0054   2.2   1.8</span><br><span class="line">  10..11     0.038  1077.6   338.4  0.3758  0.0091  0.0243   9.8   8.2</span><br><span class="line">  11..8      0.303  1077.6   338.4  0.3758  0.0723  0.1924  77.9  65.1</span><br><span class="line">  11..7      0.013  1077.6   338.4  0.3758  0.0032  0.0085   3.4   2.9</span><br><span class="line">  10..5      0.025  1077.6   338.4  0.3758  0.0059  0.0157   6.4   5.3</span><br><span class="line">   9..12     0.007  1077.6   338.4  0.3758  0.0018  0.0047   1.9   1.6</span><br><span class="line">  12..4      0.033  1077.6   338.4  0.3758  0.0079  0.0210   8.5   7.1</span><br><span class="line">  12..13     0.021  1077.6   338.4  0.3758  0.0049  0.0130   5.3   4.4</span><br><span class="line">  13..14     0.022  1077.6   338.4  0.3758  0.0052  0.0139   5.6   4.7</span><br><span class="line">  14..1      0.003  1077.6   338.4  0.3758  0.0008  0.0021   0.8   0.7</span><br><span class="line">  14..2      0.005  1077.6   338.4  0.3758  0.0013  0.0035   1.4   1.2</span><br><span class="line">  13..3      0.019  1077.6   338.4  0.3758  0.0045  0.0120   4.9   4.1</span><br><span class="line">   9..6      0.029  1077.6   338.4  0.2284  0.0054  0.0234   5.8   7.9</span><br><span class="line"></span><br><span class="line">tree length for dN:       0.1243</span><br><span class="line">tree length for dS:       0.3400</span><br><span class="line">dS tree:</span><br><span class="line">(((Vicia_faba: 0.192424, Medicago_truncatula: 0.008464): 0.024298, Lotus_japonicus: 0.015720): 0.005375, (Millettia_pinnata: 0.021039, ((Vigna_angularis: 0.002094, Vigna_radiata: 0.003462): 0.013924, Glycine_max: 0.012003): 0.013043): 0.004711, Ammopiptanthus_mongolicus: 0.023436);</span><br><span class="line">dN tree:</span><br><span class="line">(((Vicia_faba: 0.072322, Medicago_truncatula: 0.003181): 0.009132, Lotus_japonicus: 0.005908): 0.002020, (Millettia_pinnata: 0.007908, ((Vigna_angularis: 0.000787, Vigna_radiata: 0.001301): 0.005233, Glycine_max: 0.004511): 0.004902): 0.001770, Ammopiptanthus_mongolicus: 0.005353);</span><br><span class="line"></span><br><span class="line">w ratios as labels for TreeView:</span><br><span class="line">(((Vicia_faba #0.3758 , Medicago_truncatula #0.3758 ) #0.3758 , Lotus_japonicus #0.3758 ) #0.3758 , (Millettia_pinnata #0.3758 , ((Vigna_angularis #0.3758 , Vigna_radiata #0.3758 ) #0.3758 , Glycine_max #0.3758 ) #0.3758 ) #0.3758 , Ammopiptanthus_mongolicus #0.2284 );</span><br><span class="line"></span><br><span class="line">Time used:  0:05</span><br></pre></td></tr></table></figure>

<p>上述 ctl 文件中 model = 2，运行得到一个结果文件。然后将其中的 model 参数改为 0，再次运行，得到新的结果文件。那么现在有三个文件是我需要的。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">all.genes.list</span><br><span class="line">result.branch_model_0.mcl</span><br><span class="line">result.branch_model_2.mcl</span><br></pre></td></tr></table></figure>

<p>我根据这三个文件，得到了下面的表格。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">$ perl chi2_calculator.pl </span><br><span class="line"># The number of &quot;genes, lnL0 number, lnL1 number&quot; are the same.</span><br><span class="line"># If you see this info, the data is probably OK.</span><br><span class="line">#Number Group   lnL0    lnL1    Omega_branch    Omega_controlP_value</span><br><span class="line">Number:1        atp1    lnL0:-3145.827397       lnL1:-3122.716855       Omega_branch:0.22840    Omega_control:0.37585   P_value:0.000000000</span><br><span class="line">Number:2        atp4    lnL0:-930.895378        lnL1:-930.038146        Omega_branch:0.45618    Omega_control:0.52692   P_value:0.424335018</span><br><span class="line">Number:3        atp6    lnL0:-1453.909109       lnL1:-1446.498209       Omega_branch:0.86281    Omega_control:0.37541   P_value:0.000604626</span><br><span class="line">Number:4        atp8    lnL0:-924.486129        lnL1:-914.009795        Omega_branch:1.75174    Omega_control:0.19043   P_value:0.000028196</span><br><span class="line">Number:5        atp9    lnL0:-333.236113        lnL1:-331.374307        Omega_branch:0.00010    Omega_control:0.39757   P_value:0.155391739</span><br><span class="line">Number:6        ccmB    lnL0:-1023.961013       lnL1:-1022.534291       Omega_branch:999.00000  Omega_control:0.56135   P_value:0.240094664</span><br><span class="line">Number:7        ccmC    lnL0:-1131.746218       lnL1:-1129.394721       Omega_branch:999.00000  Omega_control:0.39406   P_value:0.095226501</span><br><span class="line">Number:8        ccmFC   lnL0:-1950.197350       lnL1:-1948.956865       Omega_branch:0.42724    Omega_control:0.88398   P_value:0.289243901</span><br><span class="line">Number:9        ccmFN   lnL0:-2737.722421       lnL1:-2737.051425       Omega_branch:1.13399    Omega_control:0.65455   P_value:0.511199170</span><br><span class="line">Number:10       cob     lnL0:-1836.351940       lnL1:-1826.480046       Omega_branch:0.16318    Omega_control:0.22020   P_value:0.000051605</span><br><span class="line">Number:11       cox1    lnL0:-2381.442730       lnL1:-2354.196627       Omega_branch:0.09805    Omega_control:0.04898   P_value:0.000000000</span><br><span class="line">Number:12       cox3    lnL0:-1220.568961       lnL1:-1218.320955       Omega_branch:0.00010    Omega_control:0.53125   P_value:0.105609600</span><br><span class="line">Number:13       matR    lnL0:-3293.272506       lnL1:-3288.780620       Omega_branch:1.06998    Omega_control:0.46260   P_value:0.011199502</span><br><span class="line">Number:14       mttB    lnL0:-1110.395753       lnL1:-1109.356198       Omega_branch:1.07609    Omega_control:0.44712   P_value:0.353612004</span><br><span class="line">Number:15       nad1    lnL0:-1410.749338       lnL1:-1404.082741       Omega_branch:0.34165    Omega_control:0.13677   P_value:0.001272722</span><br><span class="line">Number:16       nad2    lnL0:-2199.786871       lnL1:-2187.259279       Omega_branch:0.80075    Omega_control:0.12534   P_value:0.000003625</span><br><span class="line">Number:17       nad3    lnL0:-537.855553        lnL1:-532.299091        Omega_branch:0.00010    Omega_control:0.27379   P_value:0.003862417</span><br><span class="line">Number:18       nad4    lnL0:-2549.647893       lnL1:-2546.428114       Omega_branch:0.25913    Omega_control:0.61043   P_value:0.039963889</span><br><span class="line">Number:19       nad4L   lnL0:-420.857873        lnL1:-413.049727        Omega_branch:0.00010    Omega_control:0.00010   P_value:0.000406411</span><br><span class="line">Number:20       nad5    lnL0:-2986.298356       lnL1:-2975.844248       Omega_branch:0.15621    Omega_control:0.26908   P_value:0.000028830</span><br><span class="line">Number:21       nad6    lnL0:-1099.896407       lnL1:-1092.361003       Omega_branch:0.54957    Omega_control:0.24816   P_value:0.000533846</span><br><span class="line">Number:22       nad7    lnL0:-1766.620230       lnL1:-1765.116602       Omega_branch:0.51002    Omega_control:0.42390   P_value:0.222322111</span><br><span class="line">Number:23       nad9    lnL0:-1032.089035       lnL1:-1031.583127       Omega_branch:999.00000  Omega_control:0.86278   P_value:0.602957841</span><br><span class="line">Number:24       rpl16   lnL0:-895.556806        lnL1:-891.758505        Omega_branch:0.27993    Omega_control:0.35303   P_value:0.022408812</span><br><span class="line">Number:25       rpl5    lnL0:-920.541028        lnL1:-919.686869        Omega_branch:0.86954    Omega_control:0.50538   P_value:0.425641005</span><br><span class="line">Number:26       rps10   lnL0:-603.758104        lnL1:-600.272712        Omega_branch:999.00000  Omega_control:0.30933   P_value:0.030641745</span><br><span class="line">Number:27       rps12   lnL0:-613.237087        lnL1:-609.335187        Omega_branch:999.00000  Omega_control:0.21780   P_value:0.020203488</span><br><span class="line">Number:28       rps14   lnL0:-493.189267        lnL1:-491.210035        Omega_branch:0.26530    Omega_control:0.31987   P_value:0.138175315</span><br><span class="line">Number:29       rps3    lnL0:-2982.678697       lnL1:-2978.575083       Omega_branch:0.54439    Omega_control:0.53311   P_value:0.016512890</span><br><span class="line">Number:30       rps4    lnL0:-1761.670696       lnL1:-1760.205396       Omega_branch:0.44191    Omega_control:0.58927   P_value:0.231008678</span><br></pre></td></tr></table></figure>

<p>附上 <code>chi2_calculator.pl</code> 程序的代码。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br></pre></td><td class="code"><pre><span class="line">#!&#x2F;usr&#x2F;bin&#x2F;perl</span><br><span class="line">#&#x2F;*************************************************************************</span><br><span class="line">#	&gt; File Name: chi2_calculator.pl</span><br><span class="line">#	&gt; Author: fenglei</span><br><span class="line">#	&gt; Mail: fengleiluck@gmail.com</span><br><span class="line">#	&gt; Created Time: Thu 07 Oct 2021 01:26:04 PM CST</span><br><span class="line"># ************************************************************************&#x2F;</span><br><span class="line"></span><br><span class="line">open GENE_LIST, &quot;&lt;30_genes.list&quot; or die &quot;Can not open the file!\n&quot;;</span><br><span class="line">open NONMODEL, &quot;&lt;mtGenes.branch_model_0.mcl&quot; or die &quot;Can not open the file!\n&quot;;  # 找到这样的行：lnL(ntime: 13  np: 16):  -3122.716855      +0.000000  </span><br><span class="line">open ALTERNATIVE, &quot;&lt;mtGenes.branch_model_2.mcl&quot; or die &quot;Can not open the file!\n&quot;;</span><br><span class="line"></span><br><span class="line">#####  perl chi2_calculator.pl &gt; out</span><br><span class="line">#####  依据上面三个文件，依次计算P值（通过似然率检验（LRT）（卡方检验）获得p值确定零假设模型和备选模型之间是否存在差异）</span><br><span class="line">#      能否同时输出 omega 值？  Alternative 文件里面有一行“w (dN&#x2F;dS) for branches:  0.37585 0.22840” 第二个数字是我需要的</span><br><span class="line">my @genes;</span><br><span class="line">while(&lt;GENE_LIST&gt;)&#123;</span><br><span class="line">	chomp $_;</span><br><span class="line">	push @genes, $_;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">my @lnL0;</span><br><span class="line">while(&lt;NONMODEL&gt;)&#123;</span><br><span class="line">	chomp $_;</span><br><span class="line">	my $record &#x3D; $_;</span><br><span class="line">	if($record &#x3D;~ m&#x2F;lnL&#x2F;)&#123;</span><br><span class="line">		my @array &#x3D; split(&#x2F;\s+&#x2F;, $record);</span><br><span class="line">		#print &quot;lnL0: $array[4]\n&quot;; exit;</span><br><span class="line">		push @lnL0, $array[4];</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">my @lnL1;</span><br><span class="line">my @omega_branch;</span><br><span class="line">my @omega_control;</span><br><span class="line">while(&lt;ALTERNATIVE&gt;)&#123;</span><br><span class="line">	chomp $_;</span><br><span class="line">	my $record &#x3D; $_;</span><br><span class="line">	if($record &#x3D;~ m&#x2F;lnL&#x2F;)&#123;   # 找到这样的行：lnL(ntime: 13  np: 16):  -3122.716855      +0.000000 </span><br><span class="line">		my @array &#x3D; split(&#x2F;\s+&#x2F;, $record);</span><br><span class="line">		push @lnL1, $array[4];</span><br><span class="line">	&#125;</span><br><span class="line">	if($record &#x3D;~ m&#x2F;w\s\(dN\&#x2F;dS\)\sfor\sbranches\:&#x2F;)&#123;   # 找到这样的行：w (dN&#x2F;dS) for branches:  0.37585 0.22840” 第二个数字是我需要的选定brach的omega</span><br><span class="line">		my @array &#x3D; split(&#x2F;\s+&#x2F;, $record);</span><br><span class="line">		push @omega_branch,  $array[-1];</span><br><span class="line">		push @omega_control, $array[-2];</span><br><span class="line">	&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">if($#genes &#x3D;&#x3D; $#lnL0 &amp;&amp; $#genes &#x3D;&#x3D; $#lnL1)&#123;</span><br><span class="line">	print &quot;# The number of \&quot;genes, lnL0 number, lnL1 number\&quot; are the same.\n# If you see this info, the data is probably OK.\n&quot;;</span><br><span class="line">	#exit;</span><br><span class="line">&#125;</span><br><span class="line">else&#123;</span><br><span class="line">	print &quot;# The number of \&quot;genes, lnL0 number, lnL1 number\&quot; are NOT the same.\n# If you see this info, the data is probably NOT OK.\n# You should have a check.\n&quot;;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">print &quot;#Number\tGroup\tlnL0\tlnL1\tOmega_branch\tOmega_control\tP_value\n&quot;;</span><br><span class="line"></span><br><span class="line">for(my $count&#x3D;0; $count&lt;&#x3D;$#genes; $count++)&#123;</span><br><span class="line">	$num &#x3D; $count + 1;</span><br><span class="line">	print &quot;Number:$num\t$genes[$count]\tlnL0:$lnL0[$count]\tlnL1:$lnL1[$count]\tOmega_branch:$omega_branch[$count]\tOmega_control:$omega_control[$count]\t&quot;;</span><br><span class="line">	my $likelihood &#x3D; 2 * abs( $lnL0[$count] - $lnL1[$count] );</span><br><span class="line">	my $df  &#x3D; 2;  ### ??? df&#x3D;1 or df&#x3D;2 ???</span><br><span class="line">	system(&quot;~&#x2F;local&#x2F;app&#x2F;PAML4.8&#x2F;bin&#x2F;chi2 $df $likelihood &gt; &#x2F;tmp&#x2F;chi2_result.txt&quot;);  # 三行的一个文件：首行空格，次行“df &#x3D;  2  prob &#x3D; 0.016512890 &#x3D; 1.651e-02”，末行空格。需要提取P值</span><br><span class="line">	open CHISQUARE, &quot;&#x2F;tmp&#x2F;chi2_result.txt&quot; or die &quot;Can not open the file!\n&quot;;</span><br><span class="line">	my $p_value &#x3D; 1;</span><br><span class="line">	while(&lt;CHISQUARE&gt;)&#123;</span><br><span class="line">		chomp $_;</span><br><span class="line">		my $record &#x3D; $_;</span><br><span class="line">		if($record &#x3D;~ m&#x2F;prob&#x2F;)&#123;</span><br><span class="line">			my @array &#x3D; split(&#x2F;\s+&#x2F;, $record);</span><br><span class="line">			$p_value &#x3D; $array[-3]&#x2F;2;  ### PAML文档（http:&#x2F;&#x2F;abacus.gene.ucl.ac.uk&#x2F;software&#x2F;pamlDOC.pdf ; Page 30），p值应该是chi2程序的结果除以2。chenlianfu也有如此记录）</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	print &quot;P_value:$p_value\n&quot;;</span><br><span class="line">	close CHISQUARE;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">&#x3D;pod</span><br><span class="line">system(&quot;~&#x2F;local&#x2F;app&#x2F;PAML4.8&#x2F;bin&#x2F;chi2 2 55&quot;);</span><br><span class="line">$df &#x3D; 2;</span><br><span class="line">$likelihood &#x3D; 23;</span><br><span class="line">system(&quot;~&#x2F;local&#x2F;app&#x2F;PAML4.8&#x2F;bin&#x2F;chi2 $df $likelihood &gt;&gt; P_value.txt&quot;);</span><br><span class="line">print &quot;test\n&quot;;</span><br><span class="line"># print &quot;$a\n&quot;;  # 这命令输出的是 0，也就是 system 命令输出在屏幕的内容，不能传递给变量。</span><br><span class="line">&#x3D;cut</span><br><span class="line"></span><br><span class="line">close GENE_LIST;</span><br><span class="line">close NONMODEL;</span><br><span class="line">close ALTERNATIVE;</span><br></pre></td></tr></table></figure>

<p><u><font color=blue>按照前文 Science 杂志那个反刍动物的研究论文资料，我这里的 branch model 实际上找到的是指定分支上的快速进化基因 Rapidly evolved genes。如果要做正向选择分析，要使用 branch-site mode，即参考下文的PAML四大模型之第三个模型，分别设置ModelA：“model = 2, NSsites = 2, fix_omega = 0, omega = 1.5” 和 Model A Null：“model = 2, NSsites = 2, fix_omega = 1, omega = 1”。</font></u></p>
<p>附<a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?m=201910">chenlianfu</a>关于正向选择分析步骤的总结：</p>
<p>(1) 收集该基因在多个物种中的 CDS 和 protein 序列（可以使用 orthoMCL 分析结果）。<br>(2) 对该基因的蛋白序列进行多序列比对，再根据 CDS 序列转换成 Codon 序列比对结果。这个步骤需要自己编写一些程序来实现。<br>(3) 根据 Codon 序列的比对结果构建系统发育树。推荐使用 RAxML 软件使用 ML 算法对所有的单拷贝同源基因进行物种树构建。然后，输入物种树信息、Codon 多序列比对信息，使用 codeml 进行 omega 计算和选择压模型检验。<br>(4) 快速过滤非正选择基因：设置 runmode = -2 运行 codeml 命令对两两序列比较，使用 ML 方法计算 dNdS，若存在 omega 值大于1，则认为该基因可能属于正选择基因，进行后续分析。<br>(5) 正选择基因的初步鉴定方法：设置 model = 0、NSsites = 1 2 运行 codeml 命令（site models）分别对正选择模型(M2a)和近中性进化模型(M1a)进行检测，若两种模型的似然值相差较大，通过自由度为 2 的卡方检验，可以确定该基因是否为正选择基因。从程序结果中（例如：lnL(ntime: 11 np: 16): -870.867266 +0.000000）可以找到 M2a 模型的似然值 l1 和 M1a 模型的似然值 l0，再计算 2Δl = 2(l1-l0)，通过命令“chi2 2 2Δl”根据卡方检验算出 p 值。此外，也可以设置 NSsites = 7 8 进行计算，则表示分别对 M8(beta &amp; omega) 和 M7(beta) 模型进行检测，和前者类似，若两种模型的似然值相差较大，通过自由度为 2 的卡方检验，可以确定该基因是否为正选择基因。根据 PAML 作者所说，M2a 和 M1a 的比较，比 M7-M8 的比较更严格。即若想得到更多的正选择基因，可以使用 M7_VS_M8 分析。此外，根据 M2 和 M8 模型的 BEB（Bayes empirical Bayes）分析结果，还可以得到在多序列比对结果中第一条序列上的正选择位点。根据 PAM L作者所说，这种 site models 之间的比较，能用于检测是否在序列上不同的位点具有不同的 omega 值，而不是用于正选择检测（We suggest that The M0-M3 comparison should be used as a test of variable ω among sites rather than a test of positive selection）。<br>(6) <font color=blue>指定分化枝上的正选择基因鉴定方法：经过上一步初步鉴定后，设置 model = 2、NSsites = 2、fix_omega = 0、omega = 2.0 运行codeml命令（branch-site model A）；再设置 model = 2、NSsites = 2、fix_omega = 1、omega = 1 运行 codeml 命令（modified branch-site model A / null model）。进行这两种模型分析时，要求输入的树文件中对目标分化枝进行标注。对这两种模型进行 LRT 分析，计算 2Δl = 2(l1-l0)，注意是前者的似然值减后者（null model）的似然值；再使用自由度为 1 的卡方检验，通过命令“chi2 1 2Δl”计算出的值再除以 2，即得到 p 值。</font><br>(7) <font color=blue>目标分化枝上的快速进化基因（Rapidly evolving gene）鉴定方法：设置 model = 0、NSsites = 0 运行 codeml 命令，再设置 model = 2、NSsites = 0 运行 codeml 命令（这时要求输入的树文件中对目标分化枝进行标注）</font>。然后比较两种运行模式下的似然值，通过自由度为 1 的卡方检验，可以鉴定该基因在目标分化枝的 omega 值和背景差异显著。</p>
<h3 id="一些测试与思考"><a href="#一些测试与思考" class="headerlink" title="一些测试与思考"></a>一些测试与思考</h3><p>对于一个基因或者固定数目的少数基因，我按照一个参考流程 <a target="_blank" rel="noopener" href="https://figshare.com/articles/journal_contribution/worked_example_pdf/3839583">https://figshare.com/articles/journal_contribution/worked_example_pdf/3839583</a>，即可计算出 omega 值。</p>
<p>该 demo 用的是 6 个物种的单拷贝 orthologous，有 3000 多个基因。而我的实验里，想用 16 个物种运行单拷贝基因鉴定，只有 322 个基因，这样范围太小，怎么改善？</p>
<p>对于无根树文件而言，如果要指定运算某个分支的进化速率，需要在那个分支上加一个“#1”。这里的意思是该物种有一个进化速率 omega，而其他物种有统一的 omega。在另一个资料中，我发现还可以在tree文件中给指定的分支设定“$1”，比如<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1Qx411m7Bn?from=search&seid=9774626637377853773&spm_id_from=333.337.0.0">作者</a>用到四种海洋哺乳动物，四种陆地动物的 myoglobin 基因做的正向选择分析，把四种海洋动物标记为“$1”，这样运行结果就是指这四个鲸鱼具有统一的 omega。</p>
<p><img src="https://i.imgur.com/feqs2ew.png"></p>
<p>比对的基因序列文件中，物种数目少于tree文件中的物种数目，似乎也可以运行得到结果？（有待确认？如果真的是这样，那我如果对若干个orthologous基因的正向选择做计算，只需要一个tree文件就可以，对于个别基因而言，确实个别物种的orthologous不会影响结果。）2021.10.08 做了一个测试，如果tree文件里面有8个物种，但是phylip格式的序列文件里面只有7个物种，codeml程序会报错的。</p>
<p>Model=0; NSsites= 7 8; 这样结果里面确实有两个 1nL 值，可以用来计算 P 值，但是没有特定 branch 的 ω，而我需要的是特定分支的 ω。这个模式确实无法实现我的目的，因为 Model=0 就表示所有分支的 Omega 都是固定的。</p>
<p>对于 codeml 的配置文件而言，里面的 model 可以设置为 0、1 或 2，这三个参数有什么区别？这个参数主要设定不同分支的ω是否一致。<font color=blue>当model=0，NSsites=0，那么表示假设所有分支、所有氨基酸位点的ω都是一致的。当 model=2, NSsites=0，即假设前景支与背景支的ω不同，出来的结果特定分支ω跟其他分支ω不同，此时其他分支的ω一致。对这两个不同的假设来做卡方检验，即可推算出前者为假的可能性（即 P 值），当 P &lt; 0.05， 可信度高。</font></p>
<p><img src="https://i.imgur.com/h89xpWB.png" alt="Model=0，假设所有分支ω一致"></p>
<p><img src="https://i.imgur.com/64Ec8kB.png" alt="Model=2，假设前景支与背景支的ω不同"></p>
<p><img src="https://i.imgur.com/R4qryV2.png" alt="对上述两个假设做卡方检验"></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">...</span><br><span class="line">Model 7: beta (10 categories)</span><br><span class="line"></span><br><span class="line">TREE #  1:  (((8, 7), 5), (4, ((1, 2), 3)), 6);   MP score: 160</span><br><span class="line">lnL(ntime: 13  np: 16):  -3062.689720      +0.000000</span><br><span class="line">   9..10   10..11   11..8    11..7    10..5     9..12   12..4    12..13   13..14   14..1    14..2    13..3     9..6  </span><br><span class="line"> 0.007984 0.037210 0.329993 0.012402 0.023400 0.007531 0.031297 0.019721 0.020914 0.003075 0.005174 0.017900 0.027104 0.522045 0.005000 0.019948</span><br><span class="line"></span><br><span class="line">Note: Branch length is defined as number of nucleotide substitutions per codon (not per neucleotide site).</span><br><span class="line">...</span><br><span class="line">Model 8: beta&amp;w&gt;1 (11 categories)</span><br><span class="line"></span><br><span class="line">TREE #  1:  (((8, 7), 5), (4, ((1, 2), 3)), 6);   MP score: 160</span><br><span class="line">lnL(ntime: 13  np: 18):  -3041.864857      +0.000000</span><br><span class="line">   9..10   10..11   11..8    11..7    10..5     9..12   12..4    12..13   13..14   14..1    14..2    13..3     9..6  </span><br><span class="line"> 0.008053 0.038859 0.392264 0.011943 0.023468 0.008401 0.030708 0.020215 0.021091 0.002819 0.005071 0.017885 0.027144 0.504233 0.894161 0.007220 0.044307 4.278923</span><br><span class="line"></span><br><span class="line">Note: Branch length is defined as number of nucleotide substitutions per codon (not per neucleotide site).</span><br><span class="line"></span><br><span class="line">tree length &#x3D;   0.60792</span><br><span class="line"></span><br><span class="line">(((8: 0.392264, 7: 0.011943): 0.038859, 5: 0.023468): 0.008053, (4: 0.030708, ((1: 0.002819, 2: 0.005071): 0.021091, 3: 0.017885): 0.020215): 0.008401, 6: 0.027144);</span><br><span class="line"></span><br><span class="line">(((Vicia_faba: 0.392264, Medicago_truncatula: 0.011943): 0.038859, Lotus_japonicus: 0.023468): 0.008053, (Millettia_pinnata: 0.030708, ((Vigna_angularis: 0.002819, Vigna_radiata: 0.005071): 0.021091, Glycine_max: 0.017885): 0.020215): 0.008401, Ammopiptanthus_mongolicus: 0.027144);</span><br><span class="line"></span><br><span class="line">Detailed output identifying parameters</span><br><span class="line"></span><br><span class="line">kappa (ts&#x2F;tv) &#x3D;  0.50423</span><br><span class="line"></span><br><span class="line">Parameters in M8 (beta&amp;w&gt;1):</span><br><span class="line">  p0 &#x3D;   0.89416  p &#x3D;   0.00722 q &#x3D;   0.04431</span><br><span class="line"> (p1 &#x3D;   0.10584) w &#x3D;   4.27892</span><br><span class="line"> ...</span><br></pre></td></tr></table></figure>

<p>下面的方法（来自<a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3036">chenlianfu.com</a>）可以检测指定分支的 ω, 并且计算 P 值。我查阅了 PAML 的文档，确实是这么计算 P 值。</p>
<p><font color=blue>指定分化枝上的正选择基因鉴定方法：经过上一步初步鉴定后，设置 <code>model = 2、NSsites = 2、fix_omega = 0、omega = 2.0</code> 运行 codeml 命令（branch-site model A）；再设置 <code>model = 2、NSsites = 2、fix_omega = 1、omega = 1</code> 运行 codeml 命令（modified branch-site model A / null model）。进行这两种模型分析时，要求输入的树文件中对目标分化枝进行标注。对这两种模型进行LRT分析，计算 2Δl = 2(l1-l0)，注意是前者（branch-site model A）的似然值减后者（null model）的似然值；再使用自由度为1的卡方检验，通过命令“chi2 1 2Δl”计算出的值再除以2，即得到p值。</font></p>
<p>20211011 备注：上面卡方检验直接 <code>chi2 2 2Δl</code> 即可？不需要除以二吧？—— 查阅文档之后看到，要除以二。</p>
<p><img src="https://i.imgur.com/TV6Sr9x.jpeg"></p>
<p>上面截图中调用 <code>chi2 2 2.71</code> 命令，得到 “df =  2  prob = 0.257947294 = 2.579e-01”，再除以二就是 P 值。</p>
<p>其实也可以用 R 来做计算，在R界面输入命令 <code>pchisq(2.71, 2, lower.tail = FALSE)</code> 得到结果：“0.2579473”。跟上面调用 chi2 程序的结果是一致的。</p>
<p>卡方检验是一种用途很广的计数资料的假设检验方法，由卡尔·皮尔逊提出。它属于非参数检验的范畴，主要是比较两个及两个以上样本率( 构成比）以及两个分类变量的关联性分析。其根本思想就是在于比较理论频数和实际频数的吻合程度或拟合优度问题。比如<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/42803826">我们想知道不吃晚饭对体重下降有没有影响</a>，即探讨吃晚饭与体重的关系，将一个人群分为“吃晚饭组”与“不吃晚饭组”</p>
<h3 id="学习笔记：怎么通过-chi2-计算-P-值？"><a href="#学习笔记：怎么通过-chi2-计算-P-值？" class="headerlink" title="学习笔记：怎么通过 chi2 计算 P 值？"></a>学习笔记：怎么通过 chi2 计算 P 值？</h3><p>一开始接触 codeml 的时候，并不清楚怎么计算 P 值，经过阅读一些相关材料，获得 P 值计算方法，大致记录如下。</p>
<p>上面的 codeml 运行结果文件有这么一行：lnL(ntime: 13  np: 16):  -3122.716855      +0.000000</p>
<p>这个“-3122.716855”就有用，我们修改 ctl 文件的参数之后，再次运行 codeml 程序，将生成一个新的 lnL 值。依据前后两次 lnL 值，用下文提到的公式，就可以计算 P 值。</p>
<p><a target="_blank" rel="noopener" href="http://blog.sina.com.cn/s/blog_65ba09d90102x4um.html">http://blog.sina.com.cn/s/blog_65ba09d90102x4um.html</a> 上回讲到了用codeml计算选择位点的时候，同时选择M7和M8模型，这两个模型的差异进近在于 M8(beta &amp; ω) 模型比 M7 （beta）多了一个参数，所以可以依靠比较这两个模型来进行LRT。公式是这个：LRT = 2dl = abs(2 X (l1-l0)) （abs=绝对值）。Step1: 打开之前用codeml生成的 mcl 文件，找到 Model7, 和 Model8。两个lnL值想减，取绝对值，乘以2，这里是：2|-10302.87869-（-10332.64273）| = 59.52 (小数点后影响不大) Step2：打开PAML程序包中的 Chi2 程序, 分别输入自由度和刚才得到的数，自由度直接取 2 就好（这个是杨子恒自己说的，在PAML的文档里有写），后面写上Step1算出来的数字，得到的 p 小于 0.05 则结果比较可靠，这里的 p 非常小。</p>
<p><a target="_blank" rel="noopener" href="https://bcb.unl.edu/yyin/codeml.html">https://bcb.unl.edu/yyin/codeml.html</a> I write some simple ways to test likelihood ratio test on variable codon rates. These two models assume codon rates varied among all branchs. (1). M1a vs. M2a: M1a(model=0, NSsites=1), M2a(model=0, NSsites=2), the df = 2, LRT = 2dl = abs(2 X (l1- l0)). (2). M7 vs. M8: M1a(model=0, NSsites=7), M2a(model=0, NSsites=8), the df = 2, LRT = 2dl = abs(2 X (l1-l0)). The chi-square value can be calucated by PROGRAM CHI2.EXE in PAML.  If the p-value &lt;0.05, then we can conclude that some sites are under positive selection.</p>
<h3 id="PAML-分析四大模型"><a href="#PAML-分析四大模型" class="headerlink" title="PAML 分析四大模型"></a>PAML 分析四大模型</h3><p>在paml分析里面，主要涉及四种模型：位点模型（site-model）、支模型（branch model）、支位点模型（branch site model）以及进化支模型。位点模型通常适用于检测某一支系普遍性、广泛性的正选择，这种正选择是由于位点持续改变所引起的，例如，适应多种病原体；支模型主要是检测某一支系是否存在快速进化、选择压力约束以及正选择，但却无法检测到正选择位点；<font color=blue>支位点模型比较准确与稳定，适用于检测某一支系断点性的正选择事件，此结果是由于适应某一时期环境改变所引起的，通常会保留在后代中</font>；进化支模型（clade model），则主要是判别不同物种之间是否存受分化选择压力作用，不局限于正选择，它一次可以标记多个分支进行比较。</p>
<ol>
<li><strong>位点模型</strong></li>
</ol>
<p>M0：假设所有位点具有相同的 dN/dS 值；<br>M1a：假设存在两类位点—保守位点0 &lt; dN/dS &lt; 1，中性进化位点dN/dS = 1，并且估算这两类位点的比率（p0，p1）和ω值（ω0，ω1）；<br>M2a：假设存在三类位点——纯化选择位点 dN/dS &lt; 1，中性进化位点dN/dS = 1与正选择位点dN/dS &gt; 1，并估算三类位点的比率（p0，p1，p2）；<br>M3: 离散模型，假设所有位点的ω值呈离散分布；<br>M7：假设所有位点0 &lt; ω &lt;1且呈现beta分布；<br>M8：在M7模型的基础上，增加一类正选择位点（ω &gt;1）;<br>M8a：与M8类似，只不过是将新增的ω固定为1；</p>
<p>位点模型的 Codeml.ctl 参数设置：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">model &#x3D; 0                 </span><br><span class="line">NSsites &#x3D; 0 1 2 3 7 8</span><br><span class="line">即分别比较如下模型（* 为正选择模型，前提是需要LRTs显著）：</span><br><span class="line">M3（分散比率） vs. M0（负选择）</span><br><span class="line">M2a* （正选择）vs. M1a （选择约束放松）</span><br><span class="line">M8*（正选择） vs. M7 （选择约束放松）</span><br><span class="line">M8* （正选择）vs. M8a（选择约束放松）</span><br></pre></td></tr></table></figure>

<ol start="2">
<li><strong>支模型</strong></li>
</ol>
<p>one rario：假设所有的进化谱系都具有相同的ω值；<br>free ratio：假设所有的支系都具有独立的ω值；<br>two ratio：假设前景支与背景支的ω不同</p>
<p>分支模型的Codeml.ctl参数设置：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">two ratio: model &#x3D; 2, NSsites &#x3D; 0  </span><br><span class="line">one ratio: model &#x3D; 0, NSsites &#x3D; 0</span><br><span class="line">free ratio: model &#x3D; 1, NSsites &#x3D; 0</span><br><span class="line">               </span><br><span class="line">one ratio（负选择） vs. free ratio（自由比率）</span><br><span class="line">one ratio（负选择） vs. two ratio （正选择）</span><br><span class="line">two ratio natural（中性进化） vs. two ratio （正选择）</span><br></pre></td></tr></table></figure>

<ol start="3">
<li><strong>支位点模型</strong></li>
</ol>
<p>假定位点间的ω值是变化的，同时也假定支系间的ω值是变化的。该模型主要用于检测前景支中正选择作用对部分位点的影响。<br>modelA null（零假设）：ω值设定为固定值1<br>modelA（备择假设）：估算其ω值是否大于1<br>背景支与前景支具有相同的位点ω值：<br>K0：前景支与背景支中的位点受到纯化选择0 &lt; ω &lt; 1；<br>K1：前景支与背景支中的位点处于中性进化0 &lt; ω = 1;<br>背景支与前景支具有不同的位点ω值：<br>K2a：前景支处于中性进化，而背景支处于纯化选择；<br>K2b：前景支受到正选择压力（ ω&gt;1 ），而背景支处于中性进化；<br>支位点模型的的Codeml.ctl参数设置：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">ModelA：</span><br><span class="line">model &#x3D; 2, NSsites &#x3D; 2, fix_omega &#x3D; 0, omega &#x3D; 1.5</span><br><span class="line">ModelAnull：</span><br><span class="line">model &#x3D; 2, NSsites &#x3D; 2, fix_omega &#x3D; 1, omega &#x3D; 1</span><br><span class="line">即比较如下模型（* 为正选择模型，前提是需要LRTs显著）：</span><br><span class="line">ModelA*（正选择） vs. ModelAnull （中性进化）</span><br></pre></td></tr></table></figure>

<ol start="4">
<li><strong>进化支模型</strong></li>
</ol>
<p>与枝位点模型类型，能同时检测多个进化枝（Clade），但是该模型并没有将背景支的dN/dS值约束在（0,1）。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">M2a_model: model &#x3D; 0 NSsites &#x3D; 22</span><br><span class="line">Clade Model C: model &#x3D; 3 NSsites &#x3D; 2</span><br><span class="line">Clade Model D: model &#x3D; 3 NSsites &#x3D; 3</span><br><span class="line">Clade Model C vs. M2a_model</span><br></pre></td></tr></table></figure>

<h3 id="怎么区分-unrooted-tree-与-rooted-tree"><a href="#怎么区分-unrooted-tree-与-rooted-tree" class="headerlink" title="怎么区分 unrooted tree 与 rooted tree"></a>怎么区分 unrooted tree 与 rooted tree</h3><p>有时通过构建进化树的软件获得进化树后，软件不会对进化树是有根树，还是无根树做说明。在不做额外编写脚本的基础上，其实可以直接从newick文件的内容格式上区分有根树和无根树］</p>
<p>假设有下面三种形式的进化树：</p>
<p>[1] (a, b, c);      无根树</p>
<p>[2] (a, b, c):1;   有根树</p>
<p>[3] ((a,b),c);      有根树</p>
<p>除第一种形式的进化树外，其它两种形式的进化树都是有根树。R 语言对有根树、无根树的判断是：</p>
<p>如果最外层大括号内只有两个分枝，即为有根树，如 [3]；</p>
<p>如果最外层大括号内有三个或以上分枝，一般为无根树，如 [1]；</p>
<p>但是，如果大括号外存在枝长参数，如 [2] 中的 :1，这种情况认为三个分枝以一定的枝长连接到根上，为有根树。</p>
<p>我想通过上述三个例子，可以先看最外层是否有枝长参数，再数一下最外层括号下有几个分枝，即可简单地、直接地判断目前绝大多数获得的newick格式的进化树是有根树，还是无根树。</p>
<h3 id="实践"><a href="#实践" class="headerlink" title="实践"></a>实践</h3><p>如何收集基因？如果我想找出一个物种与相邻的若干物种相比较，有哪些基因出现了正向选择。首先，用 OrthoFinder 软件可以找出 single copy genes。这里要注意，初始的物种数目越多，single copy genes 越少，后续用于分析得到的结果越少，我看文献中有用到 7 个物种，摘录于此：“To detect genes evolving under positive selection in jujube, we used gene clustering of the seven Rosales species to identify 1,420 single-copy orthologous genes and used the optimized branch-site mode<sup>l63</sup> to identify 254 genes that are positively selected in <em>Z. jujuba</em>.” 另外，我觉得可以用 jcvi 流程，他会基于基因组位置和基因序列两个信息来做 orthologous 分析，以感兴趣物种为参考，将其他若干物种的基因集分别与之比较，就可以得这一群物种的 orthologous genes（反刍动物的进化分析其实就是类似这个策略，只不过他们用的是lastal比对软件），然后要思考的一点是，是不是必须所有物种都出现的ortholog gene，才用于后续分析。</p>
<p>2021.10.08</p>
<ol>
<li>选择7个物种，准备好CDS序列，然后翻译成pep序列，将用于orthofinder流程。注意，同时考虑jcvi流程需要的gff文件和pep文件。（别用花生。感觉序列质量不好。）</li>
<li>运行orthofinder流程，预计找到1000多个single copy genes</li>
<li>在AM与其他物种依次运行jcvi，然后整合结果，预计这种方式可以找到几千个orthologous genes</li>
<li>分别运行codeml流程，计算Ka/Ks值。并计算P值。看最终有多少个基因会出现正向选择。</li>
</ol>
<h3 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h3><p><a target="_blank" rel="noopener" href="http://yuzhenpeng.github.io/2019/12/24/paml/">于振朋: Comparative Genomics and Adaptive Evolution</a><br><a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1Qx411m7Bn?from=search&seid=9774626637377853773&spm_id_from=333.337.0.0">B 站 PAML 视频教程：利用 codeml 计算 dN/dS</a><br><a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3001">Chenlianfu: 使用 PAML 软件利用核酸序列进行系统发育参数分析</a><br><a target="_blank" rel="noopener" href="http://www.chenlianfu.com/?p=3036">Chenlianfu: 使用 PAML 软件利用密码子序列进行正选择分析</a><br><a target="_blank" rel="noopener" href="https://figshare.com/articles/journal_contribution/worked_example_pdf/3839583">A worked example of estimating ω and testing for adaptive evolution in six parasite species</a><br><a target="_blank" rel="noopener" href="http://wap.sciencenet.cn/blog-460481-1163040.html?mobile=1">高芳銮：选择压力分析之 EasyCodeML 完整篇（By Raindy)</a><br><a target="_blank" rel="noopener" href="https://static-content.springer.com/esm/art%3A10.1007%2Fs00606-019-01578-2/MediaObjects/606_2019_1578_MOESM1_ESM.pdf">Supplementary File 1 of “<u><em>Analyses of mitochondrial genomes of the genus Ammopiptanthus provide new insights into the evolution of legume plants</em></u>“</a></p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a><a class="next" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/mcscanx-%E6%B5%8B%E8%AF%95/">MCScanX 测试</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Bioinfo | LTR identification and analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>