<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>Bioinfo | Gene Ontology - Enrichment Analysis | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">Bioinfo | Gene Ontology - Enrichment Analysis</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">Bioinfo | Gene Ontology - Enrichment Analysis</h1><div class="post-meta">2021-06-29<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>下面是我做的大豆GO数据库（version Gmax_508_Wm82_a4.v1）构建与富集分析。数据文件在：<a target="_blank" rel="noopener" href="https://github.com/bioinfx/Glycine_max_Gene_Ontology">https://github.com/bioinfx/Glycine_max_Gene_Ontology</a>。</p>
<p>经过前期数据比对和整理（比如依据 swissprot、blast2GO，或 GOMAP），得到下面“基因——GO号——Ontology 类别——证据”这四列格式的文件 GMA_GO_TERM.txt，就可以在R程序里面用于 GO 数据库的构建。</p>
<h3 id="数据库源文件准备"><a href="#数据库源文件准备" class="headerlink" title="数据库源文件准备"></a>数据库源文件准备</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">$ zcat GMA_GO_TERM.txt.gz | head</span><br><span class="line">GID     GO      ONTOLOGY        EVIDENCE</span><br><span class="line">Glyma.01G000100 GO:0006412      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009063      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0009234      P       IEA</span><br><span class="line">Glyma.01G000100 GO:0016021      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0031969      C       IEA</span><br><span class="line">Glyma.01G000100 GO:0016835      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0030976      F       IEA</span><br><span class="line">Glyma.01G000100 GO:0070204      F       IEA</span><br><span class="line">Glyma.01G000137 GO:0016043      P       IEA</span><br></pre></td></tr></table></figure>

<h3 id="数据库导入-R"><a href="#数据库导入-R" class="headerlink" title="数据库导入 R"></a>数据库导入 R</h3><p>构建过程如下，调用 makeOrgPackage 函数构建 GO 数据库。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><span class="line">## Makes an organism package for Glycine max data.frames:</span><br><span class="line"></span><br><span class="line">library(RSQLite)</span><br><span class="line">library(AnnotationForge)</span><br><span class="line"></span><br><span class="line">## Now prepare some data.frames</span><br><span class="line">fGO &lt;- read.table(&quot;..&#x2F;3_final-result&#x2F;GMA_GO_TERM.txt&quot;, sep&#x3D;&quot;\t&quot;, head&#x3D;TRUE)</span><br><span class="line"></span><br><span class="line">## remove duplication in the table</span><br><span class="line">go_df &lt;- fGO[!duplicated(fGO),]</span><br><span class="line"></span><br><span class="line">## only tree coloums are needed: GID, GO, EVIDENCE</span><br><span class="line">go_df &lt;- go_df[,c(1,2,4)]</span><br><span class="line"></span><br><span class="line">## Making Organism packages </span><br><span class="line">makeOrgPackage(go&#x3D;go_df,</span><br><span class="line">    version &#x3D; &quot;0.0.1&quot;,</span><br><span class="line">    maintainer &#x3D; &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">    author &#x3D; &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">    outputDir &#x3D; &quot;.&#x2F;&quot;,</span><br><span class="line">    tax_id &#x3D; &quot;3847&quot;,</span><br><span class="line">    genus &#x3D; &quot;Glycine&quot;,</span><br><span class="line">    species &#x3D; &quot;max&quot;,</span><br><span class="line">    goTable &#x3D; &quot;go&quot;</span><br><span class="line">    )</span><br><span class="line"></span><br><span class="line">## then you can call install.packages based on the return value</span><br><span class="line">install.packages(&quot;.&#x2F;org.Gmax.eg.db&quot;, repos &#x3D; NULL, type&#x3D;&quot;source&quot;)</span><br><span class="line">library(org.Gmax.eg.db)</span><br><span class="line">org &lt;- org.Gmax.eg.db</span><br><span class="line"></span><br><span class="line">## A demo for using the database</span><br><span class="line"># BiocManager::install(&quot;Rgraphviz&quot;)</span><br><span class="line"># BiocManager::install(&quot;topGO&quot;)</span><br><span class="line">library()</span><br><span class="line">geneTable &#x3D; read.table(&quot;.&#x2F;Gmax_ABA_exp.xls&quot;, head&#x3D;T)</span><br><span class="line">geneList&#x3D;geneTable[,5]</span><br><span class="line">geneList&#x3D;geneList[!duplicated(geneList)]</span><br><span class="line"># Biology process</span><br><span class="line">ego_up &lt;-enrichGO(gene &#x3D; geneList, OrgDb &#x3D; org, keyType &#x3D; &quot;GID&quot;, ont &#x3D; &quot;BP&quot;)</span><br><span class="line">pdf(&quot;up_GO_BP_barplot.pdf&quot;, height&#x3D;10, width&#x3D;15)</span><br><span class="line">barplot(ego_up, drop&#x3D;TRUE, showCategory&#x3D;12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;up_GO_BP_dotPlot.pdf&quot;, height&#x3D;10, width&#x3D;15)</span><br><span class="line">dotplot(ego_up)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;up_GO_BP_tree.pdf&quot;)</span><br><span class="line">plotGOgraph(ego_up)</span><br><span class="line">dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_up@result ), file&#x3D;&quot;up_GO_BP.csv&quot;)</span><br></pre></td></tr></table></figure>

<h1 id="数据库开启与富集分析"><a href="#数据库开启与富集分析" class="headerlink" title="数据库开启与富集分析"></a>数据库开启与富集分析</h1><p>下面代码示范如何引用上构建的数据库，并且开展 GO 分析：BP、MF 和 CC 三个类别。<br>图片可输出为柱状图 barplot 和点状图 dotplot，</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br></pre></td><td class="code"><pre><span class="line">library(org.Gmax.eg.db)</span><br><span class="line">org &lt;- org.Gmax.eg.db</span><br><span class="line"></span><br><span class="line">## A demo for using the database</span><br><span class="line"># BiocManager::install(&quot;Rgraphviz&quot;)</span><br><span class="line"># BiocManager::install(&quot;topGO&quot;)</span><br><span class="line">library(topGO)</span><br><span class="line">library(clusterProfiler)</span><br><span class="line">geneTable &#x3D; read.table(&quot;.&#x2F;gene_table.txt&quot;, head&#x3D;T, sep&#x3D;&quot;\t&quot;)</span><br><span class="line">#geneTable &#x3D; geneTable[!apply(geneTable &#x3D;&#x3D; &quot;&quot;, 1, all),]</span><br><span class="line">geneList&#x3D;geneTable[,5]</span><br><span class="line"></span><br><span class="line">geneList&#x3D;geneList[!duplicated(geneList)]</span><br><span class="line"># Biology process</span><br><span class="line">b&#x3D;geneList</span><br><span class="line"></span><br><span class="line"># biological progress</span><br><span class="line">ego_up &lt;-enrichGO(</span><br><span class="line">                  gene &#x3D; b, </span><br><span class="line">                  OrgDb &#x3D; org, </span><br><span class="line">                  keyType &#x3D; &quot;GID&quot;, </span><br><span class="line">                  ont &#x3D; &quot;BP&quot;</span><br><span class="line">                  )</span><br><span class="line">pdf(&quot;test_GO_BP_barplot.pdf&quot;, height&#x3D;10, width&#x3D;15)</span><br><span class="line">barplot(ego_up, drop&#x3D;TRUE, showCategory&#x3D;12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;test_GO_BP_dotPlot.pdf&quot;, height&#x3D;10, width&#x3D;15)</span><br><span class="line">dotplot(ego_up)</span><br><span class="line">dev.off()</span><br><span class="line">#pdf(&quot;up_GO_BP_tree.pdf&quot;)</span><br><span class="line">#plotGOgraph(ego_up)</span><br><span class="line">#dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_up@result ), file&#x3D;&quot;test_GO_BP.csv&quot;)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># molecular function</span><br><span class="line">ego_down &lt;-enrichGO(</span><br><span class="line">                    gene &#x3D; b,</span><br><span class="line">                    OrgDb &#x3D; org,</span><br><span class="line">                    keyType &#x3D; &quot;GID&quot;,</span><br><span class="line">                    ont &#x3D; &quot;MF&quot;</span><br><span class="line">                    )</span><br><span class="line">pdf(&quot;test_GO_MF_barplot.pdf&quot;, height&#x3D;10, width&#x3D;15)</span><br><span class="line">barplot(ego_down, drop&#x3D;TRUE, showCategory&#x3D;12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;test_GO_MF_dotPlot.pdf&quot;, height&#x3D;10, width&#x3D;15)</span><br><span class="line">dotplot(ego_down)</span><br><span class="line">dev.off()</span><br><span class="line">#pdf(&quot;test_GO_MF_tree.pdf&quot;)</span><br><span class="line">#plotGOgraph(ego_down)</span><br><span class="line">#dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_down@result), file&#x3D;&quot;test_GO_MF.csv&quot;)</span><br><span class="line"></span><br><span class="line"># cellular component</span><br><span class="line">ego_down &lt;-enrichGO(</span><br><span class="line">                    gene &#x3D; b,</span><br><span class="line">                    OrgDb &#x3D; org,</span><br><span class="line">                    keyType &#x3D; &quot;GID&quot;,</span><br><span class="line">                    ont &#x3D; &quot;CC&quot;</span><br><span class="line">                    )</span><br><span class="line">pdf(&quot;test_GO_CC_barplot.pdf&quot;, height&#x3D;10, width&#x3D;15)</span><br><span class="line">barplot(ego_down, drop&#x3D;TRUE, showCategory&#x3D;12)</span><br><span class="line">dev.off()</span><br><span class="line">pdf(&quot;test_GO_CC_dotPlot.pdf&quot;, height&#x3D;10, width&#x3D;15)</span><br><span class="line">dotplot(ego_down)</span><br><span class="line">dev.off()</span><br><span class="line">#pdf(&quot;test_GO_CC_tree.pdf&quot;)</span><br><span class="line">#plotGOgraph(ego_down)</span><br><span class="line">#dev.off()</span><br><span class="line">write.csv(as.data.frame( ego_down@result), file&#x3D;&quot;test_GO_CC.csv&quot;)</span><br></pre></td></tr></table></figure>

<p>有时候 GO 项目的名称过长，基于默认作图代码出来的图片会出现文字将图形区域挤压得无法显示。</p>
<p><img src="https://i.imgur.com/FlARCCO.jpeg"></p>
<p>有两种优化方案，一是将长的文字换行；二是只截取一部分文字显示。</p>
<h3 id="一：过长的文字换行显示，"><a href="#一：过长的文字换行显示，" class="headerlink" title="一：过长的文字换行显示，"></a>一：过长的文字换行显示，</h3><p>参考资料：<a target="_blank" rel="noopener" href="https://www.cnblogs.com/songbiao/articles/12717670.html">songbiao: GO term name太长的处理</a><br>在使用clusterProfiler进行绘制GO dotplot时，会遇到 GO term 的 name 过长，压缩了图形区域的情况。<br>因为 clusterProfiler 作图是基于 ggplot2 的，所以这个处理办法也适用于其他 ggplot2 作图遇到此种情况。<br>解决文本过长的办法就是，用stringr包的str_wrap来完成文本自动换行。<br>效果如下图所示，其实还是不太完美，需要后期手工调整才美观。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)  # 需要开启 ggplot2 才可以使用 scale_y_discrete 函数</span><br><span class="line">library(stringr)</span><br><span class="line">dotplot(mf, showCategory&#x3D;30) + scale_y_discrete(labels&#x3D;function(mf) str_wrap(mf, width&#x3D;40))</span><br></pre></td></tr></table></figure>

<p><img src="https://i.imgur.com/JRkomgm.jpeg"></p>
<h3 id="二：只截取一部分文字显示"><a href="#二：只截取一部分文字显示" class="headerlink" title="二：只截取一部分文字显示"></a>二：只截取一部分文字显示</h3><p>参考资料：<a target="_blank" rel="noopener" href="https://r.bio-spring.info/2018/11/26/enrichplot-fancy-label/">在enrichplot中使用更好的Label</a><br>该方法是定义一个截取短 label 的函数, 将其传递给 scale_y_discrete() 即可. 在输出 ggplot 对象时做修改。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line">#&#39; Truncate string vector of ggplot axis label</span><br><span class="line">#&#39;</span><br><span class="line">#&#39; @param label    a ordered string vector</span><br><span class="line">#&#39; @param maxLen   max length of character (nchar) to show in label</span><br><span class="line">#&#39; @param maxWord  max count of words allowed to show in label</span><br><span class="line">#&#39; @param pattern  Word separater</span><br><span class="line">#&#39; @param dot      If true, three dots will added to truncated label</span><br><span class="line">#&#39;</span><br><span class="line">#&#39; @return a vector of truncated strings</span><br><span class="line">#&#39; @export</span><br><span class="line">#&#39;</span><br><span class="line">#&#39; @examples</span><br><span class="line"># 默认最多显示 50 个字符, 5 个单词.</span><br><span class="line">short_label &lt;- function(label, maxLen &#x3D; 50, maxWord &#x3D; 5, pattern &#x3D; &quot; &quot;, dot &#x3D; TRUE)&#123;</span><br><span class="line">  l &lt;- strsplit(label, pattern)</span><br><span class="line">  short_label &lt;- vector(&quot;character&quot;,length(l))</span><br><span class="line">  </span><br><span class="line">  for (i in seq_along(l))&#123;</span><br><span class="line">    truncated &lt;- FALSE</span><br><span class="line">    s &lt;- l[[i]]</span><br><span class="line">    if (length(s) &gt; maxWord)&#123;</span><br><span class="line">      ll &lt;- paste(s[1:maxWord], collapse &#x3D; &quot; &quot;)</span><br><span class="line">      truncated &lt;- TRUE</span><br><span class="line">    &#125;</span><br><span class="line">    else&#123;</span><br><span class="line">      ll &lt;- paste(s, collapse &#x3D; &quot; &quot;)</span><br><span class="line">    &#125;</span><br><span class="line">    </span><br><span class="line">    if (nchar(ll) &gt; maxLen)&#123;</span><br><span class="line">      ll &lt;- substr(ll, 1, maxLen)</span><br><span class="line">      truncated &lt;- TRUE</span><br><span class="line">    &#125;</span><br><span class="line">    </span><br><span class="line">    if (dot &amp; truncated) ll &lt;- paste(ll, &quot;...&quot;,sep &#x3D; &quot; &quot;)</span><br><span class="line">    </span><br><span class="line">    short_label[[i]] &lt;- ll</span><br><span class="line">  &#125;</span><br><span class="line">  attr(short_label, &quot;pos&quot;) &lt;- attr(label,&quot;pos&quot;)</span><br><span class="line">  return(short_label)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>画图代码：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)</span><br><span class="line">dotplot(mf) + scale_y_discrete(label&#x3D;short_label)</span><br></pre></td></tr></table></figure>

<p><img src="https://i.imgur.com/pALA2H9.jpeg"></p>
<p>注意上面的图，其实很多项目都是高度近似的，怎么优化？对于模式物种而言，CytoScape的插件ClueGO有一个选项可以实现，就是设置Ontology的级别。因为Ontology的结构就是树状，从根部到.</p>
<p>构建GO数据库的过程中，如果成功会输出下面的信息。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(go_df)</span><br><span class="line">           GID         GO EVIDENCE</span><br><span class="line">1 Chr01.g00016 GO:0003676      IEA</span><br><span class="line">2 Chr01.g00016 GO:0003964      IEA</span><br><span class="line">3 Chr01.g00016 GO:0004190      IEA</span><br><span class="line">4 Chr01.g00016 GO:0004519      IEA</span><br><span class="line">5 Chr01.g00016 GO:0006310      IEA</span><br><span class="line">6 Chr01.g00016 GO:0015074      IEA</span><br><span class="line">&gt; makeOrgPackage(go&#x3D;go_df,</span><br><span class="line">+                version &#x3D; &quot;0.0.1&quot;,</span><br><span class="line">+                maintainer &#x3D; &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">+                author &#x3D; &quot;FENG Lei &lt;fengleiluck@gmail.com&gt;&quot;,</span><br><span class="line">+                outputDir &#x3D; &quot;.&quot;,</span><br><span class="line">+                tax_id &#x3D; &quot;126911&quot;,</span><br><span class="line">+                genus &#x3D; &quot;Ammopiptanthus&quot;,</span><br><span class="line">+                species &#x3D; &quot;mongolicus&quot;,</span><br><span class="line">+                goTable &#x3D; &quot;go&quot;</span><br><span class="line">+ )</span><br><span class="line">Populating genes table:</span><br><span class="line">genes table filled</span><br><span class="line">Populating go table:</span><br><span class="line">go table filled</span><br><span class="line">table metadata filled</span><br><span class="line">&#39;select()&#39; returned many:1 mapping between keys and columns</span><br><span class="line">Dropping GO IDs that are too new for the current GO.db</span><br><span class="line">Populating go table:</span><br><span class="line">go table filled</span><br><span class="line">Populating go_bp table:</span><br><span class="line">go_bp table filled</span><br><span class="line">Populating go_cc table:</span><br><span class="line">go_cc table filled</span><br><span class="line">Populating go_mf table:</span><br><span class="line">go_mf table filled</span><br><span class="line">&#39;select()&#39; returned many:1 mapping between keys and columns</span><br><span class="line">Populating go_bp_all table:</span><br><span class="line">go_bp_all table filled</span><br><span class="line">Populating go_cc_all table:</span><br><span class="line">go_cc_all table filled</span><br><span class="line">Populating go_mf_all table:</span><br><span class="line">go_mf_all table filled</span><br><span class="line">Populating go_all table:</span><br><span class="line">go_all table filled</span><br><span class="line">Creating package in .&#x2F;org.Amongolicus.eg.db </span><br><span class="line">Now deleting temporary database file</span><br><span class="line">[1] &quot;.&#x2F;org.Amongolicus.eg.db&quot;</span><br><span class="line">There were 50 or more warnings (use warnings() to see the first 50)</span><br></pre></td></tr></table></figure></div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Bioinfo | LTR identification and analysis</a><a class="next" href="/2021/06/17/Linux-Gene-Ontology-GOMAP/">Bioinfo | Gene Ontology - GOMAP</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Bioinfo | LTR identification and analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/17/Linux-Gene-Ontology-GOMAP/">Bioinfo | Gene Ontology - GOMAP</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Linux | 认识 PKG_CONFIG_PATH</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">jcvi 安装与使用，报错以及解决办法</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">Bioinfo | Installation of OrthoMCL-pipeline</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Bioinfo | Using Circos for genome visualization</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">Bioinfo | Repeat sequence analysis by RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer)</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>