<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>生物进化之选择压力分析：Ka/Ks比率 | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">生物进化之选择压力分析：Ka/Ks比率</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">生物进化之选择压力分析：Ka/Ks比率</h1><div class="post-meta">2016-06-09<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><h3 id="背景知识"><a href="#背景知识" class="headerlink" title="背景知识"></a><strong>背景知识</strong></h3><p>In <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Genetics" title="Genetics">genetics</a>, the <strong>Ka/Ks ratio</strong> is an indicator of <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Selective_pressure" title="Selective pressure">selective pressure</a> (the force applied by <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Natural_selection" title="Natural selection">natural selection</a>) acting on a protein-coding gene. It is calculated as the ratio of the number of <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Nonsynonymous_substitution" title="Nonsynonymous substitution">nonsynonymous substitutions</a> per non-synonymous site (Ka), in a given period of time, to the number of <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Synonymous_substitution" title="Synonymous substitution">synonymous substitutions</a> per synonymous site (Ks), in the same period. <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Homologous_gene" title="Homologous gene">Homologous genes</a> with a Ka/Ks ratio above 1 are evolving under <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Directional_selection" title="Directional selection">positive selection</a>, meaning that at least some of the <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Mutation" title="Mutation">mutations</a> concerned must be advantageous. If all the mutations are neutral or disadvantageous, the ratio will be in the range 0 to 1. However, if some of the mutations are advantageous and some disadvantageous, the ratio could be less than 1.</p>
<h3 id="方法-软件"><a href="#方法-软件" class="headerlink" title="方法/软件"></a>方法/软件</h3><p>Conclusions one may draw from dN/dS ratios (aka. ~Ka/Ks): <strong><em>Neutral Evolution (drift):</em></strong> dN/dS ratio = 1 implies there has been equal numbers of synonymous (dna substitutions that do not affect the protein sequence) and non-synonymous changes (dna substitutions that do affect the protein sequence) during the time between ancestral to the modern versions of the protein. <strong><em>Positive Selection (adaptive evolution):</em></strong> dN/dS ratio &gt; 1 implies there has been more non-synonymous changes than synonymous changes. There has been evolutionary pressure to escape from the ancestral state - i.e. positive selection pressure. This can occur for example in paralogues that are required to serve a novel function, or in proteins of parasites that need to escape host immune recognition (e.g. changes to avoid MHC-1 binding to evade T-cell attack). <strong><em>Negative Selection (conservation):</em></strong> dN/dS ratio &lt; 1 implies there has been more synonymous changes than non-synonymous changes. There has been evolutionary pressure to conserve the ancestral state - i.e. negative selection pressure. This can occur for example in orthologues that are required to maintain (conserve) some function encoded in the protein sequence, since changes from this state would lead to disruption of function. <strong>Useful Tips:</strong></p>
<ul>
<li>  Algorithms can either run on multiple sequences, or just a pair of sequences. In either case the input sequences used to derive a dN/dS ratio must share ancestry - too divergent and there is a problem with multiple substitutions, too recent and you will not have sufficient enough observed changes to draw conclusions from.</li>
<li>  dN/dS can be used to compare whole proteins or regions within proteins (a sliding dN/dS value across the protein)</li>
<li>  A dN/dS ratio calculated for a whole protein is often an underestimate (lower than it should be) due to the variety of domains that constitute each protein, for instance a alpha-helix structure may always be required in a set of proteins that perform a variety of different functions.</li>
<li>  The only sequence changes considered are substitutions (not duplications, or inversions etc.)</li>
<li>  Significance of a given dN/dS ratio can be assessed using Fishers exact test: <a target="_blank" rel="noopener" href="https://code.google.com/p/kaks-calculator/wiki/FAQ">read this</a></li>
</ul>
<p><strong>Software:</strong> Here are my recommendations for software ordered by how flexible they are:</p>
<ul>
<li>  <a target="_blank" rel="noopener" href="http://www.mathworks.co.uk/help/bioinfo/ref/dnds.html">MATLAB’s Bioinformatics Toolbox:</a> Here you have the greatest variety of alternative algorithms, operating system compatibility, sliding vs. whole protein analysis, API to Genbank, etc (<a target="_blank" rel="noopener" href="http://www.mathworks.co.uk/products/bioinfo/code-examples.html?file=/products/demos/shipping/bioinfo/dndsdemo.html">Here’s</a> a great tutorial for using their dN/dS tool). Just remember MATLAB is not free.</li>
<li>  <a target="_blank" rel="noopener" href="https://code.google.com/p/kaks-calculator/wiki/KaKs_Calculator">KaKs Calculator:</a> If you only care about <strong>whole protein</strong> dN/dS, many options are available with the Ka/Ks calculator - they also compute statistical significance using Fisher’s exact test. I can also provide an R script that generates error bars from the output, just ask.</li>
<li>  <a target="_blank" rel="noopener" href="http://abacus.gene.ucl.ac.uk/software/paml.html">PAML:</a> If you have &gt;2 sequences per protein that you wish to get a dN/dS value from, then many options are available with PAML. This is often used in published papers, but it’s not recommended if you only have a pair of sequences per protein.</li>
<li><ul>
<li><a target="_blank" rel="noopener" href="http://code.google.com/p/kaks-calculator">KaKs_Calculator</a><ul>
<li>  <a target="_blank" rel="noopener" href="http://services.cbu.uib.no/tools/kaks">Free online server tool that calculates KaKs ratios among multiple sequences</a></li>
<li>  <a target="_blank" rel="noopener" href="http://cran.r-project.org/web/packages/seqinr/index.html">SeqinR: A free and open biological sequence analysis package for the R language that includes KaKs calculation</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<h3 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h3><p><a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Ka/Ks/_ratio">https://en.wikipedia.org/wiki/Ka/Ks\_ratio</a><br><a target="_blank" rel="noopener" href="https://www.biostars.org/p/5817/">https://www.biostars.org/p/5817/</a></p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2016/06/14/t%E6%A3%80%E9%AA%8C%E4%B8%AD%EF%BC%8C%E5%8D%95%E5%B0%BE%E5%92%8C%E5%8F%8C%E5%B0%BE%EF%BC%8C%E6%88%90%E5%AF%B9%E3%80%81%E7%AD%89%E6%96%B9%E5%B7%AE%E5%92%8C%E5%BC%82%E6%96%B9%E5%B7%AE%E7%9A%84%E5%B7%AE/">T检验中，单尾和双尾，成对、等方差和异方差的差异</a><a class="next" href="/2016/06/01/%E6%8B%9F%E5%8D%97%E8%8A%A5arabidopsis%E7%9A%84%E5%9F%BA%E5%9B%A0%E8%9B%8B%E7%99%BD%E7%AA%81%E5%8F%98%E4%BD%93%E4%B9%8B%E4%B9%A6%E5%86%99%E8%A7%84%E8%8C%83/">拟南芥Arabidopsis的基因/蛋白/突变体之书写规范</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/17/Linux-Gene-Ontology-GOMAP/">Bioinfo | Gene Ontology - GOMAP</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Linux | 认识 PKG_CONFIG_PATH</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">jcvi 安装与使用，报错以及解决办法</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">Bioinfo | Installation of OrthoMCL-pipeline</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Bioinfo | Using Circos for genome visualization</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">Bioinfo | Repeat sequence analysis by RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer)</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/16/Bioinfo-understanding-sequence-identity/">Bioinfo | 怎么理解序列一致性（Sequence identity）</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>