<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>MCScanX 测试 | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">MCScanX 测试</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">MCScanX 测试</h1><div class="post-meta">2022-04-01<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a><a href="/categories/Linux/">Linux</a></span></div><div class="post-content"><p>原文记录于 2019-07-05，因 2022-04-01 再次测试，遇到问题，再次编辑记录。</p>
<p>安装如果遇到下面的问题：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">\[fenglei@localhost MCScanX\]$ make</span><br><span class="line">g++ struct.cc mcscan.cc read\_data.cc out\_utils.cc dagchainer.cc msa.cc permutation.cc -o MCScanX</span><br><span class="line">msa.cc: In function ‘void msa\_main(const char\*)’:</span><br><span class="line">msa.cc:289:22: error: ‘chdir’ was not declared in this scope</span><br><span class="line">if (chdir(html\_fn)&lt;0)</span><br><span class="line">^</span><br><span class="line">make: \*\*\* \[makefile:2: mcscanx\] Error 1</span><br></pre></td></tr></table></figure>

<p>解决办法：</p>
<p>错误的原因是MCScanX 不支持64位系统, 如果要在 64位上运行, 需要在<code>msa.h</code>, <code>dissect_multiple_alignment.h</code>, and <code>detect_collinear_tandem_arrays.h</code>第一行加上</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">#include &lt;unistd.h&gt;</span><br></pre></td></tr></table></figure>

<p>如果要对两个基因组（以soybean和medicago为例）进行共线性分析，需要收集两个物种的数据编码基因的gff3文件和CDS序列文件，然后进行下面的操作。</p>
<p>首先利用gff3文件，得到每一个基因的起始和终止坐标信息，注意基因有多个transcripts，只需要保留“xxxx.1”的转录本，不同物种的序列用awk等工具灵活处理。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.formats.gff bed --type&#x3D;mRNA --key&#x3D;Name Gmax\_275\_Wm82.a2.v1.gene.gff3.gz -o soybean.bed</span><br><span class="line">python -m jcvi.formats.gff bed --type&#x3D;mRNA --key&#x3D;Name Mtruncatula\_285\_Mt4.0v1.gene.gff3 -o medicago.bed</span><br><span class="line"></span><br><span class="line">cat soybean.bed  awk &#39;BEGIN&#123;OFS&#x3D;&quot;\\t&quot;&#125;&#123;if(length($4) &#x3D;&#x3D; 17 &amp;&amp; $4 ~ &#x2F;\\.1$&#x2F;)&#123;gsub(&#x2F;\\.1$&#x2F;,&quot;&quot;,$4);gsub(&#x2F;Chr&#x2F;,&quot;gm&quot;,$1); print $1,$4,$2,$3&#125;&#125;&#39; &gt; soybean\_gff.tab</span><br><span class="line">cat medicago.bed  awk &#39;BEGIN&#123;OFS&#x3D;&quot;\\t&quot;&#125;&#123;if(length($4) &#x3D;&#x3D; 15 &amp;&amp; $4 ~ &#x2F;\\.1$&#x2F;)&#123;gsub(&#x2F;\\.1&#x2F;,&quot;&quot;,$4);gsub(&#x2F;chr&#x2F;,&quot;mt&quot;,$1); print $1,$4,$2,$3&#125;&#125;&#39; &gt; medicago\_gff.tab</span><br></pre></td></tr></table></figure>

<p>得到gff格式文件如下。 soybean.gff</p>
<p><strong>注意</strong> gff 文件中别包含 contigs 和 scaffolds，最好只包含 chromosomes。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">gm01 Glyma.01G000100 27354 28320</span><br><span class="line">gm01 Glyma.01G000200 58974 67527</span><br><span class="line">gm01 Glyma.01G000300 67769 69968</span><br><span class="line">gm01 Glyma.01G000400 90151 95947</span><br><span class="line">gm01 Glyma.01G000500 90288 91197</span><br><span class="line">gm01 Glyma.01G000600 116093 127845</span><br><span class="line">gm01 Glyma.01G000700 143466 155573</span><br><span class="line">gm01 Glyma.01G000800 157029 157772</span><br><span class="line">gm01 Glyma.01G000900 170533 193342</span><br><span class="line">gm01 Glyma.01G001000 196255 201895</span><br></pre></td></tr></table></figure>

<p>medicago.gff</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">mt1 Medtr1g004930 688 7332</span><br><span class="line">mt1 Medtr1g004940 6523 7366</span><br><span class="line">mt1 Medtr1g004950 14513 15729</span><br><span class="line">mt1 Medtr1g004960 16282 18382</span><br><span class="line">mt1 Medtr1g006530 28972 29214</span><br><span class="line">mt1 Medtr1g004980 31972 32344</span><br><span class="line">mt1 Medtr1g004990 35909 40554</span><br><span class="line">mt1 Medtr1g006490 44429 46280</span><br><span class="line">mt1 Medtr1g006590 46547 50437</span><br><span class="line">mt1 Medtr1g006600 51360 54977</span><br></pre></td></tr></table></figure>

<p>随后对 CDS 序列文件进行整理，保留 primary transcripts 的序列。序列名称跟上面 gff 文件里面要一致。并灵活选用工具将 CDS 序列转换成蛋白序列，或者一开始就用蛋白序列也可以。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">awk &#39;&#123;print $1&#125;&#39; ..&#x2F;jcvi\_190704&#x2F;Gmax\_275\_Wm82.a2.v1.cds.fa &gt; Gmax\_275\_Wm82.a2.v1.cds.fa\_simplified</span><br><span class="line">awk &#39;&#123;print $1&#125;&#39; ..&#x2F;jcvi\_190704&#x2F;Mtruncatula\_285\_Mt4.0v1.cds.fa &gt; Mtruncatula\_285\_Mt4.0v1.cds.fa\_simplified</span><br><span class="line"></span><br><span class="line">fasta\_formatter -t -i Mtruncatula\_285\_Mt4.0v1.cds.fa\_simplified -o Mtruncatula\_285\_Mt4.0v1.cds.tab</span><br><span class="line">fasta\_formatter -t -i Gmax\_275\_Wm82.a2.v1.cds.fa\_simplified -o Gmax\_275\_Wm82.a2.v1.cds.tab</span><br><span class="line"></span><br><span class="line">cat Gmax\_275\_Wm82.a2.v1.cds.tab  awk &#39;BEGIN&#123;OFS&#x3D;&quot;\\t&quot;&#125;&#123;if(length($1) &#x3D;&#x3D; 17 &amp;&amp; $1 ~ &#x2F;\\.1$&#x2F;)&#123;gsub(&#x2F;\\.1$&#x2F;,&quot;&quot;,$1); print &quot;\\&gt;&quot;$1&quot;\\n&quot;$2&#125;&#125;&#39; &gt; soybean\_cds.fa</span><br><span class="line">cat Mtruncatula\_285\_Mt4.0v1.cds.tab  awk &#39;BEGIN&#123;OFS&#x3D;&quot;\\t&quot;&#125;&#123;if(length($1) &#x3D;&#x3D; 15 &amp;&amp; $1 ~ &#x2F;\\.1$&#x2F;)&#123;gsub(&#x2F;\\.1&#x2F;,&quot;&quot;,$1); print &quot;\\&gt;&quot;$1&quot;\\n&quot;$2&#125;&#125;&#39; &gt; medicago.cds.fa</span><br><span class="line">iTools Fatools CDS2Pep -InCDS soybean\_cds.fa -OutPut soybean\_pep.fa</span><br><span class="line">iTools Fatools CDS2Pep -InCDS medicago.cds.fa -OutPut medicago.pep.fa</span><br></pre></td></tr></table></figure>

<p>然后用 blastp 对两个物种的蛋白序列进行比对，并用 MCScanX 命令检测共线性区间。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">makeblastdb -in medicago.pep.fa -dbtype prot -out index&#x2F;mt -parse\_seqids</span><br><span class="line">## diamond makedb --in medicago.pep.fa --db .&#x2F;diamond\_index&#x2F;mt</span><br><span class="line">## diamond blastp --query soybean\_pep.fa --evalue 1e-5 --outfmt 6 --threads 24 --db .&#x2F;diamond\_index&#x2F;mt.dmnd --out soybean\_medicago.blast</span><br><span class="line">blastp -query soybean\_pep.fa -db index&#x2F;mt -out gm\_mt.blast -evalue 1e-5 -num\_threads 24 -outfmt 6 -num\_alignments 5</span><br><span class="line">mkdir mcdata</span><br><span class="line">cp gm\_mt.blast mcdata</span><br><span class="line">cat soybean.gff medicago.gff &gt; mcdata&#x2F;gm\_mt.gff</span><br><span class="line">MCScanX .&#x2F;mcdata&#x2F;gm\_mt</span><br></pre></td></tr></table></figure>

<p>产生的线性文件（gm_mt.collinearity）如下，可以用于下游画图。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">\[fenglei@localhost gm\_mt\_test\]$ head -n 50 gm\_mt.collinearity </span><br><span class="line">############### Parameters ###############</span><br><span class="line"># MATCH\_SCORE: 50</span><br><span class="line"># MATCH\_SIZE: 5</span><br><span class="line"># GAP\_PENALTY: -1</span><br><span class="line"># OVERLAP\_WINDOW: 5</span><br><span class="line"># E\_VALUE: 1e-05</span><br><span class="line"># MAX GAPS: 25</span><br><span class="line">############### Statistics ###############</span><br><span class="line"># Number of collinear genes: 51572, Percentage: 49.62</span><br><span class="line"># Number of all genes: 103927</span><br><span class="line">##########################################</span><br><span class="line">## Alignment 0: score&#x3D;1008.0 e\_value&#x3D;1.4e-69 N&#x3D;23 gm01&amp;mt1 plus</span><br><span class="line">0- 0: Glyma01G232800 Medtr1g103160 0</span><br><span class="line">0- 1: Glyma01G234900 Medtr1g103450 8e-39</span><br><span class="line">0- 2: Glyma01G235100 Medtr1g103490 0.0001</span><br><span class="line">0- 3: Glyma01G236800 Medtr1g103500 2e-87</span><br><span class="line">0- 4: Glyma01G237000 Medtr1g103550 9e-47</span><br><span class="line">0- 5: Glyma01G237300 Medtr1g103570 2e-45</span><br><span class="line">0- 6: Glyma01G237400 Medtr1g103600 1e-144</span><br><span class="line">0- 7: Glyma01G237900 Medtr1g103690 9e-39</span><br><span class="line">0- 8: Glyma01G238200 Medtr1g103830 0</span><br><span class="line">0- 9: Glyma01G238400 Medtr1g104500 1e-170</span><br><span class="line">0- 10: Glyma01G238500 Medtr1g104520 6e-33</span><br><span class="line">0- 11: Glyma01G238700 Medtr1g104680 0</span><br><span class="line">0- 12: Glyma01G238800 Medtr1g104750 0</span><br><span class="line">0- 13: Glyma01G238900 Medtr1g104800 2e-80</span><br><span class="line">0- 14: Glyma01G239200 Medtr1g104870 0</span><br><span class="line">0- 15: Glyma01G239400 Medtr1g104930 1e-50</span><br><span class="line">0- 16: Glyma01G239800 Medtr1g105075 1e-32</span><br><span class="line">0- 17: Glyma01G239900 Medtr1g105305 5e-101</span><br><span class="line">0- 18: Glyma01G240000 Medtr1g105415 3e-91</span><br><span class="line">0- 19: Glyma01G240500 Medtr1g105555 5e-62</span><br><span class="line">0- 20: Glyma01G240600 Medtr1g105570 6e-06</span><br><span class="line">0- 21: Glyma01G240800 Medtr1g105650 2e-175</span><br><span class="line">0- 22: Glyma01G240900 Medtr1g105655 4e-164</span><br><span class="line">## Alignment 1: score&#x3D;741.0 e\_value&#x3D;1e-39 N&#x3D;16 gm01&amp;mt1 plus</span><br><span class="line">1- 0: Glyma01G084800 Medtr1g093020 0</span><br><span class="line">1- 1: Glyma01G086700 Medtr1g093080 1e-138</span><br><span class="line">1- 2: Glyma01G086800 Medtr1g093095 2e-108</span><br><span class="line">1- 3: Glyma01G087500 Medtr1g093600 5e-175</span><br><span class="line">1- 4: Glyma01G087700 Medtr1g093630 0</span><br><span class="line">1- 5: Glyma01G088000 Medtr1g093650 8e-07</span><br><span class="line">1- 6: Glyma01G088200 Medtr1g093670 9e-115</span><br><span class="line">1- 7: Glyma01G088300 Medtr1g093750 1e-118</span><br><span class="line">1- 8: Glyma01G089100 Medtr1g093770 0</span><br><span class="line">1- 9: Glyma01G089300 Medtr1g093850 0</span><br><span class="line">1- 10: Glyma01G089900 Medtr1g093860 0</span><br><span class="line">1- 11: Glyma01G090200 Medtr1g093900 0</span><br><span class="line">1- 12: Glyma01G090400 Medtr1g093920 9e-22</span><br><span class="line">1- 13: Glyma01G091000 Medtr1g094010 0</span><br></pre></td></tr></table></figure>

<p>注意画图的配置文件不要出错，双斜杠（//）表示注释，注释号前面用tab符号跟染色体名称分隔，或者干脆去掉注释，我删掉注释才顺利画图，否则报错。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">[fenglei@localhost gm\_mt\_test\]$ cat dot.ctl</span><br><span class="line"></span><br><span class="line">2000</span><br><span class="line">2000</span><br><span class="line">gm01,gm02,gm03,gm04,gm05,gm06,gm07,gm08,gm09,gm10,gm11,gm12,gm13,gm14,gm15,gm16,gm17,gm18,gm19,gm20</span><br><span class="line">mt1,mt2,mt3,mt4,mt5,mt6,mt7,mt8</span><br></pre></td></tr></table></figure>

<p>画图命令。注意 dot_plotter.java 与 dot_plotter .class 要在一个文件夹下面。其他环状图做法也类似。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cd &#x2F;PATH_TO_MCScanX&#x2F;downstream_analyses</span><br><span class="line">DATA_PATH&#x3D;&#x2F;xx&#x2F;my&#x2F;mcdata  # 根据实际情况</span><br><span class="line">java dot_plotter -g $DATA_PATH&#x2F;gm_mt.gff -s $DATA_PATH&#x2F;gm_mt.collinearity -c $DATA_PATH&#x2F;dot.ctl -o gm_mt.test.png</span><br></pre></td></tr></table></figure>

<p>画图结果：</p>
<p><img src="https://genehub.files.wordpress.com/2019/07/gm_mt.test_.png" alt="gm_mt.test"> </p>
<p>与文献（ Wang Jinpeng et al., 2017, Plant Physiol.）中的图片对比，基本是一样的效果。</p>
<p><img src="https://genehub.files.wordpress.com/2019/07/mt4_gm.order_.png" alt="mt4_gm.order"></p>
<p>这个 MCScanX 共线性图跟更新开发的 jcvi 相比，画质要差一些。下面是 jcvi 的结果。</p>
<p><img src="https://i.imgur.com/Ly4FSpT.png" alt="jcvi 绘制的基因组共线性图实例"></p>
<p>MCScanX 还可以将图画成圆形。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java circle_plotter -g &#x2F;home&#x2F;fenglei&#x2F;projects&#x2F;genome_synteny_20190704&#x2F;MCScanX_soybean_medicago_190705&#x2F;gm_mt_test&#x2F;MADS-master&#x2F;MADS.gff -s &#x2F;home&#x2F;fenglei&#x2F;projects&#x2F;genome_synteny_20190704&#x2F;MCScanX_soybean_medicago_190705&#x2F;gm_mt_test&#x2F;MADS-master&#x2F;MADS.collinearity -c &#x2F;home&#x2F;fenglei&#x2F;projects&#x2F;genome_synteny_20190704&#x2F;MCScanX_soybean_medicago_190705&#x2F;gm_mt_test&#x2F;MADS-master&#x2F;circle.ctl -o MADS.png</span><br></pre></td></tr></table></figure>
<p><strong>注意</strong> circle_plotter 的 dot.ctl 文件与 dot_plotter 的 dot.ctl 文件不一样。 </p>
<p>For circle_plotter , put chromosome IDs in one line in dot.ctl file. </p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">2000</span><br><span class="line">Gm01,Gm02,Gm03,Gs01,Gs02,Gs03</span><br></pre></td></tr></table></figure>
<p>For dot_plotter, put chromosome IDs in two lines in dot.ctl file</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">2000</span><br><span class="line">2000</span><br><span class="line">Gm01,Gm02,Gm03</span><br><span class="line">Gs01,Gs02,Gs03</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ cat &#x2F;home&#x2F;fenglei&#x2F;projects&#x2F;genome_synteny_20190704&#x2F;MCScanX_soybean_medicago_190705&#x2F;gm_mt_test&#x2F;MADS-master&#x2F;circle.ctl</span><br><span class="line">1600    &#x2F;&#x2F;plot width and height (in pixels)</span><br><span class="line">Chr1,Chr2,Chr3,Chr4,Chr5,Chr6,Chr7,Chr8,Chr9,Chr10,Chr11,Chr12,Chr13,Chr14,Chr15,Chr16,Chr17    &#x2F;&#x2F;chromosomes in the circle</span><br></pre></td></tr></table></figure>

<p><img src="https://i.imgur.com/lRwZIhA.png"></p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2022/04/01/Bioinformatics-Ancestral-Genome-evolution/">Bioinformatics | Ancestral Genome evolution</a><a class="next" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/Bioinformatics-Ancestral-Genome-evolution/">Bioinformatics | Ancestral Genome evolution</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/mcscanx-%E6%B5%8B%E8%AF%95/">MCScanX 测试</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>