<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>R：corrplot 相关系数矩阵 | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">R：corrplot 相关系数矩阵</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">R：corrplot 相关系数矩阵</h1><div class="post-meta">2020-11-16<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">install.packages(&quot;corrplot&quot;)</span><br><span class="line">library(&quot;corrplot&quot;)</span><br><span class="line">M &#x3D; cor(mtcars)</span><br><span class="line">#corrplot(M, order &#x3D; &quot;hclust&quot;, addrect &#x3D; 2)</span><br><span class="line">#head(mtcars)</span><br><span class="line">#</span><br><span class="line">#corrplot(M, order &#x3D; &quot;hclust&quot;, addrect &#x3D; 2)</span><br><span class="line"># 只需要下半部分三角形</span><br><span class="line">corrplot(M, order &#x3D; &quot;hclust&quot;, addrect &#x3D; 2, type&#x3D;&quot;lower&quot;)</span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116095756.png"><img src="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116095756.png?w=894"></a></p>
<p>参考资料：</p>
<p>[1] corrplot函数参数调整 <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/28076189">https://zhuanlan.zhihu.com/p/28076189</a></p>
<p>[2] 图形元素设定 <a target="_blank" rel="noopener" href="https://www.cnblogs.com/Mao1518202/p/11270767.html">https://www.cnblogs.com/Mao1518202/p/11270767.html</a></p>
<p>搜资料的时候看到与上述图形类似的点阵图（dot plot），还可以结合聚类分析，顺便记录一下。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"># install.packages(&#39;tidyverse&#39;)</span><br><span class="line">library(&#39;tidyverse&#39;)</span><br><span class="line"># -- Attaching packages ----------------------------------------------------# -------------------------------------------------------- tidyverse 1.3.0</span><br><span class="line"># √ tibble  3.0.4     √ stringr 1.4.0</span><br><span class="line"># √ readr   1.4.0     √ forcats 0.5.0</span><br><span class="line"># √ purrr   0.3.4     </span><br><span class="line"># -- Conflicts -------------------------------------------------------------#-------------------------------------------------- tidyverse_conflicts() --</span><br><span class="line"># x dplyr::filter() masks stats::filter()</span><br><span class="line"># x dplyr::lag()    masks stats::lag()</span><br><span class="line"></span><br><span class="line"># install.packages(&#39;ggdendro&#39;)</span><br><span class="line">library(ggdendro)</span><br><span class="line"></span><br><span class="line">library(cowplot)</span><br><span class="line"></span><br><span class="line"># BiocManager::install(&#39;ggtree&#39;)</span><br><span class="line">library(ggtree)</span><br><span class="line"></span><br><span class="line">library(patchwork) </span><br><span class="line"></span><br><span class="line">gene_cluster &lt;- read_tsv(&#39;https:&#x2F;&#x2F;github.com&#x2F;davemcg&#x2F;davemcg.github.io&#x2F;raw&#x2F;master&#x2F;content&#x2F;post&#x2F;scRNA_dotplot_data.tsv.gz&#39;)</span><br><span class="line"></span><br><span class="line">gene_cluster %&gt;% sample_n(5)</span><br><span class="line"></span><br><span class="line">markers &lt;- gene_cluster$Gene %&gt;% unique()</span><br></pre></td></tr></table></figure>

<p>#    gene_cluster %&gt;% sample_n(5) </p>
<h1 id="gene-cluster-6列-888行-的矩阵，下面显示前5行"><a href="#gene-cluster-6列-888行-的矩阵，下面显示前5行" class="headerlink" title="gene_cluster 6列 888行 的矩阵，下面显示前5行"></a>gene_cluster 6列 888行 的矩阵，下面显示前5行</h1><p>A tibble: 5 x 6<br>   Gene    cluster cell_ct cell_exp_ct   count Group    </p>
<p>   1 CELA1   c16        1278          87 0.0826  Neuron<br>   2 NDUFA1  c13        1725           5 0.00290 Neuron<br>   3 IGFBP7  c13        1725        1013 1.54    Neuron<br>   4 IGFBP7  c09        2124          29 0.0154  Astrocyte<br>   5 SLC10A7 c13        1725          59 0.0413  Neuron </p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">   gene_cluster %&gt;% filter(Gene %in% markers) %&gt;% </span><br><span class="line">      mutate(% Expressing &#x3D; (cell_exp_ct&#x2F;cell_ct) * 100) %&gt;% </span><br><span class="line">   ggplot(aes(x&#x3D;cluster, y &#x3D; Gene, color &#x3D; count, size &#x3D; % Expressing)) + </span><br><span class="line">   geom_point()     </span><br><span class="line"># mutate函数来自dplyr，功能是Create, modify, and delete columns</span><br><span class="line"># mutate函数计算一个比例数值，作为矩阵第七列，名称为“% Expression”。</span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/01.png"><img src="https://genehub.files.wordpress.com/2020/11/01.png?w=1024"></a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 用dplyr包的filter功能，筛选条件：矩阵中“count”列中大于0，“% expression”列中大于1%。</span><br><span class="line">gene_cluster %&gt;% filter(Gene %in% markers) %&gt;%    </span><br><span class="line">mutate(&#96;% Expressing&#96; &#x3D; (cell_exp_ct&#x2F;cell_ct) * 100) %&gt;%    </span><br><span class="line">filter(count &gt; 0, &#96;% Expressing&#96; &gt; 1) %&gt;%    </span><br><span class="line">ggplot(aes(x&#x3D;cluster, y &#x3D; Gene, color &#x3D; count, size &#x3D; &#96;% Expressing&#96;)) +    geom_point() </span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/02.png"><img src="https://genehub.files.wordpress.com/2020/11/02.png?w=1024"></a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"># 图形调整与美化。使用“theme(axis.line  &#x3D; element_blank())”取消显示xy坐标轴。“theme(axis.ticks &#x3D; element_blank()) ”去掉坐标轴的刻度值。</span><br><span class="line"># cowplot包是ggplot2的简单附加组件。它旨在为ggplot2提供一个出版物就绪的主题，一个需要最小量的轴标签尺寸，情节背景等。它的主要目的是方便制作符合要求的图片。除了提供修改的绘图主题外，此包还提供了对ggplot2绘图的自定义注释的功能。</span><br><span class="line"></span><br><span class="line">gene_cluster %&gt;% filter(Gene %in% markers) %&gt;%    </span><br><span class="line">mutate(&#96;% Expressing&#96; &#x3D; (cell_exp_ct&#x2F;cell_ct) * 100) %&gt;%    </span><br><span class="line">filter(count &gt; 0, &#96;% Expressing&#96; &gt; 1) %&gt;%    </span><br><span class="line">ggplot(aes(x&#x3D;cluster, y &#x3D; Gene, color &#x3D; count, size &#x3D; &#96;% Expressing&#96;)) +    geom_point() +    </span><br><span class="line">scale_color_viridis_c(name &#x3D; &#39;log2 (count + 1)&#39;) +    cowplot::theme_cowplot() +    </span><br><span class="line">theme(axis.line  &#x3D; element_blank()) +   </span><br><span class="line">theme(axis.text.x &#x3D; element_text(angle &#x3D; 90, vjust &#x3D; 0.5, hjust&#x3D;1)) +   ylab(&#39;&#39;) +   </span><br><span class="line">theme(axis.ticks &#x3D; element_blank()) </span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/03.png"><img src="https://genehub.files.wordpress.com/2020/11/03.png?w=1024"></a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">gene_cluster %&gt;% </span><br><span class="line">filter(Gene %in% markers) %&gt;%    </span><br><span class="line">mutate(&#96;% Expressing&#96; &#x3D; (cell_exp_ct&#x2F;cell_ct) * 100) %&gt;%    </span><br><span class="line">filter(count &gt; 0, &#96;% Expressing&#96; &gt; 1) %&gt;%    </span><br><span class="line">ggplot(aes(x&#x3D;cluster, y &#x3D; Gene, color &#x3D; count, size &#x3D; &#96;% Expressing&#96;)) +    geom_point() +    </span><br><span class="line">cowplot::theme_cowplot() +    </span><br><span class="line">theme(axis.line  &#x3D; element_blank()) +   </span><br><span class="line">theme(axis.text.x &#x3D; element_text(angle &#x3D; 90, vjust &#x3D; 0.5, hjust&#x3D;1)) +   ylab(&#39;&#39;) +   </span><br><span class="line">theme(axis.ticks &#x3D; element_blank()) +   </span><br><span class="line">scale_color_gradientn(colours &#x3D; viridis::viridis(20), limits &#x3D; c(0,4), oob &#x3D; scales::squish, name &#x3D; &#39;log2 (count + 1)&#39;)</span><br></pre></td></tr></table></figure>

<p>前面的图中，对所有圆点颜色进行观察，可见整体偏向深蓝，只有一个位于坐标c15 * KCNQ1 的圆点是黄色，其count数值4.94，比其他坐标的数值大了不少。通常绘制热图的时候也会因为这种值（甚至更大的极值，因为大部分基因的表达值FPKM通常小于100，但是少数基因的表达值会接近10000）导致画面颜色不够均匀。 参考原文的做法，调用 <code>scale_color_gradientn</code> 将 <code>c(0,4)</code> 数字范围的圆点进行着色，对于大于4的极值， 调用scales函数的squish功能将其“压碎”成为4（即“squished” down） <code>oob = scales::squish</code></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/05.png"><img src="https://genehub.files.wordpress.com/2020/11/05.png?w=1024"></a></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/04.png"><img src="https://genehub.files.wordpress.com/2020/11/04.png?w=1024"></a></p>
<p>进一步的，可以将基因进行聚类，并调用ggtree进行树形图绘制。先对原始数据进行整理，提取三列：Gene、cluster、count。调用pivot_wider功能，将其转换为行为gene，列为cluster的矩阵。（tidyr包的 <code>pivot_wider</code> 函数可以将长表变成宽表。 这适用于将多个变量保存到了一列的情况。—— 摘抄自<a target="_blank" rel="noopener" href="https://www.math.pku.edu.cn/teachers/lidf/docs/Rbook/html//_Rbook/summary-manip.html%EF%BC%89">https://www.math.pku.edu.cn/teachers/lidf/docs/Rbook/html/\_Rbook/summary-manip.html）</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"># make data square to calculate euclidean distance </span><br><span class="line">mat &lt;- gene_cluster %&gt;%    </span><br><span class="line">filter(Gene %in% markers) %&gt;%    </span><br><span class="line">select(-cell_ct, -cell_exp_ct, -Group) %&gt;%  # drop unused columns to faciliate widening   </span><br><span class="line">pivot_wider(names_from &#x3D; cluster, values_from &#x3D; count) %&gt;%    </span><br><span class="line">data.frame() # make df as tibbles -&gt; matrix annoying </span><br><span class="line"></span><br><span class="line">row.names(mat) &lt;- mat$Gene  # put gene in &#96;row&#96; </span><br><span class="line"></span><br><span class="line">mat &lt;- mat[,-1] #drop gene column as now in rows </span><br><span class="line"></span><br><span class="line">clust &lt;- hclust(dist(mat %&gt;% as.matrix())) # hclust with distance matrix</span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/06.png"><img src="https://genehub.files.wordpress.com/2020/11/06.png?w=1024"></a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">ddgram &lt;- as.dendrogram(clust) # create dendrogram </span><br><span class="line">ggtree_plot &lt;- ggtree::ggtree(ddgram) </span><br><span class="line">ggtree_plot</span><br><span class="line">ggtree::ggtree(ddgram, size&#x3D;1) + geom_tiplab() </span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/07-1.png"><img src="https://genehub.files.wordpress.com/2020/11/07-1.png?w=1024"></a></p>
<p>（右侧的名称有问题）</p>
<p>将树形图和点状图拼接组合。下面的操作只是将两幅图拼起来，基因名并没有对应关系。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">dotplot &lt;- gene_cluster %&gt;% filter(Gene %in% markers) %&gt;% </span><br><span class="line">   mutate(% Expressing &#x3D; (cell_exp_ct&#x2F;cell_ct) * 100) %&gt;% </span><br><span class="line">   filter(count &gt; 0, % Expressing &gt; 1) %&gt;% </span><br><span class="line">   ggplot(aes(x&#x3D;cluster, y &#x3D; Gene, color &#x3D; count, size &#x3D; % Expressing)) + </span><br><span class="line">   geom_point() + </span><br><span class="line">   cowplot::theme_cowplot() + </span><br><span class="line">   theme(axis.line  &#x3D; element_blank()) +</span><br><span class="line">   theme(axis.text.x &#x3D; element_text(angle &#x3D; 90, vjust &#x3D; 0.5, hjust&#x3D;1)) +</span><br><span class="line">   ylab(&#39;&#39;) +</span><br><span class="line">   theme(axis.ticks &#x3D; element_blank()) +</span><br><span class="line">   scale_color_gradientn(colours &#x3D; viridis::viridis(20), limits &#x3D; c(0,4), oob &#x3D; scales::squish, name &#x3D; &#39;log2 (count + 1)&#39;)   </span><br><span class="line"></span><br><span class="line"></span><br><span class="line">   plot_grid(ggtree_plot, dotplot, nrow &#x3D; 1, rel_widths &#x3D; c(0.5,2), align &#x3D; &#39;h&#39;) </span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/08.png"><img src="https://genehub.files.wordpress.com/2020/11/08.png?w=1024"></a></p>
<p>使用</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">dotplot &lt;- gene_cluster %&gt;% filter(Gene %in% markers) %&gt;% </span><br><span class="line">mutate(% Expressing &#x3D; (cell_exp_ct&#x2F;cell_ct) * 100,</span><br><span class="line">   Gene &#x3D; factor(Gene, levels &#x3D; clust$labels[clust$order])) %&gt;% </span><br><span class="line">   #filter(count &gt; 0, % Expressing &gt; 1) %&gt;% </span><br><span class="line">   ggplot(aes(x&#x3D;cluster, y &#x3D; Gene, color &#x3D; count, size &#x3D; % Expressing)) + </span><br><span class="line">   geom_point() + </span><br><span class="line">   cowplot::theme_cowplot() + </span><br><span class="line">   theme(axis.line  &#x3D; element_blank()) +</span><br><span class="line">   theme(axis.text.x &#x3D; element_text(angle &#x3D; 90, vjust &#x3D; 0.5, hjust&#x3D;1)) +</span><br><span class="line">   ylab(&#39;&#39;) +</span><br><span class="line">   theme(axis.ticks &#x3D; element_blank()) +</span><br><span class="line">   scale_color_gradientn(colours &#x3D; viridis::viridis(20), limits &#x3D; c(0,4), oob &#x3D; scales::squish, name &#x3D; &#39;log2 (count + 1)&#39;)   </span><br><span class="line">   plot_grid(ggtree_plot, NULL, dotplot, nrow &#x3D; 1, rel_widths &#x3D; c(0.5,-0.05, 2), align &#x3D; &#39;h&#39;) </span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/10.png"><img src="https://genehub.files.wordpress.com/2020/11/10.png?w=1024"></a></p>
<p>代码摘抄自：<a target="_blank" rel="noopener" href="https://davemcg.github.io/post/lets-plot-scrna-dotplots/">https://davemcg.github.io/post/lets-plot-scrna-dotplots/</a></p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2020/12/05/shell%E8%84%9A%E6%9C%AC%E5%92%8Cperl%E8%84%9A%E6%9C%AC%E7%9A%84%E6%AE%B5%E6%B3%A8%E9%87%8A/">shell脚本和perl脚本的段注释</a><a class="next" href="/2020/11/16/R%EF%BC%9A%E7%BB%8F%E9%AA%8C%E5%88%86%E5%B8%83%E5%87%BD%E6%95%B0-/">R：经验分布函数</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/Bioinfo-Ancestral-Genome-evolution/">Bioinfo | Ancestral Genome evolution</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/31/Bioinfo-MCScanX-genome-synteny-blocks/">Bioinfo | MCScanX genome synteny blocks and 4dtv of paralogous genes</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>