<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="jcvi 是 Haibao Tang 开发的基因组组装注释与比较基因组分析工具，比如可以用于可视化 synteny blocks。原理是依据物种之间基因编码序列相似性比较，如果若干个连续的基因在两个基因组之间都有出现，那这若干个基因就是一个 synteny block，可能是从祖先染色体上继承下来的，物种之间的亲缘关系越近，synteny blocks 数目越多。 之前测试过 MCScanX，其实">
<meta property="og:type" content="article">
<meta property="og:title" content="jcvi 安装与使用，报错以及解决办法">
<meta property="og:url" content="http://example.com/2021/06/03/Bioinfo-jcvi-installation/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:description" content="jcvi 是 Haibao Tang 开发的基因组组装注释与比较基因组分析工具，比如可以用于可视化 synteny blocks。原理是依据物种之间基因编码序列相似性比较，如果若干个连续的基因在两个基因组之间都有出现，那这若干个基因就是一个 synteny block，可能是从祖先染色体上继承下来的，物种之间的亲缘关系越近，synteny blocks 数目越多。 之前测试过 MCScanX，其实">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://i.imgur.com/RYAdtK3.jpg">
<meta property="og:image" content="https://i.imgur.com/Ly4FSpT.png">
<meta property="og:image" content="https://i.imgur.com/BB2UsVk.jpg">
<meta property="og:image" content="https://i.imgur.com/Bugyz9W.jpeg">
<meta property="og:image" content="https://i.imgur.com/9QhsL89.jpg">
<meta property="og:image" content="https://camo.githubusercontent.com/833e373b369f1c221132d729e1cf0128a8cd638fc04c90782ad7d5e9d30867db/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f39766c337973336e6476696d6734632f67726170652d70656163682d636163616f2e706e673f7261773d31">
<meta property="og:image" content="https://camo.githubusercontent.com/5b300c541bd65669e40cf15f7e496fd0a63a0a47ecb4c80a8f5641e9253c9981/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f736f3274643065716b6c626b6c6e6c2f67726170652e70656163682e6b6172796f747970652d677265656e2e706e673f7261773d31">
<meta property="og:image" content="https://camo.githubusercontent.com/7a878ab0e1aa1221d9c4af36a4d0f131eb43f8ddb10caea79e53366d6797cd92/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f696d646f7539756b6e74743474616d2f67726170652d70656163682d636163616f2d626c6f636b732e706e673f7261773d31">
<meta property="og:image" content="https://genehub.files.wordpress.com/2019/09/e5beaee4bfa1e688aae59bbe_20190906173207.png">
<meta property="article:published_time" content="2021-06-03T09:36:50.000Z">
<meta property="article:modified_time" content="2021-06-07T15:04:42.294Z">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://i.imgur.com/RYAdtK3.jpg">

<link rel="canonical" href="http://example.com/2021/06/03/Bioinfo-jcvi-installation/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>jcvi 安装与使用，报错以及解决办法 | FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/06/03/Bioinfo-jcvi-installation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          jcvi 安装与使用，报错以及解决办法
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-06-03 17:36:50" itemprop="dateCreated datePublished" datetime="2021-06-03T17:36:50+08:00">2021-06-03</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-06-07 23:04:42" itemprop="dateModified" datetime="2021-06-07T23:04:42+08:00">2021-06-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>jcvi 是 Haibao Tang 开发的基因组组装注释与比较基因组分析工具，比如可以用于可视化 synteny blocks。原理是依据物种之间基因编码序列相似性比较，如果若干个连续的基因在两个基因组之间都有出现，那这若干个基因就是一个 synteny block，可能是从祖先染色体上继承下来的，物种之间的亲缘关系越近，synteny blocks 数目越多。</p>
<p>之前<a target="_blank" rel="noopener" href="https://genehub.wordpress.com/2019/07/05/mcscanx-%e6%b5%8b%e8%af%95/">测试过 MCScanX</a>，其实 jcvi 和 MCScanX 就是一回事，是同一个作者开发的，jcvi 是升级版，操作更加便利，生成的图片更加精美。回去 <a target="_blank" rel="noopener" href="https://github.com/wyp1125/MCScanX#installation">MCScanX</a> 看了一下开发时间，最后更新大约是 2014 年，但是里面有的程序还是很实用的，比如可以计算共线性区块之间的基因对的 Ka/Ks value（add_kaks_to_synteny.pl）.</p>
<h3 id="安装-jcvi"><a class="markdownIt-Anchor" href="#安装-jcvi"></a> 安装 jcvi</h3>
<p>jcvi 软件推荐使用 pip 安装，也可以选择使用 conda 安装。我是 anaconda3 下面建立了 python 2.7 和 python 3.7 两个环境，分别进行了测试，最后发现 python2.7 环境下 jcvi 可以正常运行。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">source activate python27</span><br><span class="line">conda install -c bioconda jcvi</span><br><span class="line">conda install -c bioconda last</span><br><span class="line">conda install scipy</span><br><span class="line"># 但是运行序列比对时候报错！--&gt; 经过查询是 lastal 的问题，也就是别用 conda 安装的 last，自己去 last 网站下载源码，安装到用户目录下，然后将其bin路径写入环境变量bashrc文件，然后就可以解决比对报错的问题。</span><br></pre></td></tr></table></figure>
<h3 id="安装-lastal"><a class="markdownIt-Anchor" href="#安装-lastal"></a> 安装 lastal</h3>
<p>手动安装 last 软件，并将其路径写入环境变量。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">git clone https://gitlab.com/mcfrith/last.git</span><br><span class="line">cd last</span><br><span class="line">mdke</span><br><span class="line"># then we can see lastal in ./last/bin/</span><br><span class="line"># put /PATH/TO/last/bin into ~/.bashrc</span><br><span class="line">which lastal</span><br><span class="line">(python37) [fenglei@localhost cds]$ which lastal</span><br><span class="line">~/local/app/anaconda3/envs/python37/bin/lastal</span><br><span class="line">(python37) [fenglei@localhost cds]$ source ~/.bashrc</span><br><span class="line">(base) [fenglei@localhost cds]$ which lastal</span><br><span class="line">~/local/app/last/bin/lastal</span><br></pre></td></tr></table></figure>
<p>下面开始做一个数据测试，参考：<a target="_blank" rel="noopener" href="https://github.com/tanghaibao/jcvi/wiki/MCscan-(Python-version)">https://github.com/tanghaibao/jcvi/wiki/MCscan-(Python-version)</a></p>
<h3 id="基因坐标与基因序列数据准备"><a class="markdownIt-Anchor" href="#基因坐标与基因序列数据准备"></a> 基因坐标与基因序列数据准备</h3>
<p>原始数据：（1）gff3 格式的基因坐标；（2）fasta 格式的基因 CDS 序列。使用 jcvi 内置程序对这两个文件做格式整理。以大豆（Glycine max）和野生豆（Glycine soja）为例，这一步需要得到四个文件：gma.bed gma.cds gso.bed gso.cds</p>
<p>如下图所示，要注意同一个物种的 bed 文件和 cds 文件的基因 ID 要相互匹配。由于基因 ID 通常带有“.1”或“.2p”这样的后缀，需要注意，必要的时候自己写脚本处理数据。</p>
<p><img src="https://i.imgur.com/RYAdtK3.jpg" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line">### BED FORMAT FILES FOR GENES</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=transcript_id Gmax_508_Wm82.a4.v1.gene.gff3.gz &gt; gma.bed</span><br><span class="line">11:12:31 [base] Load file `Gmax_508_Wm82.a4.v1.gene.gff3.gz`</span><br><span class="line">11:13:30 [gff] Extracted 86256 features (type=mRNA id=transcript_id)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Gmax_508_Wm82.a4.v1.gene.gff3.gz &gt; gma.bed2</span><br><span class="line">11:20:14 [base] Load file `Gmax_508_Wm82.a4.v1.gene.gff3.gz`</span><br><span class="line">11:21:13 [gff] Extracted 86256 features (type=mRNA id=Name)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=transcript_id Gsoja.W05.gene.gff.gz &gt; gso.bed</span><br><span class="line">11:14:15 [base] Load file `Gsoja.W05.gene.gff.gz`</span><br><span class="line">11:15:31 [gff] Extracted 89477 features (type=mRNA id=transcript_id)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Gsoja.W05.gene.gff.gz &gt; gso.bed2             </span><br><span class="line">11:22:16 [base] Load file `Gsoja.W05.gene.gff.gz`</span><br><span class="line">11:23:32 [gff] Extracted 89477 features (type=mRNA id=Name)</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=ID Gsoja.W05.gene.gff.gz &gt; gso.bed3    </span><br><span class="line">11:27:42 [base] Load file `Gsoja.W05.gene.gff.gz`</span><br><span class="line">11:28:58 [gff] Extracted 89477 features (type=mRNA id=ID)</span><br><span class="line"></span><br><span class="line">### REMOVE DUPLICATIONS</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.bed uniq gma.bed2 </span><br><span class="line">11:35:34 [base] Load file `gma.bed2`</span><br><span class="line">11:35:37 [bed] Imported: 86256, Exported: 51013</span><br><span class="line"></span><br><span class="line">python -m jcvi.formats.bed uniq gso.bed3</span><br><span class="line">11:37:25 [base] Load file `gso.bed3`</span><br><span class="line">11:37:29 [bed] Imported: 89477, Exported: 55103</span><br><span class="line"></span><br><span class="line">### CDS FORMAT FILE FOR UNIQ GENES</span><br><span class="line"></span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gma.uniq.bed ) Gmax_508_Wm82.a4.v1.cds.fa.gz |  seqkit seq -i &gt; gma.cds</span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gso.uniq.bed ) Gsoja.W05.gene.cds.fa.gz |  seqkit seq -i &gt; gso.cds</span><br><span class="line"></span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gma.uniq.bed ) Gmax_508_Wm82.a4.v1.protein.fa |  seqkit seq -i &gt; gma.pep  # ERROR due to id isuues</span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gso.uniq.bed ) Gsoja.W05.gene.protein.fa.gz |  seqkit seq -i &gt; gso.pep</span><br><span class="line"></span><br><span class="line"># TROUBLESHOOTING</span><br><span class="line">awk &#x27;&#123;if($0~/&gt;/)&#123;gsub(/\.p/,&quot;&quot;,$1)&#125;; &#123;print&#125;&#125;&#x27; Gmax_508_Wm82.a4.v1.protein.fa &gt; Gmax_508_Wm82.a4.v1.protein.new.fa</span><br><span class="line">seqkit grep -f &lt;(cut -f 4 gma.uniq.bed ) Gmax_508_Wm82.a4.v1.protein.new.fa |  seqkit seq -i &gt; gma.pep</span><br></pre></td></tr></table></figure>
<h3 id="apairwise_synteny_searchsh"><a class="markdownIt-Anchor" href="#apairwise_synteny_searchsh"></a> a.pairwise_synteny_search.sh</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~/local/app/anaconda3/envs/python27/bin/python -m jcvi.compara.catalog ortholog gma gso --no_strip_names</span><br></pre></td></tr></table></figure>
<p>该命令直接对物种间基因序列进行比较，将生成一个 gma.gso.pdf 文件，打开如下。</p>
<p><img src="https://i.imgur.com/Ly4FSpT.png" alt="" /></p>
<p>同时生成一堆文件：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">-rw-rw-r-- 1 fenglei fenglei    139507 Jun  3 12:09 gma.gso.pdf</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   5333469 Jun  3 12:09 gma.gso.lifted.anchors</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   2931001 Jun  3 12:08 gma.gso.anchors</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   5953786 Jun  3 12:08 gma.gso.last.filtered</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  97488440 Jun  3 12:08 gma.gso.last</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  58121140 Jun  3 12:08 gso.bck</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei       515 Jun  3 12:08 gso.prj</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei 264000792 Jun  3 12:08 gso.suf</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei   1101303 Jun  3 12:07 gso.des</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    220416 Jun  3 12:07 gso.sds</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei    220416 Jun  3 12:07 gso.ssp</span><br><span class="line">-rw-rw-r-- 1 fenglei fenglei  66055302 Jun  3 12:07 gso.tis</span><br></pre></td></tr></table></figure>
<p>The <code>.last</code> file is raw LAST output, <code>.last.filtered</code> is filtered LAST output, <code>.anchors</code> is the seed synteny blocks (high quality), <code>.lifted.anchors</code> recruits additional anchors to form the final synteny blocks.</p>
<p>查看了一下 .lifted.anchors 与 .anchors 的区别，如下看到两个文件都有 551 个 synteny blocks，只是 blocks 内部的基因配对数目不同，应该就是后者有更严格的筛选条件。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.lifted.anchors | grep -v &#x27;###&#x27; | wc -l</span><br><span class="line">124133</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.lifted.anchors | grep &#x27;###&#x27; | wc -l   </span><br><span class="line">551</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.anchors | grep -v &#x27;###&#x27; | wc -l        </span><br><span class="line">68549</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ cat gma.gso.anchors | grep &#x27;###&#x27; | wc -l   </span><br><span class="line">551</span><br></pre></td></tr></table></figure>
<h3 id="bdepth_visualiztionsh"><a class="markdownIt-Anchor" href="#bdepth_visualiztionsh"></a> b.depth_visualiztion.sh</h3>
<p>We could also quick test if the synteny pattern is indeed 1:1, by running:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.compara.synteny depth --histogram gma.gso.anchors</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/BB2UsVk.jpg" alt="" /></p>
<h3 id="canchorssimple_file_generationsh"><a class="markdownIt-Anchor" href="#canchorssimple_file_generationsh"></a> c.anchors.simple_file_generation.sh</h3>
<p>This command generates a .simple file which is a more succinct form than the .anchors file.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.compara.synteny screen --minspan=30 --simple gma.gso.anchors gma.gso.anchors.new</span><br></pre></td></tr></table></figure>
<p>将 <code>gma.gso.anchors</code> 文件转化成 <code>gma.gso.anchors.simple</code> 文件。有什么不同？见下面的文本框。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">(python27) [fenglei@localhost gma-gso]$ head gma.gso.anchors</span><br><span class="line">###</span><br><span class="line">Glyma.01G000100.1       Glysoja.01G000001.1     496</span><br><span class="line">Glyma.01G000322.1       Glysoja.01G000002.1     775</span><br><span class="line">Glyma.01G000400.1       Glysoja.01G000003.1     2600</span><br><span class="line">Glyma.01G001000.2       Glysoja.01G000010.2     984</span><br><span class="line">Glyma.01G001100.3       Glysoja.01G000011.2     1490</span><br><span class="line">Glyma.01G001300.3       Glysoja.01G000013.2     3560</span><br><span class="line">Glyma.01G001350.1       Glysoja.01G000014.1     381</span><br><span class="line">Glyma.01G001500.1       Glysoja.01G000015.3     2880</span><br><span class="line">Glyma.01G001700.1       Glysoja.01G000017.1     1650</span><br><span class="line">(python27) [fenglei@localhost gma-gso]$ head gma.gso.anchors.simple</span><br><span class="line">Glyma.01G000100.1       Glyma.01G245700.6       Glysoja.01G000001.1     Glysoja.01G002529.1     2354    +</span><br><span class="line">Glyma.01G023500.1       Glyma.01G044800.1       Glysoja.02G002717.1     Glysoja.02G002929.1     209     -</span><br><span class="line">Glyma.01G045100.2       Glyma.01G071700.1       Glysoja.02G003545.1     Glysoja.02G003757.1     210     +</span><br><span class="line">Glyma.01G083700.1       Glyma.01G090200.1       Glysoja.02G003470.1     Glysoja.02G003510.1     46      +</span><br><span class="line">Glyma.01G098400.6       Glyma.01G106600.2       Glysoja.03G006294.1     Glysoja.03G006391.1     76      +</span><br><span class="line">Glyma.01G125600.2       Glyma.01G145000.1       Glysoja.03G005845.1     Glysoja.03G006096.1     198     -</span><br><span class="line">Glyma.01G118000.1       Glyma.01G122800.1       Glysoja.03G006181.1     Glysoja.03G006214.2     37      +</span><br><span class="line">Glyma.01G107000.1       Glyma.01G113400.1       Glysoja.03G006412.2     Glysoja.03G006472.1     56      +</span><br><span class="line">Glyma.01G003900.1       Glyma.01G008500.1       Glysoja.05G013057.1     Glysoja.05G013097.1     40      +</span><br><span class="line">Glyma.01G001900.2       Glyma.01G007300.1       Glysoja.07G017673.1     Glysoja.07G017734.1     56      -</span><br></pre></td></tr></table></figure>
<h3 id="dmacrosynteny_visualizationsh"><a class="markdownIt-Anchor" href="#dmacrosynteny_visualizationsh"></a> d.Macrosynteny_visualization.sh</h3>
<p>准备一个 <code>seqids</code> 文件，定义哪些染色体需要展示出来。文件内容是：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Gm01,Gm02,Gm03,Gm04,Gm05,Gm06,Gm07,Gm08,Gm09,Gm10,Gm11,Gm12,Gm13,Gm14,Gm15,Gm16,Gm17,Gm18,Gm19,Gm20</span><br><span class="line">Chr01,Chr02,Chr03,Chr04,Chr05,Chr06,Chr07,Chr08,Chr09,Chr10,Chr11,Chr12,Chr13,Chr14,Chr15,Chr16,Chr17,Chr18,Chr19,Chr20</span><br></pre></td></tr></table></figure>
<p>可以用 shell 代码生成上面的文件，我的命令如下。注意 seqids 内容上下行代表两个物种，跟下文出现的 layout 文件要顺序对应。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">for i in $(seq 1 20); do printf &quot;Gm&quot;%02d&quot;,&quot; $i ; done; echo </span><br><span class="line">for i in $(seq 1 20); do printf &quot;Chr&quot;%02d&quot;,&quot; $i ; done; echo</span><br></pre></td></tr></table></figure>
<p>再准备一个 <code>layout</code> 文件。定义染色体出现的位置，角度。文件内容是：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># y, xstart, xend, rotation, color, label, va,  bed</span><br><span class="line"> .6,     .1,    .8,       0,      , gma, top, gma.bed</span><br><span class="line"> .4,     .1,    .8,       0,      , gso, top, gso.bed</span><br><span class="line"># edges</span><br><span class="line">e, 0, 1, gma.gso.anchors.simple</span><br></pre></td></tr></table></figure>
<h3 id="ekaryotypesh"><a class="markdownIt-Anchor" href="#ekaryotypesh"></a> <a target="_blank" rel="noopener" href="http://e.karyotype.sh">e.karyotype.sh</a></h3>
<p>画图文件准备完毕，运行画图命令：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.graphics.karyotype seqids layout</span><br></pre></td></tr></table></figure>
<p>将生成一个 karyotype.pdf 文件，打开如下。</p>
<p><img src="https://i.imgur.com/Bugyz9W.jpeg" alt="" /></p>
<p>以上就是两个物种间的共线性比较分析。</p>
<h3 id="同一个物种内部的-synteny-呢"><a class="markdownIt-Anchor" href="#同一个物种内部的-synteny-呢"></a> 同一个物种内部的 synteny 呢？</h3>
<p>由于大豆在近期经历了 WGD 事件，导致很多基因都有双拷贝，所以内部就有 synteny，我做了测试分析，用与上文相同的代码，得到图片如下。看来 jcvi 程序会自动滤掉基因的自我比较，而是做 paralog 的比较。</p>
<p><img src="https://i.imgur.com/9QhsL89.jpg" alt="" /></p>
<h3 id="其他案例"><a class="markdownIt-Anchor" href="#其他案例"></a> 其他案例</h3>
<p>软件的文档里面还列出了三个物种的比较、对某些 synteny 区段进行高亮、只展示特定染色体区间的方法。下面直接摘抄原作者的图片，留作参考。</p>
<p><img src="https://camo.githubusercontent.com/833e373b369f1c221132d729e1cf0128a8cd638fc04c90782ad7d5e9d30867db/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f39766c337973336e6476696d6734632f67726170652d70656163682d636163616f2e706e673f7261773d31" alt="" /></p>
<p><img src="https://camo.githubusercontent.com/5b300c541bd65669e40cf15f7e496fd0a63a0a47ecb4c80a8f5641e9253c9981/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f736f3274643065716b6c626b6c6e6c2f67726170652e70656163682e6b6172796f747970652d677265656e2e706e673f7261773d31" alt="" /></p>
<p><img src="https://camo.githubusercontent.com/7a878ab0e1aa1221d9c4af36a4d0f131eb43f8ddb10caea79e53366d6797cd92/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f696d646f7539756b6e74743474616d2f67726170652d70656163682d636163616f2d626c6f636b732e706e673f7261773d31" alt="" /></p>
<h3 id="其他笔记"><a class="markdownIt-Anchor" href="#其他笔记"></a> 其他笔记</h3>
<p>下面是软件安装调试的一些笔记。python3.7 环境我也测试了，可以安装 jcvi，而且我发现 python3.7 的 jcvi 与 python2.7 下面的 jcvi 的屏幕输出不同。但是 pyton2.7 jcvi 运行过程中有些步骤报错（[Errno 13] Permission denied: ‘dvipng’）导致缺失一些数据结果，所以我放弃了python3.7 jcvi，转而用 python2.7 jcvi 来做分析。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">conda install scipy</span><br><span class="line">conda install -c bioconda last</span><br><span class="line">conda install -c bioconda jcvi  # 这一条就报错！</span><br><span class="line"># 于是从github手动下载软件尝试</span><br><span class="line">cd ~/local/app  # or any directory of your choice</span><br><span class="line">git clone git://github.com/tanghaibao/jcvi.git</span><br><span class="line">pip install -e .</span><br><span class="line"># 安装成功。能否运行数据比对？</span><br><span class="line"># 依然不能，经查是 lastal 的问题！！！</span><br><span class="line">## lastal: I was installed here with multi-threading disabled  # 运行到lastal就报错，根本没有比对结果输出。</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">conda create -n jcvi python=2.7</span><br><span class="line">source activate jcvi</span><br><span class="line"># 安装前需要确认存在bioconda和r channel，安装前先更新conda</span><br><span class="line"># conda --add-channel r</span><br><span class="line"># conda install jcvi last</span><br><span class="line">conda install -c bioconda last</span><br></pre></td></tr></table></figure>
<p>后续还可能会由于缺少 scipy 在绘图时报错 ImportError: No module named scipy.spatial ，因此可提<br />
前安装</p>
<p><strong>安装scipy</strong><br />
#conda-forge scipy-1.2.0<br />
conda install scipy</p>
<p><strong>问题一</strong></p>
<p>File “/biosoft/python/Python-v2.7.11/lib/python2.7/site-packages/subprocess32.py”, line 1393, in _execute_child<br />
    raise child_exception_type(errno_num, err_msg)<br />
OSError: [Errno 2] No such file or directory: ‘kpsewhich’</p>
<p>解决办法：linux中缺少kpsewhich命令导致的，于是安装该命令，CentOS系统方法如下：</p>
<p>sudo yum install tetex-latex tetex-doc tetex-fonts<br />
sudo yum install texlive tex-live-latex</p>
<p>安装好后再次运行，顺利通过。</p>
<p>我的系统是redhat，没有安yum，所以下载了textlive的源码编译，但是一直不能通过，后来重装了一个centOS系统才可以用。</p>
<p><strong>问题二</strong></p>
<p>原来在用户目录下安装了GCC-6.5.0，可以正常使用jcvi的；但是昨天升级到了GCC-9.2.0之后，今天用jcvi 画synteny图的时候遇到下面的问题。</p>
<p>ImportError: /home/fenglei/local/lib64/libstdc++.so.6: undefined symbol: libiconv</p>
<p>将libstdc++.so.6这个软连接重新指向旧版GCC对应的libstdc++.so.6.0.22，然后jcvi就没有报错了。</p>
<p><img src="https://genehub.files.wordpress.com/2019/09/e5beaee4bfa1e688aae59bbe_20190906173207.png" alt="微信截图_20190906173207" /></p>

    </div>

    
    
    

      <footer class="post-footer">

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/" rel="prev" title="Bioinfo | Installation of OrthoMCL-pipeline">
      <i class="fa fa-chevron-left"></i> Bioinfo | Installation of OrthoMCL-pipeline
    </a></div>
      <div class="post-nav-item">
    <a href="/2021/06/08/Linux-PKG-CONFIG-PATH/" rel="next" title="Linux | 认识 PKG_CONFIG_PATH">
      Linux | 认识 PKG_CONFIG_PATH <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%AE%89%E8%A3%85-jcvi"><span class="nav-number">1.</span> <span class="nav-text"> 安装 jcvi</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%AE%89%E8%A3%85-lastal"><span class="nav-number">2.</span> <span class="nav-text"> 安装 lastal</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%9F%BA%E5%9B%A0%E5%9D%90%E6%A0%87%E4%B8%8E%E5%9F%BA%E5%9B%A0%E5%BA%8F%E5%88%97%E6%95%B0%E6%8D%AE%E5%87%86%E5%A4%87"><span class="nav-number">3.</span> <span class="nav-text"> 基因坐标与基因序列数据准备</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#apairwise_synteny_searchsh"><span class="nav-number">4.</span> <span class="nav-text"> a.pairwise_synteny_search.sh</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#bdepth_visualiztionsh"><span class="nav-number">5.</span> <span class="nav-text"> b.depth_visualiztion.sh</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#canchorssimple_file_generationsh"><span class="nav-number">6.</span> <span class="nav-text"> c.anchors.simple_file_generation.sh</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#dmacrosynteny_visualizationsh"><span class="nav-number">7.</span> <span class="nav-text"> d.Macrosynteny_visualization.sh</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#ekaryotypesh"><span class="nav-number">8.</span> <span class="nav-text"> e.karyotype.sh</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%90%8C%E4%B8%80%E4%B8%AA%E7%89%A9%E7%A7%8D%E5%86%85%E9%83%A8%E7%9A%84-synteny-%E5%91%A2"><span class="nav-number">9.</span> <span class="nav-text"> 同一个物种内部的 synteny 呢？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%85%B6%E4%BB%96%E6%A1%88%E4%BE%8B"><span class="nav-number">10.</span> <span class="nav-text"> 其他案例</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%85%B6%E4%BB%96%E7%AC%94%E8%AE%B0"><span class="nav-number">11.</span> <span class="nav-text"> 其他笔记</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">173</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">10</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
