<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="Illumina 测序产生大量的短片段　DNA　序列，即 Reads。如果是下机原始数据 Raw data，通常需要进行过滤得到 Clean data，才能用于正式的组装和比对等数据分析。好几年没做过过滤，因为好多次收到公司交付的数据已经是过滤后的高质量数据，今年收到一批 Raw data，需要自己过滤，这里参考其他资料做个小结。">
<meta property="og:type" content="article">
<meta property="og:title" content="Bioinfo | 使用 Trimmomatic 软件对 Illumina 测序数据进行过滤">
<meta property="og:url" content="http://example.com/2022/11/07/Illumina-raw-reads-filtering/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:description" content="Illumina 测序产生大量的短片段　DNA　序列，即 Reads。如果是下机原始数据 Raw data，通常需要进行过滤得到 Clean data，才能用于正式的组装和比对等数据分析。好几年没做过过滤，因为好多次收到公司交付的数据已经是过滤后的高质量数据，今年收到一批 Raw data，需要自己过滤，这里参考其他资料做个小结。">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://i.imgur.com/SMmxiuX.png">
<meta property="og:image" content="https://i.imgur.com/ppkg0Pu.png">
<meta property="article:published_time" content="2022-11-07T10:10:07.000Z">
<meta property="article:modified_time" content="2022-11-07T10:29:39.131Z">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://i.imgur.com/SMmxiuX.png">

<link rel="canonical" href="http://example.com/2022/11/07/Illumina-raw-reads-filtering/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>Bioinfo | 使用 Trimmomatic 软件对 Illumina 测序数据进行过滤 | FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/11/07/Illumina-raw-reads-filtering/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          Bioinfo | 使用 Trimmomatic 软件对 Illumina 测序数据进行过滤
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2022-11-07 18:10:07 / Modified: 18:29:39" itemprop="dateCreated datePublished" datetime="2022-11-07T18:10:07+08:00">2022-11-07</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>Illumina 测序产生大量的短片段　DNA　序列，即 Reads。如果是下机原始数据 Raw data，通常需要进行过滤得到 Clean data，才能用于正式的组装和比对等数据分析。好几年没做过过滤，因为好多次收到公司交付的数据已经是过滤后的高质量数据，今年收到一批 Raw data，需要自己过滤，这里参考其他资料做个小结。</p>
<span id="more"></span>
<h3 id="为什么需要过滤低质量-reads"><a class="markdownIt-Anchor" href="#为什么需要过滤低质量-reads"></a> 为什么需要过滤低质量 Reads？</h3>
<p>下图为例，可以用来理解 Illumina 的测序原理，待测 DNA 片段两端连接有 Adapter（这就是通常所称的“文库”），上机的过程中，测序引物结合在 Adapter 的特定区域并且通过“边合成、边测序”的原理逐个碱基向未知的待测片段读取碱基信息，从而获得所谓的 Reads。目前，拿到的 Reads 长度通常为 100 bp 和 150 bp 两种规格。设想，如果文库中某些片段小于 Reads 长度，那么测序的过程中将“测通”整个待测 DNA，甚至读到另一端的 Adapter，这就使我们获得的 Reads 序列包含 Adapter 序列。另一个情况就是 Illumina 对每个碱基会生成质量值，如果某个碱基质量值很低，就不可靠，可以过滤掉。还有一个情况，根据我的经验，基因组测序的 Reads 集中会包括一部分简单重复序列，比如大量重复的 AAAAAA，这种简单序列对组装可能没什么用，可以滤掉。</p>
<p><img src="https://i.imgur.com/SMmxiuX.png" alt="" /></p>
<p>其实对于转录组测序而言，如果不经过过滤，对基因组比对的影响也很小，因为低质量的 Reads 在比对的过程中很可能使无法比对到基因组序列的。</p>
<h3 id="过滤软件的选择"><a class="markdownIt-Anchor" href="#过滤软件的选择"></a> 过滤软件的选择</h3>
<p>早期（2014年）的文献中有各种不同的过滤软件，我甚至拿不少软件做过测试，后来 Trimmomatic 软件发表，我发现非常好用。该文章至今已被大量引用，是一个广受欢迎的 Illumina 平台数据过滤工具。Trimmomatic 支持多线程，处理数据速度快，主要用来去除 Illumina 平台的 Fastq 序列中的接头，并根据碱基质量值对 Fastq 进行修剪。软件有两种过滤模式，分别对应 SE 和 PE 测序数据，同时支持 gzip 和 bzip2 压缩文件。另外也支持 phred-33 和 phred-64 格式互相转化，现在之所以会出现 phred-33 和 phred-64 格式的不同，是 Illumina 公司的不同质量体系，不过现在绝大部分 Illumina 平台的产出数据也都转为使用 phred-33 格式。下面的脚本代码可以快速识别具体的质量格式。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ cat 00.fq_qual_type.sh </span><br><span class="line">less $1 | head -n 1000 | awk &#x27;&#123;if(NR%4==0) printf(&quot;%s&quot;,$0);&#125;&#x27; \</span><br><span class="line">        | od -A n -t u1 -v \</span><br><span class="line">        | awk &#x27;BEGIN&#123;min=100;max=0;&#125; \</span><br><span class="line">        &#123;for(i=1;i&lt;=NF;i++) &#123;if($i&gt;max) max=$i; if($i&lt;min) min=$i;&#125;&#125;END \</span><br><span class="line">                &#123;if(max&lt;=126 &amp;&amp; min&lt;59) print &quot;Phred33&quot;; \</span><br><span class="line">                        else if(max&gt;73 &amp;&amp; min&gt;=64) print &quot;Phred64&quot;; \</span><br><span class="line">                        else if(min&gt;=59 &amp;&amp; min&lt;64 &amp;&amp; max&gt;73) print &quot;Solexa64&quot;; \</span><br><span class="line">                        else print &quot;Unknown score encoding&quot;; \</span><br><span class="line">                print &quot;( &quot; min &quot;, &quot; max, &quot;)&quot;;&#125;&#x27;</span><br></pre></td></tr></table></figure>
<h3 id="trimmomatic-的安装与运行"><a class="markdownIt-Anchor" href="#trimmomatic-的安装与运行"></a> Trimmomatic 的安装与运行</h3>
<p>官网：<a target="_blank" rel="noopener" href="http://www.usadellab.org/cms/?page=trimmomatic">http://www.usadellab.org/cms/?page=trimmomatic</a>，我直接下载<a target="_blank" rel="noopener" href="http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.39.zip">二进制文件</a>，解压即可使用。官网也提供详细的<a target="_blank" rel="noopener" href="http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf">《Trimmomatic 使用说明书》</a>。</p>
<p>官网提供的一个示范：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"># for reference only (less sensitive for adapters)</span><br><span class="line"></span><br><span class="line">java -jar trimmomatic-0.35.jar PE -phred33 \</span><br><span class="line">    input_forward.fq.gz input_reverse.fq.gz \</span><br><span class="line">    output_forward_paired.fq.gz output_forward_unpaired.fq.gz \</span><br><span class="line">    output_reverse_paired.fq.gz output_reverse_unpaired.fq.gz \</span><br><span class="line">    ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36</span><br><span class="line"></span><br><span class="line"># This will perform the following:</span><br><span class="line"># Remove adapters (ILLUMINACLIP:TruSeq3-PE.fa:2:30:10)</span><br><span class="line"># Remove leading low quality or N bases (below quality 3) (LEADING:3)</span><br><span class="line"># Remove trailing low quality or N bases (below quality 3) (TRAILING:3)</span><br><span class="line"># Scan the read with a 4-base wide sliding window, cutting when the average quality per base drops below 15 (SLIDINGWINDOW:4:15)</span><br><span class="line"># Drop reads below the 36 bases long (MINLEN:36)</span><br></pre></td></tr></table></figure>
<p>我根据该示范的实际操作代码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">java -jar /home/fenglei/local/app/Trimmomatic/trimmomatic-0.39.jar PE \</span><br><span class="line">        -threads $NUMCPUS -phred33 -summary $CLEANDATAPATH/$&#123;sample&#125;_trim.summary \</span><br><span class="line">        $&#123;prefix&#125;_1.fq.gz $&#123;prefix&#125;_2.fq.gz \</span><br><span class="line">        $CLEANDATAPATH/$&#123;sample&#125;_pass_1.fq.gz $CLEANDATAPATH/$&#123;sample&#125;_unpass_1.fq.gz \</span><br><span class="line">        $CLEANDATAPATH/$&#123;sample&#125;_pass_2.fq.gz $CLEANDATAPATH/$&#123;sample&#125;_unpass_2.fq.gz \</span><br><span class="line">        ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:True LEADING:3 TRAILING:3 MINLEN:36</span><br><span class="line"># 这个参数别用：-trimlog $CLEANDATAPATH/$&#123;sample&#125;_trim.log  ### DO NOT USE THIS PARAMETER! IT WILL GENERATE A HUGE FILE WITH ALL RAED NAMES.</span><br><span class="line"># 这个参数可用：SLIDINGWINDOW:4:15</span><br></pre></td></tr></table></figure>
<p>注意：</p>
<ol>
<li>Adpater 序列文件 <code>TruSeq3-PE.fa</code> 默认在当前目录，如果文件在其他位置就需要添加路径，比如 <code>ILLUMINACLIP:/PATH/TO/TruSeq3-PE.fa:2:30:10:2:True</code>。</li>
<li>Adpater 序列文件里面序列名称有讲究，Naming of the sequences indicates how they should be used. For ‘Palindrome’ clipping, the sequence names should both start with ‘Prefix’, and end in ‘/1’ for the forward adapter and ‘/2’ for the reverse adapter. All other sequences are checked using ‘simple’ mode. Sequences with names ending in ‘/1’ or ‘/2’ will be checked only against the forward or reverse read. Sequences not ending in ‘/1’ or ‘/2’ will be checked against both the forward and reverse read. If you want to check for the reverse-complement of a specific sequence, you need to specifically include the reverse-complemented form of the sequence as well, with another name.</li>
<li>生成的质量过关的文件，包括 <code>$CLEANDATAPATH/$&#123;sample&#125;_pass_1.fq.gz</code> 和 <code>$CLEANDATAPATH/$&#123;sample&#125;_pass_2.fq.gz</code>，这两个文件的序列数目是一致的。但是里面有部分序列的长度是经过修剪的，因为设置了 <code>MINLEN:36</code> 参数，去掉 Adapter 污染的序列变短。</li>
<li>生成的 <code>$CLEANDATAPATH/$&#123;sample&#125;_unpass_1.fq.gz</code> 和 <code>$CLEANDATAPATH/$&#123;sample&#125;_unpass_2.fq.gz</code> 两个文件，大小是不一样的，打开查看，这两个文件包含主要是一些低质量的简单重复序列。</li>
</ol>
<p><img src="https://i.imgur.com/ppkg0Pu.png" alt="Flow of reads in Trimmomatic Paired End mode" /></p>
<h3 id="trimmomatic-软件参数的解释"><a class="markdownIt-Anchor" href="#trimmomatic-软件参数的解释"></a> Trimmomatic 软件参数的解释</h3>
<p>Trimmomatic 过滤数据的步骤与命令行中过滤参数的顺序有关，通常的过滤步骤如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">ILLUMINACLIP: 过滤 reads 中的 Illumina 测序接头和引物序列，并决定是否去除反向互补的 R1/R2 中的 R2。</span><br><span class="line">SLIDINGWINDOW: 从 reads 的 5’ 端开始，进行滑窗质量过滤，切掉碱基质量平均值低于阈值的滑窗。</span><br><span class="line">MAXINFO: 一个自动调整的过滤选项，在保证 reads 长度的情况下尽量降低测序错误率，最大化 reads 的使用价值。</span><br><span class="line">LEADING: 从 reads 的开头切除质量值低于阈值的碱基。</span><br><span class="line">TRAILING: 从 reads 的末尾开始切除质量值低于阈值的碱基。</span><br><span class="line">CROP: 从 reads 的末尾切掉部分碱基使得 reads 达到指定长度。</span><br><span class="line">HEADCROP: 从 reads 的开头切掉指定数量的碱基。</span><br><span class="line">MINLEN: 如果经过剪切后 reads 的长度低于阈值则丢弃这条 reads。</span><br><span class="line">AVGQUAL: 如果 reads 的平均碱基质量值低于阈值则丢弃这条 reads。</span><br><span class="line">TOPHRED33: 将 reads 的碱基质量值体系转为 phred-33。</span><br><span class="line">TOPHRED64: 将 reads 的碱基质量值体系转为 phred-64。</span><br></pre></td></tr></table></figure>
<h3 id="参考资料"><a class="markdownIt-Anchor" href="#参考资料"></a> 参考资料</h3>
<ol>
<li><a target="_blank" rel="noopener" href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic: A flexible read trimming tool for Illumina NGS data</a></li>
<li><a target="_blank" rel="noopener" href="http://www.biotrainee.com/thread-1484-1-1.html">生信技能树：NGS 数据过滤之 Trimmomatic 详细说明</a></li>
<li><a target="_blank" rel="noopener" href="https://www.modb.pro/db/376116">生信洞：Trimmomatic 使用简介</a></li>
<li><a target="_blank" rel="noopener" href="https://www.cnblogs.com/Sunny-King/p/Bioinformatics-Trimmomatic.html">Sunny King：Trimmomatic 安装和使用</a></li>
</ol>

    </div>

    
    
    

      <footer class="post-footer">

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/11/07/submit-proteomics-raw-data-to-PRIDE/" rel="prev" title="Bioinfo | 提交蛋白质组学原始数据到公共数据库">
      <i class="fa fa-chevron-left"></i> Bioinfo | 提交蛋白质组学原始数据到公共数据库
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/11/09/R-draw-high-quality-figure-in-RStudio/" rel="next" title="R | 如何用 RStudio 绘制适合投稿的柱状图（柱状图 + 散点 + 显著性标记）">
      R | 如何用 RStudio 绘制适合投稿的柱状图（柱状图 + 散点 + 显著性标记） <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E4%B8%BA%E4%BB%80%E4%B9%88%E9%9C%80%E8%A6%81%E8%BF%87%E6%BB%A4%E4%BD%8E%E8%B4%A8%E9%87%8F-reads"><span class="nav-number">1.</span> <span class="nav-text"> 为什么需要过滤低质量 Reads？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E8%BF%87%E6%BB%A4%E8%BD%AF%E4%BB%B6%E7%9A%84%E9%80%89%E6%8B%A9"><span class="nav-number">2.</span> <span class="nav-text"> 过滤软件的选择</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#trimmomatic-%E7%9A%84%E5%AE%89%E8%A3%85%E4%B8%8E%E8%BF%90%E8%A1%8C"><span class="nav-number">3.</span> <span class="nav-text"> Trimmomatic 的安装与运行</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#trimmomatic-%E8%BD%AF%E4%BB%B6%E5%8F%82%E6%95%B0%E7%9A%84%E8%A7%A3%E9%87%8A"><span class="nav-number">4.</span> <span class="nav-text"> Trimmomatic 软件参数的解释</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%8F%82%E8%80%83%E8%B5%84%E6%96%99"><span class="nav-number">5.</span> <span class="nav-text"> 参考资料</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">184</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
