<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="直接对两个基因组进行比较，用Mummer： 例如，将细菌1（H_pyloriJ99_Eslice.fasta）与细菌2（H_pylori26695_Eslice.fasta）进行比较： mummer -mum -b -c H_pylori26695_Eslice.fasta H_pyloriJ99_Eslice.fasta &gt; mummer.mums mummerplot -x [0,275">
<meta property="og:type" content="article">
<meta property="og:title" content="两个基因组序列相互比较：Mummer">
<meta property="og:url" content="http://example.com/2020/12/05/Bioinfo-compare-two-genome-sequences-by-mummer/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:description" content="直接对两个基因组进行比较，用Mummer： 例如，将细菌1（H_pyloriJ99_Eslice.fasta）与细菌2（H_pylori26695_Eslice.fasta）进行比较： mummer -mum -b -c H_pylori26695_Eslice.fasta H_pyloriJ99_Eslice.fasta &gt; mummer.mums mummerplot -x [0,275">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://genehub.files.wordpress.com/2019/07/mummer.png">
<meta property="og:image" content="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome_i89_l1000_1.jpg?w=1024">
<meta property="og:image" content="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome.jpg?w=1024">
<meta property="og:image" content="https://genehub.files.wordpress.com/2020/12/gm508_gsoja_genome_i89_l1000_1.jpg?w=1024">
<meta property="article:published_time" content="2020-12-05T13:11:00.000Z">
<meta property="article:modified_time" content="2021-02-25T09:42:04.411Z">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://genehub.files.wordpress.com/2019/07/mummer.png">

<link rel="canonical" href="http://example.com/2020/12/05/Bioinfo-compare-two-genome-sequences-by-mummer/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>两个基因组序列相互比较：Mummer | FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/05/Bioinfo-compare-two-genome-sequences-by-mummer/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          两个基因组序列相互比较：Mummer
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-05 21:11:00" itemprop="dateCreated datePublished" datetime="2020-12-05T21:11:00+08:00">2020-12-05</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-02-25 17:42:04" itemprop="dateModified" datetime="2021-02-25T17:42:04+08:00">2021-02-25</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>直接对两个基因组进行比较，用Mummer：</p>
<p>例如，将细菌1（H_pyloriJ99_Eslice.fasta）与细菌2（H_pylori26695_Eslice.fasta）进行比较：</p>
<p>mummer -mum -b -c H_pylori26695_Eslice.fasta H_pyloriJ99_Eslice.fasta &gt; mummer.mums<br />
mummerplot -x [0,275287] -y [0,265111] -postscript -p mummer mummer.mums</p>
<p>结果：</p>
<p><img src="https://genehub.files.wordpress.com/2019/07/mummer.png" alt="mummer.png" /></p>
<p>如果换成两个大基因组呢？800M以上的基因组，结果报错了，尺寸太大，程序默认不支持这么多，只允许400M。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># reading input file &quot;Jxxx.fasta&quot; of length 854295539</span><br><span class="line"># construct suffix tree for sequence of length 854295539</span><br><span class="line"># (maximum reference length is 536870908)</span><br><span class="line"># (maximum query length is 4294967295)</span><br><span class="line"># process 8542955 characters per dot</span><br><span class="line">mummer: suffix tree construction failed: textlen=854295539 larger than maximal textlen=536870908</span><br></pre></td></tr></table></figure>
<p>我的版本是：MUMmer3.23</p>
<p>需要对mummer添加参数重新编译：</p>
<p>The size of your reference sequence is larger than is supported by default.<br />
You can either break your reference into multiple files and run each<br />
individually, or recompile the mummer package with:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">make clean</span><br><span class="line">make CPPFLAGS=&quot;-O3 -DSIXTYFOURBITS&quot;</span><br></pre></td></tr></table></figure>
<p><strong>20201204 更新</strong></p>
<p>MUMMER v4 可以直接对大型基因组进行比较，他的功能已经升级了。<a target="_blank" rel="noopener" href="https://github.com/mummer4/mummer/releases/download/v4.0.0rc1/mummer-4.0.0rc1.tar.gz">mummer-4.0.0rc1.tar.gz下载链接</a></p>
<p>如何对同一个物种的不同版本基因组序列比较？</p>
<p>以大豆<em>Glycine max</em>为例，大豆20条染色体，基因组尺寸约1 Gb，最新基因组版本Gmax_508_a4.v1，上一个版本是Gmax_275_a2.v1。</p>
<p><strong>一、先用nucmer对两个genome序列进行比较，生成delta文件。</strong><br />
我设定的最小匹配单元长度是1 kb，也就是1 kb内不允许错配。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">nucmer --prefix=GM508_GM275_genome --minmatch=1000 --threads=12 ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa ~/databases/Glycine_max/phytozome.12/Gmax_275_v2.0.softmasked.fa</span><br></pre></td></tr></table></figure>
<p><strong>二、然后可以delta-filter自定义筛选。</strong><br />
比如说我只想看contig和reference两者唯一匹配，并且长度在1000，相似度大于90.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">delta-filter -i 89 -l 1000 -1 GM508_GM275_genome.delta &gt;GM508_GM275_genome_i89_l1000_1.delta</span><br></pre></td></tr></table></figure>
<p><strong>三、再用mummerplot命令画图。</strong><br />
注意Rfile和Qfile主要功能是提供两个基因组各自包含的若干染色体（或者scaffold、contig）序列，不加这个参数，出来的图片没有染色体名称标记的。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mummerplot -Rfile ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa -Qfile ~/databases/Glycine_max/phytozome.12/Gmax_275_v2.0.softmasked.fa -postscript -p GM508_GM275_genome_i89_l1000_1 GM508_GM275_genome_i89_l1000_1.delta</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># `gnuplot 5.4 patchlevel 1  </span><br><span class="line"># Reading delta file GM508_GM275_genome_i89_l1000_1.delta  </span><br><span class="line"># Writing plot files GM508_GM275_genome_i89_l1000_1.fplot, GM508_GM275_genome_i89_l1000_1.rplot  </span><br><span class="line"># Writing gnuplot script GM508_GM275_genome_i89_l1000_1.gp  </span><br><span class="line"># Rendering plot GM508_GM275_genome_i89_l1000_1.ps`</span><br></pre></td></tr></table></figure>
<p><strong>20201230 备注</strong></p>
<p>更换一台机器之后，运行画图遇到错误。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ mummerplot -Rfile ./AM_genome_V5_20201211.fa -Qfile ./20201228_3ddna.fa -postscript -p plot_test V5_vs_3dV1_genome_i89_l1000_1.delta </span><br><span class="line">gnuplot 4.6 patchlevel 2</span><br><span class="line">Reading delta file V5_vs_3dV1_genome_i89_l1000_1.delta</span><br><span class="line">Writing plot files plot_test.fplot, plot_test.rplot</span><br><span class="line">Writing gnuplot script plot_test.gp</span><br><span class="line">Rendering plot plot_test.ps</span><br><span class="line">WARNING: Unable to run &#x27;false plot_test.gp&#x27;, Inappropriate ioctl for device</span><br></pre></td></tr></table></figure>
<p>切换到 python37 或者 python27 环境，用conda安装了更高版本的 gnuplot ，再安装 libjpeg-turbo 就可以运行 gnuplot 了。但是仍然无法运行mummerplot。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"># 用conda安装gnuplot</span><br><span class="line">conda install -c bioconda gnuplot</span><br><span class="line"></span><br><span class="line"># 启动gnuplot遇到错误</span><br><span class="line">$ gnuplot --version</span><br><span class="line">gnuplot: error while loading shared libraries: libjpeg.so.8: cannot open shared object file: No such file or directory</span><br><span class="line"></span><br><span class="line"># 安装 libjpeg-turbo，就可以运行gnuplot了</span><br><span class="line">$ conda install -c conda-forge libjpeg-turbo</span><br><span class="line"></span><br><span class="line"># 成功运行gnuplot</span><br><span class="line">(python27) [fenglei@localhost ~]$ gnuplot --version</span><br><span class="line">gnuplot 5.0 patchlevel 3</span><br></pre></td></tr></table></figure>
<p><strong>四、生成比对图片</strong><br />
把ps文件放到windows系统，用Photoshop软件打开上面生成的ps文件即可看到图。当然也可以直接在linux里面用convert命令将ps文件转换为pdf/png文件，但是我的linux系统没有convert命令。如图可以见到两个版本基因组的差异。11和14号染色体有明显的结构差异。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome_i89_l1000_1.jpg"><img src="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome_i89_l1000_1.jpg?w=1024" alt="" /></a></p>
<p><strong>五、如果不对delta文件做自定义筛选呢？</strong><br />
可见到大量的点，浅蓝色是反向匹配，深蓝色是正向匹配。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome.jpg"><img src="https://genehub.files.wordpress.com/2020/12/gm508_gm275_genome.jpg?w=1024" alt="" /></a></p>
<p><strong>六、如果两个物种相比较呢？</strong><br />
要注意minmatch参数不能设太大，我把两个物种（Glycine max and Glycine soja）的minmatch设置为1000000，虽然运行时间比较短，但是结果输出一个空白delta文件。设置为minmatch=100测试，需要的时间更长，得到一个满屏铺满蓝点的图，根本看不清。设置minmatch=1000，并用delta-filter对比对的序列做筛选，进而得到下图。可见到11和13号染色体的区别。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">nucmer --prefix=GM508_Gsoja_genome --minmatch=1000 --threads=12 ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa ~/databases/03_Glycine_soja_W05_2019/Gsoja.W05.genome.fadelta-filter -i 89 -l 1000 -1 GM508_Gsoja_genome.delta &gt; GM508_Gsoja_genome_i89_l1000_1.deltamummerplot -Rfile ~/databases/Glycine_max/508_Wm82.a4.v1/assembly/Gmax_508_v4.0.softmasked.fa -Qfile ~/databases/03_Glycine_soja_W05_2019/Gsoja.W05.genome.fa -postscript -p GM508_Gsoja_genome_i89_l1000_1 GM508_Gsoja_genome_i89_l1000_1.delta</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/gm508_gsoja_genome_i89_l1000_1.jpg"><img src="https://genehub.files.wordpress.com/2020/12/gm508_gsoja_genome_i89_l1000_1.jpg?w=1024" alt="" /></a></p>
<p>补充：画图可以用dotPlotly包，参考 <a target="_blank" rel="noopener" href="https://github.com/tpoorten/dotPlotly/tree/master/example">https://github.com/tpoorten/dotPlotly/tree/master/example</a></p>
<p>主要是R语言：<a target="_blank" rel="noopener" href="https://github.com/tpoorten/dotPlotly/blob/master/mummerCoordsDotPlotly.R">https://github.com/tpoorten/dotPlotly/blob/master/mummerCoordsDotPlotly.R</a></p>
<p>补充：如何理解上述作图过程中的几个文件呢？</p>
<p>delta文件（在<a target="_blank" rel="noopener" href="http://mummer.sourceforge.net/manual/#HumanCompare">http://mummer.sourceforge.net/manual/#HumanCompare</a> 页面的The delta format section下面有详细解释）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">/path/Gmax_508_v4.0.softmasked.fa /path/G.max-mitochondrion.fasta</span><br><span class="line"> NUCMER</span><br><span class="line">   Gm13 ChrM 45225048 402558     # 匹配区段所在的染色体</span><br><span class="line">   21773198 21776178 114322 117301 3 3 0   # 前四个数字：两个染色体上面的区段坐标。后三个数字：错误数(不相同碱基+indel碱基数)，相似错误(非正匹配得分) 终止密码子(NUCMER为0)。 如果结束大于起始，表示在负链。英文版解释：The three digits following the location coordinates are the number of errors (non-identities + indels), similarity errors (non-positive match scores), and stop codons (does not apply to DNA alignments, will be &quot;0&quot;)</span><br><span class="line">   8  # query的8位置有“缺失”？？？如果前面有负号，则表示此处为“插入”？相对ref序列而言。</span><br><span class="line">   0  # 表示当前匹配结束</span><br><span class="line">   Gm20 ChrM 47846026 402558</span><br><span class="line">   14299558 14301282 188578 190302 3 3 0</span><br><span class="line">   0 </span><br><span class="line">   Gm17 ChrM 41740656 402558</span><br><span class="line">   23669908 23674616 270387 265667 78 78 0</span><br><span class="line">   -499</span><br><span class="line">   -1   # 这连续的“-1”是什么意思？</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -106</span><br><span class="line">   -2</span><br><span class="line">   -1</span><br><span class="line">   -1</span><br><span class="line">   -2</span><br><span class="line">   -2182</span><br><span class="line">   1263</span><br><span class="line">   -230</span><br><span class="line">   0 </span><br><span class="line">(output continues ...)</span><br></pre></td></tr></table></figure>
<p><strong>七、输出适合肉眼观察的Tabular格式比对结果</strong><br />
上面的delta文件不便于我们肉眼阅读，可用show-coords -r ${prefix}.delta &gt; ${prefix}.coords命令之后得到更适合阅读的格式（coords文件）。<br />
有点类似BLAST的“-outfmt 6”参数输出的Tabular格式<br />
如下，以ref序列的染色体排序的。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">   [S1]     [E1]       [S2]     [E2]    [LEN 1]  [LEN 2]    [% IDY]   [TAGS]</span><br><span class="line">41311147 41314402     161362   158107       3256     3256      99.91   Gm08 ChrM</span><br><span class="line">24034371 24036869     210827   213325       2499     2499      99.88   Gm12 ChrM</span><br><span class="line">24038545 24043058      13087     8578       4514     4510      99.78   Gm12 ChrM</span><br><span class="line">21773198 21776178     114322   117301       2981     2980      99.90   Gm13 ChrM</span><br><span class="line">23669908 23674616     270387   265667       4709     4721      98.35   Gm17 ChrM</span><br><span class="line">14299558 14301282     188578   190302       1725     1725      99.83   Gm20 ChrM</span><br></pre></td></tr></table></figure>
<p><strong>八、输出适合肉眼观察的Pairwise格式比对结果</strong><br />
可指定特定的染色体或scaffold用于输出到屏幕，输出格式类似BLAST的“-outfmt 1”参数。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">show-aligns seqA_vs_seqB_genome_i89_l1000_1.delta Chrxxx Chryyy | less</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">============================================================</span><br><span class="line">-- Alignments between xxx and yyy</span><br><span class="line"></span><br><span class="line">-- BEGIN alignment [ +1 1 - 245034 | -1 94962171 - 94717077 ]</span><br><span class="line"></span><br><span class="line">                  10|       20|       30|       40|</span><br><span class="line">1          cccaaccccgagccctaaaccctaaaccccaaacccccaaccctgagcc</span><br><span class="line">94962171   cccaaccccgagccctaaaccctaaaccccaaacccccaaccctgagcc</span><br><span class="line">                                                            </span><br><span class="line">                  10|       20|       30|       40|</span><br><span class="line">50         tcgagccctaagccttgagcccgagccctaagccccgatccctaagccc</span><br><span class="line">94962122   tcgagccctaagccttgagcccgagccctaagccccgatccctaagccc</span><br><span class="line">                                                            </span><br><span class="line">                  10|       20|       30|       40|</span><br><span class="line">99         tgaaccccgaaccctcaaccccgaaccccaaaccctcaaccccaaaccc</span><br><span class="line">94962073   tgaaccccgaaccctcaaccccgaaccccaaaccctcaaccccaaaccc</span><br></pre></td></tr></table></figure>
<p>参考：</p>
<ol>
<li>MUMMER 两个基因组间比较 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/c12f2a117892">https://www.jianshu.com/p/c12f2a117892</a></li>
<li>MUMMER 两个基因组间比较 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/9701c9add82d">https://www.jianshu.com/p/9701c9add82d</a></li>
<li>最详细的MUMMER中文文档 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/2e184e5c15b7">https://www.jianshu.com/p/2e184e5c15b7</a></li>
<li>How to understand MUMMER delta format？<a target="_blank" rel="noopener" href="http://mummer.sourceforge.net/manual/#HumanCompare">http://mummer.sourceforge.net/manual/#HumanCompare</a></li>
<li>用R绘制点状图：<a target="_blank" rel="noopener" href="https://github.com/tpoorten/dotPlotly">https://github.com/tpoorten/dotPlotly</a></li>
<li>用R绘制mummer的结果图 <a target="_blank" rel="noopener" href="https://blog.csdn.net/u012110870/article/details/102492711?utm%5C_medium=distribute.pc%5C_relevant.none-task-blog-title-3&amp;spm=1001.2101.3001.4242">https://blog.csdn.net/u012110870/article/details/102492711?utm\_medium=distribute.pc\_relevant.none-task-blog-title-3&amp;spm=1001.2101.3001.4242</a></li>
</ol>

    </div>

    
    
    

      <footer class="post-footer">

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2020/12/05/Linux-shell-perl-mute-multiple-lines/" rel="prev" title="shell脚本和perl脚本的段注释">
      <i class="fa fa-chevron-left"></i> shell脚本和perl脚本的段注释
    </a></div>
      <div class="post-nav-item">
    <a href="/2020/12/06/Bioinfo-HiC-pro-installation/" rel="next" title="安装HiC-Pro （HiC数据预处理工具）与绘制HiC染色体互作图谱">
      安装HiC-Pro （HiC数据预处理工具）与绘制HiC染色体互作图谱 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">174</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">10</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
