<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta property="og:type" content="website">
<meta property="og:title" content="FENG&#39;s Notes">
<meta property="og:url" content="http://example.com/page/5/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/page/5/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/03/10/Bioinfo-Extracting-sequence-by-names/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/03/10/Bioinfo-Extracting-sequence-by-names/" class="post-title-link" itemprop="url">Bioinfo | 依据序列名字列表从 fasta 文件提取序列</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-03-10 22:37:33" itemprop="dateCreated datePublished" datetime="2021-03-10T22:37:33+08:00">2021-03-10</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-14 18:28:12" itemprop="dateModified" datetime="2022-07-14T18:28:12+08:00">2022-07-14</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>依据序列名称从fasta文件提取序列是一个很常见的操作。seqkit 这个工具可以很方便地实现。<br />
比如 input.fa 包含若干条序列，名称分别为 1、2、3……100 。而 name list 文件包含 40——49 这十个ID，那么通过下面的代码就可以提取编号 40——49 对应的序列。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seqkit grep -r -f name_list.txt input.fa -o out.fa</span><br></pre></td></tr></table></figure>
<p>上面只是一种简单的情况，而大部分原始 fasta 文件中，序列名除了 ID，还可能会包含其他信息（比如基因注释）。当输入的 name list 文件只有 ID，默认的“名称完全匹配”就提取不了，就需要加上 <code>-r</code> 参数来申明使用正则匹配（regular expression）进行提取。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seqkit grep -r -f name_list.txt input.fa -o out.fa</span><br></pre></td></tr></table></figure>
<p>另一种情况就是，对于拟南芥，每个基因有多个 isoforms，如果我们只需要依据基因 ID 输出一个 isoform，那么可以通过 <code>--delete-matched</code> 参数来控制，即当某个 ID 出现了一次匹配，则立即停止继续检索该 ID，进入下一个 ID 检索。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seqkit grep -r --delete-matched -f Arabidopsis_name_list.txt Araport11_pep_20160703.pep.fa -o Arabidopsis_pep_list_nonrebundant.fa</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/03/05/Proteomics-Serotonylation-A-novel-histone-H3-marker/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/03/05/Proteomics-Serotonylation-A-novel-histone-H3-marker/" class="post-title-link" itemprop="url">Proteomics | 组蛋白五羟色胺化修饰 H3Q5ser 及其生物学功能</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-03-05 10:43:58 / Modified: 13:55:58" itemprop="dateCreated datePublished" datetime="2021-03-05T10:43:58+08:00">2021-03-05</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Proteomics/" itemprop="url" rel="index"><span itemprop="name">Proteomics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>五羟色胺是人体重要的一种神经递质与激素，在中枢神经系统中控制着认知、学习、感情、情绪等脑神经活动；在外周神经系统中，五羟色胺控制着生殖、代谢、血管收缩、骨骼发育等生理功能。因此，五羟色胺系统一直是药物研究与开发的热点，作用于该系统的药物广泛用于抗精神分裂症、偏头痛、抗呕吐及肥胖症等疾病。</p>
<p><img src="https://cdn.shopify.com/s/files/1/0511/1981/6889/articles/a58d3fb0a0a9ce4fa1c5d180b499130c_1440x.png" alt="" /></p>
<p>五羟色胺，又称血清素 (serotonin)，是一类神经递质，对情绪等神经生物学事件有重要的调控作用。五羟色胺广泛存在于哺乳动物组织中，除中枢神经组织外，其主要外周产生部位是胃肠道，约 90% 合成和分布于肠嗜铬细胞。Ian Maze 课题组曾于 2019 年在 Nature 杂志发表论文，首次报道了 H3Q5ser 修饰的存在，以及“H3K4me3Q5ser”修饰组合对基因表达和神经发育的影响（Nature，2019，567：535–539）。该工作把五羟色胺这一内源生物活性小分子与表观遗传修饰调控联系了起来，启示了营养代谢除为细胞活动提供物质、能量和信号之外的“信息”功能。由于 H3Q5 紧邻 H3K4 位点，因此 H3Q5ser 修饰很可能对 H3K4me3 修饰的催化和识别产生影响。</p>
<p><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41392-019-0048-7/MediaObjects/41392_2019_48_Fig1_HTML.png" alt="Regulation model of serotonin. a Serotonin, a neurotransmitter, functions through the serotonergic signaling pathway. b MLL and SET1 mediate H3K4me3 methylation and regulate gene expression. c The H3K4me3Q5ser double modification enhances the interaction of histone with certain H3K4me3 ‘reader’ proteins, such as TFIID, and then reinforces permissive patterns of gene expression" /></p>
<p>2021年2月2日，清华大学医学院李海涛课题组携手普林斯顿大学 Tom Muir 课题组和西奈山医学院 Ian Maze 课题组于《美国科学院院刊》（PNAS）在线发表题为“Histone H3Q5 serotonylation stabilizes H3K4 methylation and potentiates its readout”（组蛋白 H3Q5 五羟色胺化修饰稳定 H3K4 甲基化修饰并强化其识别）的研究论文，报道了组蛋白 H3Q5 五羟色胺化修饰（H3Q5ser）对于转录起始标志 H3K4 三甲基化（H3K4me3）修饰的影响。该研究发现 H3Q5Ser 不影响邻近 H3K4me3 的酶促产生，可促进 H3K4me3 下游阅读器的识别，但抑制了 H3K4me3 的酶促消除，从而稳定 H3K4me3 修饰并强化其基因活化功能。</p>
<p><img src="http://www.med.tsinghua.edu.cn/__local/F/EE/A8/7E72FE89E5984EFB6A2E622C1AE_39F33F28_37F347.png" alt="组蛋白H3Q5ser稳定H3K4me3修饰并强化其功能" /></p>
<p>该研究探究 H3Q5ser 修饰与 H3K4me3 修饰的交叉会话（crosstalk）机制。首先，研究人员利用表面等离子体共振成像（SPRi）技术，对目前已知的代表性 H3K4me3 阅读器（如BPTF，CHD1，DIDO1，ING2，KDM5A，KDM4A，MLL5，PHF13，PHF2，RAG2，SET3，SGF29，SPIN1，TAF3，和ZCWPW1)，进行了基于蛋白质芯片阵列的互作表征。研究发现，组蛋白 H3Q5ser 修饰并不会打破这些下游效应因子对于 H3K4me3 修饰的识别。其中，对于 TAF3 的 PHD 结构域来说，无论 K4 的甲基化状态如何（K4me0，K4me1，K4me2，K4me3），H3Q5ser 修饰都可以增强结合强度，甚至达到 9 倍。如果将五羟色胺化修饰替换五羟色胺的类似物（如 H3Q5（tryptamine），H3Q5（5-methoxytryptamine），H3Q5（7-azatryptamine），H3Q5（6-hydroxytryptamine），则会不同程度地削弱与 TAF3 的结合。这表明组蛋白同 TAF3 结合强度的增强是五羟色胺修饰依赖的。</p>
<p>对于 H3K4me3 修饰因子 MLL1 甲基转移酶复合物来讲，H3Q5ser 修饰不会对其活性产生明显影响。更进一步的酶反应动力学分析发现，H3Q5ser 修饰对于 MLL1 催化 H3K4me0 到 H3K4me1 没有影响，对于催化 H3K4me1 到 H3K4me2 有轻微的抑制作用。因此，总体来讲H3Q5ser修饰对于MLL1复合物的酶活没有产生明显的影响。</p>
<p>有趣的是，对于 H3K4me3 的酶促消除而言，由于存在空间位阻效应，H3Q5ser 修饰可以完全打破 H3K4 去甲基化酶 KDM5 和 LSD1 的活性。在多肽和重构染色质水平的实验都表明 H3Q5ser 修饰可以使 LSD1 及 KDM5 无法催化 H3K4me3 的去甲基化反应。进一步的重构染色质水平的酶活实验表明 H3Q5ser 修饰影响了 KDM5 催化结构域的活性而非 PHD 结构域的结合能力。值得一提的是，KDM5 家族成员通过负调控增强子与启动子区 H3K4me3 水平，发挥着转录抑制分子功能，在肿瘤转录异质性调控、肿瘤免疫逃逸、肠道微生物组稳态调节、神经系统发育等过程中发挥着重要作用。本研究所揭示的 H3Q5ser 修饰对 KDM5 的拮抗作用对于进一步理解 KDM5 去甲基化酶的生理功能与调控提供了新视角。</p>
<p>组蛋白修饰交叉会话是一类重要的表观调控机制，在表观景观建立和基因开关切换等细胞进程中扮演着重要角色，彰显了表观调控的层次性和精密性。李海涛课题组与 Josefowicz 课题组合作，曾在去年报道组蛋白 H3.3S31ph 修饰同 H3.3K36me3 之间的交叉会话（Nature, 2020, 583：852-857），揭示 H3.3S31ph 通过强化 H3.3K36me3 的转录延伸功能，实现刺激诱导型基因的快速转录延伸。在本工作中，研究人员揭示了H3Q5ser对转录起始标志性修饰H3K4me3的调制作用，即在H3Q5ser存在情况下，H3K4me3无法被“擦除器”（eraser）有效去除，而H3K4me3“阅读器”（reader）如 TAF3 却可以被更有效的招募，提示 H3Q5ser 可以通过强化 H3K4me3 的功能，实现 H3K4me3 靶基因的转录起始活化。</p>
<p><font color=blue>越来越多的研究表明，类似 H3K4me3 和 H3K36me3 等甲基化修饰单独不足以代表基因的激活表达，通过产生 H3Q5ser 和 H3.3S31ph 等邻近修饰而拮抗负调控因子的沉默作用，对于完全释放 H3K4me3 和 H3K36me3 修饰的转录激活功能发挥着重要正向调控作用。</font></p>
<p>参考资料</p>
<ol>
<li><a target="_blank" rel="noopener" href="https://biokplus.com/blogs/news/serotonin-the-happy-hormone-produced-in-our-gut">Serotonin: The Happy Hormone Produced In Our Gut</a></li>
<li><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41586-019-1024-7">2019 Nature: Histone serotonylation is a permissive modification that enhances TFIID binding to H3K4me3</a></li>
<li><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/59800502">2019 知乎专栏：Nature重磅+点评丨你的情绪信号或可印记到组蛋白上——首次报道组蛋白五羟色胺化修饰及其生物学功能</a></li>
<li><a target="_blank" rel="noopener" href="http://www.med.tsinghua.edu.cn/info/1028/2827.htm">2021 清华大学：李海涛团队揭示五羟色胺活化基因表达的组蛋白修饰会话机制</a></li>
<li><a target="_blank" rel="noopener" href="https://www.thepaper.cn/newsDetail_forward_11197874">2021 澎湃新闻：李海涛团队揭示五羟色胺活化基因表达的组蛋白修饰会话机制</a></li>
</ol>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/03/04/Proteomics-What-the-differences-between-DDA-and-DIA-in-mass-spectrometer/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/03/04/Proteomics-What-the-differences-between-DDA-and-DIA-in-mass-spectrometer/" class="post-title-link" itemprop="url">Proteomics | 质谱分析中DDA与DIA有什么不同？</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-03-04 17:16:31 / Modified: 19:02:28" itemprop="dateCreated datePublished" datetime="2021-03-04T17:16:31+08:00">2021-03-04</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Proteomics/" itemprop="url" rel="index"><span itemprop="name">Proteomics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>液相色谱与串联质谱（LC-MS/MS）相结合的技术成为蛋白质组研究的主要研究手段。在自下而上的蛋白质组学中，蛋白质通过蛋白水解酶消化成肽段，这些肽段通过高效液相分离后经电喷雾电离源（ESI）加上电荷，发射到质谱仪中进行一级质谱（MS）和二级质谱（MS/MS）的分析。一级质谱由肽段离子的质荷比和信号强度组成，二级质谱由肽段碎裂后的碎片离子质荷比和信号强度组成。</p>
<p>DDA 为数据依赖性采集方式，选择一级质谱中特定数量的肽段分子（如信号强度最强的前 10 个离子）进行高能碰撞碎裂，产生的碎片离子送入二级质谱检测。</p>
<p>DIA 为数据非依赖性采集方式，随着时间连续设置一定范围的质荷比窗口，对将通过窗口的全部肽段离子进行碎裂并用二级质谱检测碎片离子的信号。相比之下，DIA 模式不涉及到对肽段离子的限制性筛选，<u>定量准确性和重复性都要优于 DDA 模式</u>。但 DIA 模式产生的碎片离子谱图过于复杂，丢失了肽段及碎片离子的对应关系，对其结果的解析较为困难。目前的方法是通过建立样本的参考库（如同一样本的 DDA 数据），对 DIA 的混合数据进行去卷积分析来鉴定和定量肽段。</p>
<p>DDA 模式对肽段离子的选择具有限制性和随机性，存在较多的数据信息缺失，而 DIA 模式的目标则是获得完整的数据，实现蛋白质的深度覆盖和精准定量。此外在复杂样品中，二级信号的定量比一级信号更为灵敏（例如一级质谱信号可能存在相似质荷比离子的干扰）。DDA模 式为一级质谱信号定量，而 DIA 模式为二级质谱信号定量，因此 DIA 模式的定量准确性和重现性都更好。</p>
<p>DDA 模式下肽段分子和其二级碎片离子有着对应关系，通过对已知数据库的搜索匹配，即可得到样本中的肽段及相关蛋白质信息。而 DIA 模式下肽段离子与其碎片离子之间的直接关系丢失（DIA 光谱中的碎片离子可能是由多个肽段离子），复杂的谱图仍然给后续的数据分析和统计学检验带来很多挑战。对此目前已有多个团队开发出 DIA 分析的软件，例如 OpenSWATH 和 Skyline，这些软件通过建立样本的参考库，对色谱峰进行抽提并分析。此外也有研发团队提出不需要参考库的数据分析方法 DIA-Umpire，通过对肽段离子和碎片离子的匹配合成产生虚拟谱图进而搜库分析。因此，DIA 技术非常适合一次性获取数百上千种蛋白质更为全面信息（高覆盖率，高准确度，高重复性）的研究。</p>
<p><img src="https://pic1.zhimg.com/80/v2-42e5990e613ffb77de82db5ad6aca80c_720w.jpg" alt="DIA与DDA对比" /></p>
<p><img src="http://wayen.ezweb1-3.35.com/htdocs/d/2/wayen/resource/rich/99c6d7e97d73a0c753a8e5ed2f3c4756.png" alt="DIA技术原理图（来源：LCGC North America.Volume 35, Issue 10, pg 756–759）" /></p>
<p><img src="https://f1000researchdata.s3.amazonaws.com/manuscripts/7580/6631ee4c-6910-44ec-a8cc-c8bdb77fc401_figure1.gif" alt="" /></p>
<p>参考资料</p>
<ol>
<li><a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/news/131977">2018 一篇SCI来看蛋白DIA技术如何完胜DDA</a> （这里面有个DIA与DDA对比的图，也可帮助理解，但是因网站设置导致图片链接过不来。）</li>
<li><a target="_blank" rel="noopener" href="http://html.rhhz.net/SWJSTB/html/2017-9-73.htm">2017 《生物技术通报》基于质谱的定量蛋白质组学技术发展现状</a></li>
<li><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/47994226">2018 知乎专栏：学会DIA技术</a></li>
<li><a target="_blank" rel="noopener" href="http://www.wayenbiotech.com/danbaizhizuxue-316395-151150-item-683817.html">简介：DIA蛋白质组学技术</a></li>
<li><a target="_blank" rel="noopener" href="https://chemistry.fudan.edu.cn/a9/84/c21871a240004/page.htm">2020 复旦化学系：基于深度学习的DIA蛋白质组分析新方法</a></li>
</ol>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/03/04/Proteomics-Protein-Persulfidation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/03/04/Proteomics-Protein-Persulfidation/" class="post-title-link" itemprop="url">Proteomics | 蛋白质硫巯基化修饰</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2021-03-04 17:02:00 / Modified: 17:14:19" itemprop="dateCreated datePublished" datetime="2021-03-04T17:02:00+08:00">2021-03-04</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Proteomics/" itemprop="url" rel="index"><span itemprop="name">Proteomics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>随着对硫化氢（hydrogen sulfide，H<sub>2</sub>S）生理效应的研究，蛋白质硫巯基化（S-sulfhydration，Protein Persulfidation）修饰已进入人们的视野。已知依赖于 H<sub>2</sub>S 的蛋白质硫巯基化是继磷酸化（phosphorylation）、泛素化（ubiquitylation）、乙酰化（acetylation）和 S-亚硝基化（S-nitrosylation）等之后的一种新的蛋白质翻译后修饰方式。对动物的研究表明，蛋白质硫巯基化修饰通过影响蛋白质活性和功能，从而在细胞内信号通路中发挥重要的调控作用。最近的研究结果提示，硫巯基化修饰还参与调节植物新陈代谢和形态建成。硫巯基化修饰也可能参与植物细胞信号转导。</p>
<p>硫化氢（hydrogen sulfide, H<sub>2</sub>S）是一种无色、具有臭鸡蛋气味的气体。在人们的传统观念中，硫化氢一直被认为是一种有毒有害的气体。然而随着科技发展，关于生物体内产生硫化氢的现象被陆续报道，这暗示着硫化氢在某些生理代谢过程中可能发挥着重要作用。硫化氢可自由通过生物膜，以信号分子的角色参与信号传递。最近的发现表明，H<sub>2</sub>S 具有促使蛋白质 Cys 残基的硫醇基团（R-SH）转变为过硫化基团（R-SSH）的能力，这种翻译后修饰作用被称为 蛋白质过硫化/硫巯基化修饰（persulfidation）。蛋白质过硫化修饰是一种基于氧化还原状态的修饰，然而在高等植物中这一翻译后修饰的机制仍然知之甚少。</p>
<p>当植物受到干旱、高低温、高盐、重金属等非生物胁迫刺激时，体内的硫化氢含量会显著提高，暗示硫化氢参与植物非生物胁迫响应过程。研究表明，在这一过程中硫化氢与多种信号分子，如活性氧（ROS）、一氧化氮（NO）、脱落酸（ABA）、乙烯等信号交叉对话，共同调节植物对胁迫的耐受性。</p>
<p>脱落酸（ABA）作为一种重要的植物激素，在种子的成熟和休眠，气孔关闭和植物生长中起关键作用。同时，ABA也是增强植物对非生物胁迫如干旱，盐碱和寒冷的适应性的核心调节因子，而这些非生物胁迫是导致全球农作物减产的主要原因，因此研究植物体内ABA信号转导途径具有重要的实践意义。ABA促进气孔关闭并抑制气孔开放，从而在植物缺水时保留水分。SnRK2.6蛋白在植物气孔的保卫细胞中表达，是ABA信号通路中调节植物气孔开合的关键调节因子。硫化氢（H2S）是一种水溶性的气体信号分子，在调节植物对环境胁迫和生长的适应性方面发挥重要功能。Cys残基的翻译后修饰（Cys-SH基团转换为Cys-SSH基团）引起的过硫基团的形成直接受硫化氢调节，被称为S-硫巯基化修饰作用。研究首先在SnRK2.6中发现了两个S-硫巯基化修饰位点，并且发现这两个Cys残基暴露在SnRK2.6表面并且与激酶的催化环和关键的磷酸化位点相邻。研究表明当这两个Cys残基被S-硫巯基化修饰后，促进了SnRK2.6的活性及其与ABA信号下游的转录因子的相互作用。当将SnRK2.6中的Cys131，Cys137或两者均被丝氨酸部分或者全部取代时：SnRK2.6C131S和SnRK2.6C137S或SnRK2.6C131SC137S突变基因转入到ost1-3突变体中，这些点突变的基因无法恢复ost1-3突变体表型，显示出对ABA和H2S诱导的气孔关闭和Ca2+流入的敏感性降低，并增加了水分流失和抗旱性下降。此外，该研究还表明ABA诱导保卫细胞中的DES1酶催化产生硫化氢，随后硫化氢又通过SnRK2.6的过硫化反应过来正调控ABA信号传导，并影响钙离子流，最终调控气孔运动。因此该研究提出了一种新的调控ABA信号通路的机制，其中硫化氢通过S-硫巯基化SnRK2.6在保卫细胞中正调控ABA信号传导。</p>
<p><img src="https://news.nwafu.edu.cn/images/content/2020-02/20200219193146422506.png" alt="" /></p>
<p>参考资料</p>
<ol>
<li><a target="_blank" rel="noopener" href="https://www.sohu.com/a/426629232_732029">2020 南农综述：气体信号分子——硫化氢在植物适应非生物胁迫中的生理功能及作用机制 </a></li>
<li><a target="_blank" rel="noopener" href="https://news.nwafu.edu.cn/xnxw/95411.htm">2020 西农文章：硫化氢介导的S-硫巯基化修饰促进气孔闭合关键调控蛋白SnRK2.6/OST1的活性从而参与调控脱落酸信号诱导植物气孔闭合的分子机制</a></li>
<li><a target="_blank" rel="noopener" href="http://cjbmb.bjmu.edu.cn/CN/abstract/abstract24425.shtml">2018 生物化学与分子生物学报：《硫巯基化：一种新的蛋白质翻译后修饰》</a></li>
</ol>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/03/03/Proteomics-Common-reagents-and-their-functions-in-LC-MS-MS-proteomics/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/03/03/Proteomics-Common-reagents-and-their-functions-in-LC-MS-MS-proteomics/" class="post-title-link" itemprop="url">Proteomics | Common reagents and their functions in LC-MS/MS proteomics</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-03-03 23:08:17" itemprop="dateCreated datePublished" datetime="2021-03-03T23:08:17+08:00">2021-03-03</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-03-04 14:04:42" itemprop="dateModified" datetime="2021-03-04T14:04:42+08:00">2021-03-04</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Proteomics/" itemprop="url" rel="index"><span itemprop="name">Proteomics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>蛋白质组学实验 LC-MS/MS “液相色谱法-质谱联用”中，常见的试剂的功能是什么？</p>
<h3 id="acetone-丙酮"><a class="markdownIt-Anchor" href="#acetone-丙酮"></a> Acetone 丙酮</h3>
<p>使用场景：提取组蛋白的过程中，当用 TCA 三氯乙酸沉淀组蛋白之后，加入预冷的丙酮清洗组蛋白沉淀。</p>
<p>丙酮是最简单的酮，化学式 CH<sub>3</sub>COCH<sub>3</sub>。由于甲醇、乙醇、丙酮等有机溶剂加入水中使容积介电常数降低，增加了相反电荷的吸引力，另一方面因为这些有机溶剂是强亲水试剂，争夺蛋白质分子表面的水化水，破坏蛋白质胶体分子表面的水化层而使分子聚集沉淀。当蛋白沉淀后，立即分离蛋白和丙酮并复溶蛋白完成提取。丙酮在提取蛋白质中的主要作用是通过有机溶剂破坏蛋白质的水化层，降低介电常数 增强带电蛋白质分子的相互作用，促进蛋白颗粒的聚集沉淀。<br />
丙酮要求低温操作且尽量缩短时间，避免蛋白变性。丙酮容易挥发。在提取大分子蛋白质时，TCA使蛋白沉淀之后，使用丙酮尽量将TCA抽提干净，</p>
<h3 id="acetonitrile-乙腈"><a class="markdownIt-Anchor" href="#acetonitrile-乙腈"></a> Acetonitrile 乙腈</h3>
<p>使用场景一：HPLC 实验中，ACN 可以与水分别作为流动相 B 与 A，用来洗脱不同种类的蛋白分子。<br />
使用场景二：SDS-PAGE 胶，切下目的条带之后，加入ACN，无色透明的胶块会迅速失水，皱缩成白色不透明状。</p>
<p>乙腈（英语：Acetonitrile，又称氰基甲烷），化学式 CH<sub>3</sub>CN，英文缩写：ACN 或 AeCN。乙腈是无色的液体，是最简单的有机腈，并广泛用作极性非质子溶剂。对于 HPLC 级溶剂（我们在 HPLC 分析中应始终使用 HPLC 级溶液），<u>乙腈的吸光度在这两种溶剂中最低</u>，非常适合低紫外光分析。作为比较，甲醇在 205–210 nm 左右有较高的紫外光吸收值，在非常低的紫外光范围内略有限制。注意乙腈可以燃烧，并且有毒性。</p>
<h3 id="ammonium-acetate"><a class="markdownIt-Anchor" href="#ammonium-acetate"></a> Ammonium acetate</h3>
<p>乙酸铵是一个有机盐，分子式为CH<sub>3</sub>COONH<sub>4</sub>。</p>
<h3 id="ammonium-sulfate"><a class="markdownIt-Anchor" href="#ammonium-sulfate"></a> Ammonium sulfate</h3>
<p>使用场景：在配制考马斯亮蓝染色液的时候，需要准备2% (wt/vol) phosphoric acid and 20% (wt/vol) ammonium sulfate 溶液。</p>
<p>硫酸铵，简称硫铵，又称硫酸铔，化学式为(NH<sub>4</sub>)<sub>2</sub>SO<sub>4</sub>，无色斜方晶体，易溶于水，水溶液为酸性。</p>
<h3 id="ammonium-hydrogen-carbonate-abc"><a class="markdownIt-Anchor" href="#ammonium-hydrogen-carbonate-abc"></a> Ammonium hydrogen carbonate （ABC）</h3>
<p>使用场景： In-gel digestion 的 buffer，即 100 mM ABC</p>
<p>碳酸氢铵 (Ammonium bicarbonate), 碳酸氢铵是氨的碳酸盐，分子式为NH<sub>4</sub>HCO<sub>3</sub>。用 Trypsin 酶解蛋白质时，碳酸氢氨的浓度应该是多少？酶解时碳酸氢氨的浓度一般在 10 ～ 50 mM 之间。碳酸氢氨的作用主要是提供一个碱性的水解环境，因为 Typsin 的活力在 pH8 ～ 9 之间最高。最常用的碳酸氢氨浓度有 20 mM ， 40 mM 。盐浓度过大时，后续的冻干过程中会有较多的盐析出；盐浓度过小时，则不能有效的提供碱性 pH 的水解环境（尤其是在样品的 pH 值低，而且体积大时）。</p>
<h3 id="ammonium-persulfate"><a class="markdownIt-Anchor" href="#ammonium-persulfate"></a> Ammonium persulfate</h3>
<p>过硫酸铵（APS）是一种白色晶体，分子式为(NH<sub>4</sub>)<sub>2</sub>S<sub>2</sub>O<sub>8</sub>，具有强氧化性和腐蚀性，易溶于水。其干品具有良好的稳定性，保存在防潮箱中。聚丙烯酰胺（PAGE）是由丙烯酰胺聚合形成的高分子，这个聚合反应是自由基聚合，需要有引发剂产生自由基将反应引发过硫酸铵就是引发剂，而 TEMED 可以催化过硫酸铵产生自由基，从而加速丙烯酰胺凝胶的聚合。过硫酸铵配制成 10%溶液后，应当-20℃保存。同时应尽量减少室温存放时间，以防失效。保存不能超过两周。最好现配现用。</p>
<h3 id="tca-三氯乙酸"><a class="markdownIt-Anchor" href="#tca-三氯乙酸"></a> TCA 三氯乙酸</h3>
<p>使用场景：用硫酸溶液破裂细胞核，释放组蛋白之后，加入 TCA 使组蛋白沉淀，为什么用 TCA 呢？</p>
<p>TCA 化学式 CCl<sub>3</sub>COOH，其作用主要是在酸性条件下可与蛋白质形成不溶性盐。同时作为蛋白质变性剂使蛋白质构象发生改变，暴露出较多的疏水性基团，使之聚集沉淀。但是 TCA 可能渗入蛋白质分子内部而使之较难被完全除去，在电泳前样品加热处理时可能使蛋白质结构发生酸水解而形成碎片，所以在提取大分子蛋白质时要使用丙酮（Acetone）尽量将 TCA 抽提干净。</p>
<h3 id="acetic-acid-乙酸"><a class="markdownIt-Anchor" href="#acetic-acid-乙酸"></a> Acetic acid 乙酸</h3>
<p>使用场景：SDS-PAGE胶脱色液，即考马斯亮蓝染色脱色液，配方为：40%乙醇，10%乙酸，50%蒸馏水。也可以使用其它适当的脱色液。</p>
<p>乙酸也叫做醋酸，化学式 CH<sub>3</sub>COOH，甲醇是为了更容易洗脱考马斯亮蓝，加得越多洗脱越容易，结合在蛋白上的染料也容易洗下来。冰乙酸是固定蛋白的，防止蛋白扩散。两者作用完全不同。</p>
<h3 id="tfa-trifluoroacetic-acid-三氟乙酸"><a class="markdownIt-Anchor" href="#tfa-trifluoroacetic-acid-三氟乙酸"></a> TFA (trifluoroacetic acid ) 三氟乙酸</h3>
<p>使用场景：组蛋白的 HPLC 纯化过程中，流动相 A 和 B 里面都会添加 0.1% TFA。</p>
<p>三氟乙酸化学式 CF<sub>3</sub>COOH，在反相色谱分离多肽和蛋白质的实验中，使用三氟乙酸 (TFA) 作为离子对试剂是常见的手段。流动相中的三氟乙酸通过与疏水键合相和残留的极性表面以多种模式相互作用，来改善峰形、克服峰展宽和拖尾问题。三氟乙酸与多肽上的正电荷及极性基团相结合以减少极性保留，并把多肽带回到疏水的反相表面。以同样的方式，三氟乙酸屏蔽了固定相上残留的极性表面。三氟乙酸的行为可以理解为它滞留在反相固定相的表面，同时与多肽及柱床作用。三氟乙酸优于其他离子修饰剂的原因是它容易挥发，可以方便地从制备样品中除去。另一方面，三氟乙酸的紫外最大吸收峰低于 200 nm，对多肽在低波长处的检测干扰很小。改变三氟乙酸的浓度，可以细微地调整多肽在反相色谱上的选择性。这一影响对于优化分离条件、增大复杂色谱分析（如多肽的指纹图谱）的信息量是非常有益的。三氟乙酸添加在流动相中的浓度一般为 0.1%，在这个浓度下，大部分的反相色谱柱都可以产生良好的峰形，当三氟乙酸浓度大大低于这个水平时，峰的展宽和拖尾就变得十分明显。</p>
<h3 id="fa-formic-acid-甲酸"><a class="markdownIt-Anchor" href="#fa-formic-acid-甲酸"></a> FA (Formic acid) 甲酸</h3>
<p>使用场景：经过酶解和纯化的短肽，溶解于 0.1% FA 然后在 Orbitrap 质谱仪进行 LC-MS/MS 分析。</p>
<p>甲酸又称作蚁酸，化学式为HCOOH。在反相色谱分离多肽和蛋白质的实验中，使用三氟乙酸(TFA) 作为离子对试剂是常见的手段，流动相中的三氟乙酸通过与疏水键合相和残留的极性表面以多种模式相互作用，来改善峰形、克服峰展宽和拖尾问题。同时反相色谱与电喷雾质谱联用已成为多肽和蛋白质的分子量测定和结构分析的重要工具，但是含有三氟乙酸的流动相对质谱离子的产生具有抑制作用，一定程度上降低了液质联用技术的灵敏性和分析可靠性。目前常用的方法是降低流动相中 TFA 浓度， 达到降低这种抑制作用，但是也会造成色谱分析质量的降低；而如果用甲酸代替 TFA, 会造成色谱分离柱效降低，达不到理想的分离效果。本文采用双三元液相色谱系统 (DGLC)，充分利用双泵的优势，将两个泵分别作为流动相的输液泵和柱后溶液的传输泵，实现与质谱的全自动连接。采用 TFA-Fix 技术或者抑制器技术降低流动相中 TFA 对 MS 的离子产生的抑制作用，尤其是柱后抑制器技术大大提高了多肽样品在质谱上的灵敏度。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/03/03/R-Using-ggrepel-to-repel-overlapping-text-labels-in-ggplot2/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/03/03/R-Using-ggrepel-to-repel-overlapping-text-labels-in-ggplot2/" class="post-title-link" itemprop="url">R | Using ggrepel to repel overlapping text labels in ggplot2</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-03-03 16:33:13" itemprop="dateCreated datePublished" datetime="2021-03-03T16:33:13+08:00">2021-03-03</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-03-04 16:57:18" itemprop="dateModified" datetime="2021-03-04T16:57:18+08:00">2021-03-04</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>在使用ggplot画图的过程中，如果加上文字标签，比较常见的现象是文字标签之间会出现重叠，影响展示效果。如下图所示。</p>
<p><img src="https://miro.medium.com/max/875/1*CRx8s9NXxVLQG5xXP1Z_gQ.png" alt="" /></p>
<p>通过ggrepel则可以对文字标签进行优化。<br />
如下方案例中使用的 geom_text_repel 和 geom_text_repel 两个函数。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)</span><br><span class="line">#使用数据集mtcars演示</span><br><span class="line">ggplot(mtcars)+ geom_point(aes(wt, mpg), color=&quot;red&quot;)+ </span><br><span class="line">  geom_text(aes(wt, mpg, label=rownames(mtcars)))+ </span><br><span class="line">  theme_classic(base_size = 16)</span><br><span class="line"></span><br><span class="line">install.packages(&#x27;ggrepel&#x27;)</span><br><span class="line">library(ggrepel)</span><br><span class="line"></span><br><span class="line">ggplot(mtcars)+ geom_point(aes(wt, mpg), color=&quot;red&quot;)+ </span><br><span class="line">  geom_text_repel(aes(wt, mpg, label=rownames(mtcars)))+</span><br><span class="line">  theme_classic(base_size = 16)</span><br><span class="line">head(mtcars)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">ggplot(mtcars)+ geom_point(aes(wt, mpg), color=&quot;grey&quot;, size=5)+</span><br><span class="line">geom_label_repel(aes(wt, mpg, fill=factor(cyl), </span><br><span class="line">   label=rownames(mtcars)), fontface=&quot;bold&quot;, color=&quot;white&quot;, </span><br><span class="line">   box.padding=unit(0.35, &quot;lines&quot;), point.padding=unit(0.5, &quot;lines&quot;), </span><br><span class="line">   segment.colour = &quot;grey50&quot;)+ theme_classic(base_size = 16)</span><br></pre></td></tr></table></figure>
<p><img src="https://exts.ggplot2.tidyverse.org/ggrepel_files/figure-html/unnamed-chunk-1-1.png" alt="" /></p>
<p><img src="https://exts.ggplot2.tidyverse.org/ggrepel_files/figure-html/unnamed-chunk-1-2.png" alt="" /></p>
<p>资料来源：</p>
<ol>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/84e7e0e029cd">https://www.jianshu.com/p/84e7e0e029cd</a></li>
<li><a target="_blank" rel="noopener" href="https://exts.ggplot2.tidyverse.org/ggrepel.html">https://exts.ggplot2.tidyverse.org/ggrepel.html</a></li>
<li><a target="_blank" rel="noopener" href="https://blog.exploratory.io/ggrepel-when-things-get-too-crowded-ffefd845665f">https://blog.exploratory.io/ggrepel-when-things-get-too-crowded-ffefd845665f</a></li>
</ol>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/01/08/Bioinfo-Installation-of-BUSCO-v4/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/01/08/Bioinfo-Installation-of-BUSCO-v4/" class="post-title-link" itemprop="url">Bioinfo | Installation of BUSCO v4</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-01-08 10:25:45" itemprop="dateCreated datePublished" datetime="2021-01-08T10:25:45+08:00">2021-01-08</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-03-03 23:10:18" itemprop="dateModified" datetime="2021-03-03T23:10:18+08:00">2021-03-03</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h3 id="environment"><a class="markdownIt-Anchor" href="#environment"></a> Environment</h3>
<p>BUSCO is now Python 3.3+ only. I activate python 3.7 in Anaconda 3.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br></pre></td></tr></table></figure>
<h3 id="third-party-components"><a class="markdownIt-Anchor" href="#third-party-components"></a> Third-party components</h3>
<p>A full installation of BUSCO requires Python 3.3+ (2.7 is not supported from v4 onwards), BioPython, tBLASTn 2.2+, Augustus 3.2, Prodigal, Metaeuk, HMMER3.1+, SEPP, and R + ggplot2 for the plotting companion script. Some of these tools are necessary only for analysing certain type of organisms and input data, or for specific run modes.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br></pre></td><td class="code"><pre><span class="line"># https://biopython.org/</span><br><span class="line">conda install -c biopython</span><br><span class="line"></span><br><span class="line"># https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST</span><br><span class="line"># WARNING:        You are using BLAST version 2.11.0+. This is known to yield inconsistent results when multithreading. BLAST will run on a single core as a result. </span><br><span class="line"># For performance improvement, please upgrade to BLAST 2.10.1+</span><br><span class="line">cd ~/local/app</span><br><span class="line">wget https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.11.0+-x64-linux.tar.gz</span><br><span class="line">tar -zxvf ncbi-blast-2.11.0+-x64-linux.tar.gz</span><br><span class="line">mv ncbi-blast-2.11.0+ blast-2.11.0</span><br><span class="line">export PATH=/home/fenglei/local/app/blast-2.11.0/bin:$PATH</span><br><span class="line"></span><br><span class="line"># http://bioinf.uni-greifswald.de/augustus/</span><br><span class="line"># The following NEW packages will be INSTALLED:</span><br><span class="line"># augustus    bioconda/linux-64::augustus-3.2.2-0</span><br><span class="line"># conda install -c bioconda augustus   # ERROR when running it</span><br><span class="line">git clone https://github.com/Gaius-Augustus/Augustus</span><br><span class="line">sudo yum install boost-devel          # libboost-all-dev</span><br><span class="line">sudo yum install -y zlib zlib-devel   # zlib1g-dev</span><br><span class="line">sudo yum install gsl-devel            # libgsl-dev</span><br><span class="line">sudo yum install suitesparse-devel    # libsuitesparse-dev</span><br><span class="line">sudo yum install lpsolve              # liblpsolve55-dev</span><br><span class="line">sudo yum install sqlite-devel         # sqlite-devel - Development tools for the sqlite3 embeddable SQL database engine</span><br><span class="line"># sudo yum whatprovides libmysqlclient*</span><br><span class="line"># sudo yum install mariadb-libs-5.5.56-2.el7.i686</span><br><span class="line"></span><br><span class="line"># https://github.com/soedinglab/metaeuk</span><br><span class="line"># conda install -c bioconda Metaeuk # ERROR -- Package openssl conflicts</span><br><span class="line"># git clone git@github.com:soedinglab/metaeuk.git</span><br><span class="line"># downloading a statically compiled version:</span><br><span class="line">wget https://mmseqs.com/metaeuk/metaeuk-linux-sse41.tar.gz</span><br><span class="line">tar -zxvf metaeuk-linux-sse41.tar.gz</span><br><span class="line">export PATH=/home/fenglei/local/app/metaeuk/bin:$PATH</span><br><span class="line"></span><br><span class="line"># https://github.com/hyattpd/Prodigal</span><br><span class="line"># The following NEW packages will be INSTALLED:</span><br><span class="line"># prodigal    bioconda/linux-64::prodigal-2.6.3-h516909a_2</span><br><span class="line">conda install -c bioconda Prodigal</span><br><span class="line"></span><br><span class="line"># http://hmmer.org/</span><br><span class="line"># The following NEW packages will be INSTALLED:</span><br><span class="line"># hmmer    bioconda/linux-64::hmmer-3.3.1-he1b5a44_0</span><br><span class="line"># conda install -c bioconda hmmer  # X</span><br><span class="line">wget http://eddylab.org/software/hmmer/hmmer.tar.gz</span><br><span class="line">tar -zxvf hmmer.tar.gz</span><br><span class="line">cd hmmer-3.3.2</span><br><span class="line">./configure --prefix=/home/fenglei/local</span><br><span class="line">make</span><br><span class="line">make install</span><br><span class="line"></span><br><span class="line"># https://github.com/smirarab/sepp/</span><br><span class="line"># Successfully installed sepp-0.9.0</span><br><span class="line"># git clone git@github.com:smirarab/sepp.git</span><br><span class="line"># cd sepp</span><br><span class="line"># python setup.py  install</span><br><span class="line"># Error was reported when running setup.py</span><br><span class="line">conda install -c bioconda sepp</span><br><span class="line"># sepp  bioconda/linux-64::sepp-4.3.10-py37hef7ab6a_2</span><br><span class="line"></span><br><span class="line">#https://www.r-project.org/</span><br><span class="line"># R version 3.5.1 (2018-07-02) -- &quot;Feather Spray&quot;</span><br><span class="line">conda install R</span><br></pre></td></tr></table></figure>
<p><strong>samtools, bcftools, htslib</strong><br />
Download below files from <a target="_blank" rel="noopener" href="http://www.htslib.org/download/">http://www.htslib.org/download/</a>: <a target="_blank" rel="noopener" href="https://github.com/samtools/samtools/releases/download/1.11/samtools-1.11.tar.bz2">samtools-1.11.tar.bz2</a>; <a target="_blank" rel="noopener" href="https://github.com/samtools/bcftools/releases/download/1.11/bcftools-1.11.tar.bz2">bcftools-1.11.tar.bz2</a>; <a target="_blank" rel="noopener" href="https://github.com/samtools/htslib/releases/download/1.11/htslib-1.11.tar.bz2">htslib-1.10.2.tar.bz2</a></p>
<p><strong>cmake 3.19.2</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget https://github.com/Kitware/CMake/releases/download/v3.19.2/cmake-3.19.2.tar.gz</span><br><span class="line">tar -zxvf cmake-3.19.2.tar.gz</span><br><span class="line">cd cmake-3.19.2</span><br><span class="line">sudo yum install gcc g++</span><br><span class="line">./bootstrap --prefix=/home/fenglei/local</span><br><span class="line">gmake &amp;&amp; make install</span><br></pre></td></tr></table></figure>
<p><strong>bamtools</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">git clone git://github.com/pezmaster31/bamtools.git</span><br><span class="line">mkdir build ； cd build</span><br><span class="line">cmake -DCMAKE_INSTALL_PREFIX=/your/path/to/bamtools ..</span><br><span class="line">#cmake ..</span><br><span class="line">make</span><br><span class="line">sudo make  install</span><br></pre></td></tr></table></figure>
<p><strong>augustus 3.3.1</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wget http://bioinf.uni-greifswald.de/augustus/binaries/old/augustus-3.3.1.tar.gz</span><br><span class="line">tar -zxvf augustus-3.3.1.tar.gz</span><br><span class="line">cd augustus-3.3.1</span><br><span class="line"># edit some make files as below</span><br></pre></td></tr></table></figure>
<p>Edit make files of <a target="_blank" rel="noopener" href="https://iamphioxus.org/2017/05/08/installing-augustus-with-manual-bamtools-installation/">bam2hints, and filterBam.</a> Download binary <a target="_blank" rel="noopener" href="https://github.com/MikeAxtell/bam2wig">bam2wig</a> to replace source file of bam2wig. Then we can run <code>make</code> for Augustus. Augustus requires environment variables to be declared as follows:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">export PATH=&quot;/path/to/AUGUSTUS/augustus-3.2.3/bin:$PATH&quot;</span><br><span class="line">export PATH=&quot;/path/to/AUGUSTUS/augustus-3.2.3/scripts:$PATH&quot;</span><br><span class="line">export AUGUSTUS_CONFIG_PATH=&quot;/path/to/AUGUSTUS/augustus-3.2.3/config/&quot;</span><br></pre></td></tr></table></figure>
<h3 id="edit-config-file-of-busco"><a class="markdownIt-Anchor" href="#edit-config-file-of-busco"></a> Edit config file of BUSCO</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">git clone https://gitlab.com/ezlab/busco.git</span><br><span class="line">cd busco/</span><br><span class="line">#Note: v4.1.4 is the latest stable release. To access v5.beta clone this repository and checkout the v5 branch with git checkout v5.beta</span><br><span class="line">sudo python3 setup.py</span><br><span class="line"># or</span><br><span class="line">python3 setup.py install --user</span><br><span class="line"># or: If you are running BUSCO in a Python virtual environment, the following should suffice:</span><br><span class="line">python3 setup.py install</span><br></pre></td></tr></table></figure>
<p>Edit <code>config.ini</code> as below.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line"># some words were not shown here due to space limitation</span><br><span class="line"></span><br><span class="line">[tblastn]</span><br><span class="line">path = /home/fenglei/local/app/blast-2.10.1/bin</span><br><span class="line">command = tblastn</span><br><span class="line"></span><br><span class="line">[makeblastdb]</span><br><span class="line">path = /home/fenglei/local/app/blast-2.10.1/bin</span><br><span class="line">command = makeblastdb</span><br><span class="line"></span><br><span class="line">[augustus]</span><br><span class="line">path = /home/fenglei/local/app/augustus-3.3.1/bin</span><br><span class="line">command = augustus</span><br><span class="line"></span><br><span class="line">[etraining]</span><br><span class="line">path = /home/fenglei/local/app/augustus-3.3.1/bin</span><br><span class="line">command = etraining</span><br><span class="line"></span><br><span class="line">[gff2gbSmallDNA.pl]</span><br><span class="line">path = /home/fenglei/local/app/augustus-3.3.1/scripts</span><br><span class="line">command = gff2gbSmallDNA.pl</span><br><span class="line"></span><br><span class="line">[new_species.pl]</span><br><span class="line">path = /home/fenglei/local/app/augustus-3.3.1/scripts</span><br><span class="line">command = new_species.pl</span><br><span class="line"></span><br><span class="line">[optimize_augustus.pl]</span><br><span class="line">path = /home/fenglei/local/app/augustus-3.3.1/scripts</span><br><span class="line">command = optimize_augustus.pl</span><br><span class="line"></span><br><span class="line">[hmmsearch]</span><br><span class="line">path = /home/fenglei/local/bin</span><br><span class="line">command = hmmsearch</span><br><span class="line"></span><br><span class="line">[sepp]</span><br><span class="line">path = /home/fenglei/local/app/anaconda3/envs/python37/bin</span><br><span class="line">command = run_sepp.py</span><br><span class="line"></span><br><span class="line">[prodigal]</span><br><span class="line">path = /home/fenglei/local/app/anaconda3/envs/python37/bin</span><br><span class="line">command = prodigal</span><br></pre></td></tr></table></figure>
<p>Test run of BUSCO.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost config]$ which busco</span><br><span class="line">~/.local/bin/busco</span><br><span class="line">(python37) [fenglei@localhost config]$ busco --version</span><br><span class="line">BUSCO 4.1.4</span><br><span class="line">(python37) [fenglei@localhost config]$ busco --help</span><br><span class="line">usage: busco -i [SEQUENCE_FILE] -l [LINEAGE] -o [OUTPUT_NAME] -m [MODE] [OTHER OPTIONS]</span><br><span class="line"></span><br><span class="line">Welcome to BUSCO 4.1.4: the Benchmarking Universal Single-Copy Ortholog assessment tool.</span><br><span class="line">For more detailed usage information, please review the README file provided with this distribution and the BUSCO user guide.</span><br><span class="line"></span><br><span class="line">optional arguments:</span><br><span class="line">  -i FASTA FILE, --in FASTA FILE</span><br><span class="line">                        Input sequence file in FASTA format. Can be an assembled genome or transcriptome (DNA), or protein sequences from an annotated gene set.</span><br><span class="line">  -c N, --cpu N         Specify the number (N=integer) of threads/cores to use.</span><br><span class="line">  -o OUTPUT, --out OUTPUT</span><br><span class="line">                        Give your analysis run a recognisable short name. Output folders and files will be labelled with this name. WARNING: do not provide a path</span><br><span class="line">  --out_path OUTPUT_PATH</span><br><span class="line">                        Optional location for results folder, excluding results folder name. Default is current working directory.</span><br><span class="line">  -e N, --evalue N      E-value cutoff for BLAST searches. Allowed formats, 0.001 or 1e-03 (Default: 1e-03)</span><br><span class="line">  -m MODE, --mode MODE  Specify which BUSCO analysis mode to run.</span><br><span class="line">                        There are three valid modes:</span><br><span class="line">                        - geno or genome, for genome assemblies (DNA)</span><br><span class="line">                        - tran or transcriptome, for transcriptome assemblies (DNA)</span><br><span class="line">                        - prot or proteins, for annotated gene sets (protein)</span><br><span class="line">  -l LINEAGE, --lineage_dataset LINEAGE</span><br><span class="line">                        Specify the name of the BUSCO lineage to be used.</span><br><span class="line">  -f, --force           Force rewriting of existing files. Must be used when output files with the provided name already exist.</span><br><span class="line">  -r, --restart         Continue a run that had already partially completed.</span><br><span class="line">  --limit REGION_LIMIT  How many candidate regions (contig or transcript) to consider per BUSCO (default: 3)</span><br><span class="line">  --long                Optimization mode Augustus self-training (Default: Off) adds considerably to the run time, but can improve results for some non-model organisms</span><br><span class="line">  -q, --quiet           Disable the info logs, displays only errors</span><br><span class="line">  --augustus_parameters AUGUSTUS_PARAMETERS</span><br><span class="line">                        Pass additional arguments to Augustus. All arguments should be contained within a single pair of quotation marks, separated by commas. E.g. &#x27;--param1=1,--param2=2&#x27;</span><br><span class="line">  --augustus_species AUGUSTUS_SPECIES</span><br><span class="line">                        Specify a species for Augustus training.</span><br><span class="line">  --auto-lineage        Run auto-lineage to find optimum lineage path</span><br><span class="line">  --auto-lineage-prok   Run auto-lineage just on non-eukaryote trees to find optimum lineage path</span><br><span class="line">  --auto-lineage-euk    Run auto-placement just on eukaryote tree to find optimum lineage path</span><br><span class="line">  --update-data         Download and replace with last versions all lineages datasets and files necessary to their automated selection</span><br><span class="line">  --offline             To indicate that BUSCO cannot attempt to download files</span><br><span class="line">  --config CONFIG_FILE  Provide a config file</span><br><span class="line">  -v, --version         Show this version and exit</span><br><span class="line">  -h, --help            Show this help message and exit</span><br><span class="line">  --list-datasets       Print the list of available BUSCO datasets</span><br></pre></td></tr></table></figure>
<h3 id="running-busco"><a class="markdownIt-Anchor" href="#running-busco"></a> Running BUSCO</h3>
<p>The munnual is <a target="_blank" rel="noopener" href="https://busco.ezlab.org/busco_userguide.html#manual-installation"><em><u>busco_userguide.html</u></em></a></p>
<p>You have to set the environment variable <code>BUSCO_CONFIG_FILE</code> with the path to the file, for BUSCO to be able to locate it.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">export BUSCO_CONFIG_FILE=&quot;/home/fenglei/local/app/busco-4/config/config.ini&quot;</span><br><span class="line">export AUGUSTUS_CONFIG_PATH=&quot;/home/fenglei/local/app/augustus-3.3.1/config/&quot;</span><br><span class="line"></span><br><span class="line">busco -i [SEQUENCE_FILE] -l [LINEAGE] -o [OUTPUT_NAME] -m [MODE] [OTHER OPTIONS]</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line">cd /home/fenglei/local/app/busco-4/test_data/eukaryota</span><br><span class="line"></span><br><span class="line">busco -i genome.fna -c 8 -m geno -f --out test_eukaryota</span><br><span class="line"></span><br><span class="line">####   BUSCO result   ####</span><br><span class="line">INFO:</span><br><span class="line"></span><br><span class="line">        --------------------------------------------------</span><br><span class="line">        |Results from generic domain eukaryota_odb10      |</span><br><span class="line">        --------------------------------------------------</span><br><span class="line">        |C:18.8%[S:18.8%,D:0.0%],F:0.4%,M:80.8%,n:255     |</span><br><span class="line">        |48     Complete BUSCOs (C)                       |</span><br><span class="line">        |48     Complete and single-copy BUSCOs (S)       |</span><br><span class="line">        |0      Complete and duplicated BUSCOs (D)        |</span><br><span class="line">        |1      Fragmented BUSCOs (F)                     |</span><br><span class="line">        |206    Missing BUSCOs (M)                        |</span><br><span class="line">        |255    Total BUSCO groups searched               |</span><br><span class="line">        --------------------------------------------------</span><br><span class="line"></span><br><span class="line">        --------------------------------------------------</span><br><span class="line">        |Results from dataset saccharomycetes_odb10       |</span><br><span class="line">        --------------------------------------------------</span><br><span class="line">        |C:2.0%[S:2.0%,D:0.0%],F:0.1%,M:97.9%,n:2137      |</span><br><span class="line">        |42     Complete BUSCOs (C)                       |</span><br><span class="line">        |42     Complete and single-copy BUSCOs (S)       |</span><br><span class="line">        |0      Complete and duplicated BUSCOs (D)        |</span><br><span class="line">        |3      Fragmented BUSCOs (F)                     |</span><br><span class="line">        |2092   Missing BUSCOs (M)                        |</span><br><span class="line">        |2137   Total BUSCO groups searched               |</span><br><span class="line">        --------------------------------------------------</span><br><span class="line">INFO:   BUSCO analysis done with WARNING(s). Total running time: 570 seconds</span><br><span class="line"></span><br><span class="line">#### files generated ####</span><br><span class="line">(python37) [fenglei@localhost eukaryota]$ tree -L 1</span><br><span class="line">.</span><br><span class="line">├── busco_1816363720.log</span><br><span class="line">├── busco_3672791891.log</span><br><span class="line">├── busco_downloads</span><br><span class="line">├── expected_log.txt</span><br><span class="line">├── genome.fna</span><br><span class="line">├── info.txt</span><br><span class="line">└── test_eukaryota</span><br><span class="line"></span><br><span class="line">2 directories, 5 files</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="case-study-soybean"><a class="markdownIt-Anchor" href="#case-study-soybean"></a> Case study: soybean</h3>
<p>The genome sequence of Glycine max v4.0 <code>Gmax_508_v4.0.softmasked.fa</code> was downloaded from Phytozome database.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">export BUSCO_CONFIG_FILE=&quot;/home/fenglei/local/app/busco-4/config/config.ini&quot;</span><br><span class="line">export AUGUSTUS_CONFIG_PATH=&quot;/home/fenglei/local/app/augustus-3.3.1/config/&quot;</span><br><span class="line">busco -i Gmax_508_v4.0.softmasked.fa -m genome -o BUSCO_soybean -c 16 --auto-lineage-euk</span><br><span class="line"></span><br><span class="line"># Running time: 10 hours. Below is the result:</span><br><span class="line"></span><br><span class="line">INFO:   </span><br><span class="line"></span><br><span class="line">        --------------------------------------------------</span><br><span class="line">        |Results from generic domain eukaryota_odb10      |</span><br><span class="line">        --------------------------------------------------</span><br><span class="line">        |C:98.1%[S:22.0%,D:76.1%],F:1.2%,M:0.7%,n:255     |</span><br><span class="line">        |250    Complete BUSCOs (C)                       |</span><br><span class="line">        |56     Complete and single-copy BUSCOs (S)       |</span><br><span class="line">        |194    Complete and duplicated BUSCOs (D)        |</span><br><span class="line">        |3      Fragmented BUSCOs (F)                     |</span><br><span class="line">        |2      Missing BUSCOs (M)                        |</span><br><span class="line">        |255    Total BUSCO groups searched               |</span><br><span class="line">        --------------------------------------------------</span><br><span class="line"></span><br><span class="line">        --------------------------------------------------</span><br><span class="line">        |Results from dataset fabales_odb10               |</span><br><span class="line">        --------------------------------------------------</span><br><span class="line">        |C:97.7%[S:48.7%,D:49.0%],F:0.4%,M:1.9%,n:5366    |</span><br><span class="line">        |5245   Complete BUSCOs (C)                       |</span><br><span class="line">        |2615   Complete and single-copy BUSCOs (S)       |</span><br><span class="line">        |2630   Complete and duplicated BUSCOs (D)        |</span><br><span class="line">        |19     Fragmented BUSCOs (F)                     |</span><br><span class="line">        |102    Missing BUSCOs (M)                        |</span><br><span class="line">        |5366   Total BUSCO groups searched               |</span><br><span class="line">        --------------------------------------------------</span><br><span class="line">INFO:   BUSCO analysis done. Total running time: 34378 seconds</span><br><span class="line"></span><br></pre></td></tr></table></figure>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/01/05/Linux-Installation-of-Tomcat-9/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/01/05/Linux-Installation-of-Tomcat-9/" class="post-title-link" itemprop="url">Linux | Installation of Tomcat 9, Java 8, and MySQL 5.7.20</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-01-05 20:42:21" itemprop="dateCreated datePublished" datetime="2021-01-05T20:42:21+08:00">2021-01-05</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-01-07 09:48:35" itemprop="dateModified" datetime="2021-01-07T09:48:35+08:00">2021-01-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>Tomcat 服务器是一个免费的开放源代码的 Web 应用服务器，属于轻量级应用服务器，在中小型系统和并发访问用户不是很多的场合下被普遍使用，是开发和调试 JSP 程序的首选。Tomcat 是一个 JSP/Servlet 容器。其作为 Servlet 容器，有三种工作模式：独立的 Servlet 容器、进程内的 Servlet 容器和进程外的 Servlet 容器。</p>
<p>tomcat 与 nginx、apache的区别是什么？这三者都是 web server，那他们各自有什么特点呢？他们之间的区别是什么呢？Apache Tomcat 是 Apache 基金会下的一个项目，与 Apache HTTP Server 相比，Tomcat 能够动态的生成资源并返回到客户端。Apache HTTP Server 和 Nginx 都能够将某一个文本文件的内容通过 HTTP 协议返回到客户端，但是这个文本文件的内容是固定的——也就是说无论何时、任何人访问它得到的内容都是完全相同的，这样的资源我们称之为静态资源。动态资源则与之相反，在不同的时间、不同的客户端访问得到的内容是不同的，例如：包含显示当前时间的页面显示当前IP地址的页面。（摘抄自：<a target="_blank" rel="noopener" href="https://www.zhihu.com/question/32212996/answer/87524617">David</a>）</p>
<h2 id="一-下载安装-java"><a class="markdownIt-Anchor" href="#一-下载安装-java"></a> 一、下载安装 Java</h2>
<p>Tomcat是应用（Java）服务器，它只是一个Servlet(JSP也翻译成Servlet)容器，可以认为是Apache的扩展，但是可以独立于Apache运行。要安装 Tomcat 9，我们将需要 JAVA 8 或更高版本。 所以首先我们需要在我们的系统上安装和设置 JAVA。 您可以使用以下命令安装 JAVA。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"># tomcat 9 版本需要java8以上才能支持</span><br><span class="line">cd /tmp &amp;&amp; wget --no-check-certificate --no-cookies --header &quot;Cookie: oraclelicense=accept-securebackup-cookie&quot; http://download.oracle.com/otn-pub/java/jdk/8u131-b11/d54c1d3a095b4ff2b6607d096fa80163/jdk-8u131-linux-x64.rpm </span><br><span class="line">rpm -ivh jdk-8u131-linux-x64.rpm</span><br><span class="line"></span><br><span class="line"># 修改环境变量：</span><br><span class="line">vi + /etc/profile </span><br><span class="line"></span><br><span class="line"># 加入下列信息</span><br><span class="line">JAVA_HOME=/usr/java/jdk1.8.0_131 </span><br><span class="line">CLASSPATH=.:$JAVA_HOME/lib.tools.jar </span><br><span class="line">PATH=$JAVA_HOME/bin:$PATH</span><br><span class="line">export JAVA_HOME CLASSPATH PATH </span><br><span class="line"></span><br><span class="line"># 保存退出之后加载环境变量</span><br><span class="line">source /etc/profile</span><br></pre></td></tr></table></figure>
<p>Optional:</p>
<p>CentOS/RHEL</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ sudo yum install java-1.8.0</span><br></pre></td></tr></table></figure>
<p>Ubuntu/Debian</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ sudo apt-get install openjdk-8*</span><br></pre></td></tr></table></figure>
<h2 id="二-linux-上安装-apache-tomcat"><a class="markdownIt-Anchor" href="#二-linux-上安装-apache-tomcat"></a> 二、Linux 上安装 Apache Tomcat</h2>
<p>对于 Apache Tomcat 安装，我们将使用以下命令从<a target="_blank" rel="noopener" href="https://tomcat.apache.org/download-90.cgi">官方网站</a>下载已存档的软件包，</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">wget https://apache.website-solution.net/tomcat/tomcat-9/v9.0.41/bin/apache-tomcat-9.0.41.tar.gz</span><br><span class="line"># 一旦文件完成下载，解压缩/opt目录中的包（我们也可以使用一些其他的目录/数据等）。 在终端中运行以下命令来提取文件，</span><br><span class="line">sudo tar -xvzf pache-tomcat-8.5.23.tar.gz -C /opt</span><br><span class="line">sudo mv /opt/apache-tomcat-8.5.23/ /opt/tomcat</span><br><span class="line"># Optional: 更改tomcat目录的权限</span><br><span class="line">sudo chown -hR fenglei /opt/tomcat</span><br><span class="line">cd /opt/tomcat/bin</span><br><span class="line">sh startup.sh</span><br></pre></td></tr></table></figure>
<p>如果上面的 startup 成功运行，会收到以下提示。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">Using CATALINA_BASE:   /opt/tomcat</span><br><span class="line">Using CATALINA_HOME:   /opt/tomcat</span><br><span class="line">Using CATALINA_TMPDIR: /opt/tomcat/temp</span><br><span class="line">Using JRE_HOME:        /usr</span><br><span class="line">Using CLASSPATH:       /opt/tomcat/bin/bootstrap.jar:/opt/tomcat/bin/tomcat-juli.jar</span><br><span class="line">Using CATALINA_OPTS:   </span><br><span class="line">Tomcat started.</span><br></pre></td></tr></table></figure>
<p>使用如下命令开放 Linux 的 8080 端口,再访问 <a target="_blank" rel="noopener" href="http://192.168.1.100:8080">http://192.168.1.100:8080</a> 即可打开一个 tomcat 网页。192.168.1.100 是 apache tomcat 服务器的 IP 地址，根据实际情况更改该地址。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sudo iptables -A INPUT -ptcp --dport 8080 -j ACCEPT</span><br></pre></td></tr></table></figure>
<p><img src="https://ask.qcloudimg.com/http-save/yehe-1000017/eeqqalywve.jpeg?imageView2/2/w/1620" alt="tomcat初始化页面示意图" /></p>
<p>此时在这个tomcat初始页面里，点击“Manager App”按钮会显示无法进入，需要设定conf/tomcat-users.xml文件。</p>
<p>Tomcat 的目录结构</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">bin  # 脚本启动目录</span><br><span class="line">conf #配置文件目录</span><br><span class="line">lib  #tomcat运行的依赖包</span><br><span class="line">logs #日志文件目录</span><br><span class="line">temp #临时文件目录</span><br><span class="line">webapps  # 用来存放应用程序</span><br><span class="line">    --docs  # 文档</span><br><span class="line">    --examples # 示例</span><br><span class="line">    --host-manager # 虚拟主机web管理界面</span><br><span class="line">    --manager # 管理界面</span><br><span class="line">    --ROOT # 默认访问的应用程序</span><br><span class="line">work # 部署的Web应用程序的临时工作目录，当我们需要清除缓存的时候可以将该目录删除，然后重启动 tomcat。</span><br></pre></td></tr></table></figure>
<p>这里有个详细的介绍。包括后面怎么连接的网页：<a target="_blank" rel="noopener" href="https://www.cnblogs.com/operationhome/p/10021645.html">tomcat 安装以及常用配置</a></p>
<h3 id="tomcat-9-manager-配置"><a class="markdownIt-Anchor" href="#tomcat-9-manager-配置"></a> Tomcat 9 Manager 配置</h3>
<ol>
<li>首先配置两种角色：</li>
</ol>
<p>找到 tomcat 解压目录下的 conf/tomcat-users.xml 文件，用记事本打开，如下图添加角色，自定义处我都填了 “tomcat”。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&lt;role rolename=&quot;manager-gui&quot;/&gt; </span><br><span class="line">&lt;role rolename=&quot;admin-gui&quot;/&gt; </span><br><span class="line">&lt;user username=&quot;自定义&quot; password=&quot;自定义&quot; roles=&quot;manager-gui,admin-gui&quot;/&gt;</span><br></pre></td></tr></table></figure>
<p><img src="https://img-blog.csdnimg.cn/20190315150949838.png" alt="注意图中蓝色部分不用去掉，否则会出现 401 错误" /></p>
<ol start="2">
<li>更改 tomcat 目录下的 webapps\manager\META-INF/context.xml 文件</li>
</ol>
<p>如下图，对<Valve  /> 和<Manager  />进行注释</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&lt;!--</span><br><span class="line">  &lt;Valve className=&quot;org.apache.catalina.valves.RemoteAddrValve&quot;</span><br><span class="line">     allow=&quot;127\.\d+\.\d+\.\d+|::1|0:0:0:0:0:0:0:1&quot; /&gt;</span><br><span class="line">  &lt;Manager sessionAttributeValueClassNameFilter=&quot;java\.lang\.(?:Boolean|Integer|Long|Number|String)|org\.apache\.catalina\.filters\.CsrfPreventionFilter\$LruCache(?:\$1)?|java\.util\.(?:Linked)?HashMap&quot;/&gt;</span><br><span class="line">--&gt;</span><br></pre></td></tr></table></figure>
<p><img src="https://img-blog.csdnimg.cn/20190315151458443.png" alt="" /></p>
<p>然后重启 tomcat，从网页端点击 “Manager App”，就可以进入 “Tomcat Web 应用程序管理者”。</p>
<p><img src="https://img2020.cnblogs.com/blog/1552500/202010/1552500-20201015101304750-1675230814.png" alt="" /></p>
<h3 id="端口设置与编码配置文件设置"><a class="markdownIt-Anchor" href="#端口设置与编码配置文件设置"></a> 端口设置与编码配置文件设置</h3>
<p><strong>更改端口</strong></p>
<p>编辑配置文件 server.xml，大约在第 69 行，可以将 port=“8080” 更改成我们想绑定的端口。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&lt;Connector port=&quot;8080&quot; protocol=&quot;HTTP/1.1&quot;</span><br><span class="line">       connectionTimeout=&quot;20000&quot;</span><br><span class="line">       redirectPort=&quot;8443&quot; /&gt;</span><br></pre></td></tr></table></figure>
<p><strong>配置编码</strong></p>
<p>为了避免程序上的一些请求和返回的中文乱码问题，我们需要配置。编辑配置文件 server.xml。</p>
<p>大约在69行后,添加 useBodyEncodingForURI=“true” URIEncoding=“UTF-8”</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&lt;Connector port=&quot;8080&quot; protocol=&quot;HTTP/1.1&quot;</span><br><span class="line">               useBodyEncodingForURI=&quot;true&quot; URIEncoding=&quot;UTF-8&quot;</span><br><span class="line">               connectionTimeout=&quot;20000&quot;</span><br><span class="line">               redirectPort=&quot;8443&quot; /&gt;</span><br></pre></td></tr></table></figure>
<h3 id="tomcat-启动与关闭"><a class="markdownIt-Anchor" href="#tomcat-启动与关闭"></a> Tomcat 启动与关闭</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"># 启动 tomcat</span><br><span class="line">/opt/tomcat/bin/startup.sh </span><br><span class="line">#或者</span><br><span class="line">/opt/tomcat/bin/catalina.sh start</span><br><span class="line"># </span><br><span class="line"># 关闭tomcat</span><br><span class="line">/opt/tomcat/bin/shutdown.sh</span><br><span class="line">#或者</span><br><span class="line">/opt/tomcat/bin/catalina.sh stop</span><br><span class="line">#还可以直接kill 进程号，当我们的tomcat是有用户进行访问的时候，我们杀死不了这个进程，只能通过强制杀死 kill -9</span><br><span class="line"># 查看当前的版本</span><br><span class="line">/opt/tomcat/bin/version.sh</span><br></pre></td></tr></table></figure>
<h3 id="tomcat-部署项目"><a class="markdownIt-Anchor" href="#tomcat-部署项目"></a> Tomcat 部署项目</h3>
<p>方法1：直接将 war 包放在 ./webapps 下，tomcat 会自动将war解压。</p>
<p><img src="https://djxblog.oss-cn-shenzhen.aliyuncs.com/picture/tomcat%E5%AE%89%E8%A3%85%E4%BB%A5%E5%8F%8A%E5%B8%B8%E7%94%A8%E9%85%8D%E7%BD%AE/tomcat%20%E9%83%A8%E7%BD%B2%E6%96%B9%E6%B3%95%E4%B8%80.png" alt="" /></p>
<p>方法2：更改配置文件 server.xml(官方不建议使用该方法)</p>
<p>我们可以在 ./conf/ 目录下找到 server.xml 文件，我们需要在里面添加上我们的项目的访问路径和存放路径配置。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">&lt;Context  path=&quot;/djx&quot; docBase=&quot;/tmp/djx/&quot; reloadable=&quot;true&quot; /&gt;</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">path 指的是 访问路径  </span><br><span class="line">docBase 指的是项目存放路径</span><br><span class="line">reloadable 设置为true表示Catalina将监视 /WEB-INF/classes/和/WEB-INF/lib的文件，一旦发生更改将自动重新加载Web应用程序。它可以在在开发的时候使用，但需要消耗更多资源，官方文档写到不建议在生产环境上使用，所以它的默认值是false.</span><br><span class="line">8.0版本文档</span><br><span class="line">Set to true if you want Catalina to monitor classes in /WEB-INF/classes/ and /WEB-INF/lib for changes, and automatically reload the web application if a change is detected. This feature is very useful during application development, but it requires significant runtime overhead and is not recommended for use on deployed production applications. That&#x27;s why the default setting for this attribute is false. You can use the Manager web application, however, to trigger reloads of deployed applications on demand.</span><br></pre></td></tr></table></figure>
<p>配置示例：<br />
<img src="https://djxblog.oss-cn-shenzhen.aliyuncs.com/picture/tomcat%E5%AE%89%E8%A3%85%E4%BB%A5%E5%8F%8A%E5%B8%B8%E7%94%A8%E9%85%8D%E7%BD%AE/tomcat%20%E9%83%A8%E7%BD%B2%E6%96%B9%E6%B3%95%E4%BA%8C.png" alt="" /></p>
<p>方法3：使用xml配置文件(建议方法)</p>
<p>在 ./conf/Catalina/localhost/ 下添加xml配置文件。示例：我要让 /tmp/www/ 里的网站部署在tomcat上 。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">vim  djx.xml    # 文件的名称也就是我们后面访问时要在ip端口加的后缀</span><br></pre></td></tr></table></figure>
<p>编辑 djx.xml 内容如下，docBase 后面填入项目的路径 /var/www/html/jbrowse/，这是我之前安装JBrowser的路径，路径下有 index.html 文件和一堆其他文件。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">&lt;Context  docBase=&quot;/var/www/html/jbrowse/&quot; /&gt;</span><br></pre></td></tr></table></figure>
<p>然后我们重启tomcat，然后在浏览器 192.168.1.198/djx/访问的文件名 即可访问到我们的项目文件了。<br />
（原来JBrowser放在Tomcat上面也是一样的可以用于浏览。）</p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003184221.jpg?w=1024" alt="" /></p>
<h3 id="设置开机启动-tomcat"><a class="markdownIt-Anchor" href="#设置开机启动-tomcat"></a> 设置开机启动 Tomcat</h3>
<p>服务器关机重启的时候，服务不能随计算机的启动而自己启动，那么我们可以将 tomcat 服务设置为开机启动。</p>
<p>打开 Linux 设置开机启动的文件，将下面的配置文件写入此文件的最后，注意，是文件的最后，以后若还要增加其他开机启动，只能加在这前面，总之一句话，要保持下面这句话一直在文件的最后。</p>
<p>打开开机启动文件 <code>/etc/rc.d/rc.local</code></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">vi /etc/rc.d/rc.local</span><br></pre></td></tr></table></figure>
<p>添加如下代码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"># java配置 环境变量。根据自己的实际情况修改路径和软件版本号</span><br><span class="line">export JAVA_HOME=/usr/local/java/jdk1.8.0_161</span><br><span class="line">export CLASSPATH=.:$JAVA_HOME/jre/lib/rt.jar:$JAVA_HOME/lib/dt.jar:$JAVA_HOME/lib/tools.jar</span><br><span class="line">export PATH=$PATH:$JAVA_HOME/bin</span><br><span class="line"># tomcat 配置环境变量，根据自己的实际情况修改路径</span><br><span class="line">export CATALINA_HOME=/usr/local/tomcat7</span><br><span class="line"># tomcat自启动</span><br><span class="line">/opt/tomcat/bin/startup.sh</span><br></pre></td></tr></table></figure>
<p>Tomcat依赖于Java的jdk，所以jdk也同步导入。完成上面的步骤之后可以将CentOS关机重启检查一下。</p>
<h2 id="三-安装-mysql"><a class="markdownIt-Anchor" href="#三-安装-mysql"></a> 三、安装 MySQL</h2>
<p>Linux 系统（CentOS7）安装 mysql (5.7.25)</p>
<p><strong>卸载centos7自带的 MariaDB</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">rpm -qa|grep mariadb  // 查询出来已安装的mariadb</span><br><span class="line"># 删除查询出来的mariadb</span><br><span class="line">rpm -e --nodeps mariadb-5.5.64-1.el7.x86_64</span><br><span class="line">rpm -e --nodeps mariadb-libs-5.5.64-1.el7.x86_64</span><br><span class="line">rpm -e --nodeps mariadb-server-5.5.64-1.el7.x86_64</span><br><span class="line"># 安装依赖</span><br><span class="line">yum -y install numactl.x86_64</span><br></pre></td></tr></table></figure>
<p><strong>MySQL依赖libaio，所以安装libaio</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">yum install libaio</span><br></pre></td></tr></table></figure>
<p><strong>检查MySQL是否安装</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">yum list installed | grep mysql</span><br></pre></td></tr></table></figure>
<p>如果已安装，先卸载</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">yum -y remove mysql-libs.x86_64</span><br></pre></td></tr></table></figure>
<p>检查其他方式安装的 mysql</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">rpm -qa | grep mysql</span><br></pre></td></tr></table></figure>
<p>有的话，通过 <code>rpm -e</code> 命令，或者 <code>rpm -e --nodeps</code> 命令来卸载掉。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">rpm -e mysql 检查出的名称　　#普通删除模式</span><br><span class="line">rpm -e --nodeps 检查出的名称　　#强力删除模式，如果使用上面命令删除时，提示有依赖的其它文件，则用该命令可以对其进行强力删除</span><br></pre></td></tr></table></figure>
<p>在删除完以后我们可以通过 <code>rpm -qa | grep mysql</code> 命令来查看 mysql 是否已经卸载成功！！</p>
<p><strong>安装 MySQL</strong></p>
<p>进入MySQL下载官网，下载页面：<a target="_blank" rel="noopener" href="https://dev.mysql.com/downloads/mysql/">https://dev.mysql.com/downloads/mysql/</a>，选择 MySQL Community Server 下的 Linux generic 版本。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"># 下载 mysql </span><br><span class="line">wget http://dev.mysql.com/get/Downloads/MySQL-5.7/mysql-5.7.20-linux-glibc2.12-x86_64.tar.gz</span><br><span class="line"></span><br><span class="line"># 在/usr/local/中解压压缩包，并改名为mysql</span><br><span class="line">cd /usr/local/</span><br><span class="line">tar -xzvf /data/software/mysql-5.7.13-linux-glibc2.5-x86_64.tar.gz (我的下载目录为 data/software)</span><br><span class="line">mv mysql-5.7.13-linux-glibc2.5-x86_64 mysql</span><br><span class="line"></span><br><span class="line"># 创建数据文件存放目录, 进入mysql目录，新建data目录</span><br><span class="line">cd /usr/local/mysql</span><br><span class="line">mkdir data</span><br><span class="line"></span><br><span class="line"># 创建用户组mysql，创建用户mysql并将其添加到用户组mysql中，并赋予读写权限</span><br><span class="line">groupadd mysql</span><br><span class="line">useradd -r -g mysql mysql</span><br><span class="line">chown -R mysql mysql/</span><br><span class="line">chgrp -R mysql mysql/</span><br><span class="line"></span><br><span class="line"># 修改当前目录权限</span><br><span class="line">chown -R mysql:mysql /usr/local/mysql</span><br></pre></td></tr></table></figure>
<p><strong>初始化数据库</strong></p>
<p>注：可以先按照下文配置 my.cnf，再初始化数据库。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cd /usr/local/mysql/bin       //进入bin目录</span><br><span class="line">./mysqld --initialize --user=mysql --basedir=/usr/local/mysql --datadir=/usr/local/mysql/data</span><br></pre></td></tr></table></figure>
<p>注意屏幕上的输出，包含了密码信息。摘抄如下。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost bin]# ./mysqld --initialize --user=mysql --basedir=/usr/local/mysql --datadir=/usr/local/mysql/data</span><br><span class="line">2021-01-06T03:34:49.698565Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).</span><br><span class="line">2021-01-06T03:34:51.569548Z 0 [Warning] InnoDB: New log files created, LSN=45790</span><br><span class="line">2021-01-06T03:34:52.006540Z 0 [Warning] InnoDB: Creating foreign key constraint system tables.</span><br><span class="line">2021-01-06T03:34:52.194069Z 0 [Warning] No existing UUID has been found, so we assume that this is the first time that this server has been started. Generating a new UUID: 22e1b266-4fd0-11eb-81c9-989096c937ac.</span><br><span class="line">2021-01-06T03:34:52.218950Z 0 [Warning] Gtid table is not ready to be used. Table &#x27;mysql.gtid_executed&#x27; cannot be opened.</span><br><span class="line">2021-01-06T03:34:52.219393Z 1 [Note] A temporary password is generated for root@localhost: /-hyrfM3Huy&lt;</span><br></pre></td></tr></table></figure>
<p><strong>配置my.cnf</strong><br />
find / -name my.cnf 找到后复制到 /etc/my.cnf，或者直接新建 /etc/my.cnf，写入以下内容。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">[mysqld]</span><br><span class="line">character_set_server=utf8</span><br><span class="line">init_connect=&#x27;SET NAMES utf8&#x27;</span><br><span class="line">basedir=/usr/local/mysql</span><br><span class="line">datadir=/usr/local/mysql/data</span><br><span class="line">socket=/tmp/mysql.sock</span><br><span class="line">#不区分大小写 </span><br><span class="line">lower_case_table_names = 1</span><br><span class="line">#不开启sql严格模式</span><br><span class="line">sql_mode = &quot;STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION&quot;</span><br><span class="line">log-error=/var/log/mysqld.log</span><br><span class="line">pid-file=/usr/local/mysql/data/mysqld.pid</span><br></pre></td></tr></table></figure>
<p><strong>添加开机自启</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cd /usr/local/mysql/support-files</span><br><span class="line">cp mysql.server /etc/init.d/mysqld</span><br><span class="line">chkconfig --add mysqld</span><br></pre></td></tr></table></figure>
<p>chkconfig是管理系统服务(service)的命令行工具。所谓系统服务(service)，就是随系统启动而启动，随系统关闭而关闭的程序。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">chkconfig --add mysqld</span><br><span class="line">chkconfig mysqld on</span><br></pre></td></tr></table></figure>
<p>检查是否添加成功。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost bin]# chkconfig --list mysqld</span><br><span class="line"></span><br><span class="line">Note: This output shows SysV services only and does not include native</span><br><span class="line">      systemd services. SysV configuration data might be overridden by native</span><br><span class="line">      systemd configuration.</span><br><span class="line"></span><br><span class="line">      If you want to list systemd services use &#x27;systemctl list-unit-files&#x27;.</span><br><span class="line">      To see services enabled on particular target use</span><br><span class="line">      &#x27;systemctl list-dependencies [target]&#x27;.</span><br><span class="line"></span><br><span class="line">mysqld          0:off   1:off   2:on    3:on    4:on    5:on    6:off</span><br></pre></td></tr></table></figure>
<p><strong>关闭与启动 MySQL</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">/usr/local/mysql/support-files/mysql.server stop</span><br><span class="line"># 屏幕显示：Shutting down MySQL.. SUCCESS!</span><br><span class="line"></span><br><span class="line">/usr/local/mysql/support-files/mysql.server start</span><br><span class="line"># 屏幕显示：Starting MySQL. SUCCESS!</span><br><span class="line"></span><br><span class="line"># 查看 MySQL 状态</span><br><span class="line">/usr/local/mysql/support-files/mysql.server status</span><br><span class="line"># 屏幕显示：SUCCESS! MySQL running (15291)</span><br></pre></td></tr></table></figure>
<p>也可以使用 service mysqld 命令启动/停止服务</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 5.0版本：</span><br><span class="line">service mysqld status</span><br><span class="line">service mysqld start</span><br><span class="line">service mysqld stop</span><br><span class="line">service mysqld restart</span><br></pre></td></tr></table></figure>
<p><strong>修改root本地登录密码</strong></p>
<p>上文配置my.cnf操作是在MySQL初始化之后，所以初始化MySQL的时候密码输出在屏幕上。如果配置my.cnf在MySQL初始化之前，密码会保存至 /var/log/mysqld.log 中。</p>
<p>mysql安装完成之后，在 /var/log/mysqld.log 文件中给root生成了一个默认密码。通过下面的方式找到root默认密码，然后登录mysql进行修改：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">grep &#x27;temporary password&#x27; /var/log/mysqld.log</span><br></pre></td></tr></table></figure>
<p>到这里终于可以登陆进入mysql了，摘抄如下。（备注：一开始没有卸载 Centos 7 自带的 mariadb，运行<code>mysql -uroot -p</code>命令根本不是启动新安装的MySQL，难怪输入密码一直报错。）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost support-files]# /usr/local/mysql/bin/mysql -uroot -p     </span><br><span class="line">Enter password: </span><br><span class="line">Welcome to the MySQL monitor.  Commands end with ; or \g.</span><br><span class="line">Your MySQL connection id is 3</span><br><span class="line">Server version: 5.7.20</span><br><span class="line"></span><br><span class="line">Copyright (c) 2000, 2017, Oracle and/or its affiliates. All rights reserved.</span><br><span class="line"></span><br><span class="line">Oracle is a registered trademark of Oracle Corporation and/or its</span><br><span class="line">affiliates. Other names may be trademarks of their respective</span><br><span class="line">owners.</span><br><span class="line"></span><br><span class="line">Type &#x27;help;&#x27; or &#x27;\h&#x27; for help. Type &#x27;\c&#x27; to clear the current input statement.</span><br><span class="line"></span><br><span class="line">mysql&gt; </span><br></pre></td></tr></table></figure>
<p>登录之后修改密码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">cd /usr/local/mysql/bin/</span><br><span class="line">./mysql -uroot -p</span><br><span class="line"></span><br><span class="line">按照屏幕提示输入默认密码，即可登录。登录后可用下面的命令修改成新密码，以 MyNewPass4! 为例</span><br><span class="line">alter user &#x27;root&#x27;@&#x27;localhost&#x27; identified by &#x27;新密码&#x27;;</span><br><span class="line"># e.g. ALTER USER &#x27;root&#x27;@&#x27;localhost&#x27; IDENTIFIED BY &#x27;MyNewPass4!&#x27;;</span><br><span class="line">flush privileges; </span><br><span class="line"># 刷新权限，之后退出可用新密码重新登录</span><br></pre></td></tr></table></figure>
<p>也可以用下面这种方法修改密码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">set password for &#x27;root&#x27;@&#x27;localhost&#x27;=password(&#x27;MyNewPass4!&#x27;);</span><br></pre></td></tr></table></figure>
<p><strong>创建用户并授权</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"># 添加用户允许其在任意地址登录</span><br><span class="line">CREATE USER &#x27;dog&#x27;@&#x27;%&#x27; IDENTIFIED BY &#x27;123456&#x27;; </span><br><span class="line"></span><br><span class="line">#授予所有库的所有权限 </span><br><span class="line">GRANT ALL PRIVILEGES ON  *.* TO &#x27;dog&#x27;@&#x27;%&#x27; IDENTIFIED BY &#x27;123456&#x27;;</span><br><span class="line"></span><br><span class="line"># 刷新权限</span><br><span class="line">flush privileges;</span><br></pre></td></tr></table></figure>
<h2 id="四-发布项目"><a class="markdownIt-Anchor" href="#四-发布项目"></a> 四、发布项目</h2>
<p>将项目文件放到 /usr/local/tomcat7/webapps 文件夹下即可通过 110.120.18.18:8080 访问。</p>
<p>默认端口是8080，可以在 /usr/local/tomcat7/conf/server.xml 中进行修改。</p>
<p>找到如下代码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&lt;Connector port=&quot;8080&quot; protocol=&quot;HTTP/1.1&quot;</span><br><span class="line">               connectionTimeout=&quot;20000&quot;</span><br><span class="line">               redirectPort=&quot;8443&quot; /&gt;</span><br></pre></td></tr></table></figure>
<p>将8080改为80，即可直接通过 110.120.18.18 访问，代码如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&lt;Connector port=&quot;80&quot; protocol=&quot;HTTP/1.1&quot;</span><br><span class="line">               connectionTimeout=&quot;20000&quot;</span><br><span class="line">               redirectPort=&quot;8443&quot; /&gt;</span><br></pre></td></tr></table></figure>
<p><strong>参考资料</strong><br />
1 <a target="_blank" rel="noopener" href="https://blog.csdn.net/dvdstd/article/details/88575025">Tomcat 9 Manager 配置</a><br />
2 <a target="_blank" rel="noopener" href="https://www.cnblogs.com/operationhome/p/10021645.html">tomcat 安装以及常用配置</a><br />
3 <a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/news/97003">Apache Tomcat 安装</a><br />
4 <a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1538210">Linux 安装 JDK+Tomcat+MySQL 及发布项目教程</a><br />
5 <a target="_blank" rel="noopener" href="https://blog.csdn.net/lanyang123456/article/details/54695567">系统服务管理工具 chkconfig 使用说明</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/01/04/Visualization-Remove-Background-from-Image/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/01/04/Visualization-Remove-Background-from-Image/" class="post-title-link" itemprop="url">Remove Background from Image</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-01-04 09:33:30" itemprop="dateCreated datePublished" datetime="2021-01-04T09:33:30+08:00">2021-01-04</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2021-01-05 21:02:29" itemprop="dateModified" datetime="2021-01-05T21:02:29+08:00">2021-01-05</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Visualization/" itemprop="url" rel="index"><span itemprop="name">Visualization</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>没有扫描仪的情况下，往往会用手机拍摄文档照片，但是其背景通常是不均匀的灰色，不适合打印/复印。如何通过调整使背景变成白色或接近白色？</p>
<p>对于背景均匀的情况，在 PS 中用 ctrl+L 调整色阶或者曲线就可以满足要求。但是受到拍照的时候光线的原因，照片的背景颜色一般会深浅不一，而且手机拍摄的照片往往会有畸变。</p>
<p>下面记录一实用的办法：“高斯模糊”配合“划分”模式做出扫描的效果</p>
<ol>
<li>PS 打开图片文件。由于手机拍摄的照片角度可能不够理想，往往使文档有一定畸变，如果需要调整，选中左侧工具栏“剪裁工具”里面有一个“透视剪裁工具”，鼠标左键依次单击文档的四角，按下回车键，就可以把“透视”一定程度上矫正回来。</li>
<li>crtl+J复制一个图层出来，点击菜单栏：滤镜 -&gt; 模糊 -&gt; 高斯模糊，参数输入100，点击确定。此时视野中的图片变成了模糊的一片。</li>
<li>软件右下角图层设置区域，“不透明度”字样左边有个“正常”字样，将其重设为“划分”。此时视野中的图片背景变成了白色，文字颜色深灰色。</li>
<li>按下Ctrl+Alt+Shift+E，盖印一个图层。把之前的图层合并成另一新图层。再将其复制新图层，右下角图层栏目下“正常”模式更改成“颜色加深”，即可看到文字从深灰色变成黑色。</li>
</ol>
<p>参考资料：B 站有一个视频教学：<a target="_blank" rel="noopener" href="https://www.bilibili.com/s/video/BV1i7411Z718">https://www.bilibili.com/s/video/BV1i7411Z718</a></p>
<p><img src="https://199508.com/wp-content/uploads/2016/03/31db1439868311.jpg" alt="复制图层" /><br />
<img src="https://199508.com/wp-content/uploads/2016/03/be261439868309.jpg" alt="高斯模糊" /><br />
<img src="https://199508.com/wp-content/uploads/2016/03/ac5a1439868306.jpg" alt="划分" /><br />
<img src="https://pic2.zhimg.com/80/v2-b7fd16a3ea91d8285cb903aede434689_720w.jpg?source=1940ef5c" alt="效果图（其实可以更好）" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2021/01/01/Linux-Using-Nginx-to-host-Hexo-blog-in-PC/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2021/01/01/Linux-Using-Nginx-to-host-Hexo-blog-in-PC/" class="post-title-link" itemprop="url">在Linux服务器上使用Nginx部署Hexo站点</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2021-01-01 14:14:56" itemprop="dateCreated datePublished" datetime="2021-01-01T14:14:56+08:00">2021-01-01</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-15 10:21:28" itemprop="dateModified" datetime="2022-07-15T10:21:28+08:00">2022-07-15</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>之前，我在内网的 Linux 计算机上通过使用 Nginx（web服务器），然后安装了 JBrowse 基因组浏览器，从而可以使基因组数据以 web 页面形式展示出来。通过内网其他电脑的浏览器访问 Linux 计算机的 IP 地址即可浏览基因组数据。由于没有设置外网IP，外网是不能访问的。</p>
<p>后来，由于尝试个人博客，我在 github 网站上通过 hexo（一种博客框架）搭建了自己的博客。这个博客可以在互联网访问，但是由于托管在 GitHub 仓库，所以有空间限制，不能上传太多大文件到仓库的。</p>
<p>最近买了一个虚拟主机（VPS），其实就是一个具有公网 IP 的 Linux 服务器。理论上，这台虚拟主机也可以用来自建一个 web 服务器，并建立个人博客或网站。下面是实践过程。</p>
<h3 id="一-在虚拟主机上安装-nginx-web服务器"><a class="markdownIt-Anchor" href="#一-在虚拟主机上安装-nginx-web服务器"></a> 一、在虚拟主机上安装 Nginx （web服务器）</h3>
<p>虚拟主机是在 vultr 购买的。基本配置是 CPU: 1 vCore；RAM: 1024 MB； Storage: 25 GB SSD；Bandwidth: 0.26 GB of 1000 GB；OS: CentOS SELinux 8 x64。</p>
<p><strong>Nginx安装</strong></p>
<p>以root用户登陆服务器，新建目录<code>/usr/local/nginx</code>。</p>
<p>Nginx 安装包下载地址：<a target="_blank" rel="noopener" href="http://nginx.org/download/">http://nginx.org/download/</a>，我选了 <a target="_blank" rel="noopener" href="http://nginx.org/download/nginx-1.19.6.tar.gz">nginx-1.19.6.tar.gz</a>。</p>
<p>安装步骤如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">cd /usr/local/nginx</span><br><span class="line">wget http://nginx.org/download/nginx-1.19.6.tar.gz</span><br><span class="line">tar -zxvf nginx-1.19.6.tar.gz</span><br><span class="line">cd nginx-1.19.6</span><br><span class="line">./configure --user=fenglei --group=fenglei --prefix=/usr/local/nginx/</span><br><span class="line"># 或者 ./configure --prefix=/usr/local/nginx/</span><br><span class="line">make &amp;&amp; make install</span><br></pre></td></tr></table></figure>
<p>备注：由于是新购买的主机，很多依赖软件没有安装，上面 <code>configure</code> 和 <code>make</code> 的过程中遇到系统报错信息，安装对应的编译工具就可以了。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">yum -y install gcc</span><br><span class="line">yum -y install gcc-c++</span><br><span class="line">yum -y install openssl openssl-devel</span><br><span class="line">yum -y install pcre-devel</span><br><span class="line">#useradd fenglei</span><br><span class="line">yum -y install automake autoconf libtool make</span><br></pre></td></tr></table></figure>
<p><strong>Nginx启动、关闭、重启</strong></p>
<p>执行下方的 <code>nginx -s reload</code> 命令，即启动 Nginx web 服务器，理论上就可以在任何位置访问到这个 nginx 的初始网页，比如在其他 PC 或手机浏览器输入 <a href="http://123.456.78.89:80">http://123.456.78.89:80</a>，会看到一个如下图所示的静态网页。Nginx 配置文件中默认端口号是 80，如果 80 端口被其他应用占用，就需要更改 nginx 端口号。</p>
<p>更改端口号或者其他相关的配置需要修改 nginx 的配置文件：<code>/usr/local/nginx/conf/nginx.conf</code>，root 权限下 vim 这个配置文件，在 http 内容里面加入下面的文字。端口号可以自己设置，比如 8080、8082 等。location 字样则表示指向主页文件，比如我之前的 JBrowse 浏览器就设定为<code>/var/www/html/jbrowse/index.html</code>。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"># 停止运行 nginx </span><br><span class="line">/usr/local/nginx/sbin/nginx -s stop  </span><br><span class="line"></span><br><span class="line"># 重启（reload）如果报错，就要先输入下面的命令`</span><br><span class="line">/usr/local/nginx/sbin/nginx -c /usr/local/nginx/conf/nginx.conf  </span><br><span class="line"></span><br><span class="line"># 重启 nginx</span><br><span class="line">/usr/local/nginx/sbin/nginx -s reload</span><br></pre></td></tr></table></figure>
<p><img src="https://genehub.files.wordpress.com/2020/10/nginx-default-page.jpg?w=834" alt="nginx初始页面示例" /></p>
<h3 id="二-安装-nodejs"><a class="markdownIt-Anchor" href="#二-安装-nodejs"></a> 二、安装 Node.js</h3>
<p>Node.js 是什么？<a target="_blank" rel="noopener" href="https://www.zhihu.com/question/33578075/answer/122448279">某网站</a>的解释是：Node.js 是一项服务器技术。我们都知道客户端提出服务请求，而服务器端负责处理请求并提供服务。而对于互联网来说，在Node.js 之前 JavaScript 是一项完全的客户端技术，被用于浏览器中实现各种动画，对 DOM 的操作等等。而后端，即服务端则是由 PHP、Python、Ruby、Java 等等语言来实现。Node.js 的出现，使得前后端使用同一种语言，统一模型的梦想得以实现。</p>
<p>npm 是随同 Node.js 一起安装的包管理工具，能解决 Node.js 代码部署上的很多问题，常见的使用场景有以下几种：允许用户从NPM服务器下载别人编写的第三方包到本地使用；允许用户从NPM服务器下载并安装别人编写的命令行程序到本地使用；允许用户将自己编写的包或命令行程序上传到NPM服务器供别人使用。</p>
<p>Node.js 软件包下载 <a target="_blank" rel="noopener" href="https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz">https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz</a></p>
<p>按照下方命令行下载之，直接解压，没有编译步骤。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">mkdir /usr/local/app</span><br><span class="line">cd /usr/local/app</span><br><span class="line">https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz</span><br><span class="line">tar -xvJf node-v8.11.1-linux-x64.tar.xz</span><br></pre></td></tr></table></figure>
<p>解压目录里面会包含bin目录（node，npm，npx 三个程序），将 bin 目录路径加入环境变量（.bashrc）。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">export PATH=/usr/local/app/node-v12.18.4-linux-x64/bin:$PATH</span><br><span class="line"># source ~/.bashrc 之后即可测试 node -h 与 npm -h，均可以运行。</span><br></pre></td></tr></table></figure>
<p>测试安装是否成功。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">node --version</span><br><span class="line">npm --version</span><br></pre></td></tr></table></figure>
<h3 id="三-安装-hexo-博客框架"><a class="markdownIt-Anchor" href="#三-安装-hexo-博客框架"></a> 三、安装 Hexo （博客框架）</h3>
<p>在虚拟机上直接用 npm 安装 hexo。（什么情况下需要：sudo yum install git-core）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">npm install -g hexo-cli</span><br><span class="line">hexo --version  # 测试是否安装成功</span><br></pre></td></tr></table></figure>
<p>新建hexo项目</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cd  /usr/local/app # 设置新建web项目所在路径</span><br><span class="line">hexo init hexo     # 初始化新建一个项目名称，随便起名 hexo init myname 都可以</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>通过 <code>Hexo g</code> 命令生成的静态站点，Hexo 站点目录中将出现一个 public 文件夹。如下方代码块所示。将新生成的静态站点（也就是 public/ 目录），拷贝至 nginx 目录下的 html 文件夹中。然后修改Nginx配置文件 nginx.conf，只修改 root 字段为 public 目录，其他地方都不变（参考下面的代码块所示例），然后重新加载 nginx，执行 <code>nginx -s reload</code>，重新访问 <a href="http://123.456.78.89:8000">http://123.456.78.89:8000</a>，就可以看到 Hexo 静态站点了。这里要注意浏览器缓存的问题，另外应该也可以不用将 public 拷贝到 <code>/usr/local/nginx/html/</code>，直接将 location 定义成绝对路径（/usr/local/app/hexo/public/）可能也能访问。</p>
<p>执行 <code>tree -l 1 /usr/local/app/hexo</code> 可见到其目录下包含的文件列表：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">/usr/local/app/hexo/</span><br><span class="line">├── _config.landscape.yml</span><br><span class="line">├── _config.yml</span><br><span class="line">├── db.json</span><br><span class="line">├── node_modules</span><br><span class="line">├── package.json</span><br><span class="line">├── package-lock.json</span><br><span class="line">├── public</span><br><span class="line">├── scaffolds</span><br><span class="line">├── source</span><br><span class="line">└── themes</span><br></pre></td></tr></table></figure>
<p><code>/usr/local/nginx/conf/nginx.conf</code> 配置文件中的部分内容：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">server &#123;</span><br><span class="line">    listen       8000;</span><br><span class="line">    server_name  localhost;</span><br><span class="line"></span><br><span class="line">    #charset koi8-r;</span><br><span class="line"></span><br><span class="line">    #access_log  logs/host.access.log  main;</span><br><span class="line"></span><br><span class="line">    location / &#123;</span><br><span class="line">        root   html/public;</span><br><span class="line">        index  index.html;</span><br><span class="line">    &#125;</span><br></pre></td></tr></table></figure>
<p>至此就可以看到 Hexo 的静态站点了。就是下图所示的界面。</p>
<p><img src="https://d33wubrfki0l68.cloudfront.net/5997a40576f3beca7bbbd86fe79a795e9d520d8e/87f88/themes/screenshots/landscape.png" alt="Hexo 初始页面示意图" /></p>
<p>Hexo 提供 <code>hexo s</code> 命令来实时查看页面，如果在服务器端启动 <code>hexo s</code>，在浏览器中同样也可以通过服务器 IP 来访问站点，并且能实时显示当前状态。但是，<code>hexo s</code> 效率低，只适合在线调试，不适合作为站点访问。更新的时候，需要使用 <code>hexo g</code> 命令重新生成站点，然后将更新后的 public 目录拷贝到 nginx 目录中 html 文件夹下即可。</p>
<h3 id="四-更改网站主题"><a class="markdownIt-Anchor" href="#四-更改网站主题"></a> 四、更改网站主题</h3>
<p>Hexo 有大量个性化主题，我在 GitHub 博客用的 ，因为简洁轻便，所以在访问博客的时候加载快速，又因为它自动适应不同设备，在电脑、Pad 或手机访问均能良好展示，我非常中意。这次测试 Academia 主题，是一个学术主题，例如展示实验室课题组，或者研究人员个人主页。该主题网址：<a target="_blank" rel="noopener" href="https://phosphorw.github.io/">https://phosphorw.github.io/</a></p>
<p>The simplest way to install is to clone the entire repository:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/PhosphorW/hexo-theme-academia.git themes/Academia</span><br></pre></td></tr></table></figure>
<p>Some required renderers:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">npm install hexo-renderer-pug hexo-renderer-stylus</span><br></pre></td></tr></table></figure>
<p>Set theme in hexo work folder’s _config.yml</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">theme: Academia</span><br></pre></td></tr></table></figure>
<p>以上就是安装 Academia 主题的最基本操作，设置 hexo 的 _config.yml 之后，执行 <code>hexo</code> 更新 public 文件夹，再将 public/ 拷贝到 Nginx 安装目录的 html 下面。就可以看到新主题了（类似下图），内容区域还是个空白的，需要参考 Academia 的说明书进行添加页面。</p>
<p><img src="https://raw.githubusercontent.com/PhosphorW/phower-img-folder/master/hexo-theme-academia_mockup.jpg" alt="Academia 示意图" /></p>
<h3 id="五-参考资料"><a class="markdownIt-Anchor" href="#五-参考资料"></a> 五、参考资料</h3>
<ol>
<li><a target="_blank" rel="noopener" href="https://bioinfx.github.io/2020/10/03/%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%B5%8F%E8%A7%88%E5%99%A8-jbrowse-%E5%AE%89%E8%A3%85%EF%BC%88%E5%9F%BA%E4%BA%8E-nginx-%E7%9A%84-web-%E6%9C%8D%E5%8A%A1%E5%99%A8%EF%BC%89/">基因组浏览器 JBrowse 安装（基于 Nginx 的 Web 服务器）</a></li>
<li><a target="_blank" rel="noopener" href="https://genehub.wordpress.com/2020/12/07/%e4%bd%bf%e7%94%a8-github-hexo-%e5%bb%ba%e7%ab%8b%e4%b8%aa%e4%ba%ba%e5%8d%9a%e5%ae%a2/">使用 GitHub + Hexo 建立个人博客</a></li>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/31eb5c754c01">使用Nginx部署Hexo站点</a></li>
<li><a target="_blank" rel="noopener" href="https://segmentfault.com/a/1190000009723457">使用Nginx+Hexo光速搭建博客并实现服务器自动部署</a></li>
<li><a target="_blank" rel="noopener" href="https://blog.csdn.net/t8116189520/article/details/81909574">LINUX安装nginx详细步骤</a></li>
</ol>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/4/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/4/">4</a><span class="page-number current">5</span><a class="page-number" href="/page/6/">6</a><span class="space">&hellip;</span><a class="page-number" href="/page/19/">19</a><a class="extend next" rel="next" href="/page/6/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">185</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2023</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
