<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2020/08/09/%E8%BD%AC%EF%BC%9A%E5%A6%82%E4%BD%95%E4%B8%8A%E4%BC%A0%E4%BB%A3%E7%A0%81%E5%88%B0github/">转：如何上传代码到github</a></h1><div class="post-meta">2020-08-09</div><div class="post-content"><p>自从使用github以来，一直都是在<a target="_blank" rel="noopener" href="https://link.jianshu.com/?t=https://github.com/">github网站</a>在线上传文件到仓库中，但是有时因为网络或者电脑的原因上传失败。最重要的原因是我习惯本地编辑，完成以后再一起上传github。看过了几个教程，总结出最适合自己的比较简单的方法。</p></div><p class="readmore"><a href="/2020/08/09/%E8%BD%AC%EF%BC%9A%E5%A6%82%E4%BD%95%E4%B8%8A%E4%BC%A0%E4%BB%A3%E7%A0%81%E5%88%B0github/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2020/08/08/R%EF%BC%9A%E5%A6%82%E4%BD%95%E5%AF%B9%E5%9D%90%E6%A0%87%E8%BD%B4%E5%AF%B9%E5%BA%94%E5%85%83%E7%B4%A0%E7%9A%84%E9%A1%BA%E5%BA%8F%E8%BF%9B%E8%A1%8C%E8%B0%83%E6%95%B4-/">R：如何对坐标轴对应元素的顺序进行调整</a></h1><div class="post-meta">2020-08-08</div><div class="post-content"><p>R作图中，坐标轴上的元素默认按照数字大小或者 Alphabetical order（字母顺序）排列。如果需要对坐标轴元素顺序进行调整，可以使用 reorder 命令。</p>
<p>例1. 柱状图</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"># Create data</span><br><span class="line">data &lt;- data.frame(</span><br><span class="line">  name&#x3D;c(&quot;A&quot;,&quot;B&quot;,&quot;C&quot;,&quot;D&quot;,&quot;E&quot;) ,  </span><br><span class="line">  value&#x3D;c(3,12,5,18,45)</span><br><span class="line">  )</span><br><span class="line"></span><br><span class="line"># Barplot</span><br><span class="line">ggplot(data, aes(x&#x3D;name, y&#x3D;value)) + </span><br><span class="line">geom_bar(stat &#x3D; &quot;identity&quot;)</span><br><span class="line"># Barplot</span><br><span class="line">ggplot(data, aes(x&#x3D;reorder(name, -value), y&#x3D;value)) + </span><br><span class="line">geom_bar(stat &#x3D; &quot;identity&quot;)</span><br></pre></td></tr></table></figure></div><p class="readmore"><a href="/2020/08/08/R%EF%BC%9A%E5%A6%82%E4%BD%95%E5%AF%B9%E5%9D%90%E6%A0%87%E8%BD%B4%E5%AF%B9%E5%BA%94%E5%85%83%E7%B4%A0%E7%9A%84%E9%A1%BA%E5%BA%8F%E8%BF%9B%E8%A1%8C%E8%B0%83%E6%95%B4-/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2020/08/05/%E4%B8%BA%E4%BB%80%E4%B9%88%E8%A6%81%E5%AF%B9%E5%A4%9A%E9%87%8D%E6%A3%80%E9%AA%8C%E5%BE%97%E5%88%B0%E7%9A%84p%E5%80%BC%E8%BF%9B%E8%A1%8C%E6%A0%A1%E6%AD%A3/">为什么要对多重检验得到的P值进行校正</a></h1><div class="post-meta">2020-08-05</div><div class="post-content"><p>在生物学数据（例如转录组）分析中，对两组数据做比较，比较的结果包含数万个基因的倍数关系（Fold change）与 P 值，通常还会有一个 Q value，即对每个基因的 P value 进行校正。为什么要做校正？</p>
<p>上述提到的数万个基因分别做比较与统计学检验，就是一个多重检验的过程，而多重检验会增加一类错误的概率。通俗地说，本来原假设是小概率事件， 但是夜路走多了也会遇到鬼。抽样多了小概率事件发生几率就高了。</p></div><p class="readmore"><a href="/2020/08/05/%E4%B8%BA%E4%BB%80%E4%B9%88%E8%A6%81%E5%AF%B9%E5%A4%9A%E9%87%8D%E6%A3%80%E9%AA%8C%E5%BE%97%E5%88%B0%E7%9A%84p%E5%80%BC%E8%BF%9B%E8%A1%8C%E6%A0%A1%E6%AD%A3/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2020/04/24/linux%E7%BC%96%E8%BE%91%E5%99%A8vim%E9%85%8D%E7%BD%AE%EF%BC%9Aspf13-vim/">Linux编辑器vim配置：spf13-vim</a></h1><div class="post-meta">2020-04-24</div><div class="post-content"><p>spf13懒人工具（<a target="_blank" rel="noopener" href="http://vim.spf13.com/%EF%BC%89%E6%98%AF%E4%B8%80%E5%A5%97vim%E6%8F%92%E4%BB%B6%E7%BB%84%E6%88%90%E7%9A%84%E6%8F%92%E4%BB%B6%E9%9B%86%EF%BC%8C%E5%AE%89%E8%A3%85spf13%E5%8D%B3%E6%98%AF%E5%AE%89%E8%A3%85%E4%B8%80%E5%A5%97%E6%8F%92%E4%BB%B6">http://vim.spf13.com/）是一套vim插件组成的插件集，安装spf13即是安装一套插件</a>. 安装过程：</p></div><p class="readmore"><a href="/2020/04/24/linux%E7%BC%96%E8%BE%91%E5%99%A8vim%E9%85%8D%E7%BD%AE%EF%BC%9Aspf13-vim/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2020/04/21/Python%EF%BC%9A%E9%80%9A%E8%BF%87Anaconda%E7%AE%A1%E7%90%86python2%E5%92%8Cpython3/">Python：通过Anaconda管理python2和python3</a></h1><div class="post-meta">2020-04-21</div><div class="post-content"><h3 id="一、安装anaconda3"><a href="#一、安装anaconda3" class="headerlink" title="一、安装anaconda3"></a>一、安装anaconda3</h3><p>anaconda3 的下载页面：<a target="_blank" rel="noopener" href="https://www.anaconda.com/distribution/#download-section">https://www.anaconda.com/distribution/#download-section</a> 我在 Linux 系统里下载的版本：<a target="_blank" rel="noopener" href="https://repo.anaconda.com/archive/Anaconda3-2020.02-Linux-x86/_64.sh">https://repo.anaconda.com/archive/Anaconda3-2020.02-Linux-x86\_64.sh</a> 用sh命令安装即可，安装路径可以自定义，比如<code>/home/username/xxx/yyy</code> 此时输入 <code>conda</code> 命令没有反应。 安装完毕之后，将 bin 目录加入到 <code>.bashrc</code>，并执行 <code>source ~/.bashrc</code>，马上看到命令行前面出现<code>(base)</code>的字符，说明现在已经在 conda 的环境中了。这时候输入<code>conda list</code> 命令就有反应了。</p></div><p class="readmore"><a href="/2020/04/21/Python%EF%BC%9A%E9%80%9A%E8%BF%87Anaconda%E7%AE%A1%E7%90%86python2%E5%92%8Cpython3/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2020/03/05/R%EF%BC%9A%E6%89%B9%E9%87%8F%E8%AF%BB%E5%85%A5%E6%96%87%E6%9C%AC%E6%96%87%E4%BB%B6%E5%B9%B6%E5%BE%AA%E7%8E%AF%E5%A4%84%E7%90%86%EF%BC%8C%E5%88%86%E5%88%AB%E8%BE%93%E5%87%BA%E7%BB%9F%E8%AE%A1%E5%9B%BE-/">R: 批量读入文本文件并循环处理，分别输出统计图</a></h1><div class="post-meta">2020-03-05</div><div class="post-content"><p>对于高通量测序数据，通常都是数个或数十个样品，有时候要调用R packages对测序分析结果（例如bed格式文本文件）就行统计绘图，如何批量处理？   一、批量文件读取，逐个统计 demo：</p>
<p>##读取同一目录下的所有文件<br>path &lt;- “E:/实验数据/UseData/2013”<br>fileNames &lt;- dir(path)<br>filePath &lt;- sapply(fileNames, function(x){<br>paste(path,x,sep=’/‘)})<br>data &lt;- lapply(filePath, function(x){<br>read.csv(x, header=T)})<br>————————————————<br>版权声明：本文为CSDN博主「HuFeiHu-Blog」的原创文章，遵循 CC 4.0 BY-SA 版权协议，转载请附上原文出处链接及本声明。<br>原文链接：<a target="_blank" rel="noopener" href="https://blog.csdn.net/u011596455/article/details/79601113">https://blog.csdn.net/u011596455/article/details/79601113</a></p></div><p class="readmore"><a href="/2020/03/05/R%EF%BC%9A%E6%89%B9%E9%87%8F%E8%AF%BB%E5%85%A5%E6%96%87%E6%9C%AC%E6%96%87%E4%BB%B6%E5%B9%B6%E5%BE%AA%E7%8E%AF%E5%A4%84%E7%90%86%EF%BC%8C%E5%88%86%E5%88%AB%E8%BE%93%E5%87%BA%E7%BB%9F%E8%AE%A1%E5%9B%BE-/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2020/03/05/linux%EF%BC%9Azsh%E7%9A%84%E5%AE%89%E8%A3%85%E4%B8%8E%E4%BD%BF%E7%94%A8%EF%BC%88oh-my-zsh%EF%BC%89/">Linux：zsh的安装与使用（oh-my-zsh）</a></h1><div class="post-meta">2020-03-05</div><div class="post-content"><p>  Mac OS更新系统之后，进入iTerm2收到提示说可以使用zsh，查资料比较了一下它和bash的区别，然后就安装了。 简要过程：在客户端（Mac OS）安装zsh并启动，通过oh-my-zsh配置zsh；随后在服务器（Linux CentOS 7）安装zsh，从bash切换到zsh，同样通过oh-my-zsh配置zsh，由于<del>/.bashrc和</del>/.bash_profile里面保存了环境变量和alias等信息，切换成zsh之后也需要在~/.zshrc文件里面做相应设置；还有服务器的Tmux软件也要做设置。 具体操作都是参考网络资料，下面直接转载一篇比较详细的文章。 ==========</p></div><p class="readmore"><a href="/2020/03/05/linux%EF%BC%9Azsh%E7%9A%84%E5%AE%89%E8%A3%85%E4%B8%8E%E4%BD%BF%E7%94%A8%EF%BC%88oh-my-zsh%EF%BC%89/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2020/03/02/chipseeker%E7%9A%84%E5%AE%89%E8%A3%85/">ChipSeeker的安装</a></h1><div class="post-meta">2020-03-02</div><div class="post-content"><p>我使用Linux CentOS 7系统，在ChipSeeker的安装过程中遇到了许多问题，这里做个记录。 <img src="https://genehub.files.wordpress.com/2020/03/e688aae5b18f2020-03-02e4b88be58d887.59.11.png" alt="截屏2020-03-02下午7.59.11"> 在R （version 3.6）输入如下命令开始自动安装。安装命令：</p></div><p class="readmore"><a href="/2020/03/02/chipseeker%E7%9A%84%E5%AE%89%E8%A3%85/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2020/01/20/%E8%B4%A8%E7%B2%92dna%E6%B5%8B%E5%BA%8F%E7%BB%93%E6%9E%9C%E5%88%86%E6%9E%90/">质粒DNA测序结果分析</a></h1><div class="post-meta">2020-01-20</div><div class="post-content"><p>  目的基因的克隆，简要过程如下，构建的阳性克隆需要经过 Sanger 测序验证。对于 Sanger 测序，一般一个反应（一个引物往后延伸）测的比较准确的长度大概是 800 bp，如果大于这个长度，就需要设计多条引物，用多个反应分区段测定。 <img src="https://genehub.files.wordpress.com/2020/01/cloning_dna-_science_learning_hub.jpg" alt="cloning_dna-_Science_Learning_Hub">   测序结果解压后一般有两个类型的文件：ab1后缀的谱图文件和seq后缀的序列文件。 ab1格式的谱图可以用BioEdit软件打开，界面如下。一般在 20 bp ~ 900 bp 的测序区间是比较准确的。 <img src="https://genehub.files.wordpress.com/2020/01/ab1.png" alt="ab1"> 以.seq为后缀的序列文件可以直接用 EditPlus、SnapGene 之类的序列编辑软件打开，也可以用电脑自带的记事本和写字板打开，拷贝出来用于序列分析。 可用 DNAstar 软件中的 MegAlign 软件将测序文件和模板序列进行比对，如果没有问题就可以放心就行后面的操作了。依次点击 MegAlign -&gt; File -&gt; Enter Sequences -&gt; 选择seq后缀的文件 -&gt; 打开，即可将测序文件导入软件，再以同样的方式导入fasta格式的参考序列，并选择 Align -&gt; By Clustal W Method，即可看到序列比对结果。注意反向引物的测序序列需要自己先反转（网页工具：<a target="_blank" rel="noopener" href="http://www.bioinformatics.org/sms/rev/_comp.html%EF%BC%89%EF%BC%8C%E5%8F%8D%E8%BD%AC%E5%90%8E%E7%9A%84%E5%BA%8F%E5%88%97%E5%8F%A6%E5%AD%98%E4%B8%80%E4%B8%AA%E6%96%B0%E6%96%87%E4%BB%B6%E6%89%8D%E5%8F%AF%E4%BB%A5%E5%AF%BC%E5%85%A5%EF%BC%8C%E5%90%A6%E5%88%99%E5%8F%8D%E5%90%91%E5%BC%95%E7%89%A9%E6%B5%8B%E5%BA%8F%E7%9A%84%E7%BB%93%E6%9E%9C%E4%B8%8E%E5%8F%82%E8%80%83%E5%BA%8F%E5%88%97%E6%96%B9%E5%90%91%E4%B8%8D%E4%B8%80%E8%87%B4%EF%BC%8C%E8%BD%AF%E4%BB%B6%E4%B8%8D%E8%83%BD%E6%AF%94%E5%AF%B9%E3%80%82%E9%99%A4%E4%BA%86">http://www.bioinformatics.org/sms/rev\_comp.html），反转后的序列另存一个新文件才可以导入，否则反向引物测序的结果与参考序列方向不一致，软件不能比对。除了</a> MegAlign，还有其他序列比对软件可以做，例如windows系统的ClustalX；或者网页版的 ClustalW（<a target="_blank" rel="noopener" href="https://www.genome.jp/tools-bin/clustalw%EF%BC%89%E4%B9%9F%E5%8F%AF%E4%BB%A5%E5%81%9A%E5%8F%8C%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9%E3%80%82">https://www.genome.jp/tools-bin/clustalw）也可以做双序列比对。</a> <img src="https://genehub.files.wordpress.com/2020/01/e5beaee4bfa1e688aae59bbe_20200120113453.png" alt="微信截图_20200120113453">   如果要对比对结果进行图形化展示，可以用 Color Align Conservation （<a target="_blank" rel="noopener" href="http://www.bioinformatics.org/sms2/color/_align/_cons.html%EF%BC%89%E5%87%BA%E5%9B%BE%E3%80%82%E4%B8%8B%E9%9D%A2%E6%98%AF%E7%BD%91%E9%A1%B5%E6%8F%90%E4%BE%9B%E7%9A%84%E8%9B%8B%E7%99%BD%E8%B4%A8%E5%BA%8F%E5%88%97%E4%BD%9C%E4%B8%BA%E6%A1%88%E4%BE%8B%EF%BC%8C%E5%AF%B9%E4%BA%8EDNA%E5%BA%8F%E5%88%97%E4%B9%9F%E6%98%AF%E7%B1%BB%E4%BC%BC%E7%9A%84%E3%80%82">http://www.bioinformatics.org/sms2/color\_align\_cons.html）出图。下面是网页提供的蛋白质序列作为案例，对于DNA序列也是类似的。</a> <img src="https://genehub.files.wordpress.com/2020/01/e5beaee4bfa1e688aae59bbe_20200120114724.png" alt="微信截图_20200120114724"></p></div><p class="readmore"><a href="/2020/01/20/%E8%B4%A8%E7%B2%92dna%E6%B5%8B%E5%BA%8F%E7%BB%93%E6%9E%9C%E5%88%86%E6%9E%90/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2019/12/13/ssh%E8%BF%9C%E7%A8%8B%E7%99%BB%E5%BD%95%E8%AE%BE%E7%BD%AE/">SSH远程登录设置</a></h1><div class="post-meta">2019-12-13</div><div class="post-content"><p>如图所示，Linux服务器为C，用户客户端电脑为A。如何从A登录到C服务器。之前尝试多次不成功，原来是没有在Linux服务器正确设置参数（即下面步骤二）。   步骤一、在B路由器设定“端口转发”（或端口映射？），比如将端口22转发到服务器C；参考<a target="_blank" rel="noopener" href="https://service.tp-link.com.cn/detail/_article/_69.html">https://service.tp-link.com.cn/detail\_article\_69.html</a> 步骤二、在Linux服务器上设置远程登录，编辑/etc/ssh/sshd_config，打开22端口，再修改一些设置允许用户远程登录。具体参考：jianshu.com/p/df83a3b96f64 步骤三、在用户电脑客户端进行ssh操作即可。比如ssh <a href="mailto:&#x75;&#115;&#x65;&#x72;&#x40;&#x31;&#48;&#x31;&#46;&#x31;&#46;&#49;&#x2e;&#50;">&#x75;&#115;&#x65;&#x72;&#x40;&#x31;&#48;&#x31;&#46;&#x31;&#46;&#49;&#x2e;&#50;</a>，即可登录到服务器C。参考：<a target="_blank" rel="noopener" href="https://blog.csdn.net/qiushisoftware/article/details/55258624">https://blog.csdn.net/qiushisoftware/article/details/55258624</a>   <img src="https://genehub.files.wordpress.com/2019/12/790056-20151122112639515-536213849.png" alt="790056-20151122112639515-536213849"></p></div><p class="readmore"><a href="/2019/12/13/ssh%E8%BF%9C%E7%A8%8B%E7%99%BB%E5%BD%95%E8%AE%BE%E7%BD%AE/">Read More</a></p></div><nav class="page-navigator"><a class="extend prev" rel="prev" href="/page/4/">Previous</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/3/">3</a><a class="page-number" href="/page/4/">4</a><span class="page-number current">5</span><a class="page-number" href="/page/6/">6</a><a class="page-number" href="/page/7/">7</a><span class="space">&hellip;</span><a class="page-number" href="/page/16/">16</a><a class="extend next" rel="next" href="/page/6/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/17/Linux-Gene-Ontology-GOMAP/">Bioinfo | Gene Ontology - GOMAP</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Linux | 认识 PKG_CONFIG_PATH</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">jcvi 安装与使用，报错以及解决办法</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">Bioinfo | Installation of OrthoMCL-pipeline</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Bioinfo | Using Circos for genome visualization</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">Bioinfo | Repeat sequence analysis by RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer)</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/16/Bioinfo-understanding-sequence-identity/">Bioinfo | 怎么理解序列一致性（Sequence identity）</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>