<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="以内置的 iris 数据集为例。Iris 鸢尾花数据集是一个经典数据集，在统计学习和机器学习领域都经常被用作示例。数据集内包含 3 类共 150 条记录，每类各 50 个数据，每条记录都有 4 项特征：花萼长度、花萼宽度、花瓣长度、花瓣宽度，可以通过这4个特征预测鸢尾花卉属于（iris-setosa, iris-versicolour, iris-virginica）中的哪一品种。在博文 http">
<meta property="og:type" content="article">
<meta property="og:title" content="R | 如何用 RStudio 绘制适合投稿的柱状图（柱状图 + 散点 + 显著性标记）">
<meta property="og:url" content="http://example.com/2022/11/09/R-draw-high-quality-figure-in-RStudio/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:description" content="以内置的 iris 数据集为例。Iris 鸢尾花数据集是一个经典数据集，在统计学习和机器学习领域都经常被用作示例。数据集内包含 3 类共 150 条记录，每类各 50 个数据，每条记录都有 4 项特征：花萼长度、花萼宽度、花瓣长度、花瓣宽度，可以通过这4个特征预测鸢尾花卉属于（iris-setosa, iris-versicolour, iris-virginica）中的哪一品种。在博文 http">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://i.imgur.com/4pwGYUW.jpeg">
<meta property="og:image" content="https://i.imgur.com/Nl6NDlZ.jpeg">
<meta property="og:image" content="https://i.imgur.com/JUJP8CC.jpeg">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-8-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-8-2.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-12-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-16-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-19-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-21-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-23-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-25-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-26-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-27-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-28-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-29-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-31-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-32-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-32-2.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-33-1.png">
<meta property="og:image" content="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-33-2.png">
<meta property="article:published_time" content="2022-11-09T03:32:13.000Z">
<meta property="article:modified_time" content="2022-11-09T03:26:14.037Z">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://i.imgur.com/4pwGYUW.jpeg">

<link rel="canonical" href="http://example.com/2022/11/09/R-draw-high-quality-figure-in-RStudio/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>R | 如何用 RStudio 绘制适合投稿的柱状图（柱状图 + 散点 + 显著性标记） | FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/11/09/R-draw-high-quality-figure-in-RStudio/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          R | 如何用 RStudio 绘制适合投稿的柱状图（柱状图 + 散点 + 显著性标记）
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2022-11-09 11:32:13 / Modified: 11:26:14" itemprop="dateCreated datePublished" datetime="2022-11-09T11:32:13+08:00">2022-11-09</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>以内置的 iris 数据集为例。Iris 鸢尾花数据集是一个经典数据集，在统计学习和机器学习领域都经常被用作示例。数据集内包含 3 类共 150 条记录，每类各 50 个数据，每条记录都有 4 项特征：花萼长度、花萼宽度、花瓣长度、花瓣宽度，可以通过这4个特征预测鸢尾花卉属于（iris-setosa, iris-versicolour, iris-virginica）中的哪一品种。在博文 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/52b86c774b0b">https://www.jianshu.com/p/52b86c774b0b</a> 有对 Iris 数据集进行统计分析以及机器学习的总结。</p>
<span id="more"></span>
<h3 id="柱状图-散点"><a class="markdownIt-Anchor" href="#柱状图-散点"></a> 柱状图 + 散点</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">&gt; iris</span><br><span class="line">    Sepal.Length Sepal.Width Petal.Length Petal.Width    Species</span><br><span class="line">1            5.1         3.5          1.4         0.2     setosa</span><br><span class="line">2            4.9         3.0          1.4         0.2     setosa</span><br><span class="line">3            4.7         3.2          1.3         0.2     setosa</span><br><span class="line">4            4.6         3.1          1.5         0.2     setosa</span><br><span class="line">5            5.0         3.6          1.4         0.2     setosa</span><br><span class="line">6            5.4         3.9          1.7         0.4     setosa</span><br><span class="line">7            4.6         3.4          1.4         0.3     setosa</span><br><span class="line">8            5.0         3.4          1.5         0.2     setosa</span><br><span class="line">9            4.4         2.9          1.4         0.2     setosa</span><br><span class="line">51           7.0         3.2          4.7         1.4 versicolor</span><br><span class="line">52           6.4         3.2          4.5         1.5 versicolor</span><br><span class="line">53           6.9         3.1          4.9         1.5 versicolor</span><br><span class="line">54           5.5         2.3          4.0         1.3 versicolor</span><br><span class="line">55           6.5         2.8          4.6         1.5 versicolor</span><br><span class="line">56           5.7         2.8          4.5         1.3 versicolor</span><br><span class="line">57           6.3         3.3          4.7         1.6 versicolor</span><br><span class="line">58           4.9         2.4          3.3         1.0 versicolor</span><br><span class="line">59           6.6         2.9          4.6         1.3 versicolor</span><br><span class="line">100          5.7         2.8          4.1         1.3 versicolor</span><br><span class="line">101          6.3         3.3          6.0         2.5  virginica</span><br><span class="line">102          5.8         2.7          5.1         1.9  virginica</span><br><span class="line">103          7.1         3.0          5.9         2.1  virginica</span><br><span class="line">104          6.3         2.9          5.6         1.8  virginica</span><br><span class="line">105          6.5         3.0          5.8         2.2  virginica</span><br><span class="line">106          7.6         3.0          6.6         2.1  virginica</span><br><span class="line">107          4.9         2.5          4.5         1.7  virginica</span><br></pre></td></tr></table></figure>
<p>柱状图是生物学文献中最常见的一类图形，下面基于 Iris 数据做一个柱状图（展示 MEAN、SE 等信息）。首先想到了 dplyr 函数来进行统计。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"># 对三个植物品种的花萼 (Sepal)进行平均值和标准差统计</span><br><span class="line">iris %&gt;%</span><br><span class="line">  group_by(Species) %&gt;%</span><br><span class="line">  summarise(</span><br><span class="line">    mean.Sepal.Length = mean(Sepal.Length),</span><br><span class="line">    sd.Sepal.Length = sd(Sepal.Length)</span><br><span class="line">  )</span><br><span class="line"></span><br><span class="line"># A tibble: 3 × 3</span><br><span class="line">#Species    mean.Sepal.Length sd.Sepal.Length</span><br><span class="line">#&lt;fct&gt;                  &lt;dbl&gt;           &lt;dbl&gt;</span><br><span class="line">#1 setosa                  5.01           0.352</span><br><span class="line">#2 versicolor              5.94           0.516</span><br><span class="line">#3 virginica               6.59           0.636</span><br><span class="line"></span><br><span class="line"># 如果要对所有数值型变量计算某些统计量， 可以用summarize_if(is.numeric, list(变量后缀=~统计函数名&lt;, ...&gt;))</span><br><span class="line">a &lt;- iris %&gt;%</span><br><span class="line">  group_by(Species) %&gt;%</span><br><span class="line">  summarise_if(</span><br><span class="line">    is.numeric,</span><br><span class="line">    list(avg = ~mean(.), std = ~sd(.))</span><br><span class="line">  )</span><br><span class="line"># A tibble: 3 × 9</span><br><span class="line">#Species    Sepal.Length_avg Sepal.Width_avg Petal.Length_avg Petal.Width_avg Sepal.Length_std Sepal.Width_std Petal.Length_std</span><br><span class="line">#&lt;fct&gt;                 &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;</span><br><span class="line">#1 setosa                 5.01            3.43             1.46           0.246            0.352           0.379            0.174</span><br><span class="line">#2 versicolor             5.94            2.77             4.26           1.33             0.516           0.314            0.470</span><br><span class="line">#3 virginica              6.59            2.97             5.55           2.03             0.636           0.322            0.552</span><br><span class="line"># … with 1 more variable: Petal.Width_std &lt;dbl&gt;</span><br></pre></td></tr></table></figure>
<p>下面将三个不同植物品种各自花萼平均值来做柱状图。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ggplot() +</span><br><span class="line">  geom_col(data=a, aes(x=Species, y=Sepal.Length_avg)) +</span><br><span class="line">  geom_errorbar(data=a, aes(x=Species, ymin=Sepal.Length_avg-Sepal.Length_std, ymax=Sepal.Length_avg+Sepal.Length_std))</span><br></pre></td></tr></table></figure>
<p>下面是生成的图，可看到灰色网格背景，以及柱子的宽度都不够美观。</p>
<p><img src="https://i.imgur.com/4pwGYUW.jpeg" alt="" /></p>
<p>下面通过 theme_bw() 函数去掉灰色背景；再通过 theme() 函数去掉网格背景，设置坐标轴字体大小；通过 ylab 和 xlab 函数设置坐标轴标签文字；通过 xlim 和 ylim 设置坐标轴展示区间；通过scale_fill_manual 函数来控制柱状图颜色；通过guides 函数控制是否显示 legend。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">pdf(&quot;Sepal_length_barplot.pdf&quot;, width = 2, height = 2)</span><br><span class="line">ggplot() +</span><br><span class="line">  geom_col(data=a, aes(x=Species, y=Sepal.Length_avg, fill=Species), width=.5) +</span><br><span class="line">  geom_errorbar(data=a, aes(x=Species, ymin=Sepal.Length_avg-Sepal.Length_std, ymax=Sepal.Length_avg+Sepal.Length_std), width=.3) +</span><br><span class="line">  geom_jitter(data=iris, size=.3, alpha=.5, color=&quot;darkgrey&quot;, aes(x=Species, y=Sepal.Length), width=.3) +</span><br><span class="line">  scale_fill_manual(values=c(&#x27;#90c9e7&#x27;, &#x27;#219ebc&#x27;, &#x27;#136783&#x27;)) +</span><br><span class="line">  theme_bw() +</span><br><span class="line">  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = &quot;black&quot;)) +</span><br><span class="line">  theme(axis.text.x = element_text(size = 6, color = &quot;black&quot;)) +</span><br><span class="line">  #theme(axis.text.x = element_text(size = 6, color = &quot;black&quot;, hjust = 0.1, vjust = 0.4, angle = 70)) +</span><br><span class="line">  theme(axis.text.y = element_text(size = 6, color = &quot;black&quot;)) +</span><br><span class="line">  ylab(&quot;Sepal Length&quot;) +</span><br><span class="line">  xlab(&quot;Species&quot;) +</span><br><span class="line">  #xlim(-7, 7) +</span><br><span class="line">  ylim(0, 9) +</span><br><span class="line">  theme(axis.title.x = element_text(size = 8)) +</span><br><span class="line">  theme(axis.title.y = element_text(size = 8)) +</span><br><span class="line">  guides(fill = FALSE)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<p>用pdf函数将图片存储为PDF格式，打开就是下面的效果。</p>
<p><img src="https://i.imgur.com/Nl6NDlZ.jpeg" alt="" /></p>
<p>小结：为了制作适合投稿的图片，需要选择合适的背景色、图片尺寸、字体大小。</p>
<h3 id="坐标轴排序问题"><a class="markdownIt-Anchor" href="#坐标轴排序问题"></a> 坐标轴排序问题</h3>
<p>注意，X 坐标轴的元素顺序默认是首字母排序，有时候我们想根据实验设计的情况自定义排序，更便于读者理解数据，可以将若干个元素如下设置。应该也有其他方法可以实现。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">#df$x &lt;- factor(df$x,levels=c(&quot;B&quot;,&quot;A&quot;,&quot;D&quot;,&quot;C&quot;,&quot;E&quot;)) # 参考 https://cloud.tencent.com/developer/article/1711047</span><br><span class="line">a$Species &lt;- factor(a$Species, level=c(&quot;virginica&quot;, &quot;setosa&quot;, &quot;versicolor&quot;))</span><br></pre></td></tr></table></figure>
<p>自定义 X 轴元素顺序之后，再执行画图代码，就可得到新的图。</p>
<p><img src="https://i.imgur.com/JUJP8CC.jpeg" alt="" /></p>
<h3 id="统计显著性20221109-补充"><a class="markdownIt-Anchor" href="#统计显著性20221109-补充"></a> 统计显著性（20221109 补充）</h3>
<p>一般杂志对投稿的要求，还需要加上统计显著性。可以自己用 R 自带的 <code>t.test</code> 或 <code>anova</code> 等函数手动计算，再去 AI 软件手动添加显著性标记。</p>
<p>我最近看到一篇博客介绍用代码自动添加显著性标记，很有参考性：<a target="_blank" rel="noopener" href="https://shixiangwang.github.io/home/cn/post/ggpubr-add-pvalue-and-siglevels/">王诗翔：基于ggpubr包为ggplot添加p值和显著性标记</a></p>
<p>作者介绍基于 ggpubr 包来进行显著性分析，该包可以用于向图中添加 p 值的 R 函数：<code>compare_means()</code> 和 <code>stat_compare_means()</code>。演示数据集是　R　内置的　ToothGrowth　数据集：<code>data(&quot;ToothGrowth&quot;)</code>。</p>
<h4 id="比较均值的方法"><a class="markdownIt-Anchor" href="#比较均值的方法"></a> 比较均值的方法：</h4>
<p><a target="_blank" rel="noopener" href="http://www.sthda.com/english/wiki/comparing-means-in-r">http://www.sthda.com/english/wiki/comparing-means-in-r</a> 包含了均值方法比较的详细描述，这里我们汇总常见的均值比较方法：</p>
<table>
<thead>
<tr>
<th>方法</th>
<th>R 函数</th>
<th>描述</th>
</tr>
</thead>
<tbody>
<tr>
<td>T检验</td>
<td>t.test()</td>
<td>比较两组 (参数)</td>
</tr>
<tr>
<td>Wilcoxon 检验</td>
<td>wilcox.test()</td>
<td>比较两组 (非参数)</td>
</tr>
<tr>
<td>ANOVA</td>
<td>aov() or anova()</td>
<td>比较多组 (参数)</td>
</tr>
<tr>
<td>Kruskal-Wallis</td>
<td>kruskal.test()</td>
<td>比较多组 (非参数)</td>
</tr>
</tbody>
</table>
<h4 id="比较两个独立组别"><a class="markdownIt-Anchor" href="#比较两个独立组别"></a> 比较两个独立组别</h4>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"># 执行检验。默认执行method=&quot;wilcox.test&quot;，你可以指定method = &quot;t.test&quot;进行t检验。</span><br><span class="line">compare_means(len ~ supp, data = ToothGrowth)</span><br><span class="line">#&gt; # A tibble: 1 x 8</span><br><span class="line">#&gt;   .y.   group1 group2      p p.adj p.format p.signif method  </span><br><span class="line">#&gt;   &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;   </span><br><span class="line">#&gt; 1 len   OJ     VC     0.0645 0.064 0.064    ns       Wilcoxon</span><br><span class="line"></span><br><span class="line"># 创建一个带p值的箱线图：</span><br><span class="line">p &lt;- ggboxplot(ToothGrowth, x=&quot;supp&quot;,</span><br><span class="line">               y = &quot;len&quot;, color = &quot;supp&quot;,</span><br><span class="line">               palette = &quot;jco&quot;, add = &quot;jitter&quot;)</span><br><span class="line"># 添加p值</span><br><span class="line">p + stat_compare_means()</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-8-1.png" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># 改变方法</span><br><span class="line">p + stat_compare_means(method = &quot;t.test&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-8-2.png" alt="" /></p>
<h4 id="比较两组配对样本"><a class="markdownIt-Anchor" href="#比较两组配对样本"></a> 比较两组配对样本</h4>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"># 执行检验：</span><br><span class="line">compare_means(len ~ supp, data = ToothGrowth,</span><br><span class="line">              paired = TRUE)</span><br><span class="line">#&gt; # A tibble: 1 x 8</span><br><span class="line">#&gt;   .y.   group1 group2       p  p.adj p.format p.signif method  </span><br><span class="line">#&gt;   &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;   </span><br><span class="line">#&gt; 1 len   OJ     VC     0.00431 0.0043 0.0043   **       Wilcoxon</span><br><span class="line"></span><br><span class="line">#使用ggpaired()函数可视化：</span><br><span class="line">ggpaired(ToothGrowth, x=&quot;supp&quot;, y=&quot;len&quot;,</span><br><span class="line">         color=&quot;supp&quot;, line.color=&quot;gray&quot;,</span><br><span class="line">         line.size=0.4, palette = &quot;jco&quot;) + </span><br><span class="line">    stat_compare_means(paired = TRUE)</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-12-1.png" alt="" /></p>
<h4 id="多组样本组内两两之间的比较如果分组变量包含超过两个水平配对检验会自动执行"><a class="markdownIt-Anchor" href="#多组样本组内两两之间的比较如果分组变量包含超过两个水平配对检验会自动执行"></a> 多组样本：组内两两之间的比较，如果分组变量包含超过两个水平，配对检验会自动执行</h4>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"># 执行成对比较</span><br><span class="line">compare_means(len ~ dose, data = ToothGrowth)</span><br><span class="line">#&gt; # A tibble: 3 x 8</span><br><span class="line">#&gt;   .y.   group1 group2            p      p.adj p.format p.signif method  </span><br><span class="line">#&gt;   &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;         &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;   </span><br><span class="line">#&gt; 1 len   0.5    1      0.00000702   0.000014   7.0e-06  ****     Wilcoxon</span><br><span class="line">#&gt; 2 len   0.5    2      0.0000000841 0.00000025 8.4e-08  ****     Wilcoxon</span><br><span class="line">#&gt; 3 len   1      2      0.000177     0.00018    0.00018  ***      Wilcoxon</span><br><span class="line"># 可视化： 指定你想要比较的组别</span><br><span class="line">my_comparisons &lt;- list(c(&quot;0.5&quot;,&quot;1&quot;), c(&quot;1&quot;, &quot;2&quot;),</span><br><span class="line">                       c(&quot;0.5&quot;, &quot;2&quot;))</span><br><span class="line">ggboxplot(ToothGrowth, x=&quot;dose&quot;, y=&quot;len&quot;,</span><br><span class="line">          color=&quot;dose&quot;, palette = &quot;jco&quot;) +</span><br><span class="line">    stat_compare_means(comparisons = my_comparisons) + #添加成对p值</span><br><span class="line">    stat_compare_means(label.y = 50) # 添加全局p值</span><br><span class="line">#&gt; Warning in wilcox.test.default(c(4.2, 11.5, 7.3, 5.8, 6.4, 10, 11.2, 11.2, :</span><br><span class="line">#&gt; cannot compute exact p-value with ties</span><br><span class="line">#&gt; Warning in wilcox.test.default(c(4.2, 11.5, 7.3, 5.8, 6.4, 10, 11.2, 11.2, :</span><br><span class="line">#&gt; cannot compute exact p-value with ties</span><br><span class="line">#&gt; Warning in wilcox.test.default(c(16.5, 16.5, 15.2, 17.3, 22.5, 17.3, 13.6, :</span><br><span class="line">#&gt; cannot compute exact p-value with ties</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-16-1.png" alt="" /></p>
<p>如果你想要指定 P 值标注线精确的 Y 轴位置，使用 <code>label.y</code> 参数，参考该作者原文。</p>
<h4 id="多组样本基于参考组的多重成对比较"><a class="markdownIt-Anchor" href="#多组样本基于参考组的多重成对比较"></a> 多组样本：基于参考组的多重成对比较</h4>
<p>使用场景：比如对某个样品按照处理进行若干个时间的检测，可以用初始时间样品作为其他样品的参考点。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"># 基于参考组</span><br><span class="line">compare_means(len ~ dose, data = ToothGrowth,</span><br><span class="line">              ref.group = &quot;0.5&quot;, </span><br><span class="line">              method = &quot;t.test&quot;)</span><br><span class="line">#&gt; # A tibble: 2 x 8</span><br><span class="line">#&gt;   .y.   group1 group2        p    p.adj p.format p.signif method</span><br><span class="line">#&gt;   &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt; </span><br><span class="line">#&gt; 1 len   0.5    1      1.27e- 7 1.30e- 7 1.3e-07  ****     T-test</span><br><span class="line">#&gt; 2 len   0.5    2      4.40e-14 8.80e-14 4.4e-14  ****     T-test</span><br><span class="line"># 可视化</span><br><span class="line">ggboxplot(ToothGrowth, x=&quot;dose&quot;, y=&quot;len&quot;,</span><br><span class="line">          color=&quot;dose&quot;, palette = &quot;jco&quot;) + </span><br><span class="line">    stat_compare_means(method=&quot;anova&quot;, label.y=40) + </span><br><span class="line">    stat_compare_means(label=&quot;p.signif&quot;, method=&quot;t.test&quot;,</span><br><span class="line">                       ref.group = &quot;0.5&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-19-1.png" alt="" /></p>
<h4 id="多组样本以所有值为基准base-mean进行多个成对比较"><a class="markdownIt-Anchor" href="#多组样本以所有值为基准base-mean进行多个成对比较"></a> 多组样本：以所有值为基准（base-mean）进行多个成对比较</h4>
<p>如果出现很多组别，两两比较过于复杂，通过将所有数据汇总创建一个虚拟的样本，以它为基准进行比较。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"># Comparison of each group against base-mean</span><br><span class="line">compare_means(len ~ dose,  data = ToothGrowth, ref.group = &quot;.all.&quot;,</span><br><span class="line">              method = &quot;t.test&quot;)</span><br><span class="line">#&gt; # A tibble: 3 x 8</span><br><span class="line">#&gt;   .y.   group1 group2           p      p.adj p.format p.signif method</span><br><span class="line">#&gt;   &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt; </span><br><span class="line">#&gt; 1 len   .all.  0.5    0.000000290 0.00000087 2.9e-07  ****     T-test</span><br><span class="line">#&gt; 2 len   .all.  1      0.512       0.51       0.51     ns       T-test</span><br><span class="line">#&gt; 3 len   .all.  2      0.000000425 0.00000087 4.3e-07  ****     T-test</span><br><span class="line">可视化</span><br><span class="line"></span><br><span class="line">ggboxplot(ToothGrowth, x = &quot;dose&quot;, y = &quot;len&quot;,</span><br><span class="line">          color = &quot;dose&quot;, palette = &quot;jco&quot;)+</span><br><span class="line">  stat_compare_means(method = &quot;anova&quot;, label.y = 40)+      # Add global p-value</span><br><span class="line">  stat_compare_means(label = &quot;p.signif&quot;, method = &quot;t.test&quot;,</span><br><span class="line">                     ref.group = &quot;.all.&quot;)                  # Pairwise comparison against all</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-21-1.png" alt="" /></p>
<p>这个方法有时候会非常有用，比如下面这个例子中，我们可以通过每个样本均值与所有样本的均值进行比较，判断基因水平是过表达还是下调了。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"># Load myeloma data from GitHub</span><br><span class="line">myeloma &lt;- read.delim(&quot;https://raw.githubusercontent.com/kassambara/data/master/myeloma.txt&quot;)</span><br><span class="line"># 执行检验</span><br><span class="line">compare_means(DEPDC1 ~ molecular_group,  data = myeloma,</span><br><span class="line">              ref.group = &quot;.all.&quot;, method = &quot;t.test&quot;)</span><br><span class="line">#&gt; # A tibble: 7 x 8</span><br><span class="line">#&gt;   .y.    group1 group2                      p     p.adj p.format p.signif method</span><br><span class="line">#&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;                   &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt; </span><br><span class="line">#&gt; 1 DEPDC1 .all.  Cyclin D-1       0.288          1.00e+0 0.29     ns       T-test</span><br><span class="line">#&gt; 2 DEPDC1 .all.  Cyclin D-2       0.424          1.00e+0 0.42     ns       T-test</span><br><span class="line">#&gt; 3 DEPDC1 .all.  MMSET            0.578          1.00e+0 0.58     ns       T-test</span><br><span class="line">#&gt; 4 DEPDC1 .all.  MAF              0.254          1.00e+0 0.25     ns       T-test</span><br><span class="line">#&gt; 5 DEPDC1 .all.  Hyperdiploid     0.0000000273   1.90e-7 2.7e-08  ****     T-test</span><br><span class="line">#&gt; 6 DEPDC1 .all.  Proliferation    0.0000239      1.20e-4 2.4e-05  ****     T-test</span><br><span class="line">#&gt; 7 DEPDC1 .all.  Low bone disease 0.00000526     3.20e-5 5.3e-06  ****     T-test</span><br><span class="line"></span><br><span class="line"># 可视化表达谱</span><br><span class="line">ggboxplot(myeloma, x=&quot;molecular_group&quot;, y=&quot;DEPDC1&quot;,</span><br><span class="line">          color=&quot;molecular_group&quot;, add=&quot;jitter&quot;,</span><br><span class="line">          legend=&quot;none&quot;) + </span><br><span class="line">    rotate_x_text(angle = 45) + </span><br><span class="line">    geom_hline(yintercept = mean(myeloma$DEPDC1),</span><br><span class="line">               linetype=2) + # 添加base mean的水平线</span><br><span class="line">     stat_compare_means(method = &quot;anova&quot;, label.y = 1600)+        # Add global annova p-value</span><br><span class="line">  stat_compare_means(label = &quot;p.signif&quot;, method = &quot;t.test&quot;,</span><br><span class="line">                     ref.group = &quot;.all.&quot;)                      # Pairwise comparison against all</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-23-1.png" alt="" /></p>
<h4 id="多个分组变量根据某个变量分组后两个独立样本的比较"><a class="markdownIt-Anchor" href="#多个分组变量根据某个变量分组后两个独立样本的比较"></a> 多个分组变量：根据某个变量分组后两个独立样本的比较</h4>
<p>对于 ToothGrowth 数据集，有两种变量，supp（包含：OJ 和 VC）和 dose（包含：0.5、1 和 2）。下面的代码相当于依据 dose 分成三个数据子集，对子集内部的 OJ 和 VC 进行显著性比较。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">执行检验：</span><br><span class="line"></span><br><span class="line">compare_means(len ~ supp, data = ToothGrowth,</span><br><span class="line">              group.by = &quot;dose&quot;)</span><br><span class="line">#&gt; # A tibble: 3 x 9</span><br><span class="line">#&gt;    dose .y.   group1 group2       p p.adj p.format p.signif method  </span><br><span class="line">#&gt;   &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;   </span><br><span class="line">#&gt; 1   0.5 len   OJ     VC     0.0232  0.046 0.023    *        Wilcoxon</span><br><span class="line">#&gt; 2   1   len   OJ     VC     0.00403 0.012 0.004    **       Wilcoxon</span><br><span class="line">#&gt; 3   2   len   OJ     VC     1       1     1.000    ns       Wilcoxon</span><br><span class="line">因为生成了不同的子图，根据变量分面</span><br><span class="line"></span><br><span class="line"># 根据 &quot;dose&quot; 变量分面绘制箱线图</span><br><span class="line">p &lt;- ggboxplot(ToothGrowth, x = &quot;supp&quot;, y = &quot;len&quot;,</span><br><span class="line">          color = &quot;supp&quot;, palette = &quot;jco&quot;,</span><br><span class="line">          add = &quot;jitter&quot;,</span><br><span class="line">          facet.by = &quot;dose&quot;, short.panel.labs = FALSE)</span><br><span class="line"># Use only p.format as label. Remove method name.</span><br><span class="line">p + stat_compare_means(label = &quot;p.format&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-25-1.png" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># Or use significance symbol as label</span><br><span class="line">p + stat_compare_means(label =  &quot;p.signif&quot;, label.x = 1.5)</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-26-1.png" alt="" /></p>
<p>将这几个图绘制在单个面板内：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">p &lt;- ggboxplot(ToothGrowth, x = &quot;dose&quot;, y = &quot;len&quot;,</span><br><span class="line">          color = &quot;supp&quot;, palette = &quot;jco&quot;,</span><br><span class="line">          add = &quot;jitter&quot;)</span><br><span class="line">p + stat_compare_means(aes(group = supp))</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-27-1.png" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># 仅显示p值</span><br><span class="line">p + stat_compare_means(aes(group = supp), label = &quot;p.format&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-28-1.png" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># 使用显著性标记</span><br><span class="line">p + stat_compare_means(aes(group = supp), label = &quot;p.signif&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-29-1.png" alt="" /></p>
<h4 id="多个分组变量分组后配对样本比较"><a class="markdownIt-Anchor" href="#多个分组变量分组后配对样本比较"></a> 多个分组变量：分组后配对样本比较</h4>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">compare_means(len ~ supp, data = ToothGrowth,</span><br><span class="line">              group.by = &quot;dose&quot;, paired = TRUE)</span><br><span class="line">#&gt; # A tibble: 3 x 9</span><br><span class="line">#&gt;    dose .y.   group1 group2      p p.adj p.format p.signif method  </span><br><span class="line">#&gt;   &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;   </span><br><span class="line">#&gt; 1   0.5 len   OJ     VC     0.0330 0.066 0.033    *        Wilcoxon</span><br><span class="line">#&gt; 2   1   len   OJ     VC     0.0137 0.041 0.014    *        Wilcoxon</span><br><span class="line">#&gt; 3   2   len   OJ     VC     1      1     1.000    ns       Wilcoxon</span><br><span class="line">可视化，按分组变量dose分面创建一个多面板箱线图：</span><br><span class="line"></span><br><span class="line">p &lt;- ggpaired(ToothGrowth, x=&quot;supp&quot;, y=&quot;len&quot;,</span><br><span class="line">               color=&quot;supp&quot;, palette = &quot;jco&quot;,</span><br><span class="line">               line.color = &quot;grey&quot;, line.size =0.4,</span><br><span class="line">               facet.by = &quot;dose&quot;, short.panel.labs = FALSE)</span><br><span class="line"># Use only p.format as label. Remove method name.</span><br><span class="line">p + stat_compare_means(label = &quot;p.format&quot;, paired = TRUE)</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-31-1.png" alt="" /></p>
<h3 id="其他图形"><a class="markdownIt-Anchor" href="#其他图形"></a> 其他图形</h3>
<h4 id="条形图和线图一组变量"><a class="markdownIt-Anchor" href="#条形图和线图一组变量"></a> 条形图和线图（一组变量）</h4>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 条形图加均值标准误</span><br><span class="line">ggbarplot(ToothGrowth, x = &quot;dose&quot;, y = &quot;len&quot;, add = &quot;mean_se&quot;)+</span><br><span class="line">  stat_compare_means() +                                         # Global p-value</span><br><span class="line">  stat_compare_means(ref.group = &quot;0.5&quot;, label = &quot;p.signif&quot;,</span><br><span class="line">                     label.y = c(22, 29))                   # compare to ref.group</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-32-1.png" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 线图加均值标准误</span><br><span class="line">ggline(ToothGrowth, x = &quot;dose&quot;, y = &quot;len&quot;, add = &quot;mean_se&quot;)+</span><br><span class="line">  stat_compare_means() +                                         # Global p-value</span><br><span class="line">  stat_compare_means(ref.group = &quot;0.5&quot;, label = &quot;p.signif&quot;,</span><br><span class="line">                     label.y = c(22, 29)) </span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-32-2.png" alt="" /></p>
<h4 id="条形图和线图两组变量"><a class="markdownIt-Anchor" href="#条形图和线图两组变量"></a> 条形图和线图（两组变量）</h4>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">ggbarplot(ToothGrowth, x = &quot;dose&quot;, y = &quot;len&quot;, add = &quot;mean_se&quot;,</span><br><span class="line">          color = &quot;supp&quot;, palette = &quot;jco&quot;, </span><br><span class="line">          position = position_dodge(0.8))+</span><br><span class="line">  stat_compare_means(aes(group = supp), label = &quot;p.signif&quot;, label.y = 29)</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-33-1.png" alt="" /></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">ggline(ToothGrowth, x = &quot;dose&quot;, y = &quot;len&quot;, add = &quot;mean_se&quot;,</span><br><span class="line">          color = &quot;supp&quot;, palette = &quot;jco&quot;)+</span><br><span class="line">  stat_compare_means(aes(group = supp), label = &quot;p.signif&quot;, </span><br><span class="line">                     label.y = c(16, 25, 29))</span><br></pre></td></tr></table></figure>
<p><img src="https://shixiangwang.github.io/home/cn/post/2019-09-23-ggpubr-add-p-values-and-significance-levels_files/figure-html/unnamed-chunk-33-2.png" alt="" /></p>

    </div>

    
    
    

      <footer class="post-footer">

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/11/07/Illumina-raw-reads-filtering/" rel="prev" title="Bioinfo | 使用 Trimmomatic 软件对 Illumina 测序数据进行过滤">
      <i class="fa fa-chevron-left"></i> Bioinfo | 使用 Trimmomatic 软件对 Illumina 测序数据进行过滤
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/11/20/gene-clusters-analysis-by-plantiSMASH/" rel="next" title="Bioinfo | plantiSMASH 软件鉴定生物合成相关的基因簇">
      Bioinfo | plantiSMASH 软件鉴定生物合成相关的基因簇 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E6%9F%B1%E7%8A%B6%E5%9B%BE-%E6%95%A3%E7%82%B9"><span class="nav-number">1.</span> <span class="nav-text"> 柱状图 + 散点</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%9D%90%E6%A0%87%E8%BD%B4%E6%8E%92%E5%BA%8F%E9%97%AE%E9%A2%98"><span class="nav-number">2.</span> <span class="nav-text"> 坐标轴排序问题</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E7%BB%9F%E8%AE%A1%E6%98%BE%E8%91%97%E6%80%A720221109-%E8%A1%A5%E5%85%85"><span class="nav-number">3.</span> <span class="nav-text"> 统计显著性（20221109 补充）</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#%E6%AF%94%E8%BE%83%E5%9D%87%E5%80%BC%E7%9A%84%E6%96%B9%E6%B3%95"><span class="nav-number">3.1.</span> <span class="nav-text"> 比较均值的方法：</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E6%AF%94%E8%BE%83%E4%B8%A4%E4%B8%AA%E7%8B%AC%E7%AB%8B%E7%BB%84%E5%88%AB"><span class="nav-number">3.2.</span> <span class="nav-text"> 比较两个独立组别</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E6%AF%94%E8%BE%83%E4%B8%A4%E7%BB%84%E9%85%8D%E5%AF%B9%E6%A0%B7%E6%9C%AC"><span class="nav-number">3.3.</span> <span class="nav-text"> 比较两组配对样本</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E5%A4%9A%E7%BB%84%E6%A0%B7%E6%9C%AC%E7%BB%84%E5%86%85%E4%B8%A4%E4%B8%A4%E4%B9%8B%E9%97%B4%E7%9A%84%E6%AF%94%E8%BE%83%E5%A6%82%E6%9E%9C%E5%88%86%E7%BB%84%E5%8F%98%E9%87%8F%E5%8C%85%E5%90%AB%E8%B6%85%E8%BF%87%E4%B8%A4%E4%B8%AA%E6%B0%B4%E5%B9%B3%E9%85%8D%E5%AF%B9%E6%A3%80%E9%AA%8C%E4%BC%9A%E8%87%AA%E5%8A%A8%E6%89%A7%E8%A1%8C"><span class="nav-number">3.4.</span> <span class="nav-text"> 多组样本：组内两两之间的比较，如果分组变量包含超过两个水平，配对检验会自动执行</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E5%A4%9A%E7%BB%84%E6%A0%B7%E6%9C%AC%E5%9F%BA%E4%BA%8E%E5%8F%82%E8%80%83%E7%BB%84%E7%9A%84%E5%A4%9A%E9%87%8D%E6%88%90%E5%AF%B9%E6%AF%94%E8%BE%83"><span class="nav-number">3.5.</span> <span class="nav-text"> 多组样本：基于参考组的多重成对比较</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E5%A4%9A%E7%BB%84%E6%A0%B7%E6%9C%AC%E4%BB%A5%E6%89%80%E6%9C%89%E5%80%BC%E4%B8%BA%E5%9F%BA%E5%87%86base-mean%E8%BF%9B%E8%A1%8C%E5%A4%9A%E4%B8%AA%E6%88%90%E5%AF%B9%E6%AF%94%E8%BE%83"><span class="nav-number">3.6.</span> <span class="nav-text"> 多组样本：以所有值为基准（base-mean）进行多个成对比较</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E5%A4%9A%E4%B8%AA%E5%88%86%E7%BB%84%E5%8F%98%E9%87%8F%E6%A0%B9%E6%8D%AE%E6%9F%90%E4%B8%AA%E5%8F%98%E9%87%8F%E5%88%86%E7%BB%84%E5%90%8E%E4%B8%A4%E4%B8%AA%E7%8B%AC%E7%AB%8B%E6%A0%B7%E6%9C%AC%E7%9A%84%E6%AF%94%E8%BE%83"><span class="nav-number">3.7.</span> <span class="nav-text"> 多个分组变量：根据某个变量分组后两个独立样本的比较</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E5%A4%9A%E4%B8%AA%E5%88%86%E7%BB%84%E5%8F%98%E9%87%8F%E5%88%86%E7%BB%84%E5%90%8E%E9%85%8D%E5%AF%B9%E6%A0%B7%E6%9C%AC%E6%AF%94%E8%BE%83"><span class="nav-number">3.8.</span> <span class="nav-text"> 多个分组变量：分组后配对样本比较</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%85%B6%E4%BB%96%E5%9B%BE%E5%BD%A2"><span class="nav-number">4.</span> <span class="nav-text"> 其他图形</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#%E6%9D%A1%E5%BD%A2%E5%9B%BE%E5%92%8C%E7%BA%BF%E5%9B%BE%E4%B8%80%E7%BB%84%E5%8F%98%E9%87%8F"><span class="nav-number">4.1.</span> <span class="nav-text"> 条形图和线图（一组变量）</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%E6%9D%A1%E5%BD%A2%E5%9B%BE%E5%92%8C%E7%BA%BF%E5%9B%BE%E4%B8%A4%E7%BB%84%E5%8F%98%E9%87%8F"><span class="nav-number">4.2.</span> <span class="nav-text"> 条形图和线图（两组变量）</span></a></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">192</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2023</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
