<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>PacBio测序：专有名词的定义 | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">PacBio测序：专有名词的定义</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">PacBio测序：专有名词的定义</h1><div class="post-meta">2016-02-25<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p><a target="_blank" rel="noopener" href="http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/Portal/_PacBio/_Glossary.htm">http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/Portal\_PacBio\_Glossary.htm</a></p>
<h2 id="SMRT®-Portal-Help"><a href="#SMRT®-Portal-Help" class="headerlink" title="SMRT® Portal Help"></a>SMRT® Portal Help</h2><h1 id="Pacific-Biosciences-Terminology"><a href="#Pacific-Biosciences-Terminology" class="headerlink" title="Pacific Biosciences Terminology"></a>Pacific Biosciences Terminology</h1><h2 id="General-Terminology"><a href="#General-Terminology" class="headerlink" title="General Terminology"></a>General Terminology</h2><ul>
<li>  SMRT® Cell: Consumable substrates comprising arrays of zero-mode waveguide nanostructures.</li>
<li>  Adapters: Exogenous nucleic acids that are ligated to a nucleic acid molecule to be sequenced. For example, SMRTbell™ adapters are hairpin loops that are ligated to both ends of the double stranded DNA insert to produce a SMRTbell™ sequencing template. When adapter sequences are removed from a CCS read, the read is split into multiple subreads.</li>
<li>  Movie: Real-time observation of a SMRT® Cell.</li>
<li>  zero-mode waveguide (ZMW): A nanophotonic device for confining light to a small observation volume. This can be, for example, a small hole in a conductive layer whose diameter is too small to permit the propagation of light in the wavelength range used for detection. Physically part of a SMRT® Cell.</li>
<li>  Sequencing ZMW: A ZMW (zero-mode waveguide) that is expected to be able to produce a sequence if it is populated with a polymerase. ZMWs used for automated SMRT Cell alignment are not considered sequencing ZMWs.</li>
<li>  Run: Specifies</li>
<li>  The wells and SMRT Cells to include in the sequencing run.</li>
<li>  The collection and analysis protocols to use for the selected wells and cells.</li>
</ul>
<h2 id="Read-Terminology"><a href="#Read-Terminology" class="headerlink" title="Read Terminology"></a>Read Terminology</h2><p><img src="http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/3Part_Reads_Def.png"></p>
<ul>
<li>  polymerase read (formerly called “read”): A sequence of nucleotides incorporated by the DNA polymerase while reading a template, such as a circular SMRTbell™ template. Polymerase reads are most useful for quality control of the instrument run. Polymerase read metrics primarily reflect movie length and other run parameters rather than insert size distribution. Polymerase reads are trimmed to include only the high quality region; they include sequences from adapters; and can further include sequence from multiple passes around a circular template.</li>
<li>  subread: Each polymerase read is partitioned to form one or more subreads, which contain sequence from a single pass of a polymerase on a single strand of an insert within a SMRTbell™ template and no adapter sequences. The subreads contain the full set of quality values and kinetic measurements. Subreads are useful for applications like <em>de novo</em> assembly, resequencing, base modification analysis, and so on.</li>
<li>  circular consensus (CCS) read: The consensus sequence determined using subreads taken from a single ZMW. This is not aligned against a reference sequence. In contrast to Reads of Insert, CCS reads require at least two full-pass subreads from the insert.</li>
<li>  read of insert: Represents the highest quality single sequence for an insert, regardless of the number of passes. For example, if your template received one-and-a-half subreads, that information will be combined into a Read of Insert. CCS is an example of a special case where at least two full subreads are collected for an insert. Reads of Insert give the most accurate estimate of the length of the insert sequence loaded onto a SMRT® Cell. For long templates, Reads of Insert may be the same as Polymerase Reads.</li>
</ul>
<h2 id="Read-Length-Terminology"><a href="#Read-Length-Terminology" class="headerlink" title="Read Length Terminology"></a>Read Length Terminology</h2><p><img src="http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/Mapped_Length.png"></p>
<ul>
<li>  mapped polymerase read length: The total number of bases along a read from the first adapter or aligned subread to the last adapter or aligned subread. Approximates the sequence produced by a polymerase in a ZMW.</li>
<li>  mapped subread length: The length of the subread alignment to a target reference sequence. This does not include the adapter sequence.</li>
<li>  polymerase read length: The total number of bases produced from a ZMW after trimming. This may include the adapter sequence.</li>
</ul>
<h2 id="Primary-Analysis-Terminology"><a href="#Primary-Analysis-Terminology" class="headerlink" title="Primary Analysis Terminology"></a>Primary Analysis Terminology</h2><ul>
<li><p>  Primary analysis protocol: Specifies signal processing of the movie, base calling of the traces/pulses, and quality assessment of the base calls. Primary analysis is always performed on the instrument.</p>
</li>
<li><p>  Adapter Screening: Annotates adapter read locations. Used to break a read into subreads during secondary analysis mapping and Circular Consensus.</p>
</li>
<li><p>  High Quality Region Screening: Annotates the high quality sequencing regions of a read to be used during Raw Read Trimming.</p>
</li>
<li><p>  Insert Screening: Annotates insert DNA regions in the Polymerase Read.</p>
</li>
<li><p>  Quality Value Assignment: A prediction of the error probability of a basecall.</p>
</li>
<li><p>  Quality Value (QV): The total probability that the basecall is an insertion or substitution or is preceded by a deletion. QV = -10 * log10(p). For example, QV 20 is 99% accurate, QV 30 is 99.9% accurate, and QV 50 is 99.999% accurate.</p>
</li>
<li><p>  Insertion QV: The probability that the basecall is an insertion with respect to the true sequence.</p>
</li>
<li><p>  Deletion QV: The probability that a deletion error occurred before the current base.</p>
</li>
<li><p>  Substitution QV: The probability that the basecall is a substitution.</p>
</li>
<li><p>  raw read trimming: Extraction of high quality regions from an unfiltered read. Trimming of an unfiltered read produces a polymerase read.</p>
</li>
<li><p>  Read Quality Assignment: A trained prediction of a read’s mapped accuracy based on its pulse and base file characteristics (peak signal-to-noise ratio, average base QV, interpulse distance, and so on). This is used during secondary analysis filtering.</p>
</li>
</ul>
<h2 id="Secondary-Analysis-Terminology"><a href="#Secondary-Analysis-Terminology" class="headerlink" title="Secondary Analysis Terminology"></a>Secondary Analysis Terminology</h2><ul>
<li><p>  Secondary analysis protocol: Specifies how to</p>
</li>
<li><p>  Align a group of reads to a reference sequence to produce a consensus sequence.</p>
</li>
<li><p>  Assemble a set of reads into contigs to produce a <em>de novo</em> sequence.</p>
</li>
<li><p>  Identify insertions, deletions, and SNPs.</p>
</li>
<li><p>  Evaluate consensus quality and quality of the instrument run.</p>
</li>
<li><p>  Consensus: Generation of a consensus sequence from multiple-sequence alignment.</p>
</li>
<li><p>  <em>De Novo</em> Assembly: Assembly of all subreads without a reference sequence.</p>
</li>
<li><p>  Filtering: Removes reads that do not meet the Read Quality and Read Length parameters set by the user. The current default filtering parameters defined by Pacific Biosciences are:</p>
</li>
<li><p>  Read Quality ≥ .75 (as of SMRT Analysis v2.1)</p>
</li>
<li><p>  Read Length ≥ 50 bases</p>
</li>
<li><p>  Mapping: Local alignment of a read or subread to a reference sequence.</p>
</li>
</ul>
<h2 id="Accuracy-Terminology"><a href="#Accuracy-Terminology" class="headerlink" title="Accuracy Terminology"></a>Accuracy Terminology</h2><ul>
<li>  circular consensus accuracy: Accuracy based on multiple sequencing passes around a single circular template molecule.</li>
<li>  consensus accuracy: Accuracy based on aligning multiple sequencing reads or subreads together, optionally with a reference sequence.</li>
<li>  polymerase read quality: A trained prediction of a read’s mapped accuracy based on its pulse and base file characteristics (peak signal-to-noise ratio, average base QV, inter-pulse distance, and so on).</li>
<li>  Subread Accuracy: The post-mapping accuracy of the basecalls.</li>
<li>  Formula: [1 - (errors/subread length)], where errors = number of deletions + insertions + substitutions.</li>
</ul>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2016/02/25/python%E5%A5%97%E4%BB%B6-pbh5tools-pacbio-rawdata-%E6%A0%BC%E5%BC%8F%E8%BD%89%E6%8F%9B/">python套件 pbh5tools: PacBio rawdata 格式轉換</a><a class="next" href="/2016/02/22/linux%E7%94%A8parallel%E5%91%BD%E4%BB%A4%E5%B9%B6%E8%A1%8C%E5%A4%84%E7%90%86%E5%91%BD%E4%BB%A4/">Linux用Parallel命令并行处理命令</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/%E4%B8%80%E4%BA%9BFASTA-FASTQ%E5%BA%8F%E5%88%97%E5%A4%84%E7%90%86%E5%B7%A5%E5%85%B7/">一些FASTA/FASTQ序列处理工具</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/12/%E5%A4%A7%E8%B1%86%E4%BA%A7%E4%B8%9A%E7%8E%B0%E7%8A%B6/">大豆产业现状</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/07/1st_post/">My Fast HexoPages</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/07/hello-world/">Hello World</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/06/%E5%AE%89%E8%A3%85hic-pro/">安装HiC-Pro （HiC数据预处理工具）</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/05/%E4%B8%A4%E4%B8%AA%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%BA%8F%E5%88%97%E7%9B%B8%E4%BA%92%E6%AF%94%E8%BE%83%EF%BC%9Amummer/">两个基因组序列相互比较：Mummer</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/05/shell%E8%84%9A%E6%9C%AC%E5%92%8Cperl%E8%84%9A%E6%9C%AC%E7%9A%84%E6%AE%B5%E6%B3%A8%E9%87%8A/">shell脚本和perl脚本的段注释</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/11/16/R%EF%BC%9Acorrplot-%E7%9B%B8%E5%85%B3%E7%B3%BB%E6%95%B0%E7%9F%A9%E9%98%B5-/">R：corrplot 相关系数矩阵</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/11/16/R%EF%BC%9A%E7%BB%8F%E9%AA%8C%E5%88%86%E5%B8%83%E5%87%BD%E6%95%B0-/">R：经验分布函数</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/11/14/R%EF%BC%9A%E5%88%A0%E9%99%A4%E5%8C%85%E5%90%AB%E7%BC%BA%E5%A4%B1%E5%80%BCNA%E7%9A%84%E8%A1%8C-/">R：删除包含缺失值NA的行</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2020 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>