<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2016/03/03/R%EF%BC%9A%E6%9D%A1%E5%BD%A2%E5%9B%BEbar-plot%E4%B8%8A%E7%9A%84%E8%AF%AF%E5%B7%AE%E7%BA%BFerror-barsd-se-or-ci-/">R：条形图(bar plot)上的误差线(Error Bar):SD, SE or CI?</a></h1><div class="post-meta">2016-03-03</div><div class="post-content"><p>转载自：<a target="_blank" rel="noopener" href="http://yufree.cn/blogcn/2013/08/18/error-bar.html">http://yufree.cn/blogcn/2013/08/18/error-bar.html</a></p>
<p>经常会遇到有人问条形图上误差线画什么的问题，有人说标准差（sd），有人说标准误（se），有的直接说置信区间（CI），其实这倒也不是什么大问题，你按什么画就在文章中注明就是了。后来看到JCB上有一篇科普<a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2064100/">文章</a>，分析的比较到位，就把里面的干货跳出来翻译一下并对其中的难点进行解读，既是总结也是提高，懒得看过程可直接看文末的规则。</p></div><p class="readmore"><a href="/2016/03/03/R%EF%BC%9A%E6%9D%A1%E5%BD%A2%E5%9B%BEbar-plot%E4%B8%8A%E7%9A%84%E8%AF%AF%E5%B7%AE%E7%BA%BFerror-barsd-se-or-ci-/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/03/01/how-to-use-deseq2-to-analyse-rnaseq-data/">How to use DESeq2 to analyse RNAseq data</a></h1><div class="post-meta">2016-03-01</div><div class="post-content"><h1 id="How-to-use-DESeq2-to-analyse-RNAseq-data"><a href="#How-to-use-DESeq2-to-analyse-RNAseq-data" class="headerlink" title="How to use DESeq2 to analyse RNAseq data"></a>How to use DESeq2 to analyse RNAseq data</h1><p>（<a target="_blank" rel="noopener" href="http://dwheelerau.com/2014/02/17/how-to-use-deseq2-to-analyse-rnaseq-data/%EF%BC%89">http://dwheelerau.com/2014/02/17/how-to-use-deseq2-to-analyse-rnaseq-data/）</a></p></div><p class="readmore"><a href="/2016/03/01/how-to-use-deseq2-to-analyse-rnaseq-data/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/02/29/rnaseq-pipeline-alignment-to-de-analysis/">DESeq2: RNAseq pipeline – Alignment to DE analysis</a></h1><div class="post-meta">2016-02-29</div><div class="post-content"><h1 id="RNAseq-pipeline-–-alignment-to-DE-analysis"><a href="#RNAseq-pipeline-–-alignment-to-DE-analysis" class="headerlink" title="RNAseq pipeline – alignment to DE analysis"></a><a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/2014/11/10/rnaseq-pipeline-alignment-to-de-analysis-gsea/">RNAseq pipeline – alignment to DE analysis</a></h1><p><a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/2014/11/10/rnaseq-pipeline-alignment-to-de-analysis-gsea/#comments">1 Reply</a></p></div><p class="readmore"><a href="/2016/02/29/rnaseq-pipeline-alignment-to-de-analysis/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/02/29/python%E6%A8%A1%E5%9D%97%E7%9A%84%E5%A4%9A%E7%A7%8D%E5%AE%89%E8%A3%85%E6%96%B9%E6%B3%95/">Python：Python 模块的多种安装方法</a></h1><div class="post-meta">2016-02-29</div><div class="post-content"><p><strong>Python模块安装方法</strong><br><strong>一、方法1： 单文件模块</strong><br>直接把文件拷贝到 <code>$python\_dir/Lib</code></p>
<p><strong>二、方法2： 多文件模块，带setup.py</strong><br>下载模块包，进行解压，进入模块文件夹，执行： <code>python setup.py install</code></p></div><p class="readmore"><a href="/2016/02/29/python%E6%A8%A1%E5%9D%97%E7%9A%84%E5%A4%9A%E7%A7%8D%E5%AE%89%E8%A3%85%E6%96%B9%E6%B3%95/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/02/29/linux%E7%B3%BB%E7%BB%9F%EF%BC%9Apython2-6%E5%92%8Cpython2-7%E5%90%8C%E6%97%B6%E5%AD%98%E5%9C%A8/">Linux系统：Python2.6和Python2.7同时存在</a></h1><div class="post-meta">2016-02-29</div><div class="post-content"><p>CENTOS 6.X 系列默认安装的 Python 2.6 ，目前开发中主要是使用 Python 2.7 ，这两个版本之间还是有不少差异的，程序在 Python 2.6 下经常会出问题。 比如： <code>re.sub</code> 函数 ，2.7 支持 <code>flags</code> 参数，而 2.6 却不支持。 所以，打算安装 Python 2.7 来运行 Flask 应用程序，但 2.6 不能删除，因为系统对它有依赖。</p></div><p class="readmore"><a href="/2016/02/29/linux%E7%B3%BB%E7%BB%9F%EF%BC%9Apython2-6%E5%92%8Cpython2-7%E5%90%8C%E6%97%B6%E5%AD%98%E5%9C%A8/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/02/28/cuffmerge%E7%9A%84%E7%BB%93%E6%9E%9C%E4%B8%AD%E5%B0%86%E7%9B%B8%E9%82%BB%E7%9A%84%E5%9F%BA%E5%9B%A0merge%E6%88%90%E4%B8%80%E4%B8%AA%E5%9F%BA%E5%9B%A0/">cuffmerge的结果中将相邻的基因merge成一个“基因”</a></h1><div class="post-meta">2016-02-28</div><div class="post-content"><p>cuffmerge combining genes from adjacent loci?? <a target="_blank" rel="noopener" href="https://www.biostars.org/p/104551/">https://www.biostars.org/p/104551/</a> <a target="_blank" rel="noopener" href="http://seqanswers.com/forums/showthread.php?t=19772">http://seqanswers.com/forums/showthread.php?t=19772</a> <a target="_blank" rel="noopener" href="http://seqanswers.com/forums/showthread.php?t=48571">http://seqanswers.com/forums/showthread.php?t=48571</a> <a target="_blank" rel="noopener" href="https://groups.google.com/forum/#!searchin/tuxedo-tools-users/cuffmerge/tuxedo-tools-users/HSZmU\_KDflo/Zz5mMtZkW6wJ">https://groups.google.com/forum/#!searchin/tuxedo-tools-users/cuffmerge/tuxedo-tools-users/HSZmU\_KDflo/Zz5mMtZkW6wJ</a></p></div><p class="readmore"><a href="/2016/02/28/cuffmerge%E7%9A%84%E7%BB%93%E6%9E%9C%E4%B8%AD%E5%B0%86%E7%9B%B8%E9%82%BB%E7%9A%84%E5%9F%BA%E5%9B%A0merge%E6%88%90%E4%B8%80%E4%B8%AA%E5%9F%BA%E5%9B%A0/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/02/28/rna-seq%E8%BD%AC%E5%BD%95%E7%BB%84%E6%B5%8B%E5%BA%8F%E8%B4%A8%E6%8E%A7%E5%B7%A5%E5%85%B7%EF%BC%9Akraken/">RNA Seq转录组测序质控工具：Kraken</a></h1><div class="post-meta">2016-02-28</div><div class="post-content"><h2 id="Kraken-A-set-of-tools-for-quality-control-and-analysis-of-high-throughput-sequence-data"><a href="#Kraken-A-set-of-tools-for-quality-control-and-analysis-of-high-throughput-sequence-data" class="headerlink" title="Kraken: A set of tools for quality control and analysis of high-throughput sequence data"></a>Kraken: A set of tools for quality control and analysis of high-throughput sequence data</h2><p>Matthew P.A. Davis, Stijn van Dongen, Cei Abreu-Goodger, Nenad Bartonicek and Anton J. Enright <em>Methods (2013) - Volume 63, Issue 1, 1 September 2013, Pages 41–49</em> </p></div><p class="readmore"><a href="/2016/02/28/rna-seq%E8%BD%AC%E5%BD%95%E7%BB%84%E6%B5%8B%E5%BA%8F%E8%B4%A8%E6%8E%A7%E5%B7%A5%E5%85%B7%EF%BC%9Akraken/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/02/26/%E6%97%A0root%E6%9D%83%E9%99%90%E5%AE%89%E8%A3%85python/">无root权限安装Python</a></h1><div class="post-meta">2016-02-26</div><div class="post-content"><p>RedHat系统自带python2.6，一些用户程序的要求python2.7以上，但是2.6版本的python不能随便删除，因为系统程序会用到。 可以在用户目录下安装python2.7 方法： 1.安装python</p>
<blockquote>
<p>Python ： <a target="_blank" rel="noopener" href="http://www.python.org/getit/">http://www.python.org/getit/</a></p>
<p>cd $HOME/local/app/</p>
<p>tar -zxvf Python-2.7.11.tgz</p>
<p>mv  Python-2.7.11 python</p>
<p>cd python</p>
<p>./configure –-prefix=<strong>$HOME/local</strong>（你准备安装的路径，可自己更改）</p>
<p>make &amp;&amp; make install</p>
</blockquote>
<p>2. 将指定路径下的lib和bin分别加入环境变量中：</p>
<blockquote>
<p>echo “PYTHONPATH=$HOME/local/lib/python2.7/site-packages” &gt;&gt; .bashrc</p>
<p>echo “export PATH=$HOME/local/app/python:$PATH” &gt;&gt; .bashrc</p>
<p>source ~/.bashrc</p>
</blockquote>
<p>这样就实现了在没有root权限的情况下安装python的模块。 3. 如何在无root权限下安装python模块呢？ 按照上面的方法安装python之后，再安装setuptools，（即easy_install），从而可以使用easy_install来自动安装python包。</p>
<blockquote>
<p>tar -zxvf setuptools-0.6c11.tar.gz</p>
<p>cd setuptools-0.6c11</p>
<p>/your_python_dir/bin/python2.7 setup.py build</p>
<p>/your_python_dir/bin/python2.7 setup.py install</p>
</blockquote>
<p>4. 安装其他python模块/套件 先使用easy_install安装pip工具：</p>
<blockquote>
<p>easy_install pip</p>
</blockquote>
<p>在安裝python套件的时候用pip比較好，pip會兼顧各個依存的套件版本，少用easy-install。 查看目前已有的模块：</p>
<blockquote>
<p>pip list</p>
</blockquote>
<p>pip查看需要更新的package</p>
<blockquote>
<p>pip list –outdated</p>
</blockquote>
<h3 id="下面是更详细的pip介绍："><a href="#下面是更详细的pip介绍：" class="headerlink" title="下面是更详细的pip介绍："></a>下面是更详细的pip介绍：</h3><h3 id="pip安装"><a href="#pip安装" class="headerlink" title="pip安装"></a>pip安装</h3></div><p class="readmore"><a href="/2016/02/26/%E6%97%A0root%E6%9D%83%E9%99%90%E5%AE%89%E8%A3%85python/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/02/25/python%E5%A5%97%E4%BB%B6-pbh5tools-pacbio-rawdata-%E6%A0%BC%E5%BC%8F%E8%BD%89%E6%8F%9B/">python套件 pbh5tools: PacBio rawdata 格式轉換</a></h1><div class="post-meta">2016-02-25</div><div class="post-content"><p>2016-09-06更新：pbh5tools是PacBio公司早期推出的数据分析工具，其后已经被SMRTanalysis工具所替换。SMRTanalysis的安装与使用都更简单，Linux安装之后可以直接在网页里面进行数据分析操作。 ====================================== 转载自：<a target="_blank" rel="noopener" href="http://jeffreyy.pixnet.net/">http://jeffreyy.pixnet.net/</a> PacBio 是第三代定序技術 (Third generation sequencing)的技術公司，目前的最新機型為RSII，最長長度可到30kb。 出來的rawdata其檔案格式為hdf5，其中有兩種，一個是bas.h5和bax.h5，其PacBio的文件有寫到以下: “Due to the increased throughput and read lengths achieved by the PacBio® RS II upgrade, this information is now contained in one bas.h5 file and three bax.h5 files.” <a target="_blank" rel="noopener" href="http://files.pacb.com/software/instrument/2.0.0/bas.h5%20Reference%20Guide.pdf">http://files.pacb.com/software/instrument/2.0.0/bas.h5%20Reference%20Guide.pdf</a> hdf5 格式並無法直接觀看，如果要看序列，必須要經過格式轉換，PacBio有出一個軟體為pbh5tools來進行hdf5的格式轉換。   pbh5tools是以python寫的套件，所以安裝前需要安裝其他python套件，在安裝python套件用pip比較好，pip會兼顧各個依存的套件版本，少用easy-install。 <a target="_blank" rel="noopener" href="http://blog.longwin.com.tw/2014/08/python-setup-pip-package-2014/">安裝 PIP 來管理 Python Packages</a> 除了python套件外，其他的套件還有python-devel 和 hdf5與其相關套件。   其安裝其他requirement套件指令為:</p></div><p class="readmore"><a href="/2016/02/25/python%E5%A5%97%E4%BB%B6-pbh5tools-pacbio-rawdata-%E6%A0%BC%E5%BC%8F%E8%BD%89%E6%8F%9B/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/02/25/pacbio%E6%B5%8B%E5%BA%8F%EF%BC%9A%E4%B8%93%E6%9C%89%E5%90%8D%E8%AF%8D%E7%9A%84%E5%AE%9A%E4%B9%89/">PacBio测序：专有名词的定义</a></h1><div class="post-meta">2016-02-25</div><div class="post-content"><p><a target="_blank" rel="noopener" href="http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/Portal/_PacBio/_Glossary.htm">http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/Portal\_PacBio\_Glossary.htm</a></p></div><p class="readmore"><a href="/2016/02/25/pacbio%E6%B5%8B%E5%BA%8F%EF%BC%9A%E4%B8%93%E6%9C%89%E5%90%8D%E8%AF%8D%E7%9A%84%E5%AE%9A%E4%B9%89/">Read More</a></p></div><nav class="page-navigator"><a class="extend prev" rel="prev" href="/page/9/">Previous</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/8/">8</a><a class="page-number" href="/page/9/">9</a><span class="page-number current">10</span><a class="page-number" href="/page/11/">11</a><a class="page-number" href="/page/12/">12</a><span class="space">&hellip;</span><a class="page-number" href="/page/15/">15</a><a class="extend next" rel="next" href="/page/11/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-What-the-differences-between-DDA-and-DIA-in-mass-spectrometer/">Proteomics | 质谱分析中DDA与DIA有什么不同？</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-Protein-Persulfidation/">Proteomics | 蛋白质硫巯基化修饰</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/Proteomics-Common-reagents-and-their-functions-in-LC-MS-MS-proteomics/">Proteomics | Common reagents and their functions in LC-MS/MS proteomics</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/R-Using-ggrepel-to-repel-overlapping-text-labels-in-ggplot2/">R | Using ggrepel to repel overlapping text labels in ggplot2</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/08/Bioinfo-Installation-of-BUSCO-v4/">Bioinfo | Installation of BUSCO v4</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/05/Linux-Installation-of-Tomcat-9/">Linux | Installation of Tomcat 9, Java 8, and MySQL 5.7.20</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/04/Remove-Background-from-Image/">Remove Background from Image</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/01/%E5%9C%A8Linux%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E4%BD%BF%E7%94%A8Nginx%E9%83%A8%E7%BD%B2Hexo%E7%AB%99%E7%82%B9/">在Linux服务器上使用Nginx部署Hexo站点</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/29/Linux%EF%BC%9A%E4%BD%BF%E7%94%A8Modules%E7%AE%A1%E7%90%86%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/">Linux：使用Environment Modules管理环境变量</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E4%BD%BF%E7%94%A8Snakemake%E6%90%AD%E5%BB%BA%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/">使用Snakemake搭建生物信息分析流程</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>