<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2016/07/25/%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8mega%E6%9E%84%E5%BB%BA%E8%BF%9B%E5%8C%96%E6%A0%91/">学习笔记：使用MEGA构建进化树</a></h1><div class="post-meta">2016-07-25</div><div class="post-content"><p>参考资料： <a target="_blank" rel="noopener" href="http://mbe.oxfordjournals.org/content/early/2013/03/13/molbev.mst012.full">Building Phylogenetic Trees from Molecular Data with MEGA</a> 如果需要对某个基因做进化树，首先需要得到其homologues，MEGA自带的blast可以做，提取fasta格式文件并作整理，序列名称遵循一定的规范，去掉其中多余字符。 导入序列之后，MEGA的菜单里alignment选项下有两个工具可以选择：clustal和muscle。推荐muscle，我经过实验测试发现muscle速度更快。如果比对的是编码基因的核酸序列，可见两个选项：<strong>Align DNA</strong> and <strong>Align Codons，</strong>就必须选择后者，即以氨基酸编码的顺序进行比对。 比对完毕的序列，缺口处会以“-”显示，可另存为*.meg格式的文件，或导出为新的fasta文件。随后在file选项下点击phylogenetic analysis，进入进化分析界面。在phylogeny选项下，有各种树选择：NJ、ML、MP等。 文中例子：构建MP树时候，有一个选项是是否选择删除所有的含有gap的column，文章作者认为不必要全部删除，有的可以留下。原文表述：<strong>Gaps/Missing Data Treatment</strong> determines how gaps are handled.<strong>Complete deletion</strong> means that MEGA5 ignores all columns in which there is a gap in any sequence. Unless there are very few gaps, that option can lose a lot of information because it removes a lot of sites from consideration. I prefer the <strong>Partial Deletion</strong> option in which sites with missing data are removed only as the need arises because that option retains more information.</p></div><p class="readmore"><a href="/2016/07/25/%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8mega%E6%9E%84%E5%BB%BA%E8%BF%9B%E5%8C%96%E6%A0%91/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/06/27/preqc-of-sga/">Preqc of SGA</a></h1><div class="post-meta">2016-06-27</div><div class="post-content"><p>[Copied from <a target="_blank" rel="noopener" href="https://groups.google.com/forum/#!msg/sga-users/95dTwpJCARU/oKoq54EZqKwJ\]">https://groups.google.com/forum/#!msg/sga-users/95dTwpJCARU/oKoq54EZqKwJ\]</a></p></div><p class="readmore"><a href="/2016/06/27/preqc-of-sga/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/06/14/t%E6%A3%80%E9%AA%8C%E4%B8%AD%EF%BC%8C%E5%8D%95%E5%B0%BE%E5%92%8C%E5%8F%8C%E5%B0%BE%EF%BC%8C%E6%88%90%E5%AF%B9%E3%80%81%E7%AD%89%E6%96%B9%E5%B7%AE%E5%92%8C%E5%BC%82%E6%96%B9%E5%B7%AE%E7%9A%84%E5%B7%AE/">T检验中，单尾和双尾，成对、等方差和异方差的差异</a></h1><div class="post-meta">2016-06-14</div><div class="post-content"><p>        在做数据检验的时候，通常先要检验数据是否正态，因为正态分布下有非常详尽的理论基础，用起来非常方便，如果数据不服从正态分布怎么办呢？此时就可以用非参数检验，是基于秩（其实就是序号啦）的，也有很多种方法。你可能会问：不管数据是否正态，是否都能用非参数检验呢，其实也是可以的，不过正态下用非参数检验的话，检验功效就不如参数方法了。如果你不知道什么是检验功效的话就去查一查概率统计的课本。为了有较高的检验功效，非正态数据就用非参数方法，正态数据就用参数方法。 至于该做双尾还是单尾检验，是可以根据问题的实际意义进行判断的，如果不好判断就画画散点图，看一个是否明显在另一个之上或之下。 至于方差的要求那是因为在正态分布下，等方差会省去很多麻烦，所以通常假定方差相等，相关的检验理论也非常完善。如果你不知道方差到底是否相等，可以做检验，最直观的方法就是绘制箱线图，或计算样本标准差。</p></div><p class="readmore"><a href="/2016/06/14/t%E6%A3%80%E9%AA%8C%E4%B8%AD%EF%BC%8C%E5%8D%95%E5%B0%BE%E5%92%8C%E5%8F%8C%E5%B0%BE%EF%BC%8C%E6%88%90%E5%AF%B9%E3%80%81%E7%AD%89%E6%96%B9%E5%B7%AE%E5%92%8C%E5%BC%82%E6%96%B9%E5%B7%AE%E7%9A%84%E5%B7%AE/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/06/09/%E8%BF%9B%E5%8C%96%E5%8E%8B%E5%8A%9B%E5%88%86%E6%9E%90%EF%BC%9Akaks%E6%AF%94%E7%8E%87/">生物进化之选择压力分析：Ka/Ks比率</a></h1><div class="post-meta">2016-06-09</div><div class="post-content"><h3 id="背景知识"><a href="#背景知识" class="headerlink" title="背景知识"></a><strong>背景知识</strong></h3><p>In <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Genetics" title="Genetics">genetics</a>, the <strong>Ka/Ks ratio</strong> is an indicator of <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Selective_pressure" title="Selective pressure">selective pressure</a> (the force applied by <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Natural_selection" title="Natural selection">natural selection</a>) acting on a protein-coding gene. It is calculated as the ratio of the number of <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Nonsynonymous_substitution" title="Nonsynonymous substitution">nonsynonymous substitutions</a> per non-synonymous site (Ka), in a given period of time, to the number of <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Synonymous_substitution" title="Synonymous substitution">synonymous substitutions</a> per synonymous site (Ks), in the same period. <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Homologous_gene" title="Homologous gene">Homologous genes</a> with a Ka/Ks ratio above 1 are evolving under <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Directional_selection" title="Directional selection">positive selection</a>, meaning that at least some of the <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Mutation" title="Mutation">mutations</a> concerned must be advantageous. If all the mutations are neutral or disadvantageous, the ratio will be in the range 0 to 1. However, if some of the mutations are advantageous and some disadvantageous, the ratio could be less than 1.</p></div><p class="readmore"><a href="/2016/06/09/%E8%BF%9B%E5%8C%96%E5%8E%8B%E5%8A%9B%E5%88%86%E6%9E%90%EF%BC%9Akaks%E6%AF%94%E7%8E%87/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/06/01/%E6%8B%9F%E5%8D%97%E8%8A%A5arabidopsis%E7%9A%84%E5%9F%BA%E5%9B%A0%E8%9B%8B%E7%99%BD%E7%AA%81%E5%8F%98%E4%BD%93%E4%B9%8B%E4%B9%A6%E5%86%99%E8%A7%84%E8%8C%83/">拟南芥Arabidopsis的基因/蛋白/突变体之书写规范</a></h1><div class="post-meta">2016-06-01</div><div class="post-content"><p>（2）拟南芥（_Arabidopsis thaliana_） 基因：野生型基因名用大写的斜体字母表示，如：_EMBRYO_、_DEFECTIVE 1_。野生型基因符号由三个斜体的大写字母组成，如：_EMB1_。突变体基因名用小写斜体字母表示，如：_de-etiolated 1_。突变体基因符号用相应基因座名称的三个小写斜体字母命名，如：_det1_。 蛋白质：用相应的基因符号命名，大写，正体，如：EMB1、EMB2、DET1、DET2。 表型：用基因名或一个描述性的短语表示表型。有时也可以用基因符号来命名表型（正体，首字母大写），并用上标“＋”或“－”代表野生型或突变型，如：Abc＋、Abc－。   参考资料： <a target="_blank" rel="noopener" href="https://www.google.com.hk/url?sa=t&amp;rct=j&amp;q=&amp;esrc=s&amp;source=web&amp;cd=5&amp;cad=rja&amp;uact=8&amp;ved=0ahUKEwi-vLC3gYfNAhUGX5QKHWKoCysQFgg5MAQ&amp;url=http://www.medicine.sdu.edu.cn/0tongzhixinwen/2013/9/caoh/zhuanzhu/2.doc&amp;usg=AFQjCNFH8gNip2KCNwWmqqgY1DNmhVzQYQ&amp;sig2=4fUsnPVhVfLdB0h%5C_c3oxqA">https://www.google.com.hk/url?sa=t&amp;rct=j&amp;q=&amp;esrc=s&amp;source=web&amp;cd=5&amp;cad=rja&amp;uact=8&amp;ved=0ahUKEwi-vLC3gYfNAhUGX5QKHWKoCysQFgg5MAQ&amp;url=http%3A%2F%2Fwww.medicine.sdu.edu.cn%2F0tongzhixinwen%2F2013%2F9%2Fcaoh%2Fzhuanzhu%2F2.doc&amp;usg=AFQjCNFH8gNip2KCNwWmqqgY1DNmhVzQYQ&amp;sig2=4fUsnPVhVfLdB0h\_c3oxqA</a></p></div><p class="readmore"><a href="/2016/06/01/%E6%8B%9F%E5%8D%97%E8%8A%A5arabidopsis%E7%9A%84%E5%9F%BA%E5%9B%A0%E8%9B%8B%E7%99%BD%E7%AA%81%E5%8F%98%E4%BD%93%E4%B9%8B%E4%B9%A6%E5%86%99%E8%A7%84%E8%8C%83/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/06/01/%E4%BD%BF%E7%94%A8fastq-dump%E4%BB%8Encbi%E7%9A%84sra%E6%95%B0%E6%8D%AE%E5%BA%93%E4%B8%8B%E8%BD%BD%E6%95%B0%E6%8D%AE/">使用fastq-dump从NCBI的SRA数据库下载数据</a></h1><div class="post-meta">2016-06-01</div><div class="post-content"><p>测序类的论文，一般需要将原始测序reads数据上传到某个公开的数据库，然后在文章末尾标明数据存储位置和登录号。NCBI的SRA (Sequence Read Archive) 数据库(<a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/sra/">http://www.ncbi.nlm.nih.gov/sra/</a>) 是最常用的存储测序数据的数据库。如何从SRA数据库下载他人公开的数据，以作己用呢？</p></div><p class="readmore"><a href="/2016/06/01/%E4%BD%BF%E7%94%A8fastq-dump%E4%BB%8Encbi%E7%9A%84sra%E6%95%B0%E6%8D%AE%E5%BA%93%E4%B8%8B%E8%BD%BD%E6%95%B0%E6%8D%AE/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/05/26/dogma-chloroplast-genome-annotation/">使用DOGMA对Chloroplast genome注释并用Sequin提交到GenBank</a></h1><div class="post-meta">2016-05-26</div><div class="post-content"><h2 id="DOGMA可以对叶绿体基因组进行注释，并可以生成直接导入sequin软件的注释信息，从而生成Genbank格式的注释，可提交NCBI，也可在OGDRAW进行绘图。"><a href="#DOGMA可以对叶绿体基因组进行注释，并可以生成直接导入sequin软件的注释信息，从而生成Genbank格式的注释，可提交NCBI，也可在OGDRAW进行绘图。" class="headerlink" title="DOGMA可以对叶绿体基因组进行注释，并可以生成直接导入sequin软件的注释信息，从而生成Genbank格式的注释，可提交NCBI，也可在OGDRAW进行绘图。"></a>DOGMA可以对叶绿体基因组进行注释，并可以生成直接导入sequin软件的注释信息，从而生成Genbank格式的注释，可提交NCBI，也可在OGDRAW进行绘图。</h2><h2 id="DOGMA-Help"><a href="#DOGMA-Help" class="headerlink" title="DOGMA Help"></a>DOGMA Help</h2><ol></div><p class="readmore"><a href="/2016/05/26/dogma-chloroplast-genome-annotation/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/04/19/pacbio%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%E6%A0%BC%E5%BC%8F%E9%87%8A%E4%B9%89/">PacBio测序数据格式释义</a></h1><div class="post-meta">2016-04-19</div><div class="post-content"><p>本文摘抄/参考自：<a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Data-files-you-received-from-your-service-provider">https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Data-files-you-received-from-your-service-provider</a> 为什么原始数据中有三个“.bas.h5”文件？原因：实际上在PacBio早期的测序原始数据里，只有一个bas.h5文件，因为后来单个SMRT Cell的数据通量有所提高，从而增加了bas.h5文件用以存储base calling数据。 This page describes some of the data files you received from your service provider.</p></div><p class="readmore"><a href="/2016/04/19/pacbio%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%E6%A0%BC%E5%BC%8F%E9%87%8A%E4%B9%89/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/04/08/%E5%A4%9A%E4%B8%AA%E6%A0%B7%E5%93%81%E7%9A%84denovo-rna-seq%E7%BB%84%E8%A3%85/">多个样品的denovo RNA Seq组装</a></h1><div class="post-meta">2016-04-08</div><div class="post-content"><p>问题：先各自组装，然后合并组装的结果；还是先合并reads，再一起组装？ 应该是合并Reads再进行组装能得到更好的结果。 参考资料：<a target="_blank" rel="noopener" href="http://seqanswers.com/forums/showthread.php?t=33452">http://seqanswers.com/forums/showthread.php?t=33452</a> #9： I’ve tried assembling combined reads vs. assembling separate reads for several species, and combining reads has always given a more contiguous and accurate assembly. If you have &gt; 200 million combined reads, I strongly recommend using Trinity In-Silico Normalization, which will give a good assembly in much shorter run time when compared to non-normalized assembly. 也有人分开组装，然后合并。 合并方法： TGICL CAP3 CD-HIT</p></div><p class="readmore"><a href="/2016/04/08/%E5%A4%9A%E4%B8%AA%E6%A0%B7%E5%93%81%E7%9A%84denovo-rna-seq%E7%BB%84%E8%A3%85/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2016/03/17/%E8%8B%B1%E6%96%87email%E5%86%99%E4%BD%9C%EF%BC%9A%E5%A6%82%E4%BD%95%E5%BC%80%E5%A4%B4/">英文Email写作：如何开头</a></h1><div class="post-meta">2016-03-17</div><div class="post-content"><p><strong>Here are some example opening sentences for emails:</strong> <em>I hope you enjoyed your holiday and are finding it easy to settle back in to work.</em> <em>It was good to meet you in [place]. I hope you had a safe journey home.</em> <em>Thank you very much for your email. I am glad to hear that you and your family are well.</em> <em>Thank you for your prompt reply.</em> <em>I apologise for not replying sooner, but I have been very busy these last few weeks.</em> <em>Thank you for your email of [date]. Please find my reply to your query below.</em> <em>I am writing with regard to XXXX.</em> <em>Thank you for getting in touch with us about XXXX.</em> (Less formal, more friendly) <em>Thank you for contacting us regarding XXXX.</em> (More formal) <em>With reference to your email of [date], I would like to bring the following to your attention.</em> <em>As a follow-up to our phone call this morning, I would like summarise the key issues.</em> <em>Re the question you raised in your previous email, please find my explanation below.</em> <em>Following our meeting on [date] / in [place], please find below a summary of the points we discussed.</em> <em>In reply to your query regarding XXXX, I would like to make the following points.</em> <strong>Phrases best avoided:</strong> <em>I hope this email finds you well.</em> <em>Please be informed that…</em> <em>Please be advised as follows.</em> <em>This email concerns…</em> 参考资料：<a target="_blank" rel="noopener" href="http://blog.harwardcommunications.com/2012/11/06/how-to-start-emails/">http://blog.harwardcommunications.com/2012/11/06/how-to-start-emails/</a></p></div><p class="readmore"><a href="/2016/03/17/%E8%8B%B1%E6%96%87email%E5%86%99%E4%BD%9C%EF%BC%9A%E5%A6%82%E4%BD%95%E5%BC%80%E5%A4%B4/">Read More</a></p></div><nav class="page-navigator"><a class="extend prev" rel="prev" href="/page/9/">Previous</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/8/">8</a><a class="page-number" href="/page/9/">9</a><span class="page-number current">10</span><a class="page-number" href="/page/11/">11</a><a class="page-number" href="/page/12/">12</a><span class="space">&hellip;</span><a class="page-number" href="/page/16/">16</a><a class="extend next" rel="next" href="/page/11/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Bioinfo | LTR identification and analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/17/Linux-Gene-Ontology-GOMAP/">Bioinfo | Gene Ontology - GOMAP</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Linux | 认识 PKG_CONFIG_PATH</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">jcvi 安装与使用，报错以及解决办法</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">Bioinfo | Installation of OrthoMCL-pipeline</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Bioinfo | Using Circos for genome visualization</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">Bioinfo | Repeat sequence analysis by RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer)</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>