<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>使用 Enrichment Map或ClueGO 做基因富集分析+GO网络 | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">使用 Enrichment Map或ClueGO 做基因富集分析+GO网络</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">使用 Enrichment Map或ClueGO 做基因富集分析+GO网络</h1><div class="post-meta">2020-09-04<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>对于一个基因列表，经过通路富集（Gene Ontology 或 KEGG pathways）往往富集到多个通路，数据仍然比较庞大，可以用 Enrichment Map 或 ClueGO 进一步精简，得到下面类似的网络图。</p>
<p><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fng.3785/MediaObjects/41588_2017_Article_BFng3785_Fig15_ESM.jpg?as=webp"></p>
<p>Enrichment map for pathways enriched in susceptibility to ER-negative breast cancer. From: <a target="_blank" rel="noopener" href="https://www.nature.com/articles/ng.3785">Identification of ten variants associated with risk of estrogen-receptor-negative breast cancer</a></p>
<p><img src="https://genehub.files.wordpress.com/2020/09/36897186.png?w=640"></p>
<p>Enrichment Map方法：</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904111451.png?w=1024"></p>
<p>准备工作</p>
<p>安装 Java。我操作系统是 Windows 10，Java官网下载 jdk-14.0.2_windows-x64_bin.exe，选择默认路径安装。最后安装路径在：C:\Program Files\Java\jdk-14.0.2 ，注意此时安装 Cytoscape 会报错“EXE4J_JAVA_HOME, No JVM could be found on your system”，表示找不到 java，原来需要设置环境变量。</p>
<p>将 Java 路径加入环境变量。在桌面右键点击“我的电脑”，选择“属性”，在出来的界面中左侧有“高级系统设置”字样，点击进去，即可在“高级”栏目下见到“环境变量字样”。</p>
<p>在环境变量中，要修改两个地方，一个是添加JAVA_HOME。选择“新建”，变量名填上JAVA_HOME，变量值填上“C:\Program Files\Java\jdk-14.0.2”，在java的安装过程中，默认一直下一步安装，所以装在C盘，如果在安装过程中改了目录，那可能是D盘或者E盘，那么变量值要做相应的更改。<br>还要修改一个地方，就是Path，添加JAVA的变量值到Path中，选择Path，然后点“编辑”，在最后面添加如下语句“%JAVA_HOME%\bin”。</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904101004.png?w=810"></p>
<p>按照上面的步骤设置好环境变量之后，Cytoscape 也能顺利安装了。进入 CytoScape 软件界面，上方栏目有个 App 字样，进去可以按照应用，例如 Enrichment Map，AutoAnnotate，BinGO等。</p>
<p>下面测试基因网络分析。</p>
<p>进入g:Profiler网站，提交基因列表，将对基因的Gene Ontology、KEGG这些信息进行注释。如下图所示结果。在Detailed Results 栏目下有一个 GEM 文件，将其下载，这个文件里面就是通路注释结果，包含“GO.ID Description p.Val FDR Phenotype Genes”这些信息。另外在 Data source 栏目下有一个 combined_name.gmt 下载选项，这个 GMT 文件是“Gene Matrix Tranpose”，也可以用户自己做。（By default, the GSEA desktop software searches the MSigDB gene set database that includes pathways, published gene signatures, microRNA target genes and other gene set types. <strong>The user can also provide a custom database as a text-based ‘Gene Matrix Transposed’ (GMT) file</strong> where each line defines a pathway, with its name, identifier and a list of gene identifiers that match the input gene list.）</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904100014.png?w=1024"></p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904100026.png?w=1024"></p>
<p>将上面的两个文件下载到本地之后，提交到 CytoScape 软件里面。</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904100206.png?w=1024"></p>
<p>点击 Build 即可得到网络图。我遇到中间报错信息提示KEGG的问题，直接选择了跳过，也没影响网络生成。然后再在 Apps 里面选择“AutoAnnotate”，将网络进行 cluster 分析，便可以看到一些大圈圈将若干节点围成了不同的群落，只是图形不算美观，跟人家参考文献里面有差距。似乎不太好调整网络形状，几番尝试之后再AutoAnnotate界面有一个三横的按钮，点击进去有“Layout Clusters”，选择“CoSE Layout”，网络图好看很多。</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904100319.png?w=1024"></p>
<p>上面简单测试了一下g:Profiler + Enrichiment Map + AutoAnnotate 进行网络构建。</p>
<p>问题：</p>
<p>1 如何只绘制某一方面的网络图，例如只关心细胞器/或某个通路？</p>
<p>2 如何将多个基因列表的网络图绘制在一起，用不同颜色标记不同的基因来源？</p>
<p>3 如何自己制作GMT文件</p>
<p>4 其他富集分析软件的结果应该也可以导入Cytoscape，如何对数据框进行调整，才可导入？</p>
<p>[2020.09.26 更新] <strong>ClueGO案例：</strong></p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200916094951.png?w=1024"></p>
<p>ClueGO的操作比Enrichment Map更方便，网络结果的调整也更人性化。</p>
<p>B站有一个ClueGO操作实践的视频：<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1TK411W7TM">https://www.bilibili.com/video/BV1TK411W7TM</a> （作者：<a target="_blank" rel="noopener" href="https://space.bilibili.com/198764903">野菜茎小甜饼</a>）</p>
<p>参考资料</p>
<p>[1] <a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/news/96301">CytoScape 攻略</a></p>
<p>[2] <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/54e70ee78ef9"></a><a href="/Profiler%20%E5%9F%BA%E5%9B%A0%E5%8A%9F%E8%83%BD%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E7%9A%84%E7%A4%BA%E4%BE%8B">g:Profiler 基因功能富集分析的示例</a></p>
<p>[3] <a target="_blank" rel="noopener" href="http://www.360doc.com/content/19/1202/13/50736008_876897857.shtml">Enrichment Map 操作示例</a></p>
<p>[4] Enrichment Map tutorial: <a target="_blank" rel="noopener" href="https://enrichmentmap.readthedocs.io/en/latest/QuickTour.html">https://enrichmentmap.readthedocs.io/en/latest/QuickTour.html</a></p>
<p>[5] AutoAnnotate layout: <a target="_blank" rel="noopener" href="https://autoannotate.readthedocs.io/en/latest/GroupingAndLayout.html#cluster-aware-layouts">https://autoannotate.readthedocs.io/en/latest/GroupingAndLayout.html#cluster-aware-layouts</a></p>
<p>[6] Nature Protocol 2019: Pathway enrichment analysis and visualization of omics data using g:Profiler, GSEA, Cytoscape and EnrichmentMap <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6607905/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6607905/</a></p>
<p>[7] Enrichment Map usage, SlidShare: <a target="_blank" rel="noopener" href="https://slideplayer.com/slide/3900605/">https://slideplayer.com/slide/3900605/</a></p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2020/09/16/ai-%E4%BD%9C%E5%9B%BE%E7%AC%94%E8%AE%B0/">AI 作图笔记</a><a class="next" href="/2020/08/09/%E8%BD%AC%EF%BC%9A%E5%A6%82%E4%BD%95%E4%B8%8A%E4%BC%A0%E4%BB%A3%E7%A0%81%E5%88%B0github/">转：如何上传代码到github</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/01/04/Remove-Background-from-Image/">Remove Background from Image</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/01/%E5%9C%A8Linux%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E4%BD%BF%E7%94%A8Nginx%E9%83%A8%E7%BD%B2Hexo%E7%AB%99%E7%82%B9/">在Linux服务器上使用Nginx部署Hexo站点</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/29/Linux%EF%BC%9A%E4%BD%BF%E7%94%A8Modules%E7%AE%A1%E7%90%86%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/">Linux：使用Environment Modules管理环境变量</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E4%BD%BF%E7%94%A8Snakemake%E6%90%AD%E5%BB%BA%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/">使用Snakemake搭建生物信息分析流程</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9%E7%9A%84SAM-BAM%E6%96%87%E4%BB%B6%E4%B8%AD%EF%BC%8C%E6%80%8E%E4%B9%88%E7%90%86%E8%A7%A3MAPQ%E5%80%BC/">基因组比对的SAM/BAM文件中，怎么理解MAPQ值</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/25/Linux%EF%BC%9Avim%E8%AE%BE%E7%BD%AE/">Linux：vim设置</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/Hi-C%EF%BC%9A3d-dna%E6%8C%82%E8%BD%BD%E6%9F%93%E8%89%B2%E4%BD%93/">Hi-C：3d-dna挂载染色体</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/%E4%B8%80%E4%BA%9BFASTA-FASTQ%E5%BA%8F%E5%88%97%E5%A4%84%E7%90%86%E5%B7%A5%E5%85%B7/">一些FASTA/FASTQ序列处理工具</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/12/%E5%A4%A7%E8%B1%86%E4%BA%A7%E4%B8%9A%E7%8E%B0%E7%8A%B6/">大豆产业现状</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/07/1st_post/">My Fast HexoPages</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>