<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2019/09/04/linux%E7%B3%BB%E7%BB%9F%E4%BB%8E-redhat-4-4-7-3-%E5%8D%87%E7%BA%A7%E5%88%B0-centos-7/">Linux系统从 RedHat 4.4.7-3 升级到 CentOS 7</a></h1><div class="post-meta">2019-09-04</div><div class="post-content"><p>由于Redhat附带的GLIBC 2.12版本过低，导致许多应用（jcvi、interproscan）不能正常使用，遂决定升级到新的系统。 一、下载CentOS的iso文件 进入CentOS网站（<a target="_blank" rel="noopener" href="https://www.centos.org/download/%EF%BC%89%EF%BC%8C%E9%80%89%E6%8B%A9DVD">https://www.centos.org/download/），选择DVD</a> ISO按钮，进入下载网站之后点击如下链接即可下载。文件大小约4.3 GB。 <a target="_blank" rel="noopener" href="http://mirror.sunnyvision.com/centos/7.6.1810/isos/x86/_64/CentOS-7-x86/_64-DVD-1810.iso">http://mirror.sunnyvision.com/centos/7.6.1810/isos/x86\_64/CentOS-7-x86\_64-DVD-1810.iso</a> 二、制作光盘启动盘 我的笔记本电脑自带DVD光盘刻录机，需要安装UltraIso软件用于光盘刻录，同时空白DVD光盘空间要大于CentOS的iso文件。也可以用U盘做启动盘，但是我没有合适的U盘，所以这次用了光盘。将光盘放入刻录机中；打开软件UltraISO；在软件上找到iso文件所在的位置，并双击该iso文件；随后选中DVD；再点击“刻录”按钮即可。 <img src="https://genehub.files.wordpress.com/2019/09/iso-burn-_20190904100355.png" alt="ISO burn _20190904100355"> 三、安装新系统 旧系统在一个1.8 TB的硬盘，已经做了分区，其中50 GB分区是根目录（/），剩下的空间大部分是/home目录。home目录下我已经有多年的数据，这次安装不用动，但是我还是提前备份了。新系统直接安在原有的根目录下。 安装时候参考资料： <a target="_blank" rel="noopener" href="https://www.cnblogs.com/set-config/p/9040407.html">https://www.cnblogs.com/set-config/p/9040407.html</a> <a target="_blank" rel="noopener" href="https://blog.seisman.info/centos7-setup-1/">https://blog.seisman.info/centos7-setup-1/</a> 1. 往服务器插入光盘，重启系统，开机的时候按F2进入bios，设置光盘优先启动，进入如下界面，光标移动选择第一项Install CentOS 7，进入安装。 <img src="https://genehub.files.wordpress.com/2019/09/1349552-20180515112524220-766717605.png" alt="1349552-20180515112524220-766717605"> 2. 系统自动提醒进行一些配置 第二个界面是选择安装语言界面，我相信看这个教程的很多童鞋都是新手或者第一次安装Centos的同胞，对于新手而言，最好选择可以看懂的中文语言，往下拉到底就可以看到中文了，选中它。单机Continue/继续 【或者键盘的Tab键选中+回车确定】   <img src="https://genehub.files.wordpress.com/2019/09/1349552-20180515112948570-582636142.png" alt="1349552-20180515112948570-582636142"> 第三个界面非常关键，这涉及到你系统的基本配置，比如硬盘大小、根分区以及swap分区。 本地化：本地化这一块不需要做过多的配置，按需即可；软件：软件下面的安装源和软件选中，只有特殊需求的童鞋才会用到。这一点从里面的选项描述就可以知道，为了不错误引导新手，这里就不放图了。系统：重点是系统。先说KDUMP，对于新手来说，这个在新手阶段无使用价值，配置启用或者不启用都无关紧要。安装位置，由于安装位置涉及到你自己的磁盘数据以及分区内容，如果原分区包含有数据、有必要先备份数据。 <img src="https://genehub.files.wordpress.com/2019/09/1349552-20180515113632825-1756332663.png" alt="1349552-20180515113632825-1756332663"> 点击上图的“安装位置”，出现计算机上已有的硬盘，选中系统所在盘，并点击“手动配置分区”。（因为我这块硬盘已经被分区了，旧的home目录数据不用变，所以不必用自动分区。这里忘了拍照。）点击完成按钮之后进入下图界面，左侧显示了“新 CentOS 7 安装”和“Red Hat …”两个栏目，可以看到新的CentOS还没有挂载点，旧的Redhat系统有/home, /boot, / 和 swap 这四个分区。依次点击这四个分区，图中挂载点（P）文字下方的空白栏，就是新的CentOS系统对应的分区，要相应地填入 /home, /boot, / 和 swap；同时注意有个“重新格式化（D）”文字左侧的选择框，/home 不要点这个，其他三个要选重新格式化。设备类型和文件系统我都沿用旧系统的，没做更改，看到资料说新手需记住：除了SWAP分区外，其他分区的文件系统一律选择ext4类型,设备类型默认选LVM。 <img src="https://genehub.files.wordpress.com/2019/09/e9858de7bdaee5898d.jpg" alt="配置前"> 手动分区之后点击完成就进入了如下界面，可以看到三个挂载区（对应/boot, / 和 swap）将重新格式化，点击“接收更改”。然后新系统将进入自动安装，期间会提醒设置root账户和一个普通用户账户，然后等几分钟就安好了。取出光盘，重启的时候再将bios里面改成原来的硬盘优先启动，就可以进入新系统了。 <img src="https://genehub.files.wordpress.com/2019/09/e9858de7bdaee5908e.jpg" alt="配置后"> 四、其余配置 对新系统的IP地址等做好设置，即可远程登录了。进入home目录，文件都跟以前的一致，就是所属用户组变成了500（这是啥意思？），我用chown -R username:username ~/ 将文件属性变回了自己的用户名下。   == 20200109 更新== 由于电脑多次经历断电异常关机（没有配置UPS），卡在开机界面无法顺利开机了，进入安全模式下，通过journalctl命令查阅log文件，看到的报错信息如下。</p></div><p class="readmore"><a href="/2019/09/04/linux%E7%B3%BB%E7%BB%9F%E4%BB%8E-redhat-4-4-7-3-%E5%8D%87%E7%BA%A7%E5%88%B0-centos-7/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2019/08/08/%E6%A4%8D%E7%89%A9%E5%9F%BA%E5%9B%A0%E7%BB%84ncrna%E9%A2%84%E6%B5%8B%EF%BC%88trna%E3%80%81rrna%E3%80%81snrna%E3%80%81mirna%EF%BC%89/">植物基因组ncRNA预测（tRNA、rRNA、snRNA、miRNA）</a></h1><div class="post-meta">2019-08-08</div><div class="post-content"><p>引用文献 <img src="https://genehub.files.wordpress.com/2019/08/e883a1e69da8ncrnae989b4e5ae9ae696b9e6b395.png" alt="胡杨ncRNA鉴定方法">    </p>
<h2 id="一、tRNA"><a href="#一、tRNA" class="headerlink" title="一、tRNA"></a>一、tRNA</h2><p>先安装tRNAscan-SE，然后运行如下命令即可。</p></div><p class="readmore"><a href="/2019/08/08/%E6%A4%8D%E7%89%A9%E5%9F%BA%E5%9B%A0%E7%BB%84ncrna%E9%A2%84%E6%B5%8B%EF%BC%88trna%E3%80%81rrna%E3%80%81snrna%E3%80%81mirna%EF%BC%89/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2019/07/31/linux%E6%9B%B4%E6%96%B0git/">Linux更新git</a></h1><div class="post-meta">2019-07-31</div><div class="post-content"><p>系统自带的git无法下载：</p>
<p>error: error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protocol</p>
<p>查询之后据说是 GitHub 不再支持TLSv1/TLSv1.1了，公告地址：<a target="_blank" rel="noopener" href="https://githubengineering.com/crypto-removal-notice/">Weak cryptographic standards removal notice</a> 于是安装新版git（<a target="_blank" rel="noopener" href="https://mirrors.edge.kernel.org/pub/software/scm/git/git-2.15.0.tar.gz%EF%BC%89%EF%BC%8Cmake%E7%9A%84%E8%BF%87%E7%A8%8B%E4%B8%AD%E6%8A%A5%E9%94%99%EF%BC%8C%E6%98%BE%E7%A4%BA%E7%B3%BB%E7%BB%9F%E7%9A%84openssl%E6%9C%89%E9%97%AE%E9%A2%98%E3%80%82">https://mirrors.edge.kernel.org/pub/software/scm/git/git-2.15.0.tar.gz），make的过程中报错，显示系统的openssl有问题。</a></p></div><p class="readmore"><a href="/2019/07/31/linux%E6%9B%B4%E6%96%B0git/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2019/04/01/%E8%BD%AC%EF%BC%9A%E8%87%AA%E5%8A%A8%E5%8C%96%E8%BF%90%E8%A1%8Corthomcl/">转：自动化运行OrthoMCL</a></h1><div class="post-meta">2019-04-01</div><div class="post-content"><p>OrthoMCL (<a target="_blank" rel="noopener" href="https://link.jianshu.com/?t=http://orthomcl.org/orthomcl/">http://orthomcl.org/orthomcl/</a>) 是现在用的最多的一款来找直系同源基因（Orthologs）以及旁系同源基因 (Paralog) 的软件。</p></div><p class="readmore"><a href="/2019/04/01/%E8%BD%AC%EF%BC%9A%E8%87%AA%E5%8A%A8%E5%8C%96%E8%BF%90%E8%A1%8Corthomcl/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2019/03/30/linux%E5%AE%89%E8%A3%85mysql-5-7-25/">Linux安装mysql-5.7.25</a></h1><div class="post-meta">2019-03-30</div><div class="post-content"><p>下面的操作完全是照搬其他人的经验（<a target="_blank" rel="noopener" href="https://segmentfault.com/a/1190000012703513%EF%BC%89%EF%BC%8C%E6%AD%A4%E5%A4%84%E8%AE%B0%E5%BD%95%E5%A4%87%E4%BB%BD%E7%94%A8%E3%80%82%E4%B9%9F%E5%8F%AF%E5%8F%82%E8%80%83%EF%BC%9Ahttps://www.jellythink.com/archives/14">https://segmentfault.com/a/1190000012703513），此处记录备份用。也可参考：https://www.jellythink.com/archives/14</a> ———————————- 由于mysql 5.7.17版本以后 support_files文件夹中无 <strong>my_default.cnf</strong> 文件，所以今天给大家详细描述一下 mysql 5.7.20版本(目前官方最新版)的安装步骤。 第一步：下载mysql最新版</p></div><p class="readmore"><a href="/2019/03/30/linux%E5%AE%89%E8%A3%85mysql-5-7-25/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2019/02/28/%E5%AE%89%E8%A3%85lastz/">安装LASTZ</a></h1><div class="post-meta">2019-02-28</div><div class="post-content"><p>LASTZ是序列比对工具，可以生成AXT格式的比对文件。 安装LASTZ方法如下 下载地址：<a target="_blank" rel="noopener" href="http://www.bx.psu.edu/~rsharris/lastz/">http://www.bx.psu.edu/~rsharris/lastz/</a> If you have received the distribution as a packed archive, unpack it by whatever means are appropriate for your computer. The result should be a directory <code>/lastz‑distrib‑X.XX.XX</code> that contains a <code>src</code> subdirectory (and some others). You may find it convenient to remove the revision number (<code>‑X.XX.XX</code>) from the directory name. Before building or installing any of the programs, you will need to tell the installer where to put the executable, either by setting the shell variable<code>$LASTZ_INSTALL</code>, or by editing the <code>make‑include.mak</code> file to set the definition of <code>installDir</code>. Also, be sure to add the directory you choose to your <code>$PATH</code>. Then to build the LASTZ executable, enter the following commands from bash or a similar command-line shell (Solaris users should substitute <code>gmake</code> for <code>make</code>). This will build two executables (<code>lastz</code> and <code>lastz_D</code>) and copy them into your <code>installDir</code>.</p></div><p class="readmore"><a href="/2019/02/28/%E5%AE%89%E8%A3%85lastz/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2018/11/27/Python%EF%BC%9Apython3%E4%B8%8Epython2%E5%85%B1%E5%AD%98/">Python：python3与python2共存</a></h1><div class="post-meta">2018-11-27</div><div class="post-content"><p><strong>20201208更新：本文已过期。</strong></p>
<p>Linux系统自带了Python2.6，自己又安装了Python2.7和Anaconda3（Python3.6） 今天运行python3.6报错：</p>
<p>[fenglei@localhost lib64]$ /home/fenglei/local/app/anaconda3/bin/python3<br>Failed to import the site module<br>Traceback (most recent call last):<br>File “/usr/local/lib/python2.7/site-packages/site.py”, line 75, in <module><br>__boot()<br>File “/usr/local/lib/python2.7/site-packages/site.py”, line 27, in __boot<br>import imp # Avoid import loop in Python &gt;= 3.3<br>File “/home/fenglei/local/app/anaconda3/lib/python3.6/imp.py”, line 27, in <module><br>import tokenize<br>File “/home/fenglei/local/app/anaconda3/lib/python3.6/tokenize.py”, line 33, in <module><br>import re<br>File “/home/fenglei/local/app/anaconda3/lib/python3.6/re.py”, line 142, in <module><br>class RegexFlag(enum.IntFlag):<br>AttributeError: module ‘enum’ has no attribute ‘IntFlag’</p></div><p class="readmore"><a href="/2018/11/27/Python%EF%BC%9Apython3%E4%B8%8Epython2%E5%85%B1%E5%AD%98/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2018/07/09/diamond-%E6%AF%94blast%E6%9B%B4%E5%BF%AB%E9%80%9F%E7%9A%84%E6%AF%94%E5%AF%B9%E5%B7%A5%E5%85%B7/">Diamond: 比BLAST更快速的比对工具</a></h1><div class="post-meta">2018-07-09</div><div class="post-content"><p><strong>Diamond简介</strong> 序列比对软件，能输出与BLAST类似的结果，但是速度快了数百倍。 <strong>安装</strong> 二进制可以直接解压使用：</p>
<p>wget <a target="_blank" rel="noopener" href="http://github.com/bbuchfink/diamond/releases/download/v0.9.22/diamond-linux64.tar.gz">http://github.com/bbuchfink/diamond/releases/download/v0.9.22/diamond-linux64.tar.gz</a><br>tar xzf diamond-linux64.tar.gz</p></div><p class="readmore"><a href="/2018/07/09/diamond-%E6%AF%94blast%E6%9B%B4%E5%BF%AB%E9%80%9F%E7%9A%84%E6%AF%94%E5%AF%B9%E5%B7%A5%E5%85%B7/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2018/05/16/augustus-3-3%E7%9A%84%E5%AE%89%E8%A3%85/">Augustus-3.3的安装</a></h1><div class="post-meta">2018-05-16</div><div class="post-content"><p>Augustus 是流行的基因预测软件。在安装的过程中，遇到了 make 报错的问题，查询资料之后成功安装。错误信息提示 Augustus 在 bam2hints 编译的时候不通过，导致 make 失败。原来是bam2hints、filterBam 和 bam2wig 这三个目录下的 Makefile 中对其依赖程序的调用需要手动更改。下面部分文字摘抄自 iamphioxus.org 。 Augustus is a very popular tool for gene annotation, however its installation process can be a bit tricky. For example, if we just download and install <em>Augustus</em> like below, it will not work.</p></div><p class="readmore"><a href="/2018/05/16/augustus-3-3%E7%9A%84%E5%AE%89%E8%A3%85/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2018/04/25/R%EF%BC%9A%E7%81%AB%E5%B1%B1%E5%9B%BE-volcano-plot-/">R：火山图 volcano plot</a></h1><div class="post-meta">2018-04-25</div><div class="post-content"><p>两个实验组之间的若干个基因的差异倍数比较，可以通过 Volcano Plot 图形化展示。 <img src="https://genehub.files.wordpress.com/2018/04/qqe688aae59bbe20180425163433.jpg" alt="QQ截图20180425163433"> 例如 DEseq2 对转录组测序的样品做差异基因分析之后，会得到一个基因表达差异的表格的。案例使用数据的来源是 GitHub：<a target="_blank" rel="noopener" href="https://github.com/hbc/fabio-splicing/blob/master/new-samples/2016-07-29/_fabio-newlines/summary/control-vs-e7107-deseq2.csv">https://github.com/hbc/fabio-splicing/blob/master/new-samples/2016-07-29\_fabio-newlines/summary/control-vs-e7107-deseq2.csv</a> 下载上述数据之后，运行下面的命令即可得到文章起始的图。</p></div><p class="readmore"><a href="/2018/04/25/R%EF%BC%9A%E7%81%AB%E5%B1%B1%E5%9B%BE-volcano-plot-/">Read More</a></p></div><nav class="page-navigator"><a class="extend prev" rel="prev" href="/page/6/">Previous</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/5/">5</a><a class="page-number" href="/page/6/">6</a><span class="page-number current">7</span><a class="page-number" href="/page/8/">8</a><a class="page-number" href="/page/9/">9</a><span class="space">&hellip;</span><a class="page-number" href="/page/17/">17</a><a class="extend next" rel="next" href="/page/8/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/mcscanx-%E6%B5%8B%E8%AF%95/">MCScanX 测试</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Bioinfo | LTR identification and analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>