<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta property="og:type" content="website">
<meta property="og:title" content="FENG&#39;s Notes">
<meta property="og:url" content="http://example.com/page/7/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/page/7/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/12/05/Linux-shell-perl-mute-multiple-lines/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/12/05/Linux-shell-perl-mute-multiple-lines/" class="post-title-link" itemprop="url">shell脚本和perl脚本的段注释</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-12-05 12:53:15" itemprop="dateCreated datePublished" datetime="2020-12-05T12:53:15+08:00">2020-12-05</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
                  , 
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>shell脚本段注释</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">#!/bin/bash </span><br><span class="line">echo &quot;Say Something&quot; </span><br><span class="line">&lt;&lt;COMMENT     </span><br><span class="line">  your comment 1     </span><br><span class="line">  comment 2     </span><br><span class="line">  blah </span><br><span class="line">COMMENT </span><br><span class="line">echo &quot;Do something else&quot;</span><br></pre></td></tr></table></figure>
<p>perl脚本段注释。测试了一下，“=”符号和“pod”之间不能加空格，pod不能大写。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">=pod</span><br><span class="line">while(&lt;file&gt;)&#123;</span><br><span class="line">  chomp $_;</span><br><span class="line">  print &quot;$_\n;</span><br><span class="line">&#125;</span><br><span class="line">=cut</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/11/16/R-corrplot-coefficient-correlation-/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/11/16/R-corrplot-coefficient-correlation-/" class="post-title-link" itemprop="url">R：corrplot 相关系数矩阵</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-11-16 12:03:31" itemprop="dateCreated datePublished" datetime="2020-11-16T12:03:31+08:00">2020-11-16</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">install.packages(&quot;corrplot&quot;)</span><br><span class="line">library(&quot;corrplot&quot;)</span><br><span class="line">M = cor(mtcars)</span><br><span class="line">#corrplot(M, order = &quot;hclust&quot;, addrect = 2)</span><br><span class="line">#head(mtcars)</span><br><span class="line">#</span><br><span class="line">#corrplot(M, order = &quot;hclust&quot;, addrect = 2)</span><br><span class="line"># 只需要下半部分三角形</span><br><span class="line">corrplot(M, order = &quot;hclust&quot;, addrect = 2, type=&quot;lower&quot;)</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116095756.png"><img src="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116095756.png?w=894" alt="" /></a></p>
<p>参考资料：</p>
<p>[1] corrplot函数参数调整 <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/28076189">https://zhuanlan.zhihu.com/p/28076189</a></p>
<p>[2] 图形元素设定 <a target="_blank" rel="noopener" href="https://www.cnblogs.com/Mao1518202/p/11270767.html">https://www.cnblogs.com/Mao1518202/p/11270767.html</a></p>
<p>搜资料的时候看到与上述图形类似的点阵图（dot plot），还可以结合聚类分析，顺便记录一下。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"># install.packages(&#x27;tidyverse&#x27;)</span><br><span class="line">library(&#x27;tidyverse&#x27;)</span><br><span class="line"># -- Attaching packages ----------------------------------------------------# -------------------------------------------------------- tidyverse 1.3.0</span><br><span class="line"># √ tibble  3.0.4     √ stringr 1.4.0</span><br><span class="line"># √ readr   1.4.0     √ forcats 0.5.0</span><br><span class="line"># √ purrr   0.3.4     </span><br><span class="line"># -- Conflicts -------------------------------------------------------------#-------------------------------------------------- tidyverse_conflicts() --</span><br><span class="line"># x dplyr::filter() masks stats::filter()</span><br><span class="line"># x dplyr::lag()    masks stats::lag()</span><br><span class="line"></span><br><span class="line"># install.packages(&#x27;ggdendro&#x27;)</span><br><span class="line">library(ggdendro)</span><br><span class="line"></span><br><span class="line">library(cowplot)</span><br><span class="line"></span><br><span class="line"># BiocManager::install(&#x27;ggtree&#x27;)</span><br><span class="line">library(ggtree)</span><br><span class="line"></span><br><span class="line">library(patchwork) </span><br><span class="line"></span><br><span class="line">gene_cluster &lt;- read_tsv(&#x27;https://github.com/davemcg/davemcg.github.io/raw/master/content/post/scRNA_dotplot_data.tsv.gz&#x27;)</span><br><span class="line"></span><br><span class="line">gene_cluster %&gt;% sample_n(5)</span><br><span class="line"></span><br><span class="line">markers &lt;- gene_cluster$Gene %&gt;% unique()</span><br></pre></td></tr></table></figure>
<p>#    gene_cluster %&gt;% sample_n(5)</p>
<h1 id="gene_cluster-6列-888行-的矩阵下面显示前5行"><a class="markdownIt-Anchor" href="#gene_cluster-6列-888行-的矩阵下面显示前5行"></a> gene_cluster 6列 888行 的矩阵，下面显示前5行</h1>
<p>A tibble: 5 x 6<br />
Gene    cluster cell_ct cell_exp_ct   count Group</p>
<p>1 CELA1   c16        1278          87 0.0826  Neuron<br />
2 NDUFA1  c13        1725           5 0.00290 Neuron<br />
3 IGFBP7  c13        1725        1013 1.54    Neuron<br />
4 IGFBP7  c09        2124          29 0.0154  Astrocyte<br />
5 SLC10A7 c13        1725          59 0.0413  Neuron</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">   gene_cluster %&gt;% filter(Gene %in% markers) %&gt;% </span><br><span class="line">      mutate(% Expressing = (cell_exp_ct/cell_ct) * 100) %&gt;% </span><br><span class="line">   ggplot(aes(x=cluster, y = Gene, color = count, size = % Expressing)) + </span><br><span class="line">   geom_point()     </span><br><span class="line"># mutate函数来自dplyr，功能是Create, modify, and delete columns</span><br><span class="line"># mutate函数计算一个比例数值，作为矩阵第七列，名称为“% Expression”。</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/01.png"><img src="https://genehub.files.wordpress.com/2020/11/01.png?w=1024" alt="" /></a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 用dplyr包的filter功能，筛选条件：矩阵中“count”列中大于0，“% expression”列中大于1%。</span><br><span class="line">gene_cluster %&gt;% filter(Gene %in% markers) %&gt;%    </span><br><span class="line">mutate(`% Expressing` = (cell_exp_ct/cell_ct) * 100) %&gt;%    </span><br><span class="line">filter(count &gt; 0, `% Expressing` &gt; 1) %&gt;%    </span><br><span class="line">ggplot(aes(x=cluster, y = Gene, color = count, size = `% Expressing`)) +    geom_point() </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/02.png"><img src="https://genehub.files.wordpress.com/2020/11/02.png?w=1024" alt="" /></a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"># 图形调整与美化。使用“theme(axis.line  = element_blank())”取消显示xy坐标轴。“theme(axis.ticks = element_blank()) ”去掉坐标轴的刻度值。</span><br><span class="line"># cowplot包是ggplot2的简单附加组件。它旨在为ggplot2提供一个出版物就绪的主题，一个需要最小量的轴标签尺寸，情节背景等。它的主要目的是方便制作符合要求的图片。除了提供修改的绘图主题外，此包还提供了对ggplot2绘图的自定义注释的功能。</span><br><span class="line"></span><br><span class="line">gene_cluster %&gt;% filter(Gene %in% markers) %&gt;%    </span><br><span class="line">mutate(`% Expressing` = (cell_exp_ct/cell_ct) * 100) %&gt;%    </span><br><span class="line">filter(count &gt; 0, `% Expressing` &gt; 1) %&gt;%    </span><br><span class="line">ggplot(aes(x=cluster, y = Gene, color = count, size = `% Expressing`)) +    geom_point() +    </span><br><span class="line">scale_color_viridis_c(name = &#x27;log2 (count + 1)&#x27;) +    cowplot::theme_cowplot() +    </span><br><span class="line">theme(axis.line  = element_blank()) +   </span><br><span class="line">theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +   ylab(&#x27;&#x27;) +   </span><br><span class="line">theme(axis.ticks = element_blank()) </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/03.png"><img src="https://genehub.files.wordpress.com/2020/11/03.png?w=1024" alt="" /></a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">gene_cluster %&gt;% </span><br><span class="line">filter(Gene %in% markers) %&gt;%    </span><br><span class="line">mutate(`% Expressing` = (cell_exp_ct/cell_ct) * 100) %&gt;%    </span><br><span class="line">filter(count &gt; 0, `% Expressing` &gt; 1) %&gt;%    </span><br><span class="line">ggplot(aes(x=cluster, y = Gene, color = count, size = `% Expressing`)) +    geom_point() +    </span><br><span class="line">cowplot::theme_cowplot() +    </span><br><span class="line">theme(axis.line  = element_blank()) +   </span><br><span class="line">theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +   ylab(&#x27;&#x27;) +   </span><br><span class="line">theme(axis.ticks = element_blank()) +   </span><br><span class="line">scale_color_gradientn(colours = viridis::viridis(20), limits = c(0,4), oob = scales::squish, name = &#x27;log2 (count + 1)&#x27;)</span><br></pre></td></tr></table></figure>
<p>前面的图中，对所有圆点颜色进行观察，可见整体偏向深蓝，只有一个位于坐标c15 * KCNQ1 的圆点是黄色，其count数值4.94，比其他坐标的数值大了不少。通常绘制热图的时候也会因为这种值（甚至更大的极值，因为大部分基因的表达值FPKM通常小于100，但是少数基因的表达值会接近10000）导致画面颜色不够均匀。 参考原文的做法，调用 <code>scale_color_gradientn</code> 将 <code>c(0,4)</code> 数字范围的圆点进行着色，对于大于4的极值， 调用scales函数的squish功能将其“压碎”成为4（即“squished” down） <code>oob = scales::squish</code></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/05.png"><img src="https://genehub.files.wordpress.com/2020/11/05.png?w=1024" alt="" /></a></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/04.png"><img src="https://genehub.files.wordpress.com/2020/11/04.png?w=1024" alt="" /></a></p>
<p>进一步的，可以将基因进行聚类，并调用ggtree进行树形图绘制。先对原始数据进行整理，提取三列：Gene、cluster、count。调用pivot_wider功能，将其转换为行为gene，列为cluster的矩阵。（tidyr包的 <code>pivot_wider</code> 函数可以将长表变成宽表。 这适用于将多个变量保存到了一列的情况。—— 摘抄自<a target="_blank" rel="noopener" href="https://www.math.pku.edu.cn/teachers/lidf/docs/Rbook/html/%5C_Rbook/summary-manip.html%EF%BC%89">https://www.math.pku.edu.cn/teachers/lidf/docs/Rbook/html/\_Rbook/summary-manip.html）</a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"># make data square to calculate euclidean distance </span><br><span class="line">mat &lt;- gene_cluster %&gt;%    </span><br><span class="line">filter(Gene %in% markers) %&gt;%    </span><br><span class="line">select(-cell_ct, -cell_exp_ct, -Group) %&gt;%  # drop unused columns to faciliate widening   </span><br><span class="line">pivot_wider(names_from = cluster, values_from = count) %&gt;%    </span><br><span class="line">data.frame() # make df as tibbles -&gt; matrix annoying </span><br><span class="line"></span><br><span class="line">row.names(mat) &lt;- mat$Gene  # put gene in `row` </span><br><span class="line"></span><br><span class="line">mat &lt;- mat[,-1] #drop gene column as now in rows </span><br><span class="line"></span><br><span class="line">clust &lt;- hclust(dist(mat %&gt;% as.matrix())) # hclust with distance matrix</span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/06.png"><img src="https://genehub.files.wordpress.com/2020/11/06.png?w=1024" alt="" /></a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">ddgram &lt;- as.dendrogram(clust) # create dendrogram </span><br><span class="line">ggtree_plot &lt;- ggtree::ggtree(ddgram) </span><br><span class="line">ggtree_plot</span><br><span class="line">ggtree::ggtree(ddgram, size=1) + geom_tiplab() </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/07-1.png"><img src="https://genehub.files.wordpress.com/2020/11/07-1.png?w=1024" alt="" /></a></p>
<p>（右侧的名称有问题）</p>
<p>将树形图和点状图拼接组合。下面的操作只是将两幅图拼起来，基因名并没有对应关系。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">dotplot &lt;- gene_cluster %&gt;% filter(Gene %in% markers) %&gt;% </span><br><span class="line">   mutate(% Expressing = (cell_exp_ct/cell_ct) * 100) %&gt;% </span><br><span class="line">   filter(count &gt; 0, % Expressing &gt; 1) %&gt;% </span><br><span class="line">   ggplot(aes(x=cluster, y = Gene, color = count, size = % Expressing)) + </span><br><span class="line">   geom_point() + </span><br><span class="line">   cowplot::theme_cowplot() + </span><br><span class="line">   theme(axis.line  = element_blank()) +</span><br><span class="line">   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +</span><br><span class="line">   ylab(&#x27;&#x27;) +</span><br><span class="line">   theme(axis.ticks = element_blank()) +</span><br><span class="line">   scale_color_gradientn(colours = viridis::viridis(20), limits = c(0,4), oob = scales::squish, name = &#x27;log2 (count + 1)&#x27;)   </span><br><span class="line"></span><br><span class="line"></span><br><span class="line">   plot_grid(ggtree_plot, dotplot, nrow = 1, rel_widths = c(0.5,2), align = &#x27;h&#x27;) </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/08.png"><img src="https://genehub.files.wordpress.com/2020/11/08.png?w=1024" alt="" /></a></p>
<p>使用</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">dotplot &lt;- gene_cluster %&gt;% filter(Gene %in% markers) %&gt;% </span><br><span class="line">mutate(% Expressing = (cell_exp_ct/cell_ct) * 100,</span><br><span class="line">   Gene = factor(Gene, levels = clust$labels[clust$order])) %&gt;% </span><br><span class="line">   #filter(count &gt; 0, % Expressing &gt; 1) %&gt;% </span><br><span class="line">   ggplot(aes(x=cluster, y = Gene, color = count, size = % Expressing)) + </span><br><span class="line">   geom_point() + </span><br><span class="line">   cowplot::theme_cowplot() + </span><br><span class="line">   theme(axis.line  = element_blank()) +</span><br><span class="line">   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +</span><br><span class="line">   ylab(&#x27;&#x27;) +</span><br><span class="line">   theme(axis.ticks = element_blank()) +</span><br><span class="line">   scale_color_gradientn(colours = viridis::viridis(20), limits = c(0,4), oob = scales::squish, name = &#x27;log2 (count + 1)&#x27;)   </span><br><span class="line">   plot_grid(ggtree_plot, NULL, dotplot, nrow = 1, rel_widths = c(0.5,-0.05, 2), align = &#x27;h&#x27;) </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/10.png"><img src="https://genehub.files.wordpress.com/2020/11/10.png?w=1024" alt="" /></a></p>
<p>代码摘抄自：<a target="_blank" rel="noopener" href="https://davemcg.github.io/post/lets-plot-scrna-dotplots/">https://davemcg.github.io/post/lets-plot-scrna-dotplots/</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/11/16/R-Empirical-cumulative-distribution/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/11/16/R-Empirical-cumulative-distribution/" class="post-title-link" itemprop="url">R：经验分布函数</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-11-16 09:45:08" itemprop="dateCreated datePublished" datetime="2020-11-16T09:45:08+08:00">2020-11-16</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-08 16:47:12" itemprop="dateModified" datetime="2020-12-08T16:47:12+08:00">2020-12-08</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>经验分布函数（Empirical cumulative distribution (ECDF)），是统计学中一个与样本<strong>经验</strong>测度有关的<strong>分布函数</strong>。 该累积<strong>分布函数</strong>是在所有n个数据点上都跳跃 1/n的阶跃<strong>函数</strong>。 对被测变量的某个值而言，该值的<strong>分布函数</strong>值表示所有观测样本中小于或等于该值的样本所占的比例。</p>
<p>在一个网站看到大量ggplot的画图案例，值得收藏：<a target="_blank" rel="noopener" href="https://www.datanovia.com/en/blog/ggplot-examples-best-reference/">https://www.datanovia.com/en/blog/ggplot-examples-best-reference/</a></p>
<p>以 R 内置数据集 iris 为例。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ggplot(iris, aes(Sepal.Length)) +</span><br><span class="line">stat_ecdf(aes(color = Species), size=3) +</span><br><span class="line">scale_color_viridis_d() </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116093707.png"><img src="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116093707.png?w=708" alt="" /></a></p>
<p>便于理解，用了一个简化的数据集画图。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">k=read.table(&quot;test.txt&quot;, head=T)</span><br><span class="line">ggplot(k, aes(Sepal.Length)) +</span><br><span class="line">   stat_ecdf(aes(color = Species), size=3) +</span><br><span class="line">   scale_color_viridis_d()    </span><br></pre></td></tr></table></figure>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116094244.png"><img src="https://genehub.files.wordpress.com/2020/11/e5beaee4bfa1e688aae59bbe_20201116094244.png?w=697" alt="" /></a></p>
<p>计算公式</p>
<p><img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/68f10704c974630c3b15d009b884366d2bcbf685" alt="" /></p>
<p>可以理解为将数据按照从小到大排序，然后分布统计不同大小的数字出现的比例，（其实就是histgram？），那么横坐标就是数据从小到大的范围，纵坐标就是0~100%。图形相当于把histgram的柱形顶端和左侧画出来了。</p>
<p>附数据集test.txt：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br></pre></td><td class="code"><pre><span class="line">Sepal.Length    Species</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 10    setosa</span><br><span class="line"> 1    versicolor</span><br><span class="line"> 2    versicolor</span><br><span class="line"> 3    versicolor</span><br><span class="line"> 4    versicolor</span><br><span class="line"> 5    versicolor</span><br><span class="line"> 6    versicolor</span><br><span class="line"> 7    versicolor</span><br><span class="line"> 8    versicolor</span><br><span class="line"> 9    versicolor</span><br><span class="line"> 10    versicolor</span><br><span class="line"> 11    versicolor</span><br><span class="line"> 12    versicolor</span><br><span class="line"> 13    versicolor</span><br><span class="line"> 14    versicolor</span><br><span class="line"> 15    versicolor</span><br><span class="line"> 16    versicolor</span><br><span class="line"> 17    versicolor</span><br><span class="line"> 18    versicolor</span><br><span class="line"> 19    versicolor</span><br><span class="line"> 20    versicolor</span><br><span class="line"> 1    virginica</span><br><span class="line"> 1    virginica</span><br><span class="line"> 2    virginica</span><br><span class="line"> 2    virginica</span><br><span class="line"> 3    virginica</span><br><span class="line"> 2    virginica</span><br><span class="line"> 4    virginica</span><br><span class="line"> 5    virginica</span><br><span class="line"> 4    virginica</span><br><span class="line"> 3    virginica</span><br><span class="line"> 2    virginica</span><br><span class="line"> 4    virginica</span><br><span class="line"> 5    virginica</span><br><span class="line"> 7    virginica</span><br><span class="line"> 6    virginica</span><br><span class="line"> 8    virginica</span><br><span class="line"> 6    virginica</span><br><span class="line"> 15    virginica</span><br><span class="line"> 16    virginica</span><br><span class="line"> 19    virginica</span><br><span class="line"> 13    kaka</span><br><span class="line"> 14    kaka</span><br><span class="line"> 16    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 17    kaka</span><br><span class="line"> 18    kaka</span><br><span class="line"> 19    kaka</span><br><span class="line"> 20    kaka</span><br><span class="line"> 13    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 12    kaka</span><br><span class="line"> 16    kaka</span><br><span class="line"> 18    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 13    kaka</span><br><span class="line"> 15    kaka</span><br><span class="line"> 2    kaka</span><br><span class="line"> 4    kaka</span><br><span class="line"> 5    kaka</span><br><span class="line"> 1    kaka</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/11/14/R-How-to-remove-rows-with-NA-values/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/11/14/R-How-to-remove-rows-with-NA-values/" class="post-title-link" itemprop="url">R：删除包含缺失值NA的行</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-11-14 18:08:01" itemprop="dateCreated datePublished" datetime="2020-11-14T18:08:01+08:00">2020-11-14</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-08 17:18:12" itemprop="dateModified" datetime="2020-12-08T17:18:12+08:00">2020-12-08</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p><strong>R数据框，如何删除包含NA的行？</strong></p>
<p>在进行统计建模时，通常需要用到的因变量和自变量都不包含缺失值。tidyr 包的 drop_na() 函数可以对数据框 指定一到多个变量，删去指定的变量有缺失值的行。不指定变量时有任何变量缺失的行都会被删去。</p>
<p>例如，将 NHANES 中所有存在缺失值的行删去后数出保留的行数，原来有 10000 行：</p>
<p>NHANES <strong>%&gt;%</strong> <strong>drop_na</strong>() <strong>%&gt;% nrow</strong>()</p>
<p>## [1] 0 可见所有行都有缺失值。</p>
<p>下面仅剔除 AlcoholDay 缺失的观测并计数：</p>
<p>NHANES <strong>%&gt;% drop_na</strong>(AlcoholDay) <strong>%&gt;% nrow</strong>()</p>
<p>## [1] 4914</p>
<p>基本 stats 包的 complete.cases 函数返回是否无缺失值的逻辑向量，na.omit 函数则返回无缺失值的观测的子集。</p>
<p><strong>如何删除全部为NA的行？只要有非缺省值，可保留对应行。</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line"> testmatrix &lt;- matrix(nrow=6, ncol=4)</span><br><span class="line">testmatrix</span><br><span class="line">[,1] [,2] [,3] [,4]</span><br><span class="line">[1,] NA NA NA NA</span><br><span class="line">[2,] NA NA NA NA</span><br><span class="line">[3,] NA NA NA NA</span><br><span class="line">[4,] NA NA NA NA</span><br><span class="line">[5,] NA NA NA NA</span><br><span class="line">[6,] NA NA NA NA</span><br><span class="line">testmatrix[2:5,2:3] &lt;- seq(2)</span><br><span class="line">testmatrix</span><br><span class="line">[,1] [,2] [,3] [,4]</span><br><span class="line">[1,] NA NA NA NA</span><br><span class="line">[2,] NA 1 1 NA</span><br><span class="line">[3,] NA 2 2 NA</span><br><span class="line">[4,] NA 1 1 NA</span><br><span class="line">[5,] NA 2 2 NA</span><br><span class="line">[6,] NA NA NA NA</span><br><span class="line">tm1&lt;-testmatrix[,-which(apply(testmatrix,2,function(x)all(is.na(x))))]</span><br><span class="line">tm1</span><br><span class="line">[,1] [,2]</span><br><span class="line">[1,] NA NA</span><br><span class="line">[2,] 1 1</span><br><span class="line">[3,] 2 2</span><br><span class="line">[4,] 1 1</span><br><span class="line">[5,] 2 2</span><br><span class="line">[6,] NA NA</span><br><span class="line">tm2&lt;-tm1[-which(apply(testmatrix,1,function(x)all(is.na(x)))),]</span><br><span class="line">tm2</span><br><span class="line">[,1] [,2]</span><br><span class="line">[1,] 1 1</span><br><span class="line">[2,] 2 2</span><br><span class="line">[3,] 1 1</span><br><span class="line">[4,] 2 2</span><br><span class="line">testmatrix[-which(apply(testmatrix,1,function(x)all(is.na(x)))), -which(apply(testmatrix,2,function(x)all(is.na(x))))]</span><br><span class="line">[,1] [,2]</span><br><span class="line">[1,] 1 1</span><br><span class="line">[2,] 2 2</span><br><span class="line">[3,] 1 1</span><br><span class="line">[4,] 2 2 </span><br></pre></td></tr></table></figure>
<p>Ref: <a target="_blank" rel="noopener" href="http://azaleasays.com/2008/02/06/r-remove-row-column-all-na-value/">http://azaleasays.com/2008/02/06/r-remove-row-column-all-na-value/</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/10/28/Linux-Run-windows-programs-inlinux-mono/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/10/28/Linux-Run-windows-programs-inlinux-mono/" class="post-title-link" itemprop="url">在Linux平台运行windows程序：mono</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-10-28 11:11:10" itemprop="dateCreated datePublished" datetime="2020-10-28T11:11:10+08:00">2020-10-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-20 02:11:00" itemprop="dateModified" datetime="2020-12-20T02:11:00+08:00">2020-12-20</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
                  , 
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>在Windows系统，最常见的就是Windows Forms程序（以下简称WinForm程序），也就是窗体应用程序，比如MS Office，媒体播放器软件等，WinForm程序比起Web应用程序，明显在实时性，用户亲和度和操作便捷能力方面强很多。在Linux系统下，虽然也有桌面窗体应用程序，但应用很少。能否将WinForm程序移植到Linux系统上面呢？目前，Novell公司的开源项目mono提供了将.NET程序运行在各种非Windows平台的能力，例如控制台程序，WinForm程序和ASP.NET等。(摘抄自：<a target="_blank" rel="noopener" href="https://blog.csdn.net/qq%5C_21153619/article/details/81459359">https://blog.csdn.net/qq\_21153619/article/details/81459359</a>)</p>
<p>如何在linux安装mono？我主要目的是在Linux下调用Windows程序做批量文件的处理。因为windows的cmd窗口操作真的不好用，相比之下Linux命令行就好用多了。</p>
<h3 id="1-安装依赖组件"><a class="markdownIt-Anchor" href="#1-安装依赖组件"></a> 1 安装依赖组件</h3>
<p>yum -y install gcc gcc-c++ bison pkgconfig glib2-devel gettext make libpng-devel libjpeg-devel libtiff-devel libexif-devel giflib-devel libX11-devel freetype-devel fontconfig-devel  cairo-devel</p>
<h3 id="2-安装libgdiplus"><a class="markdownIt-Anchor" href="#2-安装libgdiplus"></a> 2 安装libgdiplus</h3>
<ul>
<li><code>wget http://download.mono-project.com/sources/libgdiplus/libgdiplus-4.2.tar.gz</code></li>
<li><code>tar zxf libgdiplus-4.2.tar.gz</code></li>
<li><code>cd libgdiplus-4.2</code></li>
<li><code>./configure --prefix=/home/fenglei/local</code></li>
<li><code>make</code></li>
<li><code>make install</code></li>
</ul>
<h3 id="3-安装mono"><a class="markdownIt-Anchor" href="#3-安装mono"></a> 3 安装mono</h3>
<p>版本选择：<a target="_blank" rel="noopener" href="http://download.mono-project.com/sources/mono/">http://download.mono-project.com/sources/mono/</a></p>
<ul>
<li><code>wget http://download.mono-project.com/sources/mono/mono-4.6.0.125.tar.bz2</code></li>
<li><code>tar jxf mono-4.6.0.125.tar.bz2</code></li>
<li><code>cd mono-4.6.0</code></li>
<li><code>./configure --prefix=/home/fenglei/local</code></li>
<li><code>make -j2</code></li>
<li><code>make install</code></li>
</ul>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028105649.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028105649.png?w=1015" alt="" /></a></p>
<h3 id="4-软件测试我用了informedproteomics和rawtools软件做测试"><a class="markdownIt-Anchor" href="#4-软件测试我用了informedproteomics和rawtools软件做测试"></a> 4 软件测试，我用了InformedProteomics和rawTools软件做测试</h3>
<p>下载rawTools: <a target="_blank" rel="noopener" href="https://github.com/kevinkovalchik/RawTools/releases/download/2.0.2/RawTools-2.0.2.zip">https://github.com/kevinkovalchik/RawTools/releases/download/2.0.2/RawTools-2.0.2.zip</a></p>
<p>rawTools其他版本选择：<a target="_blank" rel="noopener" href="https://github.com/kevinkovalchik/RawTools/releases/tag/2.0.2">https://github.com/kevinkovalchik/RawTools/releases/tag/2.0.2</a></p>
<p><code>mkdir RawTools</code></p>
<p><code>unzip RawTools-2.0.2.zip -d RawTools</code></p>
<p><code>mono ~/local/app/RawTools/RawTools.exe</code></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110306.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110306.png?w=1024" alt="" /></a></p>
<p>下载InformedProteomics：<a target="_blank" rel="noopener" href="https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics/releases/download/v1.0.7569/InformedProteomics%5C_Program.zip">https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics/releases/download/v1.0.7569/InformedProteomics\_Program.zip</a></p>
<p>版本选择：<a target="_blank" rel="noopener" href="https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics/releases">https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics/releases</a></p>
<p><code>mkdir InformedProteomics</code></p>
<p><code>unzip InformedProteomics_Program.zip -d InformedProteomics</code></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110657.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110657.png?w=1024" alt="" /></a></p>
<p>mono ~/local/app/InformedProteomics/ProMex.exe</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110926.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110926.png?w=1024" alt="" /></a></p>
<p>mono ~/local/app/InformedProteomics/MSPathFinderT.exe</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110955.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201028110955.png?w=1024" alt="" /></a></p>
<h3 id="报错"><a class="markdownIt-Anchor" href="#报错"></a> 报错</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">mono ~/local/app/InformedProteomics/ProMex.exe -i xxx.raw -csv n -maxThreads 6                                                       </span><br><span class="line">************ ProMex version 1.0.7569 (September 21, 2020) ************</span><br><span class="line">InputPath    xxx.raw</span><br><span class="line">OutputPath   </span><br><span class="line">MinMass       2,000 Da</span><br><span class="line">MaxMass      50,000 Da</span><br><span class="line">MinCharge     1</span><br><span class="line">MaxCharge    60</span><br><span class="line">FeatureMap   True</span><br><span class="line">ScoreReport  False</span><br><span class="line">LikelihoodRatioThreshold -10</span><br><span class="line">MaxThreads   4</span><br><span class="line">Exception while processing: The type initializer for &#x27;InformedProteomics.Backend.MassSpecData.MassSpecDataReaderFactory&#x27; threw an exception.</span><br><span class="line">  at InformedProteomics.FeatureFinding.LcMsFeatureFinderLauncher.Run () [0x00011] in &lt;filename unknown&gt;:0 </span><br><span class="line">  at ProMex.Program.Main (System.String[] args) [0x000f1] in &lt;filename unknown&gt;:0 </span><br><span class="line">[MVID] 0604fddcd448487d905c1e71f6809d88 0</span><br><span class="line">[MVID] 072679dfee414eaab286f6a82270167b 1</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/10/15/Others-NAS-set-up/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/10/15/Others-NAS-set-up/" class="post-title-link" itemprop="url">NAS 安装与使用</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-10-15 11:38:06" itemprop="dateCreated datePublished" datetime="2020-10-15T11:38:06+08:00">2020-10-15</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Life/" itemprop="url" rel="index"><span itemprop="name">Life</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>我主要用NAS做数据存储和备份，一些生物实验的原始数据比较大。<br />
<strong>一 组装</strong><br />
NAS型号：群晖 DS918+，有四个硬盘卡槽，底部还有一个内存卡插槽。买了两块IronWolf的NAS专用硬盘，各8 TB，一起16 TB。网络上有大量教程参考，组装完毕就是下图。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/81yfruavezl._ac_sl1500_.jpg"><img src="https://genehub.files.wordpress.com/2020/10/81yfruavezl._ac_sl1500_.jpg?w=1024" alt="" /></a></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/synology-ds918-installing-hard-drive.jpg"><img src="https://genehub.files.wordpress.com/2020/10/synology-ds918-installing-hard-drive.jpg?w=800" alt="" /></a></p>
<p><strong>二 系统安装与配置</strong></p>
<p>在官网下载相应的系统安装文件 <a target="_blank" rel="noopener" href="https://www.synology.com/zh-tw/support/download/DS918+#firmware">https://www.synology.com/zh-tw/support/download/DS918+#firmware</a> 。安裝 DSM （DiskStation Manager）。</p>
<p>[安装过程：<a target="_blank" rel="noopener" href="https://ningselect.com/11397/40/%5C">https://ningselect.com/11397/40/\</a>]</p>
<p><strong>三 使用</strong></p>
<p>1 在任意位置使用web登入NAS服务器</p>
<p>2 在内网web浏览器输入IP即可登录使用</p>
<p>3 FTP连接。在FTP客户端（FileZilla）输入 sftp://admin@192.168.x.y 即可，后面不用加端口号。</p>
<p>4 将NAS挂载到本地网络的一个个人电脑下。这样相当于将NAS作为个人电脑的硬盘使用。先在NAS管理界面设置NFS，参考：<a target="_blank" rel="noopener" href="https://www.synology.com/zh-cn/knowledgebase/DSM/tutorial/File%5C_Sharing/How%5C_to%5C_access%5C_files%5C_on%5C_Synology%5C_NAS%5C_within%5C_the%5C_local%5C_network%5C_NFS%E3%80%82%E7%84%B6%E5%90%8E%E5%9C%A8%E4%B8%AA%E4%BA%BA%E7%94%B5%E8%84%91%EF%BC%88%E6%88%91%E7%94%A8%E7%9A%84windows">https://www.synology.com/zh-cn/knowledgebase/DSM/tutorial/File\_Sharing/How\_to\_access\_files\_on\_Synology\_NAS\_within\_the\_local\_network\_NFS。然后在个人电脑（我用的windows</a> 10）设置，将群晖NAS映射为网络驱动器。在“我的电脑”右键，选择“映射网络驱动器”，驱动器选“Y”（随便选都可，一般考虑Z-&gt;A的顺序），文件夹地址输入“\\192.168.x.x\share_name”（192.168.x.x是NAS的IP地址，视具体情况而定；share_name就是NAS里面盘阵的名称，视具体情况而定），点击“完成”并输入账户名和密码。依次映射了两个硬盘，我得到了下面的界面，可见“网络位置”下方出现了相应的磁盘，这样就像使用本机硬盘一样的，很方便。</p>
<p>参考 <a target="_blank" rel="noopener" href="https://jingyan.baidu.com/article/e75057f22d7901ebc81a8966.html">https://jingyan.baidu.com/article/e75057f22d7901ebc81a8966.html</a></p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201015112922.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201015112922.png?w=702" alt="" /></a></p>
<p><strong>四 空间清理</strong></p>
<p>收到系统提醒说存储空间将满，删除了一部分文件，并清空回收站，但是空间还是没有释放。查询资料之后发现是 Synology Drive 这个应用的缘故，它会将已删除的文件保留备份，所以占用大量空间，将其卸载可以释放哪些空间。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201026162623.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201026162623.png?w=831" alt="" /></a></p>
<p>Use a terminal (eg. Putty, SecureCRT) to connect Synology through SSH.<br />
ie: 192.168.x.x port 22<br />
Login ID: admin; Password: admin password.<br />
Switch to ROOT by “sudo -i”, type a new password for root user. Then detect space usage by “du -h -d 1”</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201026160121.png"><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201026160121.png?w=1024" alt="" /></a></p>
<p>进入 Synology Drive Sever，选择“版本资源管理器”，勾选“显示已删除文件”，可见到那些被删除的文件实际上还保存在硬盘里，所以占用大量空间。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/10/03/Bioinfo-Genome-Browser-jbrowse-based-on-nginx-web-server/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/10/03/Bioinfo-Genome-Browser-jbrowse-based-on-nginx-web-server/" class="post-title-link" itemprop="url">基因组浏览器 JBrowse 安装（基于 Nginx 的 Web 服务器）</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-10-03 18:44:47" itemprop="dateCreated datePublished" datetime="2020-10-03T18:44:47+08:00">2020-10-03</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-31 14:58:03" itemprop="dateModified" datetime="2020-12-31T14:58:03+08:00">2020-12-31</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
                  , 
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>需求：将基因组浏览器 JBrowse 安装在 Linux 服务器，在内网中用客户端电脑的网页浏览器可以打开 JBrowse 进行操作。安装过程如下，以 CentOS 7 系统为例，需要 root 权限。</p>
<p>JBrowse 是一款快速的，可以嵌入的基因组浏览器，它采用 JavaScript 和 HTML5 开发，同时提供了 Perl 脚本处理数据。JBrowse 采用了 Ajax 技术，主要的计算在客户端上完成， 对服务器资源的占用非常低，速度非常快，可以轻松部署上 G 的数据。JBrowse 支持 GFF3, BED, FASTA, Wiggle, BigWig, BAM, VCF (with tabix), REST 等众多数据格式，BAM, BigWig，VCF 格式的数据可以不需要转换而直接使用。 JBrowse 对系统依赖非常低，只需要一个 web 服务器，因此部署起来非常方便。Nginx 是一款轻量级、高性能 web 服务器，它占用内存少，并发能力强。越来越多的网站使用 Nginx 作为 web 服务器。</p>
<h3 id="1-安装-jbrowse-依赖的软件"><a class="markdownIt-Anchor" href="#1-安装-jbrowse-依赖的软件"></a> 1 安装 Jbrowse 依赖的软件</h3>
<p>参考资料：<a target="_blank" rel="noopener" href="https://jbrowse.org/docs/installation.html">https://jbrowse.org/docs/installation.html</a></p>
<p>需要先安装依赖的一些工具，即“system pre-requisites”。我用的 CentOS 7系统，操作如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sudo yum groupinstall &quot;Development Tools&quot;</span><br><span class="line">sudo yum install zlib-devel perl-ExtUtils-MakeMaker</span><br></pre></td></tr></table></figure>
<p>安装 node 和 npm<br />
源代码下载 <a target="_blank" rel="noopener" href="https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz">https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz</a><br />
xz文件解压：<code>tar -xvJf node-v8.11.1-linux-x64.tar.xz</code><br />
解压之后文件目录里面包含bin目录（node，npm，npx 三个程序），将bin目录加入环境变量。<br />
<code>export PATH=/xxx/xxx/bin:$PATH</code><br />
source ~/.bashrc 之后即可测试 node -h 与 npm -h，均可以运行。</p>
<h3 id="2-安装基因组浏览器-jbrowe"><a class="markdownIt-Anchor" href="#2-安装基因组浏览器-jbrowe"></a> 2 安装基因组浏览器 Jbrowe</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"># make a directory that this user can write to</span><br><span class="line"># for ubuntu/Debian that is /var/www/; for centos is /var/www/html</span><br><span class="line">sudo mkdir /var/www/html/jbrowse;</span><br><span class="line">sudo chown `whoami` /var/www/html/jbrowse;</span><br><span class="line"></span><br><span class="line"># cd into it</span><br><span class="line">cd /var/www/html/jbrowse;</span><br><span class="line"></span><br><span class="line"># fetch a JBrowse release zip file</span><br><span class="line">curl -O http://jbrowse.org/releases/jbrowse-1.16.10-release.tar.gz</span><br><span class="line"></span><br><span class="line"># unzip it and cd into it</span><br><span class="line">unzip JBrowse-1.12.3.zip</span><br><span class="line">cd JBrowse-1.12.3</span><br><span class="line"></span><br><span class="line"># run setup.sh, quick start with example data</span><br><span class="line">./setup.sh</span><br></pre></td></tr></table></figure>
<p>运行 <a target="_blank" rel="noopener" href="http://setup.sh">setup.sh</a> 自动安装成功，setup.log 文件如下所示。./setup.sh 这一步需要安装很多 perl 依赖库，需要的时间可能有点久。如果报错，请查看 setup.log 文件，一般是系统可能缺某个依赖库。</p>
<p><img src="https://genehub.files.wordpress.com/2020/10/jbrowse.jpg?w=1024" alt="" /></p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5ae89e8a385jbrowse.png?w=1024" alt="" /></p>
<p>到这一步是无法从其他客户端访问服务器上的 JBrowse 的，需要在服务器上面安装 Web 服务器。比如 appache/xampp/nginx 等。</p>
<p>xampp案例见：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/1dddec8dc29c">https://www.jianshu.com/p/1dddec8dc29c</a><br />
nginx案例见：<a target="_blank" rel="noopener" href="https://yq.aliyun.com/articles/650480">https://yq.aliyun.com/articles/650480</a></p>
<h3 id="3-安装-web-服务器-nginx"><a class="markdownIt-Anchor" href="#3-安装-web-服务器-nginx"></a> 3 安装 Web 服务器 Nginx</h3>
<p>我使用 nginx 做测试的。下载 ngnix 源文件 <a target="_blank" rel="noopener" href="http://nginx.org/download/nginx-1.18.0.tar.gz">nginx-1.18.0.tar.gz</a>，root 权限下，解压在 <code>/usr/local/ngnix</code> 下。</p>
<p>运行 <code>./configure &amp;&amp; make &amp;&amp; make install</code> 成功安装nginx（默认web端口是80），然后就可以在客户端浏览器输入服务器地址，看到初始页面：</p>
<p><img src="https://genehub.files.wordpress.com/2020/10/nginx-default-page.jpg?w=834" alt="" /></p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003182915.jpg?w=795" alt="" /></p>
<p>上方的截图所显示的 <em>Welcome to Nginx</em> 初始网页，实际上是由 nginx 安装目录下的 nginx.conf 配置文件控制的，红色标记处，<code>location</code> 字样下的 <code>index.html</code> 是指默认的 <code>/usr/local/nginx/html/index.html</code>。</p>
<p>nginx安装参考：<a target="_blank" rel="noopener" href="https://blog.csdn.net/t8116189520/article/details/81909574">https://blog.csdn.net/t8116189520/article/details/81909574</a> （实测：修改nginx.conf里面端口号之后暂时无法访问。需要linux系开放相应的端口。见下文。）</p>
<p>下面是启动与关闭 nginx 的操作。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"># 停止运行 nginx </span><br><span class="line">/usr/local/nginx/sbin/nginx -s stop  </span><br><span class="line"></span><br><span class="line"># 重启（reload）如果报错，就要先输入下面的命令`</span><br><span class="line">/usr/local/nginx/sbin/nginx -c /usr/local/nginx/conf/nginx.conf  </span><br><span class="line"></span><br><span class="line"># 重启 nginx</span><br><span class="line">/usr/local/nginx/sbin/nginx -s reload</span><br></pre></td></tr></table></figure>
<h3 id="4-设置-nginx使-jbrowse-可以在-web-浏览器访问"><a class="markdownIt-Anchor" href="#4-设置-nginx使-jbrowse-可以在-web-浏览器访问"></a> 4 设置 nginx，使 Jbrowse 可以在 Web 浏览器访问</h3>
<p>那么如何将 Jbrowse 相关文档经过 Nginx 平台呈现在客户端的web界面？需要修改 nginx的配置文件：<code>/usr/local/nginx/conf/nginx.conf</code>，root权限下vim这个配置文件，在 http 内容里面加入下面的文字。端口号可以自己设置，比如8080、8082等。location指向<code>/var/www/html/jbrowse/index.html</code></p>
<p>查询资料的过程中发现其他人安装nginx之后有这两个目录 <code>/etc/nginx/sites-available/</code> 和 <code>/etc/nginx/sites-enabled</code>，我安装之后是没有的，后面发现并没有影响。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">server &#123;</span><br><span class="line">    listen     8000;</span><br><span class="line">    server\_name  192.168.1.108;</span><br><span class="line">    location / &#123;</span><br><span class="line">        root   /var/www/html/jbrowse;</span><br><span class="line">        index  index.html;</span><br><span class="line">    &#125;</span><br></pre></td></tr></table></figure>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003183445.jpg" alt="" /></p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003183654.jpg" alt="" /></p>
<p>按照上图修改并保存，重启 nginx 理论上可以在客户端打开 192.168.1.108:8000 即可看到 jbrowse，可是我遇到了打不开的问题。原来不仅需要在 nginx 里面设置端口，还需要linux系开放相应的端口：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">iptables -A INPUT -ptcp --dport 8000 -j ACCEPT</span><br></pre></td></tr></table></figure>
<p>然后就顺利打开了 Jbrowse 浏览器，可以看到欢迎页面。</p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003184126.jpg?w=852" alt="" /></p>
<p>访问 <a target="_blank" rel="noopener" href="http://192.168.1.108/index.html?data=sample_data/json/volvox">http://192.168.1.108/index.html?data=sample_data/json/volvox</a> 可以看到示例页面：</p>
<p><img src="http://blog.biochen.com/wp-content/uploads/2016/06/JBrowse_1.png" alt="" /></p>
<h3 id="5-如何配置自己的基因组"><a class="markdownIt-Anchor" href="#5-如何配置自己的基因组"></a> 5 如何配置自己的基因组</h3>
<p>基因组浏览器一般展示如下文件</p>
<ol>
<li>reference genome sequence (.fasta)</li>
<li>gene/mRNA/ncRNA annotation file (.gff3)</li>
<li>ChIP Peak file (.bed)</li>
<li>RNA expression file (?)</li>
</ol>
<p>下图所示，JBrowse 浏览器窗口左侧是一个目录，即 Available Tracks，下面通常包括 Reference、Annotation、RNA expression、Histone PTM等栏目，是需要逐个配置的。控制这些“栏目”的配置文件是<code>/var/www/html/jbrowse/data/trackList.json</code>。</p>
<p><img src="https://genehub.files.wordpress.com/2020/10/e5beaee4bfa1e688aae59bbe_20201003184221.jpg?w=1024" alt="" /></p>
<p><strong>载入参考基因组</strong></p>
<p>我们需要准备一个物种的基因组 fasta 文件，在其他应用场景下 RNA 和 protein 都没有问题。通过 <code>prepare-refseqs.pl</code> 格式化生成的track，这为后续所有添加的文件提供一个坐标系，在网页端参考序列就可以被加载浏览，放大后连碱基序列可以显示出来。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># jbrowse家目录：/var/www/html/jbrowse</span><br><span class="line"># 以下操作在jbrowse家目录，序列文件根据实际情况修改</span><br><span class="line">./bin/prepare-refseqs.pl --fasta ~/database/Glycine_max/Gmax_275_Wm82_a2.v1/Gmax_275_v2.0.softmasked.fa</span><br><span class="line"># ./bin/prepare-refseqs.pl --fasta ~/lyrata/Sequence/Alyrata_384_v1.fa</span><br></pre></td></tr></table></figure>
<p>这就会在当前目录（/var/www/html/jbrowse）生成一个 data 文件夹，直接访问IP地址所看到的序列就来源于该文件夹。</p>
<p><strong>载入基因注释信息</strong></p>
<p>特征序列一般以&quot;gff|gbk|bed&quot;格式存放，用于注明序列的信息。所需工具为 <a target="_blank" rel="noopener" href="http://flatfile-to-json.pl">flatfile-to-json.pl</a>，注意 <code>--trackLabel</code> 参数可以设置 Tracks。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">./bin/flatfile-to-json.pl --gff /home/fenglei/database/Glycine_max/Gmax_275_Wm82_a2.v1/Gmax_275_Wm82.a2.v1.gene.gff3 --trackLabel &#x27;Gene_a2.v1&#x27;</span><br><span class="line"># ./bin/flatfile-to-json.pl --gff ~/Athalina/TAIR10/TAIR10_GFF3_genes.gtf --tracklabel gene --out ./Athaliana/   </span><br><span class="line"># 结果是在当前目录下生成data，data里包括序列track配置文件. 同样可以用--out参数输出到指定文件夹。</span><br></pre></td></tr></table></figure>
<p>注意，基因组浏览器有一个搜索框，在完成上述操作之后可以浏览基因组和基因信息，也可以输入 <code>Chr02:123-456</code> 这样的坐标查看我们感兴趣的区域，但是还不能使用搜索功能。需要在上两步的基础上运行下面程序才可。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">./bin/generate-names.pl</span><br></pre></td></tr></table></figure>
<p><strong>载入bed格式数据</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">./bin/flatfile-to-json.pl --bed ./ChIP_peaks/CCK9_peaks.xls.bed --trackType CanvasFeatures --trackLabel BED_CCK9_p0.01</span><br><span class="line">./bin/flatfile-to-json.pl --bed ./ChIP_peaks/CSK9_peaks.xls.bed --trackType CanvasFeatures --trackLabel BED_CSK9_p0.01</span><br><span class="line">./bin/flatfile-to-json.pl --bed ./ChIP_peaks/CCK9_peaks.unfilter.bed --trackType CanvasFeatures --trackLabel BED_CCK9_all</span><br><span class="line">./bin/flatfile-to-json.pl --bed ./ChIP_peaks/CSK9_peaks.unfilter.bed --trackType CanvasFeatures --trackLabel BED_CSK9_all</span><br></pre></td></tr></table></figure>
<p><strong>载入bigwig格式的peak信息</strong></p>
<p>直接在data目录下新建一个 histone_BigWig 文件夹，把需要展示的bigwig文件放置在里面。然后编辑 /var/www/html/jbrowse/data/trackList.json 文件，逐个设置。</p>
<p>每一个bigwig文件，都需要增加下面这样的代码块。我针对样品设计和网页端展示效果，做了很多次调整。下面仅仅列出某个文件的配置信息。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">&#123;</span><br><span class="line">   &quot;autoscale&quot; : &quot;local&quot;,</span><br><span class="line">   &quot;category&quot; : &quot;Histone&quot;,  # 设定文件所属的 Track</span><br><span class="line">   &quot;key&quot; : &quot;BigWig input control&quot;,</span><br><span class="line">   &quot;label&quot; : &quot;BigWig input control&quot;, # 设定在网页上显示的标签</span><br><span class="line">   &quot;logScaleOption&quot; : true,</span><br><span class="line">   &quot;maxExportSpan&quot; : 500000,</span><br><span class="line">   &quot;max_score&quot; : 6,  # 设定所需展示的最大数值。大于这个阈值的数字就按照该值展示。消除极大值带来的展示问题。</span><br><span class="line">   &quot;metadata&quot; : &#123;&#125;,</span><br><span class="line">   &quot;min_score&quot; : 0,</span><br><span class="line">   &quot;scale&quot; : &quot;log&quot;,  # 设定是否以log形式展示数值。因为测序数据经常有个别极大值，log处理更利于展示</span><br><span class="line">   &quot;storeClass&quot; : &quot;JBrowse/Store/SeqFeature/BigWig&quot;,  # 什么意思？表示这个代码块需要展示的是bigWig格式的文件。</span><br><span class="line">   &quot;style&quot; : &#123;</span><br><span class="line">      &quot;height&quot; : 60,  # 设定展示条带的高度。因为有多个文件要展示，如何设定其高度，将使浏览器打开后的视角比较舒适。</span><br><span class="line">      &quot;neg_color&quot; : &quot;red&quot;,  # 负值的颜色。</span><br><span class="line">      &quot;origin_color&quot; : &quot;#888&quot;,</span><br><span class="line">      &quot;pos_color&quot; : &quot;#DDB6A1&quot;,  # 正数值的颜色。</span><br><span class="line">      &quot;variance_band_color&quot; : &quot;rgba(0,0,0,0.3)&quot;</span><br><span class="line">   &#125;,</span><br><span class="line">   &quot;type&quot; : &quot;JBrowse/View/Track/Wiggle/XYPlot&quot;,</span><br><span class="line">   &quot;urlTemplate&quot; : &quot;histone_BigWig/CCinput.bw&quot;   # 这里指定文件路径。</span><br><span class="line">&#125;,</span><br></pre></td></tr></table></figure>
<p>配置完成之后就类似下面的结果了。</p>
<p><img src="https://ics.hutton.ac.uk/wp-content/uploads/2015/07/jbrowse.png" alt="" /></p>
<p><strong>参考资料</strong><br />
<a target="_blank" rel="noopener" href="https://www.plob.org/article/14590.html">JBrowse使用说明：参考基因组格式化</a></p>
<p>-END-</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/09/16/Visualization-Adobe-Illustrator-Learning-Usage/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/09/16/Visualization-Adobe-Illustrator-Learning-Usage/" class="post-title-link" itemprop="url">Adobe Illustrator (AI) 生物研究的作图笔记</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-09-16 10:30:27" itemprop="dateCreated datePublished" datetime="2020-09-16T10:30:27+08:00">2020-09-16</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-07-14 17:43:03" itemprop="dateModified" datetime="2022-07-14T17:43:03+08:00">2022-07-14</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Visualization/" itemprop="url" rel="index"><span itemprop="name">Visualization</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>Adobe Illustrator 在科研日常修图中应用广泛，而且可以用来绘制细胞、个体、通路等示意图，这里对学习过程中一些元素做个笔记。</p>
<p>关键词：Adobe Illustrator；画图；绘图；AI</p>
<p><strong>1 如何画一个圆圈，且边缘是虚线？</strong></p>
<p>在左侧工具栏选择“椭圆工具”（快捷键：L），在画板上先点击鼠标左键，然后按住shift键不松开，拖动鼠标，在另一个位置点一下鼠标左键，即可得到圆形。单击鼠标左键选中圆圈，上方工具栏有“描边”字样，单击进入菜单，选中“虚线”前面的框框，即可设置虚线，虚线粗细/间隔都可以设置。</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200916100211.png?w=306" alt="" /></p>
<p><strong>2 如何进行填色或描边？如何结合吸管工具进行颜色设定？</strong></p>
<p>在软件左侧工具栏，“填色”和“描边”两个正方形图标有部分重叠，哪个图标位于上层，就表示将对所选目标的相应部位进行颜色操作，可以用鼠标点击这两个图标进行选择，也可以按键“x”进行快捷切换。以描边为例，我希望圆形边缘颜色与另一个图形A的颜色一致，那么我先点击“描边”，表示将对边缘着色，再点击选中圆形，再点击左侧工具栏中“吸管”工具（快捷键：I），此时按住shift键不松开，将吸管工具在图形A的色彩区域点一下，即可看到圆形边缘颜色与图形A颜色一致。这里关键是shift键，之前一直不知道 T_T</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200916101800.png?w=406" alt="" /></p>
<p>Adobe 官网也有教程：<a target="_blank" rel="noopener" href="https://helpx.adobe.com/cn/illustrator/using/painting-fills-strokes.html">https://helpx.adobe.com/cn/illustrator/using/painting-fills-strokes.html</a></p>
<p><img src="https://i.imgur.com/01EbFL6.jpeg" alt="" /></p>
<p><strong>3 如何设定渐变效果，使一个形状具有立体感？</strong></p>
<p>选中软件左侧工具栏“渐变工具”按钮（快捷键：G），再点击需要设定渐变色的图形，图形上就会出现一个渐变条。软件右侧有一个“渐变”按钮，点击即可对所选图形的渐变颜色/范围/模式进行设定。渐变类型有线性和径向，径向即下图右侧所示，可以做出很好的立体感。颜色设定，可看到渐变条下方有几个可以移动的颜色快，里面可以设定颜色。</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200916102331.png?w=624" alt="" /></p>
<p><strong>4 如何对一个完整的形状分割成若干部分？如果将重叠的图形进行分割？</strong></p>
<p>如下图所示，A为基本图形元素。B为将椭圆形切成两半，选中图形之后，左键按住软件左侧的“橡皮擦工具图标”，可以跳出“剪刀”图标（快捷键：C），在图形的边缘上选择两个点各自点击，就可以依据两点形成的直线将形状分开。C为长按“橡皮擦工具”图标并选择跳出来的“刻刀”工具，在图形上随便划几下，就可按照划过的痕迹分割，看到C图效果。D是两个形状的重叠，快捷键Ctrl+Shift+F9，出来“路径选择器”，有不同的剪切模式，剪切之后可看到重叠边缘出现交点，右键并选择“取消编组”，就可得到E/F这样的分割效果。</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200916153510.png?w=1024" alt="" /></p>
<p><strong>5 设置渐变效果的时候如果只有黑白颜色选项怎么办？</strong></p>
<p>在AI中使用渐变填充的时候会发现颜色只有黑白色，这是因为颜色模式选择的是灰度，这就需要更改为RGB模式。打开渐变的属性框，我们在渐变条上双击色标只显示黑白颜色。这是因为颜色模式为灰度，我们鼠标点击右上角的“三横线”按钮。这个时候就会弹出下拉菜单，找到并点击——切换为RGB模式。就可以选择彩色渐变了。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/10/e6b890e58f98e58faae69c89e9bb91e799bde889b2.png"><img src="https://genehub.files.wordpress.com/2020/10/e6b890e58f98e58faae69c89e9bb91e799bde889b2.png?w=378" alt="" /></a></p>
<p><strong>6 如何将单个元素旋转复制</strong></p>
<p>应用场景：一个椭圆，旋转复制，成为韦恩图。平时也可以用来将单个花瓣旋转复制成为一朵花，或者单个细胞膜结构重复形成细胞膜。如下图所示，先画一个椭圆，并单击选中，再点击左侧工具栏“旋转”按钮，此时可见椭圆中心出现小圆斑，按住Alt键的同时可以挪动小圆斑，小圆斑表示旋转中心，松开鼠标左键之后，跳出一个旋转对话框，设定旋转角度，比如我希望得到元素的5次重复，设定72度，点击“复制”，然后按Ctrl+D即可逐次复制，得到右边的效果。</p>
<p><img src="https://i.imgur.com/jBv5JEK.jpeg" alt="" /></p>
<p><strong>7 细胞膜的画法</strong></p>
<p>细胞膜，即磷脂双分子层，由磷脂分子紧密排列而成。网上的教程是先画一个磷脂分子：即一个圆圈加两个尾巴，上下对称翻转复制，组合之后成为加入画笔，然后用其他形状，比如直线、曲线、圆形，只要点击画笔中的“磷脂双分子”元素，就可以得到下图下方的细胞膜效果。我希望将圆形的磷脂分子做一个颜色渐变，圆形就有了球形的效果，当把这个立体效果的磷脂分子添加到画笔，遇到了一个软件报错：“所选图稿包含不能在散点画笔中使用的元素”，原来渐变是不能用来作为画笔的设置的，但是文献里面明明可以看到别人画的细胞膜有这个效果。在 <a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1db41127zL/?spm_id_from=333.788.recommend_more_video.0">B 站</a> 找到了一个解决方案，就是不用将磷脂分子加入画笔，而是直接复制，按住Alt键的同时向右拖动磷脂分子，就得到一个复制品，然后Ctrl+D，就会重复执行上一步的复制操作，多次Ctrl+D，就得到下图所示的立体效果细胞膜。</p>
<p><img src="https://i.imgur.com/lhagU69.jpeg" alt="" /></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/09/04/%E4%BD%BF%E7%94%A8-enrichment-map-%E5%81%9A%E5%9F%BA%E5%9B%A0%E7%BD%91%E7%BB%9C/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/09/04/%E4%BD%BF%E7%94%A8-enrichment-map-%E5%81%9A%E5%9F%BA%E5%9B%A0%E7%BD%91%E7%BB%9C/" class="post-title-link" itemprop="url">使用 Enrichment Map或ClueGO 做基因富集分析+GO网络</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-09-04 10:49:45" itemprop="dateCreated datePublished" datetime="2020-09-04T10:49:45+08:00">2020-09-04</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>对于一个基因列表，经过通路富集（Gene Ontology 或 KEGG pathways）往往富集到多个通路，数据仍然比较庞大，可以用 Enrichment Map 或 ClueGO 进一步精简，得到下面类似的网络图。</p>
<p><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fng.3785/MediaObjects/41588_2017_Article_BFng3785_Fig15_ESM.jpg?as=webp" alt="" /></p>
<p>Enrichment map for pathways enriched in susceptibility to ER-negative breast cancer. From: <a target="_blank" rel="noopener" href="https://www.nature.com/articles/ng.3785">Identification of ten variants associated with risk of estrogen-receptor-negative breast cancer</a></p>
<p><img src="https://genehub.files.wordpress.com/2020/09/36897186.png?w=640" alt="" /></p>
<p>Enrichment Map方法：</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904111451.png?w=1024" alt="" /></p>
<p>准备工作</p>
<p>安装 Java。我操作系统是 Windows 10，Java官网下载 jdk-14.0.2_windows-x64_bin.exe，选择默认路径安装。最后安装路径在：C:\Program Files\Java\jdk-14.0.2 ，注意此时安装 Cytoscape 会报错“EXE4J_JAVA_HOME, No JVM could be found on your system”，表示找不到 java，原来需要设置环境变量。</p>
<p>将 Java 路径加入环境变量。在桌面右键点击“我的电脑”，选择“属性”，在出来的界面中左侧有“高级系统设置”字样，点击进去，即可在“高级”栏目下见到“环境变量字样”。</p>
<p>在环境变量中，要修改两个地方，一个是添加JAVA_HOME。选择“新建”，变量名填上JAVA_HOME，变量值填上“C:\Program Files\Java\jdk-14.0.2”，在java的安装过程中，默认一直下一步安装，所以装在C盘，如果在安装过程中改了目录，那可能是D盘或者E盘，那么变量值要做相应的更改。<br />
还要修改一个地方，就是Path，添加JAVA的变量值到Path中，选择Path，然后点“编辑”，在最后面添加如下语句“%JAVA_HOME%\bin”。</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904101004.png?w=810" alt="" /></p>
<p>按照上面的步骤设置好环境变量之后，Cytoscape 也能顺利安装了。进入 CytoScape 软件界面，上方栏目有个 App 字样，进去可以按照应用，例如 Enrichment Map，AutoAnnotate，BinGO等。</p>
<p>下面测试基因网络分析。</p>
<p>进入g:Profiler网站，提交基因列表，将对基因的Gene Ontology、KEGG这些信息进行注释。如下图所示结果。在Detailed Results 栏目下有一个 GEM 文件，将其下载，这个文件里面就是通路注释结果，包含“<a target="_blank" rel="noopener" href="http://GO.ID">GO.ID</a> Description p.Val FDR Phenotype Genes”这些信息。另外在 Data source 栏目下有一个 combined_name.gmt 下载选项，这个 GMT 文件是“Gene Matrix Tranpose”，也可以用户自己做。（By default, the GSEA desktop software searches the MSigDB gene set database that includes pathways, published gene signatures, microRNA target genes and other gene set types. <strong>The user can also provide a custom database as a text-based ‘Gene Matrix Transposed’ (GMT) file</strong> where each line defines a pathway, with its name, identifier and a list of gene identifiers that match the input gene list.）</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904100014.png?w=1024" alt="" /></p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904100026.png?w=1024" alt="" /></p>
<p>将上面的两个文件下载到本地之后，提交到 CytoScape 软件里面。</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904100206.png?w=1024" alt="" /></p>
<p>点击 Build 即可得到网络图。我遇到中间报错信息提示KEGG的问题，直接选择了跳过，也没影响网络生成。然后再在 Apps 里面选择“AutoAnnotate”，将网络进行 cluster 分析，便可以看到一些大圈圈将若干节点围成了不同的群落，只是图形不算美观，跟人家参考文献里面有差距。似乎不太好调整网络形状，几番尝试之后再AutoAnnotate界面有一个三横的按钮，点击进去有“Layout Clusters”，选择“CoSE Layout”，网络图好看很多。</p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200904100319.png?w=1024" alt="" /></p>
<p>上面简单测试了一下g:Profiler + Enrichiment Map + AutoAnnotate 进行网络构建。</p>
<p>问题：</p>
<p>1 如何只绘制某一方面的网络图，例如只关心细胞器/或某个通路？</p>
<p>2 如何将多个基因列表的网络图绘制在一起，用不同颜色标记不同的基因来源？</p>
<p>3 如何自己制作GMT文件</p>
<p>4 其他富集分析软件的结果应该也可以导入Cytoscape，如何对数据框进行调整，才可导入？</p>
<p>[2020.09.26 更新] <strong>ClueGO案例：</strong></p>
<p><img src="https://genehub.files.wordpress.com/2020/09/e5beaee4bfa1e688aae59bbe_20200916094951.png?w=1024" alt="" /></p>
<p>ClueGO的操作比Enrichment Map更方便，网络结果的调整也更人性化。</p>
<p>B站有一个ClueGO操作实践的视频：<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1TK411W7TM">https://www.bilibili.com/video/BV1TK411W7TM</a> （作者：<a target="_blank" rel="noopener" href="https://space.bilibili.com/198764903">野菜茎小甜饼</a>）</p>
<p>参考资料</p>
<p>[1] <a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/news/96301">CytoScape 攻略</a></p>
<p>[2] <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/54e70ee78ef9"></a><a href="/Profiler%20%E5%9F%BA%E5%9B%A0%E5%8A%9F%E8%83%BD%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E7%9A%84%E7%A4%BA%E4%BE%8B">g:Profiler 基因功能富集分析的示例</a></p>
<p>[3] <a target="_blank" rel="noopener" href="http://www.360doc.com/content/19/1202/13/50736008_876897857.shtml">Enrichment Map 操作示例</a></p>
<p>[4] Enrichment Map tutorial: <a target="_blank" rel="noopener" href="https://enrichmentmap.readthedocs.io/en/latest/QuickTour.html">https://enrichmentmap.readthedocs.io/en/latest/QuickTour.html</a></p>
<p>[5] AutoAnnotate layout: <a target="_blank" rel="noopener" href="https://autoannotate.readthedocs.io/en/latest/GroupingAndLayout.html#cluster-aware-layouts">https://autoannotate.readthedocs.io/en/latest/GroupingAndLayout.html#cluster-aware-layouts</a></p>
<p>[6] Nature Protocol 2019: Pathway enrichment analysis and visualization of omics data using g:Profiler, GSEA, Cytoscape and EnrichmentMap <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6607905/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6607905/</a></p>
<p>[7] Enrichment Map usage, SlidShare: <a target="_blank" rel="noopener" href="https://slideplayer.com/slide/3900605/">https://slideplayer.com/slide/3900605/</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2020/08/09/%E8%BD%AC%EF%BC%9A%E5%A6%82%E4%BD%95%E4%B8%8A%E4%BC%A0%E4%BB%A3%E7%A0%81%E5%88%B0github/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/08/09/%E8%BD%AC%EF%BC%9A%E5%A6%82%E4%BD%95%E4%B8%8A%E4%BC%A0%E4%BB%A3%E7%A0%81%E5%88%B0github/" class="post-title-link" itemprop="url">转：如何上传代码到github</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-08-09 14:01:23" itemprop="dateCreated datePublished" datetime="2020-08-09T14:01:23+08:00">2020-08-09</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
                  , 
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Notes/" itemprop="url" rel="index"><span itemprop="name">Notes</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>自从使用github以来，一直都是在<a target="_blank" rel="noopener" href="https://link.jianshu.com?t=https://github.com/">github网站</a>在线上传文件到仓库中，但是有时因为网络或者电脑的原因上传失败。最重要的原因是我习惯本地编辑，完成以后再一起上传github。看过了几个教程，总结出最适合自己的比较简单的方法。</p>
<p>两种方法上传本地文件到github</p>
<h4 id="1-github在线上传文件夹"><a class="markdownIt-Anchor" href="#1-github在线上传文件夹"></a> 1. github在线上传文件夹</h4>
<p>在线上传也可以上传完整的文件夹结构，直接拖拽到上传文件页面的框中即可。</p>
<h5 id="11点击上传文件"><a class="markdownIt-Anchor" href="#11点击上传文件"></a> 1.1点击上传文件</h5>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150553.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150553.png?w=1024" alt="" /></a></p>
<p>点击上传</p>
<h5 id="12-直接拖拽"><a class="markdownIt-Anchor" href="#12-直接拖拽"></a> 1.2 直接拖拽</h5>
<p>直接拖拽即可上传文件夹及文件夹里面的文件。如果点击* choose your files *就只能上传单个文件。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150605.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150605.png?w=1024" alt="" /></a></p>
<p>直接拖拽</p>
<h4 id="2-通过git工具上传本地文件夹本地项目"><a class="markdownIt-Anchor" href="#2-通过git工具上传本地文件夹本地项目"></a> 2. 通过git工具上传本地文件夹（本地项目）</h4>
<h5 id="21-下载git工具"><a class="markdownIt-Anchor" href="#21-下载git工具"></a> 2.1 下载<a target="_blank" rel="noopener" href="https://link.jianshu.com?t=https://git-scm.com/downloads">git工具</a></h5>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207145151.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207145151.png?w=465" alt="" /></a></p>
<p>选择对应版本下载</p>
<h5 id="22-下载完成后安装完成注意在安装过程中可以选择创建桌面快捷方式"><a class="markdownIt-Anchor" href="#22-下载完成后安装完成注意在安装过程中可以选择创建桌面快捷方式"></a> 2.2 下载完成后安装完成，注意在安装过程中可以选择创建桌面快捷方式</h5>
<p><img src="//upload-images.jianshu.io/upload_images/3067059-fa7d131432a1232e.png?imageMogr2/auto-orient/stripimageView2/2/w/503/format/webp" alt="" /></p>
<p>桌面快捷方式</p>
<h5 id="23-绑定用户"><a class="markdownIt-Anchor" href="#23-绑定用户"></a> 2.3 绑定用户</h5>
<p>打开git-bash.exe（直接在桌面上点击右键，或者点击开始按钮找到Git Bash）</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207144957.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207144957.png?w=395" alt="" /></a></p>
<p>运行gitBash.gif</p>
<p>在打开的GIt Bash中输入以下命令（用户和邮箱为你github注册的账号和邮箱）</p>
<p>$ git config --global <a target="_blank" rel="noopener" href="http://user.name">user.name</a> “hanyuntao”<br />
$ git config --global user.email “<a href="mailto:hanyuntaocn@163.com">hanyuntaocn@163.com</a>”</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207145430.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207145430.png?w=763" alt="" /></a></p>
<p>Paste_Image.png</p>
<h5 id="24-设置ssh-keygit中sshkey有何作用"><a class="markdownIt-Anchor" href="#24-设置ssh-keygit中sshkey有何作用"></a> 2.4 设置SSH key（<a target="_blank" rel="noopener" href="https://link.jianshu.com?t=https://segmentfault.com/q/1010000000118744">git中sshkey有何作用？</a>）</h5>
<h6 id="241-生成ssh-key"><a class="markdownIt-Anchor" href="#241-生成ssh-key"></a> 2.4.1 生成ssh key</h6>
<p>首先检查是否已生成密钥<code>cd ~/.ssh</code>，如果返回的<code>ls</code>有3个文件,则密钥已经生成。</p>
<p>密钥生成</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207145728.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207145728.png?w=874" alt="" /></a></p>
<p>如果没有密钥，则通过</p>
<p>$ ssh-keygen -t rsa -C “<a href="mailto:hanyuntaocn@163.com">hanyuntaocn@163.com</a>”</p>
<p>生成，生成过程中一路按3次回车键就好了。（默认路径，默认没有密码登录）<br />
生成成功后，去对应目录C:\Users\hyt.ssh里（hyt为电脑用户名，每个人不同）用记事本打开id_rsa.pub，得到ssh key公钥。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150754.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150754.png?w=768" alt="" /></a></p>
<p>ssh key公钥</p>
<h6 id="242-为github账号配置ssh-key"><a class="markdownIt-Anchor" href="#242-为github账号配置ssh-key"></a> 2.4.2 为github账号配置ssh key</h6>
<p>切换到github，展开个人头像的小三角，点击settings，然后打开SSH keys菜单， 点击Add SSH key新增密钥，填上标题（最好跟本地仓库保持一致）。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150300.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150300.png?w=1024" alt="" /></a></p>
<p>设置sshkey.gif</p>
<p>接着将id_rsa.pub文件中key粘贴到此，最后Add key生成密钥吧。\</p>
<h4 id="25-上传本地项目到github"><a class="markdownIt-Anchor" href="#25-上传本地项目到github"></a> 2.5 上传本地项目到github</h4>
<h5 id="251-创建一个本地项目"><a class="markdownIt-Anchor" href="#251-创建一个本地项目"></a> 2.5.1 创建一个本地项目</h5>
<p>这是我自己创建的几个文件夹及文件。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150936.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207150936.png?w=778" alt="" /></a></p>
<p>本地项目</p>
<h5 id="252-建立本地仓库"><a class="markdownIt-Anchor" href="#252-建立本地仓库"></a> 2.5.2 建立本地仓库</h5>
<p>1.首先进入text文件夹</p>
<p>cd d:text</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207151042.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207151042.png?w=480" alt="" /></a></p>
<p>首先进入text文件夹</p>
<p>2.执行指令：<code>git init</code></p>
<p>执行git init</p>
<p>初始化成功后你会发现项目里多了一个隐藏文件夹.git</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207151137.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207151137.png?w=799" alt="" /></a></p>
<p>隐藏的文件夹</p>
<p>3.执行指令：<code>git add .</code><br />
将所有文件添加到仓库</p>
<p>执行git add .</p>
<p>4.执行指令：<code>git commit -m &quot;submitted&quot;</code><br />
双引号内是提交注释。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207151233.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207151233.png?w=1024" alt="" /></a></p>
<p>提交文件</p>
<h5 id="253-关联github仓库"><a class="markdownIt-Anchor" href="#253-关联github仓库"></a> 2.5.3 关联github仓库</h5>
<p>1.到github text仓库复制仓库地址</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207151406.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207151406.png?w=1024" alt="" /></a></p>
<p>复制仓库地址<br />
2.执行指令：<code>git remote add origin https://github.com/hanyuntao/text.git</code></p>
<p>执行指令</p>
<h5 id="254-上传本地代码"><a class="markdownIt-Anchor" href="#254-上传本地代码"></a> 2.5.4 上传本地代码</h5>
<p>执行指令：<code>git push -u origin master</code></p>
<p>执行指令</p>
<blockquote>
<p>备注：github对应的文件夹里面不能有相同的文件名或文件夹，否则出错。我之前在网页版本上传了几个文件，但是网页版本无法上传文件夹，所以使用windows桌面客户端，要清理掉网页版已经上传的几个文件，才能用上面的代码上传。</p>
</blockquote>
<h5 id="255完成了"><a class="markdownIt-Anchor" href="#255完成了"></a> 2.5.5完成了</h5>
<p>可以看到我们的本地项目已经上传到了github上了。</p>
<p><a target="_blank" rel="noopener" href="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207152041.png"><img src="https://genehub.files.wordpress.com/2020/12/e5beaee4bfa1e688aae59bbe_20201207152041.png?w=1024" alt="" /></a></p>
<p>完成了</p>
<p>** 注意：git是不能管理空的文件夹的，文件夹里必须有文件才能上传。 **</p>
<p>参考资料：?</p>
<ul>
<li>如何<strong>更新代码？</strong></li>
</ul>
<p>第一步：查看当前的git仓库状态，可以使用git status，会列出当前目录下的文件与本地git仓库里面列出的文件的不同之处。</p>
<p>git status</p>
<p>第二步：更新本地git仓库</p>
<p>git add *</p>
<p>第三步：接着输入git commit -m “Updated”</p>
<p>git commit -m “Updated”</p>
<p>第四步：先git pull,拉取当前分支最新代码（也就是获取GitHub上的最新代码信息，更新本地代码）</p>
<p>git pull</p>
<p>第五步：push到远程master分支上（修改本地代码后，再更新GitHub上的代码）</p>
<p>git push origin master</p>
<p>不出意外，打开GitHub已经同步了。</p>
<p>总之，先pull，再push。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/6/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/6/">6</a><span class="page-number current">7</span><a class="page-number" href="/page/8/">8</a><span class="space">&hellip;</span><a class="page-number" href="/page/19/">19</a><a class="extend next" rel="next" href="/page/8/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">185</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2023</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
