<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>Hi-C：3d-dna挂载染色体 | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">Hi-C：3d-dna挂载染色体</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">Hi-C：3d-dna挂载染色体</h1><div class="post-meta">2020-12-15<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>使用二代数据或三代数据得到contig后，下一步就是将contig提升到染色体水平。有很多策略可以做到这一点，比如说遗传图谱，BioNano, HiC, 参考近源物种。如果利用HiC进行准染色体水平，那么目前常见的组装软件有下面几个。</p>
<ul>
<li>HiRise: 2015年后的GitHub就不再更新</li>
<li>LACHESIS: 发表在NBT，2017年后不再更新</li>
<li>SALSA: 发表在BMC genomics, 仍在更新中</li>
<li>3D-DNA: 发表在science，仍在更新中</li>
<li>ALLHiC: 发表在Nature Plants, 用于解决植物多倍体组装问题</li>
</ul>
<p>对于二倍体物种而言，目前3D-DNA应该是组装效果最好的一个软件。<a target="_blank" rel="noopener" href="https://github.com/theaidenlab/3d-dna">3d-dna</a>是一款简单、方便的处理Hi-C软件，可将contig提升到染色体水平。3D-DNA流程简介:</p>
<ul>
<li>将Hi-C数据比对到draft.genome.fa。（利用Juicer分析Hi-C数据）</li>
<li>利用自动化流程进行纠错（misjoin），排序（order），确定正确方向（orient），最后scaffolding，得到染色体水平的组装结果（3D-DNA分析）</li>
<li>Juicebox 进行人工纠错</li>
</ul>
<h3 id="一、分析软件安装"><a href="#一、分析软件安装" class="headerlink" title="一、分析软件安装"></a>一、分析软件安装</h3><p>依赖的软件</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">LastZ (version 1.03.73 released 20150708) – for diploid mode only</span><br><span class="line">Java version &gt;&#x3D;1.8</span><br><span class="line">Bash &gt;&#x3D;4</span><br><span class="line">GNU Awk &gt;&#x3D;4.0.2</span><br><span class="line">GNU coreutils sort &gt;&#x3D;8.11</span><br><span class="line">Python &gt;&#x3D;2.7 - for chromosome number-aware splitter module only</span><br><span class="line">scipy numpy matplotlib - for chromosome number-aware splitter module only</span><br><span class="line">GUN Parallel &gt;&#x3D;20150322 (可选，建议装)</span><br><span class="line">bwa</span><br></pre></td></tr></table></figure>

<p>安装Juicer</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">git clone https:&#x2F;&#x2F;github.com&#x2F;theaidenlab&#x2F;juicer.git</span><br><span class="line">cd juicer</span><br><span class="line">ln -s CPU scripts</span><br><span class="line">cd scripts&#x2F;common</span><br><span class="line">wget https:&#x2F;&#x2F;hicfiles.tc4ga.com&#x2F;public&#x2F;juicer&#x2F;juicer_tools.1.9.9_jcuda.0.8.jar</span><br><span class="line"># wget https:&#x2F;&#x2F;s3.amazonaws.com&#x2F;hicfiles.tc4ga.com&#x2F;public&#x2F;juicer&#x2F;juicer_tools_1.22.01.jar</span><br><span class="line">ln -s juicer_tools.1.9.9_jcuda.0.8.jar  juicer_tools.jar</span><br></pre></td></tr></table></figure>

<p>安装3D-DNA</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https:&#x2F;&#x2F;github.com&#x2F;theaidenlab&#x2F;3d-dna.git</span><br></pre></td></tr></table></figure>

<p>数据准备<br><code>ref</code>文件夹： 存放<code>draft.genome.fa</code><br><code>fastq</code>文件夹: 存放Hi-C测序双端reads, 注意reads文件名的格式 <code>*_R1.fastq, *_R2.fastq</code>。<br>否则报错<code>Failed to find any files matching /xxx/fastq/*_R*.fastq*</code>。</p>
<h3 id="二、利用-Juicer-分析将Hi-C数据比对到contigs-scaffolds"><a href="#二、利用-Juicer-分析将Hi-C数据比对到contigs-scaffolds" class="headerlink" title="二、利用 Juicer 分析将Hi-C数据比对到contigs/scaffolds"></a>二、利用 Juicer 分析将Hi-C数据比对到contigs/scaffolds</h3><ol>
<li><p>基因组建立索引</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bwa index draft.genome.fa</span><br></pre></td></tr></table></figure>
</li>
<li><p>创建可能的酶切位点文件</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">python ~&#x2F;software&#x2F;juicer&#x2F;misc&#x2F;generate_site_positions.py  MboI  draft.genome  draft.genome.fa</span><br><span class="line">\# 如果使用的是 HindIII 进行酶切，就设定为HindIII；这里我设定为MboI</span><br></pre></td></tr></table></figure>
</li>
<li><p>获取每条contig的长度</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk &#39;BEGIN&#123;OFS&#x3D;&quot;\t&quot;&#125;&#123;print $1, $NF&#125;&#39; draft.genome_MboI.txt &gt; draft.genome.chrom.sizes</span><br></pre></td></tr></table></figure>

</li>
</ol>
<p>设置当前工作目录的文件结构如下，目录存在fastq和ref文件夹。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">ref</span><br><span class="line">├── draft.genome.chrom.sizes</span><br><span class="line">├── draft.genome.fa -&gt; ..&#x2F;draft.genome.fa</span><br><span class="line">├── draft.genome.fa.amb</span><br><span class="line">├── draft.genome.fa.ann</span><br><span class="line">├── draft.genome.fa.bwt</span><br><span class="line">├── draft.genome.fa.pac</span><br><span class="line">├── draft.genome.fa.sa</span><br><span class="line">├── draft.genome_MboI.txt</span><br><span class="line">└── opt.sh</span><br><span class="line">fastq</span><br><span class="line">├── XXX_R1.fastq.gz -&gt; &#x2F;PATH&#x2F;TO&#x2F;XXX_R1.fastq.gz</span><br><span class="line">└── XXX_R2.fastq.gz -&gt; &#x2F;PATH&#x2F;TO&#x2F;XXX_R2.fastq.gz</span><br></pre></td></tr></table></figure>

<ol start="4">
<li>运行juicer<br>注意：必须在当前目录存在fastq和ref文件夹， -z,-p,-y,-D是必须参数</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">~&#x2F;software&#x2F;juicer&#x2F;scripts&#x2F;juicer.sh -g draft_genome -s HindIII -z .&#x2F;ref&#x2F;draft.genome.fa -y .&#x2F;ref&#x2F;draft.genome_HindIII.txt -p .&#x2F;ref&#x2F;draft.genome.chrom.sizes -D &#x2F;PATH&#x2F;TO&#x2F;juicer -t 8</span><br><span class="line"></span><br><span class="line">\## juicer 参数</span><br><span class="line">-g： 定义一个物种名</span><br><span class="line">-s：酶切类型, HindIII(AAGCTAGCTT), MboI(GATCGATC) , DpnII(GATCGATC), NcoI(CCATGCATGG)</span><br><span class="line">-z : 参考基因组文件</span><br><span class="line">-y: 限制性酶切位点可能出现位置文件</span><br><span class="line">-p: 染色体大小文件</span><br><span class="line">-C: 将原来的文件进行拆分，必须是4的倍数，默认是90000000, 即22.5M reads</span><br><span class="line">-S: 和任务重运行有关，从中途的某一步开始,&quot;merge&quot;, &quot;dedup&quot;, &quot;final&quot;, &quot;postproc&quot; 或 &quot;early&quot;</span><br><span class="line">-d: juicer的目录</span><br><span class="line">-D: juicer scripts的目录</span><br><span class="line">-t: 线程数</span><br></pre></td></tr></table></figure>
<p>运行记录。（5G Hi-C数据 + 900 Mb contigs 耗时6 h。）</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">Using .&#x2F;ref&#x2F;draft.genome_MboI.txt as site file</span><br><span class="line">(-: Looking for fastq files...fastq files exist</span><br><span class="line">(-: Aligning files matching &#x2F;PATH&#x2F;TO&#x2F;WORK&#x2F;DIRECTORY&#x2F;fastq&#x2F;\*_R\*.fastq*</span><br><span class="line"> to genome draft_genome with site file .&#x2F;ref&#x2F;draft.genome_MboI.txt</span><br><span class="line">(-: Created &#x2F;PATH&#x2F;TO&#x2F;WORK&#x2F;DIRECTORY&#x2F;splits and &#x2F;PATH&#x2F;TO&#x2F;WORK&#x2F;DIRECTORY&#x2F;aligned.</span><br><span class="line">(...  ...)</span><br><span class="line">(-: Postprocessing successfully completed, maps too sparse to annotate or GPUs unavailable (-:</span><br><span class="line">(-: Pipeline successfully completed (-:</span><br><span class="line">Run cleanup.sh to remove the splits directory</span><br><span class="line">Check &#x2F;PATH&#x2F;TO&#x2F;WORK&#x2F;DIRECTORY&#x2F;aligned for results</span><br></pre></td></tr></table></figure>
<p>结果：结果文件在aligned目录下，其中“<code>merged_nodups.txt</code>”就是下一步3D-DNA的输入文件之一。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">splits&#x2F;</span><br><span class="line">├── MySample.fastq.gz_abnorm.sam</span><br><span class="line">├── MySample.fastq.gz_linecount.txt</span><br><span class="line">├── MySample.fastq.gz_norm.txt.res.txt</span><br><span class="line">├── MySample.fastq.gz.sam</span><br><span class="line">├── MySample.fastq.gz.sort.txt</span><br><span class="line">├── MySample.fastq.gz_unmapped.sam</span><br><span class="line">├── MySample_R1.fastq.gz -&gt; &#x2F;PATH&#x2F;TO&#x2F;WORK&#x2F;DIRECTORY&#x2F;fastq&#x2F;MySample_R1.fastq.gz</span><br><span class="line">└── MySample_R2.fastq.gz -&gt; &#x2F;PATH&#x2F;TO&#x2F;WORK&#x2F;DIRECTORY&#x2F;fastq&#x2F;MySample_R2.fastq.gz</span><br><span class="line">aligned&#x2F;</span><br><span class="line">├── abnormal.sam</span><br><span class="line">├── collisions_nodups.txt</span><br><span class="line">├── collisions.txt</span><br><span class="line">├── dups.txt</span><br><span class="line">├── header</span><br><span class="line">├── inter_30_contact_domains</span><br><span class="line">├── inter_30.hic</span><br><span class="line">├── inter_30_hists.m</span><br><span class="line">├── inter_30.txt</span><br><span class="line">├── inter.hic</span><br><span class="line">├── inter_hists.m</span><br><span class="line">├── inter.txt</span><br><span class="line">├── merged_nodups.txt</span><br><span class="line">├── merged_sort.txt</span><br><span class="line">├── opt_dups.txt</span><br><span class="line">└── unmapped.sam</span><br></pre></td></tr></table></figure>

<p>Note: 测试的过程中遇到报错信息，原来是默认的“/opt/juicer/scripts/”需要根据实际安装路径，通过设定<code>-D</code>参数修改之。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">(-:  Align of &#x2F;PATH&#x2F;TO&#x2F;WORK&#x2F;DIRECTORY&#x2F;splits&#x2F;XXX.fastq.gz.sam done successfully</span><br><span class="line">awk: fatal: can&#39;t open source file &#96;&#x2F;opt&#x2F;juicer&#x2F;scripts&#x2F;common&#x2F;chimeric_blacklist.awk&#39; for reading (No such file or directory)</span><br><span class="line">***! Failure during chimera handling of &#x2F;PATH&#x2F;TO&#x2F;WORK&#x2F;DIRECTORY&#x2F;splits&#x2F;MySample.fastq.gz</span><br></pre></td></tr></table></figure>
<p>原来是<code>-D</code>参数指定juicer目录（包含script文件夹那个路径），<code>vi juicer.sh</code> 可见到：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">75 # Juicer directory, contains scripts&#x2F;, references&#x2F;, and restriction_sites&#x2F;</span><br><span class="line">76 # can also be set in options via -D</span><br><span class="line">77 juiceDir&#x3D;&quot;&#x2F;opt&#x2F;juicer&quot;</span><br></pre></td></tr></table></figure>

<h3 id="三、运行3D-DNA，对contig-scaffolds进行组装"><a href="#三、运行3D-DNA，对contig-scaffolds进行组装" class="headerlink" title="三、运行3D-DNA，对contig/scaffolds进行组装"></a>三、运行3D-DNA，对contig/scaffolds进行组装</h3><p>使用默认参数进行3D-DNA</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">~&#x2F;software&#x2F;3d-dna&#x2F;run-asm-pipeline.sh .&#x2F;ref&#x2F;draft.genome.fa .&#x2F;aligned&#x2F;merged_nodups.txt</span><br><span class="line"># 通过-i参数设定允许挂载的最小尺寸</span><br><span class="line">~&#x2F;software&#x2F;3d-dna&#x2F;run-asm-pipeline.sh -i 8000 .&#x2F;ref&#x2F;draft.genome.fa .&#x2F;aligned&#x2F;merged_nodups.txt</span><br></pre></td></tr></table></figure>

<p>最后输出文件中，包含 final 字样的文件就是我们需要的结果：<code>draft.genome.final.hic</code> 和 <code>draft.genome.final.assembly</code>。将这两个文件从服务器下载到客户端（我是win 10），并导入 JuiceBox 浏览器。</p>
<p>注意，生成的文件中，draft.genome.final.assembly 与 draft.genome.FINAL.assembly 是不同的文件。“FINAL” – chromosome-length scaffolds; “final” – input with all the misjoin correction introduced。<br>这两个文件： draft.genome.final.hic 和 draft.genome.final.assembly，同时导入 JuiceBox 才可以进行人工手动矫正。我最初错误地导入了 draft.genome.FINAL.assembly 文件，就遇到了问题。</p>
<h3 id="四、JuiceBox进行手动纠错"><a href="#四、JuiceBox进行手动纠错" class="headerlink" title="四、JuiceBox进行手动纠错"></a>四、JuiceBox进行手动纠错</h3><p>不少人将软件名写成juicerbox，其实正确写法是juicebox。下载该软件：<a target="_blank" rel="noopener" href="https://github.com/aidenlab/Juicebox/wiki/Download">https://github.com/aidenlab/Juicebox/wiki/Download</a>。直接下载exe可执行程序，不用安装，双击即可进入界面。</p>
<p><img src="https://oscimg.oschina.net/oscnet/d8e47314-9ca7-409e-9293-614ba5f561b3.png" alt="juicebox"></p>
<p>一般组装错误为：misjoin、translocations、inversions 和 chromosome boundaries。<br>关于该软件用法，可看视频：油管地址 <a target="_blank" rel="noopener" href="https://www.youtube.com/watch?v=Nj7RhQZHM18">Tutorial video for Juicebox Assembly Tools</a> 或 <a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av65134634">B 站</a></p>
<p>通过<code>File</code>-&gt;<code>open</code>可以导入后缀为.hic的文件<code>draft.genome.final.hic</code>，随后<code>Assembly</code>-&gt;<code>select</code>，选择<code>draft.genome.final.assembly</code>文件，这时候能看到“染色体”组装图（还不是完美的染色体，只是初步的scaffolds排序，但是我的数据里可看到9条染色体雏形，此外还有一部分没有定位到染色体上的contigs/scaffolds，但是从heatmap中可以看到他们中的大部分与某条染色体会有互作）。</p>
<p>怎么手动调整？左上角到右下角这一条线作为对角线，在对角线任选取两个点，就可以形成一个正方形，两点之间的线段就是“染色体”的一段。如果需要移动某个方形区域，按住shift键不动，在需要挪动的染色体区域左上角点击一次，即可看到一个黑色十字固定在此处，随后在右下角再单击一次，再次出现一个黑色十字，松开shift键，两个黑色十字即包围了这一段待调整的染色体，可以适当调整黑框大小。此时将鼠标移到另一个染色体位置，可见到鼠标白色图标变成黑色箭头，单击，上述的黑色框便移动到了这里，完成了移动选取的操作。如果需要对选择进行翻转呢？则只需要将鼠标移动到黑色框的右上角区域，这时候鼠标图标会变成黑色环状，单击，即可使之翻转。</p>
<p>纠错完以后，单击上方功能区<code>Assembly</code>-&gt;<code>Export Assembly</code>，导出<code>draft.genome.final.review.assembly</code>，或命名为<code>genome.review.assembly</code>，用于下一步的分析</p>
<h3 id="五、再次运行3D-DNA"><a href="#五、再次运行3D-DNA" class="headerlink" title="五、再次运行3D-DNA"></a>五、再次运行3D-DNA</h3><p>Generate a fasta sequence after JBAT review. In addition to the chromosome-length genome sequence (<em>.FINAL.fasta</em>), the pipeline will produce a new <em>.final.hic</em> file for concluding quality control.</p>
<p>会重新生成 draft.genome.rawchrom.* 和 draft.genome.final.* 这些文件，覆盖掉原来的文件。如果要避免覆盖。就需要另起一个目录运行程序。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">~&#x2F;software&#x2F;3d-dna&#x2F;run-asm-pipeline-post-review.sh -r genome.review.assembly .&#x2F;ref&#x2F;draft.genome.fa aligned&#x2F;merged_nodups.txt</span><br></pre></td></tr></table></figure>
<p>结果文件：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">draft.genome.FINAL.fasta</span><br><span class="line">draft.genome.FINAL.hic</span><br><span class="line">draft.genome.FINAL_asm.superscaf_track.txt</span><br><span class="line">draft.genome.FINAL_asm.scaffold_track.txt</span><br><span class="line">draft.genome.FINAL.assembly</span><br><span class="line">draft.genome.rawchrom.fasta</span><br><span class="line">draft.genome.rawchrom.hic</span><br><span class="line">draft.genome.final_asm.superscaf_track.txt</span><br><span class="line">draft.genome.final_asm.scaffold_track.txt</span><br><span class="line">draft.genome.final.asm</span><br><span class="line">draft.genome.mnd.txt</span><br><span class="line">draft.genome.rawchrom.asm</span><br><span class="line">draft.genome.rawchrom.cprops</span><br><span class="line">draft.genome.60G.final.review.assembly</span><br><span class="line">draft.genome.cprops</span><br></pre></td></tr></table></figure>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2020/12/25/Linux%EF%BC%9Avim%E8%AE%BE%E7%BD%AE/">Linux：vim设置</a><a class="next" href="/2020/12/15/%E4%B8%80%E4%BA%9BFASTA-FASTQ%E5%BA%8F%E5%88%97%E5%A4%84%E7%90%86%E5%B7%A5%E5%85%B7/">一些FASTA/FASTQ序列处理工具</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/17/Linux-Gene-Ontology-GOMAP/">Bioinfo | Gene Ontology - GOMAP</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Linux | 认识 PKG_CONFIG_PATH</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">jcvi 安装与使用，报错以及解决办法</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">Bioinfo | Installation of OrthoMCL-pipeline</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Bioinfo | Using Circos for genome visualization</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">Bioinfo | Repeat sequence analysis by RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer)</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/16/Bioinfo-understanding-sequence-identity/">Bioinfo | 怎么理解序列一致性（Sequence identity）</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>