<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>如何将RNA格式的fasta文件转换为DNA格式 | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">如何将RNA格式的fasta文件转换为DNA格式</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">如何将RNA格式的fasta文件转换为DNA格式</h1><div class="post-meta">2014-10-14<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>从miRBase数据库下载的pre-miRNA序列是RNA格式的，如下所示。 &gt;cel-mir-37 MI0000008 Caenorhabditis elegans miR-37 stem-loop UUCUAGAAACCCUUGGACCAGUGUGGGUGUCCGUUGCGGUGCUACAUUCUCUAAUCUGUA UCACCGGGUGAACACUUGCAGUGGUCCUCGUGGUUUCU 因数据分析的需要将其转换为DNA格式。</p>
<a id="more"></a>
<p>可使用FASTX-toolkit的工具。下载地址是：<a target="_blank" rel="noopener" href="http://hannonlab.cshl.edu/fastx/_toolkit/index.html">http://hannonlab.cshl.edu/fastx\_toolkit/index.html</a> 分别下载如下两个安装包。 <a target="_blank" rel="noopener" href="https://github.com/agordon/libgtextutils">libGTextUtils on Github</a> <a target="_blank" rel="noopener" href="https://github.com/agordon/fastx_toolkit">Fastx-Toolkit on Github</a> 下载完毕，先安装libGTextUtils，然后安装Fastx-toolkit。 在安装Fastx-toolkit的过程中，./configure时候遇到报错： checking for GTEXTUTILS… configure: error: Package requirements (gtextutils) were not met: No package ‘gtextutils’ found Consider adjusting the PKG_CONFIG_PATH environment variable if you installed software in a non-standard prefix. Alternatively, you may set the environment variables GTEXTUTILS_CFLAGS and GTEXTUTILS_LIBS to avoid the need to call pkg-config. See the pkg-config man page for more details.   解决方案是：</p>
<p>$ export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig:$PKG_CONFIG_PATH</p>
<p>然后就可以顺利./configure了。随后 make &amp; make install 就搞定这个工具包了。 ======== 参考资料： [1] <a target="_blank" rel="noopener" href="http://hannonlab.cshl.edu/fastx/_toolkit/index.html">http://hannonlab.cshl.edu/fastx\_toolkit/index.html</a> [2] <a target="_blank" rel="noopener" href="http://hannonlab.cshl.edu/fastx/_toolkit/pkg/_config/_email.txt">http://hannonlab.cshl.edu/fastx\_toolkit/pkg\_config\_email.txt</a></p>
<p>-——- Original Message ——–<br>Subject: Re: FASTX-Toolkit installation<br>Date: Tue, 28 Apr 2009 14:53:21 -0400<br>From: Assaf Gordon <a href="mailto:&#x67;&#111;&#114;&#x64;&#111;&#x6e;&#x40;&#99;&#115;&#x68;&#108;&#46;&#x65;&#x64;&#117;">&#x67;&#111;&#114;&#x64;&#111;&#x6e;&#x40;&#99;&#115;&#x68;&#108;&#46;&#x65;&#x64;&#117;</a></p>
<p>… wrote, On 04/28/2009 02:30 PM:</p>
<blockquote>
<p>I’m trying to install the fastx-toolkit on my Mac Leopard workstation.<br>Ultimately, I’d like to integrate it into my local copy of Galaxy.</p>
<p>However, I ran into a snag trying to compile the latest version.  I was able to configure,<br>make and install libgtextutils-0.1.  At least there were no errors after I did that.<br>However, I run into the following when trying to configure fastx-toolkit-0.0.7:</p>
<pre><code>checking for pkg-config... /sw/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for GTEXTUTILS... configure: error: Package requirements
(gtextutils-0.1) were not met:

No package &#39;gtextutils-0.1&#39; found

Consider adjusting the PKG\_CONFIG\_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables GTEXTUTILS\_CFLAGS
and GTEXTUTILS\_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.

gen-biocomp-lp:fastx\_toolkit-0.0.7 lparsons$ make
make: \*\*\* No targets specified and no makefile found.  Stop.</code></pre>
<p>Any suggestions would be appreciated.</p>
</blockquote>
<p>The short answer is that you probably need to set the environment<br>variable PKG_CONFIG_PATH, like so:<br>  $ export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig:$PKG_CONFIG_PATH</p>
<p>Before running FASTX’s “configure” script.</p>
<p>But there’s also the long answer:</p>
<p>When you run a standard “./configure” script, the default installation<br>path is “/usr/local”. On some systems (mostly debian/ubuntu), the<br>default path for most programs is just “/usr” (without “local”).</p>
<p>On the other hand, some system administrators like the separation of<br>‘standard’ packages into “/usr” and manually built packages into<br>“/usr/local”.</p>
<p>I’m telling you that because it is likely that “gtextutils” was<br>installed into “/usr/local”, but you system doesn’t look there at all.</p>
<p>So here’s how to check:<br>go back to the gtextutil directory.<br>Assuming you already run “configure” + “make”, run “make install” as a<br>REGULAR user (not root).<br>This will surely fail, but you will see the path, where the installation<br>process tried to put the library files.<br>Make a note if the path starts with /usr/local or just /usr (or<br>strangely - something else).</p>
<p>The installation of GTEXTUTILS puts a file (named “gtextutils-0.1.pc”)<br>somewhere on you disk.<br>On my computer, it is found on:<br>  /usr/local/lib/pkgconfig/gtextutils-0.1.pc</p>
<p>The program “pkg-config” is a standard program which reads those “.pc”<br>files and tells the compiler how to use the library (and it is used by<br>“configure”).</p>
<p>From the command line, run:<br>  $ pkg-config –cflags gtextutils-0.1</p>
<p>If your file was installed into “/usr/local/lib…” but pkg-config shows<br>the following error:<br>  Package gtextutils-0.1 was not found in the pkg-config search path.<br>  Perhaps you should add the directory containing `gtextutils-0.1.pc’<br>  to the PKG_CONFIG_PATH environment variable<br>  No package ‘gtextutils-0.1’ found</p>
<p>It is like that pkg-config simply doesn’t look in “/usr/local/“ for the<br>needed files.</p>
<p>The solution is set a special environment variable, that will instruct<br>pkg-config where to look:<br>  $ export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig:$PKG_CONFIG_PATH<br>  $ pkg-config –cflags gtextutils-0.1<br>  -I/usr/local/include/gtextutils-0.1/</p>
<p>Setting “PKG_CONFIG_PATH” to where the “.pc” file is stored will help<br>the configure process find this library.<br>After running the “export PKG_CONFIG_PATH” (and assuming the pkg-config<br>command succeeded), you should be able to run FASTX’s “./configure”<br>successfully.</p>
<hr>
<p>A different solution is to configure GTEXTUTILS to be installed in<br>“/usr” to begin with (and then, most of the PATH related problems should<br>be gone).<br>To do this, run:<br>    $ ./configure –prefix=/usr<br>    $ make<br>    $ sudo make install</p>
<p>In the GTEXTUTILS directory.</p>
<hr>
<p>I hope these tips were helpful.<br>If you still encounter problems, please let me know.</p>
<p>-Gordon.</p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2014/12/12/importerror-no-module-named-toolz/">Python报错：ImportError: No module named toolz</a><a class="next" href="/2014/09/15/tophat-attributeerror-nonetype-object-has-no-attribute-group/">Tophat: AttributeError: 'NoneType' object has no attribute 'group'</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2020/12/30/Demo-blog/">Demo blog</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/29/Linux%EF%BC%9A%E4%BD%BF%E7%94%A8Modules%E7%AE%A1%E7%90%86%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/">Linux：使用Environment Modules管理环境变量</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E4%BD%BF%E7%94%A8Snakemake%E6%90%AD%E5%BB%BA%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/">使用Snakemake搭建生物信息分析流程</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9%E7%9A%84SAM-BAM%E6%96%87%E4%BB%B6%E4%B8%AD%EF%BC%8C%E6%80%8E%E4%B9%88%E7%90%86%E8%A7%A3MAPQ%E5%80%BC/">基因组比对的SAM/BAM文件中，怎么理解MAPQ值</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/25/Linux%EF%BC%9Avim%E8%AE%BE%E7%BD%AE/">Linux：vim设置</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/%E4%B8%80%E4%BA%9BFASTA-FASTQ%E5%BA%8F%E5%88%97%E5%A4%84%E7%90%86%E5%B7%A5%E5%85%B7/">一些FASTA/FASTQ序列处理工具</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/12/%E5%A4%A7%E8%B1%86%E4%BA%A7%E4%B8%9A%E7%8E%B0%E7%8A%B6/">大豆产业现状</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/07/1st_post/">My Fast HexoPages</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/07/hello-world/">Hello World</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/06/%E5%AE%89%E8%A3%85hic-pro/">安装HiC-Pro （HiC数据预处理工具）</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2020 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>