<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2015/09/25/%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/">转录组数据分析</a></h1><div class="post-meta">2015-09-25</div><div class="post-content"><ul>
<li>  <strong>1. filtering</strong></li>
</ul>
<blockquote>
<p>fastqc, iTools, …</p></div><p class="readmore"><a href="/2015/09/25/%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2015/08/28/%E5%85%8D%E8%B4%B9%E4%B8%8B%E8%BD%BD%E8%8B%B1%E6%96%87%E6%96%87%E7%8C%AE%E7%9A%84%E6%96%B9%E6%B3%95/">免费下载英文文献的方法</a></h1><div class="post-meta">2015-08-28</div><div class="post-content"><p>一、国外代理下载。只需在文献网址里面插入代理网站的地址（_sci-hub.org 或者 sci-hub.club_），如下操作。</p>
<blockquote>
<p>原文献地址：<a target="_blank" rel="noopener" href="http://www.plantcell.org/content/27/6/1595.full">http://www.plantcell.org/content/27/6/1595.full</a> 下载的地址：<a target="_blank" rel="noopener" href="http://www.plantcell.org._sci-hub.org_/content/27/6/1595.full">http://www.plantcell.org._sci-hub.org_/content/27/6/1595.full</a> （代理网址可能无法长期使用。）</p></div><p class="readmore"><a href="/2015/08/28/%E5%85%8D%E8%B4%B9%E4%B8%8B%E8%BD%BD%E8%8B%B1%E6%96%87%E6%96%87%E7%8C%AE%E7%9A%84%E6%96%B9%E6%B3%95/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2015/08/25/fastq-quality-control-and-filtering/">Fastq quality control and filtering</a></h1><div class="post-meta">2015-08-25</div><div class="post-content"><p><strong>数据质量QC：<a target="_blank" rel="noopener" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastqc</a></strong> fastqc <strong>去除adapter：<a target="_blank" rel="noopener" href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a></strong> （Trimmomatic: A flexible read trimming tool for Illumina NGS data）</p></div><p class="readmore"><a href="/2015/08/25/fastq-quality-control-and-filtering/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2015/06/23/fasta%E6%96%87%E4%BB%B6%E6%93%8D%E4%BD%9Cfasta-file-statstics/">Fasta与Fastq文件操作</a></h1><div class="post-meta">2015-06-23</div><div class="post-content"><p><strong>1 fastx_toolkits</strong></p>
<blockquote>
<p>fastx_artifacts_filter fastx_nucleotide_distribution_graph.sh fastx_reverse_complement<br> fastx_barcode_splitter.pl fastx_nucleotide_distribution_line_graph.sh fastx_trimmer<br> fastx_clipper fastx_quality_stats fastx_uncollapser<br> fastx_collapser fastx_renamer</p></div><p class="readmore"><a href="/2015/06/23/fasta%E6%96%87%E4%BB%B6%E6%93%8D%E4%BD%9Cfasta-file-statstics/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2015/06/15/%E5%A6%82%E4%BD%95%E7%94%A8gatk-call-snp/">如何用GATK call snp</a></h1><div class="post-meta">2015-06-15</div><div class="post-content"><h2 id="如何用GATK-call-snp（转自http-www-cnblogs-com-freemao-p-3763885-html）"><a href="#如何用GATK-call-snp（转自http-www-cnblogs-com-freemao-p-3763885-html）" class="headerlink" title="如何用GATK call snp（转自http://www.cnblogs.com/freemao/p/3763885.html）"></a><a target="_blank" rel="noopener" href="http://www.cnblogs.com/freemao/p/3763885.html">如何用GATK call snp</a>（转自<a target="_blank" rel="noopener" href="http://www.cnblogs.com/freemao/p/3763885.html%EF%BC%89">http://www.cnblogs.com/freemao/p/3763885.html）</a></h2><p>1， 什么是GATK？ The Genome Analysis Toolkit or GATK is a software package developed at the Broad Institute to analyse next-generation resequencing data. The toolkit offers a wide variety of tools, with a primary focus on variant discovery and genotyping as well as strong emphasis on data quality assurance. Its robust architecture, powerful processing engine and high-performance computing features make it capable of taking on projects of any size. 2， 如何用GATK call SNP？</p></div><p class="readmore"><a href="/2015/06/15/%E5%A6%82%E4%BD%95%E7%94%A8gatk-call-snp/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2015/05/10/%E6%A4%8D%E7%89%A9%E7%A7%8D%E5%AD%90%E8%90%8C%E5%8F%91/">植物种子萌发</a></h1><div class="post-meta">2015-05-10</div><div class="post-content"><p>萌发实验 选取发育良好的 30 粒成熟种子或果实，进行消毒（用 0.05%高锰酸甲浸泡 15 分 钟）后再用无菌水冲洗干净，播种于花盆中，播深为 0.5cm 左右，在室温下测定种子 的萌发率，重复 3 次。 贺慧. <em>阿拉善主要荒漠植物种子 (果实) 形态结构和萌发特性的适应性研究</em>. Diss. 内蒙古农业大学, 2008.</p></div><p class="readmore"><a href="/2015/05/10/%E6%A4%8D%E7%89%A9%E7%A7%8D%E5%AD%90%E8%90%8C%E5%8F%91/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2015/04/29/sql%E5%AF%BC%E5%85%A5blast2go/">SQL导入blast2go</a></h1><div class="post-meta">2015-04-29</div><div class="post-content"><p>1 安装mysql，参考前面博文。 2 进入mysql [fenglei@localhost scripts]$ <strong>mysql -P 33060 -u root -p</strong> Enter password: Welcome to the MySQL monitor. Commands end with ; or g. Your MySQL connection id is 5 Server version: 5.6.20 Source distribution Copyright (c) 2000, 2014, Oracle and/or its affiliates. All rights reserved. Oracle is a registered trademark of Oracle Corporation and/or its affiliates. Other names may be trademarks of their respective owners. Type ‘help;’ or ‘h’ for help. Type ‘c’ to clear the current input statement. mysql&gt; <strong>insert into mysql.user(Host,User,Password) values(“localhost”,”blast2go”,password(“blast4it”));</strong> Query OK, 1 row affected, 3 warnings (0.05 sec) mysql&gt; <strong>flush privileges;</strong> Query OK, 0 rows affected (0.02 sec) mysql&gt;<strong>exit</strong> Bye [fenglei@localhost scripts]$ <strong>mysql -hlocalhost -u blast2go -pblast4it</strong> Warning: Using a password on the command line interface can be insecure. Welcome to the MySQL monitor. Commands end with ; or g. Your MySQL connection id is 6 Server version: 5.6.20 Source distribution Copyright (c) 2000, 2014, Oracle and/or its affiliates. All rights reserved. Oracle is a registered trademark of Oracle Corporation and/or its affiliates. Other names may be trademarks of their respective owners. Type ‘help;’ or ‘h’ for help. Type ‘c’ to clear the current input statement.</p></div><p class="readmore"><a href="/2015/04/29/sql%E5%AF%BC%E5%85%A5blast2go/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2015/03/21/%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E4%B9%8Bhisat%E6%AF%94%E5%AF%B9/">转录组数据分析之HISAT比对</a></h1><div class="post-meta">2015-03-21</div><div class="post-content"><p>HISAT全称为Hierarchical Indexing for Spliced Alignment of Transcripts，由约翰霍普金斯大学开发。它取代Bowtie/TopHat程序，能够将RNA-Seq的读取与基因组进行快速比对。这项成果发表在3月9日的《Nature Methods》上。 HISAT利用大量FM索引，以覆盖整个基因组。以人类基因组为例，它需要48,000个索引，每个索引代表~64,000 bp的基因组区域。这些小的索引结合几种比对策略，实现了RNA-Seq读取的高效比对，特别是那些跨越多个外显子的读取。尽管它利用大量索引，但HISAT只需要4.3 GB的内存。这种应用程序支持任何规模的基因组，包括那些超过40亿个碱基的。 第一步：建立数据库索引 命令：hisat-build genome.fa genome 将得到<code>NAME.1.bt2</code>, <code>NAME.2.bt2</code>,<code>NAME.3.bt2</code>, <code>NAME.4.bt2</code>, <code>NAME.5.bt2</code>, <code>NAME.6.bt2</code>, <code>NAME.rev.1.bt2</code>, <code>NAME.rev.2.bt2</code>, <code>NAME.rev.5.bt2</code>, and <code>NAME.rev.6.bt2 这六个文件。</code> 第二步：将转录组数据比对到基因组 命令：hisat -p 8 –phred64 -x genome -1 reads.1.fq.gz -2 reads.2.fq.gz -S out.sam [使用拟南芥转录组数据，该步骤6分钟即可完成，而原来的TopHat需要1小时。] 第三步：将sam转为bam文件，并对bam文件进行排序。 第四部：使用cufflinks进行下游分析。</p></div><p class="readmore"><a href="/2015/03/21/%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E4%B9%8Bhisat%E6%AF%94%E5%AF%B9/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2015/02/05/could-not-find-function-sqlitequicksql/">could not find function \&quot;sqliteQuickSQL\&quot;</a></h1><div class="post-meta">2015-02-05</div><div class="post-content"><p>问题描述：转录组测序数据分析，使用cummeRbund的过程中报错： &gt; library(cummeRbund) Loading required package: BiocGenerics Loading required package: parallel Attaching package: ‘BiocGenerics’ The following objects are masked from ‘package:parallel’: clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap, parApply, parCapply, parLapply, parLapplyLB, parRapply, parSapply, parSapplyLB The following object is masked from ‘package:stats’: xtabs The following objects are masked from ‘package:base’: anyDuplicated, append, as.data.frame, as.vector, cbind, colnames, do.call, duplicated, eval, evalq, Filter, Find, get, intersect, is.unsorted, lapply, Map, mapply, match, mget, order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rep.int, rownames, sapply, setdiff, sort, table, tapply, union, unique, unlist Loading required package: RSQLite Loading required package: DBI Loading required package: ggplot2 Loading required package: reshape2 Loading required package: fastcluster Attaching package: ‘fastcluster’ The following object is masked from ‘package:stats’: hclust Loading required package: rtracklayer Loading required package: GenomicRanges Loading required package: IRanges Loading required package: GenomeInfoDb Loading required package: Gviz Loading required package: grid Attaching package: ‘cummeRbund’ The following object is masked from ‘package:GenomicRanges’: promoters The following object is masked from ‘package:IRanges’: promoters &gt; cuff_data &lt;- readCufflinks(‘diff_out_CM_c.vs.m’) Creating database diff_out_CM_c.vs.m/cuffData.db Error in FUN(c(“n– Creator: MySQL Workbench 5.2.33/ExportSQLite plugin 2009.12.02n– Author: Loyal Goffn– Caption: New Modeln– Project: Name of the projectn– Changed: 2012-04-30 22:21n– Created: 2011-05-02 12:52nPRAGMA foreign_keys = OFF”, : could not find function “sqliteQuickSQL” ======================================= 失败的解决方法： &gt; remove.packages(“RSQLite”) &gt;install.packages(“/home/fenglei/Downloads/RSQLite_0.9-0.tar.gz”) &gt; library(RSQLite) Loading required package: DBI &gt; install.packages(“DBI”) 成功解决方法 First, upgrade biocLite: &gt;source(“<a target="_blank" rel="noopener" href="http://www.bioconductor.org/biocLite.R&quot;">http://www.bioconductor.org/biocLite.R&quot;</a>) &gt;biocLite(“BiocUpgrade”) Then, install “cummeRbund” &gt;biocLite(“cummeRbund”) Last, install or reinstall “RSQLite” &gt;biocLite(“RSQLite”) I hope it works for you.</p></div><p class="readmore"><a href="/2015/02/05/could-not-find-function-sqlitequicksql/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2015/01/02/%E4%BD%BF%E7%94%A8reseqtoolsitools%E8%BF%9B%E8%A1%8Cfastq%E6%95%B0%E6%8D%AE%E7%BB%9F%E8%AE%A1%E3%80%81%E8%B4%A8%E6%A3%80%E3%80%81%E8%BF%87%E6%BB%A4/">使用ReSeqTools/iTools进行fastq数据统计、质检、过滤</a></h1><div class="post-meta">2015-01-02</div><div class="post-content"><p>重测序分析集成工具包 linux 64位平台 静态编译 chmod 便可以直接运行了。 （ReSeqTools/iTools） 即在linux64机上只须</p>
<blockquote>
<ul>
<li>  tar -zxvf ReSeqTools_XXX.tar.gz ;</li>
<li>  cd iTools_Code; chmod 775 iTools ;</li>
<li>  ./ iTools -h ;</li>
<li>  无须重新make</li>
</ul>
</blockquote>
<p>QQ交流群 群名称是Reseqtools QQ群号： 125293663</p></div><p class="readmore"><a href="/2015/01/02/%E4%BD%BF%E7%94%A8reseqtoolsitools%E8%BF%9B%E8%A1%8Cfastq%E6%95%B0%E6%8D%AE%E7%BB%9F%E8%AE%A1%E3%80%81%E8%B4%A8%E6%A3%80%E3%80%81%E8%BF%87%E6%BB%A4/">Read More</a></p></div><nav class="page-navigator"><a class="extend prev" rel="prev" href="/page/12/">Previous</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/11/">11</a><a class="page-number" href="/page/12/">12</a><span class="page-number current">13</span><a class="page-number" href="/page/14/">14</a><a class="page-number" href="/page/15/">15</a><a class="extend next" rel="next" href="/page/14/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/16/Bioinfo-understanding-sequence-identity/">Bioinfo | 怎么理解序列一致性（Sequence identity）</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/10/Bioinfo-Extracting-sequence-by-names/">Bioinfo | 依据序列名字列表从 fasta 文件提取序列</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/05/Proteomics-Serotonylation-A-novel-histone-H3-marker/">Proteomics | 组蛋白五羟色胺化修饰 H3Q5ser 及其生物学功能</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-What-the-differences-between-DDA-and-DIA-in-mass-spectrometer/">Proteomics | 质谱分析中DDA与DIA有什么不同？</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-Protein-Persulfidation/">Proteomics | 蛋白质硫巯基化修饰</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/Proteomics-Common-reagents-and-their-functions-in-LC-MS-MS-proteomics/">Proteomics | Common reagents and their functions in LC-MS/MS proteomics</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/R-Using-ggrepel-to-repel-overlapping-text-labels-in-ggplot2/">R | Using ggrepel to repel overlapping text labels in ggplot2</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/08/Bioinfo-Installation-of-BUSCO-v4/">Bioinfo | Installation of BUSCO v4</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/05/Linux-Installation-of-Tomcat-9/">Linux | Installation of Tomcat 9, Java 8, and MySQL 5.7.20</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>