<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>MCScanX 测试 | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">MCScanX 测试</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">MCScanX 测试</h1><div class="post-meta">2019-07-05<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a><a href="/categories/Linux/">Linux</a></span></div><div class="post-content"><p>安装</p>
<p>[fenglei@localhost MCScanX]$ make<br>g++ struct.cc mcscan.cc read_data.cc out_utils.cc dagchainer.cc msa.cc permutation.cc -o MCScanX<br>msa.cc: In function ‘void msa_main(const char*)’:<br>msa.cc:289:22: error: ‘chdir’ was not declared in this scope<br>if (chdir(html_fn)&lt;0)<br>^<br>make: *** [makefile:2: mcscanx] Error 1</p>
<p>解决办法</p>
<p>错误的原因是MCScanX 不支持64位系统, 如果要在 64位上运行, 需要在<code>msa.h</code>, <code>dissect_multiple_alignment.h</code>, and <code>detect_collinear_tandem_arrays.h</code>第一行加上</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">#include &lt;unistd.h&gt;</span><br></pre></td></tr></table></figure>

<p>如果要对两个基因组（以soybean和medicago为例）进行共线性分析，需要收集两个物种的数据编码基因的gff3文件和CDS序列文件，然后进行下面的操作。</p>
<p>  首先利用gff3文件，得到每一个基因的起始和终止坐标信息，注意基因有多个transcripts，只需要保留“xxxx.1”的转录本，不同物种的序列用awk等工具灵活处理。</p>
<p>python -m jcvi.formats.gff bed –type=mRNA –key=Name Gmax_275_Wm82.a2.v1.gene.gff3.gz -o soybean.bed<br>python -m jcvi.formats.gff bed –type=mRNA –key=Name Mtruncatula_285_Mt4.0v1.gene.gff3 -o medicago.bed</p>
<p>cat soybean.bed  awk ‘BEGIN{OFS=”\t”}{if(length($4) == 17 &amp;&amp; $4 ~ /\.1$/){gsub(/\.1$/,””,$4);gsub(/Chr/,”gm”,$1); print $1,$4,$2,$3}}’ &gt; soybean_gff.tab<br>cat medicago.bed  awk ‘BEGIN{OFS=”\t”}{if(length($4) == 15 &amp;&amp; $4 ~ /\.1$/){gsub(/\.1/,””,$4);gsub(/chr/,”mt”,$1); print $1,$4,$2,$3}}’ &gt; medicago_gff.tab</p>
<p>得到gff格式文件如下。 soybean.gff</p>
<p>gm01 Glyma.01G000100 27354 28320<br>gm01 Glyma.01G000200 58974 67527<br>gm01 Glyma.01G000300 67769 69968<br>gm01 Glyma.01G000400 90151 95947<br>gm01 Glyma.01G000500 90288 91197<br>gm01 Glyma.01G000600 116093 127845<br>gm01 Glyma.01G000700 143466 155573<br>gm01 Glyma.01G000800 157029 157772<br>gm01 Glyma.01G000900 170533 193342<br>gm01 Glyma.01G001000 196255 201895</p>
<p>  medicago.gff</p>
<p>mt1 Medtr1g004930 688 7332<br>mt1 Medtr1g004940 6523 7366<br>mt1 Medtr1g004950 14513 15729<br>mt1 Medtr1g004960 16282 18382<br>mt1 Medtr1g006530 28972 29214<br>mt1 Medtr1g004980 31972 32344<br>mt1 Medtr1g004990 35909 40554<br>mt1 Medtr1g006490 44429 46280<br>mt1 Medtr1g006590 46547 50437<br>mt1 Medtr1g006600 51360 54977</p>
<p>  随后对CDS序列文件进行整理，保留primary transcripts的序列。序列名称跟上面gff文件里面要一致。并灵活选用工具将CDS序列转换成蛋白序列，或者一开始就用蛋白序列也可以。</p>
<p>awk ‘{print $1}’ ../jcvi_190704/Gmax_275_Wm82.a2.v1.cds.fa &gt; Gmax_275_Wm82.a2.v1.cds.fa_simplified<br>awk ‘{print $1}’ ../jcvi_190704/Mtruncatula_285_Mt4.0v1.cds.fa &gt; Mtruncatula_285_Mt4.0v1.cds.fa_simplified</p>
<p>fasta_formatter -t -i Mtruncatula_285_Mt4.0v1.cds.fa_simplified -o Mtruncatula_285_Mt4.0v1.cds.tab<br>fasta_formatter -t -i Gmax_275_Wm82.a2.v1.cds.fa_simplified -o Gmax_275_Wm82.a2.v1.cds.tab</p>
<p>cat Gmax_275_Wm82.a2.v1.cds.tab  awk ‘BEGIN{OFS=”\t”}{if(length($1) == 17 &amp;&amp; $1 ~ /\.1$/){gsub(/\.1$/,””,$1); print “\&gt;”$1”\n”$2}}’ &gt; soybean_cds.fa<br>cat Mtruncatula_285_Mt4.0v1.cds.tab  awk ‘BEGIN{OFS=”\t”}{if(length($1) == 15 &amp;&amp; $1 ~ /\.1$/){gsub(/\.1/,””,$1); print “\&gt;”$1”\n”$2}}’ &gt; medicago.cds.fa<br>iTools Fatools CDS2Pep -InCDS soybean_cds.fa -OutPut soybean_pep.fa<br>iTools Fatools CDS2Pep -InCDS medicago.cds.fa -OutPut medicago.pep.fa</p>
<p>  然后用blastp对两个物种的蛋白序列进行比对，并用MCScanX命令检测共线性区间。</p>
<p>makeblastdb -in medicago.pep.fa -dbtype prot -out index/mt -parse_seqids</p>
<h2 id="diamond-makedb-–in-medicago-pep-fa-–db-diamond-index-mt"><a href="#diamond-makedb-–in-medicago-pep-fa-–db-diamond-index-mt" class="headerlink" title="diamond makedb –in medicago.pep.fa –db ./diamond_index/mt"></a>diamond makedb –in medicago.pep.fa –db ./diamond_index/mt</h2><h2 id="diamond-blastp-–query-soybean-pep-fa-–evalue-1e-5-–outfmt-6-–threads-24-–db-diamond-index-mt-dmnd-–out-soybean-medicago-blast"><a href="#diamond-blastp-–query-soybean-pep-fa-–evalue-1e-5-–outfmt-6-–threads-24-–db-diamond-index-mt-dmnd-–out-soybean-medicago-blast" class="headerlink" title="diamond blastp –query soybean_pep.fa –evalue 1e-5 –outfmt 6 –threads 24 –db ./diamond_index/mt.dmnd –out soybean_medicago.blast"></a>diamond blastp –query soybean_pep.fa –evalue 1e-5 –outfmt 6 –threads 24 –db ./diamond_index/mt.dmnd –out soybean_medicago.blast</h2><p>blastp -query soybean_pep.fa -db index/mt -out gm_mt.blast -evalue 1e-5 -num_threads 24 -outfmt 6 -num_alignments 5<br>mkdir mcdata<br>cp gm_mt.blast mcdata<br>cat soybean.gff medicago.gff &gt; mcdata/gm_mt.gff<br>MCScanX ./mcdata/gm_mt</p>
<p>产生的线性文件（gm_mt.collinearity）如下，可以用于下游画图。</p>
<p>[fenglei@localhost gm_mt_test]$ head -n 50 gm_mt.collinearity<br>############### Parameters ###############</p>
<h1 id="MATCH-SCORE-50"><a href="#MATCH-SCORE-50" class="headerlink" title="MATCH_SCORE: 50"></a>MATCH_SCORE: 50</h1><h1 id="MATCH-SIZE-5"><a href="#MATCH-SIZE-5" class="headerlink" title="MATCH_SIZE: 5"></a>MATCH_SIZE: 5</h1><h1 id="GAP-PENALTY-1"><a href="#GAP-PENALTY-1" class="headerlink" title="GAP_PENALTY: -1"></a>GAP_PENALTY: -1</h1><h1 id="OVERLAP-WINDOW-5"><a href="#OVERLAP-WINDOW-5" class="headerlink" title="OVERLAP_WINDOW: 5"></a>OVERLAP_WINDOW: 5</h1><h1 id="E-VALUE-1e-05"><a href="#E-VALUE-1e-05" class="headerlink" title="E_VALUE: 1e-05"></a>E_VALUE: 1e-05</h1><h1 id="MAX-GAPS-25"><a href="#MAX-GAPS-25" class="headerlink" title="MAX GAPS: 25"></a>MAX GAPS: 25</h1><p>############### Statistics ###############</p>
<h1 id="Number-of-collinear-genes-51572-Percentage-49-62"><a href="#Number-of-collinear-genes-51572-Percentage-49-62" class="headerlink" title="Number of collinear genes: 51572, Percentage: 49.62"></a>Number of collinear genes: 51572, Percentage: 49.62</h1><h1 id="Number-of-all-genes-103927"><a href="#Number-of-all-genes-103927" class="headerlink" title="Number of all genes: 103927"></a>Number of all genes: 103927</h1><p>##########################################</p>
<h2 id="Alignment-0-score-1008-0-e-value-1-4e-69-N-23-gm01-amp-mt1-plus"><a href="#Alignment-0-score-1008-0-e-value-1-4e-69-N-23-gm01-amp-mt1-plus" class="headerlink" title="Alignment 0: score=1008.0 e_value=1.4e-69 N=23 gm01&amp;mt1 plus"></a>Alignment 0: score=1008.0 e_value=1.4e-69 N=23 gm01&amp;mt1 plus</h2><p>0- 0: Glyma01G232800 Medtr1g103160 0<br>0- 1: Glyma01G234900 Medtr1g103450 8e-39<br>0- 2: Glyma01G235100 Medtr1g103490 0.0001<br>0- 3: Glyma01G236800 Medtr1g103500 2e-87<br>0- 4: Glyma01G237000 Medtr1g103550 9e-47<br>0- 5: Glyma01G237300 Medtr1g103570 2e-45<br>0- 6: Glyma01G237400 Medtr1g103600 1e-144<br>0- 7: Glyma01G237900 Medtr1g103690 9e-39<br>0- 8: Glyma01G238200 Medtr1g103830 0<br>0- 9: Glyma01G238400 Medtr1g104500 1e-170<br>0- 10: Glyma01G238500 Medtr1g104520 6e-33<br>0- 11: Glyma01G238700 Medtr1g104680 0<br>0- 12: Glyma01G238800 Medtr1g104750 0<br>0- 13: Glyma01G238900 Medtr1g104800 2e-80<br>0- 14: Glyma01G239200 Medtr1g104870 0<br>0- 15: Glyma01G239400 Medtr1g104930 1e-50<br>0- 16: Glyma01G239800 Medtr1g105075 1e-32<br>0- 17: Glyma01G239900 Medtr1g105305 5e-101<br>0- 18: Glyma01G240000 Medtr1g105415 3e-91<br>0- 19: Glyma01G240500 Medtr1g105555 5e-62<br>0- 20: Glyma01G240600 Medtr1g105570 6e-06<br>0- 21: Glyma01G240800 Medtr1g105650 2e-175<br>0- 22: Glyma01G240900 Medtr1g105655 4e-164</p>
<h2 id="Alignment-1-score-741-0-e-value-1e-39-N-16-gm01-amp-mt1-plus"><a href="#Alignment-1-score-741-0-e-value-1e-39-N-16-gm01-amp-mt1-plus" class="headerlink" title="Alignment 1: score=741.0 e_value=1e-39 N=16 gm01&amp;mt1 plus"></a>Alignment 1: score=741.0 e_value=1e-39 N=16 gm01&amp;mt1 plus</h2><p>1- 0: Glyma01G084800 Medtr1g093020 0<br>1- 1: Glyma01G086700 Medtr1g093080 1e-138<br>1- 2: Glyma01G086800 Medtr1g093095 2e-108<br>1- 3: Glyma01G087500 Medtr1g093600 5e-175<br>1- 4: Glyma01G087700 Medtr1g093630 0<br>1- 5: Glyma01G088000 Medtr1g093650 8e-07<br>1- 6: Glyma01G088200 Medtr1g093670 9e-115<br>1- 7: Glyma01G088300 Medtr1g093750 1e-118<br>1- 8: Glyma01G089100 Medtr1g093770 0<br>1- 9: Glyma01G089300 Medtr1g093850 0<br>1- 10: Glyma01G089900 Medtr1g093860 0<br>1- 11: Glyma01G090200 Medtr1g093900 0<br>1- 12: Glyma01G090400 Medtr1g093920 9e-22<br>1- 13: Glyma01G091000 Medtr1g094010 0</p>
<p>注意画图的配置文件不要出错，双斜杠（//）表示注释，注释号前面用tab符号跟染色体名称分隔，或者干脆去掉注释，我删掉注释才顺利画图，否则报错。</p>
<p>[fenglei@localhost gm_mt_test]$ cat dot.ctl<br>2000<br>2000<br>gm01,gm02,gm03,gm04,gm05,gm06,gm07,gm08,gm09,gm10,gm11,gm12,gm13,gm14,gm15,gm16,gm17,gm18,gm19,gm20<br>mt1,mt2,mt3,mt4,mt5,mt6,mt7,mt8</p>
<p>画图命令。注意 dot_plotter.java 与 dot_plotter .class 要在一个文件夹下面。其他环状图做法也类似。</p>
<p>java dot_plotter -g gm_mt.gff -s gm_mt.collinearity -c dot.ctl -o gm_mt.test.png</p>
<p>画图结果： <img src="https://genehub.files.wordpress.com/2019/07/gm_mt.test_.png" alt="gm_mt.test"> 与文献（ Wang Jinpeng et al., 2017, Plant Physiol.）中的图片对比，基本是一样的效果。 <img src="https://genehub.files.wordpress.com/2019/07/mt4_gm.order_.png" alt="mt4_gm.order"></p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2019/07/31/linux%E6%9B%B4%E6%96%B0git/">Linux更新git</a><a class="next" href="/2019/04/01/%E8%BD%AC%EF%BC%9A%E8%87%AA%E5%8A%A8%E5%8C%96%E8%BF%90%E8%A1%8Corthomcl/">转：自动化运行OrthoMCL</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2020/12/29/Linux%EF%BC%9A%E4%BD%BF%E7%94%A8Modules%E7%AE%A1%E7%90%86%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/">Linux：使用Environment Modules管理环境变量</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E4%BD%BF%E7%94%A8Snakemake%E6%90%AD%E5%BB%BA%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/">使用Snakemake搭建生物信息分析流程</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9%E7%9A%84SAM-BAM%E6%96%87%E4%BB%B6%E4%B8%AD%EF%BC%8C%E6%80%8E%E4%B9%88%E7%90%86%E8%A7%A3MAPQ%E5%80%BC/">基因组比对的SAM/BAM文件中，怎么理解MAPQ值</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/25/Linux%EF%BC%9Avim%E8%AE%BE%E7%BD%AE/">Linux：vim设置</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/Hi-C%EF%BC%9A3d-dna%E6%8C%82%E8%BD%BD%E6%9F%93%E8%89%B2%E4%BD%93/">Hi-C：3d-dna挂载染色体</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/%E4%B8%80%E4%BA%9BFASTA-FASTQ%E5%BA%8F%E5%88%97%E5%A4%84%E7%90%86%E5%B7%A5%E5%85%B7/">一些FASTA/FASTQ序列处理工具</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/12/%E5%A4%A7%E8%B1%86%E4%BA%A7%E4%B8%9A%E7%8E%B0%E7%8A%B6/">大豆产业现状</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/07/1st_post/">My Fast HexoPages</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/07/hello-world/">Hello World</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/06/%E5%AE%89%E8%A3%85hic-pro/">安装HiC-Pro （HiC数据预处理工具）</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2020 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>