<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Bioinfo | LTR identification and analysis</a></h1><div class="post-meta">2021-07-17</div><div class="post-content"><p>在植物基因组中，I 类转座因子，LTR-RT（LTR retrotransposons）是基因组扩张的主要原因。完整的LTR长度在 85 – 5000 bp 之间，下图表示的是一个完整的 LTR-RT，灰色框表示 TSD（target site duplications）, 红色三角形表示 LTR motif（长度在 2 bp 左右），蓝色框表示 LTR。LTR 中间序列长度在 1,000 – 15,000 之间波动。</p></div><p class="readmore"><a href="/2021/07/17/Bioinfo-LTR-identification-and-analysis/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Bioinfo | Gene Ontology - Enrichment Analysis</a></h1><div class="post-meta">2021-06-29</div><div class="post-content"><p>下面是我做的大豆GO数据库（version Gmax_508_Wm82_a4.v1）构建与富集分析。数据文件在：<a target="_blank" rel="noopener" href="https://github.com/bioinfx/Glycine_max_Gene_Ontology">https://github.com/bioinfx/Glycine_max_Gene_Ontology</a>。</p></div><p class="readmore"><a href="/2021/06/29/Bioinfo-Gene-Ontology-Enrichment-Analysis/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/06/17/Linux-Gene-Ontology-GOMAP/">Bioinfo | Gene Ontology - GOMAP</a></h1><div class="post-meta">2021-06-17</div><div class="post-content"><p>Gene Ontology Meta Annotator for Plants</p>
<p>GOMAP 主要是给植物基因集做 Gene Ontology 注释信息匹配，它综合了多个数据库，所以得到的信息会全面一些。早年将基因对应到GO用的工具是Blast2GO，这个软件linux版本安装比较麻烦，基于mysql数据库，而且后面它商业化之后要收费。后来的策略是用diamond的快速blastp将基因集与SwissProt数据库进行比较，得到每个基因最相近的Uniprot ID，再依靠id-mapping文件将基因与GO对应起来，本文最后有简单介绍基于SwissProt做GO注释的流程。</p></div><p class="readmore"><a href="/2021/06/17/Linux-Gene-Ontology-GOMAP/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Linux | 认识 PKG_CONFIG_PATH</a></h1><div class="post-meta">2021-06-08</div><div class="post-content"><p>安装 <a target="_blank" rel="noopener" href="https://github.com/agordon/fastx_toolkit/releases/download/0.0.14/fastx_toolkit-0.0.14.tar.bz2">Fastx-Toolkit</a> 的过程中遇问题：No package ‘gtextutils’ found. Consider adjusting the PKG_CONFIG_PATH environment variable if you.</p></div><p class="readmore"><a href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">jcvi 安装与使用，报错以及解决办法</a></h1><div class="post-meta">2021-06-03</div><div class="post-content"><p>jcvi 是 Haibao Tang 开发的基因组组装注释与比较基因组分析工具，比如可以用于可视化 synteny blocks。原理是依据物种之间基因编码序列相似性比较，如果若干个连续的基因在两个基因组之间都有出现，那这若干个基因就是一个 synteny block，可能是从祖先染色体上继承下来的，物种之间的亲缘关系越近，synteny blocks 数目越多。</p></div><p class="readmore"><a href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">Bioinfo | Installation of OrthoMCL-pipeline</a></h1><div class="post-meta">2021-06-02</div><div class="post-content"><p>OrthoMCL (<a target="_blank" rel="noopener" href="http://orthomcl.org/orthomcl/">http://orthomcl.org/orthomcl/</a>) 是现在用的最多的一款来找直系同源基因（Orthologs）以及旁系同源基因 (Paralog) 的软件。OrthoMCL 原始软件安装比较繁琐，根据官网的教程至少得十多步才能完成整个运行流程，包括Mysql数据库配置、修改OrthoMCL配置文件、转换序列格式、过滤、比对、解析结果和聚类等步骤，特别麻烦。有更方便的途径来完成安装：OrthoMCL Pipeline (<a target="_blank" rel="noopener" href="https://github.com/apetkau/orthomcl-pipeline">https://github.com/apetkau/orthomcl-pipeline</a>)。Pipeline安装有点复杂，但是安装完成后，使用就方便了。</p></div><p class="readmore"><a href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Bioinfo | Using Circos for genome visualization</a></h1><div class="post-meta">2021-05-28</div><div class="post-content"><p>circos 是以环形展示基因组序列和注释信息的工具，我首先按照 circos 官网的指引进行手动安装，即手动逐个安装 Perl 模块，最后再安装 circos 文件，但是卡在 GD 库，没法成功运行 circos。但是换了一台服务器，进入 anaconda3 下面的 python3.7 环境，直接用 conda 不到一分钟就成功安装 circos。</p></div><p class="readmore"><a href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">Bioinfo | Repeat sequence analysis by RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer)</a></h1><div class="post-meta">2021-04-23</div><div class="post-content"><p>学习重复序列的过程中，读到2019年发表在《Mobile DNA》杂志的文献 “Systematic survey of plant LTR-retrotransposons elucidates phylogenetic relationships of their polyprotein domains and provides a reference for element classification”。</p></div><p class="readmore"><a href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></h1><div class="post-meta">2021-04-16</div><div class="post-content"><p>##Crosslinking</p>
<ol>
<li>   Prepare crosslinking buffer (Extraction buffer 1). For each gram of leaves, 36 ml buffer is needed.<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Reagent  Amount (500 mL)</span><br><span class="line">0.4 M sucrose 68.46 g</span><br><span class="line">10 mM Tris-HCl, pH 8.0 5 mL of 1 M stock</span><br><span class="line">5 mM beta-ME  (Add prior to use) 176 uL of 14.2 M stock</span><br><span class="line">0.1 mM PMSF (Add prior to use) 500 uL of 0.1 M stock</span><br><span class="line">1% Formaldehyde (Add prior to use) 13.5 mL of 37% stock</span><br></pre></td></tr></table></figure></li></div><p class="readmore"><a href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></h1><div class="post-meta">2021-04-08</div><div class="post-content"><p>RepeatMasker是重复序列检测的常用工具，通过与参考数据库的相似性比对来准确识别或屏蔽基因组中的重复序列，属于同源预测注释的方式。基因组组装完成后，进行基因预测和注释。由于基因组中存在重复序列结构区，特别是高等真核生物，重复序列占了相当大的比例，会影响基因预测的质量，也会带来不必要的资源消耗。因此在基因预测前，首先要检测并屏蔽基因组中的重复序列（真核生物很必要，原核生物可忽略）。对于这种情况，只是为了屏蔽重复序列降低后续基因预测的压力，重复序列的检测可以不必太纠结。但有时候我们注释基因组的重复序列结构，也可能是专注于某些特定研究，例如，某些重复元件可能参与了重要功能，我们期望定位它们的位置。这种情况下，需要识别精准。</p></div><p class="readmore"><a href="/2021/04/08/Bioinfo-RepeatModeler/">Read More</a></p></div><nav class="page-navigator"><a class="extend prev" rel="prev" href="/">Previous</a><a class="page-number" href="/">1</a><span class="page-number current">2</span><a class="page-number" href="/page/3/">3</a><a class="page-number" href="/page/4/">4</a><span class="space">&hellip;</span><a class="page-number" href="/page/17/">17</a><a class="extend next" rel="next" href="/page/3/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/Bioinfo-Ancestral-Genome-evolution/">Bioinfo | Ancestral Genome evolution</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/31/Bioinfo-MCScanX-genome-synteny-blocks/">Bioinfo | MCScanX genome synteny blocks and 4dtv of paralogous genes</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>