<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>PacBio测序数据格式释义 | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">PacBio测序数据格式释义</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">PacBio测序数据格式释义</h1><div class="post-meta">2016-04-19<span> | </span><span class="category"><a href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></span></div><div class="post-content"><p>本文摘抄/参考自：<a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Data-files-you-received-from-your-service-provider">https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Data-files-you-received-from-your-service-provider</a> 为什么原始数据中有三个“.bas.h5”文件？原因：实际上在PacBio早期的测序原始数据里，只有一个bas.h5文件，因为后来单个SMRT Cell的数据通量有所提高，从而增加了bas.h5文件用以存储base calling数据。 This page describes some of the data files you received from your service provider.</p>
<h2 id="Primary-Analysis-Data"><a href="#Primary-Analysis-Data" class="headerlink" title="Primary Analysis Data"></a><a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Data-files-you-received-from-your-service-provider#primary-analysis-data"></a>Primary Analysis Data</h2><p>This is data <strong>directly</strong> generated by a PacBio RS II run.</p>
<ul>
<li>  The <code>Primary</code> directory includes one subdirectory for <strong>each</strong> run.</li>
<li>  Each run directory includes a subdirectory for <strong>each</strong> SMRT Cell used in the run.</li>
<li>  Each SMRT Cell directory includes an <code>Analysis_Results</code> subdirectory, which contains output files of interest. <strong>Example:</strong></li>
</ul>
<p>/path/to/secondary/storage/2420294/0011<br>├── Analysis_Results<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.1.bax.h5<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.1.log<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.1.subreads.fasta<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.1.subreads.fastq<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.2.bax.h5<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.2.log<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.2.subreads.fasta<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.2.subreads.fastq<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.3.bax.h5<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.3.log<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.3.subreads.fasta<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.3.subreads.fastq<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.bas.h5<br>│   ├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.sts.csv<br>│   └── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.sts.xml<br>├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.1.xfer.xml<br>├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.2.xfer.xml<br>├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.3.xfer.xml<br>├── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.mcd.h5<br>└── m140415_143853_42175_c100635972550000001823121909121417_s1_p0.metadata.xml</p>
<h3 id="PacBio-File-Naming-Convention"><a href="#PacBio-File-Naming-Convention" class="headerlink" title="PacBio File Naming Convention"></a><a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Data-files-you-received-from-your-service-provider#pacbio-file-naming-convention"></a>PacBio File Naming Convention</h3><p>Many files and sequences used in primary and secondary analysis contain a long string identifier. Below is a typical ID string from a subread with fields defined:</p>
<blockquote>
<p> m140415_143853_42175_c100635972550000001823121909121417_s1_p0/553/3100_11230<br>└1┘└────2─────┘ └─3─┘ └────────────────4────────────────┘└5┘└6┘└7┘ └───8────┘</p>
</blockquote>
<ol>
<li> “m” = movie</li>
<li> Time of Run Start (yymmdd_hhmmss)</li>
<li> Instrument Serial Number</li>
<li> SMRT Cell Barcode</li>
<li> Set Number (a.k.a. “Look Number”. Deprecated field, used in earlier version of RS)</li>
<li> Part Number (usually “p0”, “X0” when using expired reagents)</li>
<li> ZMW hole number †</li>
<li> Subread Region (start_stop using polymerase read coordinates) †</li>
</ol>
<blockquote>
<p>† Note that Fields 7 and 8 are used as sequence IDs in FASTAFASTQ files. They are not used in filenames.</p>
</blockquote>
<p>For information on the main files of interest, see:</p>
<ul>
<li>  <a target="_blank" rel="noopener" href="https://s3.amazonaws.com/files.pacb.com/software/instrument/2.0.0/bas.h5+Reference+Guide.pdf">bas.h5 Reference Guide</a> <strong>(PDF)</strong>: Describes the main output files produced by the primary analysis pipeline: <code>bas.h5</code>,<code>.1.bax.h5</code>, <code>.2.bax.h5</code>, and <code>.3.bax.h5</code>. The <code>bax.h5</code> files contain base call information from the sequencing run. The <code>bas.h5</code> file is essentially a pointer to the three <code>bax.h5</code> files.</li>
<li>  <a target="_blank" rel="noopener" href="https://s3.amazonaws.com/files.pacb.com/software/instrument/2.0.0/Metadata+Output+Guide.pdf">Metadata Output Guide</a> <strong>(PDF)</strong>: Describes the file <code>metadata.xml</code>, which contains top-level information about the data, including what sequencing enzyme and chemistry were used, sample name, and other metadata.</li>
<li>  <a target="_blank" rel="noopener" href="https://s3.amazonaws.com/files.pacb.com/software/instrument/1.3.1/Statistics+Output+Guide.pdf">Statistics Output Guide</a> <strong>(PDF)</strong>: Describes the file <code>sts.xml</code>, which includes summary statistics from a single movie acquisition.</li>
</ul>
<h2 id="Secondary-Analysis-Data"><a href="#Secondary-Analysis-Data" class="headerlink" title="Secondary Analysis Data"></a><a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Data-files-you-received-from-your-service-provider#secondary-analysis-data"></a>Secondary Analysis Data</h2><p>This is data produced by secondary analysis, which is performed on the primary analysis data generated by the instrument.</p>
<blockquote>
<ul>
<li><p>  All files for a specific job reside in <strong>one</strong> directory that is named according to the job ID number</p>
</li>
<li><p>Every SMRT Portal job has the following structure. <strong>Example:</strong></p>
<p>  /path/to/smrtanalysis/userdata/jobs/016/016234<br>  ├── data/<br>  ├── results/<br>  ├── log/<br>  ├── workflow/<br>  ├── job.sh<br>  ├── input.xml<br>  └── settings.xml</p>
<ul>
<li>  <code>data</code> is a <strong>directory</strong> that contains intermediate and final data files for the analysis job</li>
<li>  <code>results</code> is a <strong>directory</strong> that contains summary statistics and plots for the analysis job</li>
<li>  <code>log</code> is a <strong>directory</strong> that contains all log files for the analysis job</li>
<li>  <code>workflow</code> is a <strong>directory</strong> that contains all the executables for the analysis job</li>
<li>  <code>job.sh</code> is an executable file used by SMRT Portal to run the <code>smrtpipe.py</code> analysis job</li>
<li>  <code>input.xml</code> is a .xml file containing a list of input <code>bax.h5</code> files used to run the analysis job</li>
<li>  <code>settings.xml</code> is a .xml file containing the parameters needed to perform the analysis job</li>
</ul>
</li>
<li><p>  For more detail on specific protocol outputs, see <a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Navigating-the-SMRT-Pipe-Job-Directory">Navigating the SMRT Pipe Job Directory</a>.</p>
</li>
</ul>
</blockquote>
<p>Within the <code>data</code> directory are several types of output files. You can use these data files as input for further downstream processing, pass on to collaborators, or upload to public genome sites. Depending on the protocol being performed, the <code>data</code> directory contain files in the following formats:</p>
<ul>
<li>  <strong>cmp.h5:</strong> The primary sequence alignment file for SMRT sequencing data. (Click <a target="_blank" rel="noopener" href="https://s3.amazonaws.com/files.pacb.com/software/smrtanalysis/1.4/doc/cmp.h5+Reference+Guide.pdf">here</a> **(PDF)**for further details.)</li>
<li>  <strong>H5:</strong> <code>Hierarchical Data Format</code>; a file-system-like data format. (Click <a target="_blank" rel="noopener" href="http://www.hdfgroup.org/HDF5/doc/H5.intro.html">here</a> for further details.)</li>
<li>  <strong>SAM:</strong> <code>Sequence Alignment Map</code> is a generic nucleotide alignment format that describes the alignment of query sequences or sequencing reads to a reference sequence or assembly. (Click <a target="_blank" rel="noopener" href="http://samtools.sourceforge.net/">here</a> for further details.)</li>
<li>  <strong>BAM:</strong> Binary version of the <code>Sequence Alignment Map (SAM)</code> format. (Click <a target="_blank" rel="noopener" href="http://genome.ucsc.edu/goldenPath/help/bam.html">here</a> for further details.)</li>
<li>  <strong>BAI:</strong> The index file for a file generated in the BAM format. (This is a non-standard file type.)</li>
<li>  <strong>FASTA:</strong> FASTA-formatted sequence files contains either nucleic acid sequence (such as DNA) or protein sequence information. FASTA files store multiple sequences in a single file. (Click <a target="_blank" rel="noopener" href="http://en.wikipedia.org/wiki/FASTA_format">here</a> for further details.)</li>
<li>  <strong>GFF:</strong> <code>General Feature Format</code>, used for describing genes and other features associated with DNA, RNA and Protein sequences. (Click <a target="_blank" rel="noopener" href="http://genome.ucsc.edu/FAQ/FAQformat#format3">here</a> for further details.)</li>
<li>  <strong>VCF:</strong> <code>Variant Call Format</code>, for use with the molecular visualization and analysis program VMD. (Click <a target="_blank" rel="noopener" href="http://en.wikipedia.org/wiki/Variant_Call_Format">here</a> for further details.)</li>
<li>  <strong>BED:</strong> Format that defines the data lines displayed in an annotation track. (Click <a target="_blank" rel="noopener" href="http://genome.ucsc.edu/FAQ/FAQformat#format1">here</a> for further details.)</li>
<li>  <strong>CSV:</strong> Comma-Separated Values file. Can be viewed using Microsoft Excel or a text editor.</li>
<li>  <strong>GML:</strong> An XML representation of the scaffold graph that results from scaffolding contigs using the AHA hybrid assembly algorithm.</li>
</ul>
<h2 id="SMRT-Portal-Reports"><a href="#SMRT-Portal-Reports" class="headerlink" title="SMRT Portal Reports"></a><a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Data-files-you-received-from-your-service-provider#smrt-portal-reports"></a>SMRT Portal Reports</h2><p>Your service provider included secondary analysis reports generated using SMRT Portal.</p>
<ul>
<li>  For an explanation of the report fields, click <a target="_blank" rel="noopener" href="https://s3.amazonaws.com/files.pacb.com/software/smrtanalysis/2.2.0/Reports+-+Terminology.pdf">here</a> <strong>(PDF)</strong>.</li>
</ul>
<h2 id="Downloading-SMRT-Analysis-Software"><a href="#Downloading-SMRT-Analysis-Software" class="headerlink" title="Downloading SMRT Analysis Software"></a><a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/SMRT-Analysis/wiki/Data-files-you-received-from-your-service-provider#downloading-smrt-analysis-software"></a>Downloading SMRT Analysis Software</h2><ul>
<li>  The latest version of the SMRT Analysis software is available <a target="_blank" rel="noopener" href="http://www.pacb.com/devnet/">here</a>.</li>
<li>  Pacific Biosciences provides a free Amazon Machine Image that you can use to run SMRT Portal in the cloud. See <a target="_blank" rel="noopener" href="https://github.com/PacificBiosciences/Bioinformatics-Training/wiki/%22Installing%22-SMRT-Portal-the-easy-way---Launching-A-SMRT-Portal-AMI">here</a> for details.</li>
</ul>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2016/05/26/dogma-chloroplast-genome-annotation/">使用DOGMA对Chloroplast genome注释并用Sequin提交到GenBank</a><a class="next" href="/2016/04/08/%E5%A4%9A%E4%B8%AA%E6%A0%B7%E5%93%81%E7%9A%84denovo-rna-seq%E7%BB%84%E8%A3%85/">多个样品的denovo RNA Seq组装</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/01/01/%E5%9C%A8Linux%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E4%BD%BF%E7%94%A8Nginx%E9%83%A8%E7%BD%B2Hexo%E7%AB%99%E7%82%B9/">在Linux服务器上使用Nginx部署Hexo站点</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/29/Linux%EF%BC%9A%E4%BD%BF%E7%94%A8Modules%E7%AE%A1%E7%90%86%E7%8E%AF%E5%A2%83%E5%8F%98%E9%87%8F/">Linux：使用Environment Modules管理环境变量</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E4%BD%BF%E7%94%A8Snakemake%E6%90%AD%E5%BB%BA%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/">使用Snakemake搭建生物信息分析流程</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/28/%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9%E7%9A%84SAM-BAM%E6%96%87%E4%BB%B6%E4%B8%AD%EF%BC%8C%E6%80%8E%E4%B9%88%E7%90%86%E8%A7%A3MAPQ%E5%80%BC/">基因组比对的SAM/BAM文件中，怎么理解MAPQ值</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/25/Linux%EF%BC%9Avim%E8%AE%BE%E7%BD%AE/">Linux：vim设置</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/Hi-C%EF%BC%9A3d-dna%E6%8C%82%E8%BD%BD%E6%9F%93%E8%89%B2%E4%BD%93/">Hi-C：3d-dna挂载染色体</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/15/%E4%B8%80%E4%BA%9BFASTA-FASTQ%E5%BA%8F%E5%88%97%E5%A4%84%E7%90%86%E5%B7%A5%E5%85%B7/">一些FASTA/FASTQ序列处理工具</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/12/%E5%A4%A7%E8%B1%86%E4%BA%A7%E4%B8%9A%E7%8E%B0%E7%8A%B6/">大豆产业现状</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/07/1st_post/">My Fast HexoPages</a></li><li class="post-list-item"><a class="post-list-link" href="/2020/12/07/hello-world/">Hello World</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>