<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>Proteomics | Top-down proteomics analysis of histones | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">Proteomics | Top-down proteomics analysis of histones</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">Proteomics | Top-down proteomics analysis of histones</h1><div class="post-meta">2021-09-20<span> | </span><span class="category"><a href="/categories/Proteomics/">Proteomics</a></span></div><div class="post-content"><h3 id="1-workflow"><a href="#1-workflow" class="headerlink" title="1 workflow"></a>1 workflow</h3><ol>
<li>crude histone isolation from cells or tissues;</li>
<li>offline HPLC to separete core histones  (H2, H3.1, H3.2, H4)</li>
<li>LC-MS (raw data)</li>
<li>MS data analysis</li>
</ol>
<p>Note: 图见 RGC proposal。</p>
<p><img src="https://ars.els-cdn.com/content/image/1-s2.0-S1046202319301732-gr1.jpg" alt="图例"></p>
<h3 id="2-Offline-HPLC"><a href="#2-Offline-HPLC" class="headerlink" title="2 Offline HPLC"></a>2 Offline HPLC</h3><p>Note：图见 RGC proposal</p>
<p><img src="https://www.researchgate.net/profile/Yifan-Liu-41/publication/235775342/figure/fig2/AS:931921148276736@1599198724085/HPLC-and-SDS-PAGE-separation-of-core-histones_W640.jpg" alt="图例"></p>
<h3 id="3-MS1-chromatograph"><a href="#3-MS1-chromatograph" class="headerlink" title="3 MS1 chromatograph"></a>3 MS1 chromatograph</h3><p>首先要安装 Thermo Fisher 的 Xcalibur 软件。安装方法，<a target="_blank" rel="noopener" href="https://thermo.flexnetoperations.com/control/thmo/login">https://thermo.flexnetoperations.com/control/thmo/login</a> ，使用教育邮箱注册，<a href="mailto:&#120;&#x78;&#64;&#x6c;&#105;&#110;&#107;&#x2e;&#99;&#x75;&#x68;&#x6b;&#x2e;&#101;&#100;&#x75;&#46;&#104;&#x6b;">&#120;&#x78;&#64;&#x6c;&#105;&#110;&#107;&#x2e;&#99;&#x75;&#x68;&#x6b;&#x2e;&#101;&#100;&#x75;&#46;&#104;&#x6b;</a>，注册并登入，即可看到一系列软件的列表，我们选择 Xcalibur <a target="_blank" rel="noopener" href="https://thermo.flexnetoperations.com/control/thmo/product?child_plneID=820487&amp;ver=CURRENT">https://thermo.flexnetoperations.com/control/thmo/product?child_plneID=820487&amp;ver=CURRENT</a>，可执行文件 exe 格式，将文件下载到windows电脑，再安装，安装的软件里面就包含了 freestype。</p>
<p>将 raw data 导入 freestyle 软件，即可浏览 MS1 的图。如何导出高清图片？</p>
<p><img src="" alt="图例"></p>
<h3 id="raw-data-to-pbf-和-ms2ft-格式文件。"><a href="#raw-data-to-pbf-和-ms2ft-格式文件。" class="headerlink" title="raw data to pbf 和 ms2ft 格式文件。"></a>raw data to pbf 和 ms2ft 格式文件。</h3><p>使用 ProMex 软件，可以将 raw 格式的质谱文件转换成 pbf 和 ms2ft 格式文件。</p>
<p>已经有公开的文献报道 top down proteomics 数据的分析方法 <a target="_blank" rel="noopener" href="https://www.nature.com/articles/nmeth.4388">NATURE METHODS: Informed-Proteomics: open-source software package for top-down proteomics</a>，可以去 <a target="_blank" rel="noopener" href="https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics">github</a> 下载 Informed Proteomics，软件处于持续更新中，本笔记撰写的 2021 年仍有更新。 </p>
<p>这些格式有什么不同呢？</p>
<p>raw:<br>pbf: Creates a .pbf file This file contains spectra information and full chromatograms for MS1 and MSn data, allowing fast access to extracted ion chromatograms during the search.<br>ms2ft:</p>
<p>“插播一句，这些格式转换的软件是 windows 平台的，如果要再 linux 下面跨平台运行，则需要用到 mono 软件，具体可以参考 <a target="_blank" rel="noopener" href="https://www.cnblogs.com/jessepeng/p/12452276.html">https://www.cnblogs.com/jessepeng/p/12452276.html</a>。mono 的安装并不容易，可能有些普通用户会碰到问题，主要是依赖libgdiplus库，只要这个解决了，源码安装三部曲就行了，这里不加以赘述。”</p>
<p>转换代码如下（参考：<a target="_blank" rel="noopener" href="https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics">https://github.com/PNNL-Comp-Mass-Spec/Informed-Proteomics</a>）</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">PbfGen.exe -s MyDataset.raw</span><br><span class="line">ProMex.exe -i MyDataset.pbf -minCharge 2 -maxCharge 60 -minMass 3000 -maxMass 50000 -score n -csv n -maxThreads 0</span><br></pre></td></tr></table></figure>

<h3 id="sequence-alignment-and-idntification-of-PTMs"><a href="#sequence-alignment-and-idntification-of-PTMs" class="headerlink" title="sequence alignment and idntification of PTMs"></a>sequence alignment and idntification of PTMs</h3><p>得到pbf 和 ms2ft 文件之后，就可以检索蛋白序列和可能的修饰情况。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">MSPathFinderT.exe -s MyDataset.pbf -feature MyDataset.ms1ft -d C:\FASTA\ProteinList.fasta -o C:\WorkDir -t 10 -f 10 -m 1 -tda 1 -minLength 21 -maxLength 300 -minCharge 2 -maxCharge 30 -minFragCharge 1 -maxFragCharge 15 -minMass 3000 -maxMass 50000 -mod MSPathFinder_Mods.txt</span><br></pre></td></tr></table></figure>

<p>上面的 <code>Mods.txt</code> 文件即是可能的修饰情况，以大豆组蛋白为例，文件内容如下。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"># This file is used to specify modifications for MSPathFinder</span><br><span class="line"></span><br><span class="line"># Max Number of Modifications per peptide</span><br><span class="line">NumMods&#x3D;9</span><br><span class="line"></span><br><span class="line"># Static mods</span><br><span class="line"># None</span><br><span class="line"></span><br><span class="line"># Dynamic mods</span><br><span class="line"># Functional group, Amino acid, opt?, Position, Modification type</span><br><span class="line">O1,MC,opt,any,Oxidation</span><br><span class="line">HO3P,ST,opt,any,Phospho</span><br><span class="line">C2H2O1,K,opt,any,Acetyl</span><br><span class="line">CH2,KR,opt,any,Methyl</span><br><span class="line">C2H4,KR,opt,any,Dimethyl</span><br><span class="line">C3H6,K,opt,any,Trimethyl</span><br><span class="line">C2H2O,*,opt,Prot-N-term,Acetyl     # Acetylation Protein N-term (C2H2O can be replaced with &quot;H(2) C(2) O&quot;)  </span><br></pre></td></tr></table></figure>

<p>运行结果，将得到一个 tsv （表格）文本文件，列出了可能的蛋白序列以及修饰情况，另有一个图片文件 <code>xxx.ms1ft.png</code>，意思？。</p>
<h3 id="图片"><a href="#图片" class="headerlink" title="图片"></a>图片</h3><p>上面的分析中，mods 文件中设定了一条蛋白序列上修饰的数目为 9 个，怎么确定这个数字是合理的？我设定为 10 个，跑了一天也没有结果，所以少于这10 会比较合适。另外需要画一个统计图，即从1到9个修饰的密度图，如果类似正态分布，就可以说明我们的设定是合理的。</p>
<p><img src="https://www.researchgate.net/publication/329379093/figure/fig4/AS:699921351049216@1543885662248/Degree-distribution-All-the-networks-show-a-Poisson-distribution-as-corresponds-to_W640.jpg" alt="图例"></p>
<h3 id="图片-1"><a href="#图片-1" class="headerlink" title="图片"></a>图片</h3><p>各个位点的修饰情况统计。</p>
<p>（D:\博后课题文件\PHD5 project\200831 top-down proteomics in soybean\R代码-冯磊\ptm_freq_diagram\version 2 custom color）</p>
<p><img src="https://ars.els-cdn.com/content/image/1-s2.0-S1046202319301732-gr5.jpg" alt="图例"></p>
<h3 id="Interplay-Score"><a href="#Interplay-Score" class="headerlink" title="Interplay Score"></a>Interplay Score</h3><p>何谓 interplay score：<a target="_blank" rel="noopener" href="https://www.sciencedirect.com/science/article/pii/S1046202319301732">https://www.sciencedirect.com/science/article/pii/S1046202319301732</a><br>何为 co-existence？<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5240223/bin/nihms840427f5.jpg">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5240223/bin/nihms840427f5.jpg</a></p>
<p>两个修饰是否存在coexisence？</p>
<p><img src="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5240223/bin/nihms840427f5.jpg" alt="图例"></p>
<h3 id="样品之间的比较，定量的问题？"><a href="#样品之间的比较，定量的问题？" class="headerlink" title="样品之间的比较，定量的问题？"></a>样品之间的比较，定量的问题？</h3></div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a><a class="next" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/Bioinformatics-Ancestral-Genome-evolution/">Bioinformatics | Ancestral Genome evolution</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/04/01/mcscanx-%E6%B5%8B%E8%AF%95/">MCScanX 测试</a></li><li class="post-list-item"><a class="post-list-link" href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/">Peoteomics | Using MaxQuant to analyze LC-MS/MS data</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/31/Plant-Rice-Oryza-sativa-subsp-japonica-and-indica/">Plant | Rice: Oryza sativa subsp. japonica and indica</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/27/Linux-Meet-I-O-error-when-using-harddisk-in-Linux/">Linux | Meet I/O error when using harddisk in Linux</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/20/Bioinfo-CUT-Tag-sequencing-and-ChIP-sequencing/">Bioinfo | CUT&Tag sequencing and ChIP sequencing</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/11/Bioinfo-Local-BLAST-search-of-NCBI-NR-database/">Bioinfo | Local BLAST search of NCBI NR database</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/10/01/Bioinfo-Positive-Selection-Ka-Ks-calculation/">Bioinfo | Positive Selection Ka/Ks calculation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/20/Proteomics-Top-down-proteomics-analysis-of-histones/">Proteomics | Top-down proteomics analysis of histones</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/09/16/Bioinfo-TCseq-Time-Course-sequencing-data-analysis/">Bioinfo | TCseq Time Course sequencing data analysis</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2022 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>