<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>Tophat: AttributeError: 'NoneType' object has no attribute 'group' | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">Tophat: AttributeError: 'NoneType' object has no attribute 'group'</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">Tophat: AttributeError: 'NoneType' object has no attribute 'group'</h1><div class="post-meta">2014-09-15<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>I have installed boost (version 1.38), Eigen (eigen-eigen-1306d75b4a21.tar.gz) and Samtools(version 1.0). [Ref:<a target="_blank" rel="noopener" href="http://ccb.jhu.edu/software/tophat/tutorial.shtml/]">http://ccb.jhu.edu/software/tophat/tutorial.shtml\]</a> Then I downlaod Tophat 2.0.12 (binary), But error occured when I test the pipeline on a simple test data set. tophat -r 20 test_ref reads_1.fq reads_2.fq</p>
<a id="more"></a>
<p>===============ERROR MESSAGE==========</p>
<p>[fenglei@localhost test_data]$ sh tophat.sh</p>
<p>[2014-09-15 10:49:27] Beginning TopHat run (v2.0.12)</p>
<p>-———————————————-</p>
<p>[2014-09-15 10:49:27] Checking for Bowtie</p>
<p>                  Bowtie version:        2.2.3.0</p>
<p>[2014-09-15 10:49:27] Checking for Samtools</p>
<p>Traceback (most recent call last):</p>
<p>  File “/home/fenglei/local/app/tophat~/tophat”, line 4087, in</p>
<p>    sys.exit(main())</p>
<p>  File “/home/fenglei/local/app/tophat~/tophat”, line 3885, in main</p>
<p>    check_samtools()</p>
<p>  File “/home/fenglei/local/app/tophat~/tophat”, line 1559, in check_samtools</p>
<p>    samtools_version_str, samtools_version_arr = get_samtools_version()</p>
<p>  File “/home/fenglei/local/app/tophat~/tophat”, line 1541, in get_samtools_version</p>
<p>    samtools_version_arr = [int(version_match.group(x)) for x in [1,2,3]]</p>
<p>AttributeError: ‘NoneType’ object has no attribute ‘group’</p>
<p>[fenglei@localhost test_data]$</p>
<p>====================================</p>
<p><strong>The reason: Tophat doesn’t support the most recent version of samtools.</strong></p>
<p>See this other thread: <a target="_blank" rel="noopener" href="https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/tuxedo-tools-users/b7I6YH5_0Ck">https://groups.google.com/forum/?utm_medium=email&amp;utm_source=footer#!topic/tuxedo-tools-users/b7I6YH5_0Ck</a></p>
<p><strong>Thanks to Jeff Hussmann. He had also run into this issue and attached a workaround patch to tophat’s samtools version detection.</strong></p>
<p>===========workaround=================</p>
<p><strong>1537,1542c1537,1543</strong></p>
<p>&lt; version_match = re.search(r’Version:s+(d+).(d+).(d+)([a-zA-Z]?)’, samtools_out) &lt; samtools_version_arr = [int(version_match.group(x)) for x in [1,2,3]] &lt; if version_match.group(4): &lt; samtools_version_arr.append(version_match.group(4)) &lt; else: &lt; samtools_version_arr.append(0) — &gt; version_match = re.search(r’Version:s+(d+).(d+).?(d*)([a-zA-Z]?)’, samtools_out) &gt; samtools_version_arr = [int(version_match.group(x)) for x in [1,2]] &gt; for possibly_blank_index in [3, 4]: &gt; if version_match.group(possibly_blank_index): &gt; samtools_version_arr.append(version_match.group(possibly_blank_index)) &gt; else: &gt; samtools_version_arr.append(0)</p>
<p><strong>1559c1560</strong></p>
<p>&lt; elif samtools_version_arr[1] &lt; 1 or samtools_version_arr[2] &lt; 7: — &gt; elif samtools_version_arr[0] &lt; 1 and (samtools_version_arr[1] &lt; 1 or samtools_version_arr[2] &lt; 7):</p>
<p>======================================</p>
<p>Now the tophat works well:</p>
<p>(15 Sep 2014)</p>
<p>======================================</p>
<p>[fenglei@localhost test_data]$ sh tophat.sh[2014-09-15 11:42:40] Beginning TopHat run (v2.0.12) ———————————————– [2014-09-15 11:42:40] Checking for Bowtie Bowtie version: 2.2.3.0 [2014-09-15 11:42:40] Checking for Samtools Samtools version: 1.0.0.0 [2014-09-15 11:42:40] Checking for Bowtie index files (genome).. Found both Bowtie1 and Bowtie2 indexes. [2014-09-15 11:42:40] Checking for reference FASTA file [2014-09-15 11:42:40] Generating SAM header for test_ref [2014-09-15 11:42:40] Preparing reads left reads: min. length=75, max. length=75, 100 kept reads (0 discarded) right reads: min. length=75, max. length=75, 100 kept reads (0 discarded) [2014-09-15 11:42:40] Mapping left_kept_reads to genome test_ref with Bowtie2 [2014-09-15 11:42:40] Mapping left_kept_reads_seg1 to genome test_ref with Bowtie2 (1/3) [2014-09-15 11:42:40] Mapping left_kept_reads_seg2 to genome test_ref with Bowtie2 (2/3) [2014-09-15 11:42:40] Mapping left_kept_reads_seg3 to genome test_ref with Bowtie2 (3/3) [2014-09-15 11:42:41] Mapping right_kept_reads to genome test_ref with Bowtie2 [2014-09-15 11:42:41] Mapping right_kept_reads_seg1 to genome test_ref with Bowtie2 (1/3) [2014-09-15 11:42:41] Mapping right_kept_reads_seg2 to genome test_ref with Bowtie2 (2/3) [2014-09-15 11:42:41] Mapping right_kept_reads_seg3 to genome test_ref with Bowtie2 (3/3) [2014-09-15 11:42:41] Searching for junctions via segment mapping [2014-09-15 11:42:41] Retrieving sequences for splices [2014-09-15 11:42:41] Indexing splices Building a SMALL index [2014-09-15 11:42:41] Mapping left_kept_reads_seg1 to genome segment_juncs with Bowtie2 (1/3) [2014-09-15 11:42:41] Mapping left_kept_reads_seg2 to genome segment_juncs with Bowtie2 (2/3) [2014-09-15 11:42:41] Mapping left_kept_reads_seg3 to genome segment_juncs with Bowtie2 (3/3) [2014-09-15 11:42:42] Joining segment hits [2014-09-15 11:42:42] Mapping right_kept_reads_seg1 to genome segment_juncs with Bowtie2 (1/3) [2014-09-15 11:42:42] Mapping right_kept_reads_seg2 to genome segment_juncs with Bowtie2 (2/3) [2014-09-15 11:42:42] Mapping right_kept_reads_seg3 to genome segment_juncs with Bowtie2 (3/3) [2014-09-15 11:42:42] Joining segment hits [2014-09-15 11:42:42] Reporting output tracks ———————————————– [2014-09-15 11:42:42] A summary of the alignment counts can be found in ./tophat_out/align_summary.txt [2014-09-15 11:42:42] Run complete: 00:00:02 elapsed [fenglei@localhost test_data]$ </p>
<p>======================================</p>
<p><strong>Reference:</strong></p>
<p>[1] <a target="_blank" rel="noopener" href="https://groups.google.com/forum/#!msg/tuxedo-tools-users/IMJucC83K60/n1--0DGLmxMJ">https://groups.google.com/forum/#!msg/tuxedo-tools-users/IMJucC83K60/n1--0DGLmxMJ</a></p>
<p>[2] <a target="_blank" rel="noopener" href="https://groups.google.com/forum/#!msg/tuxedo-tools-users/IMJucC83K60/lHRE7K3V97QJ">https://groups.google.com/forum/#!msg/tuxedo-tools-users/IMJucC83K60/lHRE7K3V97QJ</a></p>
<p>[3] <a target="_blank" rel="noopener" href="https://groups.google.com/forum/?utm%5C_medium=email&amp;utm%5C_source=footer#!topic/tuxedo-tools-users/BVNDDmInU0A">https://groups.google.com/forum/?utm\_medium=email&amp;utm\_source=footer#!topic/tuxedo-tools-users/BVNDDmInU0A</a></p>
</div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2014/10/14/rna-fasta-dna/">如何将RNA格式的fasta文件转换为DNA格式</a><a class="next" href="/2014/09/12/vim-error-while-loading-shared-libraries-libperl-so/">vim: error while loading shared libraries: libperl.so</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/16/Bioinfo-understanding-sequence-identity/">Bioinfo | 怎么理解序列一致性（Sequence identity）</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/10/Bioinfo-Extracting-sequence-by-names/">Bioinfo | 依据序列名字列表从 fasta 文件提取序列</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/05/Proteomics-Serotonylation-A-novel-histone-H3-marker/">Proteomics | 组蛋白五羟色胺化修饰 H3Q5ser 及其生物学功能</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-What-the-differences-between-DDA-and-DIA-in-mass-spectrometer/">Proteomics | 质谱分析中DDA与DIA有什么不同？</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-Protein-Persulfidation/">Proteomics | 蛋白质硫巯基化修饰</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/Proteomics-Common-reagents-and-their-functions-in-LC-MS-MS-proteomics/">Proteomics | Common reagents and their functions in LC-MS/MS proteomics</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/R-Using-ggrepel-to-repel-overlapping-text-labels-in-ggplot2/">R | Using ggrepel to repel overlapping text labels in ggplot2</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/08/Bioinfo-Installation-of-BUSCO-v4/">Bioinfo | Installation of BUSCO v4</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>