<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="原文记录于 2019-07-05，因 2022-04-01 再次测试，遇到问题，再次编辑记录。 安装如果遇到下面的问题： 1234567[fenglei@localhost MCScanX]$ makeg++ struct.cc mcscan.cc read_data.cc out_utils.cc dagchainer.cc msa.cc permutation.cc -o MCScanXmsa">
<meta property="og:type" content="article">
<meta property="og:title" content="Bioinfo | MCScanX genome synteny blocks and 4dtv of paralogous genes">
<meta property="og:url" content="http://example.com/2022/03/31/Bioinfo-MCScanX-genome-synteny-blocks/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:description" content="原文记录于 2019-07-05，因 2022-04-01 再次测试，遇到问题，再次编辑记录。 安装如果遇到下面的问题： 1234567[fenglei@localhost MCScanX]$ makeg++ struct.cc mcscan.cc read_data.cc out_utils.cc dagchainer.cc msa.cc permutation.cc -o MCScanXmsa">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://genehub.files.wordpress.com/2019/07/gm_mt.test_.png">
<meta property="og:image" content="https://genehub.files.wordpress.com/2019/07/mt4_gm.order_.png">
<meta property="og:image" content="https://i.imgur.com/Ly4FSpT.png">
<meta property="og:image" content="https://i.imgur.com/lRwZIhA.png">
<meta property="og:image" content="https://i.imgur.com/cAzbMcg.png">
<meta property="og:image" content="https://i.imgur.com/6P2ig8r.jpeg">
<meta property="article:published_time" content="2022-03-31T07:32:13.000Z">
<meta property="article:modified_time" content="2022-04-02T15:28:56.273Z">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://genehub.files.wordpress.com/2019/07/gm_mt.test_.png">

<link rel="canonical" href="http://example.com/2022/03/31/Bioinfo-MCScanX-genome-synteny-blocks/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>Bioinfo | MCScanX genome synteny blocks and 4dtv of paralogous genes | FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2022/03/31/Bioinfo-MCScanX-genome-synteny-blocks/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          Bioinfo | MCScanX genome synteny blocks and 4dtv of paralogous genes
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-03-31 15:32:13" itemprop="dateCreated datePublished" datetime="2022-03-31T15:32:13+08:00">2022-03-31</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-04-02 23:28:56" itemprop="dateModified" datetime="2022-04-02T23:28:56+08:00">2022-04-02</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>原文记录于 2019-07-05，因 2022-04-01 再次测试，遇到问题，再次编辑记录。</p>
<p>安装如果遇到下面的问题：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">[fenglei@localhost MCScanX]$ make</span><br><span class="line">g++ struct.cc mcscan.cc read_data.cc out_utils.cc dagchainer.cc msa.cc permutation.cc -o MCScanX</span><br><span class="line">msa.cc: In function ‘void msa_main(const char*)’:</span><br><span class="line">msa.cc:289:22: error: ‘chdir’ was not declared in this scope</span><br><span class="line">if (chdir(html_fn)&lt;0)</span><br><span class="line">^</span><br><span class="line">make: *** [makefile:2: mcscanx] Error 1</span><br></pre></td></tr></table></figure>
<p>解决办法：</p>
<p>错误的原因是 MCScanX 不支持64位系统, 如果要在 64 位上运行, 需要在<code>msa.h</code>, <code>dissect_multiple_alignment.h</code>, and <code>detect_collinear_tandem_arrays.h</code>第一行加上</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">#include &lt;unistd.h&gt;</span><br></pre></td></tr></table></figure>
<p>如果要对两个基因组（以soybean和medicago为例）进行共线性分析，需要收集两个物种的数据编码基因的gff3文件和CDS序列文件，然后进行下面的操作。</p>
<p>首先利用gff3文件，得到每一个基因的起始和终止坐标信息，注意基因有多个transcripts，只需要保留“xxxx.1”的转录本，不同物种的序列用awk等工具灵活处理。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Gmax\_275\_Wm82.a2.v1.gene.gff3.gz -o soybean.bed</span><br><span class="line">python -m jcvi.formats.gff bed --type=mRNA --key=Name Mtruncatula\_285\_Mt4.0v1.gene.gff3 -o medicago.bed</span><br><span class="line"></span><br><span class="line">cat soybean.bed  awk &#x27;BEGIN&#123;OFS=&quot;\\t&quot;&#125;&#123;if(length($4) == 17 &amp;&amp; $4 ~ /\\.1$/)&#123;gsub(/\\.1$/,&quot;&quot;,$4);gsub(/Chr/,&quot;gm&quot;,$1); print $1,$4,$2,$3&#125;&#125;&#x27; &gt; soybean\_gff.tab</span><br><span class="line">cat medicago.bed  awk &#x27;BEGIN&#123;OFS=&quot;\\t&quot;&#125;&#123;if(length($4) == 15 &amp;&amp; $4 ~ /\\.1$/)&#123;gsub(/\\.1/,&quot;&quot;,$4);gsub(/chr/,&quot;mt&quot;,$1); print $1,$4,$2,$3&#125;&#125;&#x27; &gt; medicago\_gff.tab</span><br></pre></td></tr></table></figure>
<p>得到gff格式文件如下。 soybean.gff</p>
<p><strong>注意</strong> gff 文件中别包含 contigs 和 scaffolds，最好只包含 chromosomes。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">gm01 Glyma.01G000100 27354 28320</span><br><span class="line">gm01 Glyma.01G000200 58974 67527</span><br><span class="line">gm01 Glyma.01G000300 67769 69968</span><br><span class="line">gm01 Glyma.01G000400 90151 95947</span><br><span class="line">gm01 Glyma.01G000500 90288 91197</span><br><span class="line">gm01 Glyma.01G000600 116093 127845</span><br><span class="line">gm01 Glyma.01G000700 143466 155573</span><br><span class="line">gm01 Glyma.01G000800 157029 157772</span><br><span class="line">gm01 Glyma.01G000900 170533 193342</span><br><span class="line">gm01 Glyma.01G001000 196255 201895</span><br></pre></td></tr></table></figure>
<p>medicago.gff</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">mt1 Medtr1g004930 688 7332</span><br><span class="line">mt1 Medtr1g004940 6523 7366</span><br><span class="line">mt1 Medtr1g004950 14513 15729</span><br><span class="line">mt1 Medtr1g004960 16282 18382</span><br><span class="line">mt1 Medtr1g006530 28972 29214</span><br><span class="line">mt1 Medtr1g004980 31972 32344</span><br><span class="line">mt1 Medtr1g004990 35909 40554</span><br><span class="line">mt1 Medtr1g006490 44429 46280</span><br><span class="line">mt1 Medtr1g006590 46547 50437</span><br><span class="line">mt1 Medtr1g006600 51360 54977</span><br></pre></td></tr></table></figure>
<p>随后对 CDS 序列文件进行整理，保留 primary transcripts 的序列。序列名称跟上面 gff 文件里面要一致。并灵活选用工具将 CDS 序列转换成蛋白序列，或者一开始就用蛋白序列也可以。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">awk &#x27;&#123;print $1&#125;&#x27; ../jcvi\_190704/Gmax\_275\_Wm82.a2.v1.cds.fa &gt; Gmax\_275\_Wm82.a2.v1.cds.fa\_simplified</span><br><span class="line">awk &#x27;&#123;print $1&#125;&#x27; ../jcvi\_190704/Mtruncatula\_285\_Mt4.0v1.cds.fa &gt; Mtruncatula\_285\_Mt4.0v1.cds.fa\_simplified</span><br><span class="line"></span><br><span class="line">fasta\_formatter -t -i Mtruncatula\_285\_Mt4.0v1.cds.fa\_simplified -o Mtruncatula\_285\_Mt4.0v1.cds.tab</span><br><span class="line">fasta\_formatter -t -i Gmax\_275\_Wm82.a2.v1.cds.fa\_simplified -o Gmax\_275\_Wm82.a2.v1.cds.tab</span><br><span class="line"></span><br><span class="line">cat Gmax\_275\_Wm82.a2.v1.cds.tab  awk &#x27;BEGIN&#123;OFS=&quot;\\t&quot;&#125;&#123;if(length($1) == 17 &amp;&amp; $1 ~ /\\.1$/)&#123;gsub(/\\.1$/,&quot;&quot;,$1); print &quot;\\&gt;&quot;$1&quot;\\n&quot;$2&#125;&#125;&#x27; &gt; soybean\_cds.fa</span><br><span class="line">cat Mtruncatula\_285\_Mt4.0v1.cds.tab  awk &#x27;BEGIN&#123;OFS=&quot;\\t&quot;&#125;&#123;if(length($1) == 15 &amp;&amp; $1 ~ /\\.1$/)&#123;gsub(/\\.1/,&quot;&quot;,$1); print &quot;\\&gt;&quot;$1&quot;\\n&quot;$2&#125;&#125;&#x27; &gt; medicago.cds.fa</span><br><span class="line">iTools Fatools CDS2Pep -InCDS soybean\_cds.fa -OutPut soybean\_pep.fa</span><br><span class="line">iTools Fatools CDS2Pep -InCDS medicago.cds.fa -OutPut medicago.pep.fa</span><br></pre></td></tr></table></figure>
<p>然后用 blastp 对两个物种的蛋白序列进行比对，并用 MCScanX 命令检测共线性区间。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">makeblastdb -in medicago.pep.fa -dbtype prot -out index/mt -parse\_seqids</span><br><span class="line">## diamond makedb --in medicago.pep.fa --db ./diamond\_index/mt</span><br><span class="line">## diamond blastp --query soybean\_pep.fa --evalue 1e-5 --outfmt 6 --threads 24 --db ./diamond\_index/mt.dmnd --out soybean\_medicago.blast</span><br><span class="line">blastp -query soybean\_pep.fa -db index/mt -out gm\_mt.blast -evalue 1e-5 -num\_threads 24 -outfmt 6 -num\_alignments 5</span><br><span class="line">mkdir mcdata</span><br><span class="line">cp gm\_mt.blast mcdata</span><br><span class="line">cat soybean.gff medicago.gff &gt; mcdata/gm\_mt.gff</span><br><span class="line">MCScanX ./mcdata/gm\_mt</span><br></pre></td></tr></table></figure>
<p>产生的线性文件（gm_mt.collinearity）如下，可以用于下游画图。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">\[fenglei@localhost gm\_mt\_test\]$ head -n 50 gm\_mt.collinearity </span><br><span class="line">############### Parameters ###############</span><br><span class="line"># MATCH\_SCORE: 50</span><br><span class="line"># MATCH\_SIZE: 5</span><br><span class="line"># GAP\_PENALTY: -1</span><br><span class="line"># OVERLAP\_WINDOW: 5</span><br><span class="line"># E\_VALUE: 1e-05</span><br><span class="line"># MAX GAPS: 25</span><br><span class="line">############### Statistics ###############</span><br><span class="line"># Number of collinear genes: 51572, Percentage: 49.62</span><br><span class="line"># Number of all genes: 103927</span><br><span class="line">##########################################</span><br><span class="line">## Alignment 0: score=1008.0 e\_value=1.4e-69 N=23 gm01&amp;mt1 plus</span><br><span class="line">0- 0: Glyma01G232800 Medtr1g103160 0</span><br><span class="line">0- 1: Glyma01G234900 Medtr1g103450 8e-39</span><br><span class="line">0- 2: Glyma01G235100 Medtr1g103490 0.0001</span><br><span class="line">0- 3: Glyma01G236800 Medtr1g103500 2e-87</span><br><span class="line">0- 4: Glyma01G237000 Medtr1g103550 9e-47</span><br><span class="line">0- 5: Glyma01G237300 Medtr1g103570 2e-45</span><br><span class="line">0- 6: Glyma01G237400 Medtr1g103600 1e-144</span><br><span class="line">0- 7: Glyma01G237900 Medtr1g103690 9e-39</span><br><span class="line">0- 8: Glyma01G238200 Medtr1g103830 0</span><br><span class="line">0- 9: Glyma01G238400 Medtr1g104500 1e-170</span><br><span class="line">0- 10: Glyma01G238500 Medtr1g104520 6e-33</span><br><span class="line">0- 11: Glyma01G238700 Medtr1g104680 0</span><br><span class="line">0- 12: Glyma01G238800 Medtr1g104750 0</span><br><span class="line">0- 13: Glyma01G238900 Medtr1g104800 2e-80</span><br><span class="line">0- 14: Glyma01G239200 Medtr1g104870 0</span><br><span class="line">0- 15: Glyma01G239400 Medtr1g104930 1e-50</span><br><span class="line">0- 16: Glyma01G239800 Medtr1g105075 1e-32</span><br><span class="line">0- 17: Glyma01G239900 Medtr1g105305 5e-101</span><br><span class="line">0- 18: Glyma01G240000 Medtr1g105415 3e-91</span><br><span class="line">0- 19: Glyma01G240500 Medtr1g105555 5e-62</span><br><span class="line">0- 20: Glyma01G240600 Medtr1g105570 6e-06</span><br><span class="line">0- 21: Glyma01G240800 Medtr1g105650 2e-175</span><br><span class="line">0- 22: Glyma01G240900 Medtr1g105655 4e-164</span><br><span class="line">## Alignment 1: score=741.0 e\_value=1e-39 N=16 gm01&amp;mt1 plus</span><br><span class="line">1- 0: Glyma01G084800 Medtr1g093020 0</span><br><span class="line">1- 1: Glyma01G086700 Medtr1g093080 1e-138</span><br><span class="line">1- 2: Glyma01G086800 Medtr1g093095 2e-108</span><br><span class="line">1- 3: Glyma01G087500 Medtr1g093600 5e-175</span><br><span class="line">1- 4: Glyma01G087700 Medtr1g093630 0</span><br><span class="line">1- 5: Glyma01G088000 Medtr1g093650 8e-07</span><br><span class="line">1- 6: Glyma01G088200 Medtr1g093670 9e-115</span><br><span class="line">1- 7: Glyma01G088300 Medtr1g093750 1e-118</span><br><span class="line">1- 8: Glyma01G089100 Medtr1g093770 0</span><br><span class="line">1- 9: Glyma01G089300 Medtr1g093850 0</span><br><span class="line">1- 10: Glyma01G089900 Medtr1g093860 0</span><br><span class="line">1- 11: Glyma01G090200 Medtr1g093900 0</span><br><span class="line">1- 12: Glyma01G090400 Medtr1g093920 9e-22</span><br><span class="line">1- 13: Glyma01G091000 Medtr1g094010 0</span><br></pre></td></tr></table></figure>
<p>注意画图的配置文件不要出错，双斜杠（//）表示注释，注释号前面用tab符号跟染色体名称分隔，或者干脆去掉注释，我删掉注释才顺利画图，否则报错。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">[fenglei@localhost gm\_mt\_test\]$ cat dot.ctl</span><br><span class="line">2000</span><br><span class="line">2000</span><br><span class="line">gm01,gm02,gm03,gm04,gm05,gm06,gm07,gm08,gm09,gm10,gm11,gm12,gm13,gm14,gm15,gm16,gm17,gm18,gm19,gm20</span><br><span class="line">mt1,mt2,mt3,mt4,mt5,mt6,mt7,mt8</span><br></pre></td></tr></table></figure>
<p>画图命令。注意 dot_plotter.java 与 dot_plotter .class 要在一个文件夹下面。其他环状图做法也类似。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cd /PATH_TO_MCScanX/downstream_analyses</span><br><span class="line">DATA_PATH=/xx/my/mcdata  # 根据实际情况</span><br><span class="line">java dot_plotter -g $DATA_PATH/gm_mt.gff -s $DATA_PATH/gm_mt.collinearity -c $DATA_PATH/dot.ctl -o gm_mt.test.png</span><br></pre></td></tr></table></figure>
<p>画图结果：</p>
<p><img src="https://genehub.files.wordpress.com/2019/07/gm_mt.test_.png" alt="gm_mt.test" /></p>
<p>与文献（ Wang Jinpeng et al., 2017, Plant Physiol.）中的图片对比，基本是一样的效果。</p>
<p><img src="https://genehub.files.wordpress.com/2019/07/mt4_gm.order_.png" alt="mt4_gm.order" /></p>
<p>这个 MCScanX 共线性图跟更新开发的 jcvi 相比，画质要差一些。下面是 jcvi 的结果。</p>
<p><img src="https://i.imgur.com/Ly4FSpT.png" alt="jcvi 绘制的基因组共线性图实例" /></p>
<p>MCScanX 还可以将图画成圆形。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java circle_plotter -g /home/fenglei/projects/genome_synteny_20190704/MCScanX_soybean_medicago_190705/gm_mt_test/MADS-master/MADS.gff -s /home/fenglei/projects/genome_synteny_20190704/MCScanX_soybean_medicago_190705/gm_mt_test/MADS-master/MADS.collinearity -c /home/fenglei/projects/genome_synteny_20190704/MCScanX_soybean_medicago_190705/gm_mt_test/MADS-master/circle.ctl -o MADS.png</span><br></pre></td></tr></table></figure>
<p><strong>注意</strong> circle_plotter 的 dot.ctl 文件与 dot_plotter 的 dot.ctl 文件不一样。</p>
<p>For circle_plotter , put chromosome IDs in one line in dot.ctl file.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">2000</span><br><span class="line">Gm01,Gm02,Gm03,Gs01,Gs02,Gs03</span><br></pre></td></tr></table></figure>
<p>For dot_plotter, put chromosome IDs in two lines in dot.ctl file</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">2000</span><br><span class="line">2000</span><br><span class="line">Gm01,Gm02,Gm03</span><br><span class="line">Gs01,Gs02,Gs03</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ cat /home/fenglei/projects/genome_synteny_20190704/MCScanX_soybean_medicago_190705/gm_mt_test/MADS-master/circle.ctl</span><br><span class="line">1600    //plot width and height (in pixels)</span><br><span class="line">Chr1,Chr2,Chr3,Chr4,Chr5,Chr6,Chr7,Chr8,Chr9,Chr10,Chr11,Chr12,Chr13,Chr14,Chr15,Chr16,Chr17    //chromosomes in the circle</span><br></pre></td></tr></table></figure>
<p><img src="https://i.imgur.com/lRwZIhA.png" alt="" /></p>
<h3 id="mcscanx-与-jcvi-的比较"><a class="markdownIt-Anchor" href="#mcscanx-与-jcvi-的比较"></a> MCScanX 与 jcvi 的比较</h3>
<p>在基因组进化分析的 paralogous genes 鉴定中，MCScanX 与 jcvi 都可以用，有什么区别？从下面两幅图来看，jcvi 绘制的图片更有观赏性，而且 jcvi 能识别出这是一个物种内部（intra-genomic comparison）的 paralogous genes。</p>
<p>MCScanX：2019年基于a2v1的基因组数据。一共 1112 synteny blocks，35094 个两两配对。Number of collinear genes: 37190, Percentage: 66.90，Number of all genes: 55589。</p>
<p><img src="https://i.imgur.com/cAzbMcg.png" alt="MCScanX 分析 soybean 基因组内部的 synteny blocks" /></p>
<p>jcvi：2021年基于a4v1的基因组数据。一共 472 synteny blocks，18439个两两配对。</p>
<p><img src="https://i.imgur.com/6P2ig8r.jpeg" alt="jcvi 分析 soybean 基因组内部的 synteny blocks" /></p>
<p>看来 MCScanX 的结果更全面？以基因 <code>Glyma.01G210400</code> 为例，通过 MCScanX 可以找到另外三个旁系同源基因（paralogous），而通过 jcvi 只找到一个旁系同源基因。难怪在 2021 年的时候测试 4dtv 计算的时候，我发现 MCScanX 的结果更适合做 4dtv 计算，原因是它鉴定的 paralogous 更合理。当然，也有可能是 jcvi 的参数设计原因，我用的默认参数，没有进行额外设置。</p>
<p>MCScanX：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">(base) [fenglei@localhost mcdata]$ cat soybean.collinearity | grep &#x27;Glyma.01G210400&#x27;</span><br><span class="line"> 28- 43:        Glyma.01G210400 Glyma.05G058100       0</span><br><span class="line"> 71-432:        Glyma.01G210400 Glyma.11G031600   5e-82</span><br><span class="line"> 88-129:        Glyma.01G210400 Glyma.17G140600       0</span><br></pre></td></tr></table></figure>
<p>jcvi：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">(base) [fenglei@localhost gmo-gmo]$ cat gma.gma.anchors | grep &#x27;Glyma.01G210400&#x27;</span><br><span class="line">Glyma.01G210400.2       Glyma.17G140600.1       1030</span><br></pre></td></tr></table></figure>
<p><em>END</em></p>

    </div>

    
    
    

      <footer class="post-footer">

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/03/09/Peoteomics-Using-MaxQuant-to-analyze-LC-MS-MS-data/" rel="prev" title="Peoteomics | Using MaxQuant to analyze LC-MS/MS data">
      <i class="fa fa-chevron-left"></i> Peoteomics | Using MaxQuant to analyze LC-MS/MS data
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/04/20/Bioinfo-DRIMM-synteny-lab-notes/" rel="next" title="Bioinfo | DRIMM synteny and ancestor genomes">
      Bioinfo | DRIMM synteny and ancestor genomes <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-3"><a class="nav-link" href="#mcscanx-%E4%B8%8E-jcvi-%E7%9A%84%E6%AF%94%E8%BE%83"><span class="nav-number">1.</span> <span class="nav-text"> MCScanX 与 jcvi 的比较</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">173</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">10</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
