<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>Bioinfo | Installation of OrthoMCL-pipeline | FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">Bioinfo | Installation of OrthoMCL-pipeline</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title">Bioinfo | Installation of OrthoMCL-pipeline</h1><div class="post-meta">2021-06-02<span> | </span><span class="category"><a href="/categories/Bioinformatics/">Bioinformatics</a></span></div><div class="post-content"><p>OrthoMCL (<a target="_blank" rel="noopener" href="http://orthomcl.org/orthomcl/">http://orthomcl.org/orthomcl/</a>) 是现在用的最多的一款来找直系同源基因（Orthologs）以及旁系同源基因 (Paralog) 的软件。OrthoMCL 原始软件安装比较繁琐，根据官网的教程至少得十多步才能完成整个运行流程，包括Mysql数据库配置、修改OrthoMCL配置文件、转换序列格式、过滤、比对、解析结果和聚类等步骤，特别麻烦。有更方便的途径来完成安装：OrthoMCL Pipeline (<a target="_blank" rel="noopener" href="https://github.com/apetkau/orthomcl-pipeline">https://github.com/apetkau/orthomcl-pipeline</a>)。Pipeline安装有点复杂，但是安装完成后，使用就方便了。</p>
<p>两年多前使用过 OrthoMCL-pipeline 流程来寻找一个物种内部的同源基因（paralogs）和物种间的同源基因（othologs）。物种内部的同源基因之间的碱基差别可能是若干年演化的结果，比如染色体倍增或基因扩张，导致出现同源基因，随后由于演化导致某些突变，尤其是同义替换（Ks 或称为 synonymous）不会引起氨基酸序列变化，常作为分子钟，也比如基因组分析种常见的 4dtv（四倍简并位点），也是用需要依靠同源基因的序列来做分析。</p>
<p>有哪些文章的示例？可以便于记忆与理解。</p>
<p><img src="https://i.imgur.com/aDN8Pcn.jpg" alt="Fig. 1"></p>
<p>上图是 2011 年发表于 Nature 期刊的土豆基因组（<a target="_blank" rel="noopener" href="https://www.nature.com/articles/nature10158">https://www.nature.com/articles/nature10158</a>），使用 OrthoMCL 分析 12 个植物之间的直系同源基因和旁系同源基因，统计数目并制作了条形图。</p>
<p><img src="https://i.imgur.com/XpDMtM8.jpg" alt="Fig. 2"></p>
<p>上图是 2018 年发表于 PBJ 期刊的西葫芦基因组（<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5978595/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5978595/</a>）。使用 OrthoMCL 分析五个植物的同源基因，并且基于同源基因信息计算四倍简并位点（4tdv），推算基因组是否加倍。</p>
<h3 id="安装-OrthoMLC-pipeline"><a href="#安装-OrthoMLC-pipeline" class="headerlink" title="安装 OrthoMLC-pipeline"></a>安装 OrthoMLC-pipeline</h3><p>先下载 <strong>OrthoMCL-pipeline</strong>。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https:&#x2F;&#x2F;github.com&#x2F;apetkau&#x2F;orthomcl-pipeline.git</span><br></pre></td></tr></table></figure>

<p>再安装 <strong>Perl 以及若干相关模块</strong>。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cpanm BioPerl DBD::mysql DBI Parallel::ForkManager YAML::Tiny Set::Scalar Text::Table Exception::Class Test::Most Test::Warn Test::Exception Test::Deep Moose SVG Algorithm::Combinatorics</span><br></pre></td></tr></table></figure>

<p>实际上在安装 Perl 模块的过程中遇到问题，主要是 BioPerl 安装失败，报错信息：<code>Installing the dependencies failed: Module &#39;Alien::Libxml2&#39; is not installed</code>。并且我发现不能通过 cpanm 来安装 Alien::Libxml2。经过检索，在一个论坛（<a target="_blank" rel="noopener" href="https://github.com/PerlAlien/Alien-Libxml2/issues/7">https://github.com/PerlAlien/Alien-Libxml2/issues/7</a>）找到方法。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">$ which perl</span><br><span class="line"> ~&#x2F;local&#x2F;bin&#x2F;perl</span><br><span class="line"></span><br><span class="line">$ perl --version</span><br><span class="line"> This is perl 5, version 30, subversion 3 (v5.30.3) built for x86_64-linux</span><br><span class="line"></span><br><span class="line">$ curl -O https:&#x2F;&#x2F;cpan.metacpan.org&#x2F;authors&#x2F;id&#x2F;P&#x2F;PL&#x2F;PLICEASE&#x2F;Alien-Build-1.71_01.tar.gz</span><br><span class="line">  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current</span><br><span class="line">                                 Dload  Upload   Total   Spent    Left  Speed</span><br><span class="line">100  257k  100  257k    0     0  1764k      0 --:--:-- --:--:-- --:--:-- 1775k</span><br><span class="line">$ tar xf Alien-Build-1.71_01.tar.gz </span><br><span class="line">$ cd Alien-Build-1.71_01</span><br><span class="line">$ perl Makefile.PL </span><br><span class="line">Checking if your kit is complete...</span><br><span class="line">Looks good</span><br><span class="line">Generating a Unix-style Makefile</span><br><span class="line">Writing Makefile for Alien::Build</span><br><span class="line">Writing MYMETA.yml and MYMETA.json</span><br><span class="line">...</span><br><span class="line">$ make</span><br><span class="line">$ make install</span><br><span class="line"># 最后这样就安装成功了</span><br><span class="line"># Appending installation info to &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;lib&#x2F;perl5&#x2F;5.30.3&#x2F;x86_64-linux&#x2F;perllocal.pod</span><br></pre></td></tr></table></figure>

<p><strong><a target="_blank" rel="noopener" href="http://orthomcl.org/common/downloads/software/v2.0/">OrthoMCL</a></strong> 或 OrthoMCL Custom （能够自定义序列识别符的修改版本）。OrhtoMCL 下载并解压之后，bin目录下的Perl程序首行是 <code>#!/usr/bin/perl</code>，我将其全部改为用户的 perl 了，因为我自己安装的perl位置是 <code>/home/fenglei/local/bin/perl</code>。</p>
<p><strong><a target="_blank" rel="noopener" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download">BLAST</a></strong> (blastall, formatdb) 注意不是NCBI-blast+, 推荐使用 2.2.26 版本。由于我之前下载过 interproscan，在 interproscan 安装目录下游 blastall 和 formatdb 等程序，直接将其路径加入环境变量即可！</p>
<p><strong><a target="_blank" rel="noopener" href="https://www.micans.org/mcl/src/">MCL</a></strong> The MCL algorithm is short for the Markov Cluster Algorithm, a fast and scalable unsupervised cluster algorithm for graphs (also known as networks) based on simulation of (stochastic) flow in graphs.</p>
<p>依赖环境解决后能够顺利运行设置脚本:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">$ perl scripts&#x2F;orthomcl-pipeline-setup.pl</span><br><span class="line">Checking for Software dependencies...</span><br><span class="line">Checking for OthoMCL ... OK</span><br><span class="line">Checking for formatdb ... OK</span><br><span class="line">Checking for blastall ... OK</span><br><span class="line">Checking for mcl ... OK</span><br><span class="line">Wrote new configuration to orthomcl-pipeline&#x2F;scripts&#x2F;..&#x2F;etc&#x2F;orthomcl-pipeline.conf</span><br><span class="line">Wrote executable file to orthomcl-pipeline&#x2F;scripts&#x2F;..&#x2F;bin&#x2F;orthomcl-pipeline</span><br><span class="line">Please add directory orthomcl-pipeline&#x2F;scripts&#x2F;..&#x2F;bin to PATH</span><br></pre></td></tr></table></figure>

<p>然而，我重复以前的流程失败，还是卡在 BioPerl 安装失败。看来问题跟我使用 Anaconda3 有关，下面的错误提示信息种出现的 make 是来自 python37 环境。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">make: *** [Makefile:3290: test_dynamic] Error 255</span><br><span class="line">  CJFIELDS&#x2F;BioPerl-1.7.8.tar.gz</span><br><span class="line">  &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;bin&#x2F;make test -- NOT OK</span><br><span class="line">&#x2F;&#x2F;hint&#x2F;&#x2F; to see the cpan-testers results for installing this module, try:</span><br><span class="line">  reports CJFIELDS&#x2F;BioPerl-1.7.8.tar.gz</span><br><span class="line">Failed during this command:</span><br><span class="line"> (optional) CJFIELDS&#x2F;BioPerl-1.7.8.tar.gz     : make_test NO</span><br></pre></td></tr></table></figure>
<h3 id="20210615-再次尝试安装测试"><a href="#20210615-再次尝试安装测试" class="headerlink" title="20210615 再次尝试安装测试"></a>20210615 再次尝试安装测试</h3><p>由于系统自带了 Perl（/usr/bin/perl），我安装的Anaconda3下面的不同环境也分别有 Perl（<del>/local/app/anaconda3/envs/python37/bin/perl、</del>/local/app/anaconda3/envs/python27/bin/perl），另外自己又在用户目录下安装了 Perl（~/local/bin/perl），多个版本导致冲突。</p>
<p>软件测试：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">source activate python37</span><br><span class="line">PERL5LIB&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0  </span><br><span class="line"># 上面其实应该设为 &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;site_perl&#x2F;5.26.2&#x2F;</span><br><span class="line"># 此时直接用 cpan -i 或者 perl -MCPAN -s shell 均无法安装 perl 模块，于是尝试 conda，conda基本可以安装这些模块</span><br><span class="line"> conda install -c bioconda perl-Moo</span><br><span class="line"> conda install -c bioconda perl-DBD::mysql # ERROR: Wrong command</span><br><span class="line"> conda install -c bioconda perl-DBD-mysql</span><br><span class="line"> conda install -c bioconda perl-Parallel-ForkManager</span><br><span class="line"> conda install -c bioconda perl-YAML-Tiny</span><br><span class="line"> conda install -c bioconda perl-Set-Scalar</span><br><span class="line"> conda install -c bioconda perl-Text-Table  # ERROR: PackagesNotFoundError</span><br><span class="line"> conda install -c bioconda perl-Exception-Class</span><br><span class="line"> conda install -c bioconda perl-Test-Most</span><br><span class="line"> conda install -c bioconda perl-Test-Warn</span><br><span class="line"> conda install -c bioconda perl-Test-Exception</span><br><span class="line"> conda install -c bioconda perl-Test-Deep</span><br><span class="line"> conda install -c bioconda perl-Moose</span><br><span class="line"> conda install -c bioconda perl-SVG</span><br><span class="line"> conda install -c bioconda perl-Algorithm-Combinatorics  # ERROR: PackagesNotFoundError</span><br></pre></td></tr></table></figure>

<p>conda 安装perl模块有两个安不上，所以手动尝试</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">wget https:&#x2F;&#x2F;cpan.metacpan.org&#x2F;authors&#x2F;id&#x2F;S&#x2F;SH&#x2F;SHLOMIF&#x2F;Text-Aligner-0.16.tar.gz</span><br><span class="line">perl Makefile.PL &amp;&amp; make &amp;&amp; make install</span><br><span class="line">## Installing &#x2F;home&#x2F;fenglei&#x2F;perl5&#x2F;man&#x2F;man3&#x2F;Text::Aligner.3</span><br><span class="line">## Appending installation info to &#x2F;home&#x2F;fenglei&#x2F;perl5&#x2F;lib&#x2F;perl5&#x2F;x86_64-linux-thread-multi&#x2F;perllocal.pod</span><br><span class="line"></span><br><span class="line">conda install -c bioconda  perl-Module-Build</span><br><span class="line"></span><br><span class="line">wget https:&#x2F;&#x2F;cpan.metacpan.org&#x2F;authors&#x2F;id&#x2F;S&#x2F;SH&#x2F;SHLOMIF&#x2F;Text-Table-1.134.tar.gz</span><br><span class="line">perl Makefile.PL &amp;&amp; make &amp;&amp; make install</span><br><span class="line">## Installing &#x2F;home&#x2F;fenglei&#x2F;perl5&#x2F;man&#x2F;man3&#x2F;Text::Table.3</span><br><span class="line">## Appending installation info to &#x2F;home&#x2F;fenglei&#x2F;perl5&#x2F;lib&#x2F;perl5&#x2F;x86_64-linux-thread-multi&#x2F;perllocal.pod</span><br><span class="line"></span><br><span class="line">## 安装上述模块之后， perl lib 路径下根本没有出现pm文件，我手动拷贝过去，为了保险，拷贝到两个位置：</span><br><span class="line">cp Aligner.pm &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0&#x2F;x86_64-linux-thread-multi&#x2F;auto&#x2F;Text&#x2F;</span><br><span class="line">cp Aligner.pm &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0&#x2F;Text&#x2F;</span><br><span class="line">cp Table.pm &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0&#x2F;x86_64-linux-thread-multi&#x2F;auto&#x2F;Text&#x2F;</span><br><span class="line">cp Table.pm &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0&#x2F;Text&#x2F;</span><br><span class="line">PERL5LIB&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;site_perl&#x2F;5.26.2&#x2F;x86_64-linux-thread-multi&#x2F;auto&#x2F;:$PERL5LIB</span><br><span class="line"></span><br><span class="line">wget https:&#x2F;&#x2F;cpan.metacpan.org&#x2F;authors&#x2F;id&#x2F;F&#x2F;FX&#x2F;FXN&#x2F;Algorithm-Combinatorics-0.27.tar.gz</span><br><span class="line">## make 失败</span><br><span class="line">## 报错：unable to execute &#39;x86_64-conda_cos6-linux-gnu-gcc&#39;: No such file or directory  </span><br></pre></td></tr></table></figure>

<p>这次卡在了 Algorithm-Combinatorics。</p>
<p>在base环境下，自己安装的perl可以编译Algorithm-Combinatorics。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">source activate base</span><br><span class="line">which perl</span><br><span class="line"># ~&#x2F;local&#x2F;bin&#x2F;perl</span><br><span class="line">PERL5LIB&#x3D;&#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0  </span><br><span class="line">cp Combinatorics.pm &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0&#x2F;x86_64-linux-thread-multi&#x2F;auto&#x2F;</span><br><span class="line">cp Combinatorics.pm &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0&#x2F;x86_64-linux-thread-multi&#x2F;auto&#x2F;Algorithm&#x2F;</span><br><span class="line">cp Combinatorics.pm &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;site_perl&#x2F;5.26.2&#x2F;Algorithm&#x2F;</span><br></pre></td></tr></table></figure>

<p>更改前后，都有报错，看来直接复制Combinatorics.pm也不行。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">(python37) [fenglei@localhost orthomcl-pipeline]$ perl .&#x2F;scripts&#x2F;orthomcl-pipeline-setup.pl</span><br><span class="line">Can&#39;t locate Algorithm&#x2F;Combinatorics.pm in @INC (you may need to install the Algorithm::Combinatorics module) (@INC contains: &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;site_perl&#x2F;5.26.2&#x2F;x86_64-linux-thread-multi&#x2F;auto&#x2F; &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0&#x2F;x86_64-linux-thread-multi &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0 &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;site_perl&#x2F;5.26.2&#x2F;x86_64-linux-thread-multi &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;site_perl&#x2F;5.26.2 &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;5.26.2&#x2F;x86_64-linux-thread-multi &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;5.26.2 .) at .&#x2F;scripts&#x2F;orthomcl-pipeline-setup.pl line 21.</span><br><span class="line">BEGIN failed--compilation aborted at .&#x2F;scripts&#x2F;orthomcl-pipeline-setup.pl line 21.</span><br><span class="line"></span><br><span class="line">(python37) [fenglei@localhost orthomcl-pipeline]$ perl .&#x2F;scripts&#x2F;orthomcl-pipeline-setup.pl</span><br><span class="line">Can&#39;t locate loadable object for module Algorithm::Combinatorics in @INC (@INC contains: &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;site_perl&#x2F;5.26.2&#x2F;x86_64-linux-thread-multi&#x2F;auto&#x2F; &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0&#x2F;x86_64-linux-thread-multi &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;perl5&#x2F;site_perl&#x2F;5.22.0 &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;site_perl&#x2F;5.26.2&#x2F;x86_64-linux-thread-multi &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;site_perl&#x2F;5.26.2 &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;5.26.2&#x2F;x86_64-linux-thread-multi &#x2F;home&#x2F;fenglei&#x2F;local&#x2F;app&#x2F;anaconda3&#x2F;envs&#x2F;python37&#x2F;lib&#x2F;5.26.2 .) at .&#x2F;scripts&#x2F;orthomcl-pipeline-setup.pl line 21.</span><br><span class="line">Compilation failed in require at .&#x2F;scripts&#x2F;orthomcl-pipeline-setup.pl line 21.</span><br><span class="line">BEGIN failed--compilation aborted at .&#x2F;scripts&#x2F;orthomcl-pipeline-setup.pl line 21.</span><br></pre></td></tr></table></figure></div><div class="tags"></div><div class="post-nav"><a class="pre" href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">jcvi 安装与使用，报错以及解决办法</a><a class="next" href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Bioinfo | Using Circos for genome visualization</a></div></div></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Linux | 认识 PKG_CONFIG_PATH</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">jcvi 安装与使用，报错以及解决办法</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">Bioinfo | Installation of OrthoMCL-pipeline</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Bioinfo | Using Circos for genome visualization</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">Bioinfo | Repeat sequence analysis by RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer)</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/16/Bioinfo-understanding-sequence-identity/">Bioinfo | 怎么理解序列一致性（Sequence identity）</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/10/Bioinfo-Extracting-sequence-by-names/">Bioinfo | 依据序列名字列表从 fasta 文件提取序列</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/05/Proteomics-Serotonylation-A-novel-histone-H3-marker/">Proteomics | 组蛋白五羟色胺化修饰 H3Q5ser 及其生物学功能</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>