<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon32.ico">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon16.ico">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta property="og:type" content="website">
<meta property="og:title" content="FENG&#39;s Notes">
<meta property="og:url" content="http://example.com/page/15/index.html">
<meta property="og:site_name" content="FENG&#39;s Notes">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="FENG Lei">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/page/15/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>FENG's Notes</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">FENG's Notes</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Genomics/Bioinformatics/Plants</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/02/29/rnaseq-pipeline-alignment-to-de-analysis/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/02/29/rnaseq-pipeline-alignment-to-de-analysis/" class="post-title-link" itemprop="url">DESeq2: RNAseq pipeline – Alignment to DE analysis</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-02-29 21:57:13" itemprop="dateCreated datePublished" datetime="2016-02-29T21:57:13+08:00">2016-02-29</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h1 id="rnaseq-pipeline-alignment-to-de-analysis"><a class="markdownIt-Anchor" href="#rnaseq-pipeline-alignment-to-de-analysis"></a> <a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/2014/11/10/rnaseq-pipeline-alignment-to-de-analysis-gsea/">RNAseq pipeline – alignment to DE analysis</a></h1>
<p><a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/2014/11/10/rnaseq-pipeline-alignment-to-de-analysis-gsea/#comments">1 Reply</a></p>
<p>RNA sequencing has quickly surpassed microarrays for differential gene expression analysis in most labs.  I wanted to post a RNAseq pipelines from Raw sequencing reads to differential expression analysis and various other downstream analyses.  There are quite a few pipelines for RNAseq analysis and each has its pro/cons.  One of the most popular pipelines is alignment to the genome with <a target="_blank" rel="noopener" href="https://code.google.com/p/rna-star/">STAR aligner</a>, binning of sequencing reads to genes/exons with <a target="_blank" rel="noopener" href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">HTseq-count</a>, and finally differential analysis with <a target="_blank" rel="noopener" href="http://bioconductor.org/packages/release/bioc/html/DESeq.html">DESeq</a>/<a target="_blank" rel="noopener" href="http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a>.  I linked the online manuals for each tool here but I would highly suggest the initial publications for each as those will include benchmarking compared to other programs and the general theory behind of each algorithm (<a target="_blank" rel="noopener" href="http://bioinformatics.oxfordjournals.org/content/early/2012/10/25/bioinformatics.bts635">here</a>, <a target="_blank" rel="noopener" href="http://biorxiv.org/content/early/2014/08/19/002824">here</a> and <a target="_blank" rel="noopener" href="http://genomebiology.com/2010/11/10/r106">here</a> – the original Deseq paper now has over 1800 citations on google scholar!).  I’d suggest DESeq2. This example assumes that you starting with RNA sequencing reads in fastq format after quality control (FastQC, etc.) and demultiplexing (fastq_multx, etc.).  Those steps will be a future post.  Depending on who is doing your actual sequencing (in-house core, BGI, etc.), they will probably release the fastq reads to you after each of these steps have been done. <strong>1. Alignment to genome with STAR aligner v2.4.0.1.</strong> STAR is wonderfully fast and accurate alignment program written by Alexander Dobin.  Alignment speeds with STAR are advertised at  400 million pairs per hour for 100 bp paired end reads on 12-core server (compared to Tophat at 20 million reads per hour).   In our hands, on 50 bp paired end reads, we are aligning at about 600 million pairs per hour on average with ~75-80% unique alignment rate.   The key with STAR is that it needs A LOT of memory, at least 30 GB.  Our cluster has dedicated job queues to allocate up to 64 GB memory for each STAR alignment job. <strong>1a) Alignment generate genome.</strong> Before alignment can be done, you must prepare the necessary genome index files.  The basic input for genome index files are 1) genome in fasta format and 2) annotation gtf file (needed in later steps).  You can download the genome files for human/mouse/etc from <a target="_blank" rel="noopener" href="http://support.illumina.com/sequencing/sequencing_software/igenome.html">igenome</a> from Illumina.  For the new version of STAR, there are pre-made <a target="_blank" rel="noopener" href="http://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STARgenomes/">index files</a>, depending on which reference you want to build.  They have various builds and model organisms.  We have been using the mouse mm10 reference from UCSC and have been happy with alignment rates.  I downloaded the mouse mm10 reference from igenome and renamed it “mm10_index/” .  You will also have to designate a directory for the genome index files STAR produces (“~/STAR_mm10/” or /path/to/STAR_mm10/”).</p>
<p>1</p>
<p>2</p>
<p>3</p>
<p>4</p>
<p><code>STAR --runMode genomeGenerate \</code></p>
<p><code>--genomeDir</code> <code>/path/to/STAR_mm10/</code> <code>\</code></p>
<p><code>--genomeFastaFiles ~``/mm10_index/Mus_musculus/UCSC/mm10/Sequence/WholeGenomeFasta/genome``.fa \</code></p>
<p><code>--runThreadN 16</code></p>
<p>After STAR runs, you should have a few files in the /STAR_mm10/ directory:</p>
<ol>
<li>Genome</li>
<li>genomeParameters.txt</li>
<li>chrLength.txt, chrNameLength.txt, chrStart.txt</li>
<li>SA, SAindex</li>
<li>Log.out</li>
</ol>
<p>STAR will use these files for alignment in the next step. <strong>1b) Alignment to genome (or transcriptome).</strong>  The newer version (2.4.0). has the option to align specifically to the transcriptome and not the genome.  This option is for downstream analysis with RSEM ( another future post).  I’ll detail the basic STAR alignment job for now.</p>
<p>1</p>
<p>2</p>
<p>3</p>
<p>4</p>
<p><code>STAR --genomeDir</code> <code>/path/to/STAR_mm10/</code> <code>\</code></p>
<p><code>--readFilesIn read_1.fq read_2.fq \</code></p>
<p><code>--runThreadN 16 \</code></p>
<p><code>--outSAMtype BAM SortedByCoordinate</code></p>
<p>STAR will generate a few files:</p>
<ol>
<li>Log.out</li>
<li>Log.progress.out</li>
<li>Log.final.out – summary mapping statistics – will give you an idea on qc</li>
<li>Aligned.sortedByCoord.out.bam – your reads aligned in standard BAM format sorted by coordinate – similar to samtools sort command</li>
<li>SJ.out.tab – high confidence splice junction in tab-delimited format</li>
</ol>
<p>To generate an alignment file to the transcriptome with STAR v2.4.0:</p>
<p>1</p>
<p>2</p>
<p>3</p>
<p>4</p>
<p><code>STAR --genomeDir</code> <code>/path/to/STAR_mm10/</code> <code>\</code></p>
<p><code>--readFilesIn read_1.fq read_2.fq \</code></p>
<p><code>--runThreadN 16 \</code></p>
<p><code>--quantMode TranscriptomeSAM</code></p>
<p>I haven’t had a chance to really run through this option yet (more for RSEM instead of DESeq2) but I am excited to try. <strong>Optional) Sorting, munging, etc.</strong>  The new version of STAR allows for generating alignment into a BAM file sorted by genomic coordinate.  Earlier versions didn’t and one would have to run the Aligned.out.sam files through various samtools munging steps to get alignment file into the correct format for downstream read counting.  For this example, it is necessary to have samtools-0.1.19.  First you convert the Aligned.out.sam to bam format with samtools view and pipe the output directly to samtools sort and sorting by genome position.  Finally converting back from bam to sam format.  If you have the python module pysam installed you can use a BAM file in the next steps.</p>
<p>1</p>
<p>2</p>
<p><code>samtools view -uS Aligned.out.sam -b samtools</code> <code>sort</code> <code>-@ 8 -m 800000000 - Aligned.sorted</code></p>
<p><code>samtools view Aligned.sorted.bam &gt; Aligned.sorted.sam</code></p>
<p>**3) raw counts generation using HTSeq-count. ** htseq-count has a few options:</p>
<ul>
<li>-m union: union is one of three options for htseq count for binning of reads to features (genes/exons).  it is the recommended option per <a target="_blank" rel="noopener" href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">htseq count’s manual</a>.  It is the most stringent option as it will only count a read if it aligns to only 1 gene.</li>
<li>-r pos : meaning that the BAM/SAM file is sorted by genomic coordinate/position</li>
<li>-i gene_name : tells htseq-count to use the gene name in output file</li>
<li>-a 10 : tells htseq-count to skip reads with alignment score less than 10.</li>
<li>–stranded=no : the RNAseq reads are not strand specific</li>
</ul>
<p>1</p>
<p>2</p>
<p>3</p>
<p>4</p>
<p>5</p>
<p>6</p>
<p>7</p>
<p><code>htseq-count -m union \</code></p>
<p><code>-r pos \</code></p>
<p><code>-i gene_name \</code></p>
<p><code>-a 10 \</code></p>
<p><code>--stranded=no \</code></p>
<p><code>Aligned.sortedByCoord.out.bam \</code></p>
<p><code>~``/mm10_index/Mus_musculus/UCSC/mm10/Annotation/Genes/genes``.gtf &gt; output_basename.counts</code></p>
<p>Output of htseq count is a tab-delimited text file containing two columns: 1) gene id 2) number of read counts. Below is just an example of various cadherins with number of reads mapped to each.</p>
<p>1</p>
<p>2</p>
<p>3</p>
<p>4</p>
<p>5</p>
<p>6</p>
<p>7</p>
<p>8</p>
<p>9</p>
<p>10</p>
<p>11</p>
<p>12</p>
<p>13</p>
<p>14</p>
<p>15</p>
<p>16</p>
<p>17</p>
<p>18</p>
<p>19</p>
<p>20</p>
<p><code>Cdh1 731</code></p>
<p><code>Cdh10 41</code></p>
<p><code>Cdh11 129</code></p>
<p><code>Cdh12 0</code></p>
<p><code>Cdh13 50538</code></p>
<p><code>Cdh15 0</code></p>
<p><code>Cdh16 0</code></p>
<p><code>Cdh17 648</code></p>
<p><code>Cdh18 0</code></p>
<p><code>Cdh19 0</code></p>
<p><code>Cdh2 570</code></p>
<p><code>Cdh20 0</code></p>
<p><code>Cdh22 0</code></p>
<p><code>Cdh23 11</code></p>
<p><code>Cdh24 3</code></p>
<p><code>Cdh26 140</code></p>
<p><code>Cdh3 3</code></p>
<p><code>Cdh4 0</code></p>
<p><code>Cdh5 237</code></p>
<p><code>Cdh6 496</code></p>
<p><strong>4) differential gene expression analysis using DESeq2.</strong></p>
<p>1</p>
<p>2</p>
<p>3</p>
<p>4</p>
<p>5</p>
<p>6</p>
<p>7</p>
<p>8</p>
<p>9</p>
<p>10</p>
<p>11</p>
<p>12</p>
<p>13</p>
<p>14</p>
<p>15</p>
<p>16</p>
<p>17</p>
<p>18</p>
<p>19</p>
<p>20</p>
<p>21</p>
<p>22</p>
<p>23</p>
<p>24</p>
<p>25</p>
<p>26</p>
<p>27</p>
<p>28</p>
<p>29</p>
<p>30</p>
<p>31</p>
<p>32</p>
<p>33</p>
<p>34</p>
<p>35</p>
<p>36</p>
<p>37</p>
<p>38</p>
<p>39</p>
<p>40</p>
<p>41</p>
<p>42</p>
<p>43</p>
<p>44</p>
<p>45</p>
<p>46</p>
<p>47</p>
<p>48</p>
<p>49</p>
<p>50</p>
<p>51</p>
<p>52</p>
<p>53</p>
<p>54</p>
<p>55</p>
<p>56</p>
<p>57</p>
<p>58</p>
<p>59</p>
<p>60</p>
<p>61</p>
<p>62</p>
<p>63</p>
<p>64</p>
<p>65</p>
<p>66</p>
<p>67</p>
<p>68</p>
<p>69</p>
<p>70</p>
<p>71</p>
<p>72</p>
<p>73</p>
<p>74</p>
<p>75</p>
<p>76</p>
<p>77</p>
<p>78</p>
<p>79</p>
<p>80</p>
<p>81</p>
<p>82</p>
<p>83</p>
<p>84</p>
<p>85</p>
<p>86</p>
<p>87</p>
<p>88</p>
<p>89</p>
<p>90</p>
<p>91</p>
<p>92</p>
<p>93</p>
<p>94</p>
<p>95</p>
<p>96</p>
<p>97</p>
<p>98</p>
<p>99</p>
<p>100</p>
<p>101</p>
<p>102</p>
<p>103</p>
<p>104</p>
<p>105</p>
<p>106</p>
<p>107</p>
<p>108</p>
<p>109</p>
<p>110</p>
<p>111</p>
<p>112</p>
<p>113</p>
<p>114</p>
<p>115</p>
<p>116</p>
<p>117</p>
<p>118</p>
<p>119</p>
<p>120</p>
<p>121</p>
<p>122</p>
<p>123</p>
<p>124</p>
<p>125</p>
<p>126</p>
<p>127</p>
<p>128</p>
<p>129</p>
<p>130</p>
<p>131</p>
<p>132</p>
<p>133</p>
<p>134</p>
<p>135</p>
<p>136</p>
<p>137</p>
<p>138</p>
<p>139</p>
<p>140</p>
<p>141</p>
<p>142</p>
<p>143</p>
<p>144</p>
<p>145</p>
<p>146</p>
<p>147</p>
<p>148</p>
<p>149</p>
<p>150</p>
<p>151</p>
<p>152</p>
<p>153</p>
<p>154</p>
<p>155</p>
<p>156</p>
<p>157</p>
<p>158</p>
<p>159</p>
<p>160</p>
<p>161</p>
<p>162</p>
<p>163</p>
<p>164</p>
<p>165</p>
<p>166</p>
<p>167</p>
<p>168</p>
<p>169</p>
<p>170</p>
<p>171</p>
<p>172</p>
<p>173</p>
<p>174</p>
<p>175</p>
<p>176</p>
<p>177</p>
<p>178</p>
<p>179</p>
<p>180</p>
<p>181</p>
<p>182</p>
<p>183</p>
<p>184</p>
<p>185</p>
<p>186</p>
<p>187</p>
<p>188</p>
<p>189</p>
<p>190</p>
<p>191</p>
<p>192</p>
<p>193</p>
<p>194</p>
<p>195</p>
<p>196</p>
<p>197</p>
<p>198</p>
<p>199</p>
<p>200</p>
<p>201</p>
<p>202</p>
<p>203</p>
<p>204</p>
<p>205</p>
<p>206</p>
<p>207</p>
<p>208</p>
<p>209</p>
<p>210</p>
<p>211</p>
<p>212</p>
<p>213</p>
<p>214</p>
<p>215</p>
<p>216</p>
<p>217</p>
<p>218</p>
<p>219</p>
<p>220</p>
<p>221</p>
<p>222</p>
<p>223</p>
<p>224</p>
<p>225</p>
<p>226</p>
<p>227</p>
<p>228</p>
<p>229</p>
<p>230</p>
<p>231</p>
<p>232</p>
<p>233</p>
<p>234</p>
<p>235</p>
<p>236</p>
<p>237</p>
<p>238</p>
<p>239</p>
<p>240</p>
<p>241</p>
<p>242</p>
<p>243</p>
<p>244</p>
<p>245</p>
<p>246</p>
<p>247</p>
<p>248</p>
<p>249</p>
<p>250</p>
<p>251</p>
<p>252</p>
<p>253</p>
<p>254</p>
<p>255</p>
<p>256</p>
<p>257</p>
<p>258</p>
<p>259</p>
<p>260</p>
<p>261</p>
<p>262</p>
<p>263</p>
<p>264</p>
<p>265</p>
<p>266</p>
<p><code>&lt;pre&gt;``####################################################################################</code></p>
<p><code># bioinformatic analysis of RNA-seq data using DESeq2</code></p>
<p><code>#</code></p>
<p><code># combined from DESeq2 manual and vignette</code></p>
<p><code># [http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf)</code></p>
<p><code># and from Dave Wheeler's blog at Massey University</code></p>
<p><code># [http://dwheelerau.com/2014/02/17/how-to-use-deseq2-to-analyse-rnaseq-data/](http://dwheelerau.com/2014/02/17/how-to-use-deseq2-to-analyse-rnaseq-data/)</code></p>
<p><code>#</code></p>
<p><code># analysis per htseq count -intersections-nonempty</code></p>
<p><code># controlling for batch (different RNAseq prep libraries)</code></p>
<p><code>#</code></p>
<p><code># notes:</code></p>
<p><code>#</code></p>
<p><code>####################################################################################</code></p>
<p><code>library``(``&quot;DESeq2&quot;``)</code></p>
<p><code>setwd``(``&quot;~/path/to/working/directory/&quot;``)</code></p>
<p><code>directory &lt;-</code> <code>&quot;/path/to/counts/directory/&quot;</code></p>
<p><code># can merge individual sample files (i.e. ctrl1.counts, ctrl2.counts, etc.)</code></p>
<p><code>sampleFiles &lt;-</code> <code>grep``(``&quot;counts&quot;``list.files``(directory),value=T)</code></p>
<p><code># view sampleFiles</code></p>
<p><code>sampleFiles</code></p>
<p><code># can designate different batches of samples (i.e. different sequencers,</code></p>
<p><code># PE vs SE, different library preps (eg. BATCH1 vs BATCH2))</code></p>
<p><code>sampleBatch &lt;-</code> <code>c``(``&quot;Batch1&quot;``,``&quot;Batch1&quot;``,``&quot;Batch1&quot;``,``&quot;Batch1&quot;``,``&quot;Batch1&quot;``,``&quot;Batch1&quot;``,</code></p>
<p><code>&quot;Batch2&quot;``,``&quot;Batch2&quot;``,``&quot;Batch2&quot;``,``&quot;Batch2&quot;``)</code></p>
<p><code># set sampleConditions and sampleTable for experimental conditions</code></p>
<p><code>sampleCondition &lt;-</code> <code>c``(``&quot;Control, Experimental&quot;``)</code></p>
<p><code>sampleTable &lt;-</code> <code>data.frame``(sampleName = sampleFiles,</code></p>
<p><code>fileName = sampleFiles,</code></p>
<p><code>condition = sampleCondition,</code></p>
<p><code>Batch = sampleBatch)</code></p>
<p><code># view sampleTable</code></p>
<p><code>sampleTable</code></p>
<p><code>ddsHTseq &lt;-</code> <code>DESeqDataSetFromHTSeqCount``( sampleTable = sampleTable,</code></p>
<p><code>directory = directory,</code></p>
<p><code>design = ~condition)</code></p>
<p><code>## view ddsHTseq - should give summary of class, data, etc.</code></p>
<p><code>ddsHTseq</code></p>
<p><code>colData``(ddsHTseq)$condition &lt;-</code> <code>factor``(``colData``(ddsHTseq)$condition,</code></p>
<p><code>levels =</code> <code>c``(``'Control'``,``'Experimental'``))</code></p>
<p><code># gut of DESeq2 analysis</code></p>
<p><code>dds &lt;-</code> <code>DESeq``(ddsHTseq)</code></p>
<p><code>res &lt;-</code> <code>results``(dds)</code></p>
<p><code># order results by padj value (most significant to least)</code></p>
<p><code>res &lt;- res[``order``(res$padj),]</code></p>
<p><code>head``(res)</code></p>
<p><code># should see DataFrame of baseMean, log2Foldchange, stat, pval, padj</code></p>
<p><code># save data results and normalized reads to csv!</code></p>
<p><code>resdata &lt;-</code> <code>merge``(``as.data.frame``(res),</code> <code>as.data.frame``(``counts``(dds,normalized=T)), by=``'row.names'``,sort=F)</code></p>
<p><code>names``(resdata)[1] &lt;-</code> <code>'gene'</code></p>
<p><code>head``(resdata)</code></p>
<p><code>write.csv``(resdata, file=``&quot;DATE-DESeq2-results-with-normalized.csv&quot;``)</code></p>
<p><code># send normalized counts to tab delimited file for GSEA, etc.</code></p>
<p><code>write.table``(``as.data.frame``(``counts``(dds),normalized=T), file =</code> <code>'DATE_DESeq2_normalized_counts.txt'``, sep =</code> <code>'\t'``)</code></p>
<p><code># produce DataFrame of results of statistical tests</code></p>
<p><code># could way to record experimental design</code></p>
<p><code>mcols``(res, use.names = T)</code></p>
<p><code>write.csv``(``as.data.frame``(``mcols``(res, use.name = T)),file =</code> <code>&quot;DATE-DESeq2-test-conditions.csv&quot;``)</code></p>
<p><code># replacing outlier value with estimated value as predicted by distrubution using</code></p>
<p><code># &quot;trimmed mean&quot; approach. recommended if you have several replicates per treatment</code></p>
<p><code># DESeq2 will automatically do this if you have 7 or more replicates</code></p>
<p><code>ddsClean &lt;-</code> <code>replaceOutliersWithTrimmedMean``(dds)</code></p>
<p><code>ddsClean &lt;-</code> <code>DESeq``(ddsClean)</code></p>
<p><code>tab &lt;-</code> <code>table``(initial =</code> <code>results``(dds)$padj &lt; 0.1,</code></p>
<p><code>cleaned =</code> <code>results``(ddsClean)$padj &lt; 0.1)</code></p>
<p><code>addmargins``(tab)</code></p>
<p><code>write.csv``(``as.data.frame``(tab),file =</code> <code>'DATE-DESeq2-replaceoutliers.csv'``)</code></p>
<p><code>resClean &lt;-</code> <code>results``(ddsClean)</code></p>
<p><code>resClean &lt;- resClean[``order``(resClean$padj),]</code></p>
<p><code>head``(resClean)</code></p>
<p><code>write.csv``(``as.data.frame``(resClean),file =</code> <code>'DATE-DESeq2-replaceoutliers-results.csv'``)</code></p>
<p><code>####################################################################################</code></p>
<p><code># Exploritory data analysis of RNAseq data with DESeq2</code></p>
<p><code>#</code></p>
<p><code># these next R scripts are for a variety of visualization, QC and other plots to</code></p>
<p><code># get a sense of what the RNAseq data looks like based on DESEq2 analysis</code></p>
<p><code>#</code></p>
<p><code># 1) MA plot</code></p>
<p><code># 2) rlog stabilization and variance stabiliazation</code></p>
<p><code># 3) variance stabilization plot</code></p>
<p><code># 4) heatmap of clustering analysis</code></p>
<p><code># 5) PCA plot</code></p>
<p><code>#</code></p>
<p><code>#</code></p>
<p><code>####################################################################################</code></p>
<p><code># MA plot of RNAseq data for entire dataset</code></p>
<p><code># [http://en.wikipedia.org/wiki/MA_plot](http://en.wikipedia.org/wiki/MA_plot)</code></p>
<p><code># genes with padj &lt; 0.1 are colored Red</code></p>
<p><code>plotMA``(dds, ylim=``c``(-8,8),main =</code> <code>&quot;RNAseq experiment&quot;``)</code></p>
<p><code>dev.copy``(png,</code> <code>&quot;DATE-DESeq2_MAplot_initial_analysis.png&quot;``)</code></p>
<p><code>dev.off``()</code></p>
<p><code># transform raw counts into normalized values</code></p>
<p><code># DESeq2 has two options:  1) rlog transformed and 2) variance stabilization</code></p>
<p><code># variance stabilization is very good for heatmaps, etc.</code></p>
<p><code>rld &lt;-</code> <code>rlogTransformation``(dds, blind=T)</code></p>
<p><code>vsd &lt;-</code> <code>varianceStabilizingTransformation``(dds, blind=T)</code></p>
<p><code># save normalized values</code></p>
<p><code>write.table``(``as.data.frame``(``assay``(rld),file=``'DATE-DESeq2-rlog-transformed-counts.txt'``, sep=``'\t'``)</code></p>
<p><code>write.table``(``as.data.frame``(``assay``(vsd),file=``'DATE-DESeq2-vst-transformed-counts.txt'``, sep=``'\t'``)</code></p>
<p><code># plot to show effect of transformation</code></p>
<p><code># axis is square root of variance over the mean for all samples</code></p>
<p><code>par``(mai =</code> <code>ifelse``(1:4 &lt;= 2,</code> <code>par``(``'mai'``),0))</code></p>
<p><code>px &lt;-</code> <code>counts``(dds)[,1] /</code> <code>sizeFactors``(dds)[1]</code></p>
<p><code>ord &lt;-</code> <code>order``(px)</code></p>
<p><code>ord &lt;- ord[px[ord] &lt; 150]</code></p>
<p><code>ord &lt;- ord[``seq``(1,``length``(ord),length=50)]</code></p>
<p><code>last &lt;- ord[``length``(ord)]</code></p>
<p><code>vstcol &lt;-</code> <code>c``(``'blue'``,``'black'``)</code></p>
<p><code>matplot``(px[ord],</code> <code>cbind``(``assay``(vsd)[,1],</code> <code>log2``(px))[ord, ],type=``'l'``, lty = 1, col=vstcol, xlab =</code> <code>'n'``, ylab =</code> <code>'f(n)'``)</code></p>
<p><code>legend``(``'bottomright'``,legend=``c``(``expression``(``'variance stabilizing transformation'``),</code> <code>expression``(log[2](n/s[1]))), fill=vstcol)</code></p>
<p><code>dev.copy``(png,``&quot;DATE-DESeq2_variance_stabilizing.png&quot;``)</code></p>
<p><code>dev.off``()</code></p>
<p><code># clustering analysis</code></p>
<p><code>library``(``&quot;gplots&quot;``)</code></p>
<p><code>distsRL &lt;-</code> <code>dist``(``t``(``assay``(rld)))</code></p>
<p><code>mat &lt;-</code> <code>as.matrix``(distsRL)</code></p>
<p><code>rownames``(mat) &lt;-</code> <code>colnames``(mat) &lt;-</code> <code>with``(``colData``(dds),</code> <code>paste``(condition, type, sep=``&quot; : &quot;``))</code></p>
<p><code># OR</code></p>
<p><code># if you want the conditions used</code></p>
<p><code># rownames(mat) &lt;- colnames(mat) &lt;- with(colData(dds),condition)</code></p>
<p><code>heatmap.2``(mat, trace =</code> <code>&quot;none&quot;``, col =</code> <code>rev``(hmcol), margin =</code> <code>c``(13, 13))</code></p>
<p><code>dev.copy``(png,</code> <code>&quot;DATE-DESeq2-clustering.png&quot;``)</code></p>
<p><code>dev.off``()</code></p>
<p><code># Principal components plot</code></p>
<p><code># will show additional clustering of samples</code></p>
<p><code># showing basic PCA function in R from DESeq2 package</code></p>
<p><code># this lacks sample IDs and only broad sense of sample clustering</code></p>
<p><code># its not nice - but it does the job</code></p>
<p><code>print``(``plotPCA``(rld, intgroup =</code> <code>c``(``&quot;condition&quot;``)))</code></p>
<p><code>dev.copy``(png,</code> <code>&quot;DATE-DESeq2_PCA_initial_analysis.png&quot;``)</code></p>
<p><code>dev.off``()</code></p>
<p><code># or ggplot PCA plot</code></p>
<p><code>library``(``&quot;grDevices&quot;``)</code></p>
<p><code>library``(``'ggplot2'``)</code></p>
<p><code>library``(``&quot;genefilter&quot;``)</code></p>
<p><code>rv &lt;-</code> <code>rowVars``(``assay``(rld))</code></p>
<p><code>select &lt;-</code> <code>order``(rv, decreasing=T)[``seq_len``(``min``(500,``length``(rv)))]</code></p>
<p><code>pc &lt;-</code> <code>prcomp``(``t``(``assay``(vsdMF)[select,]))</code></p>
<p><code># set condition</code></p>
<p><code>condition &lt;-</code> <code>c``(``&quot;condition1&quot;``,</code> <code>'condition2'``)</code></p>
<p><code>scores &lt;-</code> <code>data.frame``(sampleFiles, pca$x, condition)</code></p>
<p><code>(pcaplot &lt;-</code> <code>ggplot``(scores,</code> <code>aes``(x = PC1, y = PC2, col = (``factor``(condition))))</code></p>
<p><code>+</code> <code>geom_point``(size = 5)</code></p>
<p><code>+</code> <code>ggtitle``(``&quot;Principal Components&quot;``)</code></p>
<p><code>+</code> <code>scale_colour_brewer``(name =</code> <code>&quot; &quot;``, palette =</code> <code>&quot;Set1&quot;``)</code></p>
<p><code>+</code> <code>theme``(</code></p>
<p><code>plot.title =</code> <code>element_text``(face =</code> <code>'bold'``),</code></p>
<p><code>legend.position =</code> <code>c``(0,0),</code></p>
<p><code>legend.key =</code> <code>element_rect``(fill =</code> <code>'NA'``),</code></p>
<p><code>legend.text =</code> <code>element_text``(size = 10, face =</code> <code>&quot;bold&quot;``),</code></p>
<p><code>axis.text.y =</code> <code>element_text``(colour =</code> <code>&quot;Black&quot;``),</code></p>
<p><code>axis.text.x =</code> <code>element_text``(colour =</code> <code>&quot;Black&quot;``),</code></p>
<p><code>axis.title.x =</code> <code>element_text``(face =</code> <code>&quot;bold&quot;``),</code></p>
<p><code>axis.title.y =</code> <code>element_text``(face =</code> <code>'bold'``),</code></p>
<p><code>panel.grid.major.x =</code> <code>element_blank``(),</code></p>
<p><code>panel.grid.major.y =</code> <code>element_blank``(),</code></p>
<p><code>panel.grid.minor.x =</code> <code>element_blank``(),</code></p>
<p><code>panel.grid.minor.y =</code> <code>element_blank``(),</code></p>
<p><code>panel.background =</code> <code>element_rect``(color =</code> <code>'black'``,fill =</code> <code>NA``)</code></p>
<p><code>))</code></p>
<p><code>ggsave``(pcaplot,file=``&quot;DATE-PCA_ggplot2.pdf&quot;``)</code></p>
<p><code># scatter plot of rlog transformations between Sample conditions</code></p>
<p><code># nice way to compare control and experimental samples</code></p>
<p><code>head``(``assay``(rld))</code></p>
<p><code>plot``(``log2``(1+``counts``(dds,normalized=T)[,1:2]),col=``'black'``,pch=20,cex=0.3, main=``'Log2 transformed'``)</code></p>
<p><code>plot``(``assay``(rld)[,1:2],col=``'#00000020'``,pch=20,cex=0.3, main =</code> <code>&quot;rlog transformed&quot;``)</code></p>
<p><code>plot``(``assay``(rld)[,3:4],col=``'#00000020'``,pch=20,cex=0.3, main =</code> <code>&quot;rlog transformed&quot;``)</code></p>
<p><code>plot``(``assay``(rld)[,5:6],col=``'#00000020'``,pch=20,cex=0.3, main =</code> <code>&quot;rlog transformed&quot;``)</code></p>
<p><code>plot``(``assay``(rld)[,7:8],col=``'#00000020'``,pch=20,cex=0.3, main =</code> <code>&quot;rlog transformed&quot;``)</code></p>
<p><code>plot``(``assay``(rld)[,9:10],col=``'#00000020'``,pch=20,cex=0.3, main =</code> <code>&quot;rlog transformed&quot;``)</code></p>
<p><code>plot``(``assay``(rld)[,11:12],col=``'#00000020'``,pch=20,cex=0.3, main =</code> <code>&quot;rlog transformed&quot;``)</code></p>
<p><code>plot``(``assay``(rld)[,13:14],col=``'#00000020'``,pch=20,cex=0.3, main =</code> <code>&quot;rlog transformed&quot;``)</code></p>
<p><code>plot``(``assay``(rld)[,15:16],col=``'#00000020'``,pch=20,cex=0.3, main =</code> <code>&quot;rlog transformed&quot;``)</code></p>
<p><code># heatmap of data</code></p>
<p><code>library``(``&quot;RColorBrewer&quot;``)</code></p>
<p><code>library``(``&quot;gplots&quot;``)</code></p>
<p><code># 1000 top expressed genes with heatmap.2</code></p>
<p><code>select &lt;-</code> <code>order``(``rowMeans``(``counts``(dds,normalized=T)),decreasing=T)[1:1000]</code></p>
<p><code>my_palette &lt;-</code> <code>colorRampPalette``(``c``(``&quot;blue&quot;``,``'white'``,``'red'``))(n=1000)</code></p>
<p><code>heatmap.2``(``assay``(vsd)[select,], col=my_palette,</code></p>
<p><code>scale=``&quot;row&quot;``, key=T, keysize=1, symkey=T,</code></p>
<p><code>density.info=``&quot;none&quot;``, trace=``&quot;none&quot;``,</code></p>
<p><code>cexCol=0.6, labRow=F,</code></p>
<p><code>main=``&quot;TITLE&quot;``)</code></p>
<p><code>dev.copy``(png,</code> <code>&quot;DATE-DESeq2-HEATMAP.png&quot;``)</code></p>
<p><code>dev.off``()</code></p>
<p><code># top 2000 genes based on row variance with heatmap3</code></p>
<p><code>library``(heatmap3)</code></p>
<p><code>colsidecolors =</code> <code>c``(``&quot;darkgreen&quot;``,``&quot;darkgreen&quot;``,</code></p>
<p><code>&quot;mediumpurple2&quot;``,``&quot;mediumpurple2&quot;``,</code></p>
<p><code>&quot;mediumpurple2&quot;``,``&quot;mediumpurple2&quot;``,</code></p>
<p><code>&quot;darkgreen&quot;``,``&quot;darkgreen&quot;``,</code></p>
<p><code>&quot;mediumpurple2&quot;``,``&quot;mediumpurple2&quot;``,</code></p>
<p><code>&quot;darkgreen&quot;``,``&quot;darkgreen&quot;``,</code></p>
<p><code>&quot;mediumpurple2&quot;``,``&quot;mediumpurple2&quot;``,</code></p>
<p><code>&quot;mediumpurple2&quot;``,``&quot;mediumpurple2&quot;``)</code></p>
<p><code>rv &lt;-</code> <code>rowVars``(``assay``(vsd))</code></p>
<p><code>select &lt;-</code> <code>order``(rv, decreasing=T)[``seq_len``(``min``(2000,``length``(rv)))]</code></p>
<p><code>my_palette &lt;-</code> <code>colorRampPalette``(``c``(``&quot;blue&quot;``,</code> <code>&quot;white&quot;``,</code> <code>&quot;red&quot;``))(1024)</code></p>
<p><code>heatmap3``(``assay``(vsd)[select,],col=my_palette,</code></p>
<p><code>labRow = F,cexCol = 0.8,margins=``c``(6,6))</code></p>
<p><code>dev.copy``(pdf,</code> <code>&quot;DATE-DESeq2-heatmap3.pdf&quot;``)</code></p>
<p><code>dev.off``()</code></p>
<p><code>sessionInfo``()</code></p>
<p><code>###############################################################</code></p>
<p><code>#</code></p>
<p><code># Optional analyses</code></p>
<p><code>#</code></p>
<p><code>###############################################################</code></p>
<p><code># multifactor designs</code></p>
<p><code># can analysis with more than one factor influencing the counts  (e.g., different library preps, PE vs. SE)</code></p>
<p><code># from manual section 1.5 for more information and rationale</code></p>
<p><code># make copy of DESeq2 results</code></p>
<p><code>ddsMF &lt;- dds</code></p>
<p><code># change design formulate controlling for Batch</code></p>
<p><code>design``(ddsMF) &lt;-</code> <code>formula``(~ Batch + condition)</code></p>
<p><code># rerun DESeq analysis</code></p>
<p><code>ddsMF &lt;-</code> <code>DESeq``(ddsMF)</code></p>
<p><code>resMF &lt;-</code> <code>results``(ddsMF)</code></p>
<p><code># order by padj values</code></p>
<p><code>resMF &lt;- resMF[``order``(resMF$padj),]</code></p>
<p><code>head``(resMF)</code></p>
<p><code># should see DataFrame of baseMean, log2Foldchange, stat, pval, padj</code></p>
<p><code># save data 'resMF' to csv!</code></p>
<p><code>write.csv``(``as.data.frame``(resMF),file=``'DATE-DESeq2_batchcontroll_initial_analysis.csv'``)</code></p>
<p>future post idea: pairing HTSeq-DESeq2 for DE gene expression with DEXSeq for splice variants as this pipeline is agnostic for differential exon usage.  *May miss true DE genes if total exon counts are the same but use different exons in different conditions (per Lior Pachter) Future post idea: comparing STAR-HTSeq-DESeq2 to Bowtie-RSEM for gene expression analysis on the same dataset.</p>
<p>This entry was posted in <a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/category/bioinformatics/">bioinformatics</a>, <a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/category/deseq2/">DESeq2</a>, <a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/category/dexseq/">DEXSeq</a>, <a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/category/ggplot2/">ggplot2</a>, <a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/category/htseq-count/">HTSeq-count</a>, <a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/category/r/">R</a>, <a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/category/rnaseq/">RNAseq</a>, <a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/category/star/">STAR</a> on<a target="_blank" rel="noopener" href="https://benchtobioinformatics.wordpress.com/2014/11/10/rnaseq-pipeline-alignment-to-de-analysis-gsea/" title="8:33 pm">November 10, 2014</a>.</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/02/29/Python-How-to-install-packages/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/02/29/Python-How-to-install-packages/" class="post-title-link" itemprop="url">Python：Python 模块的多种安装方法</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-02-29 21:10:07" itemprop="dateCreated datePublished" datetime="2016-02-29T21:10:07+08:00">2016-02-29</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-09-21 11:34:00" itemprop="dateModified" datetime="2022-09-21T11:34:00+08:00">2022-09-21</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Python/" itemprop="url" rel="index"><span itemprop="name">Python</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p><strong>Python模块安装方法</strong><br />
<strong>一、方法1： 单文件模块</strong><br />
直接把文件拷贝到 <code>$python\_dir/Lib</code></p>
<p><strong>二、方法2： 多文件模块，<a target="_blank" rel="noopener" href="http://xn--setup-3o1j.py">带setup.py</a></strong><br />
下载模块包，进行解压，进入模块文件夹，执行： <code>python setup.py install</code></p>
<p><strong>三、 方法3：easy_install 方式</strong><br />
先下载ez_setup.py,运行<code>python ez\_setup</code> 进行easy_install工具的安装，之后就可以使用<code>easy\_install</code>进行安装package了。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&gt;  easy\_install  packageName </span><br><span class="line"># 安装  easy\_install -m packageName </span><br><span class="line"># 卸载  easy\_install  package.egg #安装</span><br></pre></td></tr></table></figure>
<p><strong>四、方法4：pip 方式</strong><br />
先进行pip工具的安裝：easy_install pip（pip 可以通过easy_install 安裝，而且也会装到 Scripts 文件夹下。）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; 安裝：pip install PackageName </span><br><span class="line">更新：pip install -U PackageName </span><br><span class="line">移除：pip uninstall PackageName </span><br><span class="line">搜索：pip search PackageName </span><br><span class="line">帮助：pip help</span><br></pre></td></tr></table></figure>
<hr />
<p>注：虽然Python的模块可以拷贝安装，但是一般情况下推荐制作一个安装包，即写一个setup.py文件来安装。<br />
setup.py文件的使用如下:</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">python setup.py build     </span><br><span class="line">#编译 % python setup.py install    </span><br><span class="line">#安装 % python setup.py sdist     </span><br><span class="line">#制作分发包 % python setup.py bdist\_wininst    </span><br><span class="line">#制作windows下的分发包 % python setup.py bdist\_rpm</span><br></pre></td></tr></table></figure>
<p>setup.py文件的编写 setup.py中主要执行一个 setup函数，该函数中大部分是描述性东西，最主要的是packages参数，列出所有的package，可以用自带的find_packages来动态获取package。<br />
所以setup.py文件的编写实际是很简单的。 简单的例子: setup.py文件：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">from setuptools import setup, find\_packages setup(       name = &quot; mytest &quot; ,       version = &quot; 0.10 &quot; ,       description = &quot; My test module &quot; ,       author = &quot; Robin Hood &quot; ,       url = &quot; http://www.csdn.net &quot; ,       license = &quot; LGPL &quot; ,       packages = find\_packages(),       scripts = \[ &quot; scripts/test.py &quot; \],       ) mytest.py import sys def get():     return sys.path scripts/test.py import os print os.environ.keys() </span><br></pre></td></tr></table></figure>
<p>setup中的scripts表示将该文件放到 Python的Scripts目录下，可以直接用。OK，简单的安装成功，可以运行所列举的命令生成安装包，或者安装该python包。本机测试成功(win32-python25)！</p>
<p><strong>附注：setuptools工具安装方法</strong></p>
<h3 id="方法一-使用ez_setuppy安装setuptools"><a class="markdownIt-Anchor" href="#方法一-使用ez_setuppy安装setuptools"></a> （方法一）. 使用ez_setup.py安装<strong>setuptools</strong></h3>
<p>进入<a target="_blank" rel="noopener" href="https://pypi.python.org/pypi/setuptools%E4%B8%8B%E8%BD%BDez%5C_setup.py">https://pypi.python.org/pypi/setuptools下载ez\_setup.py</a> 这是 setuptools 自豪的一种安装方式，只需要一个大约 8K 作为的脚本ez_setup.py，就能自动为用户安装包括 setuptools 自身在内的许多 Python 包。 使用这种方式，用户只需要下载 ez_setup。py 并运行，就可以自动下载和安装适合用户当前 Python 版本的适当的 setuptools egg 文件(当然，用户需要 Python 2.3.5 以上的版本，64 位操作系统的用户则需要 Python 2.4 以上的版本)。此外，这段脚本还会将可执行的 easy_install 脚本安装到用户所有的操作系统 Python 可执行脚本正常应该安装的位置(例如，Windows 用户会安装到 Python 安装目录下的 Scripts 目录中)。关于这种安装方法的更详细说明和注意事项，请参考其官方说明（见扩展阅读）。简单的安装命令如下： 　　wget -q ez_setup。py下载地址（见扩展阅读） 安装完后，最好确保</p>
<h3 id="方法二-使用完整的安装包安装setuptools"><a class="markdownIt-Anchor" href="#方法二-使用完整的安装包安装setuptools"></a> （方法二）. 使用完整的安装包安装<strong>setuptools</strong></h3>
<p>当然，用户也可以直接使用 setuptools发布版本来安装。<br />
对于使用 Windows 的用户，这也是挺方便的方法，许多 Linux 发行版的官方包管理仓库都包含 setuptools 的某个版本。<br />
例如，如果你跟我一样使用 Ubuntu ，那安装 setuptools 只是简单的进行如下操作： # apt-get install python-setuptools</p>
<p>安装 easy_install package-name，比如 easy_install pylab</p>
<p>模块卸载 easy_install -m package-name， 比如easy_install -m pylab</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">easy\_install -m 包名，可以卸载软件包，但是卸载后还要手动删除遗留文件。</span><br></pre></td></tr></table></figure>
<p>setuptools它可以自动的安装模块，只需要你提供给它一个模块名字就可以，并且自动帮你解决模块的依赖问题。一般情况下用setuptools给安装的模块会自动放到一个后缀是.egg的目录里。<br />
在Windows里，easy_install这个命令在python安装目录下的scripts里面，所以需要把scripts加到环境变量的PATH里，这样用起来就更方便，linux下不需要注意这个问题。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/02/29/Linux-python2-6-python2-7-in-a-system/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/02/29/Linux-python2-6-python2-7-in-a-system/" class="post-title-link" itemprop="url">Linux系统：Python2.6和Python2.7同时存在</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-02-29 21:01:19" itemprop="dateCreated datePublished" datetime="2016-02-29T21:01:19+08:00">2016-02-29</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>CENTOS 6.X 系列默认安装的 Python 2.6 ，目前开发中主要是使用 Python 2.7 ，这两个版本之间还是有不少差异的，程序在 Python 2.6 下经常会出问题。 比如： <code>re.sub</code> 函数 ，2.7 支持 <code>flags</code> 参数，而 2.6 却不支持。 所以，打算安装 Python 2.7 来运行 Flask 应用程序，但 2.6 不能删除，因为系统对它有依赖。</p>
<h2 id="1-安装-sqlite-devel"><a class="markdownIt-Anchor" href="#1-安装-sqlite-devel"></a> 1、安装 sqlite-devel</h2>
<p>因为 Flask 应用程序可能使用能 Sqlite 数据库，所以这个得装上（之前因为没装这个，导致 Python 无法导入 sqlite3 库。 当然，也可以从源码编译安装。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">yum install sqlite-devel -y</span><br></pre></td></tr></table></figure>
<h2 id="2-安装-python-27"><a class="markdownIt-Anchor" href="#2-安装-python-27"></a> 2、安装 Python 2.7</h2>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">wget https://www.python.org/ftp/python/2.7.8/Python-2.7.8.tgz</span><br><span class="line">tar xf Python-2.7.8.tgz</span><br><span class="line">cd Python-2.7.8</span><br><span class="line">./configure --prefix=/usr/local</span><br><span class="line">make &amp;&amp; make install</span><br></pre></td></tr></table></figure>
<p>安装成功之后，你可以在 <code>/usr/local/bin/python2.7</code> 找到 Python 2.7。</p>
<h2 id="3-安装-setuptools-pip"><a class="markdownIt-Anchor" href="#3-安装-setuptools-pip"></a> 3、安装 setuptools + pip</h2>
<p>这里需要注意，一定要使用 python2.7 来执行相关命令。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"># First get the setup script for Setuptools:</span><br><span class="line">wget https://bitbucket.org/pypa/setuptools/raw/bootstrap/ez_setup.py</span><br><span class="line"></span><br><span class="line"># Then install it for Python 2.7 :</span><br><span class="line">python2.7 ez_setup.py</span><br><span class="line"></span><br><span class="line"># Now install pip using the newly installed setuptools:</span><br><span class="line">easy_install-2.7 pip</span><br><span class="line"></span><br><span class="line"># With pip installed you can now do things like this:</span><br><span class="line">pip2.7 install [packagename]</span><br><span class="line">pip2.7 install --upgrade [packagename]</span><br><span class="line">pip2.7 uninstall [packagename]</span><br></pre></td></tr></table></figure>
<h2 id="4-使用-virtualenv"><a class="markdownIt-Anchor" href="#4-使用-virtualenv"></a> 4、使用 virtualenv</h2>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"># Install virtualenv for Python 2.7 and create a sandbox called my27project:</span><br><span class="line">pip2.7 install virtualenv</span><br><span class="line">virtualenv-2.7 my27project</span><br><span class="line"></span><br><span class="line"># Check the system Python interpreter version:</span><br><span class="line">python --version</span><br><span class="line"># This will show Python 2.6.6</span><br><span class="line"></span><br><span class="line"># Activate the my27project sandbox and check the version of the default Python interpreter in it:</span><br><span class="line">source my27project/bin/activate</span><br><span class="line">python --version</span><br><span class="line"># This will show Python 2.7.X</span><br><span class="line">deactivate</span><br></pre></td></tr></table></figure>
<p>基本就是这些了，网上很多教程都说要做软链接，但我感觉那样做或多或少会对系统有一些未知的影响。这个方法能尽量保持系统的完整性，很多自带 Python 程序其实在头部都指定了 <code>#!/usr/bin/python</code> ，所以它们用的其实是 Python 2.6 ，而不是新安装的 Python 2.7 。 原文：<a target="_blank" rel="noopener" href="http://digwtx.duapp.com/54.html">http://digwtx.duapp.com/54.html</a> 参考： <a target="_blank" rel="noopener" href="http://toomuchdata.com/2014/02/16/how-to-install-python-on-centos/">http://toomuchdata.com/2014/02/16/how-to-install-python-on-centos/</a></p>
<ul>
<li><a target="_blank" rel="noopener" href="https://segmentfault.com/a/1190000000654227">2014年08月31日发布</a></li>
</ul>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/02/28/Bioinfo-cuffmerge-merge-genes/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/02/28/Bioinfo-cuffmerge-merge-genes/" class="post-title-link" itemprop="url">cuffmerge的结果中将相邻的基因merge成一个“基因”</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-02-28 22:23:27" itemprop="dateCreated datePublished" datetime="2016-02-28T22:23:27+08:00">2016-02-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>cuffmerge combining genes from adjacent loci?? <a target="_blank" rel="noopener" href="https://www.biostars.org/p/104551/">https://www.biostars.org/p/104551/</a> <a target="_blank" rel="noopener" href="http://seqanswers.com/forums/showthread.php?t=19772">http://seqanswers.com/forums/showthread.php?t=19772</a> <a target="_blank" rel="noopener" href="http://seqanswers.com/forums/showthread.php?t=48571">http://seqanswers.com/forums/showthread.php?t=48571</a> <a target="_blank" rel="noopener" href="https://groups.google.com/forum/#!searchin/tuxedo-tools-users/cuffmerge/tuxedo-tools-users/HSZmU%5C_KDflo/Zz5mMtZkW6wJ">https://groups.google.com/forum/#!searchin/tuxedo-tools-users/cuffmerge/tuxedo-tools-users/HSZmU\_KDflo/Zz5mMtZkW6wJ</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/02/28/Bioinfo-RNA-seq-QC-by-kraken/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/02/28/Bioinfo-RNA-seq-QC-by-kraken/" class="post-title-link" itemprop="url">RNA Seq转录组测序质控工具：Kraken</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-02-28 19:53:06" itemprop="dateCreated datePublished" datetime="2016-02-28T19:53:06+08:00">2016-02-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h2 id="kraken-a-set-of-tools-for-quality-control-and-analysis-of-high-throughput-sequence-data"><a class="markdownIt-Anchor" href="#kraken-a-set-of-tools-for-quality-control-and-analysis-of-high-throughput-sequence-data"></a> Kraken: A set of tools for quality control and analysis of high-throughput sequence data</h2>
<p>Matthew P.A. Davis, Stijn van Dongen, Cei Abreu-Goodger, Nenad Bartonicek and Anton J. Enright <em>Methods (2013) - Volume 63, Issue 1, 1 September 2013, Pages 41–49</em></p>
<h3 id="11-october-2013"><a class="markdownIt-Anchor" href="#11-october-2013"></a> <strong>11 October 2013</strong></h3>
<h3 id="version-13-274-of-kraken-has-been-released-in-the-form-of-seqimp-13-274tgz-and-reaper-13-274tgz-see-below-this-fixes-the-following-bugs"><a class="markdownIt-Anchor" href="#version-13-274-of-kraken-has-been-released-in-the-form-of-seqimp-13-274tgz-and-reaper-13-274tgz-see-below-this-fixes-the-following-bugs"></a> <strong>Version 13-274 of Kraken has been released, in the form of seqimp-13-274.tgz and reaper-13-274.tgz (see below). This fixes the following bugs:</strong></h3>
<ul>
<li>Versions 2.14 and 3.0 of R contain a bug in the RUtils package. This is affecting how the Kraken tools pass options to R. The R developers are aware and are fixing this bug. The previous version of Kraken (13-095) will not work with these versions of R.</li>
<li>Tally would hang indefinitely on very small input files, due to a missed check on unrealistically small hash sizes.</li>
</ul>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/02/26/%E6%97%A0root%E6%9D%83%E9%99%90%E5%AE%89%E8%A3%85python/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/02/26/%E6%97%A0root%E6%9D%83%E9%99%90%E5%AE%89%E8%A3%85python/" class="post-title-link" itemprop="url">无root权限安装Python</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-02-26 16:21:28" itemprop="dateCreated datePublished" datetime="2016-02-26T16:21:28+08:00">2016-02-26</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>RedHat系统自带python2.6，一些用户程序的要求python2.7以上，但是2.6版本的python不能随便删除，因为系统程序会用到。 可以在用户目录下安装python2.7 方法： 1.安装python</p>
<blockquote>
<p>Python ： <a target="_blank" rel="noopener" href="http://www.python.org/getit/">http://www.python.org/getit/</a></p>
<p>cd $HOME/local/app/</p>
<p>tar -zxvf Python-2.7.11.tgz</p>
<p>mv  Python-2.7.11 python</p>
<p>cd python</p>
<p>./configure –-prefix=<strong>$HOME/local</strong>（你准备安装的路径，可自己更改）</p>
<p>make &amp;&amp; make install</p>
</blockquote>
<p>2. 将指定路径下的lib和bin分别加入环境变量中：</p>
<blockquote>
<p>echo “PYTHONPATH=$HOME/local/lib/python2.7/site-packages” &gt;&gt; .bashrc</p>
<p>echo “export PATH=<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi><mi>O</mi><mi>M</mi><mi>E</mi><mi mathvariant="normal">/</mi><mi>l</mi><mi>o</mi><mi>c</mi><mi>a</mi><mi>l</mi><mi mathvariant="normal">/</mi><mi>a</mi><mi>p</mi><mi>p</mi><mi mathvariant="normal">/</mi><mi>p</mi><mi>y</mi><mi>t</mi><mi>h</mi><mi>o</mi><mi>n</mi><mo>:</mo></mrow><annotation encoding="application/x-tex">HOME/local/app/python:</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.08125em;">H</span><span class="mord mathnormal" style="margin-right:0.02778em;">O</span><span class="mord mathnormal" style="margin-right:0.10903em;">M</span><span class="mord mathnormal" style="margin-right:0.05764em;">E</span><span class="mord">/</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">o</span><span class="mord mathnormal">c</span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">/</span><span class="mord mathnormal">a</span><span class="mord mathnormal">p</span><span class="mord mathnormal">p</span><span class="mord">/</span><span class="mord mathnormal">p</span><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="mord mathnormal">t</span><span class="mord mathnormal">h</span><span class="mord mathnormal">o</span><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">:</span></span></span></span>PATH” &gt;&gt; .bashrc</p>
<p>source ~/.bashrc</p>
</blockquote>
<p>这样就实现了在没有root权限的情况下安装python的模块。 3. 如何在无root权限下安装python模块呢？ 按照上面的方法安装python之后，再安装setuptools，（即easy_install），从而可以使用easy_install来自动安装python包。</p>
<blockquote>
<p>tar -zxvf setuptools-0.6c11.tar.gz</p>
<p>cd setuptools-0.6c11</p>
<p>/your_python_dir/bin/python2.7 <a target="_blank" rel="noopener" href="http://setup.py">setup.py</a> build</p>
<p>/your_python_dir/bin/python2.7 <a target="_blank" rel="noopener" href="http://setup.py">setup.py</a> install</p>
</blockquote>
<p>4. 安装其他python模块/套件 先使用easy_install安装pip工具：</p>
<blockquote>
<p>easy_install pip</p>
</blockquote>
<p>在安裝python套件的时候用pip比較好，pip會兼顧各個依存的套件版本，少用easy-install。 查看目前已有的模块：</p>
<blockquote>
<p>pip list</p>
</blockquote>
<p>pip查看需要更新的package</p>
<blockquote>
<p>pip list --outdated</p>
</blockquote>
<h3 id="下面是更详细的pip介绍"><a class="markdownIt-Anchor" href="#下面是更详细的pip介绍"></a> 下面是更详细的pip介绍：</h3>
<h3 id="pip安装"><a class="markdownIt-Anchor" href="#pip安装"></a> pip安装</h3>
          <!--noindex-->
            <div class="post-button">
              <a class="btn" href="/2016/02/26/%E6%97%A0root%E6%9D%83%E9%99%90%E5%AE%89%E8%A3%85python/#more" rel="contents">
                Read more &raquo;
              </a>
            </div>
          <!--/noindex-->
        
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/02/25/Python-pbh5tools-pacbio-rawdata-convert/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/02/25/Python-pbh5tools-pacbio-rawdata-convert/" class="post-title-link" itemprop="url">python套件 pbh5tools: PacBio rawdata 格式轉換</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-02-25 21:58:46" itemprop="dateCreated datePublished" datetime="2016-02-25T21:58:46+08:00">2016-02-25</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
                  , 
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>2016-09-06更新：pbh5tools是PacBio公司早期推出的数据分析工具，其后已经被SMRTanalysis工具所替换。SMRTanalysis的安装与使用都更简单，Linux安装之后可以直接在网页里面进行数据分析操作。 ====================================== 转载自：<a target="_blank" rel="noopener" href="http://jeffreyy.pixnet.net/">http://jeffreyy.pixnet.net/</a> PacBio 是第三代定序技術 (Third generation sequencing)的技術公司，目前的最新機型為RSII，最長長度可到30kb。 出來的rawdata其檔案格式為hdf5，其中有兩種，一個是bas.h5和bax.h5，其PacBio的文件有寫到以下: “Due to the increased throughput and read lengths achieved by the PacBio® RS II upgrade, this information is now contained in one bas.h5 file and three bax.h5 files.” <a target="_blank" rel="noopener" href="http://files.pacb.com/software/instrument/2.0.0/bas.h5%20Reference%20Guide.pdf">http://files.pacb.com/software/instrument/2.0.0/bas.h5 Reference Guide.pdf</a> hdf5 格式並無法直接觀看，如果要看序列，必須要經過格式轉換，PacBio有出一個軟體為pbh5tools來進行hdf5的格式轉換。   pbh5tools是以python寫的套件，所以安裝前需要安裝其他python套件，在安裝python套件用pip比較好，pip會兼顧各個依存的套件版本，少用easy-install。 <a target="_blank" rel="noopener" href="http://blog.longwin.com.tw/2014/08/python-setup-pip-package-2014/">安裝 PIP 來管理 Python Packages</a> 除了python套件外，其他的套件還有python-devel 和 hdf5與其相關套件。   其安裝其他requirement套件指令為:</p>
<blockquote>
<p><code>sudo yum install python-pip python-devel.x86_64 hdf5-devel.x86_64 hdf5.x86_64 hdf5-mpich-devel.x86_64 hdf5-mpich.x86_64 hdf5-openmpi.x86_64 hdf5-openmpi-devel.x86_64</code> # 上述yum指令，对于没有root权限的Linux用户，该如何解决？？？（FL 2016-02-06备注） <code>sudo pip install --upgrade pip sudo pip install --upgrade nose sudo pip install --upgrade Cython sudo pip install --upgrade numpy sudo pip install --upgrade six sudo pip install --upgrade setuptools sudo pip install --upgrade pkgconfig sudo pip install --upgrade pysam sudo pip install --upgrade h5py sudo pip install --upgrade pbcore</code></p>
</blockquote>
<p>其中，pbcore需要python 2.7，故如果系統為python 2.6又無法利用yum升級上去，就必須手動安裝python 2.7，並注意需雙版本依存。 參考資料: <a target="_blank" rel="noopener" href="http://ruiaylin.github.io/2014/12/12/python%20update/">http://ruiaylin.github.io/2014/12/12/python update/</a> <a target="_blank" rel="noopener" href="https://github.com/h2oai/h2o-2/wiki/Installing-python-2.7-on-centos-6.3.-Follow-this-sequence-exactly-for-centos-machine-only">https://github.com/h2oai/h2o-2/wiki/Installing-python-2.7-on-centos-6.3.-Follow-this-sequence-exactly-for-centos-machine-only</a> <a target="_blank" rel="noopener" href="http://toomuchdata.com/2014/02/16/how-to-install-python-on-centos/">How to install Python 2.7 and Python 3.3 on CentOS 6</a> <a target="_blank" rel="noopener" href="http://tecadmin.net/install-python-2-7-on-centos-rhel/">How to Install Python 2.7.10 on CentOS/RHEL 7/6 and Fedora</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/02/25/Bioinfo-pacbio-sequencing/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/02/25/Bioinfo-pacbio-sequencing/" class="post-title-link" itemprop="url">PacBio测序：专有名词的定义</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-02-25 14:32:42" itemprop="dateCreated datePublished" datetime="2016-02-25T14:32:42+08:00">2016-02-25</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p><a target="_blank" rel="noopener" href="http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/Portal%5C_PacBio%5C_Glossary.htm">http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/Portal\_PacBio\_Glossary.htm</a></p>
<h2 id="smrt-portal-help"><a class="markdownIt-Anchor" href="#smrt-portal-help"></a> SMRT® Portal Help</h2>
<h1 id="pacific-biosciences-terminology"><a class="markdownIt-Anchor" href="#pacific-biosciences-terminology"></a> Pacific Biosciences Terminology</h1>
<h2 id="general-terminology"><a class="markdownIt-Anchor" href="#general-terminology"></a> General Terminology</h2>
<ul>
<li>SMRT® Cell: Consumable substrates comprising arrays of zero-mode waveguide nanostructures.</li>
<li>Adapters: Exogenous nucleic acids that are ligated to a nucleic acid molecule to be sequenced. For example, SMRTbell™ adapters are hairpin loops that are ligated to both ends of the double stranded DNA insert to produce a SMRTbell™ sequencing template. When adapter sequences are removed from a CCS read, the read is split into multiple subreads.</li>
<li>Movie: Real-time observation of a SMRT® Cell.</li>
<li>zero-mode waveguide (ZMW): A nanophotonic device for confining light to a small observation volume. This can be, for example, a small hole in a conductive layer whose diameter is too small to permit the propagation of light in the wavelength range used for detection. Physically part of a SMRT® Cell.</li>
<li>Sequencing ZMW: A ZMW (zero-mode waveguide) that is expected to be able to produce a sequence if it is populated with a polymerase. ZMWs used for automated SMRT Cell alignment are not considered sequencing ZMWs.</li>
<li>Run: Specifies</li>
<li>The wells and SMRT Cells to include in the sequencing run.</li>
<li>The collection and analysis protocols to use for the selected wells and cells.</li>
</ul>
<h2 id="read-terminology"><a class="markdownIt-Anchor" href="#read-terminology"></a> Read Terminology</h2>
<p><img src="http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/3Part_Reads_Def.png" alt="" /></p>
<ul>
<li>polymerase read (formerly called “read”): A sequence of nucleotides incorporated by the DNA polymerase while reading a template, such as a circular SMRTbell™ template. Polymerase reads are most useful for quality control of the instrument run. Polymerase read metrics primarily reflect movie length and other run parameters rather than insert size distribution. Polymerase reads are trimmed to include only the high quality region; they include sequences from adapters; and can further include sequence from multiple passes around a circular template.</li>
<li>subread: Each polymerase read is partitioned to form one or more subreads, which contain sequence from a single pass of a polymerase on a single strand of an insert within a SMRTbell™ template and no adapter sequences. The subreads contain the full set of quality values and kinetic measurements. Subreads are useful for applications like <em>de novo</em> assembly, resequencing, base modification analysis, and so on.</li>
<li>circular consensus (CCS) read: The consensus sequence determined using subreads taken from a single ZMW. This is not aligned against a reference sequence. In contrast to Reads of Insert, CCS reads require at least two full-pass subreads from the insert.</li>
<li>read of insert: Represents the highest quality single sequence for an insert, regardless of the number of passes. For example, if your template received one-and-a-half subreads, that information will be combined into a Read of Insert. CCS is an example of a special case where at least two full subreads are collected for an insert. Reads of Insert give the most accurate estimate of the length of the insert sequence loaded onto a SMRT® Cell. For long templates, Reads of Insert may be the same as Polymerase Reads.</li>
</ul>
<h2 id="read-length-terminology"><a class="markdownIt-Anchor" href="#read-length-terminology"></a> Read Length Terminology</h2>
<p><img src="http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/Mapped_Length.png" alt="" /></p>
<ul>
<li>mapped polymerase read length: The total number of bases along a read from the first adapter or aligned subread to the last adapter or aligned subread. Approximates the sequence produced by a polymerase in a ZMW.</li>
<li>mapped subread length: The length of the subread alignment to a target reference sequence. This does not include the adapter sequence.</li>
<li>polymerase read length: The total number of bases produced from a ZMW after trimming. This may include the adapter sequence.</li>
</ul>
<h2 id="primary-analysis-terminology"><a class="markdownIt-Anchor" href="#primary-analysis-terminology"></a> Primary Analysis Terminology</h2>
<ul>
<li>
<p>Primary analysis protocol: Specifies signal processing of the movie, base calling of the traces/pulses, and quality assessment of the base calls. Primary analysis is always performed on the instrument.</p>
</li>
<li>
<p>Adapter Screening: Annotates adapter read locations. Used to break a read into subreads during secondary analysis mapping and Circular Consensus.</p>
</li>
<li>
<p>High Quality Region Screening: Annotates the high quality sequencing regions of a read to be used during Raw Read Trimming.</p>
</li>
<li>
<p>Insert Screening: Annotates insert DNA regions in the Polymerase Read.</p>
</li>
<li>
<p>Quality Value Assignment: A prediction of the error probability of a basecall.</p>
</li>
<li>
<p>Quality Value (QV): The total probability that the basecall is an insertion or substitution or is preceded by a deletion. QV = -10 * log10(p). For example, QV 20 is 99% accurate, QV 30 is 99.9% accurate, and QV 50 is 99.999% accurate.</p>
</li>
<li>
<p>Insertion QV: The probability that the basecall is an insertion with respect to the true sequence.</p>
</li>
<li>
<p>Deletion QV: The probability that a deletion error occurred before the current base.</p>
</li>
<li>
<p>Substitution QV: The probability that the basecall is a substitution.</p>
</li>
<li>
<p>raw read trimming: Extraction of high quality regions from an unfiltered read. Trimming of an unfiltered read produces a polymerase read.</p>
</li>
<li>
<p>Read Quality Assignment: A trained prediction of a read’s mapped accuracy based on its pulse and base file characteristics (peak signal-to-noise ratio, average base QV, interpulse distance, and so on). This is used during secondary analysis filtering.</p>
</li>
</ul>
<h2 id="secondary-analysis-terminology"><a class="markdownIt-Anchor" href="#secondary-analysis-terminology"></a> Secondary Analysis Terminology</h2>
<ul>
<li>
<p>Secondary analysis protocol: Specifies how to</p>
</li>
<li>
<p>Align a group of reads to a reference sequence to produce a consensus sequence.</p>
</li>
<li>
<p>Assemble a set of reads into contigs to produce a <em>de novo</em> sequence.</p>
</li>
<li>
<p>Identify insertions, deletions, and SNPs.</p>
</li>
<li>
<p>Evaluate consensus quality and quality of the instrument run.</p>
</li>
<li>
<p>Consensus: Generation of a consensus sequence from multiple-sequence alignment.</p>
</li>
<li>
<p><em>De Novo</em> Assembly: Assembly of all subreads without a reference sequence.</p>
</li>
<li>
<p>Filtering: Removes reads that do not meet the Read Quality and Read Length parameters set by the user. The current default filtering parameters defined by Pacific Biosciences are:</p>
</li>
<li>
<p>Read Quality ≥ .75 (as of SMRT Analysis v2.1)</p>
</li>
<li>
<p>Read Length ≥ 50 bases</p>
</li>
<li>
<p>Mapping: Local alignment of a read or subread to a reference sequence.</p>
</li>
</ul>
<h2 id="accuracy-terminology"><a class="markdownIt-Anchor" href="#accuracy-terminology"></a> Accuracy Terminology</h2>
<ul>
<li>circular consensus accuracy: Accuracy based on multiple sequencing passes around a single circular template molecule.</li>
<li>consensus accuracy: Accuracy based on aligning multiple sequencing reads or subreads together, optionally with a reference sequence.</li>
<li>polymerase read quality: A trained prediction of a read’s mapped accuracy based on its pulse and base file characteristics (peak signal-to-noise ratio, average base QV, inter-pulse distance, and so on).</li>
<li>Subread Accuracy: The post-mapping accuracy of the basecalls.</li>
<li>Formula: [1 - (errors/subread length)], where errors = number of deletions + insertions + substitutions.</li>
</ul>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/02/22/linux-using-parallel-for-a-task/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/02/22/linux-using-parallel-for-a-task/" class="post-title-link" itemprop="url">Linux用Parallel命令并行处理命令</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-02-22 16:39:16" itemprop="dateCreated datePublished" datetime="2016-02-22T16:39:16+08:00">2016-02-22</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-07 17:58:21" itemprop="dateModified" datetime="2020-12-07T17:58:21+08:00">2020-12-07</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>你是否曾经有过要计算一个非常大的数据(几百GB)的需求？或在里面搜索，或其它操作——一些无法并行的操作。数据专家们，我是在对你们说。你可能有一个4核或更多核的CPU，但我们合适的工具，例如 <strong>grep</strong>, <strong>bzip2</strong>, <strong>wc</strong>, <strong>awk</strong>, <strong>sed</strong>等等，都是单线程的，只能使用一个CPU内核。 借用卡通人物Cartman的话，“如何我能使用这些内核”? 要想让Linux命令使用所有的CPU内核，我们需要用到<a target="_blank" rel="noopener" href="https://www.gnu.org/software/parallel/" title="GNU Parallel">GNU Parallel</a>命令，它让我们所有的CPU内核在单机内做神奇的map-reduce操作，当然，这还要借助很少用到的**–pipes** 参数(也叫做**–spreadstdin**)。这样，你的负载就会平均分配到各CPU上，真的。</p>
<h3 id="bzip2"><a class="markdownIt-Anchor" href="#bzip2"></a> BZIP2</h3>
<p>bzip2是比gzip更好的压缩工具，但它很慢！别折腾了，我们有办法解决这问题。 以前的做法：</p>
<p>cat bigfile.bin  bzip2 --best &gt; compressedfile.bz2</p>
<p>现在这样：</p>
<p>cat bigfile.bin  parallel --pipe --recend ‘’ -k bzip2 --best &gt; compressedfile.bz2</p>
<p>尤其是针对bzip2，GNU parallel在多核CPU上是超级的快。你一不留神，它就执行完成了。</p>
<h3 id="grep"><a class="markdownIt-Anchor" href="#grep"></a> GREP</h3>
<p>如果你有一个非常大的文本文件，以前你可能会这样：</p>
<p>grep pattern bigfile.txt</p>
<p>现在你可以这样：</p>
<p>cat bigfile.txt  parallel  --pipe grep ‘pattern’</p>
<p>或者这样：</p>
<p>cat bigfile.txt  parallel --block 10M --pipe grep ‘pattern’</p>
<p>这第二种用法使用了 <strong>–block 10M</strong>参数，这是说每个内核处理1千万行——你可以用这个参数来调整每个CUP内核处理多少行数据。</p>
<h3 id="awk"><a class="markdownIt-Anchor" href="#awk"></a> AWK</h3>
<p>下面是一个用awk命令计算一个非常大的数据文件的例子。 常规用法：</p>
<p>cat rands20M.txt  awk ‘{s+=$1} END {print s}’</p>
<p>现在这样：</p>
<p>cat rands20M.txt  parallel --pipe awk \‘{s+=\$1} END {print s}\’  awk ‘{s+=$1} END {print s}’</p>
<p>这个有点复杂：parallel命令中的**–pipe**参数将cat输出分成多个块分派给awk调用，形成了很多子计算操作。这些子计算经过第二个管道进入了同一个awk命令，从而输出最终结果。第一个awk有三个反斜杠，这是GNU parallel调用awk的需要。</p>
<h3 id="wc"><a class="markdownIt-Anchor" href="#wc"></a> WC</h3>
<p>想要最快的速度计算一个文件的行数吗？ 传统做法：</p>
<p>wc -l bigfile.txt</p>
<p>现在你应该这样：</p>
<p>cat bigfile.txt  parallel  --pipe wc -l  awk ‘{s+=$1} END {print s}’</p>
<p>非常的巧妙，先使用parallel命令‘mapping’出大量的<code>wc -l</code>调用，形成子计算，最后通过管道发送给awk进行汇总。</p>
<h3 id="sed"><a class="markdownIt-Anchor" href="#sed"></a> SED</h3>
<p>想在一个巨大的文件里使用sed命令做大量的替换操作吗？ 常规做法：</p>
<p>sed s<sup>old</sup>new^g bigfile.txt</p>
<p>现在你可以：</p>
<p>cat bigfile.txt  parallel --pipe sed s<sup>old</sup>new^g</p>
<p>…然后你可以使用管道把输出存储到指定的文件里。 转载自：<a target="_blank" rel="noopener" href="http://www.vaikan.com/use-multiple-cpu-cores-with-your-linux-commands/">http://www.vaikan.com/use-multiple-cpu-cores-with-your-linux-commands/</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2016/01/28/Linux-How-to-use-tmux-in-Linux/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/fl-photo-202109.jpg">
      <meta itemprop="name" content="FENG Lei">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FENG's Notes">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2016/01/28/Linux-How-to-use-tmux-in-Linux/" class="post-title-link" itemprop="url">Tmux的使用</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2016-01-28 18:00:13" itemprop="dateCreated datePublished" datetime="2016-01-28T18:00:13+08:00">2016-01-28</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-12-18 11:42:08" itemprop="dateModified" datetime="2020-12-18T11:42:08+08:00">2020-12-18</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>介绍 <a target="_blank" rel="noopener" href="https://wiki.freebsdchina.org/software/t/tmux">tmux</a>是一个优秀的终端复用软件，即使非正常掉线，也能保证当前的任务运行，这一点对于 远程SSH访问特别有用，网络不好的情况下仍然能保证工作现场不丢失!此外，tmux完全使用键盘 控制窗口，实现窗口的切换功能。 简单地说，tmux对于我主要有两个功能（这应该也是tmux的主要功能）:</p>
<ol>
<li>split窗口。可以在一个terminal下打开多个终端，也可以对当前屏幕进行各种split，即可以 同时打开多个显示范围更小的终端。</li>
<li>在使用SSH的环境下，避免网络不稳定，导致工作现场的丢失。想象以下场景， 你在执行一条命令的过程中，由于网络不稳定，SSH连接断开了。这个时候，你就不知道之前 的那条命令是否执行成功。如果此时你打开了很多文件，进入了较深层次的目录，由于网络 不稳定，SSH连接断开。重新连接以后，你又不得不重新打开那些文件，进入那个深层次的 目录。如果使用了tmux，重新连接以后，就可以直接回到原来的工作环境，不但提高了工作 效率，还降低了风险，增加了安全性。</li>
</ol>
<h3 id="安装"><a class="markdownIt-Anchor" href="#安装"></a> 安装</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sudo apt-get install tmux</span><br></pre></td></tr></table></figure>
<p>安装完成后输入命令tmux即可打开软件，界面十分简单，类似一个下方带有状态栏的终端控制台； 不出意外，这时候你会跟我第一次一样，觉得tmux没什么牛逼的。</p>
<h3 id="会话窗口面板"><a class="markdownIt-Anchor" href="#会话窗口面板"></a> 会话，窗口，面板</h3>
<p>根据tmux的定义，在开启了tmux服务器后，会首先创建一个会话，而这个会话则会首先创建一个 窗口，其中仅包含一个面板；也就是说，这里看到的所谓终端控制台应该称作tmux的一个面板， 虽然其使用方法与终端控制台完全相同。 tmux使用C/S模型构建，主要包括以下单元模块：</p>
<ul>
<li>server服务器。输入tmux命令时就开启了一个服务器。</li>
<li>session会话。一个服务器可以包含多个会话</li>
<li>window窗口。一个会话可以包含多个窗口。</li>
<li>pane面板。一个窗口可以包含多个面板。</li>
</ul>
<h3 id="初体验"><a class="markdownIt-Anchor" href="#初体验"></a> 初体验</h3>
<p>我们先来看一张效果图，再来学习tmux的配置和使用我一般都是这么用，因为我的笔记本 只有13寸:-) <img src="http://mingxinglai.com/cn/image/tmux-small.jpg" alt="tmux-small" /> 当然，做为一个准IT男，我还有一个很大的屏幕，所以我偶尔也这么用： <img src="http://mingxinglai.com/cn/image/tmux-large.jpg" alt="tmux-large" /> 你所看到的只是X个会话中的一个会话，且是该会话下面的Y个窗口中的一个窗口。是的，你在一 个窗口里就可以拥有这么多个终端，这就是tmux的主要功能。好了，下面是tmux的常用按键，你 先自己玩一下再往下看吧。</p>
<h3 id="常用按键"><a class="markdownIt-Anchor" href="#常用按键"></a> 常用按键</h3>
<p>这里需要说明一点的是，tmux的任何指令，都包含一个前缀，也就是说，你按了前缀(一组按键， 默认是Ctrl+b)以后，系统才知道你接下来的指令是发送给tmux的。</p>
<ul>
<li>C-b ? 显示快捷键帮助</li>
<li>C-b C-o 调换窗口位置，类似与vim 里的C-w</li>
<li>C-b 空格键 采用下一个内置布局</li>
<li>C-b ! 把当前窗口变为新窗口</li>
<li>C-b &quot; 横向分隔窗口</li>
<li>C-b % 纵向分隔窗口</li>
<li>C-b q 显示分隔窗口的编号</li>
<li>C-b o 跳到下一个分隔窗口</li>
<li>C-b 上下键 上一个及下一个分隔窗口</li>
<li>C-b C-方向键 调整分隔窗口大小</li>
<li>C-b c 创建新窗口</li>
<li>C-b 0~9 选择几号窗口</li>
<li>C-b c 创建新窗口</li>
<li>C-b n 选择下一个窗口</li>
<li>C-b l 切换到最后使用的窗口</li>
<li>C-b p 选择前一个窗口</li>
<li>C-b w 以菜单方式显示及选择窗口</li>
<li>C-b t 显示时钟</li>
<li>C-b ; 切换到最后一个使用的面板</li>
<li>C-b x 关闭面板</li>
<li>C-b &amp; 关闭窗口</li>
<li>C-b s 以菜单方式显示和选择会话</li>
<li>C-b d 退出tumx，并保存当前会话，这时，tmux仍在后台运行，可以通过tmux attach进入 到指定的会话</li>
</ul>
<h3 id="配置"><a class="markdownIt-Anchor" href="#配置"></a> 配置</h3>
<p>参考：<a target="_blank" rel="noopener" href="http://mingxinglai.com/cn/2012/09/tmux/">http://mingxinglai.com/cn/2012/09/tmux/</a></p>
<h3 id="附安装记录"><a class="markdownIt-Anchor" href="#附安装记录"></a> 附安装记录。</h3>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br></pre></td><td class="code"><pre><span class="line">$ sudo yum install tmux</span><br><span class="line">[sudo] password for user: </span><br><span class="line">Loaded plugins: fastestmirror</span><br><span class="line">Determining fastest mirrors</span><br><span class="line">epel/x86_64/metalink                                                                                                                                                                                            | 9.1 kB  00:00:00     </span><br><span class="line"> * base: ftp.cuhk.edu.hk</span><br><span class="line"> * epel: hkg.mirror.rackspace.com</span><br><span class="line"> * extras: ftp.cuhk.edu.hk</span><br><span class="line"> * updates: mirrors.icidc.com</span><br><span class="line">base                                                                                                                                                                                                            | 3.6 kB  00:00:00     </span><br><span class="line">epel                                                                                                                                                                                                            | 4.7 kB  00:00:00     </span><br><span class="line">extras                                                                                                                                                                                                          | 2.9 kB  00:00:00     </span><br><span class="line">ius                                                                                                                                                                                                             | 1.3 kB  00:00:00     </span><br><span class="line">ius-testing                                                                                                                                                                                                     | 1.3 kB  00:00:00     </span><br><span class="line">updates                                                                                                                                                                                                         | 2.9 kB  00:00:00     </span><br><span class="line">(1/5): updates/7/x86_64/primary_db                                                                                                                                                                              | 4.0 MB  00:00:00     </span><br><span class="line">(2/5): epel/x86_64/updateinfo                                                                                                                                                                                   | 1.0 MB  00:00:00     </span><br><span class="line">(3/5): ius-testing/x86_64/primary                                                                                                                                                                               | 1.4 kB  00:00:01     </span><br><span class="line">(4/5): epel/x86_64/primary_db                                                                                                                                                                                   | 6.9 MB  00:00:02     </span><br><span class="line">(5/5): ius/x86_64/primary                                                                                                                                                                                       | 129 kB  00:00:02     </span><br><span class="line">ius                                                                                                                                                                                                                            584/584</span><br><span class="line">ius-testing                                                                                                                                                                                                                        2/2</span><br><span class="line">Resolving Dependencies</span><br><span class="line">--&gt; Running transaction check</span><br><span class="line">---&gt; Package tmux.x86_64 0:1.8-4.el7 will be installed</span><br><span class="line">--&gt; Processing Dependency: libevent-2.0.so.5()(64bit) for package: tmux-1.8-4.el7.x86_64</span><br><span class="line">--&gt; Running transaction check</span><br><span class="line">---&gt; Package libevent.x86_64 0:2.0.21-4.el7 will be installed</span><br><span class="line">--&gt; Finished Dependency Resolution</span><br><span class="line"></span><br><span class="line">Dependencies Resolved</span><br><span class="line"></span><br><span class="line">=======================================================================================================================================================================================================================================</span><br><span class="line"> Package                                                 Arch                                                  Version                                                       Repository                                           Size</span><br><span class="line">=======================================================================================================================================================================================================================================</span><br><span class="line">Installing:</span><br><span class="line"> tmux                                                    x86_64                                                1.8-4.el7                                                     base                                                243 k</span><br><span class="line">Installing for dependencies:</span><br><span class="line"> libevent                                                x86_64                                                2.0.21-4.el7                                                  base                                                214 k</span><br><span class="line"></span><br><span class="line">Transaction Summary</span><br><span class="line">=======================================================================================================================================================================================================================================</span><br><span class="line">Install  1 Package (+1 Dependent package)</span><br><span class="line"></span><br><span class="line">Total download size: 457 k</span><br><span class="line">Installed size: 1.3 M</span><br><span class="line">Is this ok [y/d/N]: y</span><br><span class="line">Downloading packages:</span><br><span class="line">(1/2): libevent-2.0.21-4.el7.x86_64.rpm                                                                                                                                                                         | 214 kB  00:00:00     </span><br><span class="line">(2/2): tmux-1.8-4.el7.x86_64.rpm                                                                                                                                                                                | 243 kB  00:00:00     </span><br><span class="line">---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------</span><br><span class="line">Total                                                                                                                                                                                                  1.1 MB/s | 457 kB  00:00:00     </span><br><span class="line">Running transaction check</span><br><span class="line">Running transaction test</span><br><span class="line">Transaction test succeeded</span><br><span class="line">Running transaction</span><br><span class="line">  Installing : libevent-2.0.21-4.el7.x86_64                                                                                                                                                                                        1/2 </span><br><span class="line">  Installing : tmux-1.8-4.el7.x86_64                                                                                                                                                                                               2/2 </span><br><span class="line">  Verifying  : tmux-1.8-4.el7.x86_64                                                                                                                                                                                               1/2 </span><br><span class="line">  Verifying  : libevent-2.0.21-4.el7.x86_64                                                                                                                                                                                        2/2 </span><br><span class="line"></span><br><span class="line">Installed:</span><br><span class="line">  tmux.x86_64 0:1.8-4.el7                                                                                                                                                                                                              </span><br><span class="line"></span><br><span class="line">Dependency Installed:</span><br><span class="line">  libevent.x86_64 0:2.0.21-4.el7                                                                                                                                                                                                       </span><br><span class="line"></span><br><span class="line">Complete!</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/14/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/14/">14</a><span class="page-number current">15</span><a class="page-number" href="/page/16/">16</a><span class="space">&hellip;</span><a class="page-number" href="/page/20/">20</a><a class="extend next" rel="next" href="/page/16/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="FENG Lei"
      src="/images/fl-photo-202109.jpg">
  <p class="site-author-name" itemprop="name">FENG Lei</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">194</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">11</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">1</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="mailto:lfeng0703@hotmail.com" title="E-Mail → mailto:lfeng0703@hotmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2024</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">FENG Lei</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://pisces.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  

</body>
</html>
