<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2014/12/17/%E5%9F%BA%E4%BA%8Eillumina%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%EF%BC%8C%E4%BD%BF%E7%94%A8jellyfish%E4%BC%B0%E8%AE%A1%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%A4%A7%E5%B0%8F/">基于illumina测序数据，使用Jellyfish估计基因组大小</a></h1><div class="post-meta">2014-12-17</div><div class="post-content"><p>一、数据 ① illumina测序数据：（细菌<strong><em>Staphylococcus aureus，</em></strong>基因组约3M，180bp文库，101PE测序，1294104条reads，测序碱基数据量约为130Mb，即基因组覆盖度45X。） <a target="_blank" rel="noopener" href="http://gage.cbcb.umd.edu/data/Staphylococcus/_aureus/Data.original/frag/_1.fastq.gz">http://gage.cbcb.umd.edu/data/Staphylococcus\_aureus/Data.original/frag\_1.fastq.gz</a> <a target="_blank" rel="noopener" href="http://gage.cbcb.umd.edu/data/Staphylococcus/_aureus/Data.original/frag/_2.fastq.gz">http://gage.cbcb.umd.edu/data/Staphylococcus\_aureus/Data.original/frag\_2.fastq.gz</a> ②  jellyfish 软件：（用来统计kmer频数。） <a href="ftp://ftp.genome.umd.edu/pub/jellyfish/jellyfish-2.1.4.tar.gz">ftp://ftp.genome.umd.edu/pub/jellyfish/jellyfish-2.1.4.tar.gz</a> ③jellyplot.pl软件：（用来画kmer图，以及估计基因组大小。） <a target="_blank" rel="noopener" href="https://github.com/josephryan/estimate/_genome/_size.pl">https://github.com/josephryan/estimate\_genome\_size.pl</a> ④fastx toolkit：（用来转换fastq和fasta格式文件。） hannonlab.cshl.edu/<strong>fastx</strong>_<strong>toolkit</strong>/   —————— 下面是实际运行代码： cat frag_1.fastq frag_2.fastq &gt; all.fastq fastq_to_fasta -i all.fastq -o all.fasta jellyfish count -m 17 -s 100M -t 10 -C all.fasta jellyfish histo mer_counts.jf -o out.2.txt jellyplot.pl out.2.txt estimate_genome_size.pl -kmer 17 -peak 15 -fastq all.fastq.gz   结果：（预估基因组为7M，比实际大了一倍。可能是原始reads没有校正和过滤的原因。） $ estimate_genome_size.pl -kmer 17 -peak 15 -fastq all.fastq.gz # running version 0.04 of estimate_genome_size.pl # run with this command: estimate_genome_size.pl TOTAL_NTS: 131998608 Estimated Coverage: 17.7906976744186 Estimated Genome Size: 7419529.6 <a target="_blank" rel="noopener" href="http://genehub.files.wordpress.com/2014/12/qqe688aae59c9620141217130925.jpg"><img src="http://genehub.files.wordpress.com/2014/12/qqe688aae59c9620141217130925.jpg?w=300" alt="QQ截圖20141217130925"></a> 使用校正之后的数据，（使用quake.py进行校正<a target="_blank" rel="noopener" href="http://gage.cbcb.umd.edu/data/Staphylococcus/_aureus%EF%BC%89">http://gage.cbcb.umd.edu/data/Staphylococcus\_aureus）</a> 估计基因组大小为： # running version 0.04 of estimate_genome_size.pl # run with this command: /home/fenglei/perl5/bin/estimate_genome_size.pl TOTAL_NTS: 59609610 Estimated Coverage: 15.3684210526316 Estimated Genome Size: 3878707.5 这次距离实际情况比较近。     =================== 附：jellyfish 参数意义 jellyfish count -m 17 -s 100M -t 10 -C all.fasta -m是kmer长度； -s是预估哈希表的大小，即G+G*c*e*k。G是Genome Size；c是coverage（genome survey测序通常低于100x）；e是测序错误率（illumina为1%）；k是kmer大小。 -C表示考虑DNA正义与反义链，遇到反义kmer时，计入正义kmer频数中。</p></div><p class="readmore"><a href="/2014/12/17/%E5%9F%BA%E4%BA%8Eillumina%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%EF%BC%8C%E4%BD%BF%E7%94%A8jellyfish%E4%BC%B0%E8%AE%A1%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%A4%A7%E5%B0%8F/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/12/16/%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85%E8%BD%AF%E4%BB%B6%E8%AF%84%E4%BC%B0/">基因组组装软件评估</a></h1><div class="post-meta">2014-12-16</div><div class="post-content"><p>比较下面8个软件，结论：AllPATHS-LG性能最优，组装的contig和scaffold比较长，错误率低；SOAP denovo组装得到的contig和scaffold几乎是几个软件里最长的，但是错误率也相对较高！   ABySS (Simpson et al. 2009) • ALLPATHS-LG (Gnerre et al. 2011) • Bambus2 (Koren et al. 2011) (<a target="_blank" rel="noopener" href="http://www.cbcb.umd.edu/software/">http://www.cbcb.umd.edu/software/</a> bambus). • CABOG (Miller et al. 2008) • MSR-CA (<a target="_blank" rel="noopener" href="http://www.genome.umd.edu/MSR/_CA/_MANUAL.htm">http://www.genome.umd.edu/MSR\_CA\_MANUAL.htm</a>) • SGA (Simpson and Durbin 2012) • SOAPdenovo (Li et al. 2010b) • Velvet (Zerbino and Birney 2008) ———————————— <a target="_blank" rel="noopener" href="http://gage.cbcb.umd.edu/index.html">http://gage.cbcb.umd.edu/index.html</a></p></div><p class="readmore"><a href="/2014/12/16/%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85%E8%BD%AF%E4%BB%B6%E8%AF%84%E4%BC%B0/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/12/12/importerror-no-module-named-toolz/">Python报错：ImportError: No module named toolz</a></h1><div class="post-meta">2014-12-12</div><div class="post-content"><p>seq_crumbs中的pair_matcher是fastq数据分析中一个小工具，依赖Python和biopython。初次运行报错： [fenglei@localhost new]$ pair_matcher -h Traceback (most recent call last): File “/usr/local/bin/pair_matcher”, line 26, in <module> from crumbs.pairs import match_pairs File “/usr/local/lib/python2.7/site-packages/crumbs/pairs.py”, line 18, in <module> from toolz import first ImportError: No module named toolz</p></div><p class="readmore"><a href="/2014/12/12/importerror-no-module-named-toolz/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/10/14/rna-fasta-dna/">如何将RNA格式的fasta文件转换为DNA格式</a></h1><div class="post-meta">2014-10-14</div><div class="post-content"><p>从miRBase数据库下载的pre-miRNA序列是RNA格式的，如下所示。 &gt;cel-mir-37 MI0000008 Caenorhabditis elegans miR-37 stem-loop UUCUAGAAACCCUUGGACCAGUGUGGGUGUCCGUUGCGGUGCUACAUUCUCUAAUCUGUA UCACCGGGUGAACACUUGCAGUGGUCCUCGUGGUUUCU 因数据分析的需要将其转换为DNA格式。</p></div><p class="readmore"><a href="/2014/10/14/rna-fasta-dna/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/09/15/tophat-attributeerror-nonetype-object-has-no-attribute-group/">Tophat: AttributeError: 'NoneType' object has no attribute 'group'</a></h1><div class="post-meta">2014-09-15</div><div class="post-content"><p>I have installed boost (version 1.38), Eigen (eigen-eigen-1306d75b4a21.tar.gz) and Samtools(version 1.0). [Ref:<a target="_blank" rel="noopener" href="http://ccb.jhu.edu/software/tophat/tutorial.shtml/]">http://ccb.jhu.edu/software/tophat/tutorial.shtml\]</a> Then I downlaod Tophat 2.0.12 (binary), But error occured when I test the pipeline on a simple test data set. tophat -r 20 test_ref reads_1.fq reads_2.fq</p></div><p class="readmore"><a href="/2014/09/15/tophat-attributeerror-nonetype-object-has-no-attribute-group/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/09/12/vim-error-while-loading-shared-libraries-libperl-so/">vim: error while loading shared libraries: libperl.so</a></h1><div class="post-meta">2014-09-12</div><div class="post-content"><p>[fenglei@localhost ~]$ <strong>vi .bashrc</strong> vim: error while loading shared libraries: libperl.so: cannot open shared object file: No such file or directory 出现类似错误是由于缺少相关的库文件（*.so文件）造成的 可能的原因是版本不匹配或者没有安装相应的rpm文件首先要执行的文件在那里，用which命令</p></div><p class="readmore"><a href="/2014/09/12/vim-error-while-loading-shared-libraries-libperl-so/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/09/08/cufflinks%E5%AE%89%E8%A3%85%E5%89%8D%E6%8F%90%E4%B9%8B%E4%B8%80%EF%BC%9A%E5%AE%89%E8%A3%85boost/">cufflinks安装前提之一：安装boost</a></h1><div class="post-meta">2014-09-08</div><div class="post-content"><p>转录组数据分析软件cufflinks的安装，前提是linux平台必须装有boost 1.47以上的版本！安装步骤如下：</p></div><p class="readmore"><a href="/2014/09/08/cufflinks%E5%AE%89%E8%A3%85%E5%89%8D%E6%8F%90%E4%B9%8B%E4%B8%80%EF%BC%9A%E5%AE%89%E8%A3%85boost/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/09/02/linux%E5%AE%89%E8%A3%85mysql/">linux安装mysql</a></h1><div class="post-meta">2014-09-02</div><div class="post-content"><p>本文转自shenwei.me</p></div><p class="readmore"><a href="/2014/09/02/linux%E5%AE%89%E8%A3%85mysql/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/08/30/cant-locate-iozlib-pm/">解决Can't locate CPAN.pm in @INC的问题</a></h1><div class="post-meta">2014-08-30</div><div class="post-content"><p>Linux <strong>RedHat</strong> 的系统，解决“Can’t locate CPAN.pm in @INC”的问题。 参考博文：<a target="_blank" rel="noopener" href="http://blog.sina.com.cn/s/blog/_645e2f0b01012npz.html">http://blog.sina.com.cn/s/blog\_645e2f0b01012npz.html</a> 实际步骤： $ wget <a href="ftp://ftp.muug.mb.ca/mirror/redhat/redhat/linux/6.2/cpan/i386/perl-CPAN-1.52-6.i386.rpm">ftp://ftp.muug.mb.ca/mirror/redhat/redhat/linux/6.2/cpan/i386/perl-CPAN-1.52-6.i386.rpm</a> $ su root # rpm -ivh perl-CPAN-1.52-6.i386.rpm 至此，CPAN模块就安装好了。 可运行 $ rpm -q perl-CPAN 命令进行检验。</p></div><p class="readmore"><a href="/2014/08/30/cant-locate-iozlib-pm/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/08/22/linux-ntfs/">Linux系统无法挂载格式为ntfs格式的移动硬盘的解决办法</a></h1><div class="post-meta">2014-08-22</div><div class="post-content"><p><strong>Linux**</strong>系统无法挂载格式为**<strong>ntfs**</strong>格式的移动硬盘的解决办法** 今天想把网上下载下来的一些软件包用移动硬盘挂载到Red Hat Enterprise Workstation系统计算机器上，可是怎么也挂载不上去，网上查了资料，原来默认 只识别 win 下的 vfat ，不识别 ntfs 格式的。我的移动硬盘就是ntfs格式的。再查了一些资料并结合自己的实际操作，最终解决了不能挂载ntfs移动硬盘的问题。具体步骤如下：</p></div><p class="readmore"><a href="/2014/08/22/linux-ntfs/">Read More</a></p></div><nav class="page-navigator"><a class="extend prev" rel="prev" href="/page/14/">Previous</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/13/">13</a><a class="page-number" href="/page/14/">14</a><span class="page-number current">15</span><a class="page-number" href="/page/16/">16</a><a class="extend next" rel="next" href="/page/16/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/06/08/Linux-PKG-CONFIG-PATH/">Linux | 认识 PKG_CONFIG_PATH</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/03/jcvi%E5%AE%89%E8%A3%85%E6%8A%A5%E9%94%99%E4%BB%A5%E5%8F%8A%E8%A7%A3%E5%86%B3%E5%8A%9E%E6%B3%95/">jcvi 安装与使用，报错以及解决办法</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/06/02/Bioinfo-Installation-of-OrthoMCL-pipeline/">Bioinfo | Installation of OrthoMCL-pipeline</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/05/28/Bioinfo-Using-Circos-for-genome-visualization/">Bioinfo | Using Circos for genome visualization</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/23/Bioinfo-Repeat-sequence-analysis-by-RepeatExplorer2-with-TAREAN-Tandem-Repeat-Analyzer/">Bioinfo | Repeat sequence analysis by RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer)</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/16/Bioinfo-understanding-sequence-identity/">Bioinfo | 怎么理解序列一致性（Sequence identity）</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/10/Bioinfo-Extracting-sequence-by-names/">Bioinfo | 依据序列名字列表从 fasta 文件提取序列</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/05/Proteomics-Serotonylation-A-novel-histone-H3-marker/">Proteomics | 组蛋白五羟色胺化修饰 H3Q5ser 及其生物学功能</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>