<!DOCTYPE html><html lang="en"><head><meta http-equiv="content-type" content="text/html; charset=utf-8"><meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport"><meta content="yes" name="apple-mobile-web-app-capable"><meta content="black-translucent" name="apple-mobile-web-app-status-bar-style"><meta content="telephone=no" name="format-detection"><meta name="description" content=""><title>FENG's notes</title><link rel="stylesheet" type="text/css" href="/css/style.css?v=0.0.0"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/normalize.css/normalize.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/pure-min.min.css"><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/purecss/build/grids-responsive-min.css"><link rel="stylesheet" href="//cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/jquery/dist/jquery.min.js"></script><link rel="icon" mask="" sizes="any" href="/favicon.ico"><link rel="Shortcut Icon" type="image/x-icon" href="/favicon.ico"><link rel="apple-touch-icon" href="/apple-touch-icon.png"><link rel="apple-touch-icon-precomposed" href="/apple-touch-icon.png"><script type="text/javascript" src="//cdn.jsdelivr.net/npm/clipboard/dist/clipboard.min.js"></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.js"></script><link rel="stylesheet" href="//cdn.jsdelivr.net/gh/codeseven/toastr/build/toastr.min.css"><meta name="generator" content="Hexo 5.2.0"></head><body><div class="body_container"><div id="header"><div class="site-name"><h1 class="hidden">FENG's notes</h1><a id="logo" href="/.">FENG's notes</a><p class="description"></p></div><div id="nav-menu"><a class="current" href="/."><i class="fa fa-home"> Home</i></a><a href="/archives/"><i class="fa fa-archive"> Archive</i></a><a href="/about/"><i class="fa fa-user"> About</i></a><a href="/atom.xml"><i class="fa fa-rss"> RSS</i></a></div></div><div class="pure-g" id="layout"><div class="pure-u-1 pure-u-md-3-4"><div class="content_container"><div class="post"><h1 class="post-title"><a href="/2014/09/15/tophat-attributeerror-nonetype-object-has-no-attribute-group/">Tophat: AttributeError: 'NoneType' object has no attribute 'group'</a></h1><div class="post-meta">2014-09-15</div><div class="post-content"><p>I have installed boost (version 1.38), Eigen (eigen-eigen-1306d75b4a21.tar.gz) and Samtools(version 1.0). [Ref:<a target="_blank" rel="noopener" href="http://ccb.jhu.edu/software/tophat/tutorial.shtml/]">http://ccb.jhu.edu/software/tophat/tutorial.shtml\]</a> Then I downlaod Tophat 2.0.12 (binary), But error occured when I test the pipeline on a simple test data set. tophat -r 20 test_ref reads_1.fq reads_2.fq</p></div><p class="readmore"><a href="/2014/09/15/tophat-attributeerror-nonetype-object-has-no-attribute-group/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/09/12/vim-error-while-loading-shared-libraries-libperl-so/">vim: error while loading shared libraries: libperl.so</a></h1><div class="post-meta">2014-09-12</div><div class="post-content"><p>[fenglei@localhost ~]$ <strong>vi .bashrc</strong> vim: error while loading shared libraries: libperl.so: cannot open shared object file: No such file or directory 出现类似错误是由于缺少相关的库文件（*.so文件）造成的 可能的原因是版本不匹配或者没有安装相应的rpm文件首先要执行的文件在那里，用which命令</p></div><p class="readmore"><a href="/2014/09/12/vim-error-while-loading-shared-libraries-libperl-so/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/09/08/cufflinks%E5%AE%89%E8%A3%85%E5%89%8D%E6%8F%90%E4%B9%8B%E4%B8%80%EF%BC%9A%E5%AE%89%E8%A3%85boost/">cufflinks安装前提之一：安装boost</a></h1><div class="post-meta">2014-09-08</div><div class="post-content"><p>转录组数据分析软件cufflinks的安装，前提是linux平台必须装有boost 1.47以上的版本！安装步骤如下：</p></div><p class="readmore"><a href="/2014/09/08/cufflinks%E5%AE%89%E8%A3%85%E5%89%8D%E6%8F%90%E4%B9%8B%E4%B8%80%EF%BC%9A%E5%AE%89%E8%A3%85boost/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/09/02/linux%E5%AE%89%E8%A3%85mysql/">linux安装mysql</a></h1><div class="post-meta">2014-09-02</div><div class="post-content"><p>本文转自shenwei.me</p></div><p class="readmore"><a href="/2014/09/02/linux%E5%AE%89%E8%A3%85mysql/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/08/30/cant-locate-iozlib-pm/">解决Can't locate CPAN.pm in @INC的问题</a></h1><div class="post-meta">2014-08-30</div><div class="post-content"><p>Linux <strong>RedHat</strong> 的系统，解决“Can’t locate CPAN.pm in @INC”的问题。 参考博文：<a target="_blank" rel="noopener" href="http://blog.sina.com.cn/s/blog/_645e2f0b01012npz.html">http://blog.sina.com.cn/s/blog\_645e2f0b01012npz.html</a> 实际步骤： $ wget <a href="ftp://ftp.muug.mb.ca/mirror/redhat/redhat/linux/6.2/cpan/i386/perl-CPAN-1.52-6.i386.rpm">ftp://ftp.muug.mb.ca/mirror/redhat/redhat/linux/6.2/cpan/i386/perl-CPAN-1.52-6.i386.rpm</a> $ su root # rpm -ivh perl-CPAN-1.52-6.i386.rpm 至此，CPAN模块就安装好了。 可运行 $ rpm -q perl-CPAN 命令进行检验。</p></div><p class="readmore"><a href="/2014/08/30/cant-locate-iozlib-pm/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/08/22/linux-ntfs/">Linux系统无法挂载格式为ntfs格式的移动硬盘的解决办法</a></h1><div class="post-meta">2014-08-22</div><div class="post-content"><p><strong>Linux**</strong>系统无法挂载格式为**<strong>ntfs**</strong>格式的移动硬盘的解决办法** 今天想把网上下载下来的一些软件包用移动硬盘挂载到Red Hat Enterprise Workstation系统计算机器上，可是怎么也挂载不上去，网上查了资料，原来默认 只识别 win 下的 vfat ，不识别 ntfs 格式的。我的移动硬盘就是ntfs格式的。再查了一些资料并结合自己的实际操作，最终解决了不能挂载ntfs移动硬盘的问题。具体步骤如下：</p></div><p class="readmore"><a href="/2014/08/22/linux-ntfs/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/08/22/linux%E7%B3%BB%E7%BB%9F%E4%B8%8B%E8%A7%A3%E5%8E%8B-tgz-%E6%96%87%E4%BB%B6/">linux系统下解压 tgz 文件</a></h1><div class="post-meta">2014-08-22</div><div class="post-content"><h3 id="解压-tgz-文件"><a href="#解压-tgz-文件" class="headerlink" title="解压 tgz 文件"></a>解压 tgz 文件</h3></div><p class="readmore"><a href="/2014/08/22/linux%E7%B3%BB%E7%BB%9F%E4%B8%8B%E8%A7%A3%E5%8E%8B-tgz-%E6%96%87%E4%BB%B6/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/08/22/%E4%BF%AE%E6%94%B9linux%E7%94%A8%E6%88%B7%E5%AF%86%E7%A0%81/">修改linux用户密码</a></h1><div class="post-meta">2014-08-22</div><div class="post-content"><p>对于初学者来说,如何修改linux用户密码也不是件容易的事,其实非常简单,下面举例说明:</p></div><p class="readmore"><a href="/2014/08/22/%E4%BF%AE%E6%94%B9linux%E7%94%A8%E6%88%B7%E5%AF%86%E7%A0%81/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/08/17/2014-plant-sci/">2014 年植物科学最新SCI影响因子前50强排行榜</a></h1><div class="post-meta">2014-08-17</div><div class="post-content"><p>2014 Journal Rankings on Plant Science:</p></div><p class="readmore"><a href="/2014/08/17/2014-plant-sci/">Read More</a></p></div><div class="post"><h1 class="post-title"><a href="/2014/08/12/illumina-pacbio/">illumina HiSeq与PacBio测序dNTP荧光标记的区别</a></h1><div class="post-meta">2014-08-12</div><div class="post-content"><p>illumina HiSeq（二代测序）测序方法采用边合成边测序的方法。向反应体系中同时添加DNA聚合酶、接头引物和带有<strong>碱基特异荧光标记的4中dNTP</strong>（如同Sanger测序法）。这些dNTP的3’-OH被化学方法所保护，因而每次只能添加一个dNTP。在dNTP被添加到合成链上后，所有未使用的游离dNTP和DNA聚合酶会被洗脱掉。接着，再加入激发荧光所需的缓冲液，用激光激发荧光信号，并有光学设备完成荧光信号的记录，最后利用计算机分析将光学信号转化为测序碱基。这样荧光信号记录完成后，再加入化学试剂淬灭荧光信号并去除dNTP 3’-OH保护基团，以便能进行下一轮的测序反应。Illumina的这种测序技术每次只添加一个dNTP的特点能够很好的地解决同聚物长度的准确测量问题，它的主要测序错误来源是碱基的替换，目前它的测序错误率在1%-1.5%之间，测序周期以人类基因组重测序为例，30x测序深度大约为1周。 <a target="_blank" rel="noopener" href="http://genehub.files.wordpress.com/2014/08/qq20140812-1.png"><img src="http://fengleiblog.files.wordpress.com/2014/08/qq20140812-1.png?w=300" alt="QQ20140812-1"></a></p></div><p class="readmore"><a href="/2014/08/12/illumina-pacbio/">Read More</a></p></div><nav class="page-navigator"><a class="extend prev" rel="prev" href="/page/14/">Previous</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/13/">13</a><a class="page-number" href="/page/14/">14</a><span class="page-number current">15</span><a class="page-number" href="/page/16/">16</a><a class="extend next" rel="next" href="/page/16/">Next</a></nav></div></div><div class="pure-u-1-4 hidden_mid_and_down"><div id="sidebar"><div class="widget"><div class="search-form"><input id="local-search-input" placeholder="Search" type="text" name="q" results="0"/><div id="local-search-result"></div></div></div><div class="widget"><div class="widget-title"><i class="fa fa-folder-o"> Categories</i></div><ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Bioinformatics/">Bioinformatics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Life/">Life</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Linux/">Linux</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Notes/">Notes</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Plant/">Plant</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Proteomics/">Proteomics</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python/">Python</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/R/">R</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/Visualization/">Visualization</a></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E6%9C%AA%E5%88%86%E7%B1%BB/">未分类</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-star-o"> Tags</i></div><div class="tagcloud"><a href="/tags/Tag/" style="font-size: 15px;">Tag</a></div></div><div class="widget"><div class="widget-title"><i class="fa fa-file-o"> Recent</i></div><ul class="post-list"><li class="post-list-item"><a class="post-list-link" href="/2021/04/16/Plant-Chromatin-immunoprecipitation/">Plant | Chromatin immunoprecipitation</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/04/08/Bioinfo-RepeatModeler/">Bioinfo | 重复序列预测工具 RepeatModeler</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/16/Bioinfo-understanding-sequence-identity/">Bioinfo | 怎么理解序列一致性（Sequence identity）</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/10/Bioinfo-Extracting-sequence-by-names/">Bioinfo | 依据序列名字列表从 fasta 文件提取序列</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/05/Proteomics-Serotonylation-A-novel-histone-H3-marker/">Proteomics | 组蛋白五羟色胺化修饰 H3Q5ser 及其生物学功能</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-What-the-differences-between-DDA-and-DIA-in-mass-spectrometer/">Proteomics | 质谱分析中DDA与DIA有什么不同？</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/04/Proteomics-Protein-Persulfidation/">Proteomics | 蛋白质硫巯基化修饰</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/Proteomics-Common-reagents-and-their-functions-in-LC-MS-MS-proteomics/">Proteomics | Common reagents and their functions in LC-MS/MS proteomics</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/03/03/R-Using-ggrepel-to-repel-overlapping-text-labels-in-ggplot2/">R | Using ggrepel to repel overlapping text labels in ggplot2</a></li><li class="post-list-item"><a class="post-list-link" href="/2021/01/08/Bioinfo-Installation-of-BUSCO-v4/">Bioinfo | Installation of BUSCO v4</a></li></ul></div><div class="widget"><div class="widget-title"><i class="fa fa-external-link"> Links</i></div><ul></ul><a href="http://www.example1.com/" title="site-name1" target="_blank">site-name1</a><ul></ul><a href="http://www.example2.com/" title="site-name2" target="_blank">site-name2</a><ul></ul><a href="http://www.example3.com/" title="site-name3" target="_blank">site-name3</a></div></div></div><div class="pure-u-1 pure-u-md-3-4"><div id="footer">Copyright © 2021 <a href="/." rel="nofollow">FENG's notes.</a> Powered by<a rel="nofollow" target="_blank" href="https://hexo.io"> Hexo.</a><a rel="nofollow" target="_blank" href="https://github.com/tufu9441/maupassant-hexo"> Theme</a> by<a rel="nofollow" target="_blank" href="https://github.com/pagecho"> Cho.</a></div></div></div><a class="show" id="rocket" href="#top"></a><script type="text/javascript" src="/js/totop.js?v=0.0.0" async></script><script type="text/javascript" src="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.js" async></script><script type="text/javascript" src="/js/fancybox.js?v=0.0.0" async></script><link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox/dist/jquery.fancybox.min.css"><link rel="stylesheet" type="text/css" href="/css/search.css?v=0.0.0"><script type="text/javascript" src="/js/search.js?v=0.0.0"></script><script>var search_path = 'search.xml';
if (search_path.length == 0) {
   search_path = 'search.xml';
}
var path = '/' + search_path;
searchFunc(path, 'local-search-input', 'local-search-result');
</script><script type="text/javascript" src="/js/copycode.js" successtext="Copy Successed!"></script><link rel="stylesheet" type="text/css" href="/css/copycode.css"><script type="text/javascript" src="/js/codeblock-resizer.js?v=0.0.0"></script><script type="text/javascript" src="/js/smartresize.js?v=0.0.0"></script></div></body></html>